<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<title>Whole Cell Simulation: src/+edu/+stanford/+covert/+cell/+sim/+analysis/CellState.m Source File</title>

<link href="tabs.css" rel="stylesheet" type="text/css"/>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
<link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="navtree.js"></script>
<script type="text/javascript">
  $(document).ready(initResizable);
</script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
  $(document).ready(function() { searchBox.OnSelectItem(0); });
</script>

</head>
<body>
<div id="top"><!-- do not remove this div! -->


<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  
  
  <td style="padding-left: 0.5em;">
   <div id="projectname">Whole Cell Simulation
   
   </div>
   
  </td>
  
  
  
 </tr>
 </tbody>
</table>
</div>

<!-- Generated by Doxygen 1.7.6.1 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "search",false,'Search');
</script>
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&#160;Page</span></a></li>
      <li><a href="annotated.html"><span>Classes</span></a></li>
      <li class="current"><a href="files.html"><span>Files</span></a></li>
      <li>
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
  <div id="navrow2" class="tabs2">
    <ul class="tablist">
      <li><a href="files.html"><span>File&#160;List</span></a></li>
      <li><a href="globals.html"><span>File&#160;Members</span></a></li>
    </ul>
  </div>
</div>
<div id="side-nav" class="ui-resizable side-nav-resizable">
  <div id="nav-tree">
    <div id="nav-tree-contents">
    </div>
  </div>
  <div id="splitbar" style="-moz-user-select:none;" 
       class="ui-resizable-handle">
  </div>
</div>
<script type="text/javascript">
  initNavTree('_09analysis_2_cell_state_8m.html','');
</script>
<div id="doc-content">
<div class="header">
  <div class="headertitle">
<div class="title">src/+edu/+stanford/+covert/+cell/+sim/+analysis/CellState.m</div>  </div>
</div><!--header-->
<div class="contents">
<a href="_09analysis_2_cell_state_8m.html">Go to the documentation of this file.</a><div class="fragment"><pre class="fragment"><a name="l00001"></a>00001 %<a class="code" href="class_cell_state.html">CellState</a>
<a name="l00002"></a>00002 %
<a name="l00003"></a>00003 % Author: Jonathan Karr, jkarr@stanford.edu
<a name="l00004"></a>00004 % Affiliation: Covert Lab, Department of Bioengineering, Stanford University
<a name="l00005"></a>00005 % Last Updated: 9/12/2010
<a name="l00006"></a><a class="code" href="class_cell_state.html">00006</a> classdef <a class="code" href="class_cell_state.html">CellState</a>
<a name="l00007"></a>00007     %printing
<a name="l00008"></a>00008     methods (Static)
<a name="l00009"></a>00009         function run(sim, fileName)
<a name="l00010"></a>00010             import edu.stanford.covert.cell.sim.analysis.<a class="code" href="class_cell_state.html">CellState</a>;
<a name="l00011"></a>00011             import edu.stanford.covert.cell.sim.util.PlotUtil;
<a name="l00012"></a>00012             import edu.stanford.covert.cell.sim.util.PrintUtil;            
<a name="l00013"></a>00013             
<a name="l00014"></a>00014             %excel file
<a name="l00015"></a>00015             [content, colLabels, indentation] = CellState.state(sim);
<a name="l00016"></a>00016             if nargin == 1
<a name="l00017"></a>00017                 PrintUtil.printToStdIO(content, colLabels, struct(&#39;indentation&#39;, indentation));
<a name="l00018"></a>00018             else                
<a name="l00019"></a>00019                 PrintUtil.printToFile(content, colLabels, [fileName &#39;.xls&#39;], &#39;<a class="code" href="class_state.html">State</a>&#39;, struct(&#39;indentation&#39;, indentation));
<a name="l00020"></a>00020             end
<a name="l00021"></a>00021             
<a name="l00022"></a>00022             [content, colLabels, indentation] = CellState.processes(sim);
<a name="l00023"></a>00023             if nargin == 1
<a name="l00024"></a>00024                 PrintUtil.printToStdIO(content, colLabels, struct(&#39;indentation&#39;, indentation));
<a name="l00025"></a>00025             else                
<a name="l00026"></a>00026                 PrintUtil.printToFile(content, colLabels, [fileName &#39;.xls&#39;], &#39;<a class="code" href="class_process.html">Process</a>&#39;, struct(&#39;indentation&#39;, indentation));
<a name="l00027"></a>00027             end
<a name="l00028"></a>00028             
<a name="l00029"></a>00029             % plots
<a name="l00030"></a>00030             if nargin == 1
<a name="l00031"></a>00031                 CellState.plotStateOnChromosome(sim, PlotUtil.newAxesHandle());
<a name="l00032"></a>00032             else
<a name="l00033"></a>00033                 [axesHandle, figHandle] = PlotUtil.newAxesHandle();
<a name="l00034"></a>00034                 
<a name="l00035"></a>00035                 cla(axesHandle);
<a name="l00036"></a>00036                 CellState.plotStateOnChromosome(sim, axesHandle);
<a name="l00037"></a>00037                 saveas(figHandle, [fileName &#39;-OnChromosome.pdf&#39;]);
<a name="l00038"></a>00038                 
<a name="l00039"></a>00039                 close(figHandle);
<a name="l00040"></a>00040             end
<a name="l00041"></a>00041         end
<a name="l00042"></a>00042         
<a name="l00043"></a>00043         function [content, colLabels, indentation] = state(sim)
<a name="l00044"></a>00044             %import classes
<a name="l00045"></a>00045             import edu.stanford.covert.util.<a class="code" href="class_constant_util.html">ConstantUtil</a>;
<a name="l00046"></a>00046             import edu.stanford.covert.cell.sim.Simulation;
<a name="l00047"></a>00047                         
<a name="l00048"></a>00048             content = cell(0, 4);
<a name="l00049"></a>00049             colLabels = {<span class="stringliteral">&#39;Quantity&#39;</span>, <span class="stringliteral">&#39;Value&#39;</span>, <span class="stringliteral">&#39;Units&#39;</span>};
<a name="l00050"></a>00050             
<a name="l00051"></a>00051             %% time
<a name="l00052"></a>00052             time = sim.state(<span class="stringliteral">&#39;Time&#39;</span>);
<a name="l00053"></a>00053             content = [content;{
<a name="l00054"></a>00054                 0 &#39;Time&#39; time.values(1)/ConstantUtil.secondsPerHour &#39;h&#39;;
<a name="l00055"></a>00055                 }];
<a name="l00056"></a>00056             
<a name="l00057"></a>00057             %% mass
<a name="l00058"></a>00058             mass = sim.state(<span class="stringliteral">&#39;Mass&#39;</span>);
<a name="l00059"></a>00059             cellDryMass = sum(mass.cellDry);
<a name="l00060"></a>00060             content = [content;{
<a name="l00061"></a>00061                 0 &#39;Mass&#39; sum(mass.cell) &#39;g&#39;;
<a name="l00062"></a>00062                     1 &#39;Dry&#39; sum(mass.cellDry) &#39;g&#39;;
<a name="l00063"></a>00063                     1 &#39;Metabolite&#39; sum(mass.metaboliteWt) &#39;g&#39;;
<a name="l00064"></a>00064                         2 &#39;Percent Dry&#39; sum(mass.metaboliteWt) / cellDryMass * 100 &#39;%&#39;;
<a name="l00065"></a>00065                     1 &#39;DNA&#39; sum(mass.dnaWt) &#39;g&#39;;
<a name="l00066"></a>00066                         2 &#39;Percent Dry&#39; sum(mass.dnaWt) / cellDryMass * 100 &#39;%&#39;;
<a name="l00067"></a>00067                     1 &#39;RNA&#39; sum(mass.rnaWt) &#39;g&#39;;
<a name="l00068"></a>00068                         2 &#39;Percent Dry&#39; sum(mass.rnaWt) / cellDryMass * 100 &#39;%&#39;;
<a name="l00069"></a>00069                     1 &#39;Protein&#39; sum(mass.proteinWt) &#39;g&#39;;
<a name="l00070"></a>00070                         2 &#39;Percent Dry&#39; sum(mass.proteinWt) / cellDryMass * 100 &#39;%&#39;;
<a name="l00071"></a>00071                     1 &#39;No. cells&#39; sum(mass.cellDry)/mass.cellInitialDryWeight &#39;cells&#39;;
<a name="l00072"></a>00072                 }];
<a name="l00073"></a>00073             
<a name="l00074"></a>00074             %% growth
<a name="l00075"></a>00075             s = sim.state(<span class="stringliteral">&#39;MetabolicReaction&#39;</span>);
<a name="l00076"></a>00076             time = sim.state(<span class="stringliteral">&#39;Time&#39;</span>);
<a name="l00077"></a>00077             content = [content;{
<a name="l00078"></a>00078                 0 &#39;Growth&#39; s.growth &#39;cells/s&#39;;
<a name="l00079"></a>00079                     1 &#39;Relative to expected rate&#39; s.growth / (1 / time.cellCycleLength * log(2) * exp(time.values / time.cellCycleLength * log(2))) [];
<a name="l00080"></a>00080                     1 &#39;Doubling Time&#39;, s.doublingTime / ConstantUtil.secondsPerHour &#39;h&#39;;
<a name="l00081"></a>00081                 }];
<a name="l00082"></a>00082             
<a name="l00083"></a>00083             %% shape
<a name="l00084"></a>00084             s = sim.state(<span class="stringliteral">&#39;Geometry&#39;</span>);
<a name="l00085"></a>00085             content = [content;{
<a name="l00086"></a>00086                 0 &#39;Shape&#39; &#39;Cyclinder with hemispherical caps&#39; []
<a name="l00087"></a>00087                     1 &#39;Width&#39; s.width &#39;m&#39;
<a name="l00088"></a>00088                     1 &#39;Pinched Diameter&#39; s.pinchedDiameter &#39;m&#39;
<a name="l00089"></a>00089                     1 &#39;Cylindrical length&#39; s.cylindricalLength &#39;m&#39;
<a name="l00090"></a>00090                     1 &#39;Total length&#39; s.totalLength &#39;m&#39;
<a name="l00091"></a>00091                     1 &#39;Pinched circumference&#39; s.pinchedCircumference &#39;m&#39;
<a name="l00092"></a>00092                     1 &#39;Volume&#39; s.volume &#39;L&#39;
<a name="l00093"></a>00093                     1 &#39;Surface area&#39; s.surfaceArea&#39; &#39;m^2&#39;
<a name="l00094"></a>00094                     1 &#39;Pinched&#39; s.pinched &#39;true/false&#39;
<a name="l00095"></a>00095                 }];
<a name="l00096"></a>00096             
<a name="l00097"></a>00097             %% FtsZ ring
<a name="l00098"></a>00098             s = sim.state(<span class="stringliteral">&#39;FtsZRing&#39;</span>);
<a name="l00099"></a>00099             content = [content;{
<a name="l00100"></a>00100                 0 &#39;FtsZ ring&#39; [] []
<a name="l00101"></a>00101                     1 &#39;No. edges&#39; s.numEdges []
<a name="l00102"></a>00102                     1 &#39;No. edges bound by 1 straight FtsZ 9-mer&#39; s.numEdgesOneStraight []
<a name="l00103"></a>00103                     1 &#39;No. edges bound by 2 straight FtsZ 9-mers&#39; s.numEdgesTwoStraight []
<a name="l00104"></a>00104                     1 &#39;No. edges bound by 2 bent FtsZ 9-mers&#39; s.numEdgesTwoBent []
<a name="l00105"></a>00105                     1 &#39;No. edges bound by 1 bent FtsZ 9-mer&#39; s.numResidualBent []
<a name="l00106"></a>00106                 }];
<a name="l00107"></a>00107             
<a name="l00108"></a>00108             %% stimuli
<a name="l00109"></a>00109             content = [content;{
<a name="l00110"></a>00110                 0 &#39;Stimuli&#39; &#39;&#39; []
<a name="l00111"></a>00111                 }];
<a name="l00112"></a>00112             stim = sim.state(<span class="stringliteral">&#39;Stimulus&#39;</span>);
<a name="l00113"></a>00113             <span class="keywordflow">for</span> i = 1:numel(stim.wholeCellModelIDs)
<a name="l00114"></a>00114                 value = stim.values(i, stim.values(i, :) ~= 0);
<a name="l00115"></a>00115                 if numel(value) == 0
<a name="l00116"></a>00116                     value = 0;
<a name="l00117"></a>00117                 elseif numel(value) &gt; 1
<a name="l00118"></a>00118                     value = mean(value);
<a name="l00119"></a>00119                 end
<a name="l00120"></a>00120                 content = [content;{
<a name="l00121"></a>00121                     1 stim.wholeCellModelIDs{i} value []
<a name="l00122"></a>00122                     }]; %#ok&lt;AGROW&gt;
<a name="l00123"></a>00123             end
<a name="l00124"></a>00124             
<a name="l00125"></a>00125             %% metabolites
<a name="l00126"></a>00126             c = sim.state(<span class="stringliteral">&#39;Chromosome&#39;</span>);
<a name="l00127"></a>00127             s = sim.state(<span class="stringliteral">&#39;Metabolite&#39;</span>);
<a name="l00128"></a>00128             
<a name="l00129"></a>00129             freedNTPs = s.counts(s.dntpIndexs, sim.compartment.cytosolIndexs);
<a name="l00130"></a>00130             freeNTPs = s.counts(s.ntpIndexs, sim.compartment.cytosolIndexs);
<a name="l00131"></a>00131             freeNDPs = s.counts(s.ndpIndexs, sim.compartment.cytosolIndexs);
<a name="l00132"></a>00132             freeNMPs = s.counts(s.nmpIndexs, sim.compartment.cytosolIndexs);
<a name="l00133"></a>00133             freeAAs = s.counts(s.aminoAcidIndexs, sim.compartment.cytosolIndexs);
<a name="l00134"></a>00134             
<a name="l00135"></a>00135             [mindNTPVal, mindNTPIdx] = min(freedNTPs);
<a name="l00136"></a>00136             [minNTPVal, minNTPIdx] = min(freeNTPs);
<a name="l00137"></a>00137             [minNDPVal, minNDPIdx] = min(freeNDPs);
<a name="l00138"></a>00138             [minNMPVal, minNMPIdx] = min(freeNMPs);
<a name="l00139"></a>00139             [minAAVal, minAAIdx] = min(freeAAs);
<a name="l00140"></a>00140             
<a name="l00141"></a>00141             freeNTPMass = freeNTPs<span class="stringliteral">&#39; * s.molecularWeights(s.ntpIndexs) / ConstantUtil.nAvogadro;</span>
<a name="l00142"></a>00142 <span class="stringliteral">            freeAAMass = freeAAs&#39;</span> * s.molecularWeights(s.aminoAcidIndexs) / <a class="code" href="class_constant_util.html">ConstantUtil</a>.<a class="code" href="class_constant_util.html#a9da93ee8ba55dd66d9406905f4c63a4b">nAvogadro</a>;
<a name="l00143"></a>00143             
<a name="l00144"></a>00144             content = [content;{
<a name="l00145"></a>00145                 0 &#39;Metabolites&#39; [] []
<a name="l00146"></a>00146                     1 &#39;Free&#39; sum(s.counts(:)) &#39;molecules&#39;
<a name="l00147"></a>00147                     1 &#39;dNTPs&#39; [] []
<a name="l00148"></a>00148                         2 &#39;Free&#39; sum(freedNTPs) &#39;molecules&#39;                        
<a name="l00149"></a>00149                         2 &#39;Min free&#39; s.wholeCellModelIDs{s.dntpIndexs(mindNTPIdx)} []
<a name="l00150"></a>00150                         2 &#39;Min free&#39; mindNTPVal &#39;molecules&#39;
<a name="l00151"></a>00151                         2 &#39;Incorporated into DNA&#39; (1 - sum(freedNTPs) / (sum(freedNTPs) + collapse(c.polymerizedRegions)))*100 &#39;%&#39;
<a name="l00152"></a>00152                     1 &#39;NTPs&#39; [] []
<a name="l00153"></a>00153                         2 &#39;Free&#39; sum(freeNTPs) &#39;molecules&#39;                        
<a name="l00154"></a>00154                         2 &#39;Min free&#39; s.wholeCellModelIDs{s.ntpIndexs(minNTPIdx)} []
<a name="l00155"></a>00155                         2 &#39;Min free&#39; minNTPVal &#39;molecules&#39;
<a name="l00156"></a>00156                         2 &#39;Incorporated into RNA&#39; (1 - freeNTPMass / (freeNTPMass + sum(sim.state(&#39;Mass&#39;).rnaWt)))*100 &#39;%&#39;
<a name="l00157"></a>00157                     1 &#39;NDPs&#39; [] []
<a name="l00158"></a>00158                         2 &#39;Free&#39; sum(freeNDPs) &#39;molecules&#39;                       
<a name="l00159"></a>00159                         2 &#39;Min free&#39; s.wholeCellModelIDs{s.ndpIndexs(minNDPIdx)} []
<a name="l00160"></a>00160                         2 &#39;Min free&#39; minNDPVal &#39;molecules&#39;
<a name="l00161"></a>00161                     1 &#39;NMPs&#39; [] []
<a name="l00162"></a>00162                         2 &#39;Free&#39; sum(freeNMPs) &#39;molecules&#39;                        
<a name="l00163"></a>00163                         2 &#39;Min free&#39; s.wholeCellModelIDs{s.nmpIndexs(minNMPIdx)} []
<a name="l00164"></a>00164                         2 &#39;Min free&#39; minNMPVal &#39;molecules&#39;
<a name="l00165"></a>00165                     1 &#39;Amino acids&#39; [] []
<a name="l00166"></a>00166                         2 &#39;Free&#39; sum(freeAAs) &#39;molecules&#39;                        
<a name="l00167"></a>00167                         2 &#39;Min free&#39; s.wholeCellModelIDs{s.aminoAcidIndexs(minAAIdx)} []
<a name="l00168"></a>00168                         2 &#39;Min free&#39; minAAVal &#39;molecules&#39;
<a name="l00169"></a>00169                         2 &#39;Incorporated into Protein&#39; (1 - freeAAMass / (freeAAMass + sum(sim.state(&#39;Mass&#39;).proteinWt)))*100 &#39;%&#39;
<a name="l00170"></a>00170                     1 &#39;Energy&#39; [] []
<a name="l00171"></a>00171                         2 &#39;ATP&#39; sum(s.counts(s.getIndexs({&#39;ATP&#39;}), :)) &#39;molecules&#39;
<a name="l00172"></a>00172                         2 &#39;ADP&#39; sum(s.counts(s.getIndexs({&#39;ADP&#39;}), :)) &#39;molecules&#39;
<a name="l00173"></a>00173                         2 &#39;AMP&#39; sum(s.counts(s.getIndexs({&#39;AMP&#39;}), :)) &#39;molecules&#39;
<a name="l00174"></a>00174                         2 &#39;PPi&#39; sum(s.counts(s.getIndexs({&#39;PPI&#39;}), :)) &#39;molecules&#39;
<a name="l00175"></a>00175                         2 &#39;Pi&#39;  sum(s.counts(s.getIndexs({&#39;PI&#39;}),  :)) &#39;molecules&#39;
<a name="l00176"></a>00176                         2 &#39;H2O&#39; sum(s.counts(s.getIndexs({&#39;H2O&#39;}), :)) &#39;molecules&#39;
<a name="l00177"></a>00177                         2 &#39;H&#39;   sum(s.counts(s.getIndexs({&#39;H&#39;}),   :)) &#39;molecules&#39;
<a name="l00178"></a>00178                 }];
<a name="l00179"></a>00179             
<a name="l00180"></a>00180             %% DNA
<a name="l00181"></a>00181             c = sim.state(<span class="stringliteral">&#39;Chromosome&#39;</span>);
<a name="l00182"></a>00182             pm = sim.state(<span class="stringliteral">&#39;ProteinMonomer&#39;</span>);
<a name="l00183"></a>00183             pc = sim.state(<span class="stringliteral">&#39;ProteinComplex&#39;</span>);
<a name="l00184"></a>00184             ri = sim.process(<span class="stringliteral">&#39;ReplicationInitiation&#39;</span>);
<a name="l00185"></a>00185             dr = sim.process(<span class="stringliteral">&#39;DNARepair&#39;</span>);
<a name="l00186"></a>00186             dnaABoxBound = ...
<a name="l00187"></a>00187                 ri.calculateDnaABoxStatus() == ri.dnaABoxStatus_DnaAATPBound | ...
<a name="l00188"></a>00188                 ri.calculateDnaABoxStatus() == ri.dnaABoxStatus_DnaAADPBound;
<a name="l00189"></a>00189             [polATPs, polADPs] = ri.calculateDnaAR1234Polymerization();
<a name="l00190"></a>00190             oriCComplexSize = ri.calculateDnaAR1234ComplexSize(polATPs, polADPs);
<a name="l00191"></a>00191             [~, boundMonomers] = find(c.monomerBoundSites);
<a name="l00192"></a>00192             [~, boundComplexs] = find(c.complexBoundSites);
<a name="l00193"></a>00193             occupiedBases = sum(c.monomerDNAFootprints(boundMonomers) + c.complexDNAFootprints(boundComplexs));
<a name="l00194"></a>00194                         
<a name="l00195"></a>00195             methylatedRMSites = nnz(c.damagedBases([
<a name="l00196"></a>00196                 dr.RM_MunI_RecognitionSites(:, dr.RM_MunI_MethylatedPositions(1))   ones(size(dr.RM_MunI_RecognitionSites, 1), 1);
<a name="l00197"></a>00197                 dr.RM_MunI_RecognitionSites(:, dr.RM_MunI_MethylatedPositions(2)) 2*ones(size(dr.RM_MunI_RecognitionSites, 1), 1);
<a name="l00198"></a>00198                 dr.RM_MunI_RecognitionSites(:, dr.RM_MunI_MethylatedPositions(1)) 3*ones(size(dr.RM_MunI_RecognitionSites, 1), 1);
<a name="l00199"></a>00199                 dr.RM_MunI_RecognitionSites(:, dr.RM_MunI_MethylatedPositions(2)) 4*ones(size(dr.RM_MunI_RecognitionSites, 1), 1);
<a name="l00200"></a>00200                 ])) / 2;
<a name="l00201"></a>00201             polymerizedRMSites = sum(c.isRegionPolymerized([
<a name="l00202"></a>00202                 dr.RM_MunI_RecognitionSites(:, 1)   ones(size(dr.RM_MunI_RecognitionSites, 1), 1);
<a name="l00203"></a>00203                 dr.RM_MunI_RecognitionSites(:, 1) 2*ones(size(dr.RM_MunI_RecognitionSites, 1), 1);
<a name="l00204"></a>00204                 dr.RM_MunI_RecognitionSites(:, 1) 3*ones(size(dr.RM_MunI_RecognitionSites, 1), 1);
<a name="l00205"></a>00205                 dr.RM_MunI_RecognitionSites(:, 1) 4*ones(size(dr.RM_MunI_RecognitionSites, 1), 1);
<a name="l00206"></a>00206                 ], 6, <span class="keyword">false</span>)) / 2;
<a name="l00207"></a>00207             
<a name="l00208"></a>00208             content = [content;{
<a name="l00209"></a>00209                 0 &#39;No. chromosomes&#39; collapse(c.polymerizedRegions)/c.sequenceLen [];
<a name="l00210"></a>00210                     1 &#39;Bases&#39; collapse(c.polymerizedRegions) &#39;nt&#39;;
<a name="l00211"></a>00211                     1 &#39;Linking number&#39; collapse(c.linkingNumbers)/2 &#39;Links&#39;;
<a name="l00212"></a>00212                     1 &#39;Superhelical density&#39;, (collapse(c.linkingNumbers)  - collapse(c.linkingNumbers_minFreeEnergy)) / collapse(c.linkingNumbers_minFreeEnergy) [];
<a name="l00213"></a>00213                     1 &#39;Bound proteins&#39; nnz(c.monomerBoundSites) + nnz(c.complexBoundSites) &#39;molecules&#39;
<a name="l00214"></a>00214                         2 &#39;Monomers&#39; nnz(c.monomerBoundSites) &#39;molecules&#39;
<a name="l00215"></a>00215                             3 &#39;GntR&#39; nnz(c.monomerBoundSites == pm.getIndexs({&#39;MG_101_MONOMER&#39;})) &#39;molecules&#39;
<a name="l00216"></a>00216                         2 &#39;Complexes&#39; nnz(c.monomerBoundSites) &#39;molecules&#39;
<a name="l00217"></a>00217                             3 &#39;DnaA&#39; nnz(dnaABoxBound) &#39;molecules&#39;
<a name="l00218"></a>00218                                 4 &#39;7-mer&#39; nnz(dnaABoxBound(ri.dnaABoxIndexs_7mer, :)) &#39;molecules&#39;
<a name="l00219"></a>00219                                 4 &#39;8-mer&#39; nnz(dnaABoxBound(ri.dnaABoxIndexs_8mer, :)) &#39;molecules&#39;
<a name="l00220"></a>00220                                 4 &#39;9-mer&#39; nnz(dnaABoxBound(ri.dnaABoxIndexs_9mer, :)) &#39;molecules&#39;
<a name="l00221"></a>00221                                 4 &#39;OriC Complex&#39; oriCComplexSize(1) &#39;molecules&#39;
<a name="l00222"></a>00222                                 4 &#39;R1&#39; nnz(dnaABoxBound(ri.dnaABoxIndexs_R12345(1), :)) &#39;molecules&#39;
<a name="l00223"></a>00223                                 4 &#39;R2&#39; nnz(dnaABoxBound(ri.dnaABoxIndexs_R12345(2), :)) &#39;molecules&#39;
<a name="l00224"></a>00224                                 4 &#39;R3&#39; nnz(dnaABoxBound(ri.dnaABoxIndexs_R12345(3), :)) &#39;molecules&#39;
<a name="l00225"></a>00225                                 4 &#39;R4&#39; nnz(dnaABoxBound(ri.dnaABoxIndexs_R12345(3), :)) &#39;molecules&#39;
<a name="l00226"></a>00226                                 4 &#39;R5&#39; nnz(dnaABoxBound(ri.dnaABoxIndexs_R12345(5), :)) &#39;molecules&#39;
<a name="l00227"></a>00227                             3 &#39;Gyrase&#39; nnz(c.complexBoundSites == pc.getIndexs({&#39;DNA_GYRASE&#39;})) &#39;molecules&#39;
<a name="l00228"></a>00228                             3 &#39;LuxR&#39; nnz(c.complexBoundSites == pc.getIndexs({&#39;MG_428_DIMER&#39;})) &#39;molecules&#39;
<a name="l00229"></a>00229                             3 &#39;SMC&#39; nnz(c.complexBoundSites == pc.getIndexs({&#39;MG_213_214_298_6MER_ADP&#39;})) &#39;molecules&#39;
<a name="l00230"></a>00230                                 4 &#39;Average spacing&#39; collapse(c.polymerizedRegions)/2 / nnz(c.complexBoundSites == pc.getIndexs({&#39;MG_213_214_298_6MER_ADP&#39;})) &#39;nt&#39;
<a name="l00231"></a>00231                         2 &#39;Occupied&#39; occupiedBases &#39;nt&#39;
<a name="l00232"></a>00232                         2 &#39;Occupied&#39; occupiedBases / collapse(c.polymerizedRegions) / 2 * 100 &#39;%&#39;
<a name="l00233"></a>00233                     1 &#39;Damage&#39; [] []
<a name="l00234"></a>00234                         2 &#39;Gap sites&#39; nnz(c.gapSites) &#39;nt&#39;
<a name="l00235"></a>00235                         2 &#39;Abasic sites&#39; nnz(c.abasicSites) &#39;nt&#39;
<a name="l00236"></a>00236                         2 &#39;Damaged bases&#39; nnz(c.damagedBases) &#39;nt&#39;
<a name="l00237"></a>00237                             3 &#39;m6AD&#39; nnz(c.damagedBases == c.metabolite.m6ADIndexs) &#39;nt&#39;
<a name="l00238"></a>00238                             3 &#39;methylated R/M sites&#39; methylatedRMSites &#39;nt&#39;
<a name="l00239"></a>00239                             3 &#39;methylated R/M sites&#39; methylatedRMSites/polymerizedRMSites*100 &#39;%&#39;
<a name="l00240"></a>00240                         2 &#39;Damaged sugar-phosphate&#39; nnz(c.damagedSugarPhosphates) &#39;nt&#39;
<a name="l00241"></a>00241                         2 &#39;Strand breaks&#39; nnz(c.strandBreaks) &#39;nt&#39;
<a name="l00242"></a>00242                         2 &#39;Intrastrand crosslinks&#39; nnz(c.intrastrandCrossLinks) &#39;nt&#39;
<a name="l00243"></a>00243                         2 &#39;Holliday junctions&#39; nnz(c.hollidayJunctions) &#39;nt&#39;
<a name="l00244"></a>00244                 }];
<a name="l00245"></a>00245             
<a name="l00246"></a>00246             %% <a class="code" href="class_r_n_a.html">RNA</a>
<a name="l00247"></a>00247             r = sim.state(<span class="stringliteral">&#39;Rna&#39;</span>);
<a name="l00248"></a>00248             
<a name="l00249"></a>00249             matureRNAs = sum(r.counts(r.matureIndexs, :), 2);
<a name="l00250"></a>00250             aminoacylatedRNAs = sum(r.counts(r.aminoacylatedIndexs, :), 2);
<a name="l00251"></a>00251             content = [content;{
<a name="l00252"></a>00252                 0 &#39;RNA&#39; sum(r.counts(:)) &#39;molecules&#39;
<a name="l00253"></a>00253                     1 &#39;mRNA&#39; sum(matureRNAs(r.matureMRNAIndexs)) &#39;molecules&#39;
<a name="l00254"></a>00254                         2 &#39;No. species expressed&#39; nnz(r.counts(r.matureIndexs(r.matureMRNAIndexs), :)) []
<a name="l00255"></a>00255                     1 &#39;rRNA&#39; sum(matureRNAs(r.matureRRNAIndexs)) &#39;molecules&#39;
<a name="l00256"></a>00256                         2 &#39;5S&#39;  matureRNAs(r.matureRibosomalRRNAIndexs(1)) &#39;molecules&#39;
<a name="l00257"></a>00257                         2 &#39;15S&#39; matureRNAs(r.matureRibosomalRRNAIndexs(2)) &#39;molecules&#39;
<a name="l00258"></a>00258                         2 &#39;23S&#39; matureRNAs(r.matureRibosomalRRNAIndexs(3)) &#39;molecules&#39;
<a name="l00259"></a>00259                     1 &#39;sRNA&#39; sum(matureRNAs(r.matureSRNAIndexs) + aminoacylatedRNAs(r.matureSRNAIndexs)) &#39;molecules&#39;
<a name="l00260"></a>00260                         2 &#39;scRNA&#39;   matureRNAs(r.matureSRNAIndexs(1))+aminoacylatedRNAs(r.matureSRNAIndexs(1)) &#39;molecules&#39;
<a name="l00261"></a>00261                         2 &#39;MG170&#39;   matureRNAs(r.matureSRNAIndexs(2))+aminoacylatedRNAs(r.matureSRNAIndexs(2)) &#39;molecules&#39;
<a name="l00262"></a>00262                         2 &#39;RNase P&#39; matureRNAs(r.matureSRNAIndexs(3))+aminoacylatedRNAs(r.matureSRNAIndexs(3)) &#39;molecules&#39;
<a name="l00263"></a>00263                         2 &#39;tmRNA&#39;   matureRNAs(r.matureSRNAIndexs(4))+aminoacylatedRNAs(r.matureSRNAIndexs(4)) &#39;molecules&#39;
<a name="l00264"></a>00264                             3 &#39;Free&#39; matureRNAs(r.matureSRNAIndexs(4)) &#39;molecules&#39;
<a name="l00265"></a>00265                             4 &#39;Aminoacylated&#39; aminoacylatedRNAs(r.matureSRNAIndexs(4)) &#39;molecules&#39;
<a name="l00266"></a>00266                     1 &#39;tRNA&#39; sum(matureRNAs(r.matureTRNAIndexs) + aminoacylatedRNAs(r.matureTRNAIndexs)) &#39;molecules&#39;
<a name="l00267"></a>00267                         2 &#39;Free&#39; sum(matureRNAs(r.matureTRNAIndexs)) &#39;molecules&#39;
<a name="l00268"></a>00268                         2 &#39;Aminoacylated&#39; sum(aminoacylatedRNAs(r.matureTRNAIndexs)) &#39;molecules&#39;
<a name="l00269"></a>00269                         2 &#39;No. species expressed&#39; nnz(matureRNAs(r.matureTRNAIndexs) + aminoacylatedRNAs(r.matureTRNAIndexs)) &#39;molecules&#39;
<a name="l00270"></a>00270                     1 &#39;Immature&#39; nnz(sim.process(&#39;Transcription&#39;).transcripts.boundTranscriptionUnits) &#39;molecules&#39;
<a name="l00271"></a>00271                     1 &#39;Aborted&#39; size(sim.state(&#39;Transcript&#39;).abortedTranscripts, 1) &#39;molecules&#39;
<a name="l00272"></a>00272                 }];
<a name="l00273"></a>00273             
<a name="l00274"></a>00274             %% <a class="code" href="class_protein.html">Protein</a>
<a name="l00275"></a>00275             m = sim.state(<span class="stringliteral">&#39;ProteinMonomer&#39;</span>);
<a name="l00276"></a>00276             c = sim.state(<span class="stringliteral">&#39;ProteinComplex&#39;</span>);
<a name="l00277"></a>00277             
<a name="l00278"></a>00278             ribosomalMonomerIdxs = sum(c.proteinComplexComposition(sim.gene.mRNAIndexs, c.getIndexs({<span class="stringliteral">&#39;RIBOSOME_70S&#39;</span>}), :),3);
<a name="l00279"></a>00279             ribosomalMonomers = sum(m.counts(m.matureIndexs(ribosomalMonomerIdxs &gt; 0), :), 2);
<a name="l00280"></a>00280             [minRibosomalMonomers, minRibosomalMonomerIdx] = min(ribosomalMonomers);
<a name="l00281"></a>00281             minRibosomalMonomerID = m.wholeCellModelIDs{m.matureIndexs(minRibosomalMonomerIdx)};
<a name="l00282"></a>00282             
<a name="l00283"></a>00283             [idxs, ~, counts] = find(sum(m.counts(m.inactivatedIndexs, :), 2));
<a name="l00284"></a>00284             inactiveMonomers = cell(numel(idxs), 4);
<a name="l00285"></a>00285             inactiveMonomers(:, 1) = {2};
<a name="l00286"></a>00286             inactiveMonomers(:, 2) = m.wholeCellModelIDs(m.inactivatedIndexs(idxs));
<a name="l00287"></a>00287             inactiveMonomers(:, 3) = num2cell(counts);
<a name="l00288"></a>00288 
<a name="l00289"></a>00289             [idxs, ~, counts] = find(sum(c.counts(c.inactivatedIndexs, :), 2));
<a name="l00290"></a>00290             inactiveComplexs = cell(numel(idxs), 4);
<a name="l00291"></a>00291             inactiveComplexs(:, 1) = {2};
<a name="l00292"></a>00292             inactiveComplexs(:, 2) = c.wholeCellModelIDs(c.inactivatedIndexs(idxs));
<a name="l00293"></a>00293             inactiveComplexs(:, 3) = reshape(num2cell(counts), [], 1);
<a name="l00294"></a>00294             
<a name="l00295"></a>00295             content = [content;{
<a name="l00296"></a>00296                 0 &#39;Protein&#39; sum(m.counts(:)) + sum(c.counts(:)) &#39;molecules&#39;
<a name="l00297"></a>00297                     1 &#39;Monomers&#39; sum(m.counts(:)) &#39;molecules&#39;
<a name="l00298"></a>00298                         2 &#39;Ribosomal proteins&#39; sum(ribosomalMonomers) &#39;molecules&#39;
<a name="l00299"></a>00299                             3 &#39;Min expressed species&#39; minRibosomalMonomers &#39;molecules&#39;
<a name="l00300"></a>00300                             3 &#39;Min expressed species&#39; minRibosomalMonomerID []
<a name="l00301"></a>00301                     1 &#39;Complexs&#39; sum(c.counts(:)) &#39;molecules&#39;
<a name="l00302"></a>00302                     1 &#39;Bound&#39; sum(sum(m.counts(m.boundIndexs, :)))+sum(sum(c.counts(c.boundIndexs, :))) &#39;molecules&#39;
<a name="l00303"></a>00303                         2 &#39;Chromosome&#39; nnz(sim.state(&#39;Chromosome&#39;).monomerBoundSites)+nnz(sim.state(&#39;Chromosome&#39;).complexBoundSites) &#39;molecules&#39;
<a name="l00304"></a>00304                         2 &#39;RNA polymerase&#39; nnz(sim.process(&#39;Transcription&#39;).transcripts.boundTranscriptionUnits) &#39;molecules&#39;
<a name="l00305"></a>00305                         2 &#39;Ribosome&#39; nnz(sim.state(&#39;Ribosome&#39;).boundMRNAs) &#39;molecules&#39;
<a name="l00306"></a>00306                     1 &#39;Immature&#39; nnz(sim.state(&#39;Ribosome&#39;).boundMRNAs) &#39;molecules&#39;
<a name="l00307"></a>00307                     1 &#39;Inactive&#39; sum(sum(m.counts(m.inactivatedIndexs, :)))+sum(sum(c.counts(c.inactivatedIndexs, :))) &#39;molecules&#39;
<a name="l00308"></a>00308                         };
<a name="l00309"></a>00309                         inactiveMonomers;
<a name="l00310"></a>00310                         inactiveComplexs;
<a name="l00311"></a>00311                         {
<a name="l00312"></a>00312                     1 &#39;Damaged&#39; sum(sum(m.counts(m.damagedIndexs, :)))+sum(sum(c.counts(c.damagedIndexs, :))) &#39;molecules&#39;
<a name="l00313"></a>00313                     1 &#39;Aborted&#39; size(sim.state(&#39;Polypeptide&#39;).abortedPolypeptides, 1) &#39;molecules&#39;
<a name="l00314"></a>00314                     1 &#39;Cytosol&#39; sum(m.counts(:, sim.compartment.cytosolIndexs))+ sum(c.counts(:, sim.compartment.cytosolIndexs)) &#39;molecules&#39;
<a name="l00315"></a>00315                     1 &#39;Membrane&#39; sum(m.counts(:, sim.compartment.membraneIndexs))+ sum(c.counts(:, sim.compartment.membraneIndexs)) &#39;molecules&#39;
<a name="l00316"></a>00316                     1 &#39;Term. Org. Cytosol&#39; sum(m.counts(:, sim.compartment.terminalOrganelleCytosolIndexs))+ sum(c.counts(:, sim.compartment.terminalOrganelleCytosolIndexs)) &#39;molecules&#39;
<a name="l00317"></a>00317                     1 &#39;Term. Org. Membrane&#39; sum(m.counts(:, sim.compartment.terminalOrganelleMembraneIndexs))+ sum(c.counts(:, sim.compartment.terminalOrganelleMembraneIndexs)) &#39;molecules&#39;
<a name="l00318"></a>00318                     1 &#39;Extracellular&#39; sum(m.counts(:, sim.compartment.extracellularIndexs))+ sum(c.counts(:, sim.compartment.extracellularIndexs)) &#39;molecules&#39;
<a name="l00319"></a>00319                 }];        
<a name="l00320"></a>00320             
<a name="l00321"></a>00321             %% DNA Polymerase
<a name="l00322"></a>00322             r = sim.process(<span class="stringliteral">&#39;Replication&#39;</span>);
<a name="l00323"></a>00323             free = r.enzymes(r.enzymeIndexs_core) + r.enzymeComposition(r.enzymeIndexs_core, :) * r.enzymes;
<a name="l00324"></a>00324             bound = r.boundEnzymes(r.enzymeIndexs_core) + r.enzymeComposition(r.enzymeIndexs_core, :) * r.boundEnzymes;            
<a name="l00325"></a>00325             content = [content;{
<a name="l00326"></a>00326                 0 &#39;DNA polymerase&#39; free+bound &#39;molecules&#39;
<a name="l00327"></a>00327                     1 &#39;Free&#39; free &#39;molecules&#39;
<a name="l00328"></a>00328                     1 &#39;Bound&#39; bound &#39;molecules&#39;
<a name="l00329"></a>00329                 }];
<a name="l00330"></a>00330             
<a name="l00331"></a>00331             %% <a class="code" href="class_transcription.html">Transcription</a> enzymes
<a name="l00332"></a>00332             t = sim.process(<span class="stringliteral">&#39;Transcription&#39;</span>);
<a name="l00333"></a>00333             s = sim.state(<span class="stringliteral">&#39;RNAPolymerase&#39;</span>);
<a name="l00334"></a>00334             rna = t.rnaPolymerases;
<a name="l00335"></a>00335             rnaPolymeraseStateOccupancies = s.stateOccupancies*100;
<a name="l00336"></a>00336             content = [content;{
<a name="l00337"></a>00337                 0 &#39;Transcription enzymes&#39; [] []                
<a name="l00338"></a>00338                     1 &#39;RNA polymerase&#39; t.enzymes(t.enzymeIndexs_rnaPolymerase)+t.boundEnzymes(t.enzymeIndexs_rnaPolymerase) &#39;molecules&#39;
<a name="l00339"></a>00339                         2 &#39;Active&#39; rnaPolymeraseStateOccupancies(rna.activelyTranscribingIndex) &#39; %&#39;
<a name="l00340"></a>00340                         2 &#39;Specifically bound&#39; rnaPolymeraseStateOccupancies(rna.specificallyBoundIndex) &#39; %&#39;
<a name="l00341"></a>00341                         2 &#39;Non-specifically bound&#39; rnaPolymeraseStateOccupancies(rna.nonSpecificallyBoundIndex) &#39; %&#39;
<a name="l00342"></a>00342                         2 &#39;Free&#39; rnaPolymeraseStateOccupancies(rna.freeIndex) &#39; %&#39;
<a name="l00343"></a>00343                     1 &#39;Initiation/Elongation factors&#39; sum(t.enzymes(t.enzymeIndexs_transcriptionFactors)+t.boundEnzymes(t.enzymeIndexs_transcriptionFactors)) &#39;molecules&#39;
<a name="l00344"></a>00344                         2 &#39;Free&#39; sum(t.enzymes(t.enzymeIndexs_transcriptionFactors)) &#39;molecules&#39;
<a name="l00345"></a>00345                         2 &#39;Bound&#39; sum(t.boundEnzymes(t.enzymeIndexs_transcriptionFactors)) &#39;molecules&#39;
<a name="l00346"></a>00346                 }];        
<a name="l00347"></a>00347             
<a name="l00348"></a>00348             %% <a class="code" href="classt_r_n_a.html">tRNA</a> aminoacylation enzymes
<a name="l00349"></a>00349             t = sim.process(<span class="stringliteral">&#39;tRNAAminoacylation&#39;</span>);
<a name="l00350"></a>00350             totEnzymes = t.enzymes + t.boundEnzymes;
<a name="l00351"></a>00351             content = [content;{
<a name="l00352"></a>00352                 0 &#39;tRNA aminoacylation enzymes&#39; sum(totEnzymes) &#39;molecules&#39;
<a name="l00353"></a>00353                     1 &#39;tRNA synthetases&#39; sum(totEnzymes(t.enzymeIndexs_tRNASynthetases)) &#39;molecules&#39;
<a name="l00354"></a>00354                     1 &#39;tRNA transferases&#39; sum(totEnzymes(t.enzymeIndexs_tRNATransferases)) &#39;molecules&#39;
<a name="l00355"></a>00355                 }];
<a name="l00356"></a>00356                                     
<a name="l00357"></a>00357             %% <a class="code" href="class_translation.html">Translation</a> enzymes
<a name="l00358"></a>00358             t = sim.process(<span class="stringliteral">&#39;Translation&#39;</span>);
<a name="l00359"></a>00359             r = sim.state(<span class="stringliteral">&#39;Ribosome&#39;</span>);
<a name="l00360"></a>00360             totEnzymes = t.enzymes + t.boundEnzymes;
<a name="l00361"></a>00361             content = [content;{
<a name="l00362"></a>00362                 0 &#39;Translation enzymes&#39; [] []
<a name="l00363"></a>00363                     1 &#39;Ribosome&#39; [] []
<a name="l00364"></a>00364                         2 &#39;Ribosome 30S&#39; totEnzymes(t.enzymeIndexs_ribosome30S) &#39;molecules&#39;
<a name="l00365"></a>00365                         2 &#39;Ribosome 30S-IF3&#39; totEnzymes(t.enzymeIndexs_ribosome30SIF3) &#39;molecules&#39;
<a name="l00366"></a>00366                         2 &#39;Ribosome 50S&#39; totEnzymes(t.enzymeIndexs_ribosome50S) &#39;molecules&#39;
<a name="l00367"></a>00367                         2 &#39;Ribosome 70S&#39; totEnzymes(t.enzymeIndexs_ribosome70S) &#39;molecules&#39;
<a name="l00368"></a>00368                             3 &#39;Free&#39; t.enzymes(t.enzymeIndexs_ribosome70S) &#39;molecules&#39;
<a name="l00369"></a>00369                             3 &#39;Bound&#39; t.boundEnzymes(t.enzymeIndexs_ribosome70S) &#39;molecules&#39;
<a name="l00370"></a>00370                             3 &#39;Active&#39; nnz(r.boundMRNAs) &#39;molecules&#39;
<a name="l00371"></a>00371                     1 &#39;Translation factors&#39; sum(totEnzymes(t.enzymeIndexs_translationFactors)) &#39;molecules&#39;
<a name="l00372"></a>00372                         1 &#39;Free&#39; sum(t.enzymes(t.enzymeIndexs_translationFactors)) &#39;molecules&#39;
<a name="l00373"></a>00373                         1 &#39;Bound&#39; sum(t.boundEnzymes(t.enzymeIndexs_translationFactors)) &#39;molecules&#39;
<a name="l00374"></a>00374                 }];                       
<a name="l00375"></a>00375             
<a name="l00376"></a>00376             %% FtsZ
<a name="l00377"></a>00377             f = sim.process(<span class="stringliteral">&#39;FtsZPolymerization&#39;</span>);
<a name="l00378"></a>00378             totEnzymes = f.enzymes + f.boundEnzymes;
<a name="l00379"></a>00379             content = [content;{
<a name="l00380"></a>00380                 0 &#39;FtsZ&#39; [1 1 1 2:9] * totEnzymes &#39;molecules&#39;
<a name="l00381"></a>00381                     1 &#39;Free monomers&#39; totEnzymes(f.enzymeIndexs_FtsZ) &#39;molecules&#39;
<a name="l00382"></a>00382                     1 &#39;GDP 1-mer&#39; totEnzymes(f.enzymeIndexs_FtsZ_GDP) &#39;molecules&#39;
<a name="l00383"></a>00383                     1 &#39;GTP 1-mer&#39; totEnzymes(f.enzymeIndexs_FtsZ_GTP) &#39;molecules&#39;
<a name="l00384"></a>00384                     1 &#39;GTP 2-mer&#39; totEnzymes(f.enzymeIndexs_FtsZ_activated(2)) &#39;molecules&#39;
<a name="l00385"></a>00385                     1 &#39;GTP 3-mer&#39; totEnzymes(f.enzymeIndexs_FtsZ_activated(3)) &#39;molecules&#39;
<a name="l00386"></a>00386                     1 &#39;GTP 4-mer&#39; totEnzymes(f.enzymeIndexs_FtsZ_activated(4)) &#39;molecules&#39;
<a name="l00387"></a>00387                     1 &#39;GTP 5-mer&#39; totEnzymes(f.enzymeIndexs_FtsZ_activated(5)) &#39;molecules&#39;
<a name="l00388"></a>00388                     1 &#39;GTP 6-mer&#39; totEnzymes(f.enzymeIndexs_FtsZ_activated(6)) &#39;molecules&#39;
<a name="l00389"></a>00389                     1 &#39;GTP 7-mer&#39; totEnzymes(f.enzymeIndexs_FtsZ_activated(7)) &#39;molecules&#39;
<a name="l00390"></a>00390                     1 &#39;GTP 8-mer&#39; totEnzymes(f.enzymeIndexs_FtsZ_activated(8)) &#39;molecules&#39;
<a name="l00391"></a>00391                     1 &#39;GTP 9-mer&#39; totEnzymes(f.enzymeIndexs_FtsZ_activated(9)) &#39;molecules&#39;
<a name="l00392"></a>00392                     1 &#39;Bound 9-mer&#39; f.boundEnzymes(f.enzymeIndexs_FtsZ_activated(9)) &#39;molecules&#39;
<a name="l00393"></a>00393                 }];            
<a name="l00394"></a>00394             
<a name="l00395"></a>00395             %% format output
<a name="l00396"></a>00396             indentation = cell2mat(content(:, 1));
<a name="l00397"></a>00397             content = content(:, 2:end);
<a name="l00398"></a>00398         end
<a name="l00399"></a>00399             
<a name="l00400"></a>00400         %diagnostics
<a name="l00401"></a>00401         <span class="keyword">function</span> [content, colLabels, indentation] = processes(sim)
<a name="l00402"></a>00402             content = cell(0, 4);
<a name="l00403"></a>00403             colLabels = {<span class="stringliteral">&#39;Quantity&#39;</span>, <span class="stringliteral">&#39;Value&#39;</span>, <span class="stringliteral">&#39;Units&#39;</span>};
<a name="l00404"></a>00404             
<a name="l00405"></a>00405             %% <a class="code" href="class_transcription.html">Transcription</a>
<a name="l00406"></a>00406             t = sim.process(<span class="stringliteral">&#39;Transcription&#39;</span>);
<a name="l00407"></a>00407             s = sim.state(<span class="stringliteral">&#39;RNAPolymerase&#39;</span>);
<a name="l00408"></a>00408             
<a name="l00409"></a>00409             %building blocks
<a name="l00410"></a>00410             availableNTPs = t.substrates(t.substrateIndexs_ntp);
<a name="l00411"></a>00411             rnaPolymeraseMaxNTPIncorporation = s.nActive * t.rnaPolymeraseElongationRate;
<a name="l00412"></a>00412             transcriptionFactors = t.enzymes(t.enzymeIndexs_transcriptionFactors);
<a name="l00413"></a>00413             
<a name="l00414"></a>00414             %limits
<a name="l00415"></a>00415             transcriptionNTPLimit = min(availableNTPs, ...
<a name="l00416"></a>00416                 availableNTPs / sum(availableNTPs) * rnaPolymeraseMaxNTPIncorporation);
<a name="l00417"></a>00417             <span class="keywordflow">if</span> abs(rnaPolymeraseMaxNTPIncorporation - sum(transcriptionNTPLimit)) &lt; 1<a class="code" href="json_parse_8m.html#ac14d26badced2970320ee4d772ad0ae9">e</a>-6
<a name="l00418"></a>00418                 transcriptionLimitingFractor = &#39;Active <a class="code" href="class_r_n_a.html">RNA</a> Polymerases&#39;;
<a name="l00419"></a>00419             <span class="keywordflow">else</span>
<a name="l00420"></a>00420                 transcriptionLimitingFractor = <span class="stringliteral">&#39;Available NMPs, NTPs&#39;</span>;
<a name="l00421"></a>00421             end
<a name="l00422"></a>00422             
<a name="l00423"></a>00423             %limiting blocks
<a name="l00424"></a>00424             [minNTPs, minNTP] = min(availableNTPs);
<a name="l00425"></a>00425             minNTP = t.substrateWholeCellModelIDs{t.substrateIndexs_ntp(minNTP)};
<a name="l00426"></a>00426             
<a name="l00427"></a>00427             [minTranscriptionFactors, minTranscriptionFactor] = min(transcriptionFactors);
<a name="l00428"></a>00428             minTranscriptionFactor = t.enzymeWholeCellModelIDs{t.enzymeIndexs_transcriptionFactors(minTranscriptionFactor)};
<a name="l00429"></a>00429             
<a name="l00430"></a>00430             %print
<a name="l00431"></a>00431             content = [content;{
<a name="l00432"></a>00432                 0 &#39;Transcription&#39; [] []
<a name="l00433"></a>00433                     1 &#39;Allocated ATP&#39; t.substrates(t.substrateIndexs_atp) &#39;molecules&#39;
<a name="l00434"></a>00434                     1 &#39;Available NTPs&#39; sum(availableNTPs) &#39;molecules&#39;
<a name="l00435"></a>00435                     1 &#39;Min NTP&#39; minNTPs &#39;molecules&#39;
<a name="l00436"></a>00436                     1 &#39;Min NTP&#39; minNTP []
<a name="l00437"></a>00437                     1 &#39;Active RNA Polymerases&#39; s.nActive &#39;molecules&#39;
<a name="l00438"></a>00438                     1 &#39;RNA Polymerase Elongation Rate&#39; t.rnaPolymeraseElongationRate &#39;nt/s&#39;
<a name="l00439"></a>00439                     1 &#39;Max RNA Polymerase NTP Incorporation&#39;  rnaPolymeraseMaxNTPIncorporation &#39;nt/s&#39;
<a name="l00440"></a>00440                     1 &#39;Transcription factors&#39; sum(transcriptionFactors) &#39;molecules&#39;
<a name="l00441"></a>00441                     1 &#39;Min transcription factors&#39; minTranscriptionFactors &#39;molecules&#39;
<a name="l00442"></a>00442                     1 &#39;Min transcription factors&#39; minTranscriptionFactor []
<a name="l00443"></a>00443                     1 &#39;Max NTP Incorporation&#39; sum(transcriptionNTPLimit) &#39;nt/s&#39;
<a name="l00444"></a>00444                     1 &#39;Limiting Factor&#39; transcriptionLimitingFractor []
<a name="l00445"></a>00445              }];
<a name="l00446"></a>00446             
<a name="l00447"></a>00447             %% <a class="code" href="class_translation.html">Translation</a>
<a name="l00448"></a>00448             t = sim.process(<span class="stringliteral">&#39;Translation&#39;</span>);
<a name="l00449"></a>00449             r = sim.state(<span class="stringliteral">&#39;Ribosome&#39;</span>);
<a name="l00450"></a>00450             
<a name="l00451"></a>00451             %building blocks
<a name="l00452"></a>00452             energy = t.substrates(t.substrateIndexs_gtp);
<a name="l00453"></a>00453             
<a name="l00454"></a>00454             aminoAcids = t.substrates(t.substrateIndexs_aminoAcids);
<a name="l00455"></a>00455             ribosomeMaxAAIncorporation = r.nActive * t.ribosomeElongationRate;
<a name="l00456"></a>00456             
<a name="l00457"></a>00457             tRNAs = t.freeTRNAs + t.aminoacylatedTRNAs;
<a name="l00458"></a>00458             translationFactors = t.enzymes(t.enzymeIndexs_translationFactors) + t.boundEnzymes(t.enzymeIndexs_translationFactors);
<a name="l00459"></a>00459             
<a name="l00460"></a>00460             %limiting blocks
<a name="l00461"></a>00461             [minAminoAcids, minAminoAcid] = min(aminoAcids);
<a name="l00462"></a>00462             minAminoAcid = t.substrateWholeCellModelIDs{t.substrateIndexs_aminoAcids(minAminoAcid)};
<a name="l00463"></a>00463             
<a name="l00464"></a>00464             [minTRNAs, minTRNA] = min(tRNAs);
<a name="l00465"></a>00465             minTRNA = t.freeTRNAWholeCellModelIDs{minTRNA};
<a name="l00466"></a>00466             
<a name="l00467"></a>00467             [minTranslationFactors, minTranslationFactor]=min(translationFactors);
<a name="l00468"></a>00468             minTranslationFactor = t.enzymeWholeCellModelIDs{t.enzymeIndexs_translationFactors(minTranslationFactor)};
<a name="l00469"></a>00469             
<a name="l00470"></a>00470             %limits            
<a name="l00471"></a>00471             content = [content;{
<a name="l00472"></a>00472                 0 &#39;Translation&#39; [] []
<a name="l00473"></a>00473                     1 &#39;Allocated ATP&#39; energy &#39;molecules&#39;
<a name="l00474"></a>00474                     1 &#39;Available AAs&#39;  sum(aminoAcids) &#39;molecules&#39;
<a name="l00475"></a>00475                     1 &#39;Min Amino Acid&#39; minAminoAcids &#39;molecules&#39;
<a name="l00476"></a>00476                     1 &#39;Min Amino Acid&#39; minAminoAcid []
<a name="l00477"></a>00477                     1 &#39;tRNAs&#39; sum(tRNAs) &#39;molecules&#39;
<a name="l00478"></a>00478                     1 &#39;Min tRNAs&#39; minTRNAs &#39;molecules&#39;
<a name="l00479"></a>00479                     1 &#39;Min tRNAs&#39; minTRNA []
<a name="l00480"></a>00480                     1 &#39;Active Ribosomes&#39; r.nActive &#39;molecules&#39;
<a name="l00481"></a>00481                     1 &#39;Ribosome Elongation Rate&#39; t.ribosomeElongationRate &#39;aa/s&#39;
<a name="l00482"></a>00482                     1 &#39;Max Ribosome AA Incorporation&#39; ribosomeMaxAAIncorporation &#39;aa/s&#39;
<a name="l00483"></a>00483                     1 &#39;Translation factors&#39; sum(translationFactors) &#39;molecules&#39;
<a name="l00484"></a>00484                     1 &#39;Min translation factors&#39; minTranslationFactors &#39;molecules&#39;
<a name="l00485"></a>00485                     1 &#39;Min translation factors&#39; minTranslationFactor []
<a name="l00486"></a>00486                 }];
<a name="l00487"></a>00487             
<a name="l00488"></a>00488             %% format output
<a name="l00489"></a>00489             indentation = cell2mat(content(:, 1));
<a name="l00490"></a>00490             content = content(:, 2:end);
<a name="l00491"></a>00491         end
<a name="l00492"></a>00492     end
<a name="l00493"></a>00493     
<a name="l00494"></a>00494     %plotting
<a name="l00495"></a>00495     methods (Static)
<a name="l00496"></a>00496         function plotStateOnChromosome(sim, axesHandle)
<a name="l00497"></a>00497             c = sim.state(&#39;<a class="code" href="class_chromosome.html">Chromosome</a>&#39;);
<a name="l00498"></a>00498             
<a name="l00499"></a>00499             cla(axesHandle);
<a name="l00500"></a>00500             hold on;
<a name="l00501"></a>00501             legendHandles = [];
<a name="l00502"></a>00502             
<a name="l00503"></a>00503             %base chromosome
<a name="l00504"></a>00504             for i = 1:4
<a name="l00505"></a>00505                 [r, xoffset] = calcRadiusXOffset(i);
<a name="l00506"></a>00506                 h = plot(axesHandle, xoffset + r*cos(pi/180 * (1:360)), r*sin(pi/180 * (1:360)), &#39;Color&#39;, [0.7 0.7 0.7], &#39;LineStyle&#39;, &#39;:&#39;);
<a name="l00507"></a>00507             end
<a name="l00508"></a>00508             legendHandles = [legendHandles; h];
<a name="l00509"></a>00509             
<a name="l00510"></a>00510             %polymerized regions
<a name="l00511"></a>00511             for i = 1:4
<a name="l00512"></a>00512                 [pos, lens] = find(c.polymerizedRegions(:, i));
<a name="l00513"></a>00513                 [r, xoffset] = calcRadiusXOffset(i);
<a name="l00514"></a>00514                 for j = 1:size(pos, 1)
<a name="l00515"></a>00515                     theta = 2*pi * (pos(j,1) + (0:lens(j)/360:lens(j)))/c.sequenceLen;
<a name="l00516"></a>00516                     h = plot(axesHandle, xoffset + r*cos(theta), r*sin(theta), &#39;Color&#39;, [0.7 0.7 0.7], &#39;LineWidth&#39;, 2);
<a name="l00517"></a>00517                 end
<a name="l00518"></a>00518             end
<a name="l00519"></a>00519             legendHandles = [legendHandles; h];
<a name="l00520"></a>00520             
<a name="l00521"></a>00521             %polymerized transcription units
<a name="l00522"></a>00522             for i = 1:2
<a name="l00523"></a>00523                 for j = 1:numel(c.transcriptionUnitWholeCellModelIDs)
<a name="l00524"></a>00524                     if ~c.polymerizedTranscriptionUnits(j, i); continue; end;
<a name="l00525"></a>00525                     [r, xoffset] = calcRadiusXOffset(2*(i-1) + c.transcriptionUnitStrands(j));
<a name="l00526"></a>00526                     theta = 2*pi * (c.transcriptionUnitStartCoordinates(j) + (0:c.transcriptionUnitLengths(j)/360:c.transcriptionUnitLengths(j))) / c.sequenceLen;
<a name="l00527"></a>00527                     h = plot(axesHandle, xoffset + r*cos(theta), r*sin(theta), &#39;Color&#39;, [0 1 0], &#39;LineWidth&#39;, 2);
<a name="l00528"></a>00528                 end                
<a name="l00529"></a>00529             end
<a name="l00530"></a>00530             legendHandles = [legendHandles; h];
<a name="l00531"></a>00531             
<a name="l00532"></a>00532             %polymerized genes
<a name="l00533"></a>00533             for i = 1:2
<a name="l00534"></a>00534                 for j = 1:numel(sim.gene.wholeCellModelIDs)
<a name="l00535"></a>00535                     if ~c.polymerizedGenes(j, i); continue; end;
<a name="l00536"></a>00536                     [r, xoffset] = calcRadiusXOffset(2*(i-1) + sim.gene.strands(j));
<a name="l00537"></a>00537                     theta = 2*pi * (sim.gene.startCoordinates(j) + (0:sim.gene.lengths(j)/360:sim.gene.lengths(j))) / c.sequenceLen;
<a name="l00538"></a>00538                     h = plot(axesHandle, xoffset + r*cos(theta), r*sin(theta), &#39;Color&#39;, [0 0 1], &#39;LineWidth&#39;, 2);
<a name="l00539"></a>00539                 end
<a name="l00540"></a>00540             end
<a name="l00541"></a>00541             legendHandles = [legendHandles; h];
<a name="l00542"></a>00542             
<a name="l00543"></a>00543             %bound proteins
<a name="l00544"></a>00544             [monomerPosStrnds, monomers] = find(c.monomerBoundSites);
<a name="l00545"></a>00545             [complexPosStrnds, complexs] = find(c.complexBoundSites);
<a name="l00546"></a>00546             posStrnds = [monomerPosStrnds; complexPosStrnds];
<a name="l00547"></a>00547             lens = [c.monomerDNAFootprints(monomers); c.complexDNAFootprints(complexs)];
<a name="l00548"></a>00548             ssStranded = [c.monomerDNAFootprintRegionStrandedness(monomers); c.complexDNAFootprintRegionStrandedness(complexs)] == c.dnaStrandedness_ssDNA;
<a name="l00549"></a>00549             for i = 1:size(posStrnds, 1)
<a name="l00550"></a>00550                 [r, xoffset] = calcRadiusXOffset(posStrnds(i,2));
<a name="l00551"></a>00551                 theta = 2*pi * (posStrnds(i,1) + (0:lens(i)/360:lens(i))) / c.sequenceLen;
<a name="l00552"></a>00552                 h = plot(axesHandle, xoffset + r*cos(theta), r*sin(theta), &#39;Color&#39;, [1 1 0], &#39;LineWidth&#39;, 4);
<a name="l00553"></a>00553                 if ~ssStranded(i)
<a name="l00554"></a>00554                     [r, xoffset] = calcRadiusXOffset(posStrnds(i, 2) -iseven(posStrnds(i, 2)) + isodd(posStrnds(i, 2)));
<a name="l00555"></a>00555                     plot(axesHandle, xoffset + r*cos(theta), r*sin(theta), &#39;Color&#39;, [1 1 0], &#39;LineWidth&#39;, 4);
<a name="l00556"></a>00556                 end
<a name="l00557"></a>00557             end
<a name="l00558"></a>00558             legendHandles = [legendHandles; h];
<a name="l00559"></a>00559             
<a name="l00560"></a>00560             %damage
<a name="l00561"></a>00561             posStrnds = find(c.getDamagedSites(1,1,1,1,1,1,1));
<a name="l00562"></a>00562             for i = 1:size(posStrnds, 1)
<a name="l00563"></a>00563                 [r, xoffset] = calcRadiusXOffset(posStrnds(i,2));
<a name="l00564"></a>00564                 theta = 2*pi * posStrnds(i,1) / c.sequenceLen;
<a name="l00565"></a>00565                 h = plot(axesHandle, xoffset + r*cos(theta), r*sin(theta), &#39;Color&#39;, [1 0 0], &#39;LineWidth&#39;, 4);
<a name="l00566"></a>00566             end
<a name="l00567"></a>00567             legendHandles = [legendHandles; h];
<a name="l00568"></a>00568             
<a name="l00569"></a>00569             %expressed RNAs
<a name="l00570"></a>00570             r = sim.state(&#39;Rna&#39;);
<a name="l00571"></a>00571             genes = r.matureRNAGeneComposition * sum(r.counts(r.matureIndexs, :) + r.counts(r.aminoacylatedIndexs, :), 2);
<a name="l00572"></a>00572             for i = 1:numel(sim.gene.wholeCellModelIDs)
<a name="l00573"></a>00573                 if genes(i) == 0; continue; end;
<a name="l00574"></a>00574                 
<a name="l00575"></a>00575                 r = 1 + [0.05;0.25] * iseven(sim.gene.strands(i)) - [0.05;0.25] * isodd(sim.gene.strands(i));
<a name="l00576"></a>00576                 xoffset = 0;
<a name="l00577"></a>00577                 
<a name="l00578"></a>00578                 theta = 2*pi * (sim.gene.startCoordinates(i) + sim.gene.lengths(i)/2) / c.sequenceLen;
<a name="l00579"></a>00579                 h = plot(axesHandle, xoffset + r*cos(theta), r*sin(theta), &#39;Color&#39;, [1 0 1], &#39;LineWidth&#39;, 2);
<a name="l00580"></a>00580             end
<a name="l00581"></a>00581             legendHandles = [legendHandles; h];
<a name="l00582"></a>00582             
<a name="l00583"></a>00583             %expressed proteins
<a name="l00584"></a>00584             pm = sim.state(&#39;<a class="code" href="class_protein_monomer.html">ProteinMonomer</a>&#39;);
<a name="l00585"></a>00585             pc = sim.state(&#39;<a class="code" href="class_protein_complex.html">ProteinComplex</a>&#39;);
<a name="l00586"></a>00586              
<a name="l00587"></a>00587             monomers = ...
<a name="l00588"></a>00588                 sum(pm.counts(pm.matureIndexs, :) + pm.counts(pm.boundIndexs, :) + pm.counts(pm.inactivatedIndexs, :), 2) + ...
<a name="l00589"></a>00589                 sum(pc.proteinComplexComposition(sim.gene.mRNAIndexs, :, :), 3) * sum(pc.counts(pc.matureIndexs, :) + pc.counts(pc.boundIndexs, :) + pc.counts(pc.inactivatedIndexs, :), 2);
<a name="l00590"></a>00590             for i = 1:numel(monomers)
<a name="l00591"></a>00591                 if monomers(i) == 0; continue; end;
<a name="l00592"></a>00592                 
<a name="l00593"></a>00593                 r = 1 + [0.05;0.25] * iseven(sim.gene.strands(sim.gene.mRNAIndexs(i))) - [0.05;0.25] * isodd(sim.gene.strands(sim.gene.mRNAIndexs(i)));
<a name="l00594"></a>00594                 xoffset = 0;
<a name="l00595"></a>00595                 
<a name="l00596"></a>00596                 theta = 2*pi * (sim.gene.startCoordinates(sim.gene.mRNAIndexs(i)) + sim.gene.lengths(sim.gene.mRNAIndexs(i))/2) / c.sequenceLen;
<a name="l00597"></a>00597                 h = plot(axesHandle, xoffset + r*cos(theta), r*sin(theta), &#39;Color&#39;, [0 1 1], &#39;LineWidth&#39;, 2);
<a name="l00598"></a>00598             end
<a name="l00599"></a>00599             legendHandles = [legendHandles; h];
<a name="l00600"></a>00600             
<a name="l00601"></a>00601             %ribosomes
<a name="l00602"></a>00602             ribosome = sim.state(&#39;Ribosome&#39;);
<a name="l00603"></a>00603             for i = 1:numel(ribosome.boundMRNAs)
<a name="l00604"></a>00604                 if ribosome.boundMRNAs(i) == 0; continue; end;
<a name="l00605"></a>00605                 
<a name="l00606"></a>00606                 r = 1 + [0.05;0.25] * iseven(sim.gene.strands(sim.gene.mRNAIndexs(ribosome.boundMRNAs(i)))) - [0.05;0.25] * isodd(sim.gene.strands(sim.gene.mRNAIndexs(ribosome.boundMRNAs(i))));
<a name="l00607"></a>00607                 xoffset = 0;
<a name="l00608"></a>00608                 
<a name="l00609"></a>00609                 theta = 2*pi * (sim.gene.startCoordinates(sim.gene.mRNAIndexs(ribosome.boundMRNAs(i))) + ribosome.states(i)) / c.sequenceLen;
<a name="l00610"></a>00610                 plot(axesHandle, xoffset + r*cos(theta), r*sin(theta), &#39;Color&#39;, [1 1 0], &#39;LineWidth&#39;, 2);
<a name="l00611"></a>00611             end
<a name="l00612"></a>00612 
<a name="l00613"></a>00613             %legend            
<a name="l00614"></a>00614             legend(legendHandles, {
<a name="l00615"></a>00615                 <span class="stringliteral">&#39;Chromosome&#39;</span>
<a name="l00616"></a>00616                 <span class="stringliteral">&#39;Polymerized&#39;</span>
<a name="l00617"></a>00617                 <span class="stringliteral">&#39;  Gene&#39;</span>
<a name="l00618"></a>00618                 <span class="stringliteral">&#39;  TU&#39;</span>
<a name="l00619"></a>00619                 <span class="stringliteral">&#39;Bnd Prot&#39;</span>
<a name="l00620"></a>00620                 <span class="stringliteral">&#39;Damage&#39;</span>
<a name="l00621"></a>00621                 <span class="stringliteral">&#39;Exp RNA&#39;</span>
<a name="l00622"></a>00622                 <span class="stringliteral">&#39;Exp Prot&#39;</span>}, ...
<a name="l00623"></a>00623                 <span class="stringliteral">&#39;Orientation&#39;</span>, <span class="stringliteral">&#39;Horizontal&#39;</span>, ...
<a name="l00624"></a>00624                 <span class="stringliteral">&#39;Location&#39;</span>, <span class="stringliteral">&#39;SouthOutside&#39;</span>);
<a name="l00625"></a>00625             
<a name="l00626"></a>00626             %format            
<a name="l00627"></a>00627             xlim([-1.4 2.5+1.4]);
<a name="l00628"></a>00628             ylim([-1.4 1.4]);
<a name="l00629"></a>00629             
<a name="l00630"></a>00630             <span class="keyword">set</span>(axesHandle, <span class="stringliteral">&#39;DataAspectRatio&#39;</span>, [1 1 1])
<a name="l00631"></a>00631             <span class="keyword">set</span>(axesHandle, <span class="stringliteral">&#39;Visible&#39;</span>, <span class="stringliteral">&#39;off&#39;</span>)
<a name="l00632"></a>00632             <span class="keyword">set</span>(<span class="keyword">get</span>(axesHandle, <span class="stringliteral">&#39;Parent&#39;</span>), <span class="stringliteral">&#39;Color&#39;</span>, [1 1 1]);
<a name="l00633"></a>00633             position = <span class="keyword">get</span>(gca, <span class="stringliteral">&#39;Position&#39;</span>);
<a name="l00634"></a>00634             scale = 1.35;
<a name="l00635"></a>00635             <span class="keyword">set</span>(gca, <span class="stringliteral">&#39;Position&#39;</span>, [position(1:2)-position(3:4)*(scale-1)/2 position(3:4)*scale]);
<a name="l00636"></a>00636             
<a name="l00637"></a>00637             <span class="keyword">function</span> [r, xoffset] = calcRadiusXOffset(strnd)
<a name="l00638"></a>00638                 r = 1 + 0.05 * iseven(strnd) - 0.05 * isodd(strnd);
<a name="l00639"></a>00639                 xoffset = 2.5 * floor((strnd-1)/2);
<a name="l00640"></a>00640             end
<a name="l00641"></a>00641         end
<a name="l00642"></a>00642     end
<a name="l00643"></a>00643 end
</pre></div></div><!-- contents -->
</div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
<a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(0)"><span class="SelectionMark">&#160;</span>All</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(1)"><span class="SelectionMark">&#160;</span>Classes</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(2)"><span class="SelectionMark">&#160;</span>Files</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(3)"><span class="SelectionMark">&#160;</span>Functions</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(4)"><span class="SelectionMark">&#160;</span>Variables</a></div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

  <div id="nav-path" class="navpath">
    <ul>
      <li class="navelem"><a class="el" href="_09analysis_2_cell_state_8m.html">CellState.m</a>      </li>

    <li class="footer">Generated on Wed Feb 8 2012 21:42:31 for Whole Cell Simulation by
    <a href="http://www.doxygen.org/index.html">
    <img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.7.6.1 </li>
   </ul>
 </div>


</body>
</html>
