<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<title>Whole Cell Simulation: src/+edu/+stanford/+covert/+cell/+sim/+process/DNADamage.m Source File</title>

<link href="tabs.css" rel="stylesheet" type="text/css"/>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
<link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="navtree.js"></script>
<script type="text/javascript">
  $(document).ready(initResizable);
</script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
  $(document).ready(function() { searchBox.OnSelectItem(0); });
</script>

</head>
<body>
<div id="top"><!-- do not remove this div! -->


<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  
  
  <td style="padding-left: 0.5em;">
   <div id="projectname">Whole Cell Simulation
   
   </div>
   
  </td>
  
  
  
 </tr>
 </tbody>
</table>
</div>

<!-- Generated by Doxygen 1.7.6.1 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "search",false,'Search');
</script>
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&#160;Page</span></a></li>
      <li><a href="annotated.html"><span>Classes</span></a></li>
      <li class="current"><a href="files.html"><span>Files</span></a></li>
      <li>
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
  <div id="navrow2" class="tabs2">
    <ul class="tablist">
      <li><a href="files.html"><span>File&#160;List</span></a></li>
      <li><a href="globals.html"><span>File&#160;Members</span></a></li>
    </ul>
  </div>
</div>
<div id="side-nav" class="ui-resizable side-nav-resizable">
  <div id="nav-tree">
    <div id="nav-tree-contents">
    </div>
  </div>
  <div id="splitbar" style="-moz-user-select:none;" 
       class="ui-resizable-handle">
  </div>
</div>
<script type="text/javascript">
  initNavTree('_09process_2_d_n_a_damage_8m.html','');
</script>
<div id="doc-content">
<div class="header">
  <div class="headertitle">
<div class="title">src/+edu/+stanford/+covert/+cell/+sim/+process/DNADamage.m</div>  </div>
</div><!--header-->
<div class="contents">
<a href="_09process_2_d_n_a_damage_8m.html">Go to the documentation of this file.</a><div class="fragment"><pre class="fragment"><a name="l00001"></a>00001 %DNA Damage
<a name="l00002"></a>00002 %
<a name="l00003"></a>00003 % @wholeCellModelID Process_DNADamage
<a name="l00004"></a>00004 % @name             DNA Damage
<a name="l00005"></a>00005 % @description
<a name="l00006"></a>00006 %   Biology
<a name="l00007"></a>00007 %   ======================
<a name="l00008"></a>00008 %   Simulates a spectrum of DNA lesions repaired by base excision repair,
<a name="l00009"></a>00009 %   nucleotide excision repair, and homologous recombination repair:
<a name="l00010"></a>00010 %   - DNA strand breaks: caused by
<a name="l00011"></a>00011 %     - hydrolysis and beta-elimination of abasic sites
<a name="l00012"></a>00012 %     - ionizing radiation (alpha, beta particles, protons, electrons,
<a name="l00013"></a>00013 %       gamma rays)
<a name="l00014"></a>00014 %   - Spontaneous base loss
<a name="l00015"></a>00015 %   - Base damage: caused by
<a name="l00016"></a>00016 %     - Spontaneous deamination
<a name="l00017"></a>00017 %     - oxidation -- oxygen radicals generated by metabolism, ionizing radiation, chemicals
<a name="l00018"></a>00018 %     - alkylation -- carbon radicals generated by metabolism, ionizing radiation, chemicals
<a name="l00019"></a>00019 %     - amination -- nitrogen radicals generated by metabolism, ionizing radiation, chemicals
<a name="l00020"></a>00020 %   - Intrastrand cross links: caused by
<a name="l00021"></a>00021 %     - UV-B induced cyclization
<a name="l00022"></a>00022 %
<a name="l00023"></a>00023 %   The simulation does not consider the following types of DNA damage
<a name="l00024"></a>00024 %   - interstrand cross links: caused by bifunctional akylating agents
<a name="l00025"></a>00025 %   - psoralens
<a name="l00026"></a>00026 %   - DNA-protein cross links
<a name="l00027"></a>00027 %   - Sugar-phosphate damage other than strand breaks
<a name="l00028"></a>00028 %
<a name="l00029"></a>00029 %   Additionally, the simulation does not consider mutiple damages per
<a name="l00030"></a>00030 %   base or sugar-phosphate.
<a name="l00031"></a>00031 %
<a name="l00032"></a>00032 %   Categorizes strand breaks in DNA segments (216 bases) according to
<a name="l00033"></a>00033 %   scheme described in the track structure model literature [PUB_0486]:
<a name="l00034"></a>00034 %   - NB     -- no strand break
<a name="l00035"></a>00035 %   - SSB    -- single strand break
<a name="l00036"></a>00036 %   - SSB+   -- multiple single strand breaks on segment all on same strand
<a name="l00037"></a>00037 %   - 2SSB   -- single strand breaks on both strand separated by at least 10 bases on DNA segment
<a name="l00038"></a>00038 %   - DSB    -- two single strand breaks on opposite strands within 10 bases
<a name="l00039"></a>00039 %   - DSB+   -- double strand break with additional strand break on DNA segment
<a name="l00040"></a>00040 %   - DSB++  -- two single strand breaks on same DNA segment
<a name="l00041"></a>00041 %
<a name="l00042"></a>00042 %   Effect of DNA damage on physiology
<a name="l00043"></a>00043 %   ++++++++++++++++++++++
<a name="l00044"></a>00044 %   The exact effect of DNA damage on physiology is not well characterized.
<a name="l00045"></a>00045 %   Certain kinds of damage are believed to arrest transcription
<a name="l00046"></a>00046 %   [PUB_0780]. Tornaletti and Hanawalt provide the most extensive review
<a name="l00047"></a>00047 %   of the effect of specific damages on transcription [PUB_0781, PUB_0782,
<a name="l00048"></a>00048 %   PUB_0783]. For simplicity we we model all damages (except m6AD involved
<a name="l00049"></a>00049 %   in R/M systems) as barriers to protein binding and function. This is
<a name="l00050"></a>00050 %   implemented with the chromosome state class.
<a name="l00051"></a>00051 %
<a name="l00052"></a>00052 %   Knowledge Base
<a name="l00053"></a>00053 %   ======================
<a name="l00054"></a>00054 %   As of 8/12/2010 the M. genitalium knowledge contained the stochiometry
<a name="l00055"></a>00055 %   and kinetics of 128 DNA damage reactions involving 15 damaging agents
<a name="l00056"></a>00056 %   and 9 spontaneous reactions. These reactions were compiled from
<a name="l00057"></a>00057 %   several sources. The rates of the most frequent types of damage are
<a name="l00058"></a>00058 %   listed below.
<a name="l00059"></a>00059 %
<a name="l00060"></a>00060 %      Type                                            Reactions
<a name="l00061"></a>00061 %      ============================================    =========
<a name="l00062"></a>00062 %      base alkylation                                   7
<a name="l00063"></a>00063 %      base amination                                    3
<a name="l00064"></a>00064 %      base ethylation                                  39
<a name="l00065"></a>00065 %      base glucosyl transfer                            2
<a name="l00066"></a>00066 %      base methylation                                 42
<a name="l00067"></a>00067 %      base reduction                                    1
<a name="l00068"></a>00068 %      photooxidation                                    2
<a name="l00069"></a>00069 %      radiation (gamma-ray) induced base oxidation     19
<a name="l00070"></a>00070 %      spontaneous base deamination                      4
<a name="l00071"></a>00071 %      spontaneous base loss                             4
<a name="l00072"></a>00072 %      strand break                                      2
<a name="l00073"></a>00073 %      UV-B photodimerization                           12
<a name="l00074"></a>00074 %
<a name="l00075"></a>00075 %      Damaging agent                                  Reactions
<a name="l00076"></a>00076 %      ============================================    =========
<a name="l00077"></a>00077 %      butylene glycol                                   1
<a name="l00078"></a>00078 %      CH3                                               4
<a name="l00079"></a>00079 %      chloroethylene oxide                              5
<a name="l00080"></a>00080 %      DES                                              13
<a name="l00081"></a>00081 %      EMS                                              13
<a name="l00082"></a>00082 %      ENU                                              13
<a name="l00083"></a>00083 %      Gamma radiation                                   1
<a name="l00084"></a>00084 %      hydrogen radical                                  1
<a name="l00085"></a>00085 %      hydroxyl radical                                 19
<a name="l00086"></a>00086 %      MMS                                               9
<a name="l00087"></a>00087 %      MMSO4                                             7
<a name="l00088"></a>00088 %      MNNG                                             10
<a name="l00089"></a>00089 %      MNU                                              12
<a name="l00090"></a>00090 %      NH3                                               1
<a name="l00091"></a>00091 %      Spontaneous                                       9
<a name="l00092"></a>00092 %      UVB radiation                                    14
<a name="l00093"></a>00093 %
<a name="l00094"></a>00094 %      Damage                          Spec Rate             Rate (1/s)   No (1/Cycle)
<a name="l00095"></a>00095 %      ============================    ===================   ==========   ============
<a name="l00096"></a>00096 %      Spontaneous base deamination    1.1e-12 1/nt/s        1.4e-06      0.1
<a name="l00097"></a>00097 %      Spontaneous base loss           7.2e-11 1/nt/s        8.4e-05      7.8
<a name="l00098"></a>00098 %      UV-B Dimer                      1.2e-07 1/nt/(J/m^2)  6.0e-01      5.6e4
<a name="l00099"></a>00099 %      \gamma-ray base oxidation       5.2e-07 1/nt/Gy       1.7e-11      1.6e-6
<a name="l00100"></a>00100 %      \gamma-ray strand break         1.7e-07 1/nt/Gy       5.6e-12      5.2e-7
<a name="l00101"></a>00101 %      base alkylation                 ?                     ?            ?
<a name="l00102"></a>00102 %      base amination                  ?                     ?            ?
<a name="l00103"></a>00103 %      base ethylation                 ?                     ?            ?
<a name="l00104"></a>00104 %      base glucosyl transfer          ?                     ?            ?
<a name="l00105"></a>00105 %      base methylation                ?                     ?            ?
<a name="l00106"></a>00106 %      base reduction                  ?                     ?            ?
<a name="l00107"></a>00107 %      photooxidation                  ?                     ?            ?
<a name="l00108"></a>00108 %
<a name="l00109"></a>00109 %   Representation
<a name="l00110"></a>00110 %   ======================
<a name="l00111"></a>00111 %   7 properties are used to keep track of the damage state of DNA. The first
<a name="l00112"></a>00112 %   dimension of each has length equal to that of the chromosome.
<a name="l00113"></a>00113 %   - gapSites                  true/false    sugar gaps
<a name="l00114"></a>00114 %   - abasicSites               true/false    abasic sites
<a name="l00115"></a>00115 %   - damagedSugarPhosphates    enumeration   damaged sugar phosphates
<a name="l00116"></a>00116 %   - damagedBases              enumeration   damaged bases
<a name="l00117"></a>00117 %   - intrastrandCrossLinks     true/false    intrastand cross link
<a name="l00118"></a>00118 %   - strandBreaks              true/false    strand break (missing sugar-phosphate bond)
<a name="l00119"></a>00119 %   - hollidayJunctions         true/false    holliday junction
<a name="l00120"></a>00120 %
<a name="l00121"></a>00121 %   substrates represents the amounts of free metabolites. initializeConstants
<a name="l00122"></a>00122 %   synchronizes the substrates of this process with that of the DNA repair
<a name="l00123"></a>00123 %   process to make it easy for these processes to communicate. That is
<a name="l00124"></a>00124 %   initializeConstants sets the substrates of both processes equal to the union
<a name="l00125"></a>00125 %   of the two, are orders the substrates the same in both processes. Thus
<a name="l00126"></a>00126 %   substrate indices defined in process are valid on properties received from
<a name="l00127"></a>00127 %   the other process, and it is unecessary to remap properties in
<a name="l00128"></a>00128 %   copyFromState.
<a name="l00129"></a>00129 %
<a name="l00130"></a>00130 %   stimuli represents the amounts of DNA damaging agents including several
<a name="l00131"></a>00131 %   kinds of radiation.
<a name="l00132"></a>00132 %
<a name="l00133"></a>00133 %   This process uses three properties rather than reactionStoichiometryMatrix
<a name="l00134"></a>00134 %   to represent DNA damaging reactions and the DNA, damaging agents, and free
<a name="l00135"></a>00135 %   metabolites that participate in them. These three matrices are a
<a name="l00136"></a>00136 %   factorization of the super class reactionStoichiometryMatrix property.
<a name="l00137"></a>00137 %   - reactionSmallMoleculeStoichiometryMatrix: the stochiometry of free
<a name="l00138"></a>00138 %     metabolites in damage reactions
<a name="l00139"></a>00139 %   - reactionDNAStoichiometryMatrix: the stoichiometry of DNA bases in
<a name="l00140"></a>00140 %     damage reactions
<a name="l00141"></a>00141 %   - reactionRadiationStoichiometryMatrix: the stoichiometry of radiation
<a name="l00142"></a>00142 %     stimuli in damage reactions
<a name="l00143"></a>00143 %
<a name="l00144"></a>00144 %   This process uses reactionVulnerableMotifs, rather the super class
<a name="l00145"></a>00145 %   reactionModificationMatrix property to represent the affects of the DNA
<a name="l00146"></a>00146 %   damage reaction on macromolecules. reactionVulnerableMotifs represents the DNA
<a name="l00147"></a>00147 %   sequence(s) vulnerable to each damage reaction.
<a name="l00148"></a>00148 %
<a name="l00149"></a>00149 %   Initialization
<a name="l00150"></a>00150 %   ======================
<a name="l00151"></a>00151 %   The DNA is initialized to the undamaged state, except that MunI
<a name="l00152"></a>00152 %   restriction/modification sites are initialized to the methylated state (see
<a name="l00153"></a>00153 %   DNA Repair process).
<a name="l00154"></a>00154 %
<a name="l00155"></a>00155 %   Simulation
<a name="l00156"></a>00156 %   ======================
<a name="l00157"></a>00157 %   For each reaction
<a name="l00158"></a>00158 %     1. Compute rate that individual vulnerable DNA regions succombs to damage
<a name="l00159"></a>00159 %     2. Randomly select among vulnerable DNA regions at rate computed in step
<a name="l00160"></a>00160 %        (1)
<a name="l00161"></a>00161 %     3. Mark DNA selected in step (2) as damaged by updating gapSites,
<a name="l00162"></a>00162 %        abasicSites, damagedSugarPhosphates, damagedBases,
<a name="l00163"></a>00163 %        intrastrandCrossLinks, strandBreaks, and hollidayJunctions
<a name="l00164"></a>00164 %     4. Decrement/increment free metabolites reaction reactants and products
<a name="l00165"></a>00165 %        using reactionSmallMoleculeStoichiometryMatrix
<a name="l00166"></a>00166 %   end
<a name="l00167"></a>00167 %
<a name="l00168"></a>00168 %   References
<a name="l00169"></a>00169 %   ======================
<a name="l00170"></a>00170 %   1. Errol C Friedberg, Graham C Walker, and Wolfram Siede (1995). DNA
<a name="l00171"></a>00171 %      Repair and Mutagenesis. ASM Press: Washington, DC. [PUB_0461]
<a name="l00172"></a>00172 %   2. Nikjoo H, O&#39;Neill P, Goodhead DT, and Terrissol M. Computational
<a name="l00173"></a>00173 %      modelling of low-energy electron-induced DNA damage by  early physical
<a name="l00174"></a>00174 %      and chemical events (1997). Int J Radiat Biol. 71(5): 467-83.
<a name="l00175"></a>00175 %      [PUB_0486]
<a name="l00176"></a>00176 %   3. Cline SD, Hanawalt PC (2003). Who&#39;s on first in the cellular
<a name="l00177"></a>00177 %      response to DNA damage? Nat Rev Mol Cell Biol. 4(5):361-72.
<a name="l00178"></a>00178 %      [PUB_0780]
<a name="l00179"></a>00179 %   4. Tornaletti S (2005). Transcription arrest at DNA damage sites. Mutat
<a name="l00180"></a>00180 %      Res. 577(1-2):131-45. [PUB_0781]
<a name="l00181"></a>00181 %   5. Tornaletti S     (2009). Transcription-coupled DNA repair: directing
<a name="l00182"></a>00182 %      your effort where it&#39;s most needed. Cell Mol Life Sci.
<a name="l00183"></a>00183 %      66(6):1010-20. [PUB_0782]
<a name="l00184"></a>00184 %   6. Tornaletti S, Hanawalt PC (1999). Effect of DNA lesions on
<a name="l00185"></a>00185 %      transcription elongation. Biochimie. 81 (1-2):139-46. [PUB_0783] 
<a name="l00186"></a>00186 %
<a name="l00187"></a>00187 % Author: Jonathan Karr, jkarr@stanford.edu
<a name="l00188"></a>00188 % Affilitation: Covert Lab, Department of Bioengineering, Stanford University
<a name="l00189"></a>00189 % Last updated: 4/25/2010
<a name="l00190"></a>00190 classdef  DNADamage &lt; edu.stanford.covert.cell.sim.ReactionProcess &amp; edu.stanford.covert.cell.sim.ChromosomeProcessAspect
<a name="l00191"></a>00191     %property annotations
<a name="l00192"></a>00192     properties (Constant)
<a name="l00193"></a>00193         optionNames__              = {}; %names of option <a class="code" href="json_parse_8m.html#a6e4256f69b751d3d7ec4ea52ff5a8dee">properties</a>
<a name="l00194"></a>00194         fixedConstantNames__       = {   %names of fixed constant <a class="code" href="json_parse_8m.html#a6e4256f69b751d3d7ec4ea52ff5a8dee">properties</a>
<a name="l00195"></a>00195             <span class="stringliteral">&#39;reactionVulnerableMotifs&#39;</span>;
<a name="l00196"></a>00196             <span class="stringliteral">&#39;reactionVulnerableMotifTypes&#39;</span>;
<a name="l00197"></a>00197             <span class="stringliteral">&#39;reactionDamageTypes&#39;</span>;
<a name="l00198"></a>00198             <span class="stringliteral">&#39;reactionSmallMoleculeStoichiometryMatrix&#39;</span>;
<a name="l00199"></a>00199             <span class="stringliteral">&#39;reactionDNAStoichiometryMatrix&#39;</span>;
<a name="l00200"></a>00200             <span class="stringliteral">&#39;reactionRadiationStoichiometryMatrix&#39;</span>;
<a name="l00201"></a>00201             <span class="stringliteral">&#39;reactionDNAProduct&#39;</span>;
<a name="l00202"></a>00202             <span class="stringliteral">&#39;reactionRadiation&#39;</span>;
<a name="l00203"></a>00203             };
<a name="l00204"></a>00204         fittedConstantNames__      = {}; %names of fitted constant <a class="code" href="json_parse_8m.html#a6e4256f69b751d3d7ec4ea52ff5a8dee">properties</a>
<a name="l00205"></a>00205         localStateNames__          = {}; %names of simulation state <a class="code" href="json_parse_8m.html#a6e4256f69b751d3d7ec4ea52ff5a8dee">properties</a> redundant with timecourses in <span class="keyword">this</span> or other processes or the simulation
<a name="l00206"></a>00206     end
<a name="l00207"></a>00207     
<a name="l00208"></a>00208     %enumerations
<a name="l00209"></a>00209     <a class="code" href="json_parse_8m.html#a6e4256f69b751d3d7ec4ea52ff5a8dee">properties</a> (Constant)
<a name="l00210"></a>00210         unmodifiedNucleobases = &#39;ACGT&#39;;           %symbols for unmodified bases
<a name="l00211"></a>00211     end
<a name="l00212"></a>00212     
<a name="l00213"></a>00213     %IDs, names, and local indices
<a name="l00214"></a>00214     <a class="code" href="json_parse_8m.html#a6e4256f69b751d3d7ec4ea52ff5a8dee">properties</a>
<a name="l00215"></a>00215         substrateIndexs_radiation                 %indices within substrates of radiation
<a name="l00216"></a>00216         substrateIndexs_hydrogen                  %index within substrates of hydrogen
<a name="l00217"></a>00217         substrateIndexs_hydroxylRadical           %index within substrates of hydroxyl radical
<a name="l00218"></a>00218         substrateIndexs_water                     %index within substrates of water
<a name="l00219"></a>00219         substrateIndexs_DR5P                      %index within substrates of deoxyribose-5-phosphate
<a name="l00220"></a>00220         substrateIndexs_dRibose5P_dRibose5P       %index within substrates of (deoxyribose-5-phosphate)-(deoxyribose-5-phosphate)
<a name="l00221"></a>00221         substrateIndexs_unmodifiedNucleobases     %indices within substrates of unmodified nucleobases
<a name="l00222"></a>00222         substrateIndexs_unmodifiedNucleosides     %indices within substrates of unmodified deoxyribonucleosides
<a name="l00223"></a>00223         substrateIndexs_unmodifiedDNMPs           %indices within substrates of unmodified dNMPs
<a name="l00224"></a>00224         substrateIndexs_modifiedNucleobases       %indices within substrates of modified nucleobase
<a name="l00225"></a>00225         substrateIndexs_modifiedNucleosides       %indices within substrates of modified deoxyribonucleoside
<a name="l00226"></a>00226         substrateIndexs_modifiedDNMPs             %indices within substrates of modified dNMPs
<a name="l00227"></a>00227         
<a name="l00228"></a>00228         reactionIndexs_spontaneousBaseLoss        %index within reactions of spontaneous base loss
<a name="l00229"></a>00229         reactionIndexs_spontaneousBaseDeamination %index within reactions of spontaneous base deamination
<a name="l00230"></a>00230         reactionIndexs_gammaRayBaseOxidation      %index within reactions of gamma ray base oxidation
<a name="l00231"></a>00231         reactionIndexs_uvbDimerization            %index within reactions of UVB dimerization
<a name="l00232"></a>00232         reactionIndexs_baseReduction              %index within reactions of base reduction
<a name="l00233"></a>00233         reactionIndexs_baseAmination              %index within reactions of base amination
<a name="l00234"></a>00234         reactionIndexs_baseGlucosylTransfer       %index within reactions of base glucosyl transfer
<a name="l00235"></a>00235         reactionIndexs_baseAlkylation             %index within reactions of base alkylation
<a name="l00236"></a>00236         reactionIndexs_baseEthylation             %index within reactions of base ethylation
<a name="l00237"></a>00237         reactionIndexs_baseMethylation            %index within reactions of base methylation
<a name="l00238"></a>00238         reactionIndexs_photooxidation             %index within reactions of photooxidation
<a name="l00239"></a>00239         reactionIndexs_strandBreak                %index within reactions of strand break
<a name="l00240"></a>00240     end
<a name="l00241"></a>00241     
<a name="l00242"></a>00242     %fixed biological constants
<a name="l00243"></a>00243     <a class="code" href="json_parse_8m.html#a6e4256f69b751d3d7ec4ea52ff5a8dee">properties</a>
<a name="l00244"></a>00244         reactionVulnerableMotifs                 %DNA base sequence / damaged bases / sugar-phosphate subject to each damage reaction
<a name="l00245"></a>00245         reactionVulnerableMotifTypes             %type of each reactionVulnerableMotif (sequence, gapSites, abasicSites, damagedSugarPhosphates, damagedBases, intrastrandCrossLinks, strandBreaks, or hollidayJunctions)
<a name="l00246"></a>00246         reactionDamageTypes                      %type of damage of reach reaction (gapSites, abasicSites, damagedSugarPhosphates, damagedBases, intrastrandCrossLinks, strandBreaks, or hollidayJunctions)
<a name="l00247"></a>00247         reactionSmallMoleculeStoichiometryMatrix %small molecule stoichiometry of reactions
<a name="l00248"></a>00248         reactionDNAStoichiometryMatrix           %DNA base stoichiometry of reactions
<a name="l00249"></a>00249         reactionRadiationStoichiometryMatrix     %DNA radiation stoichiometry of reactions
<a name="l00250"></a>00250         reactionDNAProduct                       %global index of resultant DNA product of each reaction
<a name="l00251"></a>00251         reactionRadiation                        %local index of radiation causative of each reaction
<a name="l00252"></a>00252     end
<a name="l00253"></a>00253     
<a name="l00254"></a>00254     %constructor
<a name="l00255"></a>00255     methods
<a name="l00256"></a>00256         function this = <a class="code" href="class_d_n_a_damage.html">DNADamage</a>(wholeCellModelID, name)
<a name="l00257"></a>00257             this = this@edu.stanford.covert.cell.sim.<a class="code" href="class_reaction_process.html">ReactionProcess</a>(wholeCellModelID, name);
<a name="l00258"></a>00258         end
<a name="l00259"></a>00259     end
<a name="l00260"></a>00260     
<a name="l00261"></a>00261     %communication between process/simulation
<a name="l00262"></a>00262     methods
<a name="l00263"></a>00263         %set references to state objects
<a name="l00264"></a>00264         function storeObjectReferences(this, simulation)
<a name="l00265"></a>00265             this.storeObjectReferences@edu.stanford.covert.cell.sim.<a class="code" href="class_process.html">Process</a>(simulation);
<a name="l00266"></a>00266             this.storeObjectReferences@edu.stanford.covert.cell.sim.<a class="code" href="class_chromosome_process_aspect.html">ChromosomeProcessAspect</a>(simulation);
<a name="l00267"></a>00267         end
<a name="l00268"></a>00268         
<a name="l00269"></a>00269         %initialize constants
<a name="l00270"></a>00270         function initializeConstants(this, knowledgeBase, simulation, varargin)
<a name="l00271"></a>00271             %% call super class method to get reactions, substrates, and enzymes
<a name="l00272"></a>00272             this.initializeConstants@edu.stanford.covert.cell.sim.<a class="code" href="class_reaction_process.html">ReactionProcess</a>(knowledgeBase, simulation, varargin{:});
<a name="l00273"></a>00273             this.initializeConstants@edu.stanford.covert.cell.sim.ChromosomeProcessAspect(knowledgeBase, simulation, varargin{:});
<a name="l00274"></a>00274             
<a name="l00275"></a>00275             %% <span class="keyword">remove</span> reactions with zero rate
<a name="l00276"></a>00276             this.initializeConstants_overrideReactions(simulation, this.reactionWholeCellModelIDs(this.reactionBounds(:, 2) ~= 0));
<a name="l00277"></a>00277             
<a name="l00278"></a>00278             %% additional substrates
<a name="l00279"></a>00279             substrateWholeCellModelIDs  = unique([
<a name="l00280"></a>00280                 this.substrateWholeCellModelIDs;
<a name="l00281"></a>00281                 <span class="stringliteral">&#39;DA&#39;</span>; <span class="stringliteral">&#39;DC&#39;</span>; <span class="stringliteral">&#39;DG&#39;</span>; <span class="stringliteral">&#39;DT&#39;</span>]);
<a name="l00282"></a>00282             this.initializeConstants_overrideSubstrates(substrateWholeCellModelIDs);
<a name="l00283"></a>00283             
<a name="l00284"></a>00284             %% substrate indices
<a name="l00285"></a>00285             this.substrateIndexs_radiation             = this.substrateIndexs({<span class="stringliteral">&#39;UVB_radiation&#39;</span>;<span class="stringliteral">&#39;gamma_radiation&#39;</span>});
<a name="l00286"></a>00286             this.substrateIndexs_hydrogen              = this.substrateIndexs({<span class="charliteral">&#39;H&#39;</span>});
<a name="l00287"></a>00287             this.substrateIndexs_hydroxylRadical       = this.substrateIndexs({<span class="stringliteral">&#39;hydroxyl_radical&#39;</span>});
<a name="l00288"></a>00288             this.substrateIndexs_water                 = this.substrateIndexs({<span class="stringliteral">&#39;H2O&#39;</span>});
<a name="l00289"></a>00289             this.substrateIndexs_DR5P                  = this.substrateIndexs({<span class="stringliteral">&#39;DR5P&#39;</span>});
<a name="l00290"></a>00290             this.substrateIndexs_dRibose5P_dRibose5P   = this.substrateIndexs({<span class="stringliteral">&#39;dRibose5P_dRibose5P&#39;</span>});
<a name="l00291"></a>00291             this.substrateIndexs_unmodifiedNucleobases = this.substrateIndexs({<span class="stringliteral">&#39;AD&#39;</span>;<span class="stringliteral">&#39;CSN&#39;</span>;<span class="stringliteral">&#39;GN&#39;</span>;<span class="stringliteral">&#39;THY&#39;</span>});
<a name="l00292"></a>00292             this.substrateIndexs_unmodifiedNucleosides = this.substrateIndexs({<span class="stringliteral">&#39;DA&#39;</span>;<span class="stringliteral">&#39;DC&#39;</span>;<span class="stringliteral">&#39;DG&#39;</span>;<span class="stringliteral">&#39;DT&#39;</span>});
<a name="l00293"></a>00293             this.substrateIndexs_unmodifiedDNMPs       = this.substrateIndexs({<span class="stringliteral">&#39;DAMP&#39;</span>;<span class="stringliteral">&#39;DCMP&#39;</span>;<span class="stringliteral">&#39;DGMP&#39;</span>;<span class="stringliteral">&#39;DTMP&#39;</span>});
<a name="l00294"></a>00294             this.substrateIndexs_modifiedNucleobases   = this.substrateMetaboliteLocalIndexs(...
<a name="l00295"></a>00295                 strcmp({knowledgeBase.metabolites(this.substrateMetaboliteGlobalIndexs).category},<span class="stringliteral">&#39;modified nucleobase&#39;</span>) | ...
<a name="l00296"></a>00296                 strcmp({knowledgeBase.metabolites(this.substrateMetaboliteGlobalIndexs).wholeCellModelID},<span class="stringliteral">&#39;URA&#39;</span>));
<a name="l00297"></a>00297             this.substrateIndexs_modifiedNucleosides   = this.substrateMetaboliteLocalIndexs(...
<a name="l00298"></a>00298                 strcmp({knowledgeBase.metabolites(this.substrateMetaboliteGlobalIndexs).category},<span class="stringliteral">&#39;modified deoxyribonucleoside&#39;</span>) | ...
<a name="l00299"></a>00299                 strcmp({knowledgeBase.metabolites(this.substrateMetaboliteGlobalIndexs).wholeCellModelID},<span class="stringliteral">&#39;URI&#39;</span>));
<a name="l00300"></a>00300             this.substrateIndexs_modifiedDNMPs         = this.substrateMetaboliteLocalIndexs(...
<a name="l00301"></a>00301                 strcmp({knowledgeBase.metabolites(this.substrateMetaboliteGlobalIndexs).category},<span class="stringliteral">&#39;modified deoxyribonucleotide monophosphate&#39;</span>) | ...
<a name="l00302"></a>00302                 strcmp({knowledgeBase.metabolites(this.substrateMetaboliteGlobalIndexs).wholeCellModelID},<span class="stringliteral">&#39;DUMP&#39;</span>));
<a name="l00303"></a>00303             
<a name="l00304"></a>00304             %% reaction indices
<a name="l00305"></a>00305             this.reactionIndexs_spontaneousBaseLoss        = find(strcmp(this.reactionTypes,<span class="stringliteral">&#39;spontaneous base loss&#39;</span>));
<a name="l00306"></a>00306             this.reactionIndexs_spontaneousBaseDeamination = find(strcmp(this.reactionTypes,<span class="stringliteral">&#39;spontaneous base deamination&#39;</span>));
<a name="l00307"></a>00307             this.reactionIndexs_gammaRayBaseOxidation      = find(strcmp(this.reactionTypes,<span class="stringliteral">&#39;radiation (gamma-ray) induced base oxidation&#39;</span>));
<a name="l00308"></a>00308             this.reactionIndexs_uvbDimerization            = find(strcmp(this.reactionTypes,<span class="stringliteral">&#39;UV-B photodimerization&#39;</span>));
<a name="l00309"></a>00309             this.reactionIndexs_strandBreak                = find(strcmp(this.reactionTypes,<span class="stringliteral">&#39;strand break&#39;</span>));
<a name="l00310"></a>00310             this.reactionIndexs_baseReduction              = find(strcmp(this.reactionTypes,<span class="stringliteral">&#39;base reduction&#39;</span>));
<a name="l00311"></a>00311             this.reactionIndexs_baseAmination              = find(strcmp(this.reactionTypes,<span class="stringliteral">&#39;base amination&#39;</span>));
<a name="l00312"></a>00312             this.reactionIndexs_baseGlucosylTransfer       = find(strcmp(this.reactionTypes,<span class="stringliteral">&#39;base glucosyl transfer&#39;</span>));
<a name="l00313"></a>00313             this.reactionIndexs_baseAlkylation             = find(strcmp(this.reactionTypes,<span class="stringliteral">&#39;base alkylation&#39;</span>));
<a name="l00314"></a>00314             this.reactionIndexs_baseEthylation             = find(strcmp(this.reactionTypes,<span class="stringliteral">&#39;base ethylation&#39;</span>));
<a name="l00315"></a>00315             this.reactionIndexs_baseMethylation            = find(strcmp(this.reactionTypes,<span class="stringliteral">&#39;base methylation&#39;</span>));
<a name="l00316"></a>00316             this.reactionIndexs_photooxidation             = find(strcmp(this.reactionTypes,<span class="stringliteral">&#39;photooxidation&#39;</span>));
<a name="l00317"></a>00317             
<a name="l00318"></a>00318             %% reaction types
<a name="l00319"></a>00319             this.reactionDamageTypes = cell(size(this.reactionWholeCellModelIDs));
<a name="l00320"></a>00320             this.reactionDamageTypes(this.reactionIndexs_spontaneousBaseLoss)        = {<span class="stringliteral">&#39;abasicSites&#39;</span>};
<a name="l00321"></a>00321             this.reactionDamageTypes(this.reactionIndexs_spontaneousBaseDeamination) = {<span class="stringliteral">&#39;damagedBases&#39;</span>};
<a name="l00322"></a>00322             this.reactionDamageTypes(this.reactionIndexs_gammaRayBaseOxidation)      = {<span class="stringliteral">&#39;damagedBases&#39;</span>};
<a name="l00323"></a>00323             this.reactionDamageTypes(this.reactionIndexs_uvbDimerization)            = {<span class="stringliteral">&#39;intrastrandCrossLinks&#39;</span>};
<a name="l00324"></a>00324             this.reactionDamageTypes(this.reactionIndexs_strandBreak)                = {<span class="stringliteral">&#39;strandBreaks&#39;</span>};
<a name="l00325"></a>00325             this.reactionDamageTypes(this.reactionIndexs_baseReduction)              = {<span class="stringliteral">&#39;damagedBases&#39;</span>};
<a name="l00326"></a>00326             this.reactionDamageTypes(this.reactionIndexs_baseAmination)              = {<span class="stringliteral">&#39;damagedBases&#39;</span>};
<a name="l00327"></a>00327             this.reactionDamageTypes(this.reactionIndexs_baseGlucosylTransfer)       = {<span class="stringliteral">&#39;damagedBases&#39;</span>};
<a name="l00328"></a>00328             this.reactionDamageTypes(this.reactionIndexs_baseAlkylation)             = {<span class="stringliteral">&#39;damagedBases&#39;</span>};
<a name="l00329"></a>00329             this.reactionDamageTypes(this.reactionIndexs_baseEthylation)             = {<span class="stringliteral">&#39;damagedBases&#39;</span>};
<a name="l00330"></a>00330             this.reactionDamageTypes(this.reactionIndexs_baseMethylation)            = {<span class="stringliteral">&#39;damagedBases&#39;</span>};
<a name="l00331"></a>00331             this.reactionDamageTypes(this.reactionIndexs_photooxidation)             = {<span class="stringliteral">&#39;damagedBases&#39;</span>};
<a name="l00332"></a>00332             
<a name="l00333"></a>00333             %% <a class="code" href="class_reaction.html">Reaction</a> stoichiometry matrices
<a name="l00334"></a>00334             %Factor reactionStoichiometryMatrix <span class="keyword">property</span> into 3 parts
<a name="l00335"></a>00335             %- reactionSmallMoleculeStoichiometryMatrix: the stochiometry of
<a name="l00336"></a>00336             %  free metabolites in damage reactions. Since we don&#39;t have an
<a name="l00337"></a>00337             %  explicit model of the production and amount of hydroxyl
<a name="l00338"></a>00338             %  radicals, and the rate of hydroxyl radical production from
<a name="l00339"></a>00339             %  water using radiation is implicit in the rate of the damage
<a name="l00340"></a>00340             %  reactions that use hydroxyl radicals, we replace hydroxyl
<a name="l00341"></a>00341             %  radical with water and hydrogen in reaction stoichiometries.
<a name="l00342"></a>00342             %- reactionDNAStoichiometryMatrix: the stoichiometry of DNA bases
<a name="l00343"></a>00343             %  in damage reactions
<a name="l00344"></a>00344             %- reactionRadiationStoichiometryMatrix: the stoichiometry of
<a name="l00345"></a>00345             %  radiation stimuli in damage reactions
<a name="l00346"></a>00346             numReactions = length(this.reactionWholeCellModelIDs);
<a name="l00347"></a>00347             
<a name="l00348"></a>00348             %indices of substrates in reactions that should be considered part
<a name="l00349"></a>00349             %of DNA, rather than free metabolites
<a name="l00350"></a>00350             substrateIndexs_DNA = [
<a name="l00351"></a>00351                 this.substrateIndexs_unmodifiedNucleobases;
<a name="l00352"></a>00352                 this.substrateIndexs_unmodifiedNucleosides;
<a name="l00353"></a>00353                 this.substrateIndexs_unmodifiedDNMPs;
<a name="l00354"></a>00354                 this.substrateIndexs_modifiedNucleobases;
<a name="l00355"></a>00355                 this.substrateIndexs_modifiedNucleosides;
<a name="l00356"></a>00356                 this.substrateIndexs_modifiedDNMPs;
<a name="l00357"></a>00357                 this.substrateIndexs_DR5P;
<a name="l00358"></a>00358                 this.substrateIndexs_dRibose5P_dRibose5P];
<a name="l00359"></a>00359             
<a name="l00360"></a>00360             %allocate memory
<a name="l00361"></a>00361             this.reactionSmallMoleculeStoichiometryMatrix = zeros(size(this.reactionStoichiometryMatrix));
<a name="l00362"></a>00362             this.reactionDNAStoichiometryMatrix           = zeros(size(this.reactionStoichiometryMatrix));
<a name="l00363"></a>00363             this.reactionRadiationStoichiometryMatrix     = zeros(size(this.reactionStoichiometryMatrix));
<a name="l00364"></a>00364             
<a name="l00365"></a>00365             %DNA
<a name="l00366"></a>00366             this.reactionDNAStoichiometryMatrix(substrateIndexs_DNA, :) = ...
<a name="l00367"></a>00367                 this.reactionStoichiometryMatrix(substrateIndexs_DNA, :);
<a name="l00368"></a>00368             
<a name="l00369"></a>00369             [i, j] = find(this.reactionDNAStoichiometryMatrix &gt; 0);
<a name="l00370"></a>00370             this.reactionDNAProduct = zeros(numel(this.reactionWholeCellModelIDs), 1);
<a name="l00371"></a>00371             this.reactionDNAProduct(j) = this.substrateGlobalIndexs(i);
<a name="l00372"></a>00372             
<a name="l00373"></a>00373             %radiation
<a name="l00374"></a>00374             this.reactionRadiationStoichiometryMatrix(this.substrateIndexs_radiation,:) = ...
<a name="l00375"></a>00375                 this.reactionStoichiometryMatrix(this.substrateIndexs_radiation, :);
<a name="l00376"></a>00376             
<a name="l00377"></a>00377             [i, j] = find(this.reactionRadiationStoichiometryMatrix);
<a name="l00378"></a>00378             this.reactionRadiation = zeros(numel(this.reactionWholeCellModelIDs), 1);
<a name="l00379"></a>00379             this.reactionRadiation(j) = i;
<a name="l00380"></a>00380             
<a name="l00381"></a>00381             %free metabolites
<a name="l00382"></a>00382             this.reactionSmallMoleculeStoichiometryMatrix = ...
<a name="l00383"></a>00383                 this.reactionStoichiometryMatrix - ...
<a name="l00384"></a>00384                 this.reactionDNAStoichiometryMatrix - ...
<a name="l00385"></a>00385                 this.reactionRadiationStoichiometryMatrix;
<a name="l00386"></a>00386             
<a name="l00387"></a>00387             %radicals
<a name="l00388"></a>00388             this.reactionSmallMoleculeStoichiometryMatrix([this.substrateIndexs_water;this.substrateIndexs_hydrogen],:) = ...
<a name="l00389"></a>00389                 this.reactionSmallMoleculeStoichiometryMatrix([this.substrateIndexs_water;this.substrateIndexs_hydrogen],:) + ...
<a name="l00390"></a>00390                 repmat([1;-1], 1, numReactions).*repmat(this.reactionSmallMoleculeStoichiometryMatrix(this.substrateIndexs_hydroxylRadical,:),2,1);
<a name="l00391"></a>00391             this.reactionSmallMoleculeStoichiometryMatrix(this.substrateIndexs_hydroxylRadical,:) = 0;
<a name="l00392"></a>00392             
<a name="l00393"></a>00393             %reaction-specific corrections
<a name="l00394"></a>00394             this.reactionSmallMoleculeStoichiometryMatrix(this.substrateIndexs_unmodifiedNucleobases,this.reactionIndexs_spontaneousBaseLoss) = ...
<a name="l00395"></a>00395                 this.reactionStoichiometryMatrix(this.substrateIndexs_unmodifiedNucleobases,this.reactionIndexs_spontaneousBaseLoss);
<a name="l00396"></a>00396             this.reactionDNAStoichiometryMatrix(this.substrateIndexs_unmodifiedNucleobases,this.reactionIndexs_spontaneousBaseLoss) = 0;
<a name="l00397"></a>00397             
<a name="l00398"></a>00398             %% reaction sequences
<a name="l00399"></a>00399             % represents the DNA sequence(s) vulnerable to each damage reaction
<a name="l00400"></a>00400             this.reactionVulnerableMotifs = repmat({<span class="stringliteral">&#39;&#39;</span>}, size(this.reactionStoichiometryMatrix,2), 1);
<a name="l00401"></a>00401             this.reactionVulnerableMotifTypes = repmat({<span class="stringliteral">&#39;&#39;</span>}, size(this.reactionStoichiometryMatrix,2), 1);
<a name="l00402"></a>00402             
<a name="l00403"></a>00403             <span class="keywordflow">for</span> i = 1:length(this.reactionWholeCellModelIDs)
<a name="l00404"></a>00404                 dnaSubstrateLocalIdx = find(this.reactionDNAStoichiometryMatrix(:, i) &lt; 0);
<a name="l00405"></a>00405                 if this.reactionDNAStoichiometryMatrix(dnaSubstrateLocalIdx, i) ~= -1; continue; end;
<a name="l00406"></a>00406                 
<a name="l00407"></a>00407                 [~, baseIdx] = ismember(dnaSubstrateLocalIdx, [this.substrateIndexs_unmodifiedNucleobases; this.substrateIndexs_unmodifiedNucleosides; this.substrateIndexs_unmodifiedDNMPs]);
<a name="l00408"></a>00408                 if baseIdx
<a name="l00409"></a>00409                     this.reactionVulnerableMotifs{i} = this.unmodifiedNucleobases(mod(baseIdx - 1,4) + 1);
<a name="l00410"></a>00410                     this.reactionVulnerableMotifTypes{i} = <span class="stringliteral">&#39;sequence&#39;</span>;
<a name="l00411"></a>00411                 <span class="keywordflow">else</span>
<a name="l00412"></a>00412                     this.reactionVulnerableMotifs{i} = this.substrateGlobalIndexs(dnaSubstrateLocalIdx);
<a name="l00413"></a>00413                     this.reactionVulnerableMotifTypes{i} = this.reactionDamageTypes{i};
<a name="l00414"></a>00414                 end
<a name="l00415"></a>00415             end
<a name="l00416"></a>00416             
<a name="l00417"></a>00417             this.reactionVulnerableMotifs(this.reactionIndexs_strandBreak) = {<span class="keyword">true</span>};
<a name="l00418"></a>00418             this.reactionVulnerableMotifTypes(this.reactionIndexs_strandBreak) = {<span class="stringliteral">&#39;abasicSites&#39;</span>};
<a name="l00419"></a>00419             
<a name="l00420"></a>00420             this.reactionVulnerableMotifTypes{this.reactionIndexs({<span class="stringliteral">&#39;DNADamage_CSNCSN_cyclobutane_CSNCSN_UVB_radiation&#39;</span>})}  = <span class="stringliteral">&#39;sequence&#39;</span>;
<a name="l00421"></a>00421             this.reactionVulnerableMotifTypes{this.reactionIndexs({<span class="stringliteral">&#39;DNADamage_THYTHY_cyclobutane_THYTHY_UVB_radiation&#39;</span>})}  = <span class="stringliteral">&#39;sequence&#39;</span>;
<a name="l00422"></a>00422             this.reactionVulnerableMotifTypes{this.reactionIndexs({<span class="stringliteral">&#39;DNADamage_CSNTHY_cyclobutane_CSNTHY_UVB_radiation&#39;</span>})}  = <span class="stringliteral">&#39;sequence&#39;</span>;
<a name="l00423"></a>00423             this.reactionVulnerableMotifTypes{this.reactionIndexs({<span class="stringliteral">&#39;DNADamage_THYCSN_cyclobutane_THYCSN_UVB_radiation&#39;</span>})}  = <span class="stringliteral">&#39;sequence&#39;</span>;
<a name="l00424"></a>00424             this.reactionVulnerableMotifTypes{this.reactionIndexs({<span class="stringliteral">&#39;DNADamage_CSNTHY_CSN64THY_UVB_radiation&#39;</span>})}            = <span class="stringliteral">&#39;sequence&#39;</span>;
<a name="l00425"></a>00425             %this.reactionVulnerableMotifTypes{this.reactionIndexs({<span class="stringliteral">&#39;DNADamage_CSNCSN_CSN64CSN_UVB_radiation&#39;</span>})}            = <span class="stringliteral">&#39;sequence&#39;</span>;
<a name="l00426"></a>00426             this.reactionVulnerableMotifTypes{this.reactionIndexs({<span class="stringliteral">&#39;DNADamage_THYCSN_THY64CSN_UVB_radiation&#39;</span>})}            = <span class="stringliteral">&#39;sequence&#39;</span>;
<a name="l00427"></a>00427             this.reactionVulnerableMotifTypes{this.reactionIndexs({<span class="stringliteral">&#39;DNADamage_THYTHY_THY64THY_UVB_radiation&#39;</span>})}            = <span class="stringliteral">&#39;sequence&#39;</span>;
<a name="l00428"></a>00428             
<a name="l00429"></a>00429             this.reactionVulnerableMotifs{this.reactionIndexs({<span class="stringliteral">&#39;DNADamage_CSNCSN_cyclobutane_CSNCSN_UVB_radiation&#39;</span>})}  = <span class="stringliteral">&#39;CC&#39;</span>;
<a name="l00430"></a>00430             this.reactionVulnerableMotifs{this.reactionIndexs({<span class="stringliteral">&#39;DNADamage_THYTHY_cyclobutane_THYTHY_UVB_radiation&#39;</span>})}  = <span class="stringliteral">&#39;TT&#39;</span>;
<a name="l00431"></a>00431             this.reactionVulnerableMotifs{this.reactionIndexs({<span class="stringliteral">&#39;DNADamage_CSNTHY_cyclobutane_CSNTHY_UVB_radiation&#39;</span>})}  = <span class="stringliteral">&#39;CT&#39;</span>;
<a name="l00432"></a>00432             this.reactionVulnerableMotifs{this.reactionIndexs({<span class="stringliteral">&#39;DNADamage_THYCSN_cyclobutane_THYCSN_UVB_radiation&#39;</span>})}  = <span class="stringliteral">&#39;TC&#39;</span>;
<a name="l00433"></a>00433             this.reactionVulnerableMotifs{this.reactionIndexs({<span class="stringliteral">&#39;DNADamage_CSNTHY_CSN64THY_UVB_radiation&#39;</span>})}            = <span class="stringliteral">&#39;CT&#39;</span>;
<a name="l00434"></a>00434             %this.reactionVulnerableMotifs{this.reactionIndexs({<span class="stringliteral">&#39;DNADamage_CSNCSN_CSN64CSN_UVB_radiation&#39;</span>})}            = <span class="stringliteral">&#39;CC&#39;</span>;
<a name="l00435"></a>00435             this.reactionVulnerableMotifs{this.reactionIndexs({<span class="stringliteral">&#39;DNADamage_THYCSN_THY64CSN_UVB_radiation&#39;</span>})}            = <span class="stringliteral">&#39;TC&#39;</span>;
<a name="l00436"></a>00436             this.reactionVulnerableMotifs{this.reactionIndexs({<span class="stringliteral">&#39;DNADamage_THYTHY_THY64THY_UVB_radiation&#39;</span>})}            = <span class="stringliteral">&#39;TT&#39;</span>;
<a name="l00437"></a>00437             
<a name="l00438"></a>00438             %% reaction rates
<a name="l00439"></a>00439             this.reactionBounds(:, 1) = 0;
<a name="l00440"></a>00440         end
<a name="l00441"></a>00441     end
<a name="l00442"></a>00442     
<a name="l00443"></a>00443     %model
<a name="l00444"></a>00444     methods
<a name="l00445"></a>00445         %Calculate
<a name="l00446"></a>00446         %- contribution to <a class="code" href="class_f_b_a.html">FBA</a> objective
<a name="l00447"></a>00447         %- minimum expression consistent with cell cycle length
<a name="l00448"></a>00448         function [bmProd, byProd, minEnzExp, maxEnzExp] = calcResourceRequirements_LifeCycle(this, constants, ~)
<a name="l00449"></a>00449             %substrate and byproducts: metabolic effects of damage reactions
<a name="l00450"></a>00450             rates = this.calcExpectedReactionRates() * (...
<a name="l00451"></a>00451                 + constants.states.Time.cellCycleLength + ...
<a name="l00452"></a>00452                 + 0.5 * constants.states.Time.replicationDuration ...
<a name="l00453"></a>00453                 + constants.states.Time.cytokinesisDuration);
<a name="l00454"></a>00454             bmProd = max(0, -this.reactionSmallMoleculeStoichiometryMatrix) * rates;
<a name="l00455"></a>00455             byProd = max(0,  this.reactionSmallMoleculeStoichiometryMatrix) * rates;
<a name="l00456"></a>00456             
<a name="l00457"></a>00457             %no enzymes involved in DNA damage
<a name="l00458"></a>00458             minEnzExp = zeros(size(this.enzymeWholeCellModelIDs));
<a name="l00459"></a>00459             maxEnzExp = Inf(size(this.enzymeWholeCellModelIDs));
<a name="l00460"></a>00460         end
<a name="l00461"></a>00461         
<a name="l00462"></a>00462         %initialization -- simulation initializeState initializes 1 undamaged
<a name="l00463"></a>00463         %chromosome (except methylation at restriction/modification sites which
<a name="l00464"></a>00464         %is initialized by DNA repair)
<a name="l00465"></a>00465         function initializeState(~)
<a name="l00466"></a>00466         end
<a name="l00467"></a>00467         
<a name="l00468"></a>00468         %resource requirements
<a name="l00469"></a>00469         function result = calcResourceRequirements_Current(this)
<a name="l00470"></a>00470             %metabolic effects of reactions
<a name="l00471"></a>00471             result = max(0, -this.reactionSmallMoleculeStoichiometryMatrix * this.calcExpectedReactionRates());
<a name="l00472"></a>00472         end
<a name="l00473"></a>00473         
<a name="l00474"></a>00474         <span class="keyword">function</span> rates = calcExpectedReactionRates(<span class="keyword">this</span>)
<a name="l00475"></a>00475             %number of sites vulnerable to reaction
<a name="l00476"></a>00476             nVulnerableSites = this.calcNumberVulnerableSites();
<a name="l00477"></a>00477             
<a name="l00478"></a>00478             %calculate rate, product of
<a name="l00479"></a>00479             %- number of vulnerable site / chromosome
<a name="l00480"></a>00480             %- integrated number of chromosomes over cell cycle
<a name="l00481"></a>00481             %- reaction rate per vulnerable site (and per <a class="code" href="unit_8m.html#a94c091bed34da4edc337e42e23029000">unit</a> radiation
<a name="l00482"></a>00482             %  for applicable reactions)
<a name="l00483"></a>00483             %- amount of radition, for applicable reactions
<a name="l00484"></a>00484             rates = nVulnerableSites .* this.reactionBounds(:, 2);
<a name="l00485"></a>00485             rates(this.reactionRadiation ~= 0) = ...
<a name="l00486"></a>00486                 rates(this.reactionRadiation ~= 0) .* ...
<a name="l00487"></a>00487                 this.substrates(this.reactionRadiation(this.reactionRadiation ~= 0));
<a name="l00488"></a>00488         end
<a name="l00489"></a>00489         
<a name="l00490"></a>00490         <span class="keyword">function</span> nVulnerableSites = calcNumberVulnerableSites(<span class="keyword">this</span>)
<a name="l00491"></a>00491             c = this.chromosome;
<a name="l00492"></a>00492             polDNA = collapse(c.polymerizedRegions);
<a name="l00493"></a>00493             bases = &#39;ACGT&#39;;
<a name="l00494"></a>00494             dntpComposition = [
<a name="l00495"></a>00495                 1 - c.sequenceGCContent;
<a name="l00496"></a>00496                 c.sequenceGCContent;
<a name="l00497"></a>00497                 c.sequenceGCContent;
<a name="l00498"></a>00498                 1 - c.sequenceGCContent] / 2;
<a name="l00499"></a>00499             [posStrnds, dmg] = find(c.damagedSites); %safely ignores m6AD because this process doesn&#39;t produce or modify m6AD; this is asserted in testConstants in <a class="code" href="unit_8m.html#a94c091bed34da4edc337e42e23029000">unit</a> test
<a name="l00500"></a>00500             
<a name="l00501"></a>00501             nVulnerableSites = zeros(size(this.reactionStoichiometryMatrix, 2), 1);
<a name="l00502"></a>00502             for i = 1:size(this.reactionStoichiometryMatrix, 2)
<a name="l00503"></a>00503                 %% damage would be to undamaged sequence
<a name="l00504"></a>00504                 if ischar(this.reactionVulnerableMotifs{i})
<a name="l00505"></a>00505                     nVulnerableSites(i) = polDNA * prod(dntpComposition(ismembc2(this.reactionVulnerableMotifs{i}, bases)));
<a name="l00506"></a>00506                     <span class="keywordflow">continue</span>;
<a name="l00507"></a>00507                 end
<a name="l00508"></a>00508                 
<a name="l00509"></a>00509                 %% damage would be to already damaged sequence
<a name="l00510"></a>00510                 <span class="keywordflow">if</span> isempty(posStrnds)
<a name="l00511"></a>00511                     continue;
<a name="l00512"></a>00512                 end
<a name="l00513"></a>00513                 
<a name="l00514"></a>00514                 [spcPosStrmds, spcDmg] = find(c.(this.reactionVulnerableMotifTypes{i}));
<a name="l00515"></a>00515                 <span class="keywordflow">if</span> isempty(spcPosStrmds)
<a name="l00516"></a>00516                     continue;
<a name="l00517"></a>00517                 end
<a name="l00518"></a>00518                 
<a name="l00519"></a>00519                 spcIdxs = find(spcDmg == this.reactionVulnerableMotifs{i});
<a name="l00520"></a>00520                 <span class="keywordflow">if</span> isempty(spcIdxs)
<a name="l00521"></a>00521                     continue;
<a name="l00522"></a>00522                 end
<a name="l00523"></a>00523                 
<a name="l00524"></a>00524                 idxs = find(dmg == this.reactionVulnerableMotifs{i});
<a name="l00525"></a>00525                 <span class="keywordflow">if</span> isempty(idxs)
<a name="l00526"></a>00526                     continue;
<a name="l00527"></a>00527                 end
<a name="l00528"></a>00528                 
<a name="l00529"></a>00529                 nVulnerableSites(i) = sum(edu.stanford.covert.util.<a class="code" href="class_sparse_mat.html">SparseMat</a>.ismember_subs(...
<a name="l00530"></a>00530                     spcPosStrmds(spcIdxs, :), posStrnds(idxs, :), [c.sequenceLen c.nCompartments]));
<a name="l00531"></a>00531             end
<a name="l00532"></a>00532         end
<a name="l00533"></a>00533         
<a name="l00534"></a>00534         %simulation
<a name="l00535"></a>00535         function evolveState(this)
<a name="l00536"></a>00536             %loop over reactions in random order
<a name="l00537"></a>00537             randomOrder = this.randStream.randperm(numel(this.reactionWholeCellModelIDs));
<a name="l00538"></a>00538             for i = 1:length(this.reactionWholeCellModelIDs)
<a name="l00539"></a>00539                 j = randomOrder(i);
<a name="l00540"></a>00540                 
<a name="l00541"></a>00541                 %maximum number of reactions, based on substrate availability
<a name="l00542"></a>00542                 maxReactions = floor(min(this.substrates ./ max(0, -this.reactionSmallMoleculeStoichiometryMatrix(:, j))));
<a name="l00543"></a>00543                 if maxReactions &lt;= 0
<a name="l00544"></a>00544                     continue;
<a name="l00545"></a>00545                 end
<a name="l00546"></a>00546                 
<a name="l00547"></a>00547                 % probability that valid site undergoes reaction
<a name="l00548"></a>00548                 radiationLclIdx  = this.reactionRadiation(j);
<a name="l00549"></a>00549                 if radiationLclIdx ~= 0
<a name="l00550"></a>00550                     selectionProbability = this.stepSizeSec * this.reactionBounds(j, 2) * this.substrates(radiationLclIdx);
<a name="l00551"></a>00551                 else
<a name="l00552"></a>00552                     selectionProbability = this.stepSizeSec * this.reactionBounds(j, 2);
<a name="l00553"></a>00553                 end
<a name="l00554"></a>00554                 if selectionProbability == 0
<a name="l00555"></a>00555                     continue;
<a name="l00556"></a>00556                 end
<a name="l00557"></a>00557                 
<a name="l00558"></a>00558                 %randomly damage sites
<a name="l00559"></a>00559                 [positionsStrands, sideEffects] = this.chromosome.setSiteDamaged(...
<a name="l00560"></a>00560                     this.reactionDamageTypes{j}, this.reactionDNAProduct(j), selectionProbability, ...
<a name="l00561"></a>00561                     maxReactions, this.reactionVulnerableMotifs{j}, this.reactionVulnerableMotifTypes{j});
<a name="l00562"></a>00562                 <span class="keywordflow">if</span> isempty(positionsStrands)
<a name="l00563"></a>00563                     continue;
<a name="l00564"></a>00564                 end
<a name="l00565"></a>00565                 
<a name="l00566"></a>00566                 % use small molecule reactants, produce small molecule products
<a name="l00567"></a>00567                 this.substrates = this.substrates + ...
<a name="l00568"></a>00568                     size(positionsStrands, 1) * this.reactionSmallMoleculeStoichiometryMatrix(:,j);
<a name="l00569"></a>00569                 
<a name="l00570"></a>00570                 %side effects
<a name="l00571"></a>00571                 if ~isempty(sideEffects)
<a name="l00572"></a>00572                     this.simulationStateSideEffects = [this.simulationStateSideEffects; sideEffects];
<a name="l00573"></a>00573                 end
<a name="l00574"></a>00574             end
<a name="l00575"></a>00575         end
<a name="l00576"></a>00576     end
<a name="l00577"></a>00577 end
</pre></div></div><!-- contents -->
</div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
<a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(0)"><span class="SelectionMark">&#160;</span>All</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(1)"><span class="SelectionMark">&#160;</span>Classes</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(2)"><span class="SelectionMark">&#160;</span>Files</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(3)"><span class="SelectionMark">&#160;</span>Functions</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(4)"><span class="SelectionMark">&#160;</span>Variables</a></div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

  <div id="nav-path" class="navpath">
    <ul>
      <li class="navelem"><a class="el" href="_09process_2_d_n_a_damage_8m.html">DNADamage.m</a>      </li>

    <li class="footer">Generated on Wed Feb 8 2012 21:42:31 for Whole Cell Simulation by
    <a href="http://www.doxygen.org/index.html">
    <img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.7.6.1 </li>
   </ul>
 </div>


</body>
</html>
