<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<title>Whole Cell Simulation: src/+edu/+stanford/+covert/+cell/+sim/+process/Transcription.m Source File</title>

<link href="tabs.css" rel="stylesheet" type="text/css"/>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
<link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="navtree.js"></script>
<script type="text/javascript">
  $(document).ready(initResizable);
</script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
  $(document).ready(function() { searchBox.OnSelectItem(0); });
</script>

</head>
<body>
<div id="top"><!-- do not remove this div! -->


<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  
  
  <td style="padding-left: 0.5em;">
   <div id="projectname">Whole Cell Simulation
   
   </div>
   
  </td>
  
  
  
 </tr>
 </tbody>
</table>
</div>

<!-- Generated by Doxygen 1.7.6.1 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "search",false,'Search');
</script>
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&#160;Page</span></a></li>
      <li><a href="annotated.html"><span>Classes</span></a></li>
      <li class="current"><a href="files.html"><span>Files</span></a></li>
      <li>
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
  <div id="navrow2" class="tabs2">
    <ul class="tablist">
      <li><a href="files.html"><span>File&#160;List</span></a></li>
      <li><a href="globals.html"><span>File&#160;Members</span></a></li>
    </ul>
  </div>
</div>
<div id="side-nav" class="ui-resizable side-nav-resizable">
  <div id="nav-tree">
    <div id="nav-tree-contents">
    </div>
  </div>
  <div id="splitbar" style="-moz-user-select:none;" 
       class="ui-resizable-handle">
  </div>
</div>
<script type="text/javascript">
  initNavTree('_09process_2_transcription_8m.html','');
</script>
<div id="doc-content">
<div class="header">
  <div class="headertitle">
<div class="title">src/+edu/+stanford/+covert/+cell/+sim/+process/Transcription.m</div>  </div>
</div><!--header-->
<div class="contents">
<a href="_09process_2_transcription_8m.html">Go to the documentation of this file.</a><div class="fragment"><pre class="fragment"><a name="l00001"></a>00001 %<a class="code" href="class_transcription.html">Transcription</a>
<a name="l00002"></a>00002 %
<a name="l00003"></a>00003 % @wholeCellModelID Process_Transcription
<a name="l00004"></a>00004 % @name             <a class="code" href="class_transcription.html">Transcription</a>
<a name="l00005"></a>00005 % @description
<a name="l00006"></a>00006 %   Biology
<a name="l00007"></a>00007 %   ===============
<a name="l00008"></a>00008 %   <a class="code" href="class_transcription.html">Transcription</a> is the first step in the synthesis of functional gene
<a name="l00009"></a>00009 %   products where <a class="code" href="class_r_n_a.html">RNA</a> polymerase and several accessory enzymes translate
<a name="l00010"></a>00010 %   transcription units, or regions of the DNA containing 1 or more genes, into
<a name="l00011"></a>00011 %   <a class="code" href="class_r_n_a.html">RNA</a> molecules. Following transcription the <a class="code" href="class_r_n_a.html">RNA</a> molecules follow one of
<a name="l00012"></a>00012 %   two pathways:
<a name="l00013"></a>00013 %   - mRNAs are used as templates for translation (see translation process)
<a name="l00014"></a>00014 %   - r/s/tRNAs which are transcribed as molecules containing multiple genes are
<a name="l00015"></a>00015 %     cleavaged into their individual genes (see <a class="code" href="class_r_n_a.html">RNA</a> processing process), are
<a name="l00016"></a>00016 %     modified at several bases (see <a class="code" href="class_r_n_a.html">RNA</a> modification process) to improve their
<a name="l00017"></a>00017 %     stability and enhance their catalytic activity, and finally act as
<a name="l00018"></a>00018 %     ribozymes (r/sRNAs) or as adaptors between mRNAs and the amino acids they
<a name="l00019"></a>00019 %     code for (tRNAs).
<a name="l00020"></a>00020 %
<a name="l00021"></a>00021 %   <a class="code" href="class_transcription.html">Transcription</a> begins with the recruitment of <a class="code" href="class_r_n_a.html">RNA</a> polymerase to a promoter
<a name="l00022"></a>00022 %   with the help of the sigma initiation factor and possiblity transcription
<a name="l00023"></a>00023 %   factors. Next elongation factors are recruited, <a class="code" href="class_r_n_a.html">RNA</a> begins to be
<a name="l00024"></a>00024 %   polymerized, and sigma factor is released. Finally the <a class="code" href="class_r_n_a.html">RNA</a> polymerase
<a name="l00025"></a>00025 %   reaches  a terminator at the the end of the transcription <a class="code" href="unit_8m.html#a94c091bed34da4edc337e42e23029000">unit</a>, and
<a name="l00026"></a>00026 %   with the help of termination factors releases the polymerized <a class="code" href="class_r_n_a.html">RNA</a> and
<a name="l00027"></a>00027 %   dissociates from the DNA. Termination in E. coli occurs via either a
<a name="l00028"></a>00028 %   Rho-dependent (50%) or Rho-independent mechanism (50%). Rho-dependent
<a name="l00029"></a>00029 %   termination is catalyzed by the hexameric ATP-dependent helicase Rho.
<a name="l00030"></a>00030 %   Rho is not essential in B. subtilis [PUB_0234]. Rho-independent
<a name="l00031"></a>00031 %   termination occur via the intrinsic <a class="code" href="json_parse_8m.html#a6e4256f69b751d3d7ec4ea52ff5a8dee">properties</a> <a class="code" href="class_r_n_a.html">RNA</a> which disrupt <a class="code" href="class_r_n_a.html">RNA</a>
<a name="l00032"></a>00032 %   polymerase-DNA binding. Terminator hairpins are not predicted in M.
<a name="l00033"></a>00033 %   genitalium. In E. coli termination is incompetition with
<a name="l00034"></a>00034 %   antitermination. However antitermination has not been reported in any
<a name="l00035"></a>00035 %   mcyoplasma [PUB_0182].
<a name="l00036"></a>00036 %
<a name="l00037"></a>00037 %   As soon as the <a class="code" href="class_r_n_a.html">RNA</a> begins to polymerized, even prior to termination, the
<a name="l00038"></a>00038 %   <a class="code" href="classm_r_n_a.html">mRNA</a> transcripts may be bound by ribosomes and polymerized. For simplicity,
<a name="l00039"></a>00039 %   our model doesn&#39;t represent this phenomenon.
<a name="l00040"></a>00040 %
<a name="l00041"></a>00041 %   Knowledge Base
<a name="l00042"></a>00042 %   ===============
<a name="l00043"></a>00043 %   The transcription <a class="code" href="unit_8m.html#a94c091bed34da4edc337e42e23029000">unit</a> structure was compiled from several sources:
<a name="l00044"></a>00044 %   - Primary reports of cotranscribed genes [PUB_0176, PUB_0182, PUB_0186,
<a name="l00045"></a>00045 %     PUB_0188, PUB_0188, PUB_0244, PUB_0247, PUB_0248, PUB_0249, PUB_0251]
<a name="l00046"></a>00046 %   - OperonDB database of cotranscribed genes [PUB_0250]
<a name="l00047"></a>00047 %   - Conservation of gene order across multiple species
<a name="l00048"></a>00048 %   - Related function of adjacent genes
<a name="l00049"></a>00049 %   - Expression levels as measured by microarrays of adjacent genes [PUB_0569]
<a name="l00050"></a>00050 %   - Strandedness of adjacent genes
<a name="l00051"></a>00051 %   - Weiner, Hermann, and Browning computational model of promoters and
<a name="l00052"></a>00052 %     transcription unit start sites [PUB_0411]
<a name="l00053"></a>00053 %
<a name="l00054"></a>00054 %   The transcription unit structure is organized in the knowledge base, and is
<a name="l00055"></a>00055 %   loaded into this process by the initializeConstants method.
<a name="l00056"></a>00056 %
<a name="l00057"></a>00057 %   The knowledge base also contains the measured expression and half-lives of
<a name="l00058"></a>00058 %   many transcripts. These values are loaded by initializeConstants and fit by
<a name="l00059"></a>00059 %   simulation.fitConstants to be consistent with other processes.
<a name="l00060"></a>00060 %
<a name="l00061"></a>00061 %   Representation
<a name="l00062"></a>00062 %   ===============
<a name="l00063"></a>00063 %   substrates, enzymes, boundEnzymes, and RNAs represent the counts of
<a name="l00064"></a>00064 %   metabolites, free transcription enzymes, transcription enzymes bound to RNAs
<a name="l00065"></a>00065 %   and RNA polymerases, and nascent RNAs.
<a name="l00066"></a>00066 %
<a name="l00067"></a>00067 %   rnaPolymerases.states represents the current state / pseudostate (actively
<a name="l00068"></a>00068 %   transcribing, specifically bound, non-specifically bound, free,
<a name="l00069"></a>00069 %   non-existent) of each RNA polymerase, where each state is indicated by the
<a name="l00070"></a>00070 %   enumeration:
<a name="l00071"></a>00071 %   - rnaPolymeraseActivelyTranscribingValue
<a name="l00072"></a>00072 %   - rnaPolymeraseSpecificallyBoundValue
<a name="l00073"></a>00073 %   - rnaPolymeraseNonSpecificallyBoundValue
<a name="l00074"></a>00074 %   - rnaPolymeraseFreeValue
<a name="l00075"></a>00075 %   - rnaPolymeraseNotExistValue (state exists as a way to account for memory
<a name="l00076"></a>00076 %     allocated for future <a class="code" href="class_r_n_a.html">RNA</a> polymerases)
<a name="l00077"></a>00077 %   For actively transcribing polymerases rnaPolymerases.states also represents
<a name="l00078"></a>00078 %   the position of the polymerase along the transcription unit.
<a name="l00079"></a>00079 %
<a name="l00080"></a>00080 %   That is entries of rnaPolymerases.states with the following values
<a name="l00081"></a>00081 %   corresponding to these states:
<a name="l00082"></a>00082 %     &gt;= RNAPolymerases.activelyTranscribingValue: RNA polymerases position on genome actively transcribing
<a name="l00083"></a>00083 %     == RNAPolymerases.specificallyBoundValue:    RNA polymerase specifically bound
<a name="l00084"></a>00084 %     == RNAPolymerases.nonSpecificallyBoundValue: RNA polymerase non-specifically bound
<a name="l00085"></a>00085 %     == RNAPolymerases.freeValue:                 RNA polymerase free
<a name="l00086"></a>00086 %     == RNAPolymerases.notExistValue:             RNA polymerase doesn&#39;t exist
<a name="l00087"></a>00087 %
<a name="l00088"></a>00088 %   transcripts.boundTranscriptionUnits represents the particular transcription
<a name="l00089"></a>00089 %   unit to which each actively transcribing and specifically bound polymerase
<a name="l00090"></a>00090 %   is bound.
<a name="l00091"></a>00091 %
<a name="l00092"></a>00092 %   rnaPolymeraseStateExpectations represents the expected occupancies of the
<a name="l00093"></a>00093 %   RNA polymerase states.
<a name="l00094"></a>00094 %
<a name="l00095"></a>00095 %   transcriptionUnitBindingProbabilities represents the relative affinity of
<a name="l00096"></a>00096 %   RNA polymerases for the promoters of each transcription unit.
<a name="l00097"></a>00097 %   transcriptionFactorBindingProbFoldChange represents the fold change affect of
<a name="l00098"></a>00098 %   transcription factors on the relative affinities of the RNA polymerse for
<a name="l00099"></a>00099 %   the promoters. RNA polymerases are assigned to
<a name="l00100"></a>00100 %   transcription units weighted by the product of
<a name="l00101"></a>00101 %   transcriptionUnitBindingProbabilities and
<a name="l00102"></a>00102 %   transcriptionFactorBindingProbFoldChange.
<a name="l00103"></a>00103 %
<a name="l00104"></a>00104 %   Initialization
<a name="l00105"></a>00105 %   ===============
<a name="l00106"></a>00106 %   All RNAs are initialized to the mature state. This is implemented by the
<a name="l00107"></a>00107 %   simulation class initializeState method.
<a name="l00108"></a>00108 %
<a name="l00109"></a>00109 %   RNA polymerases are initialized to their steady state:
<a name="l00110"></a>00110 %   - Each RNA polymerase is randomly assigned (with replacement) to one of the
<a name="l00111"></a>00111 %     actively transcribing, specifically bound, non-specifically bound, or free
<a name="l00112"></a>00112 %     states weighted by the expected occupancy of each state
<a name="l00113"></a>00113 %     (rnaPolymeraseStateExpectations)
<a name="l00114"></a>00114 %   - Actively transcribing and specifically bound polymerases randomly assigned
<a name="l00115"></a>00115 %     to transcription units weighted by their transcription rates
<a name="l00116"></a>00116 %     (transcriptionUnitBindingProbabilities)
<a name="l00117"></a>00117 %   - Each transcription unit to which an actively transcribing polymerase has
<a name="l00118"></a>00118 %     been assigned is divided into 1 segment for each polymerase
<a name="l00119"></a>00119 %   - Actively transcribing polymerases randomly assigned to positions within
<a name="l00120"></a>00120 %     the assigned segment of their assigned transcription unit (positions near
<a name="l00121"></a>00121 %     the segment border are not allowed to prevent polymerases from being too
<a name="l00122"></a>00122 %     close to each other) with uniform probably.
<a name="l00123"></a>00123 %
<a name="l00124"></a>00124 %   Simulation
<a name="l00125"></a>00125 %   ===============
<a name="l00126"></a>00126 %   Evolves the state of RNA polymerase using a markov chain model with four
<a name="l00127"></a>00127 %   states:
<a name="l00128"></a>00128 %   - actively translating
<a name="l00129"></a>00129 %   - specifically bound
<a name="l00130"></a>00130 %   - non-specifically bound
<a name="l00131"></a>00131 %   - free
<a name="l00132"></a>00132 %
<a name="l00133"></a>00133 %   Transition probabilities are designed to maintain the occupancy of each
<a name="l00134"></a>00134 %   state within a narrow window around their expected values. Transition
<a name="l00135"></a>00135 %   probability are determined by four logistic control functions. These can
<a name="l00136"></a>00136 %   be tuned with the constants
<a name="l00137"></a>00137 %   - rnaPolymeraseStateExpectations
<a name="l00138"></a>00138 %
<a name="l00139"></a>00139 %   RNA polymerase are created in the free state.
<a name="l00140"></a>00140 %
<a name="l00141"></a>00141 %   Actively transcribing state:
<a name="l00142"></a>00142 %   1. Release sigma factor if after first second of elongation
<a name="l00143"></a>00143 %   2. Elongate transcript according to nucleic acid limits (substrates)
<a name="l00144"></a>00144 %      if elongation factors are available
<a name="l00145"></a>00145 %   3. If transcription complete and termination factor available
<a name="l00146"></a>00146 %      - release transcript
<a name="l00147"></a>00147 %      - transition RNA polymerase to free state
<a name="l00148"></a>00148 %      - increment gene expression
<a name="l00149"></a>00149 %      Otherwise remain in active state
<a name="l00150"></a>00150 %
<a name="l00151"></a>00151 %   Specifically bound State:
<a name="l00152"></a>00152 %   - Can transition to active, specifically bound, non-specifically bound,
<a name="l00153"></a>00153 %     or free states
<a name="l00154"></a>00154 %   - Transition into state only if a free sigma factor is available
<a name="l00155"></a>00155 %     1. Decrement number of free sigma factors
<a name="l00156"></a>00156 %     2. Pick a transcription unit (tu) to bind to according to
<a name="l00157"></a>00157 %     Expression transcription unit i~prob(ribosome releases tu i|ribosome active)
<a name="l00158"></a>00158 %                      =prob(ribosome within <a class="code" href="class_r_n_a.html">RNA</a> polymerase elongation rate bases of length of tu i|ribosome active)
<a name="l00159"></a>00159 %                      =prob(ribosome within <a class="code" href="class_r_n_a.html">RNA</a> polymerase elongation rate bases of length of tu i|ribosome active, bound to tu i)*prob(ribosome bound to tu i|ribosome active)
<a name="l00160"></a>00160 %                      =[(<a class="code" href="class_r_n_a.html">RNA</a> polymerase transcription rate)/(length of tu i)] * [(length of tu i)*prob(binding tu i|binding)]
<a name="l00161"></a>00161 %     prob(binding tu i | binding)~expression tu i
<a name="l00162"></a>00162 %
<a name="l00163"></a>00163 %   Non-specifically bound state:
<a name="l00164"></a>00164 %   - Can transition to specifically bound, non-specifically bound, or free
<a name="l00165"></a>00165 %     states
<a name="l00166"></a>00166 %
<a name="l00167"></a>00167 %   Free state:
<a name="l00168"></a>00168 %   - Can transition to specifically bound, non-specifically bound, or free
<a name="l00169"></a>00169 %     states
<a name="l00170"></a>00170 %
<a name="l00171"></a>00171 %   Algorithm
<a name="l00172"></a>00172 %   +++++++++++++++
<a name="l00173"></a>00173 %   1. Randomly transition RNA polymerases among activlely transcribing,
<a name="l00174"></a>00174 %      specifically bound, non-specifically, bound, and free states weighted by
<a name="l00175"></a>00175 %      state transition probabilities. Update rnaPolymerases.states.
<a name="l00176"></a>00176 %   2. Randomly assign RNA polymerases entering the specifically bound to
<a name="l00177"></a>00177 %      specific transcription units weighted by the product of
<a name="l00178"></a>00178 %      transcriptionUnitBindingProbabilities and
<a name="l00179"></a>00179 %      transcriptionFactorBindingProbFoldChange. Update
<a name="l00180"></a>00180 %      transcripts.boundTranscriptionUnits.
<a name="l00181"></a>00181 %   3. Assign RNA polymerase entering the actively transcribing state sigma
<a name="l00182"></a>00182 %      factors. Update enzymes and boundEnzymes.
<a name="l00183"></a>00183 %   4. Simulate RNA polymerization by actively transcribing RNA polymerases with
<a name="l00184"></a>00184 %      the aid of elongation factors. Allocate available nucleic acids among the
<a name="l00185"></a>00185 %      actively transcribing RNA polymerases. Release sigma factors from RNA
<a name="l00186"></a>00186 %      polymerases that started at the beginning of the transcription unit and
<a name="l00187"></a>00187 %      progressed. Update rnaPolymerases.states. Update substrates. Update enzymes
<a name="l00188"></a>00188 %      and boundEnzymes.
<a name="l00189"></a>00189 %   5. If termination factors are available dissociate RNA polymerases which
<a name="l00190"></a>00190 %      have reached the terminus of the transcription they&#39;re bound to, and
<a name="l00191"></a>00191 %      release RNAs. Update rnaPolymerases.states and
<a name="l00192"></a>00192 %      transcripts.boundTranscriptionUnits. Increment RNAs.
<a name="l00193"></a>00193 %
<a name="l00194"></a>00194 %   References
<a name="l00195"></a>00195 %   ===========
<a name="l00196"></a>00196 %   1. McClure, W. R. 1985. Mechanism and control of transcription
<a name="l00197"></a>00197 %      initiation in prokaryotes. Annu. Rev. Biochem. 54:171-204.
<a name="l00198"></a>00198 %      [PUB_0775]
<a name="l00199"></a>00199 %   2. Ciampi MS (2006). Rho-dependent terminators and transcription
<a name="l00200"></a>00200 %      termination. Microbiology. 152(9):2515-28 [PUB_0233].
<a name="l00201"></a>00201 %   3. Nudler E, Gottesman ME (2002). Transcription termination and
<a name="l00202"></a>00202 %      anti-termination in E. coli. Genes Cells. 7(8):755-68. [PUB_0662]
<a name="l00203"></a>00203 %   4. Washio T, Sasayama J, Tomita M (1998). Analysis of complete genomes
<a name="l00204"></a>00204 %      suggests that many prokaryotes do not rely on hairpin formation in
<a name="l00205"></a>00205 %      transcription termination. Nucleic Acids Res. 26(23):5456-63
<a name="l00206"></a>00206 %      [PUB_0234]
<a name="l00207"></a>00207 %   5. Peterson JD, Umayam LA, Dickinson T, Hickey EK, White O (2001). The
<a name="l00208"></a>00208 %      Comprehensive Microbial Resource. Nucleic Acids Res. 29(1):123-5.
<a name="l00209"></a>00209 %      [PUB_0182]
<a name="l00210"></a>00210 %   6. Shepherd N, Dennis P, Bremer H (2001). Cytoplasmic RNA Polymerase in
<a name="l00211"></a>00211 %      Escherichia coli. J Bacteriol. 183(8): 2527-34. [PUB_0784]
<a name="l00212"></a>00212 %   7. Klumpp S, Hwa T (2008). Growth-rate-dependent partitioning of RNA
<a name="l00213"></a>00213 %      polymerases in bacteria. Proc Natl Acad Sci U S A. 105(21):
<a name="l00214"></a>00214 %      20245-50. [PUB_0785]
<a name="l00215"></a>00215 %   8. Grigorova IL, Phleger NJ, Mutalik VK, Gross CA (2006). Insights into
<a name="l00216"></a>00216 %      transcriptional regulation and sigma competition from an equilibrium
<a name="l00217"></a>00217 %      model of RNA polymerase binding to DNA. Proc Natl Acad Sci U S A.
<a name="l00218"></a>00218 %      103(14): 5332-7. [PUB_0786]
<a name="l00219"></a>00219 %
<a name="l00220"></a>00220 % Author: Markus Covert, mcovert@stanford.edu
<a name="l00221"></a>00221 % Author: Jayodita Sanghvi, jayodita@stanford.edu
<a name="l00222"></a>00222 % Author: Jonathan Karr, jkarr@stanford.edu
<a name="l00223"></a>00223 % Affiliation: Covert Lab, Department of Bioengineering, Stanford University
<a name="l00224"></a>00224 % Last updated: 1/4/2010
<a name="l00225"></a>00225 %
<a name="l00226"></a>00226 %TODO: require Mg2+ or Mn2+ as cofactor for transcription
<a name="l00227"></a>00227 classdef Transcription &lt; edu.stanford.covert.cell.sim.Process &amp; edu.stanford.covert.cell.sim.ChromosomeProcessAspect
<a name="l00228"></a>00228     %property annotations
<a name="l00229"></a>00229     properties (Constant)
<a name="l00230"></a>00230         optionNames__              = {}; %names of option <a class="code" href="json_parse_8m.html#a6e4256f69b751d3d7ec4ea52ff5a8dee">properties</a>
<a name="l00231"></a>00231         fixedConstantNames__       = {   %names of fixed constant <a class="code" href="json_parse_8m.html#a6e4256f69b751d3d7ec4ea52ff5a8dee">properties</a>
<a name="l00232"></a>00232             <span class="stringliteral">&#39;rnaPolymeraseElongationRate&#39;</span>;
<a name="l00233"></a>00233             <span class="stringliteral">&#39;transcriptionUnitBaseCounts&#39;</span>;
<a name="l00234"></a>00234             <span class="stringliteral">&#39;transcriptionUnitBindingProbabilities&#39;</span>;
<a name="l00235"></a>00235             };
<a name="l00236"></a>00236         fittedConstantNames__      = {   %names of fitted constant <a class="code" href="json_parse_8m.html#a6e4256f69b751d3d7ec4ea52ff5a8dee">properties</a>
<a name="l00237"></a>00237             <span class="stringliteral">&#39;transcriptionUnitBindingProbabilities&#39;</span>
<a name="l00238"></a>00238             };
<a name="l00239"></a>00239         localStateNames__          = {   %names of simulation state <a class="code" href="json_parse_8m.html#a6e4256f69b751d3d7ec4ea52ff5a8dee">properties</a> redundant with timecourses in <span class="keyword">this</span> or other processes or the simulation
<a name="l00240"></a>00240             <span class="stringliteral">&#39;RNAs&#39;</span>};
<a name="l00241"></a>00241     end
<a name="l00242"></a>00242     
<a name="l00243"></a>00243     %IDs, names, and local indices
<a name="l00244"></a>00244     <a class="code" href="json_parse_8m.html#a6e4256f69b751d3d7ec4ea52ff5a8dee">properties</a>
<a name="l00245"></a>00245         stimuliWholeCellModelIDs = {}; %whole cell model IDs of stimuli
<a name="l00246"></a>00246         
<a name="l00247"></a>00247         substrateWholeCellModelIDs = { %whole cell model IDs of substrates
<a name="l00248"></a>00248             <span class="stringliteral">&#39;ATP&#39;</span>; <span class="stringliteral">&#39;CTP&#39;</span>; <span class="stringliteral">&#39;GTP&#39;</span>; <span class="stringliteral">&#39;UTP&#39;</span>;
<a name="l00249"></a>00249             <span class="stringliteral">&#39;AMP&#39;</span>; <span class="stringliteral">&#39;CMP&#39;</span>; <span class="stringliteral">&#39;GMP&#39;</span>; <span class="stringliteral">&#39;UMP&#39;</span>;
<a name="l00250"></a>00250             <span class="stringliteral">&#39;ADP&#39;</span>; <span class="stringliteral">&#39;PPI&#39;</span>; <span class="stringliteral">&#39;H2O&#39;</span>; <span class="charliteral">&#39;H&#39;</span>};
<a name="l00251"></a>00251         substrateIndexs_ntp         = (1:4)<span class="stringliteral">&#39;; %indices within substrates of NTPs</span>
<a name="l00252"></a>00252 <span class="stringliteral">        substrateIndexs_nmp         = (5:8)&#39;</span>; %indices within substrates of NMPs
<a name="l00253"></a>00253         substrateIndexs_atp         = 1;      %index within substrates of ATP
<a name="l00254"></a>00254         substrateIndexs_adp         = 9;      %index within substrates of ADP
<a name="l00255"></a>00255         substrateIndexs_diphosphate = 10;     %index within substrates of diphosphate
<a name="l00256"></a>00256         substrateIndexs_water       = 11;     %index within substrates of water
<a name="l00257"></a>00257         substrateIndexs_hydrogen    = 12;     %index within substrates of hydrogen
<a name="l00258"></a>00258         
<a name="l00259"></a>00259         enzymeWholeCellModelIDs = {       %enzyme whole cell model ids
<a name="l00260"></a>00260             <span class="stringliteral">&#39;MG_249_MONOMER&#39;</span>;             %<a class="code" href="class_r_n_a.html">RNA</a> polymerase sigma factor RpoD
<a name="l00261"></a>00261             <span class="stringliteral">&#39;MG_282_MONOMER&#39;</span>;             %transcription elongation factor GreA
<a name="l00262"></a>00262             <span class="stringliteral">&#39;MG_141_MONOMER&#39;</span>;             %transcription termination factor NusA
<a name="l00263"></a>00263             <span class="stringliteral">&#39;MG_027_MONOMER&#39;</span>;             %transcription termination/antitermination protein NusB
<a name="l00264"></a>00264             <span class="stringliteral">&#39;RNA_POLYMERASE&#39;</span>;             %DNA-directed <a class="code" href="class_r_n_a.html">RNA</a> polymerase
<a name="l00265"></a>00265             &#39;RNA_POLYMERASE_HOLOENZYME&#39;}; %DNA-directed <a class="code" href="class_r_n_a.html">RNA</a> polymerase holoenzyme
<a name="l00266"></a>00266         enzymeIndexs_transcriptionFactors    = (1:4)&#39;; %indices within enzymes of transcription factors
<a name="l00267"></a>00267         enzymeIndexs_sigmaFactor             = 1;      %index within enzymes of sigma factor RpoD
<a name="l00268"></a>00268         enzymeIndexs_elongationFactor        = 2;      %index within enzymes of elongation factor GreA
<a name="l00269"></a>00269         enzymeIndexs_terminationFactor       = 3;      %index within enzymes of termination factor NusA
<a name="l00270"></a>00270         enzymeIndexs_antiterminationFactor   = 4;      %index within enzymes of termination/antitermination protein NusB
<a name="l00271"></a>00271         enzymeIndexs_rnaPolymerase           = 5;      %index within enzymes of DNA-directed <a class="code" href="class_r_n_a.html">RNA</a> polymerase
<a name="l00272"></a>00272         enzymeIndexs_rnaPolymeraseHoloenzyme = 6;      %index within enzymes of DNA-directed <a class="code" href="class_r_n_a.html">RNA</a> polymerase holoenzyme
<a name="l00273"></a>00273         
<a name="l00274"></a>00274         complexIndexs_DnaA_ATP                         %indices within protein complexes of DnaA-ATP
<a name="l00275"></a>00275         transcriptionUnitIndexs_DnaAR12345Boxes        %indices of transcription units containing the functional DnaA boxes R1-5
<a name="l00276"></a>00276     end
<a name="l00277"></a>00277     
<a name="l00278"></a>00278     %fixed biological constants
<a name="l00279"></a>00279     <a class="code" href="json_parse_8m.html#a6e4256f69b751d3d7ec4ea52ff5a8dee">properties</a>
<a name="l00280"></a>00280         rnaPolymeraseElongationRate                   %<a class="code" href="class_r_n_a.html">RNA</a> polymerase elongation rate (50 nucleotides per second per <a class="code" href="class_r_n_a.html">RNA</a> polymerase) [PUB_0562, PUB_0563]
<a name="l00281"></a>00281         transcriptionUnitBaseCounts                   %transcription <a class="code" href="unit_8m.html#a94c091bed34da4edc337e42e23029000">unit</a> base counts
<a name="l00282"></a>00282         transcriptionUnitBindingProbabilities         %transcription <a class="code" href="unit_8m.html#a94c091bed34da4edc337e42e23029000">unit</a> binding probabilities
<a name="l00283"></a>00283         stateTransitionProbabilities                  %transition probabilities among <a class="code" href="class_r_n_a.html">RNA</a> polymerase states
<a name="l00284"></a>00284     end
<a name="l00285"></a>00285     
<a name="l00286"></a>00286     %global state, stored locally for convenience
<a name="l00287"></a>00287     <a class="code" href="json_parse_8m.html#a6e4256f69b751d3d7ec4ea52ff5a8dee">properties</a>
<a name="l00288"></a>00288         RNAs                                          %copy number of transcription units
<a name="l00289"></a>00289     end
<a name="l00290"></a>00290     
<a name="l00291"></a>00291     %global state (referenced locally for convenience)
<a name="l00292"></a>00292     <a class="code" href="json_parse_8m.html#a6e4256f69b751d3d7ec4ea52ff5a8dee">properties</a>
<a name="l00293"></a>00293         rnaPolymerases   %<a class="code" href="class_r_n_a.html">RNA</a> Polymerase state class
<a name="l00294"></a>00294         transcripts      %New Transcripts state class
<a name="l00295"></a>00295     end
<a name="l00296"></a>00296     
<a name="l00297"></a>00297     %constructor
<a name="l00298"></a>00298     methods
<a name="l00299"></a>00299         function this = <a class="code" href="class_transcription.html">Transcription</a>(wholeCellModelID, name)
<a name="l00300"></a>00300             this = this@edu.stanford.covert.cell.sim.<a class="code" href="class_process.html">Process</a>(wholeCellModelID, name);
<a name="l00301"></a>00301         end
<a name="l00302"></a>00302     end
<a name="l00303"></a>00303     
<a name="l00304"></a>00304     %communication between process/simulation
<a name="l00305"></a>00305     methods
<a name="l00306"></a>00306         %<span class="keyword">set</span> references to state objects
<a name="l00307"></a>00307         <span class="keyword">function</span> storeObjectReferences(<span class="keyword">this</span>, simulation)
<a name="l00308"></a>00308             this.storeObjectReferences@edu.stanford.covert.cell.sim.<a class="code" href="class_process.html">Process</a>(simulation);
<a name="l00309"></a>00309             this.storeObjectReferences@edu.stanford.covert.cell.sim.<a class="code" href="class_chromosome_process_aspect.html">ChromosomeProcessAspect</a>(simulation);
<a name="l00310"></a>00310             
<a name="l00311"></a>00311             this.rnaPolymerases = simulation.state(&#39;RNAPolymerase&#39;);
<a name="l00312"></a>00312             this.transcripts = simulation.state(&#39;Transcript&#39;);
<a name="l00313"></a>00313             
<a name="l00314"></a>00314             this.states = [this.states; {this.rnaPolymerases; this.transcripts}];
<a name="l00315"></a>00315         end
<a name="l00316"></a>00316         
<a name="l00317"></a>00317         %initialize constants
<a name="l00318"></a>00318         <span class="keyword">function</span> initializeConstants(<span class="keyword">this</span>, knowledgeBase, simulation, varargin)
<a name="l00319"></a>00319             %call super <span class="keyword">class </span>method
<a name="l00320"></a>00320             this.initializeConstants@edu.stanford.covert.cell.sim.Process(knowledgeBase, simulation, varargin{:});
<a name="l00321"></a>00321             this.initializeConstants@edu.stanford.covert.cell.sim.ChromosomeProcessAspect(knowledgeBase, simulation, varargin{:});
<a name="l00322"></a>00322             
<a name="l00323"></a>00323             this.transcriptionUnitBaseCounts = reshape([knowledgeBase.transcriptionUnits.baseCount],[],length(knowledgeBase.transcriptionUnits))<span class="stringliteral">&#39;;</span>
<a name="l00324"></a>00324 <span class="stringliteral">            this.transcriptionUnitBaseCounts = this.transcriptionUnitBaseCounts(:, this.substrateMetaboliteGlobalIndexs);</span>
<a name="l00325"></a>00325 <span class="stringliteral">            </span>
<a name="l00326"></a>00326 <span class="stringliteral">            this.transcriptionUnitBindingProbabilities  = zeros(size(this.transcripts.transcriptionUnitLengths));</span>
<a name="l00327"></a>00327 <span class="stringliteral">            </span>
<a name="l00328"></a>00328 <span class="stringliteral">            this.complexIndexs_DnaA_ATP = find(ismember({knowledgeBase.proteinComplexs.wholeCellModelID}&#39;</span>, {
<a name="l00329"></a>00329                 <span class="stringliteral">&#39;MG_469_1MER_ATP&#39;</span>      %DnaA-ATP 1mer
<a name="l00330"></a>00330                 &#39;MG_469_2MER_1ATP_ADP&#39; %DnaA 2mer-(1)ATP-(1)ADP
<a name="l00331"></a>00331                 &#39;MG_469_2MER_ATP&#39;      %DnaA-ATP 2mer
<a name="l00332"></a>00332                 &#39;MG_469_3MER_2ATP_ADP&#39; %DnaA 3mer-(2)ATP-(1)ADP
<a name="l00333"></a>00333                 &#39;MG_469_3MER_ATP&#39;      %DnaA-ATP 3mer
<a name="l00334"></a>00334                 &#39;MG_469_4MER_3ATP_ADP&#39; %DnaA 4mer-(3)ATP-(1)ADP
<a name="l00335"></a>00335                 &#39;MG_469_4MER_ATP&#39;      %DnaA-ATP 4mer
<a name="l00336"></a>00336                 &#39;MG_469_5MER_4ATP_ADP&#39; %DnaA 5mer-(4)ATP-(1)ADP
<a name="l00337"></a>00337                 &#39;MG_469_5MER_ATP&#39;      %DnaA-ATP 5mer
<a name="l00338"></a>00338                 &#39;MG_469_6MER_5ATP_ADP&#39; %DnaA 6mer-(5)ATP-(1)ADP
<a name="l00339"></a>00339                 &#39;MG_469_6MER_ATP&#39;      %DnaA-ATP 6mer
<a name="l00340"></a>00340                 &#39;MG_469_7MER_6ATP_ADP&#39; %DnaA 7mer-(6)ATP-(1)ADP
<a name="l00341"></a>00341                 &#39;MG_469_7MER_ATP&#39;}));  %DnaA-ATP 7mer;
<a name="l00342"></a>00342             [~, idxs] = ismember({
<a name="l00343"></a>00343                 <span class="stringliteral">&#39;Functional box R1&#39;</span>
<a name="l00344"></a>00344                 <span class="stringliteral">&#39;Functional box R2&#39;</span>
<a name="l00345"></a>00345                 <span class="stringliteral">&#39;Functional box R3&#39;</span>
<a name="l00346"></a>00346                 <span class="stringliteral">&#39;Functional box R4&#39;</span>
<a name="l00347"></a>00347                 <span class="stringliteral">&#39;Functional box R5&#39;</span>}, {knowledgeBase.genomeFeatures.name});
<a name="l00348"></a>00348             tus = [knowledgeBase.genomeFeatures(idxs).transcriptionUnits];
<a name="l00349"></a>00349             this.transcriptionUnitIndexs_DnaAR12345Boxes = unique([tus.idx])<span class="stringliteral">&#39;;</span>
<a name="l00350"></a>00350 <span class="stringliteral">        end</span>
<a name="l00351"></a>00351 <span class="stringliteral">        </span>
<a name="l00352"></a>00352 <span class="stringliteral">        %retrieve state from simulation</span>
<a name="l00353"></a>00353 <span class="stringliteral">        function copyFromState(this)</span>
<a name="l00354"></a>00354 <span class="stringliteral">            this.copyFromState@edu.stanford.covert.cell.sim.Process();</span>
<a name="l00355"></a>00355 <span class="stringliteral">            </span>
<a name="l00356"></a>00356 <span class="stringliteral">            this.RNAs = this.rna.counts(this.rna.nascentIndexs, this.compartment.cytosolIndexs, :);</span>
<a name="l00357"></a>00357 <span class="stringliteral">        end</span>
<a name="l00358"></a>00358 <span class="stringliteral">        </span>
<a name="l00359"></a>00359 <span class="stringliteral">        %send state to simulation</span>
<a name="l00360"></a>00360 <span class="stringliteral">        function copyToState(this)</span>
<a name="l00361"></a>00361 <span class="stringliteral">            this.copyToState@edu.stanford.covert.cell.sim.Process();</span>
<a name="l00362"></a>00362 <span class="stringliteral">            </span>
<a name="l00363"></a>00363 <span class="stringliteral">            this.rna.counts(this.rna.nascentIndexs, this.compartment.cytosolIndexs, :) = this.RNAs;</span>
<a name="l00364"></a>00364 <span class="stringliteral">        end</span>
<a name="l00365"></a>00365 <span class="stringliteral">    end</span>
<a name="l00366"></a>00366 <span class="stringliteral">    </span>
<a name="l00367"></a>00367 <span class="stringliteral">    %allocate memory for state</span>
<a name="l00368"></a>00368 <span class="stringliteral">    methods</span>
<a name="l00369"></a>00369 <span class="stringliteral">        function allocateMemoryForState(this, numTimePoints)</span>
<a name="l00370"></a>00370 <span class="stringliteral">            this.allocateMemoryForState@edu.stanford.covert.cell.sim.Process(numTimePoints);</span>
<a name="l00371"></a>00371 <span class="stringliteral">            </span>
<a name="l00372"></a>00372 <span class="stringliteral">            numTranscriptionUnits = length(this.transcripts.transcriptionUnitLengths);</span>
<a name="l00373"></a>00373 <span class="stringliteral">            </span>
<a name="l00374"></a>00374 <span class="stringliteral">            this.RNAs = zeros(numTranscriptionUnits, 1, numTimePoints);</span>
<a name="l00375"></a>00375 <span class="stringliteral">        end</span>
<a name="l00376"></a>00376 <span class="stringliteral">    end</span>
<a name="l00377"></a>00377 <span class="stringliteral">    </span>
<a name="l00378"></a>00378 <span class="stringliteral">    %model</span>
<a name="l00379"></a>00379 <span class="stringliteral">    methods</span>
<a name="l00380"></a>00380 <span class="stringliteral">        %Calculate</span>
<a name="l00381"></a>00381 <span class="stringliteral">        %- contribution to FBA objective</span>
<a name="l00382"></a>00382 <span class="stringliteral">        %- minimum expression consistent with cell cycle length</span>
<a name="l00383"></a>00383 <span class="stringliteral">        function [bmProd, byProd, minEnzExp, maxEnzExp] = calcResourceRequirements_LifeCycle(this, ~, states)</span>
<a name="l00384"></a>00384 <span class="stringliteral">            %% initialize</span>
<a name="l00385"></a>00385 <span class="stringliteral">            bmProd = zeros(size(this.substrateWholeCellModelIDs));</span>
<a name="l00386"></a>00386 <span class="stringliteral">            byProd = zeros(size(this.substrateWholeCellModelIDs));</span>
<a name="l00387"></a>00387 <span class="stringliteral">            minEnzExp = zeros(size(this.enzymeWholeCellModelIDs));</span>
<a name="l00388"></a>00388 <span class="stringliteral">            maxEnzExp = Inf(size(this.enzymeWholeCellModelIDs));</span>
<a name="l00389"></a>00389 <span class="stringliteral">            </span>
<a name="l00390"></a>00390 <span class="stringliteral">            %% substrate and byproducts</span>
<a name="l00391"></a>00391 <span class="stringliteral">            %Contributions, by process, to RNA metabolism:</span>
<a name="l00392"></a>00392 <span class="stringliteral">            %                 NTPs + H2O --(transcription)--&gt; nascent RNA + PPi + H</span>
<a name="l00393"></a>00393 <span class="stringliteral">            %          nascent RNA + H2O --(processing)-----&gt; processed RNA + NMPs</span>
<a name="l00394"></a>00394 <span class="stringliteral">            %processed RNA + metabolites --(modification)---&gt; modified RNA + metabolites</span>
<a name="l00395"></a>00395 <span class="stringliteral">            %               modified RNA --(decay)----------&gt; NMPs + modified NMPs</span>
<a name="l00396"></a>00396 <span class="stringliteral">            %               NMPs + 2 ATP --(charging)-------&gt; NTPs + 2 ADP</span>
<a name="l00397"></a>00397 <span class="stringliteral">            %        modified NMPs + H2O --(catabolism)-----&gt; modified bases + Pi</span>
<a name="l00398"></a>00398 <span class="stringliteral">            %</span>
<a name="l00399"></a>00399 <span class="stringliteral">            %Here we only account for the contributions of transcription.</span>
<a name="l00400"></a>00400 <span class="stringliteral">            invMat = edu.stanford.covert.util.ComputationUtil.invertCompositionMatrix(this.rna.nascentRNAMatureRNAComposition);</span>
<a name="l00401"></a>00401 <span class="stringliteral">            </span>
<a name="l00402"></a>00402 <span class="stringliteral">            transcribedNTPs = this.transcriptionUnitBaseCounts(:, this.substrateIndexs_nmp)&#39;</span> * invMat * states.rnaProductions;
<a name="l00403"></a>00403             bmProd(this.substrateIndexs_ntp)         = bmProd(this.substrateIndexs_ntp)         + transcribedNTPs;
<a name="l00404"></a>00404             byProd(this.substrateIndexs_diphosphate) = byProd(this.substrateIndexs_diphosphate) + sum(transcribedNTPs);
<a name="l00405"></a>00405             bmProd(this.substrateIndexs_water)       = bmProd(this.substrateIndexs_water)       + sum(states.rnaProductions);
<a name="l00406"></a>00406             byProd(this.substrateIndexs_hydrogen)    = byProd(this.substrateIndexs_hydrogen)    + sum(states.rnaProductions);
<a name="l00407"></a>00407             
<a name="l00408"></a>00408             %% enzymes
<a name="l00409"></a>00409             %<a class="code" href="class_r_n_a.html">RNA</a> polymerase
<a name="l00410"></a>00410             fractionInitiating = this.rnaPolymeraseElongationRate * sum(invMat * states.rnaProductions0) / ... %fraction of active <a class="code" href="class_r_n_a.html">RNA</a> polymerases that are initiating
<a name="l00411"></a>00411                 (this.transcripts.transcriptionUnitLengths<span class="stringliteral">&#39; * invMat * states.rnaProductions0);</span>
<a name="l00412"></a>00412 <span class="stringliteral">            minEnzExp(this.enzymeIndexs_rnaPolymerase) = 4.5867 * ...</span>
<a name="l00413"></a>00413 <span class="stringliteral">                (this.transcripts.transcriptionUnitLengths&#39;</span> * invMat * states.rnaProductions0) / ...
<a name="l00414"></a>00414                 this.rnaPolymeraseElongationRate / ...
<a name="l00415"></a>00415                 this.rnaPolymerases.stateExpectations(this.rnaPolymerases.activelyTranscribingIndex) / ...
<a name="l00416"></a>00416                 (1 - fractionInitiating);
<a name="l00417"></a>00417             
<a name="l00418"></a>00418             %sigma, elongation, termination factors
<a name="l00419"></a>00419             minEnzExp(this.enzymeIndexs_sigmaFactor) = minEnzExp(this.enzymeIndexs_rnaPolymerase) * ...
<a name="l00420"></a>00420                 (this.rnaPolymerases.stateExpectations(this.rnaPolymerases.activelyTranscribingIndex) * fractionInitiating + ...
<a name="l00421"></a>00421                 this.rnaPolymerases.stateExpectations(this.rnaPolymerases.specificallyBoundIndex));
<a name="l00422"></a>00422             minEnzExp(this.enzymeIndexs_elongationFactor)  = 2;
<a name="l00423"></a>00423             minEnzExp(this.enzymeIndexs_terminationFactor) = 2;
<a name="l00424"></a>00424         end
<a name="l00425"></a>00425         
<a name="l00426"></a>00426         %Initialize <a class="code" href="class_r_n_a.html">RNA</a> polymerase, bound transcription units, transcription factor states
<a name="l00427"></a>00427         %Assumptions:
<a name="l00428"></a>00428         %- Metabolic cost of the transcription of these nucleic acids is negligible
<a name="l00429"></a>00429         %- Probability that so many <a class="code" href="class_r_n_a.html">RNA</a> polymerases bind a given transcription <a class="code" href="unit_8m.html#a94c091bed34da4edc337e42e23029000">unit</a>
<a name="l00430"></a>00430         %  that they can&#39;t pack along the transcription <a class="code" href="unit_8m.html#a94c091bed34da4edc337e42e23029000">unit</a> without steric
<a name="l00431"></a>00431         %  interactions is negligible. Thus we don&#39;t try to handle this case
<a name="l00432"></a>00432         %  separately
<a name="l00433"></a>00433         function initializeState(this)
<a name="l00434"></a>00434             rnaPols = this.rnaPolymerases;
<a name="l00435"></a>00435             trnspts = this.transcripts;
<a name="l00436"></a>00436             
<a name="l00437"></a>00437             %% states
<a name="l00438"></a>00438             trnsptStrnds = 2 - trnspts.transcriptionUnitDirections;
<a name="l00439"></a>00439             trnsptLens = trnspts.transcriptionUnitLengths;
<a name="l00440"></a>00440             trnsptDirs = 2 * trnspts.transcriptionUnitDirections - 1;
<a name="l00441"></a>00441             trnspt5Coords = trnspts.transcriptionUnitFivePrimeCoordinates;
<a name="l00442"></a>00442             trnsptStarts = ...
<a name="l00443"></a>00443                 trnspt5Coords - (1 - trnspts.transcriptionUnitDirections) .* (trnspts.transcriptionUnitLengths-1);
<a name="l00444"></a>00444             trnsptStarts(trnsptDirs == 1) = ...
<a name="l00445"></a>00445                 trnsptStarts(trnsptDirs == 1) - ...
<a name="l00446"></a>00446                 this.enzymeDNAFootprints5Prime(this.enzymeIndexs_rnaPolymerase);
<a name="l00447"></a>00447             trnsptStarts(trnsptDirs == -1) = ...
<a name="l00448"></a>00448                 trnsptStarts(trnsptDirs == -1) + ...
<a name="l00449"></a>00449                 this.enzymeDNAFootprints5Prime(this.enzymeIndexs_rnaPolymerase);
<a name="l00450"></a>00450             
<a name="l00451"></a>00451             %% enzymes
<a name="l00452"></a>00452             freTFs = this.enzymes(this.enzymeIndexs_transcriptionFactors);
<a name="l00453"></a>00453             bndTFs = this.boundEnzymes(this.enzymeIndexs_transcriptionFactors);
<a name="l00454"></a>00454             frePls = this.enzymes(this.enzymeIndexs_rnaPolymerase);
<a name="l00455"></a>00455             bndPls = this.boundEnzymes(this.enzymeIndexs_rnaPolymerase);
<a name="l00456"></a>00456             freHls = this.enzymes(this.enzymeIndexs_rnaPolymeraseHoloenzyme);
<a name="l00457"></a>00457             bndHls = this.boundEnzymes(this.enzymeIndexs_rnaPolymeraseHoloenzyme);
<a name="l00458"></a>00458             pls = frePls + bndPls + freHls + bndHls;
<a name="l00459"></a>00459             
<a name="l00460"></a>00460             %% allocate
<a name="l00461"></a>00461             rnaPols.states = repmat(rnaPols.notExistValue, [2 * pls, 1, 1]);
<a name="l00462"></a>00462             rnaPols.positionStrands = zeros([2 * pls, 2, 1]);
<a name="l00463"></a>00463             trnspts.boundTranscriptionUnits = repmat(trnspts.nullTranscriptValue, [2 * pls, 1, 1]);
<a name="l00464"></a>00464             trnspts.boundTranscriptProgress = repmat(trnspts.nullTranscriptValue, [2 * pls, 1, 1]);
<a name="l00465"></a>00465             trnspts.boundTranscriptChromosome = repmat(trnspts.nullTranscriptValue, [2 * pls, 1, 1]);
<a name="l00466"></a>00466             
<a name="l00467"></a>00467             %% initialize
<a name="l00468"></a>00468             pState = rnaPols.stateExpectations;
<a name="l00469"></a>00469             probs = this.computeRNAPolymeraseTUBindingProbabilities();
<a name="l00470"></a>00470             pBinds = trnspts.transcriptionUnitLengths .* probs(:, 1);
<a name="l00471"></a>00471             pSpBinds = pBinds;
<a name="l00472"></a>00472             
<a name="l00473"></a>00473             <span class="keywordflow">for</span> i = 1:pls
<a name="l00474"></a>00474                 %check whether conditions necessary <span class="keywordflow">for</span> actively transcribing state are met
<a name="l00475"></a>00475                 <span class="keywordflow">if</span> frePls == 0 || ~any(pBinds) || freTFs(this.enzymeIndexs_elongationFactor) == 0
<a name="l00476"></a>00476                     pState(rnaPols.activelyTranscribingIndex) = 0;
<a name="l00477"></a>00477                 end
<a name="l00478"></a>00478                 
<a name="l00479"></a>00479                 %check whether conditions necessary <span class="keywordflow">for</span> specifically bound state are met
<a name="l00480"></a>00480                 <span class="keywordflow">if</span> frePls == 0 || ~any(pSpBinds) || freTFs(this.enzymeIndexs_sigmaFactor) == 0
<a name="l00481"></a>00481                     pState(rnaPols.specificallyBoundIndex) = 0;
<a name="l00482"></a>00482                 end
<a name="l00483"></a>00483                 
<a name="l00484"></a>00484                 %check whether conditions necessary <span class="keywordflow">for</span> non-specifically bound state are met
<a name="l00485"></a>00485                 if frePls == 0
<a name="l00486"></a>00486                     pState(rnaPols.nonSpecificallyBoundIndex) = 0;
<a name="l00487"></a>00487                 end
<a name="l00488"></a>00488                 
<a name="l00489"></a>00489                 <span class="keywordflow">if</span> ~any(pState)
<a name="l00490"></a>00490                     pState(rnaPols.freeIndex) = 1;
<a name="l00491"></a>00491                 end
<a name="l00492"></a>00492                 
<a name="l00493"></a>00493                 %randomly select state
<a name="l00494"></a>00494                 state = this.randStream.randsample(4, 1, true, pState);
<a name="l00495"></a>00495                 
<a name="l00496"></a>00496                 %Actively transcribing, specifically bound: randomly select
<a name="l00497"></a>00497                 %transcription <a class="code" href="unit_8m.html#a94c091bed34da4edc337e42e23029000">unit</a>
<a name="l00498"></a>00498                 switch state
<a name="l00499"></a>00499                     case rnaPols.activelyTranscribingIndex
<a name="l00500"></a>00500                         tmp_pBinds = pBinds;
<a name="l00501"></a>00501                         posStrnds = zeros(0, 2);
<a name="l00502"></a>00502                         <a class="code" href="json_parse_8m.html#ad2fc293e8080616c8c1e2ffbd7b86742">while</a> any(tmp_pBinds)
<a name="l00503"></a>00503                             iTU = this.randStream.randsample(numel(pBinds), 1, true, tmp_pBinds);
<a name="l00504"></a>00504                             tmp_pBinds(iTU) = 0;
<a name="l00505"></a>00505                             [~, posStrnds] = this.bindProteinToChromosomeStochastically(...
<a name="l00506"></a>00506                                 this.enzymeIndexs_rnaPolymerase, 1, [trnsptStarts(iTU)+1 trnsptStrnds(iTU)], trnsptLens(iTU)-2);
<a name="l00507"></a>00507                             if ~isempty(posStrnds)
<a name="l00508"></a>00508                                 break;
<a name="l00509"></a>00509                             end
<a name="l00510"></a>00510                         end
<a name="l00511"></a>00511                         if isempty(posStrnds)
<a name="l00512"></a>00512                             throw(MException(&#39;<a class="code" href="class_transcription.html">Transcription</a>:error&#39;, &#39;unable to bind chromosome&#39;));
<a name="l00513"></a>00513                         end
<a name="l00514"></a>00514                         
<a name="l00515"></a>00515                         posStrnds( isodd(posStrnds(:, 2)), 1) = posStrnds( isodd(posStrnds(:, 2)), 1) + this.enzymeDNAFootprints5Prime(this.enzymeIndexs_rnaPolymerase);
<a name="l00516"></a>00516                         posStrnds(iseven(posStrnds(:, 2)), 1) = posStrnds(iseven(posStrnds(:, 2)), 1) + this.enzymeDNAFootprints3Prime(this.enzymeIndexs_rnaPolymerase);
<a name="l00517"></a>00517                         if trnspts.transcriptionUnitDirections(iTU)
<a name="l00518"></a>00518                             state = posStrnds(:, 1) - trnspt5Coords(iTU) + 1;
<a name="l00519"></a>00519                         else
<a name="l00520"></a>00520                             state = trnspt5Coords(iTU) - posStrnds(:, 1) + 1;
<a name="l00521"></a>00521                         end
<a name="l00522"></a>00522                         
<a name="l00523"></a>00523                         rnaPols.states(i) = state;
<a name="l00524"></a>00524                         rnaPols.positionStrands(i, :) = posStrnds;
<a name="l00525"></a>00525                         trnspts.boundTranscriptProgress(i) = state;
<a name="l00526"></a>00526                         trnspts.boundTranscriptionUnits(i) = iTU;
<a name="l00527"></a>00527                         trnspts.boundTranscriptChromosome(i) = 1; %initialize to first chromosome
<a name="l00528"></a>00528                         
<a name="l00529"></a>00529                         frePls = frePls - 1;
<a name="l00530"></a>00530                         bndPls = bndPls + 1;
<a name="l00531"></a>00531                         
<a name="l00532"></a>00532                         %freTFs(this.enzymeIndexs_elongationFactor) = freTFs(this.enzymeIndexs_elongationFactor) - 1;
<a name="l00533"></a>00533                         %bndTFs(this.enzymeIndexs_elongationFactor) = bndTFs(this.enzymeIndexs_elongationFactor) + 1;
<a name="l00534"></a>00534                     case rnaPols.specificallyBoundIndex
<a name="l00535"></a>00535                         <a class="code" href="json_parse_8m.html#ad2fc293e8080616c8c1e2ffbd7b86742">while</a> any(pSpBinds)
<a name="l00536"></a>00536                             iTU = this.randStream.randsample(numel(pSpBinds), 1, true, pSpBinds);
<a name="l00537"></a>00537                             pSpBinds(iTU) = 0;
<a name="l00538"></a>00538                             [~, ~, posStrnds] = this.bindProteinToChromosome([trnspt5Coords(iTU)-trnsptDirs(iTU) trnsptStrnds(iTU)], this.enzymeIndexs_rnaPolymeraseHoloenzyme, 1);
<a name="l00539"></a>00539                             if ~isempty(posStrnds)
<a name="l00540"></a>00540                                 break;
<a name="l00541"></a>00541                             end
<a name="l00542"></a>00542                         end
<a name="l00543"></a>00543                         if isempty(posStrnds)
<a name="l00544"></a>00544                             throw(MException(&#39;<a class="code" href="class_transcription.html">Transcription</a>:error&#39;, &#39;unable to bind chromosome&#39;));
<a name="l00545"></a>00545                         end
<a name="l00546"></a>00546                         
<a name="l00547"></a>00547                         rnaPols.states(i) = rnaPols.specificallyBoundValue;
<a name="l00548"></a>00548                         rnaPols.positionStrands(i, :) = posStrnds;
<a name="l00549"></a>00549                         trnspts.boundTranscriptProgress(i) = trnspts.nullTranscriptValue;
<a name="l00550"></a>00550                         trnspts.boundTranscriptionUnits(i) = iTU;
<a name="l00551"></a>00551                         trnspts.boundTranscriptChromosome(i) = 1; %initialize to first chromosome
<a name="l00552"></a>00552                         
<a name="l00553"></a>00553                         bndHls = bndHls + 1;
<a name="l00554"></a>00554                         frePls = frePls - 1;
<a name="l00555"></a>00555                         freTFs(this.enzymeIndexs_sigmaFactor) = freTFs(this.enzymeIndexs_sigmaFactor) - 1;
<a name="l00556"></a>00556                     case rnaPols.nonSpecificallyBoundIndex
<a name="l00557"></a>00557                         [~, posStrnds] = this.bindProteinToChromosomeStochastically(this.enzymeIndexs_rnaPolymerase, 1);
<a name="l00558"></a>00558                         if isempty(posStrnds)
<a name="l00559"></a>00559                             throw(MException(&#39;<a class="code" href="class_transcription.html">Transcription</a>:error&#39;, &#39;unable to bind chromosome&#39;));
<a name="l00560"></a>00560                         end
<a name="l00561"></a>00561                         posStrnds( isodd(posStrnds(:, 2)), 1) = posStrnds( isodd(posStrnds(:, 2)), 1) + this.enzymeDNAFootprints5Prime(this.enzymeIndexs_rnaPolymerase);
<a name="l00562"></a>00562                         posStrnds(iseven(posStrnds(:, 2)), 1) = posStrnds(iseven(posStrnds(:, 2)), 1) + this.enzymeDNAFootprints3Prime(this.enzymeIndexs_rnaPolymerase);
<a name="l00563"></a>00563                         
<a name="l00564"></a>00564                         rnaPols.states(i) = rnaPols.nonSpecificallyBoundValue;
<a name="l00565"></a>00565                         rnaPols.positionStrands(i, :) = posStrnds;
<a name="l00566"></a>00566                         trnspts.boundTranscriptionUnits(i) = trnspts.nullTranscriptValue;
<a name="l00567"></a>00567                         trnspts.boundTranscriptProgress(i) = trnspts.nullTranscriptValue;
<a name="l00568"></a>00568                         trnspts.boundTranscriptChromosome(i) = trnspts.nullTranscriptValue;
<a name="l00569"></a>00569                         
<a name="l00570"></a>00570                         frePls = frePls - 1;
<a name="l00571"></a>00571                         bndPls = bndPls + 1;
<a name="l00572"></a>00572                     case rnaPols.freeIndex
<a name="l00573"></a>00573                         rnaPols.states(i) = rnaPols.freeValue;
<a name="l00574"></a>00574                         rnaPols.positionStrands(i, :) = 0;
<a name="l00575"></a>00575                         trnspts.boundTranscriptionUnits(i) = trnspts.nullTranscriptValue;
<a name="l00576"></a>00576                         trnspts.boundTranscriptProgress(i) = trnspts.nullTranscriptValue;
<a name="l00577"></a>00577                         trnspts.boundTranscriptChromosome(i) = trnspts.nullTranscriptValue;
<a name="l00578"></a>00578                 end
<a name="l00579"></a>00579             end
<a name="l00580"></a>00580             
<a name="l00581"></a>00581             %% store states of enzymes
<a name="l00582"></a>00582             this.enzymes(this.enzymeIndexs_transcriptionFactors)      = freTFs;
<a name="l00583"></a>00583             this.boundEnzymes(this.enzymeIndexs_transcriptionFactors) = bndTFs;
<a name="l00584"></a>00584             this.enzymes(this.enzymeIndexs_rnaPolymerase)             = frePls;
<a name="l00585"></a>00585             this.boundEnzymes(this.enzymeIndexs_rnaPolymerase)        = bndPls;
<a name="l00586"></a>00586             this.enzymes(this.enzymeIndexs_rnaPolymeraseHoloenzyme)   = freHls;
<a name="l00587"></a>00587             this.boundEnzymes(this.enzymeIndexs_rnaPolymeraseHoloenzyme) = bndHls;
<a name="l00588"></a>00588             
<a name="l00589"></a>00589             %% decrement counts of RNAs
<a name="l00590"></a>00590             this.copyToState();
<a name="l00591"></a>00591             
<a name="l00592"></a>00592             comp = this.compartment;
<a name="l00593"></a>00593             rna = this.rna;
<a name="l00594"></a>00594             
<a name="l00595"></a>00595             matureRnaWt = max(0, sum(rna.dryWeight) - trnspts.dryWeight);
<a name="l00596"></a>00596             initRnaCnts = rna.counts;
<a name="l00597"></a>00597             <a class="code" href="json_parse_8m.html#ad2fc293e8080616c8c1e2ffbd7b86742">while</a> sum(rna.dryWeight) &gt; matureRnaWt
<a name="l00598"></a>00598                 idx = this.randStream.randsample(size(rna.counts, 1), 1, false, rna.counts(:, comp.cytosolIndexs));
<a name="l00599"></a>00599                 rna.counts(idx, comp.cytosolIndexs) = rna.counts(idx, comp.cytosolIndexs) - 1;
<a name="l00600"></a>00600             end
<a name="l00601"></a>00601             rna.counts(:, comp.cytosolIndexs) = ...
<a name="l00602"></a>00602                 + rna.counts(:, comp.cytosolIndexs) ...
<a name="l00603"></a>00603                 + rna.updateExternalState(-(initRnaCnts(:, comp.cytosolIndexs) - rna.counts(:, comp.cytosolIndexs)), true);
<a name="l00604"></a>00604             
<a name="l00605"></a>00605             this.copyFromState();
<a name="l00606"></a>00606         end
<a name="l00607"></a>00607         
<a name="l00608"></a>00608         %resource requirements
<a name="l00609"></a>00609         function result = calcResourceRequirements_Current(this)
<a name="l00610"></a>00610             result = zeros(size(this.substrates));
<a name="l00611"></a>00611             result(this.substrateIndexs_ntp) = 2 * ...
<a name="l00612"></a>00612                 sum(this.metabolite.nmpComposition, 2) * this.rnaPolymerases.nActive * ...
<a name="l00613"></a>00613                 this.rnaPolymeraseElongationRate;
<a name="l00614"></a>00614             result(this.substrateIndexs_water) = this.rnaPolymerases.nActive;
<a name="l00615"></a>00615         end
<a name="l00616"></a>00616         
<a name="l00617"></a>00617         %simulation
<a name="l00618"></a>00618         function evolveState(this)
<a name="l00619"></a>00619             %% define and allocate variables
<a name="l00620"></a>00620             
<a name="l00621"></a>00621             %states
<a name="l00622"></a>00622             c = this.chromosome;
<a name="l00623"></a>00623             rnaPols = this.rnaPolymerases;
<a name="l00624"></a>00624             trnspts = this.transcripts;
<a name="l00625"></a>00625             
<a name="l00626"></a>00626             %numbers of enzymes
<a name="l00627"></a>00627             nFreTfs  = this.enzymes(this.enzymeIndexs_transcriptionFactors);
<a name="l00628"></a>00628             nBndTFs  = this.boundEnzymes(this.enzymeIndexs_transcriptionFactors);
<a name="l00629"></a>00629             nFrePols = this.enzymes(this.enzymeIndexs_rnaPolymerase);
<a name="l00630"></a>00630             nBndPols = this.boundEnzymes(this.enzymeIndexs_rnaPolymerase);
<a name="l00631"></a>00631             nFreHols = this.enzymes(this.enzymeIndexs_rnaPolymeraseHoloenzyme);
<a name="l00632"></a>00632             nBndHols = this.boundEnzymes(this.enzymeIndexs_rnaPolymeraseHoloenzyme);
<a name="l00633"></a>00633             nTotPols = nFrePols + nBndPols + nFreHols + nBndHols;
<a name="l00634"></a>00634             
<a name="l00635"></a>00635             %<a class="code" href="json_parse_8m.html#a6e4256f69b751d3d7ec4ea52ff5a8dee">properties</a> of <a class="code" href="class_r_n_a.html">RNA</a> polymerases
<a name="l00636"></a>00636             pStTrns = this.stateTransitionProbabilities; %probabilities of state transitions
<a name="l00637"></a>00637             
<a name="l00638"></a>00638             %transcription <a class="code" href="unit_8m.html#a94c091bed34da4edc337e42e23029000">unit</a> <a class="code" href="json_parse_8m.html#a6e4256f69b751d3d7ec4ea52ff5a8dee">properties</a>
<a name="l00639"></a>00639             tuDirs = 2 * trnspts.transcriptionUnitDirections - 1;
<a name="l00640"></a>00640             tuStrnds = c.transcriptionUnitStrands;
<a name="l00641"></a>00641             tuLens = trnspts.transcriptionUnitLengths;
<a name="l00642"></a>00642             tu5Coords = trnspts.transcriptionUnitFivePrimeCoordinates;
<a name="l00643"></a>00643             tuSeqs = trnspts.transcriptionUnitSequences;  %sequences
<a name="l00644"></a>00644             
<a name="l00645"></a>00645             %% Update states of <a class="code" href="class_r_n_a.html">RNA</a> polymerases
<a name="l00646"></a>00646             
<a name="l00647"></a>00647             %allocate space to store states of new <a class="code" href="class_r_n_a.html">RNA</a> polymerases
<a name="l00648"></a>00648             if nTotPols &gt; size(rnaPols.states, 1)
<a name="l00649"></a>00649                 rnaPols.states = [
<a name="l00650"></a>00650                     rnaPols.states;
<a name="l00651"></a>00651                     rnaPols.notExistValue(ones(2 * nTotPols - size(rnaPols.states, 1), 1), 1)];
<a name="l00652"></a>00652                 rnaPols.positionStrands = [
<a name="l00653"></a>00653                     rnaPols.positionStrands;
<a name="l00654"></a>00654                     zeros(2 * nTotPols - size(rnaPols.positionStrands, 1), 2)];
<a name="l00655"></a>00655                 trnspts.boundTranscriptionUnits = [
<a name="l00656"></a>00656                     trnspts.boundTranscriptionUnits;
<a name="l00657"></a>00657                     trnspts.nullTranscriptValue(ones(2 * nTotPols - size(trnspts.boundTranscriptChromosome, 1), 1), 1)];
<a name="l00658"></a>00658                 trnspts.boundTranscriptProgress = [
<a name="l00659"></a>00659                     trnspts.boundTranscriptProgress;
<a name="l00660"></a>00660                     trnspts.nullTranscriptValue(ones(2 * nTotPols - size(trnspts.boundTranscriptChromosome, 1), 1), 1)];
<a name="l00661"></a>00661                 trnspts.boundTranscriptChromosome = [
<a name="l00662"></a>00662                     trnspts.boundTranscriptChromosome;
<a name="l00663"></a>00663                     trnspts.nullTranscriptValue(ones(2 * nTotPols - size(trnspts.boundTranscriptChromosome, 1), 1), 1)];
<a name="l00664"></a>00664             end
<a name="l00665"></a>00665             
<a name="l00666"></a>00666             %indices of <a class="code" href="class_r_n_a.html">RNA</a> polymerases in various states
<a name="l00667"></a>00667             actPols = find(rnaPols.states &gt;= rnaPols.activelyTranscribingValue);
<a name="l00668"></a>00668             sbPols = find(rnaPols.states == rnaPols.specificallyBoundValue);
<a name="l00669"></a>00669             freeNsbPols = find(...
<a name="l00670"></a>00670                 rnaPols.states == rnaPols.nonSpecificallyBoundValue | ...
<a name="l00671"></a>00671                 rnaPols.states == rnaPols.freeValue);
<a name="l00672"></a>00672             sbPols = sbPols(this.randStream.randperm(numel(sbPols)));
<a name="l00673"></a>00673             freeNsbPols = freeNsbPols(this.randStream.randperm(numel(freeNsbPols)));
<a name="l00674"></a>00674             
<a name="l00675"></a>00675             %number of new <a class="code" href="class_r_n_a.html">RNA</a> polymerases
<a name="l00676"></a>00676             nNewPols = nTotPols - (numel(actPols) + numel(sbPols) + numel(freeNsbPols));
<a name="l00677"></a>00677             
<a name="l00678"></a>00678             %dissociate any free <a class="code" href="class_r_n_a.html">RNA</a> polymerase holoenzymes (created say by
<a name="l00679"></a>00679             %their release from chromosome by another protein) into <a class="code" href="class_r_n_a.html">RNA</a>
<a name="l00680"></a>00680             %polymerase core and sigma factor
<a name="l00681"></a>00681             nFrePols = nFrePols + nFreHols;
<a name="l00682"></a>00682             nFreTfs(this.enzymeIndexs_sigmaFactor) = nFreTfs(this.enzymeIndexs_sigmaFactor) + nFreHols;
<a name="l00683"></a>00683             nFreHols = 0;
<a name="l00684"></a>00684             
<a name="l00685"></a>00685             %initiating specifically bound polymerases
<a name="l00686"></a>00686             nInitMax = this.randStream.stochasticRound(numel(sbPols) * pStTrns(rnaPols.activelyTranscribingIndex, rnaPols.specificallyBoundIndex)); %<span class="preprocessor">#ok&lt;*PROP&gt;</span>
<a name="l00687"></a>00687 <span class="preprocessor"></span>            <span class="keywordflow">if</span> nInitMax &gt; 0
<a name="l00688"></a>00688                 releasedProteins = this.releaseProteinFromSites(rnaPols.positionStrands(sbPols(1:nInitMax), :), <span class="keyword">false</span>, this.enzymeIndexs_rnaPolymeraseHoloenzyme, <span class="keyword">true</span>, <span class="keyword">true</span>);
<a name="l00689"></a>00689                 <span class="keywordflow">if</span> ~isequal(releasedProteins(this.enzymeIndexs_rnaPolymeraseHoloenzyme), nInitMax)
<a name="l00690"></a>00690                     throw(MException(&#39;<a class="code" href="class_transcription.html">Transcription</a>:error&#39;, &#39;Unable to release protein&#39;));
<a name="l00691"></a>00691                 end
<a name="l00692"></a>00692                 
<a name="l00693"></a>00693                 %try to initiate <a class="code" href="class_r_n_a.html">RNA</a> polymerases
<a name="l00694"></a>00694                 positions = tu5Coords(trnspts.boundTranscriptionUnits(sbPols(1:nInitMax)));
<a name="l00695"></a>00695                 tfs = this.bindProteinToChromosome([positions rnaPols.positionStrands(sbPols(1:nInitMax), 2)], ...
<a name="l00696"></a>00696                     this.enzymeIndexs_rnaPolymeraseHoloenzyme);
<a name="l00697"></a>00697                 rnaPols.states(sbPols(tfs)) = rnaPols.activelyTranscribingValue;
<a name="l00698"></a>00698                 trnspts.boundTranscriptProgress(sbPols(tfs)) = trnspts.newTranscriptValue;
<a name="l00699"></a>00699                 rnaPols.positionStrands(sbPols(tfs), 1) = positions(tfs, 1);
<a name="l00700"></a>00700                 
<a name="l00701"></a>00701                 %rebind <a class="code" href="class_r_n_a.html">RNA</a> polymerases that couldn&#39;t move forward
<a name="l00702"></a>00702                 if ~all(this.bindProteinToChromosome(rnaPols.positionStrands(sbPols(~tfs), :), ...
<a name="l00703"></a>00703                         this.enzymeIndexs_rnaPolymeraseHoloenzyme))
<a name="l00704"></a>00704                     throw(MException(&#39;<a class="code" href="class_transcription.html">Transcription</a>:error&#39;, &#39;Unable to unbind protein&#39;));
<a name="l00705"></a>00705                 end
<a name="l00706"></a>00706             end
<a name="l00707"></a>00707             
<a name="l00708"></a>00708             %specifically bound polymerases becoming non-specifically bound / free
<a name="l00709"></a>00709             nUnsb = max(0, numel(sbPols) - nInitMax - this.randStream.stochasticRound(numel(sbPols) * ...
<a name="l00710"></a>00710                 pStTrns(rnaPols.specificallyBoundIndex, rnaPols.specificallyBoundIndex)));
<a name="l00711"></a>00711             if nUnsb &gt; 0
<a name="l00712"></a>00712                 idxs = sbPols(end-nUnsb+1:end);
<a name="l00713"></a>00713                 posStrnds = rnaPols.positionStrands(idxs, :);
<a name="l00714"></a>00714                 releasedProteins = this.releaseProteinFromSites(posStrnds, false, this.enzymeIndexs_rnaPolymeraseHoloenzyme, true, true);
<a name="l00715"></a>00715                 if ~isequal(releasedProteins(this.enzymeIndexs_rnaPolymeraseHoloenzyme), nUnsb)
<a name="l00716"></a>00716                     throw(MException(&#39;<a class="code" href="class_transcription.html">Transcription</a>:error&#39;, &#39;Unable to unbind protein&#39;));
<a name="l00717"></a>00717                 end
<a name="l00718"></a>00718                 
<a name="l00719"></a>00719                 rnaPols.states(idxs) = rnaPols.freeValue;
<a name="l00720"></a>00720                 rnaPols.positionStrands(idxs, :) = 0;
<a name="l00721"></a>00721                 trnspts.boundTranscriptionUnits(idxs) = trnspts.nullTranscriptValue;
<a name="l00722"></a>00722                 trnspts.boundTranscriptProgress(idxs) = trnspts.nullTranscriptValue;
<a name="l00723"></a>00723                 trnspts.boundTranscriptChromosome(idxs) = trnspts.nullTranscriptValue;
<a name="l00724"></a>00724                 
<a name="l00725"></a>00725                 nBndHols = nBndHols - nUnsb;                
<a name="l00726"></a>00726                 nFrePols = nFrePols + nUnsb;
<a name="l00727"></a>00727                 nFreTfs(this.enzymeIndexs_sigmaFactor) = nFreTfs(this.enzymeIndexs_sigmaFactor) + nUnsb;
<a name="l00728"></a>00728             end
<a name="l00729"></a>00729             
<a name="l00730"></a>00730             %free polymerases/non-specifically bound becoming specifically bound
<a name="l00731"></a>00731             nFreeNsbPols = this.randStream.stochasticRound(min([
<a name="l00732"></a>00732                 numel(freeNsbPols) * pStTrns(rnaPols.specificallyBoundIndex, rnaPols.freeIndex)
<a name="l00733"></a>00733                 nFreTfs(this.enzymeIndexs_sigmaFactor)]));
<a name="l00734"></a>00734             if nFreeNsbPols &gt; 0
<a name="l00735"></a>00735                 %unbind from non-specifically bound site
<a name="l00736"></a>00736                 idxs = freeNsbPols(1:nFreeNsbPols);
<a name="l00737"></a>00737                 nNsbPols = sum(rnaPols.states(idxs) == rnaPols.nonSpecificallyBoundValue);
<a name="l00738"></a>00738                 releasedProteins = this.releaseProteinFromSites(rnaPols.positionStrands(idxs, :), false, this.enzymeIndexs_rnaPolymerase, true, true);
<a name="l00739"></a>00739                 if ~isequal(releasedProteins(this.enzymeIndexs_rnaPolymerase), nNsbPols)
<a name="l00740"></a>00740                     throw(MException(&#39;<a class="code" href="class_transcription.html">Transcription</a>:error&#39;, &#39;Unable to release protein&#39;));
<a name="l00741"></a>00741                 end
<a name="l00742"></a>00742                 
<a name="l00743"></a>00743                 rnaPols.states(idxs) = rnaPols.freeValue;
<a name="l00744"></a>00744                 rnaPols.positionStrands(idxs, :) = 0;
<a name="l00745"></a>00745                 trnspts.boundTranscriptionUnits(idxs) = trnspts.nullTranscriptValue;
<a name="l00746"></a>00746                 trnspts.boundTranscriptProgress(idxs) = trnspts.nullTranscriptValue;
<a name="l00747"></a>00747                 trnspts.boundTranscriptChromosome(idxs) = trnspts.nullTranscriptValue;
<a name="l00748"></a>00748                 
<a name="l00749"></a>00749                 nFrePols = nFrePols + nNsbPols;
<a name="l00750"></a>00750                 nBndPols = nBndPols - nNsbPols;
<a name="l00751"></a>00751                 
<a name="l00752"></a>00752                 %bind to new specifically bound site
<a name="l00753"></a>00753                 pBnds = this.computeRNAPolymeraseTUBindingProbabilities();
<a name="l00754"></a>00754                 
<a name="l00755"></a>00755                 if nnz(c.polymerizedRegions) == 2
<a name="l00756"></a>00756                     [~, iTU, posStrnds, nFreeNsbPols] = this.bindProteinToChromosome( ...
<a name="l00757"></a>00757                         [tu5Coords-tuDirs tuStrnds], ...
<a name="l00758"></a>00758                         this.enzymeIndexs_rnaPolymeraseHoloenzyme, nFreeNsbPols, pBnds(:, 1), ...
<a name="l00759"></a>00759                         true, true, 1, false, []);
<a name="l00760"></a>00760                     iChr = 1;
<a name="l00761"></a>00761                 else
<a name="l00762"></a>00762                     [~, iTU, posStrnds, nFreeNsbPols] = this.bindProteinToChromosome( ...
<a name="l00763"></a>00763                         [tu5Coords-tuDirs tuStrnds
<a name="l00764"></a>00764                         tu5Coords-tuDirs tuStrnds + 2], ...
<a name="l00765"></a>00765                         this.enzymeIndexs_rnaPolymeraseHoloenzyme, nFreeNsbPols, pBnds(:), ...
<a name="l00766"></a>00766                         true, true, 1, false, []);
<a name="l00767"></a>00767                     iTU = mod(iTU - 1, numel(tuLens)) + 1;
<a name="l00768"></a>00768                     iChr = ceil(posStrnds(:, 2) / 2);
<a name="l00769"></a>00769                 end
<a name="l00770"></a>00770                 
<a name="l00771"></a>00771                 idxs = freeNsbPols(1:nFreeNsbPols);
<a name="l00772"></a>00772                 rnaPols.states(idxs) = rnaPols.specificallyBoundValue;
<a name="l00773"></a>00773                 rnaPols.positionStrands(idxs, :) = posStrnds;
<a name="l00774"></a>00774                 trnspts.boundTranscriptionUnits(idxs) = iTU;
<a name="l00775"></a>00775                 trnspts.boundTranscriptProgress(idxs) = trnspts.nullTranscriptValue;
<a name="l00776"></a>00776                 trnspts.boundTranscriptChromosome(idxs) = iChr;
<a name="l00777"></a>00777                 
<a name="l00778"></a>00778                 nBndHols = nBndHols + nFreeNsbPols;
<a name="l00779"></a>00779                 nFrePols = nFrePols - nFreeNsbPols;
<a name="l00780"></a>00780                 nFreTfs(this.enzymeIndexs_sigmaFactor) = nFreTfs(this.enzymeIndexs_sigmaFactor) - nFreeNsbPols;
<a name="l00781"></a>00781             end
<a name="l00782"></a>00782             
<a name="l00783"></a>00783             %free polymerases/non-specifically bound becoming free/non-specifically bound
<a name="l00784"></a>00784             nsbPols = find(rnaPols.states == rnaPols.nonSpecificallyBoundValue);
<a name="l00785"></a>00785             if ~isempty(nsbPols)
<a name="l00786"></a>00786                 %set non-specifically bound to free
<a name="l00787"></a>00787                 releasedProteins = this.releaseProteinFromSites(rnaPols.positionStrands(nsbPols, :), false, this.enzymeIndexs_rnaPolymerase, true, true);
<a name="l00788"></a>00788                 if ~isequal(releasedProteins(this.enzymeIndexs_rnaPolymerase), numel(nsbPols))
<a name="l00789"></a>00789                     throw(MException(&#39;<a class="code" href="class_transcription.html">Transcription</a>:error&#39;, &#39;Unable to release protein&#39;));
<a name="l00790"></a>00790                 end
<a name="l00791"></a>00791                 
<a name="l00792"></a>00792                 rnaPols.states(nsbPols) = rnaPols.freeValue;
<a name="l00793"></a>00793                 rnaPols.positionStrands(nsbPols, :) = 0;
<a name="l00794"></a>00794                 trnspts.boundTranscriptionUnits(nsbPols) = trnspts.nullTranscriptValue;
<a name="l00795"></a>00795                 trnspts.boundTranscriptProgress(nsbPols) = trnspts.nullTranscriptValue;
<a name="l00796"></a>00796                 trnspts.boundTranscriptChromosome(nsbPols) = trnspts.nullTranscriptValue;
<a name="l00797"></a>00797                 
<a name="l00798"></a>00798                 nFrePols = nFrePols + numel(nsbPols);
<a name="l00799"></a>00799                 nBndPols = nBndPols - numel(nsbPols);
<a name="l00800"></a>00800             end
<a name="l00801"></a>00801             
<a name="l00802"></a>00802             freePols = find(rnaPols.states == rnaPols.freeValue);
<a name="l00803"></a>00803             nNsbPols = this.randStream.stochasticRound(min([...
<a name="l00804"></a>00804                 nFrePols
<a name="l00805"></a>00805                 numel(freePols) * pStTrns(rnaPols.nonSpecificallyBoundIndex, rnaPols.freeIndex) / (pStTrns(rnaPols.nonSpecificallyBoundIndex, rnaPols.freeIndex) + pStTrns(rnaPols.freeIndex, rnaPols.freeIndex))]));
<a name="l00806"></a>00806             if nNsbPols &gt; 0
<a name="l00807"></a>00807                 %set some to non-specifically bound state
<a name="l00808"></a>00808                 [nNsbPols posStrnds] = this.bindProteinToChromosomeStochastically(this.enzymeIndexs_rnaPolymerase, nNsbPols);
<a name="l00809"></a>00809                 posStrnds( isodd(posStrnds(:, 2)), 1) = posStrnds( isodd(posStrnds(:, 2)), 1) + this.enzymeDNAFootprints5Prime(this.enzymeIndexs_rnaPolymerase);
<a name="l00810"></a>00810                 posStrnds(iseven(posStrnds(:, 2)), 1) = posStrnds(iseven(posStrnds(:, 2)), 1) + this.enzymeDNAFootprints3Prime(this.enzymeIndexs_rnaPolymerase);
<a name="l00811"></a>00811                 
<a name="l00812"></a>00812                 rnaPols.states(freePols(1:nNsbPols)) = rnaPols.nonSpecificallyBoundValue;
<a name="l00813"></a>00813                 rnaPols.positionStrands(freePols(1:nNsbPols), :) = posStrnds;
<a name="l00814"></a>00814                 trnspts.boundTranscriptionUnits(freePols(1:nNsbPols)) = trnspts.nullTranscriptValue;
<a name="l00815"></a>00815                 trnspts.boundTranscriptProgress(freePols(1:nNsbPols)) = trnspts.nullTranscriptValue;
<a name="l00816"></a>00816                 trnspts.boundTranscriptChromosome(freePols(1:nNsbPols)) = trnspts.nullTranscriptValue;
<a name="l00817"></a>00817                 
<a name="l00818"></a>00818                 nFrePols = nFrePols - nNsbPols;
<a name="l00819"></a>00819                 nBndPols = nBndPols + nNsbPols;
<a name="l00820"></a>00820             end
<a name="l00821"></a>00821             
<a name="l00822"></a>00822             %set newly created <a class="code" href="class_r_n_a.html">RNA</a> polymerases to free state
<a name="l00823"></a>00823             if nNewPols &gt; 0
<a name="l00824"></a>00824                 newPols = find(rnaPols.states == rnaPols.notExistValue, nNewPols, &#39;first&#39;);
<a name="l00825"></a>00825                 rnaPols.states(newPols) = rnaPols.freeValue;
<a name="l00826"></a>00826                 rnaPols.positionStrands(newPols, :) = 0;
<a name="l00827"></a>00827                 trnspts.boundTranscriptionUnits(newPols) = trnspts.nullTranscriptValue;
<a name="l00828"></a>00828                 trnspts.boundTranscriptProgress(newPols) = trnspts.nullTranscriptValue;
<a name="l00829"></a>00829                 trnspts.boundTranscriptChromosome(newPols) = trnspts.nullTranscriptValue;
<a name="l00830"></a>00830             end
<a name="l00831"></a>00831             
<a name="l00832"></a>00832             %% Transcribe active sequences
<a name="l00833"></a>00833             %compute progress of transcription and cost
<a name="l00834"></a>00834             usedNTPs = zeros(4, 1);
<a name="l00835"></a>00835             if ~isempty(actPols) &amp;&amp; nFreTfs(this.enzymeIndexs_elongationFactor) &amp;&amp; this.rnaPolymeraseElongationRate &gt; 0
<a name="l00836"></a>00836                 posStrnds = rnaPols.positionStrands(actPols, :);
<a name="l00837"></a>00837                 iTUs = trnspts.boundTranscriptionUnits(actPols);
<a name="l00838"></a>00838                 
<a name="l00839"></a>00839                 %temporarily release <a class="code" href="class_r_n_a.html">RNA</a> polymerase from old position strands
<a name="l00840"></a>00840                 releasedProteins = ...
<a name="l00841"></a>00841                     this.releaseProteinFromSites(posStrnds, false, this.enzymeIndexs_rnaPolymerase, true, true) + ...
<a name="l00842"></a>00842                     this.releaseProteinFromSites(posStrnds, false, this.enzymeIndexs_rnaPolymeraseHoloenzyme, true, true);
<a name="l00843"></a>00843                 if ~isequal(releasedProteins(this.enzymeIndexs_rnaPolymerase) + releasedProteins(this.enzymeIndexs_rnaPolymeraseHoloenzyme), numel(actPols))
<a name="l00844"></a>00844                     throw(MException(&#39;<a class="code" href="class_transcription.html">Transcription</a>:error&#39;, &#39;Unable to unbind protein&#39;));
<a name="l00845"></a>00845                 end
<a name="l00846"></a>00846                 
<a name="l00847"></a>00847                 %extract active sequences (part of sequence that should be transcribed)
<a name="l00848"></a>00848                 actSeqs = cell(size(actPols)); %active sequences (those can be transcribed)
<a name="l00849"></a>00849                 for idx = 1:numel(actPols)
<a name="l00850"></a>00850                     i = actPols(idx);
<a name="l00851"></a>00851                     left = rnaPols.states(i);
<a name="l00852"></a>00852                     actSeqs{idx} = tuSeqs{trnspts.boundTranscriptionUnits(i)}(left:end);
<a name="l00853"></a>00853                 end
<a name="l00854"></a>00854                 actSeqs = char(actSeqs);
<a name="l00855"></a>00855                 actSeqs = actSeqs(:, 1:min(end, this.rnaPolymeraseElongationRate));
<a name="l00856"></a>00856                 
<a name="l00857"></a>00857                 %elongation limits by <a class="code" href="class_r_n_a.html">RNA</a> polymerase kinetic rate and transcription <a class="code" href="unit_8m.html#a94c091bed34da4edc337e42e23029000">unit</a> extent
<a name="l00858"></a>00858                 elngMax = min(this.rnaPolymeraseElongationRate, ...
<a name="l00859"></a>00859                     tuLens(trnspts.boundTranscriptionUnits(actPols)) - rnaPols.states(actPols) + 1);
<a name="l00860"></a>00860                                 
<a name="l00861"></a>00861                 %elongation limits by <a class="code" href="class_r_n_a.html">RNA</a>-polymerase / <a class="code" href="class_r_n_a.html">RNA</a>-polymerase interactions
<a name="l00862"></a>00862                 for strnd = 1 : 4
<a name="l00863"></a>00863                     idxs = find(posStrnds(:, 2) == strnd);
<a name="l00864"></a>00864                     <span class="keywordflow">if</span> numel(idxs) &lt;= 1
<a name="l00865"></a>00865                         <span class="keywordflow">continue</span>;
<a name="l00866"></a>00866                     end
<a name="l00867"></a>00867                     [~, order] = sort(posStrnds(idxs, 1));
<a name="l00868"></a>00868                     
<a name="l00869"></a>00869                     <span class="keywordflow">if</span> isodd(strnd)
<a name="l00870"></a>00870                         elngMax(idxs(order)) = min(...
<a name="l00871"></a>00871                             elngMax(idxs(order)), ...
<a name="l00872"></a>00872                             diff([posStrnds(idxs(order), 1); posStrnds(idxs(order(1)), 1) + c.sequenceLen]) - ...
<a name="l00873"></a>00873                             this.enzymeDNAFootprints(this.enzymeIndexs_rnaPolymerase));
<a name="l00874"></a>00874                     else
<a name="l00875"></a>00875                         elngMax(idxs(order)) = min(...
<a name="l00876"></a>00876                             elngMax(idxs(order)), ...
<a name="l00877"></a>00877                             diff([posStrnds(idxs(order(end)), 1) - c.sequenceLen; posStrnds(idxs(order), 1)]) - ...
<a name="l00878"></a>00878                             this.enzymeDNAFootprints(this.enzymeIndexs_rnaPolymerase));
<a name="l00879"></a>00879                     end
<a name="l00880"></a>00880                 end
<a name="l00881"></a>00881                 
<a name="l00882"></a>00882                 %elongation limits by <a class="code" href="class_r_n_a.html">RNA</a>-polymerase DNA interactions (<a class="code" href="class_note.html">Note</a>: independent of other <a class="code" href="class_r_n_a.html">RNA</a> polymerases)
<a name="l00883"></a>00883                 [~, ~, ~, tmp_elngMax] = c.isRegionAccessible(...
<a name="l00884"></a>00884                     posStrnds, (elngMax + 1) .* tuDirs(iTUs), ...
<a name="l00885"></a>00885                     [], this.enzymeGlobalIndexs(this.enzymeIndexs_rnaPolymerase), true, [], true, false);
<a name="l00886"></a>00886                 if ~all(tmp_elngMax)
<a name="l00887"></a>00887                     throw(MException(&#39;<a class="code" href="class_transcription.html">Transcription</a>:error&#39;, &#39;Unable to bind proteins&#39;));
<a name="l00888"></a>00888                 end
<a name="l00889"></a>00889                 elngMax = min(abs(tmp_elngMax) - 1, elngMax);
<a name="l00890"></a>00890                 
<a name="l00891"></a>00891                 %elongation limits by NTPs
<a name="l00892"></a>00892                 for i = 1:size(actSeqs, 1)
<a name="l00893"></a>00893                     actSeqs(i, elngMax(i) + 1:end) = &#39; &#39;;
<a name="l00894"></a>00894                 end
<a name="l00895"></a>00895                 [elngProg, this.substrates(this.substrateIndexs_ntp), usedNTPs] = ...
<a name="l00896"></a>00896                     edu.stanford.covert.cell.sim.util.polymerize(...
<a name="l00897"></a>00897                     actSeqs, this.substrates(this.substrateIndexs_ntp), &#39;ACGU&#39;, &#39; &#39;, 0, 0, this.randStream);
<a name="l00898"></a>00898                 
<a name="l00899"></a>00899                 %release sigma factor after initiation
<a name="l00900"></a>00900                 nInits = sum(...
<a name="l00901"></a>00901                     rnaPols.states(actPols) == rnaPols.activelyTranscribingValue &amp; ...
<a name="l00902"></a>00902                     elngProg &gt; 0);
<a name="l00903"></a>00903                 nBndHols = nBndHols - nInits;
<a name="l00904"></a>00904                 nFreTfs(this.enzymeIndexs_sigmaFactor) = nFreTfs(this.enzymeIndexs_sigmaFactor) + nInits;
<a name="l00905"></a>00905                 nBndPols = nBndPols + nInits;
<a name="l00906"></a>00906                 
<a name="l00907"></a>00907                 %rebind <a class="code" href="class_r_n_a.html">RNA</a> polymerase to new positions
<a name="l00908"></a>00908                 rnaPols.states(actPols) = rnaPols.states(actPols) + elngProg;
<a name="l00909"></a>00909                 rnaPols.positionStrands(actPols, 1) = posStrnds(:, 1) + elngProg .* tuDirs(iTUs);
<a name="l00910"></a>00910                 trnspts.boundTranscriptProgress(actPols) = rnaPols.states(actPols);
<a name="l00911"></a>00911                 
<a name="l00912"></a>00912                 initPols = rnaPols.states(actPols) == rnaPols.activelyTranscribingValue;
<a name="l00913"></a>00913                 if ...
<a name="l00914"></a>00914                         ~all(this.bindProteinToChromosome(rnaPols.positionStrands(actPols(~initPols), :), this.enzymeIndexs_rnaPolymerase)) || ...
<a name="l00915"></a>00915                         ~all(this.bindProteinToChromosome(rnaPols.positionStrands(actPols( initPols), :), this.enzymeIndexs_rnaPolymeraseHoloenzyme))
<a name="l00916"></a>00916                     throw(MException(&#39;<a class="code" href="class_transcription.html">Transcription</a>:error&#39;, &#39;Unable to bind proteins&#39;));
<a name="l00917"></a>00917                 end
<a name="l00918"></a>00918             end
<a name="l00919"></a>00919             
<a name="l00920"></a>00920             %release polymerases which have completed transcription, and increment
<a name="l00921"></a>00921             %expression of the corresponding transcription <a class="code" href="unit_8m.html#a94c091bed34da4edc337e42e23029000">unit</a>
<a name="l00922"></a>00922             if ~isempty(actPols) &amp;&amp; nFreTfs(this.enzymeIndexs_elongationFactor) &amp;&amp; nFreTfs(this.enzymeIndexs_terminationFactor)
<a name="l00923"></a>00923                 trmPols = actPols(rnaPols.states(actPols) &gt; ...
<a name="l00924"></a>00924                     tuLens(trnspts.boundTranscriptionUnits(actPols)));
<a name="l00925"></a>00925                 nTrmPols = numel(trmPols);
<a name="l00926"></a>00926                 if nTrmPols &gt; this.substrates(this.substrateIndexs_water)
<a name="l00927"></a>00927                    order = this.randStream.randperm(nTrmPols);
<a name="l00928"></a>00928                    nTrmPols = this.substrates(this.substrateIndexs_water);
<a name="l00929"></a>00929                    trmPols = trmPols(order(1:nTrmPols));
<a name="l00930"></a>00930                 end
<a name="l00931"></a>00931                 
<a name="l00932"></a>00932                 if nTrmPols &gt; 0
<a name="l00933"></a>00933                     if isscalar(trmPols)
<a name="l00934"></a>00934                         this.RNAs(trnspts.boundTranscriptionUnits(trmPols)) = ...
<a name="l00935"></a>00935                             this.RNAs(trnspts.boundTranscriptionUnits(trmPols)) + 1;
<a name="l00936"></a>00936                     else
<a name="l00937"></a>00937                         this.RNAs = this.RNAs + ...
<a name="l00938"></a>00938                             histc(trnspts.boundTranscriptionUnits(trmPols, 1), 1:numel(tuLens));
<a name="l00939"></a>00939                     end
<a name="l00940"></a>00940                     
<a name="l00941"></a>00941                     %release proteins from chromsome
<a name="l00942"></a>00942                     releasedProteins = this.releaseProteinFromSites(rnaPols.positionStrands(trmPols, :), false, this.enzymeIndexs_rnaPolymerase, true, true);
<a name="l00943"></a>00943                     if ~isequal(releasedProteins(this.enzymeIndexs_rnaPolymerase), numel(trmPols))
<a name="l00944"></a>00944                         throw(MException(&#39;<a class="code" href="class_transcription.html">Transcription</a>:error&#39;, &#39;Unable to unbind protein&#39;));
<a name="l00945"></a>00945                     end
<a name="l00946"></a>00946                     
<a name="l00947"></a>00947                     nFrePols = nFrePols + nTrmPols;
<a name="l00948"></a>00948                     nBndPols = nBndPols - nTrmPols;
<a name="l00949"></a>00949                     
<a name="l00950"></a>00950                     this.substrates(this.substrateIndexs_water)    = this.substrates(this.substrateIndexs_water)    - nTrmPols;
<a name="l00951"></a>00951                     this.substrates(this.substrateIndexs_hydrogen) = this.substrates(this.substrateIndexs_hydrogen) + nTrmPols;
<a name="l00952"></a>00952                     
<a name="l00953"></a>00953                     rnaPols.states(trmPols) = rnaPols.freeValue;
<a name="l00954"></a>00954                     rnaPols.positionStrands(trmPols, :) = 0;
<a name="l00955"></a>00955                     trnspts.boundTranscriptionUnits(trmPols) = trnspts.nullTranscriptValue;
<a name="l00956"></a>00956                     trnspts.boundTranscriptProgress(trmPols) = trnspts.nullTranscriptValue;
<a name="l00957"></a>00957                     trnspts.boundTranscriptChromosome(trmPols) = trnspts.nullTranscriptValue;
<a name="l00958"></a>00958                 end
<a name="l00959"></a>00959             end
<a name="l00960"></a>00960             
<a name="l00961"></a>00961             %% diphosphate released by polymerization
<a name="l00962"></a>00962             this.substrates(this.substrateIndexs_diphosphate) = ...
<a name="l00963"></a>00963                 this.substrates(this.substrateIndexs_diphosphate) + ...
<a name="l00964"></a>00964                 sum(usedNTPs);
<a name="l00965"></a>00965             
<a name="l00966"></a>00966             %% store enzyme state
<a name="l00967"></a>00967             this.enzymes(this.enzymeIndexs_transcriptionFactors) = nFreTfs;
<a name="l00968"></a>00968             this.boundEnzymes(this.enzymeIndexs_transcriptionFactors) = nBndTFs;
<a name="l00969"></a>00969             this.enzymes(this.enzymeIndexs_rnaPolymerase) = nFrePols;
<a name="l00970"></a>00970             this.boundEnzymes(this.enzymeIndexs_rnaPolymerase) = nBndPols;
<a name="l00971"></a>00971             this.enzymes(this.enzymeIndexs_rnaPolymeraseHoloenzyme) = nFreHols;
<a name="l00972"></a>00972             this.boundEnzymes(this.enzymeIndexs_rnaPolymeraseHoloenzyme) = nBndHols;
<a name="l00973"></a>00973         end
<a name="l00974"></a>00974     end
<a name="l00975"></a>00975     
<a name="l00976"></a>00976     %model helper functions
<a name="l00977"></a>00977     methods    
<a name="l00978"></a>00978         function pBinds = computeRNAPolymeraseTUBindingProbabilities(this, protectDnaABoxes)
<a name="l00979"></a>00979             c = this.chromosome;
<a name="l00980"></a>00980             r = this.rnaPolymerases;
<a name="l00981"></a>00981             
<a name="l00982"></a>00982             %relative probability <a class="code" href="class_r_n_a.html">RNA</a> polymerase binds each transcription
<a name="l00983"></a>00983             %<a class="code" href="unit_8m.html#a94c091bed34da4edc337e42e23029000">unit</a>
<a name="l00984"></a>00984             pBinds = this.transcriptionUnitBindingProbabilities(:, ones(1, 2)) .* ...
<a name="l00985"></a>00985                 r.transcriptionFactorBindingProbFoldChange .* ...
<a name="l00986"></a>00986                 r.supercoilingBindingProbFoldChange;
<a name="l00987"></a>00987             
<a name="l00988"></a>00988             %if functional DnaA boxes R1-5 are occupied by DnaA-ATP,
<a name="l00989"></a>00989             %don&#39;t allow <a class="code" href="class_r_n_a.html">RNA</a> polymerase to bind since it would knockoff
<a name="l00990"></a>00990             %DnaA-ATP
<a name="l00991"></a>00991             if nargin == 1 || protectDnaABoxes
<a name="l00992"></a>00992                 [posStrnds, complexs] = find(...
<a name="l00993"></a>00993                     c.complexBoundSites(c.transcriptionUnitStartCoordinates(this.transcriptionUnitIndexs_DnaAR12345Boxes) + ...
<a name="l00994"></a>00994                     (1:c.transcriptionUnitLengths(this.transcriptionUnitIndexs_DnaAR12345Boxes))&#39; - 1, :));
<a name="l00995"></a>00995                 if any(ismembc(complexs, this.complexIndexs_DnaA_ATP) &amp; posStrnds(:, 2) &lt;= 2)
<a name="l00996"></a>00996                     pBinds(this.transcriptionUnitIndexs_DnaAR12345Boxes, 1) = 0;
<a name="l00997"></a>00997                 end
<a name="l00998"></a>00998                 if any(ismembc(complexs, this.complexIndexs_DnaA_ATP) &amp; posStrnds(:, 2) &gt; 2)
<a name="l00999"></a>00999                     pBinds(this.transcriptionUnitIndexs_DnaAR12345Boxes, 2) = 0;
<a name="l01000"></a>01000                 end
<a name="l01001"></a>01001             end
<a name="l01002"></a>01002         end
<a name="l01003"></a>01003         
<a name="l01004"></a>01004         function pStProbs = calcStateTransitionProbabilities(this, nPols, ntpProdRate, tuBindingProbs, method)
<a name="l01005"></a>01005             import edu.stanford.covert.util.<a class="code" href="class_computation_util.html">ComputationUtil</a>;
<a name="l01006"></a>01006             import edu.stanford.covert.util.<a class="code" href="class_constant_util.html">ConstantUtil</a>;
<a name="l01007"></a>01007             
<a name="l01008"></a>01008             rna = this.rna;
<a name="l01009"></a>01009             rnaPols = this.rnaPolymerases;
<a name="l01010"></a>01010             
<a name="l01011"></a>01011             pStProbs = zeros(4, 4);
<a name="l01012"></a>01012             pStProbs(rnaPols.nonSpecificallyBoundIndex, rnaPols.activelyTranscribingIndex) = ...
<a name="l01013"></a>01013                 ntpProdRate / (nPols * rnaPols.stateExpectations(rnaPols.activelyTranscribingIndex)) / ...
<a name="l01014"></a>01014                 (tuBindingProbs&#39; * rna.lengths(rna.nascentIndexs));
<a name="l01015"></a>01015             if nargin &lt; 5 || isequal(method, &#39;handTuned&#39;)
<a name="l01016"></a>01016                 pStProbs(rnaPols.activelyTranscribingIndex, rnaPols.specificallyBoundIndex) = 1;
<a name="l01017"></a>01017                 pStProbs(rnaPols.nonSpecificallyBoundIndex, rnaPols.specificallyBoundIndex) = 0;
<a name="l01018"></a>01018                 pStProbs(rnaPols.specificallyBoundIndex, rnaPols.nonSpecificallyBoundIndex) = 0.10;
<a name="l01019"></a>01019             else
<a name="l01020"></a>01020                 pStProbs(rnaPols.activelyTranscribingIndex, rnaPols.specificallyBoundIndex) = ...
<a name="l01021"></a>01021                     pStProbs(rnaPols.nonSpecificallyBoundIndex, rnaPols.activelyTranscribingIndex) * ...
<a name="l01022"></a>01022                     rnaPols.stateExpectations(rnaPols.activelyTranscribingIndex) / ...
<a name="l01023"></a>01023                     rnaPols.stateExpectations(rnaPols.specificallyBoundIndex);
<a name="l01024"></a>01024                 pStProbs(rnaPols.specificallyBoundIndex, rnaPols.nonSpecificallyBoundIndex) = ...
<a name="l01025"></a>01025                     pStProbs(rnaPols.nonSpecificallyBoundIndex, rnaPols.activelyTranscribingIndex) * ...
<a name="l01026"></a>01026                     rnaPols.stateExpectations(rnaPols.activelyTranscribingIndex) / ...
<a name="l01027"></a>01027                     rnaPols.stateExpectations(rnaPols.nonSpecificallyBoundIndex);
<a name="l01028"></a>01028             end
<a name="l01029"></a>01029             
<a name="l01030"></a>01030             pStProbs(rnaPols.activelyTranscribingIndex, rnaPols.activelyTranscribingIndex) = ...
<a name="l01031"></a>01031                 1 - pStProbs(rnaPols.nonSpecificallyBoundIndex, rnaPols.activelyTranscribingIndex);
<a name="l01032"></a>01032             pStProbs(rnaPols.specificallyBoundIndex, rnaPols.specificallyBoundIndex) = ...
<a name="l01033"></a>01033                 + 1 ...
<a name="l01034"></a>01034                 - pStProbs(rnaPols.activelyTranscribingIndex, rnaPols.specificallyBoundIndex)...
<a name="l01035"></a>01035                 - pStProbs(rnaPols.nonSpecificallyBoundIndex, rnaPols.specificallyBoundIndex);
<a name="l01036"></a>01036             pStProbs(rnaPols.nonSpecificallyBoundIndex, rnaPols.nonSpecificallyBoundIndex) = ...
<a name="l01037"></a>01037                 1 - pStProbs(rnaPols.specificallyBoundIndex, rnaPols.nonSpecificallyBoundIndex);
<a name="l01038"></a>01038             pStProbs(:, rnaPols.freeIndex) = pStProbs(:, rnaPols.nonSpecificallyBoundIndex);
<a name="l01039"></a>01039         end
<a name="l01040"></a>01040     end
<a name="l01041"></a>01041     
<a name="l01042"></a>01042     %get methods of dependent local state
<a name="l01043"></a>01043     methods
<a name="l01044"></a>01044         function value = getDryWeight(this)
<a name="l01045"></a>01045             if size(this.RNAs, 3) == 1
<a name="l01046"></a>01046                 value = this.getDryWeight@edu.stanford.covert.cell.sim.<a class="code" href="class_process.html">Process</a>() + ...
<a name="l01047"></a>01047                     this.rna.molecularWeights(this.rna.nascentIndexs)&#39; * this.RNAs ...
<a name="l01048"></a>01048                     / edu.stanford.covert.util.ConstantUtil.nAvogadro;
<a name="l01049"></a>01049             else
<a name="l01050"></a>01050                 value = this.getDryWeight@edu.stanford.covert.cell.sim.<a class="code" href="class_process.html">Process</a>() + ...
<a name="l01051"></a>01051                     permute(this.rna.molecularWeights(this.rna.nascentIndexs)&#39; * permute(this.RNAs,[1 3 2]),[1 3 2]) ...
<a name="l01052"></a>01052                     / edu.stanford.covert.util.ConstantUtil.nAvogadro;
<a name="l01053"></a>01053             end
<a name="l01054"></a>01054         end
<a name="l01055"></a>01055     end
<a name="l01056"></a>01056 end
</pre></div></div><!-- contents -->
</div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
<a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(0)"><span class="SelectionMark">&#160;</span>All</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(1)"><span class="SelectionMark">&#160;</span>Classes</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(2)"><span class="SelectionMark">&#160;</span>Files</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(3)"><span class="SelectionMark">&#160;</span>Functions</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(4)"><span class="SelectionMark">&#160;</span>Variables</a></div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

  <div id="nav-path" class="navpath">
    <ul>
      <li class="navelem"><a class="el" href="_09process_2_transcription_8m.html">Transcription.m</a>      </li>

    <li class="footer">Generated on Wed Feb 8 2012 21:42:32 for Whole Cell Simulation by
    <a href="http://www.doxygen.org/index.html">
    <img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.7.6.1 </li>
   </ul>
 </div>


</body>
</html>
