<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<title>Whole Cell Simulation: src/+edu/+stanford/+covert/+cell/+sim/+process/Replication.m Source File</title>

<link href="tabs.css" rel="stylesheet" type="text/css"/>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
<link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="navtree.js"></script>
<script type="text/javascript">
  $(document).ready(initResizable);
</script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
  $(document).ready(function() { searchBox.OnSelectItem(0); });
</script>

</head>
<body>
<div id="top"><!-- do not remove this div! -->


<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  
  
  <td style="padding-left: 0.5em;">
   <div id="projectname">Whole Cell Simulation
   
   </div>
   
  </td>
  
  
  
 </tr>
 </tbody>
</table>
</div>

<!-- Generated by Doxygen 1.7.6.1 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "search",false,'Search');
</script>
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&#160;Page</span></a></li>
      <li><a href="annotated.html"><span>Classes</span></a></li>
      <li class="current"><a href="files.html"><span>Files</span></a></li>
      <li>
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
  <div id="navrow2" class="tabs2">
    <ul class="tablist">
      <li><a href="files.html"><span>File&#160;List</span></a></li>
      <li><a href="globals.html"><span>File&#160;Members</span></a></li>
    </ul>
  </div>
</div>
<div id="side-nav" class="ui-resizable side-nav-resizable">
  <div id="nav-tree">
    <div id="nav-tree-contents">
    </div>
  </div>
  <div id="splitbar" style="-moz-user-select:none;" 
       class="ui-resizable-handle">
  </div>
</div>
<script type="text/javascript">
  initNavTree('_09process_2_replication_8m.html','');
</script>
<div id="doc-content">
<div class="header">
  <div class="headertitle">
<div class="title">src/+edu/+stanford/+covert/+cell/+sim/+process/Replication.m</div>  </div>
</div><!--header-->
<div class="contents">
<a href="_09process_2_replication_8m.html">Go to the documentation of this file.</a><div class="fragment"><pre class="fragment"><a name="l00001"></a>00001 %<a class="code" href="class_replication.html">Replication</a>
<a name="l00002"></a>00002 %
<a name="l00003"></a>00003 % @wholeCellModelID Process_Replication
<a name="l00004"></a>00004 % @name             <a class="code" href="class_replication.html">Replication</a>
<a name="l00005"></a>00005 % @description
<a name="l00006"></a>00006 %   Biology
<a name="l00007"></a>00007 %   =================
<a name="l00008"></a>00008 %   DNA replication by replisomes is initiated by DnaA complex formation near
<a name="l00009"></a>00009 %   the oriC and proceeds bidirectionally from oriC to terC in the 5<span class="stringliteral">&#39;-&gt;3&#39;</span>
<a name="l00010"></a>00010 %   direction along leading strand as well as in Okazaki fragments in the 5<span class="stringliteral">&#39;-&gt;3&#39;</span>
<a name="l00011"></a>00011 %   direction along lagging strands. This process models the locations, kinetics,
<a name="l00012"></a>00012 %   and biochemistry of all the replication proteins:
<a name="l00013"></a>00013 %   - replicative DNA helicase
<a name="l00014"></a>00014 %   - DNA primase
<a name="l00015"></a>00015 %   - DNA polymerase core
<a name="l00016"></a>00016 %   - Beta-clamp loading complex / gamma complex
<a name="l00017"></a>00017 %   - Beta-clamp
<a name="l00018"></a>00018 %   - DNA ligase
<a name="l00019"></a>00019 %   - single stranded binding proteins
<a name="l00020"></a>00020 %
<a name="l00021"></a>00021 %   The exact mechanism replication initiation in M. genitalium is unknown.
<a name="l00022"></a>00022 %   Furthermore, because M. genitalium does not contain a DnaC homolog it is
<a name="l00023"></a>00023 %   difficult to infer M. genitalium repliciation initiation from studies of
<a name="l00024"></a>00024 %   other bacterial species.
<a name="l00025"></a>00025 %
<a name="l00026"></a>00026 %   Knowledge Base
<a name="l00027"></a>00027 %   =================
<a name="l00028"></a>00028 %   The knowledge base contains data curated from the literature and databases:
<a name="l00029"></a>00029 %   - the chromosome DNA sequence
<a name="l00030"></a>00030 %   - the DNA footprints of all replication proteins
<a name="l00031"></a>00031 %   - DNA binding protein displacement reactions (that is which proteins can
<a name="l00032"></a>00032 %     displace which other proteins from the chromosome)
<a name="l00033"></a>00033 %   - values of various structural and kinetic parameters
<a name="l00034"></a>00034 %   - subunit composition of the replication proteins
<a name="l00035"></a>00035 %
<a name="l00036"></a>00036 %   Representation
<a name="l00037"></a>00037 %   =================
<a name="l00038"></a>00038 %   The properties substrates, enzymes, and boundEnzymes represent the counts of
<a name="l00039"></a>00039 %   free metabolites, free replication proteins, and chromosally-bound
<a name="l00040"></a>00040 %   replication proteins. The chromosomes property polymerizedRegions represents
<a name="l00041"></a>00041 %   the regions of the chromosomes which have been polymerized (and the
<a name="l00042"></a>00042 %   base-pairing of strands). The chromosomes property strandBreaks represents
<a name="l00043"></a>00043 %   strand breaks 5&#39; to each base. The chromosomes property complexBoundSites
<a name="l00044"></a>00044 %   represents represents the specific chromosomal location of all chromosomally
<a name="l00045"></a>00045 %   bound proteins.
<a name="l00046"></a>00046 %
<a name="l00047"></a>00047 %   Spatial Model
<a name="l00048"></a>00048 %   +++++++++++++++++
<a name="l00049"></a>00049 %   - Helicase is centered on the boundary between ssDNA and dsDNA. The position
<a name="l00050"></a>00050 %     over which it is centered is the next position to be melted.
<a name="l00051"></a>00051 %   - Polymerase core is centered on the boundary between ssDNA and dsDNA. The
<a name="l00052"></a>00052 %     position over which it is centered is the next position to be polymerized.
<a name="l00053"></a>00053 %   - There is no gap between the helicase and polymerase core or between the
<a name="l00054"></a>00054 %     polymerase core and the beta clamp.
<a name="l00055"></a>00055 %   - Backup beta clamps bind slightly upstream of the start site of Okazaki
<a name="l00056"></a>00056 %     fragments such that there will be no gap between the polymerase and beta
<a name="l00057"></a>00057 %     clamp, and the polymerase core will be centered on the Okazaki fragment
<a name="l00058"></a>00058 %     start site
<a name="l00059"></a>00059 %   - At replication initiation a the mother strands are separated such that the
<a name="l00060"></a>00060 %     leading polymerase cores are centered at oriC+-1 and the helicases are
<a name="l00061"></a>00061 %     slightly (11 nt) ahead
<a name="l00062"></a>00062 %   - During replication initiation (and the final step of replication after the
<a name="l00063"></a>00063 %     last Okazaki fragment has completed) the lagging polymerase and primase
<a name="l00064"></a>00064 %     are accounted for as part of a complex on the leading strand (containing
<a name="l00065"></a>00065 %     also the helicase, leading polymerase, gamma complex, and leading beta
<a name="l00066"></a>00066 %     clamp). At all other times, the lagging polymerase, lagging beta clamp,
<a name="l00067"></a>00067 %     and primase are accounted for as a complex on a different strand. This
<a name="l00068"></a>00068 %     allows us to separately keep track of the leading and lagging polymerase
<a name="l00069"></a>00069 %     positions.
<a name="l00070"></a>00070 %   - Backup beta clamps can&#39;t binding until half of the previous Okazaki
<a name="l00071"></a>00071 %     fragment has been polymerized
<a name="l00072"></a>00072 %   - In general there will be approximately 1 Okazaki fragment length gap
<a name="l00073"></a>00073 %     between the progess of DNA polymerase on the leading and lagging strands
<a name="l00074"></a>00074 %
<a name="l00075"></a>00075 %   Initialization
<a name="l00076"></a>00076 %   =================
<a name="l00077"></a>00077 %   The mother chromosome is initialized by the Chromosome class and
<a name="l00078"></a>00078 %   several processes:
<a name="l00079"></a>00079 %   - completely synthesized
<a name="l00080"></a>00080 %   - mother strands base pairing
<a name="l00081"></a>00081 %   - undamged, except methylated at restriction/modification sites (DNA repair
<a name="l00082"></a>00082 %     process)
<a name="l00083"></a>00083 %   - bound by various proteins
<a name="l00084"></a>00084 %     - SMCs (DNA condensation)
<a name="l00085"></a>00085 %     - RNA polymerase (<a class="code" href="class_transcription.html">Transcription</a>)
<a name="l00086"></a>00086 %     - Transcription factors (Transcriptional regulation)
<a name="l00087"></a>00087 %     - Topoisomerases (supercoiling)
<a name="l00088"></a>00088 %     - DnaA (replication initiation)
<a name="l00089"></a>00089 %
<a name="l00090"></a>00090 %   Simulation
<a name="l00091"></a>00091 %   =================
<a name="l00092"></a>00092 %   The simulation consists of 8 subfunctions executed in a random order:
<a name="l00093"></a>00093 %   - Initiate replication (initiateReplication)
<a name="l00094"></a>00094 %     If DnaA complex assembled at OriC and sufficient protein and metabolites,
<a name="l00095"></a>00095 %     unwind small segment of DNA and binding helicase, primase, polymerase,
<a name="l00096"></a>00096 %     gamma complex, and beta clamp to chromosomes. Associate all proteins with
<a name="l00097"></a>00097 %     the leading strand. Chromosome will take care of dissassembly of the
<a name="l00098"></a>00098 %     DnaA complex. ReplicationInitiation will take care of dissociating the
<a name="l00099"></a>00099 %     released DnaA-ATP polymers.
<a name="l00100"></a>00100 %
<a name="l00101"></a>00101 %   - Advance replisomes: unwind and polymerize DNA, release SSBs (unwindAndPolymerizeDNA)
<a name="l00102"></a>00102 %     1. If first Okazaki fragment starting, associate primase, lagging polymerase
<a name="l00103"></a>00103 %        with lagging strand.
<a name="l00104"></a>00104 %     2. Advance leading and lagging polymerases and helicases up to limits
<a name="l00105"></a>00105 %        1. Polymerase and primase kinetics
<a name="l00106"></a>00106 %        2. Available dNTP for polymerization and energy for unwinding
<a name="l00107"></a>00107 %        3. Prevent leading strand progressing if no SSBs bound to lagging
<a name="l00108"></a>00108 %           strand
<a name="l00109"></a>00109 %        4. Accessibility of upstream regions to helicase and polymerases
<a name="l00110"></a>00110 %        5. Don&#39;t polymerize leading strand past terC or lagging strands past
<a name="l00111"></a>00111 %           ends of Okazaki fragments
<a name="l00112"></a>00112 %        6. Don&#39;t allow the leading strand and helicase to go way beyond the
<a name="l00113"></a>00113 %           lagging strand progress.
<a name="l00114"></a>00114 %        7. Pause progress is RNA polymerase is encountered.
<a name="l00115"></a>00115 %           If the replication loop (helicase) hits an RNAP, polymerization
<a name="l00116"></a>00116 %           pauses, the RNAP falls off, and its transcript is degraded. If it&#39;s
<a name="l00117"></a>00117 %           a head-on collision, the replication loop will not proceed for some
<a name="l00118"></a>00118 %           unknown amount of time (a fittable parameter). If it&#39;s a codirec-
<a name="l00119"></a>00119 %           tional collision, polymerization will continue at full speed the
<a name="l00120"></a>00120 %           following time step. (Mirkin 2004, Mirkin 2006, Mirkin 2007).
<a name="l00121"></a>00121 %           If occupied by RNA-polymerase, then calculate &quot;occupied DNA&quot; by base
<a name="l00122"></a>00122 %           polymerase is currently on -9 through +2. (Neidhardt 1990).
<a name="l00123"></a>00123 %
<a name="l00124"></a>00124 %   - Bind SSBs (freeAndBindSSBs)
<a name="l00125"></a>00125 %     Bind free SSB 4mers to stochastically to deterministically selected
<a name="l00126"></a>00126 %     positions within single-stranded regions as SSB 8mers
<a name="l00127"></a>00127 %
<a name="l00128"></a>00128 %   - Dissociate free SSB 8mers into 2 SSB 4mers (dissociateFreeSSBComplexes)
<a name="l00129"></a>00129 %     Dissociate released SSB 8mers into 2 SSB 4mers
<a name="l00130"></a>00130 %
<a name="l00131"></a>00131 %   - Initiate Okazaki fragment by bind beta-clamp (initiateOkazakiFragment)
<a name="l00132"></a>00132 %     Binding beta-clamp just downstream of the start position of the next
<a name="l00133"></a>00133 %     Okazaki fragment. Requires that
<a name="l00134"></a>00134 %     - There is beta-clamp monomer to form new beta-clamp dimer on chromosome
<a name="l00135"></a>00135 %     - There is energy for the gamma-complex to form the new beta-clamp
<a name="l00136"></a>00136 %     - The position is accessible to the beta-clamp
<a name="l00137"></a>00137 %     - The leading helicase has already passed the position
<a name="l00138"></a>00138 %     - The lagging strand is at least 1/2 done with the current Okazaki
<a name="l00139"></a>00139 %       fragment
<a name="l00140"></a>00140 %
<a name="l00141"></a>00141 %   - Terminate Okazaki fragment by releasing beta-clamp (terminateOkazakiFragment)
<a name="l00142"></a>00142 %     Release beta-clamp and associate lagging primase and polymerase with
<a name="l00143"></a>00143 %     backup beta-clamp (or if terminating the last Okazaki fragment,
<a name="l00144"></a>00144 %     associate the lagging primase and with the leading strand machinery).
<a name="l00145"></a>00145 %     Mark end of Okazaki fragment as having a single strand break to be ligated
<a name="l00146"></a>00146 %     by ligase.
<a name="l00147"></a>00147 %     Occurs if:
<a name="l00148"></a>00148 %     - Okazaki fragment finished polymerizing
<a name="l00149"></a>00149 %     - SSBs bound to lagging strand
<a name="l00150"></a>00150 %     - Lagging backup beta-clamp bound
<a name="l00151"></a>00151 %     - Leading machinery has advanced beyond the Okazaki fragment start site
<a name="l00152"></a>00152 %
<a name="l00153"></a>00153 %   - Terminate replication (terminateReplication)
<a name="l00154"></a>00154 %     Release bound replisomes from leading strands. Mark terC as having single
<a name="l00155"></a>00155 %     strand breaks to be ligated by ligase. Occurs if:
<a name="l00156"></a>00156 %     - Polymerization completed for both leading and lagging strands
<a name="l00157"></a>00157 %     - Lagging strand ligated (except for last ligation at terC which will
<a name="l00158"></a>00158 %       occur after the execution of this subfunction)
<a name="l00159"></a>00159 %
<a name="l00160"></a>00160 %   - Ligate DNA (ligateDNA)
<a name="l00161"></a>00161 %     Stochastically ligate single strand breaks up to
<a name="l00162"></a>00162 %     - Ligase kinetics
<a name="l00163"></a>00163 %     - Ligase availability
<a name="l00164"></a>00164 %     - NAD availability
<a name="l00165"></a>00165 %
<a name="l00166"></a>00166 %   References
<a name="l00167"></a>00167 %   =================
<a name="l00168"></a>00168 %   1. Mirkin, E.V., Mirkin, S.M. (2007). Replication fork stalling at natural
<a name="l00169"></a>00169 %      impediments. Microbiology and molecular biology reviews 71: 13-35.
<a name="l00170"></a>00170 %   2. Mirkin, E.V., Mirkin, S.M. (2005). Mechanisms of
<a name="l00171"></a>00171 %      Transcription-replication collisions in bacteria. molecular and cellular
<a name="l00172"></a>00172 %      biology 25: 888-895.
<a name="l00173"></a>00173 %   3. Mirkin, E.V., Roa, D.C., Nudler, E., Mirkin, S.M. (2006). Transcription
<a name="l00174"></a>00174 %      regulatory elements are punctuation marks for dna replication. PNAS 103:
<a name="l00175"></a>00175 %      7276-7281.
<a name="l00176"></a>00176 %   4. Miyata, M. &quot;Cell Division&quot;. Molecular biology and pathogenicity of
<a name="l00177"></a>00177 %      mycoplasmas. Razin, S., Herrmann, R. Kluwer Academic/Plenum Publishers,
<a name="l00178"></a>00178 %      2002. 117-130.
<a name="l00179"></a>00179 %   5. Rudolph, C.J., Dhillon, P., Moore, T., Lloyd, R.G. (2007). Avoiding and
<a name="l00180"></a>00180 %      resolving conflicts between DNA replication and transcription. DNA Repair
<a name="l00181"></a>00181 %      6: 981-993.
<a name="l00182"></a>00182 %   6. McGlynn, P., Guy, C.P. (2008). Replication forks blocked by protein-DNA
<a name="l00183"></a>00183 %      complexes have limited stability in vitro. J. Mol. Biol. 381: 249-255.
<a name="l00184"></a>00184 %   7. Kozlov AG and Lohman TM. (2002). Kinetic Mechanism
<a name="l00185"></a>00185 %      of Direct Transfer of Escherichia coli SSB Tetramers between
<a name="l00186"></a>00186 %      Single-Stranded DNA Molecules. Biochemistry. 41 (39): 11611-11627.
<a name="l00187"></a>00187 %      [PUB_0856]
<a name="l00188"></a>00188 %   8. Kunzelmann S, Morris C, Chavda AP, Eccleston JF, Webb MR (2010).
<a name="l00189"></a>00189 %      Mechanism of Interaction between Single-Stranded DNA Binding Protein
<a name="l00190"></a>00190 %      and DNA. Biochemistry 49(5): 843-52. [PUB_0857]
<a name="l00191"></a>00191 %   9. Roy R, Kozlov AG, Lohman TM, Ha T (2009). SSB protein diffusion on
<a name="l00192"></a>00192 %      single-stranded DNA stimulates RecA filament formation. Nature.
<a name="l00193"></a>00193 %      461(7267):1092-7. [PUB_0858]
<a name="l00194"></a>00194 %
<a name="l00195"></a>00195 % Author: Jayodita Sanghvi, jayodita@stanford.edu
<a name="l00196"></a>00196 % Author: Jonathan Karr, jkarr@stanford.edu
<a name="l00197"></a>00197 % Author: Jared Jacobs, jmjacobs@stanford.edu
<a name="l00198"></a>00198 % Affilitation: Covert Lab, Department of Bioengineering, Stanford University
<a name="l00199"></a>00199 % Last updated: 8/19/2010
<a name="l00200"></a>00200 
<a name="l00201"></a>00201 %TODO: consider moving requirement for SSBs in okazaki fragment termination to
<a name="l00202"></a>00202 %      initiation of next fragment
<a name="l00203"></a>00203 classdef Replication &lt; edu.stanford.covert.cell.sim.Process &amp; edu.stanford.covert.cell.sim.ChromosomeProcessAspect
<a name="l00204"></a>00204     %property annotations
<a name="l00205"></a>00205     properties (Constant)
<a name="l00206"></a>00206         optionNames__              = {}; %names of option <a class="code" href="json_parse_8m.html#a6e4256f69b751d3d7ec4ea52ff5a8dee">properties</a>
<a name="l00207"></a>00207         fixedConstantNames__       = {   %names of fixed constant <a class="code" href="json_parse_8m.html#a6e4256f69b751d3d7ec4ea52ff5a8dee">properties</a>
<a name="l00208"></a>00208                     <span class="stringliteral">&#39;oriCPosition&#39;</span>
<a name="l00209"></a>00209             <span class="stringliteral">&#39;terCPosition&#39;</span>
<a name="l00210"></a>00210             <span class="stringliteral">&#39;enzymeComposition&#39;</span>
<a name="l00211"></a>00211             <span class="stringliteral">&#39;primaseBindingLocations&#39;</span>
<a name="l00212"></a>00212             <span class="stringliteral">&#39;primerLength&#39;</span>
<a name="l00213"></a>00213             <span class="stringliteral">&#39;dnaPolymeraseElongationRate&#39;</span>
<a name="l00214"></a>00214             <span class="stringliteral">&#39;ssbComplexSpacing&#39;</span>
<a name="l00215"></a>00215             <span class="stringliteral">&#39;okazakiFragmentMeanLength&#39;</span>
<a name="l00216"></a>00216             <span class="stringliteral">&#39;ligaseRate&#39;</span>
<a name="l00217"></a>00217             <span class="stringliteral">&#39;laggingBackupClampReloadingLength&#39;</span>
<a name="l00218"></a>00218             <span class="stringliteral">&#39;startingOkazakiLoopLength&#39;</span>
<a name="l00219"></a>00219             <span class="stringliteral">&#39;rnaPolymeraseCollisionMeanDwellTime&#39;</span>
<a name="l00220"></a>00220             <span class="stringliteral">&#39;ssbDissociationRate&#39;</span>
<a name="l00221"></a>00221             <span class="stringliteral">&#39;dnaAFunctionalBoxStartPositions&#39;</span>
<a name="l00222"></a>00222                     };
<a name="l00223"></a>00223         fittedConstantNames__      = {}; %names of fitted constant <a class="code" href="json_parse_8m.html#a6e4256f69b751d3d7ec4ea52ff5a8dee">properties</a>
<a name="l00224"></a>00224         localStateNames__          = {}; %names of simulation state <a class="code" href="json_parse_8m.html#a6e4256f69b751d3d7ec4ea52ff5a8dee">properties</a> redundant with timecourses in <span class="keyword">this</span> or other processes or the simulation
<a name="l00225"></a>00225         
<a name="l00226"></a>00226         leadingStrandIndexs = [1 4];
<a name="l00227"></a>00227         laggingStrandIndexs = [3 2];
<a name="l00228"></a>00228     end
<a name="l00229"></a>00229     
<a name="l00230"></a>00230     %IDs, names, and local indices
<a name="l00231"></a>00231     <a class="code" href="json_parse_8m.html#a6e4256f69b751d3d7ec4ea52ff5a8dee">properties</a>
<a name="l00232"></a>00232         stimuliWholeCellModelIDs = {}; %whole cell model IDs of stimuli
<a name="l00233"></a>00233         
<a name="l00234"></a>00234         substrateWholeCellModelIDs = { %whole cell model IDs of substrates
<a name="l00235"></a>00235             <span class="stringliteral">&#39;DATP&#39;</span>; <span class="stringliteral">&#39;DCTP&#39;</span>; <span class="stringliteral">&#39;DGTP&#39;</span>; <span class="stringliteral">&#39;DTTP&#39;</span>;
<a name="l00236"></a>00236             <span class="stringliteral">&#39;ATP&#39;</span>; <span class="stringliteral">&#39;CTP&#39;</span>; <span class="stringliteral">&#39;GTP&#39;</span>; <span class="stringliteral">&#39;UTP&#39;</span>;
<a name="l00237"></a>00237             <span class="stringliteral">&#39;PPI&#39;</span>; <span class="stringliteral">&#39;H2O&#39;</span>; <span class="charliteral">&#39;H&#39;</span>; <span class="stringliteral">&#39;NAD&#39;</span>; <span class="stringliteral">&#39;NMN&#39;</span>; <span class="stringliteral">&#39;ADP&#39;</span>; <span class="stringliteral">&#39;AMP&#39;</span>; <span class="stringliteral">&#39;PI&#39;</span>;
<a name="l00238"></a>00238             };
<a name="l00239"></a>00239         substrateIndexs_dntp        = (1:4)<span class="stringliteral">&#39;;</span>
<a name="l00240"></a>00240 <span class="stringliteral">        substrateIndexs_ntp         = (5:8)&#39;</span>;
<a name="l00241"></a>00241         substrateIndexs_diphosphate = 9;
<a name="l00242"></a>00242         substrateIndexs_water       = 10;
<a name="l00243"></a>00243         substrateIndexs_hydrogen    = 11;
<a name="l00244"></a>00244         substrateIndexs_nad         = 12;
<a name="l00245"></a>00245         substrateIndexs_nmn         = 13;
<a name="l00246"></a>00246         substrateIndexs_atp         = 5;
<a name="l00247"></a>00247         substrateIndexs_adp         = 14;
<a name="l00248"></a>00248         substrateIndexs_amp         = 15;
<a name="l00249"></a>00249         substrateIndexs_phosphate   = 16;
<a name="l00250"></a>00250         
<a name="l00251"></a>00251         enzymeWholeCellModelIDs = { %enzyme whole cell model ids
<a name="l00252"></a>00252             <span class="stringliteral">&#39;REPLISOME&#39;</span>;                                            %replisome
<a name="l00253"></a>00253             <span class="stringliteral">&#39;DNA_POLYMERASE_2CORE_BETA_CLAMP_GAMMA_COMPLEX_PRIMASE&#39;</span>;%DNA-directed DNA polymerase (2) core + beta-clamp + gamma-complex + primase
<a name="l00254"></a>00254             &#39;DNA_POLYMERASE_CORE_BETA_CLAMP_GAMMA_COMPLEX&#39;;         %DNA-directed DNA polymerase core + beta-clamp + gamma-complex
<a name="l00255"></a>00255             &#39;DNA_POLYMERASE_CORE_BETA_CLAMP_PRIMASE&#39;;               %DNA-directed DNA polymerase core + beta-clamp + primase
<a name="l00256"></a>00256             &#39;DNA_POLYMERASE_CORE&#39;;                                  %DNA-directed DNA polymerase core
<a name="l00257"></a>00257             &#39;DNA_POLYMERASE_GAMMA_COMPLEX&#39;;                         %DNA-directed DNA polymerase gamma complex
<a name="l00258"></a>00258             &#39;MG_001_MONOMER&#39;;                                       %DNA polymerase III, beta subunit
<a name="l00259"></a>00259             <span class="stringliteral">&#39;MG_001_DIMER&#39;</span>;                                         %DNA polymerase III, beta clamp
<a name="l00260"></a>00260             <span class="stringliteral">&#39;MG_094_HEXAMER&#39;</span>;                                       %replicative DNA helicase
<a name="l00261"></a>00261             <span class="stringliteral">&#39;MG_254_MONOMER&#39;</span>;                                       %DNA ligase, NAD-dependent
<a name="l00262"></a>00262             &#39;MG_250_MONOMER&#39;;                                       %DNA primase
<a name="l00263"></a>00263             <span class="stringliteral">&#39;MG_091_TETRAMER&#39;</span>;                                      %single-strand binding protein family, tetramer
<a name="l00264"></a>00264             &#39;MG_091_OCTAMER&#39;;                                       %single-strand binding protein family, octamer
<a name="l00265"></a>00265             };
<a name="l00266"></a>00266         enzymeIndexs_replisome                         = 1;
<a name="l00267"></a>00267         enzymeIndexs_2coreBetaClampGammaComplexPrimase = 2;
<a name="l00268"></a>00268         enzymeIndexs_coreBetaClampGammaComplex         = 3;
<a name="l00269"></a>00269         enzymeIndexs_coreBetaClampPrimase              = 4;
<a name="l00270"></a>00270         enzymeIndexs_core                              = 5;
<a name="l00271"></a>00271         enzymeIndexs_gammaComplex                      = 6;
<a name="l00272"></a>00272         enzymeIndexs_betaClampMonomer                  = 7;
<a name="l00273"></a>00273         enzymeIndexs_betaClamp                         = 8;
<a name="l00274"></a>00274         enzymeIndexs_helicase                          = 9;
<a name="l00275"></a>00275         enzymeIndexs_ligase                            = 10;
<a name="l00276"></a>00276         enzymeIndexs_primase                           = 11;
<a name="l00277"></a>00277         enzymeIndexs_ssb4mer                           = 12;
<a name="l00278"></a>00278         enzymeIndexs_ssb8mer                           = 13;
<a name="l00279"></a>00279         
<a name="l00280"></a>00280         complexIndexs_DnaA_1mer_ATP       %index of DnaA-ATP 1mer within simulation.complexs
<a name="l00281"></a>00281         complexIndexs_DnaA_7mer_ATP       %index of DnaA-ATP 7mer within simulation.complexs
<a name="l00282"></a>00282 
<a name="l00283"></a>00283         dnaAFunctionalBoxIndexs_R1234     %index of R1-4 functional DnaA boxes within dnaAFunctionalBoxStartPositions
<a name="l00284"></a>00284         dnaAFunctionalBoxIndexs_R5        %index of R5 functional DnaA box within dnaAFunctionalBoxStartPositions
<a name="l00285"></a>00285     end
<a name="l00286"></a>00286     
<a name="l00287"></a>00287     %fixed biological constants
<a name="l00288"></a>00288     <a class="code" href="json_parse_8m.html#a6e4256f69b751d3d7ec4ea52ff5a8dee">properties</a>
<a name="l00289"></a>00289         oriCPosition                       %oriC position
<a name="l00290"></a>00290         terCPosition                       %terC position
<a name="l00291"></a>00291         
<a name="l00292"></a>00292         enzymeComposition                  %replisome subunit composition       
<a name="l00293"></a>00293         
<a name="l00294"></a>00294         primaseBindingLocations            %locations (randomly determined) where primers will bind the lagging strand to make okazaki fragments
<a name="l00295"></a>00295         
<a name="l00296"></a>00296         primerLength                       %length of primer that needs to form before an okazaki fragment can be replicated by DNA polymerase (nt) (11) [PUB_0606]
<a name="l00297"></a>00297         dnaPolymeraseElongationRate        %DNA polymerase elongation rate (nt/s) (100) [Miyata 2002]
<a name="l00298"></a>00298         ssbComplexSpacing                  %spacing between SSB octamers (nt) [PUB_0637]
<a name="l00299"></a>00299         okazakiFragmentMeanLength          %Average length of an Okazaki fragment (nt) [PUB_0623, Kornberg 1980, Xie 2008]
<a name="l00300"></a>00300         ligaseRate                         %Ligase Vmax (1/s) (0.04) [PUB_0488]
<a name="l00301"></a>00301         laggingBackupClampReloadingLength  %Length at which a second backup clamp binds the loading complex in terms of how many bases have been polymerized of the Okazaki fragment (750)
<a name="l00302"></a>00302         startingOkazakiLoopLength          %Length of DNA (bases) in the okazaki fragment loop right at the beginning of the fragment polymerization (50)
<a name="l00303"></a>00303         rnaPolymeraseCollisionMeanDwellTime%mean time of DNA polymerase stalling upon interaction with anti-directional <a class="code" href="class_r_n_a.html">RNA</a> polymerase (s) [1.7 PUB_0799] 
<a name="l00304"></a>00304         ssbDissociationRate                %dissociation rate of SSBs (1/s) [0.3 PUB_0857]
<a name="l00305"></a>00305         
<a name="l00306"></a>00306         dnaAFunctionalBoxStartPositions    %sorted start coordinates of functional DnaA boxes
<a name="l00307"></a>00307     end
<a name="l00308"></a>00308     
<a name="l00309"></a>00309     %dependent local state (implemented as dependent property for convenience)
<a name="l00310"></a>00310     <a class="code" href="json_parse_8m.html#a6e4256f69b751d3d7ec4ea52ff5a8dee">properties</a> (Dependent = true)
<a name="l00311"></a>00311         isDnaAORIComplexAssembled      %<span class="keywordtype">boolean</span> indicating whether or not a DnaA-ATP complex has formed at the oriC, if true replication can initiate
<a name="l00312"></a>00312         isAnyHelicaseBound             %<span class="keywordtype">boolean</span> indicating whether or not any helicase is bound
<a name="l00313"></a>00313         isAnyPolymeraseBound           %<span class="keywordtype">boolean</span> indicating whether or not any polymerase is bound
<a name="l00314"></a>00314         leadingStrandElongating        %whether replication elongation is in progress
<a name="l00315"></a>00315         laggingStrandElongating        %whether there is a backup clamp bound to each loading complex
<a name="l00316"></a>00316         leadingStrandPolymerized       %whether leading strands have been polymerized
<a name="l00317"></a>00317         laggingStrandPolymerized       %whether lagging strands have been polymerized
<a name="l00318"></a>00318         strandPolymerized              %whether leading and lagging strands have been polymerized
<a name="l00319"></a>00319         numLigations                   %number of ligations that have occurred (in replication order along lagging strand)
<a name="l00320"></a>00320         strandLigated                  %whether or not strands are completely ligated
<a name="l00321"></a>00321         strandDuplicated               %whether replication is complete and chromosome segregation can start
<a name="l00322"></a>00322         
<a name="l00323"></a>00323         helicasePosition               %helicase start coordinates
<a name="l00324"></a>00324         leadingPolymerasePosition      %leading polymerase start coordinates
<a name="l00325"></a>00325         laggingPolymerasePosition      %lagging polymerase start coordinates
<a name="l00326"></a>00326         leadingPosition                %next position to be polymerized on leading strands
<a name="l00327"></a>00327         laggingPosition                %next position to be polymerized on lagging strands
<a name="l00328"></a>00328         laggingBackupBetaClampPosition %lagging backup beta clamp start coordinates
<a name="l00329"></a>00329         
<a name="l00330"></a>00330         okazakiFragmentIndex           %index of the current/most recent Okazaki fragments
<a name="l00331"></a>00331         okazakiFragmentPosition        %5&#39; position of the current/most recent Okazaki fragments, including the primer
<a name="l00332"></a>00332         okazakiFragmentLength          %length of the current/most recent Okazaki fragments, including the primer
<a name="l00333"></a>00333         okazakiFragmentProgress        %number of nucleotides polymerized of the current/most recent Okazaki fragments
<a name="l00334"></a>00334         
<a name="l00335"></a>00335         leadingStrandBoundSSBs         %sparse vector indicating where SSB complexes are bound
<a name="l00336"></a>00336         laggingStrandBoundSSBs         %sparse vector indicating where SSB complexes are bound
<a name="l00337"></a>00337         numLeadingTemplateBoundSSBs    %number of SSB 8mers bound to the leading strand&#39;s template
<a name="l00338"></a>00338         numLaggingTemplateBoundSSBs    %number of SSB 8mers bound to the lagging strand&#39;s template
<a name="l00339"></a>00339         areLaggingStrandSSBSitesBound  %true/false whether or not the lagging strand&#39;s template is sufficiently covered in SSB 8mers for replication to progress
<a name="l00340"></a>00340     end
<a name="l00341"></a>00341     
<a name="l00342"></a>00342     %constructor
<a name="l00343"></a>00343     methods
<a name="l00344"></a>00344         function this = <a class="code" href="class_replication.html">Replication</a>(wholeCellModelID, name)
<a name="l00345"></a>00345             this = this@edu.stanford.covert.cell.sim.<a class="code" href="class_process.html">Process</a>(wholeCellModelID, name);
<a name="l00346"></a>00346         end
<a name="l00347"></a>00347     end
<a name="l00348"></a>00348     
<a name="l00349"></a>00349     %communication between process/simulation
<a name="l00350"></a>00350     methods
<a name="l00351"></a>00351         %<span class="keyword">set</span> references to state objects
<a name="l00352"></a>00352         <span class="keyword">function</span> storeObjectReferences(<span class="keyword">this</span>, simulation)
<a name="l00353"></a>00353             this.storeObjectReferences@edu.stanford.covert.cell.sim.<a class="code" href="class_process.html">Process</a>(simulation);
<a name="l00354"></a>00354             this.storeObjectReferences@edu.stanford.covert.cell.sim.<a class="code" href="class_chromosome_process_aspect.html">ChromosomeProcessAspect</a>(simulation);
<a name="l00355"></a>00355         end
<a name="l00356"></a>00356         
<a name="l00357"></a>00357         %initialize constants
<a name="l00358"></a>00358         function initializeConstants(this, knowledgeBase, simulation, varargin)
<a name="l00359"></a>00359             this.initializeConstants@edu.stanford.covert.cell.sim.<a class="code" href="class_process.html">Process</a>(knowledgeBase, simulation, varargin{:});
<a name="l00360"></a>00360             this.initializeConstants@edu.stanford.covert.cell.sim.ChromosomeProcessAspect(knowledgeBase, simulation, varargin{:});
<a name="l00361"></a>00361             
<a name="l00362"></a>00362             c = this.chromosome;
<a name="l00363"></a>00363             
<a name="l00364"></a>00364             %oriC position
<a name="l00365"></a>00365             oriC = findobj(knowledgeBase.genomeFeatures, <span class="stringliteral">&#39;wholeCellModelID&#39;</span>, <span class="stringliteral">&#39;oriC&#39;</span>);
<a name="l00366"></a>00366             this.oriCPosition = oriC.startCoordinate;
<a name="l00367"></a>00367             
<a name="l00368"></a>00368             %terC position
<a name="l00369"></a>00369             terC = findobj(knowledgeBase.genomeFeatures, <span class="stringliteral">&#39;wholeCellModelID&#39;</span>, <span class="stringliteral">&#39;terC&#39;</span>);
<a name="l00370"></a>00370             this.terCPosition = terC.startCoordinate;
<a name="l00371"></a>00371             
<a name="l00372"></a>00372             %enzyme subunit composition
<a name="l00373"></a>00373             this.enzymeComposition = zeros(numel(this.enzymeWholeCellModelIDs));
<a name="l00374"></a>00374             this.enzymeComposition(this.enzymeMonomerLocalIndexs, this.enzymeComplexLocalIndexs) = sum(knowledgeBase.proteinComplexMonomerComposition(<span class="keyword">this</span>.gene.mRNAIndexs(<span class="keyword">this</span>.enzymeMonomerGlobalIndexs), this.enzymeComplexGlobalIndexs, :), 3);
<a name="l00375"></a>00375             this.enzymeComposition(this.enzymeComplexLocalIndexs, this.enzymeComplexLocalIndexs) = sum(knowledgeBase.proteinComplexComplexComposition(<span class="keyword">this</span>.enzymeComplexGlobalIndexs, <span class="keyword">this</span>.enzymeComplexGlobalIndexs, :), 3);
<a name="l00376"></a>00376                          
<a name="l00377"></a>00377             %primase binding locations
<a name="l00378"></a>00378             this.primaseBindingLocations = {
<a name="l00379"></a>00379                 c.sequenceLen + 1 - this.calculatePrimaseBindingLocations(c.sequenceLen - this.terCPosition)
<a name="l00380"></a>00380                 this.calculatePrimaseBindingLocations(this.terCPosition)
<a name="l00381"></a>00381                 }&#39;;
<a name="l00382"></a>00382             
<a name="l00383"></a>00383             %DnaA and its binding sites
<a name="l00384"></a>00384             this.complexIndexs_DnaA_1mer_ATP = this.complex.getIndexs({<span class="stringliteral">&#39;MG_469_1MER_ATP&#39;</span>});
<a name="l00385"></a>00385             this.complexIndexs_DnaA_7mer_ATP = this.complex.getIndexs({<span class="stringliteral">&#39;MG_469_7MER_ATP&#39;</span>});
<a name="l00386"></a>00386             
<a name="l00387"></a>00387             dnaABoxes = findobj(knowledgeBase.genomeFeatures, <span class="stringliteral">&#39;type&#39;</span>, <span class="stringliteral">&#39;DnaA box&#39;</span>);            
<a name="l00388"></a>00388             [~, order] = sort([dnaABoxes.startCoordinate]);
<a name="l00389"></a>00389             dnaABoxes = dnaABoxes(order);
<a name="l00390"></a>00390             
<a name="l00391"></a>00391             dnaABoxNames = {dnaABoxes.name}<span class="stringliteral">&#39;;</span>
<a name="l00392"></a>00392 <span class="stringliteral">            [tfs, idxs] = ismember(dnaABoxNames, {</span>
<a name="l00393"></a>00393 <span class="stringliteral">                &#39;</span>Functional box R1<span class="stringliteral">&#39;;</span>
<a name="l00394"></a>00394 <span class="stringliteral">                &#39;</span>Functional box R2<span class="stringliteral">&#39;;</span>
<a name="l00395"></a>00395 <span class="stringliteral">                &#39;</span>Functional box R3<span class="stringliteral">&#39;;</span>
<a name="l00396"></a>00396 <span class="stringliteral">                &#39;</span>Functional box R4<span class="stringliteral">&#39;;</span>
<a name="l00397"></a>00397 <span class="stringliteral">                &#39;</span>Functional box R5<span class="stringliteral">&#39;;</span>
<a name="l00398"></a>00398 <span class="stringliteral">                });</span>
<a name="l00399"></a>00399 <span class="stringliteral">            </span>
<a name="l00400"></a>00400 <span class="stringliteral">            dnaABoxes = dnaABoxes(tfs);</span>
<a name="l00401"></a>00401 <span class="stringliteral">            this.dnaAFunctionalBoxStartPositions = ...</span>
<a name="l00402"></a>00402 <span class="stringliteral">                ceil([dnaABoxes.startCoordinate]&#39;</span> + [dnaABoxes.sequenceLength]&#39;/2 - c.complexDNAFootprints(this.complexIndexs_DnaA_1mer_ATP)/2);
<a name="l00403"></a>00403             
<a name="l00404"></a>00404             this.dnaAFunctionalBoxIndexs_R1234 = [
<a name="l00405"></a>00405                 find(idxs(tfs) == 1);
<a name="l00406"></a>00406                 find(idxs(tfs) == 2);
<a name="l00407"></a>00407                 find(idxs(tfs) == 3);
<a name="l00408"></a>00408                 find(idxs(tfs) == 4)];
<a name="l00409"></a>00409             this.dnaAFunctionalBoxIndexs_R5 = find(idxs(tfs) == 5);
<a name="l00410"></a>00410         end
<a name="l00411"></a>00411         
<a name="l00412"></a>00412         %Choose Okazaki fragment lengths randomly according to poisson distribution
<a name="l00413"></a>00413         %with mean = okazakiFragmentMeanLength.
<a name="l00414"></a>00414         <span class="keyword">function</span> pos = calculatePrimaseBindingLocations(<span class="keyword">this</span>, len)
<a name="l00415"></a>00415             pos = this.randStream.random(&#39;poiss&#39;, this.okazakiFragmentMeanLength, ceil(1.2 * len/this.okazakiFragmentMeanLength), 1);
<a name="l00416"></a>00416             <a class="code" href="json_parse_8m.html#ad2fc293e8080616c8c1e2ffbd7b86742">while</a> sum(pos) &lt; len
<a name="l00417"></a>00417                 pos = [pos; this.randStream.random(&#39;poiss&#39;, this.okazakiFragmentMeanLength, 10, 1)]; %<span class="preprocessor">#ok&lt;AGROW&gt;</span>
<a name="l00418"></a>00418 <span class="preprocessor"></span>            end
<a name="l00419"></a>00419             
<a name="l00420"></a>00420             pos = [0; cumsum(pos)];
<a name="l00421"></a>00421             pos = len - pos(end:-1:1);
<a name="l00422"></a>00422             idx = find(pos &lt;= 0, 1, <span class="stringliteral">&#39;last&#39;</span>);
<a name="l00423"></a>00423             
<a name="l00424"></a>00424             pos(1:idx) = [];
<a name="l00425"></a>00425             <span class="keywordflow">if</span> pos(1) &lt; 0.6 * this.okazakiFragmentMeanLength
<a name="l00426"></a>00426                 pos(1) = [];
<a name="l00427"></a>00427             end
<a name="l00428"></a>00428         end
<a name="l00429"></a>00429     end
<a name="l00430"></a>00430     
<a name="l00431"></a>00431     %model
<a name="l00432"></a>00432     methods
<a name="l00433"></a>00433         %Calculate
<a name="l00434"></a>00434         %- contribution to <a class="code" href="class_f_b_a.html">FBA</a> objective
<a name="l00435"></a>00435         %- minimum expression consistent with cell cycle length
<a name="l00436"></a>00436         function [bmProd, byProd, minEnzExp, maxEnzExp] = calcResourceRequirements_LifeCycle(this, constants, ~)
<a name="l00437"></a>00437             %% import classes
<a name="l00438"></a>00438             import edu.stanford.covert.util.<a class="code" href="class_constant_util.html">ConstantUtil</a>;
<a name="l00439"></a>00439             
<a name="l00440"></a>00440             %% initialize
<a name="l00441"></a>00441             bmProd = zeros(size(this.substrateWholeCellModelIDs));
<a name="l00442"></a>00442             byProd = zeros(size(this.substrateWholeCellModelIDs));
<a name="l00443"></a>00443             minEnzExp = zeros(size(this.enzymeWholeCellModelIDs));
<a name="l00444"></a>00444             maxEnzExp = Inf(size(this.enzymeWholeCellModelIDs));
<a name="l00445"></a>00445             
<a name="l00446"></a>00446             %% substrate and byproducts
<a name="l00447"></a>00447             c = this.chromosome;
<a name="l00448"></a>00448             m = this.metabolite;
<a name="l00449"></a>00449             mass = constants.states.Mass;
<a name="l00450"></a>00450             nOkazakiFragments = sum(cellfun(@numel, this.primaseBindingLocations));
<a name="l00451"></a>00451             
<a name="l00452"></a>00452             baseComp = getBaseCounts(c.sequence);
<a name="l00453"></a>00453             dnmpComp = [baseComp; nnz(c.damagedBases == m.m6ADIndexs)];
<a name="l00454"></a>00454             dnmpComp(1) = dnmpComp(1) - nnz(c.damagedBases == m.m6ADIndexs);
<a name="l00455"></a>00455             chrWt = dnmpComp<span class="stringliteral">&#39; * (m.molecularWeights([m.dnmpIndexs; m.getIndexs({&#39;</span>m6dAMP<span class="stringliteral">&#39;})]) - ConstantUtil.elements.O - ConstantUtil.elements.H) / ConstantUtil.nAvogadro;</span>
<a name="l00456"></a>00456 <span class="stringliteral">            dntpProd = ...</span>
<a name="l00457"></a>00457 <span class="stringliteral">                baseComp * min(1, mass.dryWeightFractionDNA * mass.cellInitialDryWeight / chrWt) + ...       </span>
<a name="l00458"></a>00458 <span class="stringliteral">                baseComp * max(0, mass.dryWeightFractionDNA * mass.cellInitialDryWeight - chrWt) / ...</span>
<a name="l00459"></a>00459 <span class="stringliteral">                (baseComp&#39;</span> * m.molecularWeights(m.dntpIndexs)) * <a class="code" href="class_constant_util.html">ConstantUtil</a>.<a class="code" href="class_constant_util.html#a9da93ee8ba55dd66d9406905f4c63a4b">nAvogadro</a>;
<a name="l00460"></a>00460             basesPol = c.sequenceLen * min(1, mass.dryWeightFractionDNA * mass.cellInitialDryWeight / chrWt);
<a name="l00461"></a>00461            
<a name="l00462"></a>00462             %ATP <span class="keywordflow">for</span> helicase to unwind mother chromosome
<a name="l00463"></a>00463             bmProd(this.substrateIndexs_atp)         = bmProd(this.substrateIndexs_atp)         + basesPol;
<a name="l00464"></a>00464             bmProd(this.substrateIndexs_water)       = bmProd(this.substrateIndexs_water)       + basesPol;
<a name="l00465"></a>00465             byProd(this.substrateIndexs_adp)         = byProd(this.substrateIndexs_adp)         + basesPol;
<a name="l00466"></a>00466             byProd(this.substrateIndexs_phosphate)   = byProd(this.substrateIndexs_phosphate)   + basesPol;
<a name="l00467"></a>00467             byProd(this.substrateIndexs_hydrogen)    = byProd(this.substrateIndexs_hydrogen)    + basesPol;
<a name="l00468"></a>00468             
<a name="l00469"></a>00469             %ATP used by gamma-complex to form beta-clamps
<a name="l00470"></a>00470             nBetaClamps = sum(cellfun(@(x) ~isempty(x), this.primaseBindingLocations)) + nOkazakiFragments;
<a name="l00471"></a>00471             bmProd(this.substrateIndexs_atp)         = bmProd(this.substrateIndexs_atp)         + nBetaClamps;
<a name="l00472"></a>00472             bmProd(this.substrateIndexs_water)       = bmProd(this.substrateIndexs_water)       + nBetaClamps;
<a name="l00473"></a>00473             byProd(this.substrateIndexs_adp)         = byProd(this.substrateIndexs_adp)         + nBetaClamps;
<a name="l00474"></a>00474             byProd(this.substrateIndexs_phosphate)   = byProd(this.substrateIndexs_phosphate)   + nBetaClamps;
<a name="l00475"></a>00475             byProd(this.substrateIndexs_hydrogen)    = byProd(this.substrateIndexs_hydrogen)    + nBetaClamps;            
<a name="l00476"></a>00476             
<a name="l00477"></a>00477             %dNTPs to polymerize <span class="keyword">new</span> strands by end of replication phase
<a name="l00478"></a>00478             bmProd(this.substrateIndexs_dntp)        = bmProd(this.substrateIndexs_dntp)        + dntpProd;
<a name="l00479"></a>00479             byProd(this.substrateIndexs_diphosphate) = byProd(this.substrateIndexs_diphosphate) + 2 * basesPol;
<a name="l00480"></a>00480             
<a name="l00481"></a>00481             %Ligation: (2) DR5P + NAD ==&gt; AMP + dRibose5P_dRibose5P + H + NMN
<a name="l00482"></a>00482             nLigations = sum(cellfun(@(x) ~isempty(x), this.primaseBindingLocations)) + nOkazakiFragments;
<a name="l00483"></a>00483             bmProd(this.substrateIndexs_nad)         = bmProd(this.substrateIndexs_nad)          + nLigations;
<a name="l00484"></a>00484             byProd(this.substrateIndexs_nmn)         = byProd(this.substrateIndexs_nmn)          + nLigations;
<a name="l00485"></a>00485             byProd(this.substrateIndexs_amp)         = byProd(this.substrateIndexs_amp)          + nLigations;
<a name="l00486"></a>00486             byProd(this.substrateIndexs_hydrogen)    = byProd(this.substrateIndexs_hydrogen)     + nLigations;
<a name="l00487"></a>00487             
<a name="l00488"></a>00488             %% enzymes            
<a name="l00489"></a>00489             %replisome: 
<a name="l00490"></a>00490             %- helicase
<a name="l00491"></a>00491             %- polymerase core (2)
<a name="l00492"></a>00492             %- gamma-complex
<a name="l00493"></a>00493             %- beta-clamp (2 dimers = 4 monomers)
<a name="l00494"></a>00494             %- primase
<a name="l00495"></a>00495             minEnzExp(this.enzymeIndexs_replisome) = 2;
<a name="l00496"></a>00496             
<a name="l00497"></a>00497             %lagging stand backup beta-clamp
<a name="l00498"></a>00498             minEnzExp(this.enzymeIndexs_betaClamp) = 2;
<a name="l00499"></a>00499             
<a name="l00500"></a>00500             %ligase
<a name="l00501"></a>00501             nLigations = sum(cellfun(@(x) ~isempty(x), this.primaseBindingLocations)) + sum(cellfun(@numel, this.primaseBindingLocations));
<a name="l00502"></a>00502             minEnzExp(this.enzymeIndexs_ligase) = nLigations / this.ligaseRate / constants.states.Time.replicationDuration;
<a name="l00503"></a>00503 
<a name="l00504"></a>00504             %single-stranded binding protein
<a name="l00505"></a>00505             ssbFtpt  = this.enzymeDNAFootprints(this.enzymeIndexs_ssb8mer);
<a name="l00506"></a>00506             ssDNALen = 2 * this.okazakiFragmentMeanLength;
<a name="l00507"></a>00507             minEnzExp(this.enzymeIndexs_ssb8mer) = 2 * ceil(ssDNALen / ssbFtpt);
<a name="l00508"></a>00508             
<a name="l00509"></a>00509             %account <span class="keywordflow">for</span> timing of replication
<a name="l00510"></a>00510             minEnzExp = minEnzExp / exp(log(2) * constants.states.Time.replicationInitiationDuration / constants.states.Time.cellCycleLength);
<a name="l00511"></a>00511         end
<a name="l00512"></a>00512         
<a name="l00513"></a>00513         %initialization: simulation initialize state method initializes 1
<a name="l00514"></a>00514         %chromosome
<a name="l00515"></a>00515         <span class="keyword">function</span> initializeState(~)
<a name="l00516"></a>00516         end
<a name="l00517"></a>00517         
<a name="l00518"></a>00518         %resource requirements
<a name="l00519"></a>00519         function result = calcResourceRequirements_Current(this)
<a name="l00520"></a>00520             c = this.chromosome;
<a name="l00521"></a>00521             
<a name="l00522"></a>00522             %initialize
<a name="l00523"></a>00523             result = zeros(size(this.substrates));
<a name="l00524"></a>00524             
<a name="l00525"></a>00525             %ATP, water for initiation
<a name="l00526"></a>00526             if ~this.isAnyPolymeraseBound &amp;&amp; this.isDnaAORIComplexAssembled &amp;&amp; all(this.enzymes &gt;= 2 * this.enzymeComposition(:, this.enzymeIndexs_2coreBetaClampGammaComplexPrimase))
<a name="l00527"></a>00527                 helFtpt3 = this.enzymeDNAFootprints3Prime(this.enzymeIndexs_helicase);
<a name="l00528"></a>00528                 corFtpt5 = this.enzymeDNAFootprints5Prime(this.enzymeIndexs_core);
<a name="l00529"></a>00529                 uwdLen = helFtpt3 + corFtpt5 + 1;
<a name="l00530"></a>00530                 result(this.substrateIndexs_atp)   = result(this.substrateIndexs_atp)   + 2 * (1 + uwdLen);
<a name="l00531"></a>00531                 result(this.substrateIndexs_water) = result(this.substrateIndexs_water) + 2 * (1 + uwdLen);
<a name="l00532"></a>00532             end
<a name="l00533"></a>00533             
<a name="l00534"></a>00534             %ATP for helicase to unwind mother chromosome
<a name="l00535"></a>00535             if this.isAnyHelicaseBound &amp;&amp; all(this.leadingStrandElongating)
<a name="l00536"></a>00536                 nLeadingStrandElongating = 2;
<a name="l00537"></a>00537                 nLaggingStrandElongating = nnz(this.laggingStrandElongating);
<a name="l00538"></a>00538                 isOriCCplx = this.isDnaAORIComplexAssembled;
<a name="l00539"></a>00539                 
<a name="l00540"></a>00540                 unwindLen = min(2, 2 * isOriCCplx + nLeadingStrandElongating) * this.dnaPolymeraseElongationRate;
<a name="l00541"></a>00541                 result(this.substrateIndexs_atp)   = result(this.substrateIndexs_atp)   + unwindLen;
<a name="l00542"></a>00542                 result(this.substrateIndexs_water) = result(this.substrateIndexs_water) + unwindLen;
<a name="l00543"></a>00543                 
<a name="l00544"></a>00544                 %ATP used by gamma-complex to form beta-clamps
<a name="l00545"></a>00545                 nBetaClamps = min(2, 2 * isOriCCplx + nLaggingStrandElongating) - nnz(this.laggingBackupBetaClampPosition);
<a name="l00546"></a>00546                 result(this.substrateIndexs_atp)   = result(this.substrateIndexs_atp)   + nBetaClamps;
<a name="l00547"></a>00547                 result(this.substrateIndexs_water) = result(this.substrateIndexs_water) + nBetaClamps;
<a name="l00548"></a>00548                 
<a name="l00549"></a>00549                 %dNTPs to polymerize new strands
<a name="l00550"></a>00550                 dnmpComp = [
<a name="l00551"></a>00551                     1 - c.sequenceGCContent;
<a name="l00552"></a>00552                     c.sequenceGCContent;
<a name="l00553"></a>00553                     c.sequenceGCContent;
<a name="l00554"></a>00554                     1-c.sequenceGCContent] / 2;
<a name="l00555"></a>00555                 result(this.substrateIndexs_dntp) = ...
<a name="l00556"></a>00556                     + result(this.substrateIndexs_dntp) ...
<a name="l00557"></a>00557                     + min(4, 2 * isOriCCplx + nLeadingStrandElongating + nLaggingStrandElongating) * ...
<a name="l00558"></a>00558                     this.dnaPolymeraseElongationRate * dnmpComp;
<a name="l00559"></a>00559             end
<a name="l00560"></a>00560             
<a name="l00561"></a>00561             %Ligation: (2) DR5P + NAD ==&gt; AMP + dRibose5P_dRibose5P + H + NMN
<a name="l00562"></a>00562             result(this.substrateIndexs_nad) = ...
<a name="l00563"></a>00563                 + result(this.substrateIndexs_nad) ...
<a name="l00564"></a>00564                 + min(nnz(c.strandBreaks), ceil(this.enzymes(this.enzymeIndexs_ligase) * this.stepSizeSec * this.ligaseRate));
<a name="l00565"></a>00565         end
<a name="l00566"></a>00566         
<a name="l00567"></a>00567         %simulation
<a name="l00568"></a>00568         function evolveState(this)
<a name="l00569"></a>00569             %% check that replisome state is in sync
<a name="l00570"></a>00570             cnts = this.boundEnzymes([
<a name="l00571"></a>00571                 this.enzymeIndexs_2coreBetaClampGammaComplexPrimase
<a name="l00572"></a>00572                 this.enzymeIndexs_coreBetaClampGammaComplex
<a name="l00573"></a>00573                 this.enzymeIndexs_coreBetaClampPrimase
<a name="l00574"></a>00574                 this.enzymeIndexs_helicase
<a name="l00575"></a>00575                 ]);
<a name="l00576"></a>00576             totPolCnts = [2 1 1] * cnts(1:3);
<a name="l00577"></a>00577             if ~((totPolCnts == 0 || totPolCnts == 4) &amp;&amp; ((cnts(4) == 0 &amp;&amp; totPolCnts == 0) || cnts(4) == 2))
<a name="l00578"></a>00578                 msg = [];
<a name="l00579"></a>00579                 msg = [msg sprintf(&#39;Replisome state out of sync\n&#39;)];
<a name="l00580"></a>00580                 msg = [msg sprintf(&#39;- 2coreBetaClampGammaComplexPrimase %d\n&#39;, cnts(1))];
<a name="l00581"></a>00581                 msg = [msg sprintf(&#39;- coreBetaClampGammaComplex %d\n&#39;, cnts(2))];
<a name="l00582"></a>00582                 msg = [msg sprintf(&#39;- coreBetaClampPrimase %d\n&#39;, cnts(3))];
<a name="l00583"></a>00583                 msg = [msg sprintf(&#39;- helicase %d&#39;, cnts(4))];
<a name="l00584"></a>00584                 throw(MException(&#39;<a class="code" href="class_replication.html">Replication</a>:error&#39;, msg));
<a name="l00585"></a>00585             end
<a name="l00586"></a>00586             
<a name="l00587"></a>00587             %% simulate
<a name="l00588"></a>00588             subfunctions = {
<a name="l00589"></a>00589                 @this.initiateReplication;        %Initiate replication
<a name="l00590"></a>00590                 @this.ligateDNA;                  %Ligate DNA
<a name="l00591"></a>00591                 @this.dissociateFreeSSBComplexes; %Dissociate free SSB 8mers into 2 SSB 4mers
<a name="l00592"></a>00592                 @this.freeAndBindSSBs;            %Free, Bind SSBs
<a name="l00593"></a>00593                 };
<a name="l00594"></a>00594             
<a name="l00595"></a>00595             %Only execute <span class="keywordflow">if</span> active replisomes
<a name="l00596"></a>00596             <span class="keywordflow">if</span> this.isAnyHelicaseBound &amp;&amp; all(this.leadingStrandElongating)
<a name="l00597"></a>00597                 subfunctions = [
<a name="l00598"></a>00598                     subfunctions;{
<a name="l00599"></a>00599                     @this.unwindAndPolymerizeDNA;     %Advance replisomes: unwind and polymerize DNA, release SSBs
<a name="l00600"></a>00600                     @this.initiateOkazakiFragment;    %Initiate Okazaki fragment by bind beta-clamp
<a name="l00601"></a>00601                     @this.terminateOkazakiFragment;   %Terminate Okazaki fragment by releasing beta-clamp
<a name="l00602"></a>00602                     @this.terminateReplication;       %Terminate replication
<a name="l00603"></a>00603                     }];
<a name="l00604"></a>00604             end
<a name="l00605"></a>00605             
<a name="l00606"></a>00606             order = this.randStream.randperm(numel(subfunctions));
<a name="l00607"></a>00607             <span class="keywordflow">for</span> i = 1:numel(subfunctions)
<a name="l00608"></a>00608                 subfunctions{order(i)}();
<a name="l00609"></a>00609             end
<a name="l00610"></a>00610         end
<a name="l00611"></a>00611         
<a name="l00612"></a>00612         %Initiate replication <span class="keywordflow">if</span> DnaA complex at oriC and sufficient proteins
<a name="l00613"></a>00613         %to form two replisomes:
<a name="l00614"></a>00614         %- form two replisomes at origin
<a name="l00615"></a>00615         %- indirectly (via <a class="code" href="class_chromosome.html">Chromosome</a>) disassemble DnaA complex
<a name="l00616"></a>00616         function initiateReplication(this)
<a name="l00617"></a>00617             chrLen = this.chromosome.sequenceLen;
<a name="l00618"></a>00618             holFtpt  = this.enzymeDNAFootprints(this.enzymeIndexs_2coreBetaClampGammaComplexPrimase);
<a name="l00619"></a>00619             helFtpt3 = this.enzymeDNAFootprints3Prime(this.enzymeIndexs_helicase);
<a name="l00620"></a>00620             helFtpt5 = this.enzymeDNAFootprints5Prime(this.enzymeIndexs_helicase);
<a name="l00621"></a>00621             corFtpt5 = this.enzymeDNAFootprints5Prime(this.enzymeIndexs_core);
<a name="l00622"></a>00622             helCorSpacing = helFtpt3 + corFtpt5 + 1;
<a name="l00623"></a>00623             
<a name="l00624"></a>00624             %check <span class="keywordflow">if</span> DnaA complex at oriC and sufficient protein to form 2
<a name="l00625"></a>00625             %replisomes
<a name="l00626"></a>00626             <span class="keywordflow">if</span> ~this.isDnaAORIComplexAssembled || any(this.leadingPolymerasePosition)
<a name="l00627"></a>00627                 return;
<a name="l00628"></a>00628             end
<a name="l00629"></a>00629             
<a name="l00630"></a>00630             %warn if this would be a secondary initiation
<a name="l00631"></a>00631             if collapse(this.chromosome.polymerizedRegions) &gt; 2 * chrLen
<a name="l00632"></a>00632                 warning(&#39;WholeCell:warning&#39;, &#39;Unable to represent multiple replication initiation events per cell cycle&#39;);
<a name="l00633"></a>00633                 return;
<a name="l00634"></a>00634             end
<a name="l00635"></a>00635             
<a name="l00636"></a>00636             helPos = this.helicasePosition;
<a name="l00637"></a>00637             maxSteps = floor(min(this.substrates([this.substrateIndexs_atp; this.substrateIndexs_water])) / 2);
<a name="l00638"></a>00638             if all(helPos)
<a name="l00639"></a>00639                 uwdOld = mod(-(1 - helFtpt3 - helPos(2)) - 1, chrLen) + 1;
<a name="l00640"></a>00640                 uwdLen = min(helCorSpacing - uwdOld, maxSteps);
<a name="l00641"></a>00641                 atpCost = 2 * uwdLen;
<a name="l00642"></a>00642                 if uwdLen &lt;= 0
<a name="l00643"></a>00643                     return;
<a name="l00644"></a>00644                 end
<a name="l00645"></a>00645                 this.releaseProteinFromSites([helPos; this.leadingStrandIndexs]&#39;, false);
<a name="l00646"></a>00646             elseif ~any(helPos)
<a name="l00647"></a>00647                 uwdLen = min(helCorSpacing, maxSteps - 1);
<a name="l00648"></a>00648                 uwdOld = 0;
<a name="l00649"></a>00649                 if uwdLen &lt; 0 || this.enzymes(this.enzymeIndexs_helicase) &lt; 2
<a name="l00650"></a>00650                     return;
<a name="l00651"></a>00651                 end
<a name="l00652"></a>00652                 atpCost = 2 * (1 + uwdLen);
<a name="l00653"></a>00653             else
<a name="l00654"></a>00654                 throw(MException(&#39;<a class="code" href="class_replication.html">Replication</a>:error&#39;, &#39;All or no helicases should be bound&#39;));
<a name="l00655"></a>00655             end
<a name="l00656"></a>00656             
<a name="l00657"></a>00657             %bind helicase and unwind chromosome
<a name="l00658"></a>00658             this.chromosome.setRegionUnwound(chrLen-uwdOld, -uwdLen);
<a name="l00659"></a>00659             this.chromosome.setRegionUnwound(1+uwdOld, uwdLen);
<a name="l00660"></a>00660             if ~all(this.bindProteinToChromosome([chrLen-helFtpt5-uwdOld 1-helFtpt3+uwdOld; this.leadingStrandIndexs]&#39;, this.enzymeIndexs_helicase, ...
<a name="l00661"></a>00661                     [], [], true, false, [-uwdLen-1; uwdLen+1], true, []))
<a name="l00662"></a>00662                 throw(MException(&#39;<a class="code" href="class_replication.html">Replication</a>:error&#39;, &#39;Both helicase must bind&#39;));
<a name="l00663"></a>00663             end
<a name="l00664"></a>00664             
<a name="l00665"></a>00665             %hydrolyze ATP used by gamma-complex to assemble beta-clamp, and by
<a name="l00666"></a>00666             %helicase to melt DNA
<a name="l00667"></a>00667             this.substrates(this.substrateIndexs_atp)       = this.substrates(this.substrateIndexs_atp)       - atpCost;
<a name="l00668"></a>00668             this.substrates(this.substrateIndexs_water)     = this.substrates(this.substrateIndexs_water)     - atpCost;
<a name="l00669"></a>00669             this.substrates(this.substrateIndexs_adp)       = this.substrates(this.substrateIndexs_adp)       + atpCost;
<a name="l00670"></a>00670             this.substrates(this.substrateIndexs_phosphate) = this.substrates(this.substrateIndexs_phosphate) + atpCost;
<a name="l00671"></a>00671             this.substrates(this.substrateIndexs_hydrogen)  = this.substrates(this.substrateIndexs_hydrogen)  + atpCost;
<a name="l00672"></a>00672             
<a name="l00673"></a>00673             %Form two replisomes (each with 2 cores, 1 gamma-complex, and 1 beta-clamp (MG_001 dimer)) at oriC
<a name="l00674"></a>00674             if uwdOld + uwdLen == helCorSpacing &amp;&amp; all(this.enzymes &gt;= 2 * this.enzymeComposition(:, this.enzymeIndexs_2coreBetaClampGammaComplexPrimase))
<a name="l00675"></a>00675                 this.enzymes = this.enzymes - 2 * this.enzymeComposition(:, this.enzymeIndexs_2coreBetaClampGammaComplexPrimase);
<a name="l00676"></a>00676                 this.enzymes(this.enzymeIndexs_2coreBetaClampGammaComplexPrimase) = this.enzymes(this.enzymeIndexs_2coreBetaClampGammaComplexPrimase) + 2;
<a name="l00677"></a>00677                 
<a name="l00678"></a>00678                 %bind DNA polymerase core and beta-clamp to initiate leading strand
<a name="l00679"></a>00679                 %synthesis
<a name="l00680"></a>00680                 if ~all(this.bindProteinToChromosome([chrLen-corFtpt5 1+corFtpt5-holFtpt+1; this.leadingStrandIndexs]&#39;, ...
<a name="l00681"></a>00681                         this.enzymeIndexs_2coreBetaClampGammaComplexPrimase, [], [], true, false, 1, true, []))
<a name="l00682"></a>00682                     throw(MException(&#39;<a class="code" href="class_replication.html">Replication</a>:error&#39;, &#39;Both polymerases must bind&#39;));
<a name="l00683"></a>00683                 end
<a name="l00684"></a>00684             end
<a name="l00685"></a>00685         end
<a name="l00686"></a>00686         
<a name="l00687"></a>00687         %Advance replisomes:
<a name="l00688"></a>00688         %- unwind DNA
<a name="l00689"></a>00689         %- polymerize DNA
<a name="l00690"></a>00690         %
<a name="l00691"></a>00691         %TODO: <a class="code" href="class_r_n_a.html">RNA</a> primer
<a name="l00692"></a>00692         function unwindAndPolymerizeDNA(this)
<a name="l00693"></a>00693             %% Terminate early if no active replisomes
<a name="l00694"></a>00694             if ~this.isAnyHelicaseBound || ~all(this.leadingStrandElongating)
<a name="l00695"></a>00695                 return;
<a name="l00696"></a>00696             end
<a name="l00697"></a>00697             
<a name="l00698"></a>00698             %% indices, parameters
<a name="l00699"></a>00699             helGblIdx = this.enzymeGlobalIndexs(this.enzymeIndexs_helicase);
<a name="l00700"></a>00700             leadPolLclIdx = [this.enzymeIndexs_2coreBetaClampGammaComplexPrimase; this.enzymeIndexs_coreBetaClampGammaComplex];
<a name="l00701"></a>00701             leadPolGblIdx = this.enzymeGlobalIndexs(leadPolLclIdx);
<a name="l00702"></a>00702             lagPolGblIdx = this.enzymeGlobalIndexs(this.enzymeIndexs_coreBetaClampPrimase);
<a name="l00703"></a>00703             helFtpt  = this.enzymeDNAFootprints(this.enzymeIndexs_helicase);
<a name="l00704"></a>00704             helFtpt3 = this.enzymeDNAFootprints3Prime(this.enzymeIndexs_helicase);
<a name="l00705"></a>00705             helFtpt5 = this.enzymeDNAFootprints5Prime(this.enzymeIndexs_helicase);
<a name="l00706"></a>00706             holFtpt = this.enzymeDNAFootprints(this.enzymeIndexs_coreBetaClampGammaComplex);
<a name="l00707"></a>00707             corFtpt = this.enzymeDNAFootprints(this.enzymeIndexs_core);
<a name="l00708"></a>00708             corFtpt3 = this.enzymeDNAFootprints3Prime(this.enzymeIndexs_core);
<a name="l00709"></a>00709             corFtpt5 = this.enzymeDNAFootprints5Prime(this.enzymeIndexs_core);
<a name="l00710"></a>00710             bClmpFtpt = this.enzymeDNAFootprints(this.enzymeIndexs_betaClamp);
<a name="l00711"></a>00711             
<a name="l00712"></a>00712             c = this.chromosome;
<a name="l00713"></a>00713             chrLen = c.sequenceLen;
<a name="l00714"></a>00714             firstBetaClampPos = [
<a name="l00715"></a>00715                 this.primaseBindingLocations{1}(1)-corFtpt3-bClmpFtpt
<a name="l00716"></a>00716                 this.primaseBindingLocations{2}(1)+corFtpt3+1]<span class="stringliteral">&#39;;</span>
<a name="l00717"></a>00717 <span class="stringliteral">            </span>
<a name="l00718"></a>00718 <span class="stringliteral">            %% bind lagging polymerase to lagging strand</span>
<a name="l00719"></a>00719 <span class="stringliteral">            leadingPolPos  = this.leadingPolymerasePosition;</span>
<a name="l00720"></a>00720 <span class="stringliteral">            laggingPolPos  = this.laggingPolymerasePosition;</span>
<a name="l00721"></a>00721 <span class="stringliteral">            </span>
<a name="l00722"></a>00722 <span class="stringliteral">            tfs = ...</span>
<a name="l00723"></a>00723 <span class="stringliteral">                laggingPolPos == 0 &amp; ...</span>
<a name="l00724"></a>00724 <span class="stringliteral">                this.laggingBackupBetaClampPosition == firstBetaClampPos &amp; ...</span>
<a name="l00725"></a>00725 <span class="stringliteral">                c.complexBoundSites([leadingPolPos; this.leadingStrandIndexs]&#39;</span>)<span class="stringliteral">&#39; == leadPolGblIdx(1);</span>
<a name="l00726"></a>00726 <span class="stringliteral">            </span>
<a name="l00727"></a>00727 <span class="stringliteral">            if tfs(1); laggingPolPos(1) = firstBetaClampPos(1);           end;</span>
<a name="l00728"></a>00728 <span class="stringliteral">            if tfs(2); laggingPolPos(2) = firstBetaClampPos(2) - corFtpt; end;</span>
<a name="l00729"></a>00729 <span class="stringliteral">            </span>
<a name="l00730"></a>00730 <span class="stringliteral">            n = sum(tfs);</span>
<a name="l00731"></a>00731 <span class="stringliteral">            </span>
<a name="l00732"></a>00732 <span class="stringliteral">            this.releaseProteinFromSites([leadingPolPos(:, tfs);     this.leadingStrandIndexs(:,tfs)]&#39;</span>, <span class="keyword">false</span>);
<a name="l00733"></a>00733             this.releaseProteinFromSites([firstBetaClampPos(:, tfs); this.laggingStrandIndexs(:,tfs)]<span class="stringliteral">&#39;, false);</span>
<a name="l00734"></a>00734 <span class="stringliteral">            </span>
<a name="l00735"></a>00735 <span class="stringliteral">            this.enzymes(this.enzymeIndexs_2coreBetaClampGammaComplexPrimase) = this.enzymes(this.enzymeIndexs_2coreBetaClampGammaComplexPrimase) - n;</span>
<a name="l00736"></a>00736 <span class="stringliteral">            this.enzymes(this.enzymeIndexs_coreBetaClampGammaComplex)         = this.enzymes(this.enzymeIndexs_coreBetaClampGammaComplex)  + n;</span>
<a name="l00737"></a>00737 <span class="stringliteral">            this.enzymes(this.enzymeIndexs_core)                              = this.enzymes(this.enzymeIndexs_core)                       + n;</span>
<a name="l00738"></a>00738 <span class="stringliteral">            this.enzymes(this.enzymeIndexs_primase)                           = this.enzymes(this.enzymeIndexs_primase)                    + n;</span>
<a name="l00739"></a>00739 <span class="stringliteral">            </span>
<a name="l00740"></a>00740 <span class="stringliteral">            this.enzymes(this.enzymeIndexs_betaClamp)                         = this.enzymes(this.enzymeIndexs_betaClamp)                  - n;</span>
<a name="l00741"></a>00741 <span class="stringliteral">            this.enzymes(this.enzymeIndexs_betaClampMonomer)                  = this.enzymes(this.enzymeIndexs_betaClampMonomer)           + 2*n;</span>
<a name="l00742"></a>00742 <span class="stringliteral">            </span>
<a name="l00743"></a>00743 <span class="stringliteral">            this.enzymes(this.enzymeIndexs_betaClampMonomer)                  = this.enzymes(this.enzymeIndexs_betaClampMonomer)           - 2*n;</span>
<a name="l00744"></a>00744 <span class="stringliteral">            this.enzymes(this.enzymeIndexs_core)                              = this.enzymes(this.enzymeIndexs_core)                       - n;</span>
<a name="l00745"></a>00745 <span class="stringliteral">            this.enzymes(this.enzymeIndexs_primase)                           = this.enzymes(this.enzymeIndexs_primase)                    - n;</span>
<a name="l00746"></a>00746 <span class="stringliteral">            this.enzymes(this.enzymeIndexs_coreBetaClampPrimase)              = this.enzymes(this.enzymeIndexs_coreBetaClampPrimase)       + n;</span>
<a name="l00747"></a>00747 <span class="stringliteral">            </span>
<a name="l00748"></a>00748 <span class="stringliteral">            if ~all(this.bindProteinToChromosome([leadingPolPos(:, tfs); this.leadingStrandIndexs(:,tfs)]&#39;</span>, this.enzymeIndexs_coreBetaClampGammaComplex, [], [], <span class="keyword">true</span>, <span class="keyword">false</span>, 1, <span class="keyword">true</span>, []))
<a name="l00749"></a>00749                 throw(MException(&#39;<a class="code" href="class_replication.html">Replication</a>:error&#39;, &#39;All polymerases should bind.&#39;));
<a name="l00750"></a>00750             end
<a name="l00751"></a>00751             if ~all(this.bindProteinToChromosome([laggingPolPos(:, tfs); this.laggingStrandIndexs(:,tfs)]&#39;, this.enzymeIndexs_coreBetaClampPrimase, [], [], true, false, 1, true, []))
<a name="l00752"></a>00752                 throw(MException(&#39;<a class="code" href="class_replication.html">Replication</a>:error&#39;, &#39;All polymerases should bind.&#39;));
<a name="l00753"></a>00753             end
<a name="l00754"></a>00754             
<a name="l00755"></a>00755             %% Calculate maximum unwinding and polymerization extent
<a name="l00756"></a>00756             helicasePos = this.helicasePosition;
<a name="l00757"></a>00757             leadingPolPos  = this.leadingPolymerasePosition;
<a name="l00758"></a>00758             laggingPolPos  = this.laggingPolymerasePosition;
<a name="l00759"></a>00759             leadingPos  = this.leadingPosition;
<a name="l00760"></a>00760             laggingPos  = this.laggingPosition;
<a name="l00761"></a>00761             fPos = this.okazakiFragmentPosition;
<a name="l00762"></a>00762             fPrg = this.okazakiFragmentProgress;
<a name="l00763"></a>00763             fLen = this.okazakiFragmentLength;
<a name="l00764"></a>00764             
<a name="l00765"></a>00765             %initialize limits
<a name="l00766"></a>00766             limits = zeros(2, 2); %[leading 1  leading 2; lagging 1  lagging 2]
<a name="l00767"></a>00767             
<a name="l00768"></a>00768             %primase kinetics
<a name="l00769"></a>00769             limits(1, 1) = this.primerLength - (chrLen - leadingPos(1));
<a name="l00770"></a>00770             limits(1, 2) = this.primerLength - (leadingPos(2) - 1);
<a name="l00771"></a>00771             limits(2, 1) = this.primerLength - (laggingPos(1) - fPos(1));
<a name="l00772"></a>00772             limits(2, 2) = this.primerLength - (fPos(2) - laggingPos(2));
<a name="l00773"></a>00773             
<a name="l00774"></a>00774             %DNA polymerase kinetics
<a name="l00775"></a>00775             limits(limits &lt;= 0) = this.dnaPolymeraseElongationRate;
<a name="l00776"></a>00776             
<a name="l00777"></a>00777             %polymerase must already be bound
<a name="l00778"></a>00778             limits(1, :) = limits(1, :) .* (leadingPos ~= 0);
<a name="l00779"></a>00779             limits(2, :) = limits(2, :) .* (laggingPos ~= 0);
<a name="l00780"></a>00780             
<a name="l00781"></a>00781             %prevent leading strand from progressing if no SSBs protecting lagging strand
<a name="l00782"></a>00782             limits(1, :) = limits(1, :) .* this.areLaggingStrandSSBSitesBound;
<a name="l00783"></a>00783             
<a name="l00784"></a>00784             %helicase
<a name="l00785"></a>00785             this.releaseProteinFromSites([helicasePos; this.leadingStrandIndexs]&#39;, false);
<a name="l00786"></a>00786             [~, ~, ~, extents] = c.isRegionAccessible([helicasePos + [0 helFtpt-1]; this.leadingStrandIndexs]&#39;, [-1; 1] .* (limits(1, :) + helFtpt + 1)&#39;, ...
<a name="l00787"></a>00787                 [], helGblIdx, false, [], true, true);
<a name="l00788"></a>00788             limits(1, :) = max(0, abs(extents)-helFtpt-1);
<a name="l00789"></a>00789             
<a name="l00790"></a>00790             %polymerase
<a name="l00791"></a>00791             bndLeadPolIdx = ismembc2(c.complexBoundSites([leadingPolPos; this.leadingStrandIndexs]&#39;), leadPolGblIdx);
<a name="l00792"></a>00792             this.releaseProteinFromSites([leadingPolPos; this.leadingStrandIndexs]&#39;, false);
<a name="l00793"></a>00793             [~, ~, ~, leadingExtents] = c.isRegionAccessible([leadingPolPos + [0 holFtpt-1]; this.leadingStrandIndexs]&#39;, [-1; 1] .* (limits(1, :) + holFtpt + 1)&#39;, ...
<a name="l00794"></a>00794                 [], leadPolGblIdx(1), false, [], true, true);
<a name="l00795"></a>00795             limits(1, :) = max(0, abs(leadingExtents)-holFtpt-1);
<a name="l00796"></a>00796             
<a name="l00797"></a>00797             if laggingPos(1) ~= 0
<a name="l00798"></a>00798                 this.releaseProteinFromSites([laggingPolPos(1) this.laggingStrandIndexs(1)], false);
<a name="l00799"></a>00799                 [~, ~, ~, laggingExtents] = c.isRegionAccessible([laggingPolPos(1) this.laggingStrandIndexs(1)], limits(2, 1) + holFtpt + 1, ...
<a name="l00800"></a>00800                     [], lagPolGblIdx, false, [], true, true);
<a name="l00801"></a>00801                 limits(2, 1) = max(0, abs(laggingExtents)-holFtpt-1);
<a name="l00802"></a>00802             end
<a name="l00803"></a>00803             if laggingPos(2) ~= 0
<a name="l00804"></a>00804                 this.releaseProteinFromSites([laggingPolPos(2) this.laggingStrandIndexs(2)], false);
<a name="l00805"></a>00805                 [~, ~, ~, laggingExtents] = c.isRegionAccessible([laggingPolPos(2)+holFtpt-1 this.laggingStrandIndexs(2)], -(limits(2, 2) + holFtpt + 1), ...
<a name="l00806"></a>00806                     [], lagPolGblIdx, false, [], true, true);
<a name="l00807"></a>00807                 limits(2, 2) = max(0, abs(laggingExtents)-holFtpt-1);
<a name="l00808"></a>00808             end
<a name="l00809"></a>00809             
<a name="l00810"></a>00810             %no polymerized of lagging strands past end of Okazaki fragment
<a name="l00811"></a>00811             limits(2, :) = min(limits(2, :), fLen - fPrg);
<a name="l00812"></a>00812             
<a name="l00813"></a>00813             %prevent leading strand from getting too far ahead of lagging strand
<a name="l00814"></a>00814             tmp = fPos;
<a name="l00815"></a>00815             if fPos(1) == 0; tmp(1) = chrLen; end;
<a name="l00816"></a>00816             if fPos(2) == 0; tmp(2) = 1; end;
<a name="l00817"></a>00817             limits(1, 1) = limits(1, 1) * ((tmp(1) - helicasePos(1) - helFtpt) &lt; 2 * this.okazakiFragmentMeanLength);
<a name="l00818"></a>00818             limits(1, 2) = limits(1, 2) * ((helicasePos(2) - tmp(2)) &lt; 2 * this.okazakiFragmentMeanLength);
<a name="l00819"></a>00819             
<a name="l00820"></a>00820             %no helicase progress if the <span class="keywordtype">double</span>-stranded region about the terC
<a name="l00821"></a>00821             %has non-zero linking number and would be annilated by unwinding
<a name="l00822"></a>00822             if ...
<a name="l00823"></a>00823                     helicasePos(1) + helFtpt5 &gt;= c.terCPosition + 1 &amp;&amp; ...
<a name="l00824"></a>00824                     helicasePos(1) + helFtpt5 - limits(1, 1) &lt; c.terCPosition+1 &amp;&amp; ...
<a name="l00825"></a>00825                     helicasePos(2) + helFtpt5 - 1 &lt;= c.terCPosition &amp;&amp; ...
<a name="l00826"></a>00826                     helicasePos(2) + helFtpt5 - 1 + limits(1, 2) &gt; c.terCPosition &amp;&amp; ...
<a name="l00827"></a>00827                     abs(c.linkingNumbers([min(c.terCPosition + 1, helicasePos(2) + helFtpt5 - 1) 1])) &gt; 1<a class="code" href="json_parse_8m.html#ac14d26badced2970320ee4d772ad0ae9">e</a>-6
<a name="l00828"></a>00828                 limits(1, ceil(2 * this.randStream.rand())) = 0;
<a name="l00829"></a>00829             elseif ...
<a name="l00830"></a>00830                     helicasePos(1) + helFtpt5 &gt;= c.terCPosition + 1 &amp;&amp; ...
<a name="l00831"></a>00831                     helicasePos(1) + helFtpt5 - limits(1, 1) &lt; c.terCPosition+1 &amp;&amp; ...
<a name="l00832"></a>00832                     (helicasePos(2) + helFtpt5 - 1 &gt; c.terCPosition || helicasePos(2) + helFtpt5 - 1 + limits(1, 2) &lt;= c.terCPosition) &amp;&amp; ...
<a name="l00833"></a>00833                     abs(c.linkingNumbers([min(c.terCPosition + 1, helicasePos(2) + helFtpt5 - 1) 1])) &gt; 1<a class="code" href="json_parse_8m.html#ac14d26badced2970320ee4d772ad0ae9">e</a>-6 &amp;&amp; ...
<a name="l00834"></a>00834                     ~c.isRegionPolymerized([c.terCPosition 2], 1, false)
<a name="l00835"></a>00835                 limits(1, 1) = 0;
<a name="l00836"></a>00836             elseif ...
<a name="l00837"></a>00837                     helicasePos(2) + helFtpt5 - 1 &lt;= c.terCPosition &amp;&amp; ...
<a name="l00838"></a>00838                     helicasePos(2) + helFtpt5 - 1 + limits(1, 2) &gt; c.terCPosition &amp;&amp; ...
<a name="l00839"></a>00839                     (helicasePos(1) + helFtpt5 &lt; c.terCPosition + 1 || helicasePos(1) + helFtpt5 - limits(1, 1) &gt;= c.terCPosition+1) &amp;&amp; ...
<a name="l00840"></a>00840                     abs(c.linkingNumbers([min(c.terCPosition + 1, helicasePos(2) + helFtpt5 - 1) 1])) &gt; 1<a class="code" href="json_parse_8m.html#ac14d26badced2970320ee4d772ad0ae9">e</a>-6 &amp;&amp; ...
<a name="l00841"></a>00841                     ~c.isRegionPolymerized([c.terCPosition + 1 2], 1, false)
<a name="l00842"></a>00842                 limits(1, 2) = 0;
<a name="l00843"></a>00843             end
<a name="l00844"></a>00844             
<a name="l00845"></a>00845             %pause if collision with <a class="code" href="class_r_n_a.html">RNA</a> polymerase
<a name="l00846"></a>00846             [posStrands, vals] = find(c.complexBoundSites);
<a name="l00847"></a>00847             tfs = ismembc(vals, this.complex.rnaPolymeraseIndexs);
<a name="l00848"></a>00848             if any(tfs)
<a name="l00849"></a>00849                 pos = posStrands(tfs, 1);
<a name="l00850"></a>00850                 strnds = posStrands(tfs, 2);
<a name="l00851"></a>00851                 rnaPolFtpt = c.getDNAFootprint([], this.complex.rnaPolymeraseIndexs(1));
<a name="l00852"></a>00852                 
<a name="l00853"></a>00853                 %calculate limits based on <a class="code" href="class_r_n_a.html">RNA</a> polymerase position
<a name="l00854"></a>00854                 tmpLimits = zeros(2, 2);
<a name="l00855"></a>00855                 tfs = pos &lt; helicasePos(1)   &amp; mod(strnds, 2) == 1; if any(tfs), tmpLimits(1, 1) = min(helicasePos(1) - (pos(tfs) + rnaPolFtpt));   end
<a name="l00856"></a>00856                 tfs = pos &gt; helicasePos(2)   &amp; mod(strnds, 2) == 0; if any(tfs), tmpLimits(1, 2) = min(pos(tfs) - (helicasePos(2) + helFtpt));      end
<a name="l00857"></a>00857                 tfs = pos &gt; laggingPolPos(1) &amp; mod(strnds, 2) == 0; if any(tfs), tmpLimits(2, 1) = min(pos(tfs) - (laggingPolPos(1) + holFtpt));    end
<a name="l00858"></a>00858                 tfs = pos &lt; laggingPolPos(2) &amp; mod(strnds, 2) == 1; if any(tfs), tmpLimits(2, 2) = min(laggingPolPos(2) - (pos(tfs) + rnaPolFtpt)); end
<a name="l00859"></a>00859                  
<a name="l00860"></a>00860                 %apply limits to DNA polymerase which are stalled
<a name="l00861"></a>00861                 tfs = this.randStream.random(&#39;poisson&#39;, 1 / this.rnaPolymeraseCollisionMeanDwellTime, [2 2]) == 0;
<a name="l00862"></a>00862                 limits(tfs) = min(limits(tfs), max(0, tmpLimits(tfs)));
<a name="l00863"></a>00863             end
<a name="l00864"></a>00864             
<a name="l00865"></a>00865             %no polymerization of leading strands past terC
<a name="l00866"></a>00866             %(move leading polymerization slightly past terC to make room for
<a name="l00867"></a>00867             %lagging polymerase)
<a name="l00868"></a>00868             polLimits = limits;
<a name="l00869"></a>00869             polLimits(1, 1) = min(limits(1, 1), max(0, leadingPos(1) - this.terCPosition));
<a name="l00870"></a>00870             polLimits(1, 2) = min(limits(1, 2), max(0, this.terCPosition - leadingPos(2) + 1));
<a name="l00871"></a>00871 
<a name="l00872"></a>00872             limits(1, 1) = min(limits(1, 1), max(0, leadingPos(1) - (this.terCPosition-(corFtpt5+bClmpFtpt+corFtpt5+helFtpt3))));
<a name="l00873"></a>00873             limits(1, 2) = min(limits(1, 2), max(0, (this.terCPosition+1+(corFtpt5+bClmpFtpt+corFtpt5+helFtpt3)) - leadingPos(2)));
<a name="l00874"></a>00874             
<a name="l00875"></a>00875             %dNTPs
<a name="l00876"></a>00876             sequences = <span class="keywordtype">char</span>(zeros(4, max(polLimits(:))));
<a name="l00877"></a>00877             sequences(1, 1:polLimits(1, 1)) = c.sequence(leadingPos(1) - (0:polLimits(1, 1)-1), 2)&#39;;
<a name="l00878"></a>00878             sequences(2, 1:polLimits(1, 2)) = c.sequence(leadingPos(2) + (0:polLimits(1, 2)-1), 3)&#39;;
<a name="l00879"></a>00879             sequences(3, 1:polLimits(2, 1)) = c.sequence(laggingPos(1) + (0:polLimits(2, 1)-1), 3)&#39;;
<a name="l00880"></a>00880             sequences(4, 1:polLimits(2, 2)) = c.sequence(laggingPos(2) - (0:polLimits(2, 2)-1), 2)&#39;;
<a name="l00881"></a>00881             
<a name="l00882"></a>00882             pols = min(polLimits, reshape(edu.stanford.covert.cell.sim.util.polymerize(...
<a name="l00883"></a>00883                 sequences, this.substrates(this.substrateIndexs_dntp), &#39;ACGT&#39;, &#39; &#39;, 0, 0, this.randStream), [], 2)&#39;);
<a name="l00884"></a>00884             limits(pols ~= polLimits) = pols(pols ~= polLimits);
<a name="l00885"></a>00885             polLimits = pols;
<a name="l00886"></a>00886             
<a name="l00887"></a>00887             %energy to unwind
<a name="l00888"></a>00888             unwinds = min(polLimits(1, :), max(0, [
<a name="l00889"></a>00889                 (helicasePos(1) + helFtpt5) - this.terCPosition;
<a name="l00890"></a>00890                 this.terCPosition + 1 - (helicasePos(2)+helFtpt-helFtpt5-1)]&#39;));
<a name="l00891"></a>00891             unwindLimits = min(unwinds, floor(unwinds / sum(unwinds) * min(this.substrates([this.substrateIndexs_atp; this.substrateIndexs_water]))));
<a name="l00892"></a>00892             
<a name="l00893"></a>00893             polLimits(1, unwinds ~= unwindLimits) = unwindLimits(unwinds ~= unwindLimits);
<a name="l00894"></a>00894             limits(1, unwinds ~= unwindLimits) = unwindLimits(unwinds ~= unwindLimits);
<a name="l00895"></a>00895             
<a name="l00896"></a>00896             %% Unwind DNA
<a name="l00897"></a>00897             %advance helicase
<a name="l00898"></a>00898             if ~all(this.bindProteinToChromosome([helicasePos; this.leadingStrandIndexs]&#39; , this.enzymeIndexs_helicase, ...
<a name="l00899"></a>00899                     [], [], true, false, [-1; 1] .* (limits(1,:) + 1)&#39;, true, []))
<a name="l00900"></a>00900                 throw(MException(&#39;<a class="code" href="class_replication.html">Replication</a>:error&#39;, &#39;helicases must bind&#39;));
<a name="l00901"></a>00901             end
<a name="l00902"></a>00902             
<a name="l00903"></a>00903             %unwind chromosome
<a name="l00904"></a>00904             c.setRegionUnwound(helicasePos(1)+helFtpt5,           -unwindLimits(1, 1));
<a name="l00905"></a>00905             c.setRegionUnwound(helicasePos(2)+helFtpt-helFtpt5-1,  unwindLimits(1, 2));
<a name="l00906"></a>00906             
<a name="l00907"></a>00907             %update metabolites
<a name="l00908"></a>00908             n = sum(unwindLimits(1, :));
<a name="l00909"></a>00909             this.substrates(this.substrateIndexs_atp)       = this.substrates(this.substrateIndexs_atp)       - n;
<a name="l00910"></a>00910             this.substrates(this.substrateIndexs_water)     = this.substrates(this.substrateIndexs_water)     - n;
<a name="l00911"></a>00911             this.substrates(this.substrateIndexs_adp)       = this.substrates(this.substrateIndexs_adp)       + n;
<a name="l00912"></a>00912             this.substrates(this.substrateIndexs_hydrogen)  = this.substrates(this.substrateIndexs_hydrogen)  + n;
<a name="l00913"></a>00913             this.substrates(this.substrateIndexs_phosphate) = this.substrates(this.substrateIndexs_phosphate) + n;
<a name="l00914"></a>00914             
<a name="l00915"></a>00915             %% Polymerize DNA
<a name="l00916"></a>00916             %advance DNA polymerase
<a name="l00917"></a>00917             if ~this.bindProteinToChromosome([leadingPolPos(1) this.leadingStrandIndexs(1)], leadPolLclIdx(bndLeadPolIdx(1)), [], [], true, false, -(limits(1, 1)+1), true, [])
<a name="l00918"></a>00918                 throw(MException(&#39;<a class="code" href="class_replication.html">Replication</a>:error&#39;, &#39;polymerases must bind&#39;));
<a name="l00919"></a>00919             end
<a name="l00920"></a>00920             if ~this.bindProteinToChromosome([leadingPolPos(2) this.leadingStrandIndexs(2)], leadPolLclIdx(bndLeadPolIdx(2)), [], [], true, false,  (limits(1, 2)+1), true, [])
<a name="l00921"></a>00921                 throw(MException(&#39;<a class="code" href="class_replication.html">Replication</a>:error&#39;, &#39;polymerases must bind&#39;));
<a name="l00922"></a>00922             end
<a name="l00923"></a>00923             if laggingPos(1) ~= 0
<a name="l00924"></a>00924                 if ~this.bindProteinToChromosome([laggingPolPos(1) this.laggingStrandIndexs(1)], this.enzymeIndexs_coreBetaClampPrimase, [], [], true, false, limits(2, 1)+1, true, [])
<a name="l00925"></a>00925                     throw(MException(&#39;<a class="code" href="class_replication.html">Replication</a>:error&#39;, &#39;polymerases must bind&#39;));
<a name="l00926"></a>00926                 end
<a name="l00927"></a>00927             end
<a name="l00928"></a>00928             if laggingPos(2) ~= 0
<a name="l00929"></a>00929                 if ~this.bindProteinToChromosome([laggingPolPos(2) this.laggingStrandIndexs(2)], this.enzymeIndexs_coreBetaClampPrimase, [], [], true, false, -(limits(2, 2)+1), true, [])
<a name="l00930"></a>00930                     throw(MException(&#39;<a class="code" href="class_replication.html">Replication</a>:error&#39;, &#39;polymerases must bind&#39;));
<a name="l00931"></a>00931                 end
<a name="l00932"></a>00932             end
<a name="l00933"></a>00933             
<a name="l00934"></a>00934             %set regions polymerized
<a name="l00935"></a>00935             c.setRegionPolymerized([leadingPos; 1 2]&#39;, [-1; 1] .* polLimits(1, :)&#39;);
<a name="l00936"></a>00936             c.setRegionPolymerized([laggingPos; 2 1]&#39;, [1; -1] .* polLimits(2, :)&#39;);
<a name="l00937"></a>00937             
<a name="l00938"></a>00938             %update metabolites
<a name="l00939"></a>00939             usedDNTPs = ...
<a name="l00940"></a>00940                 c.sequence.subsequenceBaseCounts(leadingPos(1) - (0:polLimits(1, 1)-1), 2) + ...
<a name="l00941"></a>00941                 c.sequence.subsequenceBaseCounts(leadingPos(2) + (0:polLimits(1, 2)-1), 3) + ...
<a name="l00942"></a>00942                 c.sequence.subsequenceBaseCounts(laggingPos(1) + (0:polLimits(2, 1)-1), 3) + ...
<a name="l00943"></a>00943                 c.sequence.subsequenceBaseCounts(laggingPos(2) - (0:polLimits(2, 2)-1), 2);
<a name="l00944"></a>00944             
<a name="l00945"></a>00945             this.substrates(this.substrateIndexs_dntp)        = this.substrates(this.substrateIndexs_dntp)        - usedDNTPs;
<a name="l00946"></a>00946             this.substrates(this.substrateIndexs_diphosphate) = this.substrates(this.substrateIndexs_diphosphate) + sum(usedDNTPs);
<a name="l00947"></a>00947         end
<a name="l00948"></a>00948         
<a name="l00949"></a>00949         %Dissociate free SSB 8mers into 2 SSB 4mers
<a name="l00950"></a>00950         function dissociateFreeSSBComplexes(this)
<a name="l00951"></a>00951             this.enzymes(this.enzymeIndexs_ssb4mer) = ...
<a name="l00952"></a>00952                 this.enzymes(this.enzymeIndexs_ssb4mer) + ...
<a name="l00953"></a>00953                 2 * this.enzymes(this.enzymeIndexs_ssb8mer);
<a name="l00954"></a>00954             this.enzymes(this.enzymeIndexs_ssb8mer) = 0;
<a name="l00955"></a>00955         end
<a name="l00956"></a>00956         
<a name="l00957"></a>00957         %Free, Bind SSBs 8mers to single-stranded DNA
<a name="l00958"></a>00958         function freeAndBindSSBs(this)
<a name="l00959"></a>00959             %% free SSBs
<a name="l00960"></a>00960             positionsStrands = this.releaseProteinFromChromosome(this.enzymeIndexs_ssb8mer, this.ssbDissociationRate, [], []);
<a name="l00961"></a>00961             
<a name="l00962"></a>00962             this.enzymes(this.enzymeIndexs_ssb4mer) = ...
<a name="l00963"></a>00963                 + this.enzymes(this.enzymeIndexs_ssb4mer) ...
<a name="l00964"></a>00964                 + 2 * size(positionsStrands, 1);
<a name="l00965"></a>00965             this.enzymes(this.enzymeIndexs_ssb8mer) = ...
<a name="l00966"></a>00966                 + this.enzymes(this.enzymeIndexs_ssb8mer) ...
<a name="l00967"></a>00967                 - size(positionsStrands, 1);
<a name="l00968"></a>00968             
<a name="l00969"></a>00969             %% bind SSBs
<a name="l00970"></a>00970             nPossibleSSB8mers = this.enzymes(this.enzymeIndexs_ssb4mer) / 2;
<a name="l00971"></a>00971             if nPossibleSSB8mers &lt; 1
<a name="l00972"></a>00972                 return;
<a name="l00973"></a>00973             end
<a name="l00974"></a>00974             
<a name="l00975"></a>00975             c = this.chromosome;
<a name="l00976"></a>00976             
<a name="l00977"></a>00977             %find accessible regions
<a name="l00978"></a>00978             [rgnPosStrnds, rgnLens] = c.getAccessibleRegions([], this.enzymeGlobalIndexs(this.enzymeIndexs_ssb8mer));
<a name="l00979"></a>00979             if isempty(rgnPosStrnds)
<a name="l00980"></a>00980                 return;
<a name="l00981"></a>00981             end
<a name="l00982"></a>00982             
<a name="l00983"></a>00983             %SSB 8mer footprint and spacing
<a name="l00984"></a>00984             ftpt = this.enzymeDNAFootprints(this.enzymeIndexs_ssb8mer);
<a name="l00985"></a>00985             spcg = this.ssbComplexSpacing;
<a name="l00986"></a>00986             
<a name="l00987"></a>00987             %exclude regions near other SSB 8mers
<a name="l00988"></a>00988             ssbPosStrnds = find(c.complexBoundSites == this.enzymeGlobalIndexs(this.enzymeIndexs_ssb8mer));
<a name="l00989"></a>00989             [rgnPosStrnds, rgnLens] = c.excludeRegions(rgnPosStrnds, rgnLens, ...
<a name="l00990"></a>00990                 [ssbPosStrnds(:, 1) - this.ssbComplexSpacing ssbPosStrnds(:, 2)], ftpt + 2*spcg);
<a name="l00991"></a>00991             if isempty(rgnPosStrnds)
<a name="l00992"></a>00992                 return;
<a name="l00993"></a>00993             end
<a name="l00994"></a>00994             
<a name="l00995"></a>00995             %split regions over terC
<a name="l00996"></a>00996             idx = reshape(find(rgnPosStrnds(:, 1) &lt;= this.terCPosition &amp; rgnPosStrnds(:, 1) + rgnLens - 1 &gt; this.terCPosition), [], 1);
<a name="l00997"></a>00997             rgnPosStrnds = [rgnPosStrnds; repmat(this.terCPosition + 1, size(idx)) rgnPosStrnds(idx, 2)];
<a name="l00998"></a>00998             rgnLens = [rgnLens; rgnLens(idx) - (this.terCPosition - rgnPosStrnds(idx, 1) + 1)];
<a name="l00999"></a>00999             rgnLens(idx) = this.terCPosition - rgnPosStrnds(idx, 1) + 1;
<a name="l01000"></a>01000             
<a name="l01001"></a>01001             %switch directions of region on far side of terC
<a name="l01002"></a>01002             tfs = rgnPosStrnds(:, 1) &gt; this.terCPosition;
<a name="l01003"></a>01003             rgnPosStrnds(tfs, 1) = rgnPosStrnds(tfs, 1) + rgnLens(tfs, 1) - ftpt;
<a name="l01004"></a>01004             rgnLens(tfs) = -rgnLens(tfs);
<a name="l01005"></a>01005             
<a name="l01006"></a>01006             %split regions into pieces large enough to bind just 1 SSB 8mer
<a name="l01007"></a>01007             tfs1 = abs(rgnLens) == ftpt;
<a name="l01008"></a>01008             tfs2 = abs(rgnLens) &gt; ftpt;
<a name="l01009"></a>01009             rgnPosStrnds = [
<a name="l01010"></a>01010                 c.splitRegions(rgnPosStrnds(tfs2, :), rgnLens(tfs2), ftpt + spcg)
<a name="l01011"></a>01011                 rgnPosStrnds(tfs1, :)];
<a name="l01012"></a>01012             if isempty(rgnPosStrnds)
<a name="l01013"></a>01013                 return;
<a name="l01014"></a>01014             end
<a name="l01015"></a>01015             
<a name="l01016"></a>01016             %bind SSB 8mers to regions
<a name="l01017"></a>01017             nBindings = min([
<a name="l01018"></a>01018                 size(rgnPosStrnds, 1);
<a name="l01019"></a>01019                 floor(nPossibleSSB8mers)]);
<a name="l01020"></a>01020             
<a name="l01021"></a>01021             this.enzymes(this.enzymeIndexs_ssb4mer) = this.enzymes(this.enzymeIndexs_ssb4mer) - 2*nBindings;
<a name="l01022"></a>01022             this.enzymes(this.enzymeIndexs_ssb8mer) = this.enzymes(this.enzymeIndexs_ssb8mer) +   nBindings;
<a name="l01023"></a>01023             
<a name="l01024"></a>01024             if nBindings ~= sum(this.bindProteinToChromosome(rgnPosStrnds, this.enzymeIndexs_ssb8mer, nBindings, [], [], false, 1, false, []))
<a name="l01025"></a>01025                 throw(MException(&#39;<a class="code" href="class_replication.html">Replication</a>:error&#39;, &#39;SSBs must bind&#39;));
<a name="l01026"></a>01026             end
<a name="l01027"></a>01027         end
<a name="l01028"></a>01028         
<a name="l01029"></a>01029         %Bind beta-clamp to start of next Okazaki fragment
<a name="l01030"></a>01030         function initiateOkazakiFragment(this)
<a name="l01031"></a>01031             %Terminate early if no active replisomes
<a name="l01032"></a>01032             if ~this.isAnyHelicaseBound || ~all(this.leadingStrandElongating)
<a name="l01033"></a>01033                 return;
<a name="l01034"></a>01034             end
<a name="l01035"></a>01035             
<a name="l01036"></a>01036             helFtpt3 = this.enzymeDNAFootprints3Prime(this.enzymeIndexs_helicase);
<a name="l01037"></a>01037             helFtpt5 = this.enzymeDNAFootprints5Prime(this.enzymeIndexs_helicase);
<a name="l01038"></a>01038             corFtpt3 = this.enzymeDNAFootprints3Prime(this.enzymeIndexs_core);
<a name="l01039"></a>01039             bClmpFtpt = this.enzymeDNAFootprints(this.enzymeIndexs_betaClamp);
<a name="l01040"></a>01040             
<a name="l01041"></a>01041             fIdx = this.okazakiFragmentIndex;
<a name="l01042"></a>01042             fPrg = this.okazakiFragmentProgress;
<a name="l01043"></a>01043             helPos = this.helicasePosition;
<a name="l01044"></a>01044             bClmpPos = this.laggingBackupBetaClampPosition;
<a name="l01045"></a>01045             strndPol = this.strandPolymerized;
<a name="l01046"></a>01046             
<a name="l01047"></a>01047             %Compute list of starting points of next Okazaki fragments that
<a name="l01048"></a>01048             %aren&#39;t yet bound by core/beta-clamps
<a name="l01049"></a>01049             posStrnds = zeros(0, 2);
<a name="l01050"></a>01050             
<a name="l01051"></a>01051             if ...
<a name="l01052"></a>01052                     fIdx(1) &lt; numel(this.primaseBindingLocations{1}) &amp;&amp; ...
<a name="l01053"></a>01053                     (fIdx(1) == 0  || (fIdx(1) &gt; 0 &amp;&amp; fPrg(1) &gt;= this.laggingBackupClampReloadingLength)) &amp;&amp; ...
<a name="l01054"></a>01054                     helPos(1) + helFtpt5 + 1 &lt; this.primaseBindingLocations{1}(fIdx(1)+1)-corFtpt3-bClmpFtpt &amp;&amp; ...
<a name="l01055"></a>01055                     bClmpPos(1) ~= this.primaseBindingLocations{1}(fIdx(1)+1)-corFtpt3-bClmpFtpt &amp;&amp; ...
<a name="l01056"></a>01056                     ~strndPol(1)
<a name="l01057"></a>01057                 
<a name="l01058"></a>01058                 posStrnds = [posStrnds; this.primaseBindingLocations{1}(fIdx(1)+1)-corFtpt3-bClmpFtpt this.laggingStrandIndexs(1)];
<a name="l01059"></a>01059             end
<a name="l01060"></a>01060             
<a name="l01061"></a>01061             <span class="keywordflow">if</span> ...
<a name="l01062"></a>01062                     fIdx(2) &lt; numel(this.primaseBindingLocations{2}) &amp;&amp; ...
<a name="l01063"></a>01063                     (fIdx(2) == 0 || (fIdx(2) &gt; 0 &amp;&amp; fPrg(2) &gt;= this.laggingBackupClampReloadingLength)) &amp;&amp; ...
<a name="l01064"></a>01064                     helPos(2)+helFtpt3 &gt; this.primaseBindingLocations{2}(fIdx(2)+1)+corFtpt3+1+bClmpFtpt &amp;&amp; ...
<a name="l01065"></a>01065                     bClmpPos(2) ~= this.primaseBindingLocations{2}(fIdx(2)+1)+corFtpt3+1 &amp;&amp; ...
<a name="l01066"></a>01066                     ~strndPol(2)
<a name="l01067"></a>01067                 
<a name="l01068"></a>01068                 posStrnds = [posStrnds; this.primaseBindingLocations{2}(fIdx(2)+1)+corFtpt3+1 this.laggingStrandIndexs(2)];
<a name="l01069"></a>01069             end
<a name="l01070"></a>01070             
<a name="l01071"></a>01071             %Bind beta-clamp to starting point of next Okazaki fragments
<a name="l01072"></a>01072             nBinding = floor(min([
<a name="l01073"></a>01073                 size(posStrnds, 1)
<a name="l01074"></a>01074                 this.substrates(this.substrateIndexs_atp)
<a name="l01075"></a>01075                 this.substrates(this.substrateIndexs_water)
<a name="l01076"></a>01076                 this.enzymes(this.enzymeIndexs_betaClampMonomer)/2]));
<a name="l01077"></a>01077             <span class="keywordflow">if</span> nBinding ~= sum(this.bindProteinToChromosome(posStrnds, this.enzymeIndexs_betaClamp, nBinding, [], <span class="keyword">true</span>, <span class="keyword">false</span>, 1, <span class="keyword">false</span>, []))
<a name="l01078"></a>01078                 throw(MException(&#39;<a class="code" href="class_replication.html">Replication</a>:error&#39;, &#39;Beta clamps must bind&#39;));
<a name="l01079"></a>01079             end
<a name="l01080"></a>01080             
<a name="l01081"></a>01081             this.enzymes(this.enzymeIndexs_betaClampMonomer) = this.enzymes(this.enzymeIndexs_betaClampMonomer) - 2 * nBinding;
<a name="l01082"></a>01082             this.enzymes(this.enzymeIndexs_betaClamp)        = this.enzymes(this.enzymeIndexs_betaClamp)        +     nBinding;
<a name="l01083"></a>01083             
<a name="l01084"></a>01084             %hydrolyze ATP used by gamma-complex to form beta-clamp
<a name="l01085"></a>01085             this.substrates(this.substrateIndexs_atp)       = this.substrates(this.substrateIndexs_atp)       - nBinding;
<a name="l01086"></a>01086             this.substrates(this.substrateIndexs_water)     = this.substrates(this.substrateIndexs_water)     - nBinding;
<a name="l01087"></a>01087             this.substrates(this.substrateIndexs_adp)       = this.substrates(this.substrateIndexs_adp)       + nBinding;
<a name="l01088"></a>01088             this.substrates(this.substrateIndexs_phosphate) = this.substrates(this.substrateIndexs_phosphate) + nBinding;
<a name="l01089"></a>01089             this.substrates(this.substrateIndexs_hydrogen)  = this.substrates(this.substrateIndexs_hydrogen)  + nBinding;
<a name="l01090"></a>01090         end
<a name="l01091"></a>01091         
<a name="l01092"></a>01092         %Terminate Okazaki fragments which have been completely synthesized
<a name="l01093"></a>01093         %- release beta-clamp
<a name="l01094"></a>01094         %- binding core to beta-clamp of next Okazaki fragment
<a name="l01095"></a>01095         function terminateOkazakiFragment(this)
<a name="l01096"></a>01096             %Terminate early if no active replisomes
<a name="l01097"></a>01097             if ~this.isAnyHelicaseBound || ~all(this.leadingStrandElongating)
<a name="l01098"></a>01098                 return;
<a name="l01099"></a>01099             end
<a name="l01100"></a>01100             
<a name="l01101"></a>01101             c = this.chromosome;
<a name="l01102"></a>01102             chrLen = c.sequenceLen;
<a name="l01103"></a>01103             fIdx = this.okazakiFragmentIndex;
<a name="l01104"></a>01104             fPrg = this.okazakiFragmentProgress;
<a name="l01105"></a>01105             fLen = this.okazakiFragmentLength;
<a name="l01106"></a>01106             helPos = this.helicasePosition;
<a name="l01107"></a>01107             helFtpt  = this.enzymeDNAFootprints(this.enzymeIndexs_helicase);
<a name="l01108"></a>01108             holFtpt  = this.enzymeDNAFootprints(this.enzymeIndexs_coreBetaClampPrimase);
<a name="l01109"></a>01109             corFtpt = this.enzymeDNAFootprints(this.enzymeIndexs_core);
<a name="l01110"></a>01110             corFtpt3 = this.enzymeDNAFootprints3Prime(this.enzymeIndexs_core);
<a name="l01111"></a>01111             corFtpt5 = this.enzymeDNAFootprints5Prime(this.enzymeIndexs_core);
<a name="l01112"></a>01112             areLaggingStrandSSBSitesBound = this.areLaggingStrandSSBSitesBound;
<a name="l01113"></a>01113             laggingBackupBetaClampPosition = this.laggingBackupBetaClampPosition;
<a name="l01114"></a>01114             leadingPolymerasePosition = this.leadingPolymerasePosition;
<a name="l01115"></a>01115             laggingPolymerasePosition = this.laggingPolymerasePosition;
<a name="l01116"></a>01116             
<a name="l01117"></a>01117             %calculate positions of core/beta-clamps that have reached end of
<a name="l01118"></a>01118             %Okazaki fragments and for which a core/beta-clamp is bound to the
<a name="l01119"></a>01119             %next Okazaki fragment
<a name="l01120"></a>01120             releasePolPosStrnds = zeros(0, 2);
<a name="l01121"></a>01121             releaseBClmpPosStrnds = zeros(0, 2);
<a name="l01122"></a>01122             bindPolPosStrnds = zeros(0, 2);
<a name="l01123"></a>01123             terminatedPosStrnds = zeros(0, 2);
<a name="l01124"></a>01124             if ...
<a name="l01125"></a>01125                     fIdx(1) &gt; 0 &amp;&amp; ...
<a name="l01126"></a>01126                     fPrg(1) == fLen(1) &amp;&amp; ...
<a name="l01127"></a>01127                     areLaggingStrandSSBSitesBound(1) &amp;&amp; ...
<a name="l01128"></a>01128                     (fIdx(1) == numel(this.primaseBindingLocations{1}) || (...
<a name="l01129"></a>01129                     (fIdx(1) == numel(this.primaseBindingLocations{1}) - 1 || ...
<a name="l01130"></a>01130                     this.primaseBindingLocations{1}(fIdx(1)+1)-(helPos(1)+helFtpt) &gt; this.startingOkazakiLoopLength) &amp;&amp; ...
<a name="l01131"></a>01131                     laggingBackupBetaClampPosition(1) == this.primaseBindingLocations{1}(fIdx(1)+1)-(holFtpt-corFtpt5)+1))
<a name="l01132"></a>01132                 
<a name="l01133"></a>01133                 releasePolPosStrnds = [releasePolPosStrnds;
<a name="l01134"></a>01134                     laggingPolymerasePosition(1) this.laggingStrandIndexs(1)];
<a name="l01135"></a>01135                 if fIdx(1) &lt; numel(this.primaseBindingLocations{1})
<a name="l01136"></a>01136                     releaseBClmpPosStrnds = [releaseBClmpPosStrnds;
<a name="l01137"></a>01137                         laggingBackupBetaClampPosition(1) this.laggingStrandIndexs(1)];
<a name="l01138"></a>01138                     bindPolPosStrnds = [bindPolPosStrnds;
<a name="l01139"></a>01139                         laggingBackupBetaClampPosition(1) this.laggingStrandIndexs(1)];
<a name="l01140"></a>01140                 else
<a name="l01141"></a>01141                     terminatedPosStrnds = [terminatedPosStrnds;
<a name="l01142"></a>01142                         leadingPolymerasePosition(1) this.leadingStrandIndexs(1)];
<a name="l01143"></a>01143                 end
<a name="l01144"></a>01144                 
<a name="l01145"></a>01145                 if fIdx(1) == 1
<a name="l01146"></a>01146                     c.strandBreaks(chrLen, 3) = 1;
<a name="l01147"></a>01147                 else
<a name="l01148"></a>01148                     c.strandBreaks(this.primaseBindingLocations{1}(fIdx(1)-1)-1, 3) = 1;
<a name="l01149"></a>01149                 end
<a name="l01150"></a>01150             end
<a name="l01151"></a>01151             <span class="keywordflow">if</span> ...
<a name="l01152"></a>01152                     fIdx(2) &gt; 0 &amp;&amp; ...
<a name="l01153"></a>01153                     fPrg(2) == fLen(2) &amp;&amp; ...
<a name="l01154"></a>01154                     areLaggingStrandSSBSitesBound(2) &amp;&amp; ...
<a name="l01155"></a>01155                     (fIdx(2) == numel(this.primaseBindingLocations{2}) || (...
<a name="l01156"></a>01156                     (fIdx(2) == numel(this.primaseBindingLocations{2})-1 || ...
<a name="l01157"></a>01157                     helPos(2) - this.primaseBindingLocations{2}(fIdx(2)+1) &gt; this.startingOkazakiLoopLength) &amp;&amp; ...
<a name="l01158"></a>01158                     laggingBackupBetaClampPosition(2) == this.primaseBindingLocations{2}(fIdx(2)+1)+corFtpt3+1))
<a name="l01159"></a>01159                 
<a name="l01160"></a>01160                 releasePolPosStrnds = [releasePolPosStrnds;
<a name="l01161"></a>01161                     laggingPolymerasePosition(2) this.laggingStrandIndexs(2)];
<a name="l01162"></a>01162                 if fIdx(2) &lt; numel(this.primaseBindingLocations{2})
<a name="l01163"></a>01163                     releaseBClmpPosStrnds = [releaseBClmpPosStrnds;
<a name="l01164"></a>01164                         laggingBackupBetaClampPosition(2) this.laggingStrandIndexs(2)];
<a name="l01165"></a>01165                     bindPolPosStrnds = [bindPolPosStrnds;
<a name="l01166"></a>01166                         laggingBackupBetaClampPosition(2)-corFtpt this.laggingStrandIndexs(2)];
<a name="l01167"></a>01167                 else
<a name="l01168"></a>01168                     terminatedPosStrnds = [terminatedPosStrnds;
<a name="l01169"></a>01169                         leadingPolymerasePosition(2) this.leadingStrandIndexs(2)];
<a name="l01170"></a>01170                 end
<a name="l01171"></a>01171                 
<a name="l01172"></a>01172                 if fIdx(2) == 1
<a name="l01173"></a>01173                     c.strandBreaks(chrLen, 2) = 1;
<a name="l01174"></a>01174                 else
<a name="l01175"></a>01175                     c.strandBreaks(this.primaseBindingLocations{2}(fIdx(2)-1), 2) = 1;
<a name="l01176"></a>01176                 end
<a name="l01177"></a>01177             end
<a name="l01178"></a>01178             
<a name="l01179"></a>01179             %release the core/beta-clamp that have reached the end of the Okazki
<a name="l01180"></a>01180             %fragment
<a name="l01181"></a>01181             this.releaseProteinFromSites(releasePolPosStrnds, false);
<a name="l01182"></a>01182             this.enzymes(this.enzymeIndexs_coreBetaClampPrimase) = this.enzymes(this.enzymeIndexs_coreBetaClampPrimase)    -   size(releasePolPosStrnds, 1);
<a name="l01183"></a>01183             this.enzymes(this.enzymeIndexs_primase)          = this.enzymes(this.enzymeIndexs_primase)          +   size(releasePolPosStrnds, 1);
<a name="l01184"></a>01184             this.enzymes(this.enzymeIndexs_core)             = this.enzymes(this.enzymeIndexs_core)             +   size(releasePolPosStrnds, 1);
<a name="l01185"></a>01185             this.enzymes(this.enzymeIndexs_betaClampMonomer) = this.enzymes(this.enzymeIndexs_betaClampMonomer) + 2*size(releasePolPosStrnds, 1);
<a name="l01186"></a>01186             
<a name="l01187"></a>01187             this.releaseProteinFromSites(releaseBClmpPosStrnds, <span class="keyword">false</span>);
<a name="l01188"></a>01188             this.enzymes(this.enzymeIndexs_betaClamp)        = this.enzymes(this.enzymeIndexs_betaClamp)        -   size(releaseBClmpPosStrnds, 1);
<a name="l01189"></a>01189             this.enzymes(this.enzymeIndexs_betaClampMonomer) = this.enzymes(this.enzymeIndexs_betaClampMonomer) + 2*size(releaseBClmpPosStrnds, 1);
<a name="l01190"></a>01190             
<a name="l01191"></a>01191             <span class="keywordflow">if</span> ~all(this.bindProteinToChromosome(bindPolPosStrnds, this.enzymeIndexs_coreBetaClampPrimase, [], [], <span class="keyword">true</span>, <span class="keyword">false</span>, 1, <span class="keyword">false</span>, []))
<a name="l01192"></a>01192                 throw(MException(&#39;<a class="code" href="class_replication.html">Replication</a>:error&#39;, &#39;lagging strand polymerases must bind&#39;));
<a name="l01193"></a>01193             end
<a name="l01194"></a>01194             this.enzymes(this.enzymeIndexs_primase)              = this.enzymes(this.enzymeIndexs_primase)              -   size(bindPolPosStrnds, 1);
<a name="l01195"></a>01195             this.enzymes(this.enzymeIndexs_core)                 = this.enzymes(this.enzymeIndexs_core)                 -   size(bindPolPosStrnds, 1);
<a name="l01196"></a>01196             this.enzymes(this.enzymeIndexs_betaClampMonomer)     = this.enzymes(this.enzymeIndexs_betaClampMonomer)     - 2*size(bindPolPosStrnds, 1);
<a name="l01197"></a>01197             this.enzymes(this.enzymeIndexs_coreBetaClampPrimase) = this.enzymes(this.enzymeIndexs_coreBetaClampPrimase) +   size(bindPolPosStrnds, 1);
<a name="l01198"></a>01198             
<a name="l01199"></a>01199             if ~isempty(terminatedPosStrnds)
<a name="l01200"></a>01200                 n = size(terminatedPosStrnds, 1);
<a name="l01201"></a>01201                 this.modifyProteinOnChromosome(terminatedPosStrnds, this.enzymeIndexs_2coreBetaClampGammaComplexPrimase(ones(n, 1)));
<a name="l01202"></a>01202                 
<a name="l01203"></a>01203                 this.enzymes(this.enzymeIndexs_coreBetaClampGammaComplex)    = this.enzymes(this.enzymeIndexs_coreBetaClampGammaComplex)  - n;
<a name="l01204"></a>01204                 this.enzymes(this.enzymeIndexs_core)                         = this.enzymes(this.enzymeIndexs_core)                       + n;
<a name="l01205"></a>01205                 this.enzymes(this.enzymeIndexs_betaClampMonomer)             = this.enzymes(this.enzymeIndexs_betaClampMonomer)           + 2*n;
<a name="l01206"></a>01206                 this.enzymes(this.enzymeIndexs_gammaComplex)                 = this.enzymes(this.enzymeIndexs_gammaComplex)               + n;
<a name="l01207"></a>01207                 
<a name="l01208"></a>01208                 this.enzymes(this.enzymeIndexs_2coreBetaClampGammaComplexPrimase)   = this.enzymes(this.enzymeIndexs_2coreBetaClampGammaComplexPrimase) + n;
<a name="l01209"></a>01209                 this.enzymes(this.enzymeIndexs_primase)                             = this.enzymes(this.enzymeIndexs_primase)                    - n;
<a name="l01210"></a>01210                 this.enzymes(this.enzymeIndexs_core)                                = this.enzymes(this.enzymeIndexs_core)                       - 2*n;
<a name="l01211"></a>01211                 this.enzymes(this.enzymeIndexs_betaClampMonomer)                    = this.enzymes(this.enzymeIndexs_betaClampMonomer)           - 2*n;
<a name="l01212"></a>01212                 this.enzymes(this.enzymeIndexs_gammaComplex)                        = this.enzymes(this.enzymeIndexs_gammaComplex)               - n;
<a name="l01213"></a>01213             end
<a name="l01214"></a>01214         end
<a name="l01215"></a>01215         
<a name="l01216"></a>01216         %Ligate DNA
<a name="l01217"></a>01217         %(2) DR5P + NAD ==&gt; AMP + dRibose5P_dRibose5P + H + NMN
<a name="l01218"></a>01218         %
<a name="l01219"></a>01219         %<a class="code" href="class_note.html">Note</a>: this function is redundant with evolveState_Ligate in DNARepair
<a name="l01220"></a>01220         function ligateDNA(this)
<a name="l01221"></a>01221             c = this.chromosome;
<a name="l01222"></a>01222             
<a name="l01223"></a>01223             positionsStrands = find(c.singleStrandBreaks);
<a name="l01224"></a>01224             if isempty(positionsStrands)
<a name="l01225"></a>01225                 return;
<a name="l01226"></a>01226             end
<a name="l01227"></a>01227                 
<a name="l01228"></a>01228             numReactions = max(0, floor(min([
<a name="l01229"></a>01229                 size(positionsStrands, 1);
<a name="l01230"></a>01230                 this.randStream.stochasticRound(this.enzymes(this.enzymeIndexs_ligase) * this.stepSizeSec * this.ligaseRate);
<a name="l01231"></a>01231                 this.substrates(this.substrateIndexs_nad)])));
<a name="l01232"></a>01232             if numReactions == 0
<a name="l01233"></a>01233                 return;
<a name="l01234"></a>01234             end
<a name="l01235"></a>01235             
<a name="l01236"></a>01236             [~, ~, positionsStrands, numReactions] = this.bindProteinToChromosome(positionsStrands, ...
<a name="l01237"></a>01237                 this.enzymeIndexs_ligase, numReactions, [], false, true, 1, false, [1 2]);
<a name="l01238"></a>01238             if numReactions == 0
<a name="l01239"></a>01239                 return;
<a name="l01240"></a>01240             end
<a name="l01241"></a>01241             
<a name="l01242"></a>01242             c.strandBreaks(positionsStrands) = 0;
<a name="l01243"></a>01243             
<a name="l01244"></a>01244             this.substrates(this.substrateIndexs_nad)      = this.substrates(this.substrateIndexs_nad)      - numReactions;
<a name="l01245"></a>01245             this.substrates(this.substrateIndexs_nmn)      = this.substrates(this.substrateIndexs_nmn)      + numReactions;
<a name="l01246"></a>01246             this.substrates(this.substrateIndexs_amp)      = this.substrates(this.substrateIndexs_amp)      + numReactions;
<a name="l01247"></a>01247             this.substrates(this.substrateIndexs_hydrogen) = this.substrates(this.substrateIndexs_hydrogen) + numReactions;
<a name="l01248"></a>01248         end
<a name="l01249"></a>01249         
<a name="l01250"></a>01250         %Terminate replication if replisomes have reached terC (meaning the
<a name="l01251"></a>01251         %chromosome has been duplicated)
<a name="l01252"></a>01252         %- dissolve replisomes
<a name="l01253"></a>01253         function terminateReplication(this)
<a name="l01254"></a>01254             if ...
<a name="l01255"></a>01255                     ~all(this.leadingStrandElongating) || ...  %Terminate early if no active replisomes
<a name="l01256"></a>01256                     ~all(this.strandPolymerized)       || ...  %check if replisomes have reached terC
<a name="l01257"></a>01257                     any(this.laggingStrandElongating)          %check if last Okazaki fragment already terminated
<a name="l01258"></a>01258                 return;
<a name="l01259"></a>01259             end
<a name="l01260"></a>01260             
<a name="l01261"></a>01261             %set strand break
<a name="l01262"></a>01262             this.chromosome.strandBreaks(this.terCPosition, 2:3) = 1;
<a name="l01263"></a>01263             
<a name="l01264"></a>01264             %dissolve replisomes
<a name="l01265"></a>01265             if size(this.releaseProteinFromChromosome(this.enzymeIndexs_helicase, Inf, [], []), 1) ~= 2
<a name="l01266"></a>01266                 throw(MException(&#39;<a class="code" href="class_replication.html">Replication</a>:error&#39;, &#39;Two helicases must be released&#39;));
<a name="l01267"></a>01267             end
<a name="l01268"></a>01268             if size(this.releaseProteinFromChromosome(this.enzymeIndexs_2coreBetaClampGammaComplexPrimase, Inf, [], []), 1) ~= 2
<a name="l01269"></a>01269                 throw(MException(&#39;<a class="code" href="class_replication.html">Replication</a>:error&#39;, &#39;Two polymerase complexes must be released&#39;));
<a name="l01270"></a>01270             end
<a name="l01271"></a>01271             
<a name="l01272"></a>01272             this.enzymes = this.enzymes + 2 * this.enzymeComposition(:, this.enzymeIndexs_2coreBetaClampGammaComplexPrimase);
<a name="l01273"></a>01273             this.enzymes(this.enzymeIndexs_2coreBetaClampGammaComplexPrimase) = this.enzymes(this.enzymeIndexs_2coreBetaClampGammaComplexPrimase) - 2;
<a name="l01274"></a>01274         end
<a name="l01275"></a>01275     end
<a name="l01276"></a>01276     
<a name="l01277"></a>01277     %get methods of dependent local state
<a name="l01278"></a>01278     methods
<a name="l01279"></a>01279         function value = get.isDnaAORIComplexAssembled(this)
<a name="l01280"></a>01280             c = this.chromosome;
<a name="l01281"></a>01281             [posStrnds, complexs] = find(c.complexBoundSites);
<a name="l01282"></a>01282             
<a name="l01283"></a>01283             %check if any DnaA-ATP 7mers bound
<a name="l01284"></a>01284             idxs7 = find(complexs == this.complexIndexs_DnaA_7mer_ATP);
<a name="l01285"></a>01285             if length(idxs7) &lt; 4
<a name="l01286"></a>01286                 value = false;
<a name="l01287"></a>01287                 return;
<a name="l01288"></a>01288             end
<a name="l01289"></a>01289             
<a name="l01290"></a>01290             %check if DnaA-ATP 1mer bound at R5 box on chromosome 1
<a name="l01291"></a>01291             idxs1 = find(posStrnds(:, 1) == this.dnaAFunctionalBoxStartPositions(this.dnaAFunctionalBoxIndexs_R5));
<a name="l01292"></a>01292             if ~any(complexs(idxs1) == this.complexIndexs_DnaA_1mer_ATP &amp; posStrnds(idxs1, 2) == 1)
<a name="l01293"></a>01293                 value = false;
<a name="l01294"></a>01294                 return;
<a name="l01295"></a>01295             end
<a name="l01296"></a>01296             
<a name="l01297"></a>01297             %check if DnaA-ATP 7mers bound at R1-4 boxes on chromosome 1
<a name="l01298"></a>01298             value = all(ismembc(this.dnaAFunctionalBoxStartPositions(this.dnaAFunctionalBoxIndexs_R1234, 1), posStrnds(idxs7(posStrnds(idxs7, 2) == 1), 1)));
<a name="l01299"></a>01299         end
<a name="l01300"></a>01300         
<a name="l01301"></a>01301         function result = get.isAnyHelicaseBound(this)
<a name="l01302"></a>01302             [~, complexs] = find(this.chromosome.complexBoundSites);
<a name="l01303"></a>01303             result = any(complexs == this.enzymeGlobalIndexs(this.enzymeIndexs_helicase));
<a name="l01304"></a>01304         end
<a name="l01305"></a>01305         
<a name="l01306"></a>01306         function result = get.isAnyPolymeraseBound(this)
<a name="l01307"></a>01307             [~, complexs] = find(this.chromosome.complexBoundSites);
<a name="l01308"></a>01308             result = ...
<a name="l01309"></a>01309                 any(complexs == this.enzymeGlobalIndexs(this.enzymeIndexs_2coreBetaClampGammaComplexPrimase)) || ...
<a name="l01310"></a>01310                 any(complexs == this.enzymeGlobalIndexs(this.enzymeIndexs_coreBetaClampGammaComplex)) || ...
<a name="l01311"></a>01311                 any(complexs == this.enzymeGlobalIndexs(this.enzymeIndexs_coreBetaClampPrimase));
<a name="l01312"></a>01312         end
<a name="l01313"></a>01313         
<a name="l01314"></a>01314         function result = get.leadingStrandElongating(this)
<a name="l01315"></a>01315             result = ...
<a name="l01316"></a>01316                 this.helicasePosition &amp; ...
<a name="l01317"></a>01317                 this.leadingPolymerasePosition;
<a name="l01318"></a>01318         end
<a name="l01319"></a>01319         
<a name="l01320"></a>01320         function result = get.laggingStrandElongating(this)
<a name="l01321"></a>01321             result = this.laggingPolymerasePosition ~= 0;
<a name="l01322"></a>01322         end
<a name="l01323"></a>01323         
<a name="l01324"></a>01324         function result = get.leadingStrandPolymerized(this)
<a name="l01325"></a>01325             c = this.chromosome;
<a name="l01326"></a>01326             result = false(1, 2);
<a name="l01327"></a>01327             
<a name="l01328"></a>01328             [pos2, lens2] = find(c.polymerizedRegions(:, 2));
<a name="l01329"></a>01329             [pos4, lens4] = find(c.polymerizedRegions(:, 4));
<a name="l01330"></a>01330             result(1) = ...
<a name="l01331"></a>01331                 any(pos2(:,1) &lt;= this.terCPosition + 1 &amp; pos2(:,1) + lens2 - 1 == c.sequenceLen) &amp;&amp; ...
<a name="l01332"></a>01332                 any(pos4(:,1) &lt;= this.terCPosition + 1 &amp; pos4(:,1) + lens4 - 1 == c.sequenceLen);
<a name="l01333"></a>01333             
<a name="l01334"></a>01334             result(2) = ...
<a name="l01335"></a>01335                 all(c.polymerizedRegions([1 1; 1 3]) &gt;= this.terCPosition);
<a name="l01336"></a>01336         end
<a name="l01337"></a>01337         
<a name="l01338"></a>01338         function result = get.laggingStrandPolymerized(this)
<a name="l01339"></a>01339             c = this.chromosome;
<a name="l01340"></a>01340             result = false(1, 2);
<a name="l01341"></a>01341             
<a name="l01342"></a>01342             [pos1, lens1] = find(c.polymerizedRegions(:, 1));
<a name="l01343"></a>01343             [pos3, lens3] = find(c.polymerizedRegions(:, 3));
<a name="l01344"></a>01344             result(1) = ...
<a name="l01345"></a>01345                 any(pos1(:,1) &lt;= this.terCPosition + 1 &amp; pos1(:,1) + lens1 - 1 == c.sequenceLen) &amp;&amp; ...
<a name="l01346"></a>01346                 any(pos3(:,1) &lt;= this.terCPosition + 1 &amp; pos3(:,1) + lens3 - 1 == c.sequenceLen);
<a name="l01347"></a>01347             
<a name="l01348"></a>01348             result(2) = ...
<a name="l01349"></a>01349                 all(c.polymerizedRegions([1 2; 1 4]) &gt;= this.terCPosition);
<a name="l01350"></a>01350         end
<a name="l01351"></a>01351         
<a name="l01352"></a>01352         function result = get.strandPolymerized(this)
<a name="l01353"></a>01353             result = ...
<a name="l01354"></a>01354                 this.leadingStrandPolymerized &amp; ...
<a name="l01355"></a>01355                 this.laggingStrandPolymerized;
<a name="l01356"></a>01356         end
<a name="l01357"></a>01357         
<a name="l01358"></a>01358         function result = get.numLigations(this)
<a name="l01359"></a>01359             fIdx = this.okazakiFragmentIndex;
<a name="l01360"></a>01360             fPrg = this.okazakiFragmentProgress;
<a name="l01361"></a>01361             
<a name="l01362"></a>01362             pos1 = [];
<a name="l01363"></a>01363             pos2 = [];
<a name="l01364"></a>01364             
<a name="l01365"></a>01365             if fIdx(1) &gt; 0; pos1 = this.primaseBindingLocations{1}(1:fIdx(1)-2)-1; end;
<a name="l01366"></a>01366             <span class="keywordflow">if</span> fIdx(2) &gt; 0; pos2 = this.primaseBindingLocations{2}(1:fIdx(2)-2)  ; end;
<a name="l01367"></a>01367             
<a name="l01368"></a>01368             <span class="keywordflow">if</span> fIdx(1) &gt; 1
<a name="l01369"></a>01369                 pos1 = [pos1; 0];
<a name="l01370"></a>01370             end
<a name="l01371"></a>01371             <span class="keywordflow">if</span> fIdx(2) &gt; 1
<a name="l01372"></a>01372                 pos2 = [pos2; 0];
<a name="l01373"></a>01373             end
<a name="l01374"></a>01374             <span class="keywordflow">if</span> fIdx(1) == numel(this.primaseBindingLocations{1}) &amp;&amp; fPrg(1) &gt; 0
<a name="l01375"></a>01375                 pos1 = [pos1; this.terCPosition];
<a name="l01376"></a>01376             end
<a name="l01377"></a>01377             <span class="keywordflow">if</span> fIdx(2) == numel(this.primaseBindingLocations{2}) &amp;&amp; fPrg(2) &gt; 0
<a name="l01378"></a>01378                 pos2 = [pos2; this.terCPosition];
<a name="l01379"></a>01379             end
<a name="l01380"></a>01380             
<a name="l01381"></a>01381             strandPolymerized = this.strandPolymerized;
<a name="l01382"></a>01382             leadingPosition = this.leadingPosition;
<a name="l01383"></a>01383             <span class="keywordflow">if</span> strandPolymerized(1)
<a name="l01384"></a>01384                 pos1 = [0; this.primaseBindingLocations{1}(1:end)-1];
<a name="l01385"></a>01385             end
<a name="l01386"></a>01386             <span class="keywordflow">if</span> strandPolymerized(2) &amp;&amp; (strandPolymerized(1) || any(leadingPosition))
<a name="l01387"></a>01387                 pos2 = [0; this.primaseBindingLocations{2}(1:end)];
<a name="l01388"></a>01388             end
<a name="l01389"></a>01389             
<a name="l01390"></a>01390             result = zeros(1, 2);
<a name="l01391"></a>01391             result(1) = numel(pos1) - nnz(this.chromosome.strandBreaks(pos1, 3));
<a name="l01392"></a>01392             result(2) = numel(pos2) - nnz(this.chromosome.strandBreaks(pos2, 2));
<a name="l01393"></a>01393         end
<a name="l01394"></a>01394         
<a name="l01395"></a>01395         %Every Okazaki fragment needs to have been ligated on one end, and
<a name="l01396"></a>01396         %the last one on each strand needs to have been ligated on both
<a name="l01397"></a>01397         %ends.
<a name="l01398"></a>01398         <span class="keyword">function</span> result = <span class="keyword">get</span>.strandLigated(<span class="keyword">this</span>)
<a name="l01399"></a>01399             result = this.numLigations == ...
<a name="l01400"></a>01400                 [numel(this.primaseBindingLocations{1}) + 1 ...
<a name="l01401"></a>01401                 numel(this.primaseBindingLocations{2}) + 1];
<a name="l01402"></a>01402         end
<a name="l01403"></a>01403         
<a name="l01404"></a>01404         <span class="keyword">function</span> result = <span class="keyword">get</span>.strandDuplicated(<span class="keyword">this</span>)
<a name="l01405"></a>01405             result = ...
<a name="l01406"></a>01406                 this.strandPolymerized &amp; ...
<a name="l01407"></a>01407                 this.strandLigated &amp; ...
<a name="l01408"></a>01408                 this.helicasePosition == 0 &amp; ...
<a name="l01409"></a>01409                 this.leadingPosition == 0 &amp; ...
<a name="l01410"></a>01410                 this.laggingPosition == 0 &amp;  ...
<a name="l01411"></a>01411                 this.laggingBackupBetaClampPosition == 0;
<a name="l01412"></a>01412         end
<a name="l01413"></a>01413         
<a name="l01414"></a>01414         %helicase position
<a name="l01415"></a>01415         <span class="keyword">function</span> result = <span class="keyword">get</span>.helicasePosition(<span class="keyword">this</span>)
<a name="l01416"></a>01416             positionsStrands = find(this.chromosome.complexBoundSites == <span class="keyword">this</span>.enzymeGlobalIndexs(<span class="keyword">this</span>.enzymeIndexs_helicase));
<a name="l01417"></a>01417             
<a name="l01418"></a>01418             idxs1 = find(positionsStrands(:, 2) == this.leadingStrandIndexs(1));
<a name="l01419"></a>01419             idxs2 = find(positionsStrands(:, 2) == this.leadingStrandIndexs(2));
<a name="l01420"></a>01420             
<a name="l01421"></a>01421             <span class="keywordflow">if</span> numel(idxs1) &gt; 1 || numel(idxs2) &gt; 1
<a name="l01422"></a>01422                 <span class="keywordflow">throw</span>(MException(<span class="stringliteral">&#39;Replication:error&#39;</span>, <span class="stringliteral">&#39;More than 1 active replisome&#39;</span>));
<a name="l01423"></a>01423             end
<a name="l01424"></a>01424             
<a name="l01425"></a>01425             result = zeros(1, 2);
<a name="l01426"></a>01426             <span class="keywordflow">if</span> ~isempty(idxs1)
<a name="l01427"></a>01427                 result(1) = positionsStrands(idxs1, 1);
<a name="l01428"></a>01428             end
<a name="l01429"></a>01429             if ~isempty(idxs2)
<a name="l01430"></a>01430                 result(2) = positionsStrands(idxs2, 1);
<a name="l01431"></a>01431             end
<a name="l01432"></a>01432         end
<a name="l01433"></a>01433         
<a name="l01434"></a>01434         function result = get.leadingPolymerasePosition(this)
<a name="l01435"></a>01435             polLclIdxs = [this.enzymeIndexs_coreBetaClampGammaComplex; this.enzymeIndexs_2coreBetaClampGammaComplexPrimase];
<a name="l01436"></a>01436             [positionsStrands, complexs] = find(this.chromosome.complexBoundSites);
<a name="l01437"></a>01437             tfs = complexs == this.enzymeGlobalIndexs(polLclIdxs(1)) | complexs == this.enzymeGlobalIndexs(polLclIdxs(2));
<a name="l01438"></a>01438             positionsStrands = positionsStrands(tfs, :);
<a name="l01439"></a>01439             
<a name="l01440"></a>01440             result = zeros(1, 2);
<a name="l01441"></a>01441             
<a name="l01442"></a>01442             idx = find(positionsStrands(:, 2) == this.leadingStrandIndexs(1));
<a name="l01443"></a>01443             if numel(idx) &gt; 1
<a name="l01444"></a>01444                 throw(MException(&#39;<a class="code" href="class_replication.html">Replication</a>:error&#39;, &#39;More than 1 active replisome&#39;));
<a name="l01445"></a>01445             elseif numel(idx) == 1
<a name="l01446"></a>01446                 result(1) = positionsStrands(idx, 1);
<a name="l01447"></a>01447             end
<a name="l01448"></a>01448             
<a name="l01449"></a>01449             idx = find(positionsStrands(:, 2) == this.leadingStrandIndexs(2));
<a name="l01450"></a>01450             if numel(idx) &gt; 1
<a name="l01451"></a>01451                 throw(MException(&#39;<a class="code" href="class_replication.html">Replication</a>:error&#39;, &#39;More than 1 active replisome&#39;));
<a name="l01452"></a>01452             elseif numel(idx) == 1
<a name="l01453"></a>01453                 result(2) = positionsStrands(idx, 1);
<a name="l01454"></a>01454             end
<a name="l01455"></a>01455         end
<a name="l01456"></a>01456         
<a name="l01457"></a>01457         function result = get.laggingPolymerasePosition(this)
<a name="l01458"></a>01458             positionsStrands = find(this.chromosome.complexBoundSites == this.enzymeGlobalIndexs(this.enzymeIndexs_coreBetaClampPrimase));
<a name="l01459"></a>01459             idxs1 = find(positionsStrands(:, 2) == this.laggingStrandIndexs(1));
<a name="l01460"></a>01460             idxs2 = find(positionsStrands(:, 2) == this.laggingStrandIndexs(2));
<a name="l01461"></a>01461             
<a name="l01462"></a>01462             result = zeros(1, 2);
<a name="l01463"></a>01463             if ~isempty(idxs1); result(1) = max(positionsStrands(idxs1, 1)); end;
<a name="l01464"></a>01464             if ~isempty(idxs2); result(2) = min(positionsStrands(idxs2, 1)); end
<a name="l01465"></a>01465             
<a name="l01466"></a>01466             if numel(idxs1) &gt; 2 || numel(idxs2) &gt; 2
<a name="l01467"></a>01467                 throw(MException(&#39;<a class="code" href="class_replication.html">Replication</a>:error&#39;, &#39;More than 1 active replisome&#39;));
<a name="l01468"></a>01468             end
<a name="l01469"></a>01469         end
<a name="l01470"></a>01470         
<a name="l01471"></a>01471         function result = get.leadingPosition(this)
<a name="l01472"></a>01472             holFtpt  = this.enzymeDNAFootprints(this.enzymeIndexs_coreBetaClampGammaComplex);
<a name="l01473"></a>01473             corFtpt5 = this.enzymeDNAFootprints5Prime(this.enzymeIndexs_core);
<a name="l01474"></a>01474             
<a name="l01475"></a>01475             result  = this.leadingPolymerasePosition;
<a name="l01476"></a>01476             if result(1) ~= 0; result(1) = result(1) + corFtpt5;               end;
<a name="l01477"></a>01477             if result(2) ~= 0; result(2) = result(2) + holFtpt - corFtpt5 - 1; end;
<a name="l01478"></a>01478             
<a name="l01479"></a>01479             result(result ~= 0) = mod(result(result ~= 0) - 1, this.chromosome.sequenceLen) + 1;
<a name="l01480"></a>01480         end
<a name="l01481"></a>01481         
<a name="l01482"></a>01482         function result = get.laggingPosition(this)
<a name="l01483"></a>01483             holFtpt  = this.enzymeDNAFootprints(this.enzymeIndexs_coreBetaClampGammaComplex);
<a name="l01484"></a>01484             corFtpt5 = this.enzymeDNAFootprints5Prime(this.enzymeIndexs_core);
<a name="l01485"></a>01485             
<a name="l01486"></a>01486             result = this.laggingPolymerasePosition;
<a name="l01487"></a>01487             if result(1) ~= 0; result(1) = result(1) + holFtpt - corFtpt5 - 1; end;
<a name="l01488"></a>01488             if result(2) ~= 0; result(2) = result(2) + corFtpt5;               end;
<a name="l01489"></a>01489             
<a name="l01490"></a>01490             result(result ~= 0) = mod(result(result ~= 0) - 1, this.chromosome.sequenceLen) + 1;
<a name="l01491"></a>01491         end
<a name="l01492"></a>01492         
<a name="l01493"></a>01493         function result = get.laggingBackupBetaClampPosition(this)
<a name="l01494"></a>01494             positionsStrands = find(this.chromosome.complexBoundSites == this.enzymeGlobalIndexs(this.enzymeIndexs_betaClamp));
<a name="l01495"></a>01495             idxs1 = find(positionsStrands(:, 2) == this.laggingStrandIndexs(1));
<a name="l01496"></a>01496             idxs2 = find(positionsStrands(:, 2) == this.laggingStrandIndexs(2));
<a name="l01497"></a>01497             
<a name="l01498"></a>01498             result = zeros(1, 2);
<a name="l01499"></a>01499             if ~isempty(idxs1); result(1) = max(positionsStrands(idxs1, 1)); end;
<a name="l01500"></a>01500             if ~isempty(idxs2); result(2) = min(positionsStrands(idxs2, 1)); end
<a name="l01501"></a>01501             
<a name="l01502"></a>01502             if numel(idxs1) &gt; 1 || numel(idxs2) &gt; 1
<a name="l01503"></a>01503                 throw(MException(&#39;<a class="code" href="class_replication.html">Replication</a>:error&#39;, &#39;More than 1 active replisome&#39;));
<a name="l01504"></a>01504             end
<a name="l01505"></a>01505         end
<a name="l01506"></a>01506         
<a name="l01507"></a>01507         function result = get.okazakiFragmentIndex(this)
<a name="l01508"></a>01508             c = this.chromosome;
<a name="l01509"></a>01509             laggingPos = this.laggingPosition;
<a name="l01510"></a>01510             
<a name="l01511"></a>01511             adjLaggingPos = laggingPos;
<a name="l01512"></a>01512             if laggingPos(1) ~= 0 &amp;&amp; laggingPos(1) &lt; 1/2 * this.terCPosition
<a name="l01513"></a>01513                 adjLaggingPos(1) = laggingPos(1) + this.chromosome.sequenceLen;
<a name="l01514"></a>01514             end
<a name="l01515"></a>01515             if laggingPos(2) ~= 0 &amp;&amp; laggingPos(2) &gt; 3/2 * this.terCPosition
<a name="l01516"></a>01516                 adjLaggingPos(2) = laggingPos(2) - this.chromosome.sequenceLen;
<a name="l01517"></a>01517             end
<a name="l01518"></a>01518             
<a name="l01519"></a>01519             result = zeros(1, 2);
<a name="l01520"></a>01520             
<a name="l01521"></a>01521             idx1 = find(this.primaseBindingLocations{1} &lt;= adjLaggingPos(1) - (c.isRegionPolymerized([adjLaggingPos(1)-1 3], 1, false) &amp;&amp; adjLaggingPos(1) &gt; this.terCPosition+1), 1, &#39;first&#39;);
<a name="l01522"></a>01522             idx2 = find(this.primaseBindingLocations{2} &gt;= adjLaggingPos(2) + (c.isRegionPolymerized([adjLaggingPos(2)+1 2], 1, false) &amp;&amp; adjLaggingPos(2) &lt; this.terCPosition),   1, &#39;first&#39;);
<a name="l01523"></a>01523             <span class="keywordflow">if</span> laggingPos(1) ~= 0 &amp;&amp; ~isempty(idx1); result(1) = idx1; end;
<a name="l01524"></a>01524             <span class="keywordflow">if</span> laggingPos(2) ~= 0 &amp;&amp; ~isempty(idx2); result(2) = idx2; end;
<a name="l01525"></a>01525         end
<a name="l01526"></a>01526         
<a name="l01527"></a>01527         <span class="keyword">function</span> result = <span class="keyword">get</span>.okazakiFragmentPosition(<span class="keyword">this</span>)
<a name="l01528"></a>01528             okazakiFragmentIndex = this.okazakiFragmentIndex;
<a name="l01529"></a>01529             result = zeros(1, 2);
<a name="l01530"></a>01530             <span class="keywordflow">if</span> okazakiFragmentIndex(1) &gt; 0
<a name="l01531"></a>01531                 result(1) = this.primaseBindingLocations{1}(okazakiFragmentIndex(1));
<a name="l01532"></a>01532             end
<a name="l01533"></a>01533             <span class="keywordflow">if</span> okazakiFragmentIndex(2) &gt; 0
<a name="l01534"></a>01534                 result(2) = this.primaseBindingLocations{2}(okazakiFragmentIndex(2));
<a name="l01535"></a>01535             end
<a name="l01536"></a>01536         end
<a name="l01537"></a>01537         
<a name="l01538"></a>01538         <span class="keyword">function</span> result = <span class="keyword">get</span>.okazakiFragmentLength(<span class="keyword">this</span>)
<a name="l01539"></a>01539             okazakiFragmentIndex = this.okazakiFragmentIndex;
<a name="l01540"></a>01540             
<a name="l01541"></a>01541             starts = zeros(1, 2);
<a name="l01542"></a>01542             <span class="keywordflow">if</span> okazakiFragmentIndex(1) &gt; 0
<a name="l01543"></a>01543                 starts(1) = this.primaseBindingLocations{1}(okazakiFragmentIndex(1));
<a name="l01544"></a>01544             end
<a name="l01545"></a>01545             <span class="keywordflow">if</span> okazakiFragmentIndex(2) &gt; 0
<a name="l01546"></a>01546                 starts(2) = this.primaseBindingLocations{2}(okazakiFragmentIndex(2));
<a name="l01547"></a>01547             end
<a name="l01548"></a>01548             
<a name="l01549"></a>01549             ends = zeros(1, 2);
<a name="l01550"></a>01550             <span class="keywordflow">if</span> okazakiFragmentIndex(1) &gt; 1
<a name="l01551"></a>01551                 ends(1) = this.primaseBindingLocations{1}(okazakiFragmentIndex(1)-1) - 1;
<a name="l01552"></a>01552             elseif okazakiFragmentIndex(1) == 1
<a name="l01553"></a>01553                 ends(1) = this.chromosome.sequenceLen;
<a name="l01554"></a>01554             end
<a name="l01555"></a>01555             <span class="keywordflow">if</span> okazakiFragmentIndex(2) &gt; 1
<a name="l01556"></a>01556                 ends(2) = this.primaseBindingLocations{2}(okazakiFragmentIndex(2)-1) + 1;
<a name="l01557"></a>01557             elseif okazakiFragmentIndex(2) == 1
<a name="l01558"></a>01558                 ends(2) = 1;
<a name="l01559"></a>01559             end
<a name="l01560"></a>01560             
<a name="l01561"></a>01561             result = abs(ends-starts)+1;
<a name="l01562"></a>01562             result(okazakiFragmentIndex == 0) = 0;
<a name="l01563"></a>01563         end
<a name="l01564"></a>01564         
<a name="l01565"></a>01565         <span class="keyword">function</span> result = <span class="keyword">get</span>.okazakiFragmentProgress(<span class="keyword">this</span>)
<a name="l01566"></a>01566             laggingPos  = this.laggingPosition;
<a name="l01567"></a>01567             fIdx = this.okazakiFragmentIndex;
<a name="l01568"></a>01568             
<a name="l01569"></a>01569             adjLaggingPos = laggingPos;
<a name="l01570"></a>01570             <span class="keywordflow">if</span> laggingPos(1) ~= 0 &amp;&amp; laggingPos(1) &lt; 1/2* this.terCPosition
<a name="l01571"></a>01571                 adjLaggingPos(1) = laggingPos(1) + this.chromosome.sequenceLen;
<a name="l01572"></a>01572             end
<a name="l01573"></a>01573             <span class="keywordflow">if</span> laggingPos(2) ~= 0 &amp;&amp; laggingPos(2) &gt; 3/2 * this.terCPosition
<a name="l01574"></a>01574                 adjLaggingPos(2) = laggingPos(2) - this.chromosome.sequenceLen;
<a name="l01575"></a>01575             end
<a name="l01576"></a>01576             
<a name="l01577"></a>01577             result = zeros(1, 2);
<a name="l01578"></a>01578             <span class="keywordflow">if</span> laggingPos(1) ~= 0 &amp;&amp; fIdx(1) ~= 0; result(1) = adjLaggingPos(1) - this.primaseBindingLocations{1}(fIdx(1)); end;
<a name="l01579"></a>01579             <span class="keywordflow">if</span> laggingPos(2) ~= 0 &amp;&amp; fIdx(2) ~= 0; result(2) = this.primaseBindingLocations{2}(fIdx(2)) - adjLaggingPos(2); end;
<a name="l01580"></a>01580         end
<a name="l01581"></a>01581         
<a name="l01582"></a>01582         <span class="keyword">function</span> result = <span class="keyword">get</span>.leadingStrandBoundSSBs(<span class="keyword">this</span>)
<a name="l01583"></a>01583             <span class="keyword">import</span> edu.stanford.covert.util.CircularSparseMat;
<a name="l01584"></a>01584             
<a name="l01585"></a>01585             helPos = this.helicasePosition;
<a name="l01586"></a>01586             polPos = this.leadingPolymerasePosition;
<a name="l01587"></a>01587             
<a name="l01588"></a>01588             posStrnds = find(this.chromosome.complexBoundSites == <span class="keyword">this</span>.enzymeGlobalIndexs(<span class="keyword">this</span>.enzymeIndexs_ssb8mer));
<a name="l01589"></a>01589             pos = posStrnds(:, 1);
<a name="l01590"></a>01590             str = posStrnds(:, 2);
<a name="l01591"></a>01591             
<a name="l01592"></a>01592             idx1 = [];
<a name="l01593"></a>01593             idx2 = [];
<a name="l01594"></a>01594             
<a name="l01595"></a>01595             <span class="keywordflow">if</span> polPos(1) ~= 0
<a name="l01596"></a>01596                 idx1 = find(...
<a name="l01597"></a>01597                     str == 1 &amp; ...
<a name="l01598"></a>01598                     pos &lt; polPos(1) &amp; pos &gt; helPos(1));
<a name="l01599"></a>01599             end
<a name="l01600"></a>01600             <span class="keywordflow">if</span> polPos(2) ~= 0
<a name="l01601"></a>01601                 idx2 = find(...
<a name="l01602"></a>01602                     str == 4 &amp; ...
<a name="l01603"></a>01603                     pos &gt; polPos(2) &amp; pos &lt; helPos(2));
<a name="l01604"></a>01604             end
<a name="l01605"></a>01605             
<a name="l01606"></a>01606             result = <a class="code" href="class_circular_sparse_mat.html">CircularSparseMat</a>([pos(idx1) ones(size(idx1)); pos(idx2) repmat(2, size(idx2))], ones(numel(idx1) + numel(idx2), 1), [this.chromosome.sequenceLen, 2], 1);
<a name="l01607"></a>01607         end
<a name="l01608"></a>01608         
<a name="l01609"></a>01609         function result = get.laggingStrandBoundSSBs(this)
<a name="l01610"></a>01610             import edu.stanford.covert.util.<a class="code" href="class_circular_sparse_mat.html">CircularSparseMat</a>;
<a name="l01611"></a>01611             
<a name="l01612"></a>01612             helPos = this.helicasePosition;
<a name="l01613"></a>01613             
<a name="l01614"></a>01614             posStrnds = find(this.chromosome.complexBoundSites == this.enzymeGlobalIndexs(this.enzymeIndexs_ssb8mer));
<a name="l01615"></a>01615             pos = posStrnds(:, 1);
<a name="l01616"></a>01616             str = posStrnds(:, 2);
<a name="l01617"></a>01617             
<a name="l01618"></a>01618             fIdx = this.okazakiFragmentIndex;
<a name="l01619"></a>01619             
<a name="l01620"></a>01620             starts = zeros(1, 2);
<a name="l01621"></a>01621             if fIdx(1) == 0
<a name="l01622"></a>01622                 starts(1) = this.chromosome.sequenceLen;
<a name="l01623"></a>01623             else
<a name="l01624"></a>01624                 starts(1) = this.primaseBindingLocations{1}(fIdx(1));
<a name="l01625"></a>01625             end
<a name="l01626"></a>01626             <span class="keywordflow">if</span> fIdx(2) == 0
<a name="l01627"></a>01627                 starts(2) = 1;
<a name="l01628"></a>01628             <span class="keywordflow">else</span>
<a name="l01629"></a>01629                 starts(2) = this.primaseBindingLocations{2}(fIdx(2));
<a name="l01630"></a>01630             end
<a name="l01631"></a>01631             
<a name="l01632"></a>01632             idx1 = find(...
<a name="l01633"></a>01633                 str == 4 &amp; ...
<a name="l01634"></a>01634                 pos &lt; starts(1) &amp; ...
<a name="l01635"></a>01635                 pos &gt; helPos(1));
<a name="l01636"></a>01636             idx2 = find(...
<a name="l01637"></a>01637                 str == 1 &amp; ...
<a name="l01638"></a>01638                 pos &gt; starts(2) &amp; ...
<a name="l01639"></a>01639                 pos &lt; helPos(2));
<a name="l01640"></a>01640             
<a name="l01641"></a>01641             result = <a class="code" href="class_circular_sparse_mat.html">CircularSparseMat</a>([pos(idx1) ones(size(idx1)); pos(idx2) repmat(2, size(idx2))], ones(numel(idx1) + numel(idx2), 1), [this.chromosome.sequenceLen, 2], 1);
<a name="l01642"></a>01642         end
<a name="l01643"></a>01643         
<a name="l01644"></a>01644         function result = get.numLeadingTemplateBoundSSBs(this)
<a name="l01645"></a>01645             result = full(sum(this.leadingStrandBoundSSBs, 1));
<a name="l01646"></a>01646         end
<a name="l01647"></a>01647         
<a name="l01648"></a>01648         function result = get.numLaggingTemplateBoundSSBs(this)
<a name="l01649"></a>01649             result = full(sum(this.laggingStrandBoundSSBs, 1));
<a name="l01650"></a>01650         end
<a name="l01651"></a>01651         
<a name="l01652"></a>01652         function result = get.areLaggingStrandSSBSitesBound(this)
<a name="l01653"></a>01653             nBndSSBs = this.numLaggingTemplateBoundSSBs;
<a name="l01654"></a>01654             ssbFtpt = this.enzymeDNAFootprints(this.enzymeIndexs_ssb8mer);
<a name="l01655"></a>01655             ssbSpcg = this.ssbComplexSpacing;
<a name="l01656"></a>01656             helPos = this.helicasePosition;
<a name="l01657"></a>01657             fIdx = this.okazakiFragmentIndex;
<a name="l01658"></a>01658             leadFtpt = this.enzymeDNAFootprints(this.enzymeIndexs_coreBetaClampGammaComplex);
<a name="l01659"></a>01659             lagFtpt = this.enzymeDNAFootprints(this.enzymeIndexs_coreBetaClampPrimase);
<a name="l01660"></a>01660             
<a name="l01661"></a>01661             starts = zeros(1, 2);
<a name="l01662"></a>01662             if fIdx(1) == 0
<a name="l01663"></a>01663                 starts(1) = this.chromosome.sequenceLen;
<a name="l01664"></a>01664             else
<a name="l01665"></a>01665                 starts(1) = this.primaseBindingLocations{1}(fIdx(1));
<a name="l01666"></a>01666             end
<a name="l01667"></a>01667             <span class="keywordflow">if</span> fIdx(2) == 0
<a name="l01668"></a>01668                 starts(2) = 1;
<a name="l01669"></a>01669             <span class="keywordflow">else</span>
<a name="l01670"></a>01670                 starts(2) = this.primaseBindingLocations{2}(fIdx(2));
<a name="l01671"></a>01671             end
<a name="l01672"></a>01672             
<a name="l01673"></a>01673             result = <span class="keyword">false</span>(1, 2);
<a name="l01674"></a>01674             <span class="keywordflow">if</span> helPos(1) ~= 0; result(1) = nBndSSBs(1) &gt;= floor((starts(1) - helPos(1) - leadFtpt - lagFtpt) / (ssbFtpt + ssbSpcg) - 2); end;
<a name="l01675"></a>01675             <span class="keywordflow">if</span> helPos(2) ~= 0; result(2) = nBndSSBs(2) &gt;= floor((helPos(2) - starts(2) - leadFtpt - lagFtpt) / (ssbFtpt + ssbSpcg) - 2); end;
<a name="l01676"></a>01676         end
<a name="l01677"></a>01677     end
<a name="l01678"></a>01678 end
</pre></div></div><!-- contents -->
</div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
<a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(0)"><span class="SelectionMark">&#160;</span>All</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(1)"><span class="SelectionMark">&#160;</span>Classes</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(2)"><span class="SelectionMark">&#160;</span>Files</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(3)"><span class="SelectionMark">&#160;</span>Functions</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(4)"><span class="SelectionMark">&#160;</span>Variables</a></div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

  <div id="nav-path" class="navpath">
    <ul>
      <li class="navelem"><a class="el" href="_09process_2_replication_8m.html">Replication.m</a>      </li>

    <li class="footer">Generated on Wed Feb 8 2012 21:42:31 for Whole Cell Simulation by
    <a href="http://www.doxygen.org/index.html">
    <img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.7.6.1 </li>
   </ul>
 </div>


</body>
</html>
