<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<title>Whole Cell Simulation: src/+edu/+stanford/+covert/+cell/+sim/+process/ChromosomeSegregation.m Source File</title>

<link href="tabs.css" rel="stylesheet" type="text/css"/>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
<link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="navtree.js"></script>
<script type="text/javascript">
  $(document).ready(initResizable);
</script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
  $(document).ready(function() { searchBox.OnSelectItem(0); });
</script>

</head>
<body>
<div id="top"><!-- do not remove this div! -->


<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  
  
  <td style="padding-left: 0.5em;">
   <div id="projectname">Whole Cell Simulation
   
   </div>
   
  </td>
  
  
  
 </tr>
 </tbody>
</table>
</div>

<!-- Generated by Doxygen 1.7.6.1 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "search",false,'Search');
</script>
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&#160;Page</span></a></li>
      <li><a href="annotated.html"><span>Classes</span></a></li>
      <li class="current"><a href="files.html"><span>Files</span></a></li>
      <li>
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
  <div id="navrow2" class="tabs2">
    <ul class="tablist">
      <li><a href="files.html"><span>File&#160;List</span></a></li>
      <li><a href="globals.html"><span>File&#160;Members</span></a></li>
    </ul>
  </div>
</div>
<div id="side-nav" class="ui-resizable side-nav-resizable">
  <div id="nav-tree">
    <div id="nav-tree-contents">
    </div>
  </div>
  <div id="splitbar" style="-moz-user-select:none;" 
       class="ui-resizable-handle">
  </div>
</div>
<script type="text/javascript">
  initNavTree('_chromosome_segregation_8m.html','');
</script>
<div id="doc-content">
<div class="header">
  <div class="headertitle">
<div class="title">src/+edu/+stanford/+covert/+cell/+sim/+process/ChromosomeSegregation.m</div>  </div>
</div><!--header-->
<div class="contents">
<a href="_chromosome_segregation_8m.html">Go to the documentation of this file.</a><div class="fragment"><pre class="fragment"><a name="l00001"></a>00001 %<a class="code" href="class_chromosome.html">Chromosome</a> Segregation
<a name="l00002"></a>00002 %
<a name="l00003"></a>00003 % @wholeCellModelID Process_ChromosomeSegregation
<a name="l00004"></a>00004 % @name             <a class="code" href="class_chromosome.html">Chromosome</a> Segregation
<a name="l00005"></a>00005 % @description
<a name="l00006"></a>00006 %   Biology
<a name="l00007"></a>00007 %   ======================
<a name="l00008"></a>00008 %   <a class="code" href="class_chromosome.html">Chromosome</a> segregation in Mycoplasma genitalium has not been well described.
<a name="l00009"></a>00009 %   <a class="code" href="class_chromosome.html">Chromosome</a> segregation in M. genitalium is believed to result from two
<a name="l00010"></a>00010 %   factors
<a name="l00011"></a>00011 %   - entropically favorable segregation throughout DNA replication
<a name="l00012"></a>00012 %   - four chromosome segregation proteins which act to decatenate the
<a name="l00013"></a>00013 %     chromosomes following the completion of replication
<a name="l00014"></a>00014 %   Little is known about the times at which these two factors contribute to
<a name="l00015"></a>00015 %   chromosome segregation.
<a name="l00016"></a>00016 %
<a name="l00017"></a>00017 %   Bacterial chromosome segregation is believed to be entropically favorable
<a name="l00018"></a>00018 %   and occur during replication (Bloom et al. 2010, Jun et al. 2006). That is,
<a name="l00019"></a>00019 %   its believed to be entropically unfavorable for two chromosomes to be
<a name="l00020"></a>00020 %   located near each other, and so as the replication fork moves, the
<a name="l00021"></a>00021 %   replicated DNA molecules migrate away from each other toward the poles. We
<a name="l00022"></a>00022 %   assume that by the end of replication the chromosomes have already migrated
<a name="l00023"></a>00023 %   toward the poles.
<a name="l00024"></a>00024 %
<a name="l00025"></a>00025 %   Four proteins are believed to be involved in chromosome decatenation in M.
<a name="l00026"></a>00026 %   genitalium:
<a name="l00027"></a>00027 %   - MG_470_MONOMER  nucleotide binding domain protein CobQ/CobB/MinD/ParA
<a name="l00028"></a>00028 %   - MG_221_OCTAMER  MraZ
<a name="l00029"></a>00029 %   - MG_387_MONOMER  GTP binding protein Era
<a name="l00030"></a>00030 %   - MG_384_MONOMER  GTPase Obg
<a name="l00031"></a>00031 %   The specific function, kinetics, and metabolic costs of these proteins are
<a name="l00032"></a>00032 %   not known. Furthemore, because M. genitalium contains a reduced complement
<a name="l00033"></a>00033 %   of segregation proteins, it is difficult to infer their functions from
<a name="l00034"></a>00034 %   studies of other bacterial species.
<a name="l00035"></a>00035 %
<a name="l00036"></a>00036 %   Knowledge Base
<a name="l00037"></a>00037 %   ======================
<a name="l00038"></a>00038 %   The knowledge base contains the value of the gtpCost parameter.
<a name="l00039"></a>00039 %
<a name="l00040"></a>00040 %   Representation
<a name="l00041"></a>00041 %   ======================
<a name="l00042"></a>00042 %   The substrates and enzymes properties represent the counts of available
<a name="l00043"></a>00043 %   metabolites and segregation enzymes. gtpCost represents the energetic
<a name="l00044"></a>00044 %   (GTP) cost of chromosome segregation. The segregation status of the
<a name="l00045"></a>00045 %   chromosomes is represented as a boolean properties (segregated) of the
<a name="l00046"></a>00046 %   Chromosome class.
<a name="l00047"></a>00047 %
<a name="l00048"></a>00048 %   Initialization
<a name="l00049"></a>00049 %   ======================
<a name="l00050"></a>00050 %   Chromosome initializes to a state with 1 chromosome (which obviously
<a name="l00051"></a>00051 %   hasn&#39;t yet segregated from the second chromosome which will be later
<a name="l00052"></a>00052 %   produced).
<a name="l00053"></a>00053 %
<a name="l00054"></a>00054 %   Simulation
<a name="l00055"></a>00055 %   ======================
<a name="l00056"></a>00056 %   Because little is known about the molecular biology of chromosome
<a name="l00057"></a>00057 %   segregation, we have chosen to implement this process as a simple boolean
<a name="l00058"></a>00058 %   rule: chromosome segregation can occur if:
<a name="l00059"></a>00059 %   - the chromosome is replicated
<a name="l00060"></a>00060 %   - the chromosome is properly supercoiled
<a name="l00061"></a>00061 %   - there is at least one free molecule of each segregation proteins, and
<a name="l00062"></a>00062 %   - there is at least gtpCost free GTP molecules
<a name="l00063"></a>00063 %
<a name="l00064"></a>00064 %   Note, although the Glass et al. gene essentiality study suggests that the
<a name="l00065"></a>00065 %   cobQ/cobB/minD/parA gene is non-essential, we model this gene as essential
<a name="l00066"></a>00066 %   because we don&#39;t know its specific function and how the other segregation
<a name="l00067"></a>00067 %   proteins compensate in its absence.
<a name="l00068"></a>00068 %
<a name="l00069"></a>00069 %   References
<a name="l00070"></a>00070 %   ======================
<a name="l00071"></a>00071 %   1. Bloom, K., Joglekar, A. (2010). Towards building a chromosome segregation
<a name="l00072"></a>00072 %      machine. Nature 463: 446-456.
<a name="l00073"></a>00073 %   2. Jun, S., Mulder, B. (2006). Entropy-driven special organization of highly
<a name="l00074"></a>00074 %      confined polymers: Lessons for the bacterial chromosome. PNAS 103:
<a name="l00075"></a>00075 %      12388-12393.
<a name="l00076"></a>00076 %   3. Glass, J.I., Assad-Garcia, N., Alperovich, N., Yooseph, S., Lewis, M.R.,
<a name="l00077"></a>00077 %      Maruf, M., Hutchison III, C.A., Smith, H.O., Venter, J.C. (2006).
<a name="l00078"></a>00078 %      Essential genes of a minimal bacterium. PNAS 103: 425-430.
<a name="l00079"></a>00079 %
<a name="l00080"></a>00080 % Author: Jayodita Sanghvi, jayodita@stanford.edu
<a name="l00081"></a>00081 % Author: Jonathan Karr, jkarr@stanford.edu
<a name="l00082"></a>00082 % Author: Jared Jacobs, jmjacobs@stanford.edu
<a name="l00083"></a>00083 % Affilitation: Covert Lab, Department of Bioengineering, Stanford University
<a name="l00084"></a>00084 % Last updated: 7/7/2010
<a name="l00085"></a>00085 classdef  ChromosomeSegregation &lt; edu.stanford.covert.cell.sim.Process &amp; edu.stanford.covert.cell.sim.ChromosomeProcessAspect
<a name="l00086"></a>00086     %property annotations
<a name="l00087"></a>00087     properties (Constant)
<a name="l00088"></a>00088         optionNames__              = {}; %names of option <a class="code" href="json_parse_8m.html#a6e4256f69b751d3d7ec4ea52ff5a8dee">properties</a>
<a name="l00089"></a>00089         fixedConstantNames__       = {   %names of fixed constant <a class="code" href="json_parse_8m.html#a6e4256f69b751d3d7ec4ea52ff5a8dee">properties</a>
<a name="l00090"></a>00090             <span class="stringliteral">&#39;gtpCost&#39;</span>};
<a name="l00091"></a>00091         fittedConstantNames__      = {}; %names of fitted constant <a class="code" href="json_parse_8m.html#a6e4256f69b751d3d7ec4ea52ff5a8dee">properties</a>
<a name="l00092"></a>00092         localStateNames__          = {}; %names of simulation state <a class="code" href="json_parse_8m.html#a6e4256f69b751d3d7ec4ea52ff5a8dee">properties</a> redundant with timecourses in <span class="keyword">this</span> or other processes or the simulation
<a name="l00093"></a>00093     end
<a name="l00094"></a>00094     
<a name="l00095"></a>00095     %IDs, names, and local indices
<a name="l00096"></a>00096     <a class="code" href="json_parse_8m.html#a6e4256f69b751d3d7ec4ea52ff5a8dee">properties</a>
<a name="l00097"></a>00097         stimuliWholeCellModelIDs   = {}; %whole cell model IDs of stimuli
<a name="l00098"></a>00098         
<a name="l00099"></a>00099         substrateWholeCellModelIDs = {   %whole cell model IDs of substrates
<a name="l00100"></a>00100             <span class="stringliteral">&#39;GTP&#39;</span>;<span class="stringliteral">&#39;GDP&#39;</span>;<span class="charliteral">&#39;H&#39;</span>;<span class="stringliteral">&#39;H2O&#39;</span>;<span class="stringliteral">&#39;PI&#39;</span>};
<a name="l00101"></a>00101         substrateIndexs_gtp       = 1; %index of ATP within substrateWholeCellModelIDs
<a name="l00102"></a>00102         substrateIndexs_gdp       = 2; %index of ADP within substrateWholeCellModelIDs
<a name="l00103"></a>00103         substrateIndexs_hydrogen  = 3; %index of hydrogen within substrateWholeCellModelIDs
<a name="l00104"></a>00104         substrateIndexs_water     = 4; %index of water within substrateWholeCellModelIDs
<a name="l00105"></a>00105         substrateIndexs_phosphate = 5; %index of inorganic phosphate within substrateWholeCellModelIDs
<a name="l00106"></a>00106         
<a name="l00107"></a>00107         enzymeWholeCellModelIDs = {      %whole cell model IDs of enzymes
<a name="l00108"></a>00108             <span class="stringliteral">&#39;MG_470_MONOMER&#39;</span>       %CobQ/CobB/MinD/ParA nucleotide binding domain
<a name="l00109"></a>00109             <span class="stringliteral">&#39;MG_221_OCTAMER&#39;</span>       %mraZ protein
<a name="l00110"></a>00110             <span class="stringliteral">&#39;MG_387_MONOMER&#39;</span>       %GTP-binding protein Era
<a name="l00111"></a>00111             &#39;MG_384_MONOMER&#39;       %GTPase1 Obg
<a name="l00112"></a>00112             &#39;MG_203_204_TETRAMER&#39;  %DNA topoisomerase IV
<a name="l00113"></a>00113             };
<a name="l00114"></a>00114         enzymeIndexs_cobQ   = 1;
<a name="l00115"></a>00115         enzymeIndexs_mraZ   = 2;
<a name="l00116"></a>00116         enzymeIndexs_obg    = 3;
<a name="l00117"></a>00117         enzymeIndexs_era    = 4;
<a name="l00118"></a>00118         enzymeIndexs_topoIV = 5;
<a name="l00119"></a>00119     end
<a name="l00120"></a>00120     
<a name="l00121"></a>00121     %fixed biological constants
<a name="l00122"></a>00122     <a class="code" href="json_parse_8m.html#a6e4256f69b751d3d7ec4ea52ff5a8dee">properties</a>
<a name="l00123"></a>00123         gtpCost %number of GTP required <span class="keywordflow">for</span> chromosome segregation (1); <span class="keyword">true</span> value unknown
<a name="l00124"></a>00124     end
<a name="l00125"></a>00125     
<a name="l00126"></a>00126     %constructor
<a name="l00127"></a>00127     methods
<a name="l00128"></a>00128         <span class="keyword">function</span> <span class="keyword">this</span> = ChromosomeSegregation(wholeCellModelID, name)
<a name="l00129"></a>00129             this = this@edu.stanford.covert.cell.sim.<a class="code" href="class_process.html">Process</a>(wholeCellModelID, name);
<a name="l00130"></a>00130         end
<a name="l00131"></a>00131     end
<a name="l00132"></a>00132     
<a name="l00133"></a>00133     %communication between process/simulation
<a name="l00134"></a>00134     methods
<a name="l00135"></a>00135         %set references to state objects
<a name="l00136"></a>00136         function storeObjectReferences(this, simulation)
<a name="l00137"></a>00137             this.storeObjectReferences@edu.stanford.covert.cell.sim.<a class="code" href="class_process.html">Process</a>(simulation);
<a name="l00138"></a>00138             this.storeObjectReferences@edu.stanford.covert.cell.sim.<a class="code" href="class_chromosome_process_aspect.html">ChromosomeProcessAspect</a>(simulation);
<a name="l00139"></a>00139         end
<a name="l00140"></a>00140         
<a name="l00141"></a>00141         %initialize constants
<a name="l00142"></a>00142         function initializeConstants(this, knowledgeBase, simulation, varargin)
<a name="l00143"></a>00143             this.initializeConstants@edu.stanford.covert.cell.sim.<a class="code" href="class_process.html">Process</a>(...
<a name="l00144"></a>00144                 knowledgeBase, simulation, varargin{:});
<a name="l00145"></a>00145             this.initializeConstants@edu.stanford.covert.cell.sim.ChromosomeProcessAspect(...
<a name="l00146"></a>00146                 knowledgeBase, simulation, varargin{:});
<a name="l00147"></a>00147         end
<a name="l00148"></a>00148     end
<a name="l00149"></a>00149     
<a name="l00150"></a>00150     %model
<a name="l00151"></a>00151     methods
<a name="l00152"></a>00152         %Calculate
<a name="l00153"></a>00153         %- contribution to <a class="code" href="class_f_b_a.html">FBA</a> objective
<a name="l00154"></a>00154         %- minimum expression consistent with cell cycle length
<a name="l00155"></a>00155         function [bmProd, byProd, minEnzExp, maxEnzExp] = calcResourceRequirements_LifeCycle(this, ~, ~)
<a name="l00156"></a>00156             %% initialize
<a name="l00157"></a>00157             bmProd = zeros(size(this.substrateWholeCellModelIDs));
<a name="l00158"></a>00158             byProd = zeros(size(this.substrateWholeCellModelIDs));
<a name="l00159"></a>00159             minEnzExp = zeros(size(this.enzymeWholeCellModelIDs));
<a name="l00160"></a>00160             maxEnzExp = Inf(size(this.enzymeWholeCellModelIDs));
<a name="l00161"></a>00161             
<a name="l00162"></a>00162             %% substrate and byproducts: GTP required to decatenate chromosomes, and its hydrolysis (GTP + H2O ==&gt; GDP + PI + H)
<a name="l00163"></a>00163             bmProd(this.substrateIndexs_gtp)       = this.gtpCost;
<a name="l00164"></a>00164             bmProd(this.substrateIndexs_water)     = this.gtpCost;
<a name="l00165"></a>00165             byProd(this.substrateIndexs_gdp)       = this.gtpCost;
<a name="l00166"></a>00166             byProd(this.substrateIndexs_phosphate) = this.gtpCost;
<a name="l00167"></a>00167             byProd(this.substrateIndexs_hydrogen)  = this.gtpCost;
<a name="l00168"></a>00168             
<a name="l00169"></a>00169             %% Segregation requires at least 1 copy of every enzyme
<a name="l00170"></a>00170             minEnzExp(:) = 2;
<a name="l00171"></a>00171         end
<a name="l00172"></a>00172         
<a name="l00173"></a>00173         %initialization: chromosome state initializes chromosome to state with 1
<a name="l00174"></a>00174         %chromosome (which trivially is not segregated)
<a name="l00175"></a>00175         function initializeState(~)
<a name="l00176"></a>00176         end
<a name="l00177"></a>00177         
<a name="l00178"></a>00178         %resource requirements
<a name="l00179"></a>00179         function result = calcResourceRequirements_Current(this)
<a name="l00180"></a>00180             c = this.chromosome;
<a name="l00181"></a>00181             result = zeros(size(this.substrates));
<a name="l00182"></a>00182             if ...
<a name="l00183"></a>00183                     ~c.segregated &amp;&amp; ...
<a name="l00184"></a>00184                     collapse(c.polymerizedRegions) == c.nCompartments * c.sequenceLen &amp;&amp; ...
<a name="l00185"></a>00185                     all(this.enzymes)
<a name="l00186"></a>00186                 
<a name="l00187"></a>00187                 result(this.substrateIndexs_gtp) = this.gtpCost;
<a name="l00188"></a>00188                 result(this.substrateIndexs_water) = result(this.substrateIndexs_gtp);
<a name="l00189"></a>00189             end
<a name="l00190"></a>00190         end
<a name="l00191"></a>00191         
<a name="l00192"></a>00192         %simulation
<a name="l00193"></a>00193         function evolveState(this)
<a name="l00194"></a>00194             c = this.chromosome;
<a name="l00195"></a>00195             
<a name="l00196"></a>00196             if ...
<a name="l00197"></a>00197                     ~c.segregated &amp;&amp; ...
<a name="l00198"></a>00198                     collapse(c.polymerizedRegions) == c.nCompartments * c.sequenceLen &amp;&amp; ...
<a name="l00199"></a>00199                     collapse(c.supercoiled) == c.nCompartments &amp;&amp; ...
<a name="l00200"></a>00200                     all(this.enzymes) &amp;&amp; ...
<a name="l00201"></a>00201                     this.substrates(this.substrateIndexs_gtp) &gt;= this.gtpCost &amp;&amp; ...
<a name="l00202"></a>00202                     this.substrates(this.substrateIndexs_water) &gt;= this.gtpCost
<a name="l00203"></a>00203                 
<a name="l00204"></a>00204                 c.segregated = true;
<a name="l00205"></a>00205                 
<a name="l00206"></a>00206                 this.substrates(this.substrateIndexs_gtp)       = this.substrates(this.substrateIndexs_gtp)       - this.gtpCost;
<a name="l00207"></a>00207                 this.substrates(this.substrateIndexs_water)     = this.substrates(this.substrateIndexs_water)     - this.gtpCost;
<a name="l00208"></a>00208                 this.substrates(this.substrateIndexs_gdp)       = this.substrates(this.substrateIndexs_gdp)       + this.gtpCost;
<a name="l00209"></a>00209                 this.substrates(this.substrateIndexs_phosphate) = this.substrates(this.substrateIndexs_phosphate) + this.gtpCost;
<a name="l00210"></a>00210                 this.substrates(this.substrateIndexs_hydrogen)  = this.substrates(this.substrateIndexs_hydrogen)  + this.gtpCost;
<a name="l00211"></a>00211             end
<a name="l00212"></a>00212         end
<a name="l00213"></a>00213     end
<a name="l00214"></a>00214 end
</pre></div></div><!-- contents -->
</div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
<a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(0)"><span class="SelectionMark">&#160;</span>All</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(1)"><span class="SelectionMark">&#160;</span>Classes</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(2)"><span class="SelectionMark">&#160;</span>Files</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(3)"><span class="SelectionMark">&#160;</span>Functions</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(4)"><span class="SelectionMark">&#160;</span>Variables</a></div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

  <div id="nav-path" class="navpath">
    <ul>
      <li class="navelem"><a class="el" href="_chromosome_segregation_8m.html">ChromosomeSegregation.m</a>      </li>

    <li class="footer">Generated on Wed Feb 8 2012 21:42:32 for Whole Cell Simulation by
    <a href="http://www.doxygen.org/index.html">
    <img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.7.6.1 </li>
   </ul>
 </div>


</body>
</html>
