<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<title>Whole Cell Simulation: src/+edu/+stanford/+covert/+cell/+sim/+process/DNARepair.m Source File</title>

<link href="tabs.css" rel="stylesheet" type="text/css"/>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
<link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="navtree.js"></script>
<script type="text/javascript">
  $(document).ready(initResizable);
</script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
  $(document).ready(function() { searchBox.OnSelectItem(0); });
</script>

</head>
<body>
<div id="top"><!-- do not remove this div! -->


<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  
  
  <td style="padding-left: 0.5em;">
   <div id="projectname">Whole Cell Simulation
   
   </div>
   
  </td>
  
  
  
 </tr>
 </tbody>
</table>
</div>

<!-- Generated by Doxygen 1.7.6.1 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "search",false,'Search');
</script>
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&#160;Page</span></a></li>
      <li><a href="annotated.html"><span>Classes</span></a></li>
      <li class="current"><a href="files.html"><span>Files</span></a></li>
      <li>
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
  <div id="navrow2" class="tabs2">
    <ul class="tablist">
      <li><a href="files.html"><span>File&#160;List</span></a></li>
      <li><a href="globals.html"><span>File&#160;Members</span></a></li>
    </ul>
  </div>
</div>
<div id="side-nav" class="ui-resizable side-nav-resizable">
  <div id="nav-tree">
    <div id="nav-tree-contents">
    </div>
  </div>
  <div id="splitbar" style="-moz-user-select:none;" 
       class="ui-resizable-handle">
  </div>
</div>
<script type="text/javascript">
  initNavTree('_d_n_a_repair_8m.html','');
</script>
<div id="doc-content">
<div class="header">
  <div class="headertitle">
<div class="title">src/+edu/+stanford/+covert/+cell/+sim/+process/DNARepair.m</div>  </div>
</div><!--header-->
<div class="contents">
<a href="_d_n_a_repair_8m.html">Go to the documentation of this file.</a><div class="fragment"><pre class="fragment"><a name="l00001"></a>00001 %DNA Repair
<a name="l00002"></a>00002 %
<a name="l00003"></a>00003 % @wholeCellModelID Process_DNARepair
<a name="l00004"></a>00004 % @name             DNA Repair
<a name="l00005"></a>00005 % @description
<a name="l00006"></a>00006 %   Biology
<a name="l00007"></a>00007 %   =====================================================================
<a name="l00008"></a>00008 %   Simulates repair of a spectrum of DNA lesions by the direct damage
<a name="l00009"></a>00009 %   repair (DDR), base excision (BER), nucleotide excision (NER), and
<a name="l00010"></a>00010 %   homologous recombination (HR) DNA repair pathways. The table below
<a name="l00011"></a>00011 %   summarizes the pathways of DNA repair modelled here.
<a name="l00012"></a>00012 %
<a name="l00013"></a>00013 %      Damage                   Repair <a class="code" href="class_pathway.html">Pathway</a>
<a name="l00014"></a>00014 %      ======                   ==============
<a name="l00015"></a>00015 %      Abasic site              BER
<a name="l00016"></a>00016 %      Damaged base             BER, GG-NER
<a name="l00017"></a>00017 %      Intrastrand crosslink    GG-NER
<a name="l00018"></a>00018 %      Single strand break      DDR
<a name="l00019"></a>00019 %      Double strand break      HR-DSBR
<a name="l00020"></a>00020 %
<a name="l00021"></a>00021 %   This process also simulates DNA damage recognition by the DNA integrity
<a name="l00022"></a>00022 %   scanning protein DisA (MG_105).
<a name="l00023"></a>00023 %
<a name="l00024"></a>00024 %   Direct Damage Reversal (DDR)
<a name="l00025"></a>00025 %   +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
<a name="l00026"></a>00026 %   The only direct damage repair homolog identified in M. genitalium is
<a name="l00027"></a>00027 %   ligase ligA (MG_254). ligA is capable of repairing single strand
<a name="l00028"></a>00028 %   breaks by a NAD-dependent mechanism:
<a name="l00029"></a>00029 %
<a name="l00030"></a>00030 %         base    O   base    O           base    O   base    O
<a name="l00031"></a>00031 %          |     ||    |     ||            |      ||   |     ||
<a name="l00032"></a>00032 %   ...-O-sugar-O-P-O-sugar-O-P-O-  +  HO-sugar-O-P-O-sugar-O-P...  + NAD
<a name="l00033"></a>00033 %                 |           |                   |           |
<a name="l00034"></a>00034 %                 O           O                   O           O
<a name="l00035"></a>00035 %
<a name="l00036"></a>00036 %                                    ||
<a name="l00037"></a>00037 %                                    ||
<a name="l00038"></a>00038 %                                    V
<a name="l00039"></a>00039 %
<a name="l00040"></a>00040 %
<a name="l00041"></a>00041 %         base    O    base   O   base    O   base   O
<a name="l00042"></a>00042 %          |     ||    |     ||    |     ||    |    ||
<a name="l00043"></a>00043 %   ...-O-sugar-O-P-O-sugar-O-P-O-sugar-O-P-O-sugar-O-P...  + AMP + NMN + H
<a name="l00044"></a>00044 %                 |           |           |           |
<a name="l00045"></a>00045 %                 O           O           O           O
<a name="l00046"></a>00046 %
<a name="l00047"></a>00047 %
<a name="l00048"></a>00048 %   Base Excision Repair (BER)
<a name="l00049"></a>00049 %   +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
<a name="l00050"></a>00050 %   Repairs specific spontaneously and endogenously produced DNA damaged
<a name="l00051"></a>00051 %   including base modifications, abasic sites, and single strand breaks
<a name="l00052"></a>00052 %   by the following mechanism:
<a name="l00053"></a>00053 %   1. Base excision by Fpg (MG_498) or Ung (MG_097) glycosylase
<a name="l00054"></a>00054 %
<a name="l00055"></a>00055 %         base    O                 OH      O
<a name="l00056"></a>00056 %          |     ||                  |     ||
<a name="l00057"></a>00057 %      -O-sugar-O-P... + H2O ==&gt; -O-sugar-O-P  +  base
<a name="l00058"></a>00058 %                 |                         |
<a name="l00059"></a>00059 %                 O                         O
<a name="l00060"></a>00060 %
<a name="l00061"></a>00061 %   2. a. i.  3&#39;-AP lyase Fpg (MG_498)
<a name="l00062"></a>00062 %         ii. 5&#39;-AP endonuclease Nfo (MG_235)
<a name="l00063"></a>00063 %      b. i.  5&#39;-deoxyribosephosphodiesterase Nfo (MG_235)
<a name="l00064"></a>00064 %         ii. 3&#39;-(deoxyribose-5&#39;-phosphate) lyase dnaN (MG_001)
<a name="l00065"></a>00065 %   3. 5&#39;-3&#39; Polymerization by dnaN (MG_001)
<a name="l00066"></a>00066 %   4. Ligation by ligA (MG_254). See DDR section for details.
<a name="l00067"></a>00067 %   5. Catabolism and export of damaged nucleobases and nucleosides by
<a name="l00068"></a>00068 %      metabolic process.
<a name="l00069"></a>00069 %
<a name="l00070"></a>00070 %   Flap endonuclease fen1 homologs have not been identified in M.
<a name="l00071"></a>00071 %   genitalium; MG_262 is a weak fen1 homolog, but is more strongly
<a name="l00072"></a>00072 %   homologous to polI exonucleases. Consequently, we do not model flap
<a name="l00073"></a>00073 %   base excision repair.
<a name="l00074"></a>00074 %
<a name="l00075"></a>00075 %   Global Genomic (GGR) Nucleotide Excision Repair (NER)
<a name="l00076"></a>00076 %   +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
<a name="l00077"></a>00077 %   NER recognizes bulky distortions in the shape of the DNA helix caused
<a name="l00078"></a>00078 %   by damaged bases and cross links caused by NO and UV light, and uses
<a name="l00079"></a>00079 %   endonucleases with lower specificity than that of BER. Consequently,
<a name="l00080"></a>00080 %   NER has broader repair capability than base excision repair. Here we
<a name="l00081"></a>00081 %   model only global genomic NER. In some bacteria such as E. coli DNA
<a name="l00082"></a>00082 %   damage is sensed by <a class="code" href="class_r_n_a.html">RNA</a> polymerase and coupled to the NER machinery by
<a name="l00083"></a>00083 %   the transcription-repair coupling factor mfd.
<a name="l00084"></a>00084 %
<a name="l00085"></a>00085 %   1. UvrABC (MG_073, MG_206, MG_421) cleaves DNA 3&#39; and 5&#39; to lesion
<a name="l00086"></a>00086 %   2. Helicase PcrA (MG_244) excises cleaved DNA
<a name="l00087"></a>00087 %   3. Cleavage excised oligonucleotide (12-13 nts) by MgpA (MG_190)
<a name="l00088"></a>00088 %   4. 5&#39;-3&#39; Polymerization by dnaN (MG_001)
<a name="l00089"></a>00089 %   5. Ligation by ligA (MG_254). See DDR section for details.
<a name="l00090"></a>00090 %   6. Catabolism and export of damaged nucleobases and nucleosides by
<a name="l00091"></a>00091 %      metabolic process.
<a name="l00092"></a>00092 %
<a name="l00093"></a>00093 %   <a class="code" href="class_note.html">Note</a> steps (3) and (4-5) occur in parallel. NER removes 2.7 x 10^3
<a name="l00094"></a>00094 %   excison products per cell per minute.
<a name="l00095"></a>00095 %
<a name="l00096"></a>00096 %   Homologous Recombination (HR) Double Strand Break Repair (DSBR)
<a name="l00097"></a>00097 %   +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
<a name="l00098"></a>00098 %   HR repairs <span class="keywordtype">double</span> strand breaks caused by ionizing radiation and
<a name="l00099"></a>00099 %   stalled replicaton forks. HR also repairs strand gaps which can be
<a name="l00100"></a>00100 %   generated by the interaction of replication forks with unrepaired
<a name="l00101"></a>00101 %   lesions. M. genitalium has a very reduced complement of recombination
<a name="l00102"></a>00102 %   repair proteins. Normally recombination repair involves the following
<a name="l00103"></a>00103 %   steps:
<a name="l00104"></a>00104 %   1. Initiation: 5&#39;-3&#39; exonuclease removes dNMPs from 5&#39; ends of strands
<a name="l00105"></a>00105 %      leaving 3&#39; overhangs. Daley and Wilson have shown homologous
<a name="l00106"></a>00106 %      recombination is most efficient with overhangs of at least 8 bases
<a name="l00107"></a>00107 %      [6]. No tradiational initiation gene has been identified in M.
<a name="l00108"></a>00108 %      genitalium. 5&#39;-3&#39; resection could be carried out by pcrA (MG_244)
<a name="l00109"></a>00109 %      or polI-like (MG_262). Here we model pcrA as the NER helicase, and
<a name="l00110"></a>00110 %      therefore choose to model polI-like as the HR initiator.
<a name="l00111"></a>00111 %   2. Strand exchange: One of the damaged 3&#39; overhang invades the
<a name="l00112"></a>00112 %      undamaged homologous chromosome, forming a holliday junction.
<a name="l00113"></a>00113 %      Catalyzed by recA (MG_339)
<a name="l00114"></a>00114 %   3. 5&#39;-3&#39; Polymerization by dnaN (MG_001)
<a name="l00115"></a>00115 %   4. 2 ligations by ligA (MG_254). See DDR section for details.
<a name="l00116"></a>00116 %   5. Second strand exchange: The second damaged 3&#39; overhang invades the
<a name="l00117"></a>00117 %      undamaged homologous chromosome, forming a second holliday
<a name="l00118"></a>00118 %      junction. Also catalyzed by recA (MG_339)
<a name="l00119"></a>00119 %   6. Holliday junction migration: ruvAB (MG_358, MG_359) widens distance
<a name="l00120"></a>00120 %      between the two strand cross over points by moving holliday
<a name="l00121"></a>00121 %      junctions to preferred sequence 5&#39;-(A/T)TT?(G/C)-3&#39; [PUB_0532].
<a name="l00122"></a>00122 %   7. Resolution: recU (MG_352) dimer nicks DNA at cross over points
<a name="l00123"></a>00123 %      creating four single strand breaks
<a name="l00124"></a>00124 %   8. 4 ligations by ligA (MG_254). See DDR section for details.
<a name="l00125"></a>00125 %   9. Catabolism and export of damaged nucleobases and nucleosides by
<a name="l00126"></a>00126 %      metabolic process.
<a name="l00127"></a>00127 %
<a name="l00128"></a>00128 %   Additionally, homologous recombination is believed to be critical
<a name="l00129"></a>00129 %   for antigenic variation in M. genitalium (Sluijter et al).
<a name="l00130"></a>00130 %
<a name="l00131"></a>00131 %   DNA integrity scanning protein DisA
<a name="l00132"></a>00132 %   +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
<a name="l00133"></a>00133 %   In B. subtilis DisA recognizes strand breaks, base damage,
<a name="l00134"></a>00134 %   and cross links (possibly through <a class="code" href="classss_d_n_a.html">ssDNA</a>, Holliday junctions, or stalled
<a name="l00135"></a>00135 %   replication forks that are generated during recombination repair of
<a name="l00136"></a>00136 %   these lesions), binds damaged DNA, and via an unknown mechanism delays
<a name="l00137"></a>00137 %   sporulation. DisA has weak, non-specific DNA binding. In addition DisA
<a name="l00138"></a>00138 %   has diadenylate cyclization (produces c-di-AMP from 2 ATPs) activity
<a name="l00139"></a>00139 %   which is abolished when bound to branched DNA such as Holliday
<a name="l00140"></a>00140 %   junctions. c-di-AMP has been speculated as the second messenger which
<a name="l00141"></a>00141 %   couples DisA damage recognition to regulation of the transcriptional
<a name="l00142"></a>00142 %   regulator Spo0A, and delayed sporulation. The rest of the B. subtilis
<a name="l00143"></a>00143 %   signaling system (spo0A, kinA, kinE) is not present in M. genitalium.
<a name="l00144"></a>00144 %
<a name="l00145"></a>00145 %   DNA Restriction/Modification
<a name="l00146"></a>00146 %   +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
<a name="l00147"></a>00147 %   Organisms employ a restriction/modification systems to distinguish
<a name="l00148"></a>00148 %   self from foreign DNA. These systems methylate hemi-methylated DNA and
<a name="l00149"></a>00149 %   cleave unmethylated DNA. The systems maintains self DNA in the fully
<a name="l00150"></a>00150 %   methylated state so that it can be distinguished from foreign
<a name="l00151"></a>00151 %   DNA (which will be unmethylated) on methylation status.
<a name="l00152"></a>00152 %
<a name="l00153"></a>00153 %   M. genitalium contains a reduced restriction/modification system
<a name="l00154"></a>00154 %   consisting of the EcoD type I DNA recognition subunit (MG_438) and the
<a name="l00155"></a>00155 %   MunI type II methylase (MG_184). M. genitalium does not contain a
<a name="l00156"></a>00156 %   type III R/M system.
<a name="l00157"></a>00157 %
<a name="l00158"></a>00158 %   EcoD is the DNA binding domain of a multimetric methylation and
<a name="l00159"></a>00159 %   restriction complex. Type I R/M systems typically have structure R2M2S
<a name="l00160"></a>00160 %   or M2S. EcoD recognizes the sequence 5&#39;-TCARTTC-3&#39;. The other subunits
<a name="l00161"></a>00161 %   methylate the N(7) position of A-3, producing N(7)-methyladenine or
<a name="l00162"></a>00162 %   cleavage DNA at this position. M. genitalium contains only the DNA
<a name="l00163"></a>00163 %   recognition subunit. 
<a name="l00164"></a>00164 %
<a name="l00165"></a>00165 %   MunI methylase (MG_184) methylates the N(6) position of A-3 in the
<a name="l00166"></a>00166 %   sequence 5&#39;-CAATTG-3&#39;, producing N(6)-methyladenine. M. genitalium
<a name="l00167"></a>00167 %   does not contain a MunI endonuclease. Type II R/M systems have been
<a name="l00168"></a>00168 %   found as monomers, dimers, and trimers.
<a name="l00169"></a>00169 %
<a name="l00170"></a>00170 %   DNA repair mechanisms not present in M. genitalium
<a name="l00171"></a>00171 %   +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
<a name="l00172"></a>00172 %   - <a class="code" href="class_transcription.html">Transcription</a>-coupled NER (TC-NER)
<a name="l00173"></a>00173 %     M. genitalium does not have a transcription-repair coupling factor
<a name="l00174"></a>00174 %     mfd homolog.
<a name="l00175"></a>00175 %   - Additional eukaryotic homologous recombination (HR) pathways
<a name="l00176"></a>00176 %     - Synthesis-dependent strand annealing (SDSA)
<a name="l00177"></a>00177 %     - Single-strand annealing (SSA)
<a name="l00178"></a>00178 %     - Break-induced replication (BIR)
<a name="l00179"></a>00179 %   - Non-homologous end joining (NHEJ)
<a name="l00180"></a>00180 %     M. genitalium does not have a dedicated NHEJ DNA ligase ligD or a
<a name="l00181"></a>00181 %     DNA-end-binding protein Ku.
<a name="l00182"></a>00182 %   - Mismatch Repair
<a name="l00183"></a>00183 %   - SOS Response
<a name="l00184"></a>00184 %
<a name="l00185"></a>00185 %   Metabolism of damaged nucleo-bases, -sides, and -tides
<a name="l00186"></a>00186 %   +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
<a name="l00187"></a>00187 %   No gene has been identified which cleaves DNA oligonucleotides;
<a name="l00188"></a>00188 %   possible genes include pcrA (MG_244), mgpA (MG_190), and polI-like
<a name="l00189"></a>00189 %   (MG_262). Here we model pcrA as the NER helicase and polI-like as the
<a name="l00190"></a>00190 %   HR resection enzyme, leaving mgpA without an assigned function.
<a name="l00191"></a>00191 %   Therefore we choose to model mgpA as capable of cleaving
<a name="l00192"></a>00192 %   oligonucleotides released by NER into single nucleotides.
<a name="l00193"></a>00193 %
<a name="l00194"></a>00194 %   Nucleotides released by polI-like or mgpA are returned to the
<a name="l00195"></a>00195 %   cytoplasmic free pool, and are available to the metabolism process.
<a name="l00196"></a>00196 %   Damaged nucleotides are degraded to nucleosides and are exported by
<a name="l00197"></a>00197 %   the metabolism process. Damaged nucleobases are also exported by the
<a name="l00198"></a>00198 %   metabolism process.
<a name="l00199"></a>00199 %
<a name="l00200"></a>00200 %   Knowledge Base
<a name="l00201"></a>00201 %   =====================================================================
<a name="l00202"></a>00202 %   As of 8/12/2010 the M. genitalium knowledge contained the stochiometry
<a name="l00203"></a>00203 %   and kinetics of 33 DNA repair reactions involving 16 enzymes. These
<a name="l00204"></a>00204 %   reactions were compiled from several sources.
<a name="l00205"></a>00205 %
<a name="l00206"></a>00206 %      <a class="code" href="class_reaction.html">Reaction</a> Type                          Reactions
<a name="l00207"></a>00207 %      ========================               ==========
<a name="l00208"></a>00208 %      base excision repair                        4
<a name="l00209"></a>00209 %      base excision repair, base excision         6
<a name="l00210"></a>00210 %      DNA catabolism                          4
<a name="l00211"></a>00211 %      DNA ligation                            1
<a name="l00212"></a>00212 %      DNA polymerization                      4
<a name="l00213"></a>00213 %      DNA restriction/modification                4
<a name="l00214"></a>00214 %      homologous recombination                7
<a name="l00215"></a>00215 %      nucleotide excision repair                  2
<a name="l00216"></a>00216 %      second messenger generation                 1
<a name="l00217"></a>00217 %
<a name="l00218"></a>00218 %   Representation
<a name="l00219"></a>00219 %   =====================================================================
<a name="l00220"></a>00220 %   8 <a class="code" href="json_parse_8m.html#a6e4256f69b751d3d7ec4ea52ff5a8dee">properties</a> are used to keep track of the damage state of DNA. These
<a name="l00221"></a>00221 %   <a class="code" href="json_parse_8m.html#a6e4256f69b751d3d7ec4ea52ff5a8dee">properties</a> are owned by the DNA damage process, and are copied locally at
<a name="l00222"></a>00222 %   each time step. The first  dimension of each has length equal to that of
<a name="l00223"></a>00223 %   the chromosome.
<a name="l00224"></a>00224 %   - occupiedDNA               enumeration   proteins bound to DNA
<a name="l00225"></a>00225 %   - gapSites                  true/false    sugar gaps
<a name="l00226"></a>00226 %   - abasicSites               true/false    abasic sites
<a name="l00227"></a>00227 %   - damagedSugarPhosphates    enumeration   damaged sugar phosphates
<a name="l00228"></a>00228 %   - damagedBases              enumeration   damaged bases
<a name="l00229"></a>00229 %   - intrastrandCrossLinks     true/false    intrastand cross link
<a name="l00230"></a>00230 %   - strandBreaks              true/false    strand break (missing sugar-phosphate bond)
<a name="l00231"></a>00231 %   - hollidayJunctions         true/false    holliday junction
<a name="l00232"></a>00232 %
<a name="l00233"></a>00233 %   substrates represents the amounts of free metabolites. initializeConstants
<a name="l00234"></a>00234 %   synchronizes the substrates of this process with that of the DNA damage
<a name="l00235"></a>00235 %   process to make it easy for these processes to communicate. That is
<a name="l00236"></a>00236 %   initializeConstants sets the substrates of both processes equal to the union
<a name="l00237"></a>00237 %   of the two, are orders the substrates the same in both processes. Thus
<a name="l00238"></a>00238 %   substrate indices defined in process are valid on <a class="code" href="json_parse_8m.html#a6e4256f69b751d3d7ec4ea52ff5a8dee">properties</a> received from
<a name="l00239"></a>00239 %   the other process, and it is unecessary to remap <a class="code" href="json_parse_8m.html#a6e4256f69b751d3d7ec4ea52ff5a8dee">properties</a> in
<a name="l00240"></a>00240 %   copyFromState.
<a name="l00241"></a>00241 %
<a name="l00242"></a>00242 %   enzymes represents the amount of DNA repair enzymes.
<a name="l00243"></a>00243 %
<a name="l00244"></a>00244 %   Rather than using reactionStoichiometryMatrix directly this process
<a name="l00245"></a>00245 %   factors the reactionStoichiometryMatrix into two parts:
<a name="l00246"></a>00246 %   - reactionSmallMoleculeStoichiometryMatrix: the stochiometry of free
<a name="l00247"></a>00247 %     metabolites in repair reactions
<a name="l00248"></a>00248 %   - reactionDNAStoichiometryMatrix: the stoichiometry of DNA bases in
<a name="l00249"></a>00249 %     repair reactions
<a name="l00250"></a>00250 %
<a name="l00251"></a>00251 %   In addition, rather than using reactionModificationReaction to
<a name="l00252"></a>00252 %   represent the macromolecule affected by each reaction we use
<a name="l00253"></a>00253 %   reactionDNABase to represent the DNA base affected by each repair reaction.
<a name="l00254"></a>00254 %
<a name="l00255"></a>00255 %   reactionCatalysisMatrix represents the enzymes which catalyze each repair
<a name="l00256"></a>00256 %   reaction. enzymeBounds represents the backward and forward enzyme rate
<a name="l00257"></a>00257 %   constant of each repair reaction.
<a name="l00258"></a>00258 %
<a name="l00259"></a>00259 %   Several additional <a class="code" href="json_parse_8m.html#a6e4256f69b751d3d7ec4ea52ff5a8dee">properties</a> represent more detailed information
<a name="l00260"></a>00260 %   about the specific action of each enzyme.
<a name="l00261"></a>00261 %
<a name="l00262"></a>00262 %   Simulation
<a name="l00263"></a>00263 %   =====================================================================
<a name="l00264"></a>00264 %   The simulation is implemented as six subfunctions which are called in a
<a name="l00265"></a>00265 %   random order from evolveState:
<a name="l00266"></a>00266 %   - base excision repair
<a name="l00267"></a>00267 %   - nucleotide excision repair
<a name="l00268"></a>00268 %   - homologous recombination <span class="keywordtype">double</span> strand break repair
<a name="l00269"></a>00269 %   - polymerization/ligation
<a name="l00270"></a>00270 %   - DisA damage recognition, and
<a name="l00271"></a>00271 %   - restriction/modification
<a name="l00272"></a>00272 %
<a name="l00273"></a>00273 %   Each subfunction carries out some variation of the algorithm:
<a name="l00274"></a>00274 %   1. Identify chromosomes sites that are substrates for the repair system
<a name="l00275"></a>00275 %   2. Eliminate sites that neighbor damage, or which are occupied by proteins
<a name="l00276"></a>00276 %      (other than the DNA damage scanning protein DisA).
<a name="l00277"></a>00277 %   3. Compute maximum number of repair reactions that can occur based on
<a name="l00278"></a>00278 %      metabolite and enzyme availability and kinetics.
<a name="l00279"></a>00279 %   4. Randomly select among chromosome sites
<a name="l00280"></a>00280 %   5. Update state of DNA. Update metabolites.
<a name="l00281"></a>00281 %   6. Free any DisA that was previously bound to the repaired site.
<a name="l00282"></a>00282 %
<a name="l00283"></a>00283 %   References
<a name="l00284"></a>00284 %   =====================================================================
<a name="l00285"></a>00285 %   1. B.A. Sokhansanj, G.R. Rodrigue, J.P. Fitch and D.M. Wilson III
<a name="l00286"></a>00286 %      (2002). A quantitative model of human DNA base excision repair: I.
<a name="l00287"></a>00287 %      Mechanistic insights. Nuc Acids Res. 30(8): 1817-25. [PUB_0488]
<a name="l00288"></a>00288 %   2. Sokhansanj BA, Wilson DM 3rd (2004). Oxidative DNA damage
<a name="l00289"></a>00289 %      background estimated by a system model of base excision repair.
<a name="l00290"></a>00290 %      Free Radic Biol Med. 37(3): 422-7. [PUB_0487]
<a name="l00291"></a>00291 %   3. Politi A, Moné MJ, Houtsmuller AB, Hoogstraten D, Vermeulen W,
<a name="l00292"></a>00292 %      Heinrich R, van Driel R (2005). Mathematical modeling of nucleotide
<a name="l00293"></a>00293 %      excision repair reveals efficiency of sequential assembly
<a name="l00294"></a>00294 %      strategies. Mol Cell. 19(5): 679-90. [PUB_0514]
<a name="l00295"></a>00295 %   4. Alain Blanchard, Glenn Browning (2005). Mycoplasmas: molecular
<a name="l00296"></a>00296 %      biology, pathogenicity and strategies for control. CRC Press.
<a name="l00297"></a>00297 %   5. Carvalho FM, Fonseca MM, Batistuzzo De Medeiros S, Scortecci KC,
<a name="l00298"></a>00298 %      Blaha CA, Agnez-Lima LF (2005). DNA repair in reduced genome: The
<a name="l00299"></a>00299 %      Mycoplasma model. <a class="code" href="class_gene.html">Gene</a>. 360(2): 111-9.
<a name="l00300"></a>00300 %   6. James M. Daley and Thomas E. Wilson (2005). Rejoining of DNA
<a name="l00301"></a>00301 %      Double-Strand Breaks as a Function of Overhang Length. Mol Cell
<a name="l00302"></a>00302 %      Biol. 25(3): 896-906.
<a name="l00303"></a>00303 %   7. Reardon JT, Sancar A (2005). Nucleotide Excision Repair. Prog
<a name="l00304"></a>00304 %      Nucleic Acid Res Mol Biol. 79: 183-235. [PUB_0084]
<a name="l00305"></a>00305 %   8. Marcelo Brocchi, Ana Tereza Ribeiro de Vasconcelos, and Arnaldo Zaha
<a name="l00306"></a>00306 %      (2007). Restriction-modification systems in Mycoplasma spp. Genet.
<a name="l00307"></a>00307 %      Mol. Biol. 30: 1. [PUB_0652]
<a name="l00308"></a>00308 %   9. Wilson GG, Murray NE (1991). Restriction and modification systems.
<a name="l00309"></a>00309 %      Annu Rev Genet. 25: 585-627. [PUB_0653]
<a name="l00310"></a>00310 %  10. Romling U (2008). Great Times for Small Molecules: c-di-AMP, a
<a name="l00311"></a>00311 %      Second Messenger Candidate in Bacteria and Archaea. Sci Signal.
<a name="l00312"></a>00312 %      1(33):39. [PUB_0195]
<a name="l00313"></a>00313 %  11. Erik Boye (2006). DisA, a Busy Bee That Monitors <a class="code" href="class_chromosome.html">Chromosome</a>
<a name="l00314"></a>00314 %      Integrity. Cell. 125(4):641-643. [PUB_0511]
<a name="l00315"></a>00315 %  12. Witte G, Hartung S, Büttner K, Hopfner KP (2008). Structural.
<a name="l00316"></a>00316 %      Biochemistry of a Bacterial Checkpoint <a class="code" href="class_protein.html">Protein</a> Reveals Diadenylate
<a name="l00317"></a>00317 %      Cyclase Activity Regulated by DNA Recombination Intermediates. Mol
<a name="l00318"></a>00318 %      Cell. 30(2):167-78. [PUB_0512]
<a name="l00319"></a>00319 %  13. Bejerano-Sagie M, Oppenheimer-Shaanan Y, Berlatzky I, Rouvinski A,
<a name="l00320"></a>00320 %      Meyerovich M, Ben-Yehuda S (2006). A checkpoint protein that scans
<a name="l00321"></a>00321 %      the chromosome for damage at the start of sporulation in Bacillus
<a name="l00322"></a>00322 %      subtilis. Cell. 125(4):679-90. [PUB_0513]
<a name="l00323"></a>00323 %  14. Murray NE (2000). Type I restriction systems: sophisticated
<a name="l00324"></a>00324 %      molecular machines (a legacy of Bertani and Weigle). Microbiol Mol
<a name="l00325"></a>00325 %      Biol Rev. 64(2):412-34. [PUB_0542]
<a name="l00326"></a>00326 %  15. Cline SD, Hanawalt PC (2003). Who&#39;s on first in the cellular
<a name="l00327"></a>00327 %      response to DNA damage? Nat Rev Mol Cell Biol. 4(5):361-72.
<a name="l00328"></a>00328 %      [PUB_0780]
<a name="l00329"></a>00329 %  16. Sluijter M, Kaptein E, Spuesens EB, Hoogenboezem T, Hartwig NG, Van 
<a name="l00330"></a>00330 %      Rossum AM, Vink C (2010). The Mycoplasma genitalium MG352-encoded protein 
<a name="l00331"></a>00331 %      is a Holliday junction resolvase that has a non-functional orthologue in 
<a name="l00332"></a>00332 %      Mycoplasma pneumoniae. Mol Microbiol. 77(5): 1261-77. [PUB_0797].
<a name="l00333"></a>00333 %
<a name="l00334"></a>00334 % Author: Jonathan Karr, jkarr@stanford.edu
<a name="l00335"></a>00335 % Affilitation: Covert Lab, Department of Bioengineering, Stanford University
<a name="l00336"></a>00336 % Last updated: 5/3/2010
<a name="l00337"></a>00337 classdef  DNARepair &lt; edu.stanford.covert.cell.sim.<a class="code" href="class_reaction_process.html">ReactionProcess</a> &amp; edu.stanford.covert.cell.sim.<a class="code" href="class_chromosome_process_aspect.html">ChromosomeProcessAspect</a>
<a name="l00338"></a>00338     %property annotations
<a name="l00339"></a>00339     <a class="code" href="json_parse_8m.html#a6e4256f69b751d3d7ec4ea52ff5a8dee">properties</a> (Constant)
<a name="l00340"></a>00340         optionNames__              = {}; %names of option <a class="code" href="json_parse_8m.html#a6e4256f69b751d3d7ec4ea52ff5a8dee">properties</a>        
<a name="l00341"></a>00341         fixedConstantNames__       = {   %names of fixed constant <a class="code" href="json_parse_8m.html#a6e4256f69b751d3d7ec4ea52ff5a8dee">properties</a>
<a name="l00342"></a>00342             <span class="stringliteral">&#39;HR_PolA_ResectionLength&#39;</span>;
<a name="l00343"></a>00343             <span class="stringliteral">&#39;HR_RecA_Spacing&#39;</span>;
<a name="l00344"></a>00344             <span class="stringliteral">&#39;HR_RuvAB_JunctionMigrationHop&#39;</span>;
<a name="l00345"></a>00345             <span class="stringliteral">&#39;HR_RecU_CleavageSequence&#39;</span>;
<a name="l00346"></a>00346             <span class="stringliteral">&#39;HR_RecU_CleavagePosition&#39;</span>;
<a name="l00347"></a>00347             <span class="stringliteral">&#39;NER_UvrABC_IncisionMargin3&#39;</span>;
<a name="l00348"></a>00348             <span class="stringliteral">&#39;NER_UvrABC_IncisionMargin5&#39;</span>;
<a name="l00349"></a>00349             <span class="stringliteral">&#39;NER_PcrA_StepSize&#39;</span>;
<a name="l00350"></a>00350             <span class="stringliteral">&#39;hollidayJunctionResolutionSites&#39;</span>;
<a name="l00351"></a>00351             <span class="stringliteral">&#39;RM_EcoD_RecognitionSequence&#39;</span>;
<a name="l00352"></a>00352             <span class="stringliteral">&#39;RM_EcoD_RecognitionSites&#39;</span>;
<a name="l00353"></a>00353             <span class="stringliteral">&#39;RM_EcoD_MethylationPosition&#39;</span>;
<a name="l00354"></a>00354             <span class="stringliteral">&#39;RM_EcoD_RestrictionPosition&#39;</span>;
<a name="l00355"></a>00355             <span class="stringliteral">&#39;RM_MunI_RecognitionSequence&#39;</span>;
<a name="l00356"></a>00356             <span class="stringliteral">&#39;RM_MunI_RecognitionSites&#39;</span>;
<a name="l00357"></a>00357             <span class="stringliteral">&#39;RM_MunI_MethylationPosition&#39;</span>;
<a name="l00358"></a>00358             <span class="stringliteral">&#39;RM_MunI_RestrictionPosition&#39;</span>;
<a name="l00359"></a>00359             <span class="stringliteral">&#39;RM_MunI_MethylatedPositions&#39;</span>;
<a name="l00360"></a>00360             <span class="stringliteral">&#39;RM_MunI_RestrictionPositions&#39;</span>;
<a name="l00361"></a>00361             <span class="stringliteral">&#39;reactionSmallMoleculeStoichiometryMatrix&#39;</span>;
<a name="l00362"></a>00362             <span class="stringliteral">&#39;reactionDNAStoichiometryMatrix&#39;</span>;
<a name="l00363"></a>00363             <span class="stringliteral">&#39;reactionDNABase&#39;</span>;
<a name="l00364"></a>00364                         }
<a name="l00365"></a>00365         fittedConstantNames__      = {}; %names of fitted constant <a class="code" href="json_parse_8m.html#a6e4256f69b751d3d7ec4ea52ff5a8dee">properties</a>
<a name="l00366"></a>00366         localStateNames__          = {}; %names of simulation state <a class="code" href="json_parse_8m.html#a6e4256f69b751d3d7ec4ea52ff5a8dee">properties</a> redundant with timecourses in <span class="keyword">this</span> or other processes or the simulation
<a name="l00367"></a>00367     end
<a name="l00368"></a>00368 
<a name="l00369"></a>00369     %IDs, names, and local indices
<a name="l00370"></a>00370     <a class="code" href="json_parse_8m.html#a6e4256f69b751d3d7ec4ea52ff5a8dee">properties</a>
<a name="l00371"></a>00371         substrateIndexs_water                       %index within substrates of water
<a name="l00372"></a>00372         substrateIndexs_hydrogen                    %index within substrates of hydrogen
<a name="l00373"></a>00373         substrateIndexs_DR5P                        %index within substrates of deoxyribose-5-phosphate
<a name="l00374"></a>00374         substrateIndexs_dRibose5P_dRibose5P         %index within substrates of (deoxyribose-5-phosphate)-(deoxyribose-5-phosphate)
<a name="l00375"></a>00375         substrateIndexs_NAD                         %index within substrates of NAD
<a name="l00376"></a>00376         substrateIndexs_AMP                         %index within substrates of AMP
<a name="l00377"></a>00377         substrateIndexs_NMN                         %index within substrates of NMN
<a name="l00378"></a>00378         substrateIndexs_PPI                         %index within substrates of diphosphate
<a name="l00379"></a>00379         substrateIndexs_m6AD                        %index within substrates of 6-methyladenine
<a name="l00380"></a>00380         substrateIndexs_undamagedNucleobases        %index within substrates of undamaged nucleobases
<a name="l00381"></a>00381         substrateIndexs_damagedNucleobases          %index within substrates of damaged nucleobases
<a name="l00382"></a>00382         substrateIndexs_damagedNucleosides          %index within substrates of damaged deoxyribonucleosides
<a name="l00383"></a>00383         substrateIndexs_damagedDNMPs                %index within substrates of damaged dNMPs
<a name="l00384"></a>00384         substrateIndexs_dNMPs                       %index within substrates of dNMPs
<a name="l00385"></a>00385         substrateIndexs_dNTPs                       %index within substrates of dNTPs
<a name="l00386"></a>00386         substrateIndexs_dNpdNps                     %index within substrates of (dNTP-dNTP)s
<a name="l00387"></a>00387 
<a name="l00388"></a>00388         enzymeIndexs_formamidopyrimidineGlycosylase %index within enzymes of formamidopyrimidine glycosylase
<a name="l00389"></a>00389         enzymeIndexs_uracilGlycosylase              %index within enzymes of uracil glycosylase
<a name="l00390"></a>00390         enzymeIndexs_apurinicEndonuclease           %index within enzymes of apurinic endonuclease
<a name="l00391"></a>00391         enzymeIndexs_incisionComplex                %index within enzymes of incision complex
<a name="l00392"></a>00392         enzymeIndexs_helicase35                     %index within enzymes of 3&#39;-&gt;5&#39; helicase
<a name="l00393"></a>00393         enzymeIndexs_recombinationStrandExchange    %index within enzymes of recombination strand exchange
<a name="l00394"></a>00394         enzymeIndexs_hollidayJunctionEndonuclease   %index within enzymes of holliday junction endonuclease
<a name="l00395"></a>00395         enzymeIndexs_hollidayJunctionHelicase       %index within enzymes of holliday junction helicase
<a name="l00396"></a>00396         enzymeIndexs_exonuclease53                  %index within enzymes of 5&#39;-&gt;3&#39; exonuclease
<a name="l00397"></a>00397         enzymeIndexs_polymerase                     %index within enzymes of polymerase
<a name="l00398"></a>00398         enzymeIndexs_ligase                         %index within enzymes of ligase
<a name="l00399"></a>00399         enzymeIndexs_phosphoesterase                %index within enzymes of phosphoesterase
<a name="l00400"></a>00400         enzymeIndexs_DisA                           %index within enzymes of DisA
<a name="l00401"></a>00401         enzymeIndexs_RM_typeI                       %index within enzymes of Restriction/Modification typeI
<a name="l00402"></a>00402         enzymeIndexs_RM_typeII                      %index within enzymes of Restriction/Modification typeII
<a name="l00403"></a>00403         enzymeIndexs_BER_baseExcision               %index within enzymes of BER base Excision
<a name="l00404"></a>00404 
<a name="l00405"></a>00405         reactionIndexs_BER                          %index within reactions of BER
<a name="l00406"></a>00406         reactionIndexs_BER_baseexcision             %index within reactions of BER base excision
<a name="l00407"></a>00407         reactionIndexs_BER_aplyase                  %index within reactions of BER aplyase
<a name="l00408"></a>00408         reactionIndexs_BER_drpesterase              %index within reactions of BER deoxyribosose-5&#39;-phosphate esterase
<a name="l00409"></a>00409         reactionIndexs_BER_apendonuclease           %index within reactions of BER AP endonuclease
<a name="l00410"></a>00410         reactionIndexs_BER_dr5plyase                %index within reactions of BER deoxyribosose-5&#39;-phosphate lyase
<a name="l00411"></a>00411         reactionIndexs_NER                          %index within reactions of NER
<a name="l00412"></a>00412         reactionIndexs_NER_incision                 %index within reactions of NER incision
<a name="l00413"></a>00413         reactionIndexs_NER_excision                 %index within reactions of NER excision
<a name="l00414"></a>00414         reactionIndexs_HR_dsbr                      %index within reactions of HR <span class="keywordtype">double</span> strand break repair
<a name="l00415"></a>00415         reactionIndexs_HR_DNAResection              %index within reactions of HR DNA resection
<a name="l00416"></a>00416         reactionIndexs_HR_strandExchange            %index within reactions of HR strand exchange
<a name="l00417"></a>00417         reactionIndexs_HR_junctionMigration         %index within reactions of HR junction migration
<a name="l00418"></a>00418         reactionIndexs_HR_junctionResolution        %index within reactions of HR junction resolution
<a name="l00419"></a>00419         reactionIndexs_ligation                     %index within reactions of ligation
<a name="l00420"></a>00420         reactionIndexs_polymerization               %index within reactions of polymerization
<a name="l00421"></a>00421         reactionIndexs_DNACleavage                  %index within reactions of DNA processive cleavage
<a name="l00422"></a>00422         reactionIndexs_RM                           %index within reactions of Restriction/Modification reactions
<a name="l00423"></a>00423         reactionIndexs_RM_EcoD_Methylation          %index within reactions of Restriction/Modification EcoD methylation
<a name="l00424"></a>00424         reactionIndexs_RM_EcoD_Restriction          %index within reactions of Restriction/Modification EcoD restriction
<a name="l00425"></a>00425         reactionIndexs_RM_MunI_Methylation          %index within reactions of Restriction/Modification MunI methylation
<a name="l00426"></a>00426         reactionIndexs_RM_MunI_Restriction          %index within reactions of Restriction/Modification MunI restriction
<a name="l00427"></a>00427     end
<a name="l00428"></a>00428     
<a name="l00429"></a>00429     %fixed biological constants
<a name="l00430"></a>00430     <a class="code" href="json_parse_8m.html#a6e4256f69b751d3d7ec4ea52ff5a8dee">properties</a>        
<a name="l00431"></a>00431         HR_PolA_ResectionLength                     %number of bases PolA resect DNA ends            8               b [PUB_0526]
<a name="l00432"></a>00432         HR_RecA_Spacing                             %spacing between RecA molecule                   3               b [PUB_0082]
<a name="l00433"></a>00433         HR_RuvAB_JunctionMigrationHop               %RuvAB junction migration hop                    2.2             b [PUB_0538]
<a name="l00434"></a>00434         HR_RecU_CleavageSequence                    %RecU cleavage sequence                          WTTS              [PUB_0538]
<a name="l00435"></a>00435         HR_RecU_CleavagePosition                    %RecU cleavage position within sequence          3                 [PUB_0538]
<a name="l00436"></a>00436         
<a name="l00437"></a>00437         NER_UvrABC_IncisionMargin3                  %3&#39; distance between damage and UvrABC cleavage  4               b [PUB_0461]
<a name="l00438"></a>00438         NER_UvrABC_IncisionMargin5                  %5&#39; distance between damage and UvrABC cleavage  7               b [PUB_0461]
<a name="l00439"></a>00439         NER_PcrA_StepSize                           %PcrA step size                                  1               b [PUB_0516]
<a name="l00440"></a>00440         
<a name="l00441"></a>00441         hollidayJunctionResolutionSites             %(no. sites) x 2 vector indicating positions of holliday junction resolution sites; first col - position; second col - strand        
<a name="l00442"></a>00442         
<a name="l00443"></a>00443         RM_EcoD_RecognitionSequence                 %EcoD recognition sequence                       TTANNNNNNNGTCY    [PUB_0542]
<a name="l00444"></a>00444         RM_EcoD_RecognitionSites                    %(no. sites) x 2 vector indicating positions of restriction/modification methylation sites; first col - position; second col - strand
<a name="l00445"></a>00445         RM_EcoD_MethylationPosition                 %EcoD methylation position within sequence       3                 [PUB_0542]
<a name="l00446"></a>00446         RM_EcoD_RestrictionPosition                 %EcoD restriction position within sequence       []                [PUB_0542]        
<a name="l00447"></a>00447         
<a name="l00448"></a>00448         RM_MunI_RecognitionSequence                 %MunI recognition sequence                       CAATG             [PUB_0096]
<a name="l00449"></a>00449         RM_MunI_RecognitionSites                    %(no. sites) x 2 vector indicating positions of restriction/modification methylation sites; first col - position; second col - strand
<a name="l00450"></a>00450         RM_MunI_MethylationPosition                 %MunI methylation position within sequence       3                 [PUB_0096]
<a name="l00451"></a>00451         RM_MunI_RestrictionPosition                 %MunI restriction position within sequence       1                 [PUB_0544]        
<a name="l00452"></a>00452         RM_MunI_MethylatedPositions                 %columns within RM_MunI_RecognitionSites that represent methylation positions
<a name="l00453"></a>00453         RM_MunI_RestrictionPositions                %columns within RM_MunI_RecognitionSites that represent cleaved phosphodiester bonds
<a name="l00454"></a>00454                
<a name="l00455"></a>00455         reactionSmallMoleculeStoichiometryMatrix    %stoichiometry of free metabolites in chemical reactions (substrates X reactions)
<a name="l00456"></a>00456         reactionDNAStoichiometryMatrix              %stoichiometry of DNA bases in chemical reactions (substrates X reactions)
<a name="l00457"></a>00457         reactionDNABase                             %substrate indices of DNA bases involved in each reaction (reactions X 1)        
<a name="l00458"></a>00458     end
<a name="l00459"></a>00459 
<a name="l00460"></a>00460     %constructor
<a name="l00461"></a>00461     methods
<a name="l00462"></a>00462         function this = DNARepair(wholeCellModelID, name)
<a name="l00463"></a>00463             this = this@edu.stanford.covert.cell.sim.<a class="code" href="class_reaction_process.html">ReactionProcess</a>(wholeCellModelID, name);
<a name="l00464"></a>00464         end
<a name="l00465"></a>00465     end
<a name="l00466"></a>00466 
<a name="l00467"></a>00467     %communication between process/simulation
<a name="l00468"></a>00468     methods
<a name="l00469"></a>00469         %<span class="keyword">set</span> references to state objects
<a name="l00470"></a>00470         <span class="keyword">function</span> storeObjectReferences(<span class="keyword">this</span>, simulation)
<a name="l00471"></a>00471             this.storeObjectReferences@edu.stanford.covert.cell.sim.<a class="code" href="class_process.html">Process</a>(simulation);
<a name="l00472"></a>00472             this.storeObjectReferences@edu.stanford.covert.cell.sim.<a class="code" href="class_chromosome_process_aspect.html">ChromosomeProcessAspect</a>(simulation);
<a name="l00473"></a>00473         end
<a name="l00474"></a>00474         
<a name="l00475"></a>00475         %initialize constants
<a name="l00476"></a>00476         function initializeConstants(this, knowledgeBase, simulation, varargin)
<a name="l00477"></a>00477             %% call super class method to get reactions, substrates, and enzymes
<a name="l00478"></a>00478             this.initializeConstants@edu.stanford.covert.cell.sim.<a class="code" href="class_reaction_process.html">ReactionProcess</a>(knowledgeBase, simulation, varargin{:});
<a name="l00479"></a>00479             this.initializeConstants@edu.stanford.covert.cell.sim.ChromosomeProcessAspect(knowledgeBase, simulation, varargin{:});
<a name="l00480"></a>00480 
<a name="l00481"></a>00481             %% additional substrates
<a name="l00482"></a>00482             additionalSubstrates = findobj(knowledgeBase.metabolites,...
<a name="l00483"></a>00483                 <span class="stringliteral">&#39;category&#39;</span>, <span class="stringliteral">&#39;modified nucleobase&#39;</span>,                        <span class="stringliteral">&#39;-or&#39;</span>,...
<a name="l00484"></a>00484                 <span class="stringliteral">&#39;category&#39;</span>, <span class="stringliteral">&#39;modified deoxyribonucleoside&#39;</span>,               <span class="stringliteral">&#39;-or&#39;</span>,...
<a name="l00485"></a>00485                 <span class="stringliteral">&#39;category&#39;</span>, <span class="stringliteral">&#39;modified deoxyribonucleotide monophosphate&#39;</span>, <span class="stringliteral">&#39;-or&#39;</span>,...
<a name="l00486"></a>00486                 <span class="stringliteral">&#39;wholeCellModelID&#39;</span>,<span class="stringliteral">&#39;URA&#39;</span>,                                 <span class="stringliteral">&#39;-or&#39;</span>,...
<a name="l00487"></a>00487                 <span class="stringliteral">&#39;wholeCellModelID&#39;</span>,<span class="stringliteral">&#39;URI&#39;</span>);
<a name="l00488"></a>00488             substrateWholeCellModelIDs = unique([
<a name="l00489"></a>00489                 this.substrateWholeCellModelIDs;
<a name="l00490"></a>00490                 {additionalSubstrates.wholeCellModelID}<span class="stringliteral">&#39;;</span>
<a name="l00491"></a>00491 <span class="stringliteral">                &#39;</span>AD<span class="stringliteral">&#39;; &#39;</span>CSN<span class="stringliteral">&#39;; &#39;</span>GN<span class="stringliteral">&#39;; &#39;</span>THY<span class="stringliteral">&#39;]);</span>
<a name="l00492"></a>00492 <span class="stringliteral">            tfs = ismember(simulation.state(&#39;</span><a class="code" href="class_metabolite.html">Metabolite</a><span class="stringliteral">&#39;).wholeCellModelIDs, substrateWholeCellModelIDs);</span>
<a name="l00493"></a>00493 <span class="stringliteral">            if sum(tfs) ~= numel(substrateWholeCellModelIDs)</span>
<a name="l00494"></a>00494 <span class="stringliteral">                throw(MException(&#39;</span>DNARepair:error<span class="stringliteral">&#39;, &#39;</span>Substrates must all be metabolites<span class="stringliteral">&#39;));</span>
<a name="l00495"></a>00495 <span class="stringliteral">            end</span>
<a name="l00496"></a>00496 <span class="stringliteral">            substrateWholeCellModelIDs = simulation.state(&#39;</span><a class="code" href="class_metabolite.html">Metabolite</a><span class="stringliteral">&#39;).wholeCellModelIDs(tfs);</span>
<a name="l00497"></a>00497 <span class="stringliteral">            this.initializeConstants_overrideSubstrates(substrateWholeCellModelIDs);</span>
<a name="l00498"></a>00498 <span class="stringliteral">            </span>
<a name="l00499"></a>00499 <span class="stringliteral">            %% additional enzymes</span>
<a name="l00500"></a>00500 <span class="stringliteral">            enzymeWholeCellModelIDs = unique([this.enzymeWholeCellModelIDs; &#39;</span>MG_105_OCTAMER<span class="stringliteral">&#39;]);</span>
<a name="l00501"></a>00501 <span class="stringliteral">            this.initializeConstants_overrideEnzymes(enzymeWholeCellModelIDs);</span>
<a name="l00502"></a>00502 <span class="stringliteral"></span>
<a name="l00503"></a>00503 <span class="stringliteral">            %% parameters</span>
<a name="l00504"></a>00504 <span class="stringliteral">            %knowledge base parameter class holds string values as cell arrays</span>
<a name="l00505"></a>00505 <span class="stringliteral">            this.HR_RecU_CleavageSequence    = this.HR_RecU_CleavageSequence{1};</span>
<a name="l00506"></a>00506 <span class="stringliteral">            this.RM_EcoD_RecognitionSequence = this.RM_EcoD_RecognitionSequence{1};</span>
<a name="l00507"></a>00507 <span class="stringliteral">            this.RM_MunI_RecognitionSequence = this.RM_MunI_RecognitionSequence{1};</span>
<a name="l00508"></a>00508 <span class="stringliteral"></span>
<a name="l00509"></a>00509 <span class="stringliteral">            %% substrate indices</span>
<a name="l00510"></a>00510 <span class="stringliteral">            this.substrateIndexs_water                = this.substrateIndexs({&#39;</span>H2O<span class="stringliteral">&#39;});</span>
<a name="l00511"></a>00511 <span class="stringliteral">            this.substrateIndexs_hydrogen             = this.substrateIndexs({&#39;</span>H<span class="stringliteral">&#39;});</span>
<a name="l00512"></a>00512 <span class="stringliteral">            this.substrateIndexs_DR5P                 = this.substrateIndexs({&#39;</span>DR5P<span class="stringliteral">&#39;});</span>
<a name="l00513"></a>00513 <span class="stringliteral">            this.substrateIndexs_dRibose5P_dRibose5P  = this.substrateIndexs({&#39;</span>dRibose5P_dRibose5P<span class="stringliteral">&#39;});</span>
<a name="l00514"></a>00514 <span class="stringliteral">            this.substrateIndexs_NAD                  = this.substrateIndexs({&#39;</span>NAD<span class="stringliteral">&#39;});</span>
<a name="l00515"></a>00515 <span class="stringliteral">            this.substrateIndexs_AMP                  = this.substrateIndexs({&#39;</span>AMP<span class="stringliteral">&#39;});</span>
<a name="l00516"></a>00516 <span class="stringliteral">            this.substrateIndexs_NMN                  = this.substrateIndexs({&#39;</span>NMN<span class="stringliteral">&#39;});</span>
<a name="l00517"></a>00517 <span class="stringliteral">            this.substrateIndexs_PPI                  = this.substrateIndexs({&#39;</span>PPI<span class="stringliteral">&#39;});</span>
<a name="l00518"></a>00518 <span class="stringliteral">            this.substrateIndexs_m6AD                 = this.substrateIndexs({&#39;</span>m6AD<span class="stringliteral">&#39;});</span>
<a name="l00519"></a>00519 <span class="stringliteral">            this.substrateIndexs_undamagedNucleobases = this.substrateIndexs({&#39;</span>AD<span class="charliteral">&#39;;&#39;</span>CSN<span class="charliteral">&#39;;&#39;</span>GN<span class="charliteral">&#39;;&#39;</span>THY<span class="stringliteral">&#39;});</span>
<a name="l00520"></a>00520 <span class="stringliteral">            this.substrateIndexs_damagedNucleobases   = this.substrateMetaboliteLocalIndexs(...</span>
<a name="l00521"></a>00521 <span class="stringliteral">                strcmp({knowledgeBase.metabolites(this.substrateMetaboliteGlobalIndexs).category},&#39;</span>modified nucleobase<span class="stringliteral">&#39;) | ...</span>
<a name="l00522"></a>00522 <span class="stringliteral">                strcmp({knowledgeBase.metabolites(this.substrateMetaboliteGlobalIndexs).wholeCellModelID},&#39;</span>URA<span class="stringliteral">&#39;));</span>
<a name="l00523"></a>00523 <span class="stringliteral">            this.substrateIndexs_damagedNucleosides = this.substrateMetaboliteLocalIndexs(...</span>
<a name="l00524"></a>00524 <span class="stringliteral">                strcmp({knowledgeBase.metabolites(this.substrateMetaboliteGlobalIndexs).category},&#39;</span>modified deoxyribonucleoside<span class="stringliteral">&#39;) | ...</span>
<a name="l00525"></a>00525 <span class="stringliteral">                strcmp({knowledgeBase.metabolites(this.substrateMetaboliteGlobalIndexs).wholeCellModelID},&#39;</span>URI<span class="stringliteral">&#39;));</span>
<a name="l00526"></a>00526 <span class="stringliteral">            this.substrateIndexs_damagedDNMPs = this.substrateMetaboliteLocalIndexs(...</span>
<a name="l00527"></a>00527 <span class="stringliteral">                strcmp({knowledgeBase.metabolites(this.substrateMetaboliteGlobalIndexs).category},&#39;</span>modified deoxyribonucleotide monophosphate<span class="stringliteral">&#39;) | ...</span>
<a name="l00528"></a>00528 <span class="stringliteral">                strcmp({knowledgeBase.metabolites(this.substrateMetaboliteGlobalIndexs).wholeCellModelID},&#39;</span>DUMP<span class="stringliteral">&#39;));</span>
<a name="l00529"></a>00529 <span class="stringliteral">            this.substrateIndexs_dNMPs                = this.substrateIndexs({&#39;</span>DAMP<span class="charliteral">&#39;;&#39;</span>DCMP<span class="charliteral">&#39;;&#39;</span>DGMP<span class="charliteral">&#39;;&#39;</span>DTMP<span class="stringliteral">&#39;});</span>
<a name="l00530"></a>00530 <span class="stringliteral">            this.substrateIndexs_dNTPs                = this.substrateIndexs({&#39;</span>DATP<span class="charliteral">&#39;;&#39;</span>DCTP<span class="charliteral">&#39;;&#39;</span>DGTP<span class="charliteral">&#39;;&#39;</span>DTTP<span class="stringliteral">&#39;});</span>
<a name="l00531"></a>00531 <span class="stringliteral">            this.substrateIndexs_dNpdNps              = this.substrateIndexs({&#39;</span>dApdAp<span class="charliteral">&#39;;&#39;</span>dCpdCp<span class="charliteral">&#39;;&#39;</span>dGpdGp<span class="charliteral">&#39;;&#39;</span>dTpdTp<span class="stringliteral">&#39;});</span>
<a name="l00532"></a>00532 <span class="stringliteral"></span>
<a name="l00533"></a>00533 <span class="stringliteral">            %% enzyme indices</span>
<a name="l00534"></a>00534 <span class="stringliteral">            this.enzymeIndexs_BER_baseExcision               = this.enzymeIndexs({&#39;</span>MG_097_MONOMER<span class="charliteral">&#39;;&#39;</span>MG_498_MONOMER<span class="stringliteral">&#39;});</span>
<a name="l00535"></a>00535 <span class="stringliteral">            this.enzymeIndexs_formamidopyrimidineGlycosylase = this.enzymeIndexs({&#39;</span>MG_498_MONOMER<span class="stringliteral">&#39;});          %BER  fpg          formamidopyrimidine-DNA glycosylase</span>
<a name="l00536"></a>00536 <span class="stringliteral">            this.enzymeIndexs_uracilGlycosylase              = this.enzymeIndexs({&#39;</span>MG_097_MONOMER<span class="stringliteral">&#39;});          %BER  ung          uracil-DNA glycosylase, putative</span>
<a name="l00537"></a>00537 <span class="stringliteral">            this.enzymeIndexs_apurinicEndonuclease           = this.enzymeIndexs({&#39;</span>MG_235_MONOMER<span class="stringliteral">&#39;});          %BER  nfo          apurinic endonuclease</span>
<a name="l00538"></a>00538 <span class="stringliteral">            this.enzymeIndexs_incisionComplex                = this.enzymeIndexs({&#39;</span>MG_073_206_421_TETRAMER<span class="stringliteral">&#39;}); %NER  uvrABC       DNA incision complex</span>
<a name="l00539"></a>00539 <span class="stringliteral">            this.enzymeIndexs_helicase35                     = this.enzymeIndexs({&#39;</span>MG_244_DIMER<span class="stringliteral">&#39;});            %NER  pcrA         3&#39;</span>-5&#39; helicase
<a name="l00540"></a>00540             this.enzymeIndexs_recombinationStrandExchange    = this.enzymeIndexs({<span class="stringliteral">&#39;MG_339_MONOMER&#39;</span>});          %HR   recA         recombination protein, strand exchange
<a name="l00541"></a>00541             this.enzymeIndexs_hollidayJunctionEndonuclease   = this.enzymeIndexs({<span class="stringliteral">&#39;MG_352_DIMER&#39;</span>});            %HR   recU         Holliday junction endonuclease
<a name="l00542"></a>00542             this.enzymeIndexs_hollidayJunctionHelicase       = this.enzymeIndexs({<span class="stringliteral">&#39;MG_358_359_10MER&#39;</span>});        %HR   ruvAB        Holliday junction DNA helicase
<a name="l00543"></a>00543             this.enzymeIndexs_exonuclease53                  = this.enzymeIndexs({<span class="stringliteral">&#39;MG_262_MONOMER&#39;</span>});          %HR   pol I-like   5&#39;-3&#39; exonuclease, putative
<a name="l00544"></a>00544             this.enzymeIndexs_polymerase                     = this.enzymeIndexs({<span class="stringliteral">&#39;MG_001_MONOMER&#39;</span>});          %     dnaN         DNA polymerase III, beta subunit
<a name="l00545"></a>00545             this.enzymeIndexs_ligase                         = this.enzymeIndexs({<span class="stringliteral">&#39;MG_254_MONOMER&#39;</span>});          %     ligA         DNA ligase, NAD-dependent
<a name="l00546"></a>00546             this.enzymeIndexs_phosphoesterase                = this.enzymeIndexs({<span class="stringliteral">&#39;MG_190_MONOMER&#39;</span>});          %     mgpA         phosphoesterase
<a name="l00547"></a>00547             this.enzymeIndexs_DisA                           = this.enzymeIndexs({<span class="stringliteral">&#39;MG_105_OCTAMER&#39;</span>});          %     disA         DNA integrity scanning protein
<a name="l00548"></a>00548             this.enzymeIndexs_RM_typeI                       = this.enzymeIndexs({<span class="stringliteral">&#39;MG_438_MONOMER&#39;</span>});          %                  type I restriction modification DNA specificity domain protein
<a name="l00549"></a>00549             this.enzymeIndexs_RM_typeII                      = this.enzymeIndexs({<span class="stringliteral">&#39;MG_184_DIMER&#39;</span>});            %                  adenine-specific DNA modification methylase
<a name="l00550"></a>00550 
<a name="l00551"></a>00551             %% reactions
<a name="l00552"></a>00552             %1. Find indices of reactions
<a name="l00553"></a>00553             %2. Factor reactionStoichiometryMatrix into
<a name="l00554"></a>00554             %   - reactionSmallMoleculeStoichiometryMatrix: the stochiometry of
<a name="l00555"></a>00555             %     free metabolites in repair reactions
<a name="l00556"></a>00556             %   - reactionDNAStoichiometryMatrix: the stoichiometry of DNA bases
<a name="l00557"></a>00557             %     in repair reactions
<a name="l00558"></a>00558             %3. Create reactionDNABase to represent the DNA base affected by
<a name="l00559"></a>00559             %   each repair reaction
<a name="l00560"></a>00560             this.reactionSmallMoleculeStoichiometryMatrix = zeros(size(this.reactionStoichiometryMatrix));
<a name="l00561"></a>00561             this.reactionDNAStoichiometryMatrix           = zeros(size(this.reactionStoichiometryMatrix));
<a name="l00562"></a>00562             this.reactionDNABase                          = zeros(size(this.reactionStoichiometryMatrix,2),1);
<a name="l00563"></a>00563 
<a name="l00564"></a>00564             %base excision repair (BER)
<a name="l00565"></a>00565             this.reactionIndexs_BER                = find(strcmp(this.reactionTypes,&#39;base excision repair&#39;) | strcmp(this.reactionTypes,&#39;base excision repair, base excision&#39;));
<a name="l00566"></a>00566             this.reactionIndexs_BER_baseexcision   = find(strcmp(this.reactionTypes,&#39;base excision repair, base excision&#39;));
<a name="l00567"></a>00567             this.reactionIndexs_BER_aplyase        = this.reactionIndexs({<span class="stringliteral">&#39;AP_lyase&#39;</span>});
<a name="l00568"></a>00568             this.reactionIndexs_BER_drpesterase    = this.reactionIndexs({<span class="stringliteral">&#39;deoxyribosephosphodiesterase&#39;</span>});            
<a name="l00569"></a>00569             this.reactionIndexs_BER_apendonuclease = this.reactionIndexs({<span class="stringliteral">&#39;AP_endonuclease&#39;</span>});
<a name="l00570"></a>00570             this.reactionIndexs_BER_dr5plyase      = this.reactionIndexs({<span class="stringliteral">&#39;dRiboseP_lyase&#39;</span>});
<a name="l00571"></a>00571 
<a name="l00572"></a>00572             dnaSubstrateIndexs = [this.substrateIndexs_damagedDNMPs; this.substrateIndexs_dRibose5P_dRibose5P; this.substrateIndexs_DR5P];
<a name="l00573"></a>00573             this.reactionDNAStoichiometryMatrix(dnaSubstrateIndexs, [this.reactionIndexs_BER_baseexcision;this.reactionIndexs_BER_aplyase;this.reactionIndexs_BER_apendonuclease]) = ...
<a name="l00574"></a>00574                 this.reactionStoichiometryMatrix(dnaSubstrateIndexs, [this.reactionIndexs_BER_baseexcision;this.reactionIndexs_BER_aplyase;this.reactionIndexs_BER_apendonuclease]);
<a name="l00575"></a>00575 
<a name="l00576"></a>00576             dnaSubstrateIndexs = [this.substrateIndexs_damagedDNMPs; this.substrateIndexs_dRibose5P_dRibose5P];
<a name="l00577"></a>00577             this.reactionDNAStoichiometryMatrix(dnaSubstrateIndexs, [this.reactionIndexs_BER_drpesterase; this.reactionIndexs_BER_dr5plyase]) = ...
<a name="l00578"></a>00578                 this.reactionStoichiometryMatrix(dnaSubstrateIndexs, [this.reactionIndexs_BER_drpesterase; this.reactionIndexs_BER_dr5plyase]);            
<a name="l00579"></a>00579             this.reactionDNAStoichiometryMatrix(this.substrateIndexs_DR5P, [this.reactionIndexs_BER_drpesterase; this.reactionIndexs_BER_dr5plyase]) = 1;
<a name="l00580"></a>00580 
<a name="l00581"></a>00581             %nucleotide excision repair (NER)
<a name="l00582"></a>00582             this.reactionIndexs_NER                   = find(strcmp(this.reactionTypes,&#39;nucleotide excision repair&#39;));
<a name="l00583"></a>00583             this.reactionIndexs_NER_incision          = this.reactionIndexs({<span class="stringliteral">&#39;NER_DNAIncision&#39;</span>});
<a name="l00584"></a>00584             this.reactionIndexs_NER_excision          = this.reactionIndexs({<span class="stringliteral">&#39;NER_DNAExcision&#39;</span>});
<a name="l00585"></a>00585 
<a name="l00586"></a>00586             dnaSubstrateIndexs = [this.substrateIndexs_damagedDNMPs; this.substrateIndexs_dRibose5P_dRibose5P; this.substrateIndexs_DR5P];
<a name="l00587"></a>00587             this.reactionDNAStoichiometryMatrix(dnaSubstrateIndexs, this.reactionIndexs_NER) = ...
<a name="l00588"></a>00588                 this.reactionStoichiometryMatrix(dnaSubstrateIndexs, this.reactionIndexs_NER);
<a name="l00589"></a>00589 
<a name="l00590"></a>00590             %cleavage
<a name="l00591"></a>00591             this.reactionIndexs_DNACleavage = find(strcmp(this.reactionTypes,<span class="stringliteral">&#39;DNA cleavage&#39;</span>));
<a name="l00592"></a>00592 
<a name="l00593"></a>00593             this.reactionDNAStoichiometryMatrix(this.substrateIndexs_dNpdNps, this.reactionIndexs_DNACleavage) = ...
<a name="l00594"></a>00594                 this.reactionStoichiometryMatrix(this.substrateIndexs_dNpdNps, this.reactionIndexs_DNACleavage);
<a name="l00595"></a>00595             this.reactionDNAStoichiometryMatrix(this.substrateIndexs_dNMPs, this.reactionIndexs_DNACleavage) = ...
<a name="l00596"></a>00596                 0.5 * this.reactionStoichiometryMatrix(this.substrateIndexs_dNMPs, this.reactionIndexs_DNACleavage);
<a name="l00597"></a>00597 
<a name="l00598"></a>00598             %homologous recombination (HR) double strand break repair (DSBR)
<a name="l00599"></a>00599             this.reactionIndexs_HR_dsbr               = find(strcmp(this.reactionTypes,&#39;homologous recombination&#39;));
<a name="l00600"></a>00600             this.reactionIndexs_HR_DNAResection       = this.reactionIndexs({<span class="stringliteral">&#39;HR_DNAResection_dAMP&#39;</span>;<span class="stringliteral">&#39;HR_DNAResection_dCMP&#39;</span>;<span class="stringliteral">&#39;HR_DNAResection_dGMP&#39;</span>;<span class="stringliteral">&#39;HR_DNAResection_dTMP&#39;</span>});
<a name="l00601"></a>00601             this.reactionIndexs_HR_strandExchange     = this.reactionIndexs({<span class="stringliteral">&#39;HR_strandExchange&#39;</span>});
<a name="l00602"></a>00602             this.reactionIndexs_HR_junctionMigration  = this.reactionIndexs({<span class="stringliteral">&#39;HR_junctionMigration&#39;</span>});
<a name="l00603"></a>00603             this.reactionIndexs_HR_junctionResolution = this.reactionIndexs({<span class="stringliteral">&#39;HR_junctionResolution&#39;</span>});
<a name="l00604"></a>00604 
<a name="l00605"></a>00605             dnaSubstrates = [this.substrateIndexs_dNMPs;this.substrateIndexs_dNpdNps;this.substrateIndexs_dRibose5P_dRibose5P;this.substrateIndexs_DR5P];
<a name="l00606"></a>00606             partitions = [0.5*ones(length(this.substrateIndexs_dNMPs),1); ones(length(dnaSubstrates)-length(this.substrateIndexs_dNMPs),1)];
<a name="l00607"></a>00607             this.reactionDNAStoichiometryMatrix(dnaSubstrates, this.reactionIndexs_HR_dsbr) = ...
<a name="l00608"></a>00608                 repmat(partitions, 1, length(this.reactionIndexs_HR_dsbr)) .* ...
<a name="l00609"></a>00609                 this.reactionStoichiometryMatrix(dnaSubstrates, this.reactionIndexs_HR_dsbr);
<a name="l00610"></a>00610             
<a name="l00611"></a>00611             this.hollidayJunctionResolutionSites = this.chromosome.sequence.findSubsequence(...
<a name="l00612"></a>00612                 this.HR_RecU_CleavageSequence, this.HR_RecU_CleavagePosition, <span class="keyword">true</span>);
<a name="l00613"></a>00613 
<a name="l00614"></a>00614             %polymerization
<a name="l00615"></a>00615             this.reactionIndexs_polymerization = find(strcmp(this.reactionTypes,<span class="stringliteral">&#39;DNA polymerization&#39;</span>));
<a name="l00616"></a>00616             dnaSubstrates = [this.substrateIndexs_dNMPs; this.substrateIndexs_dNpdNps];
<a name="l00617"></a>00617             this.reactionDNAStoichiometryMatrix(dnaSubstrates, this.reactionIndexs_polymerization) = ...
<a name="l00618"></a>00618                 this.reactionStoichiometryMatrix(dnaSubstrates, this.reactionIndexs_polymerization);
<a name="l00619"></a>00619 
<a name="l00620"></a>00620             %ligation
<a name="l00621"></a>00621             this.reactionIndexs_ligation = this.reactionIndexs({<span class="stringliteral">&#39;DNA_ligation_repair&#39;</span>});
<a name="l00622"></a>00622             dnaSubstrates = [this.substrateIndexs_DR5P; this.substrateIndexs_dRibose5P_dRibose5P];
<a name="l00623"></a>00623             this.reactionDNAStoichiometryMatrix(dnaSubstrates, this.reactionIndexs_ligation) = ...
<a name="l00624"></a>00624                 this.reactionStoichiometryMatrix(dnaSubstrates, this.reactionIndexs_ligation);
<a name="l00625"></a>00625 
<a name="l00626"></a>00626             %restriction/modification
<a name="l00627"></a>00627             this.reactionIndexs_RM                  = find(strcmp(this.reactionTypes, <span class="stringliteral">&#39;DNA restriction/modification&#39;</span>));
<a name="l00628"></a>00628             this.reactionIndexs_RM_EcoD_Methylation = this.reactionIndexs({<span class="stringliteral">&#39;DNA_RM_EcoD_Methylation&#39;</span>});
<a name="l00629"></a>00629             this.reactionIndexs_RM_EcoD_Restriction = this.reactionIndexs({<span class="stringliteral">&#39;DNA_RM_EcoD_Restriction&#39;</span>});
<a name="l00630"></a>00630             this.reactionIndexs_RM_MunI_Methylation = this.reactionIndexs({<span class="stringliteral">&#39;DNA_RM_MunI_Methylation&#39;</span>});
<a name="l00631"></a>00631             this.reactionIndexs_RM_MunI_Restriction = this.reactionIndexs({<span class="stringliteral">&#39;DNA_RM_MunI_Restriction&#39;</span>});
<a name="l00632"></a>00632 
<a name="l00633"></a>00633             dnaSubstrates = [this.substrateIndexs_undamagedNucleobases;this.substrateIndexs_damagedNucleobases;this.substrateIndexs_dRibose5P_dRibose5P;this.substrateIndexs_DR5P];
<a name="l00634"></a>00634             this.reactionDNAStoichiometryMatrix(dnaSubstrates, this.reactionIndexs_RM) = ...
<a name="l00635"></a>00635                 this.reactionStoichiometryMatrix(dnaSubstrates, this.reactionIndexs_RM);
<a name="l00636"></a>00636 
<a name="l00637"></a>00637             this.RM_MunI_MethylatedPositions = [
<a name="l00638"></a>00638                 this.RM_MunI_MethylationPosition;
<a name="l00639"></a>00639                 length(this.RM_MunI_RecognitionSequence)-this.RM_MunI_MethylationPosition+1]<span class="stringliteral">&#39;;</span>
<a name="l00640"></a>00640 <span class="stringliteral">            this.RM_MunI_RestrictionPositions = [</span>
<a name="l00641"></a>00641 <span class="stringliteral">                this.RM_MunI_RestrictionPosition;</span>
<a name="l00642"></a>00642 <span class="stringliteral">                length(this.RM_MunI_RecognitionSequence)-this.RM_MunI_RestrictionPosition]&#39;</span>;
<a name="l00643"></a>00643 
<a name="l00644"></a>00644             this.RM_EcoD_RecognitionSites = this.chromosome.sequence.findSubsequence(this.RM_EcoD_RecognitionSequence, 1);
<a name="l00645"></a>00645             this.RM_EcoD_RecognitionSites = mod(...
<a name="l00646"></a>00646                 repmat(this.RM_EcoD_RecognitionSites,1,length(this.RM_EcoD_RecognitionSequence)) + ...
<a name="l00647"></a>00647                 repmat(0:length(this.RM_EcoD_RecognitionSequence)-1,size(this.RM_EcoD_RecognitionSites,1),1) - ...
<a name="l00648"></a>00648                 1, this.chromosome.sequenceLen)+1;
<a name="l00649"></a>00649 
<a name="l00650"></a>00650             this.RM_MunI_RecognitionSites = this.chromosome.sequence.findSubsequence(this.RM_MunI_RecognitionSequence, 1);
<a name="l00651"></a>00651             this.RM_MunI_RecognitionSites = mod(...
<a name="l00652"></a>00652                 repmat(this.RM_MunI_RecognitionSites,1,length(this.RM_MunI_RecognitionSequence)) + ...
<a name="l00653"></a>00653                 repmat(0:length(this.RM_MunI_RecognitionSequence)-1,size(this.RM_MunI_RecognitionSites,1),1) - ...
<a name="l00654"></a>00654                 1, this.chromosome.sequenceLen)+1;
<a name="l00655"></a>00655 
<a name="l00656"></a>00656             % reactionSmallMoleculeStoichiometryMatrix
<a name="l00657"></a>00657             this.reactionSmallMoleculeStoichiometryMatrix = ...
<a name="l00658"></a>00658                 this.reactionStoichiometryMatrix - ...
<a name="l00659"></a>00659                 this.reactionDNAStoichiometryMatrix;
<a name="l00660"></a>00660 
<a name="l00661"></a>00661             % reactionDNABase
<a name="l00662"></a>00662             [metaboliteIdx, reactionIdx] = find(this.reactionDNAStoichiometryMatrix(this.substrateIndexs_damagedDNMPs, :)&lt;0);
<a name="l00663"></a>00663             this.reactionDNABase(reactionIdx) = this.substrateIndexs_damagedDNMPs(metaboliteIdx);
<a name="l00664"></a>00664         end
<a name="l00665"></a>00665     end
<a name="l00666"></a>00666 
<a name="l00667"></a>00667     %model
<a name="l00668"></a>00668     methods
<a name="l00669"></a>00669         %Calculate
<a name="l00670"></a>00670         %- contribution to <a class="code" href="class_f_b_a.html">FBA</a> objective
<a name="l00671"></a>00671         %- minimum expression consistent with cell cycle length
<a name="l00672"></a>00672         function [bmProd, byProd, minEnzExp, maxEnzExp] = calcResourceRequirements_LifeCycle(this, constants, ~)
<a name="l00673"></a>00673             %expected rates of reactions (repair and restriction/modification)
<a name="l00674"></a>00674             rates = this.calcExpectedReactionRates(constants);
<a name="l00675"></a>00675             
<a name="l00676"></a>00676             %substrate and byproducts: metabolic load of reactions
<a name="l00677"></a>00677             bmProd = max(0, -this.reactionSmallMoleculeStoichiometryMatrix) * rates;
<a name="l00678"></a>00678             byProd = max(0,  this.reactionSmallMoleculeStoichiometryMatrix) * rates;
<a name="l00679"></a>00679             
<a name="l00680"></a>00680             %enzymes
<a name="l00681"></a>00681             minEnzExp = max(...
<a name="l00682"></a>00682                 this.reactionCatalysisMatrix<span class="stringliteral">&#39; * (rates ./ this.enzymeBounds(:, 2)) / constants.states.Time.cellCycleLength * log(2), ...</span>
<a name="l00683"></a>00683 <span class="stringliteral">                this.reactionCatalysisMatrix(this.reactionIndexs_RM_MunI_Methylation, :)&#39;</span> * ...
<a name="l00684"></a>00684                     (rates(this.reactionIndexs_RM_MunI_Methylation) ./ this.enzymeBounds(this.reactionIndexs_RM_MunI_Methylation, 2)) / constants.states.Time.cellCycleLength * log(2) / ...
<a name="l00685"></a>00685                     (exp(log(2) * (constants.states.Time.replicationInitiationDuration + constants.states.Time.replicationDuration) / constants.states.Time.cellCycleLength) - ...
<a name="l00686"></a>00686                      exp(log(2) * constants.states.Time.replicationInitiationDuration / constants.states.Time.cellCycleLength)) ...
<a name="l00687"></a>00687                 );
<a name="l00688"></a>00688             maxEnzExp = Inf(size(this.enzymeWholeCellModelIDs));
<a name="l00689"></a>00689         end
<a name="l00690"></a>00690         
<a name="l00691"></a>00691         <span class="keyword">function</span> repairRates = calcExpectedReactionRates(<span class="keyword">this</span>, constants)
<a name="l00692"></a>00692             repairRates = zeros(numel(this.reactionWholeCellModelIDs), 1);
<a name="l00693"></a>00693             
<a name="l00694"></a>00694             %% repair of damaged bases: calculate expected number of occurences of
<a name="l00695"></a>00695             %each reaction over cell cycle
<a name="l00696"></a>00696             c = this.chromosome;
<a name="l00697"></a>00697             dmg = constants.processes.<a class="code" href="class_d_n_a_damage.html">DNADamage</a>;
<a name="l00698"></a>00698             polDNA = collapse(c.polymerizedRegions);
<a name="l00699"></a>00699             bases = &#39;ACGT&#39;;
<a name="l00700"></a>00700             dntpComposition = [
<a name="l00701"></a>00701                 1 - c.sequenceGCContent
<a name="l00702"></a>00702                 c.sequenceGCContent
<a name="l00703"></a>00703                 c.sequenceGCContent
<a name="l00704"></a>00704                 1 - c.sequenceGCContent] / 2;
<a name="l00705"></a>00705             
<a name="l00706"></a>00706             for i = 1:size(dmg.reactionStoichiometryMatrix, 2)
<a name="l00707"></a>00707                 if dmg.reactionBounds(i, 2) == 0
<a name="l00708"></a>00708                     continue;
<a name="l00709"></a>00709                 end
<a name="l00710"></a>00710                 
<a name="l00711"></a>00711                 %number of sites vulnerable to reaction per chromosome
<a name="l00712"></a>00712                 if ischar(dmg.reactionVulnerableMotifs{i})
<a name="l00713"></a>00713                     nVulnerableSites = polDNA * prod(dntpComposition(ismembc(dmg.reactionVulnerableMotifs{i}, bases)));
<a name="l00714"></a>00714                 <span class="keywordflow">else</span>
<a name="l00715"></a>00715                     nVulnerableSites = collapse(...
<a name="l00716"></a>00716                         dmg.reactionVulnerableMotifs{i} == c.(dmg.reactionVulnerableMotifTypes{i}) &amp; ...
<a name="l00717"></a>00717                         dmg.reactionVulnerableMotifs{i} == c.damagedSites_nonRedundant);
<a name="l00718"></a>00718                 end
<a name="l00719"></a>00719                 <span class="keywordflow">if</span> nVulnerableSites == 0
<a name="l00720"></a>00720                     <span class="keywordflow">continue</span>;
<a name="l00721"></a>00721                 end
<a name="l00722"></a>00722                 
<a name="l00723"></a>00723                 %calculate rate, product of
<a name="l00724"></a>00724                 %- number of vulnerable site / chromosome
<a name="l00725"></a>00725                 %- integrated number of chromosomes over cell cycle
<a name="l00726"></a>00726                 %- reaction rate per vulnerable site (and per <a class="code" href="unit_8m.html#a94c091bed34da4edc337e42e23029000">unit</a> radiation
<a name="l00727"></a>00727                 %  for applicable reactions)
<a name="l00728"></a>00728                 %- amount of radition, for applicable reactions
<a name="l00729"></a>00729                 iRadLcl = find(dmg.reactionRadiationStoichiometryMatrix(:, i));
<a name="l00730"></a>00730                 
<a name="l00731"></a>00731                 damageRate = nVulnerableSites * ...
<a name="l00732"></a>00732                     (1 * constants.states.Time.replicationInitiationDuration + 1.5 * constants.states.Time.replicationDuration + 2 * constants.states.Time.cytokinesisDuration) * ...
<a name="l00733"></a>00733                     dmg.reactionBounds(i, 2);
<a name="l00734"></a>00734                 <span class="keywordflow">if</span> ~isempty(iRadLcl)
<a name="l00735"></a>00735                     iRadGbl = dmg.substrateStimulusGlobalCompartmentIndexs(dmg.substrateStimulusLocalIndexs == iRadLcl);
<a name="l00736"></a>00736                     damageRate = damageRate * this.stimulus.values(iRadGbl);
<a name="l00737"></a>00737                 end
<a name="l00738"></a>00738                 
<a name="l00739"></a>00739                 switch dmg.reactionDamageTypes{i}
<a name="l00740"></a>00740                     <span class="keywordflow">case</span> <span class="stringliteral">&#39;abasicSites&#39;</span>
<a name="l00741"></a>00741                         repairRates(this.reactionIndexs_BER_aplyase) = ...
<a name="l00742"></a>00742                             repairRates(this.reactionIndexs_BER_aplyase) + ...
<a name="l00743"></a>00743                             damageRate / 2;
<a name="l00744"></a>00744                         repairRates(this.reactionIndexs_BER_drpesterase) = ...
<a name="l00745"></a>00745                             repairRates(this.reactionIndexs_BER_drpesterase) + ...
<a name="l00746"></a>00746                             damageRate / 2;
<a name="l00747"></a>00747                         
<a name="l00748"></a>00748                         repairRates(this.reactionIndexs_BER_apendonuclease) = ...
<a name="l00749"></a>00749                             repairRates(this.reactionIndexs_BER_apendonuclease) + ...
<a name="l00750"></a>00750                             damageRate / 2;
<a name="l00751"></a>00751                         repairRates(this.reactionIndexs_BER_dr5plyase) = ...
<a name="l00752"></a>00752                             repairRates(this.reactionIndexs_BER_dr5plyase) + ...
<a name="l00753"></a>00753                             damageRate / 2;
<a name="l00754"></a>00754                         
<a name="l00755"></a>00755                         repairRates(this.reactionIndexs_polymerization) = ...
<a name="l00756"></a>00756                             repairRates(this.reactionIndexs_polymerization) + ...
<a name="l00757"></a>00757                             damageRate * dntpComposition;
<a name="l00758"></a>00758                         repairRates(this.reactionIndexs_ligation) = ...
<a name="l00759"></a>00759                             repairRates(this.reactionIndexs_ligation) + ...
<a name="l00760"></a>00760                             damageRate;
<a name="l00761"></a>00761                     <span class="keywordflow">case</span> <span class="stringliteral">&#39;damagedBases&#39;</span>
<a name="l00762"></a>00762                         repairRates(this.reactionIndexs_BER_baseexcision) = ...
<a name="l00763"></a>00763                             repairRates(this.reactionIndexs_BER_baseexcision) + ...
<a name="l00764"></a>00764                             damageRate / 2;
<a name="l00765"></a>00765                         
<a name="l00766"></a>00766                         repairRates(this.reactionIndexs_BER_aplyase) = ...
<a name="l00767"></a>00767                             repairRates(this.reactionIndexs_BER_aplyase) + ...
<a name="l00768"></a>00768                             damageRate / 2;
<a name="l00769"></a>00769                         repairRates(this.reactionIndexs_BER_drpesterase) = ...
<a name="l00770"></a>00770                             repairRates(this.reactionIndexs_BER_drpesterase) + ...
<a name="l00771"></a>00771                             damageRate / 2;
<a name="l00772"></a>00772                         
<a name="l00773"></a>00773                         repairRates(this.reactionIndexs_BER_apendonuclease) = ...
<a name="l00774"></a>00774                             repairRates(this.reactionIndexs_BER_apendonuclease) + ...
<a name="l00775"></a>00775                             damageRate / 2;
<a name="l00776"></a>00776                         repairRates(this.reactionIndexs_BER_dr5plyase) = ...
<a name="l00777"></a>00777                             repairRates(this.reactionIndexs_BER_dr5plyase) + ...
<a name="l00778"></a>00778                             damageRate / 2;
<a name="l00779"></a>00779                         
<a name="l00780"></a>00780                         repairRates(this.reactionIndexs_polymerization) = ...
<a name="l00781"></a>00781                             repairRates(this.reactionIndexs_polymerization) + ...
<a name="l00782"></a>00782                             damageRate * dntpComposition;
<a name="l00783"></a>00783                         repairRates(this.reactionIndexs_ligation) = ...
<a name="l00784"></a>00784                             repairRates(this.reactionIndexs_ligation) + ...
<a name="l00785"></a>00785                             damageRate;
<a name="l00786"></a>00786                     <span class="keywordflow">case</span> <span class="stringliteral">&#39;intrastrandCrossLinks&#39;</span>
<a name="l00787"></a>00787                         numExcisions = (this.NER_UvrABC_IncisionMargin5 + this.NER_UvrABC_IncisionMargin3 + 1);
<a name="l00788"></a>00788                         
<a name="l00789"></a>00789                         repairRates(this.reactionIndexs_NER_incision) = ...
<a name="l00790"></a>00790                             repairRates(this.reactionIndexs_NER_incision) + ...
<a name="l00791"></a>00791                             damageRate;
<a name="l00792"></a>00792                         repairRates(this.reactionIndexs_NER_excision) = ...
<a name="l00793"></a>00793                             repairRates(this.reactionIndexs_NER_excision) + ...
<a name="l00794"></a>00794                             damageRate * ceil(numExcisions / this.NER_PcrA_StepSize);
<a name="l00795"></a>00795                         repairRates(this.reactionIndexs_DNACleavage) = ...
<a name="l00796"></a>00796                             repairRates(this.reactionIndexs_DNACleavage) + ...
<a name="l00797"></a>00797                             damageRate * numExcisions * dntpComposition;
<a name="l00798"></a>00798                         
<a name="l00799"></a>00799                         repairRates(this.reactionIndexs_polymerization) = ...
<a name="l00800"></a>00800                             repairRates(this.reactionIndexs_polymerization) + ...
<a name="l00801"></a>00801                             damageRate * 2 * dntpComposition;
<a name="l00802"></a>00802                         repairRates(this.reactionIndexs_ligation) = ...
<a name="l00803"></a>00803                             repairRates(this.reactionIndexs_ligation) + ...
<a name="l00804"></a>00804                             damageRate;
<a name="l00805"></a>00805                     <span class="keywordflow">case</span> <span class="stringliteral">&#39;strandBreaks&#39;</span>
<a name="l00806"></a>00806                         repairRates(this.reactionIndexs_ligation) = ...
<a name="l00807"></a>00807                             repairRates(this.reactionIndexs_ligation) + ...
<a name="l00808"></a>00808                             damageRate;
<a name="l00809"></a>00809                     otherwise
<a name="l00810"></a>00810                         <span class="keywordflow">throw</span>(MException(<span class="stringliteral">&#39;DNARepair:error&#39;</span>, <span class="stringliteral">&#39;undefined damage reaction type %s&#39;</span>, dmg.reactionDamageTypes{i}));
<a name="l00811"></a>00811                 end
<a name="l00812"></a>00812             end
<a name="l00813"></a>00813             
<a name="l00814"></a>00814             %methylation of restriction/modification sites
<a name="l00815"></a>00815             repairRates(this.reactionIndexs_RM_MunI_Methylation) = 2 * size(this.RM_MunI_RecognitionSites, 1);
<a name="l00816"></a>00816         end
<a name="l00817"></a>00817         
<a name="l00818"></a>00818         %initialization: 
<a name="l00819"></a>00819         %- Simulation initializeState initialize 1 undamaged chromosome
<a name="l00820"></a>00820         %- Here we methylate all restriction/modification sites
<a name="l00821"></a>00821         function initializeState(this)
<a name="l00822"></a>00822             c = this.chromosome;
<a name="l00823"></a>00823             iMet = this.substrateGlobalIndexs(this.substrateIndexs_m6AD);
<a name="l00824"></a>00824             c.damagedBases(this.RM_MunI_RecognitionSites(:, this.RM_MunI_MethylatedPositions(1)), 1) = iMet;
<a name="l00825"></a>00825             c.damagedBases(this.RM_MunI_RecognitionSites(:, this.RM_MunI_MethylatedPositions(2)), 2) = iMet;
<a name="l00826"></a>00826         end
<a name="l00827"></a>00827 
<a name="l00828"></a>00828         %resource requirements
<a name="l00829"></a>00829         <span class="keyword">function</span> result = calcResourceRequirements_Current(<span class="keyword">this</span>)
<a name="l00830"></a>00830             c = this.chromosome;
<a name="l00831"></a>00831             
<a name="l00832"></a>00832             %% numbers of damages
<a name="l00833"></a>00833             nGapSites = nnz(c.gapSites);
<a name="l00834"></a>00834             nAbasicSites = nnz(c.abasicSites);
<a name="l00835"></a>00835             if nnz(c.damagedSites)
<a name="l00836"></a>00836                 [posStrnds, bases] = find(c.damagedBases);
<a name="l00837"></a>00837                 nDamagedBases = numel(bases) - sum(...
<a name="l00838"></a>00838                     bases == this.metabolite.m6ADIndexs &amp; ...
<a name="l00839"></a>00839                     ((isodd(posStrnds(:, 2)) &amp; ismembc(posStrnds(:, 1), this.RM_MunI_RecognitionSites(:, this.RM_MunI_MethylatedPositions(1)))) | ...
<a name="l00840"></a>00840                     (iseven(posStrnds(:, 2)) &amp; ismembc(posStrnds(:, 1), this.RM_MunI_RecognitionSites(:, this.RM_MunI_MethylatedPositions(2))))));
<a name="l00841"></a>00841             else
<a name="l00842"></a>00842                 nDamagedBases = 0;
<a name="l00843"></a>00843             end
<a name="l00844"></a>00844             nIntrastrandCrossLinks = nnz(c.intrastrandCrossLinks);
<a name="l00845"></a>00845             nStrandBreaks = nnz(c.strandBreaks);
<a name="l00846"></a>00846             if nStrandBreaks &gt; 0
<a name="l00847"></a>00847                 nDSBR = nnz(c.doubleStrandBreaks);
<a name="l00848"></a>00848             else
<a name="l00849"></a>00849                 nDSBR = 0;
<a name="l00850"></a>00850             end
<a name="l00851"></a>00851             nHJ = nnz(c.hollidayJunctions) / 2;
<a name="l00852"></a>00852             
<a name="l00853"></a>00853             %% reaction rates
<a name="l00854"></a>00854             rates = zeros(size(this.reactionSmallMoleculeStoichiometryMatrix, 2), 1);
<a name="l00855"></a>00855             
<a name="l00856"></a>00856             %BER
<a name="l00857"></a>00857             rates(this.reactionIndexs_BER_baseexcision)   = nDamagedBases;
<a name="l00858"></a>00858             rates(this.reactionIndexs_BER_apendonuclease) = 0.5 * nAbasicSites;
<a name="l00859"></a>00859             rates(this.reactionIndexs_BER_aplyase)        = 0.5 * nAbasicSites;
<a name="l00860"></a>00860             rates(this.reactionIndexs_BER_drpesterase)    = 0.5 * nAbasicSites;
<a name="l00861"></a>00861             rates(this.reactionIndexs_BER_dr5plyase)      = 0.5 * nAbasicSites;
<a name="l00862"></a>00862             
<a name="l00863"></a>00863             %NER
<a name="l00864"></a>00864             nNER = nDamagedBases + nIntrastrandCrossLinks;
<a name="l00865"></a>00865             nExcisions = (this.NER_UvrABC_IncisionMargin5 + this.NER_UvrABC_IncisionMargin3 + 1);
<a name="l00866"></a>00866             rates(this.reactionIndexs_NER_incision) = nNER;
<a name="l00867"></a>00867             rates(this.reactionIndexs_NER_excision) = nNER * nExcisions;
<a name="l00868"></a>00868             rates(this.reactionIndexs_DNACleavage)  = nNER * nExcisions/4;
<a name="l00869"></a>00869             
<a name="l00870"></a>00870             %HR
<a name="l00871"></a>00871             numResections = 2 * this.HR_PolA_ResectionLength;
<a name="l00872"></a>00872             avgJunctionMigrationLength = c.sequenceLen / (size(this.hollidayJunctionResolutionSites, 2) / 2) / 2;
<a name="l00873"></a>00873             rates(this.reactionIndexs_HR_DNAResection)       = nDSBR * numResections;
<a name="l00874"></a>00874             rates(this.reactionIndexs_HR_strandExchange)     = nDSBR * 2 + nHJ * 2;
<a name="l00875"></a>00875             rates(this.reactionIndexs_HR_junctionMigration)  = nHJ * avgJunctionMigrationLength / this.HR_RuvAB_JunctionMigrationHop;
<a name="l00876"></a>00876             rates(this.reactionIndexs_HR_junctionResolution) = nHJ;
<a name="l00877"></a>00877             
<a name="l00878"></a>00878             %polymerization/ligation
<a name="l00879"></a>00879             if nGapSites &amp;&amp; nAbasicSites
<a name="l00880"></a>00880                 rates(this.reactionIndexs_polymerization) = nnz(c.gapSites &amp; c.abasicSites);
<a name="l00881"></a>00881             end
<a name="l00882"></a>00882             rates(this.reactionIndexs_ligation) = nStrandBreaks;
<a name="l00883"></a>00883             
<a name="l00884"></a>00884             %restriction / modification
<a name="l00885"></a>00885             rates(this.reactionIndexs_RM_MunI_Methylation) = nnz(c.hemiunmethylatedMunIRMSites);
<a name="l00886"></a>00886             rates(this.reactionIndexs_RM_MunI_Restriction) = nnz(c.restrictableMunIRMSites);
<a name="l00887"></a>00887             
<a name="l00888"></a>00888             %% metabolic demands
<a name="l00889"></a>00889             result = max(0, -this.reactionSmallMoleculeStoichiometryMatrix) * ...
<a name="l00890"></a>00890                 min(ceil(this.reactionCatalysisMatrix * this.enzymes * this.stepSizeSec), rates);
<a name="l00891"></a>00891         end
<a name="l00892"></a>00892 
<a name="l00893"></a>00893         %simulation
<a name="l00894"></a>00894         function evolveState(this)
<a name="l00895"></a>00895             if nnz(this.chromosome.damagedSites) == 0
<a name="l00896"></a>00896                 %there is no damage, no repair necessary
<a name="l00897"></a>00897                 if this.randStream.rand &gt; 0.5
<a name="l00898"></a>00898                     this.evolveState_Modification();
<a name="l00899"></a>00899                     this.evolveState_Restriction();
<a name="l00900"></a>00900                 else
<a name="l00901"></a>00901                     this.evolveState_Restriction();
<a name="l00902"></a>00902                     this.evolveState_Modification();
<a name="l00903"></a>00903                 end
<a name="l00904"></a>00904             else
<a name="l00905"></a>00905                 %there is damage, repair necessary
<a name="l00906"></a>00906                 subfunctions = {
<a name="l00907"></a>00907                     @this.evolveState_BER;          %base excision repair (BER)
<a name="l00908"></a>00908                     @this.evolveState_NER;          %nucleotide excision repair (NER)
<a name="l00909"></a>00909                     @this.evolveState_HR;           %homologous recombination (HR) double strand break repair (DSBR)
<a name="l00910"></a>00910                     @this.evolveState_Polymerize;   %polymerize DNA
<a name="l00911"></a>00911                     @this.evolveState_Ligate;       %ligate DNA
<a name="l00912"></a>00912                     @this.evolveState_Modification; %Modification
<a name="l00913"></a>00913                     @this.evolveState_Restriction;  %Restriction
<a name="l00914"></a>00914                     @this.evolveState_DisA;         %DNA integrity scanning protein
<a name="l00915"></a>00915                     };
<a name="l00916"></a>00916                 order = this.randStream.randperm(numel(subfunctions));
<a name="l00917"></a>00917                 for i = 1:numel(subfunctions)
<a name="l00918"></a>00918                     subfunctions{order(i)}();
<a name="l00919"></a>00919                 end
<a name="l00920"></a>00920             end
<a name="l00921"></a>00921         end
<a name="l00922"></a>00922     end
<a name="l00923"></a>00923 
<a name="l00924"></a>00924     %model helper functions
<a name="l00925"></a>00925     methods
<a name="l00926"></a>00926         %base excision repair
<a name="l00927"></a>00927         <span class="keyword">function</span> evolveState_BER(<span class="keyword">this</span>)
<a name="l00928"></a>00928             c = this.chromosome;
<a name="l00929"></a>00929             
<a name="l00930"></a>00930             %1. Base excision by
<a name="l00931"></a>00931             %   - fpg: FAPyA, FAPyG, 8-oxo-G, 8-oxo-A (MG_498_MONOMER)
<a name="l00932"></a>00932             %   - ung: uracil, ho5U (MG_097_MONOMER)
<a name="l00933"></a>00933 
<a name="l00934"></a>00934             %Loop over base excision enzymes (fpg, ung)
<a name="l00935"></a>00935             if nnz(c.damagedBases)
<a name="l00936"></a>00936                 randOrder = this.randStream.randperm(length(this.enzymeIndexs_BER_baseExcision));
<a name="l00937"></a>00937                 for idx = 1:length(this.enzymeIndexs_BER_baseExcision)
<a name="l00938"></a>00938                     i = randOrder(idx);
<a name="l00939"></a>00939                     ezIdx = this.enzymeIndexs_BER_baseExcision(i);
<a name="l00940"></a>00940                     
<a name="l00941"></a>00941                     %<span class="keywordtype">short</span> cut if no enzyme
<a name="l00942"></a>00942                     if this.enzymes(ezIdx) == 0; continue; end
<a name="l00943"></a>00943                     
<a name="l00944"></a>00944                     %Find damaged bases that enzyme can excise
<a name="l00945"></a>00945                     rxIdxs = this.reactionIndexs_BER_baseexcision(this.reactionCatalysisMatrix(this.reactionIndexs_BER_baseexcision, ezIdx) ~= 0);
<a name="l00946"></a>00946                     [positionsStrands, damagedBaseGlobalIndexs] = find(c.damagedBases);
<a name="l00947"></a>00947                     damagedBaseLocalIndexs = ismembc2(damagedBaseGlobalIndexs, this.substrateGlobalIndexs);
<a name="l00948"></a>00948                     [tfs, idxs] = ismember(damagedBaseLocalIndexs, this.reactionDNABase(rxIdxs));
<a name="l00949"></a>00949                     positionsStrands(~tfs, :) = [];
<a name="l00950"></a>00950                     reactions = rxIdxs(idxs(tfs));
<a name="l00951"></a>00951                     
<a name="l00952"></a>00952                     %<span class="keywordtype">short</span> cut if no positions to be repaired
<a name="l00953"></a>00953                     if isempty(positionsStrands); continue; end                    
<a name="l00954"></a>00954                     
<a name="l00955"></a>00955                     %Randomly select bases to excise up to available resources,
<a name="l00956"></a>00956                     %(note this is valid because each reaction has the same
<a name="l00957"></a>00957                     %enzymatic rate and free metabolite reactants)
<a name="l00958"></a>00958                     numReactions = max(0, floor(min([
<a name="l00959"></a>00959                         numel(reactions);
<a name="l00960"></a>00960                         this.randStream.stochasticRound(this.enzymes(ezIdx) * this.enzymeBounds(reactions(1), 2) * this.stepSizeSec);
<a name="l00961"></a>00961                         this.substrates ./ max(0, -this.reactionSmallMoleculeStoichiometryMatrix(:, reactions(1)))])));
<a name="l00962"></a>00962                     if numReactions == 0; continue; end         
<a name="l00963"></a>00963                     
<a name="l00964"></a>00964                     [~, idxs, positionsStrands] = this.bindProteinToChromosome(positionsStrands, ...
<a name="l00965"></a>00965                         ezIdx, numReactions, [], false, true, 1, false, 1);
<a name="l00966"></a>00966                     if isempty(idxs); continue; end
<a name="l00967"></a>00967                     reactions = reactions(idxs, :);
<a name="l00968"></a>00968                     
<a name="l00969"></a>00969                     %update state
<a name="l00970"></a>00970                     this.substrates = this.substrates + ...
<a name="l00971"></a>00971                         sum(this.reactionSmallMoleculeStoichiometryMatrix(:, reactions), 2);
<a name="l00972"></a>00972                     c.damagedBases(positionsStrands) = 0;
<a name="l00973"></a>00973                     c.abasicSites( positionsStrands) = 1;
<a name="l00974"></a>00974                 end
<a name="l00975"></a>00975             end
<a name="l00976"></a>00976 
<a name="l00977"></a>00977             %2. Sugar excision
<a name="l00978"></a>00978             %   a. i.  3&#39;-AP lyase Fpg (MG_498)
<a name="l00979"></a>00979             %      ii. 5&#39;-deoxyribosephosphodiesterase Nfo (MG_235)
<a name="l00980"></a>00980             %   b. i.  5&#39;-AP endonuclease Nfo (MG_235)
<a name="l00981"></a>00981             %      ii. 3&#39;-(deoxyribose-5&#39;-phosphate) lyase dnaN (MG_001)
<a name="l00982"></a>00982 
<a name="l00983"></a>00983             %3&#39;-AP lyase (MG_498) / 5&#39;-AP endonuclease Nfo (MG_235)
<a name="l00984"></a>00984             if nnz(c.abasicSites)
<a name="l00985"></a>00985                 rxIdxs = [this.reactionIndexs_BER_aplyase;this.reactionIndexs_BER_apendonuclease];
<a name="l00986"></a>00986                 randOrder = this.randStream.randperm(length(rxIdxs));
<a name="l00987"></a>00987                 for i = 1:length(rxIdxs)
<a name="l00988"></a>00988                     rxIdx = rxIdxs(randOrder(i));
<a name="l00989"></a>00989                     ezIdx = find(this.reactionCatalysisMatrix(rxIdx,:));
<a name="l00990"></a>00990                     if this.enzymes(ezIdx) == 0; continue; end
<a name="l00991"></a>00991                     
<a name="l00992"></a>00992                     positionsStrands = find(c.abasicSites &amp; xor(c.abasicSites,  c.strandBreaks3 | c.strandBreaks5));
<a name="l00993"></a>00993                     if isempty(positionsStrands); continue; end
<a name="l00994"></a>00994                     
<a name="l00995"></a>00995                     numReactions = max(0, floor(min([
<a name="l00996"></a>00996                         size(positionsStrands, 1);
<a name="l00997"></a>00997                         this.randStream.stochasticRound(this.enzymes(ezIdx) * this.enzymeBounds(rxIdx, 2) * this.stepSizeSec);
<a name="l00998"></a>00998                         this.substrates ./ max(0, -this.reactionSmallMoleculeStoichiometryMatrix(:, rxIdx))])));
<a name="l00999"></a>00999                     if numReactions == 0; continue; end
<a name="l01000"></a>01000                     
<a name="l01001"></a>01001                     if rxIdx == this.reactionIndexs_BER_aplyase
<a name="l01002"></a>01002                         checkPositionsStrands = positionsStrands;
<a name="l01003"></a>01003                         checkPositionsStrands(mod(positionsStrands(:,2),2)==0, 1) = checkPositionsStrands(mod(positionsStrands(:,2),2)==0, 1) - 1;
<a name="l01004"></a>01004                         [~, idxs, ~, numReactions] = this.bindProteinToChromosome(checkPositionsStrands, ...
<a name="l01005"></a>01005                             ezIdx, numReactions, [], false, true, 2, false, 1);
<a name="l01006"></a>01006                         if numReactions == 0; continue; end
<a name="l01007"></a>01007                         positionsStrands = positionsStrands(idxs, :);
<a name="l01008"></a>01008                         
<a name="l01009"></a>01009                         c.strandBreaks(c.shiftPositionsStrandsBond3Prime(positionsStrands)) = 1;
<a name="l01010"></a>01010                         c.damagedSugarPhosphates(positionsStrands) = 0;
<a name="l01011"></a>01011                     else
<a name="l01012"></a>01012                         checkPositionsStrands = positionsStrands;
<a name="l01013"></a>01013                         checkPositionsStrands(mod(positionsStrands(:,2),2)==1, 1) = checkPositionsStrands(mod(positionsStrands(:,2),2)==1, 1) - 1;
<a name="l01014"></a>01014                         [~, idxs, ~, numReactions] = this.bindProteinToChromosome(checkPositionsStrands, ...
<a name="l01015"></a>01015                             ezIdx, numReactions, [], false, true, 2, false, [1 2]);
<a name="l01016"></a>01016                         if numReactions == 0; continue; end
<a name="l01017"></a>01017                         positionsStrands = positionsStrands(idxs, :);
<a name="l01018"></a>01018                         
<a name="l01019"></a>01019                         c.strandBreaks(c.shiftPositionsStrandsBond5Prime(positionsStrands)) = 1;
<a name="l01020"></a>01020                         c.damagedSugarPhosphates(positionsStrands) = 0;
<a name="l01021"></a>01021                     end
<a name="l01022"></a>01022                     
<a name="l01023"></a>01023                     this.substrates = this.substrates + ...
<a name="l01024"></a>01024                         numReactions * this.reactionSmallMoleculeStoichiometryMatrix(:, rxIdx);
<a name="l01025"></a>01025                 end
<a name="l01026"></a>01026             end
<a name="l01027"></a>01027             
<a name="l01028"></a>01028             %5&#39;-deoxyribosephosphodiesterase Nfo (MG_235)
<a name="l01029"></a>01029             if nnz(c.abasicSites) &amp;&amp; nnz(c.strandBreaks)
<a name="l01030"></a>01030                 positionsStrands = find(...
<a name="l01031"></a>01031                     c.abasicSites &amp; ...
<a name="l01032"></a>01032                     c.strandBreaks3);
<a name="l01033"></a>01033                 if ~isempty(positionsStrands)
<a name="l01034"></a>01034                     
<a name="l01035"></a>01035                     numReactions = max(0, floor(min([
<a name="l01036"></a>01036                         size(positionsStrands, 1);
<a name="l01037"></a>01037                         this.randStream.stochasticRound(this.enzymes(this.enzymeIndexs_apurinicEndonuclease) * this.enzymeBounds(this.reactionIndexs_BER_drpesterase, 2) * this.stepSizeSec);
<a name="l01038"></a>01038                         this.substrates ./ max(0, -this.reactionSmallMoleculeStoichiometryMatrix(:, this.reactionIndexs_BER_drpesterase))])));
<a name="l01039"></a>01039                     
<a name="l01040"></a>01040                     checkPositionsStrands = positionsStrands;
<a name="l01041"></a>01041                     checkPositionsStrands(:, 1) = checkPositionsStrands(:, 1) - 1;
<a name="l01042"></a>01042                     [~, idxs, ~, numReactions] = this.bindProteinToChromosome(checkPositionsStrands, ...
<a name="l01043"></a>01043                         this.enzymeIndexs_apurinicEndonuclease, numReactions, [], false, true, 2, false, [1 2 3]);
<a name="l01044"></a>01044                     if numReactions &gt; 0
<a name="l01045"></a>01045                         positionsStrands = positionsStrands(idxs, :);
<a name="l01046"></a>01046                         
<a name="l01047"></a>01047                         c.damagedSugarPhosphates(positionsStrands) = 0;
<a name="l01048"></a>01048                         c.strandBreaks(c.shiftPositionsStrandsBond5Prime(positionsStrands)) = 1;
<a name="l01049"></a>01049                         c.gapSites(positionsStrands) = 1;
<a name="l01050"></a>01050                         
<a name="l01051"></a>01051                         this.substrates = this.substrates + ...
<a name="l01052"></a>01052                             numReactions * this.reactionSmallMoleculeStoichiometryMatrix(:,this.reactionIndexs_BER_drpesterase);
<a name="l01053"></a>01053                     end
<a name="l01054"></a>01054                 end
<a name="l01055"></a>01055             end
<a name="l01056"></a>01056             
<a name="l01057"></a>01057             %3&#39;-(deoxyribose-5&#39;-phosphate) lyase dnaN (MG_001)
<a name="l01058"></a>01058             if nnz(c.abasicSites) &amp;&amp; nnz(c.strandBreaks)
<a name="l01059"></a>01059                 positionsStrands = find(c.abasicSites &amp; c.strandBreaks5);
<a name="l01060"></a>01060                 if ~isempty(positionsStrands)
<a name="l01061"></a>01061                     
<a name="l01062"></a>01062                     numReactions = max(0, floor(min([
<a name="l01063"></a>01063                         size(positionsStrands, 1);
<a name="l01064"></a>01064                         this.randStream.stochasticRound(this.enzymes(this.enzymeIndexs_polymerase) * this.enzymeBounds(this.reactionIndexs_BER_dr5plyase, 2) * this.stepSizeSec);
<a name="l01065"></a>01065                         this.substrates ./ max(0, -this.reactionSmallMoleculeStoichiometryMatrix(:, this.reactionIndexs_BER_dr5plyase))])));
<a name="l01066"></a>01066                     
<a name="l01067"></a>01067                     checkPositionsStrands = positionsStrands;
<a name="l01068"></a>01068                     checkPositionsStrands(:, 1) = checkPositionsStrands(:, 1) - 1;
<a name="l01069"></a>01069                     [~, idxs, ~, numReactions] = this.bindProteinToChromosome(checkPositionsStrands, ...
<a name="l01070"></a>01070                         this.enzymeIndexs_polymerase, numReactions, [], false, true, 2, false, [1 2 3]);
<a name="l01071"></a>01071                     if numReactions &gt; 0
<a name="l01072"></a>01072                         positionsStrands = positionsStrands(idxs, :);
<a name="l01073"></a>01073                         
<a name="l01074"></a>01074                         c.damagedSugarPhosphates(positionsStrands) = 0;
<a name="l01075"></a>01075                         c.strandBreaks(c.shiftPositionsStrandsBond3Prime(positionsStrands)) = 1;
<a name="l01076"></a>01076                         c.gapSites(positionsStrands) = 1;
<a name="l01077"></a>01077                         
<a name="l01078"></a>01078                         this.substrates = this.substrates + ...
<a name="l01079"></a>01079                             numReactions * this.reactionSmallMoleculeStoichiometryMatrix(:,this.reactionIndexs_BER_dr5plyase);
<a name="l01080"></a>01080                     end
<a name="l01081"></a>01081                 end
<a name="l01082"></a>01082             end
<a name="l01083"></a>01083         end
<a name="l01084"></a>01084 
<a name="l01085"></a>01085         %nucleotide excision repair:
<a name="l01086"></a>01086         %1. UvrABC (MG_073, MG_206, MG_421) cleaves DNA 3&#39; and 5&#39; to lesion
<a name="l01087"></a>01087         %2. Helicase PcrA (MG_244) excises cleaved DNA
<a name="l01088"></a>01088         %3. Cleavage excised oligonucleotide by MgpA (MG_190)
<a name="l01089"></a>01089         %4. 5&#39;-3&#39; Polymerization by dnaN (MG_001)
<a name="l01090"></a>01090         %5. Ligation by ligA (MG_254). See DDR section for details.
<a name="l01091"></a>01091         %6. Catabolism and export of damaged nucleobases and nucleosides by
<a name="l01092"></a>01092         %   metabolic process.
<a name="l01093"></a>01093         %
<a name="l01094"></a>01094         %evolveState_NER models steps 1-3. evolveState_PolymeraseLigate models
<a name="l01095"></a>01095         %steps 4-6.
<a name="l01096"></a>01096         function evolveState_NER(this)            
<a name="l01097"></a>01097             %% for convenience
<a name="l01098"></a>01098             c = this.chromosome;
<a name="l01099"></a>01099             
<a name="l01100"></a>01100             %% stop early
<a name="l01101"></a>01101             %no enzyme
<a name="l01102"></a>01102             rxIdxs = [this.reactionIndexs_NER_incision; this.reactionIndexs_NER_excision; this.reactionIndexs_DNACleavage];
<a name="l01103"></a>01103             if ~all(this.reactionCatalysisMatrix(rxIdxs, :) * this.enzymes)
<a name="l01104"></a>01104                 return;
<a name="l01105"></a>01105             end
<a name="l01106"></a>01106             
<a name="l01107"></a>01107             %no damage
<a name="l01108"></a>01108             if nnz(c.damagedBases) == 0 &amp;&amp; nnz(c.intrastrandCrossLinks) == 0
<a name="l01109"></a>01109                 return;
<a name="l01110"></a>01110             end
<a name="l01111"></a>01111 
<a name="l01112"></a>01112             %% damaged sites: positions and strands
<a name="l01113"></a>01113             damagedPositionsStrands = setdiff(...
<a name="l01114"></a>01114                 find(xor(c.damagedBases, c.intrastrandCrossLinks)), ...
<a name="l01115"></a>01115                 find(c.damagedBases == this.substrateGlobalIndexs(this.substrateIndexs_m6AD)), ...
<a name="l01116"></a>01116                 &#39;rows&#39;);
<a name="l01117"></a>01117             if isempty(damagedPositionsStrands)
<a name="l01118"></a>01118                 return;
<a name="l01119"></a>01119             end
<a name="l01120"></a>01120 
<a name="l01121"></a>01121             %% exclude methylated restriction/modification sites
<a name="l01122"></a>01122             % <a class="code" href="class_note.html">Note</a>: this implements the approximation that the rate of
<a name="l01123"></a>01123             % methylation of hemimethylated restriction/modification sites is
<a name="l01124"></a>01124             % greater their rate of demethylation by the DNA repair machinery
<a name="l01125"></a>01125             rmPositionsStrands = [...
<a name="l01126"></a>01126                 reshape(this.RM_MunI_RecognitionSites(:, this.RM_MunI_MethylatedPositions), [], 1) ...
<a name="l01127"></a>01127                 reshape(repmat([1 2], size(this.RM_MunI_RecognitionSites, 1), 1), [], 1)];
<a name="l01128"></a>01128             rmPositionsStrands(...
<a name="l01129"></a>01129                 c.damagedBases(rmPositionsStrands) == this.substrateGlobalIndexs(this.substrateIndexs_m6AD), :) = [];
<a name="l01130"></a>01130             
<a name="l01131"></a>01131             damagedPositionsStrands(edu.stanford.covert.util.<a class="code" href="class_sparse_mat.html">SparseMat</a>.ismember_subs(damagedPositionsStrands, rmPositionsStrands, [c.sequenceLen c.nCompartments]), :) = [];
<a name="l01132"></a>01132             
<a name="l01133"></a>01133             %% possible excision, incision sites
<a name="l01134"></a>01134             lengths = 1 + (this.NER_UvrABC_IncisionMargin5 + this.NER_UvrABC_IncisionMargin3 + 1);
<a name="l01135"></a>01135             
<a name="l01136"></a>01136             excisedPositions = 2:lengths;
<a name="l01137"></a>01137             incisedPositions = [1 lengths];
<a name="l01138"></a>01138             numExcisions = numel(excisedPositions);
<a name="l01139"></a>01139             numIncisions = numel(incisedPositions);
<a name="l01140"></a>01140             
<a name="l01141"></a>01141             positionsStrands = damagedPositionsStrands;
<a name="l01142"></a>01142             ignoreDamageFilter = ones(size(positionsStrands, 1), 2);
<a name="l01143"></a>01143             
<a name="l01144"></a>01144             idxs = find(c.intrastrandCrossLinks(damagedPositionsStrands));
<a name="l01145"></a>01145             ignoreDamageFilter(idxs, 2) = ignoreDamageFilter(idxs, 2) + (2 * (mod(damagedPositionsStrands(idxs, 2), 2) == 1) - 1);
<a name="l01146"></a>01146 
<a name="l01147"></a>01147             %% calculate maximum number of sites that can be repaired
<a name="l01148"></a>01148             rxIdxs = [this.reactionIndexs_NER_incision; this.reactionIndexs_NER_excision; this.reactionIndexs_DNACleavage];
<a name="l01149"></a>01149             rxCounts = [1; ceil(numExcisions / this.NER_PcrA_StepSize); ceil(numExcisions / 4 * ones(4,1))];
<a name="l01150"></a>01150             
<a name="l01151"></a>01151             numReactions = max(0, floor(min([...
<a name="l01152"></a>01152                 size(positionsStrands, 1);
<a name="l01153"></a>01153                 this.randStream.stochasticRound((this.reactionCatalysisMatrix(rxIdxs, :) * this.enzymes * this.stepSizeSec) .* (this.enzymeBounds(rxIdxs, 2) .* rxCounts));
<a name="l01154"></a>01154                 this.substrates ./ (max(0, -this.reactionSmallMoleculeStoichiometryMatrix(:, rxIdxs)) * rxCounts)])));
<a name="l01155"></a>01155             if numReactions == 0
<a name="l01156"></a>01156                 return;
<a name="l01157"></a>01157             end
<a name="l01158"></a>01158             
<a name="l01159"></a>01159             %% randomly select sites to repair
<a name="l01160"></a>01160             [~, idxs, positionsStrands] = this.bindProteinToChromosome(positionsStrands, ...
<a name="l01161"></a>01161                 any(this.reactionCatalysisMatrix(rxIdxs, :), 1), numReactions, [], ...
<a name="l01162"></a>01162                 false, true, lengths, false, ignoreDamageFilter);
<a name="l01163"></a>01163             if isempty(idxs)
<a name="l01164"></a>01164                 return;
<a name="l01165"></a>01165             end
<a name="l01166"></a>01166             damagedPositionsStrands = damagedPositionsStrands(idxs, :);
<a name="l01167"></a>01167             
<a name="l01168"></a>01168             %% update state
<a name="l01169"></a>01169             numReactions = size(positionsStrands, 1);
<a name="l01170"></a>01170             positionsStrands(mod(positionsStrands(:, 2), 2) == 1, 1) = positionsStrands(mod(positionsStrands(:, 2), 2) == 1, 1) - this.NER_UvrABC_IncisionMargin5 - 1;
<a name="l01171"></a>01171             positionsStrands(mod(positionsStrands(:, 2), 2) == 0, 1) = positionsStrands(mod(positionsStrands(:, 2), 2) == 0, 1) - this.NER_UvrABC_IncisionMargin3 - 1;
<a name="l01172"></a>01172             
<a name="l01173"></a>01173             %metabolites
<a name="l01174"></a>01174             rxCounts(1) = numReactions;
<a name="l01175"></a>01175             rxCounts(2) = ceil(numExcisions / this.NER_PcrA_StepSize) * numReactions;
<a name="l01176"></a>01176             rxCounts(3:6) = c.sequence.subsequenceBaseCounts(...
<a name="l01177"></a>01177                 reshape(repmat(positionsStrands(:, 1), 1, numExcisions) + repmat(excisedPositions - 1, numReactions, 1), [], 1), ...
<a name="l01178"></a>01178                 repmat(positionsStrands(:,2), numExcisions, 1));
<a name="l01179"></a>01179             this.substrates = this.substrates + ...
<a name="l01180"></a>01180                 this.reactionSmallMoleculeStoichiometryMatrix(:, rxIdxs) * rxCounts;
<a name="l01181"></a>01181 
<a name="l01182"></a>01182             %correct metabolites for damages bases
<a name="l01183"></a>01183             damages = c.damagedBases(damagedPositionsStrands);
<a name="l01184"></a>01184             idxs = find(damages);
<a name="l01185"></a>01185             undamagedBaseCounts = c.sequence.subsequenceBaseCounts(damagedPositionsStrands(idxs, :));
<a name="l01186"></a>01186             [damages, damageCounts] = edu.stanford.covert.util.<a class="code" href="count_unique_8m.html#a410e08d5000569bcf7496572e8ce1729">countUnique</a>(damages(idxs, :));
<a name="l01187"></a>01187             damages = ismembc2(damages, this.substrateGlobalIndexs);
<a name="l01188"></a>01188             
<a name="l01189"></a>01189             this.substrates(this.substrateIndexs_dNMPs) = this.substrates(this.substrateIndexs_dNMPs) - undamagedBaseCounts;
<a name="l01190"></a>01190             this.substrates(damages) = this.substrates(damages) + damageCounts;
<a name="l01191"></a>01191             
<a name="l01192"></a>01192             %correct metabolites for intrastrand cross links
<a name="l01193"></a>01193             damages = c.intrastrandCrossLinks(damagedPositionsStrands);
<a name="l01194"></a>01194             idxs = find(damages);
<a name="l01195"></a>01195             undamagedBaseCounts = c.sequence.subsequenceBaseCounts(...
<a name="l01196"></a>01196                 repmat(damagedPositionsStrands(idxs, 1), 1, 2) + repmat(2 * (mod(damagedPositionsStrands(idxs, 2), 2) == 1) - 1, 1, 2) .* repmat(0:1, numel(idxs), 1), ...
<a name="l01197"></a>01197                 damagedPositionsStrands(idxs, 2));
<a name="l01198"></a>01198             [damages, damageCounts] = edu.stanford.covert.util.<a class="code" href="count_unique_8m.html#a410e08d5000569bcf7496572e8ce1729">countUnique</a>(damages(idxs, :));
<a name="l01199"></a>01199             damages = ismembc2(damages, this.substrateGlobalIndexs);
<a name="l01200"></a>01200             
<a name="l01201"></a>01201             this.substrates(this.substrateIndexs_dNMPs) = this.substrates(this.substrateIndexs_dNMPs) - undamagedBaseCounts;
<a name="l01202"></a>01202             this.substrates(damages) = this.substrates(damages) + damageCounts;
<a name="l01203"></a>01203             this.substrates(this.substrateIndexs_water) = this.substrates(this.substrateIndexs_water) + sum(damageCounts);
<a name="l01204"></a>01204             this.substrates(this.substrateIndexs_hydrogen) = this.substrates(this.substrateIndexs_hydrogen) - sum(damageCounts);
<a name="l01205"></a>01205             
<a name="l01206"></a>01206             %DNA
<a name="l01207"></a>01207             incisedPositionsStrands = [...
<a name="l01208"></a>01208                 reshape(repmat(positionsStrands(:,1), 1, numIncisions) + repmat(incisedPositions-1, numReactions, 1), [], 1) ...
<a name="l01209"></a>01209                 repmat(positionsStrands(:,2), numIncisions, 1)];
<a name="l01210"></a>01210             excisedPositionsStrands = [...
<a name="l01211"></a>01211                 reshape(repmat(positionsStrands(:,1), 1, numExcisions) + repmat(excisedPositions-1, numReactions, 1), [], 1) ...
<a name="l01212"></a>01212                 repmat(positionsStrands(:,2), numExcisions, 1)];
<a name="l01213"></a>01213             
<a name="l01214"></a>01214             c.strandBreaks(          incisedPositionsStrands) = 1;
<a name="l01215"></a>01215             c.gapSites(              excisedPositionsStrands) = 1;
<a name="l01216"></a>01216             c.abasicSites(           excisedPositionsStrands) = 1;
<a name="l01217"></a>01217             c.damagedBases(          excisedPositionsStrands) = 0;
<a name="l01218"></a>01218             c.intrastrandCrossLinks( excisedPositionsStrands) = 0;
<a name="l01219"></a>01219         end
<a name="l01220"></a>01220 
<a name="l01221"></a>01221         %homologous recombination repair
<a name="l01222"></a>01222         function evolveState_HR(this)
<a name="l01223"></a>01223             c = this.chromosome;                       
<a name="l01224"></a>01224 
<a name="l01225"></a>01225             %% 1. 5&#39;-3&#39; resection (polI-like, MG_262),
<a name="l01226"></a>01226             %  2. Strand exchange (recA, MG_339)
<a name="l01227"></a>01227             if nnz(c.strandBreaks) &amp;&amp; nnz(c.doubleStrandBreaks)
<a name="l01228"></a>01228                 %find <span class="keywordtype">double</span> strand breaks
<a name="l01229"></a>01229                 positionsChromosomes = find(c.doubleStrandBreaks(:, 1:2:end));
<a name="l01230"></a>01230                 
<a name="l01231"></a>01231                 %possible sites of resection
<a name="l01232"></a>01232                 numResections = 2*this.HR_PolA_ResectionLength;
<a name="l01233"></a>01233                 
<a name="l01234"></a>01234                 %randomly select <span class="keywordtype">double</span> strand breaks to repair, up to limit of
<a name="l01235"></a>01235                 %available metabolites and enzymes
<a name="l01236"></a>01236                 rxIdxs = [this.reactionIndexs_HR_strandExchange; this.reactionIndexs_HR_DNAResection(1)];
<a name="l01237"></a>01237                 rxCounts = [2; numResections];
<a name="l01238"></a>01238                 numReactions = max(0,floor(min([
<a name="l01239"></a>01239                     size(positionsChromosomes,1);
<a name="l01240"></a>01240                     this.substrates ./ (max(0,-this.reactionSmallMoleculeStoichiometryMatrix(:,rxIdxs))*rxCounts);
<a name="l01241"></a>01241                     this.randStream.stochasticRound((this.reactionCatalysisMatrix(rxIdxs,:)*this.enzymes*this.stepSizeSec).*(this.enzymeBounds(rxIdxs,2) .* rxCounts));
<a name="l01242"></a>01242                     floor(this.enzymes(this.enzymeIndexs_recombinationStrandExchange) / (2*ceil(numResections/2/this.HR_RecA_Spacing)))])));
<a name="l01243"></a>01243                 
<a name="l01244"></a>01244                 [~, idxs, ~, numReactions] = this.bindProteinToChromosome(...
<a name="l01245"></a>01245                     [positionsChromosomes(:,1) - this.HR_PolA_ResectionLength 2*positionsChromosomes(:,2)], ...
<a name="l01246"></a>01246                     this.enzymeIndexs_exonuclease53, numReactions, [], ...
<a name="l01247"></a>01247                     false, true, 2*this.HR_PolA_ResectionLength, false, this.HR_PolA_ResectionLength + [1 2]);
<a name="l01248"></a>01248                 if numReactions &gt; 0
<a name="l01249"></a>01249                     positionsChromosomes = positionsChromosomes(idxs,:);
<a name="l01250"></a>01250                     positions = positionsChromosomes(:, 1);
<a name="l01251"></a>01251                     chrs      = positionsChromosomes(:, 2);
<a name="l01252"></a>01252                     numDSBs = length(idxs);
<a name="l01253"></a>01253                     
<a name="l01254"></a>01254                     %update state
<a name="l01255"></a>01255                     resectedPositions1    = repmat(positions, 1, numResections/2) + repmat(1:this.HR_PolA_ResectionLength,    numDSBs, 1);
<a name="l01256"></a>01256                     resectedPositions2    = repmat(positions, 1, numResections/2) + repmat(-this.HR_PolA_ResectionLength+1:0, numDSBs, 1);
<a name="l01257"></a>01257                     resectedEndPositions1 =        positions                      +         this.HR_PolA_ResectionLength;
<a name="l01258"></a>01258                     resectedEndPositions2 =        positions                      +        -this.HR_PolA_ResectionLength;
<a name="l01259"></a>01259                     
<a name="l01260"></a>01260                     c.gapSites(   reshape(resectedPositions1, [], 1), 1) = 1;
<a name="l01261"></a>01261                     c.gapSites(   reshape(resectedPositions2, [], 1), 2) = 1;
<a name="l01262"></a>01262                     c.abasicSites(reshape(resectedPositions1, [], 1), 1) = 1;
<a name="l01263"></a>01263                     c.abasicSites(reshape(resectedPositions2, [], 1), 2) = 1;
<a name="l01264"></a>01264                     c.hollidayJunctions(  resectedEndPositions1,     1) = 1;
<a name="l01265"></a>01265                     c.hollidayJunctions(  resectedEndPositions2,     2) = 1;
<a name="l01266"></a>01266                     
<a name="l01267"></a>01267                     rxIdxs = [this.reactionIndexs_HR_strandExchange; this.reactionIndexs_HR_DNAResection];
<a name="l01268"></a>01268                     rxCounts = [2*numReactions;
<a name="l01269"></a>01269                         c.sequence.subsequenceBaseCounts(resectedPositions1, 1) + c.sequence.subsequenceBaseCounts(resectedPositions2, 2)];
<a name="l01270"></a>01270                     this.substrates = this.substrates + ...
<a name="l01271"></a>01271                         this.reactionSmallMoleculeStoichiometryMatrix(:, rxIdxs) * rxCounts;
<a name="l01272"></a>01272                     
<a name="l01273"></a>01273                     nRecA = ceil(numResections / 2 / this.HR_RecA_Spacing);
<a name="l01274"></a>01274                     recAPositions1 = repmat(positions, 1, nRecA) + repmat(this.randStream.stochasticRound((1:nRecA)*this.HR_PolA_ResectionLength / nRecA),            numDSBs, 1);
<a name="l01275"></a>01275                     recAPositions2 = repmat(positions, 1, nRecA) + repmat(this.randStream.stochasticRound((-nRecA+1:0)*(this.HR_PolA_ResectionLength-1) / (nRecA-1)), numDSBs, 1);
<a name="l01276"></a>01276                     this.bindProteinToChromosome(...
<a name="l01277"></a>01277                         [reshape(recAPositions1, [], 1) repmat(2*chrs-1, nRecA, 1); reshape(recAPositions2, [], 1) repmat(2*chrs, nRecA, 1)], ...
<a name="l01278"></a>01278                         this.enzymeIndexs_recombinationStrandExchange, [], [], ...
<a name="l01279"></a>01279                         true, true, 1, false, -this.HR_PolA_ResectionLength-1:this.HR_PolA_ResectionLength);
<a name="l01280"></a>01280                 end
<a name="l01281"></a>01281             end
<a name="l01282"></a>01282 
<a name="l01283"></a>01283             %% 3. 5&#39;-3&#39; polymerization (dnaN, MG_001)
<a name="l01284"></a>01284             %  4. Ligation (ligA, MG_254)
<a name="l01285"></a>01285             %    simulated by evolveState_Polymerize/Ligate
<a name="l01286"></a>01286 
<a name="l01287"></a>01287             %% 5. Second strand exchange (recA, MG_339)
<a name="l01288"></a>01288             %  6. Holliday junction migration (ruvAB, MG_358, MG_359)
<a name="l01289"></a>01289             %  7. Holliday junction resolution (recU, MG_352)
<a name="l01290"></a>01290 
<a name="l01291"></a>01291             if nnz(c.hollidayJunctions)
<a name="l01292"></a>01292                 %find holiday junctions
<a name="l01293"></a>01293                 numResections = 2*this.HR_PolA_ResectionLength;
<a name="l01294"></a>01294                 positionsChromosomes = find(...
<a name="l01295"></a>01295                     c.hollidayJunctions(:, 1:2:end) &amp; ...
<a name="l01296"></a>01296                     vertcat(c.hollidayJunctions(end-numResections+1:end,2:2:end), c.hollidayJunctions(1:end-numResections, 2:2:end)));
<a name="l01297"></a>01297                 
<a name="l01298"></a>01298                 %randomly reorder holliday junctions
<a name="l01299"></a>01299                 randomIdxs  = this.randStream.randperm(size(positionsChromosomes,1));
<a name="l01300"></a>01300                 positionsChromosomes = positionsChromosomes(randomIdxs,:);
<a name="l01301"></a>01301                 positions = positionsChromosomes(:, 1);
<a name="l01302"></a>01302                 chrs = positionsChromosomes(:, 2);
<a name="l01303"></a>01303                 
<a name="l01304"></a>01304                 %loop over holliday junctions
<a name="l01305"></a>01305                 for i = 1:numel(positions)
<a name="l01306"></a>01306                     %formation site
<a name="l01307"></a>01307                     hollidayJunctionFormationSite = mod(positions(i) + [0; -numResections] - 1, c.sequenceLen) + 1;
<a name="l01308"></a>01308                     
<a name="l01309"></a>01309                     %find nearest resolution site
<a name="l01310"></a>01310                     hollidayJunctionResolutionSite = zeros(2,1);
<a name="l01311"></a>01311                     if hollidayJunctionFormationSite(1) &gt; this.hollidayJunctionResolutionSites(find(this.hollidayJunctionResolutionSites(:, 2) == 1, 1, &#39;last&#39;), 1)
<a name="l01312"></a>01312                         hollidayJunctionResolutionSite(1) = this.hollidayJunctionResolutionSites(...
<a name="l01313"></a>01313                             find(this.hollidayJunctionResolutionSites(:, 2) == 1, 1, &#39;first&#39;), 1);
<a name="l01314"></a>01314                     else
<a name="l01315"></a>01315                         hollidayJunctionResolutionSite(1) = this.hollidayJunctionResolutionSites(...
<a name="l01316"></a>01316                             find(this.hollidayJunctionResolutionSites(:, 2) == 1 &amp; ...
<a name="l01317"></a>01317                             this.hollidayJunctionResolutionSites(:, 1) &gt; hollidayJunctionFormationSite(1), 1, &#39;first&#39;), 1);
<a name="l01318"></a>01318                     end
<a name="l01319"></a>01319                     if hollidayJunctionFormationSite(2) &lt; this.hollidayJunctionResolutionSites(find(this.hollidayJunctionResolutionSites(:, 2) == 2, 1, &#39;first&#39;), 1)
<a name="l01320"></a>01320                         hollidayJunctionResolutionSite(2) = this.hollidayJunctionResolutionSites(...
<a name="l01321"></a>01321                             find(this.hollidayJunctionResolutionSites(:, 2) == 2, 1, &#39;last&#39;), 1);
<a name="l01322"></a>01322                     else
<a name="l01323"></a>01323                         hollidayJunctionResolutionSite(2) = this.hollidayJunctionResolutionSites(...
<a name="l01324"></a>01324                             find(this.hollidayJunctionResolutionSites(:, 2) == 2 &amp; ...
<a name="l01325"></a>01325                             this.hollidayJunctionResolutionSites(:, 1) &lt; hollidayJunctionFormationSite(2), 1, &#39;last&#39;), 1);
<a name="l01326"></a>01326                     end
<a name="l01327"></a>01327                     
<a name="l01328"></a>01328                     %junction migration steps
<a name="l01329"></a>01329                     junctionMigrationLengths = abs(hollidayJunctionResolutionSite-hollidayJunctionFormationSite);
<a name="l01330"></a>01330                     junctionMigrationHops = sum(this.randStream.stochasticRound(junctionMigrationLengths / this.HR_RuvAB_JunctionMigrationHop));
<a name="l01331"></a>01331                     
<a name="l01332"></a>01332                     %stop if insufficient resources
<a name="l01333"></a>01333                     rxIdxs   = [this.reactionIndexs_HR_strandExchange; this.reactionIndexs_HR_junctionMigration; this.reactionIndexs_HR_junctionResolution];
<a name="l01334"></a>01334                     rxCounts = [2; junctionMigrationHops; 1];
<a name="l01335"></a>01335                     if ...
<a name="l01336"></a>01336                             any(this.substrates &lt; max(0,-this.reactionSmallMoleculeStoichiometryMatrix(:,rxIdxs))*rxCounts) || ...
<a name="l01337"></a>01337                             any(this.randStream.stochasticRound(this.reactionCatalysisMatrix(rxIdxs,:) * this.enzymes * this.stepSizeSec .* this.enzymeBounds(rxIdxs,2)) &lt; 1)
<a name="l01338"></a>01338                         continue;
<a name="l01339"></a>01339                     end
<a name="l01340"></a>01340                     
<a name="l01341"></a>01341                     %stop if site isn&#39;t accessible
<a name="l01342"></a>01342                     resolutionRegionLength = mod(hollidayJunctionResolutionSite(1) - hollidayJunctionResolutionSite(2) + 1 -1, c.sequenceLen) + 1;
<a name="l01343"></a>01343                     ignoreDamageFilter = mod(repmat(hollidayJunctionFormationSite, 1, 3) + repmat(-1:1, 2, 1) - hollidayJunctionResolutionSite(2) + 1 -1, c.sequenceLen) + 1;
<a name="l01344"></a>01344                     if ~this.bindProteinToChromosome([hollidayJunctionResolutionSite(2) 2*chrs(i)], ...
<a name="l01345"></a>01345                             this.enzymeIndexs_hollidayJunctionHelicase, 1, [], ...
<a name="l01346"></a>01346                             false, true, resolutionRegionLength, false, reshape(ignoreDamageFilter, 1, []))
<a name="l01347"></a>01347                         continue;
<a name="l01348"></a>01348                     end
<a name="l01349"></a>01349                     
<a name="l01350"></a>01350                     %update state
<a name="l01351"></a>01351                     this.substrates = this.substrates + ...
<a name="l01352"></a>01352                         this.reactionSmallMoleculeStoichiometryMatrix(:,rxIdxs)*rxCounts;
<a name="l01353"></a>01353                     
<a name="l01354"></a>01354                     c.hollidayJunctions([hollidayJunctionFormationSite 2*chrs(i) + [-1;0]]) = 0;
<a name="l01355"></a>01355                     c.strandBreaks([hollidayJunctionResolutionSite 2*chrs(i) + [-1;0]]) = 1;
<a name="l01356"></a>01356                 end
<a name="l01357"></a>01357             end
<a name="l01358"></a>01358 
<a name="l01359"></a>01359             %% 8. Ligation (ligA, MG_254)
<a name="l01360"></a>01360             %  9. Catabolism
<a name="l01361"></a>01361             %     simulated by evolveState_Polymerize/Ligate, metabolism process
<a name="l01362"></a>01362         end
<a name="l01363"></a>01363 
<a name="l01364"></a>01364         % polymerize DNA
<a name="l01365"></a>01365         function evolveState_Polymerize(this)
<a name="l01366"></a>01366             c = this.chromosome;
<a name="l01367"></a>01367             
<a name="l01368"></a>01368             if nnz(c.gapSites) == 0 || nnz(c.abasicSites) == 0 || nnz(c.strandBreaks) == 0
<a name="l01369"></a>01369                 return;
<a name="l01370"></a>01370             end
<a name="l01371"></a>01371             
<a name="l01372"></a>01372             %find regions that need to be polymerized
<a name="l01373"></a>01373             gapStarts = find(c.gapSites &amp; c.abasicSites &amp; c.strandBreaks5);
<a name="l01374"></a>01374             if isempty(gapStarts)
<a name="l01375"></a>01375                 return; 
<a name="l01376"></a>01376             end
<a name="l01377"></a>01377             gapNotStarts = find(c.gapSites &amp; c.abasicSites &amp; c.gapSites5 &amp; c.abasicSites5);
<a name="l01378"></a>01378             gapNotStarts(find(c.strandBreaks5(gapNotStarts)), :) = []; %<span class="preprocessor">#ok&lt;FNDSB&gt;</span>
<a name="l01379"></a>01379 <span class="preprocessor"></span>            gaps = edu.stanford.covert.util.SparseMat.sort_subs([gapStarts;gapNotStarts], [c.sequenceLen c.nCompartments]);
<a name="l01380"></a>01380             gaps(...
<a name="l01381"></a>01381                 c.damagedSugarPhosphates3(gaps) | ...
<a name="l01382"></a>01382                 c.damagedSugarPhosphates5(gaps) | ...
<a name="l01383"></a>01383                 c.damagedBases(           gaps) | ...
<a name="l01384"></a>01384                 c.intrastrandCrossLinks3( gaps) | ...
<a name="l01385"></a>01385                 c.intrastrandCrossLinks5( gaps), :)=[];
<a name="l01386"></a>01386             
<a name="l01387"></a>01387             [positions, lengths, strands, sequencesCell] = c.sequence.locateFeatures(gaps);
<a name="l01388"></a>01388             dirs = 2 * (mod(strands, 2) == 1) - 1;
<a name="l01389"></a>01389             
<a name="l01390"></a>01390             %polymerize regions in random order
<a name="l01391"></a>01391             polymerase = this.randStream.stochasticRound(this.enzymes(this.enzymeIndexs_polymerase) * this.enzymeBounds(this.reactionIndexs_polymerization(1), 2) * this.stepSizeSec);
<a name="l01392"></a>01392             randOrder = this.randStream.randperm(numel(positions));
<a name="l01393"></a>01393             [~, footprint3Prime, footprint5Prime] = c.getDNAFootprint(this.enzymeGlobalIndexs(this.enzymeIndexs_polymerase), []);
<a name="l01394"></a>01394             <span class="keywordflow">for</span> j = 1:numel(positions)
<a name="l01395"></a>01395                 %% polymerize regions in random order
<a name="l01396"></a>01396                 i = randOrder(j);
<a name="l01397"></a>01397                 
<a name="l01398"></a>01398                 %% calculate extent of polymerization
<a name="l01399"></a>01399                 %limit extent by available polymerase and metabolites
<a name="l01400"></a>01400                 polymerizedBases = edu.stanford.covert.cell.sim.util.polymerize(...
<a name="l01401"></a>01401                     sequencesCell{i}(1: min(lengths(i), polymerase)), ...
<a name="l01402"></a>01402                     this.substrates(this.substrateIndexs_dNTPs), <span class="stringliteral">&#39;ACGT&#39;</span>, char(0), 0, 0, this.randStream);                
<a name="l01403"></a>01403                 <span class="keywordflow">if</span> polymerizedBases == 0
<a name="l01404"></a>01404                     <span class="keywordflow">continue</span>;
<a name="l01405"></a>01405                 end
<a name="l01406"></a>01406                 
<a name="l01407"></a>01407                 %damage filter
<a name="l01408"></a>01408                 ignoreDamageMin = -1;
<a name="l01409"></a>01409                 ignoreDamageMax = lengths(i)+1;
<a name="l01410"></a>01410                 <span class="keywordflow">if</span> mod(strands(i), 2)
<a name="l01411"></a>01411                     ignoreDamageMin = ignoreDamageMin - footprint5Prime;
<a name="l01412"></a>01412                     ignoreDamageMax = ignoreDamageMax + footprint3Prime;
<a name="l01413"></a>01413                 else
<a name="l01414"></a>01414                     ignoreDamageMin = ignoreDamageMin - footprint3Prime;
<a name="l01415"></a>01415                     ignoreDamageMax = ignoreDamageMax + footprint5Prime;
<a name="l01416"></a>01416                 end
<a name="l01417"></a>01417                 ignoreDamageFilter = dirs(i) * (ignoreDamageMin:ignoreDamageMax) + 1;
<a name="l01418"></a>01418                 
<a name="l01419"></a>01419                 %limit extent by accessibility of sites to DNA polymerase
<a name="l01420"></a>01420                 [~, ~, ~, ~, polymerizedBases] = this.bindProteinToChromosome(c.shiftPositionsStrandsBond5Prime([positions(i) strands(i)]), this.enzymeIndexs_polymerase, 1, [], ...
<a name="l01421"></a>01421                     false, true, dirs(i) * (polymerizedBases + 1), true, ignoreDamageFilter);
<a name="l01422"></a>01422                 
<a name="l01423"></a>01423                 polymerizedBases = min(abs(polymerizedBases), lengths(i));
<a name="l01424"></a>01424                 
<a name="l01425"></a>01425                 if polymerizedBases == 0
<a name="l01426"></a>01426                     continue;
<a name="l01427"></a>01427                 end
<a name="l01428"></a>01428                 
<a name="l01429"></a>01429                 %% update state
<a name="l01430"></a>01430                 %calculate base costs
<a name="l01431"></a>01431                 [~, ~, baseCosts] = edu.stanford.covert.cell.sim.util.polymerize(...
<a name="l01432"></a>01432                     sequencesCell{i}(1:polymerizedBases), ...
<a name="l01433"></a>01433                     this.substrates(this.substrateIndexs_dNTPs), <span class="stringliteral">&#39;ACGT&#39;</span>, char(0), 0, 0, this.randStream);
<a name="l01434"></a>01434                 
<a name="l01435"></a>01435                 %decrement enzyme
<a name="l01436"></a>01436                 polymerase = polymerase - polymerizedBases;
<a name="l01437"></a>01437                 
<a name="l01438"></a>01438                 %update metabolites
<a name="l01439"></a>01439                 this.substrates = this.substrates + ...
<a name="l01440"></a>01440                     this.reactionSmallMoleculeStoichiometryMatrix(:, this.reactionIndexs_polymerization) * baseCosts;
<a name="l01441"></a>01441                 
<a name="l01442"></a>01442                 %update chromosome state
<a name="l01443"></a>01443                 pos = positions(i) + dirs(i) * (0:polymerizedBases-1);
<a name="l01444"></a>01444                 c.gapSites(   pos, strands(i)) = 0;
<a name="l01445"></a>01445                 c.abasicSites(pos, strands(i)) = 0;
<a name="l01446"></a>01446                 c.strandBreaks(c.shiftPositionsStrandsBond5Prime([pos(1)   strands(i)])) = 0;
<a name="l01447"></a>01447                 c.strandBreaks(c.shiftPositionsStrandsBond3Prime([pos(end) strands(i)])) = 1;
<a name="l01448"></a>01448             end
<a name="l01449"></a>01449         end
<a name="l01450"></a>01450 
<a name="l01451"></a>01451         % ligate DNA
<a name="l01452"></a>01452         %
<a name="l01453"></a>01453         %<a class="code" href="class_note.html">Note</a>: <span class="keyword">this</span> <span class="keyword">function</span> is redundant with ligateDNA in <a class="code" href="class_replication.html">Replication</a>
<a name="l01454"></a>01454         <span class="keyword">function</span> evolveState_Ligate(<span class="keyword">this</span>)
<a name="l01455"></a>01455             c = this.chromosome;
<a name="l01456"></a>01456 
<a name="l01457"></a>01457             if nnz(c.strandBreaks) == 0
<a name="l01458"></a>01458                 return;
<a name="l01459"></a>01459             end
<a name="l01460"></a>01460             
<a name="l01461"></a>01461             positionsStrands = find(c.singleStrandBreaks);
<a name="l01462"></a>01462             if isempty(positionsStrands)
<a name="l01463"></a>01463                 return;
<a name="l01464"></a>01464             end
<a name="l01465"></a>01465             
<a name="l01466"></a>01466             numReactions = max(0, floor(min([
<a name="l01467"></a>01467                 size(positionsStrands, 1);
<a name="l01468"></a>01468                 this.randStream.stochasticRound(this.enzymes(this.enzymeIndexs_ligase) * this.stepSizeSec * this.enzymeBounds(this.reactionIndexs_ligation, 2));
<a name="l01469"></a>01469                 this.substrates ./ max(0, -this.reactionSmallMoleculeStoichiometryMatrix(:, this.reactionIndexs_ligation))])));
<a name="l01470"></a>01470             if numReactions == 0
<a name="l01471"></a>01471                 return
<a name="l01472"></a>01472             end
<a name="l01473"></a>01473             
<a name="l01474"></a>01474             [~, ~, positionsStrands, numReactions] = this.bindProteinToChromosome(...
<a name="l01475"></a>01475                 positionsStrands, this.enzymeIndexs_ligase, numReactions, [], false, true, 1, false, [0 1 2]);            
<a name="l01476"></a>01476             if numReactions == 0
<a name="l01477"></a>01477                 return;
<a name="l01478"></a>01478             end
<a name="l01479"></a>01479             
<a name="l01480"></a>01480             c.strandBreaks(positionsStrands) = 0;
<a name="l01481"></a>01481             
<a name="l01482"></a>01482             this.substrates = this.substrates + numReactions * ...
<a name="l01483"></a>01483                 this.reactionSmallMoleculeStoichiometryMatrix(:, this.reactionIndexs_ligation);
<a name="l01484"></a>01484         end
<a name="l01485"></a>01485 
<a name="l01486"></a>01486         %bind DNA integrity scanning protein to strand breaks, base
<a name="l01487"></a>01487         %damage, and cross links
<a name="l01488"></a>01488         function evolveState_DisA(this)
<a name="l01489"></a>01489             %find locations of damages
<a name="l01490"></a>01490             positionsStrands = find(this.chromosome.damagedSites_excm6AD);
<a name="l01491"></a>01491             if isempty(positionsStrands)
<a name="l01492"></a>01492                 return;
<a name="l01493"></a>01493             end
<a name="l01494"></a>01494             
<a name="l01495"></a>01495             %bind free DisA to accessible damaged sites
<a name="l01496"></a>01496             this.bindProteinToChromosome(positionsStrands, this.enzymeIndexs_DisA, ...
<a name="l01497"></a>01497                 this.enzymes(this.enzymeIndexs_DisA), [], true, true, 1, false, 1);
<a name="l01498"></a>01498         end
<a name="l01499"></a>01499         
<a name="l01500"></a>01500         function evolveState_Modification(this)
<a name="l01501"></a>01501             c = this.chromosome;
<a name="l01502"></a>01502             
<a name="l01503"></a>01503             %% EcoD Type I restriction/modification system
<a name="l01504"></a>01504             %  M. genitalium contains only the DNA binding subunit
<a name="l01505"></a>01505             
<a name="l01506"></a>01506             %% MunI Type II restriction/modification system
<a name="l01507"></a>01507             %methylate hemi-unmethylated restriction/modification sites
<a name="l01508"></a>01508             posStrnds = find(c.hemiunmethylatedMunIRMSites);
<a name="l01509"></a>01509             if isempty(posStrnds)
<a name="l01510"></a>01510                 return;
<a name="l01511"></a>01511             end
<a name="l01512"></a>01512             
<a name="l01513"></a>01513             numReactions = max(0, floor(min([
<a name="l01514"></a>01514                 size(posStrnds, 1);
<a name="l01515"></a>01515                 this.substrates ./ max(0, -this.reactionSmallMoleculeStoichiometryMatrix(:, this.reactionIndexs_RM_MunI_Methylation));
<a name="l01516"></a>01516                 this.randStream.stochasticRound(this.enzymes(this.enzymeIndexs_RM_typeII) * this.stepSizeSec * this.enzymeBounds(this.reactionIndexs_RM_MunI_Methylation, 2))])));
<a name="l01517"></a>01517             if numReactions  == 0
<a name="l01518"></a>01518                 return;
<a name="l01519"></a>01519             end
<a name="l01520"></a>01520             
<a name="l01521"></a>01521             [~, ~, posStrnds, numReactions] = this.bindProteinToChromosome(...
<a name="l01522"></a>01522                 posStrnds, this.enzymeIndexs_RM_typeII, numReactions, [], false, true, 1, false, [0 1 2]);
<a name="l01523"></a>01523             if numReactions == 0
<a name="l01524"></a>01524                 return;
<a name="l01525"></a>01525             end
<a name="l01526"></a>01526             
<a name="l01527"></a>01527             c.damagedBases(posStrnds) = this.substrateGlobalIndexs(this.substrateIndexs_m6AD);
<a name="l01528"></a>01528             
<a name="l01529"></a>01529             this.substrates = this.substrates + ...
<a name="l01530"></a>01530                 numReactions * this.reactionSmallMoleculeStoichiometryMatrix(:, this.reactionIndexs_RM_MunI_Methylation);
<a name="l01531"></a>01531         end
<a name="l01532"></a>01532         
<a name="l01533"></a>01533         function evolveState_Restriction(this)
<a name="l01534"></a>01534             c = this.chromosome;
<a name="l01535"></a>01535             
<a name="l01536"></a>01536             %% EcoD Type I restriction/modification system
<a name="l01537"></a>01537             %  M. genitalium contains only the DNA binding subunit
<a name="l01538"></a>01538             
<a name="l01539"></a>01539             %% MunI Type II restriction/modification system
<a name="l01540"></a>01540             %cleave unmethylated restriction/modification sites
<a name="l01541"></a>01541             posStrnds = find(c.restrictableMunIRMSites);
<a name="l01542"></a>01542             if isempty(posStrnds)
<a name="l01543"></a>01543                 return;
<a name="l01544"></a>01544             end
<a name="l01545"></a>01545             
<a name="l01546"></a>01546             numReactions = max(0, floor(min([
<a name="l01547"></a>01547                 size(posStrnds, 1);
<a name="l01548"></a>01548                 this.substrates ./ max(0, -this.reactionSmallMoleculeStoichiometryMatrix(:, this.reactionIndexs_RM_MunI_Restriction));
<a name="l01549"></a>01549                 this.randStream.stochasticRound(this.enzymes(this.enzymeIndexs_RM_typeII) * this.stepSizeSec * this.enzymeBounds(this.reactionIndexs_RM_MunI_Restriction, 2))])));
<a name="l01550"></a>01550             if numReactions == 0
<a name="l01551"></a>01551                 return;
<a name="l01552"></a>01552             end
<a name="l01553"></a>01553             
<a name="l01554"></a>01554             [~, ~, posStrnds, numReactions] = this.bindProteinToChromosome(...
<a name="l01555"></a>01555                 posStrnds, this.enzymeIndexs_RM_typeII, numReactions, [], false, true, 1, false, -1:1);
<a name="l01556"></a>01556             if numReactions == 0
<a name="l01557"></a>01557                 return;
<a name="l01558"></a>01558             end
<a name="l01559"></a>01559             
<a name="l01560"></a>01560             c.strandBreaks(posStrnds) = 1;
<a name="l01561"></a>01561             
<a name="l01562"></a>01562             this.substrates = this.substrates + ...
<a name="l01563"></a>01563                 numReactions * this.reactionSmallMoleculeStoichiometryMatrix(:, this.reactionIndexs_RM_MunI_Restriction);
<a name="l01564"></a>01564         end
<a name="l01565"></a>01565     end
<a name="l01566"></a>01566 end
</pre></div></div><!-- contents -->
</div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
<a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(0)"><span class="SelectionMark">&#160;</span>All</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(1)"><span class="SelectionMark">&#160;</span>Classes</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(2)"><span class="SelectionMark">&#160;</span>Files</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(3)"><span class="SelectionMark">&#160;</span>Functions</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(4)"><span class="SelectionMark">&#160;</span>Variables</a></div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

  <div id="nav-path" class="navpath">
    <ul>
      <li class="navelem"><a class="el" href="_d_n_a_repair_8m.html">DNARepair.m</a>      </li>

    <li class="footer">Generated on Wed Feb 8 2012 21:42:32 for Whole Cell Simulation by
    <a href="http://www.doxygen.org/index.html">
    <img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.7.6.1 </li>
   </ul>
 </div>


</body>
</html>
