<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<title>Whole Cell Simulation: src/+edu/+stanford/+covert/+cell/+sim/+process/DNASupercoiling.m Source File</title>

<link href="tabs.css" rel="stylesheet" type="text/css"/>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
<link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="navtree.js"></script>
<script type="text/javascript">
  $(document).ready(initResizable);
</script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
  $(document).ready(function() { searchBox.OnSelectItem(0); });
</script>

</head>
<body>
<div id="top"><!-- do not remove this div! -->


<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  
  
  <td style="padding-left: 0.5em;">
   <div id="projectname">Whole Cell Simulation
   
   </div>
   
  </td>
  
  
  
 </tr>
 </tbody>
</table>
</div>

<!-- Generated by Doxygen 1.7.6.1 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "search",false,'Search');
</script>
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&#160;Page</span></a></li>
      <li><a href="annotated.html"><span>Classes</span></a></li>
      <li class="current"><a href="files.html"><span>Files</span></a></li>
      <li>
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
  <div id="navrow2" class="tabs2">
    <ul class="tablist">
      <li><a href="files.html"><span>File&#160;List</span></a></li>
      <li><a href="globals.html"><span>File&#160;Members</span></a></li>
    </ul>
  </div>
</div>
<div id="side-nav" class="ui-resizable side-nav-resizable">
  <div id="nav-tree">
    <div id="nav-tree-contents">
    </div>
  </div>
  <div id="splitbar" style="-moz-user-select:none;" 
       class="ui-resizable-handle">
  </div>
</div>
<script type="text/javascript">
  initNavTree('_d_n_a_supercoiling_8m.html','');
</script>
<div id="doc-content">
<div class="header">
  <div class="headertitle">
<div class="title">src/+edu/+stanford/+covert/+cell/+sim/+process/DNASupercoiling.m</div>  </div>
</div><!--header-->
<div class="contents">
<a href="_d_n_a_supercoiling_8m.html">Go to the documentation of this file.</a><div class="fragment"><pre class="fragment"><a name="l00001"></a>00001 %DNASupercoiling
<a name="l00002"></a>00002 %
<a name="l00003"></a>00003 % @wholeCellModelID Process_DNASupercoiling
<a name="l00004"></a>00004 % @name             DNA Supercoiling
<a name="l00005"></a>00005 % @description
<a name="l00006"></a>00006 %   Biology
<a name="l00007"></a>00007 %   ===========
<a name="l00008"></a>00008 %   DNA gyrase and topoisomerase IV each use 2 ATP to induce 2
<a name="l00009"></a>00009 %   negative
<a name="l00010"></a>00010 %   supercoils each time they act. Occasionally, topoisomerase IV can also
<a name="l00011"></a>00011 %   induce positive supercoils but that is not considered in our model.
<a name="l00012"></a>00012 %   Topoisomerase I acts to induce positive supercoils. Gyrase, topoisomerase
<a name="l00013"></a>00013 %   IV, and topoisomerase I act at rates of 1.2, 2.5, and 1 strand passing
<a name="l00014"></a>00014 %   events per second respectively.
<a name="l00015"></a>00015 % 
<a name="l00016"></a>00016 %   We use a calculation of the DNA<span class="stringliteral">&#39;s linking number (LK) in order to track</span>
<a name="l00017"></a>00017 <span class="stringliteral">%   the supercoiling of the DNA.  The &amp;delta;LK is the difference between the</span>
<a name="l00018"></a>00018 <span class="stringliteral">%   current level of DNA supercoiling and the relaxed level of DNA</span>
<a name="l00019"></a>00019 <span class="stringliteral">%   supercoiling. The LKrelaxed is defined as number of base pairs/10.5, in which</span>
<a name="l00020"></a>00020 <span class="stringliteral">%   10.5 is the number of bases per turn in a relaxed double helix. As the</span>
<a name="l00021"></a>00021 <span class="stringliteral">%   replication loops move, the LKcurrent deviates from this relaxed state,</span>
<a name="l00022"></a>00022 <span class="stringliteral">%   and gyrases and topoisomerases help bring the DNA back to the relaxed</span>
<a name="l00023"></a>00023 <span class="stringliteral">%   state. The superhelical density, or specific linking number density,</span>
<a name="l00024"></a>00024 <span class="stringliteral">%   &amp;sigma;&lt;sub&gt;sp&lt;/sub&gt;, is defined as the:</span>
<a name="l00025"></a>00025 <span class="stringliteral">%      (LKcurrent - LKrelaxed)/LKrelaxed. </span>
<a name="l00026"></a>00026 <span class="stringliteral">%   The activity of gyrases and topoisomerases depends on the &amp;sigma;&lt;sub&gt;sp&lt;/sub&gt;</span>
<a name="l00027"></a>00027 <span class="stringliteral">%   of the DNA. Although there is likely a more complex relationship between the</span>
<a name="l00028"></a>00028 <span class="stringliteral">%   DNA supercoiling and enzyme activity, we model the enzyme activity as a</span>
<a name="l00029"></a>00029 <span class="stringliteral">%   combination of step functions and logistic functions. </span>
<a name="l00030"></a>00030 <span class="stringliteral">%   Topoisomerase IV can only act if &amp;sigma;&lt;sub&gt;sp&lt;/sub&gt; is</span>
<a name="l00031"></a>00031 <span class="stringliteral">%   higher than 0. Topoisomerase I can only act if &amp;sigma;&lt;sub&gt;sp&lt;/sub&gt; is lower</span>
<a name="l00032"></a>00032 <span class="stringliteral">%   than zero. Gyrase can only act if &amp;sigma;&lt;sub&gt;sp&lt;/sub&gt; is higher than -0.1.</span>
<a name="l00033"></a>00033 <span class="stringliteral">%   Within, the regions where gyrase and topoI are allowed to act, we apply</span>
<a name="l00034"></a>00034 <span class="stringliteral">%   a logistic function describing the probability of activity, centered</span>
<a name="l00035"></a>00035 <span class="stringliteral">%   around equilibrium sigma, -0.06. For TopoI, the probability of activity</span>
<a name="l00036"></a>00036 <span class="stringliteral">%   approaches 1 as the sigma gets more positive. For gyrase, the</span>
<a name="l00037"></a>00037 <span class="stringliteral">%   probability of activity approaches 1 as</span>
<a name="l00038"></a>00038 <span class="stringliteral">%   </span>
<a name="l00039"></a>00039 <span class="stringliteral">%   We model up to three regions on the chromosomes and track their LKs</span>
<a name="l00040"></a>00040 <span class="stringliteral">%   separately. Before replication, &quot;unreplicated DNA&quot; is the only region</span>
<a name="l00041"></a>00041 <span class="stringliteral">%   present. The enzymes may act on the chromosome affecting the LKcurrent,</span>
<a name="l00042"></a>00042 <span class="stringliteral">%   and experimentally it has been shown that the enzymes would obtain an</span>
<a name="l00043"></a>00043 <span class="stringliteral">%   LKcurrent such that the steady state &amp;sigma;&lt;sub&gt;sp&lt;/sub&gt; is =-0.06. As the</span>
<a name="l00044"></a>00044 <span class="stringliteral">%   replication loop progresses, &quot;unreplicated DNA&quot; is the region downstream of</span>
<a name="l00045"></a>00045 <span class="stringliteral">%   the two replication loops.  The number of bases in this region decreases</span>
<a name="l00046"></a>00046 <span class="stringliteral">%   during replication, meaning that the LKrelaxed decreases. The LKcurrent then,</span>
<a name="l00047"></a>00047 <span class="stringliteral">%   is too high, and must be brought down towards the LKrelaxed by inducing</span>
<a name="l00048"></a>00048 <span class="stringliteral">%   negative supercoils.</span>
<a name="l00049"></a>00049 <span class="stringliteral">%   </span>
<a name="l00050"></a>00050 <span class="stringliteral">%   The second and third regions are the replicated DNA upstream of the</span>
<a name="l00051"></a>00051 <span class="stringliteral">%   replication loops on each of the two chromosomes. As new uncoiled DNA is</span>
<a name="l00052"></a>00052 <span class="stringliteral">%   formed, it is already coiled, but gyrases and topoisomerases can continue</span>
<a name="l00053"></a>00053 <span class="stringliteral">%   to act on this DNA, ideally maintaining the steady state &amp;sigma;&lt;sub&gt;sp&lt;/sub&gt;.</span>
<a name="l00054"></a>00054 <span class="stringliteral">%   After replication is complete, these two regions are the only two that exist.</span>
<a name="l00055"></a>00055 <span class="stringliteral">%   </span>
<a name="l00056"></a>00056 <span class="stringliteral">%   It is essential that the &amp;detlta;LK in the region downstream of the replication</span>
<a name="l00057"></a>00057 <span class="stringliteral">%   loops be brought down to 0 by the end of replication.</span>
<a name="l00058"></a>00058 <span class="stringliteral">%   </span>
<a name="l00059"></a>00059 <span class="stringliteral">%   Another consideration is the processivity of the enzymes when bound to the</span>
<a name="l00060"></a>00060 <span class="stringliteral">%   DNA. Topoisomerase IV is highly processive and will stay bound to the DNA</span>
<a name="l00061"></a>00061 <span class="stringliteral">%   as long as the &amp;sigma;&lt;sub&gt;sp&lt;/sub&gt; is greater than zero. Topoisomerase I is</span>
<a name="l00062"></a>00062 <span class="stringliteral">%   not highly processive, and essentially acts on the DNA and falls back off</span>
<a name="l00063"></a>00063 <span class="stringliteral">%   right away gyrase will stay bound to the DNA for about 30-60 seconds. We model</span>
<a name="l00064"></a>00064 <span class="stringliteral">%   gyrase processivity as a poisson distribution with &amp;lambda; = 45 seconds.</span>
<a name="l00065"></a>00065 <span class="stringliteral">%   </span>
<a name="l00066"></a>00066 <span class="stringliteral">%   In this process, we go through all free gyrases and topoisomerases in</span>
<a name="l00067"></a>00067 <span class="stringliteral">%   random order,  determine what regions they can bind in, and randomly bind</span>
<a name="l00068"></a>00068 <span class="stringliteral">%   them to a large enough open position on the DNA. We adjust the linking</span>
<a name="l00069"></a>00069 <span class="stringliteral">%   number based on all enzyme actions, and account for the usage of ATP. We</span>
<a name="l00070"></a>00070 <span class="stringliteral">%   also track the processivity of gyrase and topoisomerase IV, and unbind</span>
<a name="l00071"></a>00071 <span class="stringliteral">%   them from the DNA when appropriate.</span>
<a name="l00072"></a>00072 <span class="stringliteral">%   </span>
<a name="l00073"></a>00073 <span class="stringliteral">%   If replication is in progress, we first knock off any enzymes that the</span>
<a name="l00074"></a>00074 <span class="stringliteral">%   replication loop collides into.</span>
<a name="l00075"></a>00075 <span class="stringliteral">%</span>
<a name="l00076"></a>00076 <span class="stringliteral">%   There is an effect of supercoiling on the probabilities of gene </span>
<a name="l00077"></a>00077 <span class="stringliteral">%   transcription (Peter 2004). While fold change at differnt sigmas have</span>
<a name="l00078"></a>00078 <span class="stringliteral">%   been calculated for many E. coli genes, here, for simplicity, we are</span>
<a name="l00079"></a>00079 <span class="stringliteral">%   only including the effects on the 5 supercoiling genes: gyrB, gyrA,</span>
<a name="l00080"></a>00080 <span class="stringliteral">%   parC, parE, and topA. These genes exist in 3 transcription units. Peter</span>
<a name="l00081"></a>00081 <span class="stringliteral">%   2004 has data for the fold change of expression of each of these genes</span>
<a name="l00082"></a>00082 <span class="stringliteral">%   are various values of sigma (ranging from sigma -0.06-0.02) at various </span>
<a name="l00083"></a>00083 <span class="stringliteral">%   experimental conditions. Due to the limited data and large variation within</span>
<a name="l00084"></a>00084 <span class="stringliteral">%   the data, we have decided to use a linear fit of the fold change data and </span>
<a name="l00085"></a>00085 <span class="stringliteral">%   extrapolate the linear fit within the sigma range of -0.08 to 0.07 where </span>
<a name="l00086"></a>00086 <span class="stringliteral">%   the linear fit seems reasonable. Outside of this range, we estimate a</span>
<a name="l00087"></a>00087 <span class="stringliteral">%   constant fold change of expression. While there is a separate set of</span>
<a name="l00088"></a>00088 <span class="stringliteral">%   data for each of the 5 genes, our model requires a single probability</span>
<a name="l00089"></a>00089 <span class="stringliteral">%   of transcription for each transcription unit. The data for the first</span>
<a name="l00090"></a>00090 <span class="stringliteral">%   gene in each transcription unit is used (gyrB, and parE). TopA is</span>
<a name="l00091"></a>00091 <span class="stringliteral">%   transcribed with genes that are not supercoiling related. The</span>
<a name="l00092"></a>00092 <span class="stringliteral">%   expression of those genes (MG_119, 120, 121) will also get affected by </span>
<a name="l00093"></a>00093 <span class="stringliteral">%   this process. The general trend is that gyrase and topoIV will have an</span>
<a name="l00094"></a>00094 <span class="stringliteral">%   increased expression when sigma is higher than the equilibrium, and</span>
<a name="l00095"></a>00095 <span class="stringliteral">%   that topoI will have a higher expression when sigma is lower than the</span>
<a name="l00096"></a>00096 <span class="stringliteral">%   equilibrium. Fold changes for gyrase will vary between 0.14 and 6.57.</span>
<a name="l00097"></a>00097 <span class="stringliteral">%   Fold changes for topoIV will vary between 0.98 and 1.14. Fold changes</span>
<a name="l00098"></a>00098 <span class="stringliteral">%   for topoI will vary between 0.042 and 1.15. </span>
<a name="l00099"></a>00099 <span class="stringliteral">%   </span>
<a name="l00100"></a>00100 <span class="stringliteral">%   References</span>
<a name="l00101"></a>00101 <span class="stringliteral">%   ===============</span>
<a name="l00102"></a>00102 <span class="stringliteral">%   1. Ullsperger, C., Cozzarelli, N.R. (1996). Contrasting enzymatic activities</span>
<a name="l00103"></a>00103 <span class="stringliteral">%      of topoisomerase IV and DnA gyrase from Escherichia coli. Journal of Bio</span>
<a name="l00104"></a>00104 <span class="stringliteral">%      Chem 271: 31549-31555. [PUB_0236]</span>
<a name="l00105"></a>00105 <span class="stringliteral">%   2. Dekker, N.H., Viard, T., Bouthier de la Tour, C., Duguet, M., Bensimon,</span>
<a name="l00106"></a>00106 <span class="stringliteral">%      D., Croquette, V. (2003). Thermophilic Topoisomerase I on a single DNA</span>
<a name="l00107"></a>00107 <span class="stringliteral">%      molecule. Journal of molecular biology 329: 271-282. [PUB_0502]</span>
<a name="l00108"></a>00108 <span class="stringliteral">%   3. Gore, J., Bryant, Z., Stone, M.D., Nollmann, M., Cozzarelli, N.R.,</span>
<a name="l00109"></a>00109 <span class="stringliteral">%      Bustamante, C. (2006). Mechanochemical analysis of DNA gyrase using rotor</span>
<a name="l00110"></a>00110 <span class="stringliteral">%      bead tracking. Nature 439: 100-104. [PUB_0751]</span>
<a name="l00111"></a>00111 <span class="stringliteral">%   4. Bates, A. (2006). DNA Topoisomerases: Single Gyrase Caught in the Act.</span>
<a name="l00112"></a>00112 <span class="stringliteral">%      Current Biology 16: 204-206. [PUB_0752]</span>
<a name="l00113"></a>00113 <span class="stringliteral">%   5. Peng, H., Marians, K.J. (1995). The Interaction of Escherichia coli</span>
<a name="l00114"></a>00114 <span class="stringliteral">%      Topoisomerase IV with DNA. Journal of Biological Chemistry 42:</span>
<a name="l00115"></a>00115 <span class="stringliteral">%      25286–25290. [PUB_0694]</span>
<a name="l00116"></a>00116 <span class="stringliteral">%   6. Wang, J. (1996) DNA Topoisomerases. Annual Reviews 65: 635-692. [PUB_0693]</span>
<a name="l00117"></a>00117 <span class="stringliteral">%   7. Peter, B.J., Arsuaga, J., Breier, A.M., Khodursky, A.B., Brown,</span>
<a name="l00118"></a>00118 <span class="stringliteral">%      P.O., Cozzarelli, N.R. (2004) Genomic transcriptional response to loss</span>
<a name="l00119"></a>00119 <span class="stringliteral">%      of chromosomal supercoiling in Escherichia coli. Genome Biology 5:</span>
<a name="l00120"></a>00120 <span class="stringliteral">%      1-13. [PUB_0920]</span>
<a name="l00121"></a>00121 <span class="stringliteral">%</span>
<a name="l00122"></a>00122 <span class="stringliteral">% Author: Jayodita Sanghvi, jayodita@stanford.edu</span>
<a name="l00123"></a>00123 <span class="stringliteral">% Author: Jared Jacobs, jmjacobs@stanford.edu</span>
<a name="l00124"></a>00124 <span class="stringliteral">% Author: Jonathan Karr, jkarr@stanford.edu</span>
<a name="l00125"></a>00125 <span class="stringliteral">% Affilitation: Covert Lab, Department of Bioengineering, Stanford University</span>
<a name="l00126"></a>00126 <span class="stringliteral">% Last updated: 11/18/2010</span>
<a name="l00127"></a>00127 <span class="stringliteral">classdef DNASupercoiling &lt; edu.stanford.covert.cell.sim.Process &amp; edu.stanford.covert.cell.sim.ChromosomeProcessAspect</span>
<a name="l00128"></a>00128 <span class="stringliteral">    %property annotations</span>
<a name="l00129"></a>00129 <span class="stringliteral">    properties (Constant)</span>
<a name="l00130"></a>00130 <span class="stringliteral">        optionNames__              = {}; %names of option properties</span>
<a name="l00131"></a>00131 <span class="stringliteral">        fixedConstantNames__       = {   %names of fixed constant properties</span>
<a name="l00132"></a>00132 <span class="stringliteral">            &#39;</span>gyraseMeanDwellTime<span class="stringliteral">&#39;;</span>
<a name="l00133"></a>00133 <span class="stringliteral">            &#39;</span>gyraseSigmaLimit<span class="stringliteral">&#39;;</span>
<a name="l00134"></a>00134 <span class="stringliteral">            &#39;</span>topoISigmaLimit<span class="stringliteral">&#39;;</span>
<a name="l00135"></a>00135 <span class="stringliteral">            &#39;</span>topoIVSigmaLimit<span class="stringliteral">&#39;;</span>
<a name="l00136"></a>00136 <span class="stringliteral">            &#39;</span>gyraseDeltaLK<span class="stringliteral">&#39;;</span>
<a name="l00137"></a>00137 <span class="stringliteral">            &#39;</span>topoIDeltaLK<span class="stringliteral">&#39;;</span>
<a name="l00138"></a>00138 <span class="stringliteral">            &#39;</span>topoIVDeltaLK<span class="stringliteral">&#39;;</span>
<a name="l00139"></a>00139 <span class="stringliteral">            &#39;</span>gyraseActivityRate<span class="stringliteral">&#39;;</span>
<a name="l00140"></a>00140 <span class="stringliteral">            &#39;</span>topoIActivityRate<span class="stringliteral">&#39;;</span>
<a name="l00141"></a>00141 <span class="stringliteral">            &#39;</span>topoIVActivityRate<span class="stringliteral">&#39;;</span>
<a name="l00142"></a>00142 <span class="stringliteral">            &#39;</span>gyraseATPCost<span class="stringliteral">&#39;;</span>
<a name="l00143"></a>00143 <span class="stringliteral">            &#39;</span>topoIATPCost<span class="stringliteral">&#39;;</span>
<a name="l00144"></a>00144 <span class="stringliteral">            &#39;</span>topoIVATPCost<span class="stringliteral">&#39;;</span>
<a name="l00145"></a>00145 <span class="stringliteral">            &#39;</span>foldChangeSlopes<span class="stringliteral">&#39;; </span>
<a name="l00146"></a>00146 <span class="stringliteral">            &#39;</span>foldChangeIntercepts<span class="stringliteral">&#39;;</span>
<a name="l00147"></a>00147 <span class="stringliteral">            &#39;</span>foldChangeLowerSigmaLimit<span class="stringliteral">&#39;;</span>
<a name="l00148"></a>00148 <span class="stringliteral">            &#39;</span>foldChangeUpperSigmaLimit<span class="stringliteral">&#39;; </span>
<a name="l00149"></a>00149 <span class="stringliteral">            &#39;</span>numTranscriptionUnits<span class="stringliteral">&#39;;</span>
<a name="l00150"></a>00150 <span class="stringliteral">            &#39;</span>tuIndexs<span class="stringliteral">&#39;;</span>
<a name="l00151"></a>00151 <span class="stringliteral">            &#39;</span>tuCoordinates<span class="stringliteral">&#39;;</span>
<a name="l00152"></a>00152 <span class="stringliteral">            };</span>
<a name="l00153"></a>00153 <span class="stringliteral">        fittedConstantNames__      = {}; %names of fitted constant properties</span>
<a name="l00154"></a>00154 <span class="stringliteral">        localStateNames__          = {}; %names of simulation state properties redundant with timecourses in this or other processes or the simulation</span>
<a name="l00155"></a>00155 <span class="stringliteral">    end</span>
<a name="l00156"></a>00156 <span class="stringliteral"></span>
<a name="l00157"></a>00157 <span class="stringliteral">    %IDs, names, and local indices</span>
<a name="l00158"></a>00158 <span class="stringliteral">    properties</span>
<a name="l00159"></a>00159 <span class="stringliteral">        stimuliWholeCellModelIDs = {};   %whole cell model IDs of stimuli</span>
<a name="l00160"></a>00160 <span class="stringliteral">        substrateWholeCellModelIDs = {   %whole cell model IDs of substrates</span>
<a name="l00161"></a>00161 <span class="stringliteral">            &#39;</span>ATP<span class="stringliteral">&#39;; &#39;</span>ADP<span class="stringliteral">&#39;; &#39;</span>PI<span class="stringliteral">&#39;; &#39;</span>H2O<span class="stringliteral">&#39;; &#39;</span>H<span class="stringliteral">&#39;};</span>
<a name="l00162"></a>00162 <span class="stringliteral">        substrateIndexs_atp       = 1;</span>
<a name="l00163"></a>00163 <span class="stringliteral">        substrateIndexs_adp       = 2;</span>
<a name="l00164"></a>00164 <span class="stringliteral">        substrateIndexs_phosphate = 3;</span>
<a name="l00165"></a>00165 <span class="stringliteral">        substrateIndexs_water     = 4;</span>
<a name="l00166"></a>00166 <span class="stringliteral">        substrateIndexs_hydrogen  = 5;</span>
<a name="l00167"></a>00167 <span class="stringliteral"></span>
<a name="l00168"></a>00168 <span class="stringliteral">        enzymeWholeCellModelIDs = {      %whole cell model IDs of enzymes</span>
<a name="l00169"></a>00169 <span class="stringliteral">            &#39;</span>DNA_GYRASE<span class="stringliteral">&#39;;                %DNA gyrase</span>
<a name="l00170"></a>00170 <span class="stringliteral">            &#39;</span>MG_203_204_TETRAMER<span class="stringliteral">&#39;;       %DNA topoisomerase IV</span>
<a name="l00171"></a>00171 <span class="stringliteral">            &#39;</span>MG_122_MONOMER<span class="stringliteral">&#39;};           %DNA topoisomerase I</span>
<a name="l00172"></a>00172 <span class="stringliteral">        enzymeIndexs_gyrase = 1;</span>
<a name="l00173"></a>00173 <span class="stringliteral">        enzymeIndexs_topoIV = 2;</span>
<a name="l00174"></a>00174 <span class="stringliteral">        enzymeIndexs_topoI  = 3;</span>
<a name="l00175"></a>00175 <span class="stringliteral">    end</span>
<a name="l00176"></a>00176 <span class="stringliteral">    </span>
<a name="l00177"></a>00177 <span class="stringliteral">    %fixed biological constants</span>
<a name="l00178"></a>00178 <span class="stringliteral">    properties</span>
<a name="l00179"></a>00179 <span class="stringliteral">        gyraseMeanDwellTime       %Mean dwell time of gyrase to bind DNA [45s; PUB_0753]</span>
<a name="l00180"></a>00180 <span class="stringliteral">        </span>
<a name="l00181"></a>00181 <span class="stringliteral">        gyraseSigmaLimit          %superhelical density (\sigma_{sp}) above which gyrase acts [-0.1; PUB_0753]</span>
<a name="l00182"></a>00182 <span class="stringliteral">        topoISigmaLimit           %superhelical density (\sigma_{sp}) above which topoisomerase I acts [0; PUB_0502]</span>
<a name="l00183"></a>00183 <span class="stringliteral">        topoIVSigmaLimit          %superhelical density (\sigma_{sp}) below which topoisomerase IV acts [0; PUB_0753]</span>
<a name="l00184"></a>00184 <span class="stringliteral">        </span>
<a name="l00185"></a>00185 <span class="stringliteral">        gyraseDeltaLK             %number of supercoils introduced by each gyrase strand passing event [-2; PUB_0695]</span>
<a name="l00186"></a>00186 <span class="stringliteral">        topoIDeltaLK              %number of supercoils introduced by each topoisomerase I strand passing event [1; PUB_0502]</span>
<a name="l00187"></a>00187 <span class="stringliteral">        topoIVDeltaLK             %number of supercoils introduced by each topoisomerase IV strand passing event [-2]</span>
<a name="l00188"></a>00188 <span class="stringliteral">        </span>
<a name="l00189"></a>00189 <span class="stringliteral">        gyraseActivityRate        %strand passing events per second per enzyme [1.2; PUB_0753]</span>
<a name="l00190"></a>00190 <span class="stringliteral">        topoIActivityRate         %strand passing events per second per enzyme [1.0; PUB_0502]</span>
<a name="l00191"></a>00191 <span class="stringliteral">        topoIVActivityRate        %strand passing events per second per enzyme [2.5; PUB_0750]</span>
<a name="l00192"></a>00192 <span class="stringliteral">        </span>
<a name="l00193"></a>00193 <span class="stringliteral">        gyraseATPCost             %ATP hydrolyzed per gyrase-catalyzed strand passing event [2]</span>
<a name="l00194"></a>00194 <span class="stringliteral">        topoIATPCost              %ATP hydrolyzed per topoisomerase I-catalyzed strand passing event [0]</span>
<a name="l00195"></a>00195 <span class="stringliteral">        topoIVATPCost             %ATP hydrolyzed per topoisomerase IV-catalyzed strand passing event [2]</span>
<a name="l00196"></a>00196 <span class="stringliteral">        </span>
<a name="l00197"></a>00197 <span class="stringliteral">        topoILogisiticConst       %fittable parameter controlling the steapness of the the logistic function of topoI activity around equilibrium sigma [100]</span>
<a name="l00198"></a>00198 <span class="stringliteral">        gyrLogisiticConst         %fittable parameter controlling the steapness of the the logistic function of gyrase activity around equilibrium sigma [-100]</span>
<a name="l00199"></a>00199 <span class="stringliteral">        </span>
<a name="l00200"></a>00200 <span class="stringliteral">        enzymeProperties          %struct built by buildEnzymeProperties</span>
<a name="l00201"></a>00201 <span class="stringliteral">        </span>
<a name="l00202"></a>00202 <span class="stringliteral">        foldChangeSlopes          %slopes of linear fit of sigma-gene expression data [gyrase, topo IV, topo I] [PUB_0920] </span>
<a name="l00203"></a>00203 <span class="stringliteral">        foldChangeIntercepts      %y-intercepts of linear fit of sigma-gene expression data [gyrase, topo IV, topo I] [PUB_0920]</span>
<a name="l00204"></a>00204 <span class="stringliteral">        foldChangeLowerSigmaLimit %lower bound of sigma for applying linear fit (fittable parameter)</span>
<a name="l00205"></a>00205 <span class="stringliteral">        foldChangeUpperSigmaLimit %upper bound of sigma for applying linear fit (fittable parameter)</span>
<a name="l00206"></a>00206 <span class="stringliteral">        </span>
<a name="l00207"></a>00207 <span class="stringliteral">        numTranscriptionUnits     %number of transcription units</span>
<a name="l00208"></a>00208 <span class="stringliteral">        tuIndexs                  %transcription units whose transcription probs will be affected by supercoiling [gyrase, topo IV, topo I]</span>
<a name="l00209"></a>00209 <span class="stringliteral">        tuCoordinates             %start coords of transcription units whose transcription probs will be affected by supercoiling [gyrase, topo IV, topo I]</span>
<a name="l00210"></a>00210 <span class="stringliteral">    end</span>
<a name="l00211"></a>00211 <span class="stringliteral">    </span>
<a name="l00212"></a>00212 <span class="stringliteral">    %references to cell state</span>
<a name="l00213"></a>00213 <span class="stringliteral">    properties</span>
<a name="l00214"></a>00214 <span class="stringliteral">        rnaPolymerase             %reference to RNA polymerase state</span>
<a name="l00215"></a>00215 <span class="stringliteral">    end</span>
<a name="l00216"></a>00216 <span class="stringliteral"></span>
<a name="l00217"></a>00217 <span class="stringliteral">    %constructor</span>
<a name="l00218"></a>00218 <span class="stringliteral">    methods</span>
<a name="l00219"></a>00219 <span class="stringliteral">        function this = DNASupercoiling(wholeCellModelID, name)</span>
<a name="l00220"></a>00220 <span class="stringliteral">            this = this@edu.stanford.covert.cell.sim.Process(wholeCellModelID, name);</span>
<a name="l00221"></a>00221 <span class="stringliteral">        end</span>
<a name="l00222"></a>00222 <span class="stringliteral">    end</span>
<a name="l00223"></a>00223 <span class="stringliteral"></span>
<a name="l00224"></a>00224 <span class="stringliteral">    %communication between process/simulation</span>
<a name="l00225"></a>00225 <span class="stringliteral">    methods</span>
<a name="l00226"></a>00226 <span class="stringliteral">        %set references to state objects</span>
<a name="l00227"></a>00227 <span class="stringliteral">        function storeObjectReferences(this, simulation)</span>
<a name="l00228"></a>00228 <span class="stringliteral">            this.storeObjectReferences@edu.stanford.covert.cell.sim.Process(simulation);</span>
<a name="l00229"></a>00229 <span class="stringliteral">            this.storeObjectReferences@edu.stanford.covert.cell.sim.ChromosomeProcessAspect(simulation);</span>
<a name="l00230"></a>00230 <span class="stringliteral">            </span>
<a name="l00231"></a>00231 <span class="stringliteral">            this.rnaPolymerase = simulation.state(&#39;</span>RNAPolymerase<span class="stringliteral">&#39;);</span>
<a name="l00232"></a>00232 <span class="stringliteral">            this.states = [this.states; {this.rnaPolymerase}];</span>
<a name="l00233"></a>00233 <span class="stringliteral">        end</span>
<a name="l00234"></a>00234 <span class="stringliteral">        </span>
<a name="l00235"></a>00235 <span class="stringliteral">        %initialize constants</span>
<a name="l00236"></a>00236 <span class="stringliteral">        function initializeConstants(this, knowledgeBase, simulation, varargin)</span>
<a name="l00237"></a>00237 <span class="stringliteral">            this.initializeConstants@edu.stanford.covert.cell.sim.Process(...</span>
<a name="l00238"></a>00238 <span class="stringliteral">                knowledgeBase, simulation, varargin{:});</span>
<a name="l00239"></a>00239 <span class="stringliteral">            this.initializeConstants@edu.stanford.covert.cell.sim.ChromosomeProcessAspect(...</span>
<a name="l00240"></a>00240 <span class="stringliteral">                knowledgeBase, simulation, varargin{:});            </span>
<a name="l00241"></a>00241 <span class="stringliteral">            </span>
<a name="l00242"></a>00242 <span class="stringliteral">            c = this.chromosome;</span>
<a name="l00243"></a>00243 <span class="stringliteral">            </span>
<a name="l00244"></a>00244 <span class="stringliteral">            this.enzymeProperties = this.buildEnzymeProperties();</span>
<a name="l00245"></a>00245 <span class="stringliteral">            </span>
<a name="l00246"></a>00246 <span class="stringliteral">            %get transcription unit indices</span>
<a name="l00247"></a>00247 <span class="stringliteral">            [~, geneIndexs] = ismember({&#39;</span>MG_003<span class="stringliteral">&#39;, &#39;</span>MG_203<span class="stringliteral">&#39;, &#39;</span>MG_122<span class="stringliteral">&#39;}, this.gene.wholeCellModelIDs);</span>
<a name="l00248"></a>00248 <span class="stringliteral">            [this.tuIndexs, ~] = find(this.rna.nascentRNAGeneComposition(geneIndexs, :)&#39;</span>);            
<a name="l00249"></a>00249             this.numTranscriptionUnits = numel(c.transcriptionUnitStartCoordinates);
<a name="l00250"></a>00250             this.tuCoordinates = c.transcriptionUnitStartCoordinates(this.tuIndexs);
<a name="l00251"></a>00251         end
<a name="l00252"></a>00252     end
<a name="l00253"></a>00253 
<a name="l00254"></a>00254     %model
<a name="l00255"></a>00255     methods        
<a name="l00256"></a>00256         %Calculate sufficient metabolites and enzymes to achieve the observed
<a name="l00257"></a>00257         %mean superhelical density after replication of the daughter chromosome:
<a name="l00258"></a>00258         %
<a name="l00259"></a>00259         %  (LK after replication) = 2 * (LK before replication)
<a name="l00260"></a>00260         %                         = (LK before replication)
<a name="l00261"></a>00261         %                            + (links introduced by DNA pol)
<a name="l00262"></a>00262         %                            - (links pushed ahead of replisome and resolved by gyrase)
<a name="l00263"></a>00263         %                            - (links introduced behind replisome by gryase to return to steady-state superhelical density)
<a name="l00264"></a>00264         %               2*LK_{ss} = LK_{ss} + 2*LK_{0} - LK_{ss} - 2*(LK_{0} - LK_{ss})
<a name="l00265"></a>00265         %
<a name="l00266"></a>00266         %That is calculate:
<a name="l00267"></a>00267         %- contribution to <a class="code" href="class_f_b_a.html">FBA</a> objective
<a name="l00268"></a>00268         %- minimum expression consistent with cell cycle length
<a name="l00269"></a>00269         %
<a name="l00270"></a>00270         %FitConstants accounts <span class="keywordflow">for</span> topoisomerase I / gyrase expression balance
<a name="l00271"></a>00271         %required to achieve the observed steady state superhelical density.
<a name="l00272"></a>00272         <span class="keyword">function</span> [bmProd, byProd, minEnzExp, maxEnzExp] = calcResourceRequirements_LifeCycle(<span class="keyword">this</span>, constants, states)
<a name="l00273"></a>00273             %initialize
<a name="l00274"></a>00274             bmProd = zeros(size(this.substrateWholeCellModelIDs));
<a name="l00275"></a>00275             byProd = zeros(size(this.substrateWholeCellModelIDs));
<a name="l00276"></a>00276             minEnzExp = zeros(size(this.enzymeWholeCellModelIDs));
<a name="l00277"></a>00277             maxEnzExp = Inf(size(this.enzymeWholeCellModelIDs));
<a name="l00278"></a>00278             
<a name="l00279"></a>00279             %references
<a name="l00280"></a>00280             c = this.chromosome;
<a name="l00281"></a>00281             t = constants.states.Time;
<a name="l00282"></a>00282             
<a name="l00283"></a>00283             %number of links to be resolved by gyrase, and number of gyrase
<a name="l00284"></a>00284             %link resolution reactions
<a name="l00285"></a>00285             %- links that need to be resolved because of replication
<a name="l00286"></a>00286             %- links that need to be resolved to counter-balance the effect
<a name="l00287"></a>00287             %  of topoisomerase I
<a name="l00288"></a>00288             nRxn = 0;
<a name="l00289"></a>00289             
<a name="l00290"></a>00290             LK_0  =  c.sequenceLen / c.relaxedBasesPerTurn;
<a name="l00291"></a>00291             LK_ss = (1 + c.equilibriumSuperhelicalDensity) * LK_0;
<a name="l00292"></a>00292             dLK = - LK_ss - 2*(LK_0 - LK_ss);
<a name="l00293"></a>00293             nRxn = nRxn + dLK / this.gyraseDeltaLK;
<a name="l00294"></a>00294             
<a name="l00295"></a>00295             topoIProps = this.enzymeProperties(this.enzymeIndexs_topoI);
<a name="l00296"></a>00296             gyrProps = this.enzymeProperties(this.enzymeIndexs_gyrase);
<a name="l00297"></a>00297             nRxn = nRxn + 0.25 * ... %factor of 0.25 added based on observing simulations
<a name="l00298"></a>00298                 abs(states.monomers(this.enzymeGlobalIndexs(this.enzymeIndexs_topoI)) * ...
<a name="l00299"></a>00299                 topoIProps.activityRate * topoIProps.deltaLK / gyrProps.deltaLK);
<a name="l00300"></a>00300             
<a name="l00301"></a>00301             %substrates and byproducts
<a name="l00302"></a>00302             nATP = nRxn * this.enzymeProperties(this.enzymeIndexs_gyrase).atpCost;
<a name="l00303"></a>00303             bmProd(this.substrateIndexs_atp)       = nATP;
<a name="l00304"></a>00304             bmProd(this.substrateIndexs_water)     = nATP;
<a name="l00305"></a>00305             byProd(this.substrateIndexs_adp)       = nATP;
<a name="l00306"></a>00306             byProd(this.substrateIndexs_phosphate) = nATP;
<a name="l00307"></a>00307             byProd(this.substrateIndexs_hydrogen)  = nATP;
<a name="l00308"></a>00308             
<a name="l00309"></a>00309             %gyrase
<a name="l00310"></a>00310             minEnzExp(this.enzymeIndexs_gyrase) = ...
<a name="l00311"></a>00311                 nRxn / t.replicationDuration / this.gyraseActivityRate / ...
<a name="l00312"></a>00312                 exp(log(2) * t.replicationInitiationDuration / t.cellCycleLength);
<a name="l00313"></a>00313         end
<a name="l00314"></a>00314 
<a name="l00315"></a>00315         %initialization:
<a name="l00316"></a>00316         %- Simulation initializeState method initializes 1
<a name="l00317"></a>00317         %  chromosome with linkingNumber such that the superhelical density is
<a name="l00318"></a>00318         %  equal to the equilibriumSuperhelicalDensity
<a name="l00319"></a>00319         %- Here we bind all gyrase to the  
<a name="l00320"></a>00320         function initializeState(this)
<a name="l00321"></a>00321             %bind gyrase
<a name="l00322"></a>00322             nGyraseBinding = this.randStream.stochasticRound(...
<a name="l00323"></a>00323                 this.enzymes(this.enzymeIndexs_gyrase) * ...
<a name="l00324"></a>00324                 (1 - 1/this.gyraseMeanDwellTime/this.stepSizeSec));
<a name="l00325"></a>00325             this.bindProteinToChromosomeStochastically(this.enzymeIndexs_gyrase, ...
<a name="l00326"></a>00326                 nGyraseBinding);
<a name="l00327"></a>00327             
<a name="l00328"></a>00328             %Calculate effects of supercoiling on probability of transcription
<a name="l00329"></a>00329             this.rnaPolymerase.supercoilingBindingProbFoldChange = this.calcRNAPolymeraseBindingProbFoldChange();
<a name="l00330"></a>00330         end
<a name="l00331"></a>00331         
<a name="l00332"></a>00332         %resource requirements
<a name="l00333"></a>00333         <span class="keyword">function</span> result = calcResourceRequirements_Current(<span class="keyword">this</span>)
<a name="l00334"></a>00334             %chromosome reference
<a name="l00335"></a>00335             c = this.chromosome;
<a name="l00336"></a>00336             
<a name="l00337"></a>00337             %initialize
<a name="l00338"></a>00338             result = zeros(size(this.substrates));
<a name="l00339"></a>00339             
<a name="l00340"></a>00340             %ATP needed <span class="keywordflow">for</span> gyrase activity
<a name="l00341"></a>00341             [~, sigmas] = find(c.superhelicalDensity);
<a name="l00342"></a>00342             
<a name="l00343"></a>00343             enzProps = this.enzymeProperties(this.enzymeIndexs_gyrase);
<a name="l00344"></a>00344             nATP = enzProps.atpCost * ceil(...
<a name="l00345"></a>00345                 (this.enzymes(this.enzymeIndexs_gyrase) + this.boundEnzymes(this.enzymeIndexs_gyrase)) * ...
<a name="l00346"></a>00346                 enzProps.activityRate *  max(enzProps.probOfActivityFunc(sigmas, c)));
<a name="l00347"></a>00347             
<a name="l00348"></a>00348             result(this.substrateIndexs_atp)   = nATP;
<a name="l00349"></a>00349             result(this.substrateIndexs_water) = nATP;
<a name="l00350"></a>00350         end
<a name="l00351"></a>00351 
<a name="l00352"></a>00352         %simulation
<a name="l00353"></a>00353         <span class="keyword">function</span> evolveState(<span class="keyword">this</span>)
<a name="l00354"></a>00354             import edu.stanford.covert.util.<a class="code" href="class_circular_sparse_mat.html">CircularSparseMat</a>;
<a name="l00355"></a>00355             import edu.stanford.covert.util.<a class="code" href="class_computation_util.html">ComputationUtil</a>;
<a name="l00356"></a>00356             
<a name="l00357"></a>00357             c = this.chromosome;
<a name="l00358"></a>00358             
<a name="l00359"></a>00359             %get enzyme <a class="code" href="json_parse_8m.html#a6e4256f69b751d3d7ec4ea52ff5a8dee">properties</a>
<a name="l00360"></a>00360             enzProps = this.enzymeProperties;
<a name="l00361"></a>00361 
<a name="l00362"></a>00362             % Calculate sigmas
<a name="l00363"></a>00363             [tmpPosStrands, lengths] = find(c.doubleStrandedRegions);
<a name="l00364"></a>00364             tmpIdxs = find(mod(tmpPosStrands(:, 2), 2));
<a name="l00365"></a>00365             dsPosStrands = tmpPosStrands(tmpIdxs, :);
<a name="l00366"></a>00366             lengths = lengths(tmpIdxs, 1);
<a name="l00367"></a>00367             
<a name="l00368"></a>00368             [~, linkingNumbers] = find(c.linkingNumbers);
<a name="l00369"></a>00369             linkingNumbers = linkingNumbers(tmpIdxs, 1);
<a name="l00370"></a>00370             sigmas = (linkingNumbers - lengths/c.relaxedBasesPerTurn) ./ (lengths/c.relaxedBasesPerTurn);
<a name="l00371"></a>00371 
<a name="l00372"></a>00372             % Determine which regions accommodate which proteins.
<a name="l00373"></a>00373             legal = [...
<a name="l00374"></a>00374                 sigmas &gt; this.gyraseSigmaLimit ...
<a name="l00375"></a>00375                 sigmas &gt; this.topoIVSigmaLimit ...
<a name="l00376"></a>00376                 sigmas &lt; this.topoISigmaLimit];
<a name="l00377"></a>00377 
<a name="l00378"></a>00378             % Unbind topoIV where it can&#39;t act.
<a name="l00379"></a>00379             this.releaseProteinFromChromosome(...
<a name="l00380"></a>00380                 this.enzymeIndexs_topoIV, Inf,...
<a name="l00381"></a>00381                 dsPosStrands(legal(:, this.enzymeIndexs_topoIV), :),... %protected regions
<a name="l00382"></a>00382                 lengths(legal(:, this.enzymeIndexs_topoIV)));
<a name="l00383"></a>00383             
<a name="l00384"></a>00384             % Probabilistically unbind gyrase.
<a name="l00385"></a>00385             this.releaseProteinFromChromosome(this.enzymeIndexs_gyrase, 1 / this.gyraseMeanDwellTime, [], []);
<a name="l00386"></a>00386             
<a name="l00387"></a>00387             % Bind enzymes in regions with accommodating sigmas.
<a name="l00388"></a>00388             order = this.randStream.randperm(length(this.enzymes));
<a name="l00389"></a>00389             transientlyBinding = zeros(size(legal));
<a name="l00390"></a>00390             for i = 1:length(order)
<a name="l00391"></a>00391                 j = order(i);
<a name="l00392"></a>00392                 
<a name="l00393"></a>00393                 if enzProps(j).meanDwellTime == 0
<a name="l00394"></a>00394                     if numel(lengths) == 1 &amp;&amp; lengths == c.sequenceLen
<a name="l00395"></a>00395                         [~, mons] = find(c.monomerBoundSites);
<a name="l00396"></a>00396                         [~, cpxs] = find(c.complexBoundSites);
<a name="l00397"></a>00397                         space = lengths ...
<a name="l00398"></a>00398                             - sum(c.monomerDNAFootprints(mons, 1)) ...
<a name="l00399"></a>00399                             - sum(c.complexDNAFootprints(cpxs, 1)) ...
<a name="l00400"></a>00400                             - nnz(c.damagedSites);
<a name="l00401"></a>00401                         transientlyBinding(1, j) = min(this.enzymes(j, 1), space / this.enzymeDNAFootprints(j, 1));
<a name="l00402"></a>00402                     elseif numel(lengths) == 2 &amp;&amp; all(lengths == c.sequenceLen)
<a name="l00403"></a>00403                         [monPosStrnds, mons] = find(c.monomerBoundSites);
<a name="l00404"></a>00404                         [cpxPosStrnds, cpxs] = find(c.complexBoundSites);
<a name="l00405"></a>00405                         space = zeros(2, 1);
<a name="l00406"></a>00406                         space(1) = lengths(1) ...
<a name="l00407"></a>00407                             - sum(c.monomerDNAFootprints(mons(monPosStrnds(:, 2) &lt;=2 ), 1)) ...
<a name="l00408"></a>00408                             - sum(c.complexDNAFootprints(cpxs(cpxPosStrnds(:, 2) &lt;=2 ), 1)) ...
<a name="l00409"></a>00409                             - nnz(c.damagedSites(:, 1:2));
<a name="l00410"></a>00410                         space(2) = lengths(2) ...
<a name="l00411"></a>00411                             - sum(c.monomerDNAFootprints(mons(monPosStrnds(:, 2) &gt;2 ), 1)) ...
<a name="l00412"></a>00412                             - sum(c.complexDNAFootprints(cpxs(cpxPosStrnds(:, 2) &gt;2 ), 1)) ...
<a name="l00413"></a>00413                             - nnz(c.damagedSites(:, 3:4));
<a name="l00414"></a>00414                         transientlyBinding(:, j) = min(this.enzymes(j, 1), sum(space) / this.enzymeDNAFootprints(j, 1)) * ...
<a name="l00415"></a>00415                             space / sum(space);
<a name="l00416"></a>00416                         if this.randStream.rand &lt; 0.5
<a name="l00417"></a>00417                             transientlyBinding(1, j) = ComputationUtil.roundHalfUp(transientlyBinding(1, j));
<a name="l00418"></a>00418                             transientlyBinding(2, j) = ComputationUtil.roundHalfDown(transientlyBinding(2, j));
<a name="l00419"></a>00419                         else
<a name="l00420"></a>00420                             transientlyBinding(2, j) = ComputationUtil.roundHalfUp(transientlyBinding(2, j));
<a name="l00421"></a>00421                             transientlyBinding(1, j) = ComputationUtil.roundHalfDown(transientlyBinding(1, j));
<a name="l00422"></a>00422                         end
<a name="l00423"></a>00423                     else
<a name="l00424"></a>00424                         [monPosStrnds, mons] = find(c.monomerBoundSites);
<a name="l00425"></a>00425                         [cmpPosStrnds, cmps] = find(c.complexBoundSites);
<a name="l00426"></a>00426                         dmgPosStrnds = find(c.damagedSites);
<a name="l00427"></a>00427                         monPos = monPosStrnds(:, 1);
<a name="l00428"></a>00428                         cmpPos = cmpPosStrnds(:, 1);
<a name="l00429"></a>00429                         dmgPos = dmgPosStrnds(:, 1);
<a name="l00430"></a>00430                         monChrs = ceil(monPosStrnds(:, 2) / 2);
<a name="l00431"></a>00431                         cmpChrs = ceil(cmpPosStrnds(:, 2) / 2);
<a name="l00432"></a>00432                         dmgChrs = ceil(dmgPosStrnds(:, 2) / 2);
<a name="l00433"></a>00433                         
<a name="l00434"></a>00434                         space = zeros(size(lengths));
<a name="l00435"></a>00435                         for k = 1:numel(lengths)
<a name="l00436"></a>00436                             if ~legal(k, j)
<a name="l00437"></a>00437                                 continue;
<a name="l00438"></a>00438                             end
<a name="l00439"></a>00439                             
<a name="l00440"></a>00440                             monTfs = monPos &gt;= dsPosStrands(k, 1) &amp; monPos &lt;= dsPosStrands(k, 1) + lengths(k, 1) - 1;
<a name="l00441"></a>00441                             cmpTfs = cmpPos &gt;= dsPosStrands(k, 1) &amp; cmpPos &lt;= dsPosStrands(k, 1) + lengths(k, 1) - 1;
<a name="l00442"></a>00442                             dmgTfs = dmgPos &gt;= dsPosStrands(k, 1) &amp; dmgPos &lt;= dsPosStrands(k, 1) + lengths(k, 1) - 1;
<a name="l00443"></a>00443                             monTfs(monTfs) = monChrs(monTfs, 1) == dsPosStrands(k, 2);
<a name="l00444"></a>00444                             cmpTfs(cmpTfs) = cmpChrs(cmpTfs, 1) == dsPosStrands(k, 2);
<a name="l00445"></a>00445                             dmgTfs(dmgTfs) = dmgChrs(dmgTfs, 1) == dsPosStrands(k, 2);
<a name="l00446"></a>00446                             
<a name="l00447"></a>00447                             space(k, 1) = ...
<a name="l00448"></a>00448                                 + lengths(k, 1) ...
<a name="l00449"></a>00449                                 - sum(c.monomerDNAFootprints(mons(monTfs, 1), 1)) ...
<a name="l00450"></a>00450                                 - sum(c.monomerDNAFootprints(cmps(cmpTfs, 1), 1)) ...
<a name="l00451"></a>00451                                 - sum(dmgTfs);
<a name="l00452"></a>00452                         end
<a name="l00453"></a>00453                         
<a name="l00454"></a>00454                         if numel(lengths) == 1
<a name="l00455"></a>00455                             transientlyBinding(:, j) = min(this.enzymes(j, 1), space / this.enzymeDNAFootprints(j, 1));
<a name="l00456"></a>00456                         else
<a name="l00457"></a>00457                             transientlyBinding(:, j) = space / sum(space) * min(this.enzymes(j, 1), sum(space) / this.enzymeDNAFootprints(j, 1));
<a name="l00458"></a>00458                         end
<a name="l00459"></a>00459                     end
<a name="l00460"></a>00460                 elseif any(legal(:, j))
<a name="l00461"></a>00461                     this.bindProteinToChromosomeStochastically(...
<a name="l00462"></a>00462                         j, [], dsPosStrands(legal(:,j),:), lengths(legal(:,j)));
<a name="l00463"></a>00463                 end
<a name="l00464"></a>00464             end
<a name="l00465"></a>00465 
<a name="l00466"></a>00466             % Calculate new linking numbers and do substrate accounting.
<a name="l00467"></a>00467             enzProps = enzProps(this.randStream.randperm(length(enzProps)));
<a name="l00468"></a>00468             for i = 1:size(dsPosStrands, 1)
<a name="l00469"></a>00469                 len = lengths(i);
<a name="l00470"></a>00470                 for j = 1:length(enzProps)
<a name="l00471"></a>00471                     if legal(i, enzProps(j).idx)
<a name="l00472"></a>00472                         if enzProps(j).meanDwellTime == 0
<a name="l00473"></a>00473                             nBound = transientlyBinding(i, enzProps(j).idx);
<a name="l00474"></a>00474                         else
<a name="l00475"></a>00475                             nBound = size(this.findProteinInRegion(...
<a name="l00476"></a>00476                                 dsPosStrands(i, 1), dsPosStrands(i, 2), len, enzProps(j).idx), 1);
<a name="l00477"></a>00477                         end
<a name="l00478"></a>00478                         % use logistic function to determine the
<a name="l00479"></a>00479                         % probability of enzyme activity
<a name="l00480"></a>00480                         % Use sigmas from beginning of timestep, before any
<a name="l00481"></a>00481                         % enzymes have acted
<a name="l00482"></a>00482                         probOfActivity = enzProps(j).probOfActivityFunc(sigmas(i),c);
<a name="l00483"></a>00483                         nStrandPassingEvents = min([
<a name="l00484"></a>00484                             this.randStream.stochasticRound(nBound * enzProps(j).activityRate *  probOfActivity);
<a name="l00485"></a>00485                             fix(this.substrates([this.substrateIndexs_atp; this.substrateIndexs_water]) / enzProps(j).atpCost)
<a name="l00486"></a>00486                             ]);
<a name="l00487"></a>00487                         %adjust linking number
<a name="l00488"></a>00488                         linkingNumbers(i) = linkingNumbers(i) + enzProps(j).deltaLK * nStrandPassingEvents;
<a name="l00489"></a>00489                         
<a name="l00490"></a>00490                         %substrate accounting
<a name="l00491"></a>00491                         nATP = nStrandPassingEvents * enzProps(j).atpCost;
<a name="l00492"></a>00492                         this.substrates(this.substrateIndexs_atp)       = this.substrates(this.substrateIndexs_atp)       - nATP;
<a name="l00493"></a>00493                         this.substrates(this.substrateIndexs_water)     = this.substrates(this.substrateIndexs_water)     - nATP;
<a name="l00494"></a>00494                         this.substrates(this.substrateIndexs_adp)       = this.substrates(this.substrateIndexs_adp)       + nATP;
<a name="l00495"></a>00495                         this.substrates(this.substrateIndexs_phosphate) = this.substrates(this.substrateIndexs_phosphate) + nATP;
<a name="l00496"></a>00496                         this.substrates(this.substrateIndexs_hydrogen)  = this.substrates(this.substrateIndexs_hydrogen)  + nATP;
<a name="l00497"></a>00497                     end
<a name="l00498"></a>00498                 end
<a name="l00499"></a>00499             end
<a name="l00500"></a>00500             c.linkingNumbers = CircularSparseMat(...
<a name="l00501"></a>00501                 [dsPosStrands; dsPosStrands(:, 1) dsPosStrands(:, 2)+1], [linkingNumbers; linkingNumbers], ...
<a name="l00502"></a>00502                 [c.sequenceLen c.nCompartments], 1);    
<a name="l00503"></a>00503             %Calculate effects of supercoiling on probability of transcription
<a name="l00504"></a>00504             this.rnaPolymerase.supercoilingBindingProbFoldChange = this.calcRNAPolymeraseBindingProbFoldChange(dsPosStrands, lengths, linkingNumbers);
<a name="l00505"></a>00505         end
<a name="l00506"></a>00506         
<a name="l00507"></a>00507         function foldChange = calcRNAPolymeraseBindingProbFoldChange(this, dsPosStrands, lengths, linkingNumbers)
<a name="l00508"></a>00508             c = this.chromosome;
<a name="l00509"></a>00509             
<a name="l00510"></a>00510             if nargin == 1
<a name="l00511"></a>00511                 [tmpPosStrands, lengths] = find(c.doubleStrandedRegions);
<a name="l00512"></a>00512                 tmpIdxs = find(mod(tmpPosStrands(:, 2),2));
<a name="l00513"></a>00513                 dsPosStrands = tmpPosStrands(tmpIdxs, :);
<a name="l00514"></a>00514                 lengths = lengths(tmpIdxs, 1);
<a name="l00515"></a>00515                 
<a name="l00516"></a>00516                 [~, linkingNumbers] = find(c.linkingNumbers);
<a name="l00517"></a>00517                 linkingNumbers = linkingNumbers(tmpIdxs, 1);
<a name="l00518"></a>00518             end
<a name="l00519"></a>00519             
<a name="l00520"></a>00520             foldChange = ones(size(c.transcriptionUnitStartCoordinates, 1), 2); %initialize fold change
<a name="l00521"></a>00521             sigmas = (linkingNumbers - lengths/c.relaxedBasesPerTurn) ./ (lengths/c.relaxedBasesPerTurn);
<a name="l00522"></a>00522             for i = 1:length(this.tuCoordinates)
<a name="l00523"></a>00523                 %find <span class="keywordtype">double</span> stranded region where transcription <a class="code" href="unit_8m.html#a94c091bed34da4edc337e42e23029000">unit</a> lies
<a name="l00524"></a>00524                 rgnIdxs = find(...
<a name="l00525"></a>00525                     dsPosStrands(:,1)               &lt;= this.tuCoordinates(i) &amp; ...
<a name="l00526"></a>00526                     dsPosStrands(:,1) + lengths - 1 &gt;= this.tuCoordinates(i));
<a name="l00527"></a>00527                 
<a name="l00528"></a>00528                 for j = 1:numel(rgnIdxs)
<a name="l00529"></a>00529                     %piecewise linear function of sigma of <span class="keywordtype">double</span> stranded region
<a name="l00530"></a>00530                     %where transcription <a class="code" href="unit_8m.html#a94c091bed34da4edc337e42e23029000">unit</a> lies
<a name="l00531"></a>00531                     thresholdedSigma = min(...
<a name="l00532"></a>00532                         max(sigmas(rgnIdxs(j)), this.foldChangeLowerSigmaLimit),...
<a name="l00533"></a>00533                         this.foldChangeUpperSigmaLimit);
<a name="l00534"></a>00534                     foldChange(...
<a name="l00535"></a>00535                         this.tuIndexs(i), ceil(dsPosStrands(rgnIdxs(j), 2)/2)) = ...
<a name="l00536"></a>00536                         this.foldChangeIntercepts(i) + this.foldChangeSlopes(i) * thresholdedSigma;
<a name="l00537"></a>00537                 end
<a name="l00538"></a>00538             end
<a name="l00539"></a>00539         end
<a name="l00540"></a>00540         
<a name="l00541"></a>00541         function value = buildEnzymeProperties(this)
<a name="l00542"></a>00542             value = [...
<a name="l00543"></a>00543                 struct(...
<a name="l00544"></a>00544                     &#39;idx&#39;, this.enzymeIndexs_gyrase,...
<a name="l00545"></a>00545                     &#39;sigmaLimit&#39;, this.gyraseSigmaLimit, ...
<a name="l00546"></a>00546                     &#39;isActiveInRegion&#39;, @(sigmas) sigmas &gt; this.gyraseSigmaLimit, ...
<a name="l00547"></a>00547                     &#39;atpCost&#39;, this.gyraseATPCost,...
<a name="l00548"></a>00548                     &#39;deltaLK&#39;, this.gyraseDeltaLK,...
<a name="l00549"></a>00549                     &#39;activityRate&#39;, this.gyraseActivityRate,...
<a name="l00550"></a>00550                     &#39;meanDwellTime&#39;, this.gyraseMeanDwellTime,...
<a name="l00551"></a>00551                     &#39;probOfActivityFunc&#39;, @(sigma, c) 1 ./ (1 + exp(this.gyrLogisiticConst * (sigma - c.equilibriumSuperhelicalDensity))));
<a name="l00552"></a>00552                 struct(...
<a name="l00553"></a>00553                     &#39;idx&#39;, this.enzymeIndexs_topoIV,...
<a name="l00554"></a>00554                     &#39;sigmaLimit&#39;, this.topoIVSigmaLimit, ...
<a name="l00555"></a>00555                     &#39;isActiveInRegion&#39;, @(sigmas) sigmas &gt; this.topoIVSigmaLimit, ...
<a name="l00556"></a>00556                     &#39;atpCost&#39;, this.topoIVATPCost,...
<a name="l00557"></a>00557                     &#39;deltaLK&#39;, this.topoIVDeltaLK,...
<a name="l00558"></a>00558                     &#39;activityRate&#39;, this.topoIVActivityRate,...
<a name="l00559"></a>00559                     &#39;meanDwellTime&#39;, NaN,...
<a name="l00560"></a>00560                     &#39;probOfActivityFunc&#39;, @(sigma, c) ones(size(sigma)));
<a name="l00561"></a>00561                 struct(...
<a name="l00562"></a>00562                     &#39;idx&#39;, this.enzymeIndexs_topoI,...
<a name="l00563"></a>00563                     &#39;sigmaLimit&#39;, this.topoISigmaLimit, ...
<a name="l00564"></a>00564                     &#39;isActiveInRegion&#39;, @(sigmas) sigmas &lt; this.topoISigmaLimit, ...
<a name="l00565"></a>00565                     &#39;atpCost&#39;, this.topoIATPCost,...
<a name="l00566"></a>00566                     &#39;deltaLK&#39;, this.topoIDeltaLK,...
<a name="l00567"></a>00567                     &#39;activityRate&#39;, this.topoIActivityRate,...
<a name="l00568"></a>00568                     &#39;meanDwellTime&#39;, 0,...
<a name="l00569"></a>00569                     &#39;probOfActivityFunc&#39;, @(sigma, c) 1 ./ (1 + exp(this.topoILogisiticConst * (sigma - c.equilibriumSuperhelicalDensity))))];
<a name="l00570"></a>00570         end
<a name="l00571"></a>00571     end
<a name="l00572"></a>00572 end
</pre></div></div><!-- contents -->
</div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
<a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(0)"><span class="SelectionMark">&#160;</span>All</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(1)"><span class="SelectionMark">&#160;</span>Classes</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(2)"><span class="SelectionMark">&#160;</span>Files</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(3)"><span class="SelectionMark">&#160;</span>Functions</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(4)"><span class="SelectionMark">&#160;</span>Variables</a></div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

  <div id="nav-path" class="navpath">
    <ul>
      <li class="navelem"><a class="el" href="_d_n_a_supercoiling_8m.html">DNASupercoiling.m</a>      </li>

    <li class="footer">Generated on Wed Feb 8 2012 21:42:32 for Whole Cell Simulation by
    <a href="http://www.doxygen.org/index.html">
    <img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.7.6.1 </li>
   </ul>
 </div>


</body>
</html>
