<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<title>Whole Cell Simulation: src/+edu/+stanford/+covert/+cell/+sim/+analysis/Figures34.m Source File</title>

<link href="tabs.css" rel="stylesheet" type="text/css"/>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
<link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="navtree.js"></script>
<script type="text/javascript">
  $(document).ready(initResizable);
</script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
  $(document).ready(function() { searchBox.OnSelectItem(0); });
</script>

</head>
<body>
<div id="top"><!-- do not remove this div! -->


<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  
  
  <td style="padding-left: 0.5em;">
   <div id="projectname">Whole Cell Simulation
   
   </div>
   
  </td>
  
  
  
 </tr>
 </tbody>
</table>
</div>

<!-- Generated by Doxygen 1.7.6.1 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "search",false,'Search');
</script>
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&#160;Page</span></a></li>
      <li><a href="annotated.html"><span>Classes</span></a></li>
      <li class="current"><a href="files.html"><span>Files</span></a></li>
      <li>
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
  <div id="navrow2" class="tabs2">
    <ul class="tablist">
      <li><a href="files.html"><span>File&#160;List</span></a></li>
      <li><a href="globals.html"><span>File&#160;Members</span></a></li>
    </ul>
  </div>
</div>
<div id="side-nav" class="ui-resizable side-nav-resizable">
  <div id="nav-tree">
    <div id="nav-tree-contents">
    </div>
  </div>
  <div id="splitbar" style="-moz-user-select:none;" 
       class="ui-resizable-handle">
  </div>
</div>
<script type="text/javascript">
  initNavTree('_figures34_8m.html','');
</script>
<div id="doc-content">
<div class="header">
  <div class="headertitle">
<div class="title">src/+edu/+stanford/+covert/+cell/+sim/+analysis/Figures34.m</div>  </div>
</div><!--header-->
<div class="contents">
<a href="_figures34_8m.html">Go to the documentation of this file.</a><div class="fragment"><pre class="fragment"><a name="l00001"></a>00001 % Figures 3-4
<a name="l00002"></a>00002 %
<a name="l00003"></a>00003 % Author: Derek Macklin, macklin@stanford.edu
<a name="l00004"></a>00004 % Author: Jonathan Karr, jkarr@stanford.edu
<a name="l00005"></a>00005 % Affiliation: Covert Lab, Department of Bioengineering, Stanford University
<a name="l00006"></a>00006 % Created: 1/12/2012
<a name="l00007"></a><a class="code" href="class_figures34.html">00007</a> classdef Figures34
<a name="l00008"></a>00008     methods (Static = true)
<a name="l00009"></a>00009         function run(outDirectory)
<a name="l00010"></a>00010             import edu.stanford.covert.cell.sim.analysis.Figures34;
<a name="l00011"></a>00011             <span class="keyword">import</span> edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
<a name="l00012"></a>00012             
<a name="l00013"></a>00013             %output directory
<a name="l00014"></a>00014             <span class="keywordflow">if</span> nargin &lt; 1
<a name="l00015"></a>00015                 outDirectory = <span class="stringliteral">&#39;documentation/paper/figures&#39;</span>;
<a name="l00016"></a>00016             end
<a name="l00017"></a>00017             
<a name="l00018"></a>00018             %select simulation
<a name="l00019"></a>00019             simBatchDir = [SimulationDiskUtil.getBaseDir() filesep &#39;2011_10_25_01_45_44&#39; filesep];
<a name="l00020"></a>00020             selectedSim = 10;
<a name="l00021"></a>00021             
<a name="l00022"></a>00022             %setup linux path
<a name="l00023"></a>00023             <span class="keywordflow">if</span> isunix()
<a name="l00024"></a>00024                 a = getenv(&#39;LD_LIBRARY_PATH&#39;);
<a name="l00025"></a>00025                 a = [&#39;/usr/lib:&#39; a];
<a name="l00026"></a>00026                 setenv(&#39;LD_LIBRARY_PATH&#39;, a);
<a name="l00027"></a>00027             end
<a name="l00028"></a>00028             
<a name="l00029"></a>00029             % figure 3
<a name="l00030"></a>00030             <a class="code" href="class_figures34.html">Figures34</a>.figure3(simBatchDir, selectedSim, outDirectory)
<a name="l00031"></a>00031             
<a name="l00032"></a>00032             % figure 4
<a name="l00033"></a>00033             <a class="code" href="class_figures34.html">Figures34</a>.figure4(simBatchDir, selectedSim, outDirectory)
<a name="l00034"></a>00034         end
<a name="l00035"></a>00035         
<a name="l00036"></a>00036         function figure3(simBatchDir, selectedSim, outDirectory)
<a name="l00037"></a>00037             import edu.stanford.covert.cell.sim.analysis.<a class="code" href="class_figures34.html">Figures34</a>;
<a name="l00038"></a>00038             import edu.stanford.covert.cell.sim.util.PlotUtil;
<a name="l00039"></a>00039             
<a name="l00040"></a>00040             outDir3 = [outDirectory filesep &#39;figure3&#39; filesep];
<a name="l00041"></a>00041             
<a name="l00042"></a>00042             %% figure 3
<a name="l00043"></a>00043             w = 18.3; h = 12; %Nature 2 column
<a name="l00044"></a>00044             [~, figHandle] = PlotUtil.newAxesHandle();
<a name="l00045"></a>00045             clf(figHandle);
<a name="l00046"></a>00046             set(figHandle, &#39;PaperUnits&#39;, &#39;centimeters&#39;);
<a name="l00047"></a>00047             set(figHandle, &#39;PaperSize&#39;, [w h]);
<a name="l00048"></a>00048             set(figHandle, &#39;PaperPositionMode&#39;, &#39;manual&#39;);
<a name="l00049"></a>00049             set(figHandle, &#39;PaperPosition&#39;, [0 0 get(figHandle, &#39;PaperSize&#39;)]);
<a name="l00050"></a>00050             set(figHandle, &#39;Units&#39;, &#39;normalized&#39;);
<a name="l00051"></a>00051             
<a name="l00052"></a>00052             %A
<a name="l00053"></a>00053             PlotUtil.labelSubFigure(&#39;a&#39;, [0.05 0.992 -0.0286 -0.0208], figHandle);
<a name="l00054"></a>00054             Figures34.growthMeasurement(figHandle, [0.0315 0.73 0.33 0.26], outDir3);
<a name="l00055"></a>00055             
<a name="l00056"></a>00056             %B
<a name="l00057"></a>00057             PlotUtil.labelSubFigure(&#39;b&#39;, [0.427 0.992 -0.0286 -0.0208], figHandle);
<a name="l00058"></a>00058             Figures34.populationMass(simBatchDir, selectedSim, figHandle, [0.430 0.73 0.33 0.26], outDir3);
<a name="l00059"></a>00059             
<a name="l00060"></a>00060             %C
<a name="l00061"></a>00061             PlotUtil.labelSubFigure(&#39;c&#39;, [0.82 0.992 -0.0286 -0.0208], figHandle);
<a name="l00062"></a>00062             Figures34.dryBiomassComposition(simBatchDir, figHandle, [0.8 0.73 0.2 0.26], outDir3);
<a name="l00063"></a>00063             
<a name="l00064"></a>00064             %D: metabolic map
<a name="l00065"></a>00065             PlotUtil.labelSubFigure(&#39;d&#39;, [0.05 0.69 -0.0286 -0.0208], figHandle);
<a name="l00066"></a>00066             Figures34.metabolicMap(simBatchDir, selectedSim, outDir3, figHandle, [0.01 0.35 0.57 0.33]);
<a name="l00067"></a>00067             
<a name="l00068"></a>00068             %E: ratio of predicted to observed metabolite concentrations
<a name="l00069"></a>00069             PlotUtil.labelSubFigure(&#39;<a class="code" href="json_parse_8m.html#ac14d26badced2970320ee4d772ad0ae9">e</a>&#39;, [0.05 0.32 -0.0286 -0.0208], figHandle);
<a name="l00070"></a>00070             Figures34.metaboliteConcentrations(simBatchDir, selectedSim, outDir3, figHandle, [0.0315 0.07 0.55 0.24])
<a name="l00071"></a>00071             
<a name="l00072"></a>00072             %F: <a class="code" href="classm_r_n_a.html">mRNA</a>, protein expression dynamics
<a name="l00073"></a>00073             PlotUtil.labelSubFigure(&#39;f&#39;, [0.65 0.67 -0.0286 -0.0208], figHandle);
<a name="l00074"></a>00074             Figures34.proteinBursting(simBatchDir, selectedSim, outDir3, figHandle, [0.64 0.38 0.36 0.28]);
<a name="l00075"></a>00075             
<a name="l00076"></a>00076             %G: scatter plot of <a class="code" href="class_r_n_a.html">RNA</a> vs protein expression with fit marginals
<a name="l00077"></a>00077             PlotUtil.labelSubFigure(&#39;g&#39;, [0.65 0.32 -0.0286 -0.0208], figHandle);
<a name="l00078"></a>00078             Figures34.RNAvProtein(simBatchDir, outDir3, figHandle, [0.64 0.07 0.16 0.24]);
<a name="l00079"></a>00079             
<a name="l00080"></a>00080             %H: gene essentially predictions
<a name="l00081"></a>00081             PlotUtil.labelSubFigure(&#39;h&#39;, [0.885 0.32 -0.0286 -0.0208], figHandle);
<a name="l00082"></a>00082             Figures34.geneDeletionPrediction([], [], [], figHandle, [0.855 0.07 0.24*h/w 0.24*0.8]);
<a name="l00083"></a>00083             
<a name="l00084"></a>00084             %standarize tick lengths
<a name="l00085"></a>00085             axesHandles = findobj(figHandle, &#39;type&#39;, &#39;axes&#39;, &#39;-and&#39;, &#39;tag&#39;, &#39;&#39;);
<a name="l00086"></a>00086             axesPos = cell2mat(get(axesHandles, &#39;position&#39;));
<a name="l00087"></a>00087             tickLen = 0.01 * max(max(axesPos(:, 3:4), [], 1) .* [w h]);
<a name="l00088"></a>00088             for i = 2:numel(axesHandles)
<a name="l00089"></a>00089                 maxVal = max(axesPos(i, 3:4) .* [w h]);
<a name="l00090"></a>00090                 set(axesHandles(i), &#39;TickLen&#39;, [tickLen / maxVal  0.0250]);
<a name="l00091"></a>00091             end
<a name="l00092"></a>00092             
<a name="l00093"></a>00093             %save
<a name="l00094"></a>00094             print(figHandle, [outDirectory filesep &#39;figure3.pdf&#39;], &#39;-dpdf&#39;, &#39;-cmyk&#39;);
<a name="l00095"></a>00095             close(figHandle);
<a name="l00096"></a>00096             
<a name="l00097"></a>00097             %% supplemental figures
<a name="l00098"></a>00098             
<a name="l00099"></a>00099             %dynamics of gene, <a class="code" href="class_r_n_a.html">RNA</a>, protein copy number; reaction flux, growth rate
<a name="l00100"></a>00100             [~, figHandle] = PlotUtil.newAxesHandle();
<a name="l00101"></a>00101             Figures34.geneToPhenotype(simBatchDir, selectedSim, outDir4, figHandle, [0.1 0.1 0.8 0.8]);
<a name="l00102"></a>00102             print(figHandle, [outDir3 &#39;geneToPhenotype.pdf&#39;], &#39;-dpdf&#39;, &#39;-cmyk&#39;);
<a name="l00103"></a>00103             close(figHandle);
<a name="l00104"></a>00104 
<a name="l00105"></a>00105             %metabolic map
<a name="l00106"></a>00106             [~, figHandle] = PlotUtil.newAxesHandle();
<a name="l00107"></a>00107             Figures34.metabolicMap(simBatchDir, selectedSim, outDir3, figHandle, [0.1 0.2 0.8 0.6], true);
<a name="l00108"></a>00108             print(figHandle, [outDir3 &#39;metabolicMap.pdf&#39;], &#39;-dpdf&#39;, &#39;-cmyk&#39;);
<a name="l00109"></a>00109             close(figHandle);
<a name="l00110"></a>00110             
<a name="l00111"></a>00111             %gene expression
<a name="l00112"></a>00112             w = 8; h = w * (1-0.09)/(1-0.075);
<a name="l00113"></a>00113             [~, figHandle] = PlotUtil.newAxesHandle();            
<a name="l00114"></a>00114             set(figHandle, &#39;PaperUnits&#39;, &#39;centimeters&#39;);
<a name="l00115"></a>00115             set(figHandle, &#39;PaperSize&#39;, [w h]);
<a name="l00116"></a>00116             set(figHandle, &#39;PaperPositionMode&#39;, &#39;manual&#39;);
<a name="l00117"></a>00117             set(figHandle, &#39;PaperPosition&#39;, [0 0 get(figHandle, &#39;PaperSize&#39;)]);
<a name="l00118"></a>00118             set(figHandle, &#39;Units&#39;, &#39;normalized&#39;);
<a name="l00119"></a>00119             figPos = get(figHandle, &#39;Position&#39;);
<a name="l00120"></a>00120             set(figHandle, &#39;Position&#39;, [figPos(1:2) figPos(3) figPos(3) * h / w]);
<a name="l00121"></a>00121 
<a name="l00122"></a>00122             clf(figHandle);
<a name="l00123"></a>00123             [exptGeneExpr, modelGeneExpr] = Figures34.geneExpression(simBatchDir, selectedSim, figHandle, [0.09 0.075 1-0.09 1-0.075], true);
<a name="l00124"></a>00124             print(figHandle, [outDirectory &#39;../supplement/figures/rnaExpressionComparison_log.pdf&#39;], &#39;-dpdf&#39;, &#39;-cmyk&#39;);
<a name="l00125"></a>00125             
<a name="l00126"></a>00126             clf(figHandle);
<a name="l00127"></a>00127             [exptGeneExpr, modelGeneExpr] = Figures34.geneExpression(simBatchDir, selectedSim, figHandle, [0.09 0.075 1-0.09 1-0.075], false);
<a name="l00128"></a>00128             print(figHandle, [outDirectory &#39;../supplement/figures/rnaExpressionComparison_linear.pdf&#39;], &#39;-dpdf&#39;, &#39;-cmyk&#39;);
<a name="l00129"></a>00129             
<a name="l00130"></a>00130             close(figHandle);
<a name="l00131"></a>00131             
<a name="l00132"></a>00132             %gene expression ratios
<a name="l00133"></a>00133             [axesHandle, figHandle] = edu.stanford.covert.cell.sim.util.PlotUtil.newAxesHandle();
<a name="l00134"></a>00134             Figures34.geneExpressionRatios(exptGeneExpr, modelGeneExpr, axesHandle);
<a name="l00135"></a>00135             print(figHandle, [outDir3 &#39;RNAExprRatios.pdf&#39;], &#39;-dpdf&#39;, &#39;-cmyk&#39;);
<a name="l00136"></a>00136             close(figHandle);
<a name="l00137"></a>00137         end
<a name="l00138"></a>00138         
<a name="l00139"></a>00139         function figure4(simBatchDir, selectedSim, outDirectory)
<a name="l00140"></a>00140             import edu.stanford.covert.cell.sim.analysis.Figures34;
<a name="l00141"></a>00141             import edu.stanford.covert.cell.sim.util.PlotUtil;
<a name="l00142"></a>00142             
<a name="l00143"></a>00143             outDir4 = [outDirectory filesep &#39;figure4&#39; filesep];
<a name="l00144"></a>00144             
<a name="l00145"></a>00145             w = 13.6; h = 18; %Nature 1.5 column
<a name="l00146"></a>00146             h2 = 0.2025;
<a name="l00147"></a>00147             [~, figHandle] = PlotUtil.newAxesHandle();
<a name="l00148"></a>00148             clf(figHandle);
<a name="l00149"></a>00149             set(figHandle, &#39;PaperUnits&#39;, &#39;centimeters&#39;);
<a name="l00150"></a>00150             set(figHandle, &#39;PaperSize&#39;, [w h]);
<a name="l00151"></a>00151             set(figHandle, &#39;PaperPositionMode&#39;, &#39;manual&#39;);
<a name="l00152"></a>00152             set(figHandle, &#39;PaperPosition&#39;, [0 0 get(figHandle, &#39;PaperSize&#39;)]);
<a name="l00153"></a>00153             set(figHandle, &#39;Units&#39;, &#39;normalized&#39;);
<a name="l00154"></a>00154             
<a name="l00155"></a>00155             %A: DNA-bound protein density
<a name="l00156"></a>00156             PlotUtil.labelSubFigure(&#39;a&#39;, [0.075 1.0 -0.0286 -0.0208], figHandle);
<a name="l00157"></a>00157             Figures34.circosPlot(simBatchDir, selectedSim, outDir4, figHandle, [-0.02 0.69 0.55 0.26]);
<a name="l00158"></a>00158             
<a name="l00159"></a>00159             %B: time to DNA occupancy
<a name="l00160"></a>00160             PlotUtil.labelSubFigure(&#39;b&#39;, [0.60 1.0 -0.0286 -0.0208], figHandle);
<a name="l00161"></a>00161             Figures34.chromosomeSampling(simBatchDir, selectedSim, outDir4, figHandle, [0.61 0.88 0.38 0.11]);
<a name="l00162"></a>00162             
<a name="l00163"></a>00163             %C: time to RNA expression
<a name="l00164"></a>00164             PlotUtil.labelSubFigure(&#39;c&#39;, [0.60 0.84 -0.0286 -0.0208], figHandle);
<a name="l00165"></a>00165             Figures34.rnaSampling(simBatchDir, selectedSim, outDir4, figHandle, [0.61 0.71 0.38 0.11]);
<a name="l00166"></a>00166             
<a name="l00167"></a>00167             %D: space-time plot of DNA-bound protein
<a name="l00168"></a>00168             PlotUtil.labelSubFigure(&#39;d&#39;, [0.075 0.69 -0.0286 -0.0208], figHandle);
<a name="l00169"></a>00169             Figures34.spaceTimePlot(simBatchDir, selectedSim, outDir4, figHandle, [0.05 0.40 0.94 0.26]);
<a name="l00170"></a>00170             
<a name="l00171"></a>00171             %E: dynamics of cell mass fractions
<a name="l00172"></a>00172             PlotUtil.labelSubFigure(&#39;e&#39;, [0.075 0.36 -0.0286 -0.0208], figHandle);
<a name="l00173"></a>00173             Figures34.normCellMass(simBatchDir, selectedSim, outDir4, figHandle, [0.05 0.16 0.27 0.18]);
<a name="l00174"></a>00174             
<a name="l00175"></a>00175             %F: histograms of cell cycle phase lengths
<a name="l00176"></a>00176             PlotUtil.labelSubFigure(&#39;f&#39;, [0.405 0.36 -0.0286 -0.0208], figHandle, 8, &#39;normalized&#39;);
<a name="l00177"></a>00177             Figures34.cellCycleLength(simBatchDir, selectedSim, figHandle, [0.38 0.16 0.27 0.18]);
<a name="l00178"></a>00178             
<a name="l00179"></a>00179             %G: scatter plot of replication vs. replication initiation durations
<a name="l00180"></a>00180             PlotUtil.labelSubFigure(&#39;g&#39;, [0.735 0.36 -0.0286 -0.0208], figHandle, 8, &#39;normalized&#39;);
<a name="l00181"></a>00181             Figures34.replicationInitiationVsReplicationDuration(simBatchDir, selectedSim, figHandle, [0.71 0.16 0.27 0.18]);
<a name="l00182"></a>00182                        
<a name="l00183"></a>00183             %standarize tick lengths
<a name="l00184"></a>00184             axesHandles = findobj(figHandle, &#39;type&#39;, &#39;axes&#39;, &#39;-and&#39;, &#39;tag&#39;, &#39;&#39;);
<a name="l00185"></a>00185             axesPos = cell2mat(get(axesHandles, &#39;position&#39;));
<a name="l00186"></a>00186             tickLen = 0.0075 * max(max(axesPos(:, 3:4), [], 1) .* [w h]);
<a name="l00187"></a>00187             for i = 1:numel(axesHandles)
<a name="l00188"></a>00188                 maxVal = max(axesPos(i, 3:4) .* [w h]);
<a name="l00189"></a>00189                 set(axesHandles(i), &#39;TickLen&#39;, [tickLen / maxVal  0.0250]);
<a name="l00190"></a>00190             end
<a name="l00191"></a>00191             
<a name="l00192"></a>00192             %save
<a name="l00193"></a>00193             print(figHandle, [outDirectory filesep &#39;figure4.pdf&#39;], &#39;-dpdf&#39;, &#39;-cmyk&#39;);
<a name="l00194"></a>00194             close(figHandle);
<a name="l00195"></a>00195             
<a name="l00196"></a>00196             %% supplement
<a name="l00197"></a>00197             %fit 2C
<a name="l00198"></a>00198             [~, figHandle] = edu.stanford.covert.cell.sim.util.PlotUtil.newAxesHandle();
<a name="l00199"></a>00199             clf(figHandle);
<a name="l00200"></a>00200             Figures34.replicationInitiationVsReplicationDurationFit(simBatchDir, selectedSim, figHandle, [0.1 0.1 0.8 0.8]);
<a name="l00201"></a>00201             print(figHandle, [outDir4 filesep &#39;ReplicationInitiationVsReplicationDurationFit.pdf&#39;], &#39;-dpdf&#39;, &#39;-cmyk&#39;);
<a name="l00202"></a>00202             close(figHandle);
<a name="l00203"></a>00203             
<a name="l00204"></a>00204             %analysis of RNA polymerase exploration
<a name="l00205"></a>00205             [~, figHandle] = edu.stanford.covert.cell.sim.util.PlotUtil.newAxesHandle();
<a name="l00206"></a>00206             clf(figHandle);
<a name="l00207"></a>00207             Figures34.plotRNAPolSampling(simBatchDir, selectedSim, outDir4, figHandle, [0.1 0.1 0.8 0.8]);
<a name="l00208"></a>00208             print(figHandle, [outDir4 filesep &#39;RNAPolSampling.pdf&#39;], &#39;-dpdf&#39;, &#39;-cmyk&#39;);
<a name="l00209"></a>00209             close(figHandle);
<a name="l00210"></a>00210             return
<a name="l00211"></a>00211             
<a name="l00212"></a>00212             %F: scatter plot of density vs collisions
<a name="l00213"></a>00213             PlotUtil.labelSubFigure(&#39;F&#39;, [0.715 0.80 -0.0286 -0.0690], figHandle);
<a name="l00214"></a>00214             Figures34.dnaBoundProteinDensityVsDisplacement(simBatchDir, selectedSim, outDir4, figHandle, [0.71 0.755-h2 0.29 h2]);                        
<a name="l00215"></a>00215             
<a name="l00216"></a>00216             %more supplemental material
<a name="l00217"></a>00217             Figures34.analyzeDNABoundProteinDisplacements(simBatchDir, selectedSim, outDir4);
<a name="l00218"></a>00218             Figures34.cellShape(outDir4, simBatchDir, selectedSim);            
<a name="l00219"></a>00219             Figures34.mrnaSampling(simBatchDir, selectedSim, outDir4, figHandle, [0.71 0.755-0.12 0.29 0.12]);
<a name="l00220"></a>00220         end
<a name="l00221"></a>00221         
<a name="l00222"></a>00222         function RNAvProtein(simBatchDir, outDirectory, figHandle, position)
<a name="l00223"></a>00223             import edu.stanford.covert.cell.sim.util.SimulationEnsemble;
<a name="l00224"></a>00224             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
<a name="l00225"></a>00225             import edu.stanford.covert.cell.sim.util.PlotUtil;
<a name="l00226"></a>00226             import edu.stanford.covert.cell.sim.util.MacromoleculeUtil;
<a name="l00227"></a>00227             
<a name="l00228"></a>00228             if ~exist(&#39;simBatchDir&#39;, &#39;var&#39;) || isempty(simBatchDir)
<a name="l00229"></a>00229                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
<a name="l00230"></a>00230             end
<a name="l00231"></a>00231             selectedSimulations = 1:SimulationDiskUtil.getNumSimulations(simBatchDir);
<a name="l00232"></a>00232             
<a name="l00233"></a>00233             %% get data
<a name="l00234"></a>00234             if exist([outDirectory &#39;RNAvProtein.mat&#39;], &#39;file&#39;)
<a name="l00235"></a>00235                 load([outDirectory &#39;RNAvProtein.mat&#39;]);
<a name="l00236"></a>00236             else
<a name="l00237"></a>00237                 [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep &#39;1&#39;]);
<a name="l00238"></a>00238                 g = sim.gene;
<a name="l00239"></a>00239                 pm = sim.state(&#39;<a class="code" href="class_protein_monomer.html">ProteinMonomer</a>&#39;);
<a name="l00240"></a>00240                 pc = sim.state(&#39;<a class="code" href="class_protein_complex.html">ProteinComplex</a>&#39;);
<a name="l00241"></a>00241                 rna = sim.state(&#39;Rna&#39;);
<a name="l00242"></a>00242                 
<a name="l00243"></a>00243                 geneId = &#39;MG_218&#39;;
<a name="l00244"></a>00244                 [~, ~, ~, ~, ~, ...
<a name="l00245"></a>00245                     ~, ~, ~, ~, ...
<a name="l00246"></a>00246                     ~, ~, ~, ~, ...
<a name="l00247"></a>00247                     matureRnaIdxs, ~, ~, ~, ...
<a name="l00248"></a>00248                     monomerIdxs, ~, ~, ~, ...
<a name="l00249"></a>00249                     ~, ~, ~, ~] = ...
<a name="l00250"></a>00250                     MacromoleculeUtil.getMacromoleculeIndexsIDsNames(geneId, sim);
<a name="l00251"></a>00251                 
<a name="l00252"></a>00252                 stateNames = {
<a name="l00253"></a>00253                     <span class="stringliteral">&#39;Rna&#39;</span>               <span class="stringliteral">&#39;counts&#39;</span> rna.matureIndexs(matureRnaIdxs)         <span class="stringliteral">&#39;-sum&#39;</span>
<a name="l00254"></a>00254                     <span class="stringliteral">&#39;ProteinMonomer&#39;</span>    <span class="stringliteral">&#39;counts&#39;</span> pm.matureIndexs(monomerIdxs)            <span class="stringliteral">&#39;-sum&#39;</span>
<a name="l00255"></a>00255                     };
<a name="l00256"></a>00256                 cr = zeros(numel(selectedSimulations), 1);
<a name="l00257"></a>00257                 cp = zeros(numel(selectedSimulations), 1);
<a name="l00258"></a>00258                 <span class="keywordflow">for</span> i = 1:numel(selectedSimulations)
<a name="l00259"></a>00259                     states = SimulationEnsemble.load(simBatchDir, {<span class="stringliteral">&#39;Time&#39;</span> <span class="stringliteral">&#39;values&#39;</span>}, [], [], 1, <span class="stringliteral">&#39;extract&#39;</span>, selectedSimulations(i));
<a name="l00260"></a>00260                     
<a name="l00261"></a>00261                     tIdx = randi(numel(states.Time.values), 1);
<a name="l00262"></a>00262                     states = SimulationEnsemble.load(simBatchDir, stateNames, tIdx, tIdx, 1, <span class="stringliteral">&#39;extract&#39;</span>, selectedSimulations(i));
<a name="l00263"></a>00263                     cr(i) = states.Rna.counts;
<a name="l00264"></a>00264                     cp(i) = states.ProteinMonomer.counts;
<a name="l00265"></a>00265                 end
<a name="l00266"></a>00266                 
<a name="l00267"></a>00267                 save([outDirectory <span class="stringliteral">&#39;RNAvProtein.mat&#39;</span>], <span class="stringliteral">&#39;cp&#39;</span>, <span class="stringliteral">&#39;cr&#39;</span>);
<a name="l00268"></a>00268             end
<a name="l00269"></a>00269             
<a name="l00270"></a>00270             %% plot
<a name="l00271"></a>00271             <span class="keywordflow">if</span> nargin &lt; 4
<a name="l00272"></a>00272                 [~, figHandle] = edu.stanford.covert.cell.sim.util.PlotUtil.newAxesHandle();
<a name="l00273"></a>00273                 position = [0.1 0.1 0.8 0.8];
<a name="l00274"></a>00274             end
<a name="l00275"></a>00275             x = position(1);
<a name="l00276"></a>00276             y = position(2);
<a name="l00277"></a>00277             w = position(3);
<a name="l00278"></a>00278             h = position(4);
<a name="l00279"></a>00279             figPos = <span class="keyword">get</span>(figHandle, <span class="stringliteral">&#39;papersize&#39;</span>);
<a name="l00280"></a>00280             figW = figPos(1);
<a name="l00281"></a>00281             figH = figPos(2);
<a name="l00282"></a>00282             axesHandles = [
<a name="l00283"></a>00283                 subplot(&#39;Position&#39;, [x, y + h-0.15*w*figW/figH, 0.8*w, 0.15*w*figW/figH], &#39;Parent&#39;, figHandle)
<a name="l00284"></a>00284                 subplot(&#39;Position&#39;, [x + 0.85*w, y, 0.15*w, h-0.2*w*figW/figH], &#39;Parent&#39;, figHandle)
<a name="l00285"></a>00285                 subplot(&#39;Position&#39;, [x, y, 0.8*w, h-0.2*w*figW/figH], &#39;Parent&#39;, figHandle)
<a name="l00286"></a>00286                 ];
<a name="l00287"></a>00287             
<a name="l00288"></a>00288             xlims = [min(cr) - 0.5, max(cr) + 0.5];
<a name="l00289"></a>00289             ylims = [0 60];
<a name="l00290"></a>00290             
<a name="l00291"></a>00291             %<a class="code" href="class_r_n_a.html">RNA</a> marginal
<a name="l00292"></a>00292             parmhat = gamfit(cr);
<a name="l00293"></a>00293             
<a name="l00294"></a>00294             axesHandle = axesHandles(1);
<a name="l00295"></a>00295             hold(axesHandle, <span class="stringliteral">&#39;on&#39;</span>);
<a name="l00296"></a>00296             [n, edges] = hist(axesHandle, cr, min(cr):max(cr));
<a name="l00297"></a>00297             n = n / sum(n) * 100;
<a name="l00298"></a>00298             bar(axesHandle, edges, n, <span class="charliteral">&#39;b&#39;</span>, <span class="stringliteral">&#39;EdgeColor&#39;</span>, <span class="stringliteral">&#39;none&#39;</span>);
<a name="l00299"></a>00299             x = linspace(min(cr), max(cr), 20);
<a name="l00300"></a>00300             %plot(axesHandle, x, gampdf(x, parmhat(1), parmhat(2)) * 100, <span class="charliteral">&#39;r&#39;</span>);
<a name="l00301"></a>00301             ylim(axesHandle, [0 max(n)]);
<a name="l00302"></a>00302             xlim(axesHandle, xlims);
<a name="l00303"></a>00303             <span class="keyword">set</span>(axesHandle, <span class="stringliteral">&#39;Box&#39;</span>, <span class="stringliteral">&#39;off&#39;</span>);
<a name="l00304"></a>00304             <span class="keyword">set</span>(axesHandle, <span class="stringliteral">&#39;TickDir&#39;</span>, <span class="stringliteral">&#39;out&#39;</span>);
<a name="l00305"></a>00305             <span class="keyword">set</span>(axesHandle, <span class="stringliteral">&#39;XTickLabel&#39;</span>, <span class="stringliteral">&#39;&#39;</span>);
<a name="l00306"></a>00306             <span class="keyword">set</span>(axesHandle, <span class="stringliteral">&#39;XTick&#39;</span>, []);
<a name="l00307"></a>00307             <span class="keyword">set</span>(axesHandle, <span class="stringliteral">&#39;YTick&#39;</span>, 0:25:50);
<a name="l00308"></a>00308                         
<a name="l00309"></a>00309             tick2text(axesHandle, <span class="stringliteral">&#39;axis&#39;</span>, <span class="charliteral">&#39;y&#39;</span>, <span class="stringliteral">&#39;ytickoffset&#39;</span>, 0.05);
<a name="l00310"></a>00310             yTicks = getappdata(axesHandle, <span class="stringliteral">&#39;YTickText&#39;</span>);
<a name="l00311"></a>00311             <span class="keyword">set</span>(yTicks, <span class="stringliteral">&#39;FontSize&#39;</span>, 5, <span class="stringliteral">&#39;HorizontalAlign&#39;</span>, <span class="stringliteral">&#39;right&#39;</span>);
<a name="l00312"></a>00312             <span class="keyword">delete</span>(yTicks(2));
<a name="l00313"></a>00313             
<a name="l00314"></a>00314             ylabel(axesHandle, <span class="stringliteral">&#39;Freq&#39;</span>, <span class="stringliteral">&#39;FontSize&#39;</span>, 7);
<a name="l00315"></a>00315             ylabelPos = <span class="keyword">get</span>(<span class="keyword">get</span>(axesHandle, <span class="stringliteral">&#39;ylabel&#39;</span>), <span class="stringliteral">&#39;position&#39;</span>);
<a name="l00316"></a>00316             <span class="keyword">set</span>(<span class="keyword">get</span>(axesHandle, <span class="stringliteral">&#39;ylabel&#39;</span>), <span class="stringliteral">&#39;position&#39;</span>, [-0.87 ylabelPos(2:end)]);
<a name="l00317"></a>00317             
<a name="l00318"></a>00318             %protein marginal
<a name="l00319"></a>00319             parmhat = gamfit(cp);
<a name="l00320"></a>00320             
<a name="l00321"></a>00321             axesHandle = axesHandles(2);
<a name="l00322"></a>00322             hold(axesHandle, <span class="stringliteral">&#39;on&#39;</span>);
<a name="l00323"></a>00323             [n, xout] = hist(cp, 50);
<a name="l00324"></a>00324             n = n / sum(n) * 100;
<a name="l00325"></a>00325             barh(axesHandle, xout, n, <span class="charliteral">&#39;b&#39;</span>, <span class="stringliteral">&#39;EdgeColor&#39;</span>, <span class="stringliteral">&#39;none&#39;</span>);
<a name="l00326"></a>00326             x = linspace(0, 60, 100);
<a name="l00327"></a>00327             %plot(axesHandle, gampdf(x, parmhat(1), parmhat(2)) * 100, x, <span class="charliteral">&#39;r&#39;</span>);
<a name="l00328"></a>00328             xlim(axesHandle, [0 max(n)]);
<a name="l00329"></a>00329             ylim(axesHandle, ylims);
<a name="l00330"></a>00330             <span class="keyword">set</span>(axesHandle, <span class="stringliteral">&#39;Box&#39;</span>, <span class="stringliteral">&#39;off&#39;</span>);
<a name="l00331"></a>00331             <span class="keyword">set</span>(axesHandle, <span class="stringliteral">&#39;TickDir&#39;</span>, <span class="stringliteral">&#39;out&#39;</span>);
<a name="l00332"></a>00332             <span class="keyword">set</span>(axesHandle, <span class="stringliteral">&#39;YTickLabel&#39;</span>, []);
<a name="l00333"></a>00333             <span class="keyword">set</span>(axesHandle, <span class="stringliteral">&#39;YTick&#39;</span>, []);
<a name="l00334"></a>00334             <span class="keyword">set</span>(axesHandle, <span class="stringliteral">&#39;XTick&#39;</span>, 0:2:6);
<a name="l00335"></a>00335                         
<a name="l00336"></a>00336             tick2text(axesHandle, <span class="stringliteral">&#39;axis&#39;</span>, <span class="charliteral">&#39;x&#39;</span>, <span class="stringliteral">&#39;xtickoffset&#39;</span>, 0.09);
<a name="l00337"></a>00337             xTicks = getappdata(axesHandle, <span class="stringliteral">&#39;XTickText&#39;</span>);
<a name="l00338"></a>00338             <span class="keyword">set</span>(xTicks, <span class="stringliteral">&#39;FontSize&#39;</span>, 5);
<a name="l00339"></a>00339             <span class="keyword">delete</span>(xTicks(2:3));
<a name="l00340"></a>00340             
<a name="l00341"></a>00341             xlabel(axesHandle, <span class="stringliteral">&#39;Freq&#39;</span>, <span class="stringliteral">&#39;FontSize&#39;</span>, 7);
<a name="l00342"></a>00342             xlabelPos = <span class="keyword">get</span>(<span class="keyword">get</span>(axesHandle, <span class="stringliteral">&#39;xlabel&#39;</span>), <span class="stringliteral">&#39;position&#39;</span>);
<a name="l00343"></a>00343             <span class="keyword">set</span>(<span class="keyword">get</span>(axesHandle, <span class="stringliteral">&#39;xlabel&#39;</span>), <span class="stringliteral">&#39;position&#39;</span>, [xlabelPos(1)-0.3 -9 xlabelPos(3:end)]);
<a name="l00344"></a>00344             
<a name="l00345"></a>00345             %<a class="code" href="class_r_n_a.html">RNA</a> vs protein
<a name="l00346"></a>00346             axesHandle = axesHandles(3);
<a name="l00347"></a>00347             plot(axesHandle, cr, cp, <span class="stringliteral">&#39;b.&#39;</span>);
<a name="l00348"></a>00348             xlim(axesHandle, xlims);
<a name="l00349"></a>00349             ylim(axesHandle, ylims);
<a name="l00350"></a>00350             <span class="keyword">set</span>(axesHandle, <span class="stringliteral">&#39;XTick&#39;</span>, [0 1 2], <span class="stringliteral">&#39;YTick&#39;</span>, 0:25:50);
<a name="l00351"></a>00351             <span class="keyword">set</span>(axesHandle, <span class="stringliteral">&#39;TickDir&#39;</span>, <span class="stringliteral">&#39;out&#39;</span>, <span class="stringliteral">&#39;box&#39;</span>, <span class="stringliteral">&#39;off&#39;</span>);
<a name="l00352"></a>00352             
<a name="l00353"></a>00353             tick2text(axesHandle, <span class="stringliteral">&#39;axis&#39;</span>, <span class="stringliteral">&#39;xy&#39;</span>, <span class="stringliteral">&#39;xtickoffset&#39;</span>, 0.09, <span class="stringliteral">&#39;ytickoffset&#39;</span>, 0.05);
<a name="l00354"></a>00354             xTicks = getappdata(axesHandle, <span class="stringliteral">&#39;XTickText&#39;</span>);
<a name="l00355"></a>00355             yTicks = getappdata(axesHandle, <span class="stringliteral">&#39;YTickText&#39;</span>);
<a name="l00356"></a>00356             <span class="keyword">set</span>(xTicks, <span class="stringliteral">&#39;FontSize&#39;</span>, 5);
<a name="l00357"></a>00357             <span class="keyword">set</span>(yTicks, <span class="stringliteral">&#39;FontSize&#39;</span>, 5, <span class="stringliteral">&#39;HorizontalAlign&#39;</span>, <span class="stringliteral">&#39;right&#39;</span>);
<a name="l00358"></a>00358             <span class="keyword">delete</span>(yTicks(2));
<a name="l00359"></a>00359             
<a name="l00360"></a>00360             xlabel(axesHandle, <span class="stringliteral">&#39;mRNA count&#39;</span>, <span class="stringliteral">&#39;FontSize&#39;</span>, 7);
<a name="l00361"></a>00361             ylabel(axesHandle, <span class="stringliteral">&#39;Protein count&#39;</span>, <span class="stringliteral">&#39;FontSize&#39;</span>, 7);
<a name="l00362"></a>00362             xlabelPos = <span class="keyword">get</span>(<span class="keyword">get</span>(axesHandle, <span class="stringliteral">&#39;xlabel&#39;</span>), <span class="stringliteral">&#39;position&#39;</span>);
<a name="l00363"></a>00363             ylabelPos = <span class="keyword">get</span>(<span class="keyword">get</span>(axesHandle, <span class="stringliteral">&#39;ylabel&#39;</span>), <span class="stringliteral">&#39;position&#39;</span>);
<a name="l00364"></a>00364             <span class="keyword">set</span>(<span class="keyword">get</span>(axesHandle, <span class="stringliteral">&#39;xlabel&#39;</span>), <span class="stringliteral">&#39;position&#39;</span>, [xlabelPos(1) -9 xlabelPos(3:end)]);
<a name="l00365"></a>00365             <span class="keyword">set</span>(<span class="keyword">get</span>(axesHandle, <span class="stringliteral">&#39;ylabel&#39;</span>), <span class="stringliteral">&#39;position&#39;</span>, [-0.87 ylabelPos(2:end)]);
<a name="l00366"></a>00366         end
<a name="l00367"></a>00367         
<a name="l00368"></a>00368         <span class="keyword">function</span> cellShape(outDirectory, simBatchDir, selectedSim)
<a name="l00369"></a>00369             import edu.stanford.covert.cell.sim.util.SimulationEnsemble;
<a name="l00370"></a>00370             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
<a name="l00371"></a>00371             import edu.stanford.covert.cell.sim.analysis.<a class="code" href="class_figures34.html">Figures34</a>;
<a name="l00372"></a>00372             import edu.stanford.covert.cell.sim.state.CellGeometry;
<a name="l00373"></a>00373             
<a name="l00374"></a>00374             if ~exist(&#39;simBatchDir&#39;, &#39;var&#39;) || isempty(simBatchDir)
<a name="l00375"></a>00375                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
<a name="l00376"></a>00376             end
<a name="l00377"></a>00377             if ~exist(&#39;selectedSim&#39;, &#39;var&#39;) || isempty(selectedSim)
<a name="l00378"></a>00378                 selectedSim = 1:SimulationDiskUtil.getNumSimulations(simBatchDir);
<a name="l00379"></a>00379             end
<a name="l00380"></a>00380             
<a name="l00381"></a>00381             [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep &#39;1&#39;]);
<a name="l00382"></a>00382             
<a name="l00383"></a>00383             stateNames={<span class="stringliteral">&#39;Geometry&#39;</span> <span class="stringliteral">&#39;width&#39;</span>
<a name="l00384"></a>00384                 <span class="stringliteral">&#39;Geometry&#39;</span> <span class="stringliteral">&#39;pinchedDiameter&#39;</span>
<a name="l00385"></a>00385                 <span class="stringliteral">&#39;Chromosome&#39;</span> <span class="stringliteral">&#39;polymerizedRegions&#39;</span>
<a name="l00386"></a>00386                 <span class="stringliteral">&#39;Chromosome&#39;</span> <span class="stringliteral">&#39;monomerBoundSites&#39;</span>
<a name="l00387"></a>00387                 <span class="stringliteral">&#39;Chromosome&#39;</span> <span class="stringliteral">&#39;complexBoundSites&#39;</span>
<a name="l00388"></a>00388                 <span class="stringliteral">&#39;Time&#39;</span> <span class="stringliteral">&#39;values&#39;</span>};
<a name="l00389"></a>00389             
<a name="l00390"></a>00390             states = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, <span class="stringliteral">&#39;extract&#39;</span>, selectedSim);
<a name="l00391"></a>00391             
<a name="l00392"></a>00392             log = load([simBatchDir filesep num2str(selectedSim) filesep <span class="stringliteral">&#39;summary.mat&#39;</span>]);
<a name="l00393"></a>00393             
<a name="l00394"></a>00394             cm = sim.state(<span class="stringliteral">&#39;Mass&#39;</span>);
<a name="l00395"></a>00395             geom = sim.state(<span class="stringliteral">&#39;Geometry&#39;</span>);
<a name="l00396"></a>00396             density = geom.density;
<a name="l00397"></a>00397             mass = (log.mass(2:end) * cm.cellInitialDryWeight)/(1-cm.fractionWetWeight);
<a name="l00398"></a>00398             width = squeeze(states.Geometry.width);
<a name="l00399"></a>00399             pinchedDiameter = squeeze(states.Geometry.pinchedDiameter);
<a name="l00400"></a>00400             t = squeeze(states.Time.values);
<a name="l00401"></a>00401             septumLength = (width - pinchedDiameter)/2;
<a name="l00402"></a>00402             
<a name="l00403"></a>00403             cylindricalLength = zeros(size(t));
<a name="l00404"></a>00404             maxTotalLengths = zeros(size(t));
<a name="l00405"></a>00405             
<a name="l00406"></a>00406             <span class="keywordflow">for</span> i=1:length(cylindricalLength)
<a name="l00407"></a>00407                 if(pinchedDiameter(i) &gt; 0)
<a name="l00408"></a>00408                     [cylindricalLength(i), ~, maxTotalLengths(i)] = CellGeometry.calculateGeometry(...
<a name="l00409"></a>00409                         width(i), ...
<a name="l00410"></a>00410                         pinchedDiameter(i), ...
<a name="l00411"></a>00411                         mass(i)/density);
<a name="l00412"></a>00412                 end
<a name="l00413"></a>00413             end
<a name="l00414"></a>00414             
<a name="l00415"></a>00415             tidxs = [round(1:t(end)/4:t(end)) t(end)-1];
<a name="l00416"></a>00416             
<a name="l00417"></a>00417             scale = 1<a class="code" href="json_parse_8m.html#ac14d26badced2970320ee4d772ad0ae9">e</a>-9;
<a name="l00418"></a>00418             
<a name="l00419"></a>00419             for i = 1:numel(tidxs)
<a name="l00420"></a>00420                 
<a name="l00421"></a>00421                 svg_width = maxTotalLengths(tidxs(i)) / scale + 10;
<a name="l00422"></a>00422                 svg_height = max(width) / scale + 10;
<a name="l00423"></a>00423 
<a name="l00424"></a>00424                 fileNameSVG = sprintf(&#39;%s%sCellShape_%d.svg&#39;, outDirectory, filesep, i);
<a name="l00425"></a>00425                 fileNamePDF = sprintf(&#39;%s%sCellShape_%d.pdf&#39;, outDirectory, filesep, i);
<a name="l00426"></a>00426                 
<a name="l00427"></a>00427                 %open svg
<a name="l00428"></a>00428                 fid = fopen(fileNameSVG, &#39;w&#39;);
<a name="l00429"></a>00429                 fprintf(fid, &#39;&lt;?xml version=&quot;1.0&quot; standalone=&quot;no&quot;?&gt;\n&#39;);
<a name="l00430"></a>00430                 fprintf(fid, &#39;&lt;!DOCTYPE svg PUBLIC &quot;-<span class="comment">//W3C//DTD SVG 1.1//EN&quot; &quot;http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd&quot;&gt;\n&#39;);</span>
<a name="l00431"></a>00431                 fprintf(fid, &#39;&lt;svg width=&quot;%d&quot; height=&quot;%d&quot; version=&quot;1.1&quot; xmlns=&quot;http:<span class="comment">//www.w3.org/2000/svg&quot; viewBox=&quot;%d %d %d %d&quot;&gt;\n&#39;, svg_width, svg_height, 0, 0, svg_width, svg_height);</span>
<a name="l00432"></a>00432                 fprintf(fid, &#39;&lt;defs&gt;\n&#39;);
<a name="l00433"></a>00433                 fprintf(fid, &#39;&lt;linearGradient <span class="keywordtype">id</span>=&quot;cellBg&quot; x1=&quot;0%%&quot; y1=&quot;0%%&quot; x2=&quot;0%%&quot; y2=&quot;100%%&quot;&gt;\n&#39;);
<a name="l00434"></a>00434                 fprintf(fid, &#39;&lt;stop style=&quot;stop-color:<span class="preprocessor">#97b0fb;stop-opacity:0.05&quot; offset=&quot;0&quot;/&gt;\n&#39;);</span>
<a name="l00435"></a>00435 <span class="preprocessor"></span>                fprintf(fid, <span class="stringliteral">&#39;&lt;stop style=&quot;stop-color:#89b8d6;stop-opacity:1&quot; offset=&quot;1&quot;/&gt;\n&#39;</span>);
<a name="l00436"></a>00436                 fprintf(fid, <span class="stringliteral">&#39;&lt;/linearGradient&gt;&#39;</span>);
<a name="l00437"></a>00437                 fprintf(fid, <span class="stringliteral">&#39;&lt;/defs&gt;\n&#39;</span>);
<a name="l00438"></a>00438                 
<a name="l00439"></a>00439                 % White background
<a name="l00440"></a>00440                 fprintf(fid, <span class="stringliteral">&#39;&lt;rect x=&quot;%d&quot; y=&quot;%d&quot; width=&quot;%d&quot; height=&quot;%d&quot; style=&quot;fill:white; stroke:none&quot;/&gt;\n&#39;</span>, ...
<a name="l00441"></a>00441                     0, 0, svg_width, svg_height);
<a name="l00442"></a>00442                 
<a name="l00443"></a>00443                 fprintf(fid, <a class="code" href="class_figures34.html">Figures34</a>.<a class="code" href="class_figures34.html#a6f467b303a884c0ac1dcea668a10bba8">drawCell</a>(scale,width(tidxs(i)),cylindricalLength(tidxs(i)),septumLength(tidxs(i)),max(maxTotalLengths),max(width)));
<a name="l00444"></a>00444                 
<a name="l00445"></a>00445                 %close svg
<a name="l00446"></a>00446                 fprintf(fid, <span class="stringliteral">&#39;&lt;/svg&gt;\n&#39;</span>);
<a name="l00447"></a>00447                 fclose(fid);
<a name="l00448"></a>00448                 
<a name="l00449"></a>00449                 [status, result] = system(sprintf(<span class="stringliteral">&#39;inkscape &quot;%s&quot; --export-pdf=&quot;%s&quot;&#39;</span>, fileNameSVG, fileNamePDF));
<a name="l00450"></a>00450                 <span class="keywordflow">if</span> status ~= 0
<a name="l00451"></a>00451                     <span class="keywordflow">throw</span>(MException(<span class="stringliteral">&#39;Figures34:error&#39;</span>, <span class="stringliteral">&#39;Failed to convert svg to pdf: %s&#39;</span>, result));
<a name="l00452"></a>00452                 end            
<a name="l00453"></a>00453             end
<a name="l00454"></a>00454         end
<a name="l00455"></a>00455         
<a name="l00456"></a>00456         <span class="keyword">function</span> normCellMass(simBatchDir, selectedSim, outDirectory, figHandle, position)
<a name="l00457"></a>00457             import edu.stanford.covert.cell.sim.util.SimulationEnsemble;
<a name="l00458"></a>00458             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
<a name="l00459"></a>00459             
<a name="l00460"></a>00460             if ~exist(&#39;simBatchDir&#39;, &#39;var&#39;) || isempty(simBatchDir)
<a name="l00461"></a>00461                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
<a name="l00462"></a>00462             end
<a name="l00463"></a>00463             if ~exist(&#39;selectedSim&#39;, &#39;var&#39;) || isempty(selectedSim)
<a name="l00464"></a>00464                 selectedSim = 1:SimulationDiskUtil.getNumSimulations(simBatchDir);
<a name="l00465"></a>00465             end
<a name="l00466"></a>00466             
<a name="l00467"></a>00467             %% get constants
<a name="l00468"></a>00468             [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep &#39;1&#39;]);
<a name="l00469"></a>00469             comp = sim.compartment;
<a name="l00470"></a>00470             
<a name="l00471"></a>00471             %% get data
<a name="l00472"></a>00472             if exist([outDirectory &#39;normCellMass.mat&#39;], &#39;file&#39;)
<a name="l00473"></a>00473                 load([outDirectory &#39;normCellMass.mat&#39;])
<a name="l00474"></a>00474             else
<a name="l00475"></a>00475                 stateNames = {
<a name="l00476"></a>00476                     <span class="stringliteral">&#39;Time&#39;</span> <span class="stringliteral">&#39;values&#39;</span>       <span class="charliteral">&#39;:&#39;</span> <span class="charliteral">&#39;:&#39;</span>
<a name="l00477"></a>00477                     <span class="stringliteral">&#39;Mass&#39;</span> <span class="stringliteral">&#39;dnaWt&#39;</span>        <span class="charliteral">&#39;:&#39;</span> <span class="stringliteral">&#39;-sum&#39;</span>
<a name="l00478"></a>00478                     <span class="stringliteral">&#39;Mass&#39;</span> <span class="stringliteral">&#39;rnaWt&#39;</span>        <span class="charliteral">&#39;:&#39;</span> <span class="stringliteral">&#39;-sum&#39;</span>
<a name="l00479"></a>00479                     <span class="stringliteral">&#39;Mass&#39;</span> <span class="stringliteral">&#39;proteinWt&#39;</span>    <span class="charliteral">&#39;:&#39;</span> <span class="stringliteral">&#39;-sum&#39;</span>
<a name="l00480"></a>00480                     <span class="stringliteral">&#39;Mass&#39;</span> <span class="stringliteral">&#39;metaboliteWt&#39;</span> <span class="charliteral">&#39;:&#39;</span> [comp.cytosolIndexs; comp.membraneIndexs]
<a name="l00481"></a>00481                     <span class="stringliteral">&#39;Mass&#39;</span> <span class="stringliteral">&#39;cellDry&#39;</span>         <span class="charliteral">&#39;:&#39;</span> <span class="stringliteral">&#39;-sum&#39;</span>
<a name="l00482"></a>00482                     };
<a name="l00483"></a>00483                 states = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, <span class="stringliteral">&#39;extract&#39;</span>, selectedSim);
<a name="l00484"></a>00484                 
<a name="l00485"></a>00485                 proteinWt = permute(states.Mass.proteinWt, [1 3 2]);
<a name="l00486"></a>00486                 dnaWt = permute(states.Mass.dnaWt, [1 3 2]);
<a name="l00487"></a>00487                 membraneWt = permute(states.Mass.metaboliteWt(:, 2, :), [1 3 2]);
<a name="l00488"></a>00488                 rnaWt = permute(states.Mass.rnaWt, [1 3 2]);
<a name="l00489"></a>00489                 cellWt = permute(states.Mass.cellDry, [1 3 2]);
<a name="l00490"></a>00490                 time = permute(states.Time.values, [1 3 2]) / 3600;
<a name="l00491"></a>00491                 
<a name="l00492"></a>00492                 proteinWt = proteinWt / proteinWt(1);
<a name="l00493"></a>00493                 dnaWt = dnaWt / dnaWt(1);
<a name="l00494"></a>00494                 membraneWt = membraneWt / membraneWt(1);
<a name="l00495"></a>00495                 rnaWt = rnaWt / rnaWt(1);
<a name="l00496"></a>00496                 cellWt = cellWt / cellWt(1);
<a name="l00497"></a>00497                 
<a name="l00498"></a>00498                 save([outDirectory <span class="stringliteral">&#39;normCellMass.mat&#39;</span>], <span class="stringliteral">&#39;proteinWt&#39;</span>, <span class="stringliteral">&#39;dnaWt&#39;</span>, <span class="stringliteral">&#39;membraneWt&#39;</span>, <span class="stringliteral">&#39;rnaWt&#39;</span>, <span class="stringliteral">&#39;cellWt&#39;</span>, <span class="stringliteral">&#39;time&#39;</span>);
<a name="l00499"></a>00499             end
<a name="l00500"></a>00500             
<a name="l00501"></a>00501             %% plot
<a name="l00502"></a>00502             <span class="keywordflow">if</span> nargin &gt;= 4
<a name="l00503"></a>00503                 axesHandle = subplot(<span class="stringliteral">&#39;position&#39;</span>, position, <span class="stringliteral">&#39;parent&#39;</span>, figHandle);
<a name="l00504"></a>00504             <span class="keywordflow">else</span>
<a name="l00505"></a>00505                 axesHandle = edu.stanford.covert.cell.sim.util.PlotUtil.newAxesHandle();
<a name="l00506"></a>00506             end
<a name="l00507"></a>00507             
<a name="l00508"></a>00508             lh = plot(axesHandle, time, [cellWt; dnaWt; rnaWt; proteinWt; membraneWt]);
<a name="l00509"></a>00509             <span class="keyword">set</span>(lh(2), <span class="stringliteral">&#39;LineWidth&#39;</span>, 1, <span class="stringliteral">&#39;Color&#39;</span>, <span class="charliteral">&#39;g&#39;</span>);
<a name="l00510"></a>00510             uistack(lh(4), <span class="stringliteral">&#39;top&#39;</span>);
<a name="l00511"></a>00511             uistack(lh(1), <span class="stringliteral">&#39;top&#39;</span>);
<a name="l00512"></a>00512             uistack(lh(2), <span class="stringliteral">&#39;bottom&#39;</span>);
<a name="l00513"></a>00513             uistack(lh(3), <span class="stringliteral">&#39;bottom&#39;</span>);
<a name="l00514"></a>00514             
<a name="l00515"></a>00515             xlim(axesHandle, [0 time(end)]);
<a name="l00516"></a>00516             ylim(axesHandle, [0.92 2.05]);
<a name="l00517"></a>00517             
<a name="l00518"></a>00518             <span class="keyword">set</span>(axesHandle, <span class="stringliteral">&#39;Box&#39;</span>, <span class="stringliteral">&#39;off&#39;</span>);
<a name="l00519"></a>00519             <span class="keyword">set</span>(axesHandle, <span class="stringliteral">&#39;TickDir&#39;</span>, <span class="stringliteral">&#39;out&#39;</span>);
<a name="l00520"></a>00520             <span class="keyword">set</span>(axesHandle, <span class="stringliteral">&#39;YTick&#39;</span>, 1:0.5:2);
<a name="l00521"></a>00521             <span class="keyword">set</span>(axesHandle, <span class="stringliteral">&#39;XTick&#39;</span>, 0:2:8);
<a name="l00522"></a>00522             <span class="keyword">set</span>(axesHandle, <span class="stringliteral">&#39;FontSize&#39;</span>, 5);
<a name="l00523"></a>00523             
<a name="l00524"></a>00524             tick2text(axesHandle, <span class="stringliteral">&#39;axis&#39;</span>, <span class="stringliteral">&#39;xy&#39;</span>, <span class="stringliteral">&#39;xtickoffset&#39;</span>, 0.06, <span class="stringliteral">&#39;ytickoffset&#39;</span>, 0.015);
<a name="l00525"></a>00525             xTicks = getappdata(axesHandle, <span class="stringliteral">&#39;XTickText&#39;</span>);
<a name="l00526"></a>00526             yTicks = getappdata(axesHandle, <span class="stringliteral">&#39;YTickText&#39;</span>);
<a name="l00527"></a>00527             <span class="keyword">set</span>(xTicks, <span class="stringliteral">&#39;FontSize&#39;</span>, 5);
<a name="l00528"></a>00528             <span class="keyword">set</span>(yTicks, <span class="stringliteral">&#39;FontSize&#39;</span>, 5, <span class="stringliteral">&#39;HorizontalAlign&#39;</span>, <span class="stringliteral">&#39;right&#39;</span>);
<a name="l00529"></a>00529             <span class="keyword">delete</span>(yTicks(2));
<a name="l00530"></a>00530             <span class="keyword">delete</span>(xTicks(2:2:end));
<a name="l00531"></a>00531             
<a name="l00532"></a>00532             xlabel(axesHandle, <span class="stringliteral">&#39;Time (h)&#39;</span>, <span class="stringliteral">&#39;FontSize&#39;</span>, 7);
<a name="l00533"></a>00533             ylabel(axesHandle, <span class="stringliteral">&#39;Mass (norm)&#39;</span>, <span class="stringliteral">&#39;FontSize&#39;</span>, 7);
<a name="l00534"></a>00534             xlabelPos = <span class="keyword">get</span>(<span class="keyword">get</span>(axesHandle, <span class="stringliteral">&#39;xlabel&#39;</span>), <span class="stringliteral">&#39;position&#39;</span>);
<a name="l00535"></a>00535             ylabelPos = <span class="keyword">get</span>(<span class="keyword">get</span>(axesHandle, <span class="stringliteral">&#39;ylabel&#39;</span>), <span class="stringliteral">&#39;position&#39;</span>);
<a name="l00536"></a>00536             <span class="keyword">set</span>(<span class="keyword">get</span>(axesHandle, <span class="stringliteral">&#39;xlabel&#39;</span>), <span class="stringliteral">&#39;position&#39;</span>, [xlabelPos(1) 0.82 xlabelPos(3:end)]);
<a name="l00537"></a>00537             <span class="keyword">set</span>(<span class="keyword">get</span>(axesHandle, <span class="stringliteral">&#39;ylabel&#39;</span>), <span class="stringliteral">&#39;position&#39;</span>, [-0.35 ylabelPos(2:end)]);
<a name="l00538"></a>00538             
<a name="l00539"></a>00539             legendHandle = legend(lh, {<span class="stringliteral">&#39;Total&#39;</span>, <span class="stringliteral">&#39;DNA&#39;</span>, <span class="stringliteral">&#39;RNA&#39;</span>, <span class="stringliteral">&#39;Protein&#39;</span>, <span class="stringliteral">&#39;Membrane&#39;</span>}, <span class="stringliteral">&#39;Location&#39;</span>, <span class="stringliteral">&#39;NorthWest&#39;</span>, <span class="stringliteral">&#39;FontSize&#39;</span>, 7);
<a name="l00540"></a>00540             legend(axesHandle, <span class="stringliteral">&#39;boxoff&#39;</span>);
<a name="l00541"></a>00541             legendPos = <span class="keyword">get</span>(legendHandle, <span class="stringliteral">&#39;Position&#39;</span>);
<a name="l00542"></a>00542             <span class="keyword">set</span>(legendHandle, <span class="stringliteral">&#39;Position&#39;</span>, [0.05 0.4 legendPos(3:end)])
<a name="l00543"></a>00543             <span class="keyword">set</span>(legendHandle, <span class="stringliteral">&#39;units&#39;</span>, <span class="stringliteral">&#39;pixels&#39;</span>);
<a name="l00544"></a>00544             legendPos = <span class="keyword">get</span>(legendHandle, <span class="stringliteral">&#39;Position&#39;</span>);
<a name="l00545"></a>00545             <span class="keyword">set</span>(legendHandle, <span class="stringliteral">&#39;Position&#39;</span>, [legendPos(1:3) 49]);
<a name="l00546"></a>00546             legendChildren = <span class="keyword">get</span>(legendHandle, <span class="stringliteral">&#39;children&#39;</span>);
<a name="l00547"></a>00547             <span class="keyword">set</span>(legendChildren(1:3:end), <span class="stringliteral">&#39;visible&#39;</span>, <span class="stringliteral">&#39;off&#39;</span>);
<a name="l00548"></a>00548             <span class="keywordflow">for</span> i = 1:3:numel(legendChildren)
<a name="l00549"></a>00549                 set(legendChildren(i+1), &#39;xdata&#39;, [0.03 0.10], &#39;LineWidth&#39;, 1)
<a name="l00550"></a>00550                 
<a name="l00551"></a>00551                 pos = get(legendChildren(i+2), &#39;Position&#39;);
<a name="l00552"></a>00552                 set(legendChildren(i+2), &#39;Position&#39;, [0.13 pos(2) 0])
<a name="l00553"></a>00553             end
<a name="l00554"></a>00554         end
<a name="l00555"></a>00555         
<a name="l00556"></a>00556         function populationMass(simBatchDir, selectedSim, figHandle, position, outDirectory)
<a name="l00557"></a>00557             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
<a name="l00558"></a>00558             import edu.stanford.covert.cell.sim.util.PlotUtil;
<a name="l00559"></a>00559             
<a name="l00560"></a>00560             if ~exist(&#39;simBatchDir&#39;, &#39;var&#39;) || isempty(simBatchDir)
<a name="l00561"></a>00561                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
<a name="l00562"></a>00562             end
<a name="l00563"></a>00563             
<a name="l00564"></a>00564             if exist([outDirectory &#39;populationMassDistribution.mat&#39;], &#39;file&#39;)
<a name="l00565"></a>00565                 load([outDirectory &#39;populationMassDistribution.mat&#39;]);
<a name="l00566"></a>00566             else
<a name="l00567"></a>00567                 data = load([SimulationDiskUtil.getSimulationBatchDir([simBatchDir filesep num2str(selectedSim)]) filesep &#39;population-MassDistribution.mat&#39;], &#39;mass&#39;, &#39;time&#39;);
<a name="l00568"></a>00568                 mass = data.mass;
<a name="l00569"></a>00569                 time = data.time;
<a name="l00570"></a>00570                 save([outDirectory &#39;populationMassDistribution.mat&#39;], &#39;mass&#39;, &#39;time&#39;);
<a name="l00571"></a>00571             end
<a name="l00572"></a>00572             
<a name="l00573"></a>00573             % Determine division times
<a name="l00574"></a>00574             simEndTimes = zeros(1, size(mass, 2));
<a name="l00575"></a>00575             for i = 1:size(mass, 2)
<a name="l00576"></a>00576                 et = find(isnan(mass(:, i)), 1);
<a name="l00577"></a>00577                 if isempty(et)
<a name="l00578"></a>00578                     et = NaN;
<a name="l00579"></a>00579                 end
<a name="l00580"></a>00580                 simEndTimes(i) = et - 1;
<a name="l00581"></a>00581             end
<a name="l00582"></a>00582             
<a name="l00583"></a>00583             finalMass = zeros(1, size(mass, 2));
<a name="l00584"></a>00584             for i = 1:size(finalMass, 2)
<a name="l00585"></a>00585                 if isnan(simEndTimes(i))
<a name="l00586"></a>00586                     finalMass(i) = mass(end, i);
<a name="l00587"></a>00587                 else
<a name="l00588"></a>00588                     finalMass(i) = mass(simEndTimes(i), i);
<a name="l00589"></a>00589                 end
<a name="l00590"></a>00590             end
<a name="l00591"></a>00591             
<a name="l00592"></a>00592             %markIdx = find(simEndTimes == floor(nanmedian(simEndTimes)));
<a name="l00593"></a>00593             markIdx = 10; %52;
<a name="l00594"></a>00594                         
<a name="l00595"></a>00595             %pick every other cell
<a name="l00596"></a>00596             simEndTimes = simEndTimes(2:2:end);
<a name="l00597"></a>00597             mass = mass(:, 2:2:end);
<a name="l00598"></a>00598             markIdx = markIdx / 2;
<a name="l00599"></a>00599             
<a name="l00600"></a>00600             %% plot
<a name="l00601"></a>00601             axesSizes = [50; 5];
<a name="l00602"></a>00602             
<a name="l00603"></a>00603             options = struct;
<a name="l00604"></a>00604             if nargin &gt;= 4
<a name="l00605"></a>00605                 options.position = position;
<a name="l00606"></a>00606             else
<a name="l00607"></a>00607                 [~, figHandle] = edu.stanford.covert.cell.sim.util.PlotUtil.newAxesHandle();
<a name="l00608"></a>00608                 clf(figHandle);
<a name="l00609"></a>00609             end
<a name="l00610"></a>00610             [axesHandles, xAxisHandle] = PlotUtil.multiElementPlot(figHandle, axesSizes, time([1 end])&#39;, options);
<a name="l00611"></a>00611             xlim(xAxisHandle, [0 max(time)]);
<a name="l00612"></a>00612             set(get(xAxisHandle, &#39;xlabel&#39;), &#39;FontSize&#39;, 7);
<a name="l00613"></a>00613             set(xAxisHandle, &#39;FontSize&#39;, 5);
<a name="l00614"></a>00614             set(xAxisHandle, &#39;XTick&#39;, 0:5:10);
<a name="l00615"></a>00615             tick2text(xAxisHandle, &#39;axis&#39;, &#39;x&#39;, &#39;xtickoffset&#39;, 65);
<a name="l00616"></a>00616             xTicks = getappdata(xAxisHandle, &#39;XTickText&#39;);
<a name="l00617"></a>00617             set(xTicks, &#39;FontSize&#39;, 5);
<a name="l00618"></a>00618             xlabelPos = get(get(xAxisHandle, &#39;XLabel&#39;), &#39;Position&#39;);
<a name="l00619"></a>00619             set(get(xAxisHandle, &#39;XLabel&#39;), &#39;Position&#39;, [xlabelPos(1) -140 xlabelPos(3:end)]);
<a name="l00620"></a>00620             
<a name="l00621"></a>00621             % Downsample
<a name="l00622"></a>00622             [~, simOrder] = sort(mass(1, :));
<a name="l00623"></a>00623             timeds = time(1:100:end);
<a name="l00624"></a>00624             massds = mass(1:100:end, :);
<a name="l00625"></a>00625             
<a name="l00626"></a>00626             axesHandle = axesHandles(1);
<a name="l00627"></a>00627             cla(axesHandle)
<a name="l00628"></a>00628             hold(axesHandle, &#39;on&#39;);
<a name="l00629"></a>00629             n1 = floor(numel(simOrder) / 2);
<a name="l00630"></a>00630             n2 = ceil(numel(simOrder) / 2);            
<a name="l00631"></a>00631             color1 = 0.50 * [1 1 1];
<a name="l00632"></a>00632             color2 = 0.25 * [1 1 1];
<a name="l00633"></a>00633             color3 = 0.00 * [1 1 1];
<a name="l00634"></a>00634             colorOrder = zeros(numel(simOrder), 3);
<a name="l00635"></a>00635             colorOrder(simOrder, :) = [
<a name="l00636"></a>00636                 repmat(linspace(1, 0, n1)&#39;, 1, 3) .* repmat(color1, n1, 1)  +  repmat(linspace(0, 1, n1)&#39;, 1, 3) .* repmat(color2, n1, 1)
<a name="l00637"></a>00637                 repmat(linspace(1, 0, n2)&#39;, 1, 3) .* repmat(color2, n2, 1)  +  repmat(linspace(0, 1, n2)&#39;, 1, 3) .* repmat(color3, n2, 1)
<a name="l00638"></a>00638                 ];
<a name="l00639"></a>00639             set(axesHandle, &#39;ColorOrder&#39;, colorOrder);
<a name="l00640"></a>00640             h = plot(axesHandle, timeds, massds, &#39;LineWidth&#39;, 0.25);
<a name="l00641"></a>00641             set(h(markIdx), &#39;Color&#39;, &#39;r&#39;, &#39;LineWidth&#39;, 2);
<a name="l00642"></a>00642             uistack(h(markIdx), &#39;top&#39;);
<a name="l00643"></a>00643             set(axesHandle, &#39;XLim&#39;, [0 max(time)], &#39;FontSize&#39;, 5);
<a name="l00644"></a>00644             set(axesHandle, &#39;YTick&#39;, 20:10:30);
<a name="l00645"></a>00645             tick2text(axesHandle, &#39;axis&#39;, &#39;y&#39;, &#39;ytickoffset&#39;, 0.042);
<a name="l00646"></a>00646             yTicks = getappdata(axesHandle, &#39;YTickText&#39;);
<a name="l00647"></a>00647             set(yTicks, &#39;FontSize&#39;, 5, &#39;HorizontalAlign&#39;, &#39;right&#39;);
<a name="l00648"></a>00648             ylabel(axesHandle, &#39;Mass (fg)&#39;, &#39;FontSize&#39;, 7);
<a name="l00649"></a>00649             ylabelPos = get(get(axesHandle, &#39;ylabel&#39;), &#39;position&#39;);
<a name="l00650"></a>00650             set(get(axesHandle, &#39;ylabel&#39;), &#39;position&#39;, [-0.9 ylabelPos(2:end)]);
<a name="l00651"></a>00651             ylim(axesHandle, [min(massds(:)), max(massds(:))]);
<a name="l00652"></a>00652             
<a name="l00653"></a>00653             axesHandle = axesHandles(2);
<a name="l00654"></a>00654             hold(axesHandle, &#39;on&#39;);
<a name="l00655"></a>00655             
<a name="l00656"></a>00656             edges = linspace(time(1), time(end), 100);
<a name="l00657"></a>00657             freq = histc(simEndTimes / 3600, edges) / numel(simEndTimes) * 100;
<a name="l00658"></a>00658             freq = smooth(freq, 9, &#39;lowess&#39;);
<a name="l00659"></a>00659             line(edges, freq, &#39;Color&#39;, &#39;k&#39;, &#39;Parent&#39;, axesHandle)
<a name="l00660"></a>00660             
<a name="l00661"></a>00661             idx = find(edges &lt;= simEndTimes(markIdx) / 3600, 1, &#39;last&#39;);
<a name="l00662"></a>00662             y = freq(idx) + diff(freq(idx:idx+1)) / diff(edges(idx:idx+1)) * (simEndTimes(markIdx) / 3600 - edges(idx));
<a name="l00663"></a>00663                         
<a name="l00664"></a>00664             axesPos = get(axesHandle, &#39;Position&#39;);
<a name="l00665"></a>00665             arrowX = axesPos(1) + simEndTimes(markIdx) / 3600 / time(end) * axesPos(3);
<a name="l00666"></a>00666             arrowY1 = axesPos(2);
<a name="l00667"></a>00667             axesPos = get(axesHandles(1), &#39;Position&#39;);
<a name="l00668"></a>00668             arrowY2 = axesPos(2) + axesPos(4)*(massds(floor(simEndTimes(markIdx)/100), markIdx) - min(massds(:))) / range(massds(:));
<a name="l00669"></a>00669             y = linspace(arrowY1, arrowY2-0.01, 81);
<a name="l00670"></a>00670             for i = 1:3:numel(y)
<a name="l00671"></a>00671                 if i &gt;= 28 &amp;&amp; i &lt;= 50
<a name="l00672"></a>00672                     continue;
<a name="l00673"></a>00673                 end
<a name="l00674"></a>00674                 h = annotation(get(axesHandle, &#39;Parent&#39;), &#39;line&#39;, arrowX * [1 1], y([i+1 i]), ...
<a name="l00675"></a>00675                     &#39;Color&#39;, &#39;r&#39;);
<a name="l00676"></a>00676                 uistack(h, &#39;bottom&#39;);
<a name="l00677"></a>00677             end
<a name="l00678"></a>00678             
<a name="l00679"></a>00679             annotation(&#39;textbox&#39;, [arrowX-0.06 y(27) 0.12 0.05], &#39;String&#39;, sprintf(&#39;Median cell\n\\tau = %0.1f h&#39;, 8.9), ...
<a name="l00680"></a>00680                 &#39;HorizontalAlign&#39;, &#39;center&#39;, &#39;verticalAlign&#39;, &#39;middle&#39;, ....
<a name="l00681"></a>00681                 &#39;FontSize&#39;, 7, &#39;EdgeColor&#39;, &#39;none&#39;, &#39;BackgroundColor&#39;, &#39;none&#39;, &#39;Color&#39;, &#39;r&#39;);
<a name="l00682"></a>00682             
<a name="l00683"></a>00683             ylim(axesHandle, [0 max(freq)]);
<a name="l00684"></a>00684             xlim(axesHandle, [time(1) time(end)]);
<a name="l00685"></a>00685             
<a name="l00686"></a>00686             ylabel(axesHandle, {<span class="stringliteral">&#39;% Cell&#39;</span> <span class="stringliteral">&#39;division&#39;</span>}, <span class="stringliteral">&#39;FontSize&#39;</span>, 7);
<a name="l00687"></a>00687             ylabelPos = <span class="keyword">get</span>(<span class="keyword">get</span>(axesHandle, <span class="stringliteral">&#39;ylabel&#39;</span>), <span class="stringliteral">&#39;position&#39;</span>);
<a name="l00688"></a>00688             <span class="keyword">set</span>(<span class="keyword">get</span>(axesHandle, <span class="stringliteral">&#39;ylabel&#39;</span>), <span class="stringliteral">&#39;position&#39;</span>, [-0.9 ylabelPos(2:end)]);
<a name="l00689"></a>00689             
<a name="l00690"></a>00690             <span class="keyword">set</span>(axesHandle, <span class="stringliteral">&#39;YTick&#39;</span>, 0:4:8);
<a name="l00691"></a>00691             <span class="keyword">set</span>(axesHandle, <span class="stringliteral">&#39;XTickLabel&#39;</span>,<span class="stringliteral">&#39;&#39;</span>);
<a name="l00692"></a>00692             <span class="keyword">set</span>(axesHandle, <span class="stringliteral">&#39;XTick&#39;</span>, []);
<a name="l00693"></a>00693             <span class="keyword">set</span>(axesHandle, <span class="stringliteral">&#39;Box&#39;</span>, <span class="stringliteral">&#39;off&#39;</span>);
<a name="l00694"></a>00694             <span class="keyword">set</span>(axesHandle, <span class="stringliteral">&#39;XLim&#39;</span>, [0 max(time)]);
<a name="l00695"></a>00695             
<a name="l00696"></a>00696             tick2text(axesHandle, <span class="stringliteral">&#39;axis&#39;</span>, <span class="charliteral">&#39;y&#39;</span>, <span class="stringliteral">&#39;ytickoffset&#39;</span>, 0.042);
<a name="l00697"></a>00697             yTicks = getappdata(axesHandle, <span class="stringliteral">&#39;YTickText&#39;</span>);
<a name="l00698"></a>00698             <span class="keyword">set</span>(yTicks, <span class="stringliteral">&#39;FontSize&#39;</span>, 5, <span class="stringliteral">&#39;HorizontalAlign&#39;</span>, <span class="stringliteral">&#39;right&#39;</span>);
<a name="l00699"></a>00699             <span class="keyword">delete</span>(yTicks(2));
<a name="l00700"></a>00700             
<a name="l00701"></a>00701             %% offset axes
<a name="l00702"></a>00702             PlotUtil.offsetYAxes(axesHandles, 0.02);
<a name="l00703"></a>00703         end
<a name="l00704"></a>00704         
<a name="l00705"></a>00705         <span class="keyword">function</span> growthMeasurement(figHandle, position, outDirectory)
<a name="l00706"></a>00706             %% parameters
<a name="l00707"></a>00707             [r, host] = system(<span class="stringliteral">&#39;hostname&#39;</span>);
<a name="l00708"></a>00708 
<a name="l00709"></a>00709             <span class="keywordflow">if</span> r ~= 0
<a name="l00710"></a>00710                 warning(<span class="stringliteral">&#39;WholeCell:warning:network&#39;</span>, <span class="stringliteral">&#39;Could not determine host name&#39;</span>);
<a name="l00711"></a>00711                 value = <span class="stringliteral">&#39;&#39;</span>;
<a name="l00712"></a>00712                 <span class="keywordflow">return</span>;
<a name="l00713"></a>00713             end
<a name="l00714"></a>00714             host = deblank(host);
<a name="l00715"></a>00715             <span class="keywordflow">switch</span> host
<a name="l00716"></a>00716                 % Edit as appropriate (add jayodita and markus)
<a name="l00717"></a>00717                 <span class="keywordflow">case</span> <span class="stringliteral">&#39;hydra&#39;</span>
<a name="l00718"></a>00718                     baseDir = <span class="stringliteral">&#39;/covertlab/home/share/Mycoplasma genitalium/MGenGrowthCurves&#39;</span>;
<a name="l00719"></a>00719                 <span class="keywordflow">case</span> <span class="stringliteral">&#39;silico&#39;</span>
<a name="l00720"></a>00720                     baseDir = <span class="stringliteral">&#39;/covertlab/home/share/Mycoplasma genitalium/MGenGrowthCurves&#39;</span>;
<a name="l00721"></a>00721                 <span class="keywordflow">case</span> <span class="stringliteral">&#39;covertlab-jkarr&#39;</span>
<a name="l00722"></a>00722                     baseDir = <span class="stringliteral">&#39;Z:\share\Mycoplasma genitalium\MGenGrowthCurves&#39;</span>;
<a name="l00723"></a>00723                 otherwise
<a name="l00724"></a>00724                     <span class="keywordflow">if</span> strcmp(computer, <span class="stringliteral">&#39;GLNXA64&#39;</span>)
<a name="l00725"></a>00725                         % Probably the cluster
<a name="l00726"></a>00726                         baseDir = <span class="stringliteral">&#39;/home/share/Mycoplasma genitalium/MGenGrowthCurves&#39;</span>;
<a name="l00727"></a>00727                     <span class="keywordflow">else</span>
<a name="l00728"></a>00728                         % Assume a PC with the correct Z drive mapping
<a name="l00729"></a>00729                         baseDir = <span class="stringliteral">&#39;Z:\share\Mycoplasma genitalium\MGenGrowthCurves&#39;</span>;
<a name="l00730"></a>00730                         warning(<span class="stringliteral">&#39;WholeCell:warning:network&#39;</span>, <span class="stringliteral">&#39;Assuming M. genitalium growth curves are in %s&#39;</span>, baseDir);
<a name="l00731"></a>00731                     end
<a name="l00732"></a>00732             end
<a name="l00733"></a>00733             
<a name="l00734"></a>00734             plate = 93;
<a name="l00735"></a>00735             condition = 1;
<a name="l00736"></a>00736             replicate = 2;
<a name="l00737"></a>00737             blankCondition = 2;
<a name="l00738"></a>00738             blankReplicate = 1;
<a name="l00739"></a>00739             nDilutions = 6;
<a name="l00740"></a>00740             
<a name="l00741"></a>00741             nReplicates = 3;
<a name="l00742"></a>00742             nColumns = 10;
<a name="l00743"></a>00743             excelRows = 3:8;
<a name="l00744"></a>00744             %4: 1X
<a name="l00745"></a>00745             %5: 1/5X
<a name="l00746"></a>00746             %6: 1/25X
<a name="l00747"></a>00747             excelCols = 3:12;
<a name="l00748"></a>00748             %3: Cond 1, Rep 1;
<a name="l00749"></a>00749             %4: Cond 1, Rep 2;
<a name="l00750"></a>00750             %5: Cond 1, Rep 3;
<a name="l00751"></a>00751             %6: Cond 2, Rep 1;
<a name="l00752"></a>00752             %7: Cond 2, Rep 2;
<a name="l00753"></a>00753             %8: Cond 2, Rep 3;
<a name="l00754"></a>00754             %9: Cond 3, Rep 1;
<a name="l00755"></a>00755             %10: Cond 3, Rep 2;
<a name="l00756"></a>00756             %11: Cond 3, Rep 3;
<a name="l00757"></a>00757             %12: Blank
<a name="l00758"></a>00758             
<a name="l00759"></a>00759             %% <span class="keyword">get</span> file name and times
<a name="l00760"></a>00760             files = dir(sprintf(<span class="stringliteral">&#39;%s%sPlate%d&#39;</span>, baseDir, filesep, plate));
<a name="l00761"></a>00761             files = files(3:end);
<a name="l00762"></a>00762             [~, order] = sort(cellfun(@(x) str2double(x(end-5:end-4)), {files.name}));
<a name="l00763"></a>00763             files = files(order);
<a name="l00764"></a>00764             times = cellfun(@(x) datenum(...
<a name="l00765"></a>00765                 2000 + str2double(x(end-9:end-8)), ...   %year
<a name="l00766"></a>00766                 str2double(x(end-15:end-14)), ... %month
<a name="l00767"></a>00767                 str2double(x(end-12:end-11)), ... %day
<a name="l00768"></a>00768                 str2double(x(end-21:end-20)), ... %hour
<a name="l00769"></a>00769                 str2double(x(end-18:end-17)), ... %minute
<a name="l00770"></a>00770                 0 ... %second
<a name="l00771"></a>00771                 ), {files.name});
<a name="l00772"></a>00772             times = times - times(1);
<a name="l00773"></a>00773             
<a name="l00774"></a>00774             %% <span class="keyword">get</span> data
<a name="l00775"></a>00775             %from plate reader files
<a name="l00776"></a>00776             <span class="keywordflow">if</span> exist([outDirectory <span class="stringliteral">&#39;growthMeasurement.mat&#39;</span>], <span class="stringliteral">&#39;file&#39;</span>)
<a name="l00777"></a>00777                 load([outDirectory &#39;growthMeasurement.mat&#39;])
<a name="l00778"></a>00778             else
<a name="l00779"></a>00779                 data = zeros(nDilutions, nColumns, numel(files));
<a name="l00780"></a>00780                 for i = 1:numel(files)
<a name="l00781"></a>00781                     [~, ~, raw] = xlsread(sprintf(&#39;%s%sPlate%d%s%s&#39;, baseDir, filesep, plate, filesep, files(i).name));
<a name="l00782"></a>00782                     data(:, :, i) = cell2mat(raw(excelRows, excelCols));
<a name="l00783"></a>00783                 end
<a name="l00784"></a>00784                 save([outDirectory &#39;growthMeasurement.mat&#39;], &#39;data&#39;);
<a name="l00785"></a>00785             end
<a name="l00786"></a>00786             
<a name="l00787"></a>00787             % from Jayodita&#39;s excel sheet
<a name="l00788"></a>00788             [~, ~, raw] = xlsread([outDirectory &#39;growthMeasurement.xls&#39;]);
<a name="l00789"></a>00789             tfRows = [false; false; ~isnan(cell2mat(raw(3:end, 1)))];
<a name="l00790"></a>00790             times = cell2mat(raw(tfRows, 1))&#39;;
<a name="l00791"></a>00791             times = (times - times(1)) / 24;
<a name="l00792"></a>00792             data = zeros(nDilutions, nColumns, numel(times));
<a name="l00793"></a>00793             data(3:5, (condition-1) * nReplicates + replicate, :) = permute(cell2mat(raw(tfRows, 2:4)), [2 3 1]);
<a name="l00794"></a>00794             data(2:6, (blankCondition-1) * nReplicates + blankReplicate, :) = repmat(permute(cell2mat(raw(tfRows, 5)), [2 3 1]), [5 1 1]);
<a name="l00795"></a>00795             
<a name="l00796"></a>00796             %% fit data
<a name="l00797"></a>00797             t = NaN(nDilutions, 1);
<a name="l00798"></a>00798             for i = 1:nDilutions
<a name="l00799"></a>00799                 tmpData = permute(data(i, (condition-1) * nReplicates + replicate, :), [1 3 2]);
<a name="l00800"></a>00800                 idx1 = find(tmpData &gt; 0.35, 1, &#39;last&#39;);
<a name="l00801"></a>00801                 idx2 = find(tmpData &lt; 0.2, 1, &#39;first&#39;);
<a name="l00802"></a>00802                 if ~isempty(idx1) &amp;&amp; ~isempty(idx2)
<a name="l00803"></a>00803                     t(i) = interp1(tmpData(:, idx1:idx2), times(:, idx1:idx2) * 24, 0.28, &#39;linear&#39;);
<a name="l00804"></a>00804                 end
<a name="l00805"></a>00805             end
<a name="l00806"></a>00806             deltaT = diff(t);
<a name="l00807"></a>00807             tau = log(2) ./ (log(5) ./ deltaT);
<a name="l00808"></a>00808             
<a name="l00809"></a>00809             %% plot
<a name="l00810"></a>00810             if nargin &gt;= 2
<a name="l00811"></a>00811                 axesHandle = subplot(&#39;position&#39;, position, &#39;Parent&#39;, figHandle);
<a name="l00812"></a>00812             else
<a name="l00813"></a>00813                 axesHandle = edu.stanford.covert.cell.sim.util.PlotUtil.newAxesHandle();
<a name="l00814"></a>00814             end
<a name="l00815"></a>00815             hold(axesHandle, &#39;on&#39;);
<a name="l00816"></a>00816             
<a name="l00817"></a>00817             colors = 1 / 255 * [
<a name="l00818"></a>00818                 25 25 112
<a name="l00819"></a>00819                 0 0 255
<a name="l00820"></a>00820                 30 144 255
<a name="l00821"></a>00821                 139 137 137
<a name="l00822"></a>00822                 ];
<a name="l00823"></a>00823             
<a name="l00824"></a>00824             h = plot(axesHandle, times, permute(data(3:5, (condition-1) * nReplicates + replicate, :), [1 3 2]), &#39;LineWidth&#39;, 1); %dilutions
<a name="l00825"></a>00825             plot(axesHandle, times, permute(mean(data(2:6, (blankCondition-1) * nReplicates + blankReplicate, :), 1), [1 3 2]), &#39;Color&#39;, colors(4, :), &#39;LineWidth&#39;, 1); %blank
<a name="l00826"></a>00826             set(h(1), &#39;Color&#39;, colors(1, :)); %1x
<a name="l00827"></a>00827             set(h(2), &#39;Color&#39;, colors(2, :)); %1/5 X
<a name="l00828"></a>00828             set(h(3), &#39;Color&#39;, colors(3, :)); %1/25 X
<a name="l00829"></a>00829             xlabel(axesHandle, &#39;Time (d)&#39;, &#39;FontSize&#39;, 7);
<a name="l00830"></a>00830             ylabel(axesHandle, &#39;OD550&#39;, &#39;FontSize&#39;, 7);
<a name="l00831"></a>00831             
<a name="l00832"></a>00832             set(axesHandle, &#39;FontSize&#39;, 5, &#39;TickDir&#39;, &#39;out&#39;, &#39;YDir&#39;, &#39;Reverse&#39;, &#39;XTick&#39;, 0:5:15, &#39;YTick&#39;, 0:.2:0.6);
<a name="l00833"></a>00833             box(axesHandle, &#39;off&#39;);
<a name="l00834"></a>00834             xlim(axesHandle, [0 17]);
<a name="l00835"></a>00835             ylim(axesHandle, [0 0.65]);
<a name="l00836"></a>00836             
<a name="l00837"></a>00837             text(0.3, 0.048, &#39;\tau =&#39;, &#39;FontSize&#39;, 7, &#39;Interpreter&#39;, &#39;tex&#39;)
<a name="l00838"></a>00838             text(3.9, 0.0175, &#39;ln(2) {\Delta}{\it{t}}<span class="stringliteral">&#39;, &#39;</span>FontSize<span class="stringliteral">&#39;, 7, &#39;</span>Interpreter<span class="stringliteral">&#39;, &#39;</span>tex<span class="stringliteral">&#39;, &#39;</span>HorizontalAlign<span class="stringliteral">&#39;, &#39;</span>center<span class="stringliteral">&#39;)</span>
<a name="l00839"></a>00839 <span class="stringliteral">            text(3.9, 0.0785, &#39;</span>ln(dilution factor)<span class="stringliteral">&#39;, &#39;</span>FontSize<span class="stringliteral">&#39;, 7, &#39;</span>Interpreter<span class="stringliteral">&#39;, &#39;</span>tex<span class="stringliteral">&#39;, &#39;</span>HorizontalAlign<span class="stringliteral">&#39;, &#39;</span>center<span class="stringliteral">&#39;)</span>
<a name="l00840"></a>00840 <span class="stringliteral">            line([1.35 6.45], 0.051 * [1 1], &#39;</span>Parent<span class="stringliteral">&#39;, axesHandle, &#39;</span>Color<span class="stringliteral">&#39;, &#39;</span>k<span class="stringliteral">&#39;);</span>
<a name="l00841"></a>00841 <span class="stringliteral">            </span>
<a name="l00842"></a>00842 <span class="stringliteral">            text(0.3, .33, &#39;</span>1X dilution<span class="stringliteral">&#39;, &#39;</span>Parent<span class="stringliteral">&#39;, axesHandle, &#39;</span>FontSize<span class="stringliteral">&#39;, 7, &#39;</span>VerticalAlign<span class="stringliteral">&#39;, &#39;</span>middle<span class="stringliteral">&#39;, &#39;</span>Color<span class="stringliteral">&#39;, colors(1, :))</span>
<a name="l00843"></a>00843 <span class="stringliteral">            text(0.3, .390, &#39;</span>5X dilution<span class="stringliteral">&#39;, &#39;</span>Parent<span class="stringliteral">&#39;, axesHandle, &#39;</span>FontSize<span class="stringliteral">&#39;, 7, &#39;</span>VerticalAlign<span class="stringliteral">&#39;, &#39;</span>middle<span class="stringliteral">&#39;, &#39;</span>Color<span class="stringliteral">&#39;, colors(2, :))</span>
<a name="l00844"></a>00844 <span class="stringliteral">            text(0.3, .45, &#39;</span>25X dilution<span class="stringliteral">&#39;, &#39;</span>Parent<span class="stringliteral">&#39;, axesHandle, &#39;</span>FontSize<span class="stringliteral">&#39;, 7, &#39;</span>VerticalAlign<span class="stringliteral">&#39;, &#39;</span>middle<span class="stringliteral">&#39;, &#39;</span>Color<span class="stringliteral">&#39;, colors(3, :))</span>
<a name="l00845"></a>00845 <span class="stringliteral">            text(0.3, .510, &#39;</span>Blank<span class="stringliteral">&#39;, &#39;</span>Parent<span class="stringliteral">&#39;, axesHandle, &#39;</span>FontSize<span class="stringliteral">&#39;, 7, &#39;</span>VerticalAlign<span class="stringliteral">&#39;, &#39;</span>middle<span class="stringliteral">&#39;, &#39;</span>Color<span class="stringliteral">&#39;, colors(4, :))</span>
<a name="l00846"></a>00846 <span class="stringliteral">            </span>
<a name="l00847"></a>00847 <span class="stringliteral">            axesPos = get(axesHandle, &#39;</span>Position<span class="stringliteral">&#39;);</span>
<a name="l00848"></a>00848 <span class="stringliteral">            </span>
<a name="l00849"></a>00849 <span class="stringliteral">            y = 0.28;</span>
<a name="l00850"></a>00850 <span class="stringliteral">            x = t(3)/24-0.3;</span>
<a name="l00851"></a>00851 <span class="stringliteral">            text(x, y - 0.02, sprintf(&#39;</span>{\\Delta}{\\it{t}} = %0.1f h<span class="stringliteral">&#39;, deltaT(3)), &#39;</span>Parent<span class="stringliteral">&#39;, axesHandle, &#39;</span>FontSize<span class="stringliteral">&#39;, 6, &#39;</span>HorizontalAlign<span class="stringliteral">&#39;, &#39;</span>right<span class="stringliteral">&#39;, &#39;</span>verticalAlign<span class="stringliteral">&#39;, &#39;</span>middle<span class="stringliteral">&#39;, &#39;</span>Interpreter<span class="stringliteral">&#39;, &#39;</span>tex<span class="stringliteral">&#39;)</span>
<a name="l00852"></a>00852 <span class="stringliteral">            text(x-0.355, y + 0.02, sprintf(&#39;</span>\\tau = %0.1f h<span class="stringliteral">&#39;, tau(3)), &#39;</span>Parent<span class="stringliteral">&#39;, axesHandle, &#39;</span>FontSize<span class="stringliteral">&#39;, 6, &#39;</span>HorizontalAlign<span class="stringliteral">&#39;, &#39;</span>right<span class="stringliteral">&#39;, &#39;</span>verticalAlign<span class="stringliteral">&#39;, &#39;</span>middle<span class="stringliteral">&#39;, &#39;</span>Interpreter<span class="stringliteral">&#39;, &#39;</span>tex<span class="stringliteral">&#39;)</span>
<a name="l00853"></a>00853 <span class="stringliteral">            annotation(&#39;</span>doublearrow<span class="stringliteral">&#39;, axesPos(1) + axesPos(3) * t([3 4])/24 / max(xlim(axesHandle)), ...</span>
<a name="l00854"></a>00854 <span class="stringliteral">                axesPos(2) + axesPos(4) * (max(ylim(axesHandle))-y) / range(ylim(axesHandle)) * [1 1], ...</span>
<a name="l00855"></a>00855 <span class="stringliteral">                &#39;</span>HeadLength<span class="stringliteral">&#39;, 2, &#39;</span>HeadWidth<span class="stringliteral">&#39;, 3, &#39;</span>HeadStyle<span class="stringliteral">&#39;, &#39;</span>vback1<span class="stringliteral">&#39;)</span>
<a name="l00856"></a>00856 <span class="stringliteral">            </span>
<a name="l00857"></a>00857 <span class="stringliteral">            y = 0.28;</span>
<a name="l00858"></a>00858 <span class="stringliteral">            x = t(5)/24+0.3;</span>
<a name="l00859"></a>00859 <span class="stringliteral">            text(x, y - 0.02, sprintf(&#39;</span>{\\Delta}{\\it{t}} = %0.1f h<span class="stringliteral">&#39;, deltaT(4)), &#39;</span>Parent<span class="stringliteral">&#39;, axesHandle, &#39;</span>FontSize<span class="stringliteral">&#39;, 6, &#39;</span>HorizontalAlign<span class="stringliteral">&#39;, &#39;</span>left<span class="stringliteral">&#39;, &#39;</span>verticalAlign<span class="stringliteral">&#39;, &#39;</span>middle<span class="stringliteral">&#39;, &#39;</span>Interpreter<span class="stringliteral">&#39;, &#39;</span>tex<span class="stringliteral">&#39;)</span>
<a name="l00860"></a>00860 <span class="stringliteral">            text(x + 0.295, y + 0.02, sprintf(&#39;</span>\\tau = %0.1f h<span class="stringliteral">&#39;, tau(4)), &#39;</span>Parent<span class="stringliteral">&#39;, axesHandle, &#39;</span>FontSize<span class="stringliteral">&#39;, 6, &#39;</span>HorizontalAlign<span class="stringliteral">&#39;, &#39;</span>left<span class="stringliteral">&#39;, &#39;</span>verticalAlign<span class="stringliteral">&#39;, &#39;</span>middle<span class="stringliteral">&#39;, &#39;</span>Interpreter<span class="stringliteral">&#39;, &#39;</span>tex<span class="stringliteral">&#39;)</span>
<a name="l00861"></a>00861 <span class="stringliteral">            annotation(&#39;</span>doublearrow<span class="stringliteral">&#39;, axesPos(1) + axesPos(3) * t([4 5])/24 / max(xlim(axesHandle)), ...</span>
<a name="l00862"></a>00862 <span class="stringliteral">                axesPos(2) + axesPos(4) * (max(ylim(axesHandle))-y) / range(ylim(axesHandle)) * [1 1], ...</span>
<a name="l00863"></a>00863 <span class="stringliteral">                &#39;</span>HeadLength<span class="stringliteral">&#39;, 2, &#39;</span>HeadWidth<span class="stringliteral">&#39;, 3, &#39;</span>HeadStyle<span class="stringliteral">&#39;, &#39;</span>vback1<span class="stringliteral">&#39;)</span>
<a name="l00864"></a>00864 <span class="stringliteral">            </span>
<a name="l00865"></a>00865 <span class="stringliteral">            annotation(&#39;</span>textbox<span class="stringliteral">&#39;, [0.22 0.7850 0.12 0.05], &#39;</span>String<span class="stringliteral">&#39;, sprintf(&#39;</span>Mean\n\\tau = %0.1f h<span class="stringliteral">&#39;, 9.03), ... %Jayodita: where do this come from?</span>
<a name="l00866"></a>00866 <span class="stringliteral">                &#39;</span>HorizontalAlign<span class="stringliteral">&#39;, &#39;</span>center<span class="stringliteral">&#39;, &#39;</span>verticalAlign<span class="stringliteral">&#39;, &#39;</span>middle<span class="stringliteral">&#39;, ....</span>
<a name="l00867"></a>00867 <span class="stringliteral">                &#39;</span>FontSize<span class="stringliteral">&#39;, 7, &#39;</span>EdgeColor<span class="stringliteral">&#39;, &#39;</span>none<span class="stringliteral">&#39;, &#39;</span>BackgroundColor<span class="stringliteral">&#39;, &#39;</span>none<span class="stringliteral">&#39;, &#39;</span>Color<span class="stringliteral">&#39;, &#39;</span>r<span class="stringliteral">&#39;);</span>
<a name="l00868"></a>00868 <span class="stringliteral">            </span>
<a name="l00869"></a>00869 <span class="stringliteral">            tick2text(axesHandle, &#39;</span>axis<span class="stringliteral">&#39;, &#39;</span>xy<span class="stringliteral">&#39;, &#39;</span>xtickoffset<span class="stringliteral">&#39;, -1.065, &#39;</span>ytickoffset<span class="stringliteral">&#39;, 0.02);</span>
<a name="l00870"></a>00870 <span class="stringliteral">            xlabelPos = get(get(axesHandle, &#39;</span>xlabel<span class="stringliteral">&#39;), &#39;</span>position<span class="stringliteral">&#39;);</span>
<a name="l00871"></a>00871 <span class="stringliteral">            ylabelPos = get(get(axesHandle, &#39;</span>ylabel<span class="stringliteral">&#39;), &#39;</span>position<span class="stringliteral">&#39;);</span>
<a name="l00872"></a>00872 <span class="stringliteral">            xTicks = getappdata(axesHandle, &#39;</span>XTickText<span class="stringliteral">&#39;);</span>
<a name="l00873"></a>00873 <span class="stringliteral">            yTicks = getappdata(axesHandle, &#39;</span>YTickText<span class="stringliteral">&#39;);</span>
<a name="l00874"></a>00874 <span class="stringliteral">            set(xTicks, &#39;</span>FontSize<span class="stringliteral">&#39;, 5);</span>
<a name="l00875"></a>00875 <span class="stringliteral">            set(yTicks, &#39;</span>FontSize<span class="stringliteral">&#39;, 5, &#39;</span>horizontalAlign<span class="stringliteral">&#39;, &#39;</span>right<span class="stringliteral">&#39;);</span>
<a name="l00876"></a>00876 <span class="stringliteral">            set(get(axesHandle, &#39;</span>xlabel<span class="stringliteral">&#39;), &#39;</span>position<span class="stringliteral">&#39;, [xlabelPos(1) 0.702 xlabelPos(3:end)]);</span>
<a name="l00877"></a>00877 <span class="stringliteral">            set(get(axesHandle, &#39;</span>ylabel<span class="stringliteral">&#39;), &#39;</span>position<span class="stringliteral">&#39;, [-0.92 ylabelPos(2:end)]);</span>
<a name="l00878"></a>00878 <span class="stringliteral">        end</span>
<a name="l00879"></a>00879 <span class="stringliteral">        </span>
<a name="l00880"></a>00880 <span class="stringliteral">        function doublingTime(outDirectory)</span>
<a name="l00881"></a>00881 <span class="stringliteral">            % From</span>
<a name="l00882"></a>00882 <span class="stringliteral">            % /home/projects/WholeCell/simulation/output/runSimulation/singleGeneDeletions/summary.xls</span>
<a name="l00883"></a>00883 <span class="stringliteral">            modelMassDTMean = 8.92200510884003; % Col W</span>
<a name="l00884"></a>00884 <span class="stringliteral">            modelMassDTStddev = 0.807282508239662; % Col X</span>
<a name="l00885"></a>00885 <span class="stringliteral">            expMassDTMean = 9.03411809133663; % Col AX</span>
<a name="l00886"></a>00886 <span class="stringliteral">            expMassDTStderr = 0.889827929955399; % Col AY</span>
<a name="l00887"></a>00887 <span class="stringliteral">            </span>
<a name="l00888"></a>00888 <span class="stringliteral">            [axesHandle, figHandle] = edu.stanford.covert.cell.sim.util.PlotUtil.newAxesHandle();</span>
<a name="l00889"></a>00889 <span class="stringliteral">            hold(&#39;</span>on<span class="stringliteral">&#39;);</span>
<a name="l00890"></a>00890 <span class="stringliteral">            bar(1, modelMassDTMean);</span>
<a name="l00891"></a>00891 <span class="stringliteral">            bar(2, expMassDTMean, &#39;</span>k<span class="stringliteral">&#39;);</span>
<a name="l00892"></a>00892 <span class="stringliteral">            eh1 = errorbar([modelMassDTMean nan], [modelMassDTStddev nan], &#39;</span>.g<span class="stringliteral">&#39;);</span>
<a name="l00893"></a>00893 <span class="stringliteral">            eh2 = errorbar([nan expMassDTMean], [nan expMassDTStderr], &#39;</span>.r<span class="stringliteral">&#39;);</span>
<a name="l00894"></a>00894 <span class="stringliteral">            hold(&#39;</span>off<span class="stringliteral">&#39;);</span>
<a name="l00895"></a>00895 <span class="stringliteral">            legend([eh1 eh2], &#39;</span>Standard Deviation<span class="stringliteral">&#39;, &#39;</span>Standard Error<span class="stringliteral">&#39;, &#39;</span>Location<span class="stringliteral">&#39;, &#39;</span>BestOutside<span class="stringliteral">&#39;);</span>
<a name="l00896"></a>00896 <span class="stringliteral">            ylabel(&#39;</span>Mass Doubling Time (h)<span class="stringliteral">&#39;);</span>
<a name="l00897"></a>00897 <span class="stringliteral">            set(axesHandle, &#39;</span>XTick<span class="stringliteral">&#39;, [1 2]);</span>
<a name="l00898"></a>00898 <span class="stringliteral">            set(axesHandle, &#39;</span>XTickLabel<span class="stringliteral">&#39;, {&#39;</span>Model<span class="charliteral">&#39; &#39;</span>Experiment<span class="stringliteral">&#39;});</span>
<a name="l00899"></a>00899 <span class="stringliteral">            set(axesHandle, &#39;</span>YTick<span class="stringliteral">&#39;, [0 2 4 6 8 10]);</span>
<a name="l00900"></a>00900 <span class="stringliteral">            set(axesHandle, &#39;</span>Box<span class="stringliteral">&#39;, &#39;</span>off<span class="stringliteral">&#39;);</span>
<a name="l00901"></a>00901 <span class="stringliteral">            set(axesHandle, &#39;</span>TickDir<span class="stringliteral">&#39;, &#39;</span>out<span class="stringliteral">&#39;);</span>
<a name="l00902"></a>00902 <span class="stringliteral">            </span>
<a name="l00903"></a>00903 <span class="stringliteral">            set(figHandle, &#39;</span>Renderer<span class="stringliteral">&#39;, &#39;</span>Painters<span class="stringliteral">&#39;)</span>
<a name="l00904"></a>00904 <span class="stringliteral">            print(figHandle, [outDirectory filesep &#39;</span>MassDoublingTime.eps<span class="stringliteral">&#39;], &#39;</span>-depsc&#39;, &#39;-cmyk&#39;);
<a name="l00905"></a>00905         end
<a name="l00906"></a>00906         
<a name="l00907"></a>00907         <span class="keyword">function</span> dryBiomassComposition(simBatchDir, figHandle, position, outDirectory)
<a name="l00908"></a>00908             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
<a name="l00909"></a>00909             import edu.stanford.covert.cell.sim.util.SimulationEnsemble;
<a name="l00910"></a>00910             
<a name="l00911"></a>00911             if ~exist(&#39;simBatchDir&#39;, &#39;var&#39;) || isempty(simBatchDir)
<a name="l00912"></a>00912                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
<a name="l00913"></a>00913             end
<a name="l00914"></a>00914             selectedSimulations = 1:SimulationDiskUtil.getNumSimulations(simBatchDir);
<a name="l00915"></a>00915             
<a name="l00916"></a>00916             %% get data
<a name="l00917"></a>00917             if exist([outDirectory &#39;dryBiomassComposition.mat&#39;], &#39;file&#39;)
<a name="l00918"></a>00918                 load([outDirectory &#39;dryBiomassComposition.mat&#39;])
<a name="l00919"></a>00919             else
<a name="l00920"></a>00920                 [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep &#39;1&#39;]);
<a name="l00921"></a>00921                 comp = sim.compartment;
<a name="l00922"></a>00922                 
<a name="l00923"></a>00923                 stateNames = {
<a name="l00924"></a>00924                     <span class="stringliteral">&#39;Time&#39;</span> <span class="stringliteral">&#39;values&#39;</span>       <span class="charliteral">&#39;:&#39;</span> <span class="charliteral">&#39;:&#39;</span>
<a name="l00925"></a>00925                     <span class="stringliteral">&#39;Mass&#39;</span> <span class="stringliteral">&#39;dnaWt&#39;</span>        <span class="charliteral">&#39;:&#39;</span> <span class="stringliteral">&#39;-sum&#39;</span>
<a name="l00926"></a>00926                     <span class="stringliteral">&#39;Mass&#39;</span> <span class="stringliteral">&#39;rnaWt&#39;</span>        <span class="charliteral">&#39;:&#39;</span> <span class="stringliteral">&#39;-sum&#39;</span>
<a name="l00927"></a>00927                     <span class="stringliteral">&#39;Mass&#39;</span> <span class="stringliteral">&#39;proteinWt&#39;</span>    <span class="charliteral">&#39;:&#39;</span> <span class="stringliteral">&#39;-sum&#39;</span>
<a name="l00928"></a>00928                     <span class="stringliteral">&#39;Mass&#39;</span> <span class="stringliteral">&#39;metaboliteWt&#39;</span> <span class="charliteral">&#39;:&#39;</span> [comp.membraneIndexs]
<a name="l00929"></a>00929                     <span class="stringliteral">&#39;Mass&#39;</span> <span class="stringliteral">&#39;cellDry&#39;</span>         <span class="charliteral">&#39;:&#39;</span> <span class="stringliteral">&#39;-sum&#39;</span>
<a name="l00930"></a>00930                     };
<a name="l00931"></a>00931                 states = SimulationEnsemble.load(simBatchDir, stateNames, 3600, 3600, 1, <span class="stringliteral">&#39;extract&#39;</span>, selectedSimulations);
<a name="l00932"></a>00932                 
<a name="l00933"></a>00933                 dnaWt = permute(states.Mass.dnaWt, [4 3 2 1]) ./ permute(states.Mass.cellDry, [4 3 2 1]) * 100;
<a name="l00934"></a>00934                 rnaWt = permute(states.Mass.rnaWt, [4 3 2 1]) ./ permute(states.Mass.cellDry, [4 3 2 1]) * 100;
<a name="l00935"></a>00935                 proteinWt = permute(states.Mass.proteinWt, [4 3 2 1]) ./ permute(states.Mass.cellDry, [4 3 2 1]) * 100;
<a name="l00936"></a>00936                 lipidWt = permute(states.Mass.metaboliteWt, [4 3 2 1]) ./ permute(states.Mass.cellDry, [4 3 2 1]) * 100;
<a name="l00937"></a>00937                 
<a name="l00938"></a>00938                 modelDNA = mean(dnaWt);
<a name="l00939"></a>00939                 modelRNA = mean(rnaWt);
<a name="l00940"></a>00940                 modelProtein = mean(proteinWt);
<a name="l00941"></a>00941                 modelLipid = mean(lipidWt);
<a name="l00942"></a>00942                 
<a name="l00943"></a>00943                 stdDNA = std(dnaWt);
<a name="l00944"></a>00944                 stdRNA = std(rnaWt);
<a name="l00945"></a>00945                 stdProtein = std(proteinWt);
<a name="l00946"></a>00946                 stdLipid = std(lipidWt);
<a name="l00947"></a>00947                 
<a name="l00948"></a>00948                 save([outDirectory <span class="stringliteral">&#39;dryBiomassComposition.mat&#39;</span>], ...
<a name="l00949"></a>00949                     <span class="stringliteral">&#39;modelDNA&#39;</span>, <span class="stringliteral">&#39;modelRNA&#39;</span>, <span class="stringliteral">&#39;modelProtein&#39;</span>, <span class="stringliteral">&#39;modelLipid&#39;</span>, ...
<a name="l00950"></a>00950                     <span class="stringliteral">&#39;stdDNA&#39;</span>, <span class="stringliteral">&#39;stdRNA&#39;</span>, <span class="stringliteral">&#39;stdProtein&#39;</span>, <span class="stringliteral">&#39;stdLipid&#39;</span>);
<a name="l00951"></a>00951             end
<a name="l00952"></a>00952             
<a name="l00953"></a>00953             % From Jonathan
<a name="l00954"></a>00954             expDNA = 4;
<a name="l00955"></a>00955             expLipid = 11;
<a name="l00956"></a>00956             expProtein = 80;
<a name="l00957"></a>00957             expRNA = 8;
<a name="l00958"></a>00958             
<a name="l00959"></a>00959             <span class="keywordflow">if</span> nargin &gt;= 4
<a name="l00960"></a>00960                 axesHandle = subplot(<span class="stringliteral">&#39;Position&#39;</span>, position, <span class="stringliteral">&#39;Parent&#39;</span>, figHandle);
<a name="l00961"></a>00961             <span class="keywordflow">else</span>
<a name="l00962"></a>00962                 axesHandle = edu.stanford.covert.cell.sim.util.PlotUtil.newAxesHandle();
<a name="l00963"></a>00963             end
<a name="l00964"></a>00964             hold(axesHandle, <span class="stringliteral">&#39;on&#39;</span>);
<a name="l00965"></a>00965             
<a name="l00966"></a>00966             h = [
<a name="l00967"></a>00967                 bar(axesHandle, (0:3)*2.5-0.5, [modelDNA modelLipid modelProtein modelRNA], 0.37, &#39;b&#39;)
<a name="l00968"></a>00968                 bar(axesHandle, (0:3)*2.5+0.5, [expDNA expLipid expProtein expRNA], 0.37, &#39;k&#39;)
<a name="l00969"></a>00969                 ];
<a name="l00970"></a>00970             errorbar((0:3)*2.5-0.5, [modelDNA modelLipid modelProtein modelRNA], [stdDNA stdLipid stdProtein stdRNA], ...
<a name="l00971"></a>00971                 <span class="charliteral">&#39;r&#39;</span>, <span class="stringliteral">&#39;Parent&#39;</span>, axesHandle, <span class="stringliteral">&#39;LineStyle&#39;</span>, <span class="stringliteral">&#39;none&#39;</span>);
<a name="l00972"></a>00972             <span class="keyword">set</span>(h, <span class="stringliteral">&#39;EdgeColor&#39;</span>, <span class="stringliteral">&#39;none&#39;</span>)
<a name="l00973"></a>00973             
<a name="l00974"></a>00974             <span class="keyword">set</span>(axesHandle, <span class="stringliteral">&#39;XTick&#39;</span>, (0:3)*2.5, <span class="stringliteral">&#39;Xlim&#39;</span>, [-1.3 3*2.5+1.3]);
<a name="l00975"></a>00975             <span class="keyword">set</span>(axesHandle, <span class="stringliteral">&#39;XTickLabel&#39;</span>, {<span class="stringliteral">&#39;DNA&#39;</span> <span class="stringliteral">&#39;Lipid&#39;</span> <span class="stringliteral">&#39;Protein&#39;</span> <span class="stringliteral">&#39;RNA&#39;</span>});
<a name="l00976"></a>00976             <span class="keyword">set</span>(axesHandle, <span class="stringliteral">&#39;YTick&#39;</span>, 0:25:75);
<a name="l00977"></a>00977             tick2text(axesHandle, <span class="stringliteral">&#39;axis&#39;</span>, <span class="charliteral">&#39;y&#39;</span>, <span class="stringliteral">&#39;ytickoffset&#39;</span>, 0.020);
<a name="l00978"></a>00978             yTicks = getappdata(axesHandle, <span class="stringliteral">&#39;YTickText&#39;</span>);
<a name="l00979"></a>00979             <span class="keyword">set</span>(yTicks, <span class="stringliteral">&#39;FontSize&#39;</span>, 5, <span class="stringliteral">&#39;horizontalAlign&#39;</span>, <span class="stringliteral">&#39;right&#39;</span>);
<a name="l00980"></a>00980             
<a name="l00981"></a>00981             <span class="keyword">set</span>(axesHandle, <span class="stringliteral">&#39;Box&#39;</span>, <span class="stringliteral">&#39;off&#39;</span>);
<a name="l00982"></a>00982             <span class="keyword">set</span>(axesHandle, <span class="stringliteral">&#39;TickDir&#39;</span>, <span class="stringliteral">&#39;out&#39;</span>);
<a name="l00983"></a>00983             <span class="keyword">set</span>(axesHandle, <span class="stringliteral">&#39;FontSize&#39;</span>, 5);
<a name="l00984"></a>00984 
<a name="l00985"></a>00985             legendHandle = legend({<span class="stringliteral">&#39;Model&#39;</span>, <span class="stringliteral">&#39;Experiment&#39;</span>}, <span class="stringliteral">&#39;Location&#39;</span>, <span class="stringliteral">&#39;NorthWest&#39;</span>, <span class="stringliteral">&#39;FontSize&#39;</span>, 7);
<a name="l00986"></a>00986             legend(axesHandle, <span class="stringliteral">&#39;boxoff&#39;</span>);
<a name="l00987"></a>00987             legendPos = <span class="keyword">get</span>(legendHandle, <span class="stringliteral">&#39;Position&#39;</span>);
<a name="l00988"></a>00988             <span class="keyword">set</span>(legendHandle, <span class="stringliteral">&#39;Position&#39;</span>, [0.99 1.025 legendPos(3:end)])
<a name="l00989"></a>00989             <span class="keyword">set</span>(legendHandle, <span class="stringliteral">&#39;Units&#39;</span>, <span class="stringliteral">&#39;pixels&#39;</span>);
<a name="l00990"></a>00990             legendPos = <span class="keyword">get</span>(legendHandle, <span class="stringliteral">&#39;Position&#39;</span>);
<a name="l00991"></a>00991             <span class="keyword">set</span>(legendHandle, <span class="stringliteral">&#39;Position&#39;</span>, [legendPos(1:3) 24])
<a name="l00992"></a>00992             legendChildren = <span class="keyword">get</span>(legendHandle, <span class="stringliteral">&#39;children&#39;</span>);
<a name="l00993"></a>00993             <span class="keywordflow">for</span> i = 1:2:numel(legendChildren)
<a name="l00994"></a>00994                 ptch = get(legendChildren(i), &#39;children&#39;);
<a name="l00995"></a>00995                 set(ptch, &#39;xdata&#39;, [0.05 0.05 0.12 0.12 0.05])
<a name="l00996"></a>00996                 
<a name="l00997"></a>00997                 pos = get(legendChildren(i+1), &#39;Position&#39;);
<a name="l00998"></a>00998                 set(legendChildren(i+1), &#39;Position&#39;, [0.15 pos(2) 0])
<a name="l00999"></a>00999             end
<a name="l01000"></a>01000             
<a name="l01001"></a>01001             ylabel(axesHandle, &#39;Percent dry mass&#39;, &#39;FontSize&#39;, 7);
<a name="l01002"></a>01002             ylabelPos = get(get(axesHandle, &#39;ylabel&#39;), &#39;position&#39;);
<a name="l01003"></a>01003             set(get(axesHandle, &#39;ylabel&#39;), &#39;position&#39;, [-2.0 ylabelPos(2:end)]);
<a name="l01004"></a>01004             
<a name="l01005"></a>01005             tick2text(axesHandle, &#39;axis&#39;, &#39;x&#39;, &#39;xtickoffset&#39;, 0.07);
<a name="l01006"></a>01006             xTicks = getappdata(axesHandle, &#39;XTickText&#39;);
<a name="l01007"></a>01007             set(xTicks, &#39;FontSize&#39;, 7);
<a name="l01008"></a>01008             set(xTicks(1), &#39;String&#39;, &#39;DNA&#39;);
<a name="l01009"></a>01009             set(xTicks(2), &#39;String&#39;, &#39;Lipid&#39;);
<a name="l01010"></a>01010             set(xTicks(3), &#39;String&#39;, &#39;<a class="code" href="class_protein.html">Protein</a>&#39;);
<a name="l01011"></a>01011             set(xTicks(4), &#39;String&#39;, &#39;<a class="code" href="class_r_n_a.html">RNA</a>&#39;);
<a name="l01012"></a>01012         end
<a name="l01013"></a>01013         
<a name="l01014"></a>01014         function cellCycleLength(simBatchDir, selectedSim, figHandle, position)
<a name="l01015"></a>01015             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
<a name="l01016"></a>01016             import edu.stanford.covert.cell.sim.util.PlotUtil;
<a name="l01017"></a>01017             import edu.stanford.covert.cell.sim.util.SummaryLogger;
<a name="l01018"></a>01018             
<a name="l01019"></a>01019             if ~exist(&#39;simBatchDir&#39;, &#39;var&#39;) || isempty(simBatchDir)
<a name="l01020"></a>01020                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
<a name="l01021"></a>01021             end
<a name="l01022"></a>01022             
<a name="l01023"></a>01023             %% get data
<a name="l01024"></a>01024             data = load([SimulationDiskUtil.getSimulationBatchDir([simBatchDir filesep num2str(selectedSim)]) filesep &#39;population-CellCyclePhases.mat&#39;]);
<a name="l01025"></a>01025             stats = data.stats;
<a name="l01026"></a>01026             maxTime = data.maxTime;            
<a name="l01027"></a>01027             
<a name="l01028"></a>01028             edges = linspace(0, maxTime, 50);
<a name="l01029"></a>01029             n1 = histc(stats(:, SummaryLogger.SIM_STATUS_INDEX_CYTOKINESIS_TIME), edges);
<a name="l01030"></a>01030             n2 = histc(stats(:, SummaryLogger.SIM_STATUS_INDEX_REP_INIT_TIME), edges);
<a name="l01031"></a>01031             n3 = histc(stats(:, SummaryLogger.SIM_STATUS_INDEX_REP_TIME) - stats(:, SummaryLogger.SIM_STATUS_INDEX_REP_INIT_TIME), edges);
<a name="l01032"></a>01032             n4 = histc(stats(:, SummaryLogger.SIM_STATUS_INDEX_CYTOKINESIS_TIME) - stats(:, SummaryLogger.SIM_STATUS_INDEX_REP_TIME), edges);
<a name="l01033"></a>01033             
<a name="l01034"></a>01034             lowIdx = find(n1 &gt; 0, 1, &#39;first&#39;) - 2;
<a name="l01035"></a>01035             highIdx = find(n1 &gt; 0, 1, &#39;last&#39;) + 2;
<a name="l01036"></a>01036             n1([1:lowIdx highIdx:end]) = NaN;
<a name="l01037"></a>01037             lowIdx = find(n2 &gt; 0, 1, &#39;first&#39;) - 2;
<a name="l01038"></a>01038             highIdx = find(n2 &gt; 0, 1, &#39;last&#39;) + 2;
<a name="l01039"></a>01039             n2([1:lowIdx highIdx:end]) = NaN;
<a name="l01040"></a>01040             lowIdx = find(n3 &gt; 0, 1, &#39;first&#39;) - 2;
<a name="l01041"></a>01041             highIdx = find(n3 &gt; 0, 1, &#39;last&#39;) + 2;
<a name="l01042"></a>01042             n3([1:lowIdx highIdx:end]) = NaN;
<a name="l01043"></a>01043             lowIdx = find(n4 &gt; 0, 1, &#39;first&#39;) - 2;
<a name="l01044"></a>01044             highIdx = find(n4 &gt; 0, 1, &#39;last&#39;) + 2;
<a name="l01045"></a>01045             n4([1:lowIdx highIdx:end]) = NaN;
<a name="l01046"></a>01046  
<a name="l01047"></a>01047             %% plot data
<a name="l01048"></a>01048             if nargin &gt;= 4
<a name="l01049"></a>01049                 axesHandle = subplot(&#39;position&#39;, position, &#39;parent&#39;, figHandle);
<a name="l01050"></a>01050             else
<a name="l01051"></a>01051                 axesHandle = PlotUtil.newAxesHandle();
<a name="l01052"></a>01052             end
<a name="l01053"></a>01053             
<a name="l01054"></a>01054             n1 = smooth(n1, &#39;lowess&#39;, 13) / size(stats, 1) * 100;
<a name="l01055"></a>01055             n2 = smooth(n2, &#39;lowess&#39;, 13) / size(stats, 1) * 100; 
<a name="l01056"></a>01056             n3 = smooth(n3, &#39;lowess&#39;, 13) / size(stats, 1) * 100;
<a name="l01057"></a>01057             n4 = smooth(n4, &#39;lowess&#39;, 13) / size(stats, 1) * 100;
<a name="l01058"></a>01058             h = plot(axesHandle, edges / 3600, [n1 n2 n3 n4], &#39;LineWidth&#39;, 1);
<a name="l01059"></a>01059             set(h(2), &#39;Color&#39;, &#39;g&#39;);
<a name="l01060"></a>01060             
<a name="l01061"></a>01061             ylim(axesHandle, [0 max(max([n1 n2 n3 n4]))]);
<a name="l01062"></a>01062             set(axesHandle, &#39;XTick&#39;, 0:5:10);
<a name="l01063"></a>01063             set(axesHandle, &#39;YTick&#39;, 0:10:40);
<a name="l01064"></a>01064             set(axesHandle, &#39;Box&#39;, &#39;off&#39;);
<a name="l01065"></a>01065             set(axesHandle, &#39;TickDir&#39;, &#39;out&#39;);
<a name="l01066"></a>01066             set(axesHandle, &#39;FontSize&#39;, 5);
<a name="l01067"></a>01067             
<a name="l01068"></a>01068             xlabel(axesHandle, &#39;Duration (h)&#39;, &#39;FontSize&#39;, 7);
<a name="l01069"></a>01069             ylabel(axesHandle, &#39;% Cells&#39;, &#39;FontSize&#39;, 7);
<a name="l01070"></a>01070             xlabelPos = get(get(axesHandle, &#39;xlabel&#39;), &#39;position&#39;);
<a name="l01071"></a>01071             ylabelPos = get(get(axesHandle, &#39;ylabel&#39;), &#39;position&#39;);
<a name="l01072"></a>01072             set(get(axesHandle, &#39;xlabel&#39;), &#39;position&#39;, [xlabelPos(1) -3.7 xlabelPos(3:end)]);
<a name="l01073"></a>01073             set(get(axesHandle, &#39;ylabel&#39;), &#39;position&#39;, [-1.05 ylabelPos(2:end)]);
<a name="l01074"></a>01074             
<a name="l01075"></a>01075             tick2text(axesHandle, &#39;axis&#39;, &#39;xy&#39;, &#39;xtickoffset&#39;, 0.04, &#39;ytickoffset&#39;, 0.015);
<a name="l01076"></a>01076             xTicks = getappdata(axesHandle, &#39;XTickText&#39;);
<a name="l01077"></a>01077             yTicks = getappdata(axesHandle, &#39;YTickText&#39;);
<a name="l01078"></a>01078             set(xTicks, &#39;FontSize&#39;, 5);
<a name="l01079"></a>01079             set(yTicks, &#39;FontSize&#39;, 5, &#39;HorizontalAlign&#39;, &#39;right&#39;);
<a name="l01080"></a>01080             delete(xTicks(2));
<a name="l01081"></a>01081             delete(yTicks(2:2:end));
<a name="l01082"></a>01082             
<a name="l01083"></a>01083             legendHandle = legend(h, {<span class="stringliteral">&#39;Cell cycle&#39;</span>, <span class="stringliteral">&#39;Replication initiation&#39;</span>, <span class="stringliteral">&#39;Replication&#39;</span>, <span class="stringliteral">&#39;Cytokinesis&#39;</span>}, <span class="stringliteral">&#39;Location&#39;</span>, <span class="stringliteral">&#39;NorthEast&#39;</span>, <span class="stringliteral">&#39;FontSize&#39;</span>, 7);
<a name="l01084"></a>01084             legend(axesHandle, <span class="stringliteral">&#39;boxoff&#39;</span>);
<a name="l01085"></a>01085             legendPos = <span class="keyword">get</span>(legendHandle, <span class="stringliteral">&#39;Position&#39;</span>);            
<a name="l01086"></a>01086             <span class="keyword">set</span>(legendHandle, <span class="stringliteral">&#39;Position&#39;</span>, [0.395 0.42 legendPos(3:end)])
<a name="l01087"></a>01087             <span class="keyword">set</span>(legendHandle, <span class="stringliteral">&#39;units&#39;</span>, <span class="stringliteral">&#39;pixels&#39;</span>);
<a name="l01088"></a>01088             legendPos = <span class="keyword">get</span>(legendHandle, <span class="stringliteral">&#39;Position&#39;</span>);
<a name="l01089"></a>01089             <span class="keyword">set</span>(legendHandle, <span class="stringliteral">&#39;Position&#39;</span>, [legendPos(1:3) 40]);
<a name="l01090"></a>01090             legendChildren = <span class="keyword">get</span>(legendHandle, <span class="stringliteral">&#39;children&#39;</span>);
<a name="l01091"></a>01091             <span class="keyword">set</span>(legendChildren(1:3:end), <span class="stringliteral">&#39;visible&#39;</span>, <span class="stringliteral">&#39;off&#39;</span>);
<a name="l01092"></a>01092             <span class="keywordflow">for</span> i = 1:3:numel(legendChildren)
<a name="l01093"></a>01093                 set(legendChildren(i+1), &#39;xdata&#39;, [0.03 0.08])
<a name="l01094"></a>01094                 
<a name="l01095"></a>01095                 pos = get(legendChildren(i+2), &#39;Position&#39;);
<a name="l01096"></a>01096                 set(legendChildren(i+2), &#39;Position&#39;, [0.11 pos(2) 0])
<a name="l01097"></a>01097             end
<a name="l01098"></a>01098         end
<a name="l01099"></a>01099         
<a name="l01100"></a>01100         function replicationInitiationVsReplicationDuration(simBatchDir, selectedSim, figHandle, position)
<a name="l01101"></a>01101             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
<a name="l01102"></a>01102             import edu.stanford.covert.cell.sim.util.PlotUtil;
<a name="l01103"></a>01103             import edu.stanford.covert.cell.sim.util.SummaryLogger;
<a name="l01104"></a>01104             
<a name="l01105"></a>01105             if ~exist(&#39;simBatchDir&#39;, &#39;var&#39;) || isempty(simBatchDir)
<a name="l01106"></a>01106                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
<a name="l01107"></a>01107             end
<a name="l01108"></a>01108             
<a name="l01109"></a>01109             %% get data
<a name="l01110"></a>01110             data = load([SimulationDiskUtil.getSimulationBatchDir([simBatchDir filesep num2str(selectedSim)]) filesep &#39;population-CellCyclePhases.mat&#39;]);
<a name="l01111"></a>01111             repInitDur = data.stats(:, SummaryLogger.SIM_STATUS_INDEX_REP_INIT_TIME) / 3600;
<a name="l01112"></a>01112             repDur = (data.stats(:, SummaryLogger.SIM_STATUS_INDEX_REP_TIME) - data.stats(:, SummaryLogger.SIM_STATUS_INDEX_REP_INIT_TIME)) / 3600;            
<a name="l01113"></a>01113             
<a name="l01114"></a>01114             %% plot data
<a name="l01115"></a>01115             if nargin &gt;= 4
<a name="l01116"></a>01116                 axesHandle = subplot(&#39;position&#39;, position, &#39;parent&#39;, figHandle);
<a name="l01117"></a>01117             else
<a name="l01118"></a>01118                 axesHandle = PlotUtil.newAxesHandle();
<a name="l01119"></a>01119             end
<a name="l01120"></a>01120             
<a name="l01121"></a>01121             plot(axesHandle, repInitDur, repDur, &#39;b.&#39;)
<a name="l01122"></a>01122             
<a name="l01123"></a>01123             set(axesHandle, &#39;XTick&#39;, 0:5:10);
<a name="l01124"></a>01124             set(axesHandle, &#39;YTick&#39;, 0:4:8);
<a name="l01125"></a>01125             set(axesHandle, &#39;Box&#39;, &#39;off&#39;);
<a name="l01126"></a>01126             set(axesHandle, &#39;TickDir&#39;, &#39;out&#39;);
<a name="l01127"></a>01127             set(axesHandle, &#39;FontSize&#39;, 5);
<a name="l01128"></a>01128             
<a name="l01129"></a>01129             xlabel(axesHandle, &#39;<a class="code" href="class_replication.html">Replication</a> initiation duration (h)&#39;, &#39;FontSize&#39;, 7);
<a name="l01130"></a>01130             ylabel(axesHandle, {<span class="stringliteral">&#39;Replication&#39;</span> <span class="stringliteral">&#39;duration (h)&#39;</span>}, <span class="stringliteral">&#39;FontSize&#39;</span>, 7);
<a name="l01131"></a>01131             xlabelPos = <span class="keyword">get</span>(<span class="keyword">get</span>(axesHandle, <span class="stringliteral">&#39;xlabel&#39;</span>), <span class="stringliteral">&#39;position&#39;</span>);
<a name="l01132"></a>01132             ylabelPos = <span class="keyword">get</span>(<span class="keyword">get</span>(axesHandle, <span class="stringliteral">&#39;ylabel&#39;</span>), <span class="stringliteral">&#39;position&#39;</span>);
<a name="l01133"></a>01133             <span class="keyword">set</span>(<span class="keyword">get</span>(axesHandle, <span class="stringliteral">&#39;xlabel&#39;</span>), <span class="stringliteral">&#39;position&#39;</span>, [xlabelPos(1) -0.7 xlabelPos(3:end)]);
<a name="l01134"></a>01134             <span class="keyword">set</span>(<span class="keyword">get</span>(axesHandle, <span class="stringliteral">&#39;ylabel&#39;</span>), <span class="stringliteral">&#39;position&#39;</span>, [-0.4 4 ylabelPos(3:end)]);
<a name="l01135"></a>01135             
<a name="l01136"></a>01136             tick2text(axesHandle, <span class="stringliteral">&#39;axis&#39;</span>, <span class="stringliteral">&#39;xy&#39;</span>, <span class="stringliteral">&#39;xtickoffset&#39;</span>, 0.04, <span class="stringliteral">&#39;ytickoffset&#39;</span>, 0.015);
<a name="l01137"></a>01137             xTicks = getappdata(axesHandle, <span class="stringliteral">&#39;XTickText&#39;</span>);
<a name="l01138"></a>01138             yTicks = getappdata(axesHandle, <span class="stringliteral">&#39;YTickText&#39;</span>);
<a name="l01139"></a>01139             <span class="keyword">set</span>(xTicks, <span class="stringliteral">&#39;FontSize&#39;</span>, 5);
<a name="l01140"></a>01140             <span class="keyword">set</span>(yTicks, <span class="stringliteral">&#39;FontSize&#39;</span>, 5, <span class="stringliteral">&#39;HorizontalAlign&#39;</span>, <span class="stringliteral">&#39;right&#39;</span>);
<a name="l01141"></a>01141             <span class="keyword">delete</span>(xTicks(2));
<a name="l01142"></a>01142             <span class="keyword">delete</span>(yTicks(2:2:end));
<a name="l01143"></a>01143         end
<a name="l01144"></a>01144         
<a name="l01145"></a>01145         <span class="keyword">function</span> replicationInitiationVsReplicationDurationFit(simBatchDir, selectedSim, figHandle, position)
<a name="l01146"></a>01146             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
<a name="l01147"></a>01147             import edu.stanford.covert.cell.sim.util.PlotUtil;
<a name="l01148"></a>01148             import edu.stanford.covert.cell.sim.util.SummaryLogger;
<a name="l01149"></a>01149             import edu.stanford.covert.util.<a class="code" href="class_computation_util.html">ComputationUtil</a>;
<a name="l01150"></a>01150             
<a name="l01151"></a>01151             if ~exist(&#39;simBatchDir&#39;, &#39;var&#39;) || isempty(simBatchDir)
<a name="l01152"></a>01152                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
<a name="l01153"></a>01153             end
<a name="l01154"></a>01154             
<a name="l01155"></a>01155             %% get data
<a name="l01156"></a>01156             data = load([SimulationDiskUtil.getSimulationBatchDir([simBatchDir filesep num2str(selectedSim)]) filesep &#39;population-CellCyclePhases.mat&#39;]);
<a name="l01157"></a>01157             repInitDur = data.stats(:, SummaryLogger.SIM_STATUS_INDEX_REP_INIT_TIME) / 3600;
<a name="l01158"></a>01158             repDur = (data.stats(:, SummaryLogger.SIM_STATUS_INDEX_REP_TIME) - data.stats(:, SummaryLogger.SIM_STATUS_INDEX_REP_INIT_TIME)) / 3600;
<a name="l01159"></a>01159             
<a name="l01160"></a>01160             %% fit
<a name="l01161"></a>01161             [a, b, cutX, gof] = ComputationUtil.bilinearFit(repInitDur, repDur);            
<a name="l01162"></a>01162             
<a name="l01163"></a>01163             %% plot
<a name="l01164"></a>01164             if nargin &gt;= 4
<a name="l01165"></a>01165                 axesHandle = subplot(&#39;position&#39;, position, &#39;parent&#39;, figHandle);
<a name="l01166"></a>01166             else
<a name="l01167"></a>01167                 axesHandle = PlotUtil.newAxesHandle();
<a name="l01168"></a>01168             end
<a name="l01169"></a>01169             hold(axesHandle, &#39;on&#39;);
<a name="l01170"></a>01170             x1 = [min(repInitDur) cutX];
<a name="l01171"></a>01171             x2 = [cutX max(repInitDur)];
<a name="l01172"></a>01172             h = [
<a name="l01173"></a>01173                 plot(axesHandle, repInitDur, repDur, &#39;b.&#39;)
<a name="l01174"></a>01174                 plot(axesHandle, x1, a(1) + b(1) * x1, &#39;r&#39;)
<a name="l01175"></a>01175                 plot(axesHandle, x2, a(2) + b(2) * x2, &#39;r&#39;)
<a name="l01176"></a>01176                 ];
<a name="l01177"></a>01177             legend(h(1:2), {<span class="stringliteral">&#39;Data&#39;</span>, <span class="stringliteral">&#39;Fit&#39;</span>}, <span class="stringliteral">&#39;Location&#39;</span>, <span class="stringliteral">&#39;NorthEast&#39;</span>);
<a name="l01178"></a>01178             xlabel(axesHandle, <span class="stringliteral">&#39;Replication Initiation Duration (h)&#39;</span>)
<a name="l01179"></a>01179             ylabel(axesHandle, &#39;<a class="code" href="class_replication.html">Replication</a> Duration (h)&#39;)
<a name="l01180"></a>01180             
<a name="l01181"></a>01181             text(10, 6.5, sprintf(&#39;R^2_{adj} = %0.2f<span class="stringliteral">&#39;, gof.adjrsquare(1)), &#39;</span>Parent<span class="stringliteral">&#39;, axesHandle);</span>
<a name="l01182"></a>01182 <span class="stringliteral">            text(0.7, 4, sprintf(&#39;</span>R^2_{adj} = %0.2f<span class="stringliteral">&#39;, gof.adjrsquare(2)), &#39;</span>Parent<span class="stringliteral">&#39;, axesHandle);</span>
<a name="l01183"></a>01183 <span class="stringliteral">            text(8, 3, sprintf(&#39;</span>R^2_{adj} = %0.2f<span class="stringliteral">&#39;, gof.adjrsquare(3)), &#39;</span>Parent<span class="stringliteral">&#39;, axesHandle);</span>
<a name="l01184"></a>01184 <span class="stringliteral">        end</span>
<a name="l01185"></a>01185 <span class="stringliteral">        </span>
<a name="l01186"></a>01186 <span class="stringliteral">        function proteinBursting(simBatchDir, selectedSim, outDirectory, figHandle, position)</span>
<a name="l01187"></a>01187 <span class="stringliteral">            import edu.stanford.covert.cell.sim.util.MacromoleculeUtil;</span>
<a name="l01188"></a>01188 <span class="stringliteral">            import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;</span>
<a name="l01189"></a>01189 <span class="stringliteral">            import edu.stanford.covert.cell.sim.util.SimulationEnsemble;</span>
<a name="l01190"></a>01190 <span class="stringliteral">            import edu.stanford.covert.cell.sim.util.PlotUtil;</span>
<a name="l01191"></a>01191 <span class="stringliteral">            </span>
<a name="l01192"></a>01192 <span class="stringliteral">            if ~exist(&#39;</span>simBatchDir<span class="stringliteral">&#39;, &#39;</span>var<span class="stringliteral">&#39;) || isempty(simBatchDir)</span>
<a name="l01193"></a>01193 <span class="stringliteral">                simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();</span>
<a name="l01194"></a>01194 <span class="stringliteral">            end</span>
<a name="l01195"></a>01195 <span class="stringliteral">            </span>
<a name="l01196"></a>01196 <span class="stringliteral">            %% get data</span>
<a name="l01197"></a>01197 <span class="stringliteral">            if exist([outDirectory &#39;</span>proteinBursting.mat<span class="stringliteral">&#39;], &#39;</span>file<span class="stringliteral">&#39;)</span>
<a name="l01198"></a>01198 <span class="stringliteral">                load([outDirectory &#39;</span>proteinBursting.mat<span class="stringliteral">&#39;]);</span>
<a name="l01199"></a>01199 <span class="stringliteral">            else</span>
<a name="l01200"></a>01200 <span class="stringliteral">                [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep num2str(selectedSim)]);</span>
<a name="l01201"></a>01201 <span class="stringliteral">                </span>
<a name="l01202"></a>01202 <span class="stringliteral">                comp = sim.compartment;</span>
<a name="l01203"></a>01203 <span class="stringliteral">                rna = sim.state(&#39;</span>Rna<span class="stringliteral">&#39;);</span>
<a name="l01204"></a>01204 <span class="stringliteral">                pm = sim.state(&#39;</span><a class="code" href="class_protein_monomer.html">ProteinMonomer</a><span class="stringliteral">&#39;);</span>
<a name="l01205"></a>01205 <span class="stringliteral">                </span>
<a name="l01206"></a>01206 <span class="stringliteral">                [~, ~, ~, ~, ~, ...</span>
<a name="l01207"></a>01207 <span class="stringliteral">                ~, ~, ~, ~, ...</span>
<a name="l01208"></a>01208 <span class="stringliteral">                ~, ~, ~, ~, ...</span>
<a name="l01209"></a>01209 <span class="stringliteral">                matureRnaIdxs, ~, ~, ~, ...</span>
<a name="l01210"></a>01210 <span class="stringliteral">                monomerIdxs, ~, ~, ~, ...</span>
<a name="l01211"></a>01211 <span class="stringliteral">                complexIdxs, ~, ~, ~] = ...</span>
<a name="l01212"></a>01212 <span class="stringliteral">                MacromoleculeUtil.getMacromoleculeIndexsIDsNames(&#39;</span>MG_218<span class="stringliteral">&#39;, sim);</span>
<a name="l01213"></a>01213 <span class="stringliteral"></span>
<a name="l01214"></a>01214 <span class="stringliteral">                stateNames = {</span>
<a name="l01215"></a>01215 <span class="stringliteral">                    &#39;</span>Rna<span class="stringliteral">&#39;               &#39;</span>counts<span class="stringliteral">&#39; rna.matureIndexs(matureRnaIdxs)    comp.cytosolIndexs</span>
<a name="l01216"></a>01216 <span class="stringliteral">                    &#39;</span><a class="code" href="class_protein_monomer.html">ProteinMonomer</a><span class="stringliteral">&#39;    &#39;</span>counts<span class="stringliteral">&#39; pm.matureIndexs(monomerIdxs)       &#39;</span>-sum&#39;
<a name="l01217"></a>01217 
<a name="l01218"></a>01218                     };
<a name="l01219"></a>01219                 states = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, <span class="stringliteral">&#39;extract&#39;</span>, selectedSim);
<a name="l01220"></a>01220             
<a name="l01221"></a>01221 
<a name="l01222"></a>01222                 enzymeRNA = permute(states.Rna.counts, [1 3 2]);
<a name="l01223"></a>01223                 enzymeMonomer = permute(states.ProteinMonomer.counts, [1 3 2]);
<a name="l01224"></a>01224                 time = (1:numel(enzymeRNA)) / 3600;
<a name="l01225"></a>01225                 
<a name="l01226"></a>01226                 save([outDirectory <span class="stringliteral">&#39;proteinBursting.mat&#39;</span>], ...
<a name="l01227"></a>01227                     <span class="stringliteral">&#39;time&#39;</span>, <span class="stringliteral">&#39;enzymeRNA&#39;</span>, <span class="stringliteral">&#39;enzymeMonomer&#39;</span>);
<a name="l01228"></a>01228             end
<a name="l01229"></a>01229             
<a name="l01230"></a>01230             
<a name="l01231"></a>01231             %% plot
<a name="l01232"></a>01232             options = <span class="keyword">struct</span>;
<a name="l01233"></a>01233             <span class="keywordflow">if</span> nargin &gt;= 4
<a name="l01234"></a>01234                 options.position = position;
<a name="l01235"></a>01235             <span class="keywordflow">else</span>
<a name="l01236"></a>01236                 [~, figHandle] = edu.stanford.covert.cell.sim.util.PlotUtil.newAxesHandle();
<a name="l01237"></a>01237                 clf(figHandle);
<a name="l01238"></a>01238             end
<a name="l01239"></a>01239             
<a name="l01240"></a>01240             options.xdata = time;
<a name="l01241"></a>01241             options.ydata = {
<a name="l01242"></a>01242                 enzymeMonomer
<a name="l01243"></a>01243                 enzymeRNA
<a name="l01244"></a>01244                 };
<a name="l01245"></a>01245             options.ylabelStr = {
<a name="l01246"></a>01246                 {<span class="stringliteral">&#39;Protein (cnt)&#39;</span>}
<a name="l01247"></a>01247                 {<span class="stringliteral">&#39;mRNA (cnt)&#39;</span>}
<a name="l01248"></a>01248                 };
<a name="l01249"></a>01249             options.offsetYAxes = -0.02;
<a name="l01250"></a>01250             [axesHandles, xAxisHandle, ~, offsetAxesHandles] = PlotUtil.multiElementPlot(figHandle, 8 * ones(2, 1), [0 time(end)], options);
<a name="l01251"></a>01251             <span class="keyword">set</span>(xAxisHandle, <span class="stringliteral">&#39;XTick&#39;</span>, 0:2:8, <span class="stringliteral">&#39;FontSize&#39;</span>, 5);
<a name="l01252"></a>01252             <span class="keyword">set</span>(<span class="keyword">get</span>(xAxisHandle, <span class="stringliteral">&#39;XLabel&#39;</span>), <span class="stringliteral">&#39;FontSize&#39;</span>, 7);
<a name="l01253"></a>01253             xlabelPos = <span class="keyword">get</span>(<span class="keyword">get</span>(xAxisHandle, <span class="stringliteral">&#39;xlabel&#39;</span>), <span class="stringliteral">&#39;position&#39;</span>);
<a name="l01254"></a>01254             <span class="keyword">set</span>(<span class="keyword">get</span>(xAxisHandle, <span class="stringliteral">&#39;xlabel&#39;</span>), <span class="stringliteral">&#39;position&#39;</span>, [xlabelPos(1) -151 xlabelPos(3:end)]);
<a name="l01255"></a>01255             
<a name="l01256"></a>01256             tick2text(xAxisHandle, <span class="stringliteral">&#39;axis&#39;</span>, <span class="charliteral">&#39;x&#39;</span>, <span class="stringliteral">&#39;xtickoffset&#39;</span>, 75);
<a name="l01257"></a>01257             xTicks = getappdata(xAxisHandle, <span class="stringliteral">&#39;XTickText&#39;</span>);
<a name="l01258"></a>01258             <span class="keyword">set</span>(xTicks, <span class="stringliteral">&#39;FontSize&#39;</span>, 5);
<a name="l01259"></a>01259             <span class="keyword">delete</span>(xTicks(2:4));
<a name="l01260"></a>01260             
<a name="l01261"></a>01261             <span class="keywordflow">for</span> i = 1:numel(options.ydata)
<a name="l01262"></a>01262                 set(get(offsetAxesHandles(i), &#39;ylabel&#39;), &#39;FontSize&#39;, 7)
<a name="l01263"></a>01263                 set(offsetAxesHandles(i), &#39;FontSize&#39;, 5)
<a name="l01264"></a>01264                 set(axesHandles(i), &#39;FontSize&#39;, 5)
<a name="l01265"></a>01265                 
<a name="l01266"></a>01266                 yTicks = getappdata(offsetAxesHandles(i), &#39;YTickText&#39;);
<a name="l01267"></a>01267                 if isempty(yTicks)
<a name="l01268"></a>01268                     tick2text(offsetAxesHandles(i), &#39;axis&#39;, &#39;y&#39;, &#39;ytickoffset&#39;, 0.14)
<a name="l01269"></a>01269                     yTicks = getappdata(offsetAxesHandles(i), &#39;YTickText&#39;);
<a name="l01270"></a>01270                 end
<a name="l01271"></a>01271                 
<a name="l01272"></a>01272                 set(yTicks, &#39;FontSize&#39;, 5);
<a name="l01273"></a>01273                 for j = 1:numel(yTicks)
<a name="l01274"></a>01274                     pos = get(yTicks(j), &#39;position&#39;);
<a name="l01275"></a>01275                     set(yTicks(j), &#39;position&#39;, [-0.18 pos(2:end)], &#39;horizontalalign&#39;, &#39;right&#39;);
<a name="l01276"></a>01276                 end
<a name="l01277"></a>01277                 
<a name="l01278"></a>01278                 ylabelPos = get(get(offsetAxesHandles(i), &#39;ylabel&#39;), &#39;position&#39;);
<a name="l01279"></a>01279                 set(get(offsetAxesHandles(i), &#39;ylabel&#39;), &#39;position&#39;, [-0.6 ylabelPos(2:end)], &#39;FontSize&#39;, 7, &#39;VerticalAlign&#39;, &#39;bottom&#39;);
<a name="l01280"></a>01280             end
<a name="l01281"></a>01281             
<a name="l01282"></a>01282             burstIdxs = find(diff(enzymeRNA) &gt; 0);
<a name="l01283"></a>01283             axesPos1 = get(axesHandles(1), &#39;position&#39;);
<a name="l01284"></a>01284             axesPos2 = get(axesHandles(2), &#39;position&#39;);            
<a name="l01285"></a>01285             monYLims = ylim(axesHandles(1));
<a name="l01286"></a>01286             rnaYLims = ylim(axesHandles(2));
<a name="l01287"></a>01287             for i = 1:numel(burstIdxs)
<a name="l01288"></a>01288                 y = (axesPos2(2):0.0025:(axesPos1(2) + axesPos1(4) * (enzymeMonomer(burstIdxs(i)) - monYLims(1)) / range(monYLims)));
<a name="l01289"></a>01289                 x = (axesPos2(1) + axesPos2(3) * time(burstIdxs(i))/time(end)) * [1 1];
<a name="l01290"></a>01290                 for j = 1:3:3 * floor(numel(y) / 3)
<a name="l01291"></a>01291                     %if any(...
<a name="l01292"></a>01292                     %        (y(j:j+1)-axesPos2(2))/axesPos2(4)*range(rnaYLims) + rnaYLims(1) &gt;= enzymeRNA(burstIdxs(i)) &amp; ...
<a name="l01293"></a>01293                     %        (y(j:j+1)-axesPos2(2))/axesPos2(4)*range(rnaYLims) + rnaYLims(1) &lt;= enzymeRNA(burstIdxs(i)+1))
<a name="l01294"></a>01294                     %    continue;
<a name="l01295"></a>01295                     %end
<a name="l01296"></a>01296                     annotation(figHandle, &#39;line&#39;, x, y(j:j+1), &#39;Color&#39;, [1 0 0]);
<a name="l01297"></a>01297                 end
<a name="l01298"></a>01298             end
<a name="l01299"></a>01299         end
<a name="l01300"></a>01300         
<a name="l01301"></a>01301         function geneDeletionPrediction(~, ~, ~, figHandle, position)
<a name="l01302"></a>01302             import edu.stanford.covert.cell.sim.analysis.SingleGeneDeletions;
<a name="l01303"></a>01303             import edu.stanford.covert.cell.sim.util.CachedSimulationObjectUtil;
<a name="l01304"></a>01304             import edu.stanford.covert.cell.sim.util.PlotUtil;
<a name="l01305"></a>01305             import edu.stanford.covert.cell.sim.util.PrintUtil;
<a name="l01306"></a>01306             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
<a name="l01307"></a>01307             
<a name="l01308"></a>01308             baseDir = [SimulationDiskUtil.getBaseDir() filesep &#39;singleGeneDeletions&#39; filesep];
<a name="l01309"></a>01309 
<a name="l01310"></a>01310             sim = CachedSimulationObjectUtil.load();
<a name="l01311"></a>01311 
<a name="l01312"></a>01312             g = sim.gene;
<a name="l01313"></a>01313             [isGeneImplemented, ~, geneImplementedProcesses] = SingleGeneDeletions.getGeneImplementations(sim);
<a name="l01314"></a>01314             expEss = ismember(g.essential, {<span class="charliteral">&#39;Y&#39;</span>, <span class="stringliteral">&#39;&#39;</span>});
<a name="l01315"></a>01315             load([baseDir <span class="stringliteral">&#39;geneClasses.mat&#39;</span>]);
<a name="l01316"></a>01316             modelEss = geneClasses ~= SingleGeneDeletions.NON_ESSENTIAL;
<a name="l01317"></a>01317             
<a name="l01318"></a>01318             %% fontsizes
<a name="l01319"></a>01319             fontSizeSubfigLabel = 8;
<a name="l01320"></a>01320             fontSizeLarge = 7;
<a name="l01321"></a>01321             fontSizeMed = 6;
<a name="l01322"></a>01322             fontSizeSmall = 5;
<a name="l01323"></a>01323             
<a name="l01324"></a>01324             figPos = <span class="keyword">get</span>(figHandle, <span class="stringliteral">&#39;Position&#39;</span>);
<a name="l01325"></a>01325             
<a name="l01326"></a>01326             axesHandle = subplot(<span class="stringliteral">&#39;Position&#39;</span>, position);
<a name="l01327"></a>01327             axis(axesHandle, <span class="stringliteral">&#39;equal&#39;</span>);
<a name="l01328"></a>01328             xlim(axesHandle, [-0.5 1.5]);
<a name="l01329"></a>01329             ylim(axesHandle, [-0.5 1.5]);
<a name="l01330"></a>01330             <span class="keyword">set</span>(axesHandle, <span class="stringliteral">&#39;Box&#39;</span>, <span class="stringliteral">&#39;on&#39;</span>);
<a name="l01331"></a>01331             line([0.5 0.5], [-0.5 1.5], <span class="stringliteral">&#39;Parent&#39;</span>, axesHandle, <span class="stringliteral">&#39;Color&#39;</span>, <span class="charliteral">&#39;k&#39;</span>)
<a name="l01332"></a>01332             line([-0.5 1.5], [0.5 0.5], &#39;Parent&#39;, axesHandle, &#39;Color&#39;, &#39;k&#39;)
<a name="l01333"></a>01333             set(axesHandle, &#39;XAxisLocation&#39;, &#39;top&#39;);
<a name="l01334"></a>01334             set(axesHandle, &#39;XDir&#39;,&#39;reverse&#39;);
<a name="l01335"></a>01335             
<a name="l01336"></a>01336             set(axesHandle, &#39;TickLength&#39;, [0 0]);
<a name="l01337"></a>01337             set(axesHandle, &#39;XTick&#39;, [0 1]);
<a name="l01338"></a>01338             set(axesHandle, &#39;YTick&#39;, [0 1]);
<a name="l01339"></a>01339             tick2text(axesHandle, &#39;axis&#39;, &#39;xy&#39;, &#39;ytickoffset&#39;, 0.06)
<a name="l01340"></a>01340             xTicks = getappdata(axesHandle, &#39;XTickText&#39;);
<a name="l01341"></a>01341             yTicks = getappdata(axesHandle, &#39;YTickText&#39;);
<a name="l01342"></a>01342             set(xTicks, ...
<a name="l01343"></a>01343                 &#39;FontSize&#39;, fontSizeMed, ...
<a name="l01344"></a>01344                 &#39;HorizontalAlignment&#39;, &#39;center&#39;, ...
<a name="l01345"></a>01345                 &#39;VerticalAlignment&#39;, &#39;middle&#39;);
<a name="l01346"></a>01346             set(xTicks(1), &#39;String&#39;, &#39;Non-ess&#39;, &#39;Position&#39;, [0 1.68 -1]);
<a name="l01347"></a>01347             set(xTicks(2), &#39;String&#39;, &#39;Essential&#39;, &#39;Position&#39;, [1 1.68 -1]);
<a name="l01348"></a>01348             set(yTicks, &#39;rotation&#39;, 90, ...
<a name="l01349"></a>01349                 &#39;FontSize&#39;, fontSizeMed, ...
<a name="l01350"></a>01350                 &#39;HorizontalAlignment&#39;, &#39;center&#39;, ...
<a name="l01351"></a>01351                 &#39;VerticalAlignment&#39;, &#39;middle&#39;);
<a name="l01352"></a>01352             set(yTicks(1), &#39;String&#39;, &#39;Non-ess&#39;, &#39;Position&#39;, [1.68 0 -1]);
<a name="l01353"></a>01353             set(yTicks(2), &#39;String&#39;, &#39;Essential&#39;, &#39;Position&#39;, [1.68 1 -1]);
<a name="l01354"></a>01354             
<a name="l01355"></a>01355             xlabel(axesHandle&#39;, &#39;Model&#39;);
<a name="l01356"></a>01356             ylabel(axesHandle, &#39;Experiment&#39;);
<a name="l01357"></a>01357             xlab = get(axesHandle, &#39;xLabel&#39;);
<a name="l01358"></a>01358             ylab = get(axesHandle, &#39;yLabel&#39;);
<a name="l01359"></a>01359             set(xlab, ...
<a name="l01360"></a>01360                 &#39;FontSize&#39;, fontSizeLarge, ...
<a name="l01361"></a>01361                 &#39;HorizontalAlignment&#39;, &#39;center&#39;, ...
<a name="l01362"></a>01362                 &#39;VerticalAlignment&#39;, &#39;middle&#39;, ...
<a name="l01363"></a>01363                 &#39;Position&#39;, [0.5 1.9 1]);
<a name="l01364"></a>01364             set(ylab, ...
<a name="l01365"></a>01365                 &#39;FontSize&#39;, fontSizeLarge, ...
<a name="l01366"></a>01366                 &#39;HorizontalAlignment&#39;, &#39;center&#39;, ...
<a name="l01367"></a>01367                 &#39;VerticalAlignment&#39;, &#39;middle&#39;, ...
<a name="l01368"></a>01368                 &#39;Position&#39;, [1.9 0.5 1]);
<a name="l01369"></a>01369             
<a name="l01370"></a>01370             data = zeros(2, 2);
<a name="l01371"></a>01371             data(1, 1) = sum(~modelEss &amp; ~expEss &amp; isGeneImplemented);
<a name="l01372"></a>01372             data(1, 2) = sum(~modelEss &amp;  expEss &amp; isGeneImplemented);
<a name="l01373"></a>01373             data(2, 1) = sum( modelEss &amp; ~expEss &amp; isGeneImplemented);
<a name="l01374"></a>01374             data(2, 2) = sum( modelEss &amp;  expEss &amp; isGeneImplemented);
<a name="l01375"></a>01375             for i = 0:1
<a name="l01376"></a>01376                 for j = 0:1
<a name="l01377"></a>01377                     text(i, j, num2str(data(i+1, j+1)), ...
<a name="l01378"></a>01378                         &#39;Parent&#39;, axesHandle, ...
<a name="l01379"></a>01379                         &#39;FontSize&#39;, fontSizeLarge, ...
<a name="l01380"></a>01380                         &#39;HorizontalAlignment&#39;, &#39;center&#39;, ...
<a name="l01381"></a>01381                         &#39;VerticalAlignment&#39;, &#39;middle&#39;);
<a name="l01382"></a>01382                 end
<a name="l01383"></a>01383             end
<a name="l01384"></a>01384             
<a name="l01385"></a>01385             text(1.45, -0.68, sprintf(&#39;Correct:&#39;), ...
<a name="l01386"></a>01386                 &#39;Parent&#39;, axesHandle, ...
<a name="l01387"></a>01387                 &#39;HorizontalAlignment&#39;, &#39;left&#39;, ...
<a name="l01388"></a>01388                 &#39;VerticalAlignment&#39;, &#39;middle&#39;, ...
<a name="l01389"></a>01389                 &#39;FontSize&#39;, fontSizeLarge);
<a name="l01390"></a>01390             text(1.45, -0.93, sprintf(&#39;Incorrect:&#39;), ...
<a name="l01391"></a>01391                 &#39;Parent&#39;, axesHandle, ...
<a name="l01392"></a>01392                 &#39;HorizontalAlignment&#39;, &#39;left&#39;, ...
<a name="l01393"></a>01393                 &#39;VerticalAlignment&#39;, &#39;middle&#39;, ...
<a name="l01394"></a>01394                 &#39;FontSize&#39;, fontSizeLarge);
<a name="l01395"></a>01395             text(-0.45, -0.68, sprintf(&#39;%d (%d%%)&#39;, sum(diag(data)), round(100 * sum(diag(data)) / sum(data(:)))), ...
<a name="l01396"></a>01396                 &#39;Parent&#39;, axesHandle, ...
<a name="l01397"></a>01397                 &#39;HorizontalAlignment&#39;, &#39;right&#39;, ...
<a name="l01398"></a>01398                 &#39;VerticalAlignment&#39;, &#39;middle&#39;, ...
<a name="l01399"></a>01399                 &#39;FontSize&#39;, fontSizeLarge);
<a name="l01400"></a>01400             text(-0.45, -0.93, sprintf(&#39;%d (%d%%)&#39;, sum(data(:)) - sum(diag(data)), round(100 * (1 - sum(diag(data)) / sum(data(:))))), ...
<a name="l01401"></a>01401                 &#39;Parent&#39;, axesHandle, ...
<a name="l01402"></a>01402                 &#39;HorizontalAlignment&#39;, &#39;right&#39;, ...
<a name="l01403"></a>01403                 &#39;VerticalAlignment&#39;, &#39;middle&#39;, ...
<a name="l01404"></a>01404                 &#39;FontSize&#39;, fontSizeLarge);
<a name="l01405"></a>01405         end
<a name="l01406"></a>01406         
<a name="l01407"></a>01407         function geneToPhenotype(simBatchDir, selectedSim, outDirectory, figHandle, position)
<a name="l01408"></a>01408             import edu.stanford.covert.cell.sim.analysis.<a class="code" href="class_single_cell.html">SingleCell</a>;
<a name="l01409"></a>01409             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
<a name="l01410"></a>01410             import edu.stanford.covert.cell.sim.util.PlotUtil;
<a name="l01411"></a>01411             
<a name="l01412"></a>01412             if ~exist(&#39;simBatchDir&#39;, &#39;var&#39;) || isempty(simBatchDir)
<a name="l01413"></a>01413                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
<a name="l01414"></a>01414             end
<a name="l01415"></a>01415             
<a name="l01416"></a>01416             %% get data
<a name="l01417"></a>01417             if exist([outDirectory &#39;geneToPhenotype.mat&#39;], &#39;file&#39;)
<a name="l01418"></a>01418                 load([outDirectory &#39;geneToPhenotype.mat&#39;]);
<a name="l01419"></a>01419             else
<a name="l01420"></a>01420                 [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep num2str(selectedSim)]);
<a name="l01421"></a>01421                 
<a name="l01422"></a>01422                 [time, ~, ~, ~, ~, ~, ~, growth, reactionFlux, ...
<a name="l01423"></a>01423                     enzymeComplex, enzymeMonomer, enzymeRNA, enzymeGene] = ...
<a name="l01424"></a>01424                     SingleCell.calcGrowthData(sim, simBatchDir, selectedSim, &#39;GalE&#39;);
<a name="l01425"></a>01425                 enzymeGene = permute(enzymeGene, [1 3 2]);
<a name="l01426"></a>01426                 enzymeRNA = permute(enzymeRNA, [1 3 2]);
<a name="l01427"></a>01427                 enzymeMonomer = permute(enzymeMonomer, [1 3 2]);
<a name="l01428"></a>01428                 enzymeComplex = permute(enzymeComplex, [1 3 2]);
<a name="l01429"></a>01429                 reactionFlux = permute(reactionFlux, [1 3 2]);
<a name="l01430"></a>01430                 growth = permute(growth, [1 3 2]);
<a name="l01431"></a>01431                 
<a name="l01432"></a>01432                 save([outDirectory &#39;geneToPhenotype.mat&#39;], ...
<a name="l01433"></a>01433                     &#39;time&#39;, &#39;enzymeGene&#39;, &#39;enzymeRNA&#39;, &#39;enzymeMonomer&#39;, &#39;enzymeComplex&#39;, &#39;reactionFlux&#39;, &#39;growth&#39;);
<a name="l01434"></a>01434             end
<a name="l01435"></a>01435             reactionFlux = -reactionFlux * 1000;
<a name="l01436"></a>01436             
<a name="l01437"></a>01437             %% plot
<a name="l01438"></a>01438             options = struct;
<a name="l01439"></a>01439             if nargin &gt;= 4
<a name="l01440"></a>01440                 options.position = position;
<a name="l01441"></a>01441             else
<a name="l01442"></a>01442                 [~, figHandle] = edu.stanford.covert.cell.sim.util.PlotUtil.newAxesHandle();
<a name="l01443"></a>01443                 clf(figHandle);
<a name="l01444"></a>01444             end
<a name="l01445"></a>01445             
<a name="l01446"></a>01446             options.xdata = time;
<a name="l01447"></a>01447             options.ydata = {
<a name="l01448"></a>01448                 growth
<a name="l01449"></a>01449                 reactionFlux
<a name="l01450"></a>01450                 enzymeComplex
<a name="l01451"></a>01451                 enzymeMonomer
<a name="l01452"></a>01452                 enzymeRNA
<a name="l01453"></a>01453                 enzymeGene
<a name="l01454"></a>01454                 };
<a name="l01455"></a>01455             options.ylabelStr = {
<a name="l01456"></a>01456                 {<span class="stringliteral">&#39;Growth&#39;</span> <span class="stringliteral">&#39;(fg/h)&#39;</span>}
<a name="l01457"></a>01457                 {<span class="stringliteral">&#39;Flux&#39;</span> <span class="stringliteral">&#39;(10^3 rxn/h)&#39;</span>}
<a name="l01458"></a>01458                 {<span class="stringliteral">&#39;Cpx&#39;</span> <span class="stringliteral">&#39;(cnt)&#39;</span>}
<a name="l01459"></a>01459                 {<span class="stringliteral">&#39;Mon&#39;</span> <span class="stringliteral">&#39;(cnt)&#39;</span>}
<a name="l01460"></a>01460                 {<span class="stringliteral">&#39;mRNA&#39;</span> <span class="stringliteral">&#39;(cnt)&#39;</span>}
<a name="l01461"></a>01461                 {<span class="stringliteral">&#39;Gene&#39;</span> <span class="stringliteral">&#39;(cnt)&#39;</span>}
<a name="l01462"></a>01462                 };
<a name="l01463"></a>01463             options.offsetYAxes = <span class="keyword">false</span>;
<a name="l01464"></a>01464             [axesHandles, xAxisHandle, ~, offsetAxesHandles] = PlotUtil.multiElementPlot(figHandle, 2 * ones(6, 1), time([1 end])<span class="stringliteral">&#39;, options);</span>
<a name="l01465"></a>01465 <span class="stringliteral">            set(xAxisHandle, &#39;</span>XTick<span class="stringliteral">&#39;, 0:2:8, &#39;</span>FontSize<span class="stringliteral">&#39;, 5);</span>
<a name="l01466"></a>01466 <span class="stringliteral">            set(get(xAxisHandle, &#39;</span>XLabel<span class="stringliteral">&#39;), &#39;</span>FontSize<span class="stringliteral">&#39;, 7);</span>
<a name="l01467"></a>01467 <span class="stringliteral">            xlabelPos = get(get(xAxisHandle, &#39;</span>xlabel<span class="stringliteral">&#39;), &#39;</span>position<span class="stringliteral">&#39;);</span>
<a name="l01468"></a>01468 <span class="stringliteral">            set(get(xAxisHandle, &#39;</span>xlabel<span class="stringliteral">&#39;), &#39;</span>position<span class="stringliteral">&#39;, [xlabelPos(1) -80 xlabelPos(3:end)]);</span>
<a name="l01469"></a>01469 <span class="stringliteral">            </span>
<a name="l01470"></a>01470 <span class="stringliteral">            tick2text(xAxisHandle, &#39;</span>axis<span class="stringliteral">&#39;, &#39;</span>x<span class="stringliteral">&#39;, &#39;</span>xtickoffset<span class="stringliteral">&#39;, 60);</span>
<a name="l01471"></a>01471 <span class="stringliteral">            xTicks = getappdata(xAxisHandle, &#39;</span>XTickText<span class="stringliteral">&#39;);</span>
<a name="l01472"></a>01472 <span class="stringliteral">            set(xTicks, &#39;</span>FontSize<span class="stringliteral">&#39;, 5);</span>
<a name="l01473"></a>01473 <span class="stringliteral">            delete(xTicks(2:4));</span>
<a name="l01474"></a>01474 <span class="stringliteral">            </span>
<a name="l01475"></a>01475 <span class="stringliteral">            for i = 1:numel(axesHandles)</span>
<a name="l01476"></a>01476 <span class="stringliteral">                set(get(axesHandles(i), &#39;</span>ylabel<span class="stringliteral">&#39;), &#39;</span>FontSize<span class="stringliteral">&#39;, 7)</span>
<a name="l01477"></a>01477 <span class="stringliteral">                set(axesHandles(i), &#39;</span>FontSize<span class="stringliteral">&#39;, 5)</span>
<a name="l01478"></a>01478 <span class="stringliteral">                </span>
<a name="l01479"></a>01479 <span class="stringliteral">                yTicks = getappdata(axesHandles(i), &#39;</span>YTickText<span class="stringliteral">&#39;);                </span>
<a name="l01480"></a>01480 <span class="stringliteral">                if isempty(yTicks)</span>
<a name="l01481"></a>01481 <span class="stringliteral">                    tick2text(axesHandles(i), &#39;</span>axis<span class="stringliteral">&#39;, &#39;</span>y<span class="stringliteral">&#39;, &#39;</span>ytickoffset<span class="stringliteral">&#39;, 0.06)</span>
<a name="l01482"></a>01482 <span class="stringliteral">                    yTicks = getappdata(axesHandles(i), &#39;</span>YTickText<span class="stringliteral">&#39;);                    </span>
<a name="l01483"></a>01483 <span class="stringliteral">                end</span>
<a name="l01484"></a>01484 <span class="stringliteral">                    </span>
<a name="l01485"></a>01485 <span class="stringliteral">                set(yTicks, &#39;</span>FontSize<span class="stringliteral">&#39;, 5);</span>
<a name="l01486"></a>01486 <span class="stringliteral">                for j = 1:numel(yTicks)</span>
<a name="l01487"></a>01487 <span class="stringliteral">                    pos = get(yTicks(j), &#39;</span>position<span class="stringliteral">&#39;);</span>
<a name="l01488"></a>01488 <span class="stringliteral">                    set(yTicks(j), &#39;</span>position<span class="stringliteral">&#39;, [-0.15 pos(2:end)], &#39;</span>horizontalalign<span class="stringliteral">&#39;, &#39;</span>right<span class="stringliteral">&#39;);</span>
<a name="l01489"></a>01489 <span class="stringliteral">                end</span>
<a name="l01490"></a>01490 <span class="stringliteral">                </span>
<a name="l01491"></a>01491 <span class="stringliteral">                ylabelPos = get(get(axesHandles(i), &#39;</span>ylabel<span class="stringliteral">&#39;), &#39;</span>position<span class="stringliteral">&#39;);</span>
<a name="l01492"></a>01492 <span class="stringliteral">                set(get(axesHandles(i), &#39;</span>ylabel<span class="stringliteral">&#39;), &#39;</span>position<span class="stringliteral">&#39;, [-0.8 ylabelPos(2:end)], &#39;</span>FontSize<span class="stringliteral">&#39;, 5, &#39;</span>VerticalAlign<span class="stringliteral">&#39;, &#39;</span>bottom<span class="stringliteral">&#39;);</span>
<a name="l01493"></a>01493 <span class="stringliteral">            end</span>
<a name="l01494"></a>01494 <span class="stringliteral">            </span>
<a name="l01495"></a>01495 <span class="stringliteral">            ylabelPos = get(get(axesHandles(1), &#39;</span>ylabel<span class="stringliteral">&#39;), &#39;</span>position<span class="stringliteral">&#39;);</span>
<a name="l01496"></a>01496 <span class="stringliteral">            set(get(axesHandles(1), &#39;</span>ylabel<span class="stringliteral">&#39;), &#39;</span>position<span class="stringliteral">&#39;, [ylabelPos(1) 1.65 ylabelPos(3)]);</span>
<a name="l01497"></a>01497 <span class="stringliteral">        end</span>
<a name="l01498"></a>01498 <span class="stringliteral">        </span>
<a name="l01499"></a>01499 <span class="stringliteral">        function spaceTimePlot(simBatchDir, selectedSim, outDirectory, figHandle, position)</span>
<a name="l01500"></a>01500 <span class="stringliteral">            import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;</span>
<a name="l01501"></a>01501 <span class="stringliteral">            import edu.stanford.covert.cell.sim.util.DiskLogger;</span>
<a name="l01502"></a>01502 <span class="stringliteral">            import edu.stanford.covert.cell.sim.util.PlotUtil;</span>
<a name="l01503"></a>01503 <span class="stringliteral">            import edu.stanford.covert.cell.sim.analysis.ChromosomeSpaceTimePlot;</span>
<a name="l01504"></a>01504 <span class="stringliteral">            </span>
<a name="l01505"></a>01505 <span class="stringliteral">            if ~exist(&#39;</span>simBatchDir<span class="stringliteral">&#39;, &#39;</span>var<span class="stringliteral">&#39;) || isempty(simBatchDir)</span>
<a name="l01506"></a>01506 <span class="stringliteral">                simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();</span>
<a name="l01507"></a>01507 <span class="stringliteral">            end</span>
<a name="l01508"></a>01508 <span class="stringliteral">            if ~exist(&#39;</span>selectedSim<span class="stringliteral">&#39;, &#39;</span>var<span class="stringliteral">&#39;) || isempty(selectedSim)</span>
<a name="l01509"></a>01509 <span class="stringliteral">                selectedSim = 1;</span>
<a name="l01510"></a>01510 <span class="stringliteral">            end</span>
<a name="l01511"></a>01511 <span class="stringliteral">            </span>
<a name="l01512"></a>01512 <span class="stringliteral">            [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep num2str(selectedSim)]);</span>
<a name="l01513"></a>01513 <span class="stringliteral">            c = sim.state(&#39;</span><a class="code" href="class_chromosome.html">Chromosome</a><span class="stringliteral">&#39;);</span>
<a name="l01514"></a>01514 <span class="stringliteral">            </span>
<a name="l01515"></a>01515 <span class="stringliteral">            %% get data</span>
<a name="l01516"></a>01516 <span class="stringliteral">            if exist([outDirectory &#39;</span>spaceTimePlot.mat<span class="stringliteral">&#39;], &#39;</span>file<span class="stringliteral">&#39;)</span>
<a name="l01517"></a>01517 <span class="stringliteral">                load([outDirectory &#39;</span>spaceTimePlot.mat<span class="stringliteral">&#39;])                </span>
<a name="l01518"></a>01518 <span class="stringliteral">            else</span>
<a name="l01519"></a>01519 <span class="stringliteral">                [axesHandle, figHandle] = edu.stanford.covert.cell.sim.util.PlotUtil.newAxesHandle();</span>
<a name="l01520"></a>01520 <span class="stringliteral">                [ensemble, states] = ChromosomeSpaceTimePlot.plotSpaceTime(...</span>
<a name="l01521"></a>01521 <span class="stringliteral">                    axesHandle, simBatchDir, selectedSim, [], false, [], false, true); %#ok&lt;*NASGU,*ASGLU&gt;</span>
<a name="l01522"></a>01522 <span class="stringliteral">                save([outDirectory &#39;</span>spaceTimePlot.mat<span class="stringliteral">&#39;], &#39;</span>ensemble<span class="stringliteral">&#39;, &#39;</span>states<span class="stringliteral">&#39;)</span>
<a name="l01523"></a>01523 <span class="stringliteral">                close(figHandle);</span>
<a name="l01524"></a>01524 <span class="stringliteral">            end</span>
<a name="l01525"></a>01525 <span class="stringliteral">            </span>
<a name="l01526"></a>01526 <span class="stringliteral">            %% space-time density plots</span>
<a name="l01527"></a>01527 <span class="stringliteral">            %create axes</span>
<a name="l01528"></a>01528 <span class="stringliteral">            if nargin &lt; 4</span>
<a name="l01529"></a>01529 <span class="stringliteral">                [~, figHandle] = edu.stanford.covert.cell.sim.util.PlotUtil.newAxesHandle();</span>
<a name="l01530"></a>01530 <span class="stringliteral">                position = [0.1 0.1 0.8 0.8];</span>
<a name="l01531"></a>01531 <span class="stringliteral">            end</span>
<a name="l01532"></a>01532 <span class="stringliteral">            </span>
<a name="l01533"></a>01533 <span class="stringliteral">            x = position(1);</span>
<a name="l01534"></a>01534 <span class="stringliteral">            y = position(2);</span>
<a name="l01535"></a>01535 <span class="stringliteral">            w = position(3);</span>
<a name="l01536"></a>01536 <span class="stringliteral">            h = position(4);</span>
<a name="l01537"></a>01537 <span class="stringliteral">            spcg = 0.08;</span>
<a name="l01538"></a>01538 <span class="stringliteral">            figW = 1.45 * (0.5 - x);</span>
<a name="l01539"></a>01539 <span class="stringliteral">            </span>
<a name="l01540"></a>01540 <span class="stringliteral">            axesHandles = [</span>
<a name="l01541"></a>01541 <span class="stringliteral">                subplot(&#39;</span>position<span class="stringliteral">&#39;, [x            y  figW  h], &#39;</span>parent<span class="stringliteral">&#39;, figHandle)</span>
<a name="l01542"></a>01542 <span class="stringliteral">                subplot(&#39;</span>position<span class="stringliteral">&#39;, [x+figW+spcg  y  w-figW-spcg  h], &#39;</span>parent<span class="stringliteral">&#39;, figHandle)</span>
<a name="l01543"></a>01543 <span class="stringliteral">                ];</span>
<a name="l01544"></a>01544 <span class="stringliteral">            </span>
<a name="l01545"></a>01545 <span class="stringliteral">            zoomTime = [1.34 1.41];</span>
<a name="l01546"></a>01546 <span class="stringliteral">            zoomPos = [267000 273000];</span>
<a name="l01547"></a>01547 <span class="stringliteral">            time = ensemble.stateData.time / 3600;</span>
<a name="l01548"></a>01548 <span class="stringliteral">            </span>
<a name="l01549"></a>01549 <span class="stringliteral">            zoomBoxHandle = [</span>
<a name="l01550"></a>01550 <span class="stringliteral">                annotation(figHandle, &#39;</span>line<span class="stringliteral">&#39;, x + [figW*zoomTime(2)/time(end) figW+spcg], y + [h*zoomPos(end)/c.sequenceLen h])</span>
<a name="l01551"></a>01551 <span class="stringliteral">                annotation(figHandle, &#39;</span>line<span class="stringliteral">&#39;, x + [figW*zoomTime(2)/time(end) figW+spcg], y + [h*zoomPos(1)/c.sequenceLen 0])</span>
<a name="l01552"></a>01552 <span class="stringliteral">                annotation(figHandle, &#39;</span>rectangle<span class="stringliteral">&#39;, [</span>
<a name="l01553"></a>01553 <span class="stringliteral">                    x + figW*zoomTime(1)/time(end) </span>
<a name="l01554"></a>01554 <span class="stringliteral">                    y + h*zoomPos(1)/c.sequenceLen</span>
<a name="l01555"></a>01555 <span class="stringliteral">                    figW*diff(zoomTime)/time(end) </span>
<a name="l01556"></a>01556 <span class="stringliteral">                    h*diff(zoomPos)/c.sequenceLen</span>
<a name="l01557"></a>01557 <span class="stringliteral">                    ]&#39;</span>)
<a name="l01558"></a>01558                 ];
<a name="l01559"></a>01559             <span class="keyword">set</span>(zoomBoxHandle, <span class="stringliteral">&#39;LineWidth&#39;</span>, 1);
<a name="l01560"></a>01560             
<a name="l01561"></a>01561             %% plot 1
<a name="l01562"></a>01562             axesHandle = axesHandles(1);
<a name="l01563"></a>01563             
<a name="l01564"></a>01564             %plot
<a name="l01565"></a>01565             
<a name="l01566"></a>01566             %TODO: Fix positioning
<a name="l01567"></a>01567             [~, ~, legendHandle] = ChromosomeSpaceTimePlot.plotSpaceTime(...
<a name="l01568"></a>01568                 axesHandle, simBatchDir, selectedSim, [], <span class="keyword">false</span>, [], ...
<a name="l01569"></a>01569                 <span class="keyword">false</span>, <span class="keyword">true</span>, ensemble, states, 2, 0.25, <span class="keyword">false</span>);
<a name="l01570"></a>01570             
<a name="l01571"></a>01571             %style
<a name="l01572"></a>01572             xlim(axesHandle, [0 time(end)]);
<a name="l01573"></a>01573             ylim(axesHandle, [1 c.sequenceLen]);
<a name="l01574"></a>01574             <span class="keyword">set</span>(axesHandle, <span class="stringliteral">&#39;TickDir&#39;</span>, <span class="stringliteral">&#39;out&#39;</span>, <span class="stringliteral">&#39;XTick&#39;</span>, 0:2:8, <span class="stringliteral">&#39;FontSize&#39;</span>, 5);
<a name="l01575"></a>01575             
<a name="l01576"></a>01576             tick2text(axesHandle, <span class="stringliteral">&#39;axis&#39;</span>, <span class="stringliteral">&#39;xy&#39;</span>, <span class="stringliteral">&#39;xtickoffset&#39;</span>, 0.035, <span class="stringliteral">&#39;ytickoffset&#39;</span>, 0.008);
<a name="l01577"></a>01577             xTicks = getappdata(axesHandle, <span class="stringliteral">&#39;XTickText&#39;</span>);
<a name="l01578"></a>01578             yTicks = getappdata(axesHandle, <span class="stringliteral">&#39;YTickText&#39;</span>);
<a name="l01579"></a>01579             <span class="keyword">set</span>(xTicks, <span class="stringliteral">&#39;FontSize&#39;</span>, 5);
<a name="l01580"></a>01580             <span class="keyword">set</span>(yTicks, <span class="stringliteral">&#39;FontSize&#39;</span>, 5, <span class="stringliteral">&#39;HorizontalAlign&#39;</span>, <span class="stringliteral">&#39;right&#39;</span>);
<a name="l01581"></a>01581             <span class="keyword">delete</span>(xTicks(2:4));
<a name="l01582"></a>01582             <span class="keyword">set</span>(yTicks([1 3]), <span class="stringliteral">&#39;String&#39;</span>, <span class="stringliteral">&#39;terC&#39;</span>);
<a name="l01583"></a>01583             <span class="keyword">set</span>(yTicks(2), <span class="stringliteral">&#39;String&#39;</span>, <span class="stringliteral">&#39;oriC&#39;</span>);
<a name="l01584"></a>01584             
<a name="l01585"></a>01585             xlabel(axesHandle, <span class="stringliteral">&#39;Time (h)&#39;</span>, <span class="stringliteral">&#39;FontSize&#39;</span>, 7);
<a name="l01586"></a>01586             ylabel(axesHandle, <span class="stringliteral">&#39;Position&#39;</span>, <span class="stringliteral">&#39;FontSize&#39;</span>, 7);
<a name="l01587"></a>01587             xLabelPos = <span class="keyword">get</span>(<span class="keyword">get</span>(axesHandle, <span class="stringliteral">&#39;xlabel&#39;</span>), <span class="stringliteral">&#39;position&#39;</span>);
<a name="l01588"></a>01588             yLabelPos = <span class="keyword">get</span>(<span class="keyword">get</span>(axesHandle, <span class="stringliteral">&#39;ylabel&#39;</span>), <span class="stringliteral">&#39;position&#39;</span>);
<a name="l01589"></a>01589             <span class="keyword">set</span>(<span class="keyword">get</span>(axesHandle, <span class="stringliteral">&#39;xlabel&#39;</span>), <span class="stringliteral">&#39;position&#39;</span>, [xLabelPos(1) -25000 xLabelPos(3)])
<a name="l01590"></a>01590             set(get(axesHandle, &#39;ylabel&#39;), &#39;position&#39;, [-0.4 yLabelPos(2:3)])
<a name="l01591"></a>01591   
<a name="l01592"></a>01592             %legend
<a name="l01593"></a>01593             legendChildren = get(legendHandle, &#39;children&#39;);
<a name="l01594"></a>01594             <span class="keyword">set</span>(legendHandle, <span class="stringliteral">&#39;FontSize&#39;</span>, 7);
<a name="l01595"></a>01595             legendPos = <span class="keyword">get</span>(legendHandle, <span class="stringliteral">&#39;Position&#39;</span>);
<a name="l01596"></a>01596             <span class="keyword">set</span>(legendHandle, <span class="stringliteral">&#39;Position&#39;</span>, [0.035 0.649 legendPos(3:4)]);
<a name="l01597"></a>01597             <span class="keyword">set</span>(legendHandle, <span class="stringliteral">&#39;units&#39;</span>, <span class="stringliteral">&#39;pixels&#39;</span>);
<a name="l01598"></a>01598             legendPos = <span class="keyword">get</span>(legendHandle, <span class="stringliteral">&#39;Position&#39;</span>);
<a name="l01599"></a>01599             <span class="keyword">set</span>(legendHandle, <span class="stringliteral">&#39;Position&#39;</span>, [legendPos(1:2) 106 30]);
<a name="l01600"></a>01600             <span class="keyword">set</span>(legendChildren(1:3:end), <span class="stringliteral">&#39;visible&#39;</span>, <span class="stringliteral">&#39;off&#39;</span>)
<a name="l01601"></a>01601             <span class="keyword">set</span>(legendChildren(2:3:end), <span class="stringliteral">&#39;visible&#39;</span>, <span class="stringliteral">&#39;on&#39;</span>, <span class="stringliteral">&#39;LineWidth&#39;</span>, 4)
<a name="l01602"></a>01602             <span class="keyword">set</span>(legendChildren(2:3:end), <span class="stringliteral">&#39;XData&#39;</span>, [0.07 0.18])
<a name="l01603"></a>01603             <span class="keywordflow">for</span> i = 3:3:numel(legendChildren)
<a name="l01604"></a>01604                 pos = <span class="keyword">get</span>(legendChildren(i), <span class="stringliteral">&#39;Position&#39;</span>);
<a name="l01605"></a>01605                 <span class="keyword">set</span>(legendChildren(i), <span class="stringliteral">&#39;Position&#39;</span>, [0.24 pos(2) 0])
<a name="l01606"></a>01606             end
<a name="l01607"></a>01607             <span class="keyword">set</span>(legendHandle, <span class="stringliteral">&#39;units&#39;</span>, <span class="stringliteral">&#39;pixels&#39;</span>);
<a name="l01608"></a>01608             legendPos = <span class="keyword">get</span>(legendHandle, <span class="stringliteral">&#39;position&#39;</span>);
<a name="l01609"></a>01609             <span class="keyword">set</span>(legendHandle, <span class="stringliteral">&#39;position&#39;</span>, [[1.4 1.42] .* legendPos(1:2) 0.57*legendPos(3) legendPos(4)]);
<a name="l01610"></a>01610             
<a name="l01611"></a>01611             %% plot 2
<a name="l01612"></a>01612             axesHandle = axesHandles(2);
<a name="l01613"></a>01613             
<a name="l01614"></a>01614             %plot
<a name="l01615"></a>01615             ChromosomeSpaceTimePlot.plotSpaceTime(axesHandle, simBatchDir, selectedSim, [], ...
<a name="l01616"></a>01616                 <span class="keyword">false</span>, [], <span class="keyword">false</span>, <span class="keyword">false</span>, ensemble, states, 0.5, 0.5, <span class="keyword">true</span>); %#ok&lt;*NODEF&gt;
<a name="l01617"></a>01617             
<a name="l01618"></a>01618             xlim(axesHandle, zoomTime);
<a name="l01619"></a>01619             ylim(axesHandle, zoomPos);
<a name="l01620"></a>01620             
<a name="l01621"></a>01621             %style
<a name="l01622"></a>01622             <span class="keyword">set</span>(axesHandle, <span class="stringliteral">&#39;TickDir&#39;</span>, <span class="stringliteral">&#39;out&#39;</span>, ...
<a name="l01623"></a>01623                 <span class="stringliteral">&#39;XTick&#39;</span>, linspace(zoomTime(1), zoomTime(2), 8), ...
<a name="l01624"></a>01624                 <span class="stringliteral">&#39;YTick&#39;</span>, linspace(zoomPos(1), zoomPos(2), 7)-38, ...
<a name="l01625"></a>01625                 &#39;FontSize&#39;, 5);
<a name="l01626"></a>01626             
<a name="l01627"></a>01627             tick2text(axesHandle, <span class="stringliteral">&#39;axis&#39;</span>, <span class="stringliteral">&#39;xy&#39;</span>, <span class="stringliteral">&#39;xtickoffset&#39;</span>, 0.035, <span class="stringliteral">&#39;ytickoffset&#39;</span>, 0.015);
<a name="l01628"></a>01628             xTicks = getappdata(axesHandle, <span class="stringliteral">&#39;XTickText&#39;</span>);
<a name="l01629"></a>01629             yTicks = getappdata(axesHandle, <span class="stringliteral">&#39;YTickText&#39;</span>);
<a name="l01630"></a>01630             <span class="keyword">set</span>(xTicks, <span class="stringliteral">&#39;FontSize&#39;</span>, 5);
<a name="l01631"></a>01631             <span class="keyword">set</span>(yTicks, <span class="stringliteral">&#39;FontSize&#39;</span>, 5, <span class="stringliteral">&#39;HorizontalAlign&#39;</span>, <span class="stringliteral">&#39;right&#39;</span>);
<a name="l01632"></a>01632             <span class="keyword">set</span>(yTicks(2), <span class="stringliteral">&#39;String&#39;</span>, <span class="stringliteral">&#39;558000&#39;</span>)
<a name="l01633"></a>01633             <span class="keyword">set</span>(yTicks(4), <span class="stringliteral">&#39;String&#39;</span>, <span class="stringliteral">&#39;560000&#39;</span>)
<a name="l01634"></a>01634             <span class="keyword">set</span>(yTicks(6), <span class="stringliteral">&#39;String&#39;</span>, <span class="stringliteral">&#39;562000&#39;</span>)
<a name="l01635"></a>01635             <span class="keyword">delete</span>(xTicks(setdiff(1:end, [2 7])));
<a name="l01636"></a>01636             <span class="keyword">delete</span>(yTicks(1:2:end));
<a name="l01637"></a>01637             
<a name="l01638"></a>01638             xlabel(axesHandle, <span class="stringliteral">&#39;Time (h)&#39;</span>, <span class="stringliteral">&#39;FontSize&#39;</span>, 7);
<a name="l01639"></a>01639             ylabel(axesHandle, <span class="stringliteral">&#39;Position (nt)&#39;</span>, <span class="stringliteral">&#39;FontSize&#39;</span>, 7);
<a name="l01640"></a>01640             xLabelPos = <span class="keyword">get</span>(<span class="keyword">get</span>(axesHandle, <span class="stringliteral">&#39;xlabel&#39;</span>), <span class="stringliteral">&#39;position&#39;</span>);
<a name="l01641"></a>01641             yLabelPos = <span class="keyword">get</span>(<span class="keyword">get</span>(axesHandle, <span class="stringliteral">&#39;ylabel&#39;</span>), <span class="stringliteral">&#39;position&#39;</span>);
<a name="l01642"></a>01642             <span class="keyword">set</span>(<span class="keyword">get</span>(axesHandle, <span class="stringliteral">&#39;xlabel&#39;</span>), <span class="stringliteral">&#39;position&#39;</span>, [xLabelPos(1) 266750 xLabelPos(3)])
<a name="l01643"></a>01643             <span class="keyword">set</span>(<span class="keyword">get</span>(axesHandle, <span class="stringliteral">&#39;yLabel&#39;</span>), <span class="stringliteral">&#39;position&#39;</span>, [1.322 yLabelPos(2:end)])
<a name="l01644"></a>01644             
<a name="l01645"></a>01645             figPos = <span class="keyword">get</span>(figHandle, <span class="stringliteral">&#39;PaperSize&#39;</span>);
<a name="l01646"></a>01646             figW = figPos(1);
<a name="l01647"></a>01647             figH = figPos(2);
<a name="l01648"></a>01648             axisPos = <span class="keyword">get</span>(axesHandle, <span class="stringliteral">&#39;position&#39;</span>);
<a name="l01649"></a>01649             x = axisPos(1);
<a name="l01650"></a>01650             y = axisPos(2);
<a name="l01651"></a>01651             w = axisPos(3);
<a name="l01652"></a>01652             h = axisPos(4);
<a name="l01653"></a>01653             annotation(figHandle, <span class="stringliteral">&#39;ellipse&#39;</span>, [x+0.595*w  y+0.645*h  0.01 0.01*figW/figH], <span class="stringliteral">&#39;Color&#39;</span>, <span class="charliteral">&#39;k&#39;</span>)
<a name="l01654"></a>01654             annotation(figHandle, &#39;ellipse&#39;, [x+0.775*w  y+0.245*h  0.01 0.01*figW/figH], &#39;Color&#39;, &#39;k&#39;)
<a name="l01655"></a>01655             annotation(figHandle, &#39;ellipse&#39;, [x+0.48*w  y+0.875*h  0.01 0.01*figW/figH], &#39;Color&#39;, &#39;k&#39;)
<a name="l01656"></a>01656             annotation(figHandle, &#39;ellipse&#39;, [x+0.645*w  y+0.84*h  0.01 0.01*figW/figH], &#39;Color&#39;, &#39;k&#39;)
<a name="l01657"></a>01657             annotation(figHandle, &#39;ellipse&#39;, [x+0.655*w  y+0.86*h  0.01 0.01*figW/figH], &#39;Color&#39;, &#39;k&#39;)
<a name="l01658"></a>01658             annotation(figHandle, &#39;arrow&#39;, [x+0.535*w  x+0.60*w], [y+0.57*h  y+0.63*h], &#39;Color&#39;, &#39;k&#39;, ...
<a name="l01659"></a>01659                 &#39;HeadLength&#39;, 3, &#39;HeadWidth&#39;, 5, &#39;HeadStyle&#39;, &#39;vback1&#39;)
<a name="l01660"></a>01660             text(1.364, 270100, {<span class="stringliteral">&#39;DNA pol-RNA&#39;</span> <span class="stringliteral">&#39;pol collision&#39;</span>}, <span class="stringliteral">&#39;parent&#39;</span>, axesHandle, ...
<a name="l01661"></a>01661                 <span class="stringliteral">&#39;FontSize&#39;</span>, 6, <span class="stringliteral">&#39;VerticalAlign&#39;</span>, <span class="stringliteral">&#39;middle&#39;</span>, <span class="stringliteral">&#39;HorizontalAlign&#39;</span>, <span class="stringliteral">&#39;center&#39;</span>, <span class="stringliteral">&#39;Color&#39;</span>, <span class="charliteral">&#39;k&#39;</span>);
<a name="l01662"></a>01662             text(1.3875, 267400, {<span class="stringliteral">&#39;Leading&#39;</span> <span class="stringliteral">&#39;DNA pol&#39;</span>}, <span class="stringliteral">&#39;parent&#39;</span>, axesHandle, ...
<a name="l01663"></a>01663                 <span class="stringliteral">&#39;FontSize&#39;</span>, 6, <span class="stringliteral">&#39;VerticalAlign&#39;</span>, <span class="stringliteral">&#39;middle&#39;</span>, <span class="stringliteral">&#39;HorizontalAlign&#39;</span>, <span class="stringliteral">&#39;center&#39;</span>, <span class="stringliteral">&#39;Color&#39;</span>, <span class="charliteral">&#39;g&#39;</span>);
<a name="l01664"></a>01664             text(1.400, 272500, {<span class="stringliteral">&#39;Lagging&#39;</span> <span class="stringliteral">&#39;DNA pol&#39;</span>}, <span class="stringliteral">&#39;parent&#39;</span>, axesHandle, ...
<a name="l01665"></a>01665                 <span class="stringliteral">&#39;FontSize&#39;</span>, 6, <span class="stringliteral">&#39;VerticalAlign&#39;</span>, <span class="stringliteral">&#39;middle&#39;</span>, <span class="stringliteral">&#39;HorizontalAlign&#39;</span>, <span class="stringliteral">&#39;center&#39;</span>, <span class="stringliteral">&#39;Color&#39;</span>, <span class="charliteral">&#39;g&#39;</span>);
<a name="l01666"></a>01666             
<a name="l01667"></a>01667             %% zoom box
<a name="l01668"></a>01668             uistack(zoomBoxHandle, <span class="stringliteral">&#39;top&#39;</span>);
<a name="l01669"></a>01669         end
<a name="l01670"></a>01670         
<a name="l01671"></a>01671         <span class="keyword">function</span> metaboliteConcentrations(simBatchDir, selectedSim, outDirectory, figHandle, position)
<a name="l01672"></a>01672             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
<a name="l01673"></a>01673             import edu.stanford.covert.cell.sim.util.PlotUtil;
<a name="l01674"></a>01674             import edu.stanford.covert.cell.sim.util.PrintUtil;
<a name="l01675"></a>01675             import edu.stanford.covert.cell.sim.util.SummaryLogger;
<a name="l01676"></a>01676             
<a name="l01677"></a>01677             if ~exist(&#39;simBatchDir&#39;, &#39;var&#39;) || isempty(simBatchDir)
<a name="l01678"></a>01678                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
<a name="l01679"></a>01679             end
<a name="l01680"></a>01680             
<a name="l01681"></a>01681             %% get constants
<a name="l01682"></a>01682             [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep num2str(selectedSim)]);            
<a name="l01683"></a>01683             m = sim.state(&#39;<a class="code" href="class_metabolite.html">Metabolite</a>&#39;);
<a name="l01684"></a>01684             metIdxs = [m.ntpIndexs; m.ndpIndexs; m.nmpIndexs; m.dntpIndexs; m.aminoAcidIndexs; m.phosphateIndexs; m.diphosphateIndexs; m.hydrogenIndexs];
<a name="l01685"></a>01685             
<a name="l01686"></a>01686             metIDs = m.wholeCellModelIDs;
<a name="l01687"></a>01687             metIDs(m.dntpIndexs) = {<span class="stringliteral">&#39;dATP&#39;</span>, <span class="stringliteral">&#39;dCTP&#39;</span>, <span class="stringliteral">&#39;dGTP&#39;</span>, <span class="stringliteral">&#39;dTTP&#39;</span>};
<a name="l01688"></a>01688             metIDs(m.aminoAcidIndexs) = {
<a name="l01689"></a>01689                 <span class="stringliteral">&#39;Ala&#39;</span>
<a name="l01690"></a>01690                 <span class="stringliteral">&#39;Arg&#39;</span>
<a name="l01691"></a>01691                 <span class="stringliteral">&#39;Asn&#39;</span>
<a name="l01692"></a>01692                 <span class="stringliteral">&#39;Asp&#39;</span>
<a name="l01693"></a>01693                 <span class="stringliteral">&#39;Cys&#39;</span>
<a name="l01694"></a>01694                 <span class="stringliteral">&#39;Gln&#39;</span>
<a name="l01695"></a>01695                 <span class="stringliteral">&#39;Glu&#39;</span>
<a name="l01696"></a>01696                 <span class="stringliteral">&#39;Gly&#39;</span>
<a name="l01697"></a>01697                 <span class="stringliteral">&#39;His&#39;</span>
<a name="l01698"></a>01698                 <span class="stringliteral">&#39;Ile&#39;</span>
<a name="l01699"></a>01699                 <span class="stringliteral">&#39;Leu&#39;</span>
<a name="l01700"></a>01700                 <span class="stringliteral">&#39;Lys&#39;</span>
<a name="l01701"></a>01701                 <span class="stringliteral">&#39;Met&#39;</span>
<a name="l01702"></a>01702                 <span class="stringliteral">&#39;Phe&#39;</span>
<a name="l01703"></a>01703                 <span class="stringliteral">&#39;Pro&#39;</span>
<a name="l01704"></a>01704                 <span class="stringliteral">&#39;Ser&#39;</span>
<a name="l01705"></a>01705                 <span class="stringliteral">&#39;Thr&#39;</span>
<a name="l01706"></a>01706                 <span class="stringliteral">&#39;Trp&#39;</span>
<a name="l01707"></a>01707                 <span class="stringliteral">&#39;Tyr&#39;</span>
<a name="l01708"></a>01708                 <span class="stringliteral">&#39;Val&#39;</span>
<a name="l01709"></a>01709                 <span class="stringliteral">&#39;fMet&#39;</span>
<a name="l01710"></a>01710                 };
<a name="l01711"></a>01711             metIDs{m.phosphateIndexs} = <span class="stringliteral">&#39;Pi&#39;</span>;
<a name="l01712"></a>01712             metIDs{m.diphosphateIndexs} = <span class="stringliteral">&#39;PPi&#39;</span>;
<a name="l01713"></a>01713             metIDs{m.hydrogenIndexs} = <span class="stringliteral">&#39;H^+&#39;</span>; 
<a name="l01714"></a>01714             
<a name="l01715"></a>01715             %% <span class="keyword">get</span> data
<a name="l01716"></a>01716             tmp = load([SimulationDiskUtil.getSimulationBatchDir([simBatchDir filesep <span class="charliteral">&#39;1&#39;</span>]) filesep <span class="stringliteral">&#39;population-MetaboliteConcentrations.mat&#39;</span>]);
<a name="l01717"></a>01717             tmp2 = [tmp.meanConcs tmp.stdConcs];
<a name="l01718"></a>01718             modelConcs = zeros(numel(m.wholeCellModelIDs), 2);
<a name="l01719"></a>01719             modelConcs(m.aminoAcidIndexs, :) = tmp2(17:37, :);
<a name="l01720"></a>01720             modelConcs(m.ntpIndexs, :) = tmp2(1:4, :);
<a name="l01721"></a>01721             modelConcs(m.ndpIndexs, :) = tmp2(5:8, :);
<a name="l01722"></a>01722             modelConcs(m.nmpIndexs, :) = tmp2(9:12, :);
<a name="l01723"></a>01723             modelConcs(m.dntpIndexs, :) = tmp2(13:16, :);
<a name="l01724"></a>01724             modelConcs(m.phosphateIndexs, :) = tmp2(38, :);
<a name="l01725"></a>01725             modelConcs(m.diphosphateIndexs, :) = tmp2(39, :);
<a name="l01726"></a>01726             modelConcs(m.hydrogenIndexs, :) = tmp2(40, :);
<a name="l01727"></a>01727             
<a name="l01728"></a>01728             [~, ~, data] = xlsread(<span class="stringliteral">&#39;documentation/reconstruction/Metabolite Concentrations.xls&#39;</span>);            
<a name="l01729"></a>01729             tfs = ~cellfun(@(x) any(isnan(x)), data(:, 1));
<a name="l01730"></a>01730             tfs(1:2) = <span class="keyword">false</span>;
<a name="l01731"></a>01731             data = data(tfs, :);
<a name="l01732"></a>01732             
<a name="l01733"></a>01733             [~, idxs] = ismember(data(:, 1), m.wholeCellModelIDs);
<a name="l01734"></a>01734             expConcs = NaN(size(modelConcs, 1), 3);
<a name="l01735"></a>01735             cyberCellConc = NaN(size(modelConcs, 1), 1);
<a name="l01736"></a>01736             expConcs(idxs, :) = cell2mat(data(:, 2:4));
<a name="l01737"></a>01737             cyberCellConc(idxs, :) = cell2mat(data(:, 5));
<a name="l01738"></a>01738             
<a name="l01739"></a>01739             sortedIdxs = [
<a name="l01740"></a>01740                 m.aminoAcidIndexs(1:end-1)
<a name="l01741"></a>01741                 m.ntpIndexs
<a name="l01742"></a>01742                 m.ndpIndexs
<a name="l01743"></a>01743                 m.nmpIndexs
<a name="l01744"></a>01744                 m.dntpIndexs
<a name="l01745"></a>01745                 m.phosphateIndexs
<a name="l01746"></a>01746                 m.diphosphateIndexs
<a name="l01747"></a>01747                 m.hydrogenIndexs
<a name="l01748"></a>01748                 ];
<a name="l01749"></a>01749             allSortedIdxs = [
<a name="l01750"></a>01750                 m.dntpIndexs
<a name="l01751"></a>01751                 m.ndpIndexs
<a name="l01752"></a>01752                 m.nmpIndexs
<a name="l01753"></a>01753                 m.ntpIndexs
<a name="l01754"></a>01754                 m.aminoAcidIndexs
<a name="l01755"></a>01755                 m.phosphateIndexs
<a name="l01756"></a>01756                 m.diphosphateIndexs
<a name="l01757"></a>01757                 m.hydrogenIndexs
<a name="l01758"></a>01758                 ];
<a name="l01759"></a>01759             
<a name="l01760"></a>01760             %% plot
<a name="l01761"></a>01761             <span class="keywordflow">if</span> nargin &gt;= 4
<a name="l01762"></a>01762                 axesHandle = subplot(<span class="stringliteral">&#39;position&#39;</span>, position, <span class="stringliteral">&#39;parent&#39;</span>, figHandle);
<a name="l01763"></a>01763             <span class="keywordflow">else</span>
<a name="l01764"></a>01764                 axesHandle = edu.stanford.covert.cell.sim.util.PlotUtil.newAxesHandle();
<a name="l01765"></a>01765             end
<a name="l01766"></a>01766             
<a name="l01767"></a>01767             hold(axesHandle, <span class="stringliteral">&#39;on&#39;</span>);
<a name="l01768"></a>01768             <span class="keyword">set</span>(axesHandle, ...
<a name="l01769"></a>01769                 <span class="stringliteral">&#39;YScale&#39;</span>, <span class="stringliteral">&#39;log&#39;</span>, ...
<a name="l01770"></a>01770                 <span class="stringliteral">&#39;YMinorTick&#39;</span>, <span class="stringliteral">&#39;on&#39;</span>, ...
<a name="l01771"></a>01771                 <span class="stringliteral">&#39;XTick&#39;</span>, 1:numel(sortedIdxs), ...
<a name="l01772"></a>01772                 <span class="stringliteral">&#39;XTickLabel&#39;</span>, m.wholeCellModelIDs(sortedIdxs), ...
<a name="l01773"></a>01773                 <span class="stringliteral">&#39;YTick&#39;</span>, logspace(-2, 2, 5), ...
<a name="l01774"></a>01774                 <span class="stringliteral">&#39;FontSize&#39;</span>, 5, ...
<a name="l01775"></a>01775                 <span class="stringliteral">&#39;TickDir&#39;</span>, <span class="stringliteral">&#39;out&#39;</span>);
<a name="l01776"></a>01776             xlim(axesHandle, [0.5 numel(sortedIdxs)+0.5]);
<a name="l01777"></a>01777             ylim(axesHandle, [1e-2/1.5 1.5e2]);
<a name="l01778"></a>01778             
<a name="l01779"></a>01779             plot(axesHandle, 1:numel(sortedIdxs), expConcs(sortedIdxs, 1) ./ modelConcs(sortedIdxs, 1), ...
<a name="l01780"></a>01780                 <span class="stringliteral">&#39;rd&#39;</span>, <span class="stringliteral">&#39;MarkerSize&#39;</span>, 3, <span class="stringliteral">&#39;MarkerFaceColor&#39;</span>, <span class="charliteral">&#39;r&#39;</span>, <span class="stringliteral">&#39;MarkerEdgeColor&#39;</span>, <span class="charliteral">&#39;r&#39;</span>);
<a name="l01781"></a>01781             plot(axesHandle, 1:numel(sortedIdxs), cyberCellConc(sortedIdxs) ./ modelConcs(sortedIdxs, 1), ...
<a name="l01782"></a>01782                 <span class="stringliteral">&#39;gs&#39;</span>, <span class="stringliteral">&#39;MarkerSize&#39;</span>, 3, <span class="stringliteral">&#39;MarkerFaceColor&#39;</span>, <span class="charliteral">&#39;g&#39;</span>, <span class="stringliteral">&#39;MarkerEdgeColor&#39;</span>, <span class="charliteral">&#39;g&#39;</span>);
<a name="l01783"></a>01783             nModel = numel(sortedIdxs);
<a name="l01784"></a>01784             line([1:nModel; 1:nModel], max([ 1 + (modelConcs(sortedIdxs, 2) ./ modelConcs(sortedIdxs, 1))<span class="stringliteral">&#39;; 1 - (modelConcs(sortedIdxs, 2) ./ modelConcs(sortedIdxs, 1))&#39;</span>], eps), <span class="stringliteral">&#39;Color&#39;</span>, <span class="charliteral">&#39;b&#39;</span>);
<a name="l01785"></a>01785             
<a name="l01786"></a>01786             tick2text(axesHandle, <span class="stringliteral">&#39;axis&#39;</span>, <span class="stringliteral">&#39;xy&#39;</span>, <span class="stringliteral">&#39;xtickoffset&#39;</span>, 0.13, <span class="stringliteral">&#39;ytickoffset&#39;</span>, 0.013);
<a name="l01787"></a>01787             xTicks = getappdata(axesHandle, <span class="stringliteral">&#39;XTickText&#39;</span>);
<a name="l01788"></a>01788             yTicks = getappdata(axesHandle, <span class="stringliteral">&#39;YTickText&#39;</span>);
<a name="l01789"></a>01789             <span class="keyword">set</span>(xTicks, <span class="stringliteral">&#39;FontSize&#39;</span>, 5, <span class="stringliteral">&#39;HorizontalAlign&#39;</span>, <span class="stringliteral">&#39;left&#39;</span>, <span class="stringliteral">&#39;VerticalAlign&#39;</span>, <span class="stringliteral">&#39;middle&#39;</span>, <span class="stringliteral">&#39;Interpreter&#39;</span>, <span class="stringliteral">&#39;tex&#39;</span>, <span class="stringliteral">&#39;rotation&#39;</span>, 270);
<a name="l01790"></a>01790             <span class="keyword">set</span>(yTicks, <span class="stringliteral">&#39;FontSize&#39;</span>, 5, <span class="stringliteral">&#39;HorizontalAlign&#39;</span>, <span class="stringliteral">&#39;right&#39;</span>, <span class="stringliteral">&#39;Interpreter&#39;</span>, <span class="stringliteral">&#39;tex&#39;</span>);
<a name="l01791"></a>01791             <span class="keywordflow">for</span> i = 1:numel(xTicks)
<a name="l01792"></a>01792                 set(xTicks(i), &#39;String&#39;, metIDs{sortedIdxs(i)}, <span class="stringliteral">&#39;position&#39;</span>, [i 4.45e-3 -1], <span class="stringliteral">&#39;Margin&#39;</span>, 1e-6);
<a name="l01793"></a>01793             end
<a name="l01794"></a>01794             <span class="keywordflow">for</span> i = 1:numel(yTicks)
<a name="l01795"></a>01795                 set(yTicks(i), &#39;String&#39;, sprintf(&#39;10^{%d}<span class="stringliteral">&#39;, log10(str2double(get(yTicks(i), &#39;</span><span class="keywordtype">string</span><span class="stringliteral">&#39;)))));</span>
<a name="l01796"></a>01796 <span class="stringliteral">            end</span>
<a name="l01797"></a>01797 <span class="stringliteral">            </span>
<a name="l01798"></a>01798 <span class="stringliteral">            ylabel(axesHandle, &#39;</span>Experiment / model concentration<span class="stringliteral">&#39;, &#39;</span>FontSize<span class="stringliteral">&#39;, 7);</span>
<a name="l01799"></a>01799 <span class="stringliteral">            yLabelPos = get(get(axesHandle, &#39;</span>ylabel<span class="stringliteral">&#39;), &#39;</span>position<span class="stringliteral">&#39;);</span>
<a name="l01800"></a>01800 <span class="stringliteral">            set(get(axesHandle, &#39;</span>ylabel<span class="stringliteral">&#39;), &#39;</span>position<span class="stringliteral">&#39;, [-1.8 yLabelPos(2:3)])</span>
<a name="l01801"></a>01801 <span class="stringliteral">  </span>
<a name="l01802"></a>01802 <span class="stringliteral">            legendHandle = legend({&#39;</span>Bennett {\it{et al.}}, 2009<span class="stringliteral">&#39;, &#39;</span>Literature (CCDB)<span class="stringliteral">&#39;, &#39;</span>Model s.d.<span class="stringliteral">&#39;}, &#39;</span>Location<span class="stringliteral">&#39;, &#39;</span>SouthEast<span class="stringliteral">&#39;, &#39;</span>FontSize<span class="stringliteral">&#39;, 6);</span>
<a name="l01803"></a>01803 <span class="stringliteral">            legendPos = get(legendHandle, &#39;</span>Position<span class="stringliteral">&#39;);            </span>
<a name="l01804"></a>01804 <span class="stringliteral">            set(legendHandle, &#39;</span>Position<span class="stringliteral">&#39;, [0.048 0.09 legendPos(3:end)])</span>
<a name="l01805"></a>01805 <span class="stringliteral">            set(legendHandle, &#39;</span>units<span class="stringliteral">&#39;, &#39;</span>pixels<span class="stringliteral">&#39;);</span>
<a name="l01806"></a>01806 <span class="stringliteral">            legendPos = get(legendHandle, &#39;</span>Position<span class="stringliteral">&#39;);</span>
<a name="l01807"></a>01807 <span class="stringliteral">            set(legendHandle, &#39;</span>Position<span class="stringliteral">&#39;, [legendPos(1:2) 82 28]);</span>
<a name="l01808"></a>01808 <span class="stringliteral">            legendChildren = get(legendHandle, &#39;</span>children<span class="stringliteral">&#39;);            </span>
<a name="l01809"></a>01809 <span class="stringliteral">            for i = 1:3:numel(legendChildren)</span>
<a name="l01810"></a>01810 <span class="stringliteral">                if strcmp(get(legendChildren(i), &#39;</span>marker<span class="stringliteral">&#39;), &#39;</span>none<span class="stringliteral">&#39;)</span>
<a name="l01811"></a>01811 <span class="stringliteral">                    set(legendChildren(i), &#39;</span>visible<span class="stringliteral">&#39;, &#39;</span>off<span class="stringliteral">&#39;);</span>
<a name="l01812"></a>01812 <span class="stringliteral">                    set(legendChildren(i+1), &#39;</span>xdata<span class="stringliteral">&#39;, [0.03 0.10])</span>
<a name="l01813"></a>01813 <span class="stringliteral">                else</span>
<a name="l01814"></a>01814 <span class="stringliteral">                    set(legendChildren(i+1), &#39;</span>visible<span class="stringliteral">&#39;, &#39;</span>off<span class="stringliteral">&#39;);</span>
<a name="l01815"></a>01815 <span class="stringliteral">                    set(legendChildren(i), &#39;</span>xdata<span class="stringliteral">&#39;, 0.06)</span>
<a name="l01816"></a>01816 <span class="stringliteral">                end</span>
<a name="l01817"></a>01817 <span class="stringliteral">                </span>
<a name="l01818"></a>01818 <span class="stringliteral">                pos = get(legendChildren(i+2), &#39;</span>Position<span class="stringliteral">&#39;);</span>
<a name="l01819"></a>01819 <span class="stringliteral">                set(legendChildren(i+2), &#39;</span>Position<span class="stringliteral">&#39;, [0.13 pos(2) 0])</span>
<a name="l01820"></a>01820 <span class="stringliteral">            end</span>
<a name="l01821"></a>01821 <span class="stringliteral">            </span>
<a name="l01822"></a>01822 <span class="stringliteral">            axisPos = get(axesHandle, &#39;</span>Position<span class="stringliteral">&#39;);</span>
<a name="l01823"></a>01823 <span class="stringliteral">            x = axisPos(1);</span>
<a name="l01824"></a>01824 <span class="stringliteral">            w = axisPos(3);</span>
<a name="l01825"></a>01825 <span class="stringliteral">            y = 0.021;</span>
<a name="l01826"></a>01826 <span class="stringliteral">            annotation(figHandle, &#39;</span>line<span class="stringliteral">&#39;, x + ([ 1 20]-0.5 + [-0.2 0.2])/39*w, y * [1 1])</span>
<a name="l01827"></a>01827 <span class="stringliteral">            annotation(figHandle, &#39;</span>line<span class="stringliteral">&#39;, x + ([21 24]-0.5 + [-0.2 0.2])/39*w, y * [1 1])</span>
<a name="l01828"></a>01828 <span class="stringliteral">            annotation(figHandle, &#39;</span>line<span class="stringliteral">&#39;, x + ([25 28]-0.5 + [-0.2 0.2])/39*w, y * [1 1])</span>
<a name="l01829"></a>01829 <span class="stringliteral">            annotation(figHandle, &#39;</span>line<span class="stringliteral">&#39;, x + ([29 32]-0.5 + [-0.2 0.2])/39*w, y * [1 1])</span>
<a name="l01830"></a>01830 <span class="stringliteral">            annotation(figHandle, &#39;</span>line<span class="stringliteral">&#39;, x + ([33 36]-0.5 + [-0.2 0.2])/39*w, y * [1 1])</span>
<a name="l01831"></a>01831 <span class="stringliteral">            annotation(figHandle, &#39;</span>line<span class="stringliteral">&#39;, x + ([37 39]-0.5 + [-0.2 0.2])/39*w, y * [1 1])            </span>
<a name="l01832"></a>01832 <span class="stringliteral">            </span>
<a name="l01833"></a>01833 <span class="stringliteral">            y = 0.53e-3;</span>
<a name="l01834"></a>01834 <span class="stringliteral">            text(mean([ 1 20]), y, &#39;</span>Amino acid<span class="stringliteral">&#39;, &#39;</span>FontSize<span class="stringliteral">&#39;, 7, &#39;</span>HorizontalAlign<span class="stringliteral">&#39;, &#39;</span>center<span class="stringliteral">&#39;, &#39;</span>Parent<span class="stringliteral">&#39;, axesHandle)            </span>
<a name="l01835"></a>01835 <span class="stringliteral">            text(mean([21 24]), y, &#39;</span>NTP<span class="stringliteral">&#39;,  &#39;</span>FontSize<span class="stringliteral">&#39;, 7, &#39;</span>HorizontalAlign<span class="stringliteral">&#39;, &#39;</span>center<span class="stringliteral">&#39;, &#39;</span>Parent<span class="stringliteral">&#39;, axesHandle)</span>
<a name="l01836"></a>01836 <span class="stringliteral">            text(mean([25 28]), y, &#39;</span>NDP<span class="stringliteral">&#39;,  &#39;</span>FontSize<span class="stringliteral">&#39;, 7, &#39;</span>HorizontalAlign<span class="stringliteral">&#39;, &#39;</span>center<span class="stringliteral">&#39;, &#39;</span>Parent<span class="stringliteral">&#39;, axesHandle)</span>
<a name="l01837"></a>01837 <span class="stringliteral">            text(mean([29 32]), y, &#39;</span>NMP<span class="stringliteral">&#39;,  &#39;</span>FontSize<span class="stringliteral">&#39;, 7, &#39;</span>HorizontalAlign<span class="stringliteral">&#39;, &#39;</span>center<span class="stringliteral">&#39;, &#39;</span>Parent<span class="stringliteral">&#39;, axesHandle)</span>
<a name="l01838"></a>01838 <span class="stringliteral">            text(mean([33 36]), y, &#39;</span>dNTP<span class="stringliteral">&#39;, &#39;</span>FontSize<span class="stringliteral">&#39;, 7, &#39;</span>HorizontalAlign<span class="stringliteral">&#39;, &#39;</span>center<span class="stringliteral">&#39;, &#39;</span>Parent<span class="stringliteral">&#39;, axesHandle)</span>
<a name="l01839"></a>01839 <span class="stringliteral">            text(mean([37 39]), y, &#39;</span>Ion<span class="stringliteral">&#39;,  &#39;</span>FontSize<span class="stringliteral">&#39;, 7, &#39;</span>HorizontalAlign<span class="stringliteral">&#39;, &#39;</span>center<span class="stringliteral">&#39;, &#39;</span>Parent<span class="stringliteral">&#39;, axesHandle)</span>
<a name="l01840"></a>01840 <span class="stringliteral">            </span>
<a name="l01841"></a>01841 <span class="stringliteral">            ptch = patch([0.5 numel(sortedIdxs)+0.5 numel(sortedIdxs)+0.5 0.5], ...</span>
<a name="l01842"></a>01842 <span class="stringliteral">                [1e1 1e1 1e-1 1e-1], 1, ...</span>
<a name="l01843"></a>01843 <span class="stringliteral">                &#39;</span>Parent<span class="stringliteral">&#39;, axesHandle, &#39;</span>EdgeColor<span class="stringliteral">&#39;, &#39;</span>none<span class="stringliteral">&#39;);</span>
<a name="l01844"></a>01844 <span class="stringliteral">            set(ptch, &#39;</span>FaceColor<span class="stringliteral">&#39;, 0.925 * [1 1 1], &#39;</span>FaceAlpha<span class="stringliteral">&#39;, 1, &#39;</span>EdgeAlpha<span class="stringliteral">&#39;, 1);</span>
<a name="l01845"></a>01845 <span class="stringliteral">            h = line([0.5 0.5], [1/2*1e-1 2*1e1], &#39;</span>LineWidth<span class="stringliteral">&#39;, 0.5, &#39;</span>Color<span class="stringliteral">&#39;, &#39;</span>k<span class="stringliteral">&#39;, &#39;</span>Parent<span class="stringliteral">&#39;, axesHandle);</span>
<a name="l01846"></a>01846 <span class="stringliteral">            uistack(ptch, &#39;</span>bottom<span class="stringliteral">&#39;);</span>
<a name="l01847"></a>01847 <span class="stringliteral">            uistack(h, &#39;</span>top<span class="stringliteral">&#39;);</span>
<a name="l01848"></a>01848 <span class="stringliteral">            </span>
<a name="l01849"></a>01849 <span class="stringliteral">            %% table</span>
<a name="l01850"></a>01850 <span class="stringliteral">            colLabels = {&#39;</span><a class="code" href="class_metabolite.html">Metabolite</a> ID<span class="charliteral">&#39; &#39;</span>Name<span class="stringliteral">&#39;, &#39;</span>Whole Cell Model Mean<span class="charliteral">&#39; &#39;</span>Whole Cell Model Std<span class="charliteral">&#39; &#39;</span>Bennet et al. 2009 Mean<span class="charliteral">&#39; &#39;</span>Bennet et al. 2009 Lower Bound<span class="charliteral">&#39; &#39;</span>Bennet et al. 2009 Upper Bound<span class="charliteral">&#39; &#39;</span>CyberCell<span class="stringliteral">&#39;};</span>
<a name="l01851"></a>01851 <span class="stringliteral">            content = [m.wholeCellModelIDs  m.names  ...</span>
<a name="l01852"></a>01852 <span class="stringliteral">                num2cell([modelConcs expConcs cyberCellConc])];</span>
<a name="l01853"></a>01853 <span class="stringliteral">            content = content(allSortedIdxs, :);</span>
<a name="l01854"></a>01854 <span class="stringliteral">            </span>
<a name="l01855"></a>01855 <span class="stringliteral">            PrintUtil.printToFile(content, colLabels, [outDirectory &#39;</span>metaboliteConcentrations.xls<span class="stringliteral">&#39;], &#39;</span><a class="code" href="class_metabolite.html">Metabolite</a> Concentrations<span class="stringliteral">&#39;);</span>
<a name="l01856"></a>01856 <span class="stringliteral">        end</span>
<a name="l01857"></a>01857 <span class="stringliteral">        </span>
<a name="l01858"></a>01858 <span class="stringliteral">        function [exptGeneExpr, modelGeneExp] = geneExpression(simBatchDir, selectedSim, figHandle, position, logScale)</span>
<a name="l01859"></a>01859 <span class="stringliteral">            import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;</span>
<a name="l01860"></a>01860 <span class="stringliteral">            import edu.stanford.covert.cell.sim.util.PlotUtil;</span>
<a name="l01861"></a>01861 <span class="stringliteral">            </span>
<a name="l01862"></a>01862 <span class="stringliteral">            if ~exist(&#39;</span>simBatchDir<span class="stringliteral">&#39;, &#39;</span>var<span class="stringliteral">&#39;) || isempty(simBatchDir)</span>
<a name="l01863"></a>01863 <span class="stringliteral">                simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();</span>
<a name="l01864"></a>01864 <span class="stringliteral">            end</span>
<a name="l01865"></a>01865 <span class="stringliteral">            if ~exist(&#39;</span>selectedSim<span class="stringliteral">&#39;, &#39;</span>var<span class="stringliteral">&#39;) || isempty(selectedSim)</span>
<a name="l01866"></a>01866 <span class="stringliteral">                selectedSim = 1;</span>
<a name="l01867"></a>01867 <span class="stringliteral">            end</span>
<a name="l01868"></a>01868 <span class="stringliteral">            </span>
<a name="l01869"></a>01869 <span class="stringliteral">            %% get constants</span>
<a name="l01870"></a>01870 <span class="stringliteral">            [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep num2str(selectedSim)]);</span>
<a name="l01871"></a>01871 <span class="stringliteral">            rna = sim.state(&#39;</span>Rna<span class="stringliteral">&#39;);</span>
<a name="l01872"></a>01872 <span class="stringliteral">            g = sim.gene;</span>
<a name="l01873"></a>01873 <span class="stringliteral">            </span>
<a name="l01874"></a>01874 <span class="stringliteral">            %% get data</span>
<a name="l01875"></a>01875 <span class="stringliteral">            exptGeneExpr = rna.expectedGeneExpression(:, 1);</span>
<a name="l01876"></a>01876 <span class="stringliteral">            exptGeneExpr(exptGeneExpr == 0) = NaN;</span>
<a name="l01877"></a>01877 <span class="stringliteral">            exptGeneExpr(g.mRNAIndexs) = exptGeneExpr(g.mRNAIndexs) / nansum(exptGeneExpr(g.mRNAIndexs));</span>
<a name="l01878"></a>01878 <span class="stringliteral">            exptGeneExpr(setdiff(1:end, g.mRNAIndexs)) = exptGeneExpr(setdiff(1:end, g.mRNAIndexs)) / nansum(exptGeneExpr(setdiff(1:end, g.mRNAIndexs)));</span>
<a name="l01879"></a>01879 <span class="stringliteral">            </span>
<a name="l01880"></a>01880 <span class="stringliteral">            modelGeneExp = rna.matureRNAGeneComposition * rna.expression(rna.matureIndexs);</span>
<a name="l01881"></a>01881 <span class="stringliteral">            modelGeneExp(g.mRNAIndexs) = modelGeneExp(g.mRNAIndexs) / sum(modelGeneExp(g.mRNAIndexs));</span>
<a name="l01882"></a>01882 <span class="stringliteral">            modelGeneExp(setdiff(1:end, g.mRNAIndexs)) = modelGeneExp(setdiff(1:end, g.mRNAIndexs)) / sum(modelGeneExp(setdiff(1:end, g.mRNAIndexs)));</span>
<a name="l01883"></a>01883 <span class="stringliteral">            </span>
<a name="l01884"></a>01884 <span class="stringliteral">            %% plot</span>
<a name="l01885"></a>01885 <span class="stringliteral">            if nargin &gt;= 4</span>
<a name="l01886"></a>01886 <span class="stringliteral">                axesHandle = subplot(&#39;</span>Position<span class="stringliteral">&#39;, position, &#39;</span>Parent<span class="stringliteral">&#39;, figHandle);</span>
<a name="l01887"></a>01887 <span class="stringliteral">            else</span>
<a name="l01888"></a>01888 <span class="stringliteral">                axesHandle = edu.stanford.covert.cell.sim.util.PlotUtil.newAxesHandle();</span>
<a name="l01889"></a>01889 <span class="stringliteral">            end</span>
<a name="l01890"></a>01890 <span class="stringliteral">            </span>
<a name="l01891"></a>01891 <span class="stringliteral">            if logScale</span>
<a name="l01892"></a>01892 <span class="stringliteral">                set(axesHandle, &#39;</span>XScale<span class="stringliteral">&#39;, &#39;</span>log<span class="stringliteral">&#39;);</span>
<a name="l01893"></a>01893 <span class="stringliteral">                set(axesHandle, &#39;</span>YScale<span class="stringliteral">&#39;, &#39;</span>log<span class="stringliteral">&#39;);</span>
<a name="l01894"></a>01894 <span class="stringliteral">            end</span>
<a name="l01895"></a>01895 <span class="stringliteral">            set(axesHandle, &#39;</span>TickDir<span class="stringliteral">&#39;, &#39;</span>out<span class="stringliteral">&#39;);</span>
<a name="l01896"></a>01896 <span class="stringliteral">            set(axesHandle, &#39;</span>FontSize<span class="stringliteral">&#39;, 5);</span>
<a name="l01897"></a>01897 <span class="stringliteral">            hold(axesHandle, &#39;</span>on<span class="stringliteral">&#39;);</span>
<a name="l01898"></a>01898 <span class="stringliteral">            </span>
<a name="l01899"></a>01899 <span class="stringliteral">            plot(axesHandle, exptGeneExpr(g.mRNAIndexs), modelGeneExp(g.mRNAIndexs, 1), &#39;</span>.<span class="stringliteral">&#39;, &#39;</span>Color<span class="stringliteral">&#39;, [0 1 1])</span>
<a name="l01900"></a>01900 <span class="stringliteral">            plot(axesHandle, exptGeneExpr(g.rRNAIndexs), modelGeneExp(g.rRNAIndexs, 1), &#39;</span>.<span class="stringliteral">&#39;, &#39;</span>Color<span class="stringliteral">&#39;, [0 0 1])</span>
<a name="l01901"></a>01901 <span class="stringliteral">            plot(axesHandle, exptGeneExpr(g.sRNAIndexs), modelGeneExp(g.sRNAIndexs, 1), &#39;</span>.<span class="stringliteral">&#39;, &#39;</span>Color<span class="stringliteral">&#39;, [1 0 0])</span>
<a name="l01902"></a>01902 <span class="stringliteral">            plot(axesHandle, exptGeneExpr(g.tRNAIndexs), modelGeneExp(g.tRNAIndexs, 1), &#39;</span>.<span class="stringliteral">&#39;, &#39;</span>Color<span class="stringliteral">&#39;, [0 1 0])</span>
<a name="l01903"></a>01903 <span class="stringliteral">            </span>
<a name="l01904"></a>01904 <span class="stringliteral">            if logScale</span>
<a name="l01905"></a>01905 <span class="stringliteral">                xlim(axesHandle, [6.5e-5 8e-2]);</span>
<a name="l01906"></a>01906 <span class="stringliteral">            else</span>
<a name="l01907"></a>01907 <span class="stringliteral">                xlim(axesHandle, [0 7.2e-2])</span>
<a name="l01908"></a>01908 <span class="stringliteral">            end</span>
<a name="l01909"></a>01909 <span class="stringliteral">            ylim(axesHandle, xlim(axesHandle));</span>
<a name="l01910"></a>01910 <span class="stringliteral">            </span>
<a name="l01911"></a>01911 <span class="stringliteral">            tick2text(axesHandle, &#39;</span>axis<span class="stringliteral">&#39;, &#39;</span>xy<span class="stringliteral">&#39;, &#39;</span>xtickoffset<span class="stringliteral">&#39;, 0.025, &#39;</span>ytickoffset<span class="stringliteral">&#39;, 0.035);</span>
<a name="l01912"></a>01912 <span class="stringliteral">            xTicks = getappdata(axesHandle, &#39;</span>XTickText<span class="stringliteral">&#39;);</span>
<a name="l01913"></a>01913 <span class="stringliteral">            yTicks = getappdata(axesHandle, &#39;</span>YTickText<span class="stringliteral">&#39;);</span>
<a name="l01914"></a>01914 <span class="stringliteral">            set(xTicks, &#39;</span>FontSize<span class="stringliteral">&#39;, 5);</span>
<a name="l01915"></a>01915 <span class="stringliteral">            set(yTicks, &#39;</span>FontSize<span class="stringliteral">&#39;, 5);</span>
<a name="l01916"></a>01916 <span class="stringliteral">            if logScale</span>
<a name="l01917"></a>01917 <span class="stringliteral">                for i = 1:numel(xTicks)</span>
<a name="l01918"></a>01918 <span class="stringliteral">                    xTickPos = get(xTicks(i), &#39;</span>position<span class="stringliteral">&#39;);</span>
<a name="l01919"></a>01919 <span class="stringliteral">                    set(xTicks(i), &#39;</span>position<span class="stringliteral">&#39;, [xTickPos(1) 5.5e-5 xTickPos(3)], ...</span>
<a name="l01920"></a>01920 <span class="stringliteral">                        &#39;</span>String<span class="stringliteral">&#39;, sprintf(&#39;</span>10^{%d}<span class="stringliteral">&#39;, log10(str2double(get(xTicks(i), &#39;</span><span class="keywordtype">string</span><span class="stringliteral">&#39;)))), ...</span>
<a name="l01921"></a>01921 <span class="stringliteral">                        &#39;</span>Interpreter<span class="stringliteral">&#39;, &#39;</span>tex<span class="stringliteral">&#39;);</span>
<a name="l01922"></a>01922 <span class="stringliteral">                    </span>
<a name="l01923"></a>01923 <span class="stringliteral">                    yTickPos = get(yTicks(i), &#39;</span>position<span class="stringliteral">&#39;);</span>
<a name="l01924"></a>01924 <span class="stringliteral">                    set(yTicks(i), &#39;</span>position<span class="stringliteral">&#39;, [5e-5 yTickPos(2:3)], ...</span>
<a name="l01925"></a>01925 <span class="stringliteral">                        &#39;</span>String<span class="stringliteral">&#39;, sprintf(&#39;</span>10^{%d}<span class="stringliteral">&#39;, log10(str2double(get(yTicks(i), &#39;</span><span class="keywordtype">string</span><span class="stringliteral">&#39;)))), ...</span>
<a name="l01926"></a>01926 <span class="stringliteral">                        &#39;</span>Interpreter<span class="stringliteral">&#39;, &#39;</span>tex<span class="stringliteral">&#39;);</span>
<a name="l01927"></a>01927 <span class="stringliteral">                end</span>
<a name="l01928"></a>01928 <span class="stringliteral">            end</span>
<a name="l01929"></a>01929 <span class="stringliteral">            </span>
<a name="l01930"></a>01930 <span class="stringliteral">            xlabel(axesHandle, {&#39;</span>Observed gene expression (norm)<span class="stringliteral">&#39;}, &#39;</span>FontSize<span class="stringliteral">&#39;, 7);</span>
<a name="l01931"></a>01931 <span class="stringliteral">            ylabel(axesHandle, {&#39;</span>Predicted gene expression (norm)<span class="stringliteral">&#39;}, &#39;</span>FontSize<span class="stringliteral">&#39;, 7);</span>
<a name="l01932"></a>01932 <span class="stringliteral">            xLabelPos = get(get(axesHandle, &#39;</span>xlabel<span class="stringliteral">&#39;), &#39;</span>position<span class="stringliteral">&#39;);</span>
<a name="l01933"></a>01933 <span class="stringliteral">            yLabelPos = get(get(axesHandle, &#39;</span>ylabel<span class="stringliteral">&#39;), &#39;</span>position<span class="stringliteral">&#39;);</span>
<a name="l01934"></a>01934 <span class="stringliteral">            if logScale</span>
<a name="l01935"></a>01935 <span class="stringliteral">                set(get(axesHandle, &#39;</span>xlabel<span class="stringliteral">&#39;), &#39;</span>position<span class="stringliteral">&#39;, [xLabelPos(1) 4.75e-5 xLabelPos(3)])</span>
<a name="l01936"></a>01936 <span class="stringliteral">                set(get(axesHandle, &#39;</span>ylabel<span class="stringliteral">&#39;), &#39;</span>position<span class="stringliteral">&#39;, [4.1e-5 yLabelPos(2:3)])</span>
<a name="l01937"></a>01937 <span class="stringliteral">            else</span>
<a name="l01938"></a>01938 <span class="stringliteral">                set(get(axesHandle, &#39;</span>xlabel<span class="stringliteral">&#39;), &#39;</span>position<span class="stringliteral">&#39;, [xLabelPos(1) -0.0032 xLabelPos(3)])</span>
<a name="l01939"></a>01939 <span class="stringliteral">                set(get(axesHandle, &#39;</span>ylabel<span class="stringliteral">&#39;), &#39;</span>position<span class="stringliteral">&#39;, [-0.0047 yLabelPos(2:3)])</span>
<a name="l01940"></a>01940 <span class="stringliteral">            end</span>
<a name="l01941"></a>01941 <span class="stringliteral">            </span>
<a name="l01942"></a>01942 <span class="stringliteral">            if logScale</span>
<a name="l01943"></a>01943 <span class="stringliteral">                text(3e-3, 3e-4, &#39;</span><a class="code" href="classm_r_n_a.html">mRNA</a><span class="stringliteral">&#39;, &#39;</span>Color<span class="stringliteral">&#39;, [0 1 1], &#39;</span>Parent<span class="stringliteral">&#39;, axesHandle, &#39;</span>FontSize<span class="stringliteral">&#39;, 7)</span>
<a name="l01944"></a>01944 <span class="stringliteral">                text(7.8e-2, 2.5e-2, &#39;</span><a class="code" href="classr_r_n_a.html">rRNA</a><span class="stringliteral">&#39;, &#39;</span>Color<span class="stringliteral">&#39;, [0 0 1], &#39;</span>Parent<span class="stringliteral">&#39;, axesHandle, &#39;</span>FontSize<span class="stringliteral">&#39;, 7, &#39;</span>HorizontalAlign<span class="stringliteral">&#39;, &#39;</span>right<span class="stringliteral">&#39;)</span>
<a name="l01945"></a>01945 <span class="stringliteral">                text(1.35e-2, 0.77e-2, &#39;</span><a class="code" href="classs_r_n_a.html">sRNA</a><span class="stringliteral">&#39;, &#39;</span>Color<span class="stringliteral">&#39;, [1 0 0], &#39;</span>Parent<span class="stringliteral">&#39;, axesHandle, &#39;</span>FontSize<span class="stringliteral">&#39;, 7)</span>
<a name="l01946"></a>01946 <span class="stringliteral">                text(1.7e-2, 6e-2, &#39;</span><a class="code" href="classt_r_n_a.html">tRNA</a><span class="stringliteral">&#39;, &#39;</span>Color<span class="stringliteral">&#39;, [0 1 0], &#39;</span>Parent<span class="stringliteral">&#39;, axesHandle, &#39;</span>FontSize<span class="stringliteral">&#39;, 7, &#39;</span>HorizontalAlign<span class="stringliteral">&#39;, &#39;</span>right<span class="stringliteral">&#39;)</span>
<a name="l01947"></a>01947 <span class="stringliteral">            else</span>
<a name="l01948"></a>01948 <span class="stringliteral">                text(0.012, 0.003, &#39;</span><a class="code" href="classm_r_n_a.html">mRNA</a><span class="stringliteral">&#39;, &#39;</span>Color<span class="stringliteral">&#39;, [0 1 1], &#39;</span>Parent<span class="stringliteral">&#39;, axesHandle, &#39;</span>FontSize<span class="stringliteral">&#39;, 7)</span>
<a name="l01949"></a>01949 <span class="stringliteral">                text(7e-2, 3.1e-2, &#39;</span><a class="code" href="classr_r_n_a.html">rRNA</a><span class="stringliteral">&#39;, &#39;</span>Color<span class="stringliteral">&#39;, [0 0 1], &#39;</span>Parent<span class="stringliteral">&#39;, axesHandle, &#39;</span>FontSize<span class="stringliteral">&#39;, 7, &#39;</span>HorizontalAlign<span class="stringliteral">&#39;, &#39;</span>right<span class="stringliteral">&#39;)</span>
<a name="l01950"></a>01950 <span class="stringliteral">                text(1.35e-2, 0.77e-2, &#39;</span><a class="code" href="classs_r_n_a.html">sRNA</a><span class="stringliteral">&#39;, &#39;</span>Color<span class="stringliteral">&#39;, [1 0 0], &#39;</span>Parent<span class="stringliteral">&#39;, axesHandle, &#39;</span>FontSize<span class="stringliteral">&#39;, 7)</span>
<a name="l01951"></a>01951 <span class="stringliteral">                text(4.5e-2, 2.7e-2, &#39;</span><a class="code" href="classt_r_n_a.html">tRNA</a><span class="stringliteral">&#39;, &#39;</span>Color<span class="stringliteral">&#39;, [0 1 0], &#39;</span>Parent<span class="stringliteral">&#39;, axesHandle, &#39;</span>FontSize<span class="stringliteral">&#39;, 7, &#39;</span>HorizontalAlign<span class="stringliteral">&#39;, &#39;</span>right<span class="stringliteral">&#39;)</span>
<a name="l01952"></a>01952 <span class="stringliteral">            end</span>
<a name="l01953"></a>01953 <span class="stringliteral">        end</span>
<a name="l01954"></a>01954 <span class="stringliteral">        </span>
<a name="l01955"></a>01955 <span class="stringliteral">        function geneExpressionRatios(exptGeneExpr, modelGeneExpr, axesHandle)</span>
<a name="l01956"></a>01956 <span class="stringliteral">            import edu.stanford.covert.cell.sim.util.CachedSimulationObjectUtil;</span>
<a name="l01957"></a>01957 <span class="stringliteral">            import edu.stanford.covert.cell.sim.util.PlotUtil;</span>
<a name="l01958"></a>01958 <span class="stringliteral">            </span>
<a name="l01959"></a>01959 <span class="stringliteral">            %% get constants</span>
<a name="l01960"></a>01960 <span class="stringliteral">            sim = CachedSimulationObjectUtil.load();</span>
<a name="l01961"></a>01961 <span class="stringliteral">            g = sim.gene;</span>
<a name="l01962"></a>01962 <span class="stringliteral">            </span>
<a name="l01963"></a>01963 <span class="stringliteral">            x1 = modelGeneExpr;</span>
<a name="l01964"></a>01964 <span class="stringliteral">            x2 = exptGeneExpr;</span>
<a name="l01965"></a>01965 <span class="stringliteral">            y1 = modelGeneExpr(g.mRNAIndexs);</span>
<a name="l01966"></a>01966 <span class="stringliteral">            y2 = exptGeneExpr(g.mRNAIndexs);</span>
<a name="l01967"></a>01967 <span class="stringliteral">            z1 = modelGeneExpr(setdiff(1:end, g.mRNAIndexs));</span>
<a name="l01968"></a>01968 <span class="stringliteral">            z2 = exptGeneExpr(setdiff(1:end, g.mRNAIndexs));</span>
<a name="l01969"></a>01969 <span class="stringliteral">            </span>
<a name="l01970"></a>01970 <span class="stringliteral">            corr(x1(~isnan(x1) &amp; ~isnan(x2)), x2(~isnan(x1) &amp; ~isnan(x2)))</span>
<a name="l01971"></a>01971 <span class="stringliteral">            corr(y1(~isnan(y1) &amp; ~isnan(y2)), y2(~isnan(y1) &amp; ~isnan(y2)))</span>
<a name="l01972"></a>01972 <span class="stringliteral">            corr(z1(~isnan(z1) &amp; ~isnan(z2)), z2(~isnan(z1) &amp; ~isnan(z2)))</span>
<a name="l01973"></a>01973 <span class="stringliteral">            </span>
<a name="l01974"></a>01974 <span class="stringliteral">            %% plot</span>
<a name="l01975"></a>01975 <span class="stringliteral">            ratio = modelGeneExpr ./ exptGeneExpr;</span>
<a name="l01976"></a>01976 <span class="stringliteral">            </span>
<a name="l01977"></a>01977 <span class="stringliteral">            spcg = 50;            </span>
<a name="l01978"></a>01978 <span class="stringliteral">            xPos = zeros(size(ratio));</span>
<a name="l01979"></a>01979 <span class="stringliteral">            xPos(g.mRNAIndexs) = 1:numel(g.mRNAIndexs);</span>
<a name="l01980"></a>01980 <span class="stringliteral">            xPos(g.rRNAIndexs) = (1:numel(g.rRNAIndexs)) + spcg + numel(g.mRNAIndexs);</span>
<a name="l01981"></a>01981 <span class="stringliteral">            xPos(g.sRNAIndexs) = (1:numel(g.sRNAIndexs)) + 2*spcg + numel(g.mRNAIndexs) + numel(g.rRNAIndexs);</span>
<a name="l01982"></a>01982 <span class="stringliteral">            xPos(g.tRNAIndexs) = (1:numel(g.tRNAIndexs)) + 3*spcg + numel(g.mRNAIndexs) + numel(g.rRNAIndexs) + numel(g.sRNAIndexs);</span>
<a name="l01983"></a>01983 <span class="stringliteral">            </span>
<a name="l01984"></a>01984 <span class="stringliteral">            hold(axesHandle, &#39;</span>on<span class="stringliteral">&#39;);</span>
<a name="l01985"></a>01985 <span class="stringliteral">            plot(axesHandle, xPos(g.mRNAIndexs), ratio(g.mRNAIndexs), &#39;</span>r.<span class="stringliteral">&#39;);</span>
<a name="l01986"></a>01986 <span class="stringliteral">            plot(axesHandle, xPos(g.rRNAIndexs), ratio(g.rRNAIndexs), &#39;</span>g.<span class="stringliteral">&#39;);</span>
<a name="l01987"></a>01987 <span class="stringliteral">            plot(axesHandle, xPos(g.sRNAIndexs), ratio(g.sRNAIndexs), &#39;</span>b.<span class="stringliteral">&#39;);</span>
<a name="l01988"></a>01988 <span class="stringliteral">            plot(axesHandle, xPos(g.tRNAIndexs), ratio(g.tRNAIndexs), &#39;</span>c.<span class="stringliteral">&#39;);</span>
<a name="l01989"></a>01989 <span class="stringliteral">            </span>
<a name="l01990"></a>01990 <span class="stringliteral">            xlim(axesHandle, [-spcg numel(ratio)+4*spcg]);</span>
<a name="l01991"></a>01991 <span class="stringliteral">            ylim(axesHandle, [1/2*1e-1 2*1e1]);</span>
<a name="l01992"></a>01992 <span class="stringliteral">            ylabel(axesHandle, &#39;</span>Model / Expt <a class="code" href="class_gene.html">Gene</a> Expression<span class="stringliteral">&#39;);</span>
<a name="l01993"></a>01993 <span class="stringliteral">            xlabel(axesHandle, &#39;</span><a class="code" href="class_gene.html">Gene</a><span class="stringliteral">&#39;);</span>
<a name="l01994"></a>01994 <span class="stringliteral">            set(axesHandle, &#39;</span>XTick<span class="stringliteral">&#39;, [], &#39;</span>YScale<span class="stringliteral">&#39;, &#39;</span>log<span class="stringliteral">&#39;, &#39;</span>TickDir<span class="stringliteral">&#39;, &#39;</span>out<span class="stringliteral">&#39;);</span>
<a name="l01995"></a>01995 <span class="stringliteral">            box(axesHandle, &#39;</span>off<span class="stringliteral">&#39;);</span>
<a name="l01996"></a>01996 <span class="stringliteral">            </span>
<a name="l01997"></a>01997 <span class="stringliteral">            lineY = 0.88 * max(ylim(axesHandle));</span>
<a name="l01998"></a>01998 <span class="stringliteral">            textY = 0.99 * max(ylim(axesHandle));</span>
<a name="l01999"></a>01999 <span class="stringliteral">            </span>
<a name="l02000"></a>02000 <span class="stringliteral">            line(xPos(g.mRNAIndexs([1 end])) + [-10; 10], lineY * [1 1], &#39;</span>Parent<span class="stringliteral">&#39;, axesHandle, &#39;</span>Color<span class="stringliteral">&#39;, &#39;</span>r<span class="stringliteral">&#39;)</span>
<a name="l02001"></a>02001 <span class="stringliteral">            line(xPos(g.rRNAIndexs([1 end])) + [-10; 10], lineY * [1 1], &#39;</span>Parent<span class="stringliteral">&#39;, axesHandle, &#39;</span>Color<span class="stringliteral">&#39;, &#39;</span>g<span class="stringliteral">&#39;)</span>
<a name="l02002"></a>02002 <span class="stringliteral">            line(xPos(g.sRNAIndexs([1 end])) + [-10; 10], lineY * [1 1], &#39;</span>Parent<span class="stringliteral">&#39;, axesHandle, &#39;</span>Color<span class="stringliteral">&#39;, &#39;</span>b<span class="stringliteral">&#39;)</span>
<a name="l02003"></a>02003 <span class="stringliteral">            line(xPos(g.tRNAIndexs([1 end])) + [-10; 10], lineY * [1 1], &#39;</span>Parent<span class="stringliteral">&#39;, axesHandle, &#39;</span>Color<span class="stringliteral">&#39;, &#39;</span>c<span class="stringliteral">&#39;)</span>
<a name="l02004"></a>02004 <span class="stringliteral">            text(mean(xPos(g.mRNAIndexs([1 end]))), textY, &#39;</span><a class="code" href="classm_r_n_a.html">mRNA</a><span class="stringliteral">&#39;, &#39;</span>HorizontalAlign<span class="stringliteral">&#39;, &#39;</span>center<span class="stringliteral">&#39;, &#39;</span>Color<span class="stringliteral">&#39;, &#39;</span>r<span class="stringliteral">&#39;);</span>
<a name="l02005"></a>02005 <span class="stringliteral">            text(mean(xPos(g.rRNAIndexs([1 end]))), textY, &#39;</span><a class="code" href="classr_r_n_a.html">rRNA</a><span class="stringliteral">&#39;, &#39;</span>HorizontalAlign<span class="stringliteral">&#39;, &#39;</span>center<span class="stringliteral">&#39;, &#39;</span>Color<span class="stringliteral">&#39;, &#39;</span>g<span class="stringliteral">&#39;);</span>
<a name="l02006"></a>02006 <span class="stringliteral">            text(mean(xPos(g.sRNAIndexs([1 end]))), textY, &#39;</span><a class="code" href="classs_r_n_a.html">sRNA</a><span class="stringliteral">&#39;, &#39;</span>HorizontalAlign<span class="stringliteral">&#39;, &#39;</span>center<span class="stringliteral">&#39;, &#39;</span>Color<span class="stringliteral">&#39;, &#39;</span>b<span class="stringliteral">&#39;);</span>
<a name="l02007"></a>02007 <span class="stringliteral">            text(mean(xPos(g.tRNAIndexs([1 end]))), textY, &#39;</span><a class="code" href="classt_r_n_a.html">tRNA</a><span class="stringliteral">&#39;, &#39;</span>HorizontalAlign<span class="stringliteral">&#39;, &#39;</span>center<span class="stringliteral">&#39;, &#39;</span>Color<span class="stringliteral">&#39;, &#39;</span>c<span class="stringliteral">&#39;);</span>
<a name="l02008"></a>02008 <span class="stringliteral">        end</span>
<a name="l02009"></a>02009 <span class="stringliteral">        </span>
<a name="l02010"></a>02010 <span class="stringliteral">        function circosPlot(simBatchDir, selectedSim, outDirectory, figHandle, position)</span>
<a name="l02011"></a>02011 <span class="stringliteral">            import edu.stanford.covert.cell.sim.util.SimulationEnsemble;</span>
<a name="l02012"></a>02012 <span class="stringliteral">            import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;</span>
<a name="l02013"></a>02013 <span class="stringliteral">            import edu.stanford.covert.cell.sim.util.PlotUtil;</span>
<a name="l02014"></a>02014 <span class="stringliteral">            import edu.stanford.covert.cell.sim.analysis.ChromosomePositionHistogram;</span>
<a name="l02015"></a>02015 <span class="stringliteral">            import edu.stanford.covert.cell.sim.analysis.ChromosomeSpaceTimePlot;</span>
<a name="l02016"></a>02016 <span class="stringliteral">            import edu.stanford.covert.cell.sim.analysis.Figures34;</span>
<a name="l02017"></a>02017 <span class="stringliteral">            </span>
<a name="l02018"></a>02018 <span class="stringliteral">            if ~exist(&#39;</span>simBatchDir<span class="stringliteral">&#39;, &#39;</span>var<span class="stringliteral">&#39;) || isempty(simBatchDir)</span>
<a name="l02019"></a>02019 <span class="stringliteral">                simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();</span>
<a name="l02020"></a>02020 <span class="stringliteral">            end</span>
<a name="l02021"></a>02021 <span class="stringliteral">            if ~exist(&#39;</span>selectedSim<span class="stringliteral">&#39;, &#39;</span>var<span class="stringliteral">&#39;) || isempty(selectedSim)</span>
<a name="l02022"></a>02022 <span class="stringliteral">                selectedSim = 1;</span>
<a name="l02023"></a>02023 <span class="stringliteral">            end</span>
<a name="l02024"></a>02024 <span class="stringliteral">            </span>
<a name="l02025"></a>02025 <span class="stringliteral">            [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep num2str(selectedSim)]);</span>
<a name="l02026"></a>02026 <span class="stringliteral">            ch = sim.state(&#39;</span><a class="code" href="class_chromosome.html">Chromosome</a><span class="stringliteral">&#39;);</span>
<a name="l02027"></a>02027 <span class="stringliteral">            pc = sim.state(&#39;</span><a class="code" href="class_protein_complex.html">ProteinComplex</a><span class="stringliteral">&#39;);</span>
<a name="l02028"></a>02028 <span class="stringliteral">            </span>
<a name="l02029"></a>02029 <span class="stringliteral">            %% get data</span>
<a name="l02030"></a>02030 <span class="stringliteral">            if exist([outDirectory &#39;</span>circosPlot.mat<span class="stringliteral">&#39;], &#39;</span>file<span class="stringliteral">&#39;)</span>
<a name="l02031"></a>02031 <span class="stringliteral">                load([outDirectory &#39;</span>circosPlot.mat<span class="stringliteral">&#39;]);</span>
<a name="l02032"></a>02032 <span class="stringliteral">            else</span>
<a name="l02033"></a>02033 <span class="stringliteral">                stateNames = {</span>
<a name="l02034"></a>02034 <span class="stringliteral">                    &#39;</span>Time<span class="stringliteral">&#39;       &#39;</span>values<span class="stringliteral">&#39;                   &#39;</span>:<span class="charliteral">&#39; &#39;</span>:<span class="stringliteral">&#39;</span>
<a name="l02035"></a>02035 <span class="stringliteral">                    &#39;</span><a class="code" href="class_chromosome.html">Chromosome</a><span class="charliteral">&#39; &#39;</span>complexBoundSites<span class="stringliteral">&#39;        &#39;</span>:<span class="charliteral">&#39; &#39;</span>:<span class="stringliteral">&#39;</span>
<a name="l02036"></a>02036 <span class="stringliteral">                    };</span>
<a name="l02037"></a>02037 <span class="stringliteral">                states = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, &#39;</span>extract<span class="stringliteral">&#39;, selectedSim);</span>
<a name="l02038"></a>02038 <span class="stringliteral">                </span>
<a name="l02039"></a>02039 <span class="stringliteral">                chromProteins = {...</span>
<a name="l02040"></a>02040 <span class="stringliteral">                    %whole cell ids                strands</span>
<a name="l02041"></a>02041 <span class="stringliteral">                    {&#39;</span>MG_094_HEXAMER<span class="stringliteral">&#39;},            1:4;     %DNA Pol</span>
<a name="l02042"></a>02042 <span class="stringliteral">                    {&#39;</span>MG_469_1MER_ADP<span class="stringliteral">&#39; ...                  %DnaA</span>
<a name="l02043"></a>02043 <span class="stringliteral">                     &#39;</span>MG_469_1MER_ATP<span class="stringliteral">&#39; ...</span>
<a name="l02044"></a>02044 <span class="stringliteral">                     &#39;</span>MG_469_2MER_ATP<span class="stringliteral">&#39; ...</span>
<a name="l02045"></a>02045 <span class="stringliteral">                     &#39;</span>MG_469_3MER_ATP<span class="stringliteral">&#39; ...</span>
<a name="l02046"></a>02046 <span class="stringliteral">                     &#39;</span>MG_469_4MER_ATP<span class="stringliteral">&#39; ...</span>
<a name="l02047"></a>02047 <span class="stringliteral">                     &#39;</span>MG_469_5MER_ATP<span class="stringliteral">&#39; ...</span>
<a name="l02048"></a>02048 <span class="stringliteral">                     &#39;</span>MG_469_6MER_ATP<span class="stringliteral">&#39; ...</span>
<a name="l02049"></a>02049 <span class="stringliteral">                     &#39;</span>MG_469_7MER_ATP<span class="stringliteral">&#39;},           1:2;</span>
<a name="l02050"></a>02050 <span class="stringliteral">                    {&#39;</span>RNA_POLYMERASE<span class="stringliteral">&#39;                       %RNA Pol</span>
<a name="l02051"></a>02051 <span class="stringliteral">                     &#39;</span>RNA_POLYMERASE_HOLOENZYME<span class="stringliteral">&#39;}, 1:2;</span>
<a name="l02052"></a>02052 <span class="stringliteral">                    };</span>
<a name="l02053"></a>02053 <span class="stringliteral">                </span>
<a name="l02054"></a>02054 <span class="stringliteral">                densityBEMatrix = zeros(1000, size(chromProteins, 1));</span>
<a name="l02055"></a>02055 <span class="stringliteral">                for i = 1:size(chromProteins, 1)</span>
<a name="l02056"></a>02056 <span class="stringliteral">                    for j = 1:numel(chromProteins{i, 1})</span>
<a name="l02057"></a>02057 <span class="stringliteral">                        protIdx = pc.getIndexs(chromProteins{i, 1}{j});</span>
<a name="l02058"></a>02058 <span class="stringliteral">                        </span>
<a name="l02059"></a>02059 <span class="stringliteral">                        %calculate frequency binding each position</span>
<a name="l02060"></a>02060 <span class="stringliteral">                        tmp = full(sum(sum(states.Chromosome.complexBoundSites(:, chromProteins{i, 2}, :) == protIdx, 3), 2));</span>
<a name="l02061"></a>02061 <span class="stringliteral">                        </span>
<a name="l02062"></a>02062 <span class="stringliteral">                        %circularly convolve with footprint</span>
<a name="l02063"></a>02063 <span class="stringliteral">                        tmp = cconv(tmp, ones(ch.complexDNAFootprints(protIdx), 1), ch.sequenceLen);</span>
<a name="l02064"></a>02064 <span class="stringliteral">                        </span>
<a name="l02065"></a>02065 <span class="stringliteral">                        %bin to calculate density</span>
<a name="l02066"></a>02066 <span class="stringliteral">                        densityBEMatrix(:, i) = ...</span>
<a name="l02067"></a>02067 <span class="stringliteral">                            + densityBEMatrix(:, i) ...</span>
<a name="l02068"></a>02068 <span class="stringliteral">                            + sum(reshape([tmp; tmp(1:1000 - mod(ch.sequenceLen, 1000), :)], [], 1000), 1)&#39;</span>;
<a name="l02069"></a>02069                     end
<a name="l02070"></a>02070                 end
<a name="l02071"></a>02071                 densityBEMatrix = densityBEMatrix / ceil(ch.sequenceLen/1000) / numel(states.Time.values); %proteins/(nt h)
<a name="l02072"></a>02072                 
<a name="l02073"></a>02073                 %save
<a name="l02074"></a>02074                 save([outDirectory <span class="stringliteral">&#39;circosPlot.mat&#39;</span>], <span class="stringliteral">&#39;densityBEMatrix&#39;</span>);
<a name="l02075"></a>02075             end
<a name="l02076"></a>02076             
<a name="l02077"></a>02077             totalDensityMatrix = <a class="code" href="class_figures34.html">Figures34</a>.<a class="code" href="class_figures34.html#a409ccf33890b71aa06df826926c30f74">calculateDNABoundProteinDensity</a>(outDirectory, simBatchDir, selectedSim); %1/(nt s)
<a name="l02078"></a>02078             totalDensityMatrix = sum(reshape([totalDensityMatrix; totalDensityMatrix(1:1000 - mod(numel(totalDensityMatrix), 1000), :)], [], 1000), 1)<span class="stringliteral">&#39; ...</span>
<a name="l02079"></a>02079 <span class="stringliteral">                 / ceil(ch.sequenceLen/1000); %1/(nt s)</span>
<a name="l02080"></a>02080 <span class="stringliteral">            </span>
<a name="l02081"></a>02081 <span class="stringliteral">            densityBEMatrix = [densityBEMatrix totalDensityMatrix];</span>
<a name="l02082"></a>02082 <span class="stringliteral">            </span>
<a name="l02083"></a>02083 <span class="stringliteral">            %% plot</span>
<a name="l02084"></a>02084 <span class="stringliteral">            if nargin &lt; 5</span>
<a name="l02085"></a>02085 <span class="stringliteral">                [~, figHandle] = edu.stanford.covert.cell.sim.util.PlotUtil.newAxesHandle();</span>
<a name="l02086"></a>02086 <span class="stringliteral">                clf(figHandle);</span>
<a name="l02087"></a>02087 <span class="stringliteral">                position = [0.1 0.1 0.8 0.8];</span>
<a name="l02088"></a>02088 <span class="stringliteral">            end</span>
<a name="l02089"></a>02089 <span class="stringliteral">            x = position(1);</span>
<a name="l02090"></a>02090 <span class="stringliteral">            y = position(2);</span>
<a name="l02091"></a>02091 <span class="stringliteral">            w = position(3);</span>
<a name="l02092"></a>02092 <span class="stringliteral">            h = position(4);</span>
<a name="l02093"></a>02093 <span class="stringliteral">            </span>
<a name="l02094"></a>02094 <span class="stringliteral">            figPos = get(figHandle, &#39;</span>papersize<span class="stringliteral">&#39;);</span>
<a name="l02095"></a>02095 <span class="stringliteral">            figW = figPos(1);</span>
<a name="l02096"></a>02096 <span class="stringliteral">            figH = figPos(2);</span>
<a name="l02097"></a>02097 <span class="stringliteral">            </span>
<a name="l02098"></a>02098 <span class="stringliteral">            y2 = y+(h-0.75*w*figW/figH)/2 + 0.81*w*figW/figH/2 - 0.5*h;</span>
<a name="l02099"></a>02099 <span class="stringliteral">            </span>
<a name="l02100"></a>02100 <span class="stringliteral">            colorbarHandles = [</span>
<a name="l02101"></a>02101 <span class="stringliteral">                subplot(&#39;</span>position<span class="stringliteral">&#39;, [x+0.837*w  y2            0.03*w  0.2*h], &#39;</span>parent<span class="stringliteral">&#39;, figHandle)</span>
<a name="l02102"></a>02102 <span class="stringliteral">                subplot(&#39;</span>position<span class="stringliteral">&#39;, [x+0.837*w  y2+0.8*1/3*h  0.03*w  0.2*h], &#39;</span>parent<span class="stringliteral">&#39;, figHandle)</span>
<a name="l02103"></a>02103 <span class="stringliteral">                subplot(&#39;</span>position<span class="stringliteral">&#39;, [x+0.837*w  y2+0.8*2/3*h  0.03*w  0.2*h], &#39;</span>parent<span class="stringliteral">&#39;, figHandle)</span>
<a name="l02104"></a>02104 <span class="stringliteral">                subplot(&#39;</span>position<span class="stringliteral">&#39;, [x+0.837*w  y2+0.8*h      0.03*w  0.2*h], &#39;</span>parent<span class="stringliteral">&#39;, figHandle)</span>
<a name="l02105"></a>02105 <span class="stringliteral">                ];</span>
<a name="l02106"></a>02106 <span class="stringliteral">            axesHandle = subplot(&#39;</span>position<span class="stringliteral">&#39;, [x  y+(h-0.75*w*figW/figH)/2  0.81*w  0.81*w*figW/figH], &#39;</span>parent<span class="stringliteral">&#39;, figHandle);</span>
<a name="l02107"></a>02107 <span class="stringliteral">            </span>
<a name="l02108"></a>02108 <span class="stringliteral">            cla(axesHandle);</span>
<a name="l02109"></a>02109 <span class="stringliteral">            hold(axesHandle, &#39;</span>on<span class="stringliteral">&#39;);</span>
<a name="l02110"></a>02110 <span class="stringliteral">            set(axesHandle, &#39;</span>visible<span class="stringliteral">&#39;, &#39;</span>off<span class="stringliteral">&#39;);</span>
<a name="l02111"></a>02111 <span class="stringliteral">            </span>
<a name="l02112"></a>02112 <span class="stringliteral">            cmap = cell(4, 1);</span>
<a name="l02113"></a>02113 <span class="stringliteral">            cmapLims = zeros(4, 2);</span>
<a name="l02114"></a>02114 <span class="stringliteral">            cmapTicks = cell(4, 1);</span>
<a name="l02115"></a>02115 <span class="stringliteral">            cmapFuncs = {</span>
<a name="l02116"></a>02116 <span class="stringliteral">                [0 1 0]</span>
<a name="l02117"></a>02117 <span class="stringliteral">                [1 0 0]</span>
<a name="l02118"></a>02118 <span class="stringliteral">                [0 0 1]</span>
<a name="l02119"></a>02119 <span class="stringliteral">                [0 1 1]</span>
<a name="l02120"></a>02120 <span class="stringliteral">                };</span>
<a name="l02121"></a>02121 <span class="stringliteral">            totcmap = zeros(0, 3);</span>
<a name="l02122"></a>02122 <span class="stringliteral">            </span>
<a name="l02123"></a>02123 <span class="stringliteral">            nTheta = size(densityBEMatrix, 1);</span>
<a name="l02124"></a>02124 <span class="stringliteral">            for i = 1:size(densityBEMatrix, 2)</span>
<a name="l02125"></a>02125 <span class="stringliteral">                x = zeros(10, size(densityBEMatrix, 1));</span>
<a name="l02126"></a>02126 <span class="stringliteral">                y = zeros(10, size(densityBEMatrix, 1));</span>
<a name="l02127"></a>02127 <span class="stringliteral">                c = zeros(size(densityBEMatrix, 1), 1);</span>
<a name="l02128"></a>02128 <span class="stringliteral">                </span>
<a name="l02129"></a>02129 <span class="stringliteral">                data = densityBEMatrix(:, i);</span>
<a name="l02130"></a>02130 <span class="stringliteral">                </span>
<a name="l02131"></a>02131 <span class="stringliteral">                cmapLims(i, :) = [</span>
<a name="l02132"></a>02132 <span class="stringliteral">                    max(-6, floor(log10(min(data(data &gt; 0)))))</span>
<a name="l02133"></a>02133 <span class="stringliteral">                    ceil(log10(max(data)))</span>
<a name="l02134"></a>02134 <span class="stringliteral">                    ];</span>
<a name="l02135"></a>02135 <span class="stringliteral">                </span>
<a name="l02136"></a>02136 <span class="stringliteral">                nBins = 2 + 10 * diff(cmapLims(i, :));</span>
<a name="l02137"></a>02137 <span class="stringliteral">                cmap{i} = ones(nBins, 3) - repmat([1 1 1] - cmapFuncs{i}, nBins, 1) .* repmat(linspace(0, 1, nBins)&#39;</span>, 1, 3);
<a name="l02138"></a>02138                 cmapTicks{i} = [0 logspace(cmapLims(i, 1), cmapLims(i, 2), 1 + 10 * diff(cmapLims(i, :)))];
<a name="l02139"></a>02139                 
<a name="l02140"></a>02140                 <span class="keywordflow">for</span> j = 1:nTheta
<a name="l02141"></a>02141                     x(:, j) = [
<a name="l02142"></a>02142                         (i+1)*cos(linspace(2*pi/nTheta*j, 2*pi/nTheta*(j+1), 5)+pi/2) ...
<a name="l02143"></a>02143                         (i+2)*cos(linspace(2*pi/nTheta*(j+1), 2*pi/nTheta*j, 5)+pi/2) ...
<a name="l02144"></a>02144                         ];
<a name="l02145"></a>02145                     y(:, j) = [
<a name="l02146"></a>02146                         (i+1)*sin(linspace(2*pi/nTheta*j, 2*pi/nTheta*(j+1), 5)+pi/2) ...
<a name="l02147"></a>02147                         (i+2)*sin(linspace(2*pi/nTheta*(j+1), 2*pi/nTheta*j, 5)+pi/2) ...
<a name="l02148"></a>02148                         ];
<a name="l02149"></a>02149                     c(j) = find(cmapTicks{i} &gt;= data(j), 1, <span class="stringliteral">&#39;first&#39;</span>);
<a name="l02150"></a>02150                 end
<a name="l02151"></a>02151                 
<a name="l02152"></a>02152                 c = c + size(totcmap, 1);
<a name="l02153"></a>02153                 totcmap = [totcmap; cmap{i}]; %#ok&lt;AGROW&gt;
<a name="l02154"></a>02154                 
<a name="l02155"></a>02155                 ptch = patch(x, y, ones(1, nTheta), <span class="stringliteral">&#39;EdgeColor&#39;</span>, <span class="stringliteral">&#39;none&#39;</span>, <span class="stringliteral">&#39;Parent&#39;</span>, axesHandle);
<a name="l02156"></a>02156                 <span class="keyword">set</span>(ptch, <span class="stringliteral">&#39;FaceColor&#39;</span>, <span class="stringliteral">&#39;flat&#39;</span>, <span class="stringliteral">&#39;FaceVertexCData&#39;</span>, c, <span class="stringliteral">&#39;CDataMapping&#39;</span>, <span class="stringliteral">&#39;direct&#39;</span>, <span class="stringliteral">&#39;FaceAlpha&#39;</span>, 1, <span class="stringliteral">&#39;EdgeAlpha&#39;</span>, 1);
<a name="l02157"></a>02157             end
<a name="l02158"></a>02158             colormap(axesHandle, totcmap);
<a name="l02159"></a>02159             
<a name="l02160"></a>02160             xlim(axesHandle, [-7.5 6.5]);
<a name="l02161"></a>02161             ylim(axesHandle, 7 * [-1 1]);
<a name="l02162"></a>02162             
<a name="l02163"></a>02163             %oriC label
<a name="l02164"></a>02164             line([0 0], [2 6.2], <span class="stringliteral">&#39;Color&#39;</span>, <span class="charliteral">&#39;k&#39;</span>, <span class="stringliteral">&#39;Parent&#39;</span>, axesHandle);
<a name="l02165"></a>02165             text(0, 6.25, <span class="stringliteral">&#39;oriC&#39;</span>, ...
<a name="l02166"></a>02166                 <span class="stringliteral">&#39;Parent&#39;</span>, axesHandle, ...
<a name="l02167"></a>02167                 <span class="stringliteral">&#39;VerticalAlign&#39;</span>, <span class="stringliteral">&#39;bottom&#39;</span>, ...
<a name="l02168"></a>02168                 <span class="stringliteral">&#39;HorizontalAlign&#39;</span>, <span class="stringliteral">&#39;center&#39;</span>, ...
<a name="l02169"></a>02169                 <span class="stringliteral">&#39;FontSize&#39;</span>, 5);
<a name="l02170"></a>02170             
<a name="l02171"></a>02171             %color bars
<a name="l02172"></a>02172             ylabels = {
<a name="l02173"></a>02173                 {<span class="stringliteral">&#39;DNA pol&#39;</span> <span class="stringliteral">&#39;(nt^{-1} s^{-1})&#39;</span>}
<a name="l02174"></a>02174                 {<span class="stringliteral">&#39;DnaA&#39;</span> <span class="stringliteral">&#39;(nt^{-1} s^{-1})&#39;</span>}
<a name="l02175"></a>02175                 {<span class="stringliteral">&#39;RNA pol&#39;</span> <span class="stringliteral">&#39;(nt^{-1} s^{-1})&#39;</span>}
<a name="l02176"></a>02176                 {<span class="stringliteral">&#39;All proteins&#39;</span> <span class="stringliteral">&#39;(nt^{-1} s^{-1})&#39;</span>}
<a name="l02177"></a>02177                 };
<a name="l02178"></a>02178             <span class="keywordflow">for</span> i = 1:numel(colorbarHandles)
<a name="l02179"></a>02179                 image(permute(cmap{i}, [1 3 2]), <span class="stringliteral">&#39;Parent&#39;</span>, colorbarHandles(i));
<a name="l02180"></a>02180                 
<a name="l02181"></a>02181                 <span class="keyword">set</span>(colorbarHandles(i), ...
<a name="l02182"></a>02182                     <span class="stringliteral">&#39;TickDir&#39;</span>, <span class="stringliteral">&#39;out&#39;</span>, ...
<a name="l02183"></a>02183                     <span class="stringliteral">&#39;box&#39;</span>, <span class="stringliteral">&#39;off&#39;</span>, ...
<a name="l02184"></a>02184                     <span class="stringliteral">&#39;FontSize&#39;</span>, 5, ...
<a name="l02185"></a>02185                     <span class="stringliteral">&#39;YAxisLocation&#39;</span>, <span class="stringliteral">&#39;right&#39;</span>, ...
<a name="l02186"></a>02186                     <span class="stringliteral">&#39;XTick&#39;</span>, [], ...
<a name="l02187"></a>02187                     <span class="stringliteral">&#39;YTick&#39;</span>, 2:10:size(cmap{i}, 1), ...
<a name="l02188"></a>02188                     <span class="stringliteral">&#39;XLim&#39;</span>, [0.5 1.5], ...
<a name="l02189"></a>02189                     <span class="stringliteral">&#39;visible&#39;</span>, <span class="stringliteral">&#39;on&#39;</span>, ...
<a name="l02190"></a>02190                     <span class="stringliteral">&#39;YDir&#39;</span>, <span class="stringliteral">&#39;normal&#39;</span>);
<a name="l02191"></a>02191                 ylim(colorbarHandles(i), [0 size(cmap{i}, 1)]+0.5);
<a name="l02192"></a>02192                 
<a name="l02193"></a>02193                 line([min(xlim(colorbarHandles(i))) * [1 1]   max(xlim(colorbarHandles(i)))], ...
<a name="l02194"></a>02194                     [ylim(colorbarHandles(i)) max(ylim(colorbarHandles(i)))], ...
<a name="l02195"></a>02195                     <span class="stringliteral">&#39;parent&#39;</span>, colorbarHandles(i), <span class="stringliteral">&#39;color&#39;</span>, <span class="charliteral">&#39;k&#39;</span>);
<a name="l02196"></a>02196                 
<a name="l02197"></a>02197                 tick2text(colorbarHandles(i), <span class="stringliteral">&#39;axis&#39;</span>, <span class="charliteral">&#39;y&#39;</span>, <span class="stringliteral">&#39;ytickoffset&#39;</span>, -1.25);
<a name="l02198"></a>02198                 yTicks = getappdata(colorbarHandles(i), <span class="stringliteral">&#39;YTickText&#39;</span>);
<a name="l02199"></a>02199                 <span class="keyword">set</span>(yTicks, <span class="stringliteral">&#39;FontSize&#39;</span>, 5, <span class="stringliteral">&#39;HorizontalAlign&#39;</span>, <span class="stringliteral">&#39;left&#39;</span>, <span class="stringliteral">&#39;Interpreter&#39;</span>, <span class="stringliteral">&#39;tex&#39;</span>);
<a name="l02200"></a>02200                 
<a name="l02201"></a>02201                 <span class="keywordflow">for</span> j = 1:numel(yTicks)
<a name="l02202"></a>02202                     set(yTicks(j), &#39;String&#39;, sprintf(&#39;10^{%d}<span class="stringliteral">&#39;, cmapLims(i, 1)+j-1));</span>
<a name="l02203"></a>02203 <span class="stringliteral">                end</span>
<a name="l02204"></a>02204 <span class="stringliteral">                if numel(yTicks) == 6</span>
<a name="l02205"></a>02205 <span class="stringliteral">                    delete(yTicks(1:2:5));</span>
<a name="l02206"></a>02206 <span class="stringliteral">                end</span>
<a name="l02207"></a>02207 <span class="stringliteral">                </span>
<a name="l02208"></a>02208 <span class="stringliteral">                ylabel(colorbarHandles(i), ylabels{i}, &#39;</span>FontSize<span class="stringliteral">&#39;, 5);</span>
<a name="l02209"></a>02209 <span class="stringliteral">                ylabelpos = get(get(colorbarHandles(i), &#39;</span>ylabel<span class="stringliteral">&#39;), &#39;</span>position<span class="stringliteral">&#39;);</span>
<a name="l02210"></a>02210 <span class="stringliteral">                set(get(colorbarHandles(i), &#39;</span>ylabel<span class="stringliteral">&#39;), &#39;</span>position<span class="stringliteral">&#39;, [3.3 ylabelpos(2:end)]);</span>
<a name="l02211"></a>02211 <span class="stringliteral">            end</span>
<a name="l02212"></a>02212 <span class="stringliteral">        end</span>
<a name="l02213"></a>02213 <span class="stringliteral">        </span>
<a name="l02214"></a>02214 <span class="stringliteral">        function dnaBoundProteinDensityVsDisplacement(simBatchDir, selectedSim, outDirectory, figHandle, position)</span>
<a name="l02215"></a>02215 <span class="stringliteral">            import edu.stanford.covert.cell.sim.analysis.Figures34;</span>
<a name="l02216"></a>02216 <span class="stringliteral">            import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;</span>
<a name="l02217"></a>02217 <span class="stringliteral">            </span>
<a name="l02218"></a>02218 <span class="stringliteral">            if ~exist(&#39;</span>simBatchDir<span class="stringliteral">&#39;, &#39;</span>var<span class="stringliteral">&#39;) || isempty(simBatchDir)</span>
<a name="l02219"></a>02219 <span class="stringliteral">                simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();</span>
<a name="l02220"></a>02220 <span class="stringliteral">            end</span>
<a name="l02221"></a>02221 <span class="stringliteral">            if ~exist(&#39;</span>selectedSim<span class="stringliteral">&#39;, &#39;</span>var<span class="stringliteral">&#39;) || isempty(selectedSim)</span>
<a name="l02222"></a>02222 <span class="stringliteral">                selectedSim = 1;</span>
<a name="l02223"></a>02223 <span class="stringliteral">            end</span>
<a name="l02224"></a>02224 <span class="stringliteral">            </span>
<a name="l02225"></a>02225 <span class="stringliteral">            [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep num2str(selectedSim)]);</span>
<a name="l02226"></a>02226 <span class="stringliteral">            c = sim.state(&#39;</span><a class="code" href="class_chromosome.html">Chromosome</a><span class="stringliteral">&#39;);</span>
<a name="l02227"></a>02227 <span class="stringliteral">            </span>
<a name="l02228"></a>02228 <span class="stringliteral">            %% get data</span>
<a name="l02229"></a>02229 <span class="stringliteral">            dnaBoundProteinDensity = Figures34.calculateDNABoundProteinDensity(outDirectory, simBatchDir, selectedSim);</span>
<a name="l02230"></a>02230 <span class="stringliteral">            </span>
<a name="l02231"></a>02231 <span class="stringliteral">            dnaBoundProteinDisplacements = Figures34.calculateDNABoundProteinDisplacements(outDirectory, simBatchDir, selectedSim);</span>
<a name="l02232"></a>02232 <span class="stringliteral">            dnaBoundProteinDisplacements = conv(dnaBoundProteinDisplacements, ones(1000, 1), &#39;</span>same<span class="stringliteral">&#39;);</span>
<a name="l02233"></a>02233 <span class="stringliteral">            </span>
<a name="l02234"></a>02234 <span class="stringliteral">            %% plot</span>
<a name="l02235"></a>02235 <span class="stringliteral">            if nargin &gt;= 5</span>
<a name="l02236"></a>02236 <span class="stringliteral">                x = position(1);</span>
<a name="l02237"></a>02237 <span class="stringliteral">                y = position(2);</span>
<a name="l02238"></a>02238 <span class="stringliteral">                w = position(3);</span>
<a name="l02239"></a>02239 <span class="stringliteral">                h = position(4);</span>
<a name="l02240"></a>02240 <span class="stringliteral">                axesHandle = subplot(&#39;</span>position<span class="stringliteral">&#39;, [x y 0.82*w h], &#39;</span>parent<span class="stringliteral">&#39;, figHandle);</span>
<a name="l02241"></a>02241 <span class="stringliteral">                colorBarHandle = subplot(&#39;</span>position<span class="stringliteral">&#39;, [x+0.85*w y 0.05*w h], &#39;</span>parent<span class="stringliteral">&#39;, figHandle);</span>
<a name="l02242"></a>02242 <span class="stringliteral">            else</span>
<a name="l02243"></a>02243 <span class="stringliteral">                axesHandle = edu.stanford.covert.cell.sim.util.PlotUtil.newAxesHandle();</span>
<a name="l02244"></a>02244 <span class="stringliteral">                colorBarHandle = colorbar();</span>
<a name="l02245"></a>02245 <span class="stringliteral">            end</span>
<a name="l02246"></a>02246 <span class="stringliteral">            </span>
<a name="l02247"></a>02247 <span class="stringliteral">            % data</span>
<a name="l02248"></a>02248 <span class="stringliteral">            [cnts, centers] = hist3([dnaBoundProteinDensity dnaBoundProteinDisplacements], [100 100]);</span>
<a name="l02249"></a>02249 <span class="stringliteral">            cnts = cnts / c.sequenceLen;</span>
<a name="l02250"></a>02250 <span class="stringliteral">            edges = logspace(-6, -1, 51);</span>
<a name="l02251"></a>02251 <span class="stringliteral">            colormap = jet(numel(edges) + 1);</span>
<a name="l02252"></a>02252 <span class="stringliteral">            colors = zeros(100, 100, 3);</span>
<a name="l02253"></a>02253 <span class="stringliteral">            </span>
<a name="l02254"></a>02254 <span class="stringliteral">            [i, j] = find(cnts &lt; edges(1));</span>
<a name="l02255"></a>02255 <span class="stringliteral">            colors(sub2ind(size(colors), i, j, ones(size(i)))) = colormap(1, 1);</span>
<a name="l02256"></a>02256 <span class="stringliteral">            colors(sub2ind(size(colors), i, j, 2*ones(size(i)))) = colormap(1, 2);</span>
<a name="l02257"></a>02257 <span class="stringliteral">            colors(sub2ind(size(colors), i, j, 3*ones(size(i)))) = colormap(1, 3);</span>
<a name="l02258"></a>02258 <span class="stringliteral">            </span>
<a name="l02259"></a>02259 <span class="stringliteral">            [i, j] = find(cnts &gt; edges(end-1));</span>
<a name="l02260"></a>02260 <span class="stringliteral">            colors(sub2ind(size(colors), i, j, ones(size(i)))) = colormap(end, 1);</span>
<a name="l02261"></a>02261 <span class="stringliteral">            colors(sub2ind(size(colors), i, j, 2*ones(size(i)))) = colormap(end, 2);</span>
<a name="l02262"></a>02262 <span class="stringliteral">            colors(sub2ind(size(colors), i, j, 3*ones(size(i)))) = colormap(end, 3);</span>
<a name="l02263"></a>02263 <span class="stringliteral">            </span>
<a name="l02264"></a>02264 <span class="stringliteral">            for idx = 1:numel(edges)-1</span>
<a name="l02265"></a>02265 <span class="stringliteral">                [i, j] = find(cnts &gt;= edges(idx) &amp; cnts &lt;= edges(idx+1));</span>
<a name="l02266"></a>02266 <span class="stringliteral">                colors(sub2ind(size(colors), i, j, ones(size(i)))) = colormap(idx+1, 1);</span>
<a name="l02267"></a>02267 <span class="stringliteral">                colors(sub2ind(size(colors), i, j, 2*ones(size(i)))) = colormap(idx+1, 2);</span>
<a name="l02268"></a>02268 <span class="stringliteral">                colors(sub2ind(size(colors), i, j, 3*ones(size(i)))) = colormap(idx+1, 3);</span>
<a name="l02269"></a>02269 <span class="stringliteral">            end</span>
<a name="l02270"></a>02270 <span class="stringliteral">            </span>
<a name="l02271"></a>02271 <span class="stringliteral">            image(permute(colors, [2 1 3]), &#39;</span>Parent<span class="stringliteral">&#39;, axesHandle);</span>
<a name="l02272"></a>02272 <span class="stringliteral">            set(axesHandle, &#39;</span>XDir<span class="stringliteral">&#39;, &#39;</span>normal<span class="stringliteral">&#39;);</span>
<a name="l02273"></a>02273 <span class="stringliteral">            set(axesHandle, &#39;</span>YDir<span class="stringliteral">&#39;, &#39;</span>normal<span class="stringliteral">&#39;);</span>
<a name="l02274"></a>02274 <span class="stringliteral">            </span>
<a name="l02275"></a>02275 <span class="stringliteral">            set(axesHandle, &#39;</span>Box<span class="stringliteral">&#39;, &#39;</span>on<span class="stringliteral">&#39;, &#39;</span>FontSize<span class="stringliteral">&#39;, 5, &#39;</span>TickDir<span class="stringliteral">&#39;, &#39;</span>out<span class="stringliteral">&#39;);</span>
<a name="l02276"></a>02276 <span class="stringliteral">            xlim(axesHandle, [0.5 0.4/centers{1}(end)*100])</span>
<a name="l02277"></a>02277 <span class="stringliteral">            ylim(axesHandle, [0.5 1500/centers{2}(end)*100])</span>
<a name="l02278"></a>02278 <span class="stringliteral">            </span>
<a name="l02279"></a>02279 <span class="stringliteral">            set(axesHandle, &#39;</span>XTick<span class="stringliteral">&#39;, 0.5 + (0:0.1:0.4)/centers{1}(end)*100);</span>
<a name="l02280"></a>02280 <span class="stringliteral">            set(axesHandle, &#39;</span>YTick<span class="stringliteral">&#39;, 0.5 + (0:500:1500)/centers{2}(end)*100);</span>
<a name="l02281"></a>02281 <span class="stringliteral">            tick2text(axesHandle, &#39;</span>axis<span class="stringliteral">&#39;, &#39;</span>xy<span class="stringliteral">&#39;, &#39;</span>xtickoffset<span class="stringliteral">&#39;, 0.06, &#39;</span>ytickoffset<span class="stringliteral">&#39;, 0.015);</span>
<a name="l02282"></a>02282 <span class="stringliteral">            xTicks = getappdata(axesHandle, &#39;</span>XTickText<span class="stringliteral">&#39;);</span>
<a name="l02283"></a>02283 <span class="stringliteral">            yTicks = getappdata(axesHandle, &#39;</span>YTickText<span class="stringliteral">&#39;);</span>
<a name="l02284"></a>02284 <span class="stringliteral">            set(xTicks, &#39;</span>FontSize<span class="stringliteral">&#39;, 5);</span>
<a name="l02285"></a>02285 <span class="stringliteral">            set(yTicks, &#39;</span>FontSize<span class="stringliteral">&#39;, 5, &#39;</span>HorizontalAlign<span class="stringliteral">&#39;, &#39;</span>right<span class="stringliteral">&#39;);</span>
<a name="l02286"></a>02286 <span class="stringliteral">            set(xTicks(1), &#39;</span>String<span class="stringliteral">&#39;, &#39;</span>0<span class="stringliteral">&#39;)</span>
<a name="l02287"></a>02287 <span class="stringliteral">            set(xTicks(end), &#39;</span>String<span class="stringliteral">&#39;, &#39;</span>0.4<span class="stringliteral">&#39;)</span>
<a name="l02288"></a>02288 <span class="stringliteral">            set(yTicks(1), &#39;</span>String<span class="stringliteral">&#39;, &#39;</span>0<span class="stringliteral">&#39;)</span>
<a name="l02289"></a>02289 <span class="stringliteral">            set(yTicks(end), &#39;</span>String<span class="stringliteral">&#39;, &#39;</span>1500<span class="stringliteral">&#39;)</span>
<a name="l02290"></a>02290 <span class="stringliteral">            delete(yTicks(2:3));</span>
<a name="l02291"></a>02291 <span class="stringliteral">            delete(xTicks(2:4));</span>
<a name="l02292"></a>02292 <span class="stringliteral">            </span>
<a name="l02293"></a>02293 <span class="stringliteral">            xlabel(axesHandle, &#39;</span>Density<span class="stringliteral">&#39;, &#39;</span>FontSize<span class="stringliteral">&#39;, 7);</span>
<a name="l02294"></a>02294 <span class="stringliteral">            ylabel(axesHandle, &#39;</span>Displacement<span class="stringliteral">&#39;, &#39;</span>FontSize<span class="stringliteral">&#39;, 7);</span>
<a name="l02295"></a>02295 <span class="stringliteral">            xlabelPos = get(get(axesHandle, &#39;</span>xlabel<span class="stringliteral">&#39;), &#39;</span>position<span class="stringliteral">&#39;);</span>
<a name="l02296"></a>02296 <span class="stringliteral">            ylabelPos = get(get(axesHandle, &#39;</span>ylabel<span class="stringliteral">&#39;), &#39;</span>position<span class="stringliteral">&#39;);</span>
<a name="l02297"></a>02297 <span class="stringliteral">            set(get(axesHandle, &#39;</span>xlabel<span class="stringliteral">&#39;), &#39;</span>position<span class="stringliteral">&#39;, [xlabelPos(1) -3 xlabelPos(3:end)]);</span>
<a name="l02298"></a>02298 <span class="stringliteral">            set(get(axesHandle, &#39;</span>ylabel<span class="stringliteral">&#39;), &#39;</span>position<span class="stringliteral">&#39;, [-3 ylabelPos(2:end)]);</span>
<a name="l02299"></a>02299 <span class="stringliteral">            </span>
<a name="l02300"></a>02300 <span class="stringliteral">            % color bar</span>
<a name="l02301"></a>02301 <span class="stringliteral">            image(permute(colormap, [1 3 2]), &#39;</span>Parent<span class="stringliteral">&#39;, colorBarHandle);</span>
<a name="l02302"></a>02302 <span class="stringliteral">            </span>
<a name="l02303"></a>02303 <span class="stringliteral">            box(colorBarHandle, &#39;</span>on<span class="stringliteral">&#39;);</span>
<a name="l02304"></a>02304 <span class="stringliteral">            set(colorBarHandle, &#39;</span>XTick<span class="stringliteral">&#39;, []);</span>
<a name="l02305"></a>02305 <span class="stringliteral">            set(colorBarHandle, &#39;</span>TickDir<span class="stringliteral">&#39;, &#39;</span>out<span class="stringliteral">&#39;);</span>
<a name="l02306"></a>02306 <span class="stringliteral">            set(colorBarHandle, &#39;</span>YAxisLocation<span class="stringliteral">&#39;, &#39;</span>right<span class="stringliteral">&#39;);</span>
<a name="l02307"></a>02307 <span class="stringliteral">            set(colorBarHandle, &#39;</span>YDir<span class="stringliteral">&#39;, &#39;</span>normal<span class="stringliteral">&#39;)</span>
<a name="l02308"></a>02308 <span class="stringliteral">            set(colorBarHandle, &#39;</span>FontSize<span class="stringliteral">&#39;, 5);</span>
<a name="l02309"></a>02309 <span class="stringliteral">            </span>
<a name="l02310"></a>02310 <span class="stringliteral">            set(colorBarHandle, &#39;</span>YTick<span class="stringliteral">&#39;, 2:10:52);</span>
<a name="l02311"></a>02311 <span class="stringliteral">            tick2text(colorBarHandle, &#39;</span>axis<span class="stringliteral">&#39;, &#39;</span>y<span class="stringliteral">&#39;, &#39;</span>ytickoffset<span class="stringliteral">&#39;, -1.15);</span>
<a name="l02312"></a>02312 <span class="stringliteral">            yTicks = getappdata(colorBarHandle, &#39;</span>YTickText<span class="stringliteral">&#39;);</span>
<a name="l02313"></a>02313 <span class="stringliteral">            set(yTicks, &#39;</span>FontSize<span class="stringliteral">&#39;, 5, &#39;</span>HorizontalAlign<span class="stringliteral">&#39;, &#39;</span>left<span class="stringliteral">&#39;, &#39;</span>Interpreter<span class="stringliteral">&#39;, &#39;</span>tex<span class="stringliteral">&#39;);</span>
<a name="l02314"></a>02314 <span class="stringliteral">            for i = 1:6</span>
<a name="l02315"></a>02315 <span class="stringliteral">                set(yTicks(i), &#39;</span>String<span class="stringliteral">&#39;, sprintf(&#39;</span>10^{%d}<span class="stringliteral">&#39;, -7+i));</span>
<a name="l02316"></a>02316 <span class="stringliteral">            end</span>
<a name="l02317"></a>02317 <span class="stringliteral">            </span>
<a name="l02318"></a>02318 <span class="stringliteral">            ylabel(colorBarHandle, &#39;</span>% Bases<span class="stringliteral">&#39;, &#39;</span>FontSize<span class="stringliteral">&#39;, 7);</span>
<a name="l02319"></a>02319 <span class="stringliteral">            ylabelPos = get(get(colorBarHandle, &#39;</span>ylabel<span class="stringliteral">&#39;), &#39;</span>position<span class="stringliteral">&#39;);</span>
<a name="l02320"></a>02320 <span class="stringliteral">            set(get(colorBarHandle, &#39;</span>ylabel<span class="stringliteral">&#39;), &#39;</span>position<span class="stringliteral">&#39;, [3.0 ylabelPos(2:end)]);</span>
<a name="l02321"></a>02321 <span class="stringliteral">        end</span>
<a name="l02322"></a>02322 <span class="stringliteral">        </span>
<a name="l02323"></a>02323 <span class="stringliteral">        function analyzeDNABoundProteinDisplacements(simBatchDir, selectedSim, outDirectory)</span>
<a name="l02324"></a>02324 <span class="stringliteral">            import edu.stanford.covert.cell.sim.analysis.Figures34;</span>
<a name="l02325"></a>02325 <span class="stringliteral">            import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;</span>
<a name="l02326"></a>02326 <span class="stringliteral">            </span>
<a name="l02327"></a>02327 <span class="stringliteral">            %% constants</span>
<a name="l02328"></a>02328 <span class="stringliteral">            [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep num2str(selectedSim)]);</span>
<a name="l02329"></a>02329 <span class="stringliteral">            </span>
<a name="l02330"></a>02330 <span class="stringliteral">            g = sim.gene;</span>
<a name="l02331"></a>02331 <span class="stringliteral">            transcript = sim.state(&#39;</span>Transcript<span class="stringliteral">&#39;);</span>
<a name="l02332"></a>02332 <span class="stringliteral">            rna = sim.state(&#39;</span>Rna<span class="stringliteral">&#39;);</span>
<a name="l02333"></a>02333 <span class="stringliteral">            transcription = sim.process(&#39;</span><a class="code" href="class_transcription.html">Transcription</a><span class="stringliteral">&#39;);</span>
<a name="l02334"></a>02334 <span class="stringliteral">            </span>
<a name="l02335"></a>02335 <span class="stringliteral">            %% get data</span>
<a name="l02336"></a>02336 <span class="stringliteral">            dnaBoundProteinDisplacements = Figures34.calculateDNABoundProteinDisplacements(outDirectory, simBatchDir, selectedSim);</span>
<a name="l02337"></a>02337 <span class="stringliteral">            dnaBoundProteinDisplacements = conv(dnaBoundProteinDisplacements, ones(1000, 1), &#39;</span>same<span class="stringliteral">&#39;);</span>
<a name="l02338"></a>02338 <span class="stringliteral">            </span>
<a name="l02339"></a>02339 <span class="stringliteral">            %% find meaning of peaks</span>
<a name="l02340"></a>02340 <span class="stringliteral">            [vals, peaks] = findpeaks(dnaBoundProteinDisplacements, &#39;</span>MINPEAKDISTANCE<span class="stringliteral">&#39;, 5000);</span>
<a name="l02341"></a>02341 <span class="stringliteral">            [~, order] = sort(vals);</span>
<a name="l02342"></a>02342 <span class="stringliteral">            peaks = peaks(order(end-9:end));</span>
<a name="l02343"></a>02343 <span class="stringliteral">            tuStarts = transcript.transcriptionUnitFivePrimeCoordinates - (transcript.transcriptionUnitLengths - 1) .* (transcript.transcriptionUnitDirections == 0);</span>
<a name="l02344"></a>02344 <span class="stringliteral">            tuEnds = transcript.transcriptionUnitFivePrimeCoordinates + (transcript.transcriptionUnitLengths - 1) .* (transcript.transcriptionUnitDirections == 1);</span>
<a name="l02345"></a>02345 <span class="stringliteral">            peakGenes = NaN(size(peaks));</span>
<a name="l02346"></a>02346 <span class="stringliteral">            peakTUs = NaN(size(peaks));</span>
<a name="l02347"></a>02347 <span class="stringliteral">            for i = 1:numel(peaks)</span>
<a name="l02348"></a>02348 <span class="stringliteral">                tmp = find(g.startCoordinates &lt;= peaks(i) &amp; g.startCoordinates + g.lengths - 1 &gt;= peaks(i));</span>
<a name="l02349"></a>02349 <span class="stringliteral">                if ~isempty(tmp)</span>
<a name="l02350"></a>02350 <span class="stringliteral">                    peakGenes(i) = tmp;</span>
<a name="l02351"></a>02351 <span class="stringliteral">                end</span>
<a name="l02352"></a>02352 <span class="stringliteral">                tmp = find(tuStarts &lt;= peaks(i) &amp; tuEnds &gt;= peaks(i));</span>
<a name="l02353"></a>02353 <span class="stringliteral">                if ~isempty(tmp)</span>
<a name="l02354"></a>02354 <span class="stringliteral">                    peakTUs(i) = tmp;</span>
<a name="l02355"></a>02355 <span class="stringliteral">                end</span>
<a name="l02356"></a>02356 <span class="stringliteral">            end</span>
<a name="l02357"></a>02357 <span class="stringliteral">            [g.wholeCellModelIDs(peakGenes(~isnan(peakGenes))) ...  %most displaced genes</span>
<a name="l02358"></a>02358 <span class="stringliteral">                g.names(peakGenes(~isnan(peakGenes))) ...</span>
<a name="l02359"></a>02359 <span class="stringliteral">                num2cell(dnaBoundProteinDisplacements(peaks(~isnan(peakGenes)))) ...</span>
<a name="l02360"></a>02360 <span class="stringliteral">                num2cell(rna.nascentRNAGeneComposition(peakGenes(~isnan(peakGenes)), :) * transcription.transcriptionUnitBindingProbabilities)</span>
<a name="l02361"></a>02361 <span class="stringliteral">                ] %#ok&lt;NOPRT&gt;</span>
<a name="l02362"></a>02362 <span class="stringliteral">            [transcript.transcriptionUnitWholeCellModelIDs(peakTUs(~isnan(peakTUs))) ... %most displaced transcription units</span>
<a name="l02363"></a>02363 <span class="stringliteral">                num2cell(dnaBoundProteinDisplacements(peaks(~isnan(peakTUs)))) ...</span>
<a name="l02364"></a>02364 <span class="stringliteral">                num2cell(transcription.transcriptionUnitBindingProbabilities(peakTUs(~isnan(peakTUs))))] %#ok&lt;NOPRT&gt;</span>
<a name="l02365"></a>02365 <span class="stringliteral">        end</span>
<a name="l02366"></a>02366 <span class="stringliteral">        </span>
<a name="l02367"></a>02367 <span class="stringliteral">        function totalDensityMatrix = calculateDNABoundProteinDensity(outDirectory, simBatchDir, selectedSim)</span>
<a name="l02368"></a>02368 <span class="stringliteral">            import edu.stanford.covert.cell.sim.analysis.ChromosomePositionHistogram;</span>
<a name="l02369"></a>02369 <span class="stringliteral">            import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;</span>
<a name="l02370"></a>02370 <span class="stringliteral">            import edu.stanford.covert.cell.sim.util.SimulationEnsemble;</span>
<a name="l02371"></a>02371 <span class="stringliteral">            </span>
<a name="l02372"></a>02372 <span class="stringliteral">            if exist([outDirectory &#39;</span>DNABoundProteinDensity.mat<span class="stringliteral">&#39;], &#39;</span>file<span class="stringliteral">&#39;)</span>
<a name="l02373"></a>02373 <span class="stringliteral">                load([outDirectory &#39;</span>DNABoundProteinDensity.mat<span class="stringliteral">&#39;])</span>
<a name="l02374"></a>02374 <span class="stringliteral">            else</span>
<a name="l02375"></a>02375 <span class="stringliteral">                %% get constants</span>
<a name="l02376"></a>02376 <span class="stringliteral">                [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep num2str(selectedSim)]);</span>
<a name="l02377"></a>02377 <span class="stringliteral">            </span>
<a name="l02378"></a>02378 <span class="stringliteral">                stateNames = {</span>
<a name="l02379"></a>02379 <span class="stringliteral">                    &#39;</span><a class="code" href="class_chromosome.html">Chromosome</a><span class="charliteral">&#39; &#39;</span>complexBoundSites<span class="stringliteral">&#39;        &#39;</span>:<span class="charliteral">&#39; &#39;</span>:<span class="stringliteral">&#39;</span>
<a name="l02380"></a>02380 <span class="stringliteral">                    &#39;</span><a class="code" href="class_chromosome.html">Chromosome</a><span class="charliteral">&#39; &#39;</span>monomerBoundSites<span class="stringliteral">&#39;        &#39;</span>:<span class="charliteral">&#39; &#39;</span>:<span class="stringliteral">&#39;</span>
<a name="l02381"></a>02381 <span class="stringliteral">                    };</span>
<a name="l02382"></a>02382 <span class="stringliteral">                states = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, &#39;</span>extract<span class="stringliteral">&#39;, selectedSim);</span>
<a name="l02383"></a>02383 <span class="stringliteral">                </span>
<a name="l02384"></a>02384 <span class="stringliteral">                complexIdxs = full(unique(states.Chromosome.complexBoundSites));</span>
<a name="l02385"></a>02385 <span class="stringliteral">                complexIdxs = complexIdxs(complexIdxs &gt; 0);</span>
<a name="l02386"></a>02386 <span class="stringliteral">                </span>
<a name="l02387"></a>02387 <span class="stringliteral">                monomerIdxs = full(unique(states.Chromosome.monomerBoundSites));</span>
<a name="l02388"></a>02388 <span class="stringliteral">                monomerIdxs = monomerIdxs(monomerIdxs &gt; 0);</span>
<a name="l02389"></a>02389 <span class="stringliteral">                </span>
<a name="l02390"></a>02390 <span class="stringliteral">                boundIndices = struct(&#39;</span>complex<span class="stringliteral">&#39;, {num2cell(complexIdxs&#39;</span>)}, <span class="stringliteral">&#39;monomer&#39;</span>, {num2cell(monomerIdxs<span class="stringliteral">&#39;)});</span>
<a name="l02391"></a>02391 <span class="stringliteral">                </span>
<a name="l02392"></a>02392 <span class="stringliteral">                [cDensityMatrix, mDensityMatrix] = ChromosomePositionHistogram.makeDensityMatrix(states, boundIndices, sim);</span>
<a name="l02393"></a>02393 <span class="stringliteral">                totalDensityMatrix = sum(cDensityMatrix, 2) + sum(mDensityMatrix, 2);</span>
<a name="l02394"></a>02394 <span class="stringliteral">                </span>
<a name="l02395"></a>02395 <span class="stringliteral">                save([outDirectory &#39;</span>DNABoundProteinDensity.mat<span class="stringliteral">&#39;], &#39;</span>totalDensityMatrix<span class="stringliteral">&#39;)</span>
<a name="l02396"></a>02396 <span class="stringliteral">            end</span>
<a name="l02397"></a>02397 <span class="stringliteral">        end</span>
<a name="l02398"></a>02398 <span class="stringliteral">        </span>
<a name="l02399"></a>02399 <span class="stringliteral">        function displacements = calculateDNABoundProteinDisplacements(outDirectory, simBatchDir, selectedSim, iJob, nJobs)</span>
<a name="l02400"></a>02400 <span class="stringliteral">            import edu.stanford.covert.cell.sim.util.SimulationEnsemble;</span>
<a name="l02401"></a>02401 <span class="stringliteral">            import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;</span>
<a name="l02402"></a>02402 <span class="stringliteral">            </span>
<a name="l02403"></a>02403 <span class="stringliteral">            if nargin &lt; 5</span>
<a name="l02404"></a>02404 <span class="stringliteral">                nJobs = 1;</span>
<a name="l02405"></a>02405 <span class="stringliteral">            end</span>
<a name="l02406"></a>02406 <span class="stringliteral">            </span>
<a name="l02407"></a>02407 <span class="stringliteral">            relSimBatchDir = simBatchDir;</span>
<a name="l02408"></a>02408 <span class="stringliteral">            % Make simBatchDir a relative path</span>
<a name="l02409"></a>02409 <span class="stringliteral">            while relSimBatchDir(end) == filesep</span>
<a name="l02410"></a>02410 <span class="stringliteral">                relSimBatchDir(end) = [];</span>
<a name="l02411"></a>02411 <span class="stringliteral">            end</span>
<a name="l02412"></a>02412 <span class="stringliteral">            relSimBatchDir = [relSimBatchDir filesep];</span>
<a name="l02413"></a>02413 <span class="stringliteral">            p = find(relSimBatchDir == filesep, 2, &#39;</span>last<span class="stringliteral">&#39;);</span>
<a name="l02414"></a>02414 <span class="stringliteral">            relSimBatchDir = relSimBatchDir(p(1) + 1:p(2) - 1);</span>
<a name="l02415"></a>02415 <span class="stringliteral"></span>
<a name="l02416"></a>02416 <span class="stringliteral">            outDirectory = [outDirectory filesep];</span>
<a name="l02417"></a>02417 <span class="stringliteral">            </span>
<a name="l02418"></a>02418 <span class="stringliteral">            %% name of file where results will be cached</span>
<a name="l02419"></a>02419 <span class="stringliteral">            if nJobs == 1 || nargin &lt; 4 || isempty(iJob)</span>
<a name="l02420"></a>02420 <span class="stringliteral">                fileName = sprintf(&#39;</span>%sdnaBoundProteinDisplacement-%s-%d.mat<span class="stringliteral">&#39;, outDirectory, relSimBatchDir, selectedSim);</span>
<a name="l02421"></a>02421 <span class="stringliteral">            else</span>
<a name="l02422"></a>02422 <span class="stringliteral">                fileName = sprintf(&#39;</span>%sdnaBoundProteinDisplacement-%s-%d-%d-%d.mat<span class="stringliteral">&#39;, outDirectory, relSimBatchDir, selectedSim, iJob, nJobs);</span>
<a name="l02423"></a>02423 <span class="stringliteral">            end</span>
<a name="l02424"></a>02424 <span class="stringliteral">            </span>
<a name="l02425"></a>02425 <span class="stringliteral">            %% get parameters</span>
<a name="l02426"></a>02426 <span class="stringliteral">            [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep num2str(selectedSim)]);</span>
<a name="l02427"></a>02427 <span class="stringliteral">            c = sim.state(&#39;</span><a class="code" href="class_chromosome.html">Chromosome</a><span class="stringliteral">&#39;);</span>
<a name="l02428"></a>02428 <span class="stringliteral">            pm = sim.state(&#39;</span><a class="code" href="class_protein_monomer.html">ProteinMonomer</a><span class="stringliteral">&#39;);</span>
<a name="l02429"></a>02429 <span class="stringliteral">            pc = sim.state(&#39;</span><a class="code" href="class_protein_complex.html">ProteinComplex</a><span class="stringliteral">&#39;);</span>
<a name="l02430"></a>02430 <span class="stringliteral">            </span>
<a name="l02431"></a>02431 <span class="stringliteral">            %% if data already computed, just load data</span>
<a name="l02432"></a>02432 <span class="stringliteral">            if exist(fileName, &#39;</span>file<span class="stringliteral">&#39;)</span>
<a name="l02433"></a>02433 <span class="stringliteral">                load(fileName);</span>
<a name="l02434"></a>02434 <span class="stringliteral">                return;</span>
<a name="l02435"></a>02435 <span class="stringliteral">            end</span>
<a name="l02436"></a>02436 <span class="stringliteral">            </span>
<a name="l02437"></a>02437 <span class="stringliteral">            %% stitch together jobs</span>
<a name="l02438"></a>02438 <span class="stringliteral">            if nargin &gt;= 5 &amp;&amp; isempty(iJob)</span>
<a name="l02439"></a>02439 <span class="stringliteral">                displacements = zeros(c.sequenceLen, 1);</span>
<a name="l02440"></a>02440 <span class="stringliteral">                for iJob = 1:nJobs</span>
<a name="l02441"></a>02441 <span class="stringliteral">                    tmp = load(sprintf(&#39;</span>%sdnaBoundProteinDisplacement-%s-%d-%d-%d.mat<span class="stringliteral">&#39;, outDirectory, simBatchDir, selectedSim, iJob, nJobs));</span>
<a name="l02442"></a>02442 <span class="stringliteral">                    displacements = displacements + tmp.displacements;</span>
<a name="l02443"></a>02443 <span class="stringliteral">                end</span>
<a name="l02444"></a>02444 <span class="stringliteral">                save(fileName, &#39;</span>displacements<span class="stringliteral">&#39;);</span>
<a name="l02445"></a>02445 <span class="stringliteral">                return;</span>
<a name="l02446"></a>02446 <span class="stringliteral">            end</span>
<a name="l02447"></a>02447 <span class="stringliteral">            </span>
<a name="l02448"></a>02448 <span class="stringliteral">            %% get data</span>
<a name="l02449"></a>02449 <span class="stringliteral">            stateNames = {</span>
<a name="l02450"></a>02450 <span class="stringliteral">                &#39;</span><a class="code" href="class_chromosome.html">Chromosome</a><span class="charliteral">&#39; &#39;</span>monomerBoundSites<span class="stringliteral">&#39;</span>
<a name="l02451"></a>02451 <span class="stringliteral">                &#39;</span><a class="code" href="class_chromosome.html">Chromosome</a><span class="charliteral">&#39; &#39;</span>complexBoundSites<span class="stringliteral">&#39;</span>
<a name="l02452"></a>02452 <span class="stringliteral">                };</span>
<a name="l02453"></a>02453 <span class="stringliteral">            displacements = zeros(c.sequenceLen, 1);</span>
<a name="l02454"></a>02454 <span class="stringliteral">            </span>
<a name="l02455"></a>02455 <span class="stringliteral">            states = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, &#39;</span>extract<span class="stringliteral">&#39;, selectedSim);</span>
<a name="l02456"></a>02456 <span class="stringliteral">            </span>
<a name="l02457"></a>02457 <span class="stringliteral">            monBndSites = permute(any(states.Chromosome.monomerBoundSites(:, 1:2, :), 2), [1 3 2]);</span>
<a name="l02458"></a>02458 <span class="stringliteral">            cpxBndSites = permute(any(states.Chromosome.complexBoundSites(:, 1:2, :), 2), [1 3 2]);</span>
<a name="l02459"></a>02459 <span class="stringliteral">            </span>
<a name="l02460"></a>02460 <span class="stringliteral">            monMonDisplacement = zeros(numel(pm.matureIndexs), numel(pm.matureIndexs));</span>
<a name="l02461"></a>02461 <span class="stringliteral">            monCpxDisplacement = zeros(numel(pm.matureIndexs), numel(pc.matureIndexs));</span>
<a name="l02462"></a>02462 <span class="stringliteral">            cpxMonDisplacement = zeros(numel(pc.matureIndexs), numel(pm.matureIndexs));</span>
<a name="l02463"></a>02463 <span class="stringliteral">            cpxCpxDisplacement = zeros(numel(pc.matureIndexs), numel(pc.matureIndexs));</span>
<a name="l02464"></a>02464 <span class="stringliteral">            </span>
<a name="l02465"></a>02465 <span class="stringliteral">            [i, j] = find(c.reactionMonomerCatalysisMatrix);</span>
<a name="l02466"></a>02466 <span class="stringliteral">            tfs = c.reactionBoundMonomer(i) ~= 0;</span>
<a name="l02467"></a>02467 <span class="stringliteral">            monMonDisplacement(c.reactionBoundMonomer(i(tfs)), j(tfs)) = 1;</span>
<a name="l02468"></a>02468 <span class="stringliteral">            cpxMonDisplacement(c.reactionBoundComplex(i(~tfs)), j(~tfs)) = 1;</span>
<a name="l02469"></a>02469 <span class="stringliteral">            </span>
<a name="l02470"></a>02470 <span class="stringliteral">            [i, j] = find(c.reactionComplexCatalysisMatrix);</span>
<a name="l02471"></a>02471 <span class="stringliteral">            tfs = c.reactionBoundMonomer(i) ~= 0;</span>
<a name="l02472"></a>02472 <span class="stringliteral">            monCpxDisplacement(c.reactionBoundMonomer(i(tfs)), j(tfs)) = 1;</span>
<a name="l02473"></a>02473 <span class="stringliteral">            cpxCpxDisplacement(c.reactionBoundComplex(i(~tfs)), j(~tfs)) = 1;</span>
<a name="l02474"></a>02474 <span class="stringliteral">            </span>
<a name="l02475"></a>02475 <span class="stringliteral">            for i = iJob:nJobs:size(states.Chromosome.monomerBoundSites, 3)-1</span>
<a name="l02476"></a>02476 <span class="stringliteral">                monBndSites1 = monBndSites(:, i);</span>
<a name="l02477"></a>02477 <span class="stringliteral">                monBndSites2 = monBndSites(:, i+1);</span>
<a name="l02478"></a>02478 <span class="stringliteral">                [subs1, vals1] = find(monBndSites1);</span>
<a name="l02479"></a>02479 <span class="stringliteral">                [subs2, vals2] = find(monBndSites2);</span>
<a name="l02480"></a>02480 <span class="stringliteral">                monBndSites1(subs2(monBndSites1(subs2) == vals2, :)) = 0;</span>
<a name="l02481"></a>02481 <span class="stringliteral">                monBndSites2(subs1(monBndSites2(subs1) == vals1, :)) = 0;</span>
<a name="l02482"></a>02482 <span class="stringliteral">                [monPos1, monIdxs1] = find(monBndSites1);</span>
<a name="l02483"></a>02483 <span class="stringliteral">                [monPos2, monIdxs2] = find(monBndSites2);</span>
<a name="l02484"></a>02484 <span class="stringliteral">                monStarts1 = monPos1(:, 1);</span>
<a name="l02485"></a>02485 <span class="stringliteral">                monStarts2 = monPos2(:, 1);</span>
<a name="l02486"></a>02486 <span class="stringliteral">                monEnds1 = monPos1(:, 1) + c.monomerDNAFootprints(monIdxs1);</span>
<a name="l02487"></a>02487 <span class="stringliteral">                monEnds2 = monPos2(:, 1) + c.monomerDNAFootprints(monIdxs2);</span>
<a name="l02488"></a>02488 <span class="stringliteral">                </span>
<a name="l02489"></a>02489 <span class="stringliteral">                cpxBndSites1 = cpxBndSites(:, i);</span>
<a name="l02490"></a>02490 <span class="stringliteral">                cpxBndSites2 = cpxBndSites(:, i+1);</span>
<a name="l02491"></a>02491 <span class="stringliteral">                [subs1, vals1] = find(cpxBndSites1);</span>
<a name="l02492"></a>02492 <span class="stringliteral">                [subs2, vals2] = find(cpxBndSites2);</span>
<a name="l02493"></a>02493 <span class="stringliteral">                cpxBndSites1(subs2(cpxBndSites1(subs2) == vals2, :)) = 0;</span>
<a name="l02494"></a>02494 <span class="stringliteral">                cpxBndSites2(subs1(cpxBndSites2(subs1) == vals1, :)) = 0;</span>
<a name="l02495"></a>02495 <span class="stringliteral">                [cpxPos1, cpxIdxs1] = find(cpxBndSites1);</span>
<a name="l02496"></a>02496 <span class="stringliteral">                [cpxPos2, cpxIdxs2] = find(cpxBndSites2);</span>
<a name="l02497"></a>02497 <span class="stringliteral">                cpxStarts1 = cpxPos1(:, 1);</span>
<a name="l02498"></a>02498 <span class="stringliteral">                cpxStarts2 = cpxPos2(:, 1);</span>
<a name="l02499"></a>02499 <span class="stringliteral">                cpxEnds1 = cpxPos1(:, 1) + c.complexDNAFootprints(cpxIdxs1);</span>
<a name="l02500"></a>02500 <span class="stringliteral">                cpxEnds2 = cpxPos2(:, 1) + c.complexDNAFootprints(cpxIdxs2);</span>
<a name="l02501"></a>02501 <span class="stringliteral">                </span>
<a name="l02502"></a>02502 <span class="stringliteral">                if any(any(monMonDisplacement(monIdxs1, monIdxs2)))</span>
<a name="l02503"></a>02503 <span class="stringliteral">                    for j = 1:size(monPos1, 1)</span>
<a name="l02504"></a>02504 <span class="stringliteral">                        tfs = ...</span>
<a name="l02505"></a>02505 <span class="stringliteral">                            (monStarts1(j) &lt;= monStarts2 &amp;  monStarts1(j) &gt;= monEnds2) | ...</span>
<a name="l02506"></a>02506 <span class="stringliteral">                            (monEnds1(j) &lt;= monStarts2  &amp;  monEnds1(j) &gt;= monEnds2) | ...</span>
<a name="l02507"></a>02507 <span class="stringliteral">                            (monStarts1(j) &lt;= monStarts2  &amp;  monEnds1(j) &gt;= monEnds2);</span>
<a name="l02508"></a>02508 <span class="stringliteral">                        if any(monMonDisplacement(monIdxs1(j), monIdxs2(tfs)))</span>
<a name="l02509"></a>02509 <span class="stringliteral">                            displacements(monStarts1(j)) = displacements(monStarts1(j)) + 1;</span>
<a name="l02510"></a>02510 <span class="stringliteral">                        end</span>
<a name="l02511"></a>02511 <span class="stringliteral">                    end</span>
<a name="l02512"></a>02512 <span class="stringliteral">                end</span>
<a name="l02513"></a>02513 <span class="stringliteral">                    </span>
<a name="l02514"></a>02514 <span class="stringliteral">                if any(any(monCpxDisplacement(monIdxs1, cpxIdxs2)))</span>
<a name="l02515"></a>02515 <span class="stringliteral">                    for j = 1:size(monPos1, 1)</span>
<a name="l02516"></a>02516 <span class="stringliteral">                        tfs = ...</span>
<a name="l02517"></a>02517 <span class="stringliteral">                            (monStarts1(j) &lt;= cpxStarts2 &amp;  monStarts1(j) &gt;= cpxEnds2) | ...</span>
<a name="l02518"></a>02518 <span class="stringliteral">                            (monEnds1(j) &lt;= cpxStarts2  &amp;  monEnds1(j) &gt;= cpxEnds2) | ...</span>
<a name="l02519"></a>02519 <span class="stringliteral">                            (monStarts1(j) &lt;= cpxStarts2  &amp;  monEnds1(j) &gt;= cpxEnds2);</span>
<a name="l02520"></a>02520 <span class="stringliteral">                        if any(monCpxDisplacement(monIdxs1(j), cpxIdxs2(tfs)))</span>
<a name="l02521"></a>02521 <span class="stringliteral">                            displacements(monStarts1(j)) = displacements(monStarts1(j)) + 1;</span>
<a name="l02522"></a>02522 <span class="stringliteral">                        end</span>
<a name="l02523"></a>02523 <span class="stringliteral">                    end</span>
<a name="l02524"></a>02524 <span class="stringliteral">                end</span>
<a name="l02525"></a>02525 <span class="stringliteral">                </span>
<a name="l02526"></a>02526 <span class="stringliteral">                if any(any(cpxMonDisplacement(cpxIdxs1, monIdxs2)))</span>
<a name="l02527"></a>02527 <span class="stringliteral">                    for j = 1:size(monPos2, 1)</span>
<a name="l02528"></a>02528 <span class="stringliteral">                        tfs = ...</span>
<a name="l02529"></a>02529 <span class="stringliteral">                            (cpxStarts1 &lt;= monStarts2(j) &amp;  cpxStarts1 &gt;= monEnds2(j)) | ...</span>
<a name="l02530"></a>02530 <span class="stringliteral">                            (cpxEnds1 &lt;= monStarts2(j)  &amp;  cpxEnds1 &gt;= monEnds2(j)) | ...</span>
<a name="l02531"></a>02531 <span class="stringliteral">                            (cpxStarts1 &lt;= monStarts2(j)  &amp;  cpxEnds1 &gt;= monEnds2(j));</span>
<a name="l02532"></a>02532 <span class="stringliteral">                        if any(cpxMonDisplacement(cpxIdxs1(tfs), monIdxs2(j)))</span>
<a name="l02533"></a>02533 <span class="stringliteral">                            displacements(cpxStarts1(tfs)) = displacements(cpxStarts1(tfs)) + 1;</span>
<a name="l02534"></a>02534 <span class="stringliteral">                        end</span>
<a name="l02535"></a>02535 <span class="stringliteral">                    end</span>
<a name="l02536"></a>02536 <span class="stringliteral">                end</span>
<a name="l02537"></a>02537 <span class="stringliteral">                </span>
<a name="l02538"></a>02538 <span class="stringliteral">                if any(any(cpxCpxDisplacement(cpxIdxs1, cpxIdxs2)))</span>
<a name="l02539"></a>02539 <span class="stringliteral">                    for j = 1:size(cpxPos1, 1)</span>
<a name="l02540"></a>02540 <span class="stringliteral">                        tfs = ...</span>
<a name="l02541"></a>02541 <span class="stringliteral">                            (cpxStarts1(j) &lt;= cpxStarts2 &amp;  cpxStarts1(j) &gt;= cpxEnds2) | ...</span>
<a name="l02542"></a>02542 <span class="stringliteral">                            (cpxEnds1(j) &lt;= cpxStarts2  &amp;  cpxEnds1(j) &gt;= cpxEnds2) | ...</span>
<a name="l02543"></a>02543 <span class="stringliteral">                            (cpxStarts1(j) &lt;= cpxStarts2  &amp;  cpxEnds1(j) &gt;= cpxEnds2);</span>
<a name="l02544"></a>02544 <span class="stringliteral">                        if any(cpxCpxDisplacement(cpxIdxs1(j), cpxIdxs2(tfs)))</span>
<a name="l02545"></a>02545 <span class="stringliteral">                            displacements(cpxStarts1(j)) = displacements(cpxStarts1(j)) + 1;</span>
<a name="l02546"></a>02546 <span class="stringliteral">                        end</span>
<a name="l02547"></a>02547 <span class="stringliteral">                    end</span>
<a name="l02548"></a>02548 <span class="stringliteral">                end</span>
<a name="l02549"></a>02549 <span class="stringliteral">            end            </span>
<a name="l02550"></a>02550 <span class="stringliteral">            </span>
<a name="l02551"></a>02551 <span class="stringliteral">            %% save</span>
<a name="l02552"></a>02552 <span class="stringliteral">            save(fileName, &#39;</span>displacements<span class="stringliteral">&#39;);</span>
<a name="l02553"></a>02553 <span class="stringliteral">        end</span>
<a name="l02554"></a>02554 <span class="stringliteral">        </span>
<a name="l02555"></a>02555 <span class="stringliteral">        function metabolicMap(simBatchDir, selectedSim, outDirectory, figHandle, position, showAllLabels)</span>
<a name="l02556"></a>02556 <span class="stringliteral">            import edu.stanford.covert.cell.kb.KnowledgeBaseUtil;            </span>
<a name="l02557"></a>02557 <span class="stringliteral">            import edu.stanford.covert.cell.sim.util.SimulationEnsemble;</span>
<a name="l02558"></a>02558 <span class="stringliteral">            import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;</span>
<a name="l02559"></a>02559 <span class="stringliteral">            import edu.stanford.covert.cell.sim.util.PlotUtil;</span>
<a name="l02560"></a>02560 <span class="stringliteral">            import edu.stanford.covert.cell.sim.util.PrintUtil;</span>
<a name="l02561"></a>02561 <span class="stringliteral">            import edu.stanford.covert.cell.sim.analysis.Figures34;</span>
<a name="l02562"></a>02562 <span class="stringliteral">            import edu.stanford.covert.db.MySQLDatabase;            </span>
<a name="l02563"></a>02563 <span class="stringliteral">            </span>
<a name="l02564"></a>02564 <span class="stringliteral">            if ~exist(&#39;</span>simBatchDir<span class="stringliteral">&#39;, &#39;</span>var<span class="stringliteral">&#39;) || isempty(simBatchDir)</span>
<a name="l02565"></a>02565 <span class="stringliteral">                simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();</span>
<a name="l02566"></a>02566 <span class="stringliteral">            end</span>
<a name="l02567"></a>02567 <span class="stringliteral">            if ~exist(&#39;</span>selectedSim<span class="stringliteral">&#39;, &#39;</span>var<span class="stringliteral">&#39;) || isempty(selectedSim)</span>
<a name="l02568"></a>02568 <span class="stringliteral">                selectedSim = 1;</span>
<a name="l02569"></a>02569 <span class="stringliteral">            end</span>
<a name="l02570"></a>02570 <span class="stringliteral">            if nargin &lt; 6</span>
<a name="l02571"></a>02571 <span class="stringliteral">                showAllLabels = false;</span>
<a name="l02572"></a>02572 <span class="stringliteral">            end</span>
<a name="l02573"></a>02573 <span class="stringliteral">            </span>
<a name="l02574"></a>02574 <span class="stringliteral">            %% get constants</span>
<a name="l02575"></a>02575 <span class="stringliteral">            [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep num2str(selectedSim)]);</span>
<a name="l02576"></a>02576 <span class="stringliteral">            comp = sim.compartment;</span>
<a name="l02577"></a>02577 <span class="stringliteral">            met = sim.state(&#39;</span><a class="code" href="class_metabolite.html">Metabolite</a><span class="stringliteral">&#39;);</span>
<a name="l02578"></a>02578 <span class="stringliteral">            mr = sim.state(&#39;</span><a class="code" href="class_metabolic_reaction.html">MetabolicReaction</a><span class="stringliteral">&#39;);</span>
<a name="l02579"></a>02579 <span class="stringliteral">            metabolism = sim.process(&#39;</span>Metabolism<span class="stringliteral">&#39;);</span>
<a name="l02580"></a>02580 <span class="stringliteral">            </span>
<a name="l02581"></a>02581 <span class="stringliteral">            %% get data</span>
<a name="l02582"></a>02582 <span class="stringliteral">            if exist([outDirectory &#39;</span>metabolicMap.mat<span class="stringliteral">&#39;], &#39;</span>file<span class="stringliteral">&#39;)</span>
<a name="l02583"></a>02583 <span class="stringliteral">                load([outDirectory &#39;</span>metabolicMap.mat<span class="stringliteral">&#39;])</span>
<a name="l02584"></a>02584 <span class="stringliteral">            else                </span>
<a name="l02585"></a>02585 <span class="stringliteral">                dbConnectionParameters = config();</span>
<a name="l02586"></a>02586 <span class="stringliteral">                database = MySQLDatabase(dbConnectionParameters);</span>
<a name="l02587"></a>02587 <span class="stringliteral">                database.setNullValue(0);</span>
<a name="l02588"></a>02588 <span class="stringliteral">                </span>
<a name="l02589"></a>02589 <span class="stringliteral">                kbWID = KnowledgeBaseUtil.selectLatestKnowledgeBase(database);</span>
<a name="l02590"></a>02590 <span class="stringliteral">                </span>
<a name="l02591"></a>02591 <span class="stringliteral">                database.prepareStatement(&#39;</span>CALL get_metabolicmapmetabolites(<span class="stringliteral">&quot;{Si}&quot;</span>, null)<span class="stringliteral">&#39;, kbWID);</span>
<a name="l02592"></a>02592 <span class="stringliteral">                data = database.query();</span>
<a name="l02593"></a>02593 <span class="stringliteral">                mapMets = struct;</span>
<a name="l02594"></a>02594 <span class="stringliteral">                [~, mapMets.idxs] = ismember(data.Metabolite, met.wholeCellModelIDs);</span>
<a name="l02595"></a>02595 <span class="stringliteral">                [~, mapMets.compIdxs] = ismember(data.Compartment, comp.wholeCellModelIDs);</span>
<a name="l02596"></a>02596 <span class="stringliteral">                [~, mapMets.uidxs] = ismember(mapMets.idxs, unique(mapMets.idxs));</span>
<a name="l02597"></a>02597 <span class="stringliteral">                mapMets.x = data.X;</span>
<a name="l02598"></a>02598 <span class="stringliteral">                mapMets.y = cellfun(@str2double, data.Y);</span>
<a name="l02599"></a>02599 <span class="stringliteral">                </span>
<a name="l02600"></a>02600 <span class="stringliteral">                database.prepareStatement(&#39;</span>CALL get_metabolicmapreactions(<span class="stringliteral">&quot;{Si}&quot;</span>, null)<span class="stringliteral">&#39;, kbWID);</span>
<a name="l02601"></a>02601 <span class="stringliteral">                data = database.query();</span>
<a name="l02602"></a>02602 <span class="stringliteral">                mapRxns = struct;</span>
<a name="l02603"></a>02603 <span class="stringliteral">                [~, mapRxns.idxs] = ismember(data.Reaction, mr.reactionWholeCellModelIDs);</span>
<a name="l02604"></a>02604 <span class="stringliteral">                [~, mapRxns.uidxs] = ismember(mapRxns.idxs, unique(mapRxns.idxs));</span>
<a name="l02605"></a>02605 <span class="stringliteral">                mapRxns.path = data.Path;</span>
<a name="l02606"></a>02606 <span class="stringliteral">                mapRxns.labelX = data.LabelX;</span>
<a name="l02607"></a>02607 <span class="stringliteral">                mapRxns.labelY = data.LabelY;</span>
<a name="l02608"></a>02608 <span class="stringliteral">                mapRxns.valueX = data.ValueX;</span>
<a name="l02609"></a>02609 <span class="stringliteral">                mapRxns.valueY = data.ValueY;</span>
<a name="l02610"></a>02610 <span class="stringliteral">                </span>
<a name="l02611"></a>02611 <span class="stringliteral">                %% get data</span>
<a name="l02612"></a>02612 <span class="stringliteral">                stateNames = {</span>
<a name="l02613"></a>02613 <span class="stringliteral">                    &#39;</span><a class="code" href="class_metabolic_reaction.html">MetabolicReaction</a><span class="stringliteral">&#39;  &#39;</span>fluxs<span class="stringliteral">&#39;  &#39;</span>:<span class="stringliteral">&#39;  &#39;</span>:<span class="stringliteral">&#39;</span>
<a name="l02614"></a>02614 <span class="stringliteral">                    };</span>
<a name="l02615"></a>02615 <span class="stringliteral">                states = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, &#39;</span>extract<span class="stringliteral">&#39;, selectedSim);</span>
<a name="l02616"></a>02616 <span class="stringliteral">                </span>
<a name="l02617"></a>02617 <span class="stringliteral">                states.MetabolicReaction.fluxs = full(states.MetabolicReaction.fluxs);</span>
<a name="l02618"></a>02618 <span class="stringliteral">                </span>
<a name="l02619"></a>02619 <span class="stringliteral">                %% table</span>
<a name="l02620"></a>02620 <span class="stringliteral">                colLabels = {&#39;</span><a class="code" href="class_reaction.html">Reaction</a> ID<span class="stringliteral">&#39;, &#39;</span><a class="code" href="class_reaction.html">Reaction</a> Name<span class="stringliteral">&#39;, &#39;</span>Flux Mean (rxn/s)<span class="stringliteral">&#39;, &#39;</span>Flux Std (rxn/s)<span class="stringliteral">&#39;};</span>
<a name="l02621"></a>02621 <span class="stringliteral">                content = [</span>
<a name="l02622"></a>02622 <span class="stringliteral">                    metabolism.reactionWholeCellModelIDs  ...</span>
<a name="l02623"></a>02623 <span class="stringliteral">                    metabolism.reactionNames  ...</span>
<a name="l02624"></a>02624 <span class="stringliteral">                    num2cell(nanmean(states.MetabolicReaction.fluxs, 3)) ...</span>
<a name="l02625"></a>02625 <span class="stringliteral">                    num2cell(nanstd(states.MetabolicReaction.fluxs, [], 3))</span>
<a name="l02626"></a>02626 <span class="stringliteral">                    ];</span>
<a name="l02627"></a>02627 <span class="stringliteral">                PrintUtil.printToFile(content, colLabels, [outDirectory &#39;</span>reactionFluxes.xls<span class="stringliteral">&#39;], &#39;</span><a class="code" href="class_reaction.html">Reaction</a> Fluxes<span class="stringliteral">&#39;);</span>
<a name="l02628"></a>02628 <span class="stringliteral">                </span>
<a name="l02629"></a>02629 <span class="stringliteral">                %% iso fluxs</span>
<a name="l02630"></a>02630 <span class="stringliteral">                isoFluxIDs = {</span>
<a name="l02631"></a>02631 <span class="stringliteral">                    {&#39;</span>Adk1<span class="stringliteral">&#39;, &#39;</span>Adk2<span class="stringliteral">&#39;, &#39;</span>Adk3<span class="stringliteral">&#39;, &#39;</span>Adk4<span class="stringliteral">&#39;}</span>
<a name="l02632"></a>02632 <span class="stringliteral">                    {&#39;</span>PfkA1<span class="stringliteral">&#39;, &#39;</span>PfkA2<span class="stringliteral">&#39;, &#39;</span>PfkA3<span class="stringliteral">&#39;, &#39;</span>PfkA4<span class="stringliteral">&#39;, &#39;</span>PfkA5<span class="stringliteral">&#39;}</span>
<a name="l02633"></a>02633 <span class="stringliteral">                    {&#39;</span>Pgk1<span class="stringliteral">&#39;, &#39;</span>Pgk2<span class="stringliteral">&#39;, &#39;</span>Pgk3<span class="stringliteral">&#39;, &#39;</span>Pgk4<span class="stringliteral">&#39;, &#39;</span>Pgk5<span class="stringliteral">&#39;}</span>
<a name="l02634"></a>02634 <span class="stringliteral">                    {&#39;</span>Pyk_ADP<span class="stringliteral">&#39;, &#39;</span>Pyk_CDP<span class="stringliteral">&#39;, &#39;</span>Pyk_DADP<span class="stringliteral">&#39;, &#39;</span>Pyk_DCDP<span class="stringliteral">&#39;, &#39;</span>Pyk_DGDP<span class="stringliteral">&#39;, &#39;</span>Pyk_DTDP<span class="stringliteral">&#39;, &#39;</span>Pyk_DUDP<span class="stringliteral">&#39;, &#39;</span>Pyk_GDP<span class="stringliteral">&#39;, &#39;</span>Pyk_IDP<span class="stringliteral">&#39;, &#39;</span>Pyk_TDP<span class="stringliteral">&#39;, &#39;</span>Pyk_UDP<span class="stringliteral">&#39;}</span>
<a name="l02635"></a>02635 <span class="stringliteral">                    };</span>
<a name="l02636"></a>02636 <span class="stringliteral">                isoFluxs = false(size(mr.reactionWholeCellModelIDs));</span>
<a name="l02637"></a>02637 <span class="stringliteral">                for i = 1:numel(isoFluxIDs)</span>
<a name="l02638"></a>02638 <span class="stringliteral">                    tfs2 = ismember(mr.reactionWholeCellModelIDs, isoFluxIDs{i}(2:end));</span>
<a name="l02639"></a>02639 <span class="stringliteral">                    isoFluxs = isoFluxs | tfs2;</span>
<a name="l02640"></a>02640 <span class="stringliteral">                    </span>
<a name="l02641"></a>02641 <span class="stringliteral">                    [~, idxs1] = ismember(isoFluxIDs{i}(1), mr.reactionWholeCellModelIDs);</span>
<a name="l02642"></a>02642 <span class="stringliteral">                    [~, idxs2] = ismember(isoFluxIDs{i}(2:end), mr.reactionWholeCellModelIDs);</span>
<a name="l02643"></a>02643 <span class="stringliteral">                    states.MetabolicReaction.fluxs(idxs1, :, :) = ...</span>
<a name="l02644"></a>02644 <span class="stringliteral">                        + states.MetabolicReaction.fluxs(idxs1, :, :) ...</span>
<a name="l02645"></a>02645 <span class="stringliteral">                        + sum(states.MetabolicReaction.fluxs(idxs2, :, :), 1);                    </span>
<a name="l02646"></a>02646 <span class="stringliteral">                end</span>
<a name="l02647"></a>02647 <span class="stringliteral">                </span>
<a name="l02648"></a>02648 <span class="stringliteral">                fluxMeans = nanmean(states.MetabolicReaction.fluxs(mapRxns.idxs, :, :), 3);</span>
<a name="l02649"></a>02649 <span class="stringliteral">                fluxStds = nanstd(states.MetabolicReaction.fluxs(mapRxns.idxs, :, :), [], 3) ./ abs(fluxMeans); % (Standard Deviation)/|Mean|</span>
<a name="l02650"></a>02650 <span class="stringliteral">                </span>
<a name="l02651"></a>02651 <span class="stringliteral">                %% save</span>
<a name="l02652"></a>02652 <span class="stringliteral">                save([outDirectory &#39;</span>metabolicMap.mat<span class="stringliteral">&#39;], &#39;</span>mapMets<span class="stringliteral">&#39;, &#39;</span>mapRxns<span class="stringliteral">&#39;, &#39;</span>fluxMeans<span class="stringliteral">&#39;, &#39;</span>fluxStds<span class="stringliteral">&#39;, &#39;</span>isoFluxs<span class="stringliteral">&#39;);</span>
<a name="l02653"></a>02653 <span class="stringliteral">            end</span>
<a name="l02654"></a>02654 <span class="stringliteral">            </span>
<a name="l02655"></a>02655 <span class="stringliteral">            %% plot map</span>
<a name="l02656"></a>02656 <span class="stringliteral">            H = 600;</span>
<a name="l02657"></a>02657 <span class="stringliteral">            W = 800;</span>
<a name="l02658"></a>02658 <span class="stringliteral">            metRadius = 3;</span>
<a name="l02659"></a>02659 <span class="stringliteral">            arrowRadius = 14;</span>
<a name="l02660"></a>02660 <span class="stringliteral">            barW = 20;</span>
<a name="l02661"></a>02661 <span class="stringliteral">            h = 30;</span>
<a name="l02662"></a>02662 <span class="stringliteral">            y = H - 50;</span>
<a name="l02663"></a>02663 <span class="stringliteral">            x = 20;</span>
<a name="l02664"></a>02664 <span class="stringliteral">            reactionLims = [0 2 4];</span>
<a name="l02665"></a>02665 <span class="stringliteral">            strokes = [0.25 1.125 1];</span>
<a name="l02666"></a>02666 <span class="stringliteral">            </span>
<a name="l02667"></a>02667 <span class="stringliteral">            %align</span>
<a name="l02668"></a>02668 <span class="stringliteral">            mapRxns.path{mapRxns.idxs == metabolism.reactionIndexs(&#39;</span>PfkA1<span class="stringliteral">&#39;)} = &#39;</span>M 230,145 L 230,178<span class="stringliteral">&#39;;</span>
<a name="l02669"></a>02669 <span class="stringliteral">            mapRxns.path{mapRxns.idxs == metabolism.reactionIndexs(&#39;</span>Pgk1<span class="stringliteral">&#39;)} = &#39;</span>M 270,265 L 270,293<span class="stringliteral">&#39;;</span>
<a name="l02670"></a>02670 <span class="stringliteral">            mapRxns.path{mapRxns.idxs == metabolism.reactionIndexs(&#39;</span>GpmA<span class="stringliteral">&#39;)} = &#39;</span>M 270,305 L 270,331<span class="stringliteral">&#39;;</span>
<a name="l02671"></a>02671 <span class="stringliteral">            mapRxns.path{mapRxns.idxs == metabolism.reactionIndexs(&#39;</span>Pyk_ADP<span class="stringliteral">&#39;)} = &#39;</span>M 270,385 L 270,418<span class="stringliteral">&#39;;</span>
<a name="l02672"></a>02672 <span class="stringliteral">            mapRxns.path{mapRxns.idxs == metabolism.reactionIndexs(&#39;</span>Adk1<span class="stringliteral">&#39;)} = &#39;</span>M 761,80 L 815,80<span class="stringliteral">&#39;;</span>
<a name="l02673"></a>02673 <span class="stringliteral">            </span>
<a name="l02674"></a>02674 <span class="stringliteral">            %scale</span>
<a name="l02675"></a>02675 <span class="stringliteral">            minX = min(mapMets.x)-metRadius;</span>
<a name="l02676"></a>02676 <span class="stringliteral">            minY = min(mapMets.y)+metRadius;</span>
<a name="l02677"></a>02677 <span class="stringliteral">            scaleX = W / (range(mapMets.x) + 2*metRadius);</span>
<a name="l02678"></a>02678 <span class="stringliteral">            scaleY = H / (range(mapMets.y) + 2*metRadius);</span>
<a name="l02679"></a>02679 <span class="stringliteral">            mapMets.x = scaleX * (mapMets.x - minX);</span>
<a name="l02680"></a>02680 <span class="stringliteral">            mapMets.y = scaleY * (mapMets.y - minY);</span>
<a name="l02681"></a>02681 <span class="stringliteral">            for i = 1:numel(mapRxns.path)</span>
<a name="l02682"></a>02682 <span class="stringliteral">                [starts, ends, tokens] = regexp(mapRxns.path{i}, &#39;</span>(-*\d*\.*\d*),(-*\d*\.*\d*)<span class="stringliteral">&#39;, &#39;</span>start<span class="stringliteral">&#39;, &#39;</span>end<span class="stringliteral">&#39;, &#39;</span>tokens<span class="stringliteral">&#39;);</span>
<a name="l02683"></a>02683 <span class="stringliteral">                for j = 1:numel(starts)</span>
<a name="l02684"></a>02684 <span class="stringliteral">                    valX = num2str(scaleX * (str2double(tokens{j}{1}) - minX));</span>
<a name="l02685"></a>02685 <span class="stringliteral">                    valY = num2str(scaleY * (str2double(tokens{j}{2}) - minY));</span>
<a name="l02686"></a>02686 <span class="stringliteral">                    mapRxns.path{i} = [...</span>
<a name="l02687"></a>02687 <span class="stringliteral">                        mapRxns.path{i}(1:starts(j)-1) ...</span>
<a name="l02688"></a>02688 <span class="stringliteral">                        valX ...</span>
<a name="l02689"></a>02689 <span class="stringliteral">                        &#39;</span>,<span class="stringliteral">&#39; ...</span>
<a name="l02690"></a>02690 <span class="stringliteral">                        valY ...</span>
<a name="l02691"></a>02691 <span class="stringliteral">                        mapRxns.path{i}(ends(j)+1:end) ...</span>
<a name="l02692"></a>02692 <span class="stringliteral">                        ];</span>
<a name="l02693"></a>02693 <span class="stringliteral">                    starts = starts + numel(valX) + numel(valY) - numel(tokens{j}{1}) - numel(tokens{j}{2});</span>
<a name="l02694"></a>02694 <span class="stringliteral">                    ends = ends + numel(valX) + numel(valY) - numel(tokens{j}{1}) - numel(tokens{j}{2});                    </span>
<a name="l02695"></a>02695 <span class="stringliteral">                end</span>
<a name="l02696"></a>02696 <span class="stringliteral">            end</span>
<a name="l02697"></a>02697 <span class="stringliteral">            mapRxns.labelX = scaleX * (mapRxns.labelX - minX);</span>
<a name="l02698"></a>02698 <span class="stringliteral">            mapRxns.labelY = scaleY * (mapRxns.labelY - minY);</span>
<a name="l02699"></a>02699 <span class="stringliteral">            mapRxns.valueX = scaleX * (mapRxns.valueX - minX);</span>
<a name="l02700"></a>02700 <span class="stringliteral">            mapRxns.valueY = scaleY * (mapRxns.valueY - minY);</span>
<a name="l02701"></a>02701 <span class="stringliteral">            </span>
<a name="l02702"></a>02702 <span class="stringliteral">            if nargin &gt;= 5</span>
<a name="l02703"></a>02703 <span class="stringliteral">                axesHandle = subplot(&#39;</span>position<span class="stringliteral">&#39;, position, &#39;</span>parent<span class="stringliteral">&#39;, figHandle);</span>
<a name="l02704"></a>02704 <span class="stringliteral">            else</span>
<a name="l02705"></a>02705 <span class="stringliteral">                axesHandle = edu.stanford.covert.cell.sim.util.PlotUtil.newAxesHandle();</span>
<a name="l02706"></a>02706 <span class="stringliteral">            end</span>
<a name="l02707"></a>02707 <span class="stringliteral">            hold(axesHandle, &#39;</span>on<span class="stringliteral">&#39;);</span>
<a name="l02708"></a>02708 <span class="stringliteral">            set(axesHandle, &#39;</span>YDir<span class="stringliteral">&#39;, &#39;</span>reverse<span class="stringliteral">&#39;);</span>
<a name="l02709"></a>02709 <span class="stringliteral">            xlim(axesHandle, [0 W+10]);</span>
<a name="l02710"></a>02710 <span class="stringliteral">            ylim(axesHandle, [-10 H+3]);</span>
<a name="l02711"></a>02711 <span class="stringliteral">            set(axesHandle, &#39;</span>visible<span class="stringliteral">&#39;, &#39;</span>off<span class="stringliteral">&#39;)</span>
<a name="l02712"></a>02712 <span class="stringliteral">            </span>
<a name="l02713"></a>02713 <span class="stringliteral">            %metabolites</span>
<a name="l02714"></a>02714 <span class="stringliteral">            tfs = true(size(mapMets.x));</span>
<a name="l02715"></a>02715 <span class="stringliteral">            tfs([59:63 14 39 85 53 67 18 69 35 40 33 44 37 70 83]) = false;</span>
<a name="l02716"></a>02716 <span class="stringliteral">            plot(mapMets.x(tfs), mapMets.y(tfs), &#39;</span>o<span class="stringliteral">&#39;, &#39;</span>MarkerEdgeColor<span class="stringliteral">&#39;, &#39;</span>k<span class="stringliteral">&#39;, &#39;</span>MarkerFaceColor<span class="stringliteral">&#39;, 0.2 * [1 1 1], &#39;</span>MarkerSize<span class="stringliteral">&#39;, metRadius);</span>
<a name="l02717"></a>02717 <span class="stringliteral">            if showAllLabels</span>
<a name="l02718"></a>02718 <span class="stringliteral">                for i = 1:numel(mapMets.x)</span>
<a name="l02719"></a>02719 <span class="stringliteral">                    if ~tfs(i)</span>
<a name="l02720"></a>02720 <span class="stringliteral">                        continue;</span>
<a name="l02721"></a>02721 <span class="stringliteral">                    end</span>
<a name="l02722"></a>02722 <span class="stringliteral">                    text(mapMets.x(i)+7, mapMets.y(i)+7, met.wholeCellModelIDs{mapMets.idxs(i)}, ...</span>
<a name="l02723"></a>02723 <span class="stringliteral">                        &#39;</span>FontSize<span class="stringliteral">&#39;, 5, &#39;</span>Color<span class="stringliteral">&#39;, &#39;</span>k<span class="stringliteral">&#39;, &#39;</span>Interpreter<span class="stringliteral">&#39;, &#39;</span>none<span class="stringliteral">&#39;);</span>
<a name="l02724"></a>02724 <span class="stringliteral">                end</span>
<a name="l02725"></a>02725 <span class="stringliteral">            end</span>
<a name="l02726"></a>02726 <span class="stringliteral">            </span>
<a name="l02727"></a>02727 <span class="stringliteral">            %reactions</span>
<a name="l02728"></a>02728 <span class="stringliteral">            relMean = log10(abs(fluxMeans));</span>
<a name="l02729"></a>02729 <span class="stringliteral">            relMean(isinf(relMean)) = NaN;</span>
<a name="l02730"></a>02730 <span class="stringliteral">            relMean = max(0, min(4, relMean)) / 4;</span>
<a name="l02731"></a>02731 <span class="stringliteral">            strokeWidth = 0.25 * ones(size(relMean));</span>
<a name="l02732"></a>02732 <span class="stringliteral">            color1 = 0.2 * [0 0 1] + 0.8 * (0.9 * [1 1 1]);</span>
<a name="l02733"></a>02733 <span class="stringliteral">            color2 = [0 0 1];</span>
<a name="l02734"></a>02734 <span class="stringliteral">            colors = ...</span>
<a name="l02735"></a>02735 <span class="stringliteral">                + repmat(color1, numel(relMean), 1) .* repmat(1-relMean, 1, 3) ...</span>
<a name="l02736"></a>02736 <span class="stringliteral">                + repmat(color2, numel(relMean), 1) .* repmat(relMean, 1, 3);</span>
<a name="l02737"></a>02737 <span class="stringliteral">            </span>
<a name="l02738"></a>02738 <span class="stringliteral">            [~, order] = sort(relMean);</span>
<a name="l02739"></a>02739 <span class="stringliteral">            relMean = relMean(order, :);</span>
<a name="l02740"></a>02740 <span class="stringliteral">            fluxMeans = fluxMeans(order, :);</span>
<a name="l02741"></a>02741 <span class="stringliteral">            colors = colors(order, :);</span>
<a name="l02742"></a>02742 <span class="stringliteral">            strokeWidth = strokeWidth(order, :);</span>
<a name="l02743"></a>02743 <span class="stringliteral">            mapRxns.idxs = mapRxns.idxs(order);</span>
<a name="l02744"></a>02744 <span class="stringliteral">            mapRxns.path = mapRxns.path(order);</span>
<a name="l02745"></a>02745 <span class="stringliteral">            mapRxns.labelX = mapRxns.labelX(order);</span>
<a name="l02746"></a>02746 <span class="stringliteral">            mapRxns.labelY = mapRxns.labelY(order);</span>
<a name="l02747"></a>02747 <span class="stringliteral">            mapRxns.valueX = mapRxns.valueX(order);</span>
<a name="l02748"></a>02748 <span class="stringliteral">            mapRxns.valueY = mapRxns.valueY(order);</span>
<a name="l02749"></a>02749 <span class="stringliteral">            </span>
<a name="l02750"></a>02750 <span class="stringliteral">            arrowRadii = arrowRadius * ones(size(mapRxns.idxs));</span>
<a name="l02751"></a>02751 <span class="stringliteral">            arrowRadii(mapRxns.idxs == metabolism.reactionIndexs(&#39;</span>AspC1<span class="stringliteral">&#39;)) = 0.6 * arrowRadius;</span>
<a name="l02752"></a>02752 <span class="stringliteral">            </span>
<a name="l02753"></a>02753 <span class="stringliteral">            [~, hideFluxIdxs] = ismember({&#39;</span>GlpQ1<span class="stringliteral">&#39;, &#39;</span>GlpQ2<span class="stringliteral">&#39;, &#39;</span>GlpQ3<span class="stringliteral">&#39;, &#39;</span>GlpQ4<span class="stringliteral">&#39;, &#39;</span>GlpQ5<span class="stringliteral">&#39;, &#39;</span>DeoD<span class="stringliteral">&#39;, &#39;</span>Rpe<span class="stringliteral">&#39;, &#39;</span>RpiB<span class="stringliteral">&#39;}, mr.reactionWholeCellModelIDs);</span>
<a name="l02754"></a>02754 <span class="stringliteral">            for i = 1:numel(mapRxns.idxs)</span>
<a name="l02755"></a>02755 <span class="stringliteral">                if isoFluxs(mapRxns.idxs(i)) || any(mapRxns.idxs(i) == hideFluxIdxs) || fluxMeans(i) == 0</span>
<a name="l02756"></a>02756 <span class="stringliteral">                    continue;</span>
<a name="l02757"></a>02757 <span class="stringliteral">                end</span>
<a name="l02758"></a>02758 <span class="stringliteral">                </span>
<a name="l02759"></a>02759 <span class="stringliteral">                endX = 0;</span>
<a name="l02760"></a>02760 <span class="stringliteral">                endY = 0;</span>
<a name="l02761"></a>02761 <span class="stringliteral">                endAngle = NaN;</span>
<a name="l02762"></a>02762 <span class="stringliteral">                tokens = regexp(mapRxns.path{i}, &#39;</span>([A-Za-z])( (-*\d*\.*\d*),(-*\d*\.*\d*))+<span class="stringliteral">&#39;, &#39;</span>tokens<span class="stringliteral">&#39;);</span>
<a name="l02763"></a>02763 <span class="stringliteral">                for j = 1:numel(tokens)</span>
<a name="l02764"></a>02764 <span class="stringliteral">                    tokens2 = regexp(tokens{j}{2}, &#39;</span>(-*\d*\.*\d*),(-*\d*\.*\d*)<span class="stringliteral">&#39;, &#39;</span>tokens<span class="stringliteral">&#39;);</span>
<a name="l02765"></a>02765 <span class="stringliteral">                    switch tokens{j}{1}</span>
<a name="l02766"></a>02766 <span class="stringliteral">                        case &#39;</span>M<span class="stringliteral">&#39;</span>
<a name="l02767"></a>02767 <span class="stringliteral">                            endX = str2double(tokens2{1}{1});</span>
<a name="l02768"></a>02768 <span class="stringliteral">                            endY = str2double(tokens2{1}{2});</span>
<a name="l02769"></a>02769 <span class="stringliteral">                        case &#39;</span>L<span class="stringliteral">&#39;</span>
<a name="l02770"></a>02770 <span class="stringliteral">                            startX = endX;</span>
<a name="l02771"></a>02771 <span class="stringliteral">                            startY = endY;                            </span>
<a name="l02772"></a>02772 <span class="stringliteral">                            endX2 = str2double(tokens2{1}{1});</span>
<a name="l02773"></a>02773 <span class="stringliteral">                            endY2 = str2double(tokens2{1}{2});</span>
<a name="l02774"></a>02774 <span class="stringliteral">                            line([endX endX2], [endY endY2], ...</span>
<a name="l02775"></a>02775 <span class="stringliteral">                                &#39;</span>Parent<span class="stringliteral">&#39;, axesHandle, ...</span>
<a name="l02776"></a>02776 <span class="stringliteral">                                &#39;</span>Color<span class="stringliteral">&#39;, colors(i, :) * (~showAllLabels), ...</span>
<a name="l02777"></a>02777 <span class="stringliteral">                                &#39;</span>lineWidth<span class="stringliteral">&#39;, strokeWidth(i));</span>
<a name="l02778"></a>02778 <span class="stringliteral">                            startAngle = atan2(endY-endY2, endX-endX2);</span>
<a name="l02779"></a>02779 <span class="stringliteral">                            endAngle = atan2(endY2 - endY, endX2 - endX);</span>
<a name="l02780"></a>02780 <span class="stringliteral">                            endX = endX2;</span>
<a name="l02781"></a>02781 <span class="stringliteral">                            endY = endY2;</span>
<a name="l02782"></a>02782 <span class="stringliteral">                        case &#39;</span>C<span class="stringliteral">&#39;</span>
<a name="l02783"></a>02783 <span class="stringliteral">                            startX = endX;</span>
<a name="l02784"></a>02784 <span class="stringliteral">                            startY = endY;</span>
<a name="l02785"></a>02785 <span class="stringliteral">                            endX2 = str2double(tokens2{1}{1});</span>
<a name="l02786"></a>02786 <span class="stringliteral">                            endY2 = str2double(tokens2{1}{2});</span>
<a name="l02787"></a>02787 <span class="stringliteral">                            endX3 = str2double(tokens2{2}{1});</span>
<a name="l02788"></a>02788 <span class="stringliteral">                            endY3 = str2double(tokens2{2}{2});</span>
<a name="l02789"></a>02789 <span class="stringliteral">                            endX4 = str2double(tokens2{3}{1});</span>
<a name="l02790"></a>02790 <span class="stringliteral">                            endY4 = str2double(tokens2{3}{2});</span>
<a name="l02791"></a>02791 <span class="stringliteral">                            [h, qx, qy] = Funct_Bezier([endX endX2 endX3 endX4], [endY endY2 endY3 endY4], 50, axesHandle);</span>
<a name="l02792"></a>02792 <span class="stringliteral">                            set(h, &#39;</span>color<span class="stringliteral">&#39;, colors(i, :) * (~showAllLabels), &#39;</span>lineWidth<span class="stringliteral">&#39;, strokeWidth(i));</span>
<a name="l02793"></a>02793 <span class="stringliteral">                            startAngle = atan2(qy(1)-qy(2), qx(1)-qx(2));</span>
<a name="l02794"></a>02794 <span class="stringliteral">                            endAngle = atan2(qy(end)-qy(end-1), qx(end)-qx(end-1));</span>
<a name="l02795"></a>02795 <span class="stringliteral">                            endX = endX4;</span>
<a name="l02796"></a>02796 <span class="stringliteral">                            endY = endY4;</span>
<a name="l02797"></a>02797 <span class="stringliteral">                        otherwise</span>
<a name="l02798"></a>02798 <span class="stringliteral">                            throw(MException(&#39;</span><a class="code" href="class_figures34.html">Figures34</a>:error<span class="stringliteral">&#39;, &#39;</span>undefined command %s<span class="stringliteral">&#39;, tokens{j}{1}))</span>
<a name="l02799"></a>02799 <span class="stringliteral">                    end</span>
<a name="l02800"></a>02800 <span class="stringliteral">                    if j == 2 &amp;&amp; fluxMeans(i) &lt; 0</span>
<a name="l02801"></a>02801 <span class="stringliteral">                         x = startX + 2*cos(startAngle) + [0  arrowRadii(i)*cos(startAngle+pi/3+pi/2)  arrowRadii(i)*cos(startAngle+pi*2/3+pi/2)];</span>
<a name="l02802"></a>02802 <span class="stringliteral">                         y = startY + 2*sin(startAngle) + [0  arrowRadii(i)*sin(startAngle+pi/3+pi/2)  arrowRadii(i)*sin(startAngle+pi*2/3+pi/2)];</span>
<a name="l02803"></a>02803 <span class="stringliteral">                         </span>
<a name="l02804"></a>02804 <span class="stringliteral">                         arrowPtch = patch(x, y, 1, &#39;</span>Parent<span class="stringliteral">&#39;, axesHandle, &#39;</span>EdgeColor<span class="stringliteral">&#39;, &#39;</span>none<span class="stringliteral">&#39;);</span>
<a name="l02805"></a>02805 <span class="stringliteral">                         set(arrowPtch, &#39;</span>FaceColor<span class="stringliteral">&#39;, colors(i, :) * (~showAllLabels), &#39;</span>FaceAlpha<span class="stringliteral">&#39;, 1, &#39;</span>EdgeAlpha<span class="stringliteral">&#39;, 1);</span>
<a name="l02806"></a>02806 <span class="stringliteral">                    end</span>
<a name="l02807"></a>02807 <span class="stringliteral">                end</span>
<a name="l02808"></a>02808 <span class="stringliteral">                if ~isnan(endAngle) &amp;&amp; fluxMeans(i) &gt; 0</span>
<a name="l02809"></a>02809 <span class="stringliteral">                    x = endX + 2*cos(endAngle) + [0  arrowRadii(i)*cos(endAngle+pi/3+pi/2)  arrowRadii(i)*cos(endAngle+pi*2/3+pi/2)];</span>
<a name="l02810"></a>02810 <span class="stringliteral">                    y = endY + 2*sin(endAngle) + [0  arrowRadii(i)*sin(endAngle+pi/3+pi/2)  arrowRadii(i)*sin(endAngle+pi*2/3+pi/2)];</span>
<a name="l02811"></a>02811 <span class="stringliteral">                    </span>
<a name="l02812"></a>02812 <span class="stringliteral">                    arrowPtch = patch(x, y, 1, &#39;</span>Parent<span class="stringliteral">&#39;, axesHandle, &#39;</span>EdgeColor<span class="stringliteral">&#39;, &#39;</span>none<span class="stringliteral">&#39;);</span>
<a name="l02813"></a>02813 <span class="stringliteral">                    set(arrowPtch, &#39;</span>FaceColor<span class="stringliteral">&#39;, colors(i, :) * (~showAllLabels), &#39;</span>FaceAlpha<span class="stringliteral">&#39;, 1, &#39;</span>EdgeAlpha<span class="stringliteral">&#39;, 1);</span>
<a name="l02814"></a>02814 <span class="stringliteral">                end</span>
<a name="l02815"></a>02815 <span class="stringliteral">                </span>
<a name="l02816"></a>02816 <span class="stringliteral">                if showAllLabels</span>
<a name="l02817"></a>02817 <span class="stringliteral">                    text(mapRxns.labelX(i), mapRxns.labelY(i), ...</span>
<a name="l02818"></a>02818 <span class="stringliteral">                        metabolism.reactionWholeCellModelIDs(mapRxns.idxs(i)), ...</span>
<a name="l02819"></a>02819 <span class="stringliteral">                        &#39;</span>FontSize<span class="stringliteral">&#39;, 5, &#39;</span>Interpreter<span class="stringliteral">&#39;, &#39;</span>none<span class="stringliteral">&#39;);</span>
<a name="l02820"></a>02820 <span class="stringliteral">                end</span>
<a name="l02821"></a>02821 <span class="stringliteral">            end</span>
<a name="l02822"></a>02822 <span class="stringliteral">            </span>
<a name="l02823"></a>02823 <span class="stringliteral">            %labels</span>
<a name="l02824"></a>02824 <span class="stringliteral">            labels = {</span>
<a name="l02825"></a>02825 <span class="stringliteral">                [90 90 210 210]            [60 400 400 60]          [0 1 0]  [80 360] &#39;</span>Glycolysis<span class="stringliteral">&#39;                 0.2</span>
<a name="l02826"></a>02826 <span class="stringliteral">                [230 230 370 370]          [55 230 230 55]          [1 0 0]  [365 110] &#39;</span>PPP<span class="stringliteral">&#39;                       0.2</span>
<a name="l02827"></a>02827 <span class="stringliteral">                [265 265 425 425]          [230 340 340 230]        [0 0 1]  [345 360] &#39;</span>ATP synthesis<span class="stringliteral">&#39;             0.2</span>
<a name="l02828"></a>02828 <span class="stringliteral">                [450 680 680 820 820 450]  [20 20 320 320 580 580]  [0 1 1]  [635 595] &#39;</span>Nucleotide metabolism<span class="stringliteral">&#39;     0.1</span>
<a name="l02829"></a>02829 <span class="stringliteral">                [170 390 390 170]          [420 420 810 810]        [1 0.5 0]  [280 560] {&#39;</span>Pyruvate<span class="charliteral">&#39; &#39;</span>metabolism<span class="stringliteral">&#39;} 0.3</span>
<a name="l02830"></a>02830 <span class="stringliteral">                };</span>
<a name="l02831"></a>02831 <span class="stringliteral">            h = zeros(size(labels, 1), 1);</span>
<a name="l02832"></a>02832 <span class="stringliteral">            for i = 1:size(labels, 1)</span>
<a name="l02833"></a>02833 <span class="stringliteral">                ptch = patch(labels{i, 1}, labels{i, 2}, 1, &#39;</span>Parent<span class="stringliteral">&#39;, axesHandle, &#39;</span>EdgeColor<span class="stringliteral">&#39;, &#39;</span>none<span class="stringliteral">&#39;);</span>
<a name="l02834"></a>02834 <span class="stringliteral">                set(ptch, &#39;</span>FaceColor<span class="stringliteral">&#39;, labels{i,6}*labels{i, 3} + (1-labels{i,6})*[1 1 1], &#39;</span>FaceAlpha<span class="stringliteral">&#39;, 1, &#39;</span>EdgeAlpha<span class="stringliteral">&#39;, 1);</span>
<a name="l02835"></a>02835 <span class="stringliteral">                uistack(ptch, &#39;</span>bottom<span class="stringliteral">&#39;);</span>
<a name="l02836"></a>02836 <span class="stringliteral">                </span>
<a name="l02837"></a>02837 <span class="stringliteral">                h(i) = text(labels{i, 4}(1), labels{i, 4}(2), labels{i, 5}, ...</span>
<a name="l02838"></a>02838 <span class="stringliteral">                    &#39;</span>HorizontalAlign<span class="stringliteral">&#39;, &#39;</span>center<span class="stringliteral">&#39;,  ...</span>
<a name="l02839"></a>02839 <span class="stringliteral">                    &#39;</span>VerticalAlign<span class="stringliteral">&#39;, &#39;</span>middle<span class="stringliteral">&#39;,  ...</span>
<a name="l02840"></a>02840 <span class="stringliteral">                    &#39;</span>FontSize<span class="stringliteral">&#39;, 7, ...</span>
<a name="l02841"></a>02841 <span class="stringliteral">                    &#39;</span>Color<span class="stringliteral">&#39;, labels{i, 3});</span>
<a name="l02842"></a>02842 <span class="stringliteral">            end</span>
<a name="l02843"></a>02843 <span class="stringliteral">            set(h(2), &#39;</span>HorizontalAlign<span class="stringliteral">&#39;, &#39;</span>left<span class="stringliteral">&#39;)</span>
<a name="l02844"></a>02844 <span class="stringliteral">            </span>
<a name="l02845"></a>02845 <span class="stringliteral">            %colormap</span>
<a name="l02846"></a>02846 <span class="stringliteral">            if ~showAllLabels</span>
<a name="l02847"></a>02847 <span class="stringliteral">                x1 = W-80;</span>
<a name="l02848"></a>02848 <span class="stringliteral">                x2 = x1 + 20;</span>
<a name="l02849"></a>02849 <span class="stringliteral">                h = 6.5;</span>
<a name="l02850"></a>02850 <span class="stringliteral">                y = 20 + h*40;</span>
<a name="l02851"></a>02851 <span class="stringliteral">                colorMap = ...</span>
<a name="l02852"></a>02852 <span class="stringliteral">                    + repmat(color1, 40, 1) .* repmat(1 - linspace(0, 1, 40)&#39;</span>, 1, 3) ...
<a name="l02853"></a>02853                     + repmat(color2, 40, 1) .* repmat(    linspace(0, 1, 40)&#39;, 1, 3);
<a name="l02854"></a>02854                 <span class="keywordflow">for</span> i = 1:40
<a name="l02855"></a>02855                     ptch = patch([x1 x2 x2 x1], y - h*[i-1 i-1 i i], 1, <span class="stringliteral">&#39;EdgeColor&#39;</span>, <span class="stringliteral">&#39;none&#39;</span>, <span class="stringliteral">&#39;Parent&#39;</span>, axesHandle);
<a name="l02856"></a>02856                     <span class="keyword">set</span>(ptch, <span class="stringliteral">&#39;FaceColor&#39;</span>, colorMap(i, :), <span class="stringliteral">&#39;FaceAlpha&#39;</span>, 1, <span class="stringliteral">&#39;EdgeAlpha&#39;</span>, 1);
<a name="l02857"></a>02857                 end
<a name="l02858"></a>02858                 patch([x1 x2 x2 x1], y - [0 0 40*h 40*h], 1, <span class="stringliteral">&#39;EdgeColor&#39;</span>, <span class="charliteral">&#39;k&#39;</span>, <span class="stringliteral">&#39;FaceColor&#39;</span>, <span class="stringliteral">&#39;none&#39;</span>);
<a name="l02859"></a>02859                 
<a name="l02860"></a>02860                 line([x2 x2+5], y + [0 0], <span class="stringliteral">&#39;Color&#39;</span>, <span class="charliteral">&#39;k&#39;</span>, <span class="stringliteral">&#39;parent&#39;</span>, axesHandle)
<a name="l02861"></a>02861                 line([x2 x2+5], y-20*h + [0 0], &#39;Color&#39;, &#39;k&#39;, &#39;parent&#39;, axesHandle)
<a name="l02862"></a>02862                 line([x2 x2+5], y-40*h + [0 0], &#39;Color&#39;, &#39;k&#39;, &#39;parent&#39;, axesHandle)
<a name="l02863"></a>02863                 
<a name="l02864"></a>02864                 text(x2+50, y-20*h, &#39;Flux (rxn s^{-1})&#39;, ...
<a name="l02865"></a>02865                     &#39;Color&#39;, &#39;k&#39;, &#39;parent&#39;, axesHandle, &#39;Interpreter&#39;, &#39;tex&#39;, ...
<a name="l02866"></a>02866                     &#39;HorizontalAlign&#39;, &#39;center&#39;, &#39;VerticalAlign&#39;, &#39;middle&#39;, &#39;FontSize&#39;, 7, &#39;rotation&#39;, 270);
<a name="l02867"></a>02867                 text(x2+10, y, sprintf(<span class="stringliteral">&#39;10^{%d}&#39;</span>, 0), ...
<a name="l02868"></a>02868                     <span class="stringliteral">&#39;Color&#39;</span>, <span class="charliteral">&#39;k&#39;</span>, <span class="stringliteral">&#39;parent&#39;</span>, axesHandle, <span class="stringliteral">&#39;Interpreter&#39;</span>, <span class="stringliteral">&#39;tex&#39;</span>, ...
<a name="l02869"></a>02869                     <span class="stringliteral">&#39;HorizontalAlign&#39;</span>, <span class="stringliteral">&#39;left&#39;</span>, <span class="stringliteral">&#39;VerticalAlign&#39;</span>, <span class="stringliteral">&#39;middle&#39;</span>, <span class="stringliteral">&#39;FontSize&#39;</span>, 5);
<a name="l02870"></a>02870                 text(x2+10, y-20*h, sprintf(<span class="stringliteral">&#39;10^{%d}&#39;</span>, 2), ...
<a name="l02871"></a>02871                     <span class="stringliteral">&#39;Color&#39;</span>, <span class="charliteral">&#39;k&#39;</span>, <span class="stringliteral">&#39;parent&#39;</span>, axesHandle, <span class="stringliteral">&#39;Interpreter&#39;</span>, <span class="stringliteral">&#39;tex&#39;</span>, ...
<a name="l02872"></a>02872                     <span class="stringliteral">&#39;HorizontalAlign&#39;</span>, <span class="stringliteral">&#39;left&#39;</span>, <span class="stringliteral">&#39;VerticalAlign&#39;</span>, <span class="stringliteral">&#39;middle&#39;</span>, <span class="stringliteral">&#39;FontSize&#39;</span>, 5);
<a name="l02873"></a>02873                 text(x2+10, y-40*h, sprintf(<span class="stringliteral">&#39;10^{%d}&#39;</span>, 4), ...
<a name="l02874"></a>02874                     <span class="stringliteral">&#39;Color&#39;</span>, <span class="charliteral">&#39;k&#39;</span>, <span class="stringliteral">&#39;parent&#39;</span>, axesHandle, <span class="stringliteral">&#39;Interpreter&#39;</span>, <span class="stringliteral">&#39;tex&#39;</span>, ...
<a name="l02875"></a>02875                     <span class="stringliteral">&#39;HorizontalAlign&#39;</span>, <span class="stringliteral">&#39;left&#39;</span>, <span class="stringliteral">&#39;VerticalAlign&#39;</span>, <span class="stringliteral">&#39;middle&#39;</span>, <span class="stringliteral">&#39;FontSize&#39;</span>, 5);
<a name="l02876"></a>02876             end
<a name="l02877"></a>02877             
<a name="l02878"></a>02878             %% labels
<a name="l02879"></a>02879             <span class="keywordflow">if</span> ~showAllLabels
<a name="l02880"></a>02880                 idxGpsA = find(mapRxns.idxs == sim.process(<span class="stringliteral">&#39;Metabolism&#39;</span>).reactionIndexs(<span class="stringliteral">&#39;GpsA&#39;</span>), 1, <span class="stringliteral">&#39;first&#39;</span>);
<a name="l02881"></a>02881                 idxTalA = find(mapRxns.idxs == sim.process(<span class="stringliteral">&#39;Metabolism&#39;</span>).reactionIndexs(<span class="stringliteral">&#39;TalA&#39;</span>), 1, <span class="stringliteral">&#39;first&#39;</span>);
<a name="l02882"></a>02882                 idxPgi = find(mapRxns.idxs == sim.process(<span class="stringliteral">&#39;Metabolism&#39;</span>).reactionIndexs(<span class="stringliteral">&#39;Pgi&#39;</span>), 1, <span class="stringliteral">&#39;first&#39;</span>);
<a name="l02883"></a>02883                 
<a name="l02884"></a>02884                 text(95, 230, {<span class="stringliteral">&#39;GpsA&#39;</span> sprintf(<span class="stringliteral">&#39;(%.2f%%)&#39;</span>, abs(fluxMeans(idxGpsA)/fluxMeans(idxPgi)*100))}, ...
<a name="l02885"></a>02885                     <span class="stringliteral">&#39;FontSize&#39;</span>, 5, ...
<a name="l02886"></a>02886                     <span class="stringliteral">&#39;HorizontalAlign&#39;</span>, <span class="stringliteral">&#39;right&#39;</span>)
<a name="l02887"></a>02887                 text(365, 170, {<span class="stringliteral">&#39;TalA&#39;</span> sprintf(<span class="stringliteral">&#39;(%.2f%%)&#39;</span>, abs(fluxMeans(idxTalA)/fluxMeans(idxPgi)*100))}, ...
<a name="l02888"></a>02888                     <span class="stringliteral">&#39;FontSize&#39;</span>, 5, ...
<a name="l02889"></a>02889                     <span class="stringliteral">&#39;HorizontalAlign&#39;</span>, <span class="stringliteral">&#39;left&#39;</span>)
<a name="l02890"></a>02890             end
<a name="l02891"></a>02891         end
<a name="l02892"></a>02892         
<a name="l02893"></a>02893         <span class="keyword">function</span> chromosomeSampling(simBatchDir, selectedSim, outDirectory, figHandle, position)
<a name="l02894"></a>02894             import edu.stanford.covert.cell.sim.analysis.<a class="code" href="class_figures34.html">Figures34</a>;
<a name="l02895"></a>02895             import edu.stanford.covert.cell.sim.util.PlotUtil;
<a name="l02896"></a>02896             import edu.stanford.covert.cell.sim.util.SimulationEnsemble;
<a name="l02897"></a>02897             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
<a name="l02898"></a>02898             
<a name="l02899"></a>02899             if ~exist(&#39;simBatchDir&#39;, &#39;var&#39;) || isempty(simBatchDir)
<a name="l02900"></a>02900                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
<a name="l02901"></a>02901             end
<a name="l02902"></a>02902             if ~exist(&#39;selectedSim&#39;, &#39;var&#39;) || isempty(selectedSim)
<a name="l02903"></a>02903                 selectedSim = 1;
<a name="l02904"></a>02904             end
<a name="l02905"></a>02905             
<a name="l02906"></a>02906             %% get data
<a name="l02907"></a>02907             if exist([outDirectory &#39;chromosomeSampling.mat&#39;], &#39;file&#39;)
<a name="l02908"></a>02908                 load([outDirectory &#39;chromosomeSampling.mat&#39;]);
<a name="l02909"></a>02909             else
<a name="l02910"></a>02910                 [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep num2str(selectedSim)]);
<a name="l02911"></a>02911                 ch = sim.state(&#39;<a class="code" href="class_chromosome.html">Chromosome</a>&#39;);
<a name="l02912"></a>02912                 pc = sim.state(&#39;<a class="code" href="class_protein_complex.html">ProteinComplex</a>&#39;);
<a name="l02913"></a>02913                 rnaPol = sim.state(&#39;RNAPolymerase&#39;);
<a name="l02914"></a>02914                 
<a name="l02915"></a>02915                 stateNames = {
<a name="l02916"></a>02916                     <span class="stringliteral">&#39;Time&#39;</span>        <span class="stringliteral">&#39;values&#39;</span>             <span class="charliteral">&#39;:&#39;</span>  <span class="charliteral">&#39;:&#39;</span>
<a name="l02917"></a>02917                     <span class="stringliteral">&#39;Chromosome&#39;</span>  <span class="stringliteral">&#39;monomerBoundSites&#39;</span>  <span class="charliteral">&#39;:&#39;</span>  <span class="charliteral">&#39;:&#39;</span>
<a name="l02918"></a>02918                     <span class="stringliteral">&#39;Chromosome&#39;</span>  <span class="stringliteral">&#39;complexBoundSites&#39;</span>  <span class="charliteral">&#39;:&#39;</span>  <span class="charliteral">&#39;:&#39;</span>
<a name="l02919"></a>02919                     <span class="stringliteral">&#39;RNAPolymerase&#39;</span> <span class="stringliteral">&#39;positionStrands&#39;</span>  <span class="charliteral">&#39;:&#39;</span>  <span class="charliteral">&#39;:&#39;</span>
<a name="l02920"></a>02920                     <span class="stringliteral">&#39;RNAPolymerase&#39;</span> <span class="stringliteral">&#39;states&#39;</span>           <span class="charliteral">&#39;:&#39;</span>  <span class="charliteral">&#39;:&#39;</span>
<a name="l02921"></a>02921                     };
<a name="l02922"></a>02922                 states = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, <span class="stringliteral">&#39;extract&#39;</span>, selectedSim);
<a name="l02923"></a>02923                 time = permute(states.Time.values, [3 1 2]) / 3600;
<a name="l02924"></a>02924                 
<a name="l02925"></a>02925                 %monomers
<a name="l02926"></a>02926                 [subs, vals] = find(states.Chromosome.monomerBoundSites);
<a name="l02927"></a>02927                 times = subs(:, 3);
<a name="l02928"></a>02928                 startPos = subs(:, 1);
<a name="l02929"></a>02929                 endPos = subs(:, 1) + ch.monomerDNAFootprints(vals) - 1;
<a name="l02930"></a>02930                 idxs = find(endPos &gt; ch.sequenceLen); 
<a name="l02931"></a>02931                 vals = [vals; vals(idxs)];
<a name="l02932"></a>02932                 times = [times; times(idxs)];
<a name="l02933"></a>02933                 startPos = [startPos; ones(size(idxs))];
<a name="l02934"></a>02934                 endPos = [endPos; endPos(idxs) - ch.sequenceLen];
<a name="l02935"></a>02935                 endPos(idxs) = ch.sequenceLen;
<a name="l02936"></a>02936                 monTimeBound = <a class="code" href="class_figures34.html">Figures34</a>.<a class="code" href="class_figures34.html#a1d2e3eeb8019d1d25402d337f2265bae">calcChromosomeSampling</a>(startPos, endPos, times, 1, ch.sequenceLen, ch.sequenceLen);
<a name="l02937"></a>02937                 
<a name="l02938"></a>02938                 %complexes
<a name="l02939"></a>02939                 complexDNAFootprints = ch.complexDNAFootprints;
<a name="l02940"></a>02940                 complexDNAFootprints(pc.dnaPolymeraseIndexs) = max(complexDNAFootprints(pc.dnaPolymeraseIndexs), sim.process(<span class="stringliteral">&#39;Replication&#39;</span>).dnaPolymeraseElongationRate);
<a name="l02941"></a>02941                 complexDNAFootprints(pc.getIndexs(<span class="stringliteral">&#39;MG_094_HEXAMER&#39;</span>)) = max(complexDNAFootprints(pc.getIndexs(<span class="stringliteral">&#39;MG_094_HEXAMER&#39;</span>)), sim.process(<span class="stringliteral">&#39;Replication&#39;</span>).dnaPolymeraseElongationRate);
<a name="l02942"></a>02942                 
<a name="l02943"></a>02943                 [subs, vals] = find(states.Chromosome.complexBoundSites);
<a name="l02944"></a>02944                 subs(ismember(vals, pc.rnaPolymeraseIndexs), :) = [];
<a name="l02945"></a>02945                 vals(ismember(vals, pc.rnaPolymeraseIndexs), :) = [];
<a name="l02946"></a>02946                 times = subs(:, 3);
<a name="l02947"></a>02947                 startPos = subs(:, 1);
<a name="l02948"></a>02948                 endPos = subs(:, 1) + complexDNAFootprints(vals) - 1;
<a name="l02949"></a>02949                 
<a name="l02950"></a>02950                 tmp = [reshape(states.RNAPolymerase.positionStrands(:, 1, :), [], 1)  ...
<a name="l02951"></a>02951                     reshape(repmat(permute(1:size(states.RNAPolymerase.positionStrands, 3), [1 3 2]), [size(states.RNAPolymerase.positionStrands, 1), 1 1]), [], 1) ...
<a name="l02952"></a>02952                     reshape(states.RNAPolymerase.states(:, 1, :), [], 1)];
<a name="l02953"></a>02953                 tmp(tmp(:, 3) == rnaPol.freeValue | tmp(:, 3) == rnaPol.notExistValue, :) = [];
<a name="l02954"></a>02954                 times = [times; tmp(:, 2)];
<a name="l02955"></a>02955                 vals = [vals; pc.rnaPolymeraseIndexs(ones(size(tmp, 1), 1))];
<a name="l02956"></a>02956                 startPos = [startPos; tmp(:, 1)];
<a name="l02957"></a>02957                 endPos = [endPos; 
<a name="l02958"></a>02958                     tmp(:, 1) ...
<a name="l02959"></a>02959                     + ch.complexDNAFootprints(pc.rnaPolymeraseIndexs(1)) * (tmp(:, 1) &lt; rnaPol.activelyTranscribingValue) ...                    
<a name="l02960"></a>02960                     + sim.process(&#39;Transcription&#39;).rnaPolymeraseElongationRate * (tmp(:, 1) &gt;= rnaPol.activelyTranscribingValue)
<a name="l02961"></a>02961                     ];
<a name="l02962"></a>02962                 
<a name="l02963"></a>02963                 idxs = find(endPos &gt; ch.sequenceLen);
<a name="l02964"></a>02964                 vals = [vals; vals(idxs)];
<a name="l02965"></a>02965                 times = [times; times(idxs)];
<a name="l02966"></a>02966                 startPos = [startPos; ones(size(idxs))];
<a name="l02967"></a>02967                 endPos = [endPos; endPos(idxs) - ch.sequenceLen];
<a name="l02968"></a>02968                 endPos(idxs) = ch.sequenceLen;
<a name="l02969"></a>02969                 cpxTimeBound = <a class="code" href="class_figures34.html">Figures34</a>.<a class="code" href="class_figures34.html#a1d2e3eeb8019d1d25402d337f2265bae">calcChromosomeSampling</a>(startPos, endPos, times, 1, ch.sequenceLen, ch.sequenceLen);
<a name="l02970"></a>02970                 
<a name="l02971"></a>02971                 %fraction sampled
<a name="l02972"></a>02972                 timeBound = min(monTimeBound, cpxTimeBound);
<a name="l02973"></a>02973                 fractionSampled = cumsum(histc(timeBound, (1:size(states.Time.values, 3))<span class="stringliteral">&#39;)) / ch.sequenceLen;</span>
<a name="l02974"></a>02974 <span class="stringliteral">                </span>
<a name="l02975"></a>02975 <span class="stringliteral">                %fraction sampled - individual proteins</span>
<a name="l02976"></a>02976 <span class="stringliteral">                cpxIdxs = pc.getIndexs({&#39;</span>MG_213_214_298_6MER_ADP<span class="stringliteral">&#39;, &#39;</span>RNA_POLYMERASE<span class="stringliteral">&#39;, &#39;</span>MG_203_204_TETRAMER<span class="stringliteral">&#39;, &#39;</span>MG_094_HEXAMER<span class="stringliteral">&#39;});</span>
<a name="l02977"></a>02977 <span class="stringliteral">                fractionSampled_IndivProtein = zeros(size(states.Time.values, 3), numel(cpxIdxs));</span>
<a name="l02978"></a>02978 <span class="stringliteral">                for i = 1:numel(cpxIdxs)</span>
<a name="l02979"></a>02979 <span class="stringliteral">                    timeBound = Figures34.calcChromosomeSampling(startPos(vals == cpxIdxs(i)), endPos(vals == cpxIdxs(i)), times(vals == cpxIdxs(i)), 1, ch.sequenceLen, ch.sequenceLen);</span>
<a name="l02980"></a>02980 <span class="stringliteral">                    fractionSampled_IndivProtein(:, i) = cumsum(histc(timeBound, (1:size(states.Time.values, 3))&#39;</span>)) / ch.sequenceLen;
<a name="l02981"></a>02981                 end
<a name="l02982"></a>02982                 
<a name="l02983"></a>02983                 %fraction sampled - <a class="code" href="class_r_n_a.html">RNA</a> Pol
<a name="l02984"></a>02984                 nTime = size(states.RNAPolymerase.positionStrands, 3);
<a name="l02985"></a>02985                 nPols = size(states.RNAPolymerase.positionStrands, 1);
<a name="l02986"></a>02986                 rnaPolRate = sim.process(<span class="stringliteral">&#39;Transcription&#39;</span>).rnaPolymeraseElongationRate;
<a name="l02987"></a>02987                 rnaPolFtpt = ch.complexDNAFootprints(pc.rnaPolymeraseIndexs(1));
<a name="l02988"></a>02988                 fractionSampled_RNAPol = zeros(nTime, 3);
<a name="l02989"></a>02989                 
<a name="l02990"></a>02990                 tmpPosTime = [reshape(permute(states.RNAPolymerase.positionStrands(:, 1, :), [1 3 2]), [], 1)  reshape(repmat(1:nTime, nPols, 1), [], 1)];
<a name="l02991"></a>02991                 tmpStates = reshape(permute(states.RNAPolymerase.states, [1 3 2]), [], 1);
<a name="l02992"></a>02992                 
<a name="l02993"></a>02993                 tfs = tmpStates &gt;= rnaPol.activelyTranscribingValue;
<a name="l02994"></a>02994                 timeBound = <a class="code" href="class_figures34.html">Figures34</a>.<a class="code" href="class_figures34.html#a1d2e3eeb8019d1d25402d337f2265bae">calcChromosomeSampling</a>(tmpPosTime(tfs, 1), tmpPosTime(tfs, 1)+rnaPolRate, tmpPosTime(tfs, 2), 1, ch.sequenceLen, ch.sequenceLen);
<a name="l02995"></a>02995                 fractionSampled_RNAPol(:, 1) = cumsum(histc(timeBound, (1:nTime)<span class="stringliteral">&#39;)) / ch.sequenceLen;</span>
<a name="l02996"></a>02996 <span class="stringliteral">                </span>
<a name="l02997"></a>02997 <span class="stringliteral">                tfs = false(ch.sequenceLen, 1);</span>
<a name="l02998"></a>02998 <span class="stringliteral">                for j = 1:numel(rnaPol.transcripts.transcriptionUnitFivePrimeCoordinates)</span>
<a name="l02999"></a>02999 <span class="stringliteral">                    if rnaPol.transcripts.transcriptionUnitDirections(j) == 1</span>
<a name="l03000"></a>03000 <span class="stringliteral">                        tfs(rnaPol.transcripts.transcriptionUnitFivePrimeCoordinates(j)+(1:rnaPol.transcripts.transcriptionUnitLengths(j))-1) = true;</span>
<a name="l03001"></a>03001 <span class="stringliteral">                    else</span>
<a name="l03002"></a>03002 <span class="stringliteral">                        tfs(rnaPol.transcripts.transcriptionUnitFivePrimeCoordinates(j)-(1:rnaPol.transcripts.transcriptionUnitLengths(j))+1) = true;</span>
<a name="l03003"></a>03003 <span class="stringliteral">                    end</span>
<a name="l03004"></a>03004 <span class="stringliteral">                end</span>
<a name="l03005"></a>03005 <span class="stringliteral">                timeBound = timeBound(tfs);</span>
<a name="l03006"></a>03006 <span class="stringliteral">                fractionSampled_RNAPol(:, 3) = cumsum(histc(timeBound, (1:nTime)&#39;</span>)) / numel(timeBound);
<a name="l03007"></a>03007                 
<a name="l03008"></a>03008                 tfs = tmpStates == rnaPol.nonSpecificallyBoundValue | tmpStates == rnaPol.specificallyBoundValue;
<a name="l03009"></a>03009                 timeBound = <a class="code" href="class_figures34.html">Figures34</a>.<a class="code" href="class_figures34.html#a1d2e3eeb8019d1d25402d337f2265bae">calcChromosomeSampling</a>(tmpPosTime(tfs, 1), tmpPosTime(tfs, 1)+rnaPolFtpt, tmpPosTime(tfs, 2), 1, ch.sequenceLen, ch.sequenceLen);
<a name="l03010"></a>03010                 fractionSampled_RNAPol(:, 2) = cumsum(histc(timeBound, (1:nTime)<span class="stringliteral">&#39;)) / ch.sequenceLen;</span>
<a name="l03011"></a>03011 <span class="stringliteral">                </span>
<a name="l03012"></a>03012 <span class="stringliteral">                %fit</span>
<a name="l03013"></a>03013 <span class="stringliteral">                expFittype = fittype(&#39;</span>1 - a * exp(-log(2) * t / tau)<span class="stringliteral">&#39;, ...</span>
<a name="l03014"></a>03014 <span class="stringliteral">                    &#39;</span>independent<span class="stringliteral">&#39;, &#39;</span>t<span class="stringliteral">&#39;, ...</span>
<a name="l03015"></a>03015 <span class="stringliteral">                    &#39;</span>coefficients<span class="stringliteral">&#39;, {&#39;</span>a<span class="stringliteral">&#39;, &#39;</span>tau<span class="stringliteral">&#39;});</span>
<a name="l03016"></a>03016 <span class="stringliteral">                expFitOptions = fitoptions(...</span>
<a name="l03017"></a>03017 <span class="stringliteral">                    &#39;</span>Method<span class="stringliteral">&#39;, &#39;</span>NonlinearLeastSquares<span class="stringliteral">&#39;, ...</span>
<a name="l03018"></a>03018 <span class="stringliteral">                    &#39;</span>Startpoint<span class="stringliteral">&#39;, [.25 2], ...</span>
<a name="l03019"></a>03019 <span class="stringliteral">                    &#39;</span>Lower<span class="stringliteral">&#39;, [0 0], ...</span>
<a name="l03020"></a>03020 <span class="stringliteral">                    &#39;</span>Upper<span class="stringliteral">&#39;, [1 10]);</span>
<a name="l03021"></a>03021 <span class="stringliteral">                </span>
<a name="l03022"></a>03022 <span class="stringliteral">                [fitResult, gof] = fit(time, fractionSampled, expFittype, expFitOptions);</span>
<a name="l03023"></a>03023 <span class="stringliteral">                </span>
<a name="l03024"></a>03024 <span class="stringliteral">                tau = zeros(size(fractionSampled_IndivProtein, 2), 1);</span>
<a name="l03025"></a>03025 <span class="stringliteral">                for i = 1:size(fractionSampled_IndivProtein, 2)</span>
<a name="l03026"></a>03026 <span class="stringliteral">                    tmp = fit(time, fractionSampled_IndivProtein(:, i), expFittype, expFitOptions);</span>
<a name="l03027"></a>03027 <span class="stringliteral">                    tau(i) = tmp.tau;</span>
<a name="l03028"></a>03028 <span class="stringliteral">                end</span>
<a name="l03029"></a>03029 <span class="stringliteral">                </span>
<a name="l03030"></a>03030 <span class="stringliteral">                tau_RNAPol = zeros(size(fractionSampled_RNAPol, 2), 1);</span>
<a name="l03031"></a>03031 <span class="stringliteral">                for i = 1:size(fractionSampled_RNAPol, 2)</span>
<a name="l03032"></a>03032 <span class="stringliteral">                    tmp = fit(time, fractionSampled_RNAPol(:, i), expFittype, expFitOptions);</span>
<a name="l03033"></a>03033 <span class="stringliteral">                    tau_RNAPol(i) = tmp.tau;</span>
<a name="l03034"></a>03034 <span class="stringliteral">                end</span>
<a name="l03035"></a>03035 <span class="stringliteral">                </span>
<a name="l03036"></a>03036 <span class="stringliteral">                save([outDirectory &#39;</span>chromosomeSampling.mat<span class="stringliteral">&#39;], &#39;</span>time<span class="stringliteral">&#39;, &#39;</span>fractionSampled<span class="stringliteral">&#39;, ...</span>
<a name="l03037"></a>03037 <span class="stringliteral">                    &#39;</span>fractionSampled_IndivProtein<span class="stringliteral">&#39;, &#39;</span>fractionSampled_RNAPol<span class="stringliteral">&#39;, ...</span>
<a name="l03038"></a>03038 <span class="stringliteral">                    &#39;</span>fitResult<span class="stringliteral">&#39;, &#39;</span>gof<span class="stringliteral">&#39;, &#39;</span>tau<span class="stringliteral">&#39;, &#39;</span>tau_RNAPol<span class="stringliteral">&#39;);</span>
<a name="l03039"></a>03039 <span class="stringliteral">            end</span>
<a name="l03040"></a>03040 <span class="stringliteral">            </span>
<a name="l03041"></a>03041 <span class="stringliteral">            %% plot</span>
<a name="l03042"></a>03042 <span class="stringliteral">            if nargin &gt;= 5</span>
<a name="l03043"></a>03043 <span class="stringliteral">                axesHandle = subplot(&#39;</span>position<span class="stringliteral">&#39;, position, &#39;</span>parent<span class="stringliteral">&#39;, figHandle);</span>
<a name="l03044"></a>03044 <span class="stringliteral">            else</span>
<a name="l03045"></a>03045 <span class="stringliteral">                axesHandle = edu.stanford.covert.cell.sim.util.PlotUtil.newAxesHandle();</span>
<a name="l03046"></a>03046 <span class="stringliteral">            end</span>
<a name="l03047"></a>03047 <span class="stringliteral">            </span>
<a name="l03048"></a>03048 <span class="stringliteral">            labels = {&#39;</span>All<span class="stringliteral">&#39;, &#39;</span>DNA pol<span class="stringliteral">&#39;, &#39;</span>Gyrase<span class="stringliteral">&#39;, &#39;</span><a class="code" href="class_r_n_a.html">RNA</a> pol<span class="stringliteral">&#39;, &#39;</span>SMC<span class="stringliteral">&#39;};</span>
<a name="l03049"></a>03049 <span class="stringliteral">            colors = [</span>
<a name="l03050"></a>03050 <span class="stringliteral">                0 1 1</span>
<a name="l03051"></a>03051 <span class="stringliteral">                0 1 0</span>
<a name="l03052"></a>03052 <span class="stringliteral">                1 0 0</span>
<a name="l03053"></a>03053 <span class="stringliteral">                0 0 1</span>
<a name="l03054"></a>03054 <span class="stringliteral">                1 0 1</span>
<a name="l03055"></a>03055 <span class="stringliteral">                ];</span>
<a name="l03056"></a>03056 <span class="stringliteral">            data = [fractionSampled  fractionSampled_IndivProtein(:, end:-1:1)];</span>
<a name="l03057"></a>03057 <span class="stringliteral">            tau = [fitResult.tau; tau(end:-1:1, 1)] * 60 * log(2);</span>
<a name="l03058"></a>03058 <span class="stringliteral">            </span>
<a name="l03059"></a>03059 <span class="stringliteral">            show = true(size(labels));</span>
<a name="l03060"></a>03060 <span class="stringliteral">            %show([3 5]) = false;</span>
<a name="l03061"></a>03061 <span class="stringliteral">            labels = labels(show);</span>
<a name="l03062"></a>03062 <span class="stringliteral">            colors = colors(show, :);</span>
<a name="l03063"></a>03063 <span class="stringliteral">            data = data(:, show);</span>
<a name="l03064"></a>03064 <span class="stringliteral">            tau = tau(show, :);</span>
<a name="l03065"></a>03065 <span class="stringliteral">            </span>
<a name="l03066"></a>03066 <span class="stringliteral">            h = plot(axesHandle, time, 100 * data, &#39;</span>LineWidth<span class="stringliteral">&#39;, 1);</span>
<a name="l03067"></a>03067 <span class="stringliteral"></span>
<a name="l03068"></a>03068 <span class="stringliteral">            [~, order] = sort(tau);</span>
<a name="l03069"></a>03069 <span class="stringliteral">            labels = labels(order);</span>
<a name="l03070"></a>03070 <span class="stringliteral">            tau = tau(order);</span>
<a name="l03071"></a>03071 <span class="stringliteral">            h = h(order);</span>
<a name="l03072"></a>03072 <span class="stringliteral">            colors = colors(order, :);</span>
<a name="l03073"></a>03073 <span class="stringliteral">            data = data(:, order);</span>
<a name="l03074"></a>03074 <span class="stringliteral">            </span>
<a name="l03075"></a>03075 <span class="stringliteral">            tableFont = 5;</span>
<a name="l03076"></a>03076 <span class="stringliteral">            rowSep = 12;</span>
<a name="l03077"></a>03077 <span class="stringliteral">            text(8.84, 92, &#39;</span>t_{50} (min)<span class="stringliteral">&#39;, &#39;</span>FontSize<span class="stringliteral">&#39;, tableFont, &#39;</span>HorizontalAlign<span class="stringliteral">&#39;, &#39;</span>right<span class="stringliteral">&#39;, &#39;</span>VerticalAlign<span class="stringliteral">&#39;, &#39;</span>middle<span class="stringliteral">&#39;, &#39;</span>Parent<span class="stringliteral">&#39;, axesHandle);</span>
<a name="l03078"></a>03078 <span class="stringliteral">            text(4.6, 92, &#39;</span><a class="code" href="class_protein.html">Protein</a><span class="stringliteral">&#39;, &#39;</span>FontSize<span class="stringliteral">&#39;, tableFont, &#39;</span>HorizontalAlign<span class="stringliteral">&#39;, &#39;</span>left<span class="stringliteral">&#39;, &#39;</span>VerticalAlign<span class="stringliteral">&#39;, &#39;</span>middle<span class="stringliteral">&#39;, &#39;</span>Parent<span class="stringliteral">&#39;, axesHandle);</span>
<a name="l03079"></a>03079 <span class="stringliteral">            line(4.6 + [0 2.1], 82 + [0 0], &#39;</span>Parent<span class="stringliteral">&#39;, axesHandle, &#39;</span>Color<span class="stringliteral">&#39;, &#39;</span>k<span class="stringliteral">&#39;)</span>
<a name="l03080"></a>03080 <span class="stringliteral">            line([6.9 8.85], 82 + [0 0], &#39;</span>Parent<span class="stringliteral">&#39;, axesHandle, &#39;</span>Color<span class="stringliteral">&#39;, &#39;</span>k<span class="stringliteral">&#39;)</span>
<a name="l03081"></a>03081 <span class="stringliteral">            for i = 1:numel(labels)</span>
<a name="l03082"></a>03082 <span class="stringliteral">                t_50 = round(min(time(data(:, i) &gt;= 0.50)) * 60);</span>
<a name="l03083"></a>03083 <span class="stringliteral">                set(h(i), &#39;</span>Color<span class="stringliteral">&#39;, colors(i, :));</span>
<a name="l03084"></a>03084 <span class="stringliteral">                line(4 + [0 0.3], 87-i*rowSep + [0 0], &#39;</span>Parent<span class="stringliteral">&#39;, axesHandle, &#39;</span>Color<span class="stringliteral">&#39;, colors(i, :))</span>
<a name="l03085"></a>03085 <span class="stringliteral">                text(4.6, 87-i*rowSep, labels{i}, &#39;</span>FontSize<span class="stringliteral">&#39;, tableFont, &#39;</span>HorizontalAlign<span class="stringliteral">&#39;, &#39;</span>left<span class="stringliteral">&#39;, &#39;</span>VerticalAlign<span class="stringliteral">&#39;, &#39;</span>middle<span class="stringliteral">&#39;, &#39;</span>Parent<span class="stringliteral">&#39;, axesHandle);</span>
<a name="l03086"></a>03086 <span class="stringliteral">                if isempty(t_50)</span>
<a name="l03087"></a>03087 <span class="stringliteral">                    t_50 = &#39;</span>N/A<span class="stringliteral">&#39;;</span>
<a name="l03088"></a>03088 <span class="stringliteral">                else</span>
<a name="l03089"></a>03089 <span class="stringliteral">                    t_50 = sprintf(&#39;</span>%d<span class="stringliteral">&#39;, t_50);</span>
<a name="l03090"></a>03090 <span class="stringliteral">                end</span>
<a name="l03091"></a>03091 <span class="stringliteral">                text(8.85, 87-i*rowSep, t_50, &#39;</span>FontSize<span class="stringliteral">&#39;, tableFont, &#39;</span>HorizontalAlign<span class="stringliteral">&#39;, &#39;</span>right<span class="stringliteral">&#39;, &#39;</span>VerticalAlign<span class="stringliteral">&#39;, &#39;</span>middle<span class="stringliteral">&#39;, &#39;</span>Parent<span class="stringliteral">&#39;, axesHandle);</span>
<a name="l03092"></a>03092 <span class="stringliteral">            end</span>
<a name="l03093"></a>03093 <span class="stringliteral">            </span>
<a name="l03094"></a>03094 <span class="stringliteral">            xlim(axesHandle, [0 time(end)]);</span>
<a name="l03095"></a>03095 <span class="stringliteral">            ylim(axesHandle, [0 100]);</span>
<a name="l03096"></a>03096 <span class="stringliteral">            </span>
<a name="l03097"></a>03097 <span class="stringliteral">            set(axesHandle, &#39;</span>FontSize<span class="stringliteral">&#39;, 5, &#39;</span>TickDir<span class="stringliteral">&#39;, &#39;</span>out<span class="stringliteral">&#39;, &#39;</span>XTick<span class="stringliteral">&#39;, 0:2:8, &#39;</span>YTick<span class="stringliteral">&#39;, 0:25:100, &#39;</span>box<span class="stringliteral">&#39;, &#39;</span>off<span class="stringliteral">&#39;);</span>
<a name="l03098"></a>03098 <span class="stringliteral">            </span>
<a name="l03099"></a>03099 <span class="stringliteral">            tick2text(axesHandle, &#39;</span>axis<span class="stringliteral">&#39;, &#39;</span>xy<span class="stringliteral">&#39;, &#39;</span>xtickoffset<span class="stringliteral">&#39;, 0.08, &#39;</span>ytickoffset<span class="stringliteral">&#39;, 0.025);</span>
<a name="l03100"></a>03100 <span class="stringliteral">            xTicks = getappdata(axesHandle, &#39;</span>XTickText<span class="stringliteral">&#39;);</span>
<a name="l03101"></a>03101 <span class="stringliteral">            yTicks = getappdata(axesHandle, &#39;</span>YTickText<span class="stringliteral">&#39;);</span>
<a name="l03102"></a>03102 <span class="stringliteral">            set(xTicks, &#39;</span>FontSize<span class="stringliteral">&#39;, 5);</span>
<a name="l03103"></a>03103 <span class="stringliteral">            set(yTicks, &#39;</span>FontSize<span class="stringliteral">&#39;, 5, &#39;</span>HorizontalAlign<span class="stringliteral">&#39;, &#39;</span>right<span class="stringliteral">&#39;);</span>
<a name="l03104"></a>03104 <span class="stringliteral">            delete(xTicks(2:4));</span>
<a name="l03105"></a>03105 <span class="stringliteral">            delete(yTicks(2:4));</span>
<a name="l03106"></a>03106 <span class="stringliteral">            </span>
<a name="l03107"></a>03107 <span class="stringliteral">            xlabel(axesHandle, &#39;</span>Time (h)<span class="stringliteral">&#39;, &#39;</span>FontSize<span class="stringliteral">&#39;, 7);</span>
<a name="l03108"></a>03108 <span class="stringliteral">            ylabel(axesHandle, {&#39;</span>% <a class="code" href="class_chromosome.html">Chromosome</a><span class="charliteral">&#39; &#39;</span>explored<span class="stringliteral">&#39;}, &#39;</span>FontSize<span class="stringliteral">&#39;, 7);</span>
<a name="l03109"></a>03109 <span class="stringliteral">            xlabelPos = get(get(axesHandle, &#39;</span>xlabel<span class="stringliteral">&#39;), &#39;</span>position<span class="stringliteral">&#39;);</span>
<a name="l03110"></a>03110 <span class="stringliteral">            ylabelPos = get(get(axesHandle, &#39;</span>ylabel<span class="stringliteral">&#39;), &#39;</span>position<span class="stringliteral">&#39;);</span>
<a name="l03111"></a>03111 <span class="stringliteral">            set(get(axesHandle, &#39;</span>xlabel<span class="stringliteral">&#39;), &#39;</span>position<span class="stringliteral">&#39;, [xlabelPos(1) -11 xlabelPos(3:end)]);</span>
<a name="l03112"></a>03112 <span class="stringliteral">            set(get(axesHandle, &#39;</span>ylabel<span class="stringliteral">&#39;), &#39;</span>position<span class="stringliteral">&#39;, [-0.55 ylabelPos(2:end)]);</span>
<a name="l03113"></a>03113 <span class="stringliteral">        end</span>
<a name="l03114"></a>03114 <span class="stringliteral">        </span>
<a name="l03115"></a>03115 <span class="stringliteral">        function plotRNAPolSampling(simBatchDir, selectedSim, outDirectory, figHandle, position)</span>
<a name="l03116"></a>03116 <span class="stringliteral">            import edu.stanford.covert.cell.sim.util.PlotUtil;</span>
<a name="l03117"></a>03117 <span class="stringliteral">            import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;</span>
<a name="l03118"></a>03118 <span class="stringliteral">            </span>
<a name="l03119"></a>03119 <span class="stringliteral">            %% lood constants</span>
<a name="l03120"></a>03120 <span class="stringliteral">            [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep num2str(selectedSim)]);</span>
<a name="l03121"></a>03121 <span class="stringliteral">            ch = sim.state(&#39;</span><a class="code" href="class_chromosome.html">Chromosome</a><span class="stringliteral">&#39;);</span>
<a name="l03122"></a>03122 <span class="stringliteral">            pc = sim.state(&#39;</span><a class="code" href="class_protein_complex.html">ProteinComplex</a><span class="stringliteral">&#39;);</span>
<a name="l03123"></a>03123 <span class="stringliteral">            rnaPol = sim.state(&#39;</span>RNAPolymerase<span class="stringliteral">&#39;);</span>
<a name="l03124"></a>03124 <span class="stringliteral">            </span>
<a name="l03125"></a>03125 <span class="stringliteral">            %% load data</span>
<a name="l03126"></a>03126 <span class="stringliteral">            load([outDirectory &#39;</span>chromosomeSampling.mat<span class="stringliteral">&#39;]);</span>
<a name="l03127"></a>03127 <span class="stringliteral">            load([outDirectory &#39;</span>rnaSampling.mat<span class="stringliteral">&#39;]);</span>
<a name="l03128"></a>03128 <span class="stringliteral">            </span>
<a name="l03129"></a>03129 <span class="stringliteral">            %% plot</span>
<a name="l03130"></a>03130 <span class="stringliteral">            if nargin &gt;= 5</span>
<a name="l03131"></a>03131 <span class="stringliteral">                axesHandle = subplot(&#39;</span>position<span class="stringliteral">&#39;, position, &#39;</span>parent<span class="stringliteral">&#39;, figHandle);</span>
<a name="l03132"></a>03132 <span class="stringliteral">            else</span>
<a name="l03133"></a>03133 <span class="stringliteral">                axesHandle = edu.stanford.covert.cell.sim.util.PlotUtil.newAxesHandle();</span>
<a name="l03134"></a>03134 <span class="stringliteral">            end</span>
<a name="l03135"></a>03135 <span class="stringliteral">            </span>
<a name="l03136"></a>03136 <span class="stringliteral">            h = plot(axesHandle, time, 100 * [fractionSampled_IndivProtein(:, 2) fractionSampled_RNAPol(:, [1 3 2]) fracRNASampled]); %#ok&lt;NODEF&gt;</span>
<a name="l03137"></a>03137 <span class="stringliteral">            legend(h, {</span>
<a name="l03138"></a>03138 <span class="stringliteral">                sprintf(&#39;</span>All <a class="code" href="class_r_n_a.html">RNA</a> Polymerase (\\tau = %0.1f h)<span class="stringliteral">&#39;, tau(2))</span>
<a name="l03139"></a>03139 <span class="stringliteral">                sprintf(&#39;</span>Active/SB - of whole <a class="code" href="class_chromosome.html">Chromosome</a> (\\tau = %0.1f h)<span class="stringliteral">&#39;, tau_RNAPol(1))</span>
<a name="l03140"></a>03140 <span class="stringliteral">                sprintf(&#39;</span>Active/SB - of TUs (\\tau = %0.1f h)<span class="stringliteral">&#39;, tau_RNAPol(3))</span>
<a name="l03141"></a>03141 <span class="stringliteral">                sprintf(&#39;</span>Non-SB (\\tau = %0.1f h)<span class="stringliteral">&#39;, tau_RNAPol(2))</span>
<a name="l03142"></a>03142 <span class="stringliteral">                sprintf(&#39;</span>Expressed <a class="code" href="class_r_n_a.html">RNA</a> (\\tau = %0.1f h)<span class="stringliteral">&#39;, rnaFitResult.tau)</span>
<a name="l03143"></a>03143 <span class="stringliteral">                }, ...</span>
<a name="l03144"></a>03144 <span class="stringliteral">                &#39;</span>Location<span class="stringliteral">&#39;, &#39;</span>SouthEast<span class="stringliteral">&#39;, &#39;</span>FontSize<span class="stringliteral">&#39;, 8);</span>
<a name="l03145"></a>03145 <span class="stringliteral">            xlabel(axesHandle, &#39;</span>Time (h)<span class="stringliteral">&#39;)</span>
<a name="l03146"></a>03146 <span class="stringliteral">            ylabel(axesHandle, &#39;</span>% Explored / expressed<span class="stringliteral">&#39;)</span>
<a name="l03147"></a>03147 <span class="stringliteral">            xlim(axesHandle, [0 max(time)])</span>
<a name="l03148"></a>03148 <span class="stringliteral">            ylim(axesHandle, [0 100])</span>
<a name="l03149"></a>03149 <span class="stringliteral">        end</span>
<a name="l03150"></a>03150 <span class="stringliteral">        </span>
<a name="l03151"></a>03151 <span class="stringliteral">        function timeBound = calcChromosomeSampling(startPos, endPos, times, pos1, pos2, L)</span>
<a name="l03152"></a>03152 <span class="stringliteral">            import edu.stanford.covert.cell.sim.analysis.Figures34;</span>
<a name="l03153"></a>03153 <span class="stringliteral">            </span>
<a name="l03154"></a>03154 <span class="stringliteral">            if pos2-pos1 &gt; 500</span>
<a name="l03155"></a>03155 <span class="stringliteral">                pos21 = floor((pos1+pos2)/2);</span>
<a name="l03156"></a>03156 <span class="stringliteral">                pos12 = pos21+1;</span>
<a name="l03157"></a>03157 <span class="stringliteral">                tfs1 = startPos &lt;= pos21;</span>
<a name="l03158"></a>03158 <span class="stringliteral">                tfs2 = endPos &gt;= pos12;</span>
<a name="l03159"></a>03159 <span class="stringliteral">                timeBound = min(...</span>
<a name="l03160"></a>03160 <span class="stringliteral">                    Figures34.calcChromosomeSampling(startPos(tfs1), endPos(tfs1), times(tfs1), pos1, pos21, L), ...</span>
<a name="l03161"></a>03161 <span class="stringliteral">                    Figures34.calcChromosomeSampling(startPos(tfs2), endPos(tfs2), times(tfs2), pos12, pos2, L));</span>
<a name="l03162"></a>03162 <span class="stringliteral">            else</span>
<a name="l03163"></a>03163 <span class="stringliteral">                timeBound = NaN(L, 1);</span>
<a name="l03164"></a>03164 <span class="stringliteral">                for i = pos1:pos2</span>
<a name="l03165"></a>03165 <span class="stringliteral">                    tf = startPos &lt;= i &amp; endPos &gt;= i;</span>
<a name="l03166"></a>03166 <span class="stringliteral">                    if ~any(tf)</span>
<a name="l03167"></a>03167 <span class="stringliteral">                        continue;</span>
<a name="l03168"></a>03168 <span class="stringliteral">                    end</span>
<a name="l03169"></a>03169 <span class="stringliteral">                    timeBound(i) = min(times(tf));</span>
<a name="l03170"></a>03170 <span class="stringliteral">                end</span>
<a name="l03171"></a>03171 <span class="stringliteral">            end</span>
<a name="l03172"></a>03172 <span class="stringliteral">        end</span>
<a name="l03173"></a>03173 <span class="stringliteral">        </span>
<a name="l03174"></a>03174 <span class="stringliteral">        function rnaSampling(simBatchDir, selectedSim, outDirectory, figHandle, position)</span>
<a name="l03175"></a>03175 <span class="stringliteral">            import edu.stanford.covert.cell.sim.util.PlotUtil;</span>
<a name="l03176"></a>03176 <span class="stringliteral">            import edu.stanford.covert.cell.sim.util.SimulationEnsemble;</span>
<a name="l03177"></a>03177 <span class="stringliteral">            import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;</span>
<a name="l03178"></a>03178 <span class="stringliteral">            </span>
<a name="l03179"></a>03179 <span class="stringliteral">            if ~exist(&#39;</span>simBatchDir<span class="stringliteral">&#39;, &#39;</span>var<span class="stringliteral">&#39;) || isempty(simBatchDir)</span>
<a name="l03180"></a>03180 <span class="stringliteral">                simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();</span>
<a name="l03181"></a>03181 <span class="stringliteral">            end</span>
<a name="l03182"></a>03182 <span class="stringliteral">            if ~exist(&#39;</span>selectedSim<span class="stringliteral">&#39;, &#39;</span>var<span class="stringliteral">&#39;) || isempty(selectedSim)</span>
<a name="l03183"></a>03183 <span class="stringliteral">                selectedSim = 1;</span>
<a name="l03184"></a>03184 <span class="stringliteral">            end</span>
<a name="l03185"></a>03185 <span class="stringliteral">                        </span>
<a name="l03186"></a>03186 <span class="stringliteral">            %% data</span>
<a name="l03187"></a>03187 <span class="stringliteral">            if exist([outDirectory &#39;</span>rnaSampling.mat<span class="stringliteral">&#39;], &#39;</span>file<span class="stringliteral">&#39;)</span>
<a name="l03188"></a>03188 <span class="stringliteral">                load([outDirectory &#39;</span>rnaSampling.mat<span class="stringliteral">&#39;]);</span>
<a name="l03189"></a>03189 <span class="stringliteral">            else</span>
<a name="l03190"></a>03190 <span class="stringliteral">                [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep num2str(selectedSim)]);</span>
<a name="l03191"></a>03191 <span class="stringliteral">                comp = sim.compartment;</span>
<a name="l03192"></a>03192 <span class="stringliteral">                rna = sim.state(&#39;</span>Rna<span class="stringliteral">&#39;);</span>
<a name="l03193"></a>03193 <span class="stringliteral">                </span>
<a name="l03194"></a>03194 <span class="stringliteral">                stateNames = {</span>
<a name="l03195"></a>03195 <span class="stringliteral">                    &#39;</span>Time<span class="stringliteral">&#39;       &#39;</span>values<span class="stringliteral">&#39;        &#39;</span>:<span class="stringliteral">&#39;  &#39;</span>:<span class="stringliteral">&#39;</span>
<a name="l03196"></a>03196 <span class="stringliteral">                    &#39;</span>Rna<span class="stringliteral">&#39;        &#39;</span>counts<span class="stringliteral">&#39;        rna.matureIndexs  comp.cytosolIndexs</span>
<a name="l03197"></a>03197 <span class="stringliteral">                    };</span>
<a name="l03198"></a>03198 <span class="stringliteral">                states = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, &#39;</span>extract<span class="stringliteral">&#39;, selectedSim);</span>
<a name="l03199"></a>03199 <span class="stringliteral">                time = permute(states.Time.values, [3 1 2]) / 3600;</span>
<a name="l03200"></a>03200 <span class="stringliteral">                </span>
<a name="l03201"></a>03201 <span class="stringliteral">                [i, j] = find(permute(states.Rna.counts, [1 3 2]));</span>
<a name="l03202"></a>03202 <span class="stringliteral">                timeFirstExpressed = NaN(size(rna.matureIndexs));</span>
<a name="l03203"></a>03203 <span class="stringliteral">                timeFirstExpressed(i(end:-1:1)) = j(end:-1:1);</span>
<a name="l03204"></a>03204 <span class="stringliteral">                fracRNASampled = cumsum(histc(timeFirstExpressed, (1:numel(time)))) / numel(rna.matureIndexs);</span>
<a name="l03205"></a>03205 <span class="stringliteral">                </span>
<a name="l03206"></a>03206 <span class="stringliteral">                expFittype = fittype(&#39;</span>1 - a * exp(-log(2) * t / tau)<span class="stringliteral">&#39;, ...</span>
<a name="l03207"></a>03207 <span class="stringliteral">                    &#39;</span>independent<span class="stringliteral">&#39;, &#39;</span>t<span class="stringliteral">&#39;, ...</span>
<a name="l03208"></a>03208 <span class="stringliteral">                    &#39;</span>coefficients<span class="stringliteral">&#39;, {&#39;</span>a<span class="stringliteral">&#39;, &#39;</span>tau<span class="stringliteral">&#39;});</span>
<a name="l03209"></a>03209 <span class="stringliteral">                expFitOptions = fitoptions(...</span>
<a name="l03210"></a>03210 <span class="stringliteral">                    &#39;</span>Method<span class="stringliteral">&#39;, &#39;</span>NonlinearLeastSquares<span class="stringliteral">&#39;, ...</span>
<a name="l03211"></a>03211 <span class="stringliteral">                    &#39;</span>Startpoint<span class="stringliteral">&#39;, [.25 2], ...</span>
<a name="l03212"></a>03212 <span class="stringliteral">                    &#39;</span>Lower<span class="stringliteral">&#39;, [0 0], ...</span>
<a name="l03213"></a>03213 <span class="stringliteral">                    &#39;</span>Upper<span class="stringliteral">&#39;, [1 10]);</span>
<a name="l03214"></a>03214 <span class="stringliteral">                [rnaFitResult, rnaGof] = fit(time, fracRNASampled, expFittype, expFitOptions);</span>
<a name="l03215"></a>03215 <span class="stringliteral">                </span>
<a name="l03216"></a>03216 <span class="stringliteral">                save([outDirectory &#39;</span>rnaSampling.mat<span class="stringliteral">&#39;], &#39;</span>time<span class="stringliteral">&#39;, &#39;</span>fracRNASampled<span class="stringliteral">&#39;, &#39;</span>rnaFitResult<span class="stringliteral">&#39;, &#39;</span>rnaGof<span class="stringliteral">&#39;)</span>
<a name="l03217"></a>03217 <span class="stringliteral">            end</span>
<a name="l03218"></a>03218 <span class="stringliteral">            </span>
<a name="l03219"></a>03219 <span class="stringliteral">            %% plot</span>
<a name="l03220"></a>03220 <span class="stringliteral">            if nargin &gt;= 5</span>
<a name="l03221"></a>03221 <span class="stringliteral">                axesHandle = subplot(&#39;</span>position<span class="stringliteral">&#39;, position, &#39;</span>parent<span class="stringliteral">&#39;, figHandle);</span>
<a name="l03222"></a>03222 <span class="stringliteral">            else</span>
<a name="l03223"></a>03223 <span class="stringliteral">                axesHandle = edu.stanford.covert.cell.sim.util.PlotUtil.newAxesHandle();</span>
<a name="l03224"></a>03224 <span class="stringliteral">            end</span>
<a name="l03225"></a>03225 <span class="stringliteral">            </span>
<a name="l03226"></a>03226 <span class="stringliteral">            plot(axesHandle, time, 100 * fracRNASampled, &#39;</span>Color<span class="stringliteral">&#39;, &#39;</span>b<span class="stringliteral">&#39;, &#39;</span>LineWidth<span class="stringliteral">&#39;, 1);</span>
<a name="l03227"></a>03227 <span class="stringliteral">            </span>
<a name="l03228"></a>03228 <span class="stringliteral">            xlim(axesHandle, [0 time(end)]);</span>
<a name="l03229"></a>03229 <span class="stringliteral">            ylim(axesHandle, [0 100]);</span>
<a name="l03230"></a>03230 <span class="stringliteral">                        </span>
<a name="l03231"></a>03231 <span class="stringliteral">            set(axesHandle, &#39;</span>FontSize<span class="stringliteral">&#39;, 5, &#39;</span>TickDir<span class="stringliteral">&#39;, &#39;</span>out<span class="stringliteral">&#39;, &#39;</span>XTick<span class="stringliteral">&#39;, 0:2:8, &#39;</span>YTick<span class="stringliteral">&#39;, 0:25:100, &#39;</span>box<span class="stringliteral">&#39;, &#39;</span>off<span class="stringliteral">&#39;);</span>
<a name="l03232"></a>03232 <span class="stringliteral">            </span>
<a name="l03233"></a>03233 <span class="stringliteral">            t_50 = min(time(fracRNASampled &gt;= 0.50)) * 60;</span>
<a name="l03234"></a>03234 <span class="stringliteral">            text(8.84, 98, sprintf(&#39;</span>t_{50} = %.0f min<span class="stringliteral">&#39;, t_50), ...</span>
<a name="l03235"></a>03235 <span class="stringliteral">                &#39;</span>Parent<span class="stringliteral">&#39;, axesHandle, ...</span>
<a name="l03236"></a>03236 <span class="stringliteral">                &#39;</span>FontSize<span class="stringliteral">&#39;, 5, ...</span>
<a name="l03237"></a>03237 <span class="stringliteral">                &#39;</span>horizontalalign<span class="stringliteral">&#39;, &#39;</span>right<span class="stringliteral">&#39;, ...</span>
<a name="l03238"></a>03238 <span class="stringliteral">                &#39;</span>verticalalign<span class="stringliteral">&#39;, &#39;</span>top<span class="stringliteral">&#39;);</span>
<a name="l03239"></a>03239 <span class="stringliteral">            </span>
<a name="l03240"></a>03240 <span class="stringliteral">            tick2text(axesHandle, &#39;</span>axis<span class="stringliteral">&#39;, &#39;</span>xy<span class="stringliteral">&#39;, &#39;</span>xtickoffset<span class="stringliteral">&#39;, 0.08, &#39;</span>ytickoffset<span class="stringliteral">&#39;, 0.025);</span>
<a name="l03241"></a>03241 <span class="stringliteral">            xTicks = getappdata(axesHandle, &#39;</span>XTickText<span class="stringliteral">&#39;);</span>
<a name="l03242"></a>03242 <span class="stringliteral">            yTicks = getappdata(axesHandle, &#39;</span>YTickText<span class="stringliteral">&#39;);</span>
<a name="l03243"></a>03243 <span class="stringliteral">            set(xTicks, &#39;</span>FontSize<span class="stringliteral">&#39;, 5);</span>
<a name="l03244"></a>03244 <span class="stringliteral">            set(yTicks, &#39;</span>FontSize<span class="stringliteral">&#39;, 5, &#39;</span>HorizontalAlign<span class="stringliteral">&#39;, &#39;</span>right<span class="stringliteral">&#39;);</span>
<a name="l03245"></a>03245 <span class="stringliteral">            delete(xTicks(2:4));</span>
<a name="l03246"></a>03246 <span class="stringliteral">            delete(yTicks(2:4));</span>
<a name="l03247"></a>03247 <span class="stringliteral">            </span>
<a name="l03248"></a>03248 <span class="stringliteral">            xlabel(axesHandle, &#39;</span>Time (h)<span class="stringliteral">&#39;, &#39;</span>FontSize<span class="stringliteral">&#39;, 7);</span>
<a name="l03249"></a>03249 <span class="stringliteral">            ylabel(axesHandle, {&#39;</span>% <a class="code" href="class_r_n_a.html">RNA</a><span class="charliteral">&#39; &#39;</span>expressed<span class="stringliteral">&#39;}, &#39;</span>FontSize<span class="stringliteral">&#39;, 7);</span>
<a name="l03250"></a>03250 <span class="stringliteral">            xlabelPos = get(get(axesHandle, &#39;</span>xlabel<span class="stringliteral">&#39;), &#39;</span>position<span class="stringliteral">&#39;);</span>
<a name="l03251"></a>03251 <span class="stringliteral">            ylabelPos = get(get(axesHandle, &#39;</span>ylabel<span class="stringliteral">&#39;), &#39;</span>position<span class="stringliteral">&#39;);</span>
<a name="l03252"></a>03252 <span class="stringliteral">            set(get(axesHandle, &#39;</span>xlabel<span class="stringliteral">&#39;), &#39;</span>position<span class="stringliteral">&#39;, [xlabelPos(1) -11 xlabelPos(3:end)]);</span>
<a name="l03253"></a>03253 <span class="stringliteral">            set(get(axesHandle, &#39;</span>ylabel<span class="stringliteral">&#39;), &#39;</span>position<span class="stringliteral">&#39;, [-0.55 ylabelPos(2:end)]);</span>
<a name="l03254"></a>03254 <span class="stringliteral">        end</span>
<a name="l03255"></a>03255 <span class="stringliteral">        </span>
<a name="l03256"></a>03256 <span class="stringliteral">        function mrnaSampling(simBatchDir, selectedSim, outDirectory, figHandle, position)</span>
<a name="l03257"></a>03257 <span class="stringliteral">            import edu.stanford.covert.cell.sim.util.PlotUtil;</span>
<a name="l03258"></a>03258 <span class="stringliteral">            import edu.stanford.covert.cell.sim.util.SimulationEnsemble;</span>
<a name="l03259"></a>03259 <span class="stringliteral">            import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;</span>
<a name="l03260"></a>03260 <span class="stringliteral">            </span>
<a name="l03261"></a>03261 <span class="stringliteral">            if ~exist(&#39;</span>simBatchDir<span class="stringliteral">&#39;, &#39;</span>var<span class="stringliteral">&#39;) || isempty(simBatchDir)</span>
<a name="l03262"></a>03262 <span class="stringliteral">                simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();</span>
<a name="l03263"></a>03263 <span class="stringliteral">            end</span>
<a name="l03264"></a>03264 <span class="stringliteral">            if ~exist(&#39;</span>selectedSim<span class="stringliteral">&#39;, &#39;</span>var<span class="stringliteral">&#39;) || isempty(selectedSim)</span>
<a name="l03265"></a>03265 <span class="stringliteral">                selectedSim = 1;</span>
<a name="l03266"></a>03266 <span class="stringliteral">            end</span>
<a name="l03267"></a>03267 <span class="stringliteral">            </span>
<a name="l03268"></a>03268 <span class="stringliteral">            %% data</span>
<a name="l03269"></a>03269 <span class="stringliteral">            if exist([outDirectory &#39;</span>mrnaSampling.mat<span class="stringliteral">&#39;], &#39;</span>file<span class="stringliteral">&#39;)</span>
<a name="l03270"></a>03270 <span class="stringliteral">                load([outDirectory &#39;</span>mrnaSampling.mat<span class="stringliteral">&#39;]);</span>
<a name="l03271"></a>03271 <span class="stringliteral">            else</span>
<a name="l03272"></a>03272 <span class="stringliteral">                [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep num2str(selectedSim)]);</span>
<a name="l03273"></a>03273 <span class="stringliteral">                comp = sim.compartment;</span>
<a name="l03274"></a>03274 <span class="stringliteral">                rna = sim.state(&#39;</span>Rna<span class="stringliteral">&#39;);</span>
<a name="l03275"></a>03275 <span class="stringliteral">                </span>
<a name="l03276"></a>03276 <span class="stringliteral">                stateNames = {</span>
<a name="l03277"></a>03277 <span class="stringliteral">                    &#39;</span>Time<span class="stringliteral">&#39;       &#39;</span>values<span class="stringliteral">&#39;        &#39;</span>:<span class="stringliteral">&#39;  &#39;</span>:<span class="stringliteral">&#39;</span>
<a name="l03278"></a>03278 <span class="stringliteral">                    &#39;</span>Rna<span class="stringliteral">&#39;        &#39;</span>counts<span class="stringliteral">&#39;        rna.matureIndexs  comp.cytosolIndexs</span>
<a name="l03279"></a>03279 <span class="stringliteral">                    };</span>
<a name="l03280"></a>03280 <span class="stringliteral">                states = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, &#39;</span>extract<span class="stringliteral">&#39;, selectedSim);</span>
<a name="l03281"></a>03281 <span class="stringliteral">                time = permute(states.Time.values, [3 1 2]) / 3600;</span>
<a name="l03282"></a>03282 <span class="stringliteral">                </span>
<a name="l03283"></a>03283 <span class="stringliteral">                [i, j] = find(permute(states.Rna.counts, [1 3 2]));</span>
<a name="l03284"></a>03284 <span class="stringliteral">                timeFirstExpressed = NaN(size(rna.matureIndexs));</span>
<a name="l03285"></a>03285 <span class="stringliteral">                timeFirstExpressed(i(end:-1:1)) = j(end:-1:1);</span>
<a name="l03286"></a>03286 <span class="stringliteral">                fracMRNASampled = cumsum(histc(timeFirstExpressed(rna.matureMRNAIndexs), (1:numel(time)))) / numel(rna.matureMRNAIndexs);</span>
<a name="l03287"></a>03287 <span class="stringliteral">                </span>
<a name="l03288"></a>03288 <span class="stringliteral">                expFittype = fittype(&#39;</span>1 - a * exp(-log(2) * t / tau)&#39;, ...
<a name="l03289"></a>03289                     &#39;independent&#39;, &#39;t&#39;, ...
<a name="l03290"></a>03290                     &#39;coefficients&#39;, {<span class="charliteral">&#39;a&#39;</span>, <span class="stringliteral">&#39;tau&#39;</span>});
<a name="l03291"></a>03291                 expFitOptions = fitoptions(...
<a name="l03292"></a>03292                     <span class="stringliteral">&#39;Method&#39;</span>, <span class="stringliteral">&#39;NonlinearLeastSquares&#39;</span>, ...
<a name="l03293"></a>03293                     <span class="stringliteral">&#39;Startpoint&#39;</span>, [.25 2], ...
<a name="l03294"></a>03294                     <span class="stringliteral">&#39;Lower&#39;</span>, [0 0], ...
<a name="l03295"></a>03295                     <span class="stringliteral">&#39;Upper&#39;</span>, [1 10]);
<a name="l03296"></a>03296                 [mrnaFitResult, mrnaGof] = fit(time, fracMRNASampled, expFittype, expFitOptions);
<a name="l03297"></a>03297                 
<a name="l03298"></a>03298                 save([outDirectory <span class="stringliteral">&#39;mrnaSampling.mat&#39;</span>], <span class="stringliteral">&#39;time&#39;</span>, <span class="stringliteral">&#39;fracMRNASampled&#39;</span>, <span class="stringliteral">&#39;mrnaFitResult&#39;</span>, <span class="stringliteral">&#39;mrnaGof&#39;</span>)
<a name="l03299"></a>03299             end
<a name="l03300"></a>03300             
<a name="l03301"></a>03301             %% plot <a class="code" href="classm_r_n_a.html">mRNA</a>
<a name="l03302"></a>03302             if nargin &gt;= 5
<a name="l03303"></a>03303                 axesHandle = subplot(&#39;position&#39;, position, &#39;parent&#39;, figHandle);
<a name="l03304"></a>03304             else
<a name="l03305"></a>03305                 axesHandle = edu.stanford.covert.cell.sim.util.PlotUtil.newAxesHandle();
<a name="l03306"></a>03306             end
<a name="l03307"></a>03307             
<a name="l03308"></a>03308             h = plot(axesHandle, time, 100 * [fracMRNASampled   1-mrnaFitResult.a*exp(-log(2)*time/mrnaFitResult.tau)]);
<a name="l03309"></a>03309             set(h(1), &#39;Color&#39;, &#39;b&#39;);
<a name="l03310"></a>03310             set(h(2), &#39;Color&#39;, &#39;k&#39;);
<a name="l03311"></a>03311             uistack(h(1), &#39;top&#39;);
<a name="l03312"></a>03312             
<a name="l03313"></a>03313             xlim(axesHandle, time([1 end]));
<a name="l03314"></a>03314             ylim(axesHandle, [0 100]);
<a name="l03315"></a>03315             
<a name="l03316"></a>03316             xlabel(axesHandle, &#39;Time (h)&#39;, &#39;FontSize&#39;, 7);
<a name="l03317"></a>03317             ylabel(axesHandle, &#39;% <a class="code" href="classm_r_n_a.html">mRNA</a> expressed&#39;, &#39;FontSize&#39;, 7);
<a name="l03318"></a>03318             set(axesHandle, &#39;FontSize&#39;, 5, &#39;TickDir&#39;, &#39;out&#39;, &#39;XTick&#39;, [0 5], &#39;YTick&#39;, [0 50 100], &#39;box&#39;, &#39;off&#39;);
<a name="l03319"></a>03319             
<a name="l03320"></a>03320             text(time(end), 98.5, sprintf(&#39;\\gamma = %.0f min&#39;, mrnaFitResult.tau * 60 * log(2)), ...
<a name="l03321"></a>03321                 &#39;Parent&#39;, axesHandle, ...
<a name="l03322"></a>03322                 &#39;FontSize&#39;, 7, ...
<a name="l03323"></a>03323                 &#39;horizontalalign&#39;, &#39;right&#39;, ...
<a name="l03324"></a>03324                 &#39;verticalalign&#39;, &#39;top&#39;);
<a name="l03325"></a>03325             text(time(end), 93.5, sprintf(&#39;R^2_{adj} = %.2f<span class="stringliteral">&#39;, mrnaGof.adjrsquare), ...</span>
<a name="l03326"></a>03326 <span class="stringliteral">                &#39;</span>Parent<span class="stringliteral">&#39;, axesHandle, ...</span>
<a name="l03327"></a>03327 <span class="stringliteral">                &#39;</span>FontSize<span class="stringliteral">&#39;, 7, ...</span>
<a name="l03328"></a>03328 <span class="stringliteral">                &#39;</span>horizontalalign<span class="stringliteral">&#39;, &#39;</span>right<span class="stringliteral">&#39;, ...</span>
<a name="l03329"></a>03329 <span class="stringliteral">                &#39;</span>verticalalign<span class="stringliteral">&#39;, &#39;</span>top<span class="stringliteral">&#39;);</span>
<a name="l03330"></a>03330 <span class="stringliteral">        end</span>
<a name="l03331"></a>03331 <span class="stringliteral">    end</span>
<a name="l03332"></a>03332 <span class="stringliteral">    </span>
<a name="l03333"></a>03333 <span class="stringliteral">    methods (Static = true);</span>
<a name="l03334"></a>03334 <span class="stringliteral">        function rgbStr = colorFunc(normVal)</span>
<a name="l03335"></a>03335 <span class="stringliteral">            r = 0; g = 0; b = 0;</span>
<a name="l03336"></a>03336 <span class="stringliteral">            </span>
<a name="l03337"></a>03337 <span class="stringliteral">            if normVal &lt;= 0.5</span>
<a name="l03338"></a>03338 <span class="stringliteral">                b = 255 * (1 - 2 * normVal);</span>
<a name="l03339"></a>03339 <span class="stringliteral">                g = 255 * (2 * normVal);</span>
<a name="l03340"></a>03340 <span class="stringliteral">            else</span>
<a name="l03341"></a>03341 <span class="stringliteral">                r = 255 * (2 * normVal - 1);</span>
<a name="l03342"></a>03342 <span class="stringliteral">                g = 255 * (2 - 2 * normVal);</span>
<a name="l03343"></a>03343 <span class="stringliteral">            end</span>
<a name="l03344"></a>03344 <span class="stringliteral">            </span>
<a name="l03345"></a>03345 <span class="stringliteral">            rgbStr = sprintf(&#39;</span>rgb(%d,%d,%d)<span class="stringliteral">&#39;, uint8(r), uint8(g), uint8(b));</span>
<a name="l03346"></a>03346 <span class="stringliteral">        end</span>
<a name="l03347"></a>03347 <span class="stringliteral">        </span>
<a name="l03348"></a>03348 <span class="stringliteral">        function svg = drawCell(scale, width, cylindricalLength, septumLength, maxTotalLength, maxWidth) %#ok&lt;INUSD&gt;</span>
<a name="l03349"></a>03349 <span class="stringliteral">            width = width / scale;</span>
<a name="l03350"></a>03350 <span class="stringliteral">            cylindricalLength = cylindricalLength / scale;</span>
<a name="l03351"></a>03351 <span class="stringliteral">            septumLength = septumLength / scale;</span>
<a name="l03352"></a>03352 <span class="stringliteral">            H = maxTotalLength / scale;</span>
<a name="l03353"></a>03353 <span class="stringliteral">            W = width;</span>
<a name="l03354"></a>03354 <span class="stringliteral">            </span>
<a name="l03355"></a>03355 <span class="stringliteral">            svg = [...</span>
<a name="l03356"></a>03356 <span class="stringliteral">                sprintf(&#39;</span>  &lt;path d=<span class="stringliteral">&quot;\n&#39;) ...</span>
<a name="l03357"></a>03357 <span class="stringliteral">                sprintf(&#39;    M%0.4f,%0.4f a%0.4f,%0.4f 0 1 0 0,%0.4f\n&#39;, W/2+5, 5, width/2, width/2, width), ...</span>
<a name="l03358"></a>03358 <span class="stringliteral">                sprintf(&#39;    l%0.4f,0\n&#39;, cylindricalLength/2) ...</span>
<a name="l03359"></a>03359 <span class="stringliteral">                sprintf(&#39;    a%0.4f,%0.4f 0 0 0 %0.4f,-%0.4f\n&#39;, septumLength, septumLength, septumLength, septumLength) ...</span>
<a name="l03360"></a>03360 <span class="stringliteral">                sprintf(&#39;    a%0.4f,%0.4f 0 0 0 %0.4f,%0.4f\n&#39;, septumLength, septumLength, septumLength, septumLength) ...</span>
<a name="l03361"></a>03361 <span class="stringliteral">                sprintf(&#39;    l%0.4f,0\n&#39;, cylindricalLength/2) ...</span>
<a name="l03362"></a>03362 <span class="stringliteral">                sprintf(&#39;    a%0.4f,%0.4f 0 1 0 0,-%0.4f\n&#39;, width/2,width/2, width) ...</span>
<a name="l03363"></a>03363 <span class="stringliteral">                sprintf(&#39;    l-%0.4f,0\n&#39;, cylindricalLength/2) ...</span>
<a name="l03364"></a>03364 <span class="stringliteral">                sprintf(&#39;    a%0.4f,%0.4f 0 0 0 -%0.4f,%0.4f\n&#39;, septumLength, septumLength, septumLength, septumLength) ...</span>
<a name="l03365"></a>03365 <span class="stringliteral">                sprintf(&#39;    a%0.4f,%0.4f 0 0 0 -%0.4f,-%0.4f\n&#39;, septumLength, septumLength, septumLength, septumLength) ...</span>
<a name="l03366"></a>03366 <span class="stringliteral">                sprintf(&#39;    z\n&#39;) ...</span>
<a name="l03367"></a>03367 <span class="stringliteral">                &#39;    &quot;</span> style=<span class="stringliteral">&quot;stroke:darkblue;stroke-width:2;stroke-linecap:round;stroke-linejoin:miter;fill:url(#cellBg)&quot;</span>/&gt;\n<span class="stringliteral">&#39;];</span>
<a name="l03368"></a>03368 <span class="stringliteral">        end</span>
<a name="l03369"></a>03369 <span class="stringliteral">    end</span>
<a name="l03370"></a>03370 <span class="stringliteral">end</span>
</pre></div></div><!-- contents -->
</div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
<a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(0)"><span class="SelectionMark">&#160;</span>All</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(1)"><span class="SelectionMark">&#160;</span>Classes</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(2)"><span class="SelectionMark">&#160;</span>Files</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(3)"><span class="SelectionMark">&#160;</span>Functions</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(4)"><span class="SelectionMark">&#160;</span>Variables</a></div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

  <div id="nav-path" class="navpath">
    <ul>
      <li class="navelem"><a class="el" href="_figures34_8m.html">Figures34.m</a>      </li>

    <li class="footer">Generated on Wed Feb 8 2012 21:42:31 for Whole Cell Simulation by
    <a href="http://www.doxygen.org/index.html">
    <img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.7.6.1 </li>
   </ul>
 </div>


</body>
</html>
