<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<title>Whole Cell Simulation: src/+edu/+stanford/+covert/+cell/+sim/+process/Metabolism.m Source File</title>

<link href="tabs.css" rel="stylesheet" type="text/css"/>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
<link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="navtree.js"></script>
<script type="text/javascript">
  $(document).ready(initResizable);
</script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
  $(document).ready(function() { searchBox.OnSelectItem(0); });
</script>

</head>
<body>
<div id="top"><!-- do not remove this div! -->


<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  
  
  <td style="padding-left: 0.5em;">
   <div id="projectname">Whole Cell Simulation
   
   </div>
   
  </td>
  
  
  
 </tr>
 </tbody>
</table>
</div>

<!-- Generated by Doxygen 1.7.6.1 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "search",false,'Search');
</script>
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&#160;Page</span></a></li>
      <li><a href="annotated.html"><span>Classes</span></a></li>
      <li class="current"><a href="files.html"><span>Files</span></a></li>
      <li>
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
  <div id="navrow2" class="tabs2">
    <ul class="tablist">
      <li><a href="files.html"><span>File&#160;List</span></a></li>
      <li><a href="globals.html"><span>File&#160;Members</span></a></li>
    </ul>
  </div>
</div>
<div id="side-nav" class="ui-resizable side-nav-resizable">
  <div id="nav-tree">
    <div id="nav-tree-contents">
    </div>
  </div>
  <div id="splitbar" style="-moz-user-select:none;" 
       class="ui-resizable-handle">
  </div>
</div>
<script type="text/javascript">
  initNavTree('_metabolism_8m.html','');
</script>
<div id="doc-content">
<div class="header">
  <div class="headertitle">
<div class="title">src/+edu/+stanford/+covert/+cell/+sim/+process/Metabolism.m</div>  </div>
</div><!--header-->
<div class="contents">
<a href="_metabolism_8m.html">Go to the documentation of this file.</a><div class="fragment"><pre class="fragment"><a name="l00001"></a>00001 %Metabolism
<a name="l00002"></a>00002 %
<a name="l00003"></a>00003 % @wholeCellModelID Process_Metabolism
<a name="l00004"></a>00004 % @name             Metabolism
<a name="l00005"></a>00005 % @description
<a name="l00006"></a>00006 %   Biology
<a name="l00007"></a>00007 %   ===============
<a name="l00008"></a>00008 %   To grow and replicate cells must uptake or produce their building blocks and
<a name="l00009"></a>00009 %   intermediate energy stores, particularly nucleic and amino acids and lipids,
<a name="l00010"></a>00010 %   as well as secrete metabolites such as modified nucleic acids which it
<a name="l00011"></a>00011 %   cannot catabolize and extract usable material or energy from. Furthermore,
<a name="l00012"></a>00012 %   the <span class="keyword">import</span>/synthesis and export/breakdown of metabolites must be carefully
<a name="l00013"></a>00013 %   matched to the metabolic demands of the cell to ensure that cellular
<a name="l00014"></a>00014 %   processes aren<span class="stringliteral">&#39;t limited by too few metabolites or poisoned by too many</span>
<a name="l00015"></a>00015 <span class="stringliteral">%   metabolites.</span>
<a name="l00016"></a>00016 <span class="stringliteral">%</span>
<a name="l00017"></a>00017 <span class="stringliteral">%   This process simulates</span>
<a name="l00018"></a>00018 <span class="stringliteral">%   - the uptake of nutrients from the external environment</span>
<a name="l00019"></a>00019 <span class="stringliteral">%   - the processing of imported nutrients to intermediate energy forms (ATP,</span>
<a name="l00020"></a>00020 <span class="stringliteral">%     GTP) and macromolecule building blocks (dNTPs, NTPs, and amino acids)</span>
<a name="l00021"></a>00021 <span class="stringliteral">%   - the assembly of lipids, their insertion into the membrane, and their</span>
<a name="l00022"></a>00022 <span class="stringliteral">%     maturation within the membrane</span>
<a name="l00023"></a>00023 <span class="stringliteral">%   - the catabolism of byproducts of macromolecule assembly and degradation</span>
<a name="l00024"></a>00024 <span class="stringliteral">%   - the export to the external environment of uncatabolizable chemicals such</span>
<a name="l00025"></a>00025 <span class="stringliteral">%     as modified nucleobases</span>
<a name="l00026"></a>00026 <span class="stringliteral">%</span>
<a name="l00027"></a>00027 <span class="stringliteral">%   Knowledge Base</span>
<a name="l00028"></a>00028 <span class="stringliteral">%   ===============</span>
<a name="l00029"></a>00029 <span class="stringliteral">%   M. genitalium metabolism was reconstructed from a variety of sources,</span>
<a name="l00030"></a>00030 <span class="stringliteral">%   including flux-balance analysis metabolic models of other bacterial species</span>
<a name="l00031"></a>00031 <span class="stringliteral">%   and the reaction kinetics database SABIO-RK, and was organized into 641</span>
<a name="l00032"></a>00032 <span class="stringliteral">%   reactions in the knowledge base. These reactions are loaded into this process</span>
<a name="l00033"></a>00033 <span class="stringliteral">%   by the initializeConstants method.</span>
<a name="l00034"></a>00034 <span class="stringliteral">%</span>
<a name="l00035"></a>00035 <span class="stringliteral">%      Object       No.</span>
<a name="l00036"></a>00036 <span class="stringliteral">%      ==========   ===</span>
<a name="l00037"></a>00037 <span class="stringliteral">%      substrates   580</span>
<a name="l00038"></a>00038 <span class="stringliteral">%      enzymes      100</span>
<a name="l00039"></a>00039 <span class="stringliteral">%      reactions    641</span>
<a name="l00040"></a>00040 <span class="stringliteral">%        chemical   433</span>
<a name="l00041"></a>00041 <span class="stringliteral">%        transport  208</span>
<a name="l00042"></a>00042 <span class="stringliteral">%</span>
<a name="l00043"></a>00043 <span class="stringliteral">%   Representation</span>
<a name="l00044"></a>00044 <span class="stringliteral">%   ===============</span>
<a name="l00045"></a>00045 <span class="stringliteral">%   The properties substrates and enzymes represent the counts of metabolites</span>
<a name="l00046"></a>00046 <span class="stringliteral">%   and metabolic enzymes.</span>
<a name="l00047"></a>00047 <span class="stringliteral">%</span>
<a name="l00048"></a>00048 <span class="stringliteral">%   fbaReactionStoichiometryMatrix represents the stoichiometry and compartments</span>
<a name="l00049"></a>00049 <span class="stringliteral">%   of metabolites and biomass in each of the 641 chemical/transport reactions,</span>
<a name="l00050"></a>00050 <span class="stringliteral">%   exchange pseudoreactions, and biomass production pseudoreaction.</span>
<a name="l00051"></a>00051 <span class="stringliteral">%   fbaReactionCatalysisMatrix represents the enzyme which catalyzes each</span>
<a name="l00052"></a>00052 <span class="stringliteral">%   reaction. fbaEnzymeBounds represents the foward and backward kcat of the</span>
<a name="l00053"></a>00053 <span class="stringliteral">%   catalyzing enzyme of each reaction. fbaReactionBounds represents the maximal</span>
<a name="l00054"></a>00054 <span class="stringliteral">%   import and export rates of each  metabolite. fbaObjective indicates which</span>
<a name="l00055"></a>00055 <span class="stringliteral">%   reaction represents the biomass production pseudoreaction. fbaRightHandSide</span>
<a name="l00056"></a>00056 <span class="stringliteral">%   is a vector of zeros representing the change in concentration over time of</span>
<a name="l00057"></a>00057 <span class="stringliteral">%   each metabolite and biomass. metabolismProduction is redundant with the</span>
<a name="l00058"></a>00058 <span class="stringliteral">%   biomass production reaction in fbaReactionStoichiometryMatrix.</span>
<a name="l00059"></a>00059 <span class="stringliteral">%   metabolismProduction is calculated by summing the metabolic demands of all</span>
<a name="l00060"></a>00060 <span class="stringliteral">%   the other processes over the entire cell cycle. The table below lists the</span>
<a name="l00061"></a>00061 <span class="stringliteral">%   units of several properties of this process.</span>
<a name="l00062"></a>00062 <span class="stringliteral">%</span>
<a name="l00063"></a>00063 <span class="stringliteral">%      Property                       Units</span>
<a name="l00064"></a>00064 <span class="stringliteral">%      ===========================    ==============================</span>
<a name="l00065"></a>00065 <span class="stringliteral">%      fbaEnzymeBounds                molecules/enzyme/s</span>
<a name="l00066"></a>00066 <span class="stringliteral">%      fbaReactionBounds              molecules/(gram dry biomass)/s</span>
<a name="l00067"></a>00067 <span class="stringliteral">%      metabolites                    molecules</span>
<a name="l00068"></a>00068 <span class="stringliteral">%      enzymes                        molecules</span>
<a name="l00069"></a>00069 <span class="stringliteral">%      stepSizeSec                    s</span>
<a name="l00070"></a>00070 <span class="stringliteral">%      lowerBounds                    reactions/s</span>
<a name="l00071"></a>00071 <span class="stringliteral">%      upperBounds                    reactions/s</span>
<a name="l00072"></a>00072 <span class="stringliteral">%      growth                         cell/s</span>
<a name="l00073"></a>00073 <span class="stringliteral">%      biomassComposition             molecules/cell</span>
<a name="l00074"></a>00074 <span class="stringliteral">%      metabolismProduction           molecules/cell</span>
<a name="l00075"></a>00075 <span class="stringliteral">%      chamberVolume                  L</span>
<a name="l00076"></a>00076 <span class="stringliteral">%      setValues                      molecules/chamber</span>
<a name="l00077"></a>00077 <span class="stringliteral">%      growthAssociatedMaintanence</span>
<a name="l00078"></a>00078 <span class="stringliteral">%</span>
<a name="l00079"></a>00079 <span class="stringliteral">%   Initialization</span>
<a name="l00080"></a>00080 <span class="stringliteral">%   ===============</span>
<a name="l00081"></a>00081 <span class="stringliteral">%   The simulation is initialized with 1 cell weight of macromolecules and</span>
<a name="l00082"></a>00082 <span class="stringliteral">%   water, and few free metabolites by the simulation class&#39;</span> initializeState
<a name="l00083"></a>00083 %   method. In addition <span class="keyword">this</span> process is initialized to a positive growth rate of
<a name="l00084"></a>00084 %   approximately log(2)/cellCycleLength computing <span class="keyword">using</span> the flux-balance
<a name="l00085"></a>00085 %   analysis model implemented in evolveState.
<a name="l00086"></a>00086 %
<a name="l00087"></a>00087 %   Simulation
<a name="l00088"></a>00088 %   ===============
<a name="l00089"></a>00089 %   Transport and metabolism are modeled using the constraint-based method
<a name="l00090"></a>00090 %   flux-balance analysis (<a class="code" href="class_f_b_a.html">FBA</a>). Briefly, <a class="code" href="class_f_b_a.html">FBA</a> assumes that bacteria have evolved
<a name="l00091"></a>00091 %   to maximize growth, and poses transport and metabolism as the optimization
<a name="l00092"></a>00092 %   of cellular building block and energy production subject to available
<a name="l00093"></a>00093 %   nutrients, and allowed chemical reactions. <a class="code" href="class_f_b_a.html">FBA</a> models are typically
<a name="l00094"></a>00094 %   constrained by experimentally measured transport and diffusion rates. In
<a name="l00095"></a>00095 %   addition to these constraints, we constrain constrain our <a class="code" href="class_f_b_a.html">FBA</a> process by our
<a name="l00096"></a>00096 %   model&#39;s predicted enzyme abundances, and by experimentally measured kinetic
<a name="l00097"></a>00097 %   parameters. These additional constraints yield a more accurate model of
<a name="l00098"></a>00098 %   transport and metabolism. Furthermore, we use our other processes to obtain a
<a name="l00099"></a>00099 %   more extensive and accurate metabolic objective than used in earlier <a class="code" href="class_f_b_a.html">FBA</a>
<a name="l00100"></a>00100 %   models. The optimization problem specific by <a class="code" href="class_f_b_a.html">FBA</a> is posed as a linear
<a name="l00101"></a>00101 %   optimization problem, and solving using one of several publically available
<a name="l00102"></a>00102 %   linear programming packages.
<a name="l00103"></a>00103 %
<a name="l00104"></a>00104 %   Algorithm
<a name="l00105"></a>00105 %   ++++++++++++++++
<a name="l00106"></a>00106 %   1. Compute reaction bounds based on
<a name="l00107"></a>00107 %     - enzyme kinetics,
<a name="l00108"></a>00108 %     - enzyme availability
<a name="l00109"></a>00109 %     - maximal metabolite exchange rates
<a name="l00110"></a>00110 %     - external metabolite availability
<a name="l00111"></a>00111 %     - protein availability
<a name="l00112"></a>00112 %   2. Computes optimal reaction fluxes which maximize biomass production
<a name="l00113"></a>00113 %   3. Computes integer-valued production of biomass components, and export of
<a name="l00114"></a>00114 %      byproducts
<a name="l00115"></a>00115 %   4. Updates amounts of biomass components and byproducts
<a name="l00116"></a>00116 %
<a name="l00117"></a>00117 %   References
<a name="l00118"></a>00118 %   ===============
<a name="l00119"></a>00119 %   1. Orth JD, Thiele I, Palsson BO (2010). What is flux balance analysis?
<a name="l00120"></a>00120 %      Nat Biotechnol. 28(3):245-8 [PUB_0687].
<a name="l00121"></a>00121 %   2. Thiele I, Palsson BO (2010). A protocol for generating a
<a name="l00122"></a>00122 %      high-quality genome-scale metabolic reconstruction. Nat Protoc.
<a name="l00123"></a>00123 %      5(1): 93-121. [PUB_0686].
<a name="l00124"></a>00124 %   3. Covert MW, Xiao N, Chen TJ, Karr JR (2008). Integrating metabolic,
<a name="l00125"></a>00125 %      transcriptional regulatory and signal transduction models in
<a name="l00126"></a>00126 %      Escherichia coli. Bioinformatics. 24(18):2044-50. [PUB_0684].
<a name="l00127"></a>00127 %   4. Covert MW, Knight EM, Reed JL, Herrgard MJ, Palsson BO (2004).
<a name="l00128"></a>00128 %      Integrating high-throughput and computational data elucidates
<a name="l00129"></a>00129 %      bacterial networks. Nature. 429 (6987): 92-6. [PUB_0618].
<a name="l00130"></a>00130 %   5. Covert MW, Palsson BO (2003). Constraints-based models: regulation
<a name="l00131"></a>00131 %      of gene expression reduces the steady-state solution space. J Theor
<a name="l00132"></a>00132 %      Biol. 221(3): 309-25. [PUB_0685].
<a name="l00133"></a>00133 %
<a name="l00134"></a>00134 % Author: Jonathan Karr, jkarr@stanford.edu
<a name="l00135"></a>00135 % Author: Markus Covert, mcovert@stanford.edu
<a name="l00136"></a>00136 % Author: Jayodita Sanghvi, jayodita@stanford.edu
<a name="l00137"></a>00137 % Affilitation: Covert Lab, Department of Bioengineering, Stanford University
<a name="l00138"></a>00138 % Last updated: 3/22/2011
<a name="l00139"></a>00139 classdef  Metabolism &lt; edu.stanford.covert.cell.sim.ReactionProcess
<a name="l00140"></a>00140     %property annotations
<a name="l00141"></a>00141     properties (Constant)
<a name="l00142"></a>00142         optionNames__              = {   %names of option <a class="code" href="json_parse_8m.html#a6e4256f69b751d3d7ec4ea52ff5a8dee">properties</a>
<a name="l00143"></a>00143             <span class="stringliteral">&#39;linearProgrammingOptions&#39;</span>;
<a name="l00144"></a>00144             <span class="stringliteral">&#39;macromoleculeStateInitialization&#39;</span>;
<a name="l00145"></a>00145             <span class="stringliteral">&#39;tolerance&#39;</span>
<a name="l00146"></a>00146             <span class="stringliteral">&#39;realmax&#39;</span>
<a name="l00147"></a>00147             };
<a name="l00148"></a>00148         fixedConstantNames__       = {   %names of fixed constant <a class="code" href="json_parse_8m.html#a6e4256f69b751d3d7ec4ea52ff5a8dee">properties</a>
<a name="l00149"></a>00149             <span class="stringliteral">&#39;growthAssociatedMaintenance&#39;</span>;
<a name="l00150"></a>00150             <span class="stringliteral">&#39;nonGrowthAssociatedMaintenance&#39;</span>;
<a name="l00151"></a>00151             <span class="stringliteral">&#39;cellCycleLength&#39;</span>;
<a name="l00152"></a>00152             <span class="stringliteral">&#39;exchangeRateUpperBound_carbon&#39;</span>;
<a name="l00153"></a>00153             <span class="stringliteral">&#39;exchangeRateUpperBound_noncarbon&#39;</span>;
<a name="l00154"></a>00154             <span class="stringliteral">&#39;fbaObjective&#39;</span>;
<a name="l00155"></a>00155             <span class="stringliteral">&#39;fbaRightHandSide&#39;</span>;
<a name="l00156"></a>00156             <span class="stringliteral">&#39;fbaReactionStoichiometryMatrix&#39;</span>;
<a name="l00157"></a>00157             <span class="stringliteral">&#39;fbaReactionCatalysisMatrix&#39;</span>;
<a name="l00158"></a>00158             <span class="stringliteral">&#39;fbaReactionBounds&#39;</span>;
<a name="l00159"></a>00159             <span class="stringliteral">&#39;fbaEnzymeBounds&#39;</span>;
<a name="l00160"></a>00160             <span class="stringliteral">&#39;metabolismProduction&#39;</span>;
<a name="l00161"></a>00161             <span class="stringliteral">&#39;monomerCompartments&#39;</span>;
<a name="l00162"></a>00162             };
<a name="l00163"></a>00163         fittedConstantNames__      = {   %names of fitted constant <a class="code" href="json_parse_8m.html#a6e4256f69b751d3d7ec4ea52ff5a8dee">properties</a>
<a name="l00164"></a>00164             <span class="stringliteral">&#39;macromoleculeStateInitializationGrowthFactor&#39;</span>
<a name="l00165"></a>00165             <span class="stringliteral">&#39;unaccountedEnergyConsumption&#39;</span>;
<a name="l00166"></a>00166             };
<a name="l00167"></a>00167         localStateNames__          = {}; %names of simulation state <a class="code" href="json_parse_8m.html#a6e4256f69b751d3d7ec4ea52ff5a8dee">properties</a> redundant with timecourses in <span class="keyword">this</span> or other processes or the simulation
<a name="l00168"></a>00168     end
<a name="l00169"></a>00169     
<a name="l00170"></a>00170     %options
<a name="l00171"></a>00171     <a class="code" href="json_parse_8m.html#a6e4256f69b751d3d7ec4ea52ff5a8dee">properties</a>
<a name="l00172"></a>00172         %linear programming
<a name="l00173"></a>00173         linearProgrammingOptions = <span class="keyword">struct</span>(...
<a name="l00174"></a>00174             <span class="stringliteral">&#39;solver&#39;</span>, <span class="stringliteral">&#39;glpk&#39;</span>,...
<a name="l00175"></a>00175             <span class="stringliteral">&#39;solverOptions&#39;</span>, <span class="keyword">struct</span>(...
<a name="l00176"></a>00176                 <span class="stringliteral">&#39;glpk&#39;</span>, <span class="keyword">struct</span>(<span class="stringliteral">&#39;lpsolver&#39;</span>, 1, <span class="stringliteral">&#39;presol&#39;</span>, 1, <span class="stringliteral">&#39;scale&#39;</span>, 1, <span class="stringliteral">&#39;msglev&#39;</span>, 0, <span class="stringliteral">&#39;tolbnd&#39;</span>, 10<a class="code" href="json_parse_8m.html#ac14d26badced2970320ee4d772ad0ae9">e</a>-7),...
<a name="l00177"></a>00177                 &#39;linprog&#39;, struct(&#39;Display&#39;,&#39;off&#39;),...
<a name="l00178"></a>00178                 &#39;lp_solve&#39;, struct(&#39;verbose&#39;, 0, &#39;scaling&#39;, 3 + 64 + 128, &#39;presolve&#39;, 0), ...
<a name="l00179"></a>00179                 &#39;qsopt&#39;, struct()));
<a name="l00180"></a>00180         
<a name="l00181"></a>00181         %initial macromolecule state distribution
<a name="l00182"></a>00182         macromoleculeStateInitialization = <span class="stringliteral">&#39;multinomial&#39;</span>;
<a name="l00183"></a>00183         
<a name="l00184"></a>00184         %fitting tolerance
<a name="l00185"></a>00185         tolerance = 1<a class="code" href="json_parse_8m.html#ac14d26badced2970320ee4d772ad0ae9">e</a>-3;
<a name="l00186"></a>00186         
<a name="l00187"></a>00187         %realmax
<a name="l00188"></a>00188         % Set to several orders of magnitude higher than non-cycle highest
<a name="l00189"></a>00189         % fluxs. In tests gplk will tolerates as high as 10^9. Higher than
<a name="l00190"></a>00190         % 10^9 glpk and other solvers have trouble finding feasible
<a name="l00191"></a>00191         % solutions.
<a name="l00192"></a>00192         realmax = 1e6;
<a name="l00193"></a>00193     end
<a name="l00194"></a>00194     
<a name="l00195"></a>00195     %IDs, names, and local indices
<a name="l00196"></a>00196     <a class="code" href="json_parse_8m.html#a6e4256f69b751d3d7ec4ea52ff5a8dee">properties</a>
<a name="l00197"></a>00197         substrateIndexs_amp                   %index of AMP within reactionStoichiometryMatrix
<a name="l00198"></a>00198         substrateIndexs_adp                   %index of ADP within reactionStoichiometryMatrix
<a name="l00199"></a>00199         substrateIndexs_atp                   %index of ATP within reactionStoichiometryMatrix
<a name="l00200"></a>00200         substrateIndexs_diphosphate           %index of PPi within reactionStoichiometryMatrix
<a name="l00201"></a>00201         substrateIndexs_phosphate             %index of Pi  within reactionStoichiometryMatrix
<a name="l00202"></a>00202         substrateIndexs_water                 %index of H2O within reactionStoichiometryMatrix
<a name="l00203"></a>00203         substrateIndexs_hydrogen              %index of H   within reactionStoichiometryMatrix
<a name="l00204"></a>00204         substrateIndexs_atpHydrolysis         %index of ATP, H2O, ADP, PI, H within reactionStoichiometryMatrix
<a name="l00205"></a>00205         substrateIndexs_energy                %index of NxPs, PPI, PI, H within reactionStoichiometryMatrix
<a name="l00206"></a>00206         
<a name="l00207"></a>00207         substrateIndexs_limitableProteins     %indices of limitable proteins within reactionStoichiometryMatrix
<a name="l00208"></a>00208         substrateIndexs_fba                   %indices of <a class="code" href="class_f_b_a.html">FBA</a> substrates within reactionStoichiometryMatrix
<a name="l00209"></a>00209         substrateIndexs_externalExchangedMetabolites  %indices of exchanged metabolites within reactionStoichiometryMatrix
<a name="l00210"></a>00210         substrateIndexs_internalExchangedMetabolites  %indices of exchanged metabolites within reactionStoichiometryMatrix
<a name="l00211"></a>00211         substrateIndexs_internalExchangedLimitedMetabolites %indices of limited, exchanged metabolites within reactionStoichiometryMatrix
<a name="l00212"></a>00212         
<a name="l00213"></a>00213         fbaSubstrateIndexs_substrates         %indices of substrates (metabolites, RNAs, proteins) within fbaReactionStoichiometryMatrix
<a name="l00214"></a>00214         fbaSubstrateIndexs_metaboliteInternalExchangeConstraints %indices of substrates with are internally exchanged
<a name="l00215"></a>00215         fbaSubstrateIndexs_biomass            %index of biomass within fbaReactionStoichiometryMatrix
<a name="l00216"></a>00216         
<a name="l00217"></a>00217         reactionIndexs_chemical               %indices of chemical reactions within reactionStoichiometryMatrix
<a name="l00218"></a>00218         reactionIndexs_transport              %indices of transport reactions within reactionStoichiometryMatrix
<a name="l00219"></a>00219         reactionIndexs_fba                    %indices of <a class="code" href="class_f_b_a.html">FBA</a> reactions within reactionStoichiometryMatrix
<a name="l00220"></a>00220         
<a name="l00221"></a>00221         fbaReactionIndexs_metabolicConversion %indices of metabolic reactions (chemical, transport) within fbaReactionStoichiometryMatrix
<a name="l00222"></a>00222         fbaReactionIndexs_metaboliteExternalExchange %indices of exchange reactions within fbaReactionStoichiometryMatrix
<a name="l00223"></a>00223         fbaReactionIndexs_metaboliteInternalExchange %indices of exchange reactions within fbaReactionStoichiometryMatrix
<a name="l00224"></a>00224         fbaReactionIndexs_metaboliteInternalLimitedExchange %indices of exchange reactions within fbaReactionStoichiometryMatrix
<a name="l00225"></a>00225         fbaReactionIndexs_metaboliteInternalUnlimitedExchange %indices of exchange reactions within fbaReactionStoichiometryMatrix
<a name="l00226"></a>00226         fbaReactionIndexs_biomassProduction   %index of biomass production reaction within fbaReactionStoichiometryMatrix
<a name="l00227"></a>00227         fbaReactionIndexs_biomassExchange     %index of biomass efflux reaction within fbaReactionStoichiometryMatrix
<a name="l00228"></a>00228         
<a name="l00229"></a>00229         compartmentIndexs_cytosol             %index of cytosol compartment within reactionStoichiometryMatrix
<a name="l00230"></a>00230         compartmentIndexs_extracellular       %index of extracellular compartment within reactionStoichiometryMatrix
<a name="l00231"></a>00231         compartmentIndexs_membrane            %index of membrane compartment within reactionStoichiometryMatrix
<a name="l00232"></a>00232     end
<a name="l00233"></a>00233     
<a name="l00234"></a>00234     %fixed biological constants
<a name="l00235"></a>00235     <a class="code" href="json_parse_8m.html#a6e4256f69b751d3d7ec4ea52ff5a8dee">properties</a>
<a name="l00236"></a>00236         macromoleculeStateInitializationGrowthFactor %set by running fitting once and seeing what mean growth rate is predicted; see fitEnzymes
<a name="l00237"></a>00237         
<a name="l00238"></a>00238         growthAssociatedMaintenance         %mmol ATP/gDCW [59.810; PUB_0558]
<a name="l00239"></a>00239         nonGrowthAssociatedMaintenance      %mmol ATP/gDCW [8.390; PUB_0558]
<a name="l00240"></a>00240         cellCycleLength                     %cell cycle length (s)
<a name="l00241"></a>00241         exchangeRateUpperBound_carbon       %exchange rate upper bound -- carbon-containing metabolites (mmol/gDCW/h) [12]
<a name="l00242"></a>00242         exchangeRateUpperBound_noncarbon    %exchange rate upper bound -- non-carbon-containing metabolites (mmol/gDCW/h) [20]
<a name="l00243"></a>00243         substrateExternalExchangeBounds     %substrate exchange bounds                                                          [substrates X 2]
<a name="l00244"></a>00244         substrateExchangeBounds             %substrate exchange bounds                                                          [substrates X 2]
<a name="l00245"></a>00245         fbaObjective                        %<a class="code" href="class_f_b_a.html">FBA</a> objective function corresponding to maximize growth                            [reactions;growth;biomass efflux] X 1
<a name="l00246"></a>00246         fbaRightHandSide                    %right-hand side of equation Sv=0                                                   [substrates;biomass] X 1
<a name="l00247"></a>00247         fbaReactionStoichiometryMatrix      %<a class="code" href="class_f_b_a.html">FBA</a> stoichiometry matrix                                                           [substrates;biomass] X [reactions;growth;biomass efflux] X compartments
<a name="l00248"></a>00248         fbaReactionCatalysisMatrix          %Enzymes catalyzing each <a class="code" href="class_f_b_a.html">FBA</a> reaction                                               [reactions;growth;biomass efflux] X enzymes
<a name="l00249"></a>00249         fbaReactionBounds                   %Exchange bounds associated with each <a class="code" href="class_f_b_a.html">FBA</a> reaction (molecules/gram dry biomass/s)   [reactions;growth;biomass efflux] X 2
<a name="l00250"></a>00250         fbaEnzymeBounds                     %Kinetics bounds associated with each <a class="code" href="class_f_b_a.html">FBA</a> reaction (molecules/enzyme/s)             [reactions;growth;biomass efflux] X 2
<a name="l00251"></a>00251         metabolismProduction                %biomass composition + energy + byproducts (molecules/cell)                         substrates X compartments
<a name="l00252"></a>00252         metabolismNewProduction             %metabolism output represented by biomass pseudoreaction
<a name="l00253"></a>00253         metabolismRecyclingProduction       %metabolism output represented by internal exchange reactions
<a name="l00254"></a>00254         unaccountedEnergyConsumption        %unaccounted energy consumption (ATP molecules / cell life cycle)                   ATP X 1
<a name="l00255"></a>00255         monomerCompartments                 %Indexs of compartments of protein monomers
<a name="l00256"></a>00256         proteinLimitableProteinComposition  %protein complex composition                                                        [monomers; complexs] X [monomers; limitable complexs]
<a name="l00257"></a>00257     end
<a name="l00258"></a>00258     
<a name="l00259"></a>00259     %global state (referenced locally for convenience)
<a name="l00260"></a>00260     <a class="code" href="json_parse_8m.html#a6e4256f69b751d3d7ec4ea52ff5a8dee">properties</a>
<a name="l00261"></a>00261         mass                      %cell mass
<a name="l00262"></a>00262         metabolicReaction         %metabolic flux
<a name="l00263"></a>00263     end
<a name="l00264"></a>00264     
<a name="l00265"></a>00265     %constructor
<a name="l00266"></a>00266     methods
<a name="l00267"></a>00267         function this = Metabolism(wholeCellModelID, name)
<a name="l00268"></a>00268             this = this@edu.stanford.covert.cell.sim.<a class="code" href="class_reaction_process.html">ReactionProcess</a>(wholeCellModelID, name);
<a name="l00269"></a>00269         end
<a name="l00270"></a>00270     end
<a name="l00271"></a>00271     
<a name="l00272"></a>00272     %communication between process/simulation
<a name="l00273"></a>00273     methods
<a name="l00274"></a>00274         %set references to state objects
<a name="l00275"></a>00275         function storeObjectReferences(this, simulation)
<a name="l00276"></a>00276             this.storeObjectReferences@edu.stanford.covert.cell.sim.<a class="code" href="class_reaction_process.html">ReactionProcess</a>(simulation);
<a name="l00277"></a>00277             
<a name="l00278"></a>00278             this.mass = simulation.state(&#39;Mass&#39;);
<a name="l00279"></a>00279             this.metabolicReaction = simulation.state(&#39;<a class="code" href="class_metabolic_reaction.html">MetabolicReaction</a>&#39;);
<a name="l00280"></a>00280             this.states = [this.states; {this.mass; this.metabolicReaction}];
<a name="l00281"></a>00281         end
<a name="l00282"></a>00282         
<a name="l00283"></a>00283         %initialize constants
<a name="l00284"></a>00284         <span class="keyword">function</span> initializeConstants(<span class="keyword">this</span>, knowledgeBase, simulation, varargin)
<a name="l00285"></a>00285             import edu.stanford.covert.util.<a class="code" href="class_constant_util.html">ConstantUtil</a>;
<a name="l00286"></a>00286             
<a name="l00287"></a>00287             %% options
<a name="l00288"></a>00288             options = struct(&#39;retainSubstrateCompartments&#39;, true);
<a name="l00289"></a>00289             
<a name="l00290"></a>00290             %% call super class method
<a name="l00291"></a>00291             this.initializeConstants@edu.stanford.covert.cell.sim.<a class="code" href="class_reaction_process.html">ReactionProcess</a>(...
<a name="l00292"></a>00292                 knowledgeBase, simulation, options);
<a name="l00293"></a>00293             
<a name="l00294"></a>00294             %% cell cycle length
<a name="l00295"></a>00295             this.cellCycleLength = simulation.state(&#39;Time&#39;).cellCycleLength;
<a name="l00296"></a>00296             
<a name="l00297"></a>00297             %% compartments
<a name="l00298"></a>00298             this.compartmentIndexs_cytosol       = 1;
<a name="l00299"></a>00299             this.compartmentIndexs_extracellular = 2;
<a name="l00300"></a>00300             this.compartmentIndexs_membrane      = 3;
<a name="l00301"></a>00301             
<a name="l00302"></a>00302             compartmentIndexs = [
<a name="l00303"></a>00303                 this.compartment.cytosolIndexs;
<a name="l00304"></a>00304                 this.compartment.extracellularIndexs;
<a name="l00305"></a>00305                 this.compartment.membraneIndexs];
<a name="l00306"></a>00306             
<a name="l00307"></a>00307             this.substrateStimulusGlobalIndexs        = this.substrateStimulusGlobalIndexs(  :, compartmentIndexs);
<a name="l00308"></a>00308             this.substrateMetaboliteGlobalIndexs      = this.substrateMetaboliteGlobalIndexs(:, compartmentIndexs);
<a name="l00309"></a>00309             this.substrateRNAGlobalIndexs             = this.substrateRNAGlobalIndexs(       :, compartmentIndexs);
<a name="l00310"></a>00310             this.substrateMonomerGlobalIndexs         = this.substrateMonomerGlobalIndexs(   :, compartmentIndexs);
<a name="l00311"></a>00311             this.substrateComplexGlobalIndexs         = this.substrateComplexGlobalIndexs(   :, compartmentIndexs);
<a name="l00312"></a>00312             
<a name="l00313"></a>00313             this.substrateStimulusCompartmentIndexs   = this.substrateStimulusCompartmentIndexs(  :, compartmentIndexs);
<a name="l00314"></a>00314             this.substrateMetaboliteCompartmentIndexs = this.substrateMetaboliteCompartmentIndexs(:, compartmentIndexs);
<a name="l00315"></a>00315             this.substrateRNACompartmentIndexs        = this.substrateRNACompartmentIndexs(       :, compartmentIndexs);
<a name="l00316"></a>00316             this.substrateMonomerCompartmentIndexs    = this.substrateMonomerCompartmentIndexs(   :, compartmentIndexs);
<a name="l00317"></a>00317             this.substrateComplexCompartmentIndexs    = this.substrateComplexCompartmentIndexs(   :, compartmentIndexs);
<a name="l00318"></a>00318             
<a name="l00319"></a>00319             this.substrateStimulusGlobalCompartmentIndexs = sub2ind(...
<a name="l00320"></a>00320                 [numel(this.stimulus.wholeCellModelIDs) this.compartment.count],...
<a name="l00321"></a>00321                 this.substrateStimulusGlobalIndexs,...
<a name="l00322"></a>00322                 this.substrateStimulusCompartmentIndexs);
<a name="l00323"></a>00323             this.substrateMetaboliteGlobalCompartmentIndexs = sub2ind(...
<a name="l00324"></a>00324                 [numel(this.metabolite.wholeCellModelIDs) this.compartment.count],...
<a name="l00325"></a>00325                 this.substrateMetaboliteGlobalIndexs,...
<a name="l00326"></a>00326                 this.substrateMetaboliteCompartmentIndexs);
<a name="l00327"></a>00327             this.substrateRNAGlobalCompartmentIndexs = sub2ind(...
<a name="l00328"></a>00328                 [numel(this.rna.wholeCellModelIDs) this.compartment.count],...
<a name="l00329"></a>00329                 this.rna.matureIndexs(this.substrateRNAGlobalIndexs),...
<a name="l00330"></a>00330                 this.substrateRNACompartmentIndexs);
<a name="l00331"></a>00331             this.substrateMonomerGlobalCompartmentIndexs = sub2ind(...
<a name="l00332"></a>00332                 [numel(this.monomer.wholeCellModelIDs) this.compartment.count],...
<a name="l00333"></a>00333                 this.monomer.matureIndexs(this.substrateMonomerGlobalIndexs),...
<a name="l00334"></a>00334                 this.substrateMonomerCompartmentIndexs);
<a name="l00335"></a>00335             this.substrateComplexGlobalCompartmentIndexs = sub2ind(...
<a name="l00336"></a>00336                 [numel(this.complex.wholeCellModelIDs) this.compartment.count],...
<a name="l00337"></a>00337                 this.complex.matureIndexs(this.substrateComplexGlobalIndexs),...
<a name="l00338"></a>00338                 this.substrateComplexCompartmentIndexs);
<a name="l00339"></a>00339             
<a name="l00340"></a>00340             this.reactionStoichiometryMatrix = this.reactionStoichiometryMatrix(:, :, compartmentIndexs);
<a name="l00341"></a>00341             this.reactionCoenzymeMatrix = this.reactionCoenzymeMatrix(:, :, compartmentIndexs);
<a name="l00342"></a>00342             
<a name="l00343"></a>00343             %% metabolites
<a name="l00344"></a>00344             this.substrateIndexs_amp           = this.substrateIndexs({<span class="stringliteral">&#39;AMP&#39;</span>});
<a name="l00345"></a>00345             this.substrateIndexs_adp           = this.substrateIndexs({<span class="stringliteral">&#39;ADP&#39;</span>});
<a name="l00346"></a>00346             this.substrateIndexs_atp           = this.substrateIndexs({<span class="stringliteral">&#39;ATP&#39;</span>});
<a name="l00347"></a>00347             this.substrateIndexs_diphosphate   = this.substrateIndexs({<span class="stringliteral">&#39;PPI&#39;</span>});
<a name="l00348"></a>00348             this.substrateIndexs_phosphate     = this.substrateIndexs({<span class="stringliteral">&#39;PI&#39;</span>});
<a name="l00349"></a>00349             this.substrateIndexs_water         = this.substrateIndexs({<span class="stringliteral">&#39;H2O&#39;</span>});
<a name="l00350"></a>00350             this.substrateIndexs_hydrogen      = this.substrateIndexs({<span class="charliteral">&#39;H&#39;</span>});
<a name="l00351"></a>00351             this.substrateIndexs_atpHydrolysis = this.substrateIndexs({<span class="stringliteral">&#39;ATP&#39;</span>; <span class="stringliteral">&#39;H2O&#39;</span>; <span class="stringliteral">&#39;ADP&#39;</span>; <span class="stringliteral">&#39;PI&#39;</span>; <span class="charliteral">&#39;H&#39;</span>});
<a name="l00352"></a>00352             this.substrateIndexs_energy        = sort(this.substrateIndexs({<span class="stringliteral">&#39;AMP&#39;</span>, <span class="stringliteral">&#39;CMP&#39;</span>, <span class="stringliteral">&#39;GMP&#39;</span>, <span class="stringliteral">&#39;UMP&#39;</span>, <span class="stringliteral">&#39;ADP&#39;</span>, <span class="stringliteral">&#39;CDP&#39;</span>, <span class="stringliteral">&#39;GDP&#39;</span>, <span class="stringliteral">&#39;UDP&#39;</span>, <span class="stringliteral">&#39;ATP&#39;</span>, <span class="stringliteral">&#39;CTP&#39;</span>, <span class="stringliteral">&#39;GTP&#39;</span>, <span class="stringliteral">&#39;UTP&#39;</span>, <span class="stringliteral">&#39;PI&#39;</span>, <span class="stringliteral">&#39;PPI&#39;</span>, <span class="charliteral">&#39;H&#39;</span>}));
<a name="l00353"></a>00353             
<a name="l00354"></a>00354             %% reactions
<a name="l00355"></a>00355             this.reactionIndexs_chemical  = find(strcmp(this.reactionTypes, <span class="stringliteral">&#39;chemical&#39;</span>));
<a name="l00356"></a>00356             this.reactionIndexs_transport = find(strcmp(this.reactionTypes, <span class="stringliteral">&#39;transport&#39;</span>));
<a name="l00357"></a>00357             
<a name="l00358"></a>00358             %% <a class="code" href="class_f_b_a.html">FBA</a>
<a name="l00359"></a>00359             %comlexes
<a name="l00360"></a>00360             complexFormationProcesses = [knowledgeBase.proteinComplexs(this.substrateComplexGlobalIndexs(:, 1)).complexFormationProcess];
<a name="l00361"></a>00361             this.substrateIndexs_limitableProteins = [...
<a name="l00362"></a>00362                 this.substrateMonomerLocalIndexs
<a name="l00363"></a>00363                 this.substrateComplexLocalIndexs(strcmp({complexFormationProcesses.wholeCellModelID}, &#39;Process_MacromolecularComplexation&#39;), :)];
<a name="l00364"></a>00364             
<a name="l00365"></a>00365             substratesByLimitableSubstrates = zeros(numel(this.substrateWholeCellModelIDs), numel(this.substrateWholeCellModelIDs));
<a name="l00366"></a>00366             substratesByLimitableSubstrates(this.substrateComplexLocalIndexs, this.substrateMonomerLocalIndexs) = ...
<a name="l00367"></a>00367                 any(knowledgeBase.proteinComplexMonomerComposition(<span class="keyword">this</span>.gene.mRNAIndexs(<span class="keyword">this</span>.substrateMonomerGlobalIndexs(:, 1)), this.substrateComplexGlobalIndexs(:, 1), :), 3)<span class="stringliteral">&#39;;</span>
<a name="l00368"></a>00368 <span class="stringliteral">            substratesByLimitableSubstrates(this.substrateComplexLocalIndexs, this.substrateComplexLocalIndexs) =  ...</span>
<a name="l00369"></a>00369 <span class="stringliteral">                any(knowledgeBase.proteinComplexComplexComposition(this.substrateComplexGlobalIndexs(:, 1), this.substrateComplexGlobalIndexs(:, 1), :), 3)&#39;</span>;
<a name="l00370"></a>00370             substratesByLimitableSubstrates(:, setdiff(1:end, this.substrateIndexs_limitableProteins)) = 0;
<a name="l00371"></a>00371             substratesByLimitableSubstrates(this.substrateIndexs_limitableProteins, :) = 0;
<a name="l00372"></a>00372             substratesByLimitableSubstrates(sub2ind(...
<a name="l00373"></a>00373                 size(substratesByLimitableSubstrates), ....
<a name="l00374"></a>00374                 this.substrateIndexs_limitableProteins, ...
<a name="l00375"></a>00375                 this.substrateIndexs_limitableProteins)) = 1;
<a name="l00376"></a>00376             
<a name="l00377"></a>00377             this.proteinLimitableProteinComposition = substratesByLimitableSubstrates([this.substrateMonomerLocalIndexs; this.substrateComplexLocalIndexs], this.substrateIndexs_limitableProteins);
<a name="l00378"></a>00378             
<a name="l00379"></a>00379             %exchange bounds
<a name="l00380"></a>00380             mets = knowledgeBase.metabolites(this.substrateMetaboliteGlobalIndexs(:, 1));
<a name="l00381"></a>00381             this.substrateExternalExchangeBounds = zeros(numel(this.substrateWholeCellModelIDs), 2);
<a name="l00382"></a>00382             <span class="keywordflow">for</span> i = 1:numel(mets)
<a name="l00383"></a>00383                 met = mets(i);
<a name="l00384"></a>00384                 if isfield(met.empiricalFormula, &#39;C&#39;) &amp;&amp; met.empiricalFormula.C &gt; 0
<a name="l00385"></a>00385                     this.substrateExternalExchangeBounds(this.substrateMetaboliteLocalIndexs(i), 1) = max(-this.exchangeRateUpperBound_carbon, met.exchangeLowerBound);
<a name="l00386"></a>00386                     this.substrateExternalExchangeBounds(this.substrateMetaboliteLocalIndexs(i), 2) = min( this.exchangeRateUpperBound_carbon, met.exchangeUpperBound);
<a name="l00387"></a>00387                 else
<a name="l00388"></a>00388                     this.substrateExternalExchangeBounds(this.substrateMetaboliteLocalIndexs(i), 1) = max(-this.exchangeRateUpperBound_noncarbon, met.exchangeLowerBound);
<a name="l00389"></a>00389                     this.substrateExternalExchangeBounds(this.substrateMetaboliteLocalIndexs(i), 2) = min( this.exchangeRateUpperBound_noncarbon, met.exchangeUpperBound);
<a name="l00390"></a>00390                 end
<a name="l00391"></a>00391             end
<a name="l00392"></a>00392             this.substrateExternalExchangeBounds = this.substrateExternalExchangeBounds / ...
<a name="l00393"></a>00393                 <a class="code" href="class_constant_util.html">ConstantUtil</a>.secondsPerHour * (<a class="code" href="class_constant_util.html">ConstantUtil</a>.nAvogadro / 1000);
<a name="l00394"></a>00394             
<a name="l00395"></a>00395             %% Monomers
<a name="l00396"></a>00396             this.monomerCompartments = knowledgeBase.proteinMonomerCompartments;
<a name="l00397"></a>00397         end
<a name="l00398"></a>00398         
<a name="l00399"></a>00399         %Formulate <a class="code" href="class_f_b_a.html">FBA</a> model for current biomass composition stored in
<a name="l00400"></a>00400         %metabolite state and this module&#39;s reactions
<a name="l00401"></a>00401         %- objective
<a name="l00402"></a>00402         %- stoichiometry matrix
<a name="l00403"></a>00403         %- bounds
<a name="l00404"></a>00404         %
<a name="l00405"></a>00405         %To decrease linear programming run-time and increase linear
<a name="l00406"></a>00406         %programming accuracy this method automatically reduces the size of
<a name="l00407"></a>00407         %the linear programming problem by removing
<a name="l00408"></a>00408         %- reaction and substrates disconnected from biomass
<a name="l00409"></a>00409         %- leaf metabolites (and reactions involving them) which cannot be
<a name="l00410"></a>00410         %  balanced with non-zero fluxes
<a name="l00411"></a>00411         %- metabolites (and reactions involving them) which can&#39;t move in
<a name="l00412"></a>00412         %  both the forward and backward directions and thus cannot be
<a name="l00413"></a>00413         %  balanced
<a name="l00414"></a>00414         %- other reactions (and substrates unique to them) whose flux is
<a name="l00415"></a>00415         %  zero in all solutions of the matrix problem Sv=0. That is
<a name="l00416"></a>00416         %  reactions which have only zero components in the null space of
<a name="l00417"></a>00417         %  S.
<a name="l00418"></a>00418         function [subCmpRemoved, rxnRemoved] = formulateFBA(this, metabolismProduction, unaccountedEnergyConsumption, simplifyNetwork)
<a name="l00419"></a>00419             import edu.stanford.covert.util.<a class="code" href="class_computation_util.html">ComputationUtil</a>;
<a name="l00420"></a>00420             import edu.stanford.covert.util.<a class="code" href="class_constant_util.html">ConstantUtil</a>;
<a name="l00421"></a>00421             import edu.stanford.covert.util.<a class="code" href="find_non_interacting_rows_and_columns_8m.html#a35f7ca7ac9040917387bac96a24f6b30">findNonInteractingRowsAndColumns</a>;
<a name="l00422"></a>00422             
<a name="l00423"></a>00423             %% metabolism output
<a name="l00424"></a>00424             if nargin &gt;= 2 &amp;&amp; ~isempty(metabolismProduction)
<a name="l00425"></a>00425                 metabolismProduction(abs(metabolismProduction) &lt; 1<a class="code" href="json_parse_8m.html#ac14d26badced2970320ee4d772ad0ae9">e</a>-4) = 0; %zero out metabolites that almost zero up to rounding error
<a name="l00426"></a>00426                 this.metabolismProduction = zeros(numel(this.substrateWholeCellModelIDs), size(this.substrateMetaboliteGlobalIndexs, 2));
<a name="l00427"></a>00427                 this.metabolismProduction(this.substrateMetaboliteLocalIndexs, :) = metabolismProduction(sub2ind(...
<a name="l00428"></a>00428                     size(metabolismProduction), ...
<a name="l00429"></a>00429                     this.substrateMetaboliteGlobalIndexs, ...
<a name="l00430"></a>00430                     this.substrateMetaboliteCompartmentIndexs));
<a name="l00431"></a>00431             end
<a name="l00432"></a>00432             if nargin &gt;= 3 &amp;&amp; ~isempty(unaccountedEnergyConsumption)
<a name="l00433"></a>00433                 this.unaccountedEnergyConsumption = unaccountedEnergyConsumption;
<a name="l00434"></a>00434             end
<a name="l00435"></a>00435             assert(~any(sum(this.metabolismProduction ~= 0, 2) &gt; 1));
<a name="l00436"></a>00436             
<a name="l00437"></a>00437             this.metabolismNewProduction = max(0, this.metabolismProduction);
<a name="l00438"></a>00438             this.metabolismRecyclingProduction = min(0, this.metabolismProduction);
<a name="l00439"></a>00439             
<a name="l00440"></a>00440             this.metabolismRecyclingProduction(this.substrateIndexs({<span class="stringliteral">&#39;ATP&#39;</span>; <span class="stringliteral">&#39;CTP&#39;</span>; <span class="stringliteral">&#39;GTP&#39;</span>; <span class="stringliteral">&#39;UTP&#39;</span>}), :) = ...
<a name="l00441"></a>00441                 + this.metabolismRecyclingProduction(this.substrateIndexs({<span class="stringliteral">&#39;ATP&#39;</span>; <span class="stringliteral">&#39;CTP&#39;</span>; <span class="stringliteral">&#39;GTP&#39;</span>; <span class="stringliteral">&#39;UTP&#39;</span>}), :) ...
<a name="l00442"></a>00442                 - this.metabolismRecyclingProduction(this.substrateIndexs({<span class="stringliteral">&#39;ADP&#39;</span>; <span class="stringliteral">&#39;CDP&#39;</span>; <span class="stringliteral">&#39;GDP&#39;</span>; <span class="stringliteral">&#39;UDP&#39;</span>}), :) ...
<a name="l00443"></a>00443                 - this.metabolismRecyclingProduction(this.substrateIndexs({<span class="stringliteral">&#39;AMP&#39;</span>; <span class="stringliteral">&#39;CMP&#39;</span>; <span class="stringliteral">&#39;GMP&#39;</span>; <span class="stringliteral">&#39;UMP&#39;</span>}), :);
<a name="l00444"></a>00444             this.metabolismNewProduction(this.substrateIndexs({<span class="stringliteral">&#39;ATP&#39;</span>; <span class="stringliteral">&#39;CTP&#39;</span>; <span class="stringliteral">&#39;GTP&#39;</span>; <span class="stringliteral">&#39;UTP&#39;</span>}), :) = ...
<a name="l00445"></a>00445                 + this.metabolismNewProduction(this.substrateIndexs({<span class="stringliteral">&#39;ATP&#39;</span>; <span class="stringliteral">&#39;CTP&#39;</span>; <span class="stringliteral">&#39;GTP&#39;</span>; <span class="stringliteral">&#39;UTP&#39;</span>}), :) ...
<a name="l00446"></a>00446                 + this.metabolismRecyclingProduction(this.substrateIndexs({<span class="stringliteral">&#39;ADP&#39;</span>; <span class="stringliteral">&#39;CDP&#39;</span>; <span class="stringliteral">&#39;GDP&#39;</span>; <span class="stringliteral">&#39;UDP&#39;</span>}), :) ...
<a name="l00447"></a>00447                 + this.metabolismRecyclingProduction(this.substrateIndexs({<span class="stringliteral">&#39;AMP&#39;</span>; <span class="stringliteral">&#39;CMP&#39;</span>; <span class="stringliteral">&#39;GMP&#39;</span>; <span class="stringliteral">&#39;UMP&#39;</span>}), :);
<a name="l00448"></a>00448             
<a name="l00449"></a>00449             this.metabolismRecyclingProduction(this.substrateIndexs({<span class="stringliteral">&#39;H2O&#39;</span>}), :) = ...
<a name="l00450"></a>00450                 + this.metabolismRecyclingProduction(this.substrateIndexs({<span class="stringliteral">&#39;H2O&#39;</span>}), :) ...
<a name="l00451"></a>00451                 - sum(this.metabolismRecyclingProduction(this.substrateIndexs({<span class="stringliteral">&#39;ADP&#39;</span>; <span class="stringliteral">&#39;CDP&#39;</span>; <span class="stringliteral">&#39;GDP&#39;</span>; <span class="stringliteral">&#39;UDP&#39;</span>}), :)) ...
<a name="l00452"></a>00452                 - sum(this.metabolismRecyclingProduction(this.substrateIndexs({<span class="stringliteral">&#39;AMP&#39;</span>; <span class="stringliteral">&#39;CMP&#39;</span>; <span class="stringliteral">&#39;GMP&#39;</span>; <span class="stringliteral">&#39;UMP&#39;</span>}), :));
<a name="l00453"></a>00453             this.metabolismNewProduction(this.substrateIndexs({<span class="stringliteral">&#39;H2O&#39;</span>}), :) = ...
<a name="l00454"></a>00454                 + this.metabolismNewProduction(this.substrateIndexs({<span class="stringliteral">&#39;H2O&#39;</span>}), :) ...
<a name="l00455"></a>00455                 + sum(this.metabolismRecyclingProduction(this.substrateIndexs({<span class="stringliteral">&#39;ADP&#39;</span>; <span class="stringliteral">&#39;CDP&#39;</span>; <span class="stringliteral">&#39;GDP&#39;</span>; <span class="stringliteral">&#39;UDP&#39;</span>}), :)) ...
<a name="l00456"></a>00456                 + sum(this.metabolismRecyclingProduction(this.substrateIndexs({<span class="stringliteral">&#39;AMP&#39;</span>; <span class="stringliteral">&#39;CMP&#39;</span>; <span class="stringliteral">&#39;GMP&#39;</span>; <span class="stringliteral">&#39;UMP&#39;</span>}), :));
<a name="l00457"></a>00457             
<a name="l00458"></a>00458             this.metabolismRecyclingProduction(this.substrateIndexs({<span class="stringliteral">&#39;ATP&#39;</span>; <span class="stringliteral">&#39;ADP&#39;</span>; <span class="stringliteral">&#39;PI&#39;</span>; <span class="stringliteral">&#39;H2O&#39;</span>; <span class="charliteral">&#39;H&#39;</span>}), this.compartmentIndexs_cytosol) = ...
<a name="l00459"></a>00459                 + this.metabolismRecyclingProduction(this.substrateIndexs({<span class="stringliteral">&#39;ATP&#39;</span>; <span class="stringliteral">&#39;ADP&#39;</span>; <span class="stringliteral">&#39;PI&#39;</span>; <span class="stringliteral">&#39;H2O&#39;</span>; <span class="charliteral">&#39;H&#39;</span>}), this.compartmentIndexs_cytosol) ...
<a name="l00460"></a>00460                 + this.unaccountedEnergyConsumption * [-1; 1; 1; -1; 1];
<a name="l00461"></a>00461             this.metabolismNewProduction(this.substrateIndexs({<span class="stringliteral">&#39;ATP&#39;</span>; <span class="stringliteral">&#39;ADP&#39;</span>; <span class="stringliteral">&#39;PI&#39;</span>; <span class="stringliteral">&#39;H2O&#39;</span>; <span class="charliteral">&#39;H&#39;</span>}), this.compartmentIndexs_cytosol) = ...
<a name="l00462"></a>00462                 + this.metabolismNewProduction(this.substrateIndexs({<span class="stringliteral">&#39;ATP&#39;</span>; <span class="stringliteral">&#39;ADP&#39;</span>; <span class="stringliteral">&#39;PI&#39;</span>; <span class="stringliteral">&#39;H2O&#39;</span>; <span class="charliteral">&#39;H&#39;</span>}), this.compartmentIndexs_cytosol) ...
<a name="l00463"></a>00463                 - this.unaccountedEnergyConsumption * [-1; 1; 1; -1; 1];
<a name="l00464"></a>00464             
<a name="l00465"></a>00465             this.metabolismNewProduction(this.substrateIndexs({<span class="stringliteral">&#39;MET&#39;</span>; <span class="charliteral">&#39;H&#39;</span>; <span class="stringliteral">&#39;H2O&#39;</span>}), this.compartmentIndexs_cytosol) = ...
<a name="l00466"></a>00466                 + this.metabolismNewProduction(this.substrateIndexs({<span class="stringliteral">&#39;MET&#39;</span>; <span class="charliteral">&#39;H&#39;</span>; <span class="stringliteral">&#39;H2O&#39;</span>}), this.compartmentIndexs_cytosol) ...
<a name="l00467"></a>00467                 + [1; 1; -1] .* this.metabolismRecyclingProduction(this.substrateIndexs({<span class="stringliteral">&#39;FMET&#39;</span>}), this.compartmentIndexs_cytosol);
<a name="l00468"></a>00468             this.metabolismRecyclingProduction(this.substrateIndexs({<span class="stringliteral">&#39;MET&#39;</span>; <span class="charliteral">&#39;H&#39;</span>; <span class="stringliteral">&#39;H2O&#39;</span>}), this.compartmentIndexs_cytosol) = ...
<a name="l00469"></a>00469                 + this.metabolismRecyclingProduction(this.substrateIndexs({<span class="stringliteral">&#39;MET&#39;</span>; <span class="charliteral">&#39;H&#39;</span>; <span class="stringliteral">&#39;H2O&#39;</span>}), this.compartmentIndexs_cytosol) ...
<a name="l00470"></a>00470                 - [1; 1; -1] .* this.metabolismRecyclingProduction(this.substrateIndexs({<span class="stringliteral">&#39;FMET&#39;</span>}), this.compartmentIndexs_cytosol);
<a name="l00471"></a>00471             
<a name="l00472"></a>00472             tmp = ...
<a name="l00473"></a>00473                 + this.metabolismNewProduction(this.substrateIndexs({<span class="stringliteral">&#39;FTHF10&#39;</span>; <span class="stringliteral">&#39;FOR&#39;</span>; <span class="stringliteral">&#39;THF&#39;</span>}), this.compartmentIndexs_cytosol) ...
<a name="l00474"></a>00474                 + this.metabolismRecyclingProduction(this.substrateIndexs({<span class="stringliteral">&#39;FTHF10&#39;</span>; <span class="stringliteral">&#39;FOR&#39;</span>; <span class="stringliteral">&#39;THF&#39;</span>}), this.compartmentIndexs_cytosol);
<a name="l00475"></a>00475             this.metabolismNewProduction(this.substrateIndexs({<span class="stringliteral">&#39;FTHF10&#39;</span>; <span class="stringliteral">&#39;FOR&#39;</span>; <span class="stringliteral">&#39;THF&#39;</span>}), this.compartmentIndexs_cytosol) = max(0, ...
<a name="l00476"></a>00476                 + tmp ...
<a name="l00477"></a>00477                 - [1; -1; -1] .* max(0, min(tmp ./ [1; -1; -1])));
<a name="l00478"></a>00478             this.metabolismRecyclingProduction(this.substrateIndexs({<span class="stringliteral">&#39;FTHF10&#39;</span>; <span class="stringliteral">&#39;FOR&#39;</span>; <span class="stringliteral">&#39;THF&#39;</span>}), this.compartmentIndexs_cytosol) = ...
<a name="l00479"></a>00479                 + tmp ...
<a name="l00480"></a>00480                 - this.metabolismNewProduction(this.substrateIndexs({<span class="stringliteral">&#39;FTHF10&#39;</span>; <span class="stringliteral">&#39;FOR&#39;</span>; <span class="stringliteral">&#39;THF&#39;</span>}), this.compartmentIndexs_cytosol);
<a name="l00481"></a>00481             
<a name="l00482"></a>00482             nInternalExchangeReactions = nnz(this.metabolismRecyclingProduction);
<a name="l00483"></a>00483             nInternalExchangeReactionConstraints = 7;
<a name="l00484"></a>00484             
<a name="l00485"></a>00485             %% numbers and indices of components
<a name="l00486"></a>00486             nCompartments = size(this.reactionStoichiometryMatrix, 3);
<a name="l00487"></a>00487             nSubstrates   = size(this.reactionStoichiometryMatrix, 1) + nInternalExchangeReactionConstraints + 1;
<a name="l00488"></a>00488             nEnzymes      = length(this.enzymeWholeCellModelIDs);
<a name="l00489"></a>00489             
<a name="l00490"></a>00490             subIdxs_substrates = (1:size(this.reactionStoichiometryMatrix, 1))<span class="stringliteral">&#39;;</span>
<a name="l00491"></a>00491 <span class="stringliteral">            subIdxs_metaboliteInternalExchangeConstraints = subIdxs_substrates(end) + (1:nInternalExchangeReactionConstraints)&#39;</span>;
<a name="l00492"></a>00492             subIdxs_biomass = nSubstrates;
<a name="l00493"></a>00493             
<a name="l00494"></a>00494             subCmpIdxs_metaboliteInternalExchangeConstraints = sub2ind([nSubstrates nCompartments], ...
<a name="l00495"></a>00495                 subIdxs_metaboliteInternalExchangeConstraints, repmat(this.compartmentIndexs_cytosol, size(subIdxs_metaboliteInternalExchangeConstraints)));
<a name="l00496"></a>00496             subCmpIdxs_biomass = sub2ind([nSubstrates nCompartments], subIdxs_biomass, this.compartmentIndexs_cytosol);
<a name="l00497"></a>00497             
<a name="l00498"></a>00498             nReactions = ...
<a name="l00499"></a>00499                 + size(this.reactionStoichiometryMatrix, 2) ...
<a name="l00500"></a>00500                 + length(subIdxs_substrates) ...
<a name="l00501"></a>00501                 + nInternalExchangeReactions ...
<a name="l00502"></a>00502                 + 2;
<a name="l00503"></a>00503             
<a name="l00504"></a>00504             rxnIdxs_metabolicConversion = (1:size(this.reactionStoichiometryMatrix, 2))<span class="stringliteral">&#39;;</span>
<a name="l00505"></a>00505 <span class="stringliteral">            rxnIdxs_metaboliteExternalExchange = rxnIdxs_metabolicConversion(end) + (1:length(subIdxs_substrates))&#39;</span>;
<a name="l00506"></a>00506             rxnIdxs_metaboliteInternalExchange = rxnIdxs_metaboliteExternalExchange(end) + (1:nInternalExchangeReactions)&#39;;
<a name="l00507"></a>00507             rxnIdxs_biomassProduction = nReactions - 1;
<a name="l00508"></a>00508             rxnIdxs_biomassExchange = nReactions;
<a name="l00509"></a>00509             
<a name="l00510"></a>00510             %% reactions
<a name="l00511"></a>00511             %initialize
<a name="l00512"></a>00512             fbaSMat = zeros(nSubstrates, nReactions, nCompartments);
<a name="l00513"></a>00513             
<a name="l00514"></a>00514             %chemical and transport reactions
<a name="l00515"></a>00515             fbaSMat(subIdxs_substrates, rxnIdxs_metabolicConversion, :) = ...
<a name="l00516"></a>00516                 this.reactionStoichiometryMatrix;
<a name="l00517"></a>00517             
<a name="l00518"></a>00518             %metabolite external exchange
<a name="l00519"></a>00519             fbaSMat(sub2ind(...
<a name="l00520"></a>00520                 [nSubstrates, nReactions, nCompartments],...
<a name="l00521"></a>00521                 subIdxs_substrates,...
<a name="l00522"></a>00522                 rxnIdxs_metaboliteExternalExchange,...
<a name="l00523"></a>00523                 repmat(this.compartmentIndexs_extracellular, length(subIdxs_substrates), 1))) = ...
<a name="l00524"></a>00524                 1;
<a name="l00525"></a>00525             
<a name="l00526"></a>00526             %metabolite <span class="keyword">internal</span> exchange
<a name="l00527"></a>00527             [i, j] = find(this.metabolismRecyclingProduction);
<a name="l00528"></a>00528             fbaSMat(sub2ind(...
<a name="l00529"></a>00529                 [nSubstrates, nReactions, nCompartments],...
<a name="l00530"></a>00530                 i, rxnIdxs_metaboliteInternalExchange, j)) = -1;
<a name="l00531"></a>00531             rxnIdxs_metaboliteInternalLimitedExchange = ...
<a name="l00532"></a>00532                 rxnIdxs_metaboliteInternalExchange(...
<a name="l00533"></a>00533                 this.metabolismRecyclingProduction(this.metabolismRecyclingProduction ~= 0) &lt; 0);
<a name="l00534"></a>00534             rxnIdxs_metaboliteInternalUnlimitedExchange = ...
<a name="l00535"></a>00535                 rxnIdxs_metaboliteInternalExchange(...
<a name="l00536"></a>00536                 this.metabolismRecyclingProduction(this.metabolismRecyclingProduction ~= 0) &gt; 0);
<a name="l00537"></a>00537             
<a name="l00538"></a>00538             [~, ntpIdxs] = ismember(this.substrateIndexs({<span class="stringliteral">&#39;ATP&#39;</span>; <span class="stringliteral">&#39;CTP&#39;</span>; <span class="stringliteral">&#39;GTP&#39;</span>; <span class="stringliteral">&#39;UTP&#39;</span>}), i);
<a name="l00539"></a>00539             [~, ndpIdxs] = ismember(this.substrateIndexs({<span class="stringliteral">&#39;ADP&#39;</span>; <span class="stringliteral">&#39;CDP&#39;</span>; <span class="stringliteral">&#39;GDP&#39;</span>; <span class="stringliteral">&#39;UDP&#39;</span>}), i);
<a name="l00540"></a>00540             [~, nmpIdxs] = ismember(this.substrateIndexs({<span class="stringliteral">&#39;AMP&#39;</span>; <span class="stringliteral">&#39;CMP&#39;</span>; <span class="stringliteral">&#39;GMP&#39;</span>; <span class="stringliteral">&#39;UMP&#39;</span>}), i);
<a name="l00541"></a>00541             [~, fthf10Idxs] = ismember(this.substrateIndexs({<span class="stringliteral">&#39;FTHF10&#39;</span>; <span class="stringliteral">&#39;THF&#39;</span>; <span class="stringliteral">&#39;FOR&#39;</span>}), i);
<a name="l00542"></a>00542             [~, metIdxs] = ismember(this.substrateIndexs({<span class="stringliteral">&#39;MET&#39;</span>; <span class="stringliteral">&#39;FMET&#39;</span>}), i);
<a name="l00543"></a>00543             tmp = zeros(nInternalExchangeReactionConstraints, nInternalExchangeReactions);
<a name="l00544"></a>00544             idxs = [ntpIdxs(1) ndpIdxs(1) nmpIdxs(1)]; tmp(1, idxs(idxs ~= 0)) = 1;
<a name="l00545"></a>00545             idxs = [ntpIdxs(2) ndpIdxs(2) nmpIdxs(2)]; tmp(2, idxs(idxs ~= 0)) = 1;
<a name="l00546"></a>00546             idxs = [ntpIdxs(3) ndpIdxs(3) nmpIdxs(3)]; tmp(3, idxs(idxs ~= 0)) = 1;
<a name="l00547"></a>00547             idxs = [ntpIdxs(4) ndpIdxs(4) nmpIdxs(4)]; tmp(4, idxs(idxs ~= 0)) = 1;
<a name="l00548"></a>00548             idxs = fthf10Idxs(1:2); tmp(5, idxs(idxs ~= 0)) = 1;
<a name="l00549"></a>00549             idxs = [fthf10Idxs([1 3]); metIdxs(2)]; tmp(6, idxs(idxs ~= 0)) = 1;
<a name="l00550"></a>00550             idxs = metIdxs; tmp(7, idxs(idxs ~= 0)) = 1;
<a name="l00551"></a>00551             fbaSMat(subIdxs_metaboliteInternalExchangeConstraints, rxnIdxs_metaboliteInternalExchange, this.compartmentIndexs_cytosol) = ...
<a name="l00552"></a>00552                 + fbaSMat(subIdxs_metaboliteInternalExchangeConstraints, rxnIdxs_metaboliteInternalExchange, this.compartmentIndexs_cytosol) ...
<a name="l00553"></a>00553                 + tmp;
<a name="l00554"></a>00554             
<a name="l00555"></a>00555             %biomass production
<a name="l00556"></a>00556             fbaSMat(subIdxs_substrates, rxnIdxs_biomassProduction, :) = ...
<a name="l00557"></a>00557                 -permute(this.metabolismNewProduction, [1 3 2]);
<a name="l00558"></a>00558             fbaSMat(...
<a name="l00559"></a>00559                 subIdxs_biomass, ...
<a name="l00560"></a>00560                 rxnIdxs_biomassProduction, ...
<a name="l00561"></a>00561                 this.compartmentIndexs_cytosol) = 1;
<a name="l00562"></a>00562             
<a name="l00563"></a>00563             %biomass exchange
<a name="l00564"></a>00564             fbaSMat(...
<a name="l00565"></a>00565                 subIdxs_biomass,...
<a name="l00566"></a>00566                 rxnIdxs_biomassExchange,...
<a name="l00567"></a>00567                 this.compartmentIndexs_cytosol) = 1;
<a name="l00568"></a>00568             
<a name="l00569"></a>00569             %reshape
<a name="l00570"></a>00570             fbaSMatFull = fbaSMat;
<a name="l00571"></a>00571             fbaSMatFull(subIdxs_substrates, rxnIdxs_biomassProduction, :) = ...
<a name="l00572"></a>00572                 -permute(this.metabolismProduction, [1 3 2]);
<a name="l00573"></a>00573             fbaSMat = reshape(...
<a name="l00574"></a>00574                 permute(fbaSMat, [2 1 3]), ...
<a name="l00575"></a>00575                 size(fbaSMat, 2), [])<span class="stringliteral">&#39;;</span>
<a name="l00576"></a>00576 <span class="stringliteral">            fbaSMatFull = reshape(...</span>
<a name="l00577"></a>00577 <span class="stringliteral">                permute(fbaSMatFull, [2 1 3]), ...</span>
<a name="l00578"></a>00578 <span class="stringliteral">                size(fbaSMatFull, 2), [])&#39;</span>;
<a name="l00579"></a>00579             
<a name="l00580"></a>00580             %% reaction catalysis
<a name="l00581"></a>00581             fbaCatMat = zeros(nReactions, nEnzymes);
<a name="l00582"></a>00582             fbaCatMat(rxnIdxs_metabolicConversion, :) = this.reactionCatalysisMatrix;
<a name="l00583"></a>00583             
<a name="l00584"></a>00584             %% reaction bounds
<a name="l00585"></a>00585             fbaRxnBnds = repmat([-Inf Inf], nReactions, 1);
<a name="l00586"></a>00586             
<a name="l00587"></a>00587             this.reactionBounds(this.reactionBounds &lt; 0) = -Inf;
<a name="l00588"></a>00588             this.reactionBounds(this.reactionBounds &gt; 0) =  Inf;
<a name="l00589"></a>00589             fbaRxnBnds(rxnIdxs_metabolicConversion, :) = this.reactionBounds;
<a name="l00590"></a>00590             
<a name="l00591"></a>00591             exchangeLBs = this.substrateExternalExchangeBounds(:, 1);
<a name="l00592"></a>00592             exchangeUBs = this.substrateExternalExchangeBounds(:, 2);
<a name="l00593"></a>00593             fbaRxnBnds(rxnIdxs_metaboliteExternalExchange(~isnan(exchangeLBs)), 1) = exchangeLBs(~isnan(exchangeLBs));
<a name="l00594"></a>00594             fbaRxnBnds(rxnIdxs_metaboliteExternalExchange(~isnan(exchangeUBs)), 2) = exchangeUBs(~isnan(exchangeUBs));
<a name="l00595"></a>00595             
<a name="l00596"></a>00596             fbaRxnBnds(rxnIdxs_metaboliteInternalLimitedExchange, 1) = -Inf;
<a name="l00597"></a>00597             fbaRxnBnds(rxnIdxs_metaboliteInternalLimitedExchange, 2) = 0;
<a name="l00598"></a>00598             fbaRxnBnds(rxnIdxs_metaboliteInternalUnlimitedExchange, 1) = 0;
<a name="l00599"></a>00599             fbaRxnBnds(rxnIdxs_metaboliteInternalUnlimitedExchange, 2) = Inf;
<a name="l00600"></a>00600             
<a name="l00601"></a>00601             fbaRxnBnds(rxnIdxs_biomassProduction, :) = [0 Inf];
<a name="l00602"></a>00602             fbaRxnBnds(rxnIdxs_biomassExchange, :) = [-Inf 0];
<a name="l00603"></a>00603             
<a name="l00604"></a>00604             %% enzyme kinetics
<a name="l00605"></a>00605             fbaEnzBnds = NaN(nReactions, 2);
<a name="l00606"></a>00606             fbaEnzBnds(rxnIdxs_metabolicConversion, :) = this.enzymeBounds;
<a name="l00607"></a>00607             
<a name="l00608"></a>00608             %% right hand side
<a name="l00609"></a>00609             fbaRHS = zeros(nSubstrates * nCompartments, 1);
<a name="l00610"></a>00610             
<a name="l00611"></a>00611             %% objective
<a name="l00612"></a>00612             fbaObj = zeros(nReactions, 1);
<a name="l00613"></a>00613             fbaObj(rxnIdxs_biomassProduction) = 1e3;
<a name="l00614"></a>00614             fbaObj(rxnIdxs_metaboliteInternalLimitedExchange) = 1 / sum(min(0, this.metabolismNewProduction(:)));
<a name="l00615"></a>00615             
<a name="l00616"></a>00616             %% simplify stoichiometry matrix
<a name="l00617"></a>00617             subCmpIdxs_fba = (1:size(fbaSMat, 1))<span class="stringliteral">&#39;;</span>
<a name="l00618"></a>00618 <span class="stringliteral">            rxnIdxs_fba = (1:size(fbaSMat, 2))&#39;</span>;
<a name="l00619"></a>00619             
<a name="l00620"></a>00620             subCmpRemoved = zeros(size(fbaSMat, 1), 3);
<a name="l00621"></a>00621             rxnRemoved = zeros(size(fbaSMat, 2), 3);
<a name="l00622"></a>00622             
<a name="l00623"></a>00623             <span class="keywordflow">if</span> ~exist(<span class="stringliteral">&#39;simplifyNetwork&#39;</span>, <span class="stringliteral">&#39;var&#39;</span>) || simplifyNetwork
<a name="l00624"></a>00624                 iter = 0;
<a name="l00625"></a>00625                 <span class="keywordflow">while</span> <span class="keyword">true</span>
<a name="l00626"></a>00626                     iter = iter + 1;
<a name="l00627"></a>00627                     
<a name="l00628"></a>00628                     %factor
<a name="l00629"></a>00629                     [subCmpBlockIdxs, rxnBlockIdxs] = <a class="code" href="find_non_interacting_rows_and_columns_8m.html#a35f7ca7ac9040917387bac96a24f6b30">findNonInteractingRowsAndColumns</a>(fbaSMatFull(subCmpIdxs_fba, rxnIdxs_fba));
<a name="l00630"></a>00630                     <span class="keywordflow">if</span> all(subCmpBlockIdxs == 1)
<a name="l00631"></a>00631                         break;
<a name="l00632"></a>00632                     end
<a name="l00633"></a>00633                     
<a name="l00634"></a>00634                     blockIdx = subCmpBlockIdxs(subCmpIdxs_fba == subCmpIdxs_biomass);
<a name="l00635"></a>00635                     subCmpRemoved(subCmpIdxs_fba(subCmpBlockIdxs ~= blockIdx), 1) = iter;
<a name="l00636"></a>00636                     rxnRemoved(rxnIdxs_fba(rxnBlockIdxs ~= blockIdx), 1) = iter;
<a name="l00637"></a>00637                     subCmpRemoved(subCmpIdxs_fba(subCmpBlockIdxs ~= blockIdx), 3) = 1;
<a name="l00638"></a>00638                     rxnRemoved(rxnIdxs_fba(rxnBlockIdxs ~= blockIdx), 3) = 1;
<a name="l00639"></a>00639                     subCmpIdxs_fba = subCmpIdxs_fba(subCmpBlockIdxs == blockIdx);
<a name="l00640"></a>00640                     rxnIdxs_fba = rxnIdxs_fba(rxnBlockIdxs == blockIdx);
<a name="l00641"></a>00641                     
<a name="l00642"></a>00642                     %prune
<a name="l00643"></a>00643                     flags = true(3, 1);
<a name="l00644"></a>00644                     iter2 = 0;
<a name="l00645"></a>00645                     <a class="code" href="json_parse_8m.html#ad2fc293e8080616c8c1e2ffbd7b86742">while</a> any(flags)
<a name="l00646"></a>00646                         iter2 = iter2 + 1;
<a name="l00647"></a>00647                         flags = true(3, 1);
<a name="l00648"></a>00648                         
<a name="l00649"></a>00649                         %remov leaf metabolites (and reactions involving them)
<a name="l00650"></a>00650                         %which cannot be balanced with non-zero fluxes
<a name="l00651"></a>00651                         <a class="code" href="json_parse_8m.html#ad2fc293e8080616c8c1e2ffbd7b86742">while</a> flags(1)
<a name="l00652"></a>00652                             tmpSubCmpTfs = sum(fbaSMatFull(subCmpIdxs_fba, rxnIdxs_fba) ~= 0, 2) == 1;
<a name="l00653"></a>00653                             tmpRxnTFs = any(fbaSMatFull(subCmpIdxs_fba(tmpSubCmpTfs), rxnIdxs_fba), 1);
<a name="l00654"></a>00654                             
<a name="l00655"></a>00655                             subCmpRemoved(subCmpIdxs_fba(tmpSubCmpTfs), 1) = iter;
<a name="l00656"></a>00656                             rxnRemoved(rxnIdxs_fba(tmpRxnTFs), 1) = iter;
<a name="l00657"></a>00657                             subCmpRemoved(subCmpIdxs_fba(tmpSubCmpTfs), 2) = iter2;
<a name="l00658"></a>00658                             rxnRemoved(rxnIdxs_fba(tmpRxnTFs), 2) = iter2;
<a name="l00659"></a>00659                             subCmpRemoved(subCmpIdxs_fba(tmpSubCmpTfs), 3) = 2;
<a name="l00660"></a>00660                             rxnRemoved(rxnIdxs_fba(tmpRxnTFs), 3) = 2;
<a name="l00661"></a>00661                             
<a name="l00662"></a>00662                             subCmpIdxs_fba = subCmpIdxs_fba(~tmpSubCmpTfs);
<a name="l00663"></a>00663                             rxnIdxs_fba = rxnIdxs_fba(~tmpRxnTFs);
<a name="l00664"></a>00664                             
<a name="l00665"></a>00665                             flags(1) = any(tmpSubCmpTfs);
<a name="l00666"></a>00666                         end
<a name="l00667"></a>00667                         
<a name="l00668"></a>00668                         %remove metabolites (and reactions involving them)
<a name="l00669"></a>00669                         %which can&#39;t move in both the forward and backward
<a name="l00670"></a>00670                         %directions and thus cannot be balanced
<a name="l00671"></a>00671                         tmpSubCmpTfs = ...
<a name="l00672"></a>00672                             (any(fbaSMatFull(subCmpIdxs_fba, rxnIdxs_fba) .* repmat(fbaRxnBnds(rxnIdxs_fba, 1)&#39;, numel(subCmpIdxs_fba), 1) &lt; 0, 2) | ...
<a name="l00673"></a>00673                             any(fbaSMatFull(subCmpIdxs_fba, rxnIdxs_fba) .* repmat(fbaRxnBnds(rxnIdxs_fba, 2)&#39;, numel(subCmpIdxs_fba), 1) &lt; 0, 2)) &amp; ...
<a name="l00674"></a>00674                             (any(fbaSMatFull(subCmpIdxs_fba, rxnIdxs_fba) .* repmat(fbaRxnBnds(rxnIdxs_fba, 1)&#39;, numel(subCmpIdxs_fba), 1) &gt; 0, 2) | ...
<a name="l00675"></a>00675                             any(fbaSMatFull(subCmpIdxs_fba, rxnIdxs_fba) .* repmat(fbaRxnBnds(rxnIdxs_fba, 2)&#39;, numel(subCmpIdxs_fba), 1) &gt; 0, 2));
<a name="l00676"></a>00676                         tmpRxnTFs = ~any(fbaSMatFull(subCmpIdxs_fba(~tmpSubCmpTfs), rxnIdxs_fba), 1);
<a name="l00677"></a>00677                         
<a name="l00678"></a>00678                         subCmpRemoved(subCmpIdxs_fba(~tmpSubCmpTfs), 1) = iter;
<a name="l00679"></a>00679                         rxnRemoved(rxnIdxs_fba(~tmpRxnTFs), 1) = iter;
<a name="l00680"></a>00680                         subCmpRemoved(subCmpIdxs_fba(~tmpSubCmpTfs), 2) = iter2;
<a name="l00681"></a>00681                         rxnRemoved(rxnIdxs_fba(~tmpRxnTFs), 2) = iter2;
<a name="l00682"></a>00682                         subCmpRemoved(subCmpIdxs_fba(~tmpSubCmpTfs), 3) = 3;
<a name="l00683"></a>00683                         rxnRemoved(rxnIdxs_fba(~tmpRxnTFs), 3) = 3;
<a name="l00684"></a>00684                         
<a name="l00685"></a>00685                         subCmpIdxs_fba = subCmpIdxs_fba(tmpSubCmpTfs);
<a name="l00686"></a>00686                         rxnIdxs_fba = rxnIdxs_fba(tmpRxnTFs);
<a name="l00687"></a>00687                         
<a name="l00688"></a>00688                         flags(2) = ~all(tmpSubCmpTfs);
<a name="l00689"></a>00689                         
<a name="l00690"></a>00690                         %remove other reactions (and substrates unique to them)
<a name="l00691"></a>00691                         %whose flux is zero in all solutions of the matrix
<a name="l00692"></a>00692                         %problem Sv=0. That is reactions which have only zero
<a name="l00693"></a>00693                         %components in the null space of S.
<a name="l00694"></a>00694                         tmpRxnTFs = any(null(fbaSMatFull(subCmpIdxs_fba, rxnIdxs_fba), &#39;r&#39;), 2);
<a name="l00695"></a>00695                         tmpSubCmpTfs = any(fbaSMatFull(subCmpIdxs_fba, rxnIdxs_fba(tmpRxnTFs)), 2);
<a name="l00696"></a>00696                         
<a name="l00697"></a>00697                         subCmpRemoved(subCmpIdxs_fba(~tmpSubCmpTfs), 1) = iter;
<a name="l00698"></a>00698                         rxnRemoved(rxnIdxs_fba(~tmpRxnTFs), 1) = iter;
<a name="l00699"></a>00699                         subCmpRemoved(subCmpIdxs_fba(~tmpSubCmpTfs), 2) = iter2;
<a name="l00700"></a>00700                         rxnRemoved(rxnIdxs_fba(~tmpRxnTFs), 2) = iter2;
<a name="l00701"></a>00701                         subCmpRemoved(subCmpIdxs_fba(~tmpSubCmpTfs), 3) = 4;
<a name="l00702"></a>00702                         rxnRemoved(rxnIdxs_fba(~tmpRxnTFs), 3) = 4;
<a name="l00703"></a>00703                         
<a name="l00704"></a>00704                         subCmpIdxs_fba = subCmpIdxs_fba(tmpSubCmpTfs);
<a name="l00705"></a>00705                         rxnIdxs_fba = rxnIdxs_fba(tmpRxnTFs);
<a name="l00706"></a>00706                         
<a name="l00707"></a>00707                         flags(3) = ~all(tmpSubCmpTfs);
<a name="l00708"></a>00708                     end
<a name="l00709"></a>00709                 end
<a name="l00710"></a>00710             end
<a name="l00711"></a>00711             
<a name="l00712"></a>00712             subCmpIdxs_fba = unique([subCmpIdxs_fba; subCmpIdxs_metaboliteInternalExchangeConstraints]);
<a name="l00713"></a>00713             subCmpRemoved(subCmpIdxs_metaboliteInternalExchangeConstraints, :) = 0;
<a name="l00714"></a>00714             
<a name="l00715"></a>00715             subCmpRemoved = subCmpRemoved(reshape(sub2ind([nSubstrates nCompartments], ...
<a name="l00716"></a>00716                 repmat(subIdxs_substrates, 1, nCompartments), ...
<a name="l00717"></a>00717                 repmat(1:nCompartments, numel(subIdxs_substrates), 1)), [], 1), :);
<a name="l00718"></a>00718             rxnRemoved = rxnRemoved(rxnIdxs_metabolicConversion, :);
<a name="l00719"></a>00719             
<a name="l00720"></a>00720             %substrate indices
<a name="l00721"></a>00721             [subIdxs_fba, cmpIdxs_fba] = ind2sub([nSubstrates, nCompartments], subCmpIdxs_fba);
<a name="l00722"></a>00722             this.fbaSubstrateIndexs_substrates = find(ismember(subIdxs_fba, subIdxs_substrates));
<a name="l00723"></a>00723             this.fbaSubstrateIndexs_metaboliteInternalExchangeConstraints = find(ismember(subCmpIdxs_fba, subCmpIdxs_metaboliteInternalExchangeConstraints));
<a name="l00724"></a>00724             this.substrateIndexs_fba = sub2ind(...
<a name="l00725"></a>00725                 [numel(this.substrateWholeCellModelIDs) nCompartments], ...
<a name="l00726"></a>00726                 subIdxs_fba(this.fbaSubstrateIndexs_substrates), ...
<a name="l00727"></a>00727                 cmpIdxs_fba(this.fbaSubstrateIndexs_substrates));
<a name="l00728"></a>00728             [tfs, idxs] = ismember(rxnIdxs_fba, rxnIdxs_metaboliteExternalExchange);
<a name="l00729"></a>00729             this.substrateIndexs_externalExchangedMetabolites = idxs(tfs);
<a name="l00730"></a>00730             this.substrateIndexs_internalExchangedMetabolites = ...
<a name="l00731"></a>00731                 find(this.metabolismRecyclingProduction);
<a name="l00732"></a>00732             this.substrateIndexs_internalExchangedLimitedMetabolites = ...
<a name="l00733"></a>00733                 find(this.metabolismRecyclingProduction &lt; 0);
<a name="l00734"></a>00734             
<a name="l00735"></a>00735             this.fbaSubstrateIndexs_biomass = find(ismember(subCmpIdxs_fba, subCmpIdxs_biomass));
<a name="l00736"></a>00736             
<a name="l00737"></a>00737             %reaction indices
<a name="l00738"></a>00738             assert(all(ismember(rxnIdxs_metaboliteInternalLimitedExchange, rxnIdxs_fba)))
<a name="l00739"></a>00739             
<a name="l00740"></a>00740             this.reactionIndexs_fba = find(ismember(rxnIdxs_metabolicConversion, rxnIdxs_fba));
<a name="l00741"></a>00741             this.fbaReactionIndexs_metabolicConversion = find(ismember(rxnIdxs_fba, rxnIdxs_metabolicConversion));
<a name="l00742"></a>00742             this.fbaReactionIndexs_metaboliteExternalExchange = find(ismember(rxnIdxs_fba, rxnIdxs_metaboliteExternalExchange));
<a name="l00743"></a>00743             this.fbaReactionIndexs_metaboliteInternalExchange = find(ismember(rxnIdxs_fba, rxnIdxs_metaboliteInternalExchange));
<a name="l00744"></a>00744             this.fbaReactionIndexs_metaboliteInternalLimitedExchange = find(ismember(rxnIdxs_fba, rxnIdxs_metaboliteInternalLimitedExchange));
<a name="l00745"></a>00745             this.fbaReactionIndexs_metaboliteInternalUnlimitedExchange = find(ismember(rxnIdxs_fba, rxnIdxs_metaboliteInternalUnlimitedExchange));
<a name="l00746"></a>00746             this.fbaReactionIndexs_biomassProduction = find(ismember(rxnIdxs_fba, rxnIdxs_biomassProduction));
<a name="l00747"></a>00747             this.fbaReactionIndexs_biomassExchange = find(ismember(rxnIdxs_fba, rxnIdxs_biomassExchange));
<a name="l00748"></a>00748             
<a name="l00749"></a>00749             fbaSMat = fbaSMat(subCmpIdxs_fba, rxnIdxs_fba);
<a name="l00750"></a>00750             fbaCatMat = fbaCatMat(rxnIdxs_fba, :);
<a name="l00751"></a>00751             fbaRHS = fbaRHS(subCmpIdxs_fba, :);
<a name="l00752"></a>00752             fbaRxnBnds = fbaRxnBnds(rxnIdxs_fba, :);
<a name="l00753"></a>00753             fbaEnzBnds = fbaEnzBnds(rxnIdxs_fba, :);
<a name="l00754"></a>00754             fbaObj = fbaObj(rxnIdxs_fba, :);
<a name="l00755"></a>00755             
<a name="l00756"></a>00756             %% set <a class="code" href="json_parse_8m.html#a6e4256f69b751d3d7ec4ea52ff5a8dee">properties</a>
<a name="l00757"></a>00757             this.fbaReactionStoichiometryMatrix = fbaSMat;
<a name="l00758"></a>00758             this.fbaReactionCatalysisMatrix = fbaCatMat;
<a name="l00759"></a>00759             this.fbaRightHandSide = fbaRHS;
<a name="l00760"></a>00760             this.fbaReactionBounds = fbaRxnBnds;
<a name="l00761"></a>00761             this.fbaEnzymeBounds = fbaEnzBnds;
<a name="l00762"></a>00762             this.fbaObjective = fbaObj;
<a name="l00763"></a>00763         end
<a name="l00764"></a>00764     end
<a name="l00765"></a>00765     
<a name="l00766"></a>00766     %model
<a name="l00767"></a>00767     methods
<a name="l00768"></a>00768         %Calculate
<a name="l00769"></a>00769         %- contribution to <a class="code" href="class_f_b_a.html">FBA</a> objective
<a name="l00770"></a>00770         %- minimum expression consistent with cell cycle length
<a name="l00771"></a>00771         function [bmProd, byProd, minEnzExp, maxEnzExp] = calcResourceRequirements_LifeCycle(this, ~, states)
<a name="l00772"></a>00772             %% import classes
<a name="l00773"></a>00773             import edu.stanford.covert.util.<a class="code" href="class_computation_util.html">ComputationUtil</a>;
<a name="l00774"></a>00774             
<a name="l00775"></a>00775             %% substrate and byproducts: none used by metabolism
<a name="l00776"></a>00776             bmProd = zeros(size(this.substrateWholeCellModelIDs));
<a name="l00777"></a>00777             byProd = zeros(size(this.substrateWholeCellModelIDs));
<a name="l00778"></a>00778             
<a name="l00779"></a>00779             %% fit enzyme expression consistent with growth rate
<a name="l00780"></a>00780             if strcmp(this.macromoleculeStateInitialization, &#39;expected&#39;)
<a name="l00781"></a>00781                 growth0 = this.metabolicReaction.growth0;
<a name="l00782"></a>00782             else
<a name="l00783"></a>00783                 growth0 = this.metabolicReaction.growth0 * this.macromoleculeStateInitializationGrowthFactor;
<a name="l00784"></a>00784             end
<a name="l00785"></a>00785             
<a name="l00786"></a>00786             substrates = zeros(size(this.substrates));
<a name="l00787"></a>00787             substrates(this.substrateMetaboliteLocalIndexs, :) = this.substrates(this.substrateMetaboliteLocalIndexs, :);
<a name="l00788"></a>00788             substrates(this.substrateIndexs_internalExchangedLimitedMetabolites) = ...
<a name="l00789"></a>00789                 -growth0 *  this.metabolismRecyclingProduction(this.substrateIndexs_internalExchangedLimitedMetabolites);
<a name="l00790"></a>00790             substrates(this.substrateMonomerLocalIndexs, :) = repmat(states.monomers0(this.substrateMonomerGlobalIndexs(:, 1)), 1, size(substrates, 2));
<a name="l00791"></a>00791             substrates(this.substrateComplexLocalIndexs, :) = repmat(states.complexs0(this.substrateComplexGlobalIndexs(:, 1)), 1, size(substrates, 2));
<a name="l00792"></a>00792             
<a name="l00793"></a>00793             enzymes = zeros(size(this.enzymeWholeCellModelIDs));
<a name="l00794"></a>00794             enzymes(this.enzymeMonomerLocalIndexs) = states.monomers0(this.enzymeMonomerGlobalIndexs);
<a name="l00795"></a>00795             enzymes(this.enzymeComplexLocalIndexs) = states.complexs0(this.enzymeComplexGlobalIndexs);
<a name="l00796"></a>00796             
<a name="l00797"></a>00797             enzymes = this.fitEnzymes(growth0, substrates, enzymes);
<a name="l00798"></a>00798             
<a name="l00799"></a>00799             %% metabolic enzyme levels constistent with observed growth rate
<a name="l00800"></a>00800             [minEnzExp, maxEnzExp] = this.calcMinMaxEnzymes(substrates, enzymes);
<a name="l00801"></a>00801         end
<a name="l00802"></a>00802         
<a name="l00803"></a>00803         %calculate enzyme expression consistent with growth rate
<a name="l00804"></a>00804         %- by minimizing the fractional change in fluxes by linear
<a name="l00805"></a>00805         %  optimization (similar to MOMA)
<a name="l00806"></a>00806         %- if linear optimization fails, increase/decrease all flux
<a name="l00807"></a>00807         %  bounds by a factor of growth/growth0 and check if this yields
<a name="l00808"></a>00808         %  a growth rate of growth0
<a name="l00809"></a>00809         function enzymes = fitEnzymes(this, growth0, substrates, enzymes)
<a name="l00810"></a>00810             % import classes
<a name="l00811"></a>00811             import edu.stanford.covert.util.ComputationUtil;
<a name="l00812"></a>00812             import edu.stanford.covert.util.<a class="code" href="class_constant_util.html">ConstantUtil</a>;
<a name="l00813"></a>00813             
<a name="l00814"></a>00814             %constants
<a name="l00815"></a>00815             nEnzymes = numel(enzymes);
<a name="l00816"></a>00816             enzMWs = this.enzymeMolecularWeights;
<a name="l00817"></a>00817             
<a name="l00818"></a>00818             minAvgExp = zeros(size(enzymes));
<a name="l00819"></a>00819             minAvgExp(this.enzymeMonomerLocalIndexs) = this.monomer.minimumAverageExpression;
<a name="l00820"></a>00820             minAvgExp(this.enzymeComplexLocalIndexs) = this.complex.minimumAverageExpression;
<a name="l00821"></a>00821             
<a name="l00822"></a>00822             loIdxs = find(enzymes &lt; minAvgExp &amp; any(this.fbaReactionCatalysisMatrix, 1)&#39;);
<a name="l00823"></a>00823             hiIdxs = find(enzymes &gt; minAvgExp);
<a name="l00824"></a>00824             if (minAvgExp(loIdxs)-enzymes(loIdxs))&#39; * enzMWs(loIdxs) &lt; (enzymes(hiIdxs) - minAvgExp(hiIdxs))&#39; * enzMWs(hiIdxs)
<a name="l00825"></a>00825                 enzymes(hiIdxs) = minAvgExp(hiIdxs) + (enzymes(hiIdxs) - minAvgExp(hiIdxs)) * ...
<a name="l00826"></a>00826                     ((enzymes(hiIdxs)-minAvgExp(hiIdxs))&#39; * enzMWs(hiIdxs) - ...
<a name="l00827"></a>00827                     (minAvgExp(loIdxs)-enzymes(loIdxs))&#39; * enzMWs(loIdxs)) / ...
<a name="l00828"></a>00828                     ((enzymes(hiIdxs)-minAvgExp(hiIdxs))&#39; * enzMWs(hiIdxs));
<a name="l00829"></a>00829                 enzymes(loIdxs) = minAvgExp(loIdxs);
<a name="l00830"></a>00830             else
<a name="l00831"></a>00831                 warning(&#39;WholeCell:warning&#39;, &#39;Cannot ensure minimum enzyme expression&#39;);
<a name="l00832"></a>00832             end
<a name="l00833"></a>00833             
<a name="l00834"></a>00834             initEnzymes = enzymes;
<a name="l00835"></a>00835             
<a name="l00836"></a>00836             % calculate current growth and fluxes
<a name="l00837"></a>00837             [fbaObj, fbaSMat, fbaRxnBounds] = this.calcEffectiveFBANetwork();
<a name="l00838"></a>00838             fluxBounds = this.calcFluxBounds(substrates, enzymes, fbaRxnBounds, this.fbaEnzymeBounds);
<a name="l00839"></a>00839             [growth, ~, oldFbaReactionFluxs] = this.calcGrowthRate(fluxBounds, fbaObj, fbaSMat);
<a name="l00840"></a>00840             
<a name="l00841"></a>00841             if abs(growth - growth0) / growth0 &lt; this.tolerance || abs(growth - growth0) &lt; 1<a class="code" href="json_parse_8m.html#ac14d26badced2970320ee4d772ad0ae9">e</a>-12
<a name="l00842"></a>00842                 return;
<a name="l00843"></a>00843             end
<a name="l00844"></a>00844             
<a name="l00845"></a>00845             if growth &lt; growth0 &amp;&amp; growth0 &gt; this.calcGrowthRate(this.calcFluxBounds(...
<a name="l00846"></a>00846                     substrates, enzymes, fbaRxnBounds, this.fbaEnzymeBounds, false), fbaObj, fbaSMat)
<a name="l00847"></a>00847                 throw(MException(&#39;Metabolism:error&#39;, &#39;Cannot fit enzyme expression to match growth rate&#39;));
<a name="l00848"></a>00848             end
<a name="l00849"></a>00849             
<a name="l00850"></a>00850             % reactions constrainted by enzyme expression
<a name="l00851"></a>00851             [forIdxs, revIdxs] = this.calcEnzymeKineticallyLimitedReactions(substrates, enzymes, fbaObj, fbaSMat, fbaRxnBounds);
<a name="l00852"></a>00852             unconstrainedEnzIdxs = find(~any(this.fbaReactionCatalysisMatrix([forIdxs; revIdxs], :), 1)&#39;);
<a name="l00853"></a>00853             
<a name="l00854"></a>00854             % minimize difference from current enzyme expression
<a name="l00855"></a>00855             objectiveFunc = zeros(size(fbaObj));
<a name="l00856"></a>00856             objectiveFunc([forIdxs; revIdxs]) = 1 ./ oldFbaReactionFluxs([forIdxs; revIdxs]);
<a name="l00857"></a>00857             
<a name="l00858"></a>00858             % minimum enzyme expression
<a name="l00859"></a>00859             loFluxBounds = fluxBounds(:, 1);
<a name="l00860"></a>00860             upFluxBounds = fluxBounds(:, 2);
<a name="l00861"></a>00861             
<a name="l00862"></a>00862             if growth &lt; growth0
<a name="l00863"></a>00863                 tmpFluxBounds = this.calcFluxBounds(...
<a name="l00864"></a>00864                     substrates, enzymes, fbaRxnBounds, this.fbaEnzymeBounds, false);
<a name="l00865"></a>00865                 loFluxBounds(forIdxs) = oldFbaReactionFluxs(forIdxs);
<a name="l00866"></a>00866                 upFluxBounds(forIdxs) = min(tmpFluxBounds(forIdxs, 2), oldFbaReactionFluxs(forIdxs) * (1 + this.tolerance) * growth0 / growth);
<a name="l00867"></a>00867                 loFluxBounds(revIdxs) = max(tmpFluxBounds(revIdxs, 1), oldFbaReactionFluxs(revIdxs) * (1 + this.tolerance) * growth0 / growth);
<a name="l00868"></a>00868                 upFluxBounds(revIdxs) = oldFbaReactionFluxs(revIdxs);
<a name="l00869"></a>00869                 
<a name="l00870"></a>00870                 objective = &#39;minimize&#39;;
<a name="l00871"></a>00871             else
<a name="l00872"></a>00872                 loFluxBounds(forIdxs) = 0;
<a name="l00873"></a>00873                 upFluxBounds(forIdxs) = oldFbaReactionFluxs(forIdxs);
<a name="l00874"></a>00874                 loFluxBounds(revIdxs) = oldFbaReactionFluxs(revIdxs);
<a name="l00875"></a>00875                 upFluxBounds(revIdxs) = 0;
<a name="l00876"></a>00876                 
<a name="l00877"></a>00877                 objective = &#39;maximize&#39;;
<a name="l00878"></a>00878             end
<a name="l00879"></a>00879             
<a name="l00880"></a>00880             loFluxBounds(this.fbaReactionIndexs_biomassProduction) = growth0;
<a name="l00881"></a>00881             upFluxBounds(this.fbaReactionIndexs_biomassProduction) = growth0;
<a name="l00882"></a>00882             
<a name="l00883"></a>00883             loFluxBounds = max(loFluxBounds, -this.realmax);
<a name="l00884"></a>00884             upFluxBounds = min(upFluxBounds,  this.realmax);
<a name="l00885"></a>00885             [fbaReactionFluxs, ~, ~, errFlag, errMsg] = ComputationUtil.linearProgramming(...
<a name="l00886"></a>00886                 objective, objectiveFunc, fbaSMat, ...
<a name="l00887"></a>00887                 this.fbaRightHandSide, loFluxBounds, upFluxBounds, ...
<a name="l00888"></a>00888                 &#39;S&#39;, &#39;C&#39;, this.linearProgrammingOptions);
<a name="l00889"></a>00889             if errFlag
<a name="l00890"></a>00890                 throw(Exception(&#39;Metabolism:error&#39;, &#39;Unable to optimize fluxes: %s&#39;, errMsg));
<a name="l00891"></a>00891             end
<a name="l00892"></a>00892             
<a name="l00893"></a>00893             %adjust enzyme expression
<a name="l00894"></a>00894             fbaReactionFluxs = max(min(fbaReactionFluxs, upFluxBounds), loFluxBounds);
<a name="l00895"></a>00895             
<a name="l00896"></a>00896             reactionEnzymes = zeros(size(fbaReactionFluxs));
<a name="l00897"></a>00897             reactionEnzymes(forIdxs) = fbaReactionFluxs(forIdxs) ./ this.fbaEnzymeBounds(forIdxs, 2);
<a name="l00898"></a>00898             reactionEnzymes(revIdxs) = fbaReactionFluxs(revIdxs) ./ this.fbaEnzymeBounds(revIdxs, 1);
<a name="l00899"></a>00899             
<a name="l00900"></a>00900             enzymes = max(this.fbaReactionCatalysisMatrix([forIdxs; revIdxs], :) .* ...
<a name="l00901"></a>00901                 repmat(reactionEnzymes([forIdxs; revIdxs]), 1, nEnzymes), [], 1)&#39;;
<a name="l00902"></a>00902             enzymes(unconstrainedEnzIdxs) = initEnzymes(unconstrainedEnzIdxs);
<a name="l00903"></a>00903             enzymes = this.adjustEnzymeExpressionSparsely(growth0, substrates, enzymes, initEnzymes);
<a name="l00904"></a>00904             enzymes = max(minAvgExp, enzymes);
<a name="l00905"></a>00905             
<a name="l00906"></a>00906             if growth &lt; growth0
<a name="l00907"></a>00907                 incEnzymes = find(enzymes &gt; initEnzymes);
<a name="l00908"></a>00908                 decEnzymes = setdiff(unique([find(enzymes &lt; initEnzymes); unconstrainedEnzIdxs]), incEnzymes);
<a name="l00909"></a>00909                 enzymes = max(initEnzymes, enzymes);
<a name="l00910"></a>00910                 enzymes(decEnzymes) = minAvgExp(decEnzymes) + (initEnzymes(decEnzymes) - minAvgExp(decEnzymes)) * ...
<a name="l00911"></a>00911                     ((initEnzymes(decEnzymes) - minAvgExp(decEnzymes))&#39; * enzMWs(decEnzymes) - ...
<a name="l00912"></a>00912                     (enzymes(incEnzymes) - initEnzymes(incEnzymes))&#39; * enzMWs(incEnzymes)) / ...
<a name="l00913"></a>00913                     ((initEnzymes(decEnzymes) - minAvgExp(decEnzymes))&#39; * enzMWs(decEnzymes));
<a name="l00914"></a>00914             else
<a name="l00915"></a>00915                 incEnzymes = unique([find(enzymes &gt; initEnzymes); unconstrainedEnzIdxs]);
<a name="l00916"></a>00916                 decEnzymes = setdiff(find(enzymes &lt; initEnzymes), incEnzymes);
<a name="l00917"></a>00917                 enzymes = min(initEnzymes, enzymes);
<a name="l00918"></a>00918                 enzymes(incEnzymes) = minAvgExp(incEnzymes) + (initEnzymes(incEnzymes) - minAvgExp(incEnzymes)) * ...
<a name="l00919"></a>00919                     ((initEnzymes(incEnzymes) - minAvgExp(incEnzymes))&#39; * enzMWs(incEnzymes) - ...
<a name="l00920"></a>00920                     (enzymes(decEnzymes) - initEnzymes(decEnzymes))&#39; * enzMWs(decEnzymes)) / ...
<a name="l00921"></a>00921                     ((initEnzymes(incEnzymes) - minAvgExp(incEnzymes))&#39; * enzMWs(incEnzymes));
<a name="l00922"></a>00922             end
<a name="l00923"></a>00923             
<a name="l00924"></a>00924             %check growth rate fit
<a name="l00925"></a>00925             if abs(this.calcGrowthRate(this.calcFluxBounds(...
<a name="l00926"></a>00926                     substrates, enzymes, fbaRxnBounds, this.fbaEnzymeBounds), fbaObj, fbaSMat) ...
<a name="l00927"></a>00927                     - growth0) / growth0 &gt; this.tolerance
<a name="l00928"></a>00928                 throw(MException(&#39;Metabolism:error&#39;, &#39;Growth rate should not change&#39;));
<a name="l00929"></a>00929             end
<a name="l00930"></a>00930             if abs((enzymes&#39; * enzMWs - initEnzymes&#39; * enzMWs) / (initEnzymes&#39; * enzMWs)) &gt; this.tolerance
<a name="l00931"></a>00931                 throw(MException(&#39;Metabolism:error&#39;, &#39;<a class="code" href="class_protein.html">Protein</a> mass should not change&#39;));
<a name="l00932"></a>00932             end
<a name="l00933"></a>00933         end
<a name="l00934"></a>00934         
<a name="l00935"></a>00935         function enzymes = adjustEnzymeExpressionSparsely(this, growth0, substrates, enzymes, initEnzymes)
<a name="l00936"></a>00936             [fbaObj, fbaSMat, fbaRxnBounds] = this.calcEffectiveFBANetwork();
<a name="l00937"></a>00937             
<a name="l00938"></a>00938             tfs = this.adjustEnzymeExpressionSparselyHelper(growth0, substrates, enzymes, initEnzymes, true(size(enzymes)));
<a name="l00939"></a>00939             enzymes(tfs) = initEnzymes(tfs);
<a name="l00940"></a>00940             idxs = find(~tfs);
<a name="l00941"></a>00941             
<a name="l00942"></a>00942             %one at a time
<a name="l00943"></a>00943             tf = false;
<a name="l00944"></a>00944             for i = 1:numel(idxs)
<a name="l00945"></a>00945                 tmpEnzymes = enzymes;
<a name="l00946"></a>00946                 tmpEnzymes(idxs(i)) = initEnzymes(idxs(i));
<a name="l00947"></a>00947                 
<a name="l00948"></a>00948                 if abs(this.calcGrowthRate(this.calcFluxBounds(...
<a name="l00949"></a>00949                         substrates, tmpEnzymes, fbaRxnBounds, this.fbaEnzymeBounds), fbaObj, fbaSMat) ...
<a name="l00950"></a>00950                         - growth0) / growth0 &lt;= this.tolerance
<a name="l00951"></a>00951                     tf = true;
<a name="l00952"></a>00952                     break;
<a name="l00953"></a>00953                 end
<a name="l00954"></a>00954             end
<a name="l00955"></a>00955             if ~tf
<a name="l00956"></a>00956                 return;
<a name="l00957"></a>00957             end
<a name="l00958"></a>00958             
<a name="l00959"></a>00959             %combinations
<a name="l00960"></a>00960             try
<a name="l00961"></a>00961                 tmpTfs = dec2bin(0:2^numel(idxs)-1) == &#39;1&#39;;
<a name="l00962"></a>00962             catch %<span class="preprocessor">#ok&lt;CTCH&gt;</span>
<a name="l00963"></a>00963 <span class="preprocessor"></span>                <span class="keywordflow">return</span>;
<a name="l00964"></a>00964             end
<a name="l00965"></a>00965             [~, order] = sort(sum(tmpTfs, 2), 1, <span class="stringliteral">&#39;descend&#39;</span>);
<a name="l00966"></a>00966             tmpTfs = tmpTfs(order, :);
<a name="l00967"></a>00967             
<a name="l00968"></a>00968             <span class="keywordflow">for</span> i = 1:2^numel(idxs)
<a name="l00969"></a>00969                 tmpEnzymes = enzymes;
<a name="l00970"></a>00970                 tmpEnzymes(idxs(tmpTfs(i, :))) = initEnzymes(idxs(tmpTfs(i, :)));
<a name="l00971"></a>00971                 
<a name="l00972"></a>00972                 if abs(this.calcGrowthRate(this.calcFluxBounds(...
<a name="l00973"></a>00973                         substrates, tmpEnzymes, fbaRxnBounds, this.fbaEnzymeBounds), fbaObj, fbaSMat) ...
<a name="l00974"></a>00974                         - growth0) / growth0 &lt;= this.tolerance
<a name="l00975"></a>00975                     enzymes = tmpEnzymes;
<a name="l00976"></a>00976                     return;
<a name="l00977"></a>00977                 end
<a name="l00978"></a>00978             end
<a name="l00979"></a>00979         end
<a name="l00980"></a>00980         
<a name="l00981"></a>00981         function tfs = adjustEnzymeExpressionSparselyHelper(this, growth0, substrates, enzymes, initEnzymes, tfs)
<a name="l00982"></a>00982             [fbaObj, fbaSMat, fbaRxnBounds] = this.calcEffectiveFBANetwork();
<a name="l00983"></a>00983             
<a name="l00984"></a>00984             tmpEnzymes = enzymes;
<a name="l00985"></a>00985             tmpEnzymes(tfs) = initEnzymes(tfs);
<a name="l00986"></a>00986             
<a name="l00987"></a>00987             if sum(tfs) == 0 || abs(this.calcGrowthRate(this.calcFluxBounds(...
<a name="l00988"></a>00988                     substrates, tmpEnzymes, fbaRxnBounds, this.fbaEnzymeBounds), fbaObj, fbaSMat) ...
<a name="l00989"></a>00989                     - growth0) / growth0 &lt;= this.tolerance
<a name="l00990"></a>00990                 return;
<a name="l00991"></a>00991             elseif sum(tfs) == 1
<a name="l00992"></a>00992                 tfs(:) = false;
<a name="l00993"></a>00993                 return;
<a name="l00994"></a>00994             end
<a name="l00995"></a>00995             
<a name="l00996"></a>00996             idxs = find(tfs);
<a name="l00997"></a>00997             tfs1 = tfs;
<a name="l00998"></a>00998             tfs2 = tfs;
<a name="l00999"></a>00999             tfs1(idxs(1:ceil(end/2))) = false;
<a name="l01000"></a>01000             tfs2(idxs(ceil(end/2))+1:end) = false;
<a name="l01001"></a>01001             tfs = ...
<a name="l01002"></a>01002                 this.adjustEnzymeExpressionSparselyHelper(growth0, substrates, enzymes, initEnzymes, tfs1) | ...
<a name="l01003"></a>01003                 this.adjustEnzymeExpressionSparselyHelper(growth0, substrates, enzymes, initEnzymes, tfs2);
<a name="l01004"></a>01004         end
<a name="l01005"></a>01005         
<a name="l01006"></a>01006         %calculate minimum and maximum enzyme weight consistent with growth
<a name="l01007"></a>01007         %rate
<a name="l01008"></a>01008         %- by minimizing/maximizing the total enzyme weight consistent with
<a name="l01009"></a>01009         %  the current growth rate by linear optimization
<a name="l01010"></a>01010         %- if linear optimization fails, minEnzymes/maxEnzymes is set to
<a name="l01011"></a>01011         %  the current enyzme level
<a name="l01012"></a>01012         function [minEnzymes, maxEnzymes] = calcMinMaxEnzymes(this, substrates, enzymes)
<a name="l01013"></a>01013             % import classes
<a name="l01014"></a>01014             import edu.stanford.covert.util.ComputationUtil;
<a name="l01015"></a>01015             import edu.stanford.covert.util.ConstantUtil;
<a name="l01016"></a>01016             
<a name="l01017"></a>01017             %constants
<a name="l01018"></a>01018             nEnzymes = numel(enzymes);
<a name="l01019"></a>01019             
<a name="l01020"></a>01020             minAvgExp = zeros(size(enzymes));
<a name="l01021"></a>01021             minAvgExp(this.enzymeMonomerLocalIndexs) = this.monomer.minimumAverageExpression;
<a name="l01022"></a>01022             minAvgExp(this.enzymeComplexLocalIndexs) = this.complex.minimumAverageExpression;
<a name="l01023"></a>01023             
<a name="l01024"></a>01024             % calculate current growth and fluxes
<a name="l01025"></a>01025             [fbaObj, fbaSMat, fbaRxnBounds] = this.calcEffectiveFBANetwork();
<a name="l01026"></a>01026             fluxBounds = this.calcFluxBounds(...
<a name="l01027"></a>01027                 substrates, enzymes, fbaRxnBounds, this.fbaEnzymeBounds);
<a name="l01028"></a>01028             [growth, ~, oldFbaReactionFluxs] = this.calcGrowthRate(fluxBounds, fbaObj, fbaSMat);
<a name="l01029"></a>01029             
<a name="l01030"></a>01030             % reactions constrainted by enzyme expression
<a name="l01031"></a>01031             [forIdxs, revIdxs] = this.calcEnzymeKineticallyLimitedReactions(substrates, enzymes, fbaObj, fbaSMat, fbaRxnBounds);
<a name="l01032"></a>01032             partlyUnconstrainedEnzIdxs = find(any(this.fbaReactionCatalysisMatrix(setdiff(1:end, [forIdxs; revIdxs]), :), 1));
<a name="l01033"></a>01033             unconstrainedEnzIdxs = find(~any(this.fbaReactionCatalysisMatrix([forIdxs; revIdxs], :), 1));
<a name="l01034"></a>01034             
<a name="l01035"></a>01035             if isempty(forIdxs) &amp;&amp; isempty(revIdxs)
<a name="l01036"></a>01036                 warning(&#39;WholeCell:warning&#39;, &#39;Unable to fit growth rate&#39;);
<a name="l01037"></a>01037                 minEnzymes = enzymes;
<a name="l01038"></a>01038                 maxEnzymes = enzymes;
<a name="l01039"></a>01039                 return;
<a name="l01040"></a>01040             end
<a name="l01041"></a>01041             
<a name="l01042"></a>01042             % enzyme expression range
<a name="l01043"></a>01043             objectiveFunc = zeros(size(fbaObj));
<a name="l01044"></a>01044             objectiveFunc(forIdxs) = (this.fbaReactionCatalysisMatrix(forIdxs, :) * this.enzymeMolecularWeights) ./ this.fbaEnzymeBounds(forIdxs, 2);
<a name="l01045"></a>01045             objectiveFunc(revIdxs) = (this.fbaReactionCatalysisMatrix(revIdxs, :) * this.enzymeMolecularWeights) ./ this.fbaEnzymeBounds(revIdxs, 1);
<a name="l01046"></a>01046             
<a name="l01047"></a>01047             % minimum enzyme expression
<a name="l01048"></a>01048             loFluxBounds = fluxBounds(:, 1);
<a name="l01049"></a>01049             upFluxBounds = fluxBounds(:, 2);
<a name="l01050"></a>01050             
<a name="l01051"></a>01051             loFluxBounds(this.fbaReactionIndexs_biomassProduction) = growth;
<a name="l01052"></a>01052             upFluxBounds(this.fbaReactionIndexs_biomassProduction) = growth;
<a name="l01053"></a>01053             
<a name="l01054"></a>01054             loFluxBounds(forIdxs) = 0;
<a name="l01055"></a>01055             upFluxBounds(forIdxs) = oldFbaReactionFluxs(forIdxs);
<a name="l01056"></a>01056             loFluxBounds(revIdxs) = oldFbaReactionFluxs(revIdxs);
<a name="l01057"></a>01057             upFluxBounds(revIdxs) = 0;
<a name="l01058"></a>01058             
<a name="l01059"></a>01059             loFluxBounds = max(loFluxBounds, -this.realmax);
<a name="l01060"></a>01060             upFluxBounds = min(upFluxBounds,  this.realmax);
<a name="l01061"></a>01061             [fbaReactionFluxs, ~, ~, errFlag, errMsg] = ComputationUtil.linearProgramming(...
<a name="l01062"></a>01062                 &#39;minimize&#39;, objectiveFunc, fbaSMat, ...
<a name="l01063"></a>01063                 this.fbaRightHandSide, loFluxBounds, upFluxBounds, ...
<a name="l01064"></a>01064                 &#39;S&#39;, &#39;C&#39;, this.linearProgrammingOptions);
<a name="l01065"></a>01065             if errFlag
<a name="l01066"></a>01066                 warning(&#39;WholeCell:warning&#39;, errMsg);
<a name="l01067"></a>01067                 minEnzymes = enzymes;
<a name="l01068"></a>01068             else
<a name="l01069"></a>01069                 fbaReactionFluxs = max(min(fbaReactionFluxs, upFluxBounds), loFluxBounds);
<a name="l01070"></a>01070                 
<a name="l01071"></a>01071                 reactionEnzymes = zeros(size(oldFbaReactionFluxs));
<a name="l01072"></a>01072                 reactionEnzymes(forIdxs) = fbaReactionFluxs(forIdxs) ./ this.fbaEnzymeBounds(forIdxs, 2);
<a name="l01073"></a>01073                 reactionEnzymes(revIdxs) = fbaReactionFluxs(revIdxs) ./ this.fbaEnzymeBounds(revIdxs, 1);
<a name="l01074"></a>01074                 
<a name="l01075"></a>01075                 minEnzymes = max(this.fbaReactionCatalysisMatrix([forIdxs; revIdxs], :) .* ...
<a name="l01076"></a>01076                     repmat(reactionEnzymes([forIdxs; revIdxs]), 1, nEnzymes), [], 1)&#39;;
<a name="l01077"></a>01077                 minEnzymes(unconstrainedEnzIdxs) = 1;
<a name="l01078"></a>01078                 minEnzymes(partlyUnconstrainedEnzIdxs) = max(minEnzymes(partlyUnconstrainedEnzIdxs), 1);
<a name="l01079"></a>01079                 minEnzymes = min(enzymes, minEnzymes);
<a name="l01080"></a>01080                 minEnzymes = max(minAvgExp, minEnzymes);
<a name="l01081"></a>01081             end
<a name="l01082"></a>01082             
<a name="l01083"></a>01083             if abs(this.calcGrowthRate(this.calcFluxBounds(...
<a name="l01084"></a>01084                     substrates, minEnzymes, fbaRxnBounds, this.fbaEnzymeBounds), fbaObj, fbaSMat) ...
<a name="l01085"></a>01085                     - growth) / growth &gt; this.tolerance || ...
<a name="l01086"></a>01086                     any(minEnzymes &gt; enzymes)
<a name="l01087"></a>01087                 warning(&#39;WholeCell:warning&#39;, &#39;Unable to calculate minimum enzymes&#39;);
<a name="l01088"></a>01088                 minEnzymes = enzymes;
<a name="l01089"></a>01089             end
<a name="l01090"></a>01090             
<a name="l01091"></a>01091             % maximum enzyme expression
<a name="l01092"></a>01092             loFluxBounds = fluxBounds(:, 1);
<a name="l01093"></a>01093             upFluxBounds = fluxBounds(:, 2);
<a name="l01094"></a>01094             
<a name="l01095"></a>01095             loFluxBounds(this.fbaReactionIndexs_biomassProduction) = growth;
<a name="l01096"></a>01096             upFluxBounds(this.fbaReactionIndexs_biomassProduction) = growth;
<a name="l01097"></a>01097             
<a name="l01098"></a>01098             tmpFluxBounds = this.calcFluxBounds(...
<a name="l01099"></a>01099                 substrates, enzymes, fbaRxnBounds, this.fbaEnzymeBounds, false);
<a name="l01100"></a>01100             loFluxBounds(forIdxs) = oldFbaReactionFluxs(forIdxs);
<a name="l01101"></a>01101             upFluxBounds(forIdxs) = tmpFluxBounds(forIdxs, 2);
<a name="l01102"></a>01102             loFluxBounds(revIdxs) = tmpFluxBounds(revIdxs, 1);
<a name="l01103"></a>01103             upFluxBounds(revIdxs) = oldFbaReactionFluxs(revIdxs);
<a name="l01104"></a>01104             
<a name="l01105"></a>01105             loFluxBounds = max(loFluxBounds, -this.realmax);
<a name="l01106"></a>01106             upFluxBounds = min(upFluxBounds,  this.realmax);
<a name="l01107"></a>01107             [fbaReactionFluxs, ~, ~, errFlag, errMsg] = ComputationUtil.linearProgramming(...
<a name="l01108"></a>01108                 &#39;maximize&#39;, objectiveFunc, fbaSMat, ...
<a name="l01109"></a>01109                 this.fbaRightHandSide, loFluxBounds, upFluxBounds, ...
<a name="l01110"></a>01110                 &#39;S&#39;, &#39;C&#39;, this.linearProgrammingOptions);
<a name="l01111"></a>01111             if errFlag
<a name="l01112"></a>01112                 warning(&#39;WholeCell:warning&#39;, errMsg);
<a name="l01113"></a>01113                 maxEnzymes = enzymes;
<a name="l01114"></a>01114             else
<a name="l01115"></a>01115                 fbaReactionFluxs = max(min(fbaReactionFluxs, upFluxBounds), loFluxBounds);
<a name="l01116"></a>01116                 
<a name="l01117"></a>01117                 reactionEnzymes = zeros(size(oldFbaReactionFluxs));
<a name="l01118"></a>01118                 reactionEnzymes(forIdxs) = fbaReactionFluxs(forIdxs) ./ this.fbaEnzymeBounds(forIdxs, 2);
<a name="l01119"></a>01119                 reactionEnzymes(revIdxs) = fbaReactionFluxs(revIdxs) ./ this.fbaEnzymeBounds(revIdxs, 1);
<a name="l01120"></a>01120                 
<a name="l01121"></a>01121                 tmpReactionCatalysisMatrix = this.fbaReactionCatalysisMatrix;
<a name="l01122"></a>01122                 tmpReactionCatalysisMatrix(tmpReactionCatalysisMatrix == 0) = NaN;
<a name="l01123"></a>01123                 maxEnzymes = 0*min(tmpReactionCatalysisMatrix([forIdxs; revIdxs], :) .* ...
<a name="l01124"></a>01124                     repmat(reactionEnzymes([forIdxs; revIdxs]), 1, nEnzymes), [], 1)&#39;;
<a name="l01125"></a>01125                 maxEnzymes(unconstrainedEnzIdxs) = realmax; %<span class="preprocessor">#ok&lt;CPROP,PROP&gt;</span>
<a name="l01126"></a>01126 <span class="preprocessor"></span>                maxEnzymes(partlyUnconstrainedEnzIdxs) = max(maxEnzymes(partlyUnconstrainedEnzIdxs), 1);
<a name="l01127"></a>01127                 maxEnzymes = max(enzymes, maxEnzymes);
<a name="l01128"></a>01128             end
<a name="l01129"></a>01129             
<a name="l01130"></a>01130             <span class="keywordflow">if</span> abs(this.calcGrowthRate(this.calcFluxBounds(...
<a name="l01131"></a>01131                     substrates, maxEnzymes, fbaRxnBounds, this.fbaEnzymeBounds), fbaObj, fbaSMat) ...
<a name="l01132"></a>01132                     - growth) / growth &gt; this.tolerance || ...
<a name="l01133"></a>01133                     any(maxEnzymes &lt; enzymes)
<a name="l01134"></a>01134                 warning(&#39;WholeCell:warning&#39;, &#39;Unable to calculate maximum enzymes&#39;);
<a name="l01135"></a>01135                 maxEnzymes = enzymes;
<a name="l01136"></a>01136             end
<a name="l01137"></a>01137         end
<a name="l01138"></a>01138         
<a name="l01139"></a>01139         %calculate indices of reactions whose flux is actively limited by
<a name="l01140"></a>01140         %the product of the expression of the catalyzing enzyme and its
<a name="l01141"></a>01141         %reaction kinetics
<a name="l01142"></a>01142         function [forIdxs, revIdxs] = calcEnzymeKineticallyLimitedReactions(this, substrates, enzymes, fbaObj, fbaSMat, fbaRxnBounds)
<a name="l01143"></a>01143             fluxBounds = this.calcFluxBounds(...
<a name="l01144"></a>01144                 substrates, enzymes, fbaRxnBounds, this.fbaEnzymeBounds);
<a name="l01145"></a>01145             fluxBounds_nokinetics = this.calcFluxBounds(...
<a name="l01146"></a>01146                 substrates, enzymes, fbaRxnBounds, this.fbaEnzymeBounds, false);
<a name="l01147"></a>01147             [~, ~, fbaReactionFluxs] = this.calcGrowthRate(fluxBounds, fbaObj, fbaSMat);
<a name="l01148"></a>01148             
<a name="l01149"></a>01149             rxnIdxs = this.fbaReactionIndexs_metabolicConversion;
<a name="l01150"></a>01150             forTfs = ...
<a name="l01151"></a>01151                 fbaReactionFluxs(rxnIdxs) &gt; 0 &amp; ...
<a name="l01152"></a>01152                 fluxBounds(rxnIdxs, 2) &lt; fluxBounds_nokinetics(rxnIdxs, 2);
<a name="l01153"></a>01153             revTfs = ...
<a name="l01154"></a>01154                 fbaReactionFluxs(rxnIdxs) &lt; 0 &amp; ...
<a name="l01155"></a>01155                 fluxBounds(rxnIdxs, 1) &gt; fluxBounds_nokinetics(rxnIdxs, 1);
<a name="l01156"></a>01156             
<a name="l01157"></a>01157             forIdxs = rxnIdxs(forTfs);
<a name="l01158"></a>01158             revIdxs = rxnIdxs(revTfs);
<a name="l01159"></a>01159         end
<a name="l01160"></a>01160         
<a name="l01161"></a>01161         function [fbaObj, fbaSMat, fbaRxnBounds] = calcEffectiveFBANetwork(this)
<a name="l01162"></a>01162             fbaObj = this.fbaObjective;
<a name="l01163"></a>01163             fbaObj(:) = 0;
<a name="l01164"></a>01164             fbaObj(this.fbaReactionIndexs_biomassProduction) = 1;
<a name="l01165"></a>01165             
<a name="l01166"></a>01166             fbaSMat = this.fbaReactionStoichiometryMatrix;
<a name="l01167"></a>01167             fbaSMat(this.fbaSubstrateIndexs_substrates, this.fbaReactionIndexs_biomassProduction) = ...
<a name="l01168"></a>01168                 - this.metabolismNewProduction(this.substrateIndexs_fba) ...
<a name="l01169"></a>01169                 - 4 * this.metabolismRecyclingProduction(this.substrateIndexs_fba);
<a name="l01170"></a>01170             
<a name="l01171"></a>01171             fbaRxnBounds = this.fbaReactionBounds;
<a name="l01172"></a>01172             fbaRxnBounds(this.fbaReactionIndexs_metaboliteInternalExchange, :) = 0;
<a name="l01173"></a>01173         end
<a name="l01174"></a>01174         
<a name="l01175"></a>01175         %initialization
<a name="l01176"></a>01176         %- Compute growth rate, reaction fluxs, reduced costs, duals
<a name="l01177"></a>01177         %- Don&#39;t update metabolite counts
<a name="l01178"></a>01178         function initializeState(this)
<a name="l01179"></a>01179             [this.metabolicReaction.growth, this.metabolicReaction.fluxs] = ...
<a name="l01180"></a>01180                 this.calcGrowthRate(this.calcFluxBounds(this.substrates, this.enzymes, this.fbaReactionBounds, this.fbaEnzymeBounds));
<a name="l01181"></a>01181         end
<a name="l01182"></a>01182         
<a name="l01183"></a>01183         %resource requirements
<a name="l01184"></a>01184         function result = calcResourceRequirements_Current(this)
<a name="l01185"></a>01185             result = zeros(size(this.substrates));
<a name="l01186"></a>01186             result(this.substrateIndexs_externalExchangedMetabolites, this.compartmentIndexs_extracellular) = ...
<a name="l01187"></a>01187                 this.fbaReactionBounds(this.fbaReactionIndexs_metaboliteExternalExchange, 2) * sum(this.mass.cellDry) * this.stepSizeSec;
<a name="l01188"></a>01188             result = ...
<a name="l01189"></a>01189                 + result ...
<a name="l01190"></a>01190                 - this.metabolismRecyclingProduction * 2 * max(...
<a name="l01191"></a>01191                     sum(this.mass.cellDry) / this.mass.cellInitialDryWeight * log(2) / this.cellCycleLength, ...
<a name="l01192"></a>01192                     this.metabolicReaction.growth);
<a name="l01193"></a>01193         end
<a name="l01194"></a>01194         
<a name="l01195"></a>01195         %simulation
<a name="l01196"></a>01196         function evolveState(this)
<a name="l01197"></a>01197             %% Calculate growth rate
<a name="l01198"></a>01198             [this.metabolicReaction.growth, this.metabolicReaction.fluxs, fbaReactionFluxs] = ...
<a name="l01199"></a>01199                 this.calcGrowthRate(this.calcFluxBounds(this.substrates, this.enzymes, this.fbaReactionBounds, this.fbaEnzymeBounds));
<a name="l01200"></a>01200             
<a name="l01201"></a>01201             %% Compute real-valued amount of nutrient imported, biomass (DNA,
<a name="l01202"></a>01202             %<a class="code" href="class_r_n_a.html">RNA</a>, protein, membrane, etc) precursors produced and consumed,
<a name="l01203"></a>01203             %and monomers modified. Stochastically round to:
<a name="l01204"></a>01204             %1. Instantaneously, maintain integer-valued amounts of metabolites
<a name="l01205"></a>01205             %2. Over time, maintain experimentally determined ratios of biomass
<a name="l01206"></a>01206             %   components (stored in this.metabolismProduction)
<a name="l01207"></a>01207             
<a name="l01208"></a>01208             %nutrient uptake
<a name="l01209"></a>01209             this.substrates(this.substrateIndexs_externalExchangedMetabolites, this.compartmentIndexs_extracellular) = ...
<a name="l01210"></a>01210                 + this.substrates(this.substrateIndexs_externalExchangedMetabolites, this.compartmentIndexs_extracellular) ...
<a name="l01211"></a>01211                 - this.randStream.stochasticRound(fbaReactionFluxs(this.fbaReactionIndexs_metaboliteExternalExchange) * this.stepSizeSec);
<a name="l01212"></a>01212             
<a name="l01213"></a>01213             %recycled metabolites
<a name="l01214"></a>01214             this.substrates(this.substrateIndexs_internalExchangedMetabolites) = ...
<a name="l01215"></a>01215                 + this.substrates(this.substrateIndexs_internalExchangedMetabolites) ...
<a name="l01216"></a>01216                 + this.randStream.stochasticRound(fbaReactionFluxs(this.fbaReactionIndexs_metaboliteInternalExchange));
<a name="l01217"></a>01217             
<a name="l01218"></a>01218             %new metabolites
<a name="l01219"></a>01219             this.substrates = ...
<a name="l01220"></a>01220                 + this.substrates ...
<a name="l01221"></a>01221                 + this.randStream.stochasticRound(this.metabolismNewProduction * this.metabolicReaction.growth * this.stepSizeSec);
<a name="l01222"></a>01222             
<a name="l01223"></a>01223             %unaccounted energy consumption
<a name="l01224"></a>01224             this.substrates(this.substrateIndexs_atpHydrolysis) = ...
<a name="l01225"></a>01225                 + this.substrates(this.substrateIndexs_atpHydrolysis) ...
<a name="l01226"></a>01226                 + [-1; -1; 1; 1; 1] * this.randStream.stochasticRound(...
<a name="l01227"></a>01227                 this.unaccountedEnergyConsumption * this.metabolicReaction.growth * this.stepSizeSec);
<a name="l01228"></a>01228             
<a name="l01229"></a>01229             %make metabolites counts positive (in case stochastic rounding
<a name="l01230"></a>01230             %made them slightly negative eg -1) except H2O, H+
<a name="l01231"></a>01231             if any(any(this.substrates(this.substrateMetaboliteLocalIndexs(:, 1), :) &lt; -1))
<a name="l01232"></a>01232                 [i, j] = find(this.substrates(this.substrateMetaboliteLocalIndexs(:, 1), :) &lt; -1);
<a name="l01233"></a>01233                 compIDs = cell(3, 1);
<a name="l01234"></a>01234                 compIDs(this.compartmentIndexs_cytosol) = {<span class="charliteral">&#39;c&#39;</span>};
<a name="l01235"></a>01235                 compIDs(this.compartmentIndexs_extracellular) = {<span class="charliteral">&#39;e&#39;</span>};
<a name="l01236"></a>01236                 compIDs(this.compartmentIndexs_membrane) = {<span class="charliteral">&#39;m&#39;</span>};
<a name="l01237"></a>01237                 n = max(cellfun(@length, this.substrateWholeCellModelIDs(this.substrateMetaboliteLocalIndexs(i, 1))));
<a name="l01238"></a>01238                 msg = cellfun(@(<span class="keywordtype">id</span>, comp, val) sprintf([<span class="stringliteral">&#39;%-&#39;</span> num2str(n) <span class="stringliteral">&#39;s  %-3s  %10d&#39;</span>], <span class="keywordtype">id</span>, comp, val), ...
<a name="l01239"></a>01239                     this.substrateWholeCellModelIDs(this.substrateMetaboliteLocalIndexs(i, 1)), compIDs(j), ...
<a name="l01240"></a>01240                     num2cell(this.substrates(sub2ind(size(this.substrates), this.substrateMetaboliteLocalIndexs(i, 1), j))), ...
<a name="l01241"></a>01241                     <span class="stringliteral">&#39;UniformOutput&#39;</span>, <span class="keyword">false</span>);
<a name="l01242"></a>01242                 warning(<span class="stringliteral">&#39;WholeCell:warning:negativeMetabolites&#39;</span>, ...
<a name="l01243"></a>01243                     [<span class="stringliteral">&#39;%d metabolites have negative counts\n%-&#39;</span> num2str(n) <span class="stringliteral">&#39;s  %-3s  %10s\n%-&#39;</span> num2str(n) <span class="stringliteral">&#39;s  %-3s  %10s\n%s&#39;</span>], ...
<a name="l01244"></a>01244                     numel(i), ...
<a name="l01245"></a>01245                     <span class="stringliteral">&#39;Met&#39;</span>, <span class="stringliteral">&#39;Cmp&#39;</span>, <span class="stringliteral">&#39;Val&#39;</span>, ...
<a name="l01246"></a>01246                     repmat(<span class="charliteral">&#39;=&#39;</span>, 1, n), <span class="stringliteral">&#39;===&#39;</span>, repmat(<span class="charliteral">&#39;=&#39;</span>, 1, 10), ...
<a name="l01247"></a>01247                     strjoin(sprintf(<span class="charliteral">&#39;\n&#39;</span>), msg{:}));
<a name="l01248"></a>01248             end
<a name="l01249"></a>01249             this.substrates(this.substrateMetaboliteLocalIndexs(:, 1), :) = max(0, this.substrates(this.substrateMetaboliteLocalIndexs(:, 1), :));
<a name="l01250"></a>01250             
<a name="l01251"></a>01251             %% update cell volume
<a name="l01252"></a>01252             this.mass.calcMass();
<a name="l01253"></a>01253             this.geometry.calculateVolume();
<a name="l01254"></a>01254         end
<a name="l01255"></a>01255     end
<a name="l01256"></a>01256     
<a name="l01257"></a>01257     %evolve state helper methods
<a name="l01258"></a>01258     methods
<a name="l01259"></a>01259         %calculates the growth rate given the supplied flux bounds
<a name="l01260"></a>01260         <span class="keyword">function</span> [growth, reactionFluxs, fbaReactionFluxs, ...
<a name="l01261"></a>01261                 reducedCosts, fbaReducedCosts, ...
<a name="l01262"></a>01262                 shadowPrices, fbaShadowPrices] = ...
<a name="l01263"></a>01263                 calcGrowthRate(<span class="keyword">this</span>, fluxBounds, fbaObj, fbaSMat)
<a name="l01264"></a>01264             % <span class="keyword">import</span> classes
<a name="l01265"></a>01265             <span class="keyword">import</span> edu.stanford.covert.util.ComputationUtil;
<a name="l01266"></a>01266             <span class="keyword">import</span> edu.stanford.covert.util.ConstantUtil;
<a name="l01267"></a>01267             
<a name="l01268"></a>01268             %inputs
<a name="l01269"></a>01269             <span class="keywordflow">if</span> nargin &lt; 3
<a name="l01270"></a>01270                 fbaObj = this.fbaObjective;
<a name="l01271"></a>01271             end
<a name="l01272"></a>01272             <span class="keywordflow">if</span> nargin &lt; 4
<a name="l01273"></a>01273                 fbaSMat = this.fbaReactionStoichiometryMatrix;
<a name="l01274"></a>01274             end
<a name="l01275"></a>01275             
<a name="l01276"></a>01276             %flux bounds
<a name="l01277"></a>01277             loFluxBounds = fluxBounds(:, 1);
<a name="l01278"></a>01278             upFluxBounds = fluxBounds(:, 2);
<a name="l01279"></a>01279             
<a name="l01280"></a>01280             %real-valued linear programming
<a name="l01281"></a>01281             loFluxBounds = max(loFluxBounds, -this.realmax);
<a name="l01282"></a>01282             upFluxBounds = min(upFluxBounds,  this.realmax);
<a name="l01283"></a>01283             [fbaReactionFluxs, lambda, ~, errFlag, errMsg] = ComputationUtil.linearProgramming(...
<a name="l01284"></a>01284                 <span class="stringliteral">&#39;maximize&#39;</span>, fbaObj, fbaSMat, ...
<a name="l01285"></a>01285                 this.fbaRightHandSide, loFluxBounds, upFluxBounds, ...
<a name="l01286"></a>01286                 <span class="charliteral">&#39;S&#39;</span>, <span class="charliteral">&#39;C&#39;</span>, this.linearProgrammingOptions);
<a name="l01287"></a>01287             <span class="keywordflow">if</span> errFlag
<a name="l01288"></a>01288                 warning(<span class="stringliteral">&#39;WholeCell:warning&#39;</span>, <span class="stringliteral">&#39;Linear programming error: %s. Returning feasible, but possibly non-optimal solution x=0.&#39;</span>, errMsg);
<a name="l01289"></a>01289                 fbaReactionFluxs = zeros(size(loFluxBounds));
<a name="l01290"></a>01290             end
<a name="l01291"></a>01291             
<a name="l01292"></a>01292             fbaReactionFluxs = max(min(fbaReactionFluxs, upFluxBounds), loFluxBounds);
<a name="l01293"></a>01293             growth = fbaReactionFluxs(this.fbaReactionIndexs_biomassProduction);
<a name="l01294"></a>01294             reactionFluxs = zeros(size(this.reactionStoichiometryMatrix, 2), 1);
<a name="l01295"></a>01295             reactionFluxs(this.reactionIndexs_fba) = fbaReactionFluxs(this.fbaReactionIndexs_metabolicConversion);
<a name="l01296"></a>01296             
<a name="l01297"></a>01297             <span class="keywordflow">if</span> nargout &gt; 3
<a name="l01298"></a>01298                 fbaReducedCosts = lambda.reducedCosts;
<a name="l01299"></a>01299                 reducedCosts = zeros(size(this.reactionStoichiometryMatrix, 2), 1);
<a name="l01300"></a>01300                 reducedCosts(this.reactionIndexs_fba) = fbaReducedCosts(this.fbaReactionIndexs_metabolicConversion);
<a name="l01301"></a>01301                 
<a name="l01302"></a>01302                 fbaShadowPrices = lambda.shadowPrices;
<a name="l01303"></a>01303                 shadowPrices = zeros(size(this.substrates));
<a name="l01304"></a>01304                 shadowPrices(this.substrateIndexs_fba) = fbaShadowPrices(this.fbaSubstrateIndexs_substrates);
<a name="l01305"></a>01305             end
<a name="l01306"></a>01306         end
<a name="l01307"></a>01307         
<a name="l01308"></a>01308         %Compute reaction flux upper and lower bounds based on
<a name="l01309"></a>01309         % 1. <a class="code" href="class_enzyme.html">Enzyme</a> kinetics         (fbaEnzymeBounds)
<a name="l01310"></a>01310         % 2. <a class="code" href="class_enzyme.html">Enzyme</a> availability     (enzymes)
<a name="l01311"></a>01311         % 3. Transport rates         (fbaReactionBounds)
<a name="l01312"></a>01312         % 4. <a class="code" href="class_metabolite.html">Metabolite</a> availability (substrates)
<a name="l01313"></a>01313         % 5. <a class="code" href="class_protein.html">Protein</a> availability    (substrates)
<a name="l01314"></a>01314         function bounds = calcFluxBounds(this, substrates, enzymes, fbaReactionBounds, fbaEnzymeBounds, ...
<a name="l01315"></a>01315                 applyEnzymeKineticBounds, applyEnzymeBounds, applyDirectionalityBounds, ...
<a name="l01316"></a>01316                 applyExternalMetaboliteBounds, applyInternalMetaboliteBounds, applyProteinBounds)
<a name="l01317"></a>01317             %import classes
<a name="l01318"></a>01318             import edu.stanford.covert.util.ConstantUtil;
<a name="l01319"></a>01319             
<a name="l01320"></a>01320             %options
<a name="l01321"></a>01321             if nargin &lt; 6,  applyEnzymeKineticBounds      = true; end
<a name="l01322"></a>01322             if nargin &lt; 7,  applyEnzymeBounds             = true; end
<a name="l01323"></a>01323             if nargin &lt; 8,  applyDirectionalityBounds     = true; end
<a name="l01324"></a>01324             if nargin &lt; 9,  applyExternalMetaboliteBounds = true; end
<a name="l01325"></a>01325             if nargin &lt; 10, applyInternalMetaboliteBounds = true; end
<a name="l01326"></a>01326             if nargin &lt; 11, applyProteinBounds            = true; end
<a name="l01327"></a>01327             
<a name="l01328"></a>01328             %numbers
<a name="l01329"></a>01329             nReactions = size(this.fbaReactionStoichiometryMatrix, 2);
<a name="l01330"></a>01330             
<a name="l01331"></a>01331             %initialize
<a name="l01332"></a>01332             lowerBounds = -inf(nReactions, 1);
<a name="l01333"></a>01333             upperBounds =  inf(nReactions, 1);
<a name="l01334"></a>01334             
<a name="l01335"></a>01335             %numbers of enzymes catalyzing each reaction, enzyme kinetics
<a name="l01336"></a>01336             rxnEnzymes = this.fbaReactionCatalysisMatrix * enzymes;
<a name="l01337"></a>01337             if applyEnzymeKineticBounds
<a name="l01338"></a>01338                 lowerBounds = max(lowerBounds, fbaEnzymeBounds(:, 1) .* rxnEnzymes);
<a name="l01339"></a>01339                 upperBounds = min(upperBounds, fbaEnzymeBounds(:, 2) .* rxnEnzymes);
<a name="l01340"></a>01340             end
<a name="l01341"></a>01341             
<a name="l01342"></a>01342             %numbers of enzymes catalyzing each reaction, unkown enzyme kinetics
<a name="l01343"></a>01343             if applyEnzymeBounds
<a name="l01344"></a>01344                 lowerBounds(any(this.fbaReactionCatalysisMatrix, 2) &amp; rxnEnzymes &lt;= 0) = 0;
<a name="l01345"></a>01345                 upperBounds(any(this.fbaReactionCatalysisMatrix, 2) &amp; rxnEnzymes &lt;= 0) = 0;
<a name="l01346"></a>01346             end
<a name="l01347"></a>01347             
<a name="l01348"></a>01348             %reaction directionality / thermodynamics
<a name="l01349"></a>01349             if applyDirectionalityBounds
<a name="l01350"></a>01350                 lowerBounds(this.fbaReactionIndexs_metabolicConversion) = max(lowerBounds(this.fbaReactionIndexs_metabolicConversion), fbaReactionBounds(this.fbaReactionIndexs_metabolicConversion, 1));
<a name="l01351"></a>01351                 upperBounds(this.fbaReactionIndexs_metabolicConversion) = min(upperBounds(this.fbaReactionIndexs_metabolicConversion), fbaReactionBounds(this.fbaReactionIndexs_metabolicConversion, 2));
<a name="l01352"></a>01352                 
<a name="l01353"></a>01353                 lowerBounds(this.fbaReactionIndexs_metaboliteInternalExchange) = max(lowerBounds(this.fbaReactionIndexs_metaboliteInternalExchange), fbaReactionBounds(this.fbaReactionIndexs_metaboliteInternalExchange, 1));
<a name="l01354"></a>01354                 upperBounds(this.fbaReactionIndexs_metaboliteInternalExchange) = min(upperBounds(this.fbaReactionIndexs_metaboliteInternalExchange), fbaReactionBounds(this.fbaReactionIndexs_metaboliteInternalExchange, 2));
<a name="l01355"></a>01355                 
<a name="l01356"></a>01356                 lowerBounds(this.fbaReactionIndexs_biomassExchange) = max(lowerBounds(this.fbaReactionIndexs_biomassExchange), fbaReactionBounds(this.fbaReactionIndexs_biomassExchange, 1));
<a name="l01357"></a>01357                 upperBounds(this.fbaReactionIndexs_biomassExchange) = min(upperBounds(this.fbaReactionIndexs_biomassExchange), fbaReactionBounds(this.fbaReactionIndexs_biomassExchange, 2));
<a name="l01358"></a>01358                 
<a name="l01359"></a>01359                 lowerBounds(this.fbaReactionIndexs_biomassProduction) = max(lowerBounds(this.fbaReactionIndexs_biomassProduction), fbaReactionBounds(this.fbaReactionIndexs_biomassProduction, 1));
<a name="l01360"></a>01360                 upperBounds(this.fbaReactionIndexs_biomassProduction) = min(upperBounds(this.fbaReactionIndexs_biomassProduction), fbaReactionBounds(this.fbaReactionIndexs_biomassProduction, 2));
<a name="l01361"></a>01361             end
<a name="l01362"></a>01362             
<a name="l01363"></a>01363             %external metabolite availability
<a name="l01364"></a>01364             if applyExternalMetaboliteBounds
<a name="l01365"></a>01365                 upperBounds(this.fbaReactionIndexs_metaboliteExternalExchange) = min(...
<a name="l01366"></a>01366                     upperBounds(this.fbaReactionIndexs_metaboliteExternalExchange), ...
<a name="l01367"></a>01367                     substrates(this.substrateIndexs_externalExchangedMetabolites, this.compartmentIndexs_extracellular) / this.stepSizeSec);
<a name="l01368"></a>01368                 
<a name="l01369"></a>01369                 cellDryMass = sum(this.mass.cellDry);
<a name="l01370"></a>01370                 lowerBounds(this.fbaReactionIndexs_metaboliteExternalExchange) = ...
<a name="l01371"></a>01371                     max(lowerBounds(this.fbaReactionIndexs_metaboliteExternalExchange), ...
<a name="l01372"></a>01372                     fbaReactionBounds(this.fbaReactionIndexs_metaboliteExternalExchange, 1) * cellDryMass);
<a name="l01373"></a>01373                 upperBounds(this.fbaReactionIndexs_metaboliteExternalExchange) = ...
<a name="l01374"></a>01374                     min(upperBounds(this.fbaReactionIndexs_metaboliteExternalExchange), ...
<a name="l01375"></a>01375                     fbaReactionBounds(this.fbaReactionIndexs_metaboliteExternalExchange, 2) * cellDryMass);
<a name="l01376"></a>01376             end
<a name="l01377"></a>01377             
<a name="l01378"></a>01378             %internal metabolite availability
<a name="l01379"></a>01379             if applyInternalMetaboliteBounds
<a name="l01380"></a>01380                 lowerBounds(this.fbaReactionIndexs_metaboliteInternalLimitedExchange) = max(...
<a name="l01381"></a>01381                     lowerBounds(this.fbaReactionIndexs_metaboliteInternalLimitedExchange), ...
<a name="l01382"></a>01382                     -substrates(this.substrateIndexs_internalExchangedLimitedMetabolites) / ...
<a name="l01383"></a>01383                     this.stepSizeSec);
<a name="l01384"></a>01384             end
<a name="l01385"></a>01385             
<a name="l01386"></a>01386             %protein monomers and complexes
<a name="l01387"></a>01387             if applyProteinBounds
<a name="l01388"></a>01388                 limitedReactions = any(any(...
<a name="l01389"></a>01389                     this.reactionStoichiometryMatrix([this.substrateMonomerLocalIndexs; this.substrateComplexLocalIndexs], this.reactionIndexs_fba, :) &amp; ...
<a name="l01390"></a>01390                     ~permute(repmat(this.proteinLimitableProteinComposition * substrates(this.substrateIndexs_limitableProteins, :), [1 1 numel(this.reactionIndexs_fba)]), [1 3 2]), ...
<a name="l01391"></a>01391                     3), 1);
<a name="l01392"></a>01392                 lowerBounds(this.fbaReactionIndexs_metabolicConversion(limitedReactions)) = 0;
<a name="l01393"></a>01393                 upperBounds(this.fbaReactionIndexs_metabolicConversion(limitedReactions)) = 0;
<a name="l01394"></a>01394             end
<a name="l01395"></a>01395             
<a name="l01396"></a>01396             %sanitize linear programming input: replace Infs with large numbers
<a name="l01397"></a>01397             bounds = [lowerBounds upperBounds];
<a name="l01398"></a>01398         end
<a name="l01399"></a>01399     end
<a name="l01400"></a>01400 end
</pre></div></div><!-- contents -->
</div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
<a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(0)"><span class="SelectionMark">&#160;</span>All</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(1)"><span class="SelectionMark">&#160;</span>Classes</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(2)"><span class="SelectionMark">&#160;</span>Files</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(3)"><span class="SelectionMark">&#160;</span>Functions</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(4)"><span class="SelectionMark">&#160;</span>Variables</a></div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

  <div id="nav-path" class="navpath">
    <ul>
      <li class="navelem"><a class="el" href="_metabolism_8m.html">Metabolism.m</a>      </li>

    <li class="footer">Generated on Wed Feb 8 2012 21:42:32 for Whole Cell Simulation by
    <a href="http://www.doxygen.org/index.html">
    <img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.7.6.1 </li>
   </ul>
 </div>


</body>
</html>
