<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<title>Whole Cell Simulation: src/+edu/+stanford/+covert/+cell/+sim/+analysis/Population.m Source File</title>

<link href="tabs.css" rel="stylesheet" type="text/css"/>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
<link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="navtree.js"></script>
<script type="text/javascript">
  $(document).ready(initResizable);
</script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
  $(document).ready(function() { searchBox.OnSelectItem(0); });
</script>

</head>
<body>
<div id="top"><!-- do not remove this div! -->


<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  
  
  <td style="padding-left: 0.5em;">
   <div id="projectname">Whole Cell Simulation
   
   </div>
   
  </td>
  
  
  
 </tr>
 </tbody>
</table>
</div>

<!-- Generated by Doxygen 1.7.6.1 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "search",false,'Search');
</script>
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&#160;Page</span></a></li>
      <li><a href="annotated.html"><span>Classes</span></a></li>
      <li class="current"><a href="files.html"><span>Files</span></a></li>
      <li>
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
  <div id="navrow2" class="tabs2">
    <ul class="tablist">
      <li><a href="files.html"><span>File&#160;List</span></a></li>
      <li><a href="globals.html"><span>File&#160;Members</span></a></li>
    </ul>
  </div>
</div>
<div id="side-nav" class="ui-resizable side-nav-resizable">
  <div id="nav-tree">
    <div id="nav-tree-contents">
    </div>
  </div>
  <div id="splitbar" style="-moz-user-select:none;" 
       class="ui-resizable-handle">
  </div>
</div>
<script type="text/javascript">
  initNavTree('_population_8m.html','');
</script>
<div id="doc-content">
<div class="header">
  <div class="headertitle">
<div class="title">src/+edu/+stanford/+covert/+cell/+sim/+analysis/Population.m</div>  </div>
</div><!--header-->
<div class="contents">
<a href="_population_8m.html">Go to the documentation of this file.</a><div class="fragment"><pre class="fragment"><a name="l00001"></a>00001 %<a class="code" href="class_population.html">Population</a>
<a name="l00002"></a>00002 %
<a name="l00003"></a>00003 % Author: Jonathan Karr, jkarr@stanford.edu
<a name="l00004"></a>00004 % Author: Derek Macklin, macklin@stanford.edu
<a name="l00005"></a>00005 % Affiliation: Covert Lab, Department of Bioengineering, Stanford University
<a name="l00006"></a>00006 % Last Updated: 7/6/2011
<a name="l00007"></a><a class="code" href="class_population.html">00007</a> classdef <a class="code" href="class_population.html">Population</a>
<a name="l00008"></a>00008     methods (Static = <span class="keyword">true</span>)
<a name="l00009"></a>00009         function run(simBatchDir, fileName, selectedSimulations, methodsToRun)
<a name="l00010"></a>00010             import edu.stanford.covert.cell.sim.analysis.<a class="code" href="class_population.html">Population</a>;
<a name="l00011"></a>00011             import edu.stanford.covert.cell.sim.util.PlotUtil;
<a name="l00012"></a>00012             import edu.stanford.covert.cell.sim.util.PrintUtil;
<a name="l00013"></a>00013             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
<a name="l00014"></a>00014             
<a name="l00015"></a>00015             if nargin &lt; 1 || isempty(simBatchDir)
<a name="l00016"></a>00016                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
<a name="l00017"></a>00017             end
<a name="l00018"></a>00018             if nargin &lt; 3 || isempty(selectedSimulations)
<a name="l00019"></a>00019                 selectedSimulations = SimulationDiskUtil.getCompleteSimulations(simBatchDir);
<a name="l00020"></a>00020             end
<a name="l00021"></a>00021             
<a name="l00022"></a>00022             %evaluate each method and optionally save at the result as
<a name="l00023"></a>00023             %pdf/xls files
<a name="l00024"></a>00024             metaData = meta.class.fromName(&#39;edu.stanford.covert.cell.sim.analysis.Population&#39;);
<a name="l00025"></a>00025             for i = 1:numel(metaData.Methods)
<a name="l00026"></a>00026                 %only plot methods matching the signature
<a name="l00027"></a>00027                 %  plotFcn(figHandle, simBatchDir, selectedSimulations)
<a name="l00028"></a>00028                 if ...
<a name="l00029"></a>00029                         numel(metaData.Methods{i}.InputNames) &lt; 3 || ~isequal(metaData.Methods{i}.InputNames(1:3), {<span class="stringliteral">&#39;figHandle&#39;</span>; <span class="stringliteral">&#39;simBatchDir&#39;</span>; <span class="stringliteral">&#39;selectedSimulations&#39;</span>}) || ...
<a name="l00030"></a>00030                         numel(metaData.Methods{i}.OutputNames) &lt; 1 || ~isequal(metaData.Methods{i}.OutputNames(1), {<span class="stringliteral">&#39;figData&#39;</span>})
<a name="l00031"></a>00031                     <span class="keywordflow">continue</span>;
<a name="l00032"></a>00032                 end
<a name="l00033"></a>00033                 
<a name="l00034"></a>00034                 %<span class="keyword">get</span> method name
<a name="l00035"></a>00035                 methodName = metaData.Methods{i}.Name;
<a name="l00036"></a>00036                 MethodName = [upper(methodName(1)) methodName(2:end)];
<a name="l00037"></a>00037                 <span class="keywordflow">if</span> nargin &gt;= 4 &amp;&amp; ~ismember(metaData.Methods{i}.Name, methodsToRun)
<a name="l00038"></a>00038                     <span class="keywordflow">continue</span>;
<a name="l00039"></a>00039                 end
<a name="l00040"></a>00040                 
<a name="l00041"></a>00041                 %run plot method
<a name="l00042"></a>00042                 [~, figHandle] = edu.stanford.covert.cell.sim.util.PlotUtil.newAxesHandle();
<a name="l00043"></a>00043                 <span class="keywordflow">try</span>
<a name="l00044"></a>00044                     argout = cell(size(metaData.Methods{i}.OutputNames));
<a name="l00045"></a>00045                     [argout{:}] = <a class="code" href="class_population.html">Population</a>.(methodName)(figHandle, simBatchDir, selectedSimulations);
<a name="l00046"></a>00046                     
<a name="l00047"></a>00047                     % save plot
<a name="l00048"></a>00048                     <span class="keywordflow">if</span> nargin &gt;= 2
<a name="l00049"></a>00049                         figData = argout{1}; %#ok&lt;NASGU&gt;
<a name="l00050"></a>00050                         save([fileName <span class="charliteral">&#39;-&#39;</span> MethodName <span class="stringliteral">&#39;.mat&#39;</span>], <span class="stringliteral">&#39;-struct&#39;</span>, <span class="stringliteral">&#39;figData&#39;</span>);
<a name="l00051"></a>00051                         
<a name="l00052"></a>00052                         orient(figHandle, <span class="stringliteral">&#39;portrait&#39;</span>);
<a name="l00053"></a>00053                         saveas(figHandle, [fileName <span class="charliteral">&#39;-&#39;</span> MethodName <span class="stringliteral">&#39;.pdf&#39;</span>]);
<a name="l00054"></a>00054                     end
<a name="l00055"></a>00055                     
<a name="l00056"></a>00056                     % save table
<a name="l00057"></a>00057                     <span class="keywordflow">if</span> numel(argout) &gt; 1
<a name="l00058"></a>00058                         <span class="keywordflow">if</span> nargin &gt;= 2
<a name="l00059"></a>00059                             PrintUtil.printToFile(argout{2}, argout{3}, [fileName &#39;-&#39; MethodName &#39;.xls&#39;], MethodName);
<a name="l00060"></a>00060                         <span class="keywordflow">else</span>
<a name="l00061"></a>00061                             PrintUtil.printToStdIO(argout{2}, argout{3});
<a name="l00062"></a>00062                         end
<a name="l00063"></a>00063                     end
<a name="l00064"></a>00064                 <span class="keywordflow">catch</span> exception
<a name="l00065"></a>00065                     warning(<span class="stringliteral">&#39;WholeCell:warning&#39;</span>, <span class="stringliteral">&#39;Unable to make %s plot:\n%s&#39;</span>, methodName, exception.getReport());
<a name="l00066"></a>00066                 end
<a name="l00067"></a>00067                 
<a name="l00068"></a>00068                 <span class="keywordflow">if</span> nargin &gt;= 2
<a name="l00069"></a>00069                     close(figHandle);
<a name="l00070"></a>00070                     clear figHandle;
<a name="l00071"></a>00071                 end
<a name="l00072"></a>00072                 clear argout figData;
<a name="l00073"></a>00073             end
<a name="l00074"></a>00074         end
<a name="l00075"></a>00075         
<a name="l00076"></a>00076         <span class="keyword">function</span> figData = growth(figHandle, simBatchDir, selectedSimulations)
<a name="l00077"></a>00077             import edu.stanford.covert.cell.sim.util.MacromoleculeUtil;
<a name="l00078"></a>00078             import edu.stanford.covert.cell.sim.util.SimulationEnsemble;
<a name="l00079"></a>00079             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
<a name="l00080"></a>00080             import edu.stanford.covert.cell.sim.util.PlotUtil;
<a name="l00081"></a>00081             import edu.stanford.covert.cell.sim.analysis.<a class="code" href="class_cell_overview.html">CellOverview</a>;
<a name="l00082"></a>00082             import edu.stanford.covert.util.<a class="code" href="class_constant_util.html">ConstantUtil</a>;
<a name="l00083"></a>00083             
<a name="l00084"></a>00084             if nargin &lt; 2 || isempty(simBatchDir)
<a name="l00085"></a>00085                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
<a name="l00086"></a>00086             end
<a name="l00087"></a>00087             if nargin &lt; 3 || isempty(selectedSimulations)
<a name="l00088"></a>00088                 selectedSimulations = SimulationDiskUtil.getCompleteSimulations(simBatchDir);
<a name="l00089"></a>00089             end
<a name="l00090"></a>00090             
<a name="l00091"></a>00091             [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep &#39;1&#39;]);
<a name="l00092"></a>00092             comp = sim.compartment;
<a name="l00093"></a>00093             g = sim.gene;
<a name="l00094"></a>00094             r = sim.state(&#39;Rna&#39;);
<a name="l00095"></a>00095             pm = sim.state(&#39;<a class="code" href="class_protein_monomer.html">ProteinMonomer</a>&#39;);
<a name="l00096"></a>00096             pc = sim.state(&#39;<a class="code" href="class_protein_complex.html">ProteinComplex</a>&#39;);
<a name="l00097"></a>00097             
<a name="l00098"></a>00098             %% get data
<a name="l00099"></a>00099             stateNames = {
<a name="l00100"></a>00100                 <span class="stringliteral">&#39;mass&#39;</span>              <span class="stringliteral">&#39;Mass&#39;</span>
<a name="l00101"></a>00101                 <span class="stringliteral">&#39;ploidy&#39;</span>            {<span class="stringliteral">&#39;Chromosome&#39;</span> <span class="stringliteral">&#39;Copy Number&#39;</span>}
<a name="l00102"></a>00102                 <span class="stringliteral">&#39;rnas&#39;</span>              <span class="stringliteral">&#39;RNA&#39;</span>
<a name="l00103"></a>00103                 <span class="stringliteral">&#39;proteins&#39;</span>          <span class="stringliteral">&#39;Proteins&#39;</span>
<a name="l00104"></a>00104                 <span class="stringliteral">&#39;lipids&#39;</span>            <span class="stringliteral">&#39;Lipids&#39;</span>
<a name="l00105"></a>00105                 <span class="stringliteral">&#39;growth_rate&#39;</span>       <span class="stringliteral">&#39;Growth&#39;</span>
<a name="l00106"></a>00106                 <span class="stringliteral">&#39;ribosomes&#39;</span>         <span class="stringliteral">&#39;Ribosomes&#39;</span>
<a name="l00107"></a>00107                 };
<a name="l00108"></a>00108             ensemble = SimulationEnsemble(simBatchDir, stateNames, [], selectedSimulations);
<a name="l00109"></a>00109             simEndTimes = ensemble.stateData.simulationEndTimes;
<a name="l00110"></a>00110             
<a name="l00111"></a>00111             time = ensemble.stateData.time / 3600;
<a name="l00112"></a>00112             mass = ensemble.stateData.values(ensemble.getPropertyIndices(<span class="stringliteral">&#39;mass&#39;</span>), :, :, :) * sim.state(<span class="stringliteral">&#39;Mass&#39;</span>).cellInitialDryWeight / (1-sim.state(&#39;Mass&#39;).fractionWetWeight) * 1e15;
<a name="l00113"></a>00113             growth = ensemble.stateData.values(ensemble.getPropertyIndices(<span class="stringliteral">&#39;growth_rate&#39;</span>), :, :, :) * 3600 * sim.state(<span class="stringliteral">&#39;Mass&#39;</span>).cellInitialDryWeight / (1 - sim.state(&#39;Mass&#39;).fractionWetWeight) * 1e15;
<a name="l00114"></a>00114             ploidy = ensemble.stateData.values(ensemble.getPropertyIndices(<span class="stringliteral">&#39;ploidy&#39;</span>), :, :, :);
<a name="l00115"></a>00115             rnas = ensemble.stateData.values(ensemble.getPropertyIndices(<span class="stringliteral">&#39;rnas&#39;</span>), :, :, :);
<a name="l00116"></a>00116             proteins = ensemble.stateData.values(ensemble.getPropertyIndices(<span class="stringliteral">&#39;proteins&#39;</span>), :, :, :);
<a name="l00117"></a>00117             lipids = ensemble.stateData.values(ensemble.getPropertyIndices(<span class="stringliteral">&#39;lipids&#39;</span>), :, :, :);
<a name="l00118"></a>00118             ribosomes = ensemble.stateData.values(ensemble.getPropertyIndices(<span class="stringliteral">&#39;ribosomes&#39;</span>), :, :, :);
<a name="l00119"></a>00119             
<a name="l00120"></a>00120             [~, simOrder] = sort(mass(:, 1, min(ensemble.stateData.simulationEndTimes), :));
<a name="l00121"></a>00121             simOrder = simOrder(:);
<a name="l00122"></a>00122             
<a name="l00123"></a>00123             clear ensemble;
<a name="l00124"></a>00124             
<a name="l00125"></a>00125             ribMonomerIdxs = find(any(pc.proteinComplexComposition(g.mRNAIndexs, pc.ribosome70SIndexs, :), 3));
<a name="l00126"></a>00126             [~, tmp] = min(r.matureRNAGeneComposition(g.mRNAIndexs(ribMonomerIdxs), :) * r.expression(r.matureIndexs));
<a name="l00127"></a>00127             [~, ~, ~, ~, ~, ...
<a name="l00128"></a>00128                 ~, ~, ~, ~, ...
<a name="l00129"></a>00129                 ~, ~, ~, ~, ...
<a name="l00130"></a>00130                 ribMatureRnaIdxs, ~, ~, ~, ...
<a name="l00131"></a>00131                 ribMonomerIdxs, ~, ~, ribMonomerNames] = ...
<a name="l00132"></a>00132                 MacromoleculeUtil.getMacromoleculeIndexsIDsNames(pm.wholeCellModelIDs{pm.matureIndexs(g.mRNAIndexs(ribMonomerIdxs(tmp)))}, sim);
<a name="l00133"></a>00133             ribMonomerNames = strrep(ribMonomerNames, <span class="stringliteral">&#39;ribosomal protein &#39;</span>, <span class="stringliteral">&#39;&#39;</span>);
<a name="l00134"></a>00134             stateNames = {
<a name="l00135"></a>00135                 <span class="stringliteral">&#39;Rna&#39;</span>               <span class="stringliteral">&#39;counts&#39;</span>  r.matureIndexs(ribMatureRnaIdxs)   comp.cytosolIndexs
<a name="l00136"></a>00136                 <span class="stringliteral">&#39;ProteinMonomer&#39;</span>    <span class="stringliteral">&#39;counts&#39;</span>  pm.matureIndexs(ribMonomerIdxs)    comp.cytosolIndexs
<a name="l00137"></a>00137                 <span class="stringliteral">&#39;Mass&#39;</span>              <span class="stringliteral">&#39;proteinWt&#39;</span> <span class="charliteral">&#39;:&#39;</span> <span class="charliteral">&#39;:&#39;</span>
<a name="l00138"></a>00138                 };
<a name="l00139"></a>00139             states = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, <span class="stringliteral">&#39;extract&#39;</span>, selectedSimulations);
<a name="l00140"></a>00140             ribRNAs = states.Rna.counts;
<a name="l00141"></a>00141             ribMonomers = states.ProteinMonomer.counts;
<a name="l00142"></a>00142             proteinWt = states.Mass.proteinWt;
<a name="l00143"></a>00143             
<a name="l00144"></a>00144             <span class="keywordflow">for</span> i = 1:numel(simEndTimes)
<a name="l00145"></a>00145                 ribRNAs(:, :, simEndTimes(i)+1:end, i) = NaN;
<a name="l00146"></a>00146                 ribMonomers(:, :, simEndTimes(i)+1:end, i) = NaN;
<a name="l00147"></a>00147                 proteinWt(:, :, simEndTimes(i)+1:end, i) = NaN;
<a name="l00148"></a>00148             end
<a name="l00149"></a>00149             
<a name="l00150"></a>00150             clear states;
<a name="l00151"></a>00151             
<a name="l00152"></a>00152             %% clear figure
<a name="l00153"></a>00153             clf(figHandle);
<a name="l00154"></a>00154             
<a name="l00155"></a>00155             %% left panel
<a name="l00156"></a>00156             
<a name="l00157"></a>00157             %layout plots
<a name="l00158"></a>00158             axesHandlesL = PlotUtil.multiElementPlot(figHandle, 3.5 * ones(5, 1), [0 time(end)], struct(...
<a name="l00159"></a>00159                 &#39;titleStr&#39;, &#39;Composition&#39;, ...
<a name="l00160"></a>00160                 &#39;position&#39;, [0.12 0.52 0.21 0.4], ...
<a name="l00161"></a>00161                 &#39;colorOrder&#39;, PlotUtil.getRedGreenColorOrder(simOrder)));
<a name="l00162"></a>00162             
<a name="l00163"></a>00163             %plot data
<a name="l00164"></a>00164             PlotUtil.plotLine(axesHandlesL(1), time, permute(mass, [4 3 2 1]), false, true, false);
<a name="l00165"></a>00165             ylabel(axesHandlesL(1), {<span class="stringliteral">&#39;Mass&#39;</span> <span class="stringliteral">&#39;(fg)&#39;</span>});
<a name="l00166"></a>00166             
<a name="l00167"></a>00167             PlotUtil.plotLine(axesHandlesL(2), time, permute(ploidy, [4 3 2 1]), <span class="keyword">false</span>, <span class="keyword">true</span>, <span class="keyword">false</span>);
<a name="l00168"></a>00168             ylabel(axesHandlesL(2), {<span class="stringliteral">&#39;Chrosome&#39;</span> <span class="stringliteral">&#39;Copy&#39;</span> <span class="stringliteral">&#39;Number&#39;</span>});
<a name="l00169"></a>00169             
<a name="l00170"></a>00170             PlotUtil.plotLine(axesHandlesL(3), time, permute(rnas, [4 3 2 1]), <span class="keyword">false</span>, <span class="keyword">true</span>, <span class="keyword">false</span>);
<a name="l00171"></a>00171             ylabel(axesHandlesL(3), <span class="stringliteral">&#39;RNAs&#39;</span>);
<a name="l00172"></a>00172             
<a name="l00173"></a>00173             PlotUtil.plotLine(axesHandlesL(4), time, permute(proteins, [4 3 2 1]), <span class="keyword">false</span>, <span class="keyword">true</span>, <span class="keyword">false</span>);
<a name="l00174"></a>00174             ylabel(axesHandlesL(4), <span class="stringliteral">&#39;Proteins&#39;</span>);
<a name="l00175"></a>00175             
<a name="l00176"></a>00176             PlotUtil.plotLine(axesHandlesL(5), time, permute(lipids, [4 3 2 1]), <span class="keyword">false</span>, <span class="keyword">true</span>, <span class="keyword">false</span>);
<a name="l00177"></a>00177             ylabel(axesHandlesL(5), <span class="stringliteral">&#39;Lipids&#39;</span>);
<a name="l00178"></a>00178             
<a name="l00179"></a>00179             % Format Y-axis
<a name="l00180"></a>00180             PlotUtil.alignYAxesLabels(axesHandlesL);
<a name="l00181"></a>00181             PlotUtil.offsetYAxes(axesHandlesL, 0.04);
<a name="l00182"></a>00182             PlotUtil.labelSubplots(axesHandlesL(1), <span class="charliteral">&#39;A&#39;</span>, -0.42, 1.2);
<a name="l00183"></a>00183             
<a name="l00184"></a>00184             %% right panel
<a name="l00185"></a>00185             
<a name="l00186"></a>00186             %layout plots
<a name="l00187"></a>00187             axesHandlesR = PlotUtil.multiElementPlot(figHandle, 3.5 * ones(5, 1), [0 time(end)], <span class="keyword">struct</span>(...
<a name="l00188"></a>00188                 <span class="stringliteral">&#39;titleStr&#39;</span>, <span class="stringliteral">&#39;Growth&#39;</span>, ...
<a name="l00189"></a>00189                 <span class="stringliteral">&#39;position&#39;</span>, [0.45 0.52 0.21 0.4], ...
<a name="l00190"></a>00190                 <span class="stringliteral">&#39;colorOrder&#39;</span>, PlotUtil.getRedGreenColorOrder(simOrder)));
<a name="l00191"></a>00191             
<a name="l00192"></a>00192             %plot data
<a name="l00193"></a>00193             PlotUtil.plotLine(axesHandlesR(1), time, permute(growth, [4 3 2 1]), <span class="keyword">false</span>, <span class="keyword">true</span>, <span class="keyword">false</span>);
<a name="l00194"></a>00194             ylabel(axesHandlesR(1), {<span class="stringliteral">&#39;Growth&#39;</span> <span class="stringliteral">&#39;(fg h^{-1})&#39;</span>});
<a name="l00195"></a>00195             
<a name="l00196"></a>00196             PlotUtil.plotLine(axesHandlesR(2), time(3:end), diff(permute(sum(proteinWt, 2), [4 3 2 1]), 1, 2), <span class="keyword">false</span>, <span class="keyword">true</span>, <span class="keyword">false</span>);
<a name="l00197"></a>00197             ylabel(axesHandlesR(2), {<span class="stringliteral">&#39;Protein&#39;</span> <span class="stringliteral">&#39;Synthesis&#39;</span>});
<a name="l00198"></a>00198             
<a name="l00199"></a>00199             PlotUtil.plotLine(axesHandlesR(3), time, permute(ribosomes, [4 3 2 1]), <span class="keyword">false</span>, <span class="keyword">true</span>, <span class="keyword">false</span>);
<a name="l00200"></a>00200             ylabel(axesHandlesR(3), <span class="stringliteral">&#39;Ribosomes&#39;</span>);
<a name="l00201"></a>00201             
<a name="l00202"></a>00202             PlotUtil.plotLine(axesHandlesR(4), time(2:end), permute(ribRNAs, [4 3 2 1]), <span class="keyword">false</span>, <span class="keyword">true</span>, <span class="keyword">false</span>);
<a name="l00203"></a>00203             ylabel(axesHandlesR(4), [ribMonomerNames <span class="stringliteral">&#39;Protein&#39;</span>]);
<a name="l00204"></a>00204             
<a name="l00205"></a>00205             PlotUtil.plotLine(axesHandlesR(5), time(2:end), permute(ribMonomers, [4 3 2 1]), <span class="keyword">false</span>, <span class="keyword">true</span>, <span class="keyword">false</span>);
<a name="l00206"></a>00206             ylabel(axesHandlesR(5), [ribMonomerNames <span class="stringliteral">&#39;mRNA&#39;</span>]);
<a name="l00207"></a>00207             
<a name="l00208"></a>00208             % Format Y-axis
<a name="l00209"></a>00209             PlotUtil.alignYAxesLabels(axesHandlesR);
<a name="l00210"></a>00210             PlotUtil.offsetYAxes(axesHandlesR, 0.04);
<a name="l00211"></a>00211             PlotUtil.labelSubplots(axesHandlesR(1), <span class="charliteral">&#39;B&#39;</span>, -0.34, 1.2);
<a name="l00212"></a>00212             
<a name="l00213"></a>00213             %% organize figure data
<a name="l00214"></a>00214             figData = <span class="keyword">struct</span>;
<a name="l00215"></a>00215             figData.simOrder = simOrder;
<a name="l00216"></a>00216             figData.time = time;
<a name="l00217"></a>00217             figData.mass = mass;
<a name="l00218"></a>00218             figData.ploidy = ploidy;
<a name="l00219"></a>00219             figData.rnas = rnas;
<a name="l00220"></a>00220             figData.proteins = proteins;
<a name="l00221"></a>00221             figData.lipids = lipids;
<a name="l00222"></a>00222             figData.growth = growth;
<a name="l00223"></a>00223             figData.proteinWt = proteinWt;
<a name="l00224"></a>00224             figData.ribosomes = ribosomes;
<a name="l00225"></a>00225             figData.ribRNAs = ribRNAs;
<a name="l00226"></a>00226             figData.ribMonomers = ribMonomers;
<a name="l00227"></a>00227             figData.ribMonomerNames = ribMonomerNames;
<a name="l00228"></a>00228         end
<a name="l00229"></a>00229         
<a name="l00230"></a>00230         <span class="keyword">function</span> figData = energyProduction(figHandle, simBatchDir, selectedSimulations)
<a name="l00231"></a>00231             import edu.stanford.covert.cell.sim.util.PlotUtil;
<a name="l00232"></a>00232             import edu.stanford.covert.cell.sim.util.SimulationEnsemble;
<a name="l00233"></a>00233             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
<a name="l00234"></a>00234             
<a name="l00235"></a>00235             if nargin &lt; 2 || isempty(simBatchDir)
<a name="l00236"></a>00236                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
<a name="l00237"></a>00237             end
<a name="l00238"></a>00238             if nargin &lt; 3 || isempty(selectedSimulations)
<a name="l00239"></a>00239                 selectedSimulations = SimulationDiskUtil.getCompleteSimulations(simBatchDir);
<a name="l00240"></a>00240             end
<a name="l00241"></a>00241             
<a name="l00242"></a>00242             %% constants
<a name="l00243"></a>00243             [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep &#39;1&#39;]);
<a name="l00244"></a>00244             comp = sim.compartment;
<a name="l00245"></a>00245             m = sim.state(&#39;<a class="code" href="class_metabolite.html">Metabolite</a>&#39;);
<a name="l00246"></a>00246             met = sim.process(&#39;Metabolism&#39;);
<a name="l00247"></a>00247             mass = sim.state(&#39;Mass&#39;);
<a name="l00248"></a>00248             
<a name="l00249"></a>00249             metIdxs = sub2ind([numel(m.wholeCellModelIDs) comp.count], ...
<a name="l00250"></a>00250                 [m.ntpIndexs; m.ndpIndexs; m.nmpIndexs], repmat(comp.cytosolIndexs, 12, 1));
<a name="l00251"></a>00251             metProcessIdx = sim.processIndex(met.wholeCellModelID);
<a name="l00252"></a>00252             
<a name="l00253"></a>00253             %% get data
<a name="l00254"></a>00254             stateNames = {
<a name="l00255"></a>00255                 <span class="stringliteral">&#39;Time&#39;</span>               <span class="stringliteral">&#39;values&#39;</span>         <span class="charliteral">&#39;:&#39;</span>     <span class="charliteral">&#39;:&#39;</span>
<a name="l00256"></a>00256                 <span class="stringliteral">&#39;MetabolicReaction&#39;</span>  <span class="stringliteral">&#39;growth&#39;</span>         <span class="charliteral">&#39;:&#39;</span>     <span class="charliteral">&#39;:&#39;</span>
<a name="l00257"></a>00257                 <span class="stringliteral">&#39;Mass&#39;</span>               <span class="stringliteral">&#39;cell&#39;</span>           <span class="charliteral">&#39;:&#39;</span>     <span class="stringliteral">&#39;-sum&#39;</span>
<a name="l00258"></a>00258                 <span class="stringliteral">&#39;Metabolite&#39;</span>         <span class="stringliteral">&#39;processUsages&#39;</span>  metIdxs  metProcessIdx
<a name="l00259"></a>00259                 };
<a name="l00260"></a>00260             states = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, <span class="stringliteral">&#39;extract&#39;</span>, selectedSimulations);
<a name="l00261"></a>00261             simEndTimes = permute(max(states.Time.values, [], 3), [4 1 2 3]);
<a name="l00262"></a>00262             
<a name="l00263"></a>00263             timeIdx = min(simEndTimes);
<a name="l00264"></a>00264             [~, simOrder] = sort(states.Mass.cell(:, :, timeIdx, :));
<a name="l00265"></a>00265             simOrder = simOrder(:);
<a name="l00266"></a>00266             
<a name="l00267"></a>00267             time = permute(max(states.Time.values, [], 4), [3 4 1 2]) / 3600;
<a name="l00268"></a>00268             states.Metabolite.processUsages = -full(states.<a class="code" href="class_metabolite.html">Metabolite</a>.processUsages);
<a name="l00269"></a>00269             <span class="keywordflow">for</span> i = 1:numel(simEndTimes)
<a name="l00270"></a>00270                 states.<a class="code" href="class_metabolic_reaction.html">MetabolicReaction</a>.growth(:, :, simEndTimes(i)+1:end, i) = NaN;
<a name="l00271"></a>00271                 states.<a class="code" href="class_metabolite.html">Metabolite</a>.processUsages(:, :, simEndTimes(i)+1:end, i) = NaN;
<a name="l00272"></a>00272             end
<a name="l00273"></a>00273             
<a name="l00274"></a>00274             %% plot data
<a name="l00275"></a>00275             clf(figHandle);
<a name="l00276"></a>00276             options = struct;
<a name="l00277"></a>00277             options.position = [0.02 0.32 0.95 0.4];
<a name="l00278"></a>00278             options.titleStr = {<span class="stringliteral">&#39;Growth&#39;</span>; <span class="charliteral">&#39;A&#39;</span>; <span class="charliteral">&#39;C&#39;</span>; <span class="charliteral">&#39;G&#39;</span>; <span class="charliteral">&#39;U&#39;</span>};
<a name="l00279"></a>00279             options.colorOrder = PlotUtil.getRedGreenColorOrder(simOrder);
<a name="l00280"></a>00280             options.xdata = time;
<a name="l00281"></a>00281             options.ydata = {
<a name="l00282"></a>00282                 {permute(states.MetabolicReaction.growth * 3600 * mass.cellInitialDryWeight / (1 - mass.fractionWetWeight) * 1e15, [3 4 1 2])}
<a name="l00283"></a>00283                 permute(mat2cell(permute(states.Metabolite.processUsages(1:4:end, :, :, :), [3 4 1 2]), numel(time), numel(simEndTimes), ones(3, 1)), [3 1 2])
<a name="l00284"></a>00284                 permute(mat2cell(permute(states.<a class="code" href="class_metabolite.html">Metabolite</a>.processUsages(2:4:end, :, :, :), [3 4 1 2]), numel(time), numel(simEndTimes), ones(3, 1)), [3 1 2])
<a name="l00285"></a>00285                 permute(mat2cell(permute(states.<a class="code" href="class_metabolite.html">Metabolite</a>.processUsages(3:4:end, :, :, :), [3 4 1 2]), numel(time), numel(simEndTimes), ones(3, 1)), [3 1 2])
<a name="l00286"></a>00286                 permute(mat2cell(permute(states.<a class="code" href="class_metabolite.html">Metabolite</a>.processUsages(4:4:end, :, :, :), [3 4 1 2]), numel(time), numel(simEndTimes), ones(3, 1)), [3 1 2])
<a name="l00287"></a>00287                 };
<a name="l00288"></a>00288             options.ylabelStr = {
<a name="l00289"></a>00289                 {<span class="stringliteral">&#39;Growth (fg h^{-1})&#39;</span>}
<a name="l00290"></a>00290                 {<span class="stringliteral">&#39;ATP&#39;</span>; <span class="stringliteral">&#39;ADP&#39;</span>; <span class="stringliteral">&#39;AMP&#39;</span>}
<a name="l00291"></a>00291                 {<span class="stringliteral">&#39;CTP&#39;</span>; <span class="stringliteral">&#39;CDP&#39;</span>; <span class="stringliteral">&#39;CMP&#39;</span>}
<a name="l00292"></a>00292                 {<span class="stringliteral">&#39;GTP&#39;</span>; <span class="stringliteral">&#39;GDP&#39;</span>; <span class="stringliteral">&#39;GMP&#39;</span>}
<a name="l00293"></a>00293                 {<span class="stringliteral">&#39;UTP&#39;</span>; <span class="stringliteral">&#39;UDP&#39;</span>; <span class="stringliteral">&#39;UMP&#39;</span>}
<a name="l00294"></a>00294                 };
<a name="l00295"></a>00295             PlotUtil.multiElementPlot(figHandle, {30 [10; 10; 10] [10; 10; 10] [10; 10; 10] [10; 10; 10]}, [0 time(end)], options);
<a name="l00296"></a>00296             
<a name="l00297"></a>00297             clear options;
<a name="l00298"></a>00298             
<a name="l00299"></a>00299             %% organize figure data
<a name="l00300"></a>00300             figData = <span class="keyword">struct</span>;
<a name="l00301"></a>00301             figData.simOrder = simOrder;
<a name="l00302"></a>00302             figData.time = time;
<a name="l00303"></a>00303             figData.growth = states.MetabolicReaction.growth;
<a name="l00304"></a>00304             figData.processUsages = states.<a class="code" href="class_metabolite.html#a0d150f0c40eaa35d260c8d10c9805245">Metabolite</a>.processUsages;
<a name="l00305"></a>00305         end
<a name="l00306"></a>00306         
<a name="l00307"></a>00307         <span class="keyword">function</span> figData = cellMassDistribution(figHandle, simBatchDir, selectedSimulations)
<a name="l00308"></a>00308             import edu.stanford.covert.cell.sim.util.PlotUtil;
<a name="l00309"></a>00309             import edu.stanford.covert.cell.sim.util.SimulationEnsemble;
<a name="l00310"></a>00310             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
<a name="l00311"></a>00311             import edu.stanford.covert.util.<a class="code" href="class_constant_util.html">ConstantUtil</a>;
<a name="l00312"></a>00312             
<a name="l00313"></a>00313             if nargin &lt; 2 || isempty(simBatchDir)
<a name="l00314"></a>00314                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
<a name="l00315"></a>00315             end
<a name="l00316"></a>00316             if nargin &lt; 3 || isempty(selectedSimulations)
<a name="l00317"></a>00317                 selectedSimulations = SimulationDiskUtil.getCompleteSimulations(simBatchDir);
<a name="l00318"></a>00318             end
<a name="l00319"></a>00319             
<a name="l00320"></a>00320             %% get data
<a name="l00321"></a>00321             stateNames = {
<a name="l00322"></a>00322                 <span class="stringliteral">&#39;mass&#39;</span> <span class="stringliteral">&#39;Mass&#39;</span>
<a name="l00323"></a>00323                 };
<a name="l00324"></a>00324             ensemble = SimulationEnsemble(simBatchDir, stateNames, [], selectedSimulations);
<a name="l00325"></a>00325             mass = ensemble.stateData.values(ensemble.getPropertyIndices(<span class="stringliteral">&#39;mass&#39;</span>), :, :, :);
<a name="l00326"></a>00326             
<a name="l00327"></a>00327             simEndTimes = ensemble.stateData.simulationEndTimes;
<a name="l00328"></a>00328             
<a name="l00329"></a>00329             initMass = permute(mass(:, :, 1, :), [4 3 1 2]);
<a name="l00330"></a>00330             finMass = mass(sub2ind(size(mass), ...
<a name="l00331"></a>00331                 ones(size(simEndTimes)), ...
<a name="l00332"></a>00332                 ones(size(simEndTimes)), ...
<a name="l00333"></a>00333                 simEndTimes, ...
<a name="l00334"></a>00334                 (1:numel(simEndTimes))<span class="stringliteral">&#39;));</span>
<a name="l00335"></a>00335 <span class="stringliteral">            </span>
<a name="l00336"></a>00336 <span class="stringliteral">            %% plot data</span>
<a name="l00337"></a>00337 <span class="stringliteral">            clf(figHandle);</span>
<a name="l00338"></a>00338 <span class="stringliteral">            </span>
<a name="l00339"></a>00339 <span class="stringliteral">            mass = linspace(min([initMass; finMass - 1]), max([initMass; finMass - 1]), 20);</span>
<a name="l00340"></a>00340 <span class="stringliteral">            initMassFreq = histc(initMass, mass);</span>
<a name="l00341"></a>00341 <span class="stringliteral">            finMassFreq = histc(finMass - 1, mass);</span>
<a name="l00342"></a>00342 <span class="stringliteral">            </span>
<a name="l00343"></a>00343 <span class="stringliteral">            axesHandle = subplot(2, 1, 1, &#39;</span>Parent<span class="stringliteral">&#39;, figHandle);</span>
<a name="l00344"></a>00344 <span class="stringliteral">            bar(axesHandle, mass, initMassFreq);</span>
<a name="l00345"></a>00345 <span class="stringliteral">            xlabel(axesHandle, &#39;</span>Initial Mass (cell dry weight)<span class="stringliteral">&#39;);</span>
<a name="l00346"></a>00346 <span class="stringliteral">            ylabel(axesHandle, &#39;</span>Frequency<span class="stringliteral">&#39;);</span>
<a name="l00347"></a>00347 <span class="stringliteral">            </span>
<a name="l00348"></a>00348 <span class="stringliteral">            axesHandle = subplot(2, 1, 2, &#39;</span>Parent<span class="stringliteral">&#39;, figHandle);</span>
<a name="l00349"></a>00349 <span class="stringliteral">            bar(axesHandle, mass + 1, finMassFreq);</span>
<a name="l00350"></a>00350 <span class="stringliteral">            xlabel(axesHandle, &#39;</span>Final Mass (cell dry weight)<span class="stringliteral">&#39;);</span>
<a name="l00351"></a>00351 <span class="stringliteral">            ylabel(axesHandle, &#39;</span>Frequency<span class="stringliteral">&#39;);</span>
<a name="l00352"></a>00352 <span class="stringliteral">            </span>
<a name="l00353"></a>00353 <span class="stringliteral">            %% figure data</span>
<a name="l00354"></a>00354 <span class="stringliteral">            figData = struct;</span>
<a name="l00355"></a>00355 <span class="stringliteral">            figData.initMass = initMass;</span>
<a name="l00356"></a>00356 <span class="stringliteral">            figData.finMass = finMass;</span>
<a name="l00357"></a>00357 <span class="stringliteral">        end</span>
<a name="l00358"></a>00358 <span class="stringliteral">        </span>
<a name="l00359"></a>00359 <span class="stringliteral">        function figData = massDistribution(figHandle, simBatchDir, selectedSimulations)</span>
<a name="l00360"></a>00360 <span class="stringliteral">            import edu.stanford.covert.cell.sim.util.PlotUtil;</span>
<a name="l00361"></a>00361 <span class="stringliteral">            import edu.stanford.covert.cell.sim.util.SimulationEnsemble;</span>
<a name="l00362"></a>00362 <span class="stringliteral">            import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;</span>
<a name="l00363"></a>00363 <span class="stringliteral">            import edu.stanford.covert.util.ConstantUtil;</span>
<a name="l00364"></a>00364 <span class="stringliteral">            </span>
<a name="l00365"></a>00365 <span class="stringliteral">            if nargin &lt; 2 || isempty(simBatchDir)</span>
<a name="l00366"></a>00366 <span class="stringliteral">                simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();</span>
<a name="l00367"></a>00367 <span class="stringliteral">            end</span>
<a name="l00368"></a>00368 <span class="stringliteral">            if nargin &lt; 3 || isempty(selectedSimulations)</span>
<a name="l00369"></a>00369 <span class="stringliteral">                selectedSimulations = SimulationDiskUtil.getCompleteSimulations(simBatchDir);</span>
<a name="l00370"></a>00370 <span class="stringliteral">            end</span>
<a name="l00371"></a>00371 <span class="stringliteral">            </span>
<a name="l00372"></a>00372 <span class="stringliteral">            [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep &#39;</span>1<span class="stringliteral">&#39;]);</span>
<a name="l00373"></a>00373 <span class="stringliteral">            comp = sim.compartment;</span>
<a name="l00374"></a>00374 <span class="stringliteral">            </span>
<a name="l00375"></a>00375 <span class="stringliteral">            %% get data</span>
<a name="l00376"></a>00376 <span class="stringliteral">            stateNames = {</span>
<a name="l00377"></a>00377 <span class="stringliteral">                &#39;</span>Time<span class="stringliteral">&#39;      &#39;</span>values<span class="stringliteral">&#39;             &#39;</span>:<span class="charliteral">&#39; &#39;</span>:<span class="stringliteral">&#39;</span>
<a name="l00378"></a>00378 <span class="stringliteral">                &#39;</span>Mass<span class="stringliteral">&#39;      &#39;</span>cell<span class="stringliteral">&#39;               &#39;</span>:<span class="charliteral">&#39; &#39;</span>-sum<span class="stringliteral">&#39;</span>
<a name="l00379"></a>00379 <span class="stringliteral">                &#39;</span>Mass<span class="stringliteral">&#39;      &#39;</span>dnaWt<span class="stringliteral">&#39;              &#39;</span>:<span class="stringliteral">&#39; comp.cytosolIndexs</span>
<a name="l00380"></a>00380 <span class="stringliteral">                &#39;</span>Mass<span class="stringliteral">&#39;      &#39;</span>rnaWt<span class="stringliteral">&#39;              &#39;</span>:<span class="stringliteral">&#39; comp.cytosolIndexs</span>
<a name="l00381"></a>00381 <span class="stringliteral">                &#39;</span>Mass<span class="stringliteral">&#39;      &#39;</span>proteinWt<span class="stringliteral">&#39;          &#39;</span>:<span class="charliteral">&#39; &#39;</span>-sum<span class="stringliteral">&#39;</span>
<a name="l00382"></a>00382 <span class="stringliteral">                &#39;</span>Mass<span class="stringliteral">&#39;      &#39;</span>metaboliteWt<span class="stringliteral">&#39;       &#39;</span>:<span class="stringliteral">&#39; [comp.cytosolIndexs; comp.membraneIndexs]</span>
<a name="l00383"></a>00383 <span class="stringliteral">                };</span>
<a name="l00384"></a>00384 <span class="stringliteral">            states = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, &#39;</span>extract<span class="stringliteral">&#39;, selectedSimulations);</span>
<a name="l00385"></a>00385 <span class="stringliteral">            </span>
<a name="l00386"></a>00386 <span class="stringliteral">            time = permute(max(states.Time.values, [], 4), [3 4 1 2]) / 3600;</span>
<a name="l00387"></a>00387 <span class="stringliteral">            mass = permute(states.Mass.cell, [3 4 1 2]) * 1e15;</span>
<a name="l00388"></a>00388 <span class="stringliteral">            dnaWt = permute(states.Mass.dnaWt, [3 4 1 2]) * 1e15;</span>
<a name="l00389"></a>00389 <span class="stringliteral">            rnaWt = permute(states.Mass.rnaWt, [3 4 1 2]) * 1e15;</span>
<a name="l00390"></a>00390 <span class="stringliteral">            proteinWt = permute(states.Mass.proteinWt, [3 4 1 2]) * 1e15;</span>
<a name="l00391"></a>00391 <span class="stringliteral">            membraneWt = permute(states.Mass.metaboliteWt(:, 2, :, :), [3 4 1 2]) * 1e15;</span>
<a name="l00392"></a>00392 <span class="stringliteral">            metaboliteWt = permute(states.Mass.metaboliteWt(:, 1, :, :), [3 4 1 2]) * 1e15;</span>
<a name="l00393"></a>00393 <span class="stringliteral">            </span>
<a name="l00394"></a>00394 <span class="stringliteral">            dnaWt(mass == 0) = NaN;</span>
<a name="l00395"></a>00395 <span class="stringliteral">            rnaWt(mass == 0) = NaN;</span>
<a name="l00396"></a>00396 <span class="stringliteral">            proteinWt(mass == 0) = NaN;</span>
<a name="l00397"></a>00397 <span class="stringliteral">            membraneWt(mass == 0) = NaN;</span>
<a name="l00398"></a>00398 <span class="stringliteral">            metaboliteWt(mass == 0) = NaN;</span>
<a name="l00399"></a>00399 <span class="stringliteral">            mass(mass == 0) = NaN;</span>
<a name="l00400"></a>00400 <span class="stringliteral">            </span>
<a name="l00401"></a>00401 <span class="stringliteral">            minSimEndTime = min(max(states.Time.values, [], 4));</span>
<a name="l00402"></a>00402 <span class="stringliteral">            [~, simOrder] = sort(mass(minSimEndTime, :));</span>
<a name="l00403"></a>00403 <span class="stringliteral">            </span>
<a name="l00404"></a>00404 <span class="stringliteral">            clear states;</span>
<a name="l00405"></a>00405 <span class="stringliteral">            </span>
<a name="l00406"></a>00406 <span class="stringliteral">            %% layout figure</span>
<a name="l00407"></a>00407 <span class="stringliteral">            clf(figHandle);</span>
<a name="l00408"></a>00408 <span class="stringliteral">            </span>
<a name="l00409"></a>00409 <span class="stringliteral">            options = struct();</span>
<a name="l00410"></a>00410 <span class="stringliteral">            options.titleStr = &#39;</span>Cell Mass<span class="stringliteral">&#39;;</span>
<a name="l00411"></a>00411 <span class="stringliteral">            options.colorOrder = PlotUtil.getRedGreenColorOrder(simOrder);</span>
<a name="l00412"></a>00412 <span class="stringliteral">            axesHandles = PlotUtil.multiElementPlot(figHandle, 3 * ones(6, 1), [0 time(end)], options);</span>
<a name="l00413"></a>00413 <span class="stringliteral">            </span>
<a name="l00414"></a>00414 <span class="stringliteral">            %% plot data</span>
<a name="l00415"></a>00415 <span class="stringliteral">            </span>
<a name="l00416"></a>00416 <span class="stringliteral">            PlotUtil.plotLine(axesHandles(1), time, mass, false, true, false);</span>
<a name="l00417"></a>00417 <span class="stringliteral">            ylabel(axesHandles(1), &#39;</span>Mass (fg)<span class="stringliteral">&#39;);</span>
<a name="l00418"></a>00418 <span class="stringliteral">            </span>
<a name="l00419"></a>00419 <span class="stringliteral">            PlotUtil.plotLine(axesHandles(2), time, dnaWt, false, true, false);</span>
<a name="l00420"></a>00420 <span class="stringliteral">            ylabel(axesHandles(2), &#39;</span>DNA (fg)<span class="stringliteral">&#39;);</span>
<a name="l00421"></a>00421 <span class="stringliteral">            </span>
<a name="l00422"></a>00422 <span class="stringliteral">            PlotUtil.plotLine(axesHandles(3), time, rnaWt, false, true, false);</span>
<a name="l00423"></a>00423 <span class="stringliteral">            ylabel(axesHandles(3), &#39;</span><a class="code" href="class_r_n_a.html">RNA</a> (fg)<span class="stringliteral">&#39;);</span>
<a name="l00424"></a>00424 <span class="stringliteral">            </span>
<a name="l00425"></a>00425 <span class="stringliteral">            PlotUtil.plotLine(axesHandles(4), time, proteinWt, false, true, false);</span>
<a name="l00426"></a>00426 <span class="stringliteral">            ylabel(axesHandles(4), &#39;</span><a class="code" href="class_protein.html">Protein</a> (fg)<span class="stringliteral">&#39;);</span>
<a name="l00427"></a>00427 <span class="stringliteral">            </span>
<a name="l00428"></a>00428 <span class="stringliteral">            PlotUtil.plotLine(axesHandles(5), time, membraneWt, false, true, false);</span>
<a name="l00429"></a>00429 <span class="stringliteral">            ylabel(axesHandles(5), &#39;</span>Membrane (fg)<span class="stringliteral">&#39;);</span>
<a name="l00430"></a>00430 <span class="stringliteral">            </span>
<a name="l00431"></a>00431 <span class="stringliteral">            PlotUtil.plotLine(axesHandles(6), time, metaboliteWt, false, true, false);</span>
<a name="l00432"></a>00432 <span class="stringliteral">            ylabel(axesHandles(6), &#39;</span>Metabolites (fg)<span class="stringliteral">&#39;);</span>
<a name="l00433"></a>00433 <span class="stringliteral">            </span>
<a name="l00434"></a>00434 <span class="stringliteral">            %% Format Y-axis</span>
<a name="l00435"></a>00435 <span class="stringliteral">            PlotUtil.alignYAxesLabels(axesHandles);</span>
<a name="l00436"></a>00436 <span class="stringliteral">            PlotUtil.offsetYAxes(axesHandles, 0.03);</span>
<a name="l00437"></a>00437 <span class="stringliteral">            </span>
<a name="l00438"></a>00438 <span class="stringliteral">            %% organize figure data</span>
<a name="l00439"></a>00439 <span class="stringliteral">            figData = struct;</span>
<a name="l00440"></a>00440 <span class="stringliteral">            figData.simOrder = simOrder;</span>
<a name="l00441"></a>00441 <span class="stringliteral">            figData.time = time;</span>
<a name="l00442"></a>00442 <span class="stringliteral">            figData.mass = mass;</span>
<a name="l00443"></a>00443 <span class="stringliteral">            figData.dnaWt = dnaWt;</span>
<a name="l00444"></a>00444 <span class="stringliteral">            figData.rnaWt = rnaWt;</span>
<a name="l00445"></a>00445 <span class="stringliteral">            figData.proteinWt = proteinWt;</span>
<a name="l00446"></a>00446 <span class="stringliteral">            figData.membraneWt = membraneWt;</span>
<a name="l00447"></a>00447 <span class="stringliteral">            figData.metaboliteWt = metaboliteWt;</span>
<a name="l00448"></a>00448 <span class="stringliteral">        end</span>
<a name="l00449"></a>00449 <span class="stringliteral">        </span>
<a name="l00450"></a>00450 <span class="stringliteral">        function figData = cellCyclePhases(figHandle, simBatchDir, selectedSimulations)</span>
<a name="l00451"></a>00451 <span class="stringliteral">            import edu.stanford.covert.cell.sim.util.SummaryLogger;</span>
<a name="l00452"></a>00452 <span class="stringliteral">            import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;</span>
<a name="l00453"></a>00453 <span class="stringliteral">            import edu.stanford.covert.cell.sim.util.SimulationEnsemble;</span>
<a name="l00454"></a>00454 <span class="stringliteral">            import edu.stanford.covert.cell.sim.util.PlotUtil;</span>
<a name="l00455"></a>00455 <span class="stringliteral">            </span>
<a name="l00456"></a>00456 <span class="stringliteral">            if nargin &lt; 2 || isempty(simBatchDir)</span>
<a name="l00457"></a>00457 <span class="stringliteral">                simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();</span>
<a name="l00458"></a>00458 <span class="stringliteral">            end</span>
<a name="l00459"></a>00459 <span class="stringliteral">            if nargin &lt; 3 || isempty(selectedSimulations)</span>
<a name="l00460"></a>00460 <span class="stringliteral">                selectedSimulations = SimulationDiskUtil.getCompleteSimulations(simBatchDir);</span>
<a name="l00461"></a>00461 <span class="stringliteral">            end</span>
<a name="l00462"></a>00462 <span class="stringliteral">            </span>
<a name="l00463"></a>00463 <span class="stringliteral">            %% get data</span>
<a name="l00464"></a>00464 <span class="stringliteral">            [simDir, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep &#39;</span>1<span class="stringliteral">&#39;]);</span>
<a name="l00465"></a>00465 <span class="stringliteral">            simBatchDir = simDir(1:find(simDir == filesep, 1, &#39;</span>last<span class="stringliteral">&#39;) - 1);</span>
<a name="l00466"></a>00466 <span class="stringliteral">            mass = sim.state(&#39;</span>Mass<span class="stringliteral">&#39;);</span>
<a name="l00467"></a>00467 <span class="stringliteral">            stats = SummaryLogger.getSimulationStatistics(simBatchDir);</span>
<a name="l00468"></a>00468 <span class="stringliteral">            nSims = size(stats, 1);</span>
<a name="l00469"></a>00469 <span class="stringliteral">            </span>
<a name="l00470"></a>00470 <span class="stringliteral">            ensemble = SimulationEnsemble(simBatchDir, cell(0, 2), [], selectedSimulations);</span>
<a name="l00471"></a>00471 <span class="stringliteral">            maxTime = ensemble.stateData.time(end);</span>
<a name="l00472"></a>00472 <span class="stringliteral">            </span>
<a name="l00473"></a>00473 <span class="stringliteral">            clear ensemble;</span>
<a name="l00474"></a>00474 <span class="stringliteral">            </span>
<a name="l00475"></a>00475 <span class="stringliteral">            %% clear figure</span>
<a name="l00476"></a>00476 <span class="stringliteral">            clf(figHandle);</span>
<a name="l00477"></a>00477 <span class="stringliteral">            </span>
<a name="l00478"></a>00478 <span class="stringliteral">            %% top panel</span>
<a name="l00479"></a>00479 <span class="stringliteral">            </span>
<a name="l00480"></a>00480 <span class="stringliteral">            %layout plots</span>
<a name="l00481"></a>00481 <span class="stringliteral">            edges = 15 * 60 * (0 : ceil(maxTime / (15 * 60)));</span>
<a name="l00482"></a>00482 <span class="stringliteral">            n1 = histc(stats(:, SummaryLogger.SIM_STATUS_INDEX_CYTOKINESIS_TIME), edges);</span>
<a name="l00483"></a>00483 <span class="stringliteral">            n2 = histc(stats(:, SummaryLogger.SIM_STATUS_INDEX_REP_INIT_TIME), edges);</span>
<a name="l00484"></a>00484 <span class="stringliteral">            n3 = histc(stats(:, SummaryLogger.SIM_STATUS_INDEX_REP_TIME) - stats(:, SummaryLogger.SIM_STATUS_INDEX_REP_INIT_TIME), edges);</span>
<a name="l00485"></a>00485 <span class="stringliteral">            n4 = histc(stats(:, SummaryLogger.SIM_STATUS_INDEX_CYTOKINESIS_TIME) - stats(:, SummaryLogger.SIM_STATUS_INDEX_REP_TIME), edges);</span>
<a name="l00486"></a>00486 <span class="stringliteral">            </span>
<a name="l00487"></a>00487 <span class="stringliteral">            [axesHandlesT, xAxesHandleT] = PlotUtil.multiElementPlot(figHandle, 3 * ones(4, 1), edges([1 end]) / 3600&#39;</span>, ...
<a name="l00488"></a>00488                 <span class="keyword">struct</span>(<span class="stringliteral">&#39;position&#39;</span>, [0.12 0.68 0.21 0.24], <span class="stringliteral">&#39;xlabelStr&#39;</span>, <span class="stringliteral">&#39;Cell Cycle Phase Duration (h)&#39;</span>)); %#ok&lt;NASGU&gt;
<a name="l00489"></a>00489             
<a name="l00490"></a>00490             %plot data
<a name="l00491"></a>00491             bar(axesHandlesT(1), edges / 3600, n1 / nSims * 100, <span class="stringliteral">&#39;histc&#39;</span>);
<a name="l00492"></a>00492             ylabel(axesHandlesT(1), {<span class="stringliteral">&#39;Cell&#39;</span> <span class="stringliteral">&#39;Cycle&#39;</span>});
<a name="l00493"></a>00493             ytick = <span class="keyword">get</span>(axesHandlesT(1), <span class="stringliteral">&#39;YTick&#39;</span>);
<a name="l00494"></a>00494             <span class="keyword">set</span>(axesHandlesT(1), <span class="stringliteral">&#39;YTick&#39;</span>, ytick([1 end]));
<a name="l00495"></a>00495             ylim = <span class="keyword">get</span>(axesHandlesT(1), <span class="stringliteral">&#39;YLim&#39;</span>);
<a name="l00496"></a>00496             <span class="keyword">set</span>(axesHandlesT(1), <span class="stringliteral">&#39;YLim&#39;</span>, ylim);
<a name="l00497"></a>00497             
<a name="l00498"></a>00498             bar(axesHandlesT(2), edges / 3600, n2 / nSims * 100, <span class="stringliteral">&#39;histc&#39;</span>);
<a name="l00499"></a>00499             ylabel(axesHandlesT(2), {<span class="stringliteral">&#39;Rep&#39;</span> <span class="stringliteral">&#39;Init&#39;</span>});
<a name="l00500"></a>00500             ytick = <span class="keyword">get</span>(axesHandlesT(2), <span class="stringliteral">&#39;YTick&#39;</span>);
<a name="l00501"></a>00501             <span class="keyword">set</span>(axesHandlesT(2), <span class="stringliteral">&#39;YTick&#39;</span>, ytick([1 end]));
<a name="l00502"></a>00502             ylim = <span class="keyword">get</span>(axesHandlesT(2), <span class="stringliteral">&#39;YLim&#39;</span>);
<a name="l00503"></a>00503             <span class="keyword">set</span>(axesHandlesT(2), <span class="stringliteral">&#39;YLim&#39;</span>, ylim);
<a name="l00504"></a>00504             
<a name="l00505"></a>00505             bar(axesHandlesT(3), edges / 3600, n3 / nSims * 100, <span class="stringliteral">&#39;histc&#39;</span>);
<a name="l00506"></a>00506             ylabel(axesHandlesT(3), {<span class="stringliteral">&#39;Repli-&#39;</span> <span class="stringliteral">&#39;cation&#39;</span>});
<a name="l00507"></a>00507             ytick = <span class="keyword">get</span>(axesHandlesT(3), <span class="stringliteral">&#39;YTick&#39;</span>);
<a name="l00508"></a>00508             <span class="keyword">set</span>(axesHandlesT(3), <span class="stringliteral">&#39;YTick&#39;</span>, ytick([1 end]));
<a name="l00509"></a>00509             ylim = <span class="keyword">get</span>(axesHandlesT(3), <span class="stringliteral">&#39;YLim&#39;</span>);
<a name="l00510"></a>00510             <span class="keyword">set</span>(axesHandlesT(3), <span class="stringliteral">&#39;YLim&#39;</span>, ylim);
<a name="l00511"></a>00511             
<a name="l00512"></a>00512             bar(axesHandlesT(4), edges / 3600, n4 / nSims * 100, <span class="stringliteral">&#39;histc&#39;</span>);
<a name="l00513"></a>00513             ylabel(axesHandlesT(4), {<span class="stringliteral">&#39;Cyto-&#39;</span> <span class="stringliteral">&#39;kinesis&#39;</span>});
<a name="l00514"></a>00514             ytick = <span class="keyword">get</span>(axesHandlesT(4), <span class="stringliteral">&#39;YTick&#39;</span>);
<a name="l00515"></a>00515             <span class="keyword">set</span>(axesHandlesT(4), <span class="stringliteral">&#39;YTick&#39;</span>, ytick([1 end]));
<a name="l00516"></a>00516             ylim = <span class="keyword">get</span>(axesHandlesT(4), <span class="stringliteral">&#39;YLim&#39;</span>);
<a name="l00517"></a>00517             <span class="keyword">set</span>(axesHandlesT(4), <span class="stringliteral">&#39;YLim&#39;</span>, ylim);
<a name="l00518"></a>00518             
<a name="l00519"></a>00519             %xlabelPos = <span class="keyword">get</span>(<span class="keyword">get</span>(xAxesHandleT, <span class="stringliteral">&#39;XLabel&#39;</span>), <span class="stringliteral">&#39;Position&#39;</span>);
<a name="l00520"></a>00520             %<span class="keyword">set</span>(<span class="keyword">get</span>(xAxesHandleT, <span class="stringliteral">&#39;XLabel&#39;</span>), <span class="stringliteral">&#39;Position&#39;</span>, xlabelPos+[0 17 0]);
<a name="l00521"></a>00521             
<a name="l00522"></a>00522             %% bottom panel
<a name="l00523"></a>00523             edges = linspace(min(stats(:, SummaryLogger.SIM_STATUS_INDEX_FIN_MASS)), max(stats(:, SummaryLogger.SIM_STATUS_INDEX_FIN_MASS)), 50);
<a name="l00524"></a>00524             n = histc(stats(:, SummaryLogger.SIM_STATUS_INDEX_FIN_MASS), edges);
<a name="l00525"></a>00525             
<a name="l00526"></a>00526             [axesHandlesB, xAxesHandleB] = PlotUtil.multiElementPlot(figHandle, 3, ...
<a name="l00527"></a>00527                 edges([1 end]) * mass.cellInitialDryWeight / (1-mass.fractionWetWeight) * 1e15, ...
<a name="l00528"></a>00528                 <span class="keyword">struct</span>(<span class="stringliteral">&#39;position&#39;</span>, [0.12 0.52 0.21 0.06], <span class="stringliteral">&#39;xlabelStr&#39;</span>, <span class="stringliteral">&#39;Cell Mass (fg)&#39;</span>)); %#ok&lt;NASGU&gt;
<a name="l00529"></a>00529             
<a name="l00530"></a>00530             bar(axesHandlesB(1), edges * mass.cellInitialDryWeight / (1-mass.fractionWetWeight) * 1e15, n / nSims * 100, <span class="stringliteral">&#39;histc&#39;</span>);
<a name="l00531"></a>00531             ylabel(axesHandlesB(1), <span class="stringliteral">&#39;Freq&#39;</span>);
<a name="l00532"></a>00532             ytick = <span class="keyword">get</span>(axesHandlesB(1), <span class="stringliteral">&#39;YTick&#39;</span>);
<a name="l00533"></a>00533             <span class="keyword">set</span>(axesHandlesB(1), <span class="stringliteral">&#39;YTick&#39;</span>, ytick([1 end]));
<a name="l00534"></a>00534             ylim = <span class="keyword">get</span>(axesHandlesB(1), <span class="stringliteral">&#39;YLim&#39;</span>);
<a name="l00535"></a>00535             <span class="keyword">set</span>(axesHandlesB(1), <span class="stringliteral">&#39;YLim&#39;</span>, ylim);
<a name="l00536"></a>00536             
<a name="l00537"></a>00537             %xlabelPos = <span class="keyword">get</span>(<span class="keyword">get</span>(xAxesHandleB, <span class="stringliteral">&#39;XLabel&#39;</span>), <span class="stringliteral">&#39;Position&#39;</span>);
<a name="l00538"></a>00538             %<span class="keyword">set</span>(<span class="keyword">get</span>(xAxesHandleB, <span class="stringliteral">&#39;XLabel&#39;</span>), <span class="stringliteral">&#39;Position&#39;</span>, xlabelPos+[0 15 0]);
<a name="l00539"></a>00539             
<a name="l00540"></a>00540             %% Format Y-axis
<a name="l00541"></a>00541             PlotUtil.alignYAxesLabels(axesHandlesT);
<a name="l00542"></a>00542             PlotUtil.offsetYAxes([axesHandlesT; axesHandlesB], 0.04);
<a name="l00543"></a>00543             PlotUtil.labelSubplots(axesHandlesT(1), <span class="charliteral">&#39;A&#39;</span>, -0.25, 1.2);
<a name="l00544"></a>00544             PlotUtil.labelSubplots(axesHandlesB(1), <span class="charliteral">&#39;B&#39;</span>, -0.25, 1.2);
<a name="l00545"></a>00545             
<a name="l00546"></a>00546             %% organize figure data
<a name="l00547"></a>00547             figData = <span class="keyword">struct</span>;
<a name="l00548"></a>00548             figData.nSims = nSims;
<a name="l00549"></a>00549             figData.maxTime = maxTime;
<a name="l00550"></a>00550             figData.mass = mass;
<a name="l00551"></a>00551             figData.stats = stats;
<a name="l00552"></a>00552         end
<a name="l00553"></a>00553         
<a name="l00554"></a>00554         <span class="keyword">function</span> [figData, tabContent, tabColLabels] = nucleotideDistribution(figHandle, simBatchDir, selectedSimulations)
<a name="l00555"></a>00555             import edu.stanford.covert.cell.sim.util.MacromoleculeUtil;
<a name="l00556"></a>00556             import edu.stanford.covert.cell.sim.util.SimulationEnsemble;
<a name="l00557"></a>00557             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
<a name="l00558"></a>00558             import edu.stanford.covert.cell.sim.util.PlotUtil;
<a name="l00559"></a>00559             import edu.stanford.covert.cell.sim.analysis.<a class="code" href="class_cell_overview.html">CellOverview</a>;
<a name="l00560"></a>00560             import edu.stanford.covert.util.<a class="code" href="class_constant_util.html">ConstantUtil</a>;
<a name="l00561"></a>00561             
<a name="l00562"></a>00562             if nargin &lt; 2 || isempty(simBatchDir)
<a name="l00563"></a>00563                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
<a name="l00564"></a>00564             end
<a name="l00565"></a>00565             if nargin &lt; 3 || isempty(selectedSimulations)
<a name="l00566"></a>00566                 selectedSimulations = SimulationDiskUtil.getCompleteSimulations(simBatchDir);
<a name="l00567"></a>00567             end
<a name="l00568"></a>00568             
<a name="l00569"></a>00569             [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep &#39;1&#39;]);
<a name="l00570"></a>00570             comp = sim.compartment;
<a name="l00571"></a>00571             m = sim.state(&#39;<a class="code" href="class_metabolite.html">Metabolite</a>&#39;);
<a name="l00572"></a>00572             
<a name="l00573"></a>00573             %% get data
<a name="l00574"></a>00574             stateNames = {
<a name="l00575"></a>00575                 <span class="stringliteral">&#39;mass&#39;</span> <span class="stringliteral">&#39;Mass&#39;</span>
<a name="l00576"></a>00576                 };
<a name="l00577"></a>00577             ensemble = SimulationEnsemble(simBatchDir, stateNames, [], selectedSimulations);            
<a name="l00578"></a>00578             mass = ensemble.stateData.values(ensemble.getPropertyIndices(<span class="stringliteral">&#39;mass&#39;</span>), :, :, :) * sim.state(<span class="stringliteral">&#39;Mass&#39;</span>).cellInitialDryWeight / (1-sim.state(&#39;Mass&#39;).fractionWetWeight) * 1e15;
<a name="l00579"></a>00579             
<a name="l00580"></a>00580             simEndTimes = ensemble.stateData.simulationEndTimes;
<a name="l00581"></a>00581             timeIdx = ceil(min(simEndTimes) / ensemble.stateData.downsampleStepSec);
<a name="l00582"></a>00582             [~, simOrder] = sort(mass(:, 1, timeIdx, :));
<a name="l00583"></a>00583             simOrder = simOrder(:);
<a name="l00584"></a>00584             
<a name="l00585"></a>00585             clear ensemble mass;
<a name="l00586"></a>00586             
<a name="l00587"></a>00587             stateNames = {
<a name="l00588"></a>00588                 <span class="stringliteral">&#39;Time&#39;</span>       <span class="stringliteral">&#39;values&#39;</span> <span class="charliteral">&#39;:&#39;</span>  <span class="charliteral">&#39;:&#39;</span>
<a name="l00589"></a>00589                 <span class="stringliteral">&#39;Metabolite&#39;</span> <span class="stringliteral">&#39;counts&#39;</span> [m.ntpIndexs; m.ndpIndexs; m.nmpIndexs; m.aminoAcidIndexs] comp.cytosolIndexs
<a name="l00590"></a>00590                 };
<a name="l00591"></a>00591             states = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, <span class="stringliteral">&#39;extract&#39;</span>, selectedSimulations);
<a name="l00592"></a>00592             time = max(full(states.Time.values), [], 4) / 3600;
<a name="l00593"></a>00593             ntps = full(states.Metabolite.counts(1:4, :, :, :));
<a name="l00594"></a>00594             ndps = full(states.Metabolite.counts(5:8, :, :, :));
<a name="l00595"></a>00595             nmps = full(states.Metabolite.counts(9:12, :, :, :));
<a name="l00596"></a>00596             aas = full(states.Metabolite.counts(13:end, :, :, :));
<a name="l00597"></a>00597             
<a name="l00598"></a>00598             <span class="keywordflow">for</span> i = 1:numel(simEndTimes)
<a name="l00599"></a>00599                 ntps(:, :, simEndTimes(i)+1:end, i) = NaN;
<a name="l00600"></a>00600                 ndps(:, :, simEndTimes(i)+1:end, i) = NaN;
<a name="l00601"></a>00601                 nmps(:, :, simEndTimes(i)+1:end, i) = NaN;
<a name="l00602"></a>00602                 aas(:, :, simEndTimes(i)+1:end, i) = NaN;
<a name="l00603"></a>00603             end
<a name="l00604"></a>00604             
<a name="l00605"></a>00605             clear states;
<a name="l00606"></a>00606             
<a name="l00607"></a>00607             %% clear figure
<a name="l00608"></a>00608             clf(figHandle);
<a name="l00609"></a>00609             
<a name="l00610"></a>00610             %% left panel
<a name="l00611"></a>00611             
<a name="l00612"></a>00612             %layout plots
<a name="l00613"></a>00613             options = struct();
<a name="l00614"></a>00614             options.titleStr = {<span class="stringliteral">&#39;Nucleotides&#39;</span>; <span class="stringliteral">&#39;AXP&#39;</span>; <span class="stringliteral">&#39;CXP&#39;</span>; <span class="stringliteral">&#39;GXP&#39;</span>; <span class="stringliteral">&#39;UXP&#39;</span>};
<a name="l00615"></a>00615             options.xdata = permute(time, [4 3 2 1]);
<a name="l00616"></a>00616             options.ydata = {{
<a name="l00617"></a>00617                 permute(sum(aas, 1), [4 3 2 1])
<a name="l00618"></a>00618                 permute(sum(ntps, 1), [4 3 2 1])
<a name="l00619"></a>00619                 permute(sum(ndps, 1), [4 3 2 1])
<a name="l00620"></a>00620                 permute(sum(nmps, 1), [4 3 2 1])
<a name="l00621"></a>00621                 };{
<a name="l00622"></a>00622                 []
<a name="l00623"></a>00623                 permute(ntps(1, :, :, :), [4 3 2 1])
<a name="l00624"></a>00624                 permute(ndps(1, :, :, :), [4 3 2 1])
<a name="l00625"></a>00625                 permute(nmps(1, :, :, :), [4 3 2 1])
<a name="l00626"></a>00626                 };{
<a name="l00627"></a>00627                 []
<a name="l00628"></a>00628                 permute(ntps(2, :, :, :), [4 3 2 1])
<a name="l00629"></a>00629                 []
<a name="l00630"></a>00630                 permute(nmps(2, :, :, :), [4 3 2 1])
<a name="l00631"></a>00631                 };{
<a name="l00632"></a>00632                 []
<a name="l00633"></a>00633                 permute(ntps(3, :, :, :), [4 3 2 1])
<a name="l00634"></a>00634                 permute(ndps(3, :, :, :), [4 3 2 1])
<a name="l00635"></a>00635                 permute(nmps(3, :, :, :), [4 3 2 1])
<a name="l00636"></a>00636                 }; {
<a name="l00637"></a>00637                 []
<a name="l00638"></a>00638                 permute(ntps(4, :, :, :), [4 3 2 1])
<a name="l00639"></a>00639                 []
<a name="l00640"></a>00640                 permute(nmps(4, :, :, :), [4 3 2 1])
<a name="l00641"></a>00641                 }};
<a name="l00642"></a>00642             options.ylabelStr = {
<a name="l00643"></a>00643                 {<span class="stringliteral">&#39;Amino Acid&#39;</span>; <span class="stringliteral">&#39;NTP&#39;</span>; <span class="stringliteral">&#39;NDP&#39;</span>; <span class="stringliteral">&#39;NMP&#39;</span>};
<a name="l00644"></a>00644                 {}; {}; {}; {}
<a name="l00645"></a>00645                 };
<a name="l00646"></a>00646             options.colorOrder = PlotUtil.getRedGreenColorOrder(simOrder);
<a name="l00647"></a>00647             [axesHandles, ~, ~, offsetAxesHandles] = PlotUtil.multiElementPlot(figHandle, repmat({3.5 * ones(4, 1)}, 5, 1), [0 time(end)], options);
<a name="l00648"></a>00648             
<a name="l00649"></a>00649             <span class="keyword">set</span>(cellfun(@(x) x(1), axesHandles(2:end)), <span class="stringliteral">&#39;visible&#39;</span>, <span class="stringliteral">&#39;off&#39;</span>);
<a name="l00650"></a>00650             <span class="keyword">set</span>(cellfun(@(x) x(1), offsetAxesHandles(2:end)), <span class="stringliteral">&#39;visible&#39;</span>, <span class="stringliteral">&#39;off&#39;</span>);
<a name="l00651"></a>00651             <span class="keyword">set</span>(cellfun(@(x) x(3), axesHandles([3 5])), <span class="stringliteral">&#39;visible&#39;</span>, <span class="stringliteral">&#39;off&#39;</span>);
<a name="l00652"></a>00652             <span class="keyword">set</span>(cellfun(@(x) x(3), offsetAxesHandles([3 5])), <span class="stringliteral">&#39;visible&#39;</span>, <span class="stringliteral">&#39;off&#39;</span>);
<a name="l00653"></a>00653             
<a name="l00654"></a>00654             clear options;
<a name="l00655"></a>00655             
<a name="l00656"></a>00656             %% table
<a name="l00657"></a>00657             aaSums = sum(aas, 1);
<a name="l00658"></a>00658             medianNtps = [
<a name="l00659"></a>00659                 median(ntps(sub2ind(size(ntps), 1 * ones(size(simEndTimes)), ones(size(simEndTimes)), simEndTimes, (1:numel(simEndTimes))&#39;)))
<a name="l00660"></a>00660                 median(ntps(sub2ind(size(ntps), 2 * ones(size(simEndTimes)), ones(size(simEndTimes)), simEndTimes, (1:numel(simEndTimes))&#39;)))
<a name="l00661"></a>00661                 median(ntps(sub2ind(size(ntps), 3 * ones(size(simEndTimes)), ones(size(simEndTimes)), simEndTimes, (1:numel(simEndTimes))&#39;)))
<a name="l00662"></a>00662                 median(ntps(sub2ind(size(ntps), 4 * ones(size(simEndTimes)), ones(size(simEndTimes)), simEndTimes, (1:numel(simEndTimes))&#39;)))
<a name="l00663"></a>00663                 ];
<a name="l00664"></a>00664             medianNdps = [
<a name="l00665"></a>00665                 median(ndps(sub2ind(size(ndps), 1 * ones(size(simEndTimes)), ones(size(simEndTimes)), simEndTimes, (1:numel(simEndTimes))&#39;)))
<a name="l00666"></a>00666                 median(ndps(sub2ind(size(ndps), 2 * ones(size(simEndTimes)), ones(size(simEndTimes)), simEndTimes, (1:numel(simEndTimes))&#39;)))
<a name="l00667"></a>00667                 median(ndps(sub2ind(size(ndps), 3 * ones(size(simEndTimes)), ones(size(simEndTimes)), simEndTimes, (1:numel(simEndTimes))&#39;)))
<a name="l00668"></a>00668                 median(ndps(sub2ind(size(ndps), 4 * ones(size(simEndTimes)), ones(size(simEndTimes)), simEndTimes, (1:numel(simEndTimes))&#39;)))
<a name="l00669"></a>00669                 ];
<a name="l00670"></a>00670             medianNmps = [
<a name="l00671"></a>00671                 median(nmps(sub2ind(size(nmps), 1 * ones(size(simEndTimes)), ones(size(simEndTimes)), simEndTimes, (1:numel(simEndTimes))&#39;)))
<a name="l00672"></a>00672                 median(nmps(sub2ind(size(nmps), 2 * ones(size(simEndTimes)), ones(size(simEndTimes)), simEndTimes, (1:numel(simEndTimes))&#39;)))
<a name="l00673"></a>00673                 median(nmps(sub2ind(size(nmps), 3 * ones(size(simEndTimes)), ones(size(simEndTimes)), simEndTimes, (1:numel(simEndTimes))&#39;)))
<a name="l00674"></a>00674                 median(nmps(sub2ind(size(nmps), 4 * ones(size(simEndTimes)), ones(size(simEndTimes)), simEndTimes, (1:numel(simEndTimes))&#39;)))
<a name="l00675"></a>00675                 ];
<a name="l00676"></a>00676             medianAAs = median(aaSums(sub2ind(size(aaSums), ones(size(simEndTimes)), ones(size(simEndTimes)), simEndTimes, (1:numel(simEndTimes))<span class="stringliteral">&#39;)));</span>
<a name="l00677"></a>00677 <span class="stringliteral">            nxps = [</span>
<a name="l00678"></a>00678 <span class="stringliteral">                median(ntps(:, :, 1, :), 4)&#39;</span> medianNtps<span class="stringliteral">&#39;</span>
<a name="l00679"></a>00679 <span class="stringliteral">                median(ndps(:, :, 1, :), 4)&#39;</span> medianNdps<span class="stringliteral">&#39;</span>
<a name="l00680"></a>00680 <span class="stringliteral">                median(nmps(:, :, 1, :), 4)&#39;</span> medianNmps<span class="stringliteral">&#39;</span>
<a name="l00681"></a>00681 <span class="stringliteral">                median(aaSums(:, :, 1, :)) zeros(1, 3) medianAAs zeros(1, 3)</span>
<a name="l00682"></a>00682 <span class="stringliteral">                ];</span>
<a name="l00683"></a>00683 <span class="stringliteral">            tabColLabels = {[] &#39;</span>t = 0<span class="stringliteral">&#39; [] [] [] [&#39;</span>t = <span class="stringliteral">&#39; num2str(median(simEndTimes))] [] [] []};</span>
<a name="l00684"></a>00684 <span class="stringliteral">            tabContent = [{[] &#39;</span>A<span class="charliteral">&#39; &#39;</span>C<span class="charliteral">&#39; &#39;</span>G<span class="charliteral">&#39; &#39;</span>U<span class="charliteral">&#39; &#39;</span>A<span class="charliteral">&#39; &#39;</span>C<span class="charliteral">&#39; &#39;</span>G<span class="charliteral">&#39; &#39;</span>U<span class="stringliteral">&#39;;}</span>
<a name="l00685"></a>00685 <span class="stringliteral">                {&#39;</span>NTP<span class="stringliteral">&#39;; &#39;</span>NDP<span class="stringliteral">&#39;; &#39;</span>NMP<span class="stringliteral">&#39;; &#39;</span>AAs<span class="stringliteral">&#39;} num2cell(nxps)];</span>
<a name="l00686"></a>00686 <span class="stringliteral">            tabContent(end, [3:5 7:9]) = {[]};</span>
<a name="l00687"></a>00687 <span class="stringliteral">            </span>
<a name="l00688"></a>00688 <span class="stringliteral">            %% organize figure data</span>
<a name="l00689"></a>00689 <span class="stringliteral">            figData = struct;</span>
<a name="l00690"></a>00690 <span class="stringliteral">            figData.simEndTimes = simEndTimes;</span>
<a name="l00691"></a>00691 <span class="stringliteral">            figData.simOrder = simOrder;</span>
<a name="l00692"></a>00692 <span class="stringliteral">            figData.timeIdx = timeIdx;</span>
<a name="l00693"></a>00693 <span class="stringliteral">            figData.time = time;</span>
<a name="l00694"></a>00694 <span class="stringliteral">            figData.aas = aas;</span>
<a name="l00695"></a>00695 <span class="stringliteral">            figData.ntps = ntps;</span>
<a name="l00696"></a>00696 <span class="stringliteral">            figData.ndps = ndps;</span>
<a name="l00697"></a>00697 <span class="stringliteral">            figData.nmps = nmps;</span>
<a name="l00698"></a>00698 <span class="stringliteral">        end</span>
<a name="l00699"></a>00699 <span class="stringliteral">        </span>
<a name="l00700"></a>00700 <span class="stringliteral">        function figData = aminoAcidCounts(figHandle, simBatchDir, selectedSimulations)</span>
<a name="l00701"></a>00701 <span class="stringliteral">            import edu.stanford.covert.cell.sim.util.SimulationEnsemble;</span>
<a name="l00702"></a>00702 <span class="stringliteral">            import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;</span>
<a name="l00703"></a>00703 <span class="stringliteral">            import edu.stanford.covert.cell.sim.util.PlotUtil;</span>
<a name="l00704"></a>00704 <span class="stringliteral">            import edu.stanford.covert.util.ConstantUtil;</span>
<a name="l00705"></a>00705 <span class="stringliteral">            </span>
<a name="l00706"></a>00706 <span class="stringliteral">            if nargin &lt; 2 || isempty(simBatchDir)</span>
<a name="l00707"></a>00707 <span class="stringliteral">                simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();</span>
<a name="l00708"></a>00708 <span class="stringliteral">            end</span>
<a name="l00709"></a>00709 <span class="stringliteral">            if nargin &lt; 3 || isempty(selectedSimulations)</span>
<a name="l00710"></a>00710 <span class="stringliteral">                selectedSimulations = SimulationDiskUtil.getCompleteSimulations(simBatchDir);</span>
<a name="l00711"></a>00711 <span class="stringliteral">            end</span>
<a name="l00712"></a>00712 <span class="stringliteral">            </span>
<a name="l00713"></a>00713 <span class="stringliteral">            [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep &#39;</span>1<span class="stringliteral">&#39;]);</span>
<a name="l00714"></a>00714 <span class="stringliteral">            comp = sim.compartment;</span>
<a name="l00715"></a>00715 <span class="stringliteral">            m = sim.state(&#39;</span><a class="code" href="class_metabolite.html">Metabolite</a><span class="stringliteral">&#39;);</span>
<a name="l00716"></a>00716 <span class="stringliteral">            </span>
<a name="l00717"></a>00717 <span class="stringliteral">            %% get data</span>
<a name="l00718"></a>00718 <span class="stringliteral">            stateNames = {</span>
<a name="l00719"></a>00719 <span class="stringliteral">                &#39;</span>Time<span class="stringliteral">&#39;        &#39;</span>values<span class="stringliteral">&#39;  &#39;</span>:<span class="stringliteral">&#39;                &#39;</span>:<span class="stringliteral">&#39;</span>
<a name="l00720"></a>00720 <span class="stringliteral">                &#39;</span>Mass<span class="stringliteral">&#39;        &#39;</span>cell<span class="stringliteral">&#39;    &#39;</span>:<span class="stringliteral">&#39;                &#39;</span>-sum<span class="stringliteral">&#39;</span>
<a name="l00721"></a>00721 <span class="stringliteral">                &#39;</span><a class="code" href="class_metabolite.html">Metabolite</a><span class="stringliteral">&#39;  &#39;</span>counts<span class="stringliteral">&#39;  m.aminoAcidIndexs  comp.cytosolIndexs</span>
<a name="l00722"></a>00722 <span class="stringliteral">                };</span>
<a name="l00723"></a>00723 <span class="stringliteral">            states = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, &#39;</span>extract<span class="stringliteral">&#39;, selectedSimulations);</span>
<a name="l00724"></a>00724 <span class="stringliteral">            </span>
<a name="l00725"></a>00725 <span class="stringliteral">            time = permute(max(states.Time.values, [], 4), [4 3 1 2]) / 3600;</span>
<a name="l00726"></a>00726 <span class="stringliteral">            simEndTimes = squeeze(max(states.Time.values, [], 3));</span>
<a name="l00727"></a>00727 <span class="stringliteral">            [~, simOrder] = sort(states.Mass.cell(sub2ind(size(states.Mass.cell), ones(size(simEndTimes)), ones(size(simEndTimes)), simEndTimes, (1:numel(simEndTimes))&#39;</span>)));
<a name="l00728"></a>00728             
<a name="l00729"></a>00729             aminoAcids = full(permute(states.Metabolite.counts, [4 3 1 2]));
<a name="l00730"></a>00730             <span class="keywordflow">for</span> i = 1:numel(simEndTimes)
<a name="l00731"></a>00731                 aminoAcids(i, simEndTimes(i) + 1:end, :, :) = NaN;
<a name="l00732"></a>00732             end
<a name="l00733"></a>00733             
<a name="l00734"></a>00734             %% plot data
<a name="l00735"></a>00735             clf(figHandle);
<a name="l00736"></a>00736             
<a name="l00737"></a>00737             options = struct;
<a name="l00738"></a>00738             options.colorOrder = PlotUtil.getRedGreenColorOrder(simOrder);
<a name="l00739"></a>00739             options.xdata = time;
<a name="l00740"></a>00740             options.ylabelStr = {
<a name="l00741"></a>00741                 m.wholeCellModelIDs(m.aminoAcidIndexs(1:6));
<a name="l00742"></a>00742                 cell(6, 1);
<a name="l00743"></a>00743                 m.wholeCellModelIDs(m.aminoAcidIndexs(7:12));
<a name="l00744"></a>00744                 cell(6, 1);
<a name="l00745"></a>00745                 m.wholeCellModelIDs(m.aminoAcidIndexs(13:18));
<a name="l00746"></a>00746                 cell(6, 1);
<a name="l00747"></a>00747                 [m.wholeCellModelIDs(m.aminoAcidIndexs(19:21)); {[]; []; []}]
<a name="l00748"></a>00748                 cell(6, 1);
<a name="l00749"></a>00749                 };
<a name="l00750"></a>00750             options.ydata = {
<a name="l00751"></a>00751                 permute(mat2cell(aminoAcids(:, :, 1:6), size(aminoAcids, 1), size(aminoAcids, 2), ones(1, 6)), [3 1 2])
<a name="l00752"></a>00752                 repmat({[]}, 6, 1)
<a name="l00753"></a>00753                 permute(mat2cell(aminoAcids(:, :, 7:12), size(aminoAcids, 1), size(aminoAcids, 2), ones(1, 6)), [3 1 2])
<a name="l00754"></a>00754                 repmat({[]}, 6, 1)
<a name="l00755"></a>00755                 permute(mat2cell(aminoAcids(:, :, 13:18), size(aminoAcids, 1), size(aminoAcids, 2), ones(1, 6)), [3 1 2])
<a name="l00756"></a>00756                 repmat({[]}, 6, 1)
<a name="l00757"></a>00757                 [permute(mat2cell(aminoAcids(:, :, 19:21), size(aminoAcids, 1), size(aminoAcids, 2), ones(1, 3)), [3 1 2]); {[]; []; []}]
<a name="l00758"></a>00758                 repmat({[]}, 6, 1)
<a name="l00759"></a>00759                 };
<a name="l00760"></a>00760             options.colWidths = repmat([4; 1], 4, 1);
<a name="l00761"></a>00761             options.yAxesLabelWidths = repmat([0.1; 0.01], 4, 1);
<a name="l00762"></a>00762             [axesHandles, xAxesHandles, ~, offsetAxesHandles] = PlotUtil.multiElementPlot(figHandle, repmat({10 * ones(6, 1)}, 8, 1), [0 time(end)], options);
<a name="l00763"></a>00763             
<a name="l00764"></a>00764             clear options;
<a name="l00765"></a>00765             
<a name="l00766"></a>00766             <span class="keyword">delete</span>(axesHandles{7}(4:6));
<a name="l00767"></a>00767             <span class="keyword">delete</span>(axesHandles{8}(4:6));
<a name="l00768"></a>00768             <span class="keyword">delete</span>(offsetAxesHandles{7}(4:6));
<a name="l00769"></a>00769             <span class="keyword">delete</span>(offsetAxesHandles{8}(4:6));
<a name="l00770"></a>00770             <span class="keyword">delete</span>(cell2mat(xAxesHandles(2:2:end)));
<a name="l00771"></a>00771             
<a name="l00772"></a>00772             <span class="keywordflow">for</span> i = 1:numel(m.aminoAcidIndexs)
<a name="l00773"></a>00773                 row = mod(i - 1, 6) + 1;
<a name="l00774"></a>00774                 col = 2 * ceil(i / 6);
<a name="l00775"></a>00775                 
<a name="l00776"></a>00776                 cnts = aminoAcids(sub2ind(size(aminoAcids), (1:numel(simEndTimes))&#39;, simEndTimes, i * ones(size(simEndTimes))));
<a name="l00777"></a>00777                 edges = linspace(min(min(aminoAcids(:, :, i))), max(max(aminoAcids(:, :, i))), 10);
<a name="l00778"></a>00778                 edges = edges(:);
<a name="l00779"></a>00779                 freq = histc(cnts, edges);
<a name="l00780"></a>00780                 
<a name="l00781"></a>00781                 axesHandle = axesHandles{col}(row);
<a name="l00782"></a>00782                 offsetAxesHandle = offsetAxesHandles{col}(row);
<a name="l00783"></a>00783                 
<a name="l00784"></a>00784                 cla(axesHandle)
<a name="l00785"></a>00785                 cla(offsetAxesHandle)
<a name="l00786"></a>00786                 
<a name="l00787"></a>00787                 barh(axesHandle, edges, freq)
<a name="l00788"></a>00788                 xlim(axesHandle, [0 max(freq(:))]);
<a name="l00789"></a>00789                 ylim(axesHandle, [min(min(aminoAcids(:, :, i))) max(max(aminoAcids(:, :, i)))]);
<a name="l00790"></a>00790                 set(axesHandle, &#39;YTick&#39;, [])
<a name="l00791"></a>00791                 set(axesHandle, &#39;XTick&#39;, [])
<a name="l00792"></a>00792                 set(offsetAxesHandle, &#39;YTick&#39;, [])
<a name="l00793"></a>00793                 set(offsetAxesHandle, &#39;XTick&#39;, [])
<a name="l00794"></a>00794             end
<a name="l00795"></a>00795             
<a name="l00796"></a>00796             %% data
<a name="l00797"></a>00797             figData = struct;
<a name="l00798"></a>00798             figData.simOrder = simOrder;
<a name="l00799"></a>00799             figData.time = time;
<a name="l00800"></a>00800             figData.aminoAcids = aminoAcids;
<a name="l00801"></a>00801         end
<a name="l00802"></a>00802         
<a name="l00803"></a>00803         function [figData, tabContent, tabColLabels] = metabolism(figHandle, simBatchDir, selectedSimulations)
<a name="l00804"></a>00804             import edu.stanford.covert.cell.sim.util.SimulationEnsemble;
<a name="l00805"></a>00805             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
<a name="l00806"></a>00806             import edu.stanford.covert.cell.sim.util.PlotUtil;
<a name="l00807"></a>00807             
<a name="l00808"></a>00808             if nargin &lt; 2 || isempty(simBatchDir)
<a name="l00809"></a>00809                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
<a name="l00810"></a>00810             end
<a name="l00811"></a>00811             if nargin &lt; 3 || isempty(selectedSimulations)
<a name="l00812"></a>00812                 selectedSimulations = SimulationDiskUtil.getCompleteSimulations(simBatchDir);
<a name="l00813"></a>00813             end
<a name="l00814"></a>00814             
<a name="l00815"></a>00815             [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep &#39;1&#39;]);
<a name="l00816"></a>00816             pm = sim.state(&#39;<a class="code" href="class_protein_monomer.html">ProteinMonomer</a>&#39;);
<a name="l00817"></a>00817             pc = sim.state(&#39;<a class="code" href="class_protein_complex.html">ProteinComplex</a>&#39;);
<a name="l00818"></a>00818             met = sim.process(&#39;Metabolism&#39;);
<a name="l00819"></a>00819             
<a name="l00820"></a>00820             %% get data
<a name="l00821"></a>00821             corrEnzymeFlux = zeros(numel(met.reactionWholeCellModelIDs), numel(selectedSimulations));
<a name="l00822"></a>00822             corrEnzymeGrowth = zeros(numel(met.enzymeWholeCellModelIDs), numel(selectedSimulations));
<a name="l00823"></a>00823             corrFluxGrowth = zeros(numel(met.reactionWholeCellModelIDs), numel(selectedSimulations));
<a name="l00824"></a>00824             stateNames = {
<a name="l00825"></a>00825                 <span class="stringliteral">&#39;MetabolicReaction&#39;</span>  <span class="stringliteral">&#39;fluxs&#39;</span>   <span class="charliteral">&#39;:&#39;</span>                                            <span class="charliteral">&#39;:&#39;</span>
<a name="l00826"></a>00826                 <span class="stringliteral">&#39;MetabolicReaction&#39;</span>  <span class="stringliteral">&#39;growth&#39;</span>  <span class="charliteral">&#39;:&#39;</span>                                            <span class="charliteral">&#39;:&#39;</span>
<a name="l00827"></a>00827                 <span class="stringliteral">&#39;ProteinMonomer&#39;</span>     <span class="stringliteral">&#39;counts&#39;</span>  pm.matureIndexs(met.enzymeMonomerGlobalIndexs) <span class="stringliteral">&#39;-sum&#39;</span>
<a name="l00828"></a>00828                 <span class="stringliteral">&#39;ProteinComplex&#39;</span>     <span class="stringliteral">&#39;counts&#39;</span>  pc.matureIndexs(met.enzymeComplexGlobalIndexs) <span class="stringliteral">&#39;-sum&#39;</span>
<a name="l00829"></a>00829                 };
<a name="l00830"></a>00830             <span class="keywordflow">for</span> i = 1:numel(selectedSimulations)
<a name="l00831"></a>00831                 states = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, &#39;extract&#39;, selectedSimulations(i));
<a name="l00832"></a>00832                 growth = permute(states.<a class="code" href="class_metabolic_reaction.html">MetabolicReaction</a>.growth, [1 3 2]);
<a name="l00833"></a>00833                 enzymes = zeros(numel(met.enzymeWholeCellModelIDs), size(states.ProteinMonomer.counts, 3));
<a name="l00834"></a>00834                 enzymes(met.enzymeMonomerLocalIndexs, :, :) = full(permute(states.ProteinMonomer.counts, [1 3 2]));
<a name="l00835"></a>00835                 enzymes(met.enzymeComplexLocalIndexs, :, :) = full(permute(states.ProteinComplex.counts, [1 3 2]));
<a name="l00836"></a>00836                 fluxs = permute(states.<a class="code" href="class_metabolic_reaction.html">MetabolicReaction</a>.fluxs, [1 3 2]);
<a name="l00837"></a>00837                 
<a name="l00838"></a>00838                 for j = 1:numel(met.reactionIndexs_fba)
<a name="l00839"></a>00839                     k = met.reactionIndexs_fba(j);
<a name="l00840"></a>00840                     corrEnzymeFlux(k, i) = corr(fluxs(k, :)&#39;, (met.reactionCatalysisMatrix(k, :) * enzymes)&#39;);
<a name="l00841"></a>00841                 end
<a name="l00842"></a>00842                 corrEnzymeGrowth(:, i) = corr(enzymes&#39;, growth&#39;);
<a name="l00843"></a>00843                 corrFluxGrowth(:, i) = corr(fluxs&#39;, growth&#39;);
<a name="l00844"></a>00844                 
<a name="l00845"></a>00845                 clear states;
<a name="l00846"></a>00846             end
<a name="l00847"></a>00847             
<a name="l00848"></a>00848             corrEnzymeFlux(isnan(corrEnzymeFlux)) = 0;
<a name="l00849"></a>00849             corrEnzymeGrowth(isnan(corrEnzymeGrowth)) = 0;
<a name="l00850"></a>00850             corrFluxGrowth(isnan(corrFluxGrowth)) = 0;
<a name="l00851"></a>00851             
<a name="l00852"></a>00852             corrEnzymeFluxMean = nanmean(corrEnzymeFlux, 2);
<a name="l00853"></a>00853             corrEnzymeGrowthMean = nanmean(corrEnzymeGrowth, 2);
<a name="l00854"></a>00854             corrFluxGrowthMean = nanmean(corrFluxGrowth, 2);
<a name="l00855"></a>00855             corrEnzymeFluxStd = nanstd(corrEnzymeFlux, [], 2);
<a name="l00856"></a>00856             corrEnzymeGrowthStd = nanstd(corrEnzymeGrowth, [], 2);
<a name="l00857"></a>00857             corrFluxGrowthStd = nanstd(corrFluxGrowth, [], 2);
<a name="l00858"></a>00858             
<a name="l00859"></a>00859             [~, enzIdx] = max(corrEnzymeGrowthMean);
<a name="l00860"></a>00860             [~, simOrder] = sort(corrEnzymeGrowth(enzIdx, :));
<a name="l00861"></a>00861             
<a name="l00862"></a>00862             clear corrEnzymeFlux corrEnzymeGrowth corrFluxGrowth;
<a name="l00863"></a>00863             
<a name="l00864"></a>00864             %% get temporal data
<a name="l00865"></a>00865             rxnIdx = find(met.reactionCatalysisMatrix(:, enzIdx));
<a name="l00866"></a>00866             monIdx = pm.matureIndexs(met.enzymeMonomerGlobalIndexs(met.enzymeMonomerLocalIndexs == enzIdx));
<a name="l00867"></a>00867             cpxIdx = pc.matureIndexs(met.enzymeComplexGlobalIndexs(met.enzymeComplexLocalIndexs == enzIdx));
<a name="l00868"></a>00868             
<a name="l00869"></a>00869             stateNames = {
<a name="l00870"></a>00870                 <span class="stringliteral">&#39;Time&#39;</span>               <span class="stringliteral">&#39;values&#39;</span>  <span class="charliteral">&#39;:&#39;</span>     <span class="charliteral">&#39;:&#39;</span>
<a name="l00871"></a>00871                 <span class="stringliteral">&#39;MetabolicReaction&#39;</span>  <span class="stringliteral">&#39;fluxs&#39;</span>   rxnIdx  <span class="charliteral">&#39;:&#39;</span>
<a name="l00872"></a>00872                 };
<a name="l00873"></a>00873             <span class="keywordflow">if</span> ~isempty(monIdx)
<a name="l00874"></a>00874                 stateNames = [stateNames; {<span class="stringliteral">&#39;ProteinMonomer&#39;</span> <span class="stringliteral">&#39;counts&#39;</span> monIdx <span class="stringliteral">&#39;-sum&#39;</span>}];
<a name="l00875"></a>00875                 states = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, <span class="stringliteral">&#39;extract&#39;</span>, selectedSimulations);
<a name="l00876"></a>00876                 enzymes = full(permute(states.ProteinMonomer.counts, [3 4 1 2]));
<a name="l00877"></a>00877             <span class="keywordflow">else</span>
<a name="l00878"></a>00878                 stateNames = [stateNames; {&#39;ProteinComplex&#39; &#39;counts&#39; cpxIdx &#39;-sum&#39;}];
<a name="l00879"></a>00879                 states = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, <span class="stringliteral">&#39;extract&#39;</span>, selectedSimulations);
<a name="l00880"></a>00880                 enzymes = full(permute(states.ProteinComplex.counts, [3 4 1 2]));
<a name="l00881"></a>00881             end
<a name="l00882"></a>00882             time = permute(max(states.Time.values, [], 4), [3 4 1 2]) / 3600;
<a name="l00883"></a>00883             simEndTimes = squeeze(max(states.Time.values, [], 3));
<a name="l00884"></a>00884             fluxs = full(permute(states.MetabolicReaction.fluxs, [3 4 1 2]));
<a name="l00885"></a>00885             
<a name="l00886"></a>00886             clear states;
<a name="l00887"></a>00887             
<a name="l00888"></a>00888             <span class="keywordflow">for</span> i = 1:numel(simEndTimes)
<a name="l00889"></a>00889                 enzymes(simEndTimes(i)+1:end, i) = NaN;
<a name="l00890"></a>00890                 fluxs(simEndTimes(i)+1:end, i) = NaN;
<a name="l00891"></a>00891             end
<a name="l00892"></a>00892             
<a name="l00893"></a>00893             colorOrder = PlotUtil.getRedGreenColorOrder(simOrder);
<a name="l00894"></a>00894             
<a name="l00895"></a>00895             %% plot
<a name="l00896"></a>00896             clf(figHandle);
<a name="l00897"></a>00897             
<a name="l00898"></a>00898             axesHandles = repmat({zeros(4, 1)}, 2, 1);
<a name="l00899"></a>00899             
<a name="l00900"></a>00900             %enzyme-fluxs
<a name="l00901"></a>00901             axesHandle = subplot(4, 2, 1);
<a name="l00902"></a>00902             axesHandles{1}(1) = axesHandle;
<a name="l00903"></a>00903             [~, order] = sort(corrEnzymeFluxMean);
<a name="l00904"></a>00904             errorbar((1:numel(corrEnzymeFluxMean))<span class="stringliteral">&#39;, corrEnzymeFluxMean(order), corrEnzymeFluxStd(order), ...</span>
<a name="l00905"></a>00905 <span class="stringliteral">                &#39;</span>Parent<span class="stringliteral">&#39;, axesHandle, &#39;</span>LineStyle<span class="stringliteral">&#39;, &#39;</span>none<span class="stringliteral">&#39;, &#39;</span>Marker<span class="stringliteral">&#39;, &#39;</span>.<span class="stringliteral">&#39;);</span>
<a name="l00906"></a>00906 <span class="stringliteral">            xlabel(axesHandle, &#39;</span>Flux<span class="stringliteral">&#39;, &#39;</span>FontSize<span class="stringliteral">&#39;, 8);</span>
<a name="l00907"></a>00907 <span class="stringliteral">            ylabel(axesHandle, {&#39;</span>Corr<span class="charliteral">&#39; &#39;</span>(Flux, <a class="code" href="class_enzyme.html">Enzyme</a>)<span class="stringliteral">&#39;}, &#39;</span>FontSize<span class="stringliteral">&#39;, 8);</span>
<a name="l00908"></a>00908 <span class="stringliteral">            xlim(axesHandle, [0.5 numel(corrEnzymeFluxMean) + 0.5]);</span>
<a name="l00909"></a>00909 <span class="stringliteral">            ylim(axesHandle, [min(corrEnzymeFluxMean - corrEnzymeFluxStd) max(corrEnzymeFluxMean - corrEnzymeFluxStd)]);</span>
<a name="l00910"></a>00910 <span class="stringliteral">            </span>
<a name="l00911"></a>00911 <span class="stringliteral">            axesHandle = subplot(4, 2, 2);</span>
<a name="l00912"></a>00912 <span class="stringliteral">            axesHandles{2}(1) = axesHandle;</span>
<a name="l00913"></a>00913 <span class="stringliteral">            hist(axesHandle, corrEnzymeFluxMean);</span>
<a name="l00914"></a>00914 <span class="stringliteral">            xlabel(axesHandle, &#39;</span>Corr (Flux, <a class="code" href="class_enzyme.html">Enzyme</a>)<span class="stringliteral">&#39;, &#39;</span>FontSize<span class="stringliteral">&#39;, 8);</span>
<a name="l00915"></a>00915 <span class="stringliteral">            ylabel(axesHandle, &#39;</span>Freq<span class="stringliteral">&#39;, &#39;</span>FontSize<span class="stringliteral">&#39;, 8);</span>
<a name="l00916"></a>00916 <span class="stringliteral">            </span>
<a name="l00917"></a>00917 <span class="stringliteral">            %fluxes-growth</span>
<a name="l00918"></a>00918 <span class="stringliteral">            axesHandle = subplot(4, 2, 3);</span>
<a name="l00919"></a>00919 <span class="stringliteral">            axesHandles{1}(2) = axesHandle;</span>
<a name="l00920"></a>00920 <span class="stringliteral">            [~, order] = sort(corrFluxGrowthMean);</span>
<a name="l00921"></a>00921 <span class="stringliteral">            errorbar((1:numel(corrFluxGrowthMean))&#39;</span>, corrFluxGrowthMean(order), corrFluxGrowthStd(order), ...
<a name="l00922"></a>00922                 <span class="stringliteral">&#39;Parent&#39;</span>, axesHandle, <span class="stringliteral">&#39;LineStyle&#39;</span>, <span class="stringliteral">&#39;none&#39;</span>, <span class="stringliteral">&#39;Marker&#39;</span>, <span class="charliteral">&#39;.&#39;</span>);
<a name="l00923"></a>00923             xlabel(axesHandle, <span class="stringliteral">&#39;Flux&#39;</span>, <span class="stringliteral">&#39;FontSize&#39;</span>, 8);
<a name="l00924"></a>00924             ylabel(axesHandle, {<span class="stringliteral">&#39;Corr&#39;</span> <span class="stringliteral">&#39;(Flux, Growth)&#39;</span>}, <span class="stringliteral">&#39;FontSize&#39;</span>, 8);
<a name="l00925"></a>00925             xlim(axesHandle, [0.5 numel(corrFluxGrowthMean) + 0.5]);
<a name="l00926"></a>00926             ylim(axesHandle, [min(corrFluxGrowthMean - corrFluxGrowthStd) max(corrFluxGrowthMean - corrFluxGrowthStd)]);
<a name="l00927"></a>00927             
<a name="l00928"></a>00928             axesHandle = subplot(4, 2, 4);
<a name="l00929"></a>00929             axesHandles{2}(2) = axesHandle;
<a name="l00930"></a>00930             hist(axesHandle, corrFluxGrowthMean);
<a name="l00931"></a>00931             xlabel(axesHandle, <span class="stringliteral">&#39;Corr (Flux, Growth)&#39;</span>, <span class="stringliteral">&#39;FontSize&#39;</span>, 8);
<a name="l00932"></a>00932             ylabel(axesHandle, <span class="stringliteral">&#39;Freq&#39;</span>, <span class="stringliteral">&#39;FontSize&#39;</span>, 8);
<a name="l00933"></a>00933             
<a name="l00934"></a>00934             %enzymes-growth
<a name="l00935"></a>00935             axesHandle = subplot(4, 2, 5);
<a name="l00936"></a>00936             axesHandles{1}(3) = axesHandle;
<a name="l00937"></a>00937             [~, order] = sort(corrEnzymeGrowthMean);
<a name="l00938"></a>00938             errorbar((1:numel(corrEnzymeGrowthMean))<span class="stringliteral">&#39;, corrEnzymeGrowthMean(order), corrEnzymeGrowthStd(order), ...</span>
<a name="l00939"></a>00939 <span class="stringliteral">                &#39;</span>Parent<span class="stringliteral">&#39;, axesHandle, &#39;</span>LineStyle<span class="stringliteral">&#39;, &#39;</span>none<span class="stringliteral">&#39;, &#39;</span>Marker<span class="stringliteral">&#39;, &#39;</span>.<span class="stringliteral">&#39;);</span>
<a name="l00940"></a>00940 <span class="stringliteral">            xlabel(axesHandle, &#39;</span>Flux<span class="stringliteral">&#39;, &#39;</span>FontSize<span class="stringliteral">&#39;, 8);</span>
<a name="l00941"></a>00941 <span class="stringliteral">            ylabel(axesHandle, {&#39;</span>Corr<span class="charliteral">&#39; &#39;</span>(Growth, <a class="code" href="class_enzyme.html">Enzyme</a>)<span class="stringliteral">&#39;}, &#39;</span>FontSize<span class="stringliteral">&#39;, 8);</span>
<a name="l00942"></a>00942 <span class="stringliteral">            xlim(axesHandle, [0.5 numel(corrEnzymeGrowthMean) + 0.5]);</span>
<a name="l00943"></a>00943 <span class="stringliteral">            ylim(axesHandle, [min(corrEnzymeGrowthMean - corrEnzymeGrowthStd) max(corrEnzymeGrowthMean - corrEnzymeGrowthStd)]);</span>
<a name="l00944"></a>00944 <span class="stringliteral">            </span>
<a name="l00945"></a>00945 <span class="stringliteral">            axesHandle = subplot(4, 2, 6);</span>
<a name="l00946"></a>00946 <span class="stringliteral">            axesHandles{2}(3) = axesHandle;</span>
<a name="l00947"></a>00947 <span class="stringliteral">            hist(axesHandle, corrEnzymeGrowthMean);</span>
<a name="l00948"></a>00948 <span class="stringliteral">            xlabel(axesHandle, &#39;</span>Corr (Growth, <a class="code" href="class_enzyme.html">Enzyme</a>)<span class="stringliteral">&#39;, &#39;</span>FontSize<span class="stringliteral">&#39;, 8);</span>
<a name="l00949"></a>00949 <span class="stringliteral">            ylabel(axesHandle, &#39;</span>Freq<span class="stringliteral">&#39;, &#39;</span>FontSize<span class="stringliteral">&#39;, 8);</span>
<a name="l00950"></a>00950 <span class="stringliteral">            </span>
<a name="l00951"></a>00951 <span class="stringliteral">            %flux-time</span>
<a name="l00952"></a>00952 <span class="stringliteral">            axesHandle = subplot(4, 2, 7);</span>
<a name="l00953"></a>00953 <span class="stringliteral">            axesHandles{1}(4) = axesHandle;</span>
<a name="l00954"></a>00954 <span class="stringliteral">            hold(axesHandle, &#39;</span>on<span class="stringliteral">&#39;);</span>
<a name="l00955"></a>00955 <span class="stringliteral">            set(axesHandle, &#39;</span>ColorOrder<span class="stringliteral">&#39;, colorOrder);</span>
<a name="l00956"></a>00956 <span class="stringliteral">            PlotUtil.plotLine(axesHandle, time, fluxs, true, true);</span>
<a name="l00957"></a>00957 <span class="stringliteral">            xlabel(axesHandle, &#39;</span>Time (h)<span class="stringliteral">&#39;, &#39;</span>FontSize<span class="stringliteral">&#39;, 8);</span>
<a name="l00958"></a>00958 <span class="stringliteral">            ylabel(axesHandle, [met.reactionWholeCellModelIDs(rxnIdx) &#39;</span>Flux<span class="stringliteral">&#39;], ...</span>
<a name="l00959"></a>00959 <span class="stringliteral">                &#39;</span>FontSize<span class="stringliteral">&#39;, 8, &#39;</span>Interpreter<span class="stringliteral">&#39;, &#39;</span>none<span class="stringliteral">&#39;);</span>
<a name="l00960"></a>00960 <span class="stringliteral">            </span>
<a name="l00961"></a>00961 <span class="stringliteral">            %enzyme-time</span>
<a name="l00962"></a>00962 <span class="stringliteral">            axesHandle = subplot(4, 2, 8);</span>
<a name="l00963"></a>00963 <span class="stringliteral">            axesHandles{2}(4) = axesHandle;</span>
<a name="l00964"></a>00964 <span class="stringliteral">            hold(axesHandle, &#39;</span>on<span class="stringliteral">&#39;);</span>
<a name="l00965"></a>00965 <span class="stringliteral">            set(axesHandle, &#39;</span>ColorOrder<span class="stringliteral">&#39;, colorOrder);</span>
<a name="l00966"></a>00966 <span class="stringliteral">            PlotUtil.plotLine(axesHandle, time, enzymes, true, true);</span>
<a name="l00967"></a>00967 <span class="stringliteral">            xlabel(axesHandle, &#39;</span>Time (h)<span class="stringliteral">&#39;, &#39;</span>FontSize<span class="stringliteral">&#39;, 8);</span>
<a name="l00968"></a>00968 <span class="stringliteral">            ylabel(axesHandle, [met.enzymeWholeCellModelIDs(enzIdx) &#39;</span><a class="code" href="class_enzyme.html">Enzyme</a><span class="stringliteral">&#39;], ...</span>
<a name="l00969"></a>00969 <span class="stringliteral">                &#39;</span>FontSize<span class="stringliteral">&#39;, 8, &#39;</span>Interpreter<span class="stringliteral">&#39;, &#39;</span>none<span class="stringliteral">&#39;);</span>
<a name="l00970"></a>00970 <span class="stringliteral">            </span>
<a name="l00971"></a>00971 <span class="stringliteral">            %% format plots</span>
<a name="l00972"></a>00972 <span class="stringliteral">            for i = 1:numel(axesHandles)</span>
<a name="l00973"></a>00973 <span class="stringliteral">                for j = 1:numel(axesHandles{i})</span>
<a name="l00974"></a>00974 <span class="stringliteral">                    axesHandle = axesHandles{i}(j);</span>
<a name="l00975"></a>00975 <span class="stringliteral">                    set(axesHandle, &#39;</span>FontSize<span class="stringliteral">&#39;, 6);</span>
<a name="l00976"></a>00976 <span class="stringliteral">                    set(get(axesHandle, &#39;</span>YLabel<span class="stringliteral">&#39;), &#39;</span>Units<span class="stringliteral">&#39;, &#39;</span>normalized<span class="stringliteral">&#39;)</span>
<a name="l00977"></a>00977 <span class="stringliteral">                end</span>
<a name="l00978"></a>00978 <span class="stringliteral">            end</span>
<a name="l00979"></a>00979 <span class="stringliteral">            PlotUtil.alignYAxesLabels(axesHandles);</span>
<a name="l00980"></a>00980 <span class="stringliteral">            </span>
<a name="l00981"></a>00981 <span class="stringliteral">            %% layout table</span>
<a name="l00982"></a>00982 <span class="stringliteral">            tabColLabels = {</span>
<a name="l00983"></a>00983 <span class="stringliteral">                &#39;</span><a class="code" href="class_reaction.html">Reaction</a><span class="charliteral">&#39; &#39;</span>Mean <a class="code" href="class_enzyme.html">Enzyme</a> Correlation<span class="charliteral">&#39; &#39;</span>Std <a class="code" href="class_enzyme.html">Enzyme</a> Correlation<span class="stringliteral">&#39; ...</span>
<a name="l00984"></a>00984 <span class="stringliteral">                [] ...</span>
<a name="l00985"></a>00985 <span class="stringliteral">                &#39;</span><a class="code" href="class_reaction.html">Reaction</a><span class="charliteral">&#39; &#39;</span>Mean Growth Correlation<span class="charliteral">&#39; &#39;</span>Std Growth Correlation<span class="stringliteral">&#39; ...</span>
<a name="l00986"></a>00986 <span class="stringliteral">                [] ...</span>
<a name="l00987"></a>00987 <span class="stringliteral">                &#39;</span><a class="code" href="class_enzyme.html">Enzyme</a><span class="stringliteral">&#39;   &#39;</span>Mean Growth Correlation<span class="charliteral">&#39; &#39;</span>Std Growth Correlation<span class="stringliteral">&#39; ...</span>
<a name="l00988"></a>00988 <span class="stringliteral">                };</span>
<a name="l00989"></a>00989 <span class="stringliteral">            tabContent = [</span>
<a name="l00990"></a>00990 <span class="stringliteral">                met.reactionWholeCellModelIDs num2cell(corrEnzymeFluxMean) num2cell(corrEnzymeFluxStd) ...</span>
<a name="l00991"></a>00991 <span class="stringliteral">                cell(size(met.reactionWholeCellModelIDs)) ...</span>
<a name="l00992"></a>00992 <span class="stringliteral">                met.reactionWholeCellModelIDs num2cell(corrFluxGrowthMean) num2cell(corrFluxGrowthStd) ...</span>
<a name="l00993"></a>00993 <span class="stringliteral">                cell(size(met.reactionWholeCellModelIDs)) ...</span>
<a name="l00994"></a>00994 <span class="stringliteral">                [met.enzymeWholeCellModelIDs num2cell(corrEnzymeGrowthMean) num2cell(corrEnzymeGrowthStd); cell(numel(met.reactionWholeCellModelIDs) - numel(met.enzymeWholeCellModelIDs), 3)] ...</span>
<a name="l00995"></a>00995 <span class="stringliteral">                ];</span>
<a name="l00996"></a>00996 <span class="stringliteral">            </span>
<a name="l00997"></a>00997 <span class="stringliteral">            %% organize figure data</span>
<a name="l00998"></a>00998 <span class="stringliteral">            figData = struct;</span>
<a name="l00999"></a>00999 <span class="stringliteral">            figData.simOrder = simOrder;</span>
<a name="l01000"></a>01000 <span class="stringliteral">            figData.time = time;</span>
<a name="l01001"></a>01001 <span class="stringliteral">            figData.enzymes = enzymes;</span>
<a name="l01002"></a>01002 <span class="stringliteral">            figData.fluxs = fluxs;</span>
<a name="l01003"></a>01003 <span class="stringliteral">        end</span>
<a name="l01004"></a>01004 <span class="stringliteral">        </span>
<a name="l01005"></a>01005 <span class="stringliteral">        function figData = metaboliteConcentrations(figHandle, simBatchDir, selectedSimulations)</span>
<a name="l01006"></a>01006 <span class="stringliteral">            import edu.stanford.covert.cell.sim.util.SimulationEnsemble;</span>
<a name="l01007"></a>01007 <span class="stringliteral">            import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;</span>
<a name="l01008"></a>01008 <span class="stringliteral">            import edu.stanford.covert.cell.sim.util.PlotUtil;</span>
<a name="l01009"></a>01009 <span class="stringliteral">            import edu.stanford.covert.util.ConstantUtil;</span>
<a name="l01010"></a>01010 <span class="stringliteral">            </span>
<a name="l01011"></a>01011 <span class="stringliteral">            if nargin &lt; 2 || isempty(simBatchDir)</span>
<a name="l01012"></a>01012 <span class="stringliteral">                simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();</span>
<a name="l01013"></a>01013 <span class="stringliteral">            end</span>
<a name="l01014"></a>01014 <span class="stringliteral">            if nargin &lt; 3 || isempty(selectedSimulations)</span>
<a name="l01015"></a>01015 <span class="stringliteral">                selectedSimulations = SimulationDiskUtil.getCompleteSimulations(simBatchDir);</span>
<a name="l01016"></a>01016 <span class="stringliteral">            end</span>
<a name="l01017"></a>01017 <span class="stringliteral">            </span>
<a name="l01018"></a>01018 <span class="stringliteral">            [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep &#39;</span>1<span class="stringliteral">&#39;]);</span>
<a name="l01019"></a>01019 <span class="stringliteral">            comp = sim.compartment;</span>
<a name="l01020"></a>01020 <span class="stringliteral">            m = sim.state(&#39;</span><a class="code" href="class_metabolite.html">Metabolite</a><span class="stringliteral">&#39;);</span>
<a name="l01021"></a>01021 <span class="stringliteral">            </span>
<a name="l01022"></a>01022 <span class="stringliteral">            %% get data</span>
<a name="l01023"></a>01023 <span class="stringliteral">            metIdxs = [m.ntpIndexs; m.ndpIndexs; m.nmpIndexs; m.dntpIndexs; m.aminoAcidIndexs; m.phosphateIndexs; m.diphosphateIndexs; m.hydrogenIndexs];</span>
<a name="l01024"></a>01024 <span class="stringliteral">            stateNames = {</span>
<a name="l01025"></a>01025 <span class="stringliteral">                &#39;</span>Geometry<span class="stringliteral">&#39;    &#39;</span>volume<span class="stringliteral">&#39;  &#39;</span>:<span class="stringliteral">&#39;      &#39;</span>:<span class="stringliteral">&#39;</span>
<a name="l01026"></a>01026 <span class="stringliteral">                &#39;</span><a class="code" href="class_metabolite.html">Metabolite</a><span class="stringliteral">&#39;  &#39;</span>counts<span class="stringliteral">&#39;  metIdxs  comp.cytosolIndexs</span>
<a name="l01027"></a>01027 <span class="stringliteral">                };</span>
<a name="l01028"></a>01028 <span class="stringliteral">            meanMetConcs = zeros(numel(metIdxs), 1, 1, numel(selectedSimulations));</span>
<a name="l01029"></a>01029 <span class="stringliteral">            for i = 1:numel(selectedSimulations)</span>
<a name="l01030"></a>01030 <span class="stringliteral">                states = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, &#39;</span>extract<span class="stringliteral">&#39;, selectedSimulations(i));</span>
<a name="l01031"></a>01031 <span class="stringliteral">                meanMetConcs(:, :, :, i) = mean(states.Metabolite.counts ./ states.Geometry.volume(ones(size(metIdxs)), :, :), 3) / ConstantUtil.nAvogadro * 1000;</span>
<a name="l01032"></a>01032 <span class="stringliteral">                </span>
<a name="l01033"></a>01033 <span class="stringliteral">                clear states;</span>
<a name="l01034"></a>01034 <span class="stringliteral">            end</span>
<a name="l01035"></a>01035 <span class="stringliteral">            </span>
<a name="l01036"></a>01036 <span class="stringliteral">            meanConcs = mean(meanMetConcs, 4);</span>
<a name="l01037"></a>01037 <span class="stringliteral">            stdConcs = std(meanMetConcs, [], 4);</span>
<a name="l01038"></a>01038 <span class="stringliteral">            </span>
<a name="l01039"></a>01039 <span class="stringliteral">            clear meanMetConcs;</span>
<a name="l01040"></a>01040 <span class="stringliteral">            </span>
<a name="l01041"></a>01041 <span class="stringliteral">            %% plot</span>
<a name="l01042"></a>01042 <span class="stringliteral">            clf(figHandle);</span>
<a name="l01043"></a>01043 <span class="stringliteral">            </span>
<a name="l01044"></a>01044 <span class="stringliteral">            axesHandle = subplot(1, 1, 1, &#39;</span>Position<span class="stringliteral">&#39;, [0.1300    0.1500    0.7750    0.8150]);</span>
<a name="l01045"></a>01045 <span class="stringliteral">            errorbar((1:numel(metIdxs))&#39;</span>, meanConcs, meanConcs - max(1<a class="code" href="json_parse_8m.html#ac14d26badced2970320ee4d772ad0ae9">e</a>-3, meanConcs - stdConcs), stdConcs, ...
<a name="l01046"></a>01046                 <span class="stringliteral">&#39;Parent&#39;</span>, axesHandle, <span class="stringliteral">&#39;LineStyle&#39;</span>, <span class="stringliteral">&#39;none&#39;</span>, <span class="stringliteral">&#39;Marker&#39;</span>, <span class="charliteral">&#39;.&#39;</span>);
<a name="l01047"></a>01047             <span class="keyword">set</span>(axesHandle, <span class="stringliteral">&#39;YScale&#39;</span>, <span class="stringliteral">&#39;log&#39;</span>);
<a name="l01048"></a>01048             xlim(axesHandle, [0.5  numel(metIdxs) + 0.5]);
<a name="l01049"></a>01049             <span class="keyword">set</span>(axesHandle, ...
<a name="l01050"></a>01050                 <span class="stringliteral">&#39;XTick&#39;</span>, 1:numel(metIdxs), ...
<a name="l01051"></a>01051                 <span class="stringliteral">&#39;XTickLabel&#39;</span>, [], ...
<a name="l01052"></a>01052                 <span class="stringliteral">&#39;FontSize&#39;</span>, 8, ...
<a name="l01053"></a>01053                 <span class="stringliteral">&#39;TickDir&#39;</span>, <span class="stringliteral">&#39;out&#39;</span>);
<a name="l01054"></a>01054             tick2text(axesHandle, <span class="stringliteral">&#39;axis&#39;</span>, <span class="charliteral">&#39;x&#39;</span>);
<a name="l01055"></a>01055             <span class="keywordflow">for</span> i = 1:numel(metIdxs)
<a name="l01056"></a>01056                 text((i-0.5) / numel(metIdxs), -0.02, m.wholeCellModelIDs{metIdxs(i)}, ...
<a name="l01057"></a>01057                     <span class="stringliteral">&#39;Parent&#39;</span>, axesHandle, ...
<a name="l01058"></a>01058                     <span class="stringliteral">&#39;FontSize&#39;</span>, 6, ...
<a name="l01059"></a>01059                     <span class="stringliteral">&#39;Rotation&#39;</span>, 270, ...
<a name="l01060"></a>01060                     <span class="stringliteral">&#39;VerticalAlignment&#39;</span>, <span class="stringliteral">&#39;middle&#39;</span>, ...
<a name="l01061"></a>01061                     <span class="stringliteral">&#39;HorizontalAlignment&#39;</span>, <span class="stringliteral">&#39;left&#39;</span>, ...
<a name="l01062"></a>01062                     <span class="stringliteral">&#39;Units&#39;</span>, <span class="stringliteral">&#39;normalized&#39;</span>);
<a name="l01063"></a>01063             end
<a name="l01064"></a>01064             axesPos = <span class="keyword">get</span>(axesHandle, <span class="stringliteral">&#39;Position&#39;</span>);
<a name="l01065"></a>01065             
<a name="l01066"></a>01066             annotation(figHandle, <span class="stringliteral">&#39;line&#39;</span>, axesPos(1) + axesPos(3) * ((-0.5 + [1 4] + [-0.25 0.35]) / numel(metIdxs)), axesPos(2) + axesPos(4) * (-0.09 * [1 1]), ...
<a name="l01067"></a>01067                 <span class="stringliteral">&#39;Units&#39;</span>, <span class="stringliteral">&#39;normalized&#39;</span>);
<a name="l01068"></a>01068             annotation(figHandle, <span class="stringliteral">&#39;line&#39;</span>, axesPos(1) + axesPos(3) * ((-0.5 + [5 8] + [-0.25 0.35]) / numel(metIdxs)), axesPos(2) + axesPos(4) * (-0.09 * [1 1]), ...
<a name="l01069"></a>01069                 <span class="stringliteral">&#39;Units&#39;</span>, <span class="stringliteral">&#39;normalized&#39;</span>);
<a name="l01070"></a>01070             annotation(figHandle, <span class="stringliteral">&#39;line&#39;</span>, axesPos(1) + axesPos(3) * ((-0.5 + [9 12] + [-0.25 0.35]) / numel(metIdxs)), axesPos(2) + axesPos(4) * (-0.09 * [1 1]), ...
<a name="l01071"></a>01071                 <span class="stringliteral">&#39;Units&#39;</span>, <span class="stringliteral">&#39;normalized&#39;</span>);
<a name="l01072"></a>01072             annotation(figHandle, <span class="stringliteral">&#39;line&#39;</span>, axesPos(1) + axesPos(3) * ((-0.5 + [13 16] + [-0.25 0.35]) / numel(metIdxs)), axesPos(2) + axesPos(4) * (-0.09 * [1 1]), ...
<a name="l01073"></a>01073                 <span class="stringliteral">&#39;Units&#39;</span>, <span class="stringliteral">&#39;normalized&#39;</span>);
<a name="l01074"></a>01074             annotation(figHandle, <span class="stringliteral">&#39;line&#39;</span>, axesPos(1) + axesPos(3) * ((-0.5 + [17 37] + [-0.25 0.35]) / numel(metIdxs)), axesPos(2) + axesPos(4) * (-0.09 * [1 1]), ...
<a name="l01075"></a>01075                 <span class="stringliteral">&#39;Units&#39;</span>, <span class="stringliteral">&#39;normalized&#39;</span>);
<a name="l01076"></a>01076             annotation(figHandle, <span class="stringliteral">&#39;line&#39;</span>, axesPos(1) + axesPos(3) * ((-0.5 + [38 40] + [-0.25 0.35]) / numel(metIdxs)), axesPos(2) + axesPos(4) * (-0.09 * [1 1]), ...
<a name="l01077"></a>01077                 <span class="stringliteral">&#39;Units&#39;</span>, <span class="stringliteral">&#39;normalized&#39;</span>);
<a name="l01078"></a>01078             
<a name="l01079"></a>01079             text(2 / numel(metIdxs), -0.092, <span class="stringliteral">&#39;NTPs&#39;</span>, ...
<a name="l01080"></a>01080                 <span class="stringliteral">&#39;Parent&#39;</span>, axesHandle, ...
<a name="l01081"></a>01081                 <span class="stringliteral">&#39;FontSize&#39;</span>, 6, ...
<a name="l01082"></a>01082                 <span class="stringliteral">&#39;VerticalAlignment&#39;</span>, <span class="stringliteral">&#39;top&#39;</span>, ...
<a name="l01083"></a>01083                 <span class="stringliteral">&#39;HorizontalAlignment&#39;</span>, <span class="stringliteral">&#39;center&#39;</span>, ...
<a name="l01084"></a>01084                 <span class="stringliteral">&#39;Units&#39;</span>, <span class="stringliteral">&#39;normalized&#39;</span>);
<a name="l01085"></a>01085             text(6 / numel(metIdxs), -0.092, <span class="stringliteral">&#39;NDPs&#39;</span>, ...
<a name="l01086"></a>01086                 <span class="stringliteral">&#39;Parent&#39;</span>, axesHandle, ...
<a name="l01087"></a>01087                 <span class="stringliteral">&#39;FontSize&#39;</span>, 6, ...
<a name="l01088"></a>01088                 <span class="stringliteral">&#39;VerticalAlignment&#39;</span>, <span class="stringliteral">&#39;top&#39;</span>, ...
<a name="l01089"></a>01089                 <span class="stringliteral">&#39;HorizontalAlignment&#39;</span>, <span class="stringliteral">&#39;center&#39;</span>, ...
<a name="l01090"></a>01090                 <span class="stringliteral">&#39;Units&#39;</span>, <span class="stringliteral">&#39;normalized&#39;</span>);
<a name="l01091"></a>01091             text(10 / numel(metIdxs), -0.092, <span class="stringliteral">&#39;NMPs&#39;</span>, ...
<a name="l01092"></a>01092                 <span class="stringliteral">&#39;Parent&#39;</span>, axesHandle, ...
<a name="l01093"></a>01093                 <span class="stringliteral">&#39;FontSize&#39;</span>, 6, ...
<a name="l01094"></a>01094                 <span class="stringliteral">&#39;VerticalAlignment&#39;</span>, <span class="stringliteral">&#39;top&#39;</span>, ...
<a name="l01095"></a>01095                 <span class="stringliteral">&#39;HorizontalAlignment&#39;</span>, <span class="stringliteral">&#39;center&#39;</span>, ...
<a name="l01096"></a>01096                 <span class="stringliteral">&#39;Units&#39;</span>, <span class="stringliteral">&#39;normalized&#39;</span>);
<a name="l01097"></a>01097             text(14 / numel(metIdxs), -0.092, <span class="stringliteral">&#39;dNTPs&#39;</span>, ...
<a name="l01098"></a>01098                 <span class="stringliteral">&#39;Parent&#39;</span>, axesHandle, ...
<a name="l01099"></a>01099                 <span class="stringliteral">&#39;FontSize&#39;</span>, 6, ...
<a name="l01100"></a>01100                 <span class="stringliteral">&#39;VerticalAlignment&#39;</span>, <span class="stringliteral">&#39;top&#39;</span>, ...
<a name="l01101"></a>01101                 <span class="stringliteral">&#39;HorizontalAlignment&#39;</span>, <span class="stringliteral">&#39;center&#39;</span>, ...
<a name="l01102"></a>01102                 <span class="stringliteral">&#39;Units&#39;</span>, <span class="stringliteral">&#39;normalized&#39;</span>);
<a name="l01103"></a>01103             text(26.5 / numel(metIdxs), -0.092, <span class="stringliteral">&#39;AAs&#39;</span>, ...
<a name="l01104"></a>01104                 <span class="stringliteral">&#39;Parent&#39;</span>, axesHandle, ...
<a name="l01105"></a>01105                 <span class="stringliteral">&#39;FontSize&#39;</span>, 6, ...
<a name="l01106"></a>01106                 <span class="stringliteral">&#39;VerticalAlignment&#39;</span>, <span class="stringliteral">&#39;top&#39;</span>, ...
<a name="l01107"></a>01107                 <span class="stringliteral">&#39;HorizontalAlignment&#39;</span>, <span class="stringliteral">&#39;center&#39;</span>, ...
<a name="l01108"></a>01108                 <span class="stringliteral">&#39;Units&#39;</span>, <span class="stringliteral">&#39;normalized&#39;</span>);
<a name="l01109"></a>01109             text(38.5 / numel(metIdxs), -0.092, <span class="stringliteral">&#39;Other&#39;</span>, ...
<a name="l01110"></a>01110                 <span class="stringliteral">&#39;Parent&#39;</span>, axesHandle, ...
<a name="l01111"></a>01111                 <span class="stringliteral">&#39;FontSize&#39;</span>, 6, ...
<a name="l01112"></a>01112                 <span class="stringliteral">&#39;VerticalAlignment&#39;</span>, <span class="stringliteral">&#39;top&#39;</span>, ...
<a name="l01113"></a>01113                 <span class="stringliteral">&#39;HorizontalAlignment&#39;</span>, <span class="stringliteral">&#39;center&#39;</span>, ...
<a name="l01114"></a>01114                 <span class="stringliteral">&#39;Units&#39;</span>, <span class="stringliteral">&#39;normalized&#39;</span>);
<a name="l01115"></a>01115             
<a name="l01116"></a>01116             xlabel(axesHandle, <span class="stringliteral">&#39;Metabolite&#39;</span>, <span class="stringliteral">&#39;FontSize&#39;</span>, 12);
<a name="l01117"></a>01117             <span class="keyword">set</span>(<span class="keyword">get</span>(axesHandle, <span class="stringliteral">&#39;xlabel&#39;</span>), <span class="stringliteral">&#39;Position&#39;</span>, [0.5 -0.12 1], <span class="stringliteral">&#39;Units&#39;</span>, <span class="stringliteral">&#39;normalized&#39;</span>, <span class="stringliteral">&#39;HorizontalAlignment&#39;</span>, <span class="stringliteral">&#39;center&#39;</span>, <span class="stringliteral">&#39;VerticalAlignment&#39;</span>, <span class="stringliteral">&#39;top&#39;</span>);
<a name="l01118"></a>01118             ylabel(axesHandle, <span class="stringliteral">&#39;Concentration (mM)&#39;</span>, <span class="stringliteral">&#39;FontSize&#39;</span>, 12);
<a name="l01119"></a>01119             
<a name="l01120"></a>01120             %% organize figure data
<a name="l01121"></a>01121             figData = <span class="keyword">struct</span>;
<a name="l01122"></a>01122             figData.meanConcs = meanConcs;
<a name="l01123"></a>01123             figData.stdConcs = stdConcs;
<a name="l01124"></a>01124         end
<a name="l01125"></a>01125         
<a name="l01126"></a>01126         <span class="keyword">function</span> figData = translation(figHandle, simBatchDir, selectedSimulations)
<a name="l01127"></a>01127             import edu.stanford.covert.cell.sim.util.PlotUtil;
<a name="l01128"></a>01128             import edu.stanford.covert.cell.sim.util.SimulationEnsemble;
<a name="l01129"></a>01129             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
<a name="l01130"></a>01130             import edu.stanford.covert.util.<a class="code" href="class_constant_util.html">ConstantUtil</a>;
<a name="l01131"></a>01131             
<a name="l01132"></a>01132             if nargin &lt; 2 || isempty(simBatchDir)
<a name="l01133"></a>01133                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
<a name="l01134"></a>01134             end
<a name="l01135"></a>01135             if nargin &lt; 3 || isempty(selectedSimulations)
<a name="l01136"></a>01136                 selectedSimulations = SimulationDiskUtil.getCompleteSimulations(simBatchDir);
<a name="l01137"></a>01137             end
<a name="l01138"></a>01138             
<a name="l01139"></a>01139             [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep &#39;1&#39;]);
<a name="l01140"></a>01140             comp = sim.compartment;
<a name="l01141"></a>01141             g = sim.gene;
<a name="l01142"></a>01142             
<a name="l01143"></a>01143             t = sim.state(&#39;Time&#39;);
<a name="l01144"></a>01144             mass = sim.state(&#39;Mass&#39;);
<a name="l01145"></a>01145             m = sim.state(&#39;<a class="code" href="class_metabolite.html">Metabolite</a>&#39;);
<a name="l01146"></a>01146             rna = sim.state(&#39;Rna&#39;);
<a name="l01147"></a>01147             pm = sim.state(&#39;ProteinMonomer&#39;);
<a name="l01148"></a>01148             pc = sim.state(&#39;ProteinComplex&#39;);
<a name="l01149"></a>01149             rib = sim.state(&#39;Ribosome&#39;);
<a name="l01150"></a>01150             pol = sim.state(&#39;Polypeptide&#39;);
<a name="l01151"></a>01151             
<a name="l01152"></a>01152             tl = sim.process(&#39;<a class="code" href="class_translation.html">Translation</a>&#39;);
<a name="l01153"></a>01153             ta = sim.process(&#39;tRNAAminoacylation&#39;);
<a name="l01154"></a>01154             
<a name="l01155"></a>01155             expTotRnas = mass.initialFractionNTPsInRNAs * mass.dryWeightFractionRNA * mass.cellInitialDryWeight * rna.expression(rna.matureIndexs) / ...
<a name="l01156"></a>01156                 (rna.expression(rna.matureIndexs)&#39; * rna.molecularWeights(rna.matureIndexs) / ConstantUtil.nAvogadro);
<a name="l01157"></a>01157             monExp = (rna.matureRNAGeneComposition(g.mRNAIndexs, :) * rna.expression(rna.matureIndexs)) ./ ...
<a name="l01158"></a>01158                 (log(2) / t.cellCycleLength + pm.decayRates(pm.matureIndexs));
<a name="l01159"></a>01159             expTotMons = mass.initialFractionAAsInMonomers * mass.dryWeightFractionProtein * mass.cellInitialDryWeight * ...
<a name="l01160"></a>01160                 monExp / (monExp&#39; * pm.molecularWeights(pm.matureIndexs) / ConstantUtil.nAvogadro);
<a name="l01161"></a>01161             expTotCpxs = min(repmat(expTotMons, [1 numel(pc.matureIndexs)]) ./ sum(pc.proteinComplexComposition(g.mRNAIndexs, :, :), 3), [], 1);
<a name="l01162"></a>01162             
<a name="l01163"></a>01163             %% get data
<a name="l01164"></a>01164             %mass, energy
<a name="l01165"></a>01165             ensemble = SimulationEnsemble(simBatchDir, {<span class="stringliteral">&#39;mass&#39;</span> <span class="stringliteral">&#39;mass&#39;</span>; <span class="stringliteral">&#39;atp&#39;</span> <span class="stringliteral">&#39;atp&#39;</span>}, [], selectedSimulations);
<a name="l01166"></a>01166             time = ensemble.stateData.time / 3600;
<a name="l01167"></a>01167             mass = ensemble.stateData.values(ensemble.getPropertyIndices(<span class="stringliteral">&#39;mass&#39;</span>), :, :, :) * sim.state(<span class="stringliteral">&#39;Mass&#39;</span>).cellInitialDryWeight / (1-sim.state(&#39;Mass&#39;).fractionWetWeight) * 1e15;
<a name="l01168"></a>01168             simEndTimes = ensemble.stateData.simulationEndTimes;
<a name="l01169"></a>01169             
<a name="l01170"></a>01170             [~, simOrder] = sort(mass(:, 1, min(simEndTimes), :));
<a name="l01171"></a>01171             simOrder = simOrder(:);
<a name="l01172"></a>01172             
<a name="l01173"></a>01173             %cleanup
<a name="l01174"></a>01174             clear ensemble mass;
<a name="l01175"></a>01175             
<a name="l01176"></a>01176             %metabolites, rna, proteins
<a name="l01177"></a>01177             monIdxs = [tl.enzymeMonomerGlobalIndexs; ta.enzymeMonomerGlobalIndexs];
<a name="l01178"></a>01178             cpxIdxs = [tl.enzymeComplexGlobalIndexs; ta.enzymeComplexGlobalIndexs];
<a name="l01179"></a>01179             stateNames = {
<a name="l01180"></a>01180                 <span class="stringliteral">&#39;Metabolite&#39;</span>     <span class="stringliteral">&#39;counts&#39;</span> [m.ntpIndexs([1 3]); m.ndpIndexs([1 3]); m.nmpIndexs(1); m.aminoAcidIndexs; m.getIndexs(&#39;FTHF10&#39;)] comp.cytosolIndexs
<a name="l01181"></a>01181                 <span class="stringliteral">&#39;Rna&#39;</span>            <span class="stringliteral">&#39;counts&#39;</span> [rna.matureIndexs(rna.matureTRNAIndexs); rna.aminoacylatedIndexs(rna.matureTRNAIndexs)] comp.cytosolIndexs
<a name="l01182"></a>01182                 <span class="stringliteral">&#39;ProteinMonomer&#39;</span> <span class="stringliteral">&#39;counts&#39;</span> [pm.matureIndexs(monIdxs) pm.boundIndexs(monIdxs)] comp.cytosolIndexs
<a name="l01183"></a>01183                 <span class="stringliteral">&#39;ProteinComplex&#39;</span> <span class="stringliteral">&#39;counts&#39;</span> [pc.matureIndexs(cpxIdxs) pc.boundIndexs(cpxIdxs)] comp.cytosolIndexs
<a name="l01184"></a>01184                 };
<a name="l01185"></a>01185             states = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, <span class="stringliteral">&#39;extract&#39;</span>, selectedSimulations);
<a name="l01186"></a>01186             
<a name="l01187"></a>01187             tmp = full(states.Metabolite.counts);
<a name="l01188"></a>01188             atp = permute(tmp(1, :, :, :), [4 3 1 2]);
<a name="l01189"></a>01189             gtp = permute(tmp(2, :, :, :), [4 3 1 2]);
<a name="l01190"></a>01190             adp = permute(tmp(3, :, :, :), [4 3 1 2]);
<a name="l01191"></a>01191             gdp = permute(tmp(4, :, :, :), [4 3 1 2]);
<a name="l01192"></a>01192             amp = permute(tmp(5, :, :, :), [4 3 1 2]);
<a name="l01193"></a>01193             aas = permute(sum(tmp(6:26, :, :, :), 1), [4 3 1 2]);
<a name="l01194"></a>01194             fthf10 = permute(sum(tmp(27:end, :, :, :), 1), [4 3 1 2]);
<a name="l01195"></a>01195             
<a name="l01196"></a>01196             monCnts = full(states.ProteinMonomer.counts);
<a name="l01197"></a>01197             cpxCnts = full(states.ProteinComplex.counts);
<a name="l01198"></a>01198             
<a name="l01199"></a>01199             enzProcesses = [
<a name="l01200"></a>01200                 repmat({&#39;TL&#39;}, numel(tl.enzymeWholeCellModelIDs), 1)
<a name="l01201"></a>01201                 repmat({&#39;TA&#39;}, numel(ta.enzymeWholeCellModelIDs)-numel(ta.enzymeRNALocalIndexs), 1)
<a name="l01202"></a>01202                 ];
<a name="l01203"></a>01203             enzProcessIdxs = [
<a name="l01204"></a>01204                 (1:numel(tl.enzymeWholeCellModelIDs))&#39;
<a name="l01205"></a>01205                 (1:numel(ta.enzymeWholeCellModelIDs)-numel(ta.enzymeRNALocalIndexs))&#39;
<a name="l01206"></a>01206                 ];
<a name="l01207"></a>01207             
<a name="l01208"></a>01208             enzymes = zeros(0, 1, size(monCnts, 3), size(monCnts, 4));
<a name="l01209"></a>01209             expEnzymes = zeros(0, 1);
<a name="l01210"></a>01210             
<a name="l01211"></a>01211             tmp = zeros(numel(tl.enzymes), 1, size(monCnts, 3), size(monCnts, 4));
<a name="l01212"></a>01212             tmp(tl.enzymeMonomerLocalIndexs, :, :, :) = monCnts(1:numel(tl.enzymeMonomerGlobalIndexs), :, :, :);
<a name="l01213"></a>01213             tmp(tl.enzymeComplexLocalIndexs, :, :, :) = cpxCnts(1:numel(tl.enzymeComplexGlobalIndexs), :, :, :);
<a name="l01214"></a>01214             tmp2 = zeros(numel(tl.enzymes), 1);
<a name="l01215"></a>01215             tmp2(tl.enzymeMonomerLocalIndexs) = expTotMons(tl.enzymeMonomerGlobalIndexs);
<a name="l01216"></a>01216             tmp2(tl.enzymeComplexLocalIndexs) = expTotCpxs(tl.enzymeComplexGlobalIndexs);
<a name="l01217"></a>01217             enzymes = [enzymes; tmp];
<a name="l01218"></a>01218             expEnzymes = [expEnzymes; tmp2];
<a name="l01219"></a>01219             
<a name="l01220"></a>01220             tmp = zeros(numel(ta.enzymes), 1, size(monCnts, 3), size(monCnts, 4));
<a name="l01221"></a>01221             tmp(ta.enzymeMonomerLocalIndexs, :, :, :) = monCnts(end-numel(ta.enzymeMonomerGlobalIndexs)+1:end, :, :, :);
<a name="l01222"></a>01222             tmp(ta.enzymeComplexLocalIndexs, :, :, :) = cpxCnts(end-numel(ta.enzymeComplexGlobalIndexs)+1:end, :, :, :);
<a name="l01223"></a>01223             tmp2 = zeros(numel(ta.enzymes), 1);
<a name="l01224"></a>01224             tmp2(ta.enzymeMonomerLocalIndexs) = expTotMons(ta.enzymeMonomerGlobalIndexs);
<a name="l01225"></a>01225             tmp2(ta.enzymeComplexLocalIndexs) = expTotCpxs(ta.enzymeComplexGlobalIndexs);
<a name="l01226"></a>01226             tmp(ta.enzymeRNALocalIndexs, :, :, :) = [];
<a name="l01227"></a>01227             tmp2(ta.enzymeRNALocalIndexs) = [];
<a name="l01228"></a>01228             enzymes = [enzymes; tmp];
<a name="l01229"></a>01229             expEnzymes = [expEnzymes; tmp2];
<a name="l01230"></a>01230             
<a name="l01231"></a>01231             expEnzymes = repmat(expEnzymes, [1 1 numel(time)-1 1]) .* exp(log(2) * repmat(permute(time(2:end),[1 3 2]), size(expEnzymes, 1), 1) / (t.cellCycleLength/3600));
<a name="l01232"></a>01232             
<a name="l01233"></a>01233             trnas = states.Rna.counts(1:end/2, :, :, :);
<a name="l01234"></a>01234             aatrnas = states.Rna.counts(end/2+1:end, :, :, :);
<a name="l01235"></a>01235             expTRNAs = repmat(expTotRnas(rna.matureTRNAIndexs), [1 1 numel(time)-1 1]) .* exp(log(2) * repmat(permute(time(2:end),[1 3 2]), size(rna.matureTRNAIndexs, 1), 1) / (t.cellCycleLength/3600));
<a name="l01236"></a>01236             
<a name="l01237"></a>01237             <span class="keywordflow">for</span> i = 1:numel(simEndTimes)
<a name="l01238"></a>01238                 atp(i, simEndTimes(i)+1:end) = NaN;
<a name="l01239"></a>01239                 gtp(i, simEndTimes(i)+1:end) = NaN;
<a name="l01240"></a>01240                 adp(i, simEndTimes(i)+1:end) = NaN;
<a name="l01241"></a>01241                 gdp(i, simEndTimes(i)+1:end) = NaN;
<a name="l01242"></a>01242                 amp(i, simEndTimes(i)+1:end) = NaN;
<a name="l01243"></a>01243                 aas(i, simEndTimes(i)+1:end) = NaN;
<a name="l01244"></a>01244                 fthf10(i, simEndTimes(i)+1:end) = NaN;
<a name="l01245"></a>01245                 enzymes(:, :, simEndTimes(i)+1:end, i) = NaN;
<a name="l01246"></a>01246                 trnas(:, :, simEndTimes(i)+1:end, i) = NaN;
<a name="l01247"></a>01247                 aatrnas(:, :, simEndTimes(i)+1:end, i) = NaN;
<a name="l01248"></a>01248             end
<a name="l01249"></a>01249             
<a name="l01250"></a>01250             %cleanup
<a name="l01251"></a>01251             clear states rnaCnts monCnts cpxCnts;
<a name="l01252"></a>01252             
<a name="l01253"></a>01253             %ribosomes
<a name="l01254"></a>01254             pols = NaN(4, 1, numel(time)-1, numel(selectedSimulations));
<a name="l01255"></a>01255             stateNames = {
<a name="l01256"></a>01256                 <span class="stringliteral">&#39;Ribosome&#39;</span>  <span class="stringliteral">&#39;states&#39;</span>          <span class="charliteral">&#39;:&#39;</span> <span class="charliteral">&#39;:&#39;</span>
<a name="l01257"></a>01257                 <span class="stringliteral">&#39;Ribosome&#39;</span>  <span class="stringliteral">&#39;boundMRNAs&#39;</span>      <span class="charliteral">&#39;:&#39;</span> <span class="charliteral">&#39;:&#39;</span>
<a name="l01258"></a>01258                 <span class="stringliteral">&#39;Ribosome&#39;</span>  <span class="stringliteral">&#39;mRNAPositions&#39;</span>   <span class="charliteral">&#39;:&#39;</span> <span class="charliteral">&#39;:&#39;</span>
<a name="l01259"></a>01259                 <span class="stringliteral">&#39;Ribosome&#39;</span>  <span class="stringliteral">&#39;tmRNAPositions&#39;</span>  <span class="charliteral">&#39;:&#39;</span> <span class="charliteral">&#39;:&#39;</span>
<a name="l01260"></a>01260                 };
<a name="l01261"></a>01261             <span class="keywordflow">for</span> j = 1:numel(selectedSimulations)
<a name="l01262"></a>01262                 states = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, &#39;extract&#39;, selectedSimulations(j));
<a name="l01263"></a>01263                 
<a name="l01264"></a>01264                 for i = 1:size(states.Ribosome.states, 3)
<a name="l01265"></a>01265                     tfs = states.Ribosome.states(:, :, i) == rib.activeValue | states.Ribosome.states(:, :, i) == rib.stalledValue;
<a name="l01266"></a>01266                     boundMRNAs = states.Ribosome.boundMRNAs(tfs, :, i);
<a name="l01267"></a>01267                     lengths1 = states.Ribosome.mRNAPositions(tfs, :, i);
<a name="l01268"></a>01268                     lengths2 = states.Ribosome.tmRNAPositions(tfs, :, i);
<a name="l01269"></a>01269                     
<a name="l01270"></a>01270                     pols(1, 1, i, j) = sum(pm.lengths(pm.nascentIndexs(boundMRNAs)) == lengths1 | lengths2 == pol.proteolysisTagLength);
<a name="l01271"></a>01271                     pols(2, 1, i, j) = sum(lengths1 == 0 &amp; lengths2 == 0);
<a name="l01272"></a>01272                     pols(4, 1, i, j) = sum(states.Ribosome.states(tfs, :, i) == rib.stalledValue &amp; lengths2 ~= pol.proteolysisTagLength);
<a name="l01273"></a>01273                     pols(3, 1, i, j) = sum(tfs) - sum(pols([1 2 4], 1, i, j));
<a name="l01274"></a>01274                 end
<a name="l01275"></a>01275                 
<a name="l01276"></a>01276                 clear states tfs boundMRNAs lengths1 lengths2;
<a name="l01277"></a>01277             end
<a name="l01278"></a>01278             
<a name="l01279"></a>01279             %% clear figure
<a name="l01280"></a>01280             clf(figHandle);
<a name="l01281"></a>01281             C = 70;
<a name="l01282"></a>01282             
<a name="l01283"></a>01283             %% left panel
<a name="l01284"></a>01284             
<a name="l01285"></a>01285             %layout plots
<a name="l01286"></a>01286             h = (C - 10 + 1) / 10;
<a name="l01287"></a>01287             axesHandlesL = PlotUtil.multiElementPlot(figHandle, h * ones(11, 1), [0 time(end)], struct(...
<a name="l01288"></a>01288                 &#39;titleStr&#39;, &#39;Metabolites&#39;, ...
<a name="l01289"></a>01289                 &#39;position&#39;, [0.10 0.10 0.20 0.8], ...
<a name="l01290"></a>01290                 &#39;colorOrder&#39;, PlotUtil.getRedGreenColorOrder(simOrder)));
<a name="l01291"></a>01291             
<a name="l01292"></a>01292             %plot data
<a name="l01293"></a>01293             PlotUtil.plotLine(axesHandlesL(1), time(2:end), atp, false, true, false);
<a name="l01294"></a>01294             ylabel(axesHandlesL(1), &#39;ATP&#39;);
<a name="l01295"></a>01295             
<a name="l01296"></a>01296             PlotUtil.plotLine(axesHandlesL(2), time(2:end), gtp, false, true, false);
<a name="l01297"></a>01297             ylabel(axesHandlesL(2), &#39;GTP&#39;);
<a name="l01298"></a>01298             
<a name="l01299"></a>01299             PlotUtil.plotLine(axesHandlesL(3), time(2:end), adp, false, true, false);
<a name="l01300"></a>01300             ylabel(axesHandlesL(3), &#39;ADP&#39;);
<a name="l01301"></a>01301             
<a name="l01302"></a>01302             PlotUtil.plotLine(axesHandlesL(4), time(2:end), gdp, false, true, false);
<a name="l01303"></a>01303             ylabel(axesHandlesL(4), &#39;GDP&#39;);
<a name="l01304"></a>01304             
<a name="l01305"></a>01305             PlotUtil.plotLine(axesHandlesL(5), time(2:end), amp, false, true, false);
<a name="l01306"></a>01306             ylabel(axesHandlesL(5), &#39;AMP&#39;);
<a name="l01307"></a>01307             
<a name="l01308"></a>01308             PlotUtil.plotLine(axesHandlesL(6), time(2:end), aas, false, true, false);
<a name="l01309"></a>01309             ylabel(axesHandlesL(6), &#39;AAs&#39;);
<a name="l01310"></a>01310             
<a name="l01311"></a>01311             PlotUtil.plotLine(axesHandlesL(7), time(2:end), fthf10, false, true, false);
<a name="l01312"></a>01312             ylabel(axesHandlesL(7), &#39;FTHF10&#39;);
<a name="l01313"></a>01313             
<a name="l01314"></a>01314             PlotUtil.plotLine(axesHandlesL(8), time(2:end), permute(pols(2, :, :, :), [4 3 1 2]), false, true, false);
<a name="l01315"></a>01315             ylabel(axesHandlesL(8), {<span class="stringliteral">&#39;Init&#39;</span> <span class="stringliteral">&#39;Ribs&#39;</span>});
<a name="l01316"></a>01316             
<a name="l01317"></a>01317             PlotUtil.plotLine(axesHandlesL(9), time(2:end), permute(pols(3, :, :, :), [4 3 1 2]), <span class="keyword">false</span>, <span class="keyword">true</span>, <span class="keyword">false</span>);
<a name="l01318"></a>01318             ylabel(axesHandlesL(9), {<span class="stringliteral">&#39;Elng&#39;</span> <span class="stringliteral">&#39;Ribs&#39;</span>});
<a name="l01319"></a>01319             
<a name="l01320"></a>01320             PlotUtil.plotLine(axesHandlesL(10), time(2:end), permute(pols(1, :, :, :), [4 3 1 2]), <span class="keyword">false</span>, <span class="keyword">true</span>, <span class="keyword">false</span>);
<a name="l01321"></a>01321             ylabel(axesHandlesL(10), {<span class="stringliteral">&#39;Term&#39;</span> <span class="stringliteral">&#39;Ribs&#39;</span>});
<a name="l01322"></a>01322             
<a name="l01323"></a>01323             PlotUtil.plotLine(axesHandlesL(11), time(2:end), permute(pols(4, :, :, :), [4 3 1 2]), <span class="keyword">false</span>, <span class="keyword">true</span>, <span class="keyword">false</span>);
<a name="l01324"></a>01324             ylabel(axesHandlesL(11), {<span class="stringliteral">&#39;Stalled&#39;</span> <span class="stringliteral">&#39;Ribs&#39;</span>});
<a name="l01325"></a>01325             
<a name="l01326"></a>01326             % Format Y-axis
<a name="l01327"></a>01327             PlotUtil.alignYAxesLabels(axesHandlesL);
<a name="l01328"></a>01328             PlotUtil.offsetYAxes(axesHandlesL, 0.04);
<a name="l01329"></a>01329             PlotUtil.labelSubplots(axesHandlesL(1), <span class="charliteral">&#39;A&#39;</span>, -0.42, 1.2);
<a name="l01330"></a>01330             
<a name="l01331"></a>01331             %% middle panels
<a name="l01332"></a>01332             
<a name="l01333"></a>01333             nPlots1 = ceil(size(enzymes, 1) / 2);
<a name="l01334"></a>01334             nPlots2 = floor(size(enzymes, 1) / 2);
<a name="l01335"></a>01335             
<a name="l01336"></a>01336             %layout plots
<a name="l01337"></a>01337             colorOrder = [PlotUtil.getRedGreenColorOrder(simOrder); 0.5 0.5 0.5];
<a name="l01338"></a>01338             h = (C - ceil(size(enzymes, 1)/2) + 1) / ceil(size(enzymes, 1) / 2);
<a name="l01339"></a>01339             axesHandlesM1 = PlotUtil.multiElementPlot(figHandle, h * ones(nPlots1, 1), [0 time(end)], <span class="keyword">struct</span>(...
<a name="l01340"></a>01340                 <span class="stringliteral">&#39;titleStr&#39;</span>, <span class="stringliteral">&#39;Enzymes&#39;</span>, ...
<a name="l01341"></a>01341                 <span class="stringliteral">&#39;position&#39;</span>, [0.36 0.10 0.10 0.8], ...
<a name="l01342"></a>01342                 <span class="stringliteral">&#39;colorOrder&#39;</span>, colorOrder));
<a name="l01343"></a>01343             axesHandlesM2 = PlotUtil.multiElementPlot(figHandle, h * ones(nPlots1, 1), [0 time(end)], <span class="keyword">struct</span>(...
<a name="l01344"></a>01344                 <span class="stringliteral">&#39;position&#39;</span>, [0.53 0.10 0.10 0.8], ...
<a name="l01345"></a>01345                 <span class="stringliteral">&#39;colorOrder&#39;</span>, colorOrder));
<a name="l01346"></a>01346             
<a name="l01347"></a>01347             %plot data
<a name="l01348"></a>01348             <span class="keywordflow">for</span> i = 1:nPlots1
<a name="l01349"></a>01349                 PlotUtil.plotLine(axesHandlesM1(i), time(2:end), permute(cat(4, enzymes(i, :, :, :), expEnzymes(i, :, :, :)), [4 3 1 2]), <span class="keyword">false</span>, <span class="keyword">true</span>, <span class="keyword">false</span>);
<a name="l01350"></a>01350                 ylabel(axesHandlesM1(i), [enzProcesses(i) num2str(enzProcessIdxs(i))]);
<a name="l01351"></a>01351             end
<a name="l01352"></a>01352             <span class="keywordflow">for</span> i = 1:nPlots2
<a name="l01353"></a>01353                 PlotUtil.plotLine(axesHandlesM2(i), time(2:end), permute(cat(4, enzymes(nPlots1+i, :, :, :), expEnzymes(nPlots1+i, :, :, :)), [4 3 1 2]), <span class="keyword">false</span>, <span class="keyword">true</span>, <span class="keyword">false</span>);
<a name="l01354"></a>01354                 ylabel(axesHandlesM2(i), [enzProcesses(nPlots1+i) num2str(enzProcessIdxs(nPlots1+i))]);
<a name="l01355"></a>01355             end
<a name="l01356"></a>01356             
<a name="l01357"></a>01357             % Format Y-axis
<a name="l01358"></a>01358             PlotUtil.alignYAxesLabels(axesHandlesM1);
<a name="l01359"></a>01359             PlotUtil.alignYAxesLabels(axesHandlesM2);
<a name="l01360"></a>01360             axesHandlesM12 = PlotUtil.offsetYAxes(axesHandlesM1, 0.04);
<a name="l01361"></a>01361             axesHandlesM22 = PlotUtil.offsetYAxes(axesHandlesM2, 0.04);
<a name="l01362"></a>01362             PlotUtil.labelSubplots(axesHandlesM1(1), <span class="charliteral">&#39;B&#39;</span>, -0.42, 1.2);
<a name="l01363"></a>01363             
<a name="l01364"></a>01364             <span class="keyword">set</span>(axesHandlesM1(nPlots1+1:end), <span class="stringliteral">&#39;visible&#39;</span>, <span class="stringliteral">&#39;off&#39;</span>)
<a name="l01365"></a>01365             <span class="keyword">set</span>(axesHandlesM2(nPlots2+1:end), <span class="stringliteral">&#39;visible&#39;</span>, <span class="stringliteral">&#39;off&#39;</span>)
<a name="l01366"></a>01366             <span class="keyword">set</span>(axesHandlesM12(nPlots1+1:end), <span class="stringliteral">&#39;visible&#39;</span>, <span class="stringliteral">&#39;off&#39;</span>)
<a name="l01367"></a>01367             <span class="keyword">set</span>(axesHandlesM22(nPlots2+1:end), <span class="stringliteral">&#39;visible&#39;</span>, <span class="stringliteral">&#39;off&#39;</span>)
<a name="l01368"></a>01368             
<a name="l01369"></a>01369             %% right panels
<a name="l01370"></a>01370             nPlots1 = ceil(size(trnas, 1) / 2);
<a name="l01371"></a>01371             nPlots2 = floor(size(trnas, 1) / 2);
<a name="l01372"></a>01372             
<a name="l01373"></a>01373             %layout plots
<a name="l01374"></a>01374             colorOrder = [PlotUtil.getRedGreenColorOrder(simOrder); PlotUtil.getRedGreenColorOrder(simOrder); 0.5 0.5 0.5];
<a name="l01375"></a>01375             h = (C - ceil(size(trnas, 1)/2) + 1) / ceil(size(trnas, 1) / 2);
<a name="l01376"></a>01376             axesHandlesR1 = PlotUtil.multiElementPlot(figHandle, h * ones(nPlots1, 1), [0 time(end)], <span class="keyword">struct</span>(...
<a name="l01377"></a>01377                 <span class="stringliteral">&#39;titleStr&#39;</span>, <span class="stringliteral">&#39;tRNA&#39;</span>, ...
<a name="l01378"></a>01378                 <span class="stringliteral">&#39;position&#39;</span>, [0.70 0.10 0.10 0.8], ...
<a name="l01379"></a>01379                 <span class="stringliteral">&#39;colorOrder&#39;</span>, colorOrder));
<a name="l01380"></a>01380             axesHandlesR2 = PlotUtil.multiElementPlot(figHandle, h * ones(nPlots1, 1), [0 time(end)], <span class="keyword">struct</span>(...
<a name="l01381"></a>01381                 <span class="stringliteral">&#39;position&#39;</span>, [0.87 0.10 0.10 0.8], ...
<a name="l01382"></a>01382                 <span class="stringliteral">&#39;colorOrder&#39;</span>, colorOrder));
<a name="l01383"></a>01383             
<a name="l01384"></a>01384             %plot data
<a name="l01385"></a>01385             <span class="keywordflow">for</span> i = 1:nPlots1
<a name="l01386"></a>01386                 PlotUtil.plotLine(axesHandlesR1(i), time(2:end), permute(cat(4, trnas(i, :, :, :), aatrnas(i, :, :, :), expTRNAs(i, :, :, :)), [4 3 1 2]), <span class="keyword">false</span>, <span class="keyword">true</span>, <span class="keyword">false</span>);
<a name="l01387"></a>01387                 ylabel(axesHandlesR1(i), num2str(i));
<a name="l01388"></a>01388             end
<a name="l01389"></a>01389             <span class="keywordflow">for</span> i = 1:nPlots2
<a name="l01390"></a>01390                 PlotUtil.plotLine(axesHandlesR2(i), time(2:end), permute(cat(4, trnas(nPlots1+i, :, :, :), aatrnas(nPlots1+i, :, :, :), expTRNAs(nPlots1+i, :, :, :)), [4 3 1 2]), <span class="keyword">false</span>, <span class="keyword">true</span>, <span class="keyword">false</span>);
<a name="l01391"></a>01391                 ylabel(axesHandlesR2(i), num2str(nPlots1+i));
<a name="l01392"></a>01392             end
<a name="l01393"></a>01393             
<a name="l01394"></a>01394             % Format Y-axis
<a name="l01395"></a>01395             PlotUtil.alignYAxesLabels(axesHandlesR1);
<a name="l01396"></a>01396             PlotUtil.alignYAxesLabels(axesHandlesR2);
<a name="l01397"></a>01397             axesHandlesR12 = PlotUtil.offsetYAxes(axesHandlesR1, 0.04);
<a name="l01398"></a>01398             axesHandlesR22 = PlotUtil.offsetYAxes(axesHandlesR2, 0.04);
<a name="l01399"></a>01399             PlotUtil.labelSubplots(axesHandlesR1(1), <span class="charliteral">&#39;C&#39;</span>, -0.42, 1.2);
<a name="l01400"></a>01400             
<a name="l01401"></a>01401             <span class="keyword">set</span>(axesHandlesR1(nPlots1+1:end), <span class="stringliteral">&#39;visible&#39;</span>, <span class="stringliteral">&#39;off&#39;</span>)
<a name="l01402"></a>01402             <span class="keyword">set</span>(axesHandlesR2(nPlots2+1:end), <span class="stringliteral">&#39;visible&#39;</span>, <span class="stringliteral">&#39;off&#39;</span>)
<a name="l01403"></a>01403             <span class="keyword">set</span>(axesHandlesR12(nPlots1+1:end), <span class="stringliteral">&#39;visible&#39;</span>, <span class="stringliteral">&#39;off&#39;</span>)
<a name="l01404"></a>01404             <span class="keyword">set</span>(axesHandlesR22(nPlots2+1:end), <span class="stringliteral">&#39;visible&#39;</span>, <span class="stringliteral">&#39;off&#39;</span>)
<a name="l01405"></a>01405             
<a name="l01406"></a>01406             %% organize figure data
<a name="l01407"></a>01407             figData = <span class="keyword">struct</span>;
<a name="l01408"></a>01408             figData.selectedSimulations = selectedSimulations;
<a name="l01409"></a>01409             figData.simOrder = simOrder;
<a name="l01410"></a>01410             figData.time = time;
<a name="l01411"></a>01411             figData.atp = atp;
<a name="l01412"></a>01412             figData.gtp = gtp;
<a name="l01413"></a>01413             figData.adp = adp;
<a name="l01414"></a>01414             figData.gdp = gdp;
<a name="l01415"></a>01415             figData.amp = amp;
<a name="l01416"></a>01416             figData.aas = aas;
<a name="l01417"></a>01417             figData.fthf10 = fthf10;
<a name="l01418"></a>01418             figData.enzProcesses = enzProcesses;
<a name="l01419"></a>01419             figData.enzProcessIdxs = enzProcessIdxs;
<a name="l01420"></a>01420             figData.enzymes = enzymes;
<a name="l01421"></a>01421             figData.expEnzymes = expEnzymes;
<a name="l01422"></a>01422             figData.trnas = trnas;
<a name="l01423"></a>01423             figData.aatrnas = aatrnas;
<a name="l01424"></a>01424             figData.expTRNAs = expTRNAs;
<a name="l01425"></a>01425             figData.pols = pols;
<a name="l01426"></a>01426         end
<a name="l01427"></a>01427         
<a name="l01428"></a>01428         <span class="keyword">function</span> figData = proteinSynthesis(figHandle, simBatchDir, selectedSimulations)
<a name="l01429"></a>01429             import edu.stanford.covert.cell.sim.util.PlotUtil;
<a name="l01430"></a>01430             import edu.stanford.covert.cell.sim.util.SimulationEnsemble;
<a name="l01431"></a>01431             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
<a name="l01432"></a>01432             
<a name="l01433"></a>01433             if nargin &lt; 2 || isempty(simBatchDir)
<a name="l01434"></a>01434                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
<a name="l01435"></a>01435             end
<a name="l01436"></a>01436             if nargin &lt; 3 || isempty(selectedSimulations)
<a name="l01437"></a>01437                 selectedSimulations = SimulationDiskUtil.getCompleteSimulations(simBatchDir);
<a name="l01438"></a>01438             end
<a name="l01439"></a>01439             
<a name="l01440"></a>01440             [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep &#39;1&#39;]);
<a name="l01441"></a>01441             comp = sim.compartment;
<a name="l01442"></a>01442             pm = sim.state(&#39;ProteinMonomer&#39;);
<a name="l01443"></a>01443             pc = sim.state(&#39;ProteinComplex&#39;);
<a name="l01444"></a>01444             
<a name="l01445"></a>01445             ppI = sim.process(&#39;ProteinProcessingI&#39;);
<a name="l01446"></a>01446             ppII = sim.process(&#39;ProteinProcessingII&#39;);
<a name="l01447"></a>01447             pt = sim.process(&#39;ProteinTranslocation&#39;);
<a name="l01448"></a>01448             pf = sim.process(&#39;ProteinFolding&#39;);
<a name="l01449"></a>01449             pmod = sim.process(&#39;ProteinModification&#39;);
<a name="l01450"></a>01450             pd = sim.process(&#39;ProteinDecay&#39;);
<a name="l01451"></a>01451             
<a name="l01452"></a>01452             enzProcesses = [
<a name="l01453"></a>01453                 repmat({<span class="stringliteral">&#39;PP-I&#39;</span>}, size(ppI.enzymeWholeCellModelIDs))
<a name="l01454"></a>01454                 repmat({<span class="stringliteral">&#39;PT&#39;</span>}, size(pt.enzymeWholeCellModelIDs))
<a name="l01455"></a>01455                 repmat({<span class="stringliteral">&#39;PP-II&#39;</span>}, size(ppII.enzymeWholeCellModelIDs))
<a name="l01456"></a>01456                 repmat({<span class="stringliteral">&#39;PF&#39;</span>}, size(pf.enzymeWholeCellModelIDs))
<a name="l01457"></a>01457                 repmat({<span class="stringliteral">&#39;PMod&#39;</span>}, size(pmod.enzymeWholeCellModelIDs))
<a name="l01458"></a>01458                 repmat({<span class="stringliteral">&#39;PD&#39;</span>}, size(pd.enzymeWholeCellModelIDs))
<a name="l01459"></a>01459                 ];
<a name="l01460"></a>01460             enzProcessIdxs = [
<a name="l01461"></a>01461                 (1:size(ppI.enzymeWholeCellModelIDs))&#39;
<a name="l01462"></a>01462                 (1:size(pt.enzymeWholeCellModelIDs))&#39;
<a name="l01463"></a>01463                 (1:size(ppII.enzymeWholeCellModelIDs))&#39;
<a name="l01464"></a>01464                 (1:size(pf.enzymeWholeCellModelIDs))&#39;
<a name="l01465"></a>01465                 (1:size(pmod.enzymeWholeCellModelIDs))&#39;
<a name="l01466"></a>01466                 (1:size(pd.enzymeWholeCellModelIDs))&#39;
<a name="l01467"></a>01467                 ];
<a name="l01468"></a>01468             
<a name="l01469"></a>01469             %% <span class="keyword">get</span> data
<a name="l01470"></a>01470             ensemble = SimulationEnsemble(simBatchDir, {<span class="stringliteral">&#39;mass&#39;</span> <span class="stringliteral">&#39;mass&#39;</span>; <span class="stringliteral">&#39;immatureMonomers&#39;</span> <span class="stringliteral">&#39;immatureMonomers&#39;</span>}, [], selectedSimulations);
<a name="l01471"></a>01471             time = ensemble.stateData.time / 3600;
<a name="l01472"></a>01472             mass = ensemble.stateData.values(ensemble.getPropertyIndices(<span class="stringliteral">&#39;mass&#39;</span>), :, :, :) * sim.state(<span class="stringliteral">&#39;Mass&#39;</span>).cellInitialDryWeight / (1-sim.state(&#39;Mass&#39;).fractionWetWeight) * 1e15;
<a name="l01473"></a>01473             simEndTimes = ensemble.stateData.simulationEndTimes;
<a name="l01474"></a>01474             
<a name="l01475"></a>01475             [~, simOrder] = sort(mass(:, 1, min(simEndTimes), :));
<a name="l01476"></a>01476             simOrder = simOrder(:);
<a name="l01477"></a>01477             
<a name="l01478"></a>01478             clear ensemble;
<a name="l01479"></a>01479             
<a name="l01480"></a>01480             stateNames = {
<a name="l01481"></a>01481                 <span class="stringliteral">&#39;Polypeptide&#39;</span>    <span class="stringliteral">&#39;abortedPolypeptides&#39;</span>  <span class="stringliteral">&#39;-nnz&#39;</span>           1
<a name="l01482"></a>01482                 <span class="stringliteral">&#39;ProteinMonomer&#39;</span> <span class="stringliteral">&#39;counts&#39;</span>               <span class="charliteral">&#39;:&#39;</span>              <span class="charliteral">&#39;:&#39;</span>
<a name="l01483"></a>01483                 <span class="stringliteral">&#39;ProteinComplex&#39;</span> <span class="stringliteral">&#39;counts&#39;</span>               pc.matureIndexs  <span class="stringliteral">&#39;-sum&#39;</span>
<a name="l01484"></a>01484                 };
<a name="l01485"></a>01485             nscnt = NaN(numel(selectedSimulations), max(simEndTimes));
<a name="l01486"></a>01486             procsI = NaN(numel(selectedSimulations), max(simEndTimes));
<a name="l01487"></a>01487             trans = NaN(numel(selectedSimulations), max(simEndTimes));
<a name="l01488"></a>01488             procsII = NaN(numel(selectedSimulations), max(simEndTimes));
<a name="l01489"></a>01489             folds = NaN(numel(selectedSimulations), max(simEndTimes));
<a name="l01490"></a>01490             signals = NaN(numel(selectedSimulations), max(simEndTimes));
<a name="l01491"></a>01491             inacts = NaN(numel(selectedSimulations), max(simEndTimes));
<a name="l01492"></a>01492             misfolds = NaN(numel(selectedSimulations), max(simEndTimes));
<a name="l01493"></a>01493             damages = NaN(numel(selectedSimulations), max(simEndTimes));
<a name="l01494"></a>01494             aborts = NaN(numel(selectedSimulations), max(simEndTimes));
<a name="l01495"></a>01495             enzymes = NaN(numel(enzProcesses), 1, max(simEndTimes), numel(selectedSimulations));
<a name="l01496"></a>01496             <span class="keywordflow">for</span> i = 1:numel(selectedSimulations)
<a name="l01497"></a>01497                 states = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, &#39;extract&#39;, selectedSimulations(i));
<a name="l01498"></a>01498                 
<a name="l01499"></a>01499                 monCnts = full(sum(states.ProteinMonomer.counts, 2));
<a name="l01500"></a>01500                 nscnt(i, 1:simEndTimes(i)) = permute(sum(monCnts(pm.nascentIndexs, :, :, :), 1), [4 3 1 2]);
<a name="l01501"></a>01501                 procsI(i, 1:simEndTimes(i)) = full(permute(sum(states.ProteinMonomer.counts(...
<a name="l01502"></a>01502                     pm.processedIIndexs(pm.compartments(pm.processedIIIndexs) ~= comp.cytosolIndexs), ...
<a name="l01503"></a>01503                     comp.cytosolIndexs, :, :), 1), [4 3 1 2]));
<a name="l01504"></a>01504                 trans(i, 1:simEndTimes(i)) = permute(sum(monCnts(pm.processedIIndexs, :, :, :), 1), [4 3 1 2]) - procsI(i, 1:simEndTimes(i));
<a name="l01505"></a>01505                 procsII(i, 1:simEndTimes(i)) = permute(sum(monCnts(pm.processedIIIndexs, :, :, :), 1), [4 3 1 2]);
<a name="l01506"></a>01506                 folds(i, 1:simEndTimes(i)) = permute(sum(monCnts(pm.foldedIndexs, :, :, :), 1), [4 3 1 2]);
<a name="l01507"></a>01507                 signals(i, 1:simEndTimes(i)) = permute(sum(monCnts(pm.signalSequenceIndexs, :, :, :), 1), [4 3 1 2]);
<a name="l01508"></a>01508                 inacts(i, 1:simEndTimes(i)) = permute(sum(monCnts(pm.inactivatedIndexs, :, :, :), 1), [4 3 1 2]);
<a name="l01509"></a>01509                 misfolds(i, 1:simEndTimes(i)) = permute(sum(monCnts(pm.misfoldedIndexs, :, :, :), 1), [4 3 1 2]);
<a name="l01510"></a>01510                 damages(i, 1:simEndTimes(i)) = permute(sum(monCnts(pm.damagedIndexs, :, :, :), 1), [4 3 1 2]);
<a name="l01511"></a>01511                 aborts(i, 1:simEndTimes(i)) = permute(full(states.Polypeptide.abortedPolypeptides), [4 3 1 2]);
<a name="l01512"></a>01512                 
<a name="l01513"></a>01513                 monCnts = monCnts(pm.matureIndexs, :, :, :);
<a name="l01514"></a>01514                 cpxCnts = full(states.ProteinComplex.counts);
<a name="l01515"></a>01515                 
<a name="l01516"></a>01516                 tmpEnzymes = zeros(0, 1, size(monCnts, 3), size(monCnts, 4));
<a name="l01517"></a>01517                 
<a name="l01518"></a>01518                 tmp = zeros(numel(ppI.enzymes), 1, size(monCnts, 3), size(monCnts, 4));
<a name="l01519"></a>01519                 tmp(ppI.enzymeMonomerLocalIndexs, :, :, :) = monCnts(ppI.enzymeMonomerGlobalIndexs, :, :, :);
<a name="l01520"></a>01520                 tmp(ppI.enzymeComplexLocalIndexs, :, :, :) = cpxCnts(ppI.enzymeComplexGlobalIndexs, :, :, :);
<a name="l01521"></a>01521                 tmpEnzymes = [tmpEnzymes; tmp]; %<span class="preprocessor">#ok&lt;AGROW&gt;</span>
<a name="l01522"></a>01522 <span class="preprocessor"></span>                
<a name="l01523"></a>01523                 tmp = zeros(numel(pt.enzymes), 1, size(monCnts, 3), size(monCnts, 4));
<a name="l01524"></a>01524                 tmp(pt.enzymeMonomerLocalIndexs, :, :, :) = monCnts(pt.enzymeMonomerGlobalIndexs, :, :, :);
<a name="l01525"></a>01525                 tmp(pt.enzymeComplexLocalIndexs, :, :, :) = cpxCnts(pt.enzymeComplexGlobalIndexs, :, :, :);
<a name="l01526"></a>01526                 tmpEnzymes = [tmpEnzymes; tmp]; %#ok&lt;AGROW&gt;
<a name="l01527"></a>01527                 
<a name="l01528"></a>01528                 tmp = zeros(numel(ppII.enzymes), 1, size(monCnts, 3), size(monCnts, 4));
<a name="l01529"></a>01529                 tmp(ppII.enzymeMonomerLocalIndexs, :, :, :) = monCnts(ppII.enzymeMonomerGlobalIndexs, :, :, :);
<a name="l01530"></a>01530                 tmp(ppII.enzymeComplexLocalIndexs, :, :, :) = cpxCnts(ppII.enzymeComplexGlobalIndexs, :, :, :);
<a name="l01531"></a>01531                 tmpEnzymes = [tmpEnzymes; tmp]; %#ok&lt;AGROW&gt;
<a name="l01532"></a>01532                 
<a name="l01533"></a>01533                 tmp = zeros(numel(pf.enzymes), 1, size(monCnts, 3), size(monCnts, 4));
<a name="l01534"></a>01534                 tmp(pf.enzymeMonomerLocalIndexs, :, :, :) = monCnts(pf.enzymeMonomerGlobalIndexs, :, :, :);
<a name="l01535"></a>01535                 tmp(pf.enzymeComplexLocalIndexs, :, :, :) = cpxCnts(pf.enzymeComplexGlobalIndexs, :, :, :);
<a name="l01536"></a>01536                 tmpEnzymes = [tmpEnzymes; tmp]; %#ok&lt;AGROW&gt;
<a name="l01537"></a>01537                 
<a name="l01538"></a>01538                 tmp = zeros(numel(pmod.enzymes), 1, size(monCnts, 3), size(monCnts, 4));
<a name="l01539"></a>01539                 tmp(pmod.enzymeMonomerLocalIndexs, :, :, :) = monCnts(pmod.enzymeMonomerGlobalIndexs, :, :, :);
<a name="l01540"></a>01540                 tmp(pmod.enzymeComplexLocalIndexs, :, :, :) = cpxCnts(pmod.enzymeComplexGlobalIndexs, :, :, :);
<a name="l01541"></a>01541                 tmpEnzymes = [tmpEnzymes; tmp]; %#ok&lt;AGROW&gt;
<a name="l01542"></a>01542                 
<a name="l01543"></a>01543                 tmp = zeros(numel(pd.enzymes), 1, size(monCnts, 3), size(monCnts, 4));
<a name="l01544"></a>01544                 tmp(pd.enzymeMonomerLocalIndexs, :, :, :) = monCnts(pd.enzymeMonomerGlobalIndexs, :, :, :);
<a name="l01545"></a>01545                 tmp(pd.enzymeComplexLocalIndexs, :, :, :) = cpxCnts(pd.enzymeComplexGlobalIndexs, :, :, :);
<a name="l01546"></a>01546                 tmpEnzymes = [tmpEnzymes; tmp]; %#ok&lt;AGROW&gt;
<a name="l01547"></a>01547                 
<a name="l01548"></a>01548                 enzymes(:, :, 1:simEndTimes(i), i) = tmpEnzymes;
<a name="l01549"></a>01549                 
<a name="l01550"></a>01550                 clear states monCnts cpxCnts tmpEnzymes tmp;
<a name="l01551"></a>01551             end
<a name="l01552"></a>01552             
<a name="l01553"></a>01553             %% clear figure
<a name="l01554"></a>01554             clf(figHandle);
<a name="l01555"></a>01555             
<a name="l01556"></a>01556             %% left panel
<a name="l01557"></a>01557             
<a name="l01558"></a>01558             %layout plots
<a name="l01559"></a>01559             axesHandlesL = PlotUtil.multiElementPlot(figHandle, 3.5 * ones(10, 1), [0 time(end)], <span class="keyword">struct</span>(...
<a name="l01560"></a>01560                 <span class="stringliteral">&#39;titleStr&#39;</span>, <span class="stringliteral">&#39;Monomers&#39;</span>, ...
<a name="l01561"></a>01561                 <span class="stringliteral">&#39;position&#39;</span>, [0.12 0.10 0.21 0.8], ...
<a name="l01562"></a>01562                 <span class="stringliteral">&#39;colorOrder&#39;</span>, PlotUtil.getRedGreenColorOrder(simOrder)));
<a name="l01563"></a>01563             
<a name="l01564"></a>01564             %plot data
<a name="l01565"></a>01565             PlotUtil.plotLine(axesHandlesL(1), time(2:end), nscnt, <span class="keyword">false</span>, <span class="keyword">true</span>, <span class="keyword">false</span>);
<a name="l01566"></a>01566             ylabel(axesHandlesL(1), <span class="stringliteral">&#39;Nascent&#39;</span>);
<a name="l01567"></a>01567             
<a name="l01568"></a>01568             PlotUtil.plotLine(axesHandlesL(2), time(2:end), procsI, <span class="keyword">false</span>, <span class="keyword">true</span>, <span class="keyword">false</span>);
<a name="l01569"></a>01569             ylabel(axesHandlesL(2), <span class="stringliteral">&#39;Processed-I&#39;</span>);
<a name="l01570"></a>01570             
<a name="l01571"></a>01571             PlotUtil.plotLine(axesHandlesL(3), time(2:end), trans, <span class="keyword">false</span>, <span class="keyword">true</span>, <span class="keyword">false</span>);
<a name="l01572"></a>01572             ylabel(axesHandlesL(3), <span class="stringliteral">&#39;Translocated&#39;</span>);
<a name="l01573"></a>01573             
<a name="l01574"></a>01574             PlotUtil.plotLine(axesHandlesL(4), time(2:end), procsII, <span class="keyword">false</span>, <span class="keyword">true</span>, <span class="keyword">false</span>);
<a name="l01575"></a>01575             ylabel(axesHandlesL(4), <span class="stringliteral">&#39;Processed-II&#39;</span>);
<a name="l01576"></a>01576             
<a name="l01577"></a>01577             PlotUtil.plotLine(axesHandlesL(5), time(2:end), folds, <span class="keyword">false</span>, <span class="keyword">true</span>, <span class="keyword">false</span>);
<a name="l01578"></a>01578             ylabel(axesHandlesL(5), <span class="stringliteral">&#39;Folded&#39;</span>);
<a name="l01579"></a>01579             
<a name="l01580"></a>01580             PlotUtil.plotLine(axesHandlesL(6), time(2:end), signals, <span class="keyword">false</span>, <span class="keyword">true</span>, <span class="keyword">false</span>);
<a name="l01581"></a>01581             ylabel(axesHandlesL(6), {<span class="stringliteral">&#39;Signal&#39;</span> <span class="stringliteral">&#39;Sequences&#39;</span>});
<a name="l01582"></a>01582             
<a name="l01583"></a>01583             PlotUtil.plotLine(axesHandlesL(7), time(2:end), inacts, <span class="keyword">false</span>, <span class="keyword">true</span>, <span class="keyword">false</span>);
<a name="l01584"></a>01584             ylabel(axesHandlesL(7), <span class="stringliteral">&#39;Inactivated&#39;</span>);
<a name="l01585"></a>01585             
<a name="l01586"></a>01586             PlotUtil.plotLine(axesHandlesL(8), time(2:end), misfolds, <span class="keyword">false</span>, <span class="keyword">true</span>, <span class="keyword">false</span>);
<a name="l01587"></a>01587             ylabel(axesHandlesL(8), <span class="stringliteral">&#39;Misfolded&#39;</span>);
<a name="l01588"></a>01588             
<a name="l01589"></a>01589             PlotUtil.plotLine(axesHandlesL(9), time(2:end), damages, <span class="keyword">false</span>, <span class="keyword">true</span>, <span class="keyword">false</span>);
<a name="l01590"></a>01590             ylabel(axesHandlesL(9), <span class="stringliteral">&#39;Damaged&#39;</span>);
<a name="l01591"></a>01591             
<a name="l01592"></a>01592             PlotUtil.plotLine(axesHandlesL(10), time(2:end), aborts, <span class="keyword">false</span>, <span class="keyword">true</span>, <span class="keyword">false</span>);
<a name="l01593"></a>01593             ylabel(axesHandlesL(10), <span class="stringliteral">&#39;Aborted&#39;</span>);
<a name="l01594"></a>01594             
<a name="l01595"></a>01595             % Format Y-axis
<a name="l01596"></a>01596             PlotUtil.alignYAxesLabels(axesHandlesL);
<a name="l01597"></a>01597             PlotUtil.offsetYAxes(axesHandlesL, 0.04);
<a name="l01598"></a>01598             PlotUtil.labelSubplots(axesHandlesL(1), <span class="charliteral">&#39;A&#39;</span>, -0.42, 1.2);
<a name="l01599"></a>01599             
<a name="l01600"></a>01600             %% right panels
<a name="l01601"></a>01601             
<a name="l01602"></a>01602             nPlots1 = ceil(size(enzymes, 1) / 2);
<a name="l01603"></a>01603             nPlots2 = floor(size(enzymes, 1) / 2);
<a name="l01604"></a>01604             
<a name="l01605"></a>01605             %layout plots
<a name="l01606"></a>01606             axesHandlesM = PlotUtil.multiElementPlot(figHandle, 3.5 * ones(nPlots1, 1), [0 time(end)], <span class="keyword">struct</span>(...
<a name="l01607"></a>01607                 <span class="stringliteral">&#39;titleStr&#39;</span>, <span class="stringliteral">&#39;Enzymes&#39;</span>, ...
<a name="l01608"></a>01608                 <span class="stringliteral">&#39;position&#39;</span>, [0.45 0.10 0.21 0.8], ...
<a name="l01609"></a>01609                 <span class="stringliteral">&#39;colorOrder&#39;</span>, PlotUtil.getRedGreenColorOrder(simOrder)));
<a name="l01610"></a>01610             axesHandlesR = PlotUtil.multiElementPlot(figHandle, 3.5 * ones(nPlots2, 1), [0 time(end)], <span class="keyword">struct</span>(...
<a name="l01611"></a>01611                 <span class="stringliteral">&#39;position&#39;</span>, [0.78 0.10 0.21 0.8], ...
<a name="l01612"></a>01612                 <span class="stringliteral">&#39;colorOrder&#39;</span>, PlotUtil.getRedGreenColorOrder(simOrder)));
<a name="l01613"></a>01613             
<a name="l01614"></a>01614             %plot data
<a name="l01615"></a>01615             <span class="keywordflow">for</span> i = 1:nPlots1
<a name="l01616"></a>01616                 PlotUtil.plotLine(axesHandlesM(i), time(2:end), permute(enzymes(i, :, :, :), [4 3 1 2]), <span class="keyword">false</span>, <span class="keyword">true</span>, <span class="keyword">false</span>);
<a name="l01617"></a>01617                 ylabel(axesHandlesM(i), [enzProcesses(i) num2str(enzProcessIdxs(i))]);
<a name="l01618"></a>01618             end
<a name="l01619"></a>01619             <span class="keywordflow">for</span> i = 1:nPlots2
<a name="l01620"></a>01620                 PlotUtil.plotLine(axesHandlesR(i), time(2:end), permute(enzymes(nPlots1+i, :, :, :), [4 3 1 2]), <span class="keyword">false</span>, <span class="keyword">true</span>, <span class="keyword">false</span>);
<a name="l01621"></a>01621                 ylabel(axesHandlesR(i), [enzProcesses(nPlots1+i) num2str(enzProcessIdxs(nPlots1+i))]);
<a name="l01622"></a>01622             end
<a name="l01623"></a>01623             
<a name="l01624"></a>01624             % Format Y-axis
<a name="l01625"></a>01625             PlotUtil.alignYAxesLabels(axesHandlesM);
<a name="l01626"></a>01626             PlotUtil.alignYAxesLabels(axesHandlesR);
<a name="l01627"></a>01627             PlotUtil.offsetYAxes(axesHandlesM, 0.04);
<a name="l01628"></a>01628             PlotUtil.offsetYAxes(axesHandlesR, 0.04);
<a name="l01629"></a>01629             PlotUtil.labelSubplots(axesHandlesM(1), <span class="charliteral">&#39;B&#39;</span>, -0.42, 1.2);
<a name="l01630"></a>01630             
<a name="l01631"></a>01631             %% organize figure data
<a name="l01632"></a>01632             figData = <span class="keyword">struct</span>;
<a name="l01633"></a>01633             figData.simOrder = simOrder;
<a name="l01634"></a>01634             figData.time = time;
<a name="l01635"></a>01635             figData.nscnt = nscnt;
<a name="l01636"></a>01636             figData.procsI = procsI;
<a name="l01637"></a>01637             figData.trans = trans;
<a name="l01638"></a>01638             figData.procsII = procsII;
<a name="l01639"></a>01639             figData.folds = folds;
<a name="l01640"></a>01640             figData.signals = signals;
<a name="l01641"></a>01641             figData.inacts = inacts;
<a name="l01642"></a>01642             figData.misfolds = misfolds;
<a name="l01643"></a>01643             figData.damages = damages;
<a name="l01644"></a>01644             figData.aborts = aborts;
<a name="l01645"></a>01645             figData.enzymes = enzymes;
<a name="l01646"></a>01646             figData.enzProcesses = enzProcesses;
<a name="l01647"></a>01647         end
<a name="l01648"></a>01648         
<a name="l01649"></a>01649         <span class="keyword">function</span> figData = rnaSynthesis(figHandle, simBatchDir, selectedSimulations)
<a name="l01650"></a>01650             import edu.stanford.covert.cell.sim.util.PlotUtil;
<a name="l01651"></a>01651             import edu.stanford.covert.cell.sim.util.SimulationEnsemble;
<a name="l01652"></a>01652             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
<a name="l01653"></a>01653             import edu.stanford.covert.util.<a class="code" href="class_constant_util.html">ConstantUtil</a>;
<a name="l01654"></a>01654             
<a name="l01655"></a>01655             if nargin &lt; 2 || isempty(simBatchDir)
<a name="l01656"></a>01656                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
<a name="l01657"></a>01657             end
<a name="l01658"></a>01658             if nargin &lt; 3 || isempty(selectedSimulations)
<a name="l01659"></a>01659                 selectedSimulations = SimulationDiskUtil.getCompleteSimulations(simBatchDir);
<a name="l01660"></a>01660             end
<a name="l01661"></a>01661             
<a name="l01662"></a>01662             [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep &#39;1&#39;]);
<a name="l01663"></a>01663             g = sim.gene;
<a name="l01664"></a>01664             
<a name="l01665"></a>01665             t = sim.state(&#39;Time&#39;);
<a name="l01666"></a>01666             mass = sim.state(&#39;Mass&#39;);
<a name="l01667"></a>01667             rna = sim.state(&#39;Rna&#39;);
<a name="l01668"></a>01668             pm = sim.state(&#39;ProteinMonomer&#39;);
<a name="l01669"></a>01669             pc = sim.state(&#39;ProteinComplex&#39;);
<a name="l01670"></a>01670             
<a name="l01671"></a>01671             rp = sim.process(&#39;RNAProcessing&#39;);
<a name="l01672"></a>01672             rmod = sim.process(&#39;RNAModification&#39;);
<a name="l01673"></a>01673             rd = sim.process(&#39;RNADecay&#39;);
<a name="l01674"></a>01674             
<a name="l01675"></a>01675             monExp = (rna.matureRNAGeneComposition(g.mRNAIndexs, :) * rna.expression(rna.matureIndexs)) ./ ...
<a name="l01676"></a>01676                 (log(2) / t.cellCycleLength + pm.decayRates(pm.matureIndexs));
<a name="l01677"></a>01677             expTotMons = mass.initialFractionAAsInMonomers * mass.dryWeightFractionProtein * mass.cellInitialDryWeight * ...
<a name="l01678"></a>01678                 monExp / (monExp&#39; * pm.molecularWeights(pm.matureIndexs) / ConstantUtil.nAvogadro);
<a name="l01679"></a>01679             expTotCpxs = min(repmat(expTotMons, [1 numel(pc.matureIndexs)]) ./ sum(pc.proteinComplexComposition(g.mRNAIndexs, :, :), 3), [], 1);
<a name="l01680"></a>01680             
<a name="l01681"></a>01681             enzProcesses = [
<a name="l01682"></a>01682                 repmat({<span class="stringliteral">&#39;RP&#39;</span>}, size(rp.enzymeWholeCellModelIDs))
<a name="l01683"></a>01683                 repmat({<span class="stringliteral">&#39;RMod&#39;</span>}, size(rmod.enzymeWholeCellModelIDs))
<a name="l01684"></a>01684                 repmat({<span class="stringliteral">&#39;RD&#39;</span>}, size(rd.enzymeWholeCellModelIDs))
<a name="l01685"></a>01685                 ];
<a name="l01686"></a>01686             enzProcessIdxs = [
<a name="l01687"></a>01687                 (1:size(rp.enzymeWholeCellModelIDs))&#39;
<a name="l01688"></a>01688                 (1:size(rmod.enzymeWholeCellModelIDs))&#39;
<a name="l01689"></a>01689                 (1:size(rd.enzymeWholeCellModelIDs))&#39;
<a name="l01690"></a>01690                 ];
<a name="l01691"></a>01691             
<a name="l01692"></a>01692             %% <span class="keyword">get</span> data
<a name="l01693"></a>01693             ensemble = SimulationEnsemble(simBatchDir, {<span class="stringliteral">&#39;mass&#39;</span> <span class="stringliteral">&#39;mass&#39;</span>; <span class="stringliteral">&#39;immatureRnas&#39;</span> <span class="stringliteral">&#39;immatureRnas&#39;</span>}, [], selectedSimulations);
<a name="l01694"></a>01694             time = ensemble.stateData.time / 3600;
<a name="l01695"></a>01695             mass = ensemble.stateData.values(ensemble.getPropertyIndices(<span class="stringliteral">&#39;mass&#39;</span>), :, :, :) * sim.state(<span class="stringliteral">&#39;Mass&#39;</span>).cellInitialDryWeight / (1-sim.state(&#39;Mass&#39;).fractionWetWeight) * 1e15;
<a name="l01696"></a>01696             simEndTimes = ensemble.stateData.simulationEndTimes();
<a name="l01697"></a>01697             
<a name="l01698"></a>01698             [~, simOrder] = sort(mass(:, 1, min(simEndTimes), :));
<a name="l01699"></a>01699             simOrder = simOrder(:);
<a name="l01700"></a>01700             
<a name="l01701"></a>01701             clear ensemble;
<a name="l01702"></a>01702             
<a name="l01703"></a>01703             stateNames = {
<a name="l01704"></a>01704                 <span class="stringliteral">&#39;Rna&#39;</span>            <span class="stringliteral">&#39;counts&#39;</span>  <span class="charliteral">&#39;:&#39;</span>               <span class="stringliteral">&#39;-sum&#39;</span>
<a name="l01705"></a>01705                 <span class="stringliteral">&#39;Transcript&#39;</span>     <span class="stringliteral">&#39;abortedTranscripts&#39;</span> <span class="stringliteral">&#39;-nnz&#39;</span>  1
<a name="l01706"></a>01706                 <span class="stringliteral">&#39;ProteinMonomer&#39;</span> <span class="stringliteral">&#39;counts&#39;</span>  pm.matureIndexs  <span class="stringliteral">&#39;-sum&#39;</span>
<a name="l01707"></a>01707                 <span class="stringliteral">&#39;ProteinComplex&#39;</span> <span class="stringliteral">&#39;counts&#39;</span>  pc.matureIndexs  <span class="stringliteral">&#39;-sum&#39;</span>
<a name="l01708"></a>01708                 };
<a name="l01709"></a>01709             nscnt = NaN(numel(selectedSimulations), max(simEndTimes));
<a name="l01710"></a>01710             procs = NaN(numel(selectedSimulations), max(simEndTimes));
<a name="l01711"></a>01711             misfolds = NaN(numel(selectedSimulations), max(simEndTimes));
<a name="l01712"></a>01712             damages = NaN(numel(selectedSimulations), max(simEndTimes));
<a name="l01713"></a>01713             aborts = NaN(numel(selectedSimulations), max(simEndTimes));
<a name="l01714"></a>01714             enzymes = NaN(numel(enzProcessIdxs), 1, max(simEndTimes), numel(selectedSimulations));
<a name="l01715"></a>01715             <span class="keywordflow">for</span> i = 1:numel(selectedSimulations)
<a name="l01716"></a>01716                 states = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, &#39;extract&#39;, selectedSimulations(i));
<a name="l01717"></a>01717                 
<a name="l01718"></a>01718                 rnaCnts = full(states.Rna.counts);
<a name="l01719"></a>01719                 monCnts = full(states.ProteinMonomer.counts);
<a name="l01720"></a>01720                 cpxCnts = full(states.ProteinComplex.counts);
<a name="l01721"></a>01721                 nscnt(i, 1:simEndTimes(i)) = permute(sum(rnaCnts(rna.nascentIndexs, :, :, :), 1), [4 3 1 2]);
<a name="l01722"></a>01722                 procs(i, 1:simEndTimes(i)) = permute(sum(rnaCnts(rna.processedIndexs, :, :, :), 1), [4 3 1 2]);
<a name="l01723"></a>01723                 misfolds (i, 1:simEndTimes(i))= permute(sum(rnaCnts(rna.misfoldedIndexs, :, :, :), 1), [4 3 1 2]);
<a name="l01724"></a>01724                 damages(i, 1:simEndTimes(i)) = permute(sum(rnaCnts(rna.damagedIndexs, :, :, :), 1), [4 3 1 2]);
<a name="l01725"></a>01725                 aborts(i, 1:simEndTimes(i)) = permute(full(states.Transcript.abortedTranscripts), [4 3 1 2]);
<a name="l01726"></a>01726                 
<a name="l01727"></a>01727                 tmpEnzymes = zeros(0, 1, size(monCnts, 3), size(monCnts, 4));
<a name="l01728"></a>01728                 
<a name="l01729"></a>01729                 tmp = zeros(numel(rp.enzymes), 1, size(monCnts, 3), size(monCnts, 4));
<a name="l01730"></a>01730                 tmp(rp.enzymeRNALocalIndexs, :, :, :) = rnaCnts(numel(rna.nascentIndexs)+numel(rna.processedIndexs) + rp.enzymeRNAGlobalIndexs, :, :, :);
<a name="l01731"></a>01731                 tmp(rp.enzymeMonomerLocalIndexs, :, :, :) = monCnts(rp.enzymeMonomerGlobalIndexs, :, :, :);
<a name="l01732"></a>01732                 tmp(rp.enzymeComplexLocalIndexs, :, :, :) = cpxCnts(rp.enzymeComplexGlobalIndexs, :, :, :);
<a name="l01733"></a>01733                 tmpEnzymes = [tmpEnzymes; tmp]; %<span class="preprocessor">#ok&lt;AGROW&gt;</span>
<a name="l01734"></a>01734 <span class="preprocessor"></span>                
<a name="l01735"></a>01735                 tmp = zeros(numel(rmod.enzymes), 1, size(monCnts, 3), size(monCnts, 4));
<a name="l01736"></a>01736                 tmp(rmod.enzymeRNALocalIndexs, :, :, :) = rnaCnts(numel(rna.nascentIndexs)+numel(rna.processedIndexs) + rmod.enzymeRNAGlobalIndexs, :, :, :);
<a name="l01737"></a>01737                 tmp(rmod.enzymeMonomerLocalIndexs, :, :, :) = monCnts(rmod.enzymeMonomerGlobalIndexs, :, :, :);
<a name="l01738"></a>01738                 tmp(rmod.enzymeComplexLocalIndexs, :, :, :) = cpxCnts(rmod.enzymeComplexGlobalIndexs, :, :, :);
<a name="l01739"></a>01739                 tmpEnzymes = [tmpEnzymes; tmp]; %#ok&lt;AGROW&gt;
<a name="l01740"></a>01740                 
<a name="l01741"></a>01741                 tmp = zeros(numel(rd.enzymes), 1, size(monCnts, 3), size(monCnts, 4));
<a name="l01742"></a>01742                 tmp(rd.enzymeRNALocalIndexs, :, :, :) = rnaCnts(numel(rna.nascentIndexs)+numel(rna.processedIndexs) + rd.enzymeRNAGlobalIndexs, :, :, :);
<a name="l01743"></a>01743                 tmp(rd.enzymeMonomerLocalIndexs, :, :, :) = monCnts(rd.enzymeMonomerGlobalIndexs, :, :, :);
<a name="l01744"></a>01744                 tmp(rd.enzymeComplexLocalIndexs, :, :, :) = cpxCnts(rd.enzymeComplexGlobalIndexs, :, :, :);
<a name="l01745"></a>01745                 tmpEnzymes = [tmpEnzymes; tmp]; %#ok&lt;AGROW&gt;
<a name="l01746"></a>01746                 
<a name="l01747"></a>01747                 enzymes(:, :, 1:simEndTimes(i), i) = tmpEnzymes;
<a name="l01748"></a>01748                 
<a name="l01749"></a>01749                 clear states rnaCnts monCnts cpxCnts tmpEnzymes tmp;
<a name="l01750"></a>01750             end
<a name="l01751"></a>01751             
<a name="l01752"></a>01752             expEnzymes = zeros(0, 1);
<a name="l01753"></a>01753             
<a name="l01754"></a>01754             tmp2 = zeros(numel(rp.enzymes), 1);
<a name="l01755"></a>01755             tmp2(rp.enzymeMonomerLocalIndexs) = expTotMons(rp.enzymeMonomerGlobalIndexs);
<a name="l01756"></a>01756             tmp2(rp.enzymeComplexLocalIndexs) = expTotCpxs(rp.enzymeComplexGlobalIndexs);
<a name="l01757"></a>01757             expEnzymes = [expEnzymes; tmp2];
<a name="l01758"></a>01758             
<a name="l01759"></a>01759             tmp2 = zeros(numel(rmod.enzymes), 1);
<a name="l01760"></a>01760             tmp2(rmod.enzymeMonomerLocalIndexs) = expTotMons(rmod.enzymeMonomerGlobalIndexs);
<a name="l01761"></a>01761             tmp2(rmod.enzymeComplexLocalIndexs) = expTotCpxs(rmod.enzymeComplexGlobalIndexs);
<a name="l01762"></a>01762             expEnzymes = [expEnzymes; tmp2];
<a name="l01763"></a>01763             
<a name="l01764"></a>01764             tmp2 = zeros(numel(rd.enzymes), 1);
<a name="l01765"></a>01765             tmp2(rd.enzymeMonomerLocalIndexs) = expTotMons(rd.enzymeMonomerGlobalIndexs);
<a name="l01766"></a>01766             tmp2(rd.enzymeComplexLocalIndexs) = expTotCpxs(rd.enzymeComplexGlobalIndexs);
<a name="l01767"></a>01767             expEnzymes = [expEnzymes; tmp2];
<a name="l01768"></a>01768             
<a name="l01769"></a>01769             expEnzymes = repmat(expEnzymes, [1 1 numel(time)-1 1]) .* exp(log(2) * repmat(permute(time(2:end),[1 3 2]), size(expEnzymes, 1), 1) / (t.cellCycleLength/3600));
<a name="l01770"></a>01770             
<a name="l01771"></a>01771             %% clear figure
<a name="l01772"></a>01772             clf(figHandle);
<a name="l01773"></a>01773             C = 70;
<a name="l01774"></a>01774             
<a name="l01775"></a>01775             %% left panel
<a name="l01776"></a>01776             
<a name="l01777"></a>01777             %layout plots
<a name="l01778"></a>01778             h = (C - 2 + 1) / 2;
<a name="l01779"></a>01779             axesHandlesL = PlotUtil.multiElementPlot(figHandle, h* ones(5, 1), [0 time(end)], <span class="keyword">struct</span>(...
<a name="l01780"></a>01780                 <span class="stringliteral">&#39;titleStr&#39;</span>, <span class="stringliteral">&#39;RNAs&#39;</span>, ...
<a name="l01781"></a>01781                 <span class="stringliteral">&#39;position&#39;</span>, [0.12 0.10 0.21 0.8], ...
<a name="l01782"></a>01782                 <span class="stringliteral">&#39;colorOrder&#39;</span>, PlotUtil.getRedGreenColorOrder(simOrder)));
<a name="l01783"></a>01783             
<a name="l01784"></a>01784             %plot data
<a name="l01785"></a>01785             PlotUtil.plotLine(axesHandlesL(1), time(2:end), nscnt, <span class="keyword">false</span>, <span class="keyword">true</span>, <span class="keyword">false</span>);
<a name="l01786"></a>01786             ylabel(axesHandlesL(1), <span class="stringliteral">&#39;Nascent&#39;</span>);
<a name="l01787"></a>01787             
<a name="l01788"></a>01788             PlotUtil.plotLine(axesHandlesL(2), time(2:end), procs, <span class="keyword">false</span>, <span class="keyword">true</span>, <span class="keyword">false</span>);
<a name="l01789"></a>01789             ylabel(axesHandlesL(2), <span class="stringliteral">&#39;Processed&#39;</span>);
<a name="l01790"></a>01790             
<a name="l01791"></a>01791             PlotUtil.plotLine(axesHandlesL(3), time(2:end), misfolds, <span class="keyword">false</span>, <span class="keyword">true</span>, <span class="keyword">false</span>);
<a name="l01792"></a>01792             ylabel(axesHandlesL(3), <span class="stringliteral">&#39;Misfolded&#39;</span>);
<a name="l01793"></a>01793             
<a name="l01794"></a>01794             PlotUtil.plotLine(axesHandlesL(4), time(2:end), damages, <span class="keyword">false</span>, <span class="keyword">true</span>, <span class="keyword">false</span>);
<a name="l01795"></a>01795             ylabel(axesHandlesL(4), <span class="stringliteral">&#39;Damaged&#39;</span>);
<a name="l01796"></a>01796             
<a name="l01797"></a>01797             PlotUtil.plotLine(axesHandlesL(5), time(2:end), aborts, <span class="keyword">false</span>, <span class="keyword">true</span>, <span class="keyword">false</span>);
<a name="l01798"></a>01798             ylabel(axesHandlesL(5), <span class="stringliteral">&#39;Aborted&#39;</span>);
<a name="l01799"></a>01799             
<a name="l01800"></a>01800             % Format Y-axis
<a name="l01801"></a>01801             PlotUtil.alignYAxesLabels(axesHandlesL);
<a name="l01802"></a>01802             PlotUtil.offsetYAxes(axesHandlesL, 0.04);
<a name="l01803"></a>01803             PlotUtil.labelSubplots(axesHandlesL(1), <span class="charliteral">&#39;A&#39;</span>, -0.42, 1.2);
<a name="l01804"></a>01804             
<a name="l01805"></a>01805             %% right panels
<a name="l01806"></a>01806             
<a name="l01807"></a>01807             nPlots1 = ceil(size(enzymes, 1) / 2);
<a name="l01808"></a>01808             nPlots2 = floor(size(enzymes, 1) / 2);
<a name="l01809"></a>01809             
<a name="l01810"></a>01810             %layout plots
<a name="l01811"></a>01811             colorOrder = [PlotUtil.getRedGreenColorOrder(simOrder); 0.5 0.5 0.5];
<a name="l01812"></a>01812             h = (C - ceil(size(enzymes, 1)/2) + 1) / ceil(size(enzymes, 1) / 2);
<a name="l01813"></a>01813             axesHandlesM = PlotUtil.multiElementPlot(figHandle, h * ones(nPlots1, 1), [0 time(end)], <span class="keyword">struct</span>(...
<a name="l01814"></a>01814                 <span class="stringliteral">&#39;titleStr&#39;</span>, <span class="stringliteral">&#39;Enzymes&#39;</span>, ...
<a name="l01815"></a>01815                 <span class="stringliteral">&#39;position&#39;</span>, [0.45 0.10 0.21 0.8], ...
<a name="l01816"></a>01816                 <span class="stringliteral">&#39;colorOrder&#39;</span>, colorOrder));
<a name="l01817"></a>01817             axesHandlesR = PlotUtil.multiElementPlot(figHandle, h * ones(nPlots2, 1), [0 time(end)], <span class="keyword">struct</span>(...
<a name="l01818"></a>01818                 <span class="stringliteral">&#39;position&#39;</span>, [0.78 0.10 0.21 0.8], ...
<a name="l01819"></a>01819                 <span class="stringliteral">&#39;colorOrder&#39;</span>, colorOrder));
<a name="l01820"></a>01820             
<a name="l01821"></a>01821             %plot data
<a name="l01822"></a>01822             <span class="keywordflow">for</span> i = 1:nPlots1
<a name="l01823"></a>01823                 PlotUtil.plotLine(axesHandlesM(i), time(2:end), permute(cat(4, enzymes(i, :, :, :), expEnzymes(i, :, :, :)), [4 3 1 2]), <span class="keyword">false</span>, <span class="keyword">true</span>, <span class="keyword">false</span>);
<a name="l01824"></a>01824                 ylabel(axesHandlesM(i), [enzProcesses(i) num2str(enzProcessIdxs(i))]);
<a name="l01825"></a>01825             end
<a name="l01826"></a>01826             <span class="keywordflow">for</span> i = 1:nPlots2
<a name="l01827"></a>01827                 PlotUtil.plotLine(axesHandlesR(i), time(2:end), permute(cat(4, enzymes(nPlots1+i, :, :, :), expEnzymes(nPlots1+i, :, :, :)), [4 3 1 2]), <span class="keyword">false</span>, <span class="keyword">true</span>, <span class="keyword">false</span>);
<a name="l01828"></a>01828                 ylabel(axesHandlesR(i), [enzProcesses(nPlots1+i) num2str(enzProcessIdxs(nPlots1+i))]);
<a name="l01829"></a>01829             end
<a name="l01830"></a>01830             
<a name="l01831"></a>01831             % Format Y-axis
<a name="l01832"></a>01832             PlotUtil.alignYAxesLabels(axesHandlesM);
<a name="l01833"></a>01833             PlotUtil.alignYAxesLabels(axesHandlesR);
<a name="l01834"></a>01834             PlotUtil.offsetYAxes(axesHandlesM, 0.04);
<a name="l01835"></a>01835             PlotUtil.offsetYAxes(axesHandlesR, 0.04);
<a name="l01836"></a>01836             PlotUtil.labelSubplots(axesHandlesM(1), <span class="charliteral">&#39;B&#39;</span>, -0.42, 1.2);
<a name="l01837"></a>01837             
<a name="l01838"></a>01838             %% organize figure data
<a name="l01839"></a>01839             figData = <span class="keyword">struct</span>;
<a name="l01840"></a>01840             figData.simOrder = simOrder;
<a name="l01841"></a>01841             figData.time = time;
<a name="l01842"></a>01842             figData.nscnt = nscnt;
<a name="l01843"></a>01843             figData.procs = procs;
<a name="l01844"></a>01844             figData.misfolds = misfolds;
<a name="l01845"></a>01845             figData.damages = damages;
<a name="l01846"></a>01846             figData.aborts = aborts;
<a name="l01847"></a>01847             figData.enzymes = enzymes;
<a name="l01848"></a>01848             figData.expEnzymes = expEnzymes;
<a name="l01849"></a>01849             figData.enzProcessIdxs = enzProcessIdxs;
<a name="l01850"></a>01850         end
<a name="l01851"></a>01851         
<a name="l01852"></a>01852         <span class="keyword">function</span> figData = cellShape(figHandle, simBatchDir, selectedSimulations)
<a name="l01853"></a>01853             import edu.stanford.covert.cell.sim.util.PlotUtil;
<a name="l01854"></a>01854             import edu.stanford.covert.cell.sim.util.SimulationEnsemble;
<a name="l01855"></a>01855             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
<a name="l01856"></a>01856             import edu.stanford.covert.util.<a class="code" href="class_constant_util.html">ConstantUtil</a>;
<a name="l01857"></a>01857             
<a name="l01858"></a>01858             if nargin &lt; 2 || isempty(simBatchDir)
<a name="l01859"></a>01859                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
<a name="l01860"></a>01860             end
<a name="l01861"></a>01861             if nargin &lt; 3 || isempty(selectedSimulations)
<a name="l01862"></a>01862                 selectedSimulations = SimulationDiskUtil.getCompleteSimulations(simBatchDir);
<a name="l01863"></a>01863             end
<a name="l01864"></a>01864             
<a name="l01865"></a>01865             %% get data
<a name="l01866"></a>01866             stateNames = {
<a name="l01867"></a>01867                 <span class="stringliteral">&#39;Time&#39;</span>      <span class="stringliteral">&#39;values&#39;</span>             <span class="charliteral">&#39;:&#39;</span> <span class="charliteral">&#39;:&#39;</span>
<a name="l01868"></a>01868                 <span class="stringliteral">&#39;Mass&#39;</span>      <span class="stringliteral">&#39;cell&#39;</span>               <span class="charliteral">&#39;:&#39;</span> <span class="stringliteral">&#39;-sum&#39;</span>
<a name="l01869"></a>01869                 <span class="stringliteral">&#39;Geometry&#39;</span>  <span class="stringliteral">&#39;volume&#39;</span>             <span class="charliteral">&#39;:&#39;</span> <span class="charliteral">&#39;:&#39;</span>
<a name="l01870"></a>01870                 <span class="stringliteral">&#39;Geometry&#39;</span>  <span class="stringliteral">&#39;width&#39;</span>              <span class="charliteral">&#39;:&#39;</span> <span class="charliteral">&#39;:&#39;</span>
<a name="l01871"></a>01871                 <span class="stringliteral">&#39;Geometry&#39;</span>  <span class="stringliteral">&#39;totalLength&#39;</span>        <span class="charliteral">&#39;:&#39;</span> <span class="charliteral">&#39;:&#39;</span>
<a name="l01872"></a>01872                 <span class="stringliteral">&#39;Geometry&#39;</span>  <span class="stringliteral">&#39;surfaceArea&#39;</span>        <span class="charliteral">&#39;:&#39;</span> <span class="charliteral">&#39;:&#39;</span>
<a name="l01873"></a>01873                 <span class="stringliteral">&#39;Geometry&#39;</span>  <span class="stringliteral">&#39;pinchedDiameter&#39;</span>    <span class="charliteral">&#39;:&#39;</span> <span class="charliteral">&#39;:&#39;</span>
<a name="l01874"></a>01874                 };
<a name="l01875"></a>01875             states = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, <span class="stringliteral">&#39;extract&#39;</span>, selectedSimulations);
<a name="l01876"></a>01876             simEndTimes = permute(max(states.Time.values, [], 3), [4 3 1 2]);
<a name="l01877"></a>01877             
<a name="l01878"></a>01878             states.Geometry.width(states.Geometry.width == -1) = NaN;
<a name="l01879"></a>01879             states.Geometry.totalLength(states.Geometry.totalLength == -1) = NaN;
<a name="l01880"></a>01880             states.Geometry.surfaceArea(states.Geometry.surfaceArea == -1) = NaN;
<a name="l01881"></a>01881             
<a name="l01882"></a>01882             time = permute(max(states.Time.values, [], 4), [3 4 1 2]) / 3600;
<a name="l01883"></a>01883             mass = permute(states.Mass.cell, [3 4 1 2]) * 1e15;
<a name="l01884"></a>01884             volume = permute(states.Geometry.volume, [3 4 1 2]) * 1e15;
<a name="l01885"></a>01885             width = permute(states.Geometry.width, [3 4 1 2]) * 1e6;
<a name="l01886"></a>01886             totalLength = permute(states.Geometry.totalLength, [3 4 1 2]) * 1e6;
<a name="l01887"></a>01887             surfaceArea = permute(states.Geometry.surfaceArea, [3 4 1 2]) * 1e12;
<a name="l01888"></a>01888             pinchedDiameter = permute(states.Geometry.pinchedDiameter, [3 4 1 2]) * 1e6;
<a name="l01889"></a>01889             
<a name="l01890"></a>01890             minSimEndTime = min(max(states.Time.values, [], 4));
<a name="l01891"></a>01891             [~, simOrder] = sort(mass(minSimEndTime, :));
<a name="l01892"></a>01892             
<a name="l01893"></a>01893             <span class="keywordflow">for</span> i = 1:numel(simEndTimes)
<a name="l01894"></a>01894                 mass(simEndTimes(i)+1:end, i) = NaN;
<a name="l01895"></a>01895                 volume(simEndTimes(i)+1:end, i) = NaN;
<a name="l01896"></a>01896                 width(simEndTimes(i)+1:end, i) = NaN;
<a name="l01897"></a>01897                 totalLength(simEndTimes(i)+1:end, i) = NaN;
<a name="l01898"></a>01898                 surfaceArea(simEndTimes(i)+1:end, i) = NaN;
<a name="l01899"></a>01899                 pinchedDiameter(simEndTimes(i)+1:end, i) = NaN;
<a name="l01900"></a>01900             end
<a name="l01901"></a>01901             
<a name="l01902"></a>01902             clear states;
<a name="l01903"></a>01903             
<a name="l01904"></a>01904             %% plot data
<a name="l01905"></a>01905             clf(figHandle);
<a name="l01906"></a>01906             
<a name="l01907"></a>01907             options = struct();
<a name="l01908"></a>01908             options.titleStr = &#39;Cell Shape&#39;;
<a name="l01909"></a>01909             options.colorOrder = PlotUtil.getRedGreenColorOrder(simOrder);
<a name="l01910"></a>01910             options.xdata = time;
<a name="l01911"></a>01911             options.ylabelStr = {
<a name="l01912"></a>01912                 <span class="stringliteral">&#39;Mass (fg)&#39;</span>
<a name="l01913"></a>01913                 <span class="stringliteral">&#39;Volume (fL)&#39;</span>
<a name="l01914"></a>01914                 {<span class="stringliteral">&#39;Surface Area&#39;</span> <span class="stringliteral">&#39;({\mu}m^2)&#39;</span>}
<a name="l01915"></a>01915                 <span class="stringliteral">&#39;Length ({\mu}m)&#39;</span>
<a name="l01916"></a>01916                 <span class="stringliteral">&#39;Width ({\mu}m)&#39;</span>
<a name="l01917"></a>01917                 {<span class="stringliteral">&#39;Pinched&#39;</span> <span class="stringliteral">&#39;Diameter ({\mu}m)&#39;</span>}
<a name="l01918"></a>01918                 };
<a name="l01919"></a>01919             options.ydata = {
<a name="l01920"></a>01920                 mass
<a name="l01921"></a>01921                 volume
<a name="l01922"></a>01922                 surfaceArea
<a name="l01923"></a>01923                 totalLength
<a name="l01924"></a>01924                 width
<a name="l01925"></a>01925                 pinchedDiameter
<a name="l01926"></a>01926                 };
<a name="l01927"></a>01927             PlotUtil.multiElementPlot(figHandle, 3 * ones(6, 1), [0 time(end)], options);
<a name="l01928"></a>01928             
<a name="l01929"></a>01929             clear options;
<a name="l01930"></a>01930             
<a name="l01931"></a>01931             %% organize figure data
<a name="l01932"></a>01932             figData = <span class="keyword">struct</span>;
<a name="l01933"></a>01933             figData.simOrder = simOrder;
<a name="l01934"></a>01934             figData.time = time;
<a name="l01935"></a>01935             figData.mass = mass;
<a name="l01936"></a>01936             figData.volume = volume;
<a name="l01937"></a>01937             figData.surfaceArea = surfaceArea;
<a name="l01938"></a>01938             figData.totalLength = totalLength;
<a name="l01939"></a>01939             figData.width = width;
<a name="l01940"></a>01940             figData.pinchedDiameter = pinchedDiameter;
<a name="l01941"></a>01941         end
<a name="l01942"></a>01942         
<a name="l01943"></a>01943         <span class="keyword">function</span> figData = rnaPolymerases(figHandle, simBatchDir, selectedSimulations)
<a name="l01944"></a>01944             import edu.stanford.covert.cell.sim.util.PlotUtil;
<a name="l01945"></a>01945             import edu.stanford.covert.cell.sim.util.SimulationEnsemble;
<a name="l01946"></a>01946             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
<a name="l01947"></a>01947             
<a name="l01948"></a>01948             if nargin &lt; 2 || isempty(simBatchDir)
<a name="l01949"></a>01949                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
<a name="l01950"></a>01950             end
<a name="l01951"></a>01951             if nargin &lt; 3 || isempty(selectedSimulations)
<a name="l01952"></a>01952                 selectedSimulations = SimulationDiskUtil.getCompleteSimulations(simBatchDir);
<a name="l01953"></a>01953             end
<a name="l01954"></a>01954             
<a name="l01955"></a>01955             %% constants
<a name="l01956"></a>01956             [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep &#39;1&#39;]);
<a name="l01957"></a>01957             g = sim.gene;
<a name="l01958"></a>01958             time = sim.state(&#39;Time&#39;);
<a name="l01959"></a>01959             rna = sim.state(&#39;Rna&#39;);
<a name="l01960"></a>01960             pc = sim.state(&#39;ProteinComplex&#39;);
<a name="l01961"></a>01961             rnaPol = sim.state(&#39;RNAPolymerase&#39;);
<a name="l01962"></a>01962             tr = sim.process(&#39;TranscriptionalRegulation&#39;);
<a name="l01963"></a>01963             sc = sim.process(&#39;DNASupercoiling&#39;);
<a name="l01964"></a>01964             pcComp = sum(pc.proteinComplexComposition(setdiff(1:end, g.mRNAIndexs), :, :), 3);
<a name="l01965"></a>01965             
<a name="l01966"></a>01966             fcRNAIdxs = unique([tr.tuIndexs(:); sc.tuIndexs(:)]);
<a name="l01967"></a>01967             
<a name="l01968"></a>01968             expectedRNASyn = rna.expression(rna.matureIndexs);
<a name="l01969"></a>01969             expectedRNASyn = expectedRNASyn / sum(expectedRNASyn);
<a name="l01970"></a>01970             
<a name="l01971"></a>01971             expectedRNAExp = expectedRNASyn ./ (log(2) / time.cellCycleLength + rna.decayRates(rna.matureIndexs));
<a name="l01972"></a>01972             expectedRNAExp = expectedRNAExp / sum(expectedRNAExp);
<a name="l01973"></a>01973             
<a name="l01974"></a>01974             %% get data
<a name="l01975"></a>01975             %<a class="code" href="class_r_n_a.html">RNA</a> polymerases
<a name="l01976"></a>01976             stateNames = {
<a name="l01977"></a>01977                 <span class="stringliteral">&#39;Time&#39;</span>           <span class="stringliteral">&#39;values&#39;</span>                                    <span class="charliteral">&#39;:&#39;</span> <span class="charliteral">&#39;:&#39;</span>
<a name="l01978"></a>01978                 <span class="stringliteral">&#39;Mass&#39;</span>           <span class="stringliteral">&#39;cell&#39;</span>                                      <span class="charliteral">&#39;:&#39;</span> <span class="stringliteral">&#39;-sum&#39;</span>
<a name="l01979"></a>01979                 <span class="stringliteral">&#39;RNAPolymerase&#39;</span>  <span class="stringliteral">&#39;nActive&#39;</span>                                   <span class="charliteral">&#39;:&#39;</span> <span class="charliteral">&#39;:&#39;</span>
<a name="l01980"></a>01980                 <span class="stringliteral">&#39;RNAPolymerase&#39;</span>  <span class="stringliteral">&#39;nSpecificallyBound&#39;</span>                        <span class="charliteral">&#39;:&#39;</span> <span class="charliteral">&#39;:&#39;</span>
<a name="l01981"></a>01981                 <span class="stringliteral">&#39;RNAPolymerase&#39;</span>  <span class="stringliteral">&#39;nNonSpecificallyBound&#39;</span>                     <span class="charliteral">&#39;:&#39;</span> <span class="charliteral">&#39;:&#39;</span>
<a name="l01982"></a>01982                 <span class="stringliteral">&#39;RNAPolymerase&#39;</span>  <span class="stringliteral">&#39;nFree&#39;</span>                                     <span class="charliteral">&#39;:&#39;</span> <span class="charliteral">&#39;:&#39;</span>
<a name="l01983"></a>01983                 <span class="stringliteral">&#39;RNAPolymerase&#39;</span>  <span class="stringliteral">&#39;transcriptionFactorBindingProbFoldChange&#39;</span>  fcRNAIdxs 1
<a name="l01984"></a>01984                 <span class="stringliteral">&#39;RNAPolymerase&#39;</span>  <span class="stringliteral">&#39;supercoilingBindingProbFoldChange&#39;</span>         fcRNAIdxs 1
<a name="l01985"></a>01985                 };
<a name="l01986"></a>01986             states = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, <span class="stringliteral">&#39;extract&#39;</span>, selectedSimulations);
<a name="l01987"></a>01987             time = permute(max(states.Time.values, [], 4), [3 4 1 2]) / 3600;
<a name="l01988"></a>01988             mass = permute(states.Mass.cell, [3 4 1 2]);
<a name="l01989"></a>01989             
<a name="l01990"></a>01990             simEndTimes = permute(max(states.Time.values, [], 3), [4 3 1 2]);
<a name="l01991"></a>01991             minSimEndTime = min(simEndTimes);
<a name="l01992"></a>01992             [~, simOrder] = sort(mass(minSimEndTime, :));
<a name="l01993"></a>01993             
<a name="l01994"></a>01994             foldChanges = ...
<a name="l01995"></a>01995                 states.RNAPolymerase.transcriptionFactorBindingProbFoldChange .* ...
<a name="l01996"></a>01996                 states.RNAPolymerase.supercoilingBindingProbFoldChange;
<a name="l01997"></a>01997             <span class="keywordflow">for</span> i = 1:numel(simEndTimes)
<a name="l01998"></a>01998                 foldChanges(:, :, simEndTimes(i)+1:end, i) = NaN;
<a name="l01999"></a>01999             end
<a name="l02000"></a>02000             
<a name="l02001"></a>02001             actPols = states.RNAPolymerase.nActive;
<a name="l02002"></a>02002             sbPols = states.RNAPolymerase.nSpecificallyBound;
<a name="l02003"></a>02003             nsbPols = states.RNAPolymerase.nNonSpecificallyBound;
<a name="l02004"></a>02004             freePols = states.RNAPolymerase.nFree;
<a name="l02005"></a>02005             tmp = (actPols + sbPols + nsbPols + freePols);
<a name="l02006"></a>02006             actPols = actPols ./ tmp * 100;
<a name="l02007"></a>02007             sbPols = sbPols ./ tmp * 100;
<a name="l02008"></a>02008             nsbPols = nsbPols ./ tmp * 100;
<a name="l02009"></a>02009             freePols = freePols ./ tmp * 100;
<a name="l02010"></a>02010             
<a name="l02011"></a>02011             for i = 1:numel(simEndTimes)
<a name="l02012"></a>02012                 actPols(:, :, simEndTimes(i)+1:end, i) = NaN;
<a name="l02013"></a>02013                 sbPols(:, :, simEndTimes(i)+1:end, i) = NaN;
<a name="l02014"></a>02014                 nsbPols(:, :, simEndTimes(i)+1:end, i) = NaN;
<a name="l02015"></a>02015                 freePols(:, :, simEndTimes(i)+1:end, i) = NaN;
<a name="l02016"></a>02016             end
<a name="l02017"></a>02017             
<a name="l02018"></a>02018             %cleanup
<a name="l02019"></a>02019             clear clear states mass minSimEndTime tmp;
<a name="l02020"></a>02020             
<a name="l02021"></a>02021             %transcription units
<a name="l02022"></a>02022             stateNames = {
<a name="l02023"></a>02023                 <span class="stringliteral">&#39;RNAPolymerase&#39;</span>  <span class="stringliteral">&#39;states&#39;</span>                   <span class="charliteral">&#39;:&#39;</span>  <span class="charliteral">&#39;:&#39;</span>
<a name="l02024"></a>02024                 <span class="stringliteral">&#39;Transcript&#39;</span>     <span class="stringliteral">&#39;boundTranscriptionUnits&#39;</span>  <span class="charliteral">&#39;:&#39;</span>  <span class="charliteral">&#39;:&#39;</span>
<a name="l02025"></a>02025                 <span class="stringliteral">&#39;Transcript&#39;</span>     <span class="stringliteral">&#39;boundTranscriptProgress&#39;</span>  <span class="charliteral">&#39;:&#39;</span>  <span class="charliteral">&#39;:&#39;</span>
<a name="l02026"></a>02026                 <span class="stringliteral">&#39;Rna&#39;</span>            <span class="stringliteral">&#39;counts&#39;</span>                   <span class="charliteral">&#39;:&#39;</span>  <span class="stringliteral">&#39;-sum&#39;</span>
<a name="l02027"></a>02027                 <span class="stringliteral">&#39;ProteinComplex&#39;</span> <span class="stringliteral">&#39;counts&#39;</span>                   <span class="charliteral">&#39;:&#39;</span>  <span class="stringliteral">&#39;-sum&#39;</span>
<a name="l02028"></a>02028                 };
<a name="l02029"></a>02029             rnaSyn = zeros(numel(rna.matureIndexs), 1, 1, numel(selectedSimulations));
<a name="l02030"></a>02030             rnaExp = zeros(numel(rna.matureIndexs), 1, 1, numel(selectedSimulations));
<a name="l02031"></a>02031             
<a name="l02032"></a>02032             rnaPolAbsPosDensity = zeros(numel(rna.nascentIndexs), 1 + max(rna.lengths(rna.nascentIndexs)));
<a name="l02033"></a>02033             
<a name="l02034"></a>02034             <span class="keywordflow">for</span> i = 1:numel(selectedSimulations)
<a name="l02035"></a>02035                 states = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, &#39;extract&#39;, selectedSimulations(i));
<a name="l02036"></a>02036                 
<a name="l02037"></a>02037                 tfs = cat(3, diff(states.Transcript.boundTranscriptProgress, [], 3) &lt; 0, states.Transcript.boundTranscriptionUnits(:, :, end) ~= rnaPol.notExistValue);
<a name="l02038"></a>02038                 rnaSyn(:, :, :, i) = histc(states.Transcript.boundTranscriptionUnits(tfs), (1:numel(rna.matureIndexs)));
<a name="l02039"></a>02039                 
<a name="l02040"></a>02040                 tmp = sum(states.Rna.counts, 3);
<a name="l02041"></a>02041                 rnaExp(:, :, :, i) = full(...
<a name="l02042"></a>02042                     + tmp(rna.processedIndexs, :) ...
<a name="l02043"></a>02043                     + tmp(rna.matureIndexs, :) ...
<a name="l02044"></a>02044                     + tmp(rna.aminoacylatedIndexs, :) ...
<a name="l02045"></a>02045                     + tmp(rna.boundIndexs, :) ...
<a name="l02046"></a>02046                     + tmp(rna.misfoldedIndexs, :) ...
<a name="l02047"></a>02047                     + tmp(rna.damagedIndexs, :));
<a name="l02048"></a>02048                 rnaExp(setdiff(1:end, rna.matureMRNAIndexs), :, :, i) = ...
<a name="l02049"></a>02049                     + rnaExp(setdiff(1:end, rna.matureMRNAIndexs), :, :, i) ...
<a name="l02050"></a>02050                     + pcComp * full(sum(reshape(sum(states.ProteinComplex.counts, 3), ...
<a name="l02051"></a>02051                     [numel(pc.matureIndexs)  numel(pc.wholeCellModelIDs)/numel(pc.matureIndexs)]), 2));
<a name="l02052"></a>02052                 
<a name="l02053"></a>02053                 states.RNAPolymerase.states = full(states.RNAPolymerase.states);
<a name="l02054"></a>02054                 states.Transcript.boundTranscriptionUnits = full(states.Transcript.boundTranscriptionUnits);
<a name="l02055"></a>02055                 states.Transcript.boundTranscriptProgress = full(states.Transcript.boundTranscriptProgress);
<a name="l02056"></a>02056                 states.RNAPolymerase.states = states.RNAPolymerase.states(:);
<a name="l02057"></a>02057                 states.Transcript.boundTranscriptionUnits = states.Transcript.boundTranscriptionUnits(:);
<a name="l02058"></a>02058                 states.Transcript.boundTranscriptProgress = states.Transcript.boundTranscriptProgress(:);
<a name="l02059"></a>02059                 tfs = states.RNAPolymerase.states &gt;= rnaPol.activelyTranscribingValue;
<a name="l02060"></a>02060                 rnaPolAbsPosDensity = ...
<a name="l02061"></a>02061                     + rnaPolAbsPosDensity ...
<a name="l02062"></a>02062                     + accumarray([states.Transcript.boundTranscriptionUnits(tfs)  states.Transcript.boundTranscriptProgress(tfs) + 1], ones(sum(tfs), 1), size(rnaPolAbsPosDensity));
<a name="l02063"></a>02063                 
<a name="l02064"></a>02064                 clear states tfs tmp;
<a name="l02065"></a>02065             end
<a name="l02066"></a>02066             
<a name="l02067"></a>02067             totRNASyn = sum(rnaSyn, 1);
<a name="l02068"></a>02068             rnaSyn = rnaSyn ./ totRNASyn(ones(size(rnaSyn, 1), 1), :, :, :);
<a name="l02069"></a>02069             meanRNASyn = mean(rnaSyn, 4);
<a name="l02070"></a>02070             stdRNASyn = std(rnaSyn, [], 4);
<a name="l02071"></a>02071             
<a name="l02072"></a>02072             totRNAExp = sum(rnaExp, 1);
<a name="l02073"></a>02073             rnaExp = rnaExp ./ totRNAExp(ones(size(rnaExp, 1), 1), :, :, :);
<a name="l02074"></a>02074             meanRNAExp = mean(rnaExp, 4);
<a name="l02075"></a>02075             stdRNAExp = std(rnaExp, [], 4);
<a name="l02076"></a>02076             
<a name="l02077"></a>02077             rnaPolRelPosDensity = zeros(numel(rna.nascentIndexs), 100);
<a name="l02078"></a>02078             for i = 1:numel(rna.nascentIndexs)
<a name="l02079"></a>02079                 rnaPolRelPosDensity(i, :) = accumarray(...
<a name="l02080"></a>02080                     [ones(rna.lengths(rna.nascentIndexs(i)) + 1, 1)  ceil((1:rna.lengths(rna.nascentIndexs(i)) + 1)&#39; / (rna.lengths(rna.nascentIndexs(i)) + 1) * 100)], ...
<a name="l02081"></a>02081                     rnaPolAbsPosDensity(i, 1:rna.lengths(rna.nascentIndexs(i)) + 1)&#39;, [1 100]);
<a name="l02082"></a>02082             end
<a name="l02083"></a>02083             
<a name="l02084"></a>02084             %cleanup
<a name="l02085"></a>02085             clear rnaSyn rnaExp totRNASyn totRNAExp;
<a name="l02086"></a>02086             
<a name="l02087"></a>02087             %% plot data
<a name="l02088"></a>02088             clf(figHandle);
<a name="l02089"></a>02089             
<a name="l02090"></a>02090             options = struct();
<a name="l02091"></a>02091             options.titleStr = {<span class="stringliteral">&#39;Transcriptional Regulation&#39;</span>; <span class="stringliteral">&#39;RNA Polymerase&#39;</span>; <span class="stringliteral">&#39;RNA&#39;</span>};
<a name="l02092"></a>02092             options.colorOrder = PlotUtil.getRedGreenColorOrder(simOrder);
<a name="l02093"></a>02093             axesHandles = PlotUtil.multiElementPlot(figHandle, {10 * ones(size(fcRNAIdxs)) [10 10 10 10] [10 10 10 10]}, [0 time(end)], options);
<a name="l02094"></a>02094             
<a name="l02095"></a>02095             %left
<a name="l02096"></a>02096             <span class="keywordflow">for</span> i = 1:numel(fcRNAIdxs)
<a name="l02097"></a>02097                 axesHandle = axesHandles{1}(i);
<a name="l02098"></a>02098                 PlotUtil.plotLine(axesHandle, time, permute(foldChanges(i, :, :, :), [3 4 1 2]));
<a name="l02099"></a>02099                 xlim(axesHandle, [0 time(end)]);
<a name="l02100"></a>02100                 ylabel(axesHandle, rna.wholeCellModelIDs{rna.nascentIndexs(fcRNAIdxs(i))});
<a name="l02101"></a>02101             end
<a name="l02102"></a>02102             
<a name="l02103"></a>02103             %middle
<a name="l02104"></a>02104             axesHandle = axesHandles{2}(1);
<a name="l02105"></a>02105             plot(axesHandle, time, permute(actPols, [3 4 1 2]));
<a name="l02106"></a>02106             xlim(axesHandle, [0 time(end)])
<a name="l02107"></a>02107             ylim(axesHandle, [0 100])
<a name="l02108"></a>02108             set(axesHandle, &#39;YTick&#39;, [0 50 100]);
<a name="l02109"></a>02109             ylabel(axesHandle, &#39;Active&#39;);
<a name="l02110"></a>02110             
<a name="l02111"></a>02111             axesHandle = axesHandles{2}(2);
<a name="l02112"></a>02112             plot(axesHandle, time, permute(sbPols, [3 4 1 2]));
<a name="l02113"></a>02113             xlim(axesHandle, [0 time(end)])
<a name="l02114"></a>02114             ylim(axesHandle, [0 100])
<a name="l02115"></a>02115             set(axesHandle, &#39;YTick&#39;, [0 50 100]);
<a name="l02116"></a>02116             ylabel(axesHandle, {<span class="stringliteral">&#39;Specifically&#39;</span> <span class="stringliteral">&#39;bound&#39;</span>});
<a name="l02117"></a>02117             
<a name="l02118"></a>02118             axesHandle = axesHandles{2}(3);
<a name="l02119"></a>02119             plot(axesHandle, time, permute(nsbPols, [3 4 1 2]));
<a name="l02120"></a>02120             xlim(axesHandle, [0 time(end)])
<a name="l02121"></a>02121             ylim(axesHandle, [0 100])
<a name="l02122"></a>02122             set(axesHandle, &#39;YTick&#39;, [0 50 100]);
<a name="l02123"></a>02123             ylabel(axesHandle, {<span class="stringliteral">&#39;Non-specifically&#39;</span> <span class="stringliteral">&#39;bound&#39;</span>});
<a name="l02124"></a>02124             
<a name="l02125"></a>02125             axesHandle = axesHandles{2}(4);
<a name="l02126"></a>02126             plot(axesHandle, time, permute(freePols, [3 4 1 2]));
<a name="l02127"></a>02127             xlim(axesHandle, [0 time(end)])
<a name="l02128"></a>02128             ylim(axesHandle, [0 100])
<a name="l02129"></a>02129             set(axesHandle, &#39;YTick&#39;, [0 50 100]);
<a name="l02130"></a>02130             ylabel(axesHandle, &#39;Free&#39;);
<a name="l02131"></a>02131             
<a name="l02132"></a>02132             %right
<a name="l02133"></a>02133             axesHandle = axesHandles{3}(1);
<a name="l02134"></a>02134             cla(axesHandle);
<a name="l02135"></a>02135             density = smooth(sum(rnaPolAbsPosDensity, 1) / sum(simEndTimes) * 3600, 11, <span class="stringliteral">&#39;moving&#39;</span>);
<a name="l02136"></a>02136             plot(axesHandle, 0:max(rna.lengths(rna.nascentIndexs)), density, <span class="stringliteral">&#39;Color&#39;</span>, <span class="charliteral">&#39;r&#39;</span>);
<a name="l02137"></a>02137             xlim(axesHandle, [-0.5 max(rna.lengths(rna.nascentIndexs)) + 0.5]);
<a name="l02138"></a>02138             ylim(axesHandle, [0 max(density)]);
<a name="l02139"></a>02139             <span class="keyword">set</span>(axesHandle, <span class="stringliteral">&#39;XTick&#39;</span>, [0 max(rna.lengths(rna.nascentIndexs))]);
<a name="l02140"></a>02140             <span class="keyword">set</span>(axesHandle, <span class="stringliteral">&#39;YTickMode&#39;</span>, <span class="stringliteral">&#39;auto&#39;</span>);
<a name="l02141"></a>02141             <span class="keyword">set</span>(axesHandle, <span class="stringliteral">&#39;XColor&#39;</span>, <span class="charliteral">&#39;k&#39;</span>);
<a name="l02142"></a>02142             <span class="keyword">set</span>(axesHandle, <span class="stringliteral">&#39;YScale&#39;</span>, <span class="stringliteral">&#39;log&#39;</span>)
<a name="l02143"></a>02143             xlabel(axesHandle, <span class="stringliteral">&#39;Absolute Position&#39;</span>);
<a name="l02144"></a>02144             ylabel(axesHandle, {<span class="stringliteral">&#39;Density&#39;</span>  <span class="stringliteral">&#39;(h^{-1})&#39;</span>});
<a name="l02145"></a>02145             
<a name="l02146"></a>02146             axesHandle = axesHandles{3}(2);
<a name="l02147"></a>02147             density = smooth(sum(rnaPolRelPosDensity, 1) / sum(simEndTimes) * 3600, 11, <span class="stringliteral">&#39;moving&#39;</span>);
<a name="l02148"></a>02148             plot(axesHandle, (1:100) - 0.5, density, <span class="stringliteral">&#39;Color&#39;</span>, <span class="charliteral">&#39;r&#39;</span>);
<a name="l02149"></a>02149             xlim(axesHandle, [0 100]);
<a name="l02150"></a>02150             ylim(axesHandle, [0 max(density)]);
<a name="l02151"></a>02151             <span class="keyword">set</span>(axesHandle, <span class="stringliteral">&#39;XTick&#39;</span>, [0 100]);
<a name="l02152"></a>02152             <span class="keyword">set</span>(axesHandle, <span class="stringliteral">&#39;YTickMode&#39;</span>, <span class="stringliteral">&#39;auto&#39;</span>);
<a name="l02153"></a>02153             <span class="keyword">set</span>(axesHandle, <span class="stringliteral">&#39;XColor&#39;</span>, <span class="charliteral">&#39;k&#39;</span>);
<a name="l02154"></a>02154             <span class="keyword">set</span>(axesHandle, <span class="stringliteral">&#39;YScale&#39;</span>, <span class="stringliteral">&#39;log&#39;</span>)
<a name="l02155"></a>02155             xlabel(axesHandle, <span class="stringliteral">&#39;Relative Position (%)&#39;</span>);
<a name="l02156"></a>02156             ylabel(axesHandle, {<span class="stringliteral">&#39;Density&#39;</span>  <span class="stringliteral">&#39;(h^{-1})&#39;</span>});
<a name="l02157"></a>02157             
<a name="l02158"></a>02158             axesHandle = axesHandles{3}(3);
<a name="l02159"></a>02159             [~, order] = sort(expectedRNASyn);
<a name="l02160"></a>02160             errorbar((1:numel(rna.matureIndexs))<span class="stringliteral">&#39;, meanRNASyn(order), stdRNASyn(order), &#39;</span>Parent<span class="stringliteral">&#39;, axesHandle, &#39;</span>Color<span class="stringliteral">&#39;, &#39;</span>g<span class="stringliteral">&#39;);</span>
<a name="l02161"></a>02161 <span class="stringliteral">            plot(axesHandle, (1:numel(rna.matureIndexs))&#39;</span>, expectedRNASyn(order), <span class="stringliteral">&#39;Color&#39;</span>, <span class="charliteral">&#39;b&#39;</span>);
<a name="l02162"></a>02162             xlabel(axesHandle, <span class="stringliteral">&#39;Transcription Units&#39;</span>);
<a name="l02163"></a>02163             xlim(axesHandle, [0.5 numel(rna.matureIndexs)+0.5])
<a name="l02164"></a>02164             ylabel(axesHandle, &#39;Synthesis&#39;);
<a name="l02165"></a>02165             ylim(axesHandle, [0 0.05]);
<a name="l02166"></a>02166             set(axesHandle, &#39;YTick&#39;, 0:0.05:0.05);
<a name="l02167"></a>02167             
<a name="l02168"></a>02168             axesHandle = axesHandles{3}(4);
<a name="l02169"></a>02169             [~, order] = sort(expectedRNAExp);
<a name="l02170"></a>02170             errorbar((1:numel(rna.matureIndexs))<span class="stringliteral">&#39;, meanRNAExp(order), stdRNAExp(order), &#39;</span>Parent<span class="stringliteral">&#39;, axesHandle, &#39;</span>Color<span class="stringliteral">&#39;, &#39;</span>g<span class="stringliteral">&#39;);</span>
<a name="l02171"></a>02171 <span class="stringliteral">            plot(axesHandle, (1:numel(rna.matureIndexs))&#39;</span>, expectedRNAExp(order), <span class="stringliteral">&#39;Color&#39;</span>, <span class="charliteral">&#39;b&#39;</span>);
<a name="l02172"></a>02172             xlabel(axesHandle, <span class="stringliteral">&#39;Transcription Units&#39;</span>);
<a name="l02173"></a>02173             xlim(axesHandle, [0.5 numel(rna.matureIndexs)+0.5])
<a name="l02174"></a>02174             ylabel(axesHandle, &#39;Expression&#39;);
<a name="l02175"></a>02175             ylim(axesHandle, [0 0.05]);
<a name="l02176"></a>02176             set(axesHandle, &#39;YTick&#39;, 0:0.05:0.05);
<a name="l02177"></a>02177             
<a name="l02178"></a>02178             %y axes
<a name="l02179"></a>02179             PlotUtil.alignYAxesLabels(axesHandles{1});
<a name="l02180"></a>02180             PlotUtil.alignYAxesLabels(axesHandles{2});
<a name="l02181"></a>02181             PlotUtil.alignYAxesLabels(axesHandles{3});
<a name="l02182"></a>02182             PlotUtil.offsetYAxes(axesHandles{1}, 0.03);
<a name="l02183"></a>02183             PlotUtil.offsetYAxes(axesHandles{2}, 0.03);
<a name="l02184"></a>02184             PlotUtil.offsetYAxes(axesHandles{3}, 0.03);
<a name="l02185"></a>02185             
<a name="l02186"></a>02186             %% organize figure data
<a name="l02187"></a>02187             figData = <span class="keyword">struct</span>;
<a name="l02188"></a>02188             figData.simEndTimes = simEndTimes;
<a name="l02189"></a>02189             figData.simOrder = simOrder;
<a name="l02190"></a>02190             figData.time = time;
<a name="l02191"></a>02191             figData.foldChanges = foldChanges;
<a name="l02192"></a>02192             figData.actPols = actPols;
<a name="l02193"></a>02193             figData.sbPols = sbPols;
<a name="l02194"></a>02194             figData.nsbPols = nsbPols;
<a name="l02195"></a>02195             figData.freePols = freePols;
<a name="l02196"></a>02196             figData.meanRNASyn = meanRNASyn;
<a name="l02197"></a>02197             figData.stdRNASyn = stdRNASyn;
<a name="l02198"></a>02198             figData.meanRNAExp = meanRNAExp;
<a name="l02199"></a>02199             figData.stdRNAExp = stdRNAExp;
<a name="l02200"></a>02200             figData.expectedRNASyn = expectedRNASyn;
<a name="l02201"></a>02201             figData.expectedRNAExp = expectedRNAExp;
<a name="l02202"></a>02202             figData.rnaPolAbsPosDensity = rnaPolAbsPosDensity;
<a name="l02203"></a>02203             figData.rnaPolRelPosDensity = rnaPolRelPosDensity;
<a name="l02204"></a>02204         end
<a name="l02205"></a>02205         
<a name="l02206"></a>02206         <span class="keyword">function</span> figData = ribosomes(figHandle, simBatchDir, selectedSimulations)
<a name="l02207"></a>02207             import edu.stanford.covert.cell.sim.util.PlotUtil;
<a name="l02208"></a>02208             import edu.stanford.covert.cell.sim.util.SimulationEnsemble;
<a name="l02209"></a>02209             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
<a name="l02210"></a>02210             
<a name="l02211"></a>02211             if nargin &lt; 2 || isempty(simBatchDir)
<a name="l02212"></a>02212                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
<a name="l02213"></a>02213             end
<a name="l02214"></a>02214             if nargin &lt; 3 || isempty(selectedSimulations)
<a name="l02215"></a>02215                 selectedSimulations = SimulationDiskUtil.getCompleteSimulations(simBatchDir);
<a name="l02216"></a>02216             end
<a name="l02217"></a>02217             
<a name="l02218"></a>02218             %% constants
<a name="l02219"></a>02219             [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep &#39;1&#39;]);
<a name="l02220"></a>02220             g = sim.gene;
<a name="l02221"></a>02221             time = sim.state(&#39;Time&#39;);
<a name="l02222"></a>02222             rna = sim.state(&#39;Rna&#39;);
<a name="l02223"></a>02223             pm = sim.state(&#39;ProteinMonomer&#39;);
<a name="l02224"></a>02224             pc = sim.state(&#39;ProteinComplex&#39;);
<a name="l02225"></a>02225             rib = sim.state(&#39;Ribosome&#39;);
<a name="l02226"></a>02226             pcComp = sum(pc.proteinComplexComposition(g.mRNAIndexs, :, :), 3);
<a name="l02227"></a>02227             
<a name="l02228"></a>02228             expectedMonSyn = rna.matureRNAGeneComposition(g.mRNAIndexs, :) * rna.expression(rna.matureIndexs);
<a name="l02229"></a>02229             expectedMonSyn = expectedMonSyn / sum(expectedMonSyn);
<a name="l02230"></a>02230             
<a name="l02231"></a>02231             expectedMonExp = expectedMonSyn ./ (log(2) / time.cellCycleLength + pm.decayRates(pm.matureIndexs));
<a name="l02232"></a>02232             expectedMonExp = expectedMonExp / sum(expectedMonExp);
<a name="l02233"></a>02233             
<a name="l02234"></a>02234             %% get data
<a name="l02235"></a>02235             %ribosome states
<a name="l02236"></a>02236             stateNames = {
<a name="l02237"></a>02237                 <span class="stringliteral">&#39;Time&#39;</span>           <span class="stringliteral">&#39;values&#39;</span>          <span class="charliteral">&#39;:&#39;</span> <span class="charliteral">&#39;:&#39;</span>
<a name="l02238"></a>02238                 <span class="stringliteral">&#39;Mass&#39;</span>           <span class="stringliteral">&#39;cell&#39;</span>            <span class="charliteral">&#39;:&#39;</span> <span class="stringliteral">&#39;-sum&#39;</span>
<a name="l02239"></a>02239                 <span class="stringliteral">&#39;Ribosome&#39;</span>       <span class="stringliteral">&#39;nActive&#39;</span>         <span class="charliteral">&#39;:&#39;</span> <span class="charliteral">&#39;:&#39;</span>
<a name="l02240"></a>02240                 <span class="stringliteral">&#39;Ribosome&#39;</span>       <span class="stringliteral">&#39;nStalled&#39;</span>        <span class="charliteral">&#39;:&#39;</span> <span class="charliteral">&#39;:&#39;</span>
<a name="l02241"></a>02241                 };
<a name="l02242"></a>02242             states = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, <span class="stringliteral">&#39;extract&#39;</span>, selectedSimulations);
<a name="l02243"></a>02243             time = permute(max(states.Time.values, [], 4), [3 4 1 2]) / 3600;
<a name="l02244"></a>02244             mass = permute(states.Mass.cell, [3 4 1 2]);
<a name="l02245"></a>02245             
<a name="l02246"></a>02246             simEndTimes = permute(max(states.Time.values, [], 3), [4 3 1 2]);
<a name="l02247"></a>02247             minSimEndTime = min(simEndTimes);
<a name="l02248"></a>02248             [~, simOrder] = sort(mass(minSimEndTime, :));
<a name="l02249"></a>02249             
<a name="l02250"></a>02250             actRibs = states.Ribosome.nActive;
<a name="l02251"></a>02251             stalledRibs = states.Ribosome.nStalled;
<a name="l02252"></a>02252             tmp = (actRibs + stalledRibs);
<a name="l02253"></a>02253             actRibs = actRibs ./ tmp * 100;
<a name="l02254"></a>02254             stalledRibs = stalledRibs ./ tmp * 100;
<a name="l02255"></a>02255             
<a name="l02256"></a>02256             <span class="keywordflow">for</span> i = 1:numel(simEndTimes)
<a name="l02257"></a>02257                 actRibs(:, :, simEndTimes(i)+1:end, i) = NaN;
<a name="l02258"></a>02258                 stalledRibs(:, :, simEndTimes(i)+1:end, i) = NaN;
<a name="l02259"></a>02259             end
<a name="l02260"></a>02260             
<a name="l02261"></a>02261             %cleanup
<a name="l02262"></a>02262             clear states mass minSimEndTime tmp;
<a name="l02263"></a>02263             
<a name="l02264"></a>02264             %synthesis, expression
<a name="l02265"></a>02265             stateNames = {
<a name="l02266"></a>02266                 <span class="stringliteral">&#39;Ribosome&#39;</span>       <span class="stringliteral">&#39;states&#39;</span>          <span class="charliteral">&#39;:&#39;</span>  <span class="charliteral">&#39;:&#39;</span>
<a name="l02267"></a>02267                 <span class="stringliteral">&#39;Ribosome&#39;</span>       <span class="stringliteral">&#39;boundMRNAs&#39;</span>      <span class="charliteral">&#39;:&#39;</span>  <span class="charliteral">&#39;:&#39;</span>
<a name="l02268"></a>02268                 <span class="stringliteral">&#39;Ribosome&#39;</span>       <span class="stringliteral">&#39;mRNAPositions&#39;</span>   <span class="charliteral">&#39;:&#39;</span>  <span class="charliteral">&#39;:&#39;</span>
<a name="l02269"></a>02269                 <span class="stringliteral">&#39;ProteinMonomer&#39;</span> <span class="stringliteral">&#39;counts&#39;</span>          <span class="charliteral">&#39;:&#39;</span>  <span class="stringliteral">&#39;-sum&#39;</span>
<a name="l02270"></a>02270                 <span class="stringliteral">&#39;ProteinComplex&#39;</span> <span class="stringliteral">&#39;counts&#39;</span>          <span class="charliteral">&#39;:&#39;</span>  <span class="stringliteral">&#39;-sum&#39;</span>
<a name="l02271"></a>02271                 };
<a name="l02272"></a>02272             monSyn = zeros(numel(pm.matureIndexs), 1, 1, numel(selectedSimulations));
<a name="l02273"></a>02273             monExp = zeros(numel(pm.matureIndexs), 1, 1, numel(selectedSimulations));
<a name="l02274"></a>02274             cpxExp = zeros(numel(pc.matureIndexs), 1, 1, numel(selectedSimulations));
<a name="l02275"></a>02275             ribAbsPosDensity = zeros(numel(pm.nascentIndexs), 1 + max(pm.lengths(pm.nascentIndexs)));
<a name="l02276"></a>02276             <span class="keywordflow">for</span> i = 1:numel(selectedSimulations)
<a name="l02277"></a>02277                 states = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, &#39;extract&#39;, selectedSimulations(i));
<a name="l02278"></a>02278                 
<a name="l02279"></a>02279                 boundMRNAs = states.Ribosome.boundMRNAs;
<a name="l02280"></a>02280                 mRNAPositions = states.Ribosome.mRNAPositions;
<a name="l02281"></a>02281                 tfs = cat(3, diff(mRNAPositions, [], 3) &lt; 0, boundMRNAs(:, :, end) ~= rib.notExistValue);
<a name="l02282"></a>02282                 monSyn(:, :, :, i) = histc(boundMRNAs(tfs), (1:numel(pm.matureIndexs)));
<a name="l02283"></a>02283                 
<a name="l02284"></a>02284                 monExp(:, :, :, i) = full(sum(reshape(sum(states.ProteinMonomer.counts, 3), ...
<a name="l02285"></a>02285                     [numel(pm.matureIndexs)  numel(pm.wholeCellModelIDs)/numel(pm.matureIndexs)]), 2));
<a name="l02286"></a>02286                 cpxExp(:, :, :, i) = full(sum(reshape(sum(states.ProteinComplex.counts, 3), ...
<a name="l02287"></a>02287                     [numel(pc.matureIndexs)  numel(pc.wholeCellModelIDs)/numel(pc.matureIndexs)]), 2));
<a name="l02288"></a>02288                 
<a name="l02289"></a>02289                 states.Ribosome.states = full(states.Ribosome.states);
<a name="l02290"></a>02290                 states.Ribosome.boundMRNAs = full(states.Ribosome.boundMRNAs);
<a name="l02291"></a>02291                 states.Ribosome.mRNAPositions = full(states.Ribosome.mRNAPositions);
<a name="l02292"></a>02292                 states.Ribosome.states = states.Ribosome.states(:);
<a name="l02293"></a>02293                 states.Ribosome.boundMRNAs = states.Ribosome.boundMRNAs(:);
<a name="l02294"></a>02294                 states.Ribosome.mRNAPositions = states.Ribosome.mRNAPositions(:);
<a name="l02295"></a>02295                 tfs = states.Ribosome.states == rib.activeValue;
<a name="l02296"></a>02296                 ribAbsPosDensity = ...
<a name="l02297"></a>02297                     + ribAbsPosDensity ...
<a name="l02298"></a>02298                     + accumarray([states.Ribosome.boundMRNAs(tfs)  states.Ribosome.mRNAPositions(tfs) + 1], ones(sum(tfs), 1), size(ribAbsPosDensity));
<a name="l02299"></a>02299                 
<a name="l02300"></a>02300                 clear states boundMRNAs mRNAPositions tfs;
<a name="l02301"></a>02301             end
<a name="l02302"></a>02302             
<a name="l02303"></a>02303             totMonSyn = sum(monSyn, 1);
<a name="l02304"></a>02304             monSyn = monSyn ./ totMonSyn(ones(size(monSyn, 1), 1), :, :, :);
<a name="l02305"></a>02305             meanMonSyn = mean(monSyn, 4);
<a name="l02306"></a>02306             stdMonSyn = std(monSyn, [], 4);
<a name="l02307"></a>02307             
<a name="l02308"></a>02308             for j = 1:size(monExp, 4)
<a name="l02309"></a>02309                 monExp(:, :, :, j) = ...
<a name="l02310"></a>02310                     + monExp(:, :, :, j) ...
<a name="l02311"></a>02311                     + pcComp * cpxExp(:, :, :, j);
<a name="l02312"></a>02312             end
<a name="l02313"></a>02313             totMonExp = sum(monExp, 1);
<a name="l02314"></a>02314             monExp = monExp ./ totMonExp(ones(size(monExp, 1), 1), :, :, :);
<a name="l02315"></a>02315             meanMonExp = mean(monExp, 4);
<a name="l02316"></a>02316             stdMonExp = std(monExp, [], 4);
<a name="l02317"></a>02317             
<a name="l02318"></a>02318             ribRelPosDensity = zeros(numel(pm.nascentIndexs), 100);
<a name="l02319"></a>02319             for i = 1:numel(pm.nascentIndexs)
<a name="l02320"></a>02320                 ribRelPosDensity(i, :) = accumarray(...
<a name="l02321"></a>02321                     [ones(pm.lengths(pm.nascentIndexs(i)) + 1, 1)  ceil((1:pm.lengths(pm.nascentIndexs(i)) + 1)&#39; / (pm.lengths(pm.nascentIndexs(i)) + 1) * 100)], ...
<a name="l02322"></a>02322                     ribAbsPosDensity(i, 1:pm.lengths(pm.nascentIndexs(i)) + 1)&#39;, [1 100]);
<a name="l02323"></a>02323             end
<a name="l02324"></a>02324             
<a name="l02325"></a>02325             %cleanup
<a name="l02326"></a>02326             clear monExp monSyn cpxExp totMonSyn totMonExp;
<a name="l02327"></a>02327             
<a name="l02328"></a>02328             %% plot data
<a name="l02329"></a>02329             clf(figHandle);
<a name="l02330"></a>02330             
<a name="l02331"></a>02331             options = struct();
<a name="l02332"></a>02332             options.titleStr = {<span class="stringliteral">&#39;States&#39;</span>; <span class="stringliteral">&#39;Monomers&#39;</span>};
<a name="l02333"></a>02333             options.colorOrder = PlotUtil.getRedGreenColorOrder(simOrder);
<a name="l02334"></a>02334             axesHandles = PlotUtil.multiElementPlot(figHandle, {[10 10] [10 10 10 10]}, [0 time(end)], options);
<a name="l02335"></a>02335             
<a name="l02336"></a>02336             %left
<a name="l02337"></a>02337             axesHandle = axesHandles{1}(1);
<a name="l02338"></a>02338             plot(axesHandle, time, permute(actRibs, [3 4 1 2]));
<a name="l02339"></a>02339             xlim(axesHandle, [0 time(end)])
<a name="l02340"></a>02340             ylim(axesHandle, [0 100])
<a name="l02341"></a>02341             set(axesHandle, &#39;YTick&#39;, [0 50 100]);
<a name="l02342"></a>02342             ylabel(axesHandle, &#39;Active&#39;);
<a name="l02343"></a>02343             
<a name="l02344"></a>02344             axesHandle = axesHandles{1}(2);
<a name="l02345"></a>02345             plot(axesHandle, time, permute(stalledRibs, [3 4 1 2]));
<a name="l02346"></a>02346             xlim(axesHandle, [0 time(end)])
<a name="l02347"></a>02347             ylim(axesHandle, [0 100])
<a name="l02348"></a>02348             set(axesHandle, &#39;YTick&#39;, [0 50 100]);
<a name="l02349"></a>02349             ylabel(axesHandle, &#39;Stalled&#39;);
<a name="l02350"></a>02350             
<a name="l02351"></a>02351             %right
<a name="l02352"></a>02352             axesHandle = axesHandles{2}(1);
<a name="l02353"></a>02353             density = smooth(sum(ribAbsPosDensity, 1) / sum(simEndTimes) * 3600, 11, <span class="stringliteral">&#39;moving&#39;</span>);
<a name="l02354"></a>02354             plot(axesHandle, 0:max(pm.lengths(pm.nascentIndexs)), density, <span class="stringliteral">&#39;Color&#39;</span>, <span class="charliteral">&#39;r&#39;</span>);
<a name="l02355"></a>02355             xlim(axesHandle, [-0.5 max(pm.lengths(pm.nascentIndexs)) + 0.5]);
<a name="l02356"></a>02356             ylim(axesHandle, [0 max(density)]);
<a name="l02357"></a>02357             <span class="keyword">set</span>(axesHandle, <span class="stringliteral">&#39;XTick&#39;</span>, [0 max(pm.lengths(pm.nascentIndexs))]);
<a name="l02358"></a>02358             <span class="keyword">set</span>(axesHandle, <span class="stringliteral">&#39;YTickMode&#39;</span>, <span class="stringliteral">&#39;auto&#39;</span>, <span class="stringliteral">&#39;YScale&#39;</span>, <span class="stringliteral">&#39;log&#39;</span>);
<a name="l02359"></a>02359             <span class="keyword">set</span>(axesHandle, <span class="stringliteral">&#39;XColor&#39;</span>, <span class="charliteral">&#39;k&#39;</span>);
<a name="l02360"></a>02360             xlabel(axesHandle, <span class="stringliteral">&#39;Absolute Position&#39;</span>);
<a name="l02361"></a>02361             ylabel(axesHandle, {<span class="stringliteral">&#39;Density&#39;</span>  <span class="stringliteral">&#39;(h^{-1})&#39;</span>});
<a name="l02362"></a>02362             
<a name="l02363"></a>02363             axesHandle = axesHandles{2}(2);
<a name="l02364"></a>02364             density = smooth(sum(ribRelPosDensity, 1) / sum(simEndTimes) * 3600, 11, <span class="stringliteral">&#39;moving&#39;</span>);
<a name="l02365"></a>02365             plot(axesHandle, (1:100) - 0.5, density, <span class="stringliteral">&#39;Color&#39;</span>, <span class="charliteral">&#39;r&#39;</span>);
<a name="l02366"></a>02366             xlim(axesHandle, [0 100]);
<a name="l02367"></a>02367             ylim(axesHandle, [0 max(density)]);
<a name="l02368"></a>02368             <span class="keyword">set</span>(axesHandle, <span class="stringliteral">&#39;XTick&#39;</span>, [0 100]);
<a name="l02369"></a>02369             <span class="keyword">set</span>(axesHandle, <span class="stringliteral">&#39;YTickMode&#39;</span>, <span class="stringliteral">&#39;auto&#39;</span>, <span class="stringliteral">&#39;YScale&#39;</span>, <span class="stringliteral">&#39;log&#39;</span>);
<a name="l02370"></a>02370             <span class="keyword">set</span>(axesHandle, <span class="stringliteral">&#39;XColor&#39;</span>, <span class="charliteral">&#39;k&#39;</span>);
<a name="l02371"></a>02371             xlabel(axesHandle, <span class="stringliteral">&#39;Relative Position (%)&#39;</span>);
<a name="l02372"></a>02372             ylabel(axesHandle, {<span class="stringliteral">&#39;Density&#39;</span>  <span class="stringliteral">&#39;(h^{-1})&#39;</span>});
<a name="l02373"></a>02373             
<a name="l02374"></a>02374             axesHandle = axesHandles{2}(3);
<a name="l02375"></a>02375             [~, order] = sort(expectedMonSyn);
<a name="l02376"></a>02376             errorbar((1:numel(pm.matureIndexs))<span class="stringliteral">&#39;, meanMonSyn(order), stdMonSyn(order), &#39;</span>Parent<span class="stringliteral">&#39;, axesHandle, &#39;</span>Color<span class="stringliteral">&#39;, &#39;</span>g<span class="stringliteral">&#39;);</span>
<a name="l02377"></a>02377 <span class="stringliteral">            plot(axesHandle, (1:numel(pm.matureIndexs))&#39;</span>, expectedMonSyn(order), <span class="stringliteral">&#39;Color&#39;</span>, <span class="charliteral">&#39;b&#39;</span>);
<a name="l02378"></a>02378             xlabel(axesHandle, <span class="stringliteral">&#39;Protein-Coding Genes&#39;</span>);
<a name="l02379"></a>02379             xlim(axesHandle, [0.5 numel(pm.matureIndexs)+0.5])
<a name="l02380"></a>02380             ylabel(axesHandle, &#39;Synthesis&#39;);
<a name="l02381"></a>02381             ylim(axesHandle, [min(meanMonSyn(order) - stdMonSyn(order)) max(meanMonSyn(order) + stdMonSyn(order))]);
<a name="l02382"></a>02382             set(axesHandle, &#39;YTick&#39;, 0:0.01:0.02);
<a name="l02383"></a>02383             
<a name="l02384"></a>02384             axesHandle = axesHandles{2}(4);
<a name="l02385"></a>02385             [~, order] = sort(expectedMonExp);
<a name="l02386"></a>02386             errorbar((1:numel(pm.matureIndexs))<span class="stringliteral">&#39;, meanMonExp(order), stdMonExp(order), &#39;</span>Parent<span class="stringliteral">&#39;, axesHandle, &#39;</span>Color<span class="stringliteral">&#39;, &#39;</span>g<span class="stringliteral">&#39;);</span>
<a name="l02387"></a>02387 <span class="stringliteral">            plot(axesHandle, (1:numel(pm.matureIndexs))&#39;</span>, expectedMonExp(order), <span class="stringliteral">&#39;Color&#39;</span>, <span class="charliteral">&#39;b&#39;</span>);
<a name="l02388"></a>02388             xlabel(axesHandle, <span class="stringliteral">&#39;Protein-Coding Genes&#39;</span>);
<a name="l02389"></a>02389             xlim(axesHandle, [0.5 numel(pm.matureIndexs)+0.5])
<a name="l02390"></a>02390             ylabel(axesHandle, &#39;Expression&#39;);
<a name="l02391"></a>02391             ylim(axesHandle, [min(meanMonExp(order) - stdMonExp(order)) max(meanMonExp(order) + stdMonExp(order))]);
<a name="l02392"></a>02392             set(axesHandle, &#39;YTick&#39;, 0:0.01:0.02);
<a name="l02393"></a>02393             
<a name="l02394"></a>02394             %y axes
<a name="l02395"></a>02395             PlotUtil.alignYAxesLabels(axesHandles{1});
<a name="l02396"></a>02396             PlotUtil.alignYAxesLabels(axesHandles{2});
<a name="l02397"></a>02397             PlotUtil.offsetYAxes(axesHandles{1}, 0.03);
<a name="l02398"></a>02398             PlotUtil.offsetYAxes(axesHandles{2}, 0.03);
<a name="l02399"></a>02399             
<a name="l02400"></a>02400             %% organize figure data
<a name="l02401"></a>02401             figData = <span class="keyword">struct</span>;
<a name="l02402"></a>02402             figData.simEndTimes = simEndTimes;
<a name="l02403"></a>02403             figData.simOrder = simOrder;
<a name="l02404"></a>02404             figData.time = time;
<a name="l02405"></a>02405             figData.actRibs = actRibs;
<a name="l02406"></a>02406             figData.stalledRibs = stalledRibs;
<a name="l02407"></a>02407             figData.expectedMonSyn = expectedMonSyn;
<a name="l02408"></a>02408             figData.expectedMonExp = expectedMonExp;
<a name="l02409"></a>02409             figData.meanMonSyn = meanMonSyn;
<a name="l02410"></a>02410             figData.stdMonSyn = stdMonSyn;
<a name="l02411"></a>02411             figData.meanMonExp = meanMonExp;
<a name="l02412"></a>02412             figData.stdMonExp = stdMonExp;
<a name="l02413"></a>02413             figData.ribAbsPosDensity = ribAbsPosDensity;
<a name="l02414"></a>02414             figData.ribRelPosDensity = ribRelPosDensity;
<a name="l02415"></a>02415         end
<a name="l02416"></a>02416         
<a name="l02417"></a>02417         <span class="keyword">function</span> figData = macromoleculeExpression(figHandle, simBatchDir, selectedSimulations)
<a name="l02418"></a>02418             import edu.stanford.covert.cell.sim.util.PlotUtil;
<a name="l02419"></a>02419             import edu.stanford.covert.cell.sim.util.SimulationEnsemble;
<a name="l02420"></a>02420             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
<a name="l02421"></a>02421             
<a name="l02422"></a>02422             %% get data
<a name="l02423"></a>02423             if nargin &lt; 2 || isempty(simBatchDir)
<a name="l02424"></a>02424                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
<a name="l02425"></a>02425             end
<a name="l02426"></a>02426             if nargin &lt; 3 || isempty(selectedSimulations)
<a name="l02427"></a>02427                 selectedSimulations = SimulationDiskUtil.getCompleteSimulations(simBatchDir);
<a name="l02428"></a>02428             end
<a name="l02429"></a>02429             
<a name="l02430"></a>02430             %% constants
<a name="l02431"></a>02431             [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep &#39;1&#39;]);
<a name="l02432"></a>02432             g = sim.gene;
<a name="l02433"></a>02433             m = sim.state(&#39;<a class="code" href="class_metabolite.html">Metabolite</a>&#39;);
<a name="l02434"></a>02434             rna = sim.state(&#39;Rna&#39;);
<a name="l02435"></a>02435             mon = sim.state(&#39;ProteinMonomer&#39;);
<a name="l02436"></a>02436             cpx = sim.state(&#39;ProteinComplex&#39;);
<a name="l02437"></a>02437             
<a name="l02438"></a>02438             rnaPcComp = sum(cpx.proteinComplexComposition(setdiff(1:end, g.mRNAIndexs), :, :), 3);
<a name="l02439"></a>02439             monPcComp = sum(cpx.proteinComplexComposition(g.mRNAIndexs, :, :), 3);
<a name="l02440"></a>02440             
<a name="l02441"></a>02441             essGene = strcmp(g.essential, &#39;Y&#39;);
<a name="l02442"></a>02442             
<a name="l02443"></a>02443             %% get data
<a name="l02444"></a>02444             stateNames = {
<a name="l02445"></a>02445                 <span class="stringliteral">&#39;Rna&#39;</span>             <span class="stringliteral">&#39;counts&#39;</span>
<a name="l02446"></a>02446                 <span class="stringliteral">&#39;ProteinMonomer&#39;</span>  <span class="stringliteral">&#39;counts&#39;</span>
<a name="l02447"></a>02447                 <span class="stringliteral">&#39;ProteinComplex&#39;</span>  <span class="stringliteral">&#39;counts&#39;</span>
<a name="l02448"></a>02448                 };
<a name="l02449"></a>02449             rnaFreq = zeros(1000, numel(rna.matureIndexs));
<a name="l02450"></a>02450             monFreq = zeros(10000, numel(mon.matureIndexs));
<a name="l02451"></a>02451             <span class="keywordflow">for</span> i = 1:numel(selectedSimulations)
<a name="l02452"></a>02452                 states = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, &#39;extract&#39;, selectedSimulations(i));
<a name="l02453"></a>02453                 
<a name="l02454"></a>02454                 tmp = permute(sum(states.Rna.counts, 2), [1 3 2]);
<a name="l02455"></a>02455                 rnaCnts = ...
<a name="l02456"></a>02456                     + rna.nascentRNAMatureRNAComposition * full(tmp(rna.nascentIndexs, :)) ...
<a name="l02457"></a>02457                     + full(...
<a name="l02458"></a>02458                     + tmp(rna.processedIndexs, :) ...
<a name="l02459"></a>02459                     + tmp(rna.matureIndexs, :) ...
<a name="l02460"></a>02460                     + tmp(rna.boundIndexs, :) ...
<a name="l02461"></a>02461                     + tmp(rna.misfoldedIndexs, :) ...
<a name="l02462"></a>02462                     + tmp(rna.damagedIndexs, :) ...
<a name="l02463"></a>02463                     + tmp(rna.aminoacylatedIndexs, :) ...
<a name="l02464"></a>02464                     );
<a name="l02465"></a>02465                 clear tmp;
<a name="l02466"></a>02466                 
<a name="l02467"></a>02467                 monCnts = full(permute(sum(reshape(sum(states.ProteinMonomer.counts, 2), [numel(mon.matureIndexs)  size(states.ProteinMonomer.counts, 1) / numel(mon.matureIndexs) size(states.ProteinMonomer.counts, 3)]), 2), [1 3 2]));
<a name="l02468"></a>02468                 cpxCnts = full(permute(sum(reshape(sum(states.ProteinComplex.counts, 2), [numel(cpx.matureIndexs)  size(states.ProteinComplex.counts, 1) / numel(cpx.matureIndexs) size(states.ProteinComplex.counts, 3)]), 2), [1 3 2]));
<a name="l02469"></a>02469                 rnaCnts(setdiff(1:end, rna.matureMRNAIndexs), :) = ...
<a name="l02470"></a>02470                     + rnaCnts(setdiff(1:end, rna.matureMRNAIndexs), :) ...
<a name="l02471"></a>02471                     + rnaPcComp * cpxCnts;
<a name="l02472"></a>02472                 monCnts = ...
<a name="l02473"></a>02473                     + monCnts ...
<a name="l02474"></a>02474                     + monPcComp * cpxCnts;
<a name="l02475"></a>02475                 
<a name="l02476"></a>02476                 idxs = (1:numel(rna.matureIndexs))&#39;;
<a name="l02477"></a>02477                 idxs = idxs(:, ones(size(rnaCnts, 2), 1));
<a name="l02478"></a>02478                 rnaFreq = ...
<a name="l02479"></a>02479                     + rnaFreq ...
<a name="l02480"></a>02480                     + accumarray([rnaCnts(:)+1 idxs(:)], ones(numel(rnaCnts), 1), size(rnaFreq));
<a name="l02481"></a>02481                 
<a name="l02482"></a>02482                 idxs = (1:numel(mon.matureIndexs))&#39;;
<a name="l02483"></a>02483                 idxs = idxs(:, ones(size(monCnts, 2), 1));
<a name="l02484"></a>02484                 monFreq = ...
<a name="l02485"></a>02485                     + monFreq ...
<a name="l02486"></a>02486                     + accumarray([monCnts(:)+1 idxs(:)], ones(numel(monCnts), 1), size(monFreq));
<a name="l02487"></a>02487                 
<a name="l02488"></a>02488                 clear states rnaCnts monCnts cpxCnts idxs;
<a name="l02489"></a>02489             end
<a name="l02490"></a>02490             
<a name="l02491"></a>02491             rnaFreq = rnaFreq(1:find(any(rnaFreq, 2), 1, &#39;last&#39;), :);
<a name="l02492"></a>02492             monFreq = monFreq(1:find(any(monFreq, 2), 1, &#39;last&#39;), :);
<a name="l02493"></a>02493             
<a name="l02494"></a>02494             meanRnaCnt = ((0:size(rnaFreq, 1)-1) * rnaFreq) ./ sum(rnaFreq, 1);
<a name="l02495"></a>02495             stdRnaCnt = sqrt((((1:size(rnaFreq, 1)).^2) * rnaFreq) ./ sum(rnaFreq, 1) - meanRnaCnt.^2);
<a name="l02496"></a>02496             meanMRnaCnt = meanRnaCnt * rna.matureRNAGeneComposition(g.mRNAIndexs, :)&#39;;
<a name="l02497"></a>02497             stdMRnaCnt = stdRnaCnt * rna.matureRNAGeneComposition(g.mRNAIndexs, :)&#39;;
<a name="l02498"></a>02498             meanMonCnt = ((0:size(monFreq, 1)-1) * monFreq) ./ sum(monFreq, 1);
<a name="l02499"></a>02499             stdMonCnt = sqrt((((1:size(monFreq, 1)).^2) * monFreq) ./ sum(monFreq, 1) - meanMonCnt.^2);
<a name="l02500"></a>02500             
<a name="l02501"></a>02501             rnaLognParam = NaN(numel(rna.matureIndexs), 2);
<a name="l02502"></a>02502             rnaLognParamCI = NaN(numel(rna.matureIndexs), 2, 2);
<a name="l02503"></a>02503             rnaNonNormPValue = zeros(numel(rna.matureIndexs), 1);
<a name="l02504"></a>02504             for i = 1:numel(rna.matureIndexs)
<a name="l02505"></a>02505                 cnts = rude(rnaFreq(:, i), 0:size(rnaFreq, 1) - 1);
<a name="l02506"></a>02506                 if ~rnaFreq(1, i)
<a name="l02507"></a>02507                     [rnaLognParam(i, :), rnaLognParamCI(i, :, :)] = lognfit(cnts);
<a name="l02508"></a>02508                 end
<a name="l02509"></a>02509                 if cnts &gt; 5e3
<a name="l02510"></a>02510                     cnts = cnts(randsample(numel(cnts), 5e3, false));
<a name="l02511"></a>02511                 end
<a name="l02512"></a>02512                 [~, rnaNonNormPValue(i)] = swtest(log10(cnts), 0.05, 0);
<a name="l02513"></a>02513             end
<a name="l02514"></a>02514             
<a name="l02515"></a>02515             monGamParam = zeros(numel(mon.matureIndexs), 2);
<a name="l02516"></a>02516             monGamParamCI = zeros(numel(mon.matureIndexs), 2, 2);
<a name="l02517"></a>02517             monNonNormPValue = zeros(numel(mon.matureIndexs), 1);
<a name="l02518"></a>02518             for i = 1:numel(mon.matureIndexs)
<a name="l02519"></a>02519                 cnts = rude(monFreq(:, i), 0:size(monFreq, 1) - 1);
<a name="l02520"></a>02520                 [monGamParam(i, :), monGamParamCI(i, :, :)] = gamfit(cnts);
<a name="l02521"></a>02521                 if cnts &gt; 5e3
<a name="l02522"></a>02522                     cnts = cnts(randsample(numel(cnts), 5e3, false));
<a name="l02523"></a>02523                 end
<a name="l02524"></a>02524                 [~, monNonNormPValue(i)] = swtest(cnts, 0.05, 0);
<a name="l02525"></a>02525             end
<a name="l02526"></a>02526             
<a name="l02527"></a>02527             %% plot data
<a name="l02528"></a>02528             clf(figHandle);
<a name="l02529"></a>02529             
<a name="l02530"></a>02530             [axesHandles, xAxesHandles, ~, offsetAxesHandles] = PlotUtil.multiElementPlot(figHandle, {1 * ones(3, 1); 1 * ones(2,1); 3 * ones(3, 1); 3 * ones(4, 1)}, [0 1]);
<a name="l02531"></a>02531             <span class="keyword">delete</span>(cell2mat(xAxesHandles));
<a name="l02532"></a>02532             
<a name="l02533"></a>02533             <span class="keyword">set</span>(cell2mat(offsetAxesHandles), <span class="stringliteral">&#39;visible&#39;</span>, <span class="stringliteral">&#39;on&#39;</span>);
<a name="l02534"></a>02534             <span class="keyword">set</span>(cell2mat(axesHandles), <span class="stringliteral">&#39;xcolor&#39;</span>, <span class="charliteral">&#39;k&#39;</span>, <span class="stringliteral">&#39;xtickmode&#39;</span>, <span class="stringliteral">&#39;auto&#39;</span>, <span class="stringliteral">&#39;xlimmode&#39;</span>, <span class="stringliteral">&#39;auto&#39;</span>, <span class="stringliteral">&#39;ylimmode&#39;</span>, <span class="stringliteral">&#39;auto&#39;</span>);
<a name="l02535"></a>02535             
<a name="l02536"></a>02536             %<a class="code" href="classm_r_n_a.html">mRNA</a> count distribution <span class="keywordflow">for</span> particular gene
<a name="l02537"></a>02537             [~, idx] = max(meanMRnaCnt);
<a name="l02538"></a>02538             tmp = rnaFreq(:, rna.matureMRNAIndexs(idx)) / sum(rnaFreq(:, rna.matureMRNAIndexs(idx)));
<a name="l02539"></a>02539             axesHandle = axesHandles{1}(1);
<a name="l02540"></a>02540             cla(axesHandle);
<a name="l02541"></a>02541             bar(axesHandle, (1:size(tmp, 1))<span class="stringliteral">&#39;-1, tmp, &#39;</span>EdgeColor<span class="stringliteral">&#39;, &#39;</span>none<span class="stringliteral">&#39;, &#39;</span>BarWidth<span class="stringliteral">&#39;, 0.9);</span>
<a name="l02542"></a>02542 <span class="stringliteral">            minX = find(tmp, 1, &#39;</span>first<span class="stringliteral">&#39;) - 1.5;</span>
<a name="l02543"></a>02543 <span class="stringliteral">            maxX = find(tmp, 1, &#39;</span>last<span class="stringliteral">&#39;) - 0.5;</span>
<a name="l02544"></a>02544 <span class="stringliteral">            xlim(axesHandle, [minX maxX]);</span>
<a name="l02545"></a>02545 <span class="stringliteral">            ylim(axesHandle, [0 max(tmp)]);</span>
<a name="l02546"></a>02546 <span class="stringliteral">            xlabel(axesHandle, &#39;</span><a class="code" href="classm_r_n_a.html">mRNA</a> Count<span class="stringliteral">&#39;);</span>
<a name="l02547"></a>02547 <span class="stringliteral">            ylabel(axesHandle, &#39;</span>Freq<span class="stringliteral">&#39;);</span>
<a name="l02548"></a>02548 <span class="stringliteral">            </span>
<a name="l02549"></a>02549 <span class="stringliteral">            %rRNA count distribution for particular gene</span>
<a name="l02550"></a>02550 <span class="stringliteral">            x = rnaFreq(:, rna.matureRRNAIndexs(1));</span>
<a name="l02551"></a>02551 <span class="stringliteral">            cnts = [];</span>
<a name="l02552"></a>02552 <span class="stringliteral">            for j = 0:numel(x) - 1</span>
<a name="l02553"></a>02553 <span class="stringliteral">                cnts = [cnts; j(ones(x(j+1), 1), 1)]; %#ok&lt;AGROW&gt;</span>
<a name="l02554"></a>02554 <span class="stringliteral">            end</span>
<a name="l02555"></a>02555 <span class="stringliteral">            axesHandle = axesHandles{1}(2);</span>
<a name="l02556"></a>02556 <span class="stringliteral">            cla(axesHandle);</span>
<a name="l02557"></a>02557 <span class="stringliteral">            hold(axesHandle, &#39;</span>on<span class="stringliteral">&#39;);</span>
<a name="l02558"></a>02558 <span class="stringliteral">            </span>
<a name="l02559"></a>02559 <span class="stringliteral">            edges = logspace(log10(min(cnts)), log10(max(cnts)), 200);</span>
<a name="l02560"></a>02560 <span class="stringliteral">            freq = histc(cnts, edges);</span>
<a name="l02561"></a>02561 <span class="stringliteral">            freq = freq / sum(freq);</span>
<a name="l02562"></a>02562 <span class="stringliteral">            bar(axesHandle, log10(edges), freq, &#39;</span>EdgeColor<span class="stringliteral">&#39;, &#39;</span>none<span class="stringliteral">&#39;, &#39;</span>BarWidth<span class="stringliteral">&#39;, 2);</span>
<a name="l02563"></a>02563 <span class="stringliteral">            set(axesHandle, &#39;</span>XScale<span class="stringliteral">&#39;, &#39;</span>linear<span class="stringliteral">&#39;);</span>
<a name="l02564"></a>02564 <span class="stringliteral">            xlim(axesHandle, log10(edges([1 end])))</span>
<a name="l02565"></a>02565 <span class="stringliteral">            set(axesHandle, &#39;</span>XTickMode<span class="stringliteral">&#39;, &#39;</span><span class="keyword">auto</span><span class="stringliteral">&#39;);</span>
<a name="l02566"></a>02566 <span class="stringliteral">            set(axesHandle, &#39;</span>XTick<span class="stringliteral">&#39;, 10.^get(axesHandle, &#39;</span>XTick<span class="stringliteral">&#39;));</span>
<a name="l02567"></a>02567 <span class="stringliteral">            ylim(axesHandle, [0 max(freq)])</span>
<a name="l02568"></a>02568 <span class="stringliteral">            </span>
<a name="l02569"></a>02569 <span class="stringliteral">            y = lognpdf(edges, rnaLognParam(rna.matureRRNAIndexs(2), 1), rnaLognParam(rna.matureRRNAIndexs(2), 2));</span>
<a name="l02570"></a>02570 <span class="stringliteral">            plot(axesHandle, log10(edges), y, &#39;</span>r<span class="stringliteral">&#39;);</span>
<a name="l02571"></a>02571 <span class="stringliteral">            </span>
<a name="l02572"></a>02572 <span class="stringliteral">            xlabel(axesHandle, &#39;</span><a class="code" href="classr_r_n_a.html">rRNA</a> Count<span class="stringliteral">&#39;);</span>
<a name="l02573"></a>02573 <span class="stringliteral">            ylabel(axesHandle, &#39;</span>Freq<span class="stringliteral">&#39;);</span>
<a name="l02574"></a>02574 <span class="stringliteral">            </span>
<a name="l02575"></a>02575 <span class="stringliteral">            %monomer count distribution for particular gene</span>
<a name="l02576"></a>02576 <span class="stringliteral">            tmp = monFreq(:, 1) ./ sum(monFreq(:, 1));</span>
<a name="l02577"></a>02577 <span class="stringliteral">            axesHandle = axesHandles{1}(3);</span>
<a name="l02578"></a>02578 <span class="stringliteral">            cla(axesHandle);</span>
<a name="l02579"></a>02579 <span class="stringliteral">            bar(axesHandle, (1:size(tmp, 1))&#39;</span>-1, tmp, <span class="stringliteral">&#39;EdgeColor&#39;</span>, <span class="stringliteral">&#39;none&#39;</span>, <span class="stringliteral">&#39;BarWidth&#39;</span>, 2);
<a name="l02580"></a>02580             minX = find(tmp, 1, <span class="stringliteral">&#39;first&#39;</span>) - 1.5;
<a name="l02581"></a>02581             maxX = find(tmp, 1, <span class="stringliteral">&#39;last&#39;</span>) - 0.5;
<a name="l02582"></a>02582             xlim(axesHandle, [minX maxX])
<a name="l02583"></a>02583             ylim([0 max(tmp)]);
<a name="l02584"></a>02584             
<a name="l02585"></a>02585             x = linspace(max(0, minX), maxX, 100);
<a name="l02586"></a>02586             y = gampdf(x, monGamParam(1, 1), monGamParam(1, 2));
<a name="l02587"></a>02587             plot(axesHandle, x, y, <span class="charliteral">&#39;r&#39;</span>);
<a name="l02588"></a>02588             
<a name="l02589"></a>02589             xlabel(axesHandle, <span class="stringliteral">&#39;Protein Count&#39;</span>);
<a name="l02590"></a>02590             ylabel(axesHandle, <span class="stringliteral">&#39;Freq&#39;</span>);
<a name="l02591"></a>02591             
<a name="l02592"></a>02592             %freq of strains with mean protein number
<a name="l02593"></a>02593             axesHandle = axesHandles{2}(1);
<a name="l02594"></a>02594             cla(axesHandle);
<a name="l02595"></a>02595             edges = logspace(log10(min(meanMonCnt)), log10(max(meanMonCnt)), 10);
<a name="l02596"></a>02596             cnts = [
<a name="l02597"></a>02597                 histc(meanMonCnt, edges)
<a name="l02598"></a>02598                 histc(meanMonCnt(essGene(g.mRNAIndexs)), edges)
<a name="l02599"></a>02599                 ];
<a name="l02600"></a>02600             h = bar(axesHandle, log10(edges<span class="stringliteral">&#39;), cnts&#39;</span>, <span class="stringliteral">&#39;grouped&#39;</span>);
<a name="l02601"></a>02601             <span class="keyword">set</span>(h, <span class="stringliteral">&#39;BarWidth&#39;</span>, 2);
<a name="l02602"></a>02602             <span class="keyword">set</span>(h(1), <span class="stringliteral">&#39;EdgeColor&#39;</span>, <span class="stringliteral">&#39;none&#39;</span>)
<a name="l02603"></a>02603             <span class="keyword">set</span>(h(2), <span class="stringliteral">&#39;EdgeColor&#39;</span>, <span class="stringliteral">&#39;none&#39;</span>)
<a name="l02604"></a>02604             <span class="keyword">set</span>(axesHandle, <span class="stringliteral">&#39;XScale&#39;</span>, <span class="stringliteral">&#39;linear&#39;</span>);
<a name="l02605"></a>02605             xlim(axesHandle, log10(edges([1 end])))
<a name="l02606"></a>02606             set(axesHandle, &#39;XTickMode&#39;, &#39;auto&#39;);
<a name="l02607"></a>02607             set(axesHandle, &#39;XTick&#39;, 10.^get(axesHandle, &#39;XTick&#39;));
<a name="l02608"></a>02608             xlabel(axesHandle, &#39;Prot Exp&#39;);
<a name="l02609"></a>02609             ylabel(axesHandle, &#39;Freq&#39;);
<a name="l02610"></a>02610             
<a name="l02611"></a>02611             %noise vs mean protein number
<a name="l02612"></a>02612             axesHandle = axesHandles{2}(2);
<a name="l02613"></a>02613             cla(axesHandle);
<a name="l02614"></a>02614             hold(axesHandle, <span class="stringliteral">&#39;on&#39;</span>);
<a name="l02615"></a>02615             noise = (stdMonCnt ./ meanMonCnt).^2;
<a name="l02616"></a>02616             x = quantile(meanMonCnt, [0.01 0.99])<span class="stringliteral">&#39;;</span>
<a name="l02617"></a>02617 <span class="stringliteral">            y = 5./quantile(meanMonCnt, [0.01 0.99])&#39;</span>;
<a name="l02618"></a>02618             plot(axesHandle, meanMonCnt, noise, <span class="stringliteral">&#39;k.&#39;</span>);
<a name="l02619"></a>02619             plot(axesHandle, x, y, <span class="stringliteral">&#39;r-&#39;</span>);
<a name="l02620"></a>02620             plot(axesHandle, [min(meanMonCnt) max(meanMonCnt)], quantile(noise, 0.01) * [1 1], <span class="stringliteral">&#39;b-&#39;</span>);
<a name="l02621"></a>02621             ylim(axesHandle, [min(noise) max(noise)])
<a name="l02622"></a>02622             xlim(axesHandle, [min(meanMonCnt) max(meanMonCnt)])
<a name="l02623"></a>02623             set(axesHandle, &#39;XScale&#39;, &#39;log&#39;);
<a name="l02624"></a>02624             set(axesHandle, &#39;YScale&#39;, &#39;log&#39;);
<a name="l02625"></a>02625             xlabel(axesHandle, &#39;Mean <a class="code" href="class_protein.html">Protein</a> Number (\mu)&#39;);
<a name="l02626"></a>02626             ylabel(axesHandle, &#39;<a class="code" href="class_protein.html">Protein</a> Noise (\sigma^2 / \mu^2)&#39;);
<a name="l02627"></a>02627             
<a name="l02628"></a>02628             %<a class="code" href="class_protein.html">Protein</a> vs <a class="code" href="class_r_n_a.html">RNA</a> expression
<a name="l02629"></a>02629             axesHandle = axesHandles{3}(1);
<a name="l02630"></a>02630             cla(axesHandle);
<a name="l02631"></a>02631             plot(axesHandle, meanMRnaCnt, meanMonCnt, <span class="charliteral">&#39;.&#39;</span>);
<a name="l02632"></a>02632             xlim(axesHandle, [min(meanMRnaCnt) max(meanMRnaCnt)])
<a name="l02633"></a>02633             ylim(axesHandle, [min(meanMonCnt) max(meanMonCnt)])
<a name="l02634"></a>02634             xlabel(axesHandle, &#39;Mean <a class="code" href="classm_r_n_a.html">mRNA</a> Level&#39;);
<a name="l02635"></a>02635             ylabel(axesHandle, &#39;Mean <a class="code" href="class_protein.html">Protein</a> Number&#39;);
<a name="l02636"></a>02636             set(axesHandle, &#39;XScale&#39;, &#39;log&#39;);
<a name="l02637"></a>02637             set(axesHandle, &#39;YScale&#39;, &#39;log&#39;);
<a name="l02638"></a>02638             
<a name="l02639"></a>02639             %<a class="code" href="class_r_n_a.html">RNA</a> expression vs noise
<a name="l02640"></a>02640             axesHandle = axesHandles{3}(2);
<a name="l02641"></a>02641             cla(axesHandle);
<a name="l02642"></a>02642             noise = (stdMRnaCnt ./ meanMRnaCnt) .^2;
<a name="l02643"></a>02643             plot(axesHandle, meanMRnaCnt, noise, <span class="charliteral">&#39;.&#39;</span>);
<a name="l02644"></a>02644             xlim(axesHandle, [min(meanMRnaCnt) max(meanMRnaCnt)])
<a name="l02645"></a>02645             ylim(axesHandle, [min(noise) max(noise)])
<a name="l02646"></a>02646             xlabel(axesHandle, &#39;Mean MRNA Number&#39;);
<a name="l02647"></a>02647             ylabel(axesHandle, &#39;<a class="code" href="classm_r_n_a.html">mRNA</a> Noise&#39;);
<a name="l02648"></a>02648             set(axesHandle, &#39;XScale&#39;, &#39;log&#39;);
<a name="l02649"></a>02649             set(axesHandle, &#39;YScale&#39;, &#39;log&#39;);
<a name="l02650"></a>02650             
<a name="l02651"></a>02651             %<a class="code" href="classm_r_n_a.html">mRNA</a> fano factor frequence
<a name="l02652"></a>02652             axesHandle = axesHandles{3}(3);
<a name="l02653"></a>02653             cla(axesHandle);
<a name="l02654"></a>02654             tfs = ~isnan(stdMRnaCnt) &amp; meanMRnaCnt &gt; 0;
<a name="l02655"></a>02655             hist(axesHandle, (stdMRnaCnt(tfs) .^2) ./ meanMRnaCnt(tfs), 200);
<a name="l02656"></a>02656             h = findobj(axesHandle, <span class="stringliteral">&#39;Type&#39;</span>, <span class="stringliteral">&#39;patch&#39;</span>);
<a name="l02657"></a>02657             <span class="keyword">set</span>(h, <span class="stringliteral">&#39;EdgeColor&#39;</span>, <span class="stringliteral">&#39;none&#39;</span>);
<a name="l02658"></a>02658             xlim(axesHandle, [2 20]);
<a name="l02659"></a>02659             xlabel(axesHandle, <span class="stringliteral">&#39;mRNA Fano Factor&#39;</span>);
<a name="l02660"></a>02660             ylabel(axesHandle, <span class="stringliteral">&#39;Number of Strains&#39;</span>);
<a name="l02661"></a>02661             
<a name="l02662"></a>02662             %<a class="code" href="classm_r_n_a.html">mRNA</a> gamma a vs b
<a name="l02663"></a>02663             axesHandle = axesHandles{4}(1);
<a name="l02664"></a>02664             cla(axesHandle)
<a name="l02665"></a>02665             plot(axesHandle, monGamParam(:, 1), monGamParam(:, 2), &#39;.&#39;);
<a name="l02666"></a>02666             xlim(axesHandle, [min(monGamParam(:, 1)) max(monGamParam(:, 1))])
<a name="l02667"></a>02667             ylim(axesHandle, [min(monGamParam(:, 2)) max(monGamParam(:, 2))])
<a name="l02668"></a>02668             xlabel(axesHandle, &#39;a Value&#39;);
<a name="l02669"></a>02669             ylabel(axesHandle, &#39;b Value&#39;);
<a name="l02670"></a>02670             set(axesHandle, &#39;XScale&#39;, &#39;log&#39;);
<a name="l02671"></a>02671             set(axesHandle, &#39;XScale&#39;, &#39;log&#39;);
<a name="l02672"></a>02672             
<a name="l02673"></a>02673             axesHandle = axesHandles{4}(2);
<a name="l02674"></a>02674             cla(axesHandle)
<a name="l02675"></a>02675             plot(axesHandle, meanMonCnt, monGamParam(:, 2), &#39;.&#39;);
<a name="l02676"></a>02676             xlim(axesHandle, [min(meanMonCnt) max(meanMonCnt)])
<a name="l02677"></a>02677             ylim(axesHandle, [min(monGamParam(:, 2)) max(monGamParam(:, 2))])
<a name="l02678"></a>02678             xlabel(axesHandle, &#39;Mean <a class="code" href="class_protein.html">Protein</a> Number&#39;);
<a name="l02679"></a>02679             ylabel(axesHandle, &#39;b Value&#39;);
<a name="l02680"></a>02680             set(axesHandle, &#39;XScale&#39;, &#39;log&#39;);
<a name="l02681"></a>02681             set(axesHandle, &#39;XScale&#39;, &#39;log&#39;);
<a name="l02682"></a>02682             
<a name="l02683"></a>02683             %protein expression vs gc content
<a name="l02684"></a>02684             axesHandle = axesHandles{4}(3);
<a name="l02685"></a>02685             cla(axesHandle)
<a name="l02686"></a>02686             tmp = rna.baseCounts(rna.matureIndexs, m.nmpIndexs);
<a name="l02687"></a>02687             gcContent = rna.matureRNAGeneComposition(g.mRNAIndexs, :) * (sum(tmp(:, [2 3]), 2) ./ sum(tmp, 2));
<a name="l02688"></a>02688             plot(axesHandle, meanMonCnt, gcContent, &#39;.&#39;);
<a name="l02689"></a>02689             xlim(axesHandle, [min(meanMonCnt) max(meanMonCnt)])
<a name="l02690"></a>02690             ylim(axesHandle, [min(gcContent) max(gcContent)])
<a name="l02691"></a>02691             xlabel(&#39;Mean <a class="code" href="class_protein.html">Protein</a> Number&#39;);
<a name="l02692"></a>02692             ylabel(&#39;GC Content&#39;);
<a name="l02693"></a>02693             set(axesHandle, &#39;XScale&#39;, &#39;log&#39;);
<a name="l02694"></a>02694             
<a name="l02695"></a>02695             %protein expression vs <a class="code" href="classm_r_n_a.html">mRNA</a> half life
<a name="l02696"></a>02696             axesHandle = axesHandles{4}(4);
<a name="l02697"></a>02697             cla(axesHandle)
<a name="l02698"></a>02698             plot(axesHandle, meanMonCnt, rna.matureRNAGeneComposition(g.mRNAIndexs, :) * rna.halfLives(rna.matureIndexs), &#39;.&#39;);
<a name="l02699"></a>02699             xlim(axesHandle, [min(meanMonCnt) max(meanMonCnt)])
<a name="l02700"></a>02700             ylim(axesHandle, [min(rna.halfLives(rna.matureIndexs(rna.matureMRNAIndexs))) max(rna.halfLives(rna.matureIndexs(rna.matureMRNAIndexs)))])
<a name="l02701"></a>02701             xlabel(&#39;Mean <a class="code" href="class_protein.html">Protein</a> Number&#39;);
<a name="l02702"></a>02702             ylabel(&#39;<a class="code" href="classm_r_n_a.html">mRNA</a> Life Time (min)&#39;);
<a name="l02703"></a>02703             set(axesHandle, &#39;XScale&#39;, &#39;log&#39;);
<a name="l02704"></a>02704             set(axesHandle, &#39;YScale&#39;, &#39;log&#39;);
<a name="l02705"></a>02705             
<a name="l02706"></a>02706             %% data
<a name="l02707"></a>02707             figData = struct;
<a name="l02708"></a>02708             figData.rnaFreq = rnaFreq;
<a name="l02709"></a>02709             figData.monFreq = monFreq;
<a name="l02710"></a>02710             figData.meanRnaCnt = meanRnaCnt;
<a name="l02711"></a>02711             figData.stdRnaCnt = stdRnaCnt;
<a name="l02712"></a>02712             figData.meanMRnaCnt = meanMRnaCnt;
<a name="l02713"></a>02713             figData.stdMRnaCnt = stdMRnaCnt;
<a name="l02714"></a>02714             figData.meanMonCnt = meanMonCnt;
<a name="l02715"></a>02715             figData.stdMonCnt = stdMonCnt;
<a name="l02716"></a>02716             figData.rnaLognParam = rnaLognParam;
<a name="l02717"></a>02717             figData.rnaLognParamCI = rnaLognParamCI;
<a name="l02718"></a>02718             figData.rnaNonNormPValue = rnaNonNormPValue;
<a name="l02719"></a>02719             figData.monGamParam = monGamParam;
<a name="l02720"></a>02720             figData.monGamParamCI = monGamParamCI;
<a name="l02721"></a>02721             figData.monNonNormPValue = monNonNormPValue;
<a name="l02722"></a>02722         end
<a name="l02723"></a>02723         
<a name="l02724"></a>02724         function figData = dnaBoundProteinDisplacement(figHandle, simBatchDir, selectedSimulations)
<a name="l02725"></a>02725             import edu.stanford.covert.cell.sim.util.PlotUtil;
<a name="l02726"></a>02726             import edu.stanford.covert.cell.sim.util.SimulationEnsemble;
<a name="l02727"></a>02727             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
<a name="l02728"></a>02728             
<a name="l02729"></a>02729             %% get data
<a name="l02730"></a>02730             if nargin &lt; 2 || isempty(simBatchDir)
<a name="l02731"></a>02731                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
<a name="l02732"></a>02732             end
<a name="l02733"></a>02733             if nargin &lt; 3 || isempty(selectedSimulations)
<a name="l02734"></a>02734                 selectedSimulations = SimulationDiskUtil.getCompleteSimulations(simBatchDir);
<a name="l02735"></a>02735             end
<a name="l02736"></a>02736             
<a name="l02737"></a>02737             %% constants
<a name="l02738"></a>02738             nSims = numel(selectedSimulations);
<a name="l02739"></a>02739             [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep &#39;1&#39;]);
<a name="l02740"></a>02740             c = sim.state(&#39;<a class="code" href="class_chromosome.html">Chromosome</a>&#39;);
<a name="l02741"></a>02741             pc = sim.state(&#39;ProteinComplex&#39;);
<a name="l02742"></a>02742             
<a name="l02743"></a>02743             %% get data
<a name="l02744"></a>02744             stateNames = {
<a name="l02745"></a>02745                 <span class="stringliteral">&#39;Chromosome&#39;</span> <span class="stringliteral">&#39;monomerBoundSites&#39;</span>
<a name="l02746"></a>02746                 <span class="stringliteral">&#39;Chromosome&#39;</span> <span class="stringliteral">&#39;complexBoundSites&#39;</span>
<a name="l02747"></a>02747                 };
<a name="l02748"></a>02748             displacements = zeros(numel(c.reactionBoundMonomer), nSims);
<a name="l02749"></a>02749             simEndTimes = zeros(1, nSims);
<a name="l02750"></a>02750             <span class="keywordflow">for</span> i = 1:nSims
<a name="l02751"></a>02751                 states = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, <span class="stringliteral">&#39;extract&#39;</span>, selectedSimulations(i));
<a name="l02752"></a>02752                 
<a name="l02753"></a>02753                 simEndTimes(i) = size(states.Chromosome.monomerBoundSites, 3);
<a name="l02754"></a>02754                 [monPosStrndTimes, monIdxs] = find(states.Chromosome.monomerBoundSites);
<a name="l02755"></a>02755                 [cpxPosStrndTimes, cpxIdxs] = find(states.Chromosome.complexBoundSites);
<a name="l02756"></a>02756                 
<a name="l02757"></a>02757                 rxnIdxs = find(...
<a name="l02758"></a>02758                     (ismember(c.reactionBoundComplex, [pc.dnaPolymeraseIndexs; pc.rnaPolymeraseIndexs]) | ...
<a name="l02759"></a>02759                     any(c.reactionComplexCatalysisMatrix(:, [pc.dnaPolymeraseIndexs; pc.rnaPolymeraseIndexs]), 2)) &amp; ...
<a name="l02760"></a>02760                     (ismember(c.reactionBoundMonomer, monIdxs) | ismember(c.reactionBoundComplex, cpxIdxs)) &amp; ...
<a name="l02761"></a>02761                     (any(c.reactionMonomerCatalysisMatrix(:, unique(monIdxs)), 2) | any(c.reactionComplexCatalysisMatrix(:, unique(cpxIdxs)), 2)) ...
<a name="l02762"></a>02762                     );
<a name="l02763"></a>02763                 <span class="keywordflow">for</span> j = 1:numel(rxnIdxs)
<a name="l02764"></a>02764                     if c.reactionBoundMonomer(rxnIdxs(j))
<a name="l02765"></a>02765                         tfs = monIdxs == c.reactionBoundMonomer(rxnIdxs(j));
<a name="l02766"></a>02766                         boundPosStrdTimes = monPosStrndTimes(tfs, :);
<a name="l02767"></a>02767                         boundFtpt = c.monomerDNAFootprints(c.reactionBoundMonomer(rxnIdxs(j)));
<a name="l02768"></a>02768                     else
<a name="l02769"></a>02769                         tfs = cpxIdxs == c.reactionBoundComplex(rxnIdxs(j));
<a name="l02770"></a>02770                         boundPosStrdTimes = cpxPosStrndTimes(tfs, :);
<a name="l02771"></a>02771                         boundFtpt = c.complexDNAFootprints(c.reactionBoundComplex(rxnIdxs(j)));
<a name="l02772"></a>02772                     end
<a name="l02773"></a>02773                     if isempty(boundPosStrdTimes)
<a name="l02774"></a>02774                         continue;
<a name="l02775"></a>02775                     end
<a name="l02776"></a>02776                     
<a name="l02777"></a>02777                     if any(c.reactionMonomerCatalysisMatrix(rxnIdxs(j), :))
<a name="l02778"></a>02778                         tfs = monIdxs == find(c.reactionMonomerCatalysisMatrix(rxnIdxs(j), :));
<a name="l02779"></a>02779                         bindingPosStrdTimes = monPosStrndTimes(tfs, :);
<a name="l02780"></a>02780                         bindingFtpt = c.monomerDNAFootprints(c.reactionMonomerCatalysisMatrix(rxnIdxs(j), :) ~= 0);
<a name="l02781"></a>02781                     else
<a name="l02782"></a>02782                         tfs = cpxIdxs == find(c.reactionComplexCatalysisMatrix(rxnIdxs(j), :));
<a name="l02783"></a>02783                         bindingPosStrdTimes = cpxPosStrndTimes(tfs, :);
<a name="l02784"></a>02784                         bindingFtpt = c.complexDNAFootprints(c.reactionComplexCatalysisMatrix(rxnIdxs(j), :) ~= 0);
<a name="l02785"></a>02785                     end
<a name="l02786"></a>02786                     if isempty(bindingPosStrdTimes)
<a name="l02787"></a>02787                         continue;
<a name="l02788"></a>02788                     end
<a name="l02789"></a>02789                     
<a name="l02790"></a>02790                     times = intersect(unique(boundPosStrdTimes(:, 3)), unique(bindingPosStrdTimes(:, 3) - 1));
<a name="l02791"></a>02791                     for k = 1:numel(times)
<a name="l02792"></a>02792                         tmpboundPosStrdTimes = boundPosStrdTimes(boundPosStrdTimes(:, 3) == times(k), 1:2);
<a name="l02793"></a>02793                         tmpbindingPosStrdTimes = bindingPosStrdTimes(bindingPosStrdTimes(:, 3) == times(k) + 1, 1:2);
<a name="l02794"></a>02794                         
<a name="l02795"></a>02795                         if numel(tmpboundPosStrdTimes) &lt;= numel(tmpbindingPosStrdTimes)
<a name="l02796"></a>02796                             for l = 1:size(tmpboundPosStrdTimes, 1)
<a name="l02797"></a>02797                                 if ...
<a name="l02798"></a>02798                                         any(tmpbindingPosStrdTimes(tmpbindingPosStrdTimes(:, 1) &gt;= tmpboundPosStrdTimes(l, 1), 1) - tmpboundPosStrdTimes(l, 1) &lt; boundFtpt) || ...
<a name="l02799"></a>02799                                         any(tmpbindingPosStrdTimes(:, 1) + c.sequenceLen - tmpboundPosStrdTimes(l, 1) &lt; boundFtpt) || ...
<a name="l02800"></a>02800                                         any(tmpboundPosStrdTimes(l, 1) - tmpbindingPosStrdTimes(tmpbindingPosStrdTimes(:, 1) &lt; tmpboundPosStrdTimes(l, 1), 1) &lt; bindingFtpt) || ...
<a name="l02801"></a>02801                                         any(tmpboundPosStrdTimes(l, 1) + c.sequenceLen - tmpbindingPosStrdTimes(:, 1) &lt; boundFtpt)
<a name="l02802"></a>02802                                     displacements(rxnIdxs(j), i) = ...
<a name="l02803"></a>02803                                         displacements(rxnIdxs(j), i) + 1;
<a name="l02804"></a>02804                                 end
<a name="l02805"></a>02805                             end
<a name="l02806"></a>02806                         else
<a name="l02807"></a>02807                             for l = 1:size(tmpbindingPosStrdTimes, 1)
<a name="l02808"></a>02808                                 if ...
<a name="l02809"></a>02809                                         any(tmpboundPosStrdTimes(tmpboundPosStrdTimes(:, 1) &gt;= tmpbindingPosStrdTimes(l, 1), 1) - tmpbindingPosStrdTimes(l, 1) &lt; bindingFtpt) || ...
<a name="l02810"></a>02810                                         any(tmpboundPosStrdTimes(:, 1) + c.sequenceLen - tmpbindingPosStrdTimes(l, 1) &lt; bindingFtpt) || ...
<a name="l02811"></a>02811                                         any(tmpbindingPosStrdTimes(l, 1) - tmpboundPosStrdTimes(tmpboundPosStrdTimes(:, 1) &lt; tmpbindingPosStrdTimes(l, 1), 1) &lt; boundFtpt) || ...
<a name="l02812"></a>02812                                         any(tmpbindingPosStrdTimes(l, 1) + c.sequenceLen - tmpboundPosStrdTimes(:, 1) &lt; boundFtpt)
<a name="l02813"></a>02813                                     displacements(rxnIdxs(j), i) = ...
<a name="l02814"></a>02814                                         displacements(rxnIdxs(j), i) + 1;
<a name="l02815"></a>02815                                 end
<a name="l02816"></a>02816                             end
<a name="l02817"></a>02817                         end
<a name="l02818"></a>02818                     end
<a name="l02819"></a>02819                     
<a name="l02820"></a>02820                     clear tfs boundPosStrdTimes boundFtpt bindingPosStrdTimes bindingFtpt tmpboundPosStrdTimes tmpbindingPosStrdTimes;
<a name="l02821"></a>02821                 end
<a name="l02822"></a>02822                 
<a name="l02823"></a>02823                 clear states monPosStrndTimes cpxPosStrndTimes monIdxs cpxIdxs;
<a name="l02824"></a>02824             end
<a name="l02825"></a>02825             
<a name="l02826"></a>02826             %% plot data
<a name="l02827"></a>02827             clf(figHandle);
<a name="l02828"></a>02828             
<a name="l02829"></a>02829             meanDisplacements = mean(displacements ./ simEndTimes(ones(size(displacements, 1), 1), :), 2) * 3600;
<a name="l02830"></a>02830             stdDisplacements = std(displacements ./ simEndTimes(ones(size(displacements, 1), 1), :), [], 2) * 3600;
<a name="l02831"></a>02831             axesHandles = zeros(2, 2);
<a name="l02832"></a>02832             
<a name="l02833"></a>02833             axesHandle = subplot(2, 2, 1);
<a name="l02834"></a>02834             axesHandles(1, 1) = axesHandle;
<a name="l02835"></a>02835             hold(axesHandle, &#39;on&#39;);
<a name="l02836"></a>02836             means = meanDisplacements(any(c.reactionComplexCatalysisMatrix(:, pc.dnaPolymeraseIndexs), 2));
<a name="l02837"></a>02837             stds = stdDisplacements(any(c.reactionComplexCatalysisMatrix(:, pc.dnaPolymeraseIndexs), 2));
<a name="l02838"></a>02838             bar(axesHandle, (1:numel(means))&#39;, means);
<a name="l02839"></a>02839             errorbar((1:numel(means))&#39;, means, stds, &#39;Parent&#39;, axesHandle, &#39;LineStyle&#39;, &#39;none&#39;, &#39;Marker&#39;, &#39;none&#39;);
<a name="l02840"></a>02840             title(axesHandle, &#39;DNA Pol-displaced Proteins&#39;, &#39;FontSize&#39;, 8);
<a name="l02841"></a>02841             xlabel(axesHandle, &#39;Displaced <a class="code" href="class_protein.html">Protein</a>&#39;, &#39;FontSize&#39;, 8);
<a name="l02842"></a>02842             ylabel(axesHandle, &#39;Collisions h^{-1}&#39;, &#39;FontSize&#39;, 8);
<a name="l02843"></a>02843             <span class="keyword">set</span>(axesHandle, <span class="stringliteral">&#39;XTick&#39;</span>, []);
<a name="l02844"></a>02844             ylim(axesHandle, [0 max(means + stds)])
<a name="l02845"></a>02845             
<a name="l02846"></a>02846             axesHandle = subplot(2, 2, 2);
<a name="l02847"></a>02847             axesHandles(1, 2) = axesHandle;
<a name="l02848"></a>02848             hold(axesHandle, &#39;on&#39;);
<a name="l02849"></a>02849             means = meanDisplacements(any(c.reactionComplexCatalysisMatrix(:, pc.rnaPolymeraseIndexs), 2));
<a name="l02850"></a>02850             stds = stdDisplacements(any(c.reactionComplexCatalysisMatrix(:, pc.rnaPolymeraseIndexs), 2));
<a name="l02851"></a>02851             bar(axesHandle, (1:numel(means))&#39;, means);
<a name="l02852"></a>02852             errorbar((1:numel(means))&#39;, means, stds, &#39;Parent&#39;, axesHandle, &#39;LineStyle&#39;, &#39;none&#39;, &#39;Marker&#39;, &#39;none&#39;);
<a name="l02853"></a>02853             title(axesHandle, &#39;<a class="code" href="class_r_n_a.html">RNA</a> Pol-displaced Proteins&#39;, &#39;FontSize&#39;, 8);
<a name="l02854"></a>02854             xlabel(axesHandle, &#39;Displaced Protein&#39;, &#39;FontSize&#39;, 8);
<a name="l02855"></a>02855             ylabel(axesHandle, &#39;Collisions h^{-1}&#39;, &#39;FontSize&#39;, 8);
<a name="l02856"></a>02856             <span class="keyword">set</span>(axesHandle, <span class="stringliteral">&#39;XTick&#39;</span>, []);
<a name="l02857"></a>02857             ylim(axesHandle, [0 max(means + stds)])
<a name="l02858"></a>02858             
<a name="l02859"></a>02859             axesHandle = subplot(2, 2, 3);
<a name="l02860"></a>02860             axesHandles(2, 1) = axesHandle;
<a name="l02861"></a>02861             hold(axesHandle, &#39;on&#39;);
<a name="l02862"></a>02862             means = meanDisplacements(ismember(c.reactionBoundComplex, pc.rnaPolymeraseIndexs));
<a name="l02863"></a>02863             stds = stdDisplacements(ismember(c.reactionBoundComplex, pc.rnaPolymeraseIndexs));
<a name="l02864"></a>02864             bar(axesHandle, (1:numel(means))&#39;, means);
<a name="l02865"></a>02865             errorbar((1:numel(means))&#39;, means, stds, &#39;Parent&#39;, axesHandle, &#39;LineStyle&#39;, &#39;none&#39;, &#39;Marker&#39;, &#39;none&#39;);
<a name="l02866"></a>02866             title(axesHandle, &#39;<a class="code" href="class_r_n_a.html">RNA</a> Pol Displacements&#39;, &#39;FontSize&#39;, 8);
<a name="l02867"></a>02867             xlabel(axesHandle, &#39;Binding Protein&#39;, &#39;FontSize&#39;, 8);
<a name="l02868"></a>02868             ylabel(axesHandle, &#39;Collisions h^{-1}&#39;, &#39;FontSize&#39;, 8);
<a name="l02869"></a>02869             <span class="keyword">set</span>(axesHandle, <span class="stringliteral">&#39;XTick&#39;</span>, []);
<a name="l02870"></a>02870             ylim(axesHandle, [0 max(means + stds)])
<a name="l02871"></a>02871             
<a name="l02872"></a>02872             axesHandle = subplot(2, 2, 4);
<a name="l02873"></a>02873             axesHandles(2, 2) = axesHandle;
<a name="l02874"></a>02874             hold(axesHandle, &#39;on&#39;);
<a name="l02875"></a>02875             tmp = sum(displacements(ismember(c.reactionBoundComplex, pc.rnaPolymeraseIndexs) &amp; any(c.reactionComplexCatalysisMatrix(:, pc.dnaPolymeraseIndexs), 2), :), 1) ./ simEndTimes * 3600;
<a name="l02876"></a>02876             bar(axesHandle, tmp);
<a name="l02877"></a>02877             title(axesHandle, &#39;<a class="code" href="class_r_n_a.html">RNA</a> Pol Displacements by DNA Pol&#39;, &#39;FontSize&#39;, 8);
<a name="l02878"></a>02878             xlabel(axesHandle, &#39;Collisions h^{-1}&#39;, &#39;FontSize&#39;, 8);
<a name="l02879"></a>02879             ylabel(axesHandle, <span class="stringliteral">&#39;Frequency&#39;</span>, <span class="stringliteral">&#39;FontSize&#39;</span>, 8);
<a name="l02880"></a>02880             
<a name="l02881"></a>02881             %y-axes
<a name="l02882"></a>02882             set(axesHandles, &#39;FontSize&#39;, 6);
<a name="l02883"></a>02883             <span class="keyword">set</span>(cell2mat(<span class="keyword">get</span>(axesHandles, <span class="stringliteral">&#39;Ylabel&#39;</span>)), <span class="stringliteral">&#39;Units&#39;</span>, <span class="stringliteral">&#39;normalized&#39;</span>)
<a name="l02884"></a>02884             PlotUtil.alignYAxesLabels(axesHandles(:, 1));
<a name="l02885"></a>02885             PlotUtil.alignYAxesLabels(axesHandles(:, 2));
<a name="l02886"></a>02886             
<a name="l02887"></a>02887             %% organize figure data
<a name="l02888"></a>02888             figData = <span class="keyword">struct</span>;
<a name="l02889"></a>02889             figData.simEndTimes = simEndTimes;
<a name="l02890"></a>02890             figData.displacements = displacements;
<a name="l02891"></a>02891         end
<a name="l02892"></a>02892         
<a name="l02893"></a>02893         <span class="keyword">function</span> figData = macromoleculeHalfLives(figHandle, simBatchDir, selectedSimulations)
<a name="l02894"></a>02894             import edu.stanford.covert.cell.sim.util.PlotUtil;
<a name="l02895"></a>02895             import edu.stanford.covert.cell.sim.util.SimulationEnsemble;
<a name="l02896"></a>02896             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
<a name="l02897"></a>02897             
<a name="l02898"></a>02898             if nargin &lt; 2 || isempty(simBatchDir)
<a name="l02899"></a>02899                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
<a name="l02900"></a>02900             end
<a name="l02901"></a>02901             if nargin &lt; 3 || isempty(selectedSimulations)
<a name="l02902"></a>02902                 selectedSimulations = SimulationDiskUtil.getCompleteSimulations(simBatchDir);
<a name="l02903"></a>02903             end
<a name="l02904"></a>02904             
<a name="l02905"></a>02905             %% constants
<a name="l02906"></a>02906             nSims = numel(selectedSimulations);
<a name="l02907"></a>02907             
<a name="l02908"></a>02908             [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep &#39;1&#39;]);
<a name="l02909"></a>02909             g = sim.gene;
<a name="l02910"></a>02910             rna = sim.state(&#39;Rna&#39;);
<a name="l02911"></a>02911             pm = sim.state(&#39;ProteinMonomer&#39;);
<a name="l02912"></a>02912             pc = sim.state(&#39;ProteinComplex&#39;);
<a name="l02913"></a>02913             
<a name="l02914"></a>02914             pcComp = sum(pc.proteinComplexComposition, 3);
<a name="l02915"></a>02915             
<a name="l02916"></a>02916             expRNAHalfLives = rna.halfLives(rna.matureIndexs) / 60;
<a name="l02917"></a>02917             expMonHalfLives = pm.halfLives(pm.matureIndexs) / 3600;
<a name="l02918"></a>02918             expCpxHalfLives = pc.halfLives(pc.matureIndexs) / 3600;
<a name="l02919"></a>02919             
<a name="l02920"></a>02920             expRNAHalfLives(isinf(expRNAHalfLives)) = NaN;
<a name="l02921"></a>02921             expMonHalfLives(isinf(expMonHalfLives)) = NaN;
<a name="l02922"></a>02922             expCpxHalfLives(isinf(expCpxHalfLives)) = NaN;
<a name="l02923"></a>02923             
<a name="l02924"></a>02924             %% get data
<a name="l02925"></a>02925             rnaCounts = zeros(numel(rna.matureIndexs), nSims);
<a name="l02926"></a>02926             monCounts = zeros(numel(pm.matureIndexs),  nSims);
<a name="l02927"></a>02927             cpxCounts = zeros(numel(pc.matureIndexs),  nSims);
<a name="l02928"></a>02928             rnaDecays = zeros(numel(rna.matureIndexs), nSims);
<a name="l02929"></a>02929             monDecays = zeros(numel(pm.matureIndexs),  nSims);
<a name="l02930"></a>02930             cpxDecays = zeros(numel(pc.matureIndexs),  nSims);
<a name="l02931"></a>02931             
<a name="l02932"></a>02932             stateNames = {
<a name="l02933"></a>02933                 <span class="stringliteral">&#39;Rna&#39;</span>             <span class="stringliteral">&#39;counts&#39;</span>  <span class="charliteral">&#39;:&#39;</span>  <span class="stringliteral">&#39;-sum&#39;</span>
<a name="l02934"></a>02934                 <span class="stringliteral">&#39;ProteinMonomer&#39;</span>  <span class="stringliteral">&#39;counts&#39;</span>  <span class="charliteral">&#39;:&#39;</span>  <span class="stringliteral">&#39;-sum&#39;</span>
<a name="l02935"></a>02935                 <span class="stringliteral">&#39;ProteinComplex&#39;</span>  <span class="stringliteral">&#39;counts&#39;</span>  <span class="charliteral">&#39;:&#39;</span>  <span class="stringliteral">&#39;-sum&#39;</span>
<a name="l02936"></a>02936                 };
<a name="l02937"></a>02937             <span class="keywordflow">for</span> i = 1:nSims
<a name="l02938"></a>02938                 states = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, <span class="stringliteral">&#39;extract&#39;</span>, selectedSimulations(i));
<a name="l02939"></a>02939                 
<a name="l02940"></a>02940                 <span class="keywordflow">if</span> isa(states.Rna.counts, <span class="stringliteral">&#39;edu.stanford.covert.util.SparseMat&#39;</span>)
<a name="l02941"></a>02941                     [subs, vals] = find(states.Rna.counts);
<a name="l02942"></a>02942                     tfs = ...
<a name="l02943"></a>02943                         (subs(:, 1) &gt;= rna.nascentIndexs(1) &amp; subs(:, 1) &lt;= rna.nascentIndexs(end)) | ...
<a name="l02944"></a>02944                         (subs(:, 1) &gt;= rna.intergenicIndexs(1) &amp; subs(:, 1) &lt;= rna.intergenicIndexs(end));
<a name="l02945"></a>02945                     subs = subs(~tfs, :);
<a name="l02946"></a>02946                     vals = vals(~tfs, :);
<a name="l02947"></a>02947                     
<a name="l02948"></a>02948                     subs(subs(:, 1) &gt; rna.intergenicIndexs(end)) = subs(subs(:, 1) &gt; rna.intergenicIndexs(end)) - numel(rna.intergenicIndexs);
<a name="l02949"></a>02949                     subs(subs(:, 1) &gt; rna.nascentIndexs(end)) = subs(subs(:, 1) &gt; rna.nascentIndexs(end)) - numel(rna.nascentIndexs);
<a name="l02950"></a>02950                     
<a name="l02951"></a>02951                     tmp = edu.stanford.covert.util.SparseMat(subs, vals, [6 * numel(rna.matureIndexs) 1 size(states.Rna.counts, 3)]);
<a name="l02952"></a>02952                     states.Rna.counts = full(permute(sum(reshape(tmp, [numel(rna.matureIndexs) 6 size(states.Rna.counts, 3)]), 2), [1 3 2 4]));
<a name="l02953"></a>02953                     
<a name="l02954"></a>02954                     clear tmp tfs subs vals;
<a name="l02955"></a>02955                 <span class="keywordflow">else</span>
<a name="l02956"></a>02956                     states.Rna.counts = permute(...
<a name="l02957"></a>02957                         + states.Rna.counts(rna.processedIndexs, :, :) ...
<a name="l02958"></a>02958                         + states.Rna.counts(rna.matureIndexs, :, :) ...
<a name="l02959"></a>02959                         + states.Rna.counts(rna.aminoacylatedIndexs, :, :) ...
<a name="l02960"></a>02960                         + states.Rna.counts(rna.boundIndexs, :, :) ...
<a name="l02961"></a>02961                         + states.Rna.counts(rna.misfoldedIndexs, :, :) ...
<a name="l02962"></a>02962                         + states.Rna.counts(rna.damagedIndexs, :, :)...
<a name="l02963"></a>02963                         , [1 3 2]);
<a name="l02964"></a>02964                 end
<a name="l02965"></a>02965                 
<a name="l02966"></a>02966                 states.ProteinMonomer.counts = full(permute(sum(reshape(states.ProteinMonomer.counts, ...
<a name="l02967"></a>02967                     [numel(pm.matureIndexs)  numel(pm.wholeCellModelIDs)/numel(pm.matureIndexs)  size(states.ProteinMonomer.counts, 3)]), 2), [1 3 2]));
<a name="l02968"></a>02968                 states.ProteinComplex.counts = full(permute(sum(reshape(states.ProteinComplex.counts, ...
<a name="l02969"></a>02969                     [numel(pc.matureIndexs)  numel(pc.wholeCellModelIDs)/numel(pc.matureIndexs)  size(states.ProteinComplex.counts, 3)]), 2), [1 3 2]));
<a name="l02970"></a>02970                 
<a name="l02971"></a>02971                 tmp = states.Rna.counts;
<a name="l02972"></a>02972                 tmp(setdiff(1:end, rna.matureMRNAIndexs), :) = ...
<a name="l02973"></a>02973                     + tmp(setdiff(1:end, rna.matureMRNAIndexs), :) ...
<a name="l02974"></a>02974                     + pcComp(setdiff(1:end, g.mRNAIndexs), :) * states.ProteinComplex.counts;
<a name="l02975"></a>02975                 rnaCounts(:, i) = sum(tmp(:, 1:end-1), 2);
<a name="l02976"></a>02976                 rnaDecays(:, i) = -sum(min(0, diff(tmp, 1, 2)), 2);
<a name="l02977"></a>02977                 
<a name="l02978"></a>02978                 tmp = ...
<a name="l02979"></a>02979                     + states.ProteinMonomer.counts ...
<a name="l02980"></a>02980                     + pcComp(g.mRNAIndexs, :) * states.ProteinComplex.counts;
<a name="l02981"></a>02981                 monCounts(:, i) = sum(tmp(:, 1:end-1), 2);
<a name="l02982"></a>02982                 monDecays(:, i) = -sum(min(0, diff(tmp, 1, 2)), 2);
<a name="l02983"></a>02983                 
<a name="l02984"></a>02984                 tmp = states.ProteinComplex.counts;
<a name="l02985"></a>02985                 cpxCounts(:, i) = sum(tmp(:, 1:end-1), 2);
<a name="l02986"></a>02986                 cpxDecays(:, i) = -sum(min(0, diff(tmp, 1, 2)), 2);
<a name="l02987"></a>02987                 
<a name="l02988"></a>02988                 clear states simEndTimes tmp;
<a name="l02989"></a>02989             end
<a name="l02990"></a>02990             
<a name="l02991"></a>02991             avgRnaHalfLives = 1 / 60   * log(2) ./ (sum(rnaDecays, 2) ./ sum(rnaCounts, 2));
<a name="l02992"></a>02992             avgMonHalfLives = 1 / 3600 * log(2) ./ (sum(monDecays, 2) ./ sum(monCounts, 2));
<a name="l02993"></a>02993             avgCpxHalfLives = 1 / 3600 * log(2) ./ (sum(cpxDecays, 2) ./ sum(cpxCounts, 2));
<a name="l02994"></a>02994             stdRnaHalfLives = std(1 / 60   * log(2) ./ (rnaDecays ./ rnaCounts), [], 2);
<a name="l02995"></a>02995             stdMonHalfLives = std(1 / 3600 * log(2) ./ (monDecays ./ monCounts), [], 2);
<a name="l02996"></a>02996             stdCpxHalfLives = std(1 / 3600 * log(2) ./ (cpxDecays ./ cpxCounts), [], 2);
<a name="l02997"></a>02997             
<a name="l02998"></a>02998             avgRnaHalfLives(isinf(avgRnaHalfLives) | isnan(expRNAHalfLives)) = NaN;
<a name="l02999"></a>02999             avgMonHalfLives(isinf(avgMonHalfLives) | isnan(expMonHalfLives)) = NaN;
<a name="l03000"></a>03000             avgCpxHalfLives(isinf(avgCpxHalfLives) | isnan(expCpxHalfLives)) = NaN;
<a name="l03001"></a>03001             stdRnaHalfLives(isinf(stdRnaHalfLives) | isnan(expRNAHalfLives)) = NaN;
<a name="l03002"></a>03002             stdMonHalfLives(isinf(stdMonHalfLives) | isnan(expMonHalfLives)) = NaN;
<a name="l03003"></a>03003             stdCpxHalfLives(isinf(stdCpxHalfLives) | isnan(expCpxHalfLives)) = NaN;
<a name="l03004"></a>03004             
<a name="l03005"></a>03005             avgMonHalfLive = nanmean(avgMonHalfLives);
<a name="l03006"></a>03006             avgCpxHalfLive = nanmean(avgCpxHalfLives);
<a name="l03007"></a>03007             
<a name="l03008"></a>03008             clear rnaCounts monCounts cpxCounts rnaDecays monDecays cpxDecays;
<a name="l03009"></a>03009             
<a name="l03010"></a>03010             %% plot data
<a name="l03011"></a>03011             clf(figHandle);
<a name="l03012"></a>03012             [axesHandles, xAxesHandles] = PlotUtil.multiElementPlot(figHandle, {100 100 100}, [0 1], <span class="keyword">struct</span>(...
<a name="l03013"></a>03013                 <span class="stringliteral">&#39;position&#39;</span>, [0.02 0.62 0.95 0.2417]));
<a name="l03014"></a>03014             
<a name="l03015"></a>03015             %<a class="code" href="class_r_n_a.html">RNA</a>
<a name="l03016"></a>03016             axesHandle = axesHandles{1};
<a name="l03017"></a>03017             xAxesHandle = xAxesHandles{1};
<a name="l03018"></a>03018             hold(axesHandle, <span class="stringliteral">&#39;on&#39;</span>);
<a name="l03019"></a>03019             h = zeros(4, 1);
<a name="l03020"></a>03020             h(1) = errorbar(expRNAHalfLives(rna.matureMRNAIndexs), avgRnaHalfLives(rna.matureMRNAIndexs), stdRnaHalfLives(rna.matureMRNAIndexs), ...
<a name="l03021"></a>03021                 <span class="stringliteral">&#39;Parent&#39;</span>, axesHandle, <span class="stringliteral">&#39;LineStyle&#39;</span>, <span class="stringliteral">&#39;none&#39;</span>, <span class="stringliteral">&#39;Color&#39;</span>, <span class="charliteral">&#39;r&#39;</span>, <span class="stringliteral">&#39;Marker&#39;</span>, <span class="charliteral">&#39;.&#39;</span>);
<a name="l03022"></a>03022             h(2) = errorbar(expRNAHalfLives(rna.matureRRNAIndexs), avgRnaHalfLives(rna.matureRRNAIndexs), stdRnaHalfLives(rna.matureRRNAIndexs), ...
<a name="l03023"></a>03023                 <span class="stringliteral">&#39;Parent&#39;</span>, axesHandle, <span class="stringliteral">&#39;LineStyle&#39;</span>, <span class="stringliteral">&#39;none&#39;</span>, <span class="stringliteral">&#39;Color&#39;</span>, <span class="charliteral">&#39;g&#39;</span>, <span class="stringliteral">&#39;Marker&#39;</span>, <span class="charliteral">&#39;.&#39;</span>);
<a name="l03024"></a>03024             h(3) = errorbar(expRNAHalfLives(rna.matureSRNAIndexs), avgRnaHalfLives(rna.matureSRNAIndexs), stdRnaHalfLives(rna.matureSRNAIndexs), ...
<a name="l03025"></a>03025                 <span class="stringliteral">&#39;Parent&#39;</span>, axesHandle, <span class="stringliteral">&#39;LineStyle&#39;</span>, <span class="stringliteral">&#39;none&#39;</span>, <span class="stringliteral">&#39;Color&#39;</span>, <span class="charliteral">&#39;b&#39;</span>, <span class="stringliteral">&#39;Marker&#39;</span>, <span class="charliteral">&#39;.&#39;</span>);
<a name="l03026"></a>03026             h(4) = errorbar(expRNAHalfLives(rna.matureTRNAIndexs), avgRnaHalfLives(rna.matureTRNAIndexs), stdRnaHalfLives(rna.matureTRNAIndexs), ...
<a name="l03027"></a>03027                 <span class="stringliteral">&#39;Parent&#39;</span>, axesHandle, <span class="stringliteral">&#39;LineStyle&#39;</span>, <span class="stringliteral">&#39;none&#39;</span>, <span class="stringliteral">&#39;Color&#39;</span>, <span class="charliteral">&#39;c&#39;</span>, <span class="stringliteral">&#39;Marker&#39;</span>, <span class="charliteral">&#39;.&#39;</span>);
<a name="l03028"></a>03028             minVal = min(min(avgRnaHalfLives - stdRnaHalfLives, expRNAHalfLives - 1));
<a name="l03029"></a>03029             maxVal = max(max(avgRnaHalfLives - stdRnaHalfLives, expRNAHalfLives + 1));
<a name="l03030"></a>03030             xlim(xAxesHandle, [minVal maxVal]);
<a name="l03031"></a>03031             xlim(axesHandle, [minVal maxVal]);
<a name="l03032"></a>03032             ylim(axesHandle, [minVal maxVal]);
<a name="l03033"></a>03033             title(axesHandle, <span class="stringliteral">&#39;RNAs&#39;</span>, <span class="stringliteral">&#39;FontSize&#39;</span>, 12);
<a name="l03034"></a>03034             xlabel(xAxesHandle, <span class="stringliteral">&#39;ExpHalf Life (min)&#39;</span>, <span class="stringliteral">&#39;FontSize&#39;</span>, 8);
<a name="l03035"></a>03035             ylabel(axesHandle, <span class="stringliteral">&#39;Sim Half Life (min)&#39;</span>, <span class="stringliteral">&#39;FontSize&#39;</span>, 8);
<a name="l03036"></a>03036             <span class="keyword">set</span>(axesHandle, <span class="stringliteral">&#39;FontSize&#39;</span>, 6);
<a name="l03037"></a>03037             <span class="keyword">set</span>(axesHandle, <span class="stringliteral">&#39;YTick&#39;</span>, <span class="keyword">get</span>(xAxesHandle, <span class="stringliteral">&#39;XTick&#39;</span>));
<a name="l03038"></a>03038             legend(h, {<span class="charliteral">&#39;m&#39;</span>, <span class="charliteral">&#39;r&#39;</span>, <span class="charliteral">&#39;s&#39;</span>, <span class="charliteral">&#39;t&#39;</span>}, <span class="stringliteral">&#39;Location&#39;</span>, <span class="stringliteral">&#39;SouthEast&#39;</span>);
<a name="l03039"></a>03039             
<a name="l03040"></a>03040             %Protein Monomers
<a name="l03041"></a>03041             axesHandle = axesHandles{2};
<a name="l03042"></a>03042             xAxesHandle = xAxesHandles{2};
<a name="l03043"></a>03043             [~, order] = sort(avgMonHalfLives);
<a name="l03044"></a>03044             avgMonHalfLives = avgMonHalfLives(order);
<a name="l03045"></a>03045             stdMonHalfLives = stdMonHalfLives(order);
<a name="l03046"></a>03046             expMonHalfLives = expMonHalfLives(order);
<a name="l03047"></a>03047             hold(axesHandle, <span class="stringliteral">&#39;on&#39;</span>);
<a name="l03048"></a>03048             line([0.5 numel(avgMonHalfLives)+0.5], expMonHalfLives([1 1]), <span class="stringliteral">&#39;Parent&#39;</span>, axesHandle, <span class="stringliteral">&#39;Color&#39;</span>, <span class="charliteral">&#39;r&#39;</span>);
<a name="l03049"></a>03049             line([0.5 numel(avgMonHalfLives)+0.5], avgMonHalfLive([1 1]), <span class="stringliteral">&#39;Parent&#39;</span>, axesHandle, <span class="stringliteral">&#39;Color&#39;</span>, <span class="charliteral">&#39;g&#39;</span>);
<a name="l03050"></a>03050             idxs = find(~isnan(stdMonHalfLives));
<a name="l03051"></a>03051             errorbar(idxs, avgMonHalfLives(idxs), stdMonHalfLives(idxs), ...
<a name="l03052"></a>03052                 <span class="stringliteral">&#39;Parent&#39;</span>, axesHandle, ...
<a name="l03053"></a>03053                 <span class="stringliteral">&#39;LineStyle&#39;</span>, <span class="stringliteral">&#39;none&#39;</span>, ...
<a name="l03054"></a>03054                 <span class="stringliteral">&#39;Marker&#39;</span>, <span class="charliteral">&#39;.&#39;</span>, ...
<a name="l03055"></a>03055                 <span class="stringliteral">&#39;Color&#39;</span>, <span class="charliteral">&#39;b&#39;</span>);
<a name="l03056"></a>03056             idxs = find(isnan(stdMonHalfLives) &amp; ~isnan(avgMonHalfLives));
<a name="l03057"></a>03057             plot(axesHandle, idxs, avgMonHalfLives(idxs), <span class="stringliteral">&#39;b.&#39;</span>);
<a name="l03058"></a>03058             minVal = max(0, min(min(avgMonHalfLives - stdMonHalfLives, expMonHalfLives - 1)));
<a name="l03059"></a>03059             maxVal = max(max(avgMonHalfLives - stdMonHalfLives, expMonHalfLives + 1));
<a name="l03060"></a>03060             xlim(xAxesHandle, [0.5 numel(avgMonHalfLives)+0.5]);
<a name="l03061"></a>03061             xlim(axesHandle, [0.5 numel(avgMonHalfLives)+0.5]);
<a name="l03062"></a>03062             ylim(axesHandle, [minVal maxVal]);
<a name="l03063"></a>03063             title(axesHandle, <span class="stringliteral">&#39;Monomers&#39;</span>, <span class="stringliteral">&#39;FontSize&#39;</span>, 12);
<a name="l03064"></a>03064             xlabel(xAxesHandle, <span class="stringliteral">&#39;Monomer&#39;</span>, <span class="stringliteral">&#39;FontSize&#39;</span>, 8);
<a name="l03065"></a>03065             ylabel(axesHandle, <span class="stringliteral">&#39;Sim Half Life (h)&#39;</span>, <span class="stringliteral">&#39;FontSize&#39;</span>, 8);
<a name="l03066"></a>03066             <span class="keyword">set</span>(axesHandle, <span class="stringliteral">&#39;FontSize&#39;</span>, 6);
<a name="l03067"></a>03067             
<a name="l03068"></a>03068             %Protein Complexes
<a name="l03069"></a>03069             axesHandle = axesHandles{3};
<a name="l03070"></a>03070             xAxesHandle = xAxesHandles{3};
<a name="l03071"></a>03071             [~, order] = sort(avgCpxHalfLives);
<a name="l03072"></a>03072             avgCpxHalfLives = avgCpxHalfLives(order);
<a name="l03073"></a>03073             stdCpxHalfLives = stdCpxHalfLives(order);
<a name="l03074"></a>03074             expCpxHalfLives = expCpxHalfLives(order);
<a name="l03075"></a>03075             hold(axesHandle, <span class="stringliteral">&#39;on&#39;</span>);
<a name="l03076"></a>03076             line([0.5 numel(avgCpxHalfLives)+0.5], expCpxHalfLives([1 1]), <span class="stringliteral">&#39;Parent&#39;</span>, axesHandle, <span class="stringliteral">&#39;Color&#39;</span>, <span class="charliteral">&#39;r&#39;</span>);
<a name="l03077"></a>03077             line([0.5 numel(avgCpxHalfLives)+0.5], avgCpxHalfLive([1 1]), <span class="stringliteral">&#39;Parent&#39;</span>, axesHandle, <span class="stringliteral">&#39;Color&#39;</span>, <span class="charliteral">&#39;g&#39;</span>);
<a name="l03078"></a>03078             idxs = find(~isnan(stdCpxHalfLives));
<a name="l03079"></a>03079             errorbar(idxs, avgCpxHalfLives(idxs), stdCpxHalfLives(idxs), ...
<a name="l03080"></a>03080                 <span class="stringliteral">&#39;Parent&#39;</span>, axesHandle, ...
<a name="l03081"></a>03081                 <span class="stringliteral">&#39;LineStyle&#39;</span>, <span class="stringliteral">&#39;none&#39;</span>, ...
<a name="l03082"></a>03082                 <span class="stringliteral">&#39;Marker&#39;</span>, <span class="charliteral">&#39;.&#39;</span>, ...
<a name="l03083"></a>03083                 <span class="stringliteral">&#39;Color&#39;</span>, <span class="charliteral">&#39;b&#39;</span>);
<a name="l03084"></a>03084             idxs = find(isnan(stdCpxHalfLives) &amp; ~isnan(avgCpxHalfLives));
<a name="l03085"></a>03085             plot(axesHandle, idxs, avgCpxHalfLives(idxs), <span class="stringliteral">&#39;b.&#39;</span>);
<a name="l03086"></a>03086             minVal = max(0, min(min(avgCpxHalfLives - stdCpxHalfLives, expCpxHalfLives - 1)));
<a name="l03087"></a>03087             maxVal = max(max(avgCpxHalfLives - stdCpxHalfLives, expCpxHalfLives + 1));
<a name="l03088"></a>03088             xlim(xAxesHandle, [0.5 numel(avgCpxHalfLives)+0.5]);
<a name="l03089"></a>03089             xlim(axesHandle, [0.5 numel(avgCpxHalfLives)+0.5]);
<a name="l03090"></a>03090             ylim(axesHandle, [minVal maxVal]);
<a name="l03091"></a>03091             title(axesHandle, <span class="stringliteral">&#39;Complexes&#39;</span>, <span class="stringliteral">&#39;FontSize&#39;</span>, 12);
<a name="l03092"></a>03092             xlabel(xAxesHandle, <span class="stringliteral">&#39;Complex&#39;</span>, <span class="stringliteral">&#39;FontSize&#39;</span>, 8);
<a name="l03093"></a>03093             ylabel(axesHandle, <span class="stringliteral">&#39;Sim Half Life (h)&#39;</span>, <span class="stringliteral">&#39;FontSize&#39;</span>, 8);
<a name="l03094"></a>03094             <span class="keyword">set</span>(axesHandle, <span class="stringliteral">&#39;FontSize&#39;</span>, 6);
<a name="l03095"></a>03095             
<a name="l03096"></a>03096             %align
<a name="l03097"></a>03097             PlotUtil.offsetYAxes(axesHandles, 0.04);
<a name="l03098"></a>03098             PlotUtil.labelSubplots(cell2mat(axesHandles));
<a name="l03099"></a>03099             
<a name="l03100"></a>03100             %% organize figure data
<a name="l03101"></a>03101             figData = <span class="keyword">struct</span>;
<a name="l03102"></a>03102             figData.expRNAHalfLives = expRNAHalfLives;
<a name="l03103"></a>03103             figData.avgRnaHalfLives = avgRnaHalfLives;
<a name="l03104"></a>03104             figData.stdRnaHalfLives = stdRnaHalfLives;
<a name="l03105"></a>03105             figData.avgMonHalfLives = avgMonHalfLives;
<a name="l03106"></a>03106             figData.stdMonHalfLives = stdMonHalfLives;
<a name="l03107"></a>03107             figData.expMonHalfLives = expMonHalfLives;
<a name="l03108"></a>03108             figData.avgCpxHalfLives = avgCpxHalfLives;
<a name="l03109"></a>03109             figData.stdCpxHalfLives = stdCpxHalfLives;
<a name="l03110"></a>03110             figData.expCpxHalfLives = expCpxHalfLives;
<a name="l03111"></a>03111         end
<a name="l03112"></a>03112         
<a name="l03113"></a>03113         <span class="keyword">function</span> figData = rnaSynthesisDuration(figHandle, simBatchDir, selectedSimulations)
<a name="l03114"></a>03114             import edu.stanford.covert.cell.sim.util.PlotUtil;
<a name="l03115"></a>03115             import edu.stanford.covert.cell.sim.util.SimulationEnsemble;
<a name="l03116"></a>03116             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
<a name="l03117"></a>03117             
<a name="l03118"></a>03118             if nargin &lt; 2 || isempty(simBatchDir)
<a name="l03119"></a>03119                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
<a name="l03120"></a>03120             end
<a name="l03121"></a>03121             if nargin &lt; 3 || isempty(selectedSimulations)
<a name="l03122"></a>03122                 selectedSimulations = SimulationDiskUtil.getCompleteSimulations(simBatchDir);
<a name="l03123"></a>03123             end
<a name="l03124"></a>03124             
<a name="l03125"></a>03125             %% constants
<a name="l03126"></a>03126             [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep &#39;1&#39;]);
<a name="l03127"></a>03127             g = sim.gene;
<a name="l03128"></a>03128             comp = sim.compartment;
<a name="l03129"></a>03129             c = sim.state(&#39;<a class="code" href="class_chromosome.html">Chromosome</a>&#39;);
<a name="l03130"></a>03130             r = sim.state(&#39;Rna&#39;);
<a name="l03131"></a>03131             pc = sim.state(&#39;ProteinComplex&#39;);
<a name="l03132"></a>03132             rnaPol = sim.state(&#39;RNAPolymerase&#39;);
<a name="l03133"></a>03133             transcript = sim.state(&#39;Transcript&#39;);
<a name="l03134"></a>03134             
<a name="l03135"></a>03135             nSims = numel(selectedSimulations);
<a name="l03136"></a>03136             pcComp = sum(pc.proteinComplexComposition(setdiff(1:end, g.mRNAIndexs), :, :), 3);
<a name="l03137"></a>03137             
<a name="l03138"></a>03138             %% get data
<a name="l03139"></a>03139             stateNames = {
<a name="l03140"></a>03140                 <span class="stringliteral">&#39;Rna&#39;</span>             <span class="stringliteral">&#39;counts&#39;</span>                   <span class="charliteral">&#39;:&#39;</span>  comp.cytosolIndexs
<a name="l03141"></a>03141                 <span class="stringliteral">&#39;Transcript&#39;</span>      <span class="stringliteral">&#39;boundTranscriptionUnits&#39;</span>  <span class="charliteral">&#39;:&#39;</span>  <span class="charliteral">&#39;:&#39;</span>
<a name="l03142"></a>03142                 <span class="stringliteral">&#39;Transcript&#39;</span>      <span class="stringliteral">&#39;boundTranscriptProgress&#39;</span>  <span class="charliteral">&#39;:&#39;</span>  <span class="charliteral">&#39;:&#39;</span>
<a name="l03143"></a>03143                 <span class="stringliteral">&#39;RNAPolymerase&#39;</span>   <span class="stringliteral">&#39;positionStrands&#39;</span>          <span class="charliteral">&#39;:&#39;</span>  <span class="charliteral">&#39;:&#39;</span>
<a name="l03144"></a>03144                 <span class="stringliteral">&#39;RNAPolymerase&#39;</span>   <span class="stringliteral">&#39;states&#39;</span>                   <span class="charliteral">&#39;:&#39;</span>  <span class="charliteral">&#39;:&#39;</span>
<a name="l03145"></a>03145                 <span class="stringliteral">&#39;Ribosome&#39;</span>        <span class="stringliteral">&#39;boundMRNAs&#39;</span>               <span class="charliteral">&#39;:&#39;</span>  <span class="charliteral">&#39;:&#39;</span>
<a name="l03146"></a>03146                 <span class="stringliteral">&#39;ProteinComplex&#39;</span>  <span class="stringliteral">&#39;counts&#39;</span>                   <span class="charliteral">&#39;:&#39;</span>  <span class="stringliteral">&#39;-sum&#39;</span>
<a name="l03147"></a>03147                 };
<a name="l03148"></a>03148             rnaSynthesisTimes = zeros(numel(r.matureIndexs), 3, nSims);
<a name="l03149"></a>03149             mrnaFracBounds = zeros(numel(g.mRNAIndexs), 1, nSims);
<a name="l03150"></a>03150             rnaFracAminoacylated = zeros(numel(r.matureIndexs), 1, nSims);
<a name="l03151"></a>03151             rnaPauses = zeros(4, 1000);
<a name="l03152"></a>03152             <span class="keywordflow">for</span> i = 1:nSims
<a name="l03153"></a>03153                 states = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, <span class="stringliteral">&#39;extract&#39;</span>, selectedSimulations(i));
<a name="l03154"></a>03154                 
<a name="l03155"></a>03155                 states.RNAPolymerase.states = permute(states.RNAPolymerase.states, [1 3 2]);
<a name="l03156"></a>03156                 states.RNAPolymerase.positionStrands = permute(states.RNAPolymerase.positionStrands, [1 3 2]);
<a name="l03157"></a>03157                 states.Transcript.boundTranscriptionUnits = permute(states.Transcript.boundTranscriptionUnits, [1 3 2]);
<a name="l03158"></a>03158                 states.Transcript.boundTranscriptProgress = permute(states.Transcript.boundTranscriptProgress, [1 3 2]);
<a name="l03159"></a>03159                 states.Ribosome.boundMRNAs = permute(states.Ribosome.boundMRNAs, [1 3 2]);
<a name="l03160"></a>03160                 
<a name="l03161"></a>03161                 states.Rna.counts = permute(states.Rna.counts, [1 3 2]);
<a name="l03162"></a>03162                 
<a name="l03163"></a>03163                 nascs = full(states.Rna.counts(r.nascentIndexs, :));
<a name="l03164"></a>03164                 procs = full(states.Rna.counts(r.processedIndexs, :));
<a name="l03165"></a>03165                 matrs = full(...
<a name="l03166"></a>03166                     + states.Rna.counts(r.matureIndexs, :) ...
<a name="l03167"></a>03167                     + states.Rna.counts(r.boundIndexs, :) ...
<a name="l03168"></a>03168                     + states.Rna.counts(r.misfoldedIndexs, :) ...
<a name="l03169"></a>03169                     + states.Rna.counts(r.damagedIndexs, :) ...
<a name="l03170"></a>03170                     + states.Rna.counts(r.aminoacylatedIndexs, :));
<a name="l03171"></a>03171                 matrs(setdiff(1:end, r.matureMRNAIndexs), :) = ...
<a name="l03172"></a>03172                     + matrs(setdiff(1:end, r.matureMRNAIndexs), :) ...
<a name="l03173"></a>03173                     + pcComp * full(permute(sum(reshape(states.ProteinComplex.counts, ...
<a name="l03174"></a>03174                     [numel(pc.matureIndexs)  numel(pc.wholeCellModelIDs)/numel(pc.matureIndexs)  size(states.ProteinComplex.counts, 3)]), 2), [1 3 2]));
<a name="l03175"></a>03175                 amincs = full(sum(states.Rna.counts(r.aminoacylatedIndexs, :), 2));
<a name="l03176"></a>03176                 
<a name="l03177"></a>03177                 %transcription times
<a name="l03178"></a>03178                 [startTimes, startPols] = find(cat(2, ...
<a name="l03179"></a>03179                     states.RNAPolymerase.states(:, 1) &gt;= rnaPol.activelyTranscribingValue, ...
<a name="l03180"></a>03180                     states.RNAPolymerase.states(:, 2:end)   &gt;= rnaPol.activelyTranscribingValue &amp; ...
<a name="l03181"></a>03181                     states.RNAPolymerase.states(:, 1:end-1) &lt; rnaPol.activelyTranscribingValue)<span class="stringliteral">&#39;);</span>
<a name="l03182"></a>03182 <span class="stringliteral">                [endTimes, endPols] = find(cat(2, ...</span>
<a name="l03183"></a>03183 <span class="stringliteral">                    states.RNAPolymerase.states(:, 1:end-1) &gt;= rnaPol.activelyTranscribingValue &amp; ...</span>
<a name="l03184"></a>03184 <span class="stringliteral">                    states.RNAPolymerase.states(:, 2:end)   &lt; rnaPol.activelyTranscribingValue, ...</span>
<a name="l03185"></a>03185 <span class="stringliteral">                    states.RNAPolymerase.states(:, end) &gt;= rnaPol.activelyTranscribingValue)&#39;</span>);
<a name="l03186"></a>03186                 assert(all(startPols == endPols));
<a name="l03187"></a>03187                 assert(all(endTimes &gt;= startTimes));
<a name="l03188"></a>03188                 boundTUs = states.Transcript.boundTranscriptionUnits(sub2ind(...
<a name="l03189"></a>03189                     size(states.Transcript.boundTranscriptionUnits), ...
<a name="l03190"></a>03190                     startPols, startTimes));
<a name="l03191"></a>03191                 rnaSynthesisTimes(1:numel(r.nascentIndexs), 1, i) = ...
<a name="l03192"></a>03192                     accumarray(boundTUs, endTimes - startTimes + 1, [numel(r.nascentIndexs) 1]) ./ histc(boundTUs, (1:numel(r.nascentIndexs))<span class="stringliteral">&#39;);</span>
<a name="l03193"></a>03193 <span class="stringliteral">                </span>
<a name="l03194"></a>03194 <span class="stringliteral">                %transcription pauses</span>
<a name="l03195"></a>03195 <span class="stringliteral">                pause = states.RNAPolymerase.states(:, 1:end-1) &gt;= rnaPol.activelyTranscribingValue &amp; diff(states.Transcript.boundTranscriptProgress, 1, 2) == 0;</span>
<a name="l03196"></a>03196 <span class="stringliteral">                [startTimes, startPols] = find([pause(:, 1)  diff(pause, 1, 2) &gt; 0]&#39;</span>);
<a name="l03197"></a>03197                 [endTimes, endPols] = find([diff(pause, 1, 2) &lt; 0  pause(:, end)]<span class="stringliteral">&#39;);</span>
<a name="l03198"></a>03198 <span class="stringliteral">                assert(all(startPols == endPols));</span>
<a name="l03199"></a>03199 <span class="stringliteral">                assert(all(endTimes &gt;= startTimes));</span>
<a name="l03200"></a>03200 <span class="stringliteral">                </span>
<a name="l03201"></a>03201 <span class="stringliteral">                [~, bases] = ismember(c.sequence.subsequence(states.RNAPolymerase.positionStrands(sub2ind(...</span>
<a name="l03202"></a>03202 <span class="stringliteral">                    size(states.RNAPolymerase.positionStrands), ...</span>
<a name="l03203"></a>03203 <span class="stringliteral">                    [startPols startPols], ...</span>
<a name="l03204"></a>03204 <span class="stringliteral">                    [startTimes startTimes], ...</span>
<a name="l03205"></a>03205 <span class="stringliteral">                    [ones(size(startPols, 1), 1) 2 * ones(size(startPols, 1), 1)]))), &#39;</span>ACGT<span class="stringliteral">&#39;);</span>
<a name="l03206"></a>03206 <span class="stringliteral">                </span>
<a name="l03207"></a>03207 <span class="stringliteral">                if max(endTimes - startTimes + 1) &gt; size(rnaPauses, 2)</span>
<a name="l03208"></a>03208 <span class="stringliteral">                    rnaPauses = [rnaPauses zeros(4, max(endTimes - startTimes + 1) - size(rnaPauses, 2))]; %#ok&lt;AGROW&gt;</span>
<a name="l03209"></a>03209 <span class="stringliteral">                end</span>
<a name="l03210"></a>03210 <span class="stringliteral">                rnaPauses = ...</span>
<a name="l03211"></a>03211 <span class="stringliteral">                    + rnaPauses...</span>
<a name="l03212"></a>03212 <span class="stringliteral">                    + accumarray([bases  endTimes - startTimes + 1], ones(size(startTimes)), [4 size(rnaPauses, 2)]);</span>
<a name="l03213"></a>03213 <span class="stringliteral">                </span>
<a name="l03214"></a>03214 <span class="stringliteral">                %processing time</span>
<a name="l03215"></a>03215 <span class="stringliteral">                rnaSynthesisTimes(1:numel(r.nascentIndexs), 2, i) = log(2) ./ (sum(max(0, -diff(nascs, 1, 2)), 2) ./ sum(nascs(:, 1:end-1), 2));</span>
<a name="l03216"></a>03216 <span class="stringliteral">                </span>
<a name="l03217"></a>03217 <span class="stringliteral">                %modification time</span>
<a name="l03218"></a>03218 <span class="stringliteral">                rnaSynthesisTimes(:, 3, i) = log(2) ./ (sum(max(0, -diff(procs, 1, 2)), 2) ./ sum(procs(:, 1:end-1), 2));</span>
<a name="l03219"></a>03219 <span class="stringliteral">                </span>
<a name="l03220"></a>03220 <span class="stringliteral">                %fraction time bound</span>
<a name="l03221"></a>03221 <span class="stringliteral">                [~, j, val] = find(states.Ribosome.boundMRNAs);</span>
<a name="l03222"></a>03222 <span class="stringliteral">                tmp = zeros(numel(g.mRNAIndexs), size(states.Ribosome.boundMRNAs, 2));</span>
<a name="l03223"></a>03223 <span class="stringliteral">                tmp(sub2ind(size(tmp), val, j)) = 1;</span>
<a name="l03224"></a>03224 <span class="stringliteral">                mrnaFracBounds(:, 1, i) = sum(tmp, 2) ./ sum(r.matureRNAGeneComposition(g.mRNAIndexs, :) * matrs &gt; 0, 2);</span>
<a name="l03225"></a>03225 <span class="stringliteral">                </span>
<a name="l03226"></a>03226 <span class="stringliteral">                %fraction aminoacylated bound</span>
<a name="l03227"></a>03227 <span class="stringliteral">                rnaFracAminoacylated(:, 1, i) = amincs ./ sum(matrs, 2);</span>
<a name="l03228"></a>03228 <span class="stringliteral">                </span>
<a name="l03229"></a>03229 <span class="stringliteral">                clear states nascs procs matrs amincs startTimes startPols endTimes endPols boundTUs pause j val tmp;</span>
<a name="l03230"></a>03230 <span class="stringliteral">            end</span>
<a name="l03231"></a>03231 <span class="stringliteral">            </span>
<a name="l03232"></a>03232 <span class="stringliteral">            rnaSynthesisTimes(isinf(rnaSynthesisTimes)) = NaN;</span>
<a name="l03233"></a>03233 <span class="stringliteral">            mrnaFracBounds(isinf(mrnaFracBounds)) = NaN;</span>
<a name="l03234"></a>03234 <span class="stringliteral">            rnaFracAminoacylated(isinf(rnaFracAminoacylated)) = NaN;</span>
<a name="l03235"></a>03235 <span class="stringliteral">            </span>
<a name="l03236"></a>03236 <span class="stringliteral">            %% plot</span>
<a name="l03237"></a>03237 <span class="stringliteral">            clf(figHandle);</span>
<a name="l03238"></a>03238 <span class="stringliteral">            </span>
<a name="l03239"></a>03239 <span class="stringliteral">            %synthesis times</span>
<a name="l03240"></a>03240 <span class="stringliteral">            axesHandle = subplot(3, 2, 1, &#39;</span>Parent<span class="stringliteral">&#39;, figHandle);</span>
<a name="l03241"></a>03241 <span class="stringliteral">            axesHandles(1, 1) = axesHandle;</span>
<a name="l03242"></a>03242 <span class="stringliteral">            tmp = [zeros(numel(r.matureIndexs), 2)  nanmean(rnaSynthesisTimes(:, 3, :), 3)];</span>
<a name="l03243"></a>03243 <span class="stringliteral">            for i = 1:numel(r.nascentIndexs)</span>
<a name="l03244"></a>03244 <span class="stringliteral">                tmpTfs = r.nascentRNAMatureRNAComposition(:, i) == 1;</span>
<a name="l03245"></a>03245 <span class="stringliteral">                tmp(tmpTfs, 1) = nanmean(rnaSynthesisTimes(i, 1, :), 3);</span>
<a name="l03246"></a>03246 <span class="stringliteral">                tmp(tmpTfs, 2) = nanmean(rnaSynthesisTimes(i, 2, :), 3);</span>
<a name="l03247"></a>03247 <span class="stringliteral">            end</span>
<a name="l03248"></a>03248 <span class="stringliteral">            [~, order] = sort(nansum(tmp, 2));</span>
<a name="l03249"></a>03249 <span class="stringliteral">            h = bar(axesHandle, 1:numel(r.matureIndexs),  tmp(order, :), &#39;</span>stacked<span class="stringliteral">&#39;);</span>
<a name="l03250"></a>03250 <span class="stringliteral">            xlim(axesHandle, [0.5 numel(r.matureIndexs) + 0.5]);</span>
<a name="l03251"></a>03251 <span class="stringliteral">            ylim(axesHandle, [0 max(nansum(tmp, 2))]);</span>
<a name="l03252"></a>03252 <span class="stringliteral">            legend(h, {&#39;</span><a class="code" href="class_transcription.html">Transcription</a><span class="stringliteral">&#39;, &#39;</span>Processing<span class="stringliteral">&#39;, &#39;</span>Modification<span class="stringliteral">&#39;}, &#39;</span>Location<span class="stringliteral">&#39;, &#39;</span>NorthWest<span class="stringliteral">&#39;);</span>
<a name="l03253"></a>03253 <span class="stringliteral">            xlabel(axesHandle, &#39;</span><a class="code" href="class_r_n_a.html">RNA</a><span class="stringliteral">&#39;, &#39;</span>FontSize<span class="stringliteral">&#39;, 8);</span>
<a name="l03254"></a>03254 <span class="stringliteral">            ylabel(axesHandle, {&#39;</span>Synthesis<span class="charliteral">&#39; &#39;</span>Duration (s)<span class="stringliteral">&#39;}, &#39;</span>FontSize<span class="stringliteral">&#39;, 8);</span>
<a name="l03255"></a>03255 <span class="stringliteral">            </span>
<a name="l03256"></a>03256 <span class="stringliteral">            %transcription time vs length</span>
<a name="l03257"></a>03257 <span class="stringliteral">            axesHandle = subplot(3, 2, 3, &#39;</span>Parent<span class="stringliteral">&#39;, figHandle);</span>
<a name="l03258"></a>03258 <span class="stringliteral">            axesHandles(2, 1) = axesHandle;</span>
<a name="l03259"></a>03259 <span class="stringliteral">            meanVal = nanmean(rnaSynthesisTimes(1:numel(r.nascentIndexs), 1, :), 3);</span>
<a name="l03260"></a>03260 <span class="stringliteral">            stdVal = nanstd(rnaSynthesisTimes(1:numel(r.nascentIndexs), 1, :), [], 3);</span>
<a name="l03261"></a>03261 <span class="stringliteral">            errorbar(r.lengths(r.nascentIndexs), meanVal, stdVal, ...</span>
<a name="l03262"></a>03262 <span class="stringliteral">                &#39;</span>LineStyle<span class="stringliteral">&#39;, &#39;</span>none<span class="stringliteral">&#39;, &#39;</span>Parent<span class="stringliteral">&#39;, axesHandle);</span>
<a name="l03263"></a>03263 <span class="stringliteral">            xlim(axesHandle, [min(r.lengths(r.nascentIndexs)) max(r.lengths(r.nascentIndexs))]);</span>
<a name="l03264"></a>03264 <span class="stringliteral">            ylim(axesHandle, [min(meanVal - stdVal) max(meanVal + stdVal)]);</span>
<a name="l03265"></a>03265 <span class="stringliteral">            xlabel(axesHandle, &#39;</span>Length<span class="stringliteral">&#39;, &#39;</span>FontSize<span class="stringliteral">&#39;, 8);</span>
<a name="l03266"></a>03266 <span class="stringliteral">            ylabel(axesHandle, {&#39;</span><a class="code" href="class_transcription.html">Transcription</a><span class="charliteral">&#39; &#39;</span>Duration (s)<span class="stringliteral">&#39;}, &#39;</span>FontSize<span class="stringliteral">&#39;, 8)</span>
<a name="l03267"></a>03267 <span class="stringliteral">            </span>
<a name="l03268"></a>03268 <span class="stringliteral">            %transcription time vs G/C content</span>
<a name="l03269"></a>03269 <span class="stringliteral">            axesHandle = subplot(3, 2, 5, &#39;</span>Parent<span class="stringliteral">&#39;, figHandle);</span>
<a name="l03270"></a>03270 <span class="stringliteral">            axesHandles(3, 1) = axesHandle;</span>
<a name="l03271"></a>03271 <span class="stringliteral">            gcContent = cellfun(@(seq) sum(ismember(seq, &#39;</span>GC<span class="stringliteral">&#39;)) / numel(seq), transcript.transcriptionUnitSequences(r.nascentIndexs));</span>
<a name="l03272"></a>03272 <span class="stringliteral">            meanVal = nanmean(rnaSynthesisTimes(1:numel(r.nascentIndexs), 1, :), 3);</span>
<a name="l03273"></a>03273 <span class="stringliteral">            stdVal = nanstd(rnaSynthesisTimes(1:numel(r.nascentIndexs), 1, :), [], 3);</span>
<a name="l03274"></a>03274 <span class="stringliteral">            errorbar(gcContent, ...</span>
<a name="l03275"></a>03275 <span class="stringliteral">                nanmean(rnaSynthesisTimes(1:numel(r.nascentIndexs), 1, :), 3), ...</span>
<a name="l03276"></a>03276 <span class="stringliteral">                nanstd(rnaSynthesisTimes(1:numel(r.nascentIndexs), 1, :), [], 3), ...</span>
<a name="l03277"></a>03277 <span class="stringliteral">                &#39;</span>LineStyle<span class="stringliteral">&#39;, &#39;</span>none<span class="stringliteral">&#39;, &#39;</span>Parent<span class="stringliteral">&#39;, axesHandle);</span>
<a name="l03278"></a>03278 <span class="stringliteral">            xlim(axesHandle, [min(gcContent) max(gcContent)]);</span>
<a name="l03279"></a>03279 <span class="stringliteral">            ylim(axesHandle, [min(meanVal - stdVal) max(meanVal + stdVal)]);</span>
<a name="l03280"></a>03280 <span class="stringliteral">            xlabel(axesHandle, &#39;</span>G/C Content<span class="stringliteral">&#39;, &#39;</span>FontSize<span class="stringliteral">&#39;, 8);</span>
<a name="l03281"></a>03281 <span class="stringliteral">            ylabel(axesHandle, {&#39;</span><a class="code" href="class_transcription.html">Transcription</a><span class="charliteral">&#39; &#39;</span>Duration<span class="stringliteral">&#39;}, &#39;</span>FontSize<span class="stringliteral">&#39;, 8)</span>
<a name="l03282"></a>03282 <span class="stringliteral">            </span>
<a name="l03283"></a>03283 <span class="stringliteral">            %pauses</span>
<a name="l03284"></a>03284 <span class="stringliteral">            axesHandle = subplot(3, 2, 2, &#39;</span>Parent<span class="stringliteral">&#39;, figHandle);</span>
<a name="l03285"></a>03285 <span class="stringliteral">            axesHandles(1, 2) = axesHandle;</span>
<a name="l03286"></a>03286 <span class="stringliteral">            tmp = rnaPauses ./ repmat(sum(rnaPauses, 2), 1, size(rnaPauses, 2));</span>
<a name="l03287"></a>03287 <span class="stringliteral">            h = bar(axesHandle, 1:size(rnaPauses, 2), tmp&#39;</span>, <span class="stringliteral">&#39;group&#39;</span>);
<a name="l03288"></a>03288             xlim(axesHandle, [0.5 find(any(tmp &gt; 0.01, 1), 1, <span class="stringliteral">&#39;last&#39;</span>) + 0.5]);
<a name="l03289"></a>03289             legend(h, {<span class="charliteral">&#39;A&#39;</span>, <span class="charliteral">&#39;C&#39;</span>, <span class="charliteral">&#39;G&#39;</span>, <span class="charliteral">&#39;T&#39;</span>});
<a name="l03290"></a>03290             ylim(axesHandle, [0 max(tmp(:))]);
<a name="l03291"></a>03291             xlabel(axesHandle, <span class="stringliteral">&#39;Pause Duration (s)&#39;</span>, <span class="stringliteral">&#39;FontSize&#39;</span>, 8);
<a name="l03292"></a>03292             ylabel(axesHandle, <span class="stringliteral">&#39;Freq&#39;</span>, <span class="stringliteral">&#39;FontSize&#39;</span>, 8)
<a name="l03293"></a>03293             
<a name="l03294"></a>03294             %fraction bound
<a name="l03295"></a>03295             axesHandle = subplot(3, 2, 4, <span class="stringliteral">&#39;Parent&#39;</span>, figHandle);
<a name="l03296"></a>03296             axesHandles(2, 2) = axesHandle;
<a name="l03297"></a>03297             [~, order] = sort(nanmean(mrnaFracBounds, 3));
<a name="l03298"></a>03298             errorbar((1:numel(g.mRNAIndexs))<span class="stringliteral">&#39;, ...</span>
<a name="l03299"></a>03299 <span class="stringliteral">                nanmean(mrnaFracBounds(order, :, :), 3), ...</span>
<a name="l03300"></a>03300 <span class="stringliteral">                nanstd(mrnaFracBounds(order, :, :), [], 3), ...</span>
<a name="l03301"></a>03301 <span class="stringliteral">                &#39;</span>LineStyle<span class="stringliteral">&#39;, &#39;</span>none<span class="stringliteral">&#39;, &#39;</span>Parent<span class="stringliteral">&#39;, axesHandle);</span>
<a name="l03302"></a>03302 <span class="stringliteral">            xlim(axesHandle, [0.5 numel(g.mRNAIndexs) + 0.5])</span>
<a name="l03303"></a>03303 <span class="stringliteral">            ylim(axesHandle, [</span>
<a name="l03304"></a>03304 <span class="stringliteral">                min(nanmean(mrnaFracBounds, 3) - nanstd(mrnaFracBounds, [], 3))  ...</span>
<a name="l03305"></a>03305 <span class="stringliteral">                max(nanmean(mrnaFracBounds, 3) + nanstd(mrnaFracBounds, [], 3))</span>
<a name="l03306"></a>03306 <span class="stringliteral">                ]);</span>
<a name="l03307"></a>03307 <span class="stringliteral">            xlabel(axesHandle, &#39;</span><a class="code" href="classm_r_n_a.html">mRNA</a><span class="stringliteral">&#39;, &#39;</span>FontSize<span class="stringliteral">&#39;, 8)</span>
<a name="l03308"></a>03308 <span class="stringliteral">            ylabel(axesHandle, {&#39;</span>Frac<span class="charliteral">&#39; &#39;</span>Bound<span class="stringliteral">&#39;}, &#39;</span>FontSize<span class="stringliteral">&#39;, 8);</span>
<a name="l03309"></a>03309 <span class="stringliteral">            </span>
<a name="l03310"></a>03310 <span class="stringliteral">            %fraction aminoacylated</span>
<a name="l03311"></a>03311 <span class="stringliteral">            axesHandle = subplot(3, 2, 6, &#39;</span>Parent<span class="stringliteral">&#39;, figHandle);</span>
<a name="l03312"></a>03312 <span class="stringliteral">            axesHandles(3, 2) = axesHandle;</span>
<a name="l03313"></a>03313 <span class="stringliteral">            expr = r.expression(r.matureIndexs(r.matureTRNAIndexs));</span>
<a name="l03314"></a>03314 <span class="stringliteral">            expr = expr / sum(expr);</span>
<a name="l03315"></a>03315 <span class="stringliteral">            errorbar(expr, ...</span>
<a name="l03316"></a>03316 <span class="stringliteral">                nanmean(rnaFracAminoacylated(r.matureTRNAIndexs, :, :), 3), ...</span>
<a name="l03317"></a>03317 <span class="stringliteral">                nanstd(rnaFracAminoacylated(r.matureTRNAIndexs, :, :), [], 3), ...</span>
<a name="l03318"></a>03318 <span class="stringliteral">                &#39;</span>LineStyle<span class="stringliteral">&#39;, &#39;</span>none<span class="stringliteral">&#39;, &#39;</span>Parent<span class="stringliteral">&#39;, axesHandle);</span>
<a name="l03319"></a>03319 <span class="stringliteral">            xlim(axesHandle, [min(expr) max(expr)]);</span>
<a name="l03320"></a>03320 <span class="stringliteral">            ylim(axesHandle, [</span>
<a name="l03321"></a>03321 <span class="stringliteral">                min(nanmean(rnaFracAminoacylated(r.matureTRNAIndexs, :, :), 3) - nanstd(rnaFracAminoacylated(r.matureTRNAIndexs, :, :), [], 3)) ...</span>
<a name="l03322"></a>03322 <span class="stringliteral">                max(nanmean(rnaFracAminoacylated(r.matureTRNAIndexs, :, :), 3) + nanstd(rnaFracAminoacylated(r.matureTRNAIndexs, :, :), [], 3))</span>
<a name="l03323"></a>03323 <span class="stringliteral">                ]);</span>
<a name="l03324"></a>03324 <span class="stringliteral">            xlabel(axesHandle, &#39;</span><a class="code" href="classt_r_n_a.html">tRNA</a> Exp<span class="stringliteral">&#39;, &#39;</span>FontSize<span class="stringliteral">&#39;, 8);</span>
<a name="l03325"></a>03325 <span class="stringliteral">            ylabel(axesHandle, {&#39;</span>Frac<span class="charliteral">&#39; &#39;</span>Aminoacylated<span class="stringliteral">&#39;}, &#39;</span>FontSize<span class="stringliteral">&#39;, 8);</span>
<a name="l03326"></a>03326 <span class="stringliteral">            </span>
<a name="l03327"></a>03327 <span class="stringliteral">            %align axes labels</span>
<a name="l03328"></a>03328 <span class="stringliteral">            set(axesHandles, &#39;</span>FontSize<span class="stringliteral">&#39;, 6);</span>
<a name="l03329"></a>03329 <span class="stringliteral">            PlotUtil.alignYAxesLabels(axesHandles(:, 1));</span>
<a name="l03330"></a>03330 <span class="stringliteral">            PlotUtil.alignYAxesLabels(axesHandles(:, 2));</span>
<a name="l03331"></a>03331 <span class="stringliteral">            </span>
<a name="l03332"></a>03332 <span class="stringliteral">            %% organize figure data</span>
<a name="l03333"></a>03333 <span class="stringliteral">            figData = struct;</span>
<a name="l03334"></a>03334 <span class="stringliteral">            figData.rnaSynthesisTimes = rnaSynthesisTimes;</span>
<a name="l03335"></a>03335 <span class="stringliteral">            figData.mrnaFracBounds = mrnaFracBounds;</span>
<a name="l03336"></a>03336 <span class="stringliteral">            figData.rnaFracAminoacylated = rnaFracAminoacylated;</span>
<a name="l03337"></a>03337 <span class="stringliteral">            figData.rnaPauses = rnaPauses;</span>
<a name="l03338"></a>03338 <span class="stringliteral">        end</span>
<a name="l03339"></a>03339 <span class="stringliteral">        </span>
<a name="l03340"></a>03340 <span class="stringliteral">        function figData = proteinSynthesisDuration(figHandle, simBatchDir, selectedSimulations)</span>
<a name="l03341"></a>03341 <span class="stringliteral">            import edu.stanford.covert.cell.sim.util.PlotUtil;</span>
<a name="l03342"></a>03342 <span class="stringliteral">            import edu.stanford.covert.cell.sim.util.SimulationEnsemble;</span>
<a name="l03343"></a>03343 <span class="stringliteral">            import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;</span>
<a name="l03344"></a>03344 <span class="stringliteral">            </span>
<a name="l03345"></a>03345 <span class="stringliteral">            if nargin &lt; 2 || isempty(simBatchDir)</span>
<a name="l03346"></a>03346 <span class="stringliteral">                simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();</span>
<a name="l03347"></a>03347 <span class="stringliteral">            end</span>
<a name="l03348"></a>03348 <span class="stringliteral">            if nargin &lt; 3 || isempty(selectedSimulations)</span>
<a name="l03349"></a>03349 <span class="stringliteral">                selectedSimulations = SimulationDiskUtil.getCompleteSimulations(simBatchDir);</span>
<a name="l03350"></a>03350 <span class="stringliteral">            end</span>
<a name="l03351"></a>03351 <span class="stringliteral">            </span>
<a name="l03352"></a>03352 <span class="stringliteral">            %% constants</span>
<a name="l03353"></a>03353 <span class="stringliteral">            [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep &#39;</span>1<span class="stringliteral">&#39;]);</span>
<a name="l03354"></a>03354 <span class="stringliteral">            g = sim.gene;</span>
<a name="l03355"></a>03355 <span class="stringliteral">            comp = sim.compartment;</span>
<a name="l03356"></a>03356 <span class="stringliteral">            m = sim.state(&#39;</span><a class="code" href="class_metabolite.html">Metabolite</a><span class="stringliteral">&#39;);</span>
<a name="l03357"></a>03357 <span class="stringliteral">            pm = sim.state(&#39;</span>ProteinMonomer<span class="stringliteral">&#39;);</span>
<a name="l03358"></a>03358 <span class="stringliteral">            pc = sim.state(&#39;</span>ProteinComplex<span class="stringliteral">&#39;);</span>
<a name="l03359"></a>03359 <span class="stringliteral">            rib = sim.state(&#39;</span>Ribosome<span class="stringliteral">&#39;);</span>
<a name="l03360"></a>03360 <span class="stringliteral">            pol = sim.state(&#39;</span>Polypeptide<span class="stringliteral">&#39;);</span>
<a name="l03361"></a>03361 <span class="stringliteral">            </span>
<a name="l03362"></a>03362 <span class="stringliteral">            nSims = numel(selectedSimulations);</span>
<a name="l03363"></a>03363 <span class="stringliteral">            pcComp = sum(pc.proteinComplexComposition(g.mRNAIndexs, :, :), 3);</span>
<a name="l03364"></a>03364 <span class="stringliteral">            </span>
<a name="l03365"></a>03365 <span class="stringliteral">            %% get data</span>
<a name="l03366"></a>03366 <span class="stringliteral">            stateNames = {</span>
<a name="l03367"></a>03367 <span class="stringliteral">                &#39;</span>ProteinMonomer<span class="stringliteral">&#39;  &#39;</span>counts<span class="stringliteral">&#39;                   &#39;</span>:<span class="stringliteral">&#39;  &#39;</span>:<span class="stringliteral">&#39;</span>
<a name="l03368"></a>03368 <span class="stringliteral">                &#39;</span>ProteinComplex<span class="stringliteral">&#39;  &#39;</span>counts<span class="stringliteral">&#39;                   &#39;</span>:<span class="stringliteral">&#39;  &#39;</span>-sum<span class="stringliteral">&#39;</span>
<a name="l03369"></a>03369 <span class="stringliteral">                &#39;</span>Ribosome<span class="stringliteral">&#39;        &#39;</span>states<span class="stringliteral">&#39;                   &#39;</span>:<span class="stringliteral">&#39;  &#39;</span>:<span class="stringliteral">&#39;</span>
<a name="l03370"></a>03370 <span class="stringliteral">                &#39;</span>Ribosome<span class="stringliteral">&#39;        &#39;</span>boundMRNAs<span class="stringliteral">&#39;               &#39;</span>:<span class="stringliteral">&#39;  &#39;</span>:<span class="stringliteral">&#39;</span>
<a name="l03371"></a>03371 <span class="stringliteral">                &#39;</span>Ribosome<span class="stringliteral">&#39;        &#39;</span>mRNAPositions<span class="stringliteral">&#39;            &#39;</span>:<span class="stringliteral">&#39;  &#39;</span>:<span class="stringliteral">&#39;</span>
<a name="l03372"></a>03372 <span class="stringliteral">                &#39;</span>Ribosome<span class="stringliteral">&#39;        &#39;</span>tmRNAPositions<span class="stringliteral">&#39;           &#39;</span>:<span class="stringliteral">&#39;  &#39;</span>:<span class="stringliteral">&#39;</span>
<a name="l03373"></a>03373 <span class="stringliteral">                };</span>
<a name="l03374"></a>03374 <span class="stringliteral">            synthesisTimes = zeros(numel(pm.matureIndexs), 6, nSims);</span>
<a name="l03375"></a>03375 <span class="stringliteral">            fracBounds = zeros(numel(pm.matureIndexs), 1, nSims);</span>
<a name="l03376"></a>03376 <span class="stringliteral">            trnaPauses = zeros(36, 1000);</span>
<a name="l03377"></a>03377 <span class="stringliteral">            aaPauses = zeros(21, 1000);</span>
<a name="l03378"></a>03378 <span class="stringliteral">            for i = 1:nSims</span>
<a name="l03379"></a>03379 <span class="stringliteral">                states = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, &#39;</span>extract<span class="stringliteral">&#39;, selectedSimulations(i));</span>
<a name="l03380"></a>03380 <span class="stringliteral">                states.Ribosome.states = permute(states.Ribosome.states, [1 3 2]);</span>
<a name="l03381"></a>03381 <span class="stringliteral">                states.Ribosome.boundMRNAs = permute(states.Ribosome.boundMRNAs, [1 3 2]);</span>
<a name="l03382"></a>03382 <span class="stringliteral">                states.Ribosome.mRNAPositions = permute(states.Ribosome.mRNAPositions, [1 3 2]);</span>
<a name="l03383"></a>03383 <span class="stringliteral">                states.Ribosome.tmRNAPositions = permute(states.Ribosome.tmRNAPositions, [1 3 2]);</span>
<a name="l03384"></a>03384 <span class="stringliteral">                </span>
<a name="l03385"></a>03385 <span class="stringliteral">                monCounts = permute(sum(states.ProteinMonomer.counts, 2), [1 3 2]);</span>
<a name="l03386"></a>03386 <span class="stringliteral">                cpxCounts = full(permute(sum(reshape(states.ProteinComplex.counts, ...</span>
<a name="l03387"></a>03387 <span class="stringliteral">                    [numel(pc.matureIndexs)  numel(pc.wholeCellModelIDs)/numel(pc.matureIndexs)  size(states.ProteinComplex.counts, 3)]), 2), [1 3 2]));</span>
<a name="l03388"></a>03388 <span class="stringliteral">                nascs = full(monCounts(pm.nascentIndexs, :));</span>
<a name="l03389"></a>03389 <span class="stringliteral">                procsI = zeros(numel(pm.matureIndexs), size(states.ProteinMonomer.counts, 3));</span>
<a name="l03390"></a>03390 <span class="stringliteral">                procsI(pm.compartments(pm.processedIIIndexs) ~= comp.cytosolIndexs, :) = ...</span>
<a name="l03391"></a>03391 <span class="stringliteral">                    full(permute(sum(states.ProteinMonomer.counts(...</span>
<a name="l03392"></a>03392 <span class="stringliteral">                    pm.processedIIndexs(pm.compartments(pm.processedIIIndexs) ~= comp.cytosolIndexs), ...</span>
<a name="l03393"></a>03393 <span class="stringliteral">                    comp.cytosolIndexs, :), 2), [1 3 2]));</span>
<a name="l03394"></a>03394 <span class="stringliteral">                trans = full(monCounts(pm.processedIIndexs, :)) - procsI;</span>
<a name="l03395"></a>03395 <span class="stringliteral">                procsII = full(monCounts(pm.processedIIIndexs, :));</span>
<a name="l03396"></a>03396 <span class="stringliteral">                folds = full(monCounts(pm.foldedIndexs, :));</span>
<a name="l03397"></a>03397 <span class="stringliteral">                signals = full(monCounts(pm.signalSequenceIndexs, :));</span>
<a name="l03398"></a>03398 <span class="stringliteral">                matrs = full(permute(sum(reshape(permute(monCounts, [1 3 2]), ...</span>
<a name="l03399"></a>03399 <span class="stringliteral">                    [numel(pm.matureIndexs)  numel(pm.wholeCellModelIDs)/numel(pm.matureIndexs)  size(states.ProteinMonomer.counts, 3)]), 2), [1 3 2])) ...</span>
<a name="l03400"></a>03400 <span class="stringliteral">                    - nascs ...</span>
<a name="l03401"></a>03401 <span class="stringliteral">                    - procsI ...</span>
<a name="l03402"></a>03402 <span class="stringliteral">                    - trans ...</span>
<a name="l03403"></a>03403 <span class="stringliteral">                    - procsII ...</span>
<a name="l03404"></a>03404 <span class="stringliteral">                    - folds ...</span>
<a name="l03405"></a>03405 <span class="stringliteral">                    - signals ...</span>
<a name="l03406"></a>03406 <span class="stringliteral">                    + pcComp * cpxCounts;</span>
<a name="l03407"></a>03407 <span class="stringliteral">                bounds = ...</span>
<a name="l03408"></a>03408 <span class="stringliteral">                    + monCounts(pm.boundIndexs, :) ...</span>
<a name="l03409"></a>03409 <span class="stringliteral">                    + pcComp * full(permute(states.ProteinComplex.counts(pc.boundIndexs, :, :), [1 3 2]));</span>
<a name="l03410"></a>03410 <span class="stringliteral">                </span>
<a name="l03411"></a>03411 <span class="stringliteral">                %synthesis times</span>
<a name="l03412"></a>03412 <span class="stringliteral">                [startTimes, startRibs] = find(cat(2, ...</span>
<a name="l03413"></a>03413 <span class="stringliteral">                    states.Ribosome.mRNAPositions(:, 1) &gt; 0 | ...</span>
<a name="l03414"></a>03414 <span class="stringliteral">                    states.Ribosome.tmRNAPositions(:, 1) &gt; 0, ...</span>
<a name="l03415"></a>03415 <span class="stringliteral">                    states.Ribosome.mRNAPositions(:, 1:end-1) == 0 &amp; ...</span>
<a name="l03416"></a>03416 <span class="stringliteral">                    states.Ribosome.tmRNAPositions(:, 1:end-1) == 0 &amp; ...</span>
<a name="l03417"></a>03417 <span class="stringliteral">                    (states.Ribosome.mRNAPositions(:, 2:end) &gt; 0 | ...</span>
<a name="l03418"></a>03418 <span class="stringliteral">                    states.Ribosome.tmRNAPositions(:, 2:end) &gt; 0))&#39;</span>);
<a name="l03419"></a>03419                 [endTimes, endRibs] = find(cat(2, ...
<a name="l03420"></a>03420                     (states.Ribosome.mRNAPositions(:, 1:end-1) &gt; 0 | ...
<a name="l03421"></a>03421                     states.Ribosome.tmRNAPositions(:, 1:end-1) &gt; 0) &amp; ...
<a name="l03422"></a>03422                     states.Ribosome.mRNAPositions(:, 2:end) == 0 &amp; ...
<a name="l03423"></a>03423                     states.Ribosome.tmRNAPositions(:, 2:end) == 0, ...
<a name="l03424"></a>03424                     states.Ribosome.mRNAPositions(:, end) &gt; 0 | ...
<a name="l03425"></a>03425                     states.Ribosome.tmRNAPositions(:, end) &gt; 0)<span class="stringliteral">&#39;);</span>
<a name="l03426"></a>03426 <span class="stringliteral">                assert(all(startRibs == endRibs));</span>
<a name="l03427"></a>03427 <span class="stringliteral">                assert(all(endTimes &gt;= startTimes));</span>
<a name="l03428"></a>03428 <span class="stringliteral">                boundMRNAs = states.Ribosome.boundMRNAs(sub2ind(...</span>
<a name="l03429"></a>03429 <span class="stringliteral">                    size(states.Ribosome.boundMRNAs), ...</span>
<a name="l03430"></a>03430 <span class="stringliteral">                    startRibs, startTimes));</span>
<a name="l03431"></a>03431 <span class="stringliteral">                synthesisTimes(:, 1, i) = ...</span>
<a name="l03432"></a>03432 <span class="stringliteral">                    accumarray(boundMRNAs, endTimes - startTimes + 1, ...</span>
<a name="l03433"></a>03433 <span class="stringliteral">                    [numel(pm.matureIndexs) 1]) ./ histc(boundMRNAs, (1:numel(pm.matureIndexs))&#39;</span>);
<a name="l03434"></a>03434                 
<a name="l03435"></a>03435                 %translation pauses
<a name="l03436"></a>03436                 pause = ...
<a name="l03437"></a>03437                     states.Ribosome.states(:, 1:end-1) ~= rib.notExistValue &amp; ...
<a name="l03438"></a>03438                     diff(states.Ribosome.mRNAPositions + states.Ribosome.tmRNAPositions, 1, 2) == 0;
<a name="l03439"></a>03439                 [startTimes, startRibs] = find([pause(:, 1)  diff(pause, 1, 2) &gt; 0]<span class="stringliteral">&#39;);</span>
<a name="l03440"></a>03440 <span class="stringliteral">                [endTimes, endRibs] = find([diff(pause, 1, 2) &lt; 0  pause(:, end)]&#39;</span>);
<a name="l03441"></a>03441                 assert(all(startRibs == endRibs));
<a name="l03442"></a>03442                 assert(all(endTimes &gt;= startTimes));
<a name="l03443"></a>03443                 
<a name="l03444"></a>03444                 <span class="keywordflow">for</span> j = 1:numel(startRibs)
<a name="l03445"></a>03445                     if ...
<a name="l03446"></a>03446                             states.Ribosome.mRNAPositions(startRibs(j),  startTimes(j)) == pol.monomerLengths(states.Ribosome.boundMRNAs(startRibs(j),  startTimes(j))) || ...
<a name="l03447"></a>03447                             states.Ribosome.tmRNAPositions(startRibs(j), startTimes(j)) == pol.proteolysisTagLength
<a name="l03448"></a>03448                         continue;
<a name="l03449"></a>03449                     end
<a name="l03450"></a>03450                     
<a name="l03451"></a>03451                     if states.Ribosome.tmRNAPositions(startRibs(j), startTimes(j)) == 0
<a name="l03452"></a>03452                         aa = pol.monomerAASequences{states.Ribosome.boundMRNAs(startRibs(j), startTimes(j))}(states.Ribosome.mRNAPositions(startRibs(j),  startTimes(j))+1);
<a name="l03453"></a>03453                         tnra = pol.monomerTRNASequences{states.Ribosome.boundMRNAs(startRibs(j), startTimes(j))}(states.Ribosome.mRNAPositions(startRibs(j),  startTimes(j))+1);
<a name="l03454"></a>03454                     <span class="keywordflow">else</span>
<a name="l03455"></a>03455                         aa = pol.proteolysisTagAASequence(states.Ribosome.tmRNAPositions(startRibs(j), startTimes(j))+1);
<a name="l03456"></a>03456                         tnra = pol.proteolysisTagTRNASequence(states.Ribosome.tmRNAPositions(startRibs(j), startTimes(j))+1);
<a name="l03457"></a>03457                     end
<a name="l03458"></a>03458                     <span class="keywordflow">if</span> states.Ribosome.mRNAPositions(startRibs(j), startTimes(j)) == 0
<a name="l03459"></a>03459                         aaIdx = 21;
<a name="l03460"></a>03460                     <span class="keywordflow">else</span>
<a name="l03461"></a>03461                         aaIdx = find(aa == edu.stanford.covert.cell.kb.ProteinMonomer.bases);
<a name="l03462"></a>03462                     end
<a name="l03463"></a>03463                     <span class="keywordflow">if</span> endTimes(j) - startTimes(j) + 1 &gt; size(aaPauses, 2)
<a name="l03464"></a>03464                         aaPauses = [aaPauses zeros(size(aaPauses, 1), (endTimes(j) - startTimes(j) + 1) - size(aaPauses, 2))]; %#ok&lt;AGROW&gt;
<a name="l03465"></a>03465                         trnaPauses = [trnaPauses zeros(size(trnaPauses, 1), (endTimes(j) - startTimes(j) + 1) - size(trnaPauses, 2))]; %#ok&lt;AGROW&gt;
<a name="l03466"></a>03466                     end
<a name="l03467"></a>03467                     aaPauses(aaIdx, endTimes(j) - startTimes(j) + 1) = ...
<a name="l03468"></a>03468                         aaPauses(aaIdx, endTimes(j) - startTimes(j) + 1) + 1;
<a name="l03469"></a>03469                     trnaPauses(tnra, endTimes(j) - startTimes(j) + 1) = ...
<a name="l03470"></a>03470                         trnaPauses(tnra, endTimes(j) - startTimes(j) + 1) + 1;
<a name="l03471"></a>03471                 end
<a name="l03472"></a>03472                 
<a name="l03473"></a>03473                 %processing I time
<a name="l03474"></a>03474                 synthesisTimes(:, 2, i) = log(2) ./ (sum(max(0, -diff(nascs, 1, 2)), 2) ./ sum(nascs(:, 1:end-1), 2));
<a name="l03475"></a>03475                 
<a name="l03476"></a>03476                 %translocation time
<a name="l03477"></a>03477                 synthesisTimes(:, 3, i) = log(2) ./ (sum(max(0, -diff(procsI, 1, 2)), 2) ./ sum(procsI(:, 1:end-1), 2));
<a name="l03478"></a>03478                 
<a name="l03479"></a>03479                 %processing II time
<a name="l03480"></a>03480                 synthesisTimes(:, 4, i) = log(2) ./ (sum(max(0, -diff(trans, 1, 2)), 2) ./ sum(trans(:, 1:end-1), 2));
<a name="l03481"></a>03481                 
<a name="l03482"></a>03482                 %folding time
<a name="l03483"></a>03483                 synthesisTimes(:, 5, i) = log(2) ./ (sum(max(0, -diff(procsII, 1, 2)), 2) ./ sum(procsII(:, 1:end-1), 2));
<a name="l03484"></a>03484                 
<a name="l03485"></a>03485                 %modification time
<a name="l03486"></a>03486                 synthesisTimes(:, 6, i) = log(2) ./ (sum(max(0, -diff(folds, 1, 2)), 2) ./ sum(folds(:, 1:end-1), 2));
<a name="l03487"></a>03487                 
<a name="l03488"></a>03488                 %fraction bound
<a name="l03489"></a>03489                 fracBounds(:, 1, i) = sum(bounds, 2) ./ sum(matrs, 2);
<a name="l03490"></a>03490                 
<a name="l03491"></a>03491                 clear states monCounts cpxCounts nascs procsI trans procsII folds signals matrs bounds;
<a name="l03492"></a>03492                 clear boundMRNAs startTimes endTimes startRibs endRibs pause;
<a name="l03493"></a>03493             end
<a name="l03494"></a>03494             
<a name="l03495"></a>03495             synthesisTimes(isinf(synthesisTimes)) = NaN;
<a name="l03496"></a>03496             fracBounds(isinf(fracBounds)) = NaN;
<a name="l03497"></a>03497             
<a name="l03498"></a>03498             %% plot
<a name="l03499"></a>03499             clf(figHandle);
<a name="l03500"></a>03500             axesHandles = zeros(3, 2);
<a name="l03501"></a>03501             
<a name="l03502"></a>03502             %synthesis times
<a name="l03503"></a>03503             axesHandle = subplot(3, 2, 1, <span class="stringliteral">&#39;Parent&#39;</span>, figHandle);
<a name="l03504"></a>03504             axesHandles(1, 1) = axesHandle;
<a name="l03505"></a>03505             tmp = nanmean(synthesisTimes, 3);
<a name="l03506"></a>03506             [~, order] = sort(nansum(tmp, 2));
<a name="l03507"></a>03507             h = bar(axesHandle, 1:numel(pm.matureIndexs),  tmp(order, :), <span class="stringliteral">&#39;stacked&#39;</span>);
<a name="l03508"></a>03508             xlim(axesHandle, [0.5 numel(pm.matureIndexs) + 0.5]);
<a name="l03509"></a>03509             ylim(axesHandle, [0 max(nansum(tmp, 2))]);
<a name="l03510"></a>03510             legend(h, {<span class="stringliteral">&#39;Translation&#39;</span>, <span class="stringliteral">&#39;Processing-I&#39;</span>, <span class="stringliteral">&#39;Translocation&#39;</span>, <span class="stringliteral">&#39;Processing-II&#39;</span>, <span class="stringliteral">&#39;Folding&#39;</span>, <span class="stringliteral">&#39;Modification&#39;</span>}, ...
<a name="l03511"></a>03511                 <span class="stringliteral">&#39;Location&#39;</span>, <span class="stringliteral">&#39;NorthWest&#39;</span>, <span class="stringliteral">&#39;FontSize&#39;</span>, 4);
<a name="l03512"></a>03512             xlabel(axesHandle, <span class="stringliteral">&#39;Monomer&#39;</span>, <span class="stringliteral">&#39;FontSize&#39;</span>, 8);
<a name="l03513"></a>03513             ylabel(axesHandle, {<span class="stringliteral">&#39;Synthesis&#39;</span> <span class="stringliteral">&#39;Duration (s)&#39;</span>}, <span class="stringliteral">&#39;FontSize&#39;</span>, 8);
<a name="l03514"></a>03514             
<a name="l03515"></a>03515             %transcription time vs length
<a name="l03516"></a>03516             axesHandle = subplot(3, 2, 3, <span class="stringliteral">&#39;Parent&#39;</span>, figHandle);
<a name="l03517"></a>03517             axesHandles(2, 1) = axesHandle;
<a name="l03518"></a>03518             meanVal = nanmean(synthesisTimes(:, 1, :), 3);
<a name="l03519"></a>03519             stdVal = nanstd(synthesisTimes(:, 1, :), [], 3);
<a name="l03520"></a>03520             errorbar(pm.lengths(pm.nascentIndexs), meanVal, stdVal, ...
<a name="l03521"></a>03521                 <span class="stringliteral">&#39;LineStyle&#39;</span>, <span class="stringliteral">&#39;none&#39;</span>, <span class="stringliteral">&#39;Parent&#39;</span>, axesHandle);
<a name="l03522"></a>03522             xlim(axesHandle, [min(pm.lengths(pm.nascentIndexs)) max(pm.lengths(pm.nascentIndexs))]);
<a name="l03523"></a>03523             ylim(axesHandle, [min(meanVal - stdVal) max(meanVal + stdVal)]);
<a name="l03524"></a>03524             xlabel(axesHandle, <span class="stringliteral">&#39;Length&#39;</span>, <span class="stringliteral">&#39;FontSize&#39;</span>, 8);
<a name="l03525"></a>03525             ylabel(axesHandle, {<span class="stringliteral">&#39;Translation&#39;</span> <span class="stringliteral">&#39;Duration (s)&#39;</span>}, <span class="stringliteral">&#39;FontSize&#39;</span>, 8)
<a name="l03526"></a>03526             
<a name="l03527"></a>03527             %transcription time vs G/C content
<a name="l03528"></a>03528             axesHandle = subplot(3, 2, 5, <span class="stringliteral">&#39;Parent&#39;</span>, figHandle);
<a name="l03529"></a>03529             axesHandles(3, 1) = axesHandle;
<a name="l03530"></a>03530             [~, idx] = min(m.biomassComposition(m.aminoAcidIndexs(1:20)));
<a name="l03531"></a>03531             aaContent = pm.baseCounts(pm.nascentIndexs, m.aminoAcidIndexs(idx)) ./ pm.lengths(pm.nascentIndexs);
<a name="l03532"></a>03532             meanVal = nanmean(synthesisTimes(:, 1, :), 3);
<a name="l03533"></a>03533             stdVal = nanstd(synthesisTimes(:, 1, :), [], 3);
<a name="l03534"></a>03534             errorbar(aaContent, meanVal, stdVal, ...
<a name="l03535"></a>03535                 <span class="stringliteral">&#39;LineStyle&#39;</span>, <span class="stringliteral">&#39;none&#39;</span>, <span class="stringliteral">&#39;Parent&#39;</span>, axesHandle);
<a name="l03536"></a>03536             xlim(axesHandle, [min(aaContent) max(aaContent)]);
<a name="l03537"></a>03537             ylim(axesHandle, [min(meanVal - stdVal) max(meanVal + stdVal)]);
<a name="l03538"></a>03538             xlabel(axesHandle, [m.names{m.aminoAcidIndexs(idx)} <span class="stringliteral">&#39; Content&#39;</span>], <span class="stringliteral">&#39;FontSize&#39;</span>, 8);
<a name="l03539"></a>03539             ylabel(axesHandle, {<span class="stringliteral">&#39;Translation&#39;</span> <span class="stringliteral">&#39;Duration&#39;</span>}, <span class="stringliteral">&#39;FontSize&#39;</span>, 8)
<a name="l03540"></a>03540             
<a name="l03541"></a>03541             %pauses -- <a class="code" href="classt_r_n_a.html">tRNA</a>
<a name="l03542"></a>03542             axesHandle = subplot(3, 2, 2, <span class="stringliteral">&#39;Parent&#39;</span>, figHandle);
<a name="l03543"></a>03543             axesHandles(1, 2) = axesHandle;
<a name="l03544"></a>03544             tmp = trnaPauses ./ repmat(sum(trnaPauses, 2), 1, size(trnaPauses, 2));
<a name="l03545"></a>03545             bar(axesHandle, 1:size(trnaPauses, 2), tmp<span class="stringliteral">&#39;, &#39;</span>group<span class="stringliteral">&#39;);</span>
<a name="l03546"></a>03546 <span class="stringliteral">            xlim(axesHandle, [0.5 find(any(tmp &gt; 0.01, 1), 1, &#39;</span>last<span class="stringliteral">&#39;)+0.5]);</span>
<a name="l03547"></a>03547 <span class="stringliteral">            ylim(axesHandle, [0 max(tmp(:))]);</span>
<a name="l03548"></a>03548 <span class="stringliteral">            xlabel(axesHandle, &#39;</span>Pause Duration (s)<span class="stringliteral">&#39;, &#39;</span>FontSize<span class="stringliteral">&#39;, 8);</span>
<a name="l03549"></a>03549 <span class="stringliteral">            ylabel(axesHandle, &#39;</span>Freq<span class="stringliteral">&#39;, &#39;</span>FontSize<span class="stringliteral">&#39;, 8)</span>
<a name="l03550"></a>03550 <span class="stringliteral">            </span>
<a name="l03551"></a>03551 <span class="stringliteral">            %pauses -- AA</span>
<a name="l03552"></a>03552 <span class="stringliteral">            axesHandle = subplot(3, 2, 4, &#39;</span>Parent<span class="stringliteral">&#39;, figHandle);</span>
<a name="l03553"></a>03553 <span class="stringliteral">            axesHandles(2, 2) = axesHandle;</span>
<a name="l03554"></a>03554 <span class="stringliteral">            tmp = aaPauses ./ repmat(sum(aaPauses, 2), 1, size(aaPauses, 2));</span>
<a name="l03555"></a>03555 <span class="stringliteral">            bar(axesHandle, 1:size(aaPauses, 2), tmp&#39;</span>, <span class="stringliteral">&#39;group&#39;</span>);
<a name="l03556"></a>03556             xlim(axesHandle, [0.5 find(any(tmp &gt; 0.01, 1), 1, <span class="stringliteral">&#39;last&#39;</span>) + 0.5]);
<a name="l03557"></a>03557             ylim(axesHandle, [0 max(tmp(:))]);
<a name="l03558"></a>03558             xlabel(axesHandle, <span class="stringliteral">&#39;Pause Duration (s)&#39;</span>, <span class="stringliteral">&#39;FontSize&#39;</span>, 8);
<a name="l03559"></a>03559             ylabel(axesHandle, <span class="stringliteral">&#39;Freq&#39;</span>, <span class="stringliteral">&#39;FontSize&#39;</span>, 8)
<a name="l03560"></a>03560             
<a name="l03561"></a>03561             %fraction bound
<a name="l03562"></a>03562             axesHandle = subplot(3, 2, 6, <span class="stringliteral">&#39;Parent&#39;</span>, figHandle);
<a name="l03563"></a>03563             axesHandles(3, 2) = axesHandle;
<a name="l03564"></a>03564             [~, order] = sort(nanmean(fracBounds, 3));
<a name="l03565"></a>03565             order = order(any(fracBounds(order, :) &gt; 0, 2));
<a name="l03566"></a>03566             errorbar((1:numel(order))<span class="stringliteral">&#39;, ...</span>
<a name="l03567"></a>03567 <span class="stringliteral">                nanmean(fracBounds(order, :, :), 3), ...</span>
<a name="l03568"></a>03568 <span class="stringliteral">                nanstd(fracBounds(order, :, :), [], 3), ...</span>
<a name="l03569"></a>03569 <span class="stringliteral">                &#39;</span>LineStyle<span class="stringliteral">&#39;, &#39;</span>none<span class="stringliteral">&#39;, &#39;</span>Parent<span class="stringliteral">&#39;, axesHandle);</span>
<a name="l03570"></a>03570 <span class="stringliteral">            xlim(axesHandle, [0.5 numel(order) + 0.5])</span>
<a name="l03571"></a>03571 <span class="stringliteral">            ylim(axesHandle, [</span>
<a name="l03572"></a>03572 <span class="stringliteral">                min(nanmean(fracBounds, 3) - nanstd(fracBounds, [], 3))  ...</span>
<a name="l03573"></a>03573 <span class="stringliteral">                max(nanmean(fracBounds, 3) + nanstd(fracBounds, [], 3))</span>
<a name="l03574"></a>03574 <span class="stringliteral">                ]);</span>
<a name="l03575"></a>03575 <span class="stringliteral">            xlabel(axesHandle, &#39;</span>Monomer<span class="stringliteral">&#39;, &#39;</span>FontSize<span class="stringliteral">&#39;, 8)</span>
<a name="l03576"></a>03576 <span class="stringliteral">            ylabel(axesHandle, {&#39;</span>Frac<span class="charliteral">&#39; &#39;</span>Bound<span class="stringliteral">&#39;}, &#39;</span>FontSize<span class="stringliteral">&#39;, 8);</span>
<a name="l03577"></a>03577 <span class="stringliteral">            </span>
<a name="l03578"></a>03578 <span class="stringliteral">            %align axes labels</span>
<a name="l03579"></a>03579 <span class="stringliteral">            set(axesHandles, &#39;</span>FontSize<span class="stringliteral">&#39;, 6);</span>
<a name="l03580"></a>03580 <span class="stringliteral">            PlotUtil.alignYAxesLabels(axesHandles(:, 1));</span>
<a name="l03581"></a>03581 <span class="stringliteral">            PlotUtil.alignYAxesLabels(axesHandles(:, 2));</span>
<a name="l03582"></a>03582 <span class="stringliteral">            </span>
<a name="l03583"></a>03583 <span class="stringliteral">            %% organize figure data</span>
<a name="l03584"></a>03584 <span class="stringliteral">            figData = struct;</span>
<a name="l03585"></a>03585 <span class="stringliteral">            figData.synthesisTimes = synthesisTimes;</span>
<a name="l03586"></a>03586 <span class="stringliteral">            figData.fracBounds = fracBounds;</span>
<a name="l03587"></a>03587 <span class="stringliteral">            figData.trnaPauses = trnaPauses;</span>
<a name="l03588"></a>03588 <span class="stringliteral">            figData.aaPauses = aaPauses;</span>
<a name="l03589"></a>03589 <span class="stringliteral">        end</span>
<a name="l03590"></a>03590 <span class="stringliteral">        </span>
<a name="l03591"></a>03591 <span class="stringliteral">        function figData = macromolecularComplexes(figHandle, simBatchDir, selectedSimulations)</span>
<a name="l03592"></a>03592 <span class="stringliteral">            import edu.stanford.covert.cell.sim.util.PlotUtil;</span>
<a name="l03593"></a>03593 <span class="stringliteral">            import edu.stanford.covert.cell.sim.util.SimulationEnsemble;</span>
<a name="l03594"></a>03594 <span class="stringliteral">            import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;</span>
<a name="l03595"></a>03595 <span class="stringliteral">            import edu.stanford.covert.util.ConstantUtil;</span>
<a name="l03596"></a>03596 <span class="stringliteral">            </span>
<a name="l03597"></a>03597 <span class="stringliteral">            if nargin &lt; 2 || isempty(simBatchDir)</span>
<a name="l03598"></a>03598 <span class="stringliteral">                simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();</span>
<a name="l03599"></a>03599 <span class="stringliteral">            end</span>
<a name="l03600"></a>03600 <span class="stringliteral">            if nargin &lt; 3 || isempty(selectedSimulations)</span>
<a name="l03601"></a>03601 <span class="stringliteral">                selectedSimulations = SimulationDiskUtil.getCompleteSimulations(simBatchDir);</span>
<a name="l03602"></a>03602 <span class="stringliteral">            end</span>
<a name="l03603"></a>03603 <span class="stringliteral">            </span>
<a name="l03604"></a>03604 <span class="stringliteral">            %% constants</span>
<a name="l03605"></a>03605 <span class="stringliteral">            [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep &#39;</span>1<span class="stringliteral">&#39;]);</span>
<a name="l03606"></a>03606 <span class="stringliteral">            g = sim.gene;</span>
<a name="l03607"></a>03607 <span class="stringliteral">            pm = sim.state(&#39;</span>ProteinMonomer<span class="stringliteral">&#39;);</span>
<a name="l03608"></a>03608 <span class="stringliteral">            pc = sim.state(&#39;</span>ProteinComplex<span class="stringliteral">&#39;);</span>
<a name="l03609"></a>03609 <span class="stringliteral">            </span>
<a name="l03610"></a>03610 <span class="stringliteral">            pcComp = sum(pc.proteinComplexComposition(g.mRNAIndexs, :, :), 3);</span>
<a name="l03611"></a>03611 <span class="stringliteral">            monIdxs = find(any(pcComp, 2));</span>
<a name="l03612"></a>03612 <span class="stringliteral">            pcComp = pcComp(monIdxs, :);</span>
<a name="l03613"></a>03613 <span class="stringliteral">            </span>
<a name="l03614"></a>03614 <span class="stringliteral">            %% load data</span>
<a name="l03615"></a>03615 <span class="stringliteral">            stateNames = {</span>
<a name="l03616"></a>03616 <span class="stringliteral">                &#39;</span>Geometry<span class="stringliteral">&#39;       &#39;</span>volume<span class="stringliteral">&#39;  &#39;</span>:<span class="stringliteral">&#39;  &#39;</span>:<span class="stringliteral">&#39;</span>
<a name="l03617"></a>03617 <span class="stringliteral">                &#39;</span>ProteinMonomer<span class="charliteral">&#39; &#39;</span>counts<span class="stringliteral">&#39;  &#39;</span>:<span class="stringliteral">&#39;  &#39;</span>-sum<span class="stringliteral">&#39;</span>
<a name="l03618"></a>03618 <span class="stringliteral">                &#39;</span>ProteinComplex<span class="charliteral">&#39; &#39;</span>counts<span class="stringliteral">&#39;  &#39;</span>:<span class="stringliteral">&#39;  &#39;</span>-sum<span class="stringliteral">&#39;</span>
<a name="l03619"></a>03619 <span class="stringliteral">                };</span>
<a name="l03620"></a>03620 <span class="stringliteral">            monConc = zeros(numel(monIdxs), numel(selectedSimulations));</span>
<a name="l03621"></a>03621 <span class="stringliteral">            cpxConc = zeros(numel(pc.matureIndexs), numel(selectedSimulations));</span>
<a name="l03622"></a>03622 <span class="stringliteral">            for i = 1:numel(selectedSimulations)</span>
<a name="l03623"></a>03623 <span class="stringliteral">                states = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, &#39;</span>extract<span class="stringliteral">&#39;, selectedSimulations(i));</span>
<a name="l03624"></a>03624 <span class="stringliteral">                states.ProteinMonomer.counts = full(sum(reshape(states.ProteinMonomer.counts, ...</span>
<a name="l03625"></a>03625 <span class="stringliteral">                    [numel(pm.matureIndexs)  numel(pm.wholeCellModelIDs)/numel(pm.matureIndexs)  size(states.ProteinMonomer.counts, 3)]), 2));</span>
<a name="l03626"></a>03626 <span class="stringliteral">                states.ProteinMonomer.counts = states.ProteinMonomer.counts(monIdxs, :, :);</span>
<a name="l03627"></a>03627 <span class="stringliteral">                states.ProteinComplex.counts = full(sum(reshape(states.ProteinComplex.counts, ...</span>
<a name="l03628"></a>03628 <span class="stringliteral">                    [numel(pc.matureIndexs)  numel(pc.wholeCellModelIDs)/numel(pc.matureIndexs)  size(states.ProteinComplex.counts, 3)]), 2));</span>
<a name="l03629"></a>03629 <span class="stringliteral">                </span>
<a name="l03630"></a>03630 <span class="stringliteral">                monConc(:, i) = mean((...</span>
<a name="l03631"></a>03631 <span class="stringliteral">                    + permute(states.ProteinMonomer.counts, [1 3 2]) ...</span>
<a name="l03632"></a>03632 <span class="stringliteral">                    + pcComp * permute(states.ProteinComplex.counts, [1 3 2])) ...</span>
<a name="l03633"></a>03633 <span class="stringliteral">                    ./ permute(states.Geometry.volume(ones(numel(monIdxs), 1), :, :) / ConstantUtil.nAvogadro * 1e6, [1 3 2]), ...</span>
<a name="l03634"></a>03634 <span class="stringliteral">                    2);</span>
<a name="l03635"></a>03635 <span class="stringliteral">                cpxConc(:, i) = mean((...</span>
<a name="l03636"></a>03636 <span class="stringliteral">                    permute(states.ProteinComplex.counts, [1 3 2])) ...</span>
<a name="l03637"></a>03637 <span class="stringliteral">                    ./ permute(states.Geometry.volume(ones(numel(pc.matureIndexs), 1), :, :) / ConstantUtil.nAvogadro * 1e6, [1 3 2]), ...</span>
<a name="l03638"></a>03638 <span class="stringliteral">                    2);</span>
<a name="l03639"></a>03639 <span class="stringliteral">                </span>
<a name="l03640"></a>03640 <span class="stringliteral">                clear states;</span>
<a name="l03641"></a>03641 <span class="stringliteral">            end</span>
<a name="l03642"></a>03642 <span class="stringliteral">            </span>
<a name="l03643"></a>03643 <span class="stringliteral">            meanMonConc = mean(monConc, 2);</span>
<a name="l03644"></a>03644 <span class="stringliteral">            stdMonConc = std(monConc, [], 2);</span>
<a name="l03645"></a>03645 <span class="stringliteral">            meanCpxConc = pcComp * mean(cpxConc, 2);</span>
<a name="l03646"></a>03646 <span class="stringliteral">            stdCpxConc = pcComp * std(cpxConc, [], 2);</span>
<a name="l03647"></a>03647 <span class="stringliteral">            </span>
<a name="l03648"></a>03648 <span class="stringliteral">            clear monConc cpxConc;</span>
<a name="l03649"></a>03649 <span class="stringliteral">            </span>
<a name="l03650"></a>03650 <span class="stringliteral">            %% plot</span>
<a name="l03651"></a>03651 <span class="stringliteral">            clf(figHandle);</span>
<a name="l03652"></a>03652 <span class="stringliteral">            [axesHandle, xAxesHandle] = PlotUtil.multiElementPlot(figHandle, 50, [min(meanCpxConc-stdCpxConc) max(meanCpxConc+stdCpxConc)], ...</span>
<a name="l03653"></a>03653 <span class="stringliteral">                struct(&#39;</span>xlabelStr<span class="stringliteral">&#39;, &#39;</span>Complex Concentration ({\mu}M)<span class="stringliteral">&#39;));</span>
<a name="l03654"></a>03654 <span class="stringliteral">            xlabel(xAxesHandle, &#39;</span>Complex Concentration ({\mu}M)<span class="stringliteral">&#39;, &#39;</span>FontSize<span class="stringliteral">&#39;, 12, &#39;</span>Units<span class="stringliteral">&#39;, &#39;</span>normalized<span class="stringliteral">&#39;);</span>
<a name="l03655"></a>03655 <span class="stringliteral">            </span>
<a name="l03656"></a>03656 <span class="stringliteral">            PlotUtil.plotBubbles(axesHandle, [meanCpxConc stdCpxConc/3], [meanMonConc stdMonConc/3]);</span>
<a name="l03657"></a>03657 <span class="stringliteral">            </span>
<a name="l03658"></a>03658 <span class="stringliteral">            set(axesHandle, &#39;</span>YTick<span class="stringliteral">&#39;, [0 100]);</span>
<a name="l03659"></a>03659 <span class="stringliteral">            ylim(axesHandle, [min(meanMonConc-stdMonConc) max(meanMonConc+stdMonConc)]);</span>
<a name="l03660"></a>03660 <span class="stringliteral">            ylabel(axesHandle, &#39;</span>Monomer Concentration ({\mu}M)<span class="stringliteral">&#39;, &#39;</span>FontSize<span class="stringliteral">&#39;, 12);</span>
<a name="l03661"></a>03661 <span class="stringliteral">            </span>
<a name="l03662"></a>03662 <span class="stringliteral">            PlotUtil.offsetYAxes(axesHandle, 0.02);</span>
<a name="l03663"></a>03663 <span class="stringliteral">            </span>
<a name="l03664"></a>03664 <span class="stringliteral">            %% organize figure data</span>
<a name="l03665"></a>03665 <span class="stringliteral">            figData = struct;</span>
<a name="l03666"></a>03666 <span class="stringliteral">            figData.meanMonConc = meanMonConc;</span>
<a name="l03667"></a>03667 <span class="stringliteral">            figData.stdMonConc = stdMonConc;</span>
<a name="l03668"></a>03668 <span class="stringliteral">            figData.meanCpxConc = meanCpxConc;</span>
<a name="l03669"></a>03669 <span class="stringliteral">            figData.stdCpxConc = stdCpxConc;</span>
<a name="l03670"></a>03670 <span class="stringliteral">        end</span>
<a name="l03671"></a>03671 <span class="stringliteral">        </span>
<a name="l03672"></a>03672 <span class="stringliteral">        function figData = replication(figHandle, simBatchDir, selectedSimulations)</span>
<a name="l03673"></a>03673 <span class="stringliteral">            import edu.stanford.covert.cell.sim.util.PlotUtil;</span>
<a name="l03674"></a>03674 <span class="stringliteral">            import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;</span>
<a name="l03675"></a>03675 <span class="stringliteral">            import edu.stanford.covert.cell.sim.util.SimulationEnsemble;</span>
<a name="l03676"></a>03676 <span class="stringliteral">            </span>
<a name="l03677"></a>03677 <span class="stringliteral">            if nargin &lt; 2 || isempty(simBatchDir)</span>
<a name="l03678"></a>03678 <span class="stringliteral">                simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();</span>
<a name="l03679"></a>03679 <span class="stringliteral">            end</span>
<a name="l03680"></a>03680 <span class="stringliteral">            if nargin &lt; 3 || isempty(selectedSimulations)</span>
<a name="l03681"></a>03681 <span class="stringliteral">                selectedSimulations = SimulationDiskUtil.getCompleteSimulations(simBatchDir);</span>
<a name="l03682"></a>03682 <span class="stringliteral">            end</span>
<a name="l03683"></a>03683 <span class="stringliteral">            </span>
<a name="l03684"></a>03684 <span class="stringliteral">            %% constants</span>
<a name="l03685"></a>03685 <span class="stringliteral">            [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep &#39;</span>1<span class="stringliteral">&#39;]);</span>
<a name="l03686"></a>03686 <span class="stringliteral">            comp = sim.compartment;</span>
<a name="l03687"></a>03687 <span class="stringliteral">            c = sim.state(&#39;</span><a class="code" href="class_chromosome.html">Chromosome</a><span class="stringliteral">&#39;);</span>
<a name="l03688"></a>03688 <span class="stringliteral">            pm = sim.state(&#39;</span>ProteinMonomer<span class="stringliteral">&#39;);</span>
<a name="l03689"></a>03689 <span class="stringliteral">            pc = sim.state(&#39;</span>ProteinComplex<span class="stringliteral">&#39;);</span>
<a name="l03690"></a>03690 <span class="stringliteral">            rep = sim.process(&#39;</span><a class="code" href="class_replication.html">Replication</a><span class="stringliteral">&#39;);</span>
<a name="l03691"></a>03691 <span class="stringliteral">            </span>
<a name="l03692"></a>03692 <span class="stringliteral">            cpxIdxs = [</span>
<a name="l03693"></a>03693 <span class="stringliteral">                pc.matureIndexs(rep.enzymeGlobalIndexs([rep.enzymeIndexs_ssb4mer; rep.enzymeIndexs_ssb8mer]))</span>
<a name="l03694"></a>03694 <span class="stringliteral">                pc.boundIndexs(rep.enzymeGlobalIndexs([rep.enzymeIndexs_ssb4mer; rep.enzymeIndexs_ssb8mer]))</span>
<a name="l03695"></a>03695 <span class="stringliteral">                ]&#39;</span>;
<a name="l03696"></a>03696             
<a name="l03697"></a>03697             nSims = numel(selectedSimulations);
<a name="l03698"></a>03698             
<a name="l03699"></a>03699             %% <span class="keyword">get</span> data
<a name="l03700"></a>03700             stateNames = {
<a name="l03701"></a>03701                 <span class="stringliteral">&#39;mass&#39;</span>         <span class="stringliteral">&#39;Mass (fg)&#39;</span>
<a name="l03702"></a>03702                 <span class="stringliteral">&#39;growth_rate&#39;</span>  <span class="stringliteral">&#39;Growth (fg h^{-1})&#39;</span>
<a name="l03703"></a>03703                 <span class="stringliteral">&#39;leadingPol1&#39;</span>  <span class="stringliteral">&#39;Leading Pol 1&#39;</span>
<a name="l03704"></a>03704                 <span class="stringliteral">&#39;leadingPol2&#39;</span>  <span class="stringliteral">&#39;Leading Pol 2&#39;</span>
<a name="l03705"></a>03705                 <span class="stringliteral">&#39;laggingPol1&#39;</span>  <span class="stringliteral">&#39;Lagging Pol 1&#39;</span>
<a name="l03706"></a>03706                 <span class="stringliteral">&#39;laggingPol2&#39;</span>  <span class="stringliteral">&#39;Lagging Pol 2&#39;</span>
<a name="l03707"></a>03707                 };
<a name="l03708"></a>03708             ensemble = SimulationEnsemble(simBatchDir, stateNames, [], selectedSimulations);
<a name="l03709"></a>03709             mass = permute(ensemble.stateData.values(ensemble.getPropertyIndices(<span class="stringliteral">&#39;mass&#39;</span>), :, :, :) * sim.state(<span class="stringliteral">&#39;Mass&#39;</span>).cellInitialDryWeight / (1-sim.state(&#39;Mass&#39;).fractionWetWeight) * 1e15, [4 3 1 2]);
<a name="l03710"></a>03710             growth = permute(ensemble.stateData.values(ensemble.getPropertyIndices(<span class="stringliteral">&#39;growth_rate&#39;</span>), :, :, :) * 3600 * sim.state(<span class="stringliteral">&#39;Mass&#39;</span>).cellInitialDryWeight / (1 - sim.state(&#39;Mass&#39;).fractionWetWeight) * 1e15, [4 3 1 2]);
<a name="l03711"></a>03711             ledPol1 = permute(ensemble.stateData.values(ensemble.getPropertyIndices(<span class="stringliteral">&#39;leadingPol1&#39;</span>), :, :, :), [4 3 1 2]);
<a name="l03712"></a>03712             ledPol2 = permute(ensemble.stateData.values(ensemble.getPropertyIndices(<span class="stringliteral">&#39;leadingPol2&#39;</span>), :, :, :), [4 3 1 2]);
<a name="l03713"></a>03713             lagPol1 = permute(ensemble.stateData.values(ensemble.getPropertyIndices(<span class="stringliteral">&#39;laggingPol1&#39;</span>), :, :, :), [4 3 1 2]);
<a name="l03714"></a>03714             lagPol2 = permute(ensemble.stateData.values(ensemble.getPropertyIndices(<span class="stringliteral">&#39;laggingPol2&#39;</span>), :, :, :), [4 3 1 2]);
<a name="l03715"></a>03715             
<a name="l03716"></a>03716             stateNames = {
<a name="l03717"></a>03717                 <span class="stringliteral">&#39;ProteinMonomer&#39;</span> <span class="stringliteral">&#39;counts&#39;</span> pm.matureIndexs(rep.enzymeGlobalIndexs(rep.enzymeIndexs_ligase)) comp.cytosolIndexs
<a name="l03718"></a>03718                 <span class="stringliteral">&#39;ProteinComplex&#39;</span> <span class="stringliteral">&#39;counts&#39;</span> cpxIdxs comp.cytosolIndexs
<a name="l03719"></a>03719                 <span class="stringliteral">&#39;Chromosome&#39;</span>     <span class="stringliteral">&#39;strandBreaks&#39;</span> <span class="charliteral">&#39;:&#39;</span> <span class="charliteral">&#39;:&#39;</span>
<a name="l03720"></a>03720                 };
<a name="l03721"></a>03721             states = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, <span class="stringliteral">&#39;extract&#39;</span>, selectedSimulations);
<a name="l03722"></a>03722             ssbs = permute(states.ProteinComplex.counts, [4 3 1 2]);
<a name="l03723"></a>03723             [~, ligationTimes] = find(sum(states.Chromosome.strandBreaks, 3));
<a name="l03724"></a>03724             
<a name="l03725"></a>03725             repStarts = NaN(nSims, 1);
<a name="l03726"></a>03726             repEnds = NaN(nSims, 1);
<a name="l03727"></a>03727             kineticallyLimitedEnds = NaN(nSims, 1);
<a name="l03728"></a>03728             repStartMass = NaN(nSims, 1);
<a name="l03729"></a>03729             repStartGrowth = NaN(nSims, 1);
<a name="l03730"></a>03730             repStartSSbs = NaN(nSims, 1);
<a name="l03731"></a>03731             leadLeadCorr = NaN(nSims, 1);
<a name="l03732"></a>03732             okFragSynTimes = [];
<a name="l03733"></a>03733             <span class="keywordflow">for</span> i = 1:nSims
<a name="l03734"></a>03734                 <span class="keywordflow">if</span> ~all(isnan(ledPol1(i, :)))
<a name="l03735"></a>03735                     repStarts(i) = find(~isnan(ledPol1(i, :)), 1, &#39;first&#39;);
<a name="l03736"></a>03736                     repEnds(i) = find(~isnan(ledPol1(i, :)), 1, &#39;last&#39;);
<a name="l03737"></a>03737                     repStartMass(i) = mass(i, repStarts(i));
<a name="l03738"></a>03738                     repStartGrowth(i) = growth(i, repStarts(i));
<a name="l03739"></a>03739                     repStartSSbs(i) = ssbs(i, repStarts(i) - 1);
<a name="l03740"></a>03740                     
<a name="l03741"></a>03741                     tmp = find(conv(diff(ledPol2(i, :), 1, 2), ones(1, 100), &#39;same&#39;) &gt; 0.75 * 100 * rep.dnaPolymeraseElongationRate, 1, &#39;last&#39;) + 1;
<a name="l03742"></a>03742                     if ~isempty(tmp)
<a name="l03743"></a>03743                         kineticallyLimitedEnds(i) = tmp;
<a name="l03744"></a>03744                     else
<a name="l03745"></a>03745                         kineticallyLimitedEnds(i) = repEnds(i);
<a name="l03746"></a>03746                     end
<a name="l03747"></a>03747                     
<a name="l03748"></a>03748                     leadLeadCorr(i) = -corr(ledPol1(i, ~isnan(ledPol1(i, :)))&#39;, ledPol2(i, ~isnan(ledPol1(i, :)))&#39;);
<a name="l03749"></a>03749                     
<a name="l03750"></a>03750                     okFragSynTimes = [okFragSynTimes diff(find(diff(lagPol1(i, :), 1, 2) &lt; 0)) diff(find(diff(lagPol2(i, :), 1, 2) &gt; 0))]; %<span class="preprocessor">#ok&lt;AGROW&gt;</span>
<a name="l03751"></a>03751 <span class="preprocessor"></span>                end
<a name="l03752"></a>03752             end
<a name="l03753"></a>03753             
<a name="l03754"></a>03754             clear ensemble states mass growth ssbs;
<a name="l03755"></a>03755             
<a name="l03756"></a>03756             %% plot data
<a name="l03757"></a>03757             clf(figHandle);
<a name="l03758"></a>03758             axesHandles = zeros(5, 2);
<a name="l03759"></a>03759             
<a name="l03760"></a>03760             axesHandle = subplot(5, 2, 1, <span class="stringliteral">&#39;Parent&#39;</span>, figHandle);
<a name="l03761"></a>03761             axesHandles(1, 1) = axesHandle;
<a name="l03762"></a>03762             h = plot(axesHandle, repStartMass, [(repEnds - repStarts + 1)  (kineticallyLimitedEnds - repStarts + 1)  (repEnds - kineticallyLimitedEnds + 1)] / 3600, <span class="charliteral">&#39;.&#39;</span>);
<a name="l03763"></a>03763             <span class="keywordflow">if</span> any(repStartMass) &amp;&amp; range(repStartMass)
<a name="l03764"></a>03764                 xlim(axesHandle, [min(repStartMass) max(repStartMass)])
<a name="l03765"></a>03765             end
<a name="l03766"></a>03766             if any(repEnds - repStarts) &amp;&amp; range(repEnds - repStarts)
<a name="l03767"></a>03767                 ylim(axesHandle, [min(repEnds - repStarts + 1) max(repEnds - repStarts + 1)] / 3600)
<a name="l03768"></a>03768             end
<a name="l03769"></a>03769             legend(h, {<span class="stringliteral">&#39;Replication&#39;</span>, <span class="stringliteral">&#39;Kin Lim&#39;</span>, <span class="stringliteral">&#39;Met Lim&#39;</span>}, <span class="stringliteral">&#39;FontSize&#39;</span>, 6)
<a name="l03770"></a>03770             xlabel(axesHandle, <span class="stringliteral">&#39;Mass (fg)&#39;</span>, <span class="stringliteral">&#39;FontSize&#39;</span>, 8);
<a name="l03771"></a>03771             ylabel(axesHandle, <span class="stringliteral">&#39;Duration (h)&#39;</span>, <span class="stringliteral">&#39;FontSize&#39;</span>, 8);
<a name="l03772"></a>03772             
<a name="l03773"></a>03773             axesHandle = subplot(5, 2, 3, <span class="stringliteral">&#39;Parent&#39;</span>, figHandle);
<a name="l03774"></a>03774             axesHandles(2, 1) = axesHandle;
<a name="l03775"></a>03775             plot(axesHandle, repStartGrowth, [(repEnds - repStarts + 1)  (kineticallyLimitedEnds - repStarts + 1)  (repEnds - kineticallyLimitedEnds + 1)] / 3600, <span class="charliteral">&#39;.&#39;</span>);
<a name="l03776"></a>03776             <span class="keywordflow">if</span> any(repStartGrowth) &amp;&amp; range(repStartGrowth)
<a name="l03777"></a>03777                 xlim(axesHandle, [min(repStartGrowth) max(repStartGrowth)])
<a name="l03778"></a>03778             end
<a name="l03779"></a>03779             if any(repEnds - repStarts) &amp;&amp; range(repEnds - repStarts)
<a name="l03780"></a>03780                 ylim(axesHandle, [min(repEnds - repStarts + 1) max(repEnds - repStarts + 1)] / 3600)
<a name="l03781"></a>03781             end
<a name="l03782"></a>03782             xlabel(axesHandle, &#39;Growth (fg h^{-1})&#39;, &#39;FontSize&#39;, 8);
<a name="l03783"></a>03783             ylabel(axesHandle, <span class="stringliteral">&#39;Duration (h)&#39;</span>, <span class="stringliteral">&#39;FontSize&#39;</span>, 8);
<a name="l03784"></a>03784             
<a name="l03785"></a>03785             axesHandle = subplot(5, 2, 5, <span class="stringliteral">&#39;Parent&#39;</span>, figHandle);
<a name="l03786"></a>03786             axesHandles(3, 1) = axesHandle;
<a name="l03787"></a>03787             plot(axesHandle, repStartSSbs, [(repEnds - repStarts + 1)  (kineticallyLimitedEnds - repStarts + 1)  (repEnds - kineticallyLimitedEnds + 1)] / 3600, <span class="charliteral">&#39;.&#39;</span>);
<a name="l03788"></a>03788             <span class="keywordflow">if</span> any(repStartSSbs) &amp;&amp; range(repStartSSbs)
<a name="l03789"></a>03789                 xlim(axesHandle, [min(repStartSSbs) max(repStartSSbs)])
<a name="l03790"></a>03790             end
<a name="l03791"></a>03791             if any(repEnds - repStarts) &amp;&amp; range(repEnds - repStarts)
<a name="l03792"></a>03792                 ylim(axesHandle, [min(repEnds - repStarts + 1) max(repEnds - repStarts + 1)] / 3600)
<a name="l03793"></a>03793             end
<a name="l03794"></a>03794             xlabel(axesHandle, &#39;SSBs&#39;, &#39;FontSize&#39;, 8);
<a name="l03795"></a>03795             ylabel(axesHandle, &#39;Duration (h)&#39;, &#39;FontSize&#39;, 8);
<a name="l03796"></a>03796             
<a name="l03797"></a>03797             axesHandle = subplot(5, 2, 2, &#39;Parent&#39;, figHandle);
<a name="l03798"></a>03798             axesHandles(1, 2) = axesHandle;
<a name="l03799"></a>03799             tmp = [abs(ledPol1(:) - lagPol1(:)); abs(ledPol2(:) - lagPol2(:))] * 1<a class="code" href="json_parse_8m.html#ac14d26badced2970320ee4d772ad0ae9">e</a>-3;
<a name="l03800"></a>03800             hist(tmp)
<a name="l03801"></a>03801             if any(tmp)
<a name="l03802"></a>03802                 xlim(axesHandle, [0.5 max(tmp)+0.5])
<a name="l03803"></a>03803             end
<a name="l03804"></a>03804             ylabel(axesHandle, &#39;Freq&#39;, &#39;FontSize&#39;, 8);
<a name="l03805"></a>03805             xlabel(axesHandle, {<span class="stringliteral">&#39;Lead-Lag Distance (knt)&#39;</span>}, <span class="stringliteral">&#39;FontSize&#39;</span>, 8);
<a name="l03806"></a>03806             
<a name="l03807"></a>03807             axesHandle = subplot(5, 2, 4, <span class="stringliteral">&#39;Parent&#39;</span>, figHandle);
<a name="l03808"></a>03808             axesHandles(2, 2) = axesHandle;
<a name="l03809"></a>03809             tmp = abs((c.sequenceLen - ledPol1(:)) - ledPol2(:)) * 1<a class="code" href="json_parse_8m.html#ac14d26badced2970320ee4d772ad0ae9">e</a>-3;
<a name="l03810"></a>03810             hist(tmp)
<a name="l03811"></a>03811             if any(tmp)
<a name="l03812"></a>03812                 xlim(axesHandle, [0.5 max(tmp)+0.5])
<a name="l03813"></a>03813             end
<a name="l03814"></a>03814             ylabel(axesHandle, &#39;Freq&#39;, &#39;FontSize&#39;, 8);
<a name="l03815"></a>03815             xlabel(axesHandle, {<span class="stringliteral">&#39;Lead-Lead Distance (knt)&#39;</span>}, <span class="stringliteral">&#39;FontSize&#39;</span>, 8);
<a name="l03816"></a>03816             
<a name="l03817"></a>03817             axesHandle = subplot(5, 2, 6, <span class="stringliteral">&#39;Parent&#39;</span>, figHandle);
<a name="l03818"></a>03818             axesHandles(3, 2) = axesHandle;
<a name="l03819"></a>03819             hist(leadLeadCorr)
<a name="l03820"></a>03820             if any(leadLeadCorr) &amp;&amp; range(leadLeadCorr)
<a name="l03821"></a>03821                 xlim(axesHandle, [min(leadLeadCorr) max(leadLeadCorr)])
<a name="l03822"></a>03822             end
<a name="l03823"></a>03823             ylabel(axesHandle, &#39;Freq&#39;, &#39;FontSize&#39;, 8);
<a name="l03824"></a>03824             xlabel(axesHandle, {<span class="stringliteral">&#39;Lead-Lead Corr&#39;</span>}, <span class="stringliteral">&#39;FontSize&#39;</span>, 8);
<a name="l03825"></a>03825             
<a name="l03826"></a>03826             axesHandle = subplot(5, 2, 8, <span class="stringliteral">&#39;Parent&#39;</span>, figHandle);
<a name="l03827"></a>03827             axesHandles(4, 2) = axesHandle;
<a name="l03828"></a>03828             hist(okFragSynTimes)
<a name="l03829"></a>03829             if ~isempty(okFragSynTimes) &amp;&amp; range(okFragSynTimes)
<a name="l03830"></a>03830                 xlim(axesHandle, [min(okFragSynTimes) max(okFragSynTimes)])
<a name="l03831"></a>03831             end
<a name="l03832"></a>03832             ylabel(axesHandle, &#39;Freq&#39;, &#39;FontSize&#39;, 8);
<a name="l03833"></a>03833             xlabel(axesHandle, {<span class="stringliteral">&#39;Okazaki Synthesis Time (s)&#39;</span>}, <span class="stringliteral">&#39;FontSize&#39;</span>, 8);
<a name="l03834"></a>03834             
<a name="l03835"></a>03835             axesHandle = subplot(5, 2, 10, <span class="stringliteral">&#39;Parent&#39;</span>, figHandle);
<a name="l03836"></a>03836             axesHandles(5, 2) = axesHandle;
<a name="l03837"></a>03837             hist(ligationTimes)
<a name="l03838"></a>03838             if ~isempty(ligationTimes) &amp;&amp; range(ligationTimes)
<a name="l03839"></a>03839                 xlim(axesHandle, [min(ligationTimes) max(ligationTimes)])
<a name="l03840"></a>03840             end
<a name="l03841"></a>03841             ylabel(axesHandle, &#39;Freq&#39;, &#39;FontSize&#39;, 8);
<a name="l03842"></a>03842             xlabel(axesHandle, {<span class="stringliteral">&#39;Okazaki Synthesis Time (s)&#39;</span>}, <span class="stringliteral">&#39;FontSize&#39;</span>, 8);
<a name="l03843"></a>03843             
<a name="l03844"></a>03844             %format axes
<a name="l03845"></a>03845             <span class="keywordflow">for</span> i = 1:numel(axesHandles)
<a name="l03846"></a>03846                 if ~axesHandles(i)
<a name="l03847"></a>03847                     continue;
<a name="l03848"></a>03848                 end
<a name="l03849"></a>03849                 set(axesHandles(i), &#39;FontSize&#39;, 6);
<a name="l03850"></a>03850             end
<a name="l03851"></a>03851             
<a name="l03852"></a>03852             %% organize figure data
<a name="l03853"></a>03853             figData = struct;
<a name="l03854"></a>03854             figData.ledPol1 = ledPol1;
<a name="l03855"></a>03855             figData.ledPol2 = ledPol2;
<a name="l03856"></a>03856             figData.lagPol1 = lagPol1;
<a name="l03857"></a>03857             figData.lagPol2 = lagPol2;
<a name="l03858"></a>03858             figData.ligationTimes = ligationTimes;
<a name="l03859"></a>03859             figData.repStarts = repStarts;
<a name="l03860"></a>03860             figData.repEnds = repEnds;
<a name="l03861"></a>03861             figData.kineticallyLimitedEnds = kineticallyLimitedEnds;
<a name="l03862"></a>03862             figData.repStartMass = repStartMass;
<a name="l03863"></a>03863             figData.repStartGrowth = repStartGrowth;
<a name="l03864"></a>03864             figData.repStartSSbs = repStartSSbs;
<a name="l03865"></a>03865             figData.leadLeadCorr = leadLeadCorr;
<a name="l03866"></a>03866             figData.okFragSynTimes = okFragSynTimes;
<a name="l03867"></a>03867         end
<a name="l03868"></a>03868         
<a name="l03869"></a>03869         function figData = SSBs(figHandle, simBatchDir, selectedSimulations)
<a name="l03870"></a>03870             import edu.stanford.covert.cell.sim.util.PlotUtil;
<a name="l03871"></a>03871             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
<a name="l03872"></a>03872             import edu.stanford.covert.cell.sim.util.SimulationEnsemble;
<a name="l03873"></a>03873             
<a name="l03874"></a>03874             if nargin &lt; 2 || isempty(simBatchDir)
<a name="l03875"></a>03875                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
<a name="l03876"></a>03876             end
<a name="l03877"></a>03877             if nargin &lt; 3 || isempty(selectedSimulations)
<a name="l03878"></a>03878                 selectedSimulations = SimulationDiskUtil.getCompleteSimulations(simBatchDir);
<a name="l03879"></a>03879             end
<a name="l03880"></a>03880             
<a name="l03881"></a>03881             %% constants
<a name="l03882"></a>03882             [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep &#39;1&#39;]);
<a name="l03883"></a>03883             comp = sim.compartment;
<a name="l03884"></a>03884             pc = sim.state(&#39;ProteinComplex&#39;);
<a name="l03885"></a>03885             
<a name="l03886"></a>03886             cytosolIndexs = comp.cytosolIndexs;
<a name="l03887"></a>03887             cpxIdxs = [
<a name="l03888"></a>03888                 pc.matureIndexs(pc.getIndexs(&#39;MG_091_TETRAMER&#39;))
<a name="l03889"></a>03889                 pc.matureIndexs(pc.getIndexs(&#39;MG_091_OCTAMER&#39;))
<a name="l03890"></a>03890                 pc.boundIndexs(pc.getIndexs(&#39;MG_091_OCTAMER&#39;))
<a name="l03891"></a>03891                 ];
<a name="l03892"></a>03892             
<a name="l03893"></a>03893             %% get data
<a name="l03894"></a>03894             stateNames = {
<a name="l03895"></a>03895                 <span class="stringliteral">&#39;Time&#39;</span>            <span class="stringliteral">&#39;values&#39;</span>  <span class="charliteral">&#39;:&#39;</span>      <span class="charliteral">&#39;:&#39;</span>
<a name="l03896"></a>03896                 <span class="stringliteral">&#39;Mass&#39;</span>            <span class="stringliteral">&#39;cell&#39;</span>    <span class="charliteral">&#39;:&#39;</span>      <span class="stringliteral">&#39;-sum&#39;</span>
<a name="l03897"></a>03897                 <span class="stringliteral">&#39;ProteinComplex&#39;</span>  <span class="stringliteral">&#39;counts&#39;</span>  cpxIdxs  cytosolIndexs
<a name="l03898"></a>03898                 };
<a name="l03899"></a>03899             tmpStates = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, <span class="stringliteral">&#39;extract&#39;</span>, selectedSimulations);
<a name="l03900"></a>03900             
<a name="l03901"></a>03901             mass = permute(tmpStates.Mass.cell, [3 4 1 2]);
<a name="l03902"></a>03902             simEndTimes = permute(max(tmpStates.Time.values, [], 3), [4 3 1 2]);
<a name="l03903"></a>03903             minSimEndTime = min(simEndTimes);
<a name="l03904"></a>03904             [~, simOrder] = sort(mass(minSimEndTime, :));
<a name="l03905"></a>03905             
<a name="l03906"></a>03906             states.ProteinComplex.free4mer = permute(tmpStates.ProteinComplex.counts(1, :, :, :), [3 4 1 2]);
<a name="l03907"></a>03907             states.ProteinComplex.free8mer = permute(tmpStates.ProteinComplex.counts(2, :, :, :), [3 4 1 2]);
<a name="l03908"></a>03908             states.ProteinComplex.bound8mer = permute(tmpStates.ProteinComplex.counts(3, :, :, :), [3 4 1 2]);
<a name="l03909"></a>03909             
<a name="l03910"></a>03910             time = (1:size(states.ProteinComplex.bound8mer, 1))<span class="stringliteral">&#39; / 3600;</span>
<a name="l03911"></a>03911 <span class="stringliteral">            </span>
<a name="l03912"></a>03912 <span class="stringliteral">            clear tmpStates mass;</span>
<a name="l03913"></a>03913 <span class="stringliteral">            </span>
<a name="l03914"></a>03914 <span class="stringliteral">            %% plot</span>
<a name="l03915"></a>03915 <span class="stringliteral">            clf(figHandle);</span>
<a name="l03916"></a>03916 <span class="stringliteral">            options = struct;</span>
<a name="l03917"></a>03917 <span class="stringliteral">            options.colorOrder = PlotUtil.getRedGreenColorOrder(simOrder);</span>
<a name="l03918"></a>03918 <span class="stringliteral">            options.xdata = time;</span>
<a name="l03919"></a>03919 <span class="stringliteral">            options.ylabelStr = {</span>
<a name="l03920"></a>03920 <span class="stringliteral">                {&#39;</span>Bound<span class="charliteral">&#39; &#39;</span>8mer<span class="stringliteral">&#39;}</span>
<a name="l03921"></a>03921 <span class="stringliteral">                {&#39;</span>Free<span class="charliteral">&#39; &#39;</span>8mer<span class="stringliteral">&#39;}</span>
<a name="l03922"></a>03922 <span class="stringliteral">                {&#39;</span>Free<span class="charliteral">&#39; &#39;</span>4mer<span class="stringliteral">&#39;}</span>
<a name="l03923"></a>03923 <span class="stringliteral">                };</span>
<a name="l03924"></a>03924 <span class="stringliteral">            options.ydata = {</span>
<a name="l03925"></a>03925 <span class="stringliteral">                states.ProteinComplex.bound8mer</span>
<a name="l03926"></a>03926 <span class="stringliteral">                states.ProteinComplex.free8mer</span>
<a name="l03927"></a>03927 <span class="stringliteral">                states.ProteinComplex.free4mer</span>
<a name="l03928"></a>03928 <span class="stringliteral">                };</span>
<a name="l03929"></a>03929 <span class="stringliteral">            PlotUtil.multiElementPlot(figHandle, [10; 10; 10], [0 time(end)], options);</span>
<a name="l03930"></a>03930 <span class="stringliteral">            </span>
<a name="l03931"></a>03931 <span class="stringliteral">            clear options;</span>
<a name="l03932"></a>03932 <span class="stringliteral">            </span>
<a name="l03933"></a>03933 <span class="stringliteral">            %% organize figure data</span>
<a name="l03934"></a>03934 <span class="stringliteral">            figData = struct;</span>
<a name="l03935"></a>03935 <span class="stringliteral">            figData.time = time;</span>
<a name="l03936"></a>03936 <span class="stringliteral">            figData.free4mer = states.ProteinComplex.free4mer;</span>
<a name="l03937"></a>03937 <span class="stringliteral">            figData.free8mer = states.ProteinComplex.free8mer;</span>
<a name="l03938"></a>03938 <span class="stringliteral">            figData.bound8mer = states.ProteinComplex.bound8mer;</span>
<a name="l03939"></a>03939 <span class="stringliteral">        end</span>
<a name="l03940"></a>03940 <span class="stringliteral">        </span>
<a name="l03941"></a>03941 <span class="stringliteral">        function figData = dnaRepair(figHandle, simBatchDir, selectedSimulations)</span>
<a name="l03942"></a>03942 <span class="stringliteral">            import edu.stanford.covert.cell.sim.util.PlotUtil;</span>
<a name="l03943"></a>03943 <span class="stringliteral">            import edu.stanford.covert.cell.sim.util.SimulationEnsemble;</span>
<a name="l03944"></a>03944 <span class="stringliteral">            import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;</span>
<a name="l03945"></a>03945 <span class="stringliteral">            import edu.stanford.covert.util.CircularSparseMat;</span>
<a name="l03946"></a>03946 <span class="stringliteral">            import edu.stanford.covert.util.ConstantUtil;</span>
<a name="l03947"></a>03947 <span class="stringliteral">            </span>
<a name="l03948"></a>03948 <span class="stringliteral">            if nargin &lt; 2 || isempty(simBatchDir)</span>
<a name="l03949"></a>03949 <span class="stringliteral">                simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();</span>
<a name="l03950"></a>03950 <span class="stringliteral">            end</span>
<a name="l03951"></a>03951 <span class="stringliteral">            if nargin &lt; 3 || isempty(selectedSimulations)</span>
<a name="l03952"></a>03952 <span class="stringliteral">                selectedSimulations = SimulationDiskUtil.getCompleteSimulations(simBatchDir);</span>
<a name="l03953"></a>03953 <span class="stringliteral">            end</span>
<a name="l03954"></a>03954 <span class="stringliteral">            </span>
<a name="l03955"></a>03955 <span class="stringliteral">            [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep &#39;</span>1<span class="stringliteral">&#39;]);</span>
<a name="l03956"></a>03956 <span class="stringliteral">            pm = sim.state(&#39;</span>ProteinMonomer<span class="stringliteral">&#39;);</span>
<a name="l03957"></a>03957 <span class="stringliteral">            pc = sim.state(&#39;</span>ProteinComplex<span class="stringliteral">&#39;);</span>
<a name="l03958"></a>03958 <span class="stringliteral">            dnaRepair = sim.process(&#39;</span>DNARepair<span class="stringliteral">&#39;);</span>
<a name="l03959"></a>03959 <span class="stringliteral">            </span>
<a name="l03960"></a>03960 <span class="stringliteral">            monIdxs = pm.matureIndexs(dnaRepair.enzymeMonomerGlobalIndexs);</span>
<a name="l03961"></a>03961 <span class="stringliteral">            cpxIdxs = pc.matureIndexs(dnaRepair.enzymeComplexGlobalIndexs);</span>
<a name="l03962"></a>03962 <span class="stringliteral">            </span>
<a name="l03963"></a>03963 <span class="stringliteral">            %% get data</span>
<a name="l03964"></a>03964 <span class="stringliteral">            stateNames = {</span>
<a name="l03965"></a>03965 <span class="stringliteral">                &#39;</span>Time<span class="stringliteral">&#39;            &#39;</span>values<span class="stringliteral">&#39;                  &#39;</span>:<span class="stringliteral">&#39;      &#39;</span>:<span class="stringliteral">&#39;</span>
<a name="l03966"></a>03966 <span class="stringliteral">                &#39;</span><a class="code" href="class_chromosome.html">Chromosome</a><span class="stringliteral">&#39;      &#39;</span>abasicSites<span class="stringliteral">&#39;             &#39;</span>-sum<span class="stringliteral">&#39;   &#39;</span>-sum<span class="stringliteral">&#39;</span>
<a name="l03967"></a>03967 <span class="stringliteral">                &#39;</span><a class="code" href="class_chromosome.html">Chromosome</a><span class="stringliteral">&#39;      &#39;</span>gapSites<span class="stringliteral">&#39;                &#39;</span>-sum<span class="stringliteral">&#39;   &#39;</span>-sum<span class="stringliteral">&#39;</span>
<a name="l03968"></a>03968 <span class="stringliteral">                &#39;</span><a class="code" href="class_chromosome.html">Chromosome</a><span class="stringliteral">&#39;      &#39;</span>damagedSugarPhosphates<span class="stringliteral">&#39;  &#39;</span>-sum<span class="stringliteral">&#39;   &#39;</span>-sum<span class="stringliteral">&#39;</span>
<a name="l03969"></a>03969 <span class="stringliteral">                &#39;</span><a class="code" href="class_chromosome.html">Chromosome</a><span class="stringliteral">&#39;      &#39;</span>intrastrandCrossLinks<span class="stringliteral">&#39;   &#39;</span>-sum<span class="stringliteral">&#39;   &#39;</span>-sum<span class="stringliteral">&#39;</span>
<a name="l03970"></a>03970 <span class="stringliteral">                &#39;</span><a class="code" href="class_chromosome.html">Chromosome</a><span class="stringliteral">&#39;      &#39;</span>strandBreaks<span class="stringliteral">&#39;            &#39;</span>-sum<span class="stringliteral">&#39;   &#39;</span>-sum<span class="stringliteral">&#39;</span>
<a name="l03971"></a>03971 <span class="stringliteral">                &#39;</span><a class="code" href="class_chromosome.html">Chromosome</a><span class="stringliteral">&#39;      &#39;</span>hollidayJunctions<span class="stringliteral">&#39;       &#39;</span>-sum<span class="stringliteral">&#39;   &#39;</span>-sum<span class="stringliteral">&#39;</span>
<a name="l03972"></a>03972 <span class="stringliteral">                &#39;</span>Mass<span class="stringliteral">&#39;            &#39;</span>cell<span class="stringliteral">&#39;                    &#39;</span>:<span class="stringliteral">&#39;      &#39;</span>-sum<span class="stringliteral">&#39;</span>
<a name="l03973"></a>03973 <span class="stringliteral">                &#39;</span>ProteinMonomer<span class="stringliteral">&#39;  &#39;</span>counts<span class="stringliteral">&#39;                  monIdxs  &#39;</span>-sum<span class="stringliteral">&#39;</span>
<a name="l03974"></a>03974 <span class="stringliteral">                &#39;</span>ProteinComplex<span class="stringliteral">&#39;  &#39;</span>counts<span class="stringliteral">&#39;                  cpxIdxs  &#39;</span>-sum<span class="stringliteral">&#39;</span>
<a name="l03975"></a>03975 <span class="stringliteral">                };</span>
<a name="l03976"></a>03976 <span class="stringliteral">            states = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, &#39;</span>extract<span class="stringliteral">&#39;, selectedSimulations);</span>
<a name="l03977"></a>03977 <span class="stringliteral">            </span>
<a name="l03978"></a>03978 <span class="stringliteral">            time = permute(max(states.Time.values, [], 4), [3 4 1 2]) / 3600;</span>
<a name="l03979"></a>03979 <span class="stringliteral">            mass = permute(states.Mass.cell, [3 4 1 2]);</span>
<a name="l03980"></a>03980 <span class="stringliteral">            </span>
<a name="l03981"></a>03981 <span class="stringliteral">            simEndTimes = permute(max(states.Time.values, [], 3), [4 3 1 2]);</span>
<a name="l03982"></a>03982 <span class="stringliteral">            minSimEndTime = min(simEndTimes);</span>
<a name="l03983"></a>03983 <span class="stringliteral">            [~, simOrder] = sort(mass(minSimEndTime, :));</span>
<a name="l03984"></a>03984 <span class="stringliteral">            </span>
<a name="l03985"></a>03985 <span class="stringliteral">            clear mass;</span>
<a name="l03986"></a>03986 <span class="stringliteral">            </span>
<a name="l03987"></a>03987 <span class="stringliteral">            %damage</span>
<a name="l03988"></a>03988 <span class="stringliteral">            gapSites = full(states.Chromosome.gapSites);</span>
<a name="l03989"></a>03989 <span class="stringliteral">            abasicSites = full(states.Chromosome.abasicSites);</span>
<a name="l03990"></a>03990 <span class="stringliteral">            damagedSugarPhosphates = full(states.Chromosome.damagedSugarPhosphates);</span>
<a name="l03991"></a>03991 <span class="stringliteral">            intrastrandCrossLinks = full(states.Chromosome.intrastrandCrossLinks);</span>
<a name="l03992"></a>03992 <span class="stringliteral">            strandBreaks = full(states.Chromosome.strandBreaks);</span>
<a name="l03993"></a>03993 <span class="stringliteral">            hollidayJunctions = full(states.Chromosome.hollidayJunctions);</span>
<a name="l03994"></a>03994 <span class="stringliteral">            </span>
<a name="l03995"></a>03995 <span class="stringliteral">            %enzymes</span>
<a name="l03996"></a>03996 <span class="stringliteral">            enzymes(dnaRepair.enzymeMonomerLocalIndexs, :, :, :) = states.ProteinMonomer.counts;</span>
<a name="l03997"></a>03997 <span class="stringliteral">            enzymes(dnaRepair.enzymeComplexLocalIndexs, :, :, :) = states.ProteinComplex.counts;</span>
<a name="l03998"></a>03998 <span class="stringliteral">            </span>
<a name="l03999"></a>03999 <span class="stringliteral">            %cleanup</span>
<a name="l04000"></a>04000 <span class="stringliteral">            clear states;</span>
<a name="l04001"></a>04001 <span class="stringliteral">            </span>
<a name="l04002"></a>04002 <span class="stringliteral">            %damaged bases -- calculate separately to minimize memory profile</span>
<a name="l04003"></a>04003 <span class="stringliteral">            damagedBases = zeros(size(gapSites));</span>
<a name="l04004"></a>04004 <span class="stringliteral">            for i = 1:numel(selectedSimulations)</span>
<a name="l04005"></a>04005 <span class="stringliteral">                tmpStates = SimulationEnsemble.load(simBatchDir, {&#39;</span><a class="code" href="class_chromosome.html">Chromosome</a><span class="charliteral">&#39; &#39;</span>damagedBases<span class="stringliteral">&#39;}, [], [], 1, &#39;</span>extract<span class="stringliteral">&#39;, selectedSimulations(i));</span>
<a name="l04006"></a>04006 <span class="stringliteral">                [posStrnds, vals] = find(tmpStates.Chromosome.damagedBases);</span>
<a name="l04007"></a>04007 <span class="stringliteral">                tfs = ~ismember(vals, dnaRepair.substrateGlobalIndexs(dnaRepair.substrateIndexs_m6AD));</span>
<a name="l04008"></a>04008 <span class="stringliteral">                damagedBases(:, :, 1:size(tmpStates.Chromosome.damagedBases, 3), i) = ...</span>
<a name="l04009"></a>04009 <span class="stringliteral">                    full(sum(sum(CircularSparseMat(posStrnds(tfs, :), ones(sum(tfs), 1), size(tmpStates.Chromosome.damagedBases), 1), 2), 1));</span>
<a name="l04010"></a>04010 <span class="stringliteral">                clear tmpStates posStrnds vals tfs;</span>
<a name="l04011"></a>04011 <span class="stringliteral">            end</span>
<a name="l04012"></a>04012 <span class="stringliteral">            </span>
<a name="l04013"></a>04013 <span class="stringliteral">            %NaN</span>
<a name="l04014"></a>04014 <span class="stringliteral">            for i = 1:numel(simEndTimes)</span>
<a name="l04015"></a>04015 <span class="stringliteral">                gapSites(:, :, simEndTimes(i)+1:end, i) = NaN;</span>
<a name="l04016"></a>04016 <span class="stringliteral">                abasicSites(:, :, simEndTimes(i)+1:end, i) = NaN;</span>
<a name="l04017"></a>04017 <span class="stringliteral">                damagedSugarPhosphates(:, :, simEndTimes(i)+1:end, i) = NaN;</span>
<a name="l04018"></a>04018 <span class="stringliteral">                intrastrandCrossLinks(:, :, simEndTimes(i)+1:end, i) = NaN;</span>
<a name="l04019"></a>04019 <span class="stringliteral">                strandBreaks(:, :, simEndTimes(i)+1:end, i) = NaN;</span>
<a name="l04020"></a>04020 <span class="stringliteral">                hollidayJunctions(:, :, simEndTimes(i)+1:end, i) = NaN;</span>
<a name="l04021"></a>04021 <span class="stringliteral">                enzymes(:, :, simEndTimes(i)+1:end, i) = NaN;</span>
<a name="l04022"></a>04022 <span class="stringliteral">            end</span>
<a name="l04023"></a>04023 <span class="stringliteral">            </span>
<a name="l04024"></a>04024 <span class="stringliteral">            %% plot data</span>
<a name="l04025"></a>04025 <span class="stringliteral">            clf(figHandle);</span>
<a name="l04026"></a>04026 <span class="stringliteral">            </span>
<a name="l04027"></a>04027 <span class="stringliteral">            options = struct();</span>
<a name="l04028"></a>04028 <span class="stringliteral">            options.titleStr = {&#39;</span>Damage<span class="stringliteral">&#39;; &#39;</span>Repair<span class="stringliteral">&#39;};</span>
<a name="l04029"></a>04029 <span class="stringliteral">            options.colorOrder = PlotUtil.getRedGreenColorOrder(simOrder);</span>
<a name="l04030"></a>04030 <span class="stringliteral">            options.xdata = time;</span>
<a name="l04031"></a>04031 <span class="stringliteral">            options.ylabelStr = {</span>
<a name="l04032"></a>04032 <span class="stringliteral">                {</span>
<a name="l04033"></a>04033 <span class="stringliteral">                {&#39;</span>Gap<span class="charliteral">&#39; &#39;</span>Sites<span class="stringliteral">&#39;}</span>
<a name="l04034"></a>04034 <span class="stringliteral">                {&#39;</span>Abasic<span class="charliteral">&#39; &#39;</span>Sites<span class="stringliteral">&#39;}</span>
<a name="l04035"></a>04035 <span class="stringliteral">                {&#39;</span>Damaged<span class="charliteral">&#39; &#39;</span>Sugar-Phosphates<span class="stringliteral">&#39;}</span>
<a name="l04036"></a>04036 <span class="stringliteral">                {&#39;</span>Damaged<span class="charliteral">&#39; &#39;</span>Bases<span class="stringliteral">&#39;}</span>
<a name="l04037"></a>04037 <span class="stringliteral">                {&#39;</span>Intrastrand<span class="charliteral">&#39; &#39;</span>Crosslinks<span class="stringliteral">&#39;}</span>
<a name="l04038"></a>04038 <span class="stringliteral">                {&#39;</span>Strand<span class="charliteral">&#39; &#39;</span>Breaks<span class="stringliteral">&#39;}</span>
<a name="l04039"></a>04039 <span class="stringliteral">                {&#39;</span>Holliday<span class="charliteral">&#39; &#39;</span>Junctions<span class="stringliteral">&#39;}</span>
<a name="l04040"></a>04040 <span class="stringliteral">                }</span>
<a name="l04041"></a>04041 <span class="stringliteral">                dnaRepair.enzymeWholeCellModelIDs</span>
<a name="l04042"></a>04042 <span class="stringliteral">                };</span>
<a name="l04043"></a>04043 <span class="stringliteral">            options.ydata = {</span>
<a name="l04044"></a>04044 <span class="stringliteral">                {</span>
<a name="l04045"></a>04045 <span class="stringliteral">                permute(gapSites, [3 4 1 2])</span>
<a name="l04046"></a>04046 <span class="stringliteral">                permute(abasicSites, [3 4 1 2])</span>
<a name="l04047"></a>04047 <span class="stringliteral">                permute(damagedSugarPhosphates, [3 4 1 2])</span>
<a name="l04048"></a>04048 <span class="stringliteral">                permute(damagedBases, [3 4 1 2])</span>
<a name="l04049"></a>04049 <span class="stringliteral">                permute(intrastrandCrossLinks, [3 4 1 2])</span>
<a name="l04050"></a>04050 <span class="stringliteral">                permute(strandBreaks, [3 4 1 2])</span>
<a name="l04051"></a>04051 <span class="stringliteral">                permute(hollidayJunctions, [3 4 1 2])</span>
<a name="l04052"></a>04052 <span class="stringliteral">                }</span>
<a name="l04053"></a>04053 <span class="stringliteral">                permute(mat2cell(permute(enzymes, [3 4 1 2]), size(enzymes, 3), size(enzymes, 4), ones(size(enzymes, 1), 1)), [3 1 2])</span>
<a name="l04054"></a>04054 <span class="stringliteral">                };</span>
<a name="l04055"></a>04055 <span class="stringliteral">            PlotUtil.multiElementPlot(figHandle, cellfun(@(x) 3 * ones(size(x)), options.ylabelStr, &#39;</span>UniformOutput<span class="stringliteral">&#39;, false), [0 time(end)], options);</span>
<a name="l04056"></a>04056 <span class="stringliteral">            </span>
<a name="l04057"></a>04057 <span class="stringliteral">            clear options;</span>
<a name="l04058"></a>04058 <span class="stringliteral">            </span>
<a name="l04059"></a>04059 <span class="stringliteral">            %% organize figure data</span>
<a name="l04060"></a>04060 <span class="stringliteral">            figData = struct;</span>
<a name="l04061"></a>04061 <span class="stringliteral">            figData.simOrder = simOrder;</span>
<a name="l04062"></a>04062 <span class="stringliteral">            figData.time = time;</span>
<a name="l04063"></a>04063 <span class="stringliteral">            figData.gapSites = gapSites;</span>
<a name="l04064"></a>04064 <span class="stringliteral">            figData.abasicSites = abasicSites;</span>
<a name="l04065"></a>04065 <span class="stringliteral">            figData.damagedSugarPhosphates = damagedSugarPhosphates;</span>
<a name="l04066"></a>04066 <span class="stringliteral">            figData.damagedBases = damagedBases;</span>
<a name="l04067"></a>04067 <span class="stringliteral">            figData.intrastrandCrossLinks = intrastrandCrossLinks;</span>
<a name="l04068"></a>04068 <span class="stringliteral">            figData.strandBreaks = strandBreaks;</span>
<a name="l04069"></a>04069 <span class="stringliteral">            figData.hollidayJunctions = hollidayJunctions;</span>
<a name="l04070"></a>04070 <span class="stringliteral">            figData.enzymes = enzymes;</span>
<a name="l04071"></a>04071 <span class="stringliteral">        end</span>
<a name="l04072"></a>04072 <span class="stringliteral">        </span>
<a name="l04073"></a>04073 <span class="stringliteral">        function figData = immuneActivation(figHandle, simBatchDir, selectedSimulations)</span>
<a name="l04074"></a>04074 <span class="stringliteral">            import edu.stanford.covert.cell.sim.util.PlotUtil;</span>
<a name="l04075"></a>04075 <span class="stringliteral">            import edu.stanford.covert.cell.sim.util.SimulationEnsemble;</span>
<a name="l04076"></a>04076 <span class="stringliteral">            import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;</span>
<a name="l04077"></a>04077 <span class="stringliteral">            import edu.stanford.covert.util.ConstantUtil;</span>
<a name="l04078"></a>04078 <span class="stringliteral">            </span>
<a name="l04079"></a>04079 <span class="stringliteral">            if nargin &lt; 2 || isempty(simBatchDir)</span>
<a name="l04080"></a>04080 <span class="stringliteral">                simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();</span>
<a name="l04081"></a>04081 <span class="stringliteral">            end</span>
<a name="l04082"></a>04082 <span class="stringliteral">            if nargin &lt; 3 || isempty(selectedSimulations)</span>
<a name="l04083"></a>04083 <span class="stringliteral">                selectedSimulations = SimulationDiskUtil.getCompleteSimulations(simBatchDir);</span>
<a name="l04084"></a>04084 <span class="stringliteral">            end</span>
<a name="l04085"></a>04085 <span class="stringliteral">            </span>
<a name="l04086"></a>04086 <span class="stringliteral">            [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep &#39;</span>1<span class="stringliteral">&#39;]);</span>
<a name="l04087"></a>04087 <span class="stringliteral">            comp = sim.compartment;</span>
<a name="l04088"></a>04088 <span class="stringliteral">            pm = sim.state(&#39;</span>ProteinMonomer<span class="stringliteral">&#39;);</span>
<a name="l04089"></a>04089 <span class="stringliteral">            host = sim.state(&#39;</span>Host<span class="stringliteral">&#39;);</span>
<a name="l04090"></a>04090 <span class="stringliteral">            </span>
<a name="l04091"></a>04091 <span class="stringliteral">            %% get data</span>
<a name="l04092"></a>04092 <span class="stringliteral">            monIdxs = pm.matureIndexs(ismember(pm.compartments(pm.matureIndexs), [comp.terminalOrganelleCytosolIndexs comp.terminalOrganelleMembraneIndexs]));</span>
<a name="l04093"></a>04093 <span class="stringliteral">            stateNames = {</span>
<a name="l04094"></a>04094 <span class="stringliteral">                &#39;</span>Time<span class="stringliteral">&#39;            &#39;</span>values<span class="stringliteral">&#39;                           &#39;</span>:<span class="charliteral">&#39; &#39;</span>:<span class="stringliteral">&#39;</span>
<a name="l04095"></a>04095 <span class="stringliteral">                &#39;</span>Mass<span class="stringliteral">&#39;            &#39;</span>cell<span class="stringliteral">&#39;                             &#39;</span>:<span class="charliteral">&#39; &#39;</span>-sum<span class="stringliteral">&#39;</span>
<a name="l04096"></a>04096 <span class="stringliteral">                &#39;</span>Host<span class="stringliteral">&#39;            &#39;</span>isBacteriumAdherent<span class="stringliteral">&#39;              &#39;</span>:<span class="charliteral">&#39; &#39;</span>:<span class="stringliteral">&#39;</span>
<a name="l04097"></a>04097 <span class="stringliteral">                &#39;</span>Host<span class="stringliteral">&#39;            &#39;</span>isTLRActivated<span class="stringliteral">&#39;                   &#39;</span>:<span class="charliteral">&#39; &#39;</span>:<span class="stringliteral">&#39;</span>
<a name="l04098"></a>04098 <span class="stringliteral">                &#39;</span>Host<span class="stringliteral">&#39;            &#39;</span>isNFkBActivated<span class="stringliteral">&#39;                  &#39;</span>:<span class="charliteral">&#39; &#39;</span>:<span class="stringliteral">&#39;</span>
<a name="l04099"></a>04099 <span class="stringliteral">                &#39;</span>Host<span class="stringliteral">&#39;            &#39;</span>isInflammatoryResponseActivated<span class="stringliteral">&#39;  &#39;</span>:<span class="charliteral">&#39; &#39;</span>:<span class="stringliteral">&#39;</span>
<a name="l04100"></a>04100 <span class="stringliteral">                &#39;</span>ProteinMonomer<span class="stringliteral">&#39;  &#39;</span>counts<span class="stringliteral">&#39;                           monIdxs [comp.terminalOrganelleCytosolIndexs comp.terminalOrganelleMembraneIndexs]</span>
<a name="l04101"></a>04101 <span class="stringliteral">                };</span>
<a name="l04102"></a>04102 <span class="stringliteral">            states = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, &#39;</span>extract<span class="stringliteral">&#39;, selectedSimulations);</span>
<a name="l04103"></a>04103 <span class="stringliteral">            </span>
<a name="l04104"></a>04104 <span class="stringliteral">            time = permute(max(states.Time.values, [], 4), [3 4 1 2]) / 3600;</span>
<a name="l04105"></a>04105 <span class="stringliteral">            mass = permute(states.Mass.cell, [3 4 1 2]);</span>
<a name="l04106"></a>04106 <span class="stringliteral">            adherence = double(permute(states.Host.isBacteriumAdherent, [3 4 1 2]));</span>
<a name="l04107"></a>04107 <span class="stringliteral">            tlr = double(permute(states.Host.isTLRActivated, [3 4 1 2]));</span>
<a name="l04108"></a>04108 <span class="stringliteral">            nfkb = double(permute(states.Host.isNFkBActivated, [3 4 1 2]));</span>
<a name="l04109"></a>04109 <span class="stringliteral">            inflammation = double(permute(states.Host.isInflammatoryResponseActivated, [3 4 1 2]));</span>
<a name="l04110"></a>04110 <span class="stringliteral">            proteins = permute(states.ProteinMonomer.counts, [3 4 1 2]);</span>
<a name="l04111"></a>04111 <span class="stringliteral">            </span>
<a name="l04112"></a>04112 <span class="stringliteral">            simEndTimes = permute(max(states.Time.values, [], 3), [4 3 1 2]);</span>
<a name="l04113"></a>04113 <span class="stringliteral">            minSimEndTime = min(simEndTimes);</span>
<a name="l04114"></a>04114 <span class="stringliteral">            [~, simOrder] = sort(mass(minSimEndTime, :));</span>
<a name="l04115"></a>04115 <span class="stringliteral">            </span>
<a name="l04116"></a>04116 <span class="stringliteral">            for i = 1:numel(simEndTimes)</span>
<a name="l04117"></a>04117 <span class="stringliteral">                adherence(simEndTimes(i)+1:end, i, :) = NaN;</span>
<a name="l04118"></a>04118 <span class="stringliteral">                tlr(simEndTimes(i)+1:end, i, :) = NaN;</span>
<a name="l04119"></a>04119 <span class="stringliteral">                nfkb(simEndTimes(i)+1:end, i, :) = NaN;</span>
<a name="l04120"></a>04120 <span class="stringliteral">                inflammation(simEndTimes(i)+1:end, i, :) = NaN;</span>
<a name="l04121"></a>04121 <span class="stringliteral">                proteins(simEndTimes(i)+1:end, i, :) = NaN;</span>
<a name="l04122"></a>04122 <span class="stringliteral">            end</span>
<a name="l04123"></a>04123 <span class="stringliteral">            </span>
<a name="l04124"></a>04124 <span class="stringliteral">            clear states mass;</span>
<a name="l04125"></a>04125 <span class="stringliteral">            </span>
<a name="l04126"></a>04126 <span class="stringliteral">            %% plot data</span>
<a name="l04127"></a>04127 <span class="stringliteral">            clf(figHandle);</span>
<a name="l04128"></a>04128 <span class="stringliteral">            </span>
<a name="l04129"></a>04129 <span class="stringliteral">            options = struct();</span>
<a name="l04130"></a>04130 <span class="stringliteral">            options.titleStr = {&#39;</span>{\it M. genitalium}<span class="charliteral">&#39; &#39;</span>Host Response<span class="stringliteral">&#39;};</span>
<a name="l04131"></a>04131 <span class="stringliteral">            options.colorOrder = PlotUtil.getRedGreenColorOrder(simOrder);</span>
<a name="l04132"></a>04132 <span class="stringliteral">            options.xdata = time;</span>
<a name="l04133"></a>04133 <span class="stringliteral">            options.ylabelStr = {</span>
<a name="l04134"></a>04134 <span class="stringliteral">                cellfun(@(x) strrep(x, &#39;</span>_<span class="stringliteral">&#39;, &#39;</span>\_<span class="stringliteral">&#39;), pm.wholeCellModelIDs(monIdxs), &#39;</span>UniformOutput<span class="stringliteral">&#39;, false)</span>
<a name="l04135"></a>04135 <span class="stringliteral">                {</span>
<a name="l04136"></a>04136 <span class="stringliteral">                &#39;</span>Adherence<span class="stringliteral">&#39;</span>
<a name="l04137"></a>04137 <span class="stringliteral">                &#39;</span>TLR 1<span class="stringliteral">&#39;</span>
<a name="l04138"></a>04138 <span class="stringliteral">                &#39;</span>TLR 2<span class="stringliteral">&#39;</span>
<a name="l04139"></a>04139 <span class="stringliteral">                &#39;</span>TLR 6<span class="stringliteral">&#39;</span>
<a name="l04140"></a>04140 <span class="stringliteral">                &#39;</span>NF-{\kappa}B<span class="stringliteral">&#39;</span>
<a name="l04141"></a>04141 <span class="stringliteral">                &#39;</span>Inflammation<span class="stringliteral">&#39;</span>
<a name="l04142"></a>04142 <span class="stringliteral">                }</span>
<a name="l04143"></a>04143 <span class="stringliteral">                };</span>
<a name="l04144"></a>04144 <span class="stringliteral">            options.ydata = {</span>
<a name="l04145"></a>04145 <span class="stringliteral">                mat2cell(proteins, size(proteins, 1), size(proteins, 2), ones(size(proteins, 3), 1))</span>
<a name="l04146"></a>04146 <span class="stringliteral">                {</span>
<a name="l04147"></a>04147 <span class="stringliteral">                adherence</span>
<a name="l04148"></a>04148 <span class="stringliteral">                tlr(:, :, host.tlrIndexs_1)</span>
<a name="l04149"></a>04149 <span class="stringliteral">                tlr(:, :, host.tlrIndexs_2)</span>
<a name="l04150"></a>04150 <span class="stringliteral">                tlr(:, :, host.tlrIndexs_6)</span>
<a name="l04151"></a>04151 <span class="stringliteral">                nfkb</span>
<a name="l04152"></a>04152 <span class="stringliteral">                inflammation</span>
<a name="l04153"></a>04153 <span class="stringliteral">                }};</span>
<a name="l04154"></a>04154 <span class="stringliteral">            PlotUtil.multiElementPlot(figHandle, {3 * ones(size(monIdxs)) 3 * ones(6, 1)}, [0 time(end)], options);</span>
<a name="l04155"></a>04155 <span class="stringliteral">            </span>
<a name="l04156"></a>04156 <span class="stringliteral">            clear options;</span>
<a name="l04157"></a>04157 <span class="stringliteral">            </span>
<a name="l04158"></a>04158 <span class="stringliteral">            %% organize figure data</span>
<a name="l04159"></a>04159 <span class="stringliteral">            figData = struct;</span>
<a name="l04160"></a>04160 <span class="stringliteral">            figData.simOrder = simOrder;</span>
<a name="l04161"></a>04161 <span class="stringliteral">            figData.time = time;</span>
<a name="l04162"></a>04162 <span class="stringliteral">            figData.proteins = proteins;</span>
<a name="l04163"></a>04163 <span class="stringliteral">            figData.adherence = adherence;</span>
<a name="l04164"></a>04164 <span class="stringliteral">            figData.tlr = tlr;</span>
<a name="l04165"></a>04165 <span class="stringliteral">            figData.nfkb = nfkb;</span>
<a name="l04166"></a>04166 <span class="stringliteral">            figData.inflammation = inflammation;</span>
<a name="l04167"></a>04167 <span class="stringliteral">        end</span>
<a name="l04168"></a>04168 <span class="stringliteral">        </span>
<a name="l04169"></a>04169 <span class="stringliteral">        function figData = smcSpacing(figHandle, simBatchDir, selectedSimulations)</span>
<a name="l04170"></a>04170 <span class="stringliteral">            import edu.stanford.covert.cell.sim.util.DiskLogger;</span>
<a name="l04171"></a>04171 <span class="stringliteral">            import edu.stanford.covert.cell.sim.util.SimulationEnsemble;</span>
<a name="l04172"></a>04172 <span class="stringliteral">            import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;</span>
<a name="l04173"></a>04173 <span class="stringliteral">            import edu.stanford.covert.cell.sim.util.PlotUtil;</span>
<a name="l04174"></a>04174 <span class="stringliteral">            </span>
<a name="l04175"></a>04175 <span class="stringliteral">            if nargin &lt; 2 || isempty(simBatchDir)</span>
<a name="l04176"></a>04176 <span class="stringliteral">                simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();</span>
<a name="l04177"></a>04177 <span class="stringliteral">            end</span>
<a name="l04178"></a>04178 <span class="stringliteral">            if nargin &lt; 3 || isempty(selectedSimulations)</span>
<a name="l04179"></a>04179 <span class="stringliteral">                selectedSimulations = SimulationDiskUtil.getCompleteSimulations(simBatchDir);</span>
<a name="l04180"></a>04180 <span class="stringliteral">            end</span>
<a name="l04181"></a>04181 <span class="stringliteral">            </span>
<a name="l04182"></a>04182 <span class="stringliteral">            [simDir, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep &#39;</span>1<span class="stringliteral">&#39;]);</span>
<a name="l04183"></a>04183 <span class="stringliteral">            simBatchDir = SimulationDiskUtil.getSimulationBatchDir(simDir);</span>
<a name="l04184"></a>04184 <span class="stringliteral">            pc = sim.state(&#39;</span>ProteinComplex<span class="stringliteral">&#39;);</span>
<a name="l04185"></a>04185 <span class="stringliteral">            ch = sim.state(&#39;</span><a class="code" href="class_chromosome.html">Chromosome</a><span class="stringliteral">&#39;);</span>
<a name="l04186"></a>04186 <span class="stringliteral">            </span>
<a name="l04187"></a>04187 <span class="stringliteral">            stateNames = {...</span>
<a name="l04188"></a>04188 <span class="stringliteral">                &#39;</span>Mass<span class="stringliteral">&#39;       &#39;</span>cell<span class="stringliteral">&#39;                     &#39;</span>:<span class="stringliteral">&#39;     &#39;</span>-sum&#39;
<a name="l04189"></a>04189                 &#39;Time&#39;       &#39;values&#39;                   &#39;:&#39;     &#39;:&#39;
<a name="l04190"></a>04190                 };
<a name="l04191"></a>04191             states = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, <span class="stringliteral">&#39;extract&#39;</span>, selectedSimulations);
<a name="l04192"></a>04192             simEndTimes = permute(max(states.Time.values, [], 3), [4 1 2 3]);
<a name="l04193"></a>04193             timeIdx = min(simEndTimes);
<a name="l04194"></a>04194             [~, simOrder] = sort(states.Mass.cell(:, :, timeIdx, :));
<a name="l04195"></a>04195             simOrder = simOrder(:);
<a name="l04196"></a>04196             clear(<span class="stringliteral">&#39;states&#39;</span>);
<a name="l04197"></a>04197             
<a name="l04198"></a>04198             maxTime = 0;
<a name="l04199"></a>04199             <span class="keywordflow">for</span> i = 1:length(selectedSimulations)
<a name="l04200"></a>04200                 summary = load([simBatchDir filesep num2str(selectedSimulations(i)) filesep &#39;summary.mat&#39;], &#39;time&#39;);
<a name="l04201"></a>04201                 maxTime = max(maxTime, sum(summary.time &gt;= 0));
<a name="l04202"></a>04202             end
<a name="l04203"></a>04203             
<a name="l04204"></a>04204             time = 1:maxTime;
<a name="l04205"></a>04205             smcSpacing = zeros(numel(selectedSimulations), numel(time));
<a name="l04206"></a>04206             
<a name="l04207"></a>04207             stateNames = {...
<a name="l04208"></a>04208                 <span class="stringliteral">&#39;Chromosome&#39;</span> <span class="stringliteral">&#39;complexBoundSites&#39;</span>        <span class="charliteral">&#39;:&#39;</span>     <span class="charliteral">&#39;:&#39;</span>
<a name="l04209"></a>04209                 };
<a name="l04210"></a>04210             
<a name="l04211"></a>04211             <span class="keywordflow">for</span> selSimIdx = 1:numel(selectedSimulations)
<a name="l04212"></a>04212                 % Run out of memory when <span class="keyword">using</span> SimulationEnsemble
<a name="l04213"></a>04213                 % Instead, load and analyze each simulation one at a time
<a name="l04214"></a>04214                 states = DiskLogger.load([simBatchDir filesep num2str(selectedSimulations(selSimIdx))], stateNames, [], [], 1, <span class="stringliteral">&#39;extract&#39;</span>);
<a name="l04215"></a>04215                 
<a name="l04216"></a>04216                 cbs = states.Chromosome.complexBoundSites;
<a name="l04217"></a>04217                 subs = find(cbs == pc.getIndexs(<span class="stringliteral">&#39;MG_213_214_298_6MER_ADP&#39;</span>));
<a name="l04218"></a>04218                 subs = subs(subs(:, 2) &lt;= 2, :);
<a name="l04219"></a>04219                 
<a name="l04220"></a>04220                 smcSpacing(selSimIdx, :) = ch.sequenceLen ./  histc(subs(:, 3), time);
<a name="l04221"></a>04221                 
<a name="l04222"></a>04222             end
<a name="l04223"></a>04223             
<a name="l04224"></a>04224             time = time / 3600;
<a name="l04225"></a>04225             smcSpacing(smcSpacing == Inf) = NaN;
<a name="l04226"></a>04226             
<a name="l04227"></a>04227             %% plot data
<a name="l04228"></a>04228             clf(figHandle);
<a name="l04229"></a>04229             options = <span class="keyword">struct</span>;
<a name="l04230"></a>04230             options.titleStr = {<span class="stringliteral">&#39;SMC Spacing&#39;</span>};
<a name="l04231"></a>04231             options.colorOrder = PlotUtil.getRedGreenColorOrder(simOrder);
<a name="l04232"></a>04232             options.xdata = time;
<a name="l04233"></a>04233             options.ydata = {
<a name="l04234"></a>04234                 smcSpacing
<a name="l04235"></a>04235                 };
<a name="l04236"></a>04236             options.ylabelStr = {
<a name="l04237"></a>04237                 {<span class="stringliteral">&#39;Average Distance (bp)&#39;</span>}
<a name="l04238"></a>04238                 };
<a name="l04239"></a>04239             PlotUtil.multiElementPlot(figHandle, {30}, [0 time(end)], options);
<a name="l04240"></a>04240             
<a name="l04241"></a>04241             clear options;
<a name="l04242"></a>04242             
<a name="l04243"></a>04243             %% organize figure data
<a name="l04244"></a>04244             figData = <span class="keyword">struct</span>;
<a name="l04245"></a>04245             figData.simOrder = simOrder;
<a name="l04246"></a>04246             figData.time = time;
<a name="l04247"></a>04247             figData.smcSpacing = smcSpacing;
<a name="l04248"></a>04248         end
<a name="l04249"></a>04249         
<a name="l04250"></a>04250         <span class="keyword">function</span> figData = geneCopyNumber(figHandle, simBatchDir, selectedSimulations)
<a name="l04251"></a>04251             import edu.stanford.covert.cell.sim.util.SimulationEnsemble;
<a name="l04252"></a>04252             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
<a name="l04253"></a>04253             import edu.stanford.covert.cell.sim.util.PlotUtil;
<a name="l04254"></a>04254             import edu.stanford.covert.cell.sim.util.DiskLogger;
<a name="l04255"></a>04255             if ~exist(&#39;simBatchDir&#39;, &#39;var&#39;) || isempty(simBatchDir)
<a name="l04256"></a>04256                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
<a name="l04257"></a>04257             end
<a name="l04258"></a>04258             if ~exist(&#39;selectedSimulations&#39;, &#39;var&#39;) || isempty(selectedSimulations)
<a name="l04259"></a>04259                 selectedSimulations = SimulationDiskUtil.getCompleteSimulations(simBatchDir);
<a name="l04260"></a>04260             end
<a name="l04261"></a>04261             
<a name="l04262"></a>04262             [simDir, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep &#39;1&#39;]);
<a name="l04263"></a>04263             simBatchDir = SimulationDiskUtil.getSimulationBatchDir(simDir);
<a name="l04264"></a>04264             ch = sim.state(&#39;<a class="code" href="class_chromosome.html">Chromosome</a>&#39;);
<a name="l04265"></a>04265             
<a name="l04266"></a>04266             %% get data
<a name="l04267"></a>04267             stateNames = {...
<a name="l04268"></a>04268                 <span class="stringliteral">&#39;Mass&#39;</span>       <span class="stringliteral">&#39;cell&#39;</span>                     <span class="charliteral">&#39;:&#39;</span>     <span class="stringliteral">&#39;-sum&#39;</span>
<a name="l04269"></a>04269                 <span class="stringliteral">&#39;Time&#39;</span>       <span class="stringliteral">&#39;values&#39;</span>                   <span class="charliteral">&#39;:&#39;</span>     <span class="charliteral">&#39;:&#39;</span>
<a name="l04270"></a>04270                 };
<a name="l04271"></a>04271             states = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, <span class="stringliteral">&#39;extract&#39;</span>, selectedSimulations);
<a name="l04272"></a>04272             simEndTimes = permute(max(states.Time.values, [], 3), [4 1 2 3]);
<a name="l04273"></a>04273             timeIdx = min(simEndTimes);
<a name="l04274"></a>04274             [~, simOrder] = sort(states.Mass.cell(:, :, timeIdx, :));
<a name="l04275"></a>04275             simOrder = simOrder(:);
<a name="l04276"></a>04276             clear(<span class="stringliteral">&#39;states&#39;</span>);
<a name="l04277"></a>04277             
<a name="l04278"></a>04278             stateNames = {...
<a name="l04279"></a>04279                 <span class="stringliteral">&#39;Chromosome&#39;</span> <span class="stringliteral">&#39;polymerizedRegions&#39;</span>        <span class="charliteral">&#39;:&#39;</span>     <span class="charliteral">&#39;:&#39;</span>
<a name="l04280"></a>04280                 };
<a name="l04281"></a>04281             
<a name="l04282"></a>04282             copyNumber = zeros(numel(selectedSimulations), ch.sequenceLen);
<a name="l04283"></a>04283             
<a name="l04284"></a>04284             <span class="keywordflow">for</span> i = 1:numel(selectedSimulations)
<a name="l04285"></a>04285                 % Function is ~5x faster <span class="keywordflow">if</span> I <span class="keywordflow">do</span> <span class="keyword">this</span> rather than
<a name="l04286"></a>04286                 % SimulationEnsemble.load() outside of the loop due to
<a name="l04287"></a>04287                 % avoiding <a class="code" href="class_sparse_mat.html">SparseMat</a> concatenation
<a name="l04288"></a>04288                 states = DiskLogger.load([simBatchDir filesep num2str(selectedSimulations(i))], stateNames, [], [], 1, <span class="stringliteral">&#39;extract&#39;</span>);
<a name="l04289"></a>04289                 
<a name="l04290"></a>04290                 [subs vals] = find(states.Chromosome.polymerizedRegions);
<a name="l04291"></a>04291                 
<a name="l04292"></a>04292                 startpos = subs(:, 1);
<a name="l04293"></a>04293                 endpos = startpos + vals(:, 1) - 1;
<a name="l04294"></a>04294                 
<a name="l04295"></a>04295                 copyNumber(i, :) = cumsum(histc(startpos, 1:ch.sequenceLen) - histc(endpos + 1, 1:ch.sequenceLen)) / simEndTimes(i) / 2; % 2 strands per chromosome
<a name="l04296"></a>04296             end
<a name="l04297"></a>04297             
<a name="l04298"></a>04298             %% plot data
<a name="l04299"></a>04299             clf(figHandle);
<a name="l04300"></a>04300             options = <span class="keyword">struct</span>;
<a name="l04301"></a>04301             options.titleStr = {<span class="stringliteral">&#39;Copy Number Over Cell Cycle&#39;</span>};
<a name="l04302"></a>04302             options.colorOrder = PlotUtil.getRedGreenColorOrder(simOrder);
<a name="l04303"></a>04303             options.xdata = 1:ch.sequenceLen;
<a name="l04304"></a>04304             options.xlabelStr = {<span class="stringliteral">&#39;Position on Chromosome&#39;</span>};
<a name="l04305"></a>04305             options.ydata = {
<a name="l04306"></a>04306                 copyNumber
<a name="l04307"></a>04307                 };
<a name="l04308"></a>04308             options.ylabelStr = {
<a name="l04309"></a>04309                 {<span class="stringliteral">&#39;Average Copy Number&#39;</span>}
<a name="l04310"></a>04310                 };
<a name="l04311"></a>04311             [~, xAxesHandle] = PlotUtil.multiElementPlot(figHandle, {30}, [1 ch.sequenceLen], options);
<a name="l04312"></a>04312             
<a name="l04313"></a>04313             <span class="keyword">set</span>(xAxesHandle{1}, <span class="stringliteral">&#39;XTick&#39;</span>, [1 ch.terCPosition ch.sequenceLen], <span class="stringliteral">&#39;XTickLabel&#39;</span>, {<span class="stringliteral">&#39;oriC&#39;</span>, <span class="stringliteral">&#39;terC&#39;</span>, <span class="stringliteral">&#39;oriC&#39;</span>});
<a name="l04314"></a>04314             
<a name="l04315"></a>04315             clear options;
<a name="l04316"></a>04316             
<a name="l04317"></a>04317             %% organize figure data
<a name="l04318"></a>04318             figData = <span class="keyword">struct</span>;
<a name="l04319"></a>04319             figData.simOrder = simOrder;
<a name="l04320"></a>04320             figData.copyNumber = copyNumber;
<a name="l04321"></a>04321         end
<a name="l04322"></a>04322         
<a name="l04323"></a>04323         <span class="keyword">function</span> figData = unsynchronizedPopulation(figHandle, simBatchDir, selectedSimulations)
<a name="l04324"></a>04324             import edu.stanford.covert.cell.sim.analysis.<a class="code" href="class_population.html">Population</a>;
<a name="l04325"></a>04325             import edu.stanford.covert.cell.sim.util.SimulationEnsemble;
<a name="l04326"></a>04326             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
<a name="l04327"></a>04327             import edu.stanford.covert.cell.sim.util.PlotUtil;
<a name="l04328"></a>04328             import edu.stanford.covert.cell.sim.util.DiskLogger;
<a name="l04329"></a>04329             
<a name="l04330"></a>04330             if ~exist(&#39;simBatchDir&#39;, &#39;var&#39;) || isempty(simBatchDir)
<a name="l04331"></a>04331                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
<a name="l04332"></a>04332             end
<a name="l04333"></a>04333             if ~exist(&#39;selectedSimulations&#39;, &#39;var&#39;) || isempty(selectedSimulations)
<a name="l04334"></a>04334                 selectedSimulations = SimulationDiskUtil.getCompleteSimulations(simBatchDir);
<a name="l04335"></a>04335             end
<a name="l04336"></a>04336             
<a name="l04337"></a>04337             %% get constants
<a name="l04338"></a>04338             [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep &#39;1&#39;]);
<a name="l04339"></a>04339             
<a name="l04340"></a>04340             %% get data
<a name="l04341"></a>04341             stateNames = {
<a name="l04342"></a>04342                 <span class="stringliteral">&#39;growth_rate&#39;</span>       <span class="stringliteral">&#39;Growth&#39;</span>
<a name="l04343"></a>04343                 <span class="stringliteral">&#39;mass&#39;</span>              <span class="stringliteral">&#39;Mass&#39;</span>
<a name="l04344"></a>04344                 <span class="stringliteral">&#39;ploidy&#39;</span>            {<span class="stringliteral">&#39;Chromosome&#39;</span> <span class="stringliteral">&#39;Copy Number&#39;</span>}
<a name="l04345"></a>04345                 <span class="stringliteral">&#39;ftsZRing1st&#39;</span>       <span class="stringliteral">&#39;ftsZRing&#39;</span>
<a name="l04346"></a>04346                 <span class="stringliteral">&#39;ftsZRing2st&#39;</span>       <span class="stringliteral">&#39;ftsZRing&#39;</span>
<a name="l04347"></a>04347                 <span class="stringliteral">&#39;ftsZRing2bt&#39;</span>       <span class="stringliteral">&#39;ftsZRing&#39;</span>
<a name="l04348"></a>04348                 <span class="stringliteral">&#39;ftsZRingRbt&#39;</span>       <span class="stringliteral">&#39;ftsZRing&#39;</span>
<a name="l04349"></a>04349                 <span class="stringliteral">&#39;pinchedDiameter&#39;</span>   <span class="stringliteral">&#39;pinchedDiameter&#39;</span>
<a name="l04350"></a>04350                 };
<a name="l04351"></a>04351             ensemble = SimulationEnsemble(simBatchDir, stateNames, [], selectedSimulations);
<a name="l04352"></a>04352             simEndTimes = ensemble.stateData.simulationEndTimes;
<a name="l04353"></a>04353             
<a name="l04354"></a>04354             [time, values, cellCount, simOrder] = <a class="code" href="class_population.html">Population</a>.<a class="code" href="class_population.html#a83b7d5ddb5b0d9466e930a65a7bbde96">desynchronizePopulation</a>(sim, ensemble.stateData.time, ensemble.stateData.values, simEndTimes);
<a name="l04355"></a>04355             time = time / 3600;
<a name="l04356"></a>04356             growth = permute(values(ensemble.getPropertyIndices(<span class="stringliteral">&#39;growth_rate&#39;</span>), :, :, :), [4 3 1 2]) * 3600 * sim.state(<span class="stringliteral">&#39;Mass&#39;</span>).cellInitialDryWeight / (1 - sim.state(&#39;Mass&#39;).fractionWetWeight);
<a name="l04357"></a>04357             mass = permute(values(ensemble.getPropertyIndices(<span class="stringliteral">&#39;mass&#39;</span>), :, :, :), [4 3 1 2]) * 1e15;
<a name="l04358"></a>04358             ploidy = permute(values(ensemble.getPropertyIndices(<span class="stringliteral">&#39;ploidy&#39;</span>), :, :, :), [4 3 1 2]);
<a name="l04359"></a>04359             ftsZRing1st = permute(values(ensemble.getPropertyIndices(<span class="stringliteral">&#39;ftsZRing1st&#39;</span>), :, :, :), [4 3 1 2]);
<a name="l04360"></a>04360             ftsZRing2st = permute(values(ensemble.getPropertyIndices(<span class="stringliteral">&#39;ftsZRing2st&#39;</span>), :, :, :), [4 3 1 2]);
<a name="l04361"></a>04361             ftsZRing2bt = permute(values(ensemble.getPropertyIndices(<span class="stringliteral">&#39;ftsZRing2bt&#39;</span>), :, :, :), [4 3 1 2]);
<a name="l04362"></a>04362             ftsZRingRbt = permute(values(ensemble.getPropertyIndices(<span class="stringliteral">&#39;ftsZRingRbt&#39;</span>), :, :, :), [4 3 1 2]);
<a name="l04363"></a>04363             pinchedDiameter = permute(values(ensemble.getPropertyIndices(<span class="stringliteral">&#39;pinchedDiameter&#39;</span>), :, :, :), [4 3 1 2]);
<a name="l04364"></a>04364             cellCount = permute(cellCount, [4 3 1 2]);
<a name="l04365"></a>04365             
<a name="l04366"></a>04366             %% plot data
<a name="l04367"></a>04367             clf(figHandle);
<a name="l04368"></a>04368             options = <span class="keyword">struct</span>;
<a name="l04369"></a>04369             options.colorOrder = PlotUtil.getRedGreenColorOrder(simOrder);
<a name="l04370"></a>04370             options.xdata = time;
<a name="l04371"></a>04371             options.ydata = {
<a name="l04372"></a>04372                 growth
<a name="l04373"></a>04373                 mass
<a name="l04374"></a>04374                 ploidy
<a name="l04375"></a>04375                 ftsZRing1st
<a name="l04376"></a>04376                 ftsZRing2st
<a name="l04377"></a>04377                 ftsZRing2bt
<a name="l04378"></a>04378                 ftsZRingRbt
<a name="l04379"></a>04379                 pinchedDiameter
<a name="l04380"></a>04380                 cellCount
<a name="l04381"></a>04381                 };
<a name="l04382"></a>04382             options.ylabelStr = {
<a name="l04383"></a>04383                 {<span class="stringliteral">&#39;Growth&#39;</span> <span class="stringliteral">&#39;(fg h^{-1})&#39;</span>}
<a name="l04384"></a>04384                 {<span class="stringliteral">&#39;Mass&#39;</span> <span class="stringliteral">&#39;(fg)&#39;</span>}
<a name="l04385"></a>04385                 {<span class="stringliteral">&#39;Chr Copy&#39;</span> <span class="stringliteral">&#39;Number&#39;</span>}
<a name="l04386"></a>04386                 {<span class="stringliteral">&#39;FtsZ&#39;</span> <span class="stringliteral">&#39;Ring&#39;</span> <span class="stringliteral">&#39;1-Strt&#39;</span>}
<a name="l04387"></a>04387                 {<span class="stringliteral">&#39;FtsZ&#39;</span> <span class="stringliteral">&#39;Ring&#39;</span> <span class="stringliteral">&#39;2-Strt&#39;</span>}
<a name="l04388"></a>04388                 {<span class="stringliteral">&#39;FtsZ&#39;</span> <span class="stringliteral">&#39;Ring&#39;</span> <span class="stringliteral">&#39;2-Bent&#39;</span>}
<a name="l04389"></a>04389                 {<span class="stringliteral">&#39;FtsZ&#39;</span> <span class="stringliteral">&#39;Ring&#39;</span> <span class="stringliteral">&#39;R-Bent&#39;</span>}
<a name="l04390"></a>04390                 {<span class="stringliteral">&#39;Pinched&#39;</span> <span class="stringliteral">&#39;Diameter&#39;</span>}
<a name="l04391"></a>04391                 {<span class="stringliteral">&#39;Cell&#39;</span> <span class="stringliteral">&#39;Count&#39;</span>}
<a name="l04392"></a>04392                 };
<a name="l04393"></a>04393             PlotUtil.multiElementPlot(figHandle, 10 * ones(size(options.ydata)), [0 time(end)], options);
<a name="l04394"></a>04394             
<a name="l04395"></a>04395             clear options;
<a name="l04396"></a>04396             
<a name="l04397"></a>04397             %% organize figure data
<a name="l04398"></a>04398             figData = <span class="keyword">struct</span>;
<a name="l04399"></a>04399             figData.simOrder = simOrder;
<a name="l04400"></a>04400             figData.time = time;
<a name="l04401"></a>04401             figData.growth = growth;
<a name="l04402"></a>04402             figData.mass = mass;
<a name="l04403"></a>04403             figData.ploidy = ploidy;
<a name="l04404"></a>04404             figData.cellCount = cellCount;
<a name="l04405"></a>04405         end
<a name="l04406"></a>04406         
<a name="l04407"></a>04407         <span class="keyword">function</span> figData = blockedDecayEvents(figHandle, simBatchDir, selectedSimulations)
<a name="l04408"></a>04408             import edu.stanford.covert.cell.sim.util.SimulationEnsemble;
<a name="l04409"></a>04409             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
<a name="l04410"></a>04410             import edu.stanford.covert.cell.sim.util.PlotUtil;
<a name="l04411"></a>04411             
<a name="l04412"></a>04412             if ~exist(&#39;simBatchDir&#39;, &#39;var&#39;) || isempty(simBatchDir)
<a name="l04413"></a>04413                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
<a name="l04414"></a>04414             end
<a name="l04415"></a>04415             if ~exist(&#39;selectedSimulations&#39;, &#39;var&#39;) || isempty(selectedSimulations)
<a name="l04416"></a>04416                 selectedSimulations = SimulationDiskUtil.getCompleteSimulations(simBatchDir);
<a name="l04417"></a>04417             end
<a name="l04418"></a>04418             
<a name="l04419"></a>04419             %% get data
<a name="l04420"></a>04420             nonDecays_DnaA = zeros(0, 1);
<a name="l04421"></a>04421             nonDecays_FtsZ = zeros(0, 1);
<a name="l04422"></a>04422             nonDecays_Replisome = zeros(0, 1);
<a name="l04423"></a>04423             nonDecayCnts_DnaA = zeros(size(selectedSimulations));
<a name="l04424"></a>04424             nonDecayCnts_FtsZ = zeros(size(selectedSimulations));
<a name="l04425"></a>04425             nonDecayCnts_Replisome = zeros(size(selectedSimulations));
<a name="l04426"></a>04426             for i = 1:numel(selectedSimulations)
<a name="l04427"></a>04427                 warnings = SimulationDiskUtil.getSimulationWarnings(simBatchDir, i);
<a name="l04428"></a>04428                 
<a name="l04429"></a>04429                 idxs = find(strcmp({warnings.message}, <span class="stringliteral">&#39;DnaA complex not decayed&#39;</span>));
<a name="l04430"></a>04430                 <span class="keywordflow">for</span> j = 1:numel(idxs)
<a name="l04431"></a>04431                     nonDecays_DnaA = [nonDecays_DnaA; warnings(idxs(j)).times]; %<span class="preprocessor">#ok&lt;AGROW&gt;</span>
<a name="l04432"></a>04432 <span class="preprocessor"></span>                    nonDecayCnts_DnaA(i) = nonDecayCnts_DnaA(i) + numel(warnings(idxs(j)).times);
<a name="l04433"></a>04433                 end
<a name="l04434"></a>04434                 
<a name="l04435"></a>04435                 idxs = find(strcmp({warnings.message}, <span class="stringliteral">&#39;FtsZ not decayed&#39;</span>));
<a name="l04436"></a>04436                 <span class="keywordflow">for</span> j = 1:numel(idxs)
<a name="l04437"></a>04437                     nonDecays_FtsZ = [nonDecays_FtsZ; warnings(idxs(j)).times]; %<span class="preprocessor">#ok&lt;AGROW&gt;</span>
<a name="l04438"></a>04438 <span class="preprocessor"></span>                    nonDecayCnts_FtsZ(i) = nonDecayCnts_FtsZ(i) + numel(warnings(idxs(j)).times);
<a name="l04439"></a>04439                 end
<a name="l04440"></a>04440                 
<a name="l04441"></a>04441                 idxs = find(strcmp({warnings.message}, <span class="stringliteral">&#39;DNA polymerase not decayed&#39;</span>));
<a name="l04442"></a>04442                 <span class="keywordflow">for</span> j = 1:numel(idxs)
<a name="l04443"></a>04443                     nonDecays_Replisome = [nonDecays_Replisome; warnings(idxs(j)).times]; %<span class="preprocessor">#ok&lt;AGROW&gt;</span>
<a name="l04444"></a>04444 <span class="preprocessor"></span>                    nonDecayCnts_Replisome(i) = nonDecayCnts_Replisome(i) + numel(warnings(idxs(j)).times);
<a name="l04445"></a>04445                 end
<a name="l04446"></a>04446             end
<a name="l04447"></a>04447             
<a name="l04448"></a>04448             nonDecays_DnaA = nonDecays_DnaA / 3600;
<a name="l04449"></a>04449             nonDecays_FtsZ = nonDecays_FtsZ / 3600;
<a name="l04450"></a>04450             nonDecays_Replisome = nonDecays_Replisome / 3600;
<a name="l04451"></a>04451             
<a name="l04452"></a>04452             %% plot data
<a name="l04453"></a>04453             clf(figHandle);
<a name="l04454"></a>04454             options = <span class="keyword">struct</span>();
<a name="l04455"></a>04455             options.titleStr = {
<a name="l04456"></a>04456                 <span class="stringliteral">&#39;Non Decay Event Timing&#39;</span>
<a name="l04457"></a>04457                 <span class="stringliteral">&#39;Non Decay Events Per Simulation&#39;</span>
<a name="l04458"></a>04458                 };
<a name="l04459"></a>04459             [axesHandles, xAxesHandles] = PlotUtil.multiElementPlot(figHandle, repmat({10 * ones(3, 1)}, 2, 1), [0 max([nonDecays_DnaA; nonDecays_FtsZ; nonDecays_Replisome])], options);
<a name="l04460"></a>04460             <span class="keyword">delete</span>(xAxesHandles{2});
<a name="l04461"></a>04461             
<a name="l04462"></a>04462             %Non Decay Event Timing
<a name="l04463"></a>04463             edges = linspace(0, max([nonDecays_DnaA; nonDecays_FtsZ; nonDecays_Replisome]), 50);
<a name="l04464"></a>04464             
<a name="l04465"></a>04465             axesHandle = axesHandles{1}(1);
<a name="l04466"></a>04466             hist(axesHandle, nonDecays_DnaA, edges);
<a name="l04467"></a>04467             <span class="keyword">set</span>(findobj(axesHandle, <span class="stringliteral">&#39;type&#39;</span>, <span class="stringliteral">&#39;patch&#39;</span>), <span class="stringliteral">&#39;EdgeColor&#39;</span>, <span class="charliteral">&#39;b&#39;</span>, <span class="stringliteral">&#39;FaceColor&#39;</span>, <span class="charliteral">&#39;b&#39;</span>);
<a name="l04468"></a>04468             ylabel(axesHandle, <span class="stringliteral">&#39;DnaA&#39;</span>);
<a name="l04469"></a>04469             
<a name="l04470"></a>04470             axesHandle = axesHandles{1}(2);
<a name="l04471"></a>04471             hist(axesHandle, nonDecays_FtsZ, edges);
<a name="l04472"></a>04472             <span class="keyword">set</span>(findobj(axesHandle, <span class="stringliteral">&#39;type&#39;</span>, <span class="stringliteral">&#39;patch&#39;</span>), <span class="stringliteral">&#39;EdgeColor&#39;</span>, <span class="charliteral">&#39;b&#39;</span>, <span class="stringliteral">&#39;FaceColor&#39;</span>, <span class="charliteral">&#39;b&#39;</span>);
<a name="l04473"></a>04473             ylabel(axesHandle, <span class="stringliteral">&#39;FtsZ&#39;</span>);
<a name="l04474"></a>04474             
<a name="l04475"></a>04475             axesHandle = axesHandles{1}(3);
<a name="l04476"></a>04476             hist(axesHandle, nonDecays_Replisome, edges);
<a name="l04477"></a>04477             <span class="keyword">set</span>(findobj(axesHandle, <span class="stringliteral">&#39;type&#39;</span>, <span class="stringliteral">&#39;patch&#39;</span>), <span class="stringliteral">&#39;EdgeColor&#39;</span>, <span class="charliteral">&#39;b&#39;</span>, <span class="stringliteral">&#39;FaceColor&#39;</span>, <span class="charliteral">&#39;b&#39;</span>);
<a name="l04478"></a>04478             ylabel(axesHandle, <span class="stringliteral">&#39;Replisome&#39;</span>);
<a name="l04479"></a>04479             
<a name="l04480"></a>04480             %Non Decay Events Per Simulation
<a name="l04481"></a>04481             axesHandle = axesHandles{2}(1);
<a name="l04482"></a>04482             hist(axesHandle, nonDecayCnts_DnaA, 10);
<a name="l04483"></a>04483             <span class="keyword">set</span>(findobj(axesHandle, <span class="stringliteral">&#39;type&#39;</span>, <span class="stringliteral">&#39;patch&#39;</span>), <span class="stringliteral">&#39;EdgeColor&#39;</span>, <span class="charliteral">&#39;b&#39;</span>, <span class="stringliteral">&#39;FaceColor&#39;</span>, <span class="charliteral">&#39;b&#39;</span>);
<a name="l04484"></a>04484             xlabel(axesHandle, <span class="stringliteral">&#39;Non-Decays&#39;</span>);
<a name="l04485"></a>04485             ylabel(axesHandle, <span class="stringliteral">&#39;DnaA&#39;</span>);
<a name="l04486"></a>04486             
<a name="l04487"></a>04487             axesHandle = axesHandles{2}(2);
<a name="l04488"></a>04488             hist(axesHandle, nonDecayCnts_DnaA, 10);
<a name="l04489"></a>04489             <span class="keyword">set</span>(findobj(axesHandle, <span class="stringliteral">&#39;type&#39;</span>, <span class="stringliteral">&#39;patch&#39;</span>), <span class="stringliteral">&#39;EdgeColor&#39;</span>, <span class="charliteral">&#39;b&#39;</span>, <span class="stringliteral">&#39;FaceColor&#39;</span>, <span class="charliteral">&#39;b&#39;</span>);
<a name="l04490"></a>04490             xlabel(axesHandle, <span class="stringliteral">&#39;Non-Decays&#39;</span>);
<a name="l04491"></a>04491             ylabel(axesHandle, <span class="stringliteral">&#39;FtsZ&#39;</span>);
<a name="l04492"></a>04492             
<a name="l04493"></a>04493             axesHandle = axesHandles{2}(3);
<a name="l04494"></a>04494             hist(axesHandle, nonDecayCnts_DnaA, 10);
<a name="l04495"></a>04495             <span class="keyword">set</span>(findobj(axesHandle, <span class="stringliteral">&#39;type&#39;</span>, <span class="stringliteral">&#39;patch&#39;</span>), <span class="stringliteral">&#39;EdgeColor&#39;</span>, <span class="charliteral">&#39;b&#39;</span>, <span class="stringliteral">&#39;FaceColor&#39;</span>, <span class="charliteral">&#39;b&#39;</span>);
<a name="l04496"></a>04496             xlabel(axesHandle, <span class="stringliteral">&#39;Non-Decays&#39;</span>);
<a name="l04497"></a>04497             ylabel(axesHandle, <span class="stringliteral">&#39;Replisome&#39;</span>);
<a name="l04498"></a>04498             
<a name="l04499"></a>04499             PlotUtil.alignYAxesLabels(axesHandles);
<a name="l04500"></a>04500             PlotUtil.offsetYAxes(axesHandles, 0.04);
<a name="l04501"></a>04501             
<a name="l04502"></a>04502             %% data
<a name="l04503"></a>04503             figData = <span class="keyword">struct</span>;
<a name="l04504"></a>04504             figData.nonDecays_DnaA = nonDecays_DnaA;
<a name="l04505"></a>04505             figData.nonDecays_FtsZ = nonDecays_FtsZ;
<a name="l04506"></a>04506             figData.nonDecays_Replisome = nonDecays_Replisome;
<a name="l04507"></a>04507             figData.nonDecayCnts_DnaA = nonDecayCnts_DnaA;
<a name="l04508"></a>04508             figData.nonDecayCnts_FtsZ = nonDecayCnts_FtsZ;
<a name="l04509"></a>04509             figData.nonDecayCnts_Replisome = nonDecayCnts_Replisome;
<a name="l04510"></a>04510         end
<a name="l04511"></a>04511         
<a name="l04512"></a>04512         <span class="keyword">function</span> figData = secondaryReplicationInitiation(figHandle, simBatchDir, selectedSimulations)
<a name="l04513"></a>04513             import edu.stanford.covert.cell.sim.util.SimulationEnsemble;
<a name="l04514"></a>04514             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
<a name="l04515"></a>04515             import edu.stanford.covert.cell.sim.util.SummaryLogger;
<a name="l04516"></a>04516             import edu.stanford.covert.cell.sim.util.PlotUtil;
<a name="l04517"></a>04517             
<a name="l04518"></a>04518             if ~exist(&#39;simBatchDir&#39;, &#39;var&#39;) || isempty(simBatchDir)
<a name="l04519"></a>04519                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
<a name="l04520"></a>04520             end
<a name="l04521"></a>04521             if ~exist(&#39;selectedSimulations&#39;, &#39;var&#39;) || isempty(selectedSimulations)
<a name="l04522"></a>04522                 selectedSimulations = SimulationDiskUtil.getCompleteSimulations(simBatchDir);
<a name="l04523"></a>04523             end
<a name="l04524"></a>04524             
<a name="l04525"></a>04525             %% get data
<a name="l04526"></a>04526             secondaryInitiationTimes = [];
<a name="l04527"></a>04527             secondaryInitiationInitTimes = NaN(numel(selectedSimulations), 1);
<a name="l04528"></a>04528             secondaryInitiationCnts = zeros(numel(selectedSimulations), 1);
<a name="l04529"></a>04529             for i = 1:numel(selectedSimulations)
<a name="l04530"></a>04530                 warnings = SimulationDiskUtil.getSimulationWarnings(simBatchDir, i);
<a name="l04531"></a>04531                 
<a name="l04532"></a>04532                 idxs = find(strcmp({warnings.message}, <span class="stringliteral">&#39;Unable to represent multiple replication initiation events per cell cycle&#39;</span>));
<a name="l04533"></a>04533                 <span class="keywordflow">for</span> j = 1:numel(idxs)
<a name="l04534"></a>04534                     secondaryInitiationTimes = [secondaryInitiationTimes; warnings(idxs(j)).times]; %<span class="preprocessor">#ok&lt;AGROW&gt;</span>
<a name="l04535"></a>04535 <span class="preprocessor"></span>                    secondaryInitiationInitTimes(i) = min(secondaryInitiationInitTimes(i), warnings(idxs).times(1));
<a name="l04536"></a>04536                     secondaryInitiationCnts(i) = secondaryInitiationCnts(i) + numel(warnings(idxs(j)).times);
<a name="l04537"></a>04537                 end
<a name="l04538"></a>04538             end
<a name="l04539"></a>04539             secondaryInitiationTimes = secondaryInitiationTimes / 3600;
<a name="l04540"></a>04540             secondaryInitiationInitTimes = secondaryInitiationInitTimes / 3600;
<a name="l04541"></a>04541             
<a name="l04542"></a>04542             simStats = SummaryLogger.getSimulationStatistics([SimulationDiskUtil.getBaseDir() filesep simBatchDir], selectedSimulations);
<a name="l04543"></a>04543             repInitTimes = simStats(:, SummaryLogger.SIM_STATUS_INDEX_REP_INIT_TIME) / 3600;
<a name="l04544"></a>04544             cellCycleTimes = simStats(:, SummaryLogger.SIM_STATUS_INDEX_CYTOKINESIS_TIME) / 3600;
<a name="l04545"></a>04545             
<a name="l04546"></a>04546             clear simStats;
<a name="l04547"></a>04547             
<a name="l04548"></a>04548             stateNames = {
<a name="l04549"></a>04549                 <span class="stringliteral">&#39;dnaA_box1&#39;</span>
<a name="l04550"></a>04550                 <span class="stringliteral">&#39;dnaA_box2&#39;</span>
<a name="l04551"></a>04551                 <span class="stringliteral">&#39;dnaA_box3&#39;</span>
<a name="l04552"></a>04552                 <span class="stringliteral">&#39;dnaA_box4&#39;</span>
<a name="l04553"></a>04553                 <span class="stringliteral">&#39;dnaA_box5&#39;</span>
<a name="l04554"></a>04554                 };
<a name="l04555"></a>04555             ensemble = SimulationEnsemble(simBatchDir, stateNames, [], selectedSimulations);
<a name="l04556"></a>04556             tmp = permute(sum(ensemble.stateData.values, 1), [4 3 1 2]);
<a name="l04557"></a>04557             dnaAComplexSizes = tmp(sub2ind(size(tmp), (1:numel(selectedSimulations))<span class="stringliteral">&#39;, ...</span>
<a name="l04558"></a>04558 <span class="stringliteral">                floor(ensemble.stateData.simulationEndTimes / ensemble.stateData.downsampleStepSec)));</span>
<a name="l04559"></a>04559 <span class="stringliteral">            </span>
<a name="l04560"></a>04560 <span class="stringliteral">            clear ensemble stateNames;</span>
<a name="l04561"></a>04561 <span class="stringliteral">            </span>
<a name="l04562"></a>04562 <span class="stringliteral">            %% plot data</span>
<a name="l04563"></a>04563 <span class="stringliteral">            clf(figHandle);</span>
<a name="l04564"></a>04564 <span class="stringliteral">            options = struct();</span>
<a name="l04565"></a>04565 <span class="stringliteral">            options.xlabelStr = {&#39;</span>Secondary <a class="code" href="class_replication.html">Replication</a> Initiation Time (h)<span class="stringliteral">&#39;, &#39;</span>Frequency<span class="stringliteral">&#39;};</span>
<a name="l04566"></a>04566 <span class="stringliteral">            [axesHandles, xAxesHandles] = PlotUtil.multiElementPlot(figHandle, {[25 25 25] [1 1]}, [0 max(cellCycleTimes)], options);</span>
<a name="l04567"></a>04567 <span class="stringliteral">            </span>
<a name="l04568"></a>04568 <span class="stringliteral">            axesHandle = axesHandles{1}(1);</span>
<a name="l04569"></a>04569 <span class="stringliteral">            edges = reshape(linspace(0, max(cellCycleTimes), 50), [], 1);</span>
<a name="l04570"></a>04570 <span class="stringliteral">            cla(axesHandle);</span>
<a name="l04571"></a>04571 <span class="stringliteral">            hist(axesHandle, secondaryInitiationTimes, edges);</span>
<a name="l04572"></a>04572 <span class="stringliteral">            set(findobj(axesHandle, &#39;</span>type<span class="stringliteral">&#39;, &#39;</span>patch<span class="stringliteral">&#39;), &#39;</span>EdgeColor<span class="stringliteral">&#39;, &#39;</span>b<span class="stringliteral">&#39;, &#39;</span>FaceColor<span class="stringliteral">&#39;, &#39;</span>b<span class="stringliteral">&#39;);</span>
<a name="l04573"></a>04573 <span class="stringliteral">            set(axesHandle, &#39;</span>YTick<span class="stringliteral">&#39;, [0 max(get(axesHandle, &#39;</span>YTick<span class="stringliteral">&#39;))]);</span>
<a name="l04574"></a>04574 <span class="stringliteral">            ylabel(axesHandle, {&#39;</span>Attempt Frequency<span class="stringliteral">&#39;});</span>
<a name="l04575"></a>04575 <span class="stringliteral">            </span>
<a name="l04576"></a>04576 <span class="stringliteral">            axesHandle = axesHandles{1}(2);</span>
<a name="l04577"></a>04577 <span class="stringliteral">            cla(axesHandle)</span>
<a name="l04578"></a>04578 <span class="stringliteral">            edges = reshape(linspace(0, max(cellCycleTimes), 10), [], 1);</span>
<a name="l04579"></a>04579 <span class="stringliteral">            tmp = [histc(repInitTimes, edges)  histc(secondaryInitiationInitTimes, edges)];</span>
<a name="l04580"></a>04580 <span class="stringliteral">            h = bar(axesHandle, edges, tmp);</span>
<a name="l04581"></a>04581 <span class="stringliteral">            legend(h, {&#39;</span>1st<span class="stringliteral">&#39;, &#39;</span>2nd<span class="stringliteral">&#39;}, &#39;</span>Location<span class="stringliteral">&#39;, &#39;</span>NorthWest<span class="stringliteral">&#39;);</span>
<a name="l04582"></a>04582 <span class="stringliteral">            set(h, &#39;</span>EdgeColor<span class="stringliteral">&#39;, &#39;</span>none<span class="stringliteral">&#39;);</span>
<a name="l04583"></a>04583 <span class="stringliteral">            set(h(1), &#39;</span>FaceColor<span class="stringliteral">&#39;, &#39;</span>r<span class="stringliteral">&#39;);</span>
<a name="l04584"></a>04584 <span class="stringliteral">            set(h(2), &#39;</span>FaceColor<span class="stringliteral">&#39;, &#39;</span>b<span class="stringliteral">&#39;);</span>
<a name="l04585"></a>04585 <span class="stringliteral">            xlim(axesHandle, [0 max(cellCycleTimes)]);</span>
<a name="l04586"></a>04586 <span class="stringliteral">            set(axesHandle, &#39;</span>YTick<span class="stringliteral">&#39;, [0 max(get(axesHandle, &#39;</span>YTick<span class="stringliteral">&#39;))]);</span>
<a name="l04587"></a>04587 <span class="stringliteral">            ylabel(axesHandle, {&#39;</span>Frequency<span class="stringliteral">&#39;});</span>
<a name="l04588"></a>04588 <span class="stringliteral">            </span>
<a name="l04589"></a>04589 <span class="stringliteral">            axesHandle = axesHandles{1}(3);</span>
<a name="l04590"></a>04590 <span class="stringliteral">            hold(axesHandle, &#39;</span>on<span class="stringliteral">&#39;);</span>
<a name="l04591"></a>04591 <span class="stringliteral">            plot(axesHandle, secondaryInitiationInitTimes, repInitTimes, &#39;</span>.<span class="stringliteral">&#39;);</span>
<a name="l04592"></a>04592 <span class="stringliteral">            line([0 max(cellCycleTimes)], [0 max(cellCycleTimes)], ...</span>
<a name="l04593"></a>04593 <span class="stringliteral">                &#39;</span>Color<span class="stringliteral">&#39;, &#39;</span>r<span class="stringliteral">&#39;, &#39;</span>LineStyle<span class="stringliteral">&#39;, &#39;</span>:<span class="stringliteral">&#39;, &#39;</span>Parent<span class="stringliteral">&#39;, axesHandle);</span>
<a name="l04594"></a>04594 <span class="stringliteral">            xlim(axesHandle, [0 max(cellCycleTimes)]);</span>
<a name="l04595"></a>04595 <span class="stringliteral">            ylim(axesHandle, [0 max(cellCycleTimes)]);</span>
<a name="l04596"></a>04596 <span class="stringliteral">            set(axesHandle, &#39;</span>YTick<span class="stringliteral">&#39;, get(xAxesHandles{1}, &#39;</span>XTick<span class="stringliteral">&#39;));</span>
<a name="l04597"></a>04597 <span class="stringliteral">            ylabel(axesHandle, {&#39;</span><a class="code" href="class_replication.html">Replication</a> Initiation Time<span class="stringliteral">&#39;});</span>
<a name="l04598"></a>04598 <span class="stringliteral">            </span>
<a name="l04599"></a>04599 <span class="stringliteral">            axesHandle = axesHandles{2}(1);</span>
<a name="l04600"></a>04600 <span class="stringliteral">            cla(axesHandle);</span>
<a name="l04601"></a>04601 <span class="stringliteral">            edges = linspace(min(secondaryInitiationCnts), max(max(secondaryInitiationCnts), 9), 10);</span>
<a name="l04602"></a>04602 <span class="stringliteral">            freq1 = histc(secondaryInitiationCnts, edges);</span>
<a name="l04603"></a>04603 <span class="stringliteral">            barh(axesHandle, edges, freq1, &#39;</span>EdgeColor<span class="stringliteral">&#39;, &#39;</span>b<span class="stringliteral">&#39;, &#39;</span>FaceColor<span class="stringliteral">&#39;, &#39;</span>b<span class="stringliteral">&#39;);</span>
<a name="l04604"></a>04604 <span class="stringliteral">            ylabel(axesHandle, &#39;</span>Secondary Initiation Attempts<span class="stringliteral">&#39;);</span>
<a name="l04605"></a>04605 <span class="stringliteral">            ylim(axesHandle, edges([1 end]) + [-0.5 0.5]);</span>
<a name="l04606"></a>04606 <span class="stringliteral">            set(axesHandle, &#39;</span>YTick<span class="stringliteral">&#39;, [0 max(get(axesHandle, &#39;</span>YTick<span class="stringliteral">&#39;))]);</span>
<a name="l04607"></a>04607 <span class="stringliteral">            </span>
<a name="l04608"></a>04608 <span class="stringliteral">            axesHandle = axesHandles{2}(2);</span>
<a name="l04609"></a>04609 <span class="stringliteral">            cla(axesHandle);</span>
<a name="l04610"></a>04610 <span class="stringliteral">            edges = (0:29)&#39;</span>;
<a name="l04611"></a>04611             freq2 = histc(dnaAComplexSizes, edges);
<a name="l04612"></a>04612             barh(axesHandle, edges, freq2, <span class="stringliteral">&#39;EdgeColor&#39;</span>, <span class="charliteral">&#39;b&#39;</span>, <span class="stringliteral">&#39;FaceColor&#39;</span>, <span class="charliteral">&#39;b&#39;</span>);
<a name="l04613"></a>04613             ylabel(axesHandle, <span class="stringliteral">&#39;DnaA Complex Size&#39;</span>);
<a name="l04614"></a>04614             ylim(axesHandle, edges([1 end]) + [-0.5; 0.5]);
<a name="l04615"></a>04615             <span class="keyword">set</span>(axesHandle, <span class="stringliteral">&#39;YTick&#39;</span>, 0:4:28);
<a name="l04616"></a>04616             <span class="keyword">set</span>(axesHandle, <span class="stringliteral">&#39;YTickLabel&#39;</span>, 0:7);
<a name="l04617"></a>04617             
<a name="l04618"></a>04618             xlim(xAxesHandles{2}, [min([freq1; freq2]) max([freq1; freq2])]);
<a name="l04619"></a>04619             xlim(axesHandles{2}(1), [min([freq1; freq2]) max([freq1; freq2])]);
<a name="l04620"></a>04620             xlim(axesHandles{2}(2), [min([freq1; freq2]) max([freq1; freq2])]);
<a name="l04621"></a>04621             <span class="keyword">set</span>(xAxesHandles{2}, <span class="stringliteral">&#39;XTickMode&#39;</span>, <span class="stringliteral">&#39;auto&#39;</span>);
<a name="l04622"></a>04622             
<a name="l04623"></a>04623             PlotUtil.offsetYAxes(axesHandles, 0.03);
<a name="l04624"></a>04624             
<a name="l04625"></a>04625             %% data
<a name="l04626"></a>04626             figData = <span class="keyword">struct</span>;
<a name="l04627"></a>04627             figData.repInitTimes = repInitTimes;
<a name="l04628"></a>04628             figData.secondaryInitiationTimes = secondaryInitiationTimes;
<a name="l04629"></a>04629             figData.secondaryInitiationInitTimes = secondaryInitiationInitTimes;
<a name="l04630"></a>04630             figData.secondaryInitiationCnts = secondaryInitiationCnts;
<a name="l04631"></a>04631             figData.dnaAComplexSizes = dnaAComplexSizes;
<a name="l04632"></a>04632         end
<a name="l04633"></a>04633         
<a name="l04634"></a>04634         <span class="keyword">function</span> [figData, tabContent, tabColLabels] = warnings(figHandle, simBatchDir, selectedSimulations)
<a name="l04635"></a>04635             import edu.stanford.covert.cell.sim.util.SimulationEnsemble;
<a name="l04636"></a>04636             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
<a name="l04637"></a>04637             import edu.stanford.covert.cell.sim.util.SummaryLogger;
<a name="l04638"></a>04638             import edu.stanford.covert.cell.sim.util.PlotUtil;
<a name="l04639"></a>04639             
<a name="l04640"></a>04640             if ~exist(&#39;simBatchDir&#39;, &#39;var&#39;) || isempty(simBatchDir)
<a name="l04641"></a>04641                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
<a name="l04642"></a>04642             end
<a name="l04643"></a>04643             if ~exist(&#39;selectedSimulations&#39;, &#39;var&#39;) || isempty(selectedSimulations)
<a name="l04644"></a>04644                 selectedSimulations = SimulationDiskUtil.getCompleteSimulations(simBatchDir);
<a name="l04645"></a>04645             end
<a name="l04646"></a>04646             
<a name="l04647"></a>04647             %% get data
<a name="l04648"></a>04648             warnings = [];
<a name="l04649"></a>04649             freq = zeros(10, 1);
<a name="l04650"></a>04650             for i = 1:numel(selectedSimulations)
<a name="l04651"></a>04651                 tmpWarnings = SimulationDiskUtil.getSimulationWarnings(simBatchDir, selectedSimulations(i));
<a name="l04652"></a>04652                 
<a name="l04653"></a>04653                 repeatedWarning = 0;
<a name="l04654"></a>04654                 for k = 1:numel(tmpWarnings)
<a name="l04655"></a>04655                     if strcmp(tmpWarnings(k).message, &#39;Unable to represent multiple replication initiation events per cell cycle&#39;)
<a name="l04656"></a>04656                         freq(1) = freq(1) + numel(tmpWarnings(k).times);
<a name="l04657"></a>04657                     elseif strcmp(tmpWarnings(k).message, &#39;DnaA complex not decayed&#39;)
<a name="l04658"></a>04658                         freq(2) = freq(2) + numel(tmpWarnings(k).times);
<a name="l04659"></a>04659                     elseif strcmp(tmpWarnings(k).message, &#39;FtsZ not decayed&#39;)
<a name="l04660"></a>04660                         freq(3) = freq(3) + numel(tmpWarnings(k).times);
<a name="l04661"></a>04661                     elseif strcmp(tmpWarnings(k).message, &#39;DNA polymerase not decayed&#39;)
<a name="l04662"></a>04662                         freq(4) = freq(4) + numel(tmpWarnings(k).times);
<a name="l04663"></a>04663                     elseif strcmp(tmpWarnings(k).message, &#39;Cell has divided. Cell shape undefined&#39;)
<a name="l04664"></a>04664                         freq(5) = freq(5) + numel(tmpWarnings(k).times);
<a name="l04665"></a>04665                     elseif length(tmpWarnings(k).message) &gt; 24 &amp;&amp; strcmp(tmpWarnings(k).message(1:24), &#39;Linear programming error&#39;)
<a name="l04666"></a>04666                         freq(6) = freq(6) + numel(tmpWarnings(k).times);
<a name="l04667"></a>04667                     elseif length(tmpWarnings(k).message) &gt; 50 &amp;&amp; strcmp(tmpWarnings(k).message(1:50), &#39;Simulation state side effects should preserve mass&#39;)
<a name="l04668"></a>04668                         freq(7) = freq(7) + numel(tmpWarnings(k).times);
<a name="l04669"></a>04669                     elseif length(tmpWarnings(k).message) &gt; 35 &amp;&amp; strcmp(tmpWarnings(k).message(1:35), &#39;Summary figure could not be saved: &#39;)
<a name="l04670"></a>04670                         freq(8) = freq(8) + numel(tmpWarnings(k).times);
<a name="l04671"></a>04671                     elseif regexp(tmpWarnings(k).message, &#39;metabolites have negative counts&#39;)
<a name="l04672"></a>04672                         freq(9) = freq(9) + numel(tmpWarnings(k).times);
<a name="l04673"></a>04673                     else
<a name="l04674"></a>04674                         freq(10) = freq(10) + numel(tmpWarnings(k).times);
<a name="l04675"></a>04675                     end
<a name="l04676"></a>04676                     
<a name="l04677"></a>04677                     tmpWarnings(k).simulations = repmat(i, size(tmpWarnings(k).times));
<a name="l04678"></a>04678                     
<a name="l04679"></a>04679                     for j = 1:numel(warnings)
<a name="l04680"></a>04680                         if isequal(warnings(j).message, tmpWarnings(k).message) &amp;&amp; ...
<a name="l04681"></a>04681                                 isequal(warnings(j).files, tmpWarnings(k).files) &amp;&amp; ...
<a name="l04682"></a>04682                                 isequal(warnings(j).lineNumbers, tmpWarnings(k).lineNumbers)
<a name="l04683"></a>04683                             repeatedWarning = j;
<a name="l04684"></a>04684                             break;
<a name="l04685"></a>04685                         end
<a name="l04686"></a>04686                     end
<a name="l04687"></a>04687                     
<a name="l04688"></a>04688                     if isempty(warnings)
<a name="l04689"></a>04689                         warnings = tmpWarnings(k);
<a name="l04690"></a>04690                     elseif repeatedWarning
<a name="l04691"></a>04691                         warnings(repeatedWarning).times = ...
<a name="l04692"></a>04692                             [warnings(repeatedWarning).times; tmpWarnings(k).times]; %<span class="preprocessor">#ok&lt;AGROW&gt;</span>
<a name="l04693"></a>04693 <span class="preprocessor"></span>                        warnings(repeatedWarning).simulations = ...
<a name="l04694"></a>04694                             [warnings(repeatedWarning).simulations; tmpWarnings(k).simulations]; %#ok&lt;AGROW&gt;
<a name="l04695"></a>04695                     <span class="keywordflow">else</span>
<a name="l04696"></a>04696                         warnings = [warnings; tmpWarnings(k)]; %#ok&lt;AGROW&gt;
<a name="l04697"></a>04697                     end
<a name="l04698"></a>04698                 end
<a name="l04699"></a>04699             end
<a name="l04700"></a>04700             
<a name="l04701"></a>04701             %% plot data
<a name="l04702"></a>04702             clf(figHandle);
<a name="l04703"></a>04703             axesHandle = subplot(1, 1, 1, <span class="stringliteral">&#39;Parent&#39;</span>, figHandle);
<a name="l04704"></a>04704             barh(axesHandle, freq)
<a name="l04705"></a>04705             set(axesHandle, &#39;YTick&#39;, 1:numel(freq),  &#39;YTickLabel&#39;, {<span class="stringliteral">&#39;2nd Rep Init&#39;</span>, <span class="stringliteral">&#39;DnaA Decay&#39;</span>, <span class="stringliteral">&#39;FtsZ Decay&#39;</span>, <span class="stringliteral">&#39;Replisome Decay&#39;</span>, <span class="stringliteral">&#39;Shape Undef&#39;</span>, <span class="stringliteral">&#39;Lin Prog&#39;</span>, <span class="stringliteral">&#39;Neg Mets&#39;</span>, <span class="stringliteral">&#39;Sum Fig Save&#39;</span>, <span class="stringliteral">&#39;Mass Bal&#39;</span>, <span class="stringliteral">&#39;Other&#39;</span>});
<a name="l04706"></a>04706             ylabel(axesHandle, <span class="stringliteral">&#39;Warning&#39;</span>);
<a name="l04707"></a>04707             xlabel(axesHandle, <span class="stringliteral">&#39;Frequency&#39;</span>);
<a name="l04708"></a>04708             <span class="keyword">set</span>(axesHandle, <span class="stringliteral">&#39;XScale&#39;</span>, <span class="stringliteral">&#39;log&#39;</span>);
<a name="l04709"></a>04709             ylim(axesHandle, [0.5 numel(freq) + 0.5]);
<a name="l04710"></a>04710             
<a name="l04711"></a>04711             %% table
<a name="l04712"></a>04712             warnings = warnings(~ismember({warnings.message}, {
<a name="l04713"></a>04713                 <span class="stringliteral">&#39;Unable to represent multiple replication initiation events per cell cycle&#39;</span>
<a name="l04714"></a>04714                 <span class="stringliteral">&#39;DnaA complex not decayed&#39;</span>
<a name="l04715"></a>04715                 <span class="stringliteral">&#39;FtsZ not decayed&#39;</span>
<a name="l04716"></a>04716                 <span class="stringliteral">&#39;DNA polymerase not decayed&#39;</span>
<a name="l04717"></a>04717                 <span class="stringliteral">&#39;Cell has divided. Cell shape undefined&#39;</span>
<a name="l04718"></a>04718                 }));
<a name="l04719"></a>04719             tfs = <span class="keyword">true</span>(size(warnings));
<a name="l04720"></a>04720             <span class="keywordflow">for</span> i = 1:numel(warnings)
<a name="l04721"></a>04721                 tfs(i) = length(warnings(i).message) &lt; 35 || ~isequal(warnings(i).message(1:35), &#39;Summary figure could not be saved: &#39;);
<a name="l04722"></a>04722             end
<a name="l04723"></a>04723             warnings = warnings(tfs);
<a name="l04724"></a>04724             tabColLabels = {<span class="stringliteral">&#39;Warning&#39;</span>, <span class="stringliteral">&#39;Stack Trace&#39;</span>, <span class="stringliteral">&#39;Simulation.Times&#39;</span>};
<a name="l04725"></a>04725             tabContent = cell(numel(warnings), 3);
<a name="l04726"></a>04726             <span class="keywordflow">for</span> i = 1:numel(warnings)
<a name="l04727"></a>04727                 tabContent{i, 1} = warnings(i).message;
<a name="l04728"></a>04728                 
<a name="l04729"></a>04729                 <span class="keywordflow">for</span> j = 1:numel(warnings(i).files)
<a name="l04730"></a>04730                     tabContent{i, 2} = [tabContent{i, 2} ...
<a name="l04731"></a>04731                         sprintf(&#39;%s at %d\n&#39;, warnings(i).files{j}, warnings(i).lineNumbers(j))];
<a name="l04732"></a>04732                 end
<a name="l04733"></a>04733                 tabContent{i, 2} = tabContent{i, 2}(1:end-1);
<a name="l04734"></a>04734                 
<a name="l04735"></a>04735                 <span class="keywordflow">for</span> j = 1:numel(warnings(i).simulations)
<a name="l04736"></a>04736                     tabContent{i, 3} = [tabContent{i, 3} ...
<a name="l04737"></a>04737                         sprintf(&#39;%d.%d\n&#39;, warnings(i).simulations(j), warnings(i).times(j))];
<a name="l04738"></a>04738                 end
<a name="l04739"></a>04739                 tabContent{i, 3} = tabContent{i, 3}(1:end-1);
<a name="l04740"></a>04740             end
<a name="l04741"></a>04741             
<a name="l04742"></a>04742             %% data
<a name="l04743"></a>04743             figData = <span class="keyword">struct</span>;
<a name="l04744"></a>04744             figData.freq = freq;
<a name="l04745"></a>04745         end
<a name="l04746"></a>04746     end
<a name="l04747"></a>04747     
<a name="l04748"></a>04748     %helper methods
<a name="l04749"></a>04749     methods (Static = <span class="keyword">true</span>)
<a name="l04750"></a>04750         function [unsyncTime, unsyncData, cellCount, simOrder] = desynchronizePopulation(sim, syncTime, syncData, simEndTimes)
<a name="l04751"></a>04751             finTimeIdx = min(simEndTimes);
<a name="l04752"></a>04752             finTime = syncTime(finTimeIdx);
<a name="l04753"></a>04753             
<a name="l04754"></a>04754             cellCycleLength = sim.state(&#39;Time&#39;).cellCycleLength;
<a name="l04755"></a>04755             randStream = edu.stanford.covert.util.<a class="code" href="class_rand_stream.html">RandStream</a>(&#39;mcg16807&#39;);
<a name="l04756"></a>04756             
<a name="l04757"></a>04757             unsyncData = NaN(size(syncData, 1), size(syncData, 2), min(simEndTimes), 2 * size(syncData, 4));
<a name="l04758"></a>04758             simStartTimes = zeros(size(syncData, 4), 1);
<a name="l04759"></a>04759             for i = 1:ceil(size(syncData, 4) * (1 - exp(-log(2) * finTimeIdx / cellCycleLength)))
<a name="l04760"></a>04760                 simTimeIdx = ceil(randStream.rand * finTimeIdx);
<a name="l04761"></a>04761                 unsyncData(:, :, 1:simTimeIdx, i) = syncData(:, :, simEndTimes(i)-simTimeIdx+1:simEndTimes(i), i);
<a name="l04762"></a>04762                 unsyncData(:, :, simTimeIdx+1:end, i) = syncData(:, :, 1:finTimeIdx - simTimeIdx, i);
<a name="l04763"></a>04763             end
<a name="l04764"></a>04764             for j = i+1:size(syncData, 4)
<a name="l04765"></a>04765                 simStartTimes(j) = randStream.stochasticRound(log(1 + randStream.rand * (exp(log(2) * finTime / cellCycleLength) - 1)) * cellCycleLength / log(2));
<a name="l04766"></a>04766                 simTimeIdx = find(syncTime == simStartTimes(j));
<a name="l04767"></a>04767                 if randStream.rand &gt; 0.5
<a name="l04768"></a>04768                     unsyncData(:, :, simTimeIdx+1:end, j) = syncData(:, :, 1:finTimeIdx - simTimeIdx, j);
<a name="l04769"></a>04769                 else
<a name="l04770"></a>04770                     unsyncData(:, :, simTimeIdx+1:end, j) = syncData(:, :, simEndTimes(j) - (finTimeIdx - simTimeIdx)+1:simEndTimes(j), j);
<a name="l04771"></a>04771                 end
<a name="l04772"></a>04772             end
<a name="l04773"></a>04773             [~, simOrder] = sort(simStartTimes);
<a name="l04774"></a>04774             
<a name="l04775"></a>04775             cellCount = sum(~isnan(unsyncData(1, :, :, :)), 4);
<a name="l04776"></a>04776             
<a name="l04777"></a>04777             unsyncTime = syncTime(:, 1:finTimeIdx);
<a name="l04778"></a>04778         end
<a name="l04779"></a>04779     end
<a name="l04780"></a>04780 end
</pre></div></div><!-- contents -->
</div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
<a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(0)"><span class="SelectionMark">&#160;</span>All</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(1)"><span class="SelectionMark">&#160;</span>Classes</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(2)"><span class="SelectionMark">&#160;</span>Files</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(3)"><span class="SelectionMark">&#160;</span>Functions</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(4)"><span class="SelectionMark">&#160;</span>Variables</a></div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

  <div id="nav-path" class="navpath">
    <ul>
      <li class="navelem"><a class="el" href="_population_8m.html">Population.m</a>      </li>

    <li class="footer">Generated on Wed Feb 8 2012 21:42:31 for Whole Cell Simulation by
    <a href="http://www.doxygen.org/index.html">
    <img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.7.6.1 </li>
   </ul>
 </div>


</body>
</html>
