<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<title>Whole Cell Simulation: src/+edu/+stanford/+covert/+cell/+sim/+process/ProteinProcessingI.m Source File</title>

<link href="tabs.css" rel="stylesheet" type="text/css"/>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
<link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="navtree.js"></script>
<script type="text/javascript">
  $(document).ready(initResizable);
</script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
  $(document).ready(function() { searchBox.OnSelectItem(0); });
</script>

</head>
<body>
<div id="top"><!-- do not remove this div! -->


<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  
  
  <td style="padding-left: 0.5em;">
   <div id="projectname">Whole Cell Simulation
   
   </div>
   
  </td>
  
  
  
 </tr>
 </tbody>
</table>
</div>

<!-- Generated by Doxygen 1.7.6.1 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "search",false,'Search');
</script>
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&#160;Page</span></a></li>
      <li><a href="annotated.html"><span>Classes</span></a></li>
      <li class="current"><a href="files.html"><span>Files</span></a></li>
      <li>
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
  <div id="navrow2" class="tabs2">
    <ul class="tablist">
      <li><a href="files.html"><span>File&#160;List</span></a></li>
      <li><a href="globals.html"><span>File&#160;Members</span></a></li>
    </ul>
  </div>
</div>
<div id="side-nav" class="ui-resizable side-nav-resizable">
  <div id="nav-tree">
    <div id="nav-tree-contents">
    </div>
  </div>
  <div id="splitbar" style="-moz-user-select:none;" 
       class="ui-resizable-handle">
  </div>
</div>
<script type="text/javascript">
  initNavTree('_protein_processing_i_8m.html','');
</script>
<div id="doc-content">
<div class="header">
  <div class="headertitle">
<div class="title">src/+edu/+stanford/+covert/+cell/+sim/+process/ProteinProcessingI.m</div>  </div>
</div><!--header-->
<div class="contents">
<a href="_protein_processing_i_8m.html">Go to the documentation of this file.</a><div class="fragment"><pre class="fragment"><a name="l00001"></a>00001 %<a class="code" href="class_protein.html">Protein</a> Processing I
<a name="l00002"></a>00002 %
<a name="l00003"></a>00003 % @wholeCellModelID Process_ProteinProcessingI
<a name="l00004"></a>00004 % @name             <a class="code" href="class_protein.html">Protein</a> Processing I
<a name="l00005"></a>00005 % @description
<a name="l00006"></a>00006 %   Biology
<a name="l00007"></a>00007 %   ===========
<a name="l00008"></a>00008 %   Following translation, nascent peptides are deformylated, cleaved,
<a name="l00009"></a>00009 %   translocated, folded, and modified. First, peptide deformylase (MG_106)
<a name="l00010"></a>00010 %   deformylates the N-terminal formylmethionine of each nascent peptide.
<a name="l00011"></a>00011 %   Second, methionine aminopeptidase (MG_172) cleaves the N-terminal methionine
<a name="l00012"></a>00012 %   of 35 peptides. Second (see protein translocation process), 117 integral
<a name="l00013"></a>00013 %   membrane, lipoproteins, and extracellular proteins bind the SecA translocase
<a name="l00014"></a>00014 %   and are translocated into and through the plasma membrane via the
<a name="l00015"></a>00015 %   SecYEGDF-YidC pore. Third (see protein processing II process), diacylglyceryl
<a name="l00016"></a>00016 %   is transferred to the C-terminal cysteine of the signal sequence of each
<a name="l00017"></a>00017 %   lipoprotein by diacylglyceryl transferase, and the signal sequence of each
<a name="l00018"></a>00018 %   lipoprotein is cleaved by signal peptidase II. Next (see protein
<a name="l00019"></a>00019 %   folding process), 85 peptides bind inorganic ions at particular sites
<a name="l00020"></a>00020 %   and 64 peptides fold with the assistance of the chaperones and chaperonins
<a name="l00021"></a>00021 %   DnaJ, DnaK, GroEL, GroES, and GrpE. All peptides require trigger factor
<a name="l00022"></a>00022 %   to properly fold. Finally (see protein modification process), 20 peptide
<a name="l00023"></a>00023 %   species are modified at 63 sites by 3 enzymes – serine/threonine protein
<a name="l00024"></a>00024 %   kinase, lipoate ligase, and alpha glutamate ligase.
<a name="l00025"></a>00025 %
<a name="l00026"></a>00026 %   This process simulates
<a name="l00027"></a>00027 %   - N-terminal peptide deformylation, and
<a name="l00028"></a>00028 %   - N-terminal amino acid cleavage.
<a name="l00029"></a>00029 %
<a name="l00030"></a>00030 %   Knowledge Base
<a name="l00031"></a>00031 %   ===========
<a name="l00032"></a>00032 %   Every M. genitalium protein requires deformylation, and roughly 7%
<a name="l00033"></a>00033 %   require N-terminal amino acid cleavage. N-terminal methionine cleavages were
<a name="l00034"></a>00034 %   reconstructed by mapping N-terminal methionine cleavages observed in
<a name="l00035"></a>00035 %   Shewanella oneidensis MR-1 [PUB_0280] onto homologous M. genitalium genes.
<a name="l00036"></a>00036 %   The N-terminal methionine cleavage state of each protein monomer is
<a name="l00037"></a>00037 %   organized in the knowledge base, and loaded into the
<a name="l00038"></a>00038 %   nascentMonomerNTerminalMethionineCleavages property of this process by the
<a name="l00039"></a>00039 %   initializeConstants method.
<a name="l00040"></a>00040 %
<a name="l00041"></a>00041 %   Representation
<a name="l00042"></a>00042 %   ===========
<a name="l00043"></a>00043 %   substrates, enzymes, unprocessedMonomers, and processedMonomers represent
<a name="l00044"></a>00044 %   the counts of metabolites, the deformylase and methionine aminopeptidase,
<a name="l00045"></a>00045 %   and nascent and deformylated, cleaved protein monomers. The compartment
<a name="l00046"></a>00046 %   dimension of each of these <a class="code" href="json_parse_8m.html#a6e4256f69b751d3d7ec4ea52ff5a8dee">properties</a> has length 1. That is the process only
<a name="l00047"></a>00047 %   accesses the counts of these objects in the relevant compartments. This
<a name="l00048"></a>00048 %   process doesn&#39;t represent any additional intermediate processed states. This
<a name="l00049"></a>00049 %   process treats N-terminal deformylation and methionine cleavage as an
<a name="l00050"></a>00050 %   all-or-nothing event that either proceeds to complete within a single time
<a name="l00051"></a>00051 %   step, or does not occur at all.
<a name="l00052"></a>00052 %
<a name="l00053"></a>00053 %   nascentMonomerNTerminalMethionineCleavages is a <span class="keywordtype">boolean</span> which represents
<a name="l00054"></a>00054 %   whether or not the N-terminal methionine of each protein monomer must be
<a name="l00055"></a>00055 %   cleaved.
<a name="l00056"></a>00056 %
<a name="l00057"></a>00057 %   Initialization
<a name="l00058"></a>00058 %   ===============
<a name="l00059"></a>00059 %   All protein monomers are initialized to the mature state. This is
<a name="l00060"></a>00060 %   implemented by the simulation class initializeState method.
<a name="l00061"></a>00061 %
<a name="l00062"></a>00062 %   Simulation
<a name="l00063"></a>00063 %   ===========
<a name="l00064"></a>00064 %   1. Compute the maximum number of peptides that can be processed based on the
<a name="l00065"></a>00065 %      availability of the two enzymes.
<a name="l00066"></a>00066 %   2. Randomly select peptides to be processed, weighted by the counts of each
<a name="l00067"></a>00067 %      protein species.
<a name="l00068"></a>00068 %   3. Update the counts of unprocessed and processed proteins monomers.
<a name="l00069"></a>00069 %      Decrement the counts of available enzyme activity.
<a name="l00070"></a>00070 %   4. Compute the maximum number of peptides which don&#39;t require cleavage that
<a name="l00071"></a>00071 %      can be processed based on the availability of peptide deformylase
<a name="l00072"></a>00072 %      activity.
<a name="l00073"></a>00073 %   5. Randomly select peptides to be deformylated, weighted by the counts of
<a name="l00074"></a>00074 %      each protein species.
<a name="l00075"></a>00075 %   6. Update the counts of unprocessed and processed proteins monomers.
<a name="l00076"></a>00076 %      Decrement the counts of available deformylase activity.
<a name="l00077"></a>00077 %
<a name="l00078"></a>00078 % Author: Jonathan Karr, jkarr@stanford.edu
<a name="l00079"></a>00079 % Affilitation: Covert Lab, Department of Bioengineering, Stanford University
<a name="l00080"></a>00080 % Last updated: 7/30/2010
<a name="l00081"></a>00081 classdef ProteinProcessingI &lt; edu.stanford.covert.cell.sim.Process
<a name="l00082"></a>00082     %property annotations
<a name="l00083"></a>00083     properties (Constant)
<a name="l00084"></a>00084         optionNames__              = {}; %names of option <a class="code" href="json_parse_8m.html#a6e4256f69b751d3d7ec4ea52ff5a8dee">properties</a>
<a name="l00085"></a>00085         fixedConstantNames__       = {   %names of fixed constant <a class="code" href="json_parse_8m.html#a6e4256f69b751d3d7ec4ea52ff5a8dee">properties</a>
<a name="l00086"></a>00086                     <span class="stringliteral">&#39;deformylaseSpecificRate&#39;</span>;
<a name="l00087"></a>00087             <span class="stringliteral">&#39;methionineAminoPeptidaseSpecificRate&#39;</span>;
<a name="l00088"></a>00088             <span class="stringliteral">&#39;nascentMonomerNTerminalMethionineCleavages&#39;</span>;            
<a name="l00089"></a>00089                     };
<a name="l00090"></a>00090         fittedConstantNames__      = {}; %names of fitted constant <a class="code" href="json_parse_8m.html#a6e4256f69b751d3d7ec4ea52ff5a8dee">properties</a>
<a name="l00091"></a>00091         localStateNames__          = {   %names of simulation state <a class="code" href="json_parse_8m.html#a6e4256f69b751d3d7ec4ea52ff5a8dee">properties</a> redundant with timecourses in <span class="keyword">this</span> or other processes or the simulation
<a name="l00092"></a>00092             <span class="stringliteral">&#39;unprocessedMonomers&#39;</span>;
<a name="l00093"></a>00093             <span class="stringliteral">&#39;processedMonomers&#39;</span>};
<a name="l00094"></a>00094     end
<a name="l00095"></a>00095 
<a name="l00096"></a>00096     %IDs, names, and local indices
<a name="l00097"></a>00097     <a class="code" href="json_parse_8m.html#a6e4256f69b751d3d7ec4ea52ff5a8dee">properties</a>
<a name="l00098"></a>00098         stimuliWholeCellModelIDs = {};  %whole cell model IDs of stimuli
<a name="l00099"></a>00099 
<a name="l00100"></a>00100         substrateWholeCellModelIDs = {  %whole cell model IDs of substrates
<a name="l00101"></a>00101             <span class="stringliteral">&#39;H2O&#39;</span>;
<a name="l00102"></a>00102             <span class="charliteral">&#39;H&#39;</span>;
<a name="l00103"></a>00103             <span class="stringliteral">&#39;MET&#39;</span>;
<a name="l00104"></a>00104             <span class="stringliteral">&#39;FOR&#39;</span>};
<a name="l00105"></a>00105         substrateIndexs_water      = 1; %index within substrates of water
<a name="l00106"></a>00106         substrateIndexs_hydrogen   = 2; %index within substrates of hydrogen
<a name="l00107"></a>00107         substrateIndexs_methionine = 3; %index within substrates of methionine
<a name="l00108"></a>00108         substrateIndexs_formate    = 4; %index within substrates of formate
<a name="l00109"></a>00109 
<a name="l00110"></a>00110         enzymeWholeCellModelIDs={       %enzyme whole cell model ids
<a name="l00111"></a>00111             <span class="stringliteral">&#39;MG_106_DIMER&#39;</span>;             %peptide deformylase
<a name="l00112"></a>00112             <span class="stringliteral">&#39;MG_172_MONOMER&#39;</span>};          %methionine aminopeptidase, type I
<a name="l00113"></a>00113         enzymeIndexs_deformylase              = 1; %index within enzymes of peptide deformylase
<a name="l00114"></a>00114         enzymeIndexs_methionineAminoPeptidase = 2; %index within enzymes of methionine aminopeptidase, type I
<a name="l00115"></a>00115 
<a name="l00116"></a>00116         unprocessedMonomerWholeCellModelIDs %whole cell model IDs of unprocessed protein monomers
<a name="l00117"></a>00117         processedMonomerWholeCellModelIDs   %whole cell model IDs of processed protein monomers
<a name="l00118"></a>00118     end
<a name="l00119"></a>00119     
<a name="l00120"></a>00120     %fixed biological constants
<a name="l00121"></a>00121     <a class="code" href="json_parse_8m.html#a6e4256f69b751d3d7ec4ea52ff5a8dee">properties</a>
<a name="l00122"></a>00122         deformylaseSpecificRate                     %number of reactions per second (38) [PUB_0021]
<a name="l00123"></a>00123         methionineAminoPeptidaseSpecificRate        %number of reactions per second (6) [PUB_0026]
<a name="l00124"></a>00124         nascentMonomerNTerminalMethionineCleavages  %whether N-terminal amino acid needs to be cleaved        
<a name="l00125"></a>00125     end
<a name="l00126"></a>00126 
<a name="l00127"></a>00127     %global state (stored locally for convenience)
<a name="l00128"></a>00128     <a class="code" href="json_parse_8m.html#a6e4256f69b751d3d7ec4ea52ff5a8dee">properties</a>
<a name="l00129"></a>00129         unprocessedMonomers %counts of unprocessed protein monomers
<a name="l00130"></a>00130         processedMonomers   %counts of processed protein monomers
<a name="l00131"></a>00131     end
<a name="l00132"></a>00132 
<a name="l00133"></a>00133     %constructor
<a name="l00134"></a>00134     methods
<a name="l00135"></a>00135         function this = ProteinProcessingI(wholeCellModelID, name)
<a name="l00136"></a>00136             this = this@edu.stanford.covert.cell.sim.<a class="code" href="class_process.html">Process</a>(wholeCellModelID, name);
<a name="l00137"></a>00137         end
<a name="l00138"></a>00138     end
<a name="l00139"></a>00139 
<a name="l00140"></a>00140     %communication between process/simulation
<a name="l00141"></a>00141     methods
<a name="l00142"></a>00142         %initialize constants
<a name="l00143"></a>00143         <span class="keyword">function</span> initializeConstants(<span class="keyword">this</span>, knowledgeBase, simulation, varargin)
<a name="l00144"></a>00144             this.initializeConstants@edu.stanford.covert.cell.sim.<a class="code" href="class_process.html">Process</a>(knowledgeBase, simulation, varargin{:});
<a name="l00145"></a>00145 
<a name="l00146"></a>00146             this.unprocessedMonomerWholeCellModelIDs = this.monomer.wholeCellModelIDs(this.monomer.nascentIndexs);
<a name="l00147"></a>00147             this.processedMonomerWholeCellModelIDs   = this.monomer.wholeCellModelIDs(this.monomer.processedIIndexs);
<a name="l00148"></a>00148 
<a name="l00149"></a>00149             this.nascentMonomerNTerminalMethionineCleavages = knowledgeBase.proteinMonomerNTerminalMethionineCleavages;
<a name="l00150"></a>00150         end
<a name="l00151"></a>00151 
<a name="l00152"></a>00152         %retrieve state from simulation
<a name="l00153"></a>00153         <span class="keyword">function</span> copyFromState(<span class="keyword">this</span>)
<a name="l00154"></a>00154             this.copyFromState@edu.stanford.covert.cell.sim.<a class="code" href="class_process.html">Process</a>();
<a name="l00155"></a>00155 
<a name="l00156"></a>00156             this.unprocessedMonomers = this.monomer.counts(this.monomer.nascentIndexs,    this.compartment.cytosolIndexs, :);
<a name="l00157"></a>00157             this.processedMonomers   = this.monomer.counts(this.monomer.processedIIndexs, this.compartment.cytosolIndexs, :);
<a name="l00158"></a>00158         end
<a name="l00159"></a>00159 
<a name="l00160"></a>00160         %send state to simulation
<a name="l00161"></a>00161         function copyToState(this)
<a name="l00162"></a>00162             this.copyToState@edu.stanford.covert.cell.sim.Process();
<a name="l00163"></a>00163 
<a name="l00164"></a>00164             this.monomer.counts(this.monomer.nascentIndexs,    this.compartment.cytosolIndexs, :) = this.unprocessedMonomers;
<a name="l00165"></a>00165             this.monomer.counts(this.monomer.processedIIndexs, this.compartment.cytosolIndexs, :) = this.processedMonomers;
<a name="l00166"></a>00166         end
<a name="l00167"></a>00167     end
<a name="l00168"></a>00168 
<a name="l00169"></a>00169     %allocate memory for state
<a name="l00170"></a>00170     methods
<a name="l00171"></a>00171         function allocateMemoryForState(this, numTimePoints)
<a name="l00172"></a>00172             this.allocateMemoryForState@edu.stanford.covert.cell.sim.Process(numTimePoints);
<a name="l00173"></a>00173 
<a name="l00174"></a>00174             this.unprocessedMonomers = zeros(length(this.unprocessedMonomerWholeCellModelIDs), 1, numTimePoints);
<a name="l00175"></a>00175             this.processedMonomers   = zeros(length(this.processedMonomerWholeCellModelIDs),   1, numTimePoints);
<a name="l00176"></a>00176         end
<a name="l00177"></a>00177     end
<a name="l00178"></a>00178 
<a name="l00179"></a>00179     %model
<a name="l00180"></a>00180     methods
<a name="l00181"></a>00181         %Calculate
<a name="l00182"></a>00182         %- contribution to <a class="code" href="class_f_b_a.html">FBA</a> objective
<a name="l00183"></a>00183         %- minimum expression consistent with cell cycle length
<a name="l00184"></a>00184         function [bmProd, byProd, minEnzExp, maxEnzExp] = calcResourceRequirements_LifeCycle(this, ~, states)
<a name="l00185"></a>00185             %% initialize
<a name="l00186"></a>00186             bmProd = zeros(size(this.substrateWholeCellModelIDs));
<a name="l00187"></a>00187             byProd = zeros(size(this.substrateWholeCellModelIDs));
<a name="l00188"></a>00188             minEnzExp = zeros(size(this.enzymeWholeCellModelIDs));
<a name="l00189"></a>00189             maxEnzExp = Inf(size(this.enzymeWholeCellModelIDs));
<a name="l00190"></a>00190             
<a name="l00191"></a>00191             %% substrate and byproducts
<a name="l00192"></a>00192             %deformylation
<a name="l00193"></a>00193             %formyl-L-methionyl peptide + H2O &lt;==&gt; methionyl peptide + formate + H+
<a name="l00194"></a>00194             bmProd(this.substrateIndexs_water)      = bmProd(this.substrateIndexs_water)    + sum(states.monomerProductions);
<a name="l00195"></a>00195             byProd(this.substrateIndexs_formate)    = byProd(this.substrateIndexs_formate)  + sum(states.monomerProductions);
<a name="l00196"></a>00196             byProd(this.substrateIndexs_hydrogen)   = byProd(this.substrateIndexs_hydrogen) + sum(states.monomerProductions);
<a name="l00197"></a>00197             
<a name="l00198"></a>00198             %N-terminal methionine cleavage
<a name="l00199"></a>00199             bmProd(this.substrateIndexs_water)      = bmProd(this.substrateIndexs_water)      + states.monomerProductions&#39; * this.nascentMonomerNTerminalMethionineCleavages;
<a name="l00200"></a>00200             byProd(this.substrateIndexs_methionine) = byProd(this.substrateIndexs_methionine) + states.monomerProductions&#39; * this.nascentMonomerNTerminalMethionineCleavages;
<a name="l00201"></a>00201             
<a name="l00202"></a>00202             %% enzymes
<a name="l00203"></a>00203             %deformylation
<a name="l00204"></a>00204             minEnzExp(this.enzymeIndexs_deformylase) = 2 * sum(states.monomerProductions0) / this.deformylaseSpecificRate;
<a name="l00205"></a>00205             
<a name="l00206"></a>00206             %N-terminal methionine cleavage
<a name="l00207"></a>00207             minEnzExp(this.enzymeIndexs_methionineAminoPeptidase) = ...
<a name="l00208"></a>00208                 2 * (states.monomerProductions0&#39; * this.nascentMonomerNTerminalMethionineCleavages) / ...
<a name="l00209"></a>00209                 this.methionineAminoPeptidaseSpecificRate;
<a name="l00210"></a>00210         end
<a name="l00211"></a>00211 
<a name="l00212"></a>00212         %initialization: monomers initialized to mature/bound/inactivated state
<a name="l00213"></a>00213         %by simulation initializeState method
<a name="l00214"></a>00214         function initializeState(~)
<a name="l00215"></a>00215         end
<a name="l00216"></a>00216 
<a name="l00217"></a>00217         %resource requirements
<a name="l00218"></a>00218         function result = calcResourceRequirements_Current(this)
<a name="l00219"></a>00219             %Numbers of enzymes
<a name="l00220"></a>00220             deformylaseLimit = ceil(this.enzymes(this.enzymeIndexs_deformylase) * this.deformylaseSpecificRate * this.stepSizeSec);
<a name="l00221"></a>00221             cleavageLimit = ceil(this.enzymes(this.enzymeIndexs_methionineAminoPeptidase) * ...
<a name="l00222"></a>00222                 this.methionineAminoPeptidaseSpecificRate * this.stepSizeSec);
<a name="l00223"></a>00223             
<a name="l00224"></a>00224             %initialize
<a name="l00225"></a>00225             result = zeros(size(this.substrates));
<a name="l00226"></a>00226             
<a name="l00227"></a>00227             %upper bound on metabolites
<a name="l00228"></a>00228             result(this.substrateIndexs_water) = min(...
<a name="l00229"></a>00229                 + min(deformylaseLimit, sum(this.unprocessedMonomers)) ...                                          %deformylation: formyl-L-methionyl peptide + H2O &lt;==&gt; methionyl peptide + formate + H+
<a name="l00230"></a>00230                 + min(cleavageLimit, this.nascentMonomerNTerminalMethionineCleavages&#39; * this.unprocessedMonomers)); %N-terminal methionine cleavage
<a name="l00231"></a>00231         end
<a name="l00232"></a>00232 
<a name="l00233"></a>00233         %simulation
<a name="l00234"></a>00234         %- Deformylation: formyl-L-methionyl peptide + H2O &lt;==&gt; methionyl peptide + formate + H+
<a name="l00235"></a>00235         %- N-terminal methionine cleavage
<a name="l00236"></a>00236         function evolveState(this)
<a name="l00237"></a>00237             %Numbers of enzymes
<a name="l00238"></a>00238             deformylaseLimit = this.enzymes(this.enzymeIndexs_deformylase) * this.deformylaseSpecificRate * this.stepSizeSec;
<a name="l00239"></a>00239             cleavageLimit = this.enzymes(this.enzymeIndexs_methionineAminoPeptidase) * ...
<a name="l00240"></a>00240                 this.methionineAminoPeptidaseSpecificRate * this.stepSizeSec;
<a name="l00241"></a>00241             
<a name="l00242"></a>00242             %Simulate deformylation and cleavage
<a name="l00243"></a>00243             if cleavageLimit &gt; 0 &amp;&amp; any(this.unprocessedMonomers(this.nascentMonomerNTerminalMethionineCleavages))
<a name="l00244"></a>00244                 %compute maximum of transformations that can occur, limited by
<a name="l00245"></a>00245                 %- unprocessed protein monomers
<a name="l00246"></a>00246                 %- enzymes (kinetics, availability)
<a name="l00247"></a>00247                 transformations = this.unprocessedMonomers;
<a name="l00248"></a>00248                 transformations = transformations * ...
<a name="l00249"></a>00249                     min(1, deformylaseLimit / sum(transformations));
<a name="l00250"></a>00250                 transformations(this.nascentMonomerNTerminalMethionineCleavages) = ...
<a name="l00251"></a>00251                     transformations(this.nascentMonomerNTerminalMethionineCleavages) * ...
<a name="l00252"></a>00252                     min(1, cleavageLimit / sum(transformations(this.nascentMonomerNTerminalMethionineCleavages)));
<a name="l00253"></a>00253                 transformations = this.randStream.stochasticRound(transformations);
<a name="l00254"></a>00254                 
<a name="l00255"></a>00255                 %water availability
<a name="l00256"></a>00256                 if sum(transformations) + sum(transformations(this.nascentMonomerNTerminalMethionineCleavages)) &gt; ...
<a name="l00257"></a>00257                        this.substrates(this.substrateIndexs_water)
<a name="l00258"></a>00258                    water_cleavage = this.randStream.stochasticRound(this.substrates(this.substrateIndexs_water) * ...
<a name="l00259"></a>00259                        2 * sum(transformations(this.nascentMonomerNTerminalMethionineCleavages)) / ...
<a name="l00260"></a>00260                        (sum(transformations) + sum(transformations(this.nascentMonomerNTerminalMethionineCleavages))));
<a name="l00261"></a>00261                    water_deformylations = this.substrates(this.substrateIndexs_water) - water_cleavage;
<a name="l00262"></a>00262                    if any(transformations(this.nascentMonomerNTerminalMethionineCleavages))
<a name="l00263"></a>00263                        transformations(this.nascentMonomerNTerminalMethionineCleavages) = ...
<a name="l00264"></a>00264                            min(transformations(this.nascentMonomerNTerminalMethionineCleavages), ...
<a name="l00265"></a>00265                            this.randStream.mnrnd(floor(water_cleavage / 2), ...
<a name="l00266"></a>00266                            transformations(this.nascentMonomerNTerminalMethionineCleavages) / ...
<a name="l00267"></a>00267                            sum(transformations(this.nascentMonomerNTerminalMethionineCleavages)))&#39;);
<a name="l00268"></a>00268                    end
<a name="l00269"></a>00269                    if any(transformations(~this.nascentMonomerNTerminalMethionineCleavages))
<a name="l00270"></a>00270                        transformations(~this.nascentMonomerNTerminalMethionineCleavages) = ...
<a name="l00271"></a>00271                            min(transformations(~this.nascentMonomerNTerminalMethionineCleavages), ...
<a name="l00272"></a>00272                            this.randStream.mnrnd(water_deformylations, ...
<a name="l00273"></a>00273                            transformations(~this.nascentMonomerNTerminalMethionineCleavages) / ...
<a name="l00274"></a>00274                            sum(transformations(~this.nascentMonomerNTerminalMethionineCleavages)))&#39;);
<a name="l00275"></a>00275                    end
<a name="l00276"></a>00276                 end
<a name="l00277"></a>00277                 
<a name="l00278"></a>00278                 %update numbers of monomers
<a name="l00279"></a>00279                 this.processedMonomers   = this.processedMonomers   + transformations;
<a name="l00280"></a>00280                 this.unprocessedMonomers = this.unprocessedMonomers - transformations;
<a name="l00281"></a>00281                 
<a name="l00282"></a>00282                 %update enzyme availability
<a name="l00283"></a>00283                 deformylaseLimit = max(0, deformylaseLimit - sum(transformations));
<a name="l00284"></a>00284                 
<a name="l00285"></a>00285                 %update metabolites
<a name="l00286"></a>00286                 this.substrates([this.substrateIndexs_water; this.substrateIndexs_formate; this.substrateIndexs_hydrogen]) = ...
<a name="l00287"></a>00287                     this.substrates([this.substrateIndexs_water; this.substrateIndexs_formate; this.substrateIndexs_hydrogen]) + ...
<a name="l00288"></a>00288                     [-1; 1; 1] * sum(transformations);
<a name="l00289"></a>00289                 
<a name="l00290"></a>00290                 this.substrates([this.substrateIndexs_water; this.substrateIndexs_methionine]) = ...
<a name="l00291"></a>00291                     this.substrates([this.substrateIndexs_water; this.substrateIndexs_methionine]) + ...
<a name="l00292"></a>00292                     [-1; 1] * sum(transformations(this.nascentMonomerNTerminalMethionineCleavages));
<a name="l00293"></a>00293             end
<a name="l00294"></a>00294 
<a name="l00295"></a>00295             %% Continue to simulate deformylation only
<a name="l00296"></a>00296 
<a name="l00297"></a>00297             %compute maximum of transformations that can occur, limited by
<a name="l00298"></a>00298             %- unprocessed protein monomers
<a name="l00299"></a>00299             %- enzymes (kinetics, availability)
<a name="l00300"></a>00300             transformations = this.unprocessedMonomers;
<a name="l00301"></a>00301             transformations(this.nascentMonomerNTerminalMethionineCleavages) = 0;
<a name="l00302"></a>00302             if deformylaseLimit &gt; 0 &amp;&amp; any(transformations)
<a name="l00303"></a>00303                 transformations = transformations * ...
<a name="l00304"></a>00304                     min(1, deformylaseLimit / sum(transformations));
<a name="l00305"></a>00305 
<a name="l00306"></a>00306                 transformations = this.randStream.stochasticRound(transformations);
<a name="l00307"></a>00307                 if sum(transformations) &gt; this.substrates(this.substrateIndexs_water)
<a name="l00308"></a>00308                     transformations = min(transformations, this.randStream.mnrnd(this.substrates(this.substrateIndexs_water), transformations / sum(transformations))&#39;);
<a name="l00309"></a>00309                 end
<a name="l00310"></a>00310 
<a name="l00311"></a>00311                 %update numbers of monomers
<a name="l00312"></a>00312                 this.processedMonomers   = this.processedMonomers   + transformations;
<a name="l00313"></a>00313                 this.unprocessedMonomers = this.unprocessedMonomers - transformations;
<a name="l00314"></a>00314 
<a name="l00315"></a>00315                 %update metabolites
<a name="l00316"></a>00316                 this.substrates([this.substrateIndexs_water; this.substrateIndexs_formate; this.substrateIndexs_hydrogen]) = ...
<a name="l00317"></a>00317                     this.substrates([this.substrateIndexs_water; this.substrateIndexs_formate; this.substrateIndexs_hydrogen]) + ...
<a name="l00318"></a>00318                     [-1; 1; 1] * sum(transformations);
<a name="l00319"></a>00319             end
<a name="l00320"></a>00320         end
<a name="l00321"></a>00321     end
<a name="l00322"></a>00322 
<a name="l00323"></a>00323     %get methods of dependent local state
<a name="l00324"></a>00324     methods
<a name="l00325"></a>00325         function value = getDryWeight(this)
<a name="l00326"></a>00326             if size(this.unprocessedMonomers, 3) == 1
<a name="l00327"></a>00327                 value = this.getDryWeight@edu.stanford.covert.cell.sim.Process() + (...
<a name="l00328"></a>00328                     this.monomer.molecularWeights(this.monomer.nascentIndexs)&#39;    * this.unprocessedMonomers + ...
<a name="l00329"></a>00329                     this.monomer.molecularWeights(this.monomer.processedIIndexs)&#39; * this.processedMonomers) ...
<a name="l00330"></a>00330                     / edu.stanford.covert.util.<a class="code" href="class_constant_util.html">ConstantUtil</a>.nAvogadro;
<a name="l00331"></a>00331             else
<a name="l00332"></a>00332                 value = this.getDryWeight@edu.stanford.covert.cell.sim.Process() + (...
<a name="l00333"></a>00333                     permute(this.monomer.molecularWeights(this.monomer.nascentIndexs)&#39;    * permute(this.unprocessedMonomers,[1 3 2]),[1 3 2]) + ...
<a name="l00334"></a>00334                     permute(this.monomer.molecularWeights(this.monomer.processedIIndexs)&#39; * permute(this.processedMonomers,  [1 3 2]),[1 3 2])) ...
<a name="l00335"></a>00335                     / edu.stanford.covert.util.<a class="code" href="class_constant_util.html">ConstantUtil</a>.nAvogadro;
<a name="l00336"></a>00336             end
<a name="l00337"></a>00337         end
<a name="l00338"></a>00338     end
<a name="l00339"></a>00339 end
</pre></div></div><!-- contents -->
</div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
<a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(0)"><span class="SelectionMark">&#160;</span>All</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(1)"><span class="SelectionMark">&#160;</span>Classes</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(2)"><span class="SelectionMark">&#160;</span>Files</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(3)"><span class="SelectionMark">&#160;</span>Functions</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(4)"><span class="SelectionMark">&#160;</span>Variables</a></div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

  <div id="nav-path" class="navpath">
    <ul>
      <li class="navelem"><a class="el" href="_protein_processing_i_8m.html">ProteinProcessingI.m</a>      </li>

    <li class="footer">Generated on Wed Feb 8 2012 21:42:32 for Whole Cell Simulation by
    <a href="http://www.doxygen.org/index.html">
    <img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.7.6.1 </li>
   </ul>
 </div>


</body>
</html>
