<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<title>Whole Cell Simulation: src/+edu/+stanford/+covert/+cell/+sim/+process/ProteinDecay.m Source File</title>

<link href="tabs.css" rel="stylesheet" type="text/css"/>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
<link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="navtree.js"></script>
<script type="text/javascript">
  $(document).ready(initResizable);
</script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
  $(document).ready(function() { searchBox.OnSelectItem(0); });
</script>

</head>
<body>
<div id="top"><!-- do not remove this div! -->


<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  
  
  <td style="padding-left: 0.5em;">
   <div id="projectname">Whole Cell Simulation
   
   </div>
   
  </td>
  
  
  
 </tr>
 </tbody>
</table>
</div>

<!-- Generated by Doxygen 1.7.6.1 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "search",false,'Search');
</script>
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&#160;Page</span></a></li>
      <li><a href="annotated.html"><span>Classes</span></a></li>
      <li class="current"><a href="files.html"><span>Files</span></a></li>
      <li>
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
  <div id="navrow2" class="tabs2">
    <ul class="tablist">
      <li><a href="files.html"><span>File&#160;List</span></a></li>
      <li><a href="globals.html"><span>File&#160;Members</span></a></li>
    </ul>
  </div>
</div>
<div id="side-nav" class="ui-resizable side-nav-resizable">
  <div id="nav-tree">
    <div id="nav-tree-contents">
    </div>
  </div>
  <div id="splitbar" style="-moz-user-select:none;" 
       class="ui-resizable-handle">
  </div>
</div>
<script type="text/javascript">
  initNavTree('_protein_decay_8m.html','');
</script>
<div id="doc-content">
<div class="header">
  <div class="headertitle">
<div class="title">src/+edu/+stanford/+covert/+cell/+sim/+process/ProteinDecay.m</div>  </div>
</div><!--header-->
<div class="contents">
<a href="_protein_decay_8m.html">Go to the documentation of this file.</a><div class="fragment"><pre class="fragment"><a name="l00001"></a>00001 %<a class="code" href="class_protein.html">Protein</a> Decay
<a name="l00002"></a>00002 %
<a name="l00003"></a>00003 % @wholeCellModelID Process_ProteinDecay
<a name="l00004"></a>00004 % @name             <a class="code" href="class_protein.html">Protein</a> Decay
<a name="l00005"></a>00005 % @description
<a name="l00006"></a>00006 %   Biology
<a name="l00007"></a>00007 %   ===========
<a name="l00008"></a>00008 %   Poisson simulation of protein damage, repair, and decay mediated by
<a name="l00009"></a>00009 %   - chaperone: clpB chaperone refolds/disaggregates proteins by an
<a name="l00010"></a>00010 %     energy-dependent mechanism
<a name="l00011"></a>00011 %   - proteases: lon, ftsH; each is believed to processively cleave
<a name="l00012"></a>00012 %     peptides into smaller peptides of approximately 20 amino acids
<a name="l00013"></a>00013 %     - ftsH: cleaves peptides with tmRNA stalled translation proteolysis
<a name="l00014"></a>00014 %       tag; parameterized by
<a name="l00015"></a>00015 %       - fragment length (aa)
<a name="l00016"></a>00016 %       - kinetic rate (cleaves/s)
<a name="l00017"></a>00017 %       - energy per cleavage (ATP/cleavage)
<a name="l00018"></a>00018 %     - lon: cleaves all other peptides; parameterized by
<a name="l00019"></a>00019 %       - fragment length (aa)
<a name="l00020"></a>00020 %       - kinetic rate (cleaves/s)
<a name="l00021"></a>00021 %       - energy per cleavage (ATP/cleavage)
<a name="l00022"></a>00022 %   - several peptidases: the specific functions and kinetics of the
<a name="l00023"></a>00023 %     individual peptidases are unknown and not modelled; decay proceeds as
<a name="l00024"></a>00024 %     <span class="keywordtype">long</span> as at least one of each peptidase is present
<a name="l00025"></a>00025 %
<a name="l00026"></a>00026 % 1. Misfold proteins
<a name="l00027"></a>00027 %    - <a class="code" href="class_protein.html">Protein</a> misfolding is modeled as a poisson process
<a name="l00028"></a>00028 %      with a small rate constant
<a name="l00029"></a>00029 % 2. Refold cytosolic proteins
<a name="l00030"></a>00030 %    - Occurs if clpB is present
<a name="l00031"></a>00031 % 3. Decay macromolecular complexes
<a name="l00032"></a>00032 %    - Model decay is poisson process with rate parameter given by the
<a name="l00033"></a>00033 %      inverse weighted average half life of the complex&#39;s subunits
<a name="l00034"></a>00034 %    - Salvage bound prosthetic groups
<a name="l00035"></a>00035 %    - Mark subunits as damaged to be degraded by either the
<a name="l00036"></a>00036 %      protease/peptidase or ribonuclease machinery
<a name="l00037"></a>00037 % 4. Decay protein monomers and proteolysis tagged polypeptides
<a name="l00038"></a>00038 %    - Model selection of monomers that decay as poisson process with rate
<a name="l00039"></a>00039 %      parameters equal to the decay rates (ln(2)/half life as computed by
<a name="l00040"></a>00040 %      the N-end rule) of the monomers.
<a name="l00041"></a>00041 %    - Execute decay reactions as <span class="keywordtype">long</span> as the following are available
<a name="l00042"></a>00042 %      - Lon/FtsH protease for normal / ribosome-stalled peptides
<a name="l00043"></a>00043 %      - Complete peptidase complement
<a name="l00044"></a>00044 %      - Water for hydrolysis of peptide bonds
<a name="l00045"></a>00045 %
<a name="l00046"></a>00046 %   Knowledge Base
<a name="l00047"></a>00047 %   ===============
<a name="l00048"></a>00048 %   The knowledge base contains implements the N-end rule which predicts the
<a name="l00049"></a>00049 %   half life of every protein monomer based on the average experimentally
<a name="l00050"></a>00050 %   measured (in E. coli) protein half lives for each possible N-terminal amino
<a name="l00051"></a>00051 %   acid. Exceptions:
<a name="l00052"></a>00052 %   - signal sequences are assumed to have 0 half life
<a name="l00053"></a>00053 %   - proteolysis tagged polypeptides are assumed to have 0 half life
<a name="l00054"></a>00054 %
<a name="l00055"></a>00055 %   The knowledge base predicts the half lives of complexes as the weighted mean
<a name="l00056"></a>00056 %   of that of their constituent protein monomers and RNAs.
<a name="l00057"></a>00057 %
<a name="l00058"></a>00058 %   The knowledge base ProteinMonomer and ProteinComplex classes also computes
<a name="l00059"></a>00059 %   hydrolytic degradation reactions of every protein monomer and complex:
<a name="l00060"></a>00060 %   - the (modified) amino acids released by hydrolytic cleavage of every
<a name="l00061"></a>00061 %     protein monomer
<a name="l00062"></a>00062 %   - water required for hydrolytic peptide bond cleavage
<a name="l00063"></a>00063 %   - prosthetic groups released by cleavage
<a name="l00064"></a>00064 %
<a name="l00065"></a>00065 %   The ProteinMonomer class is also used to compute hydrolytic degradation
<a name="l00066"></a>00066 %   reactions for proteolysis tagged polypeptides.
<a name="l00067"></a>00067 %
<a name="l00068"></a>00068 %   Representation
<a name="l00069"></a>00069 %   ===============
<a name="l00070"></a>00070 %   The substrates, enzymes, monomers, RNAs, and complexs properties represent
<a name="l00071"></a>00071 %   the counts of metabolites, proteases and peptidases, protein monomers,
<a name="l00072"></a>00072 %   damaged RNAs, protein complexes. The substrates and enzymes properties have
<a name="l00073"></a>00073 %   only 1 compartment (cytosol). The monomers, RNAs, and complexs properties
<a name="l00074"></a>00074 %   have 5 compartments. The monomers and complexs properties represent all
<a name="l00075"></a>00075 %   forms of proteins (nascent, mature, damaged, folded, misfolded, etc.)
<a name="l00076"></a>00076 %
<a name="l00077"></a>00077 %   abortedSequences represents the sequences of every proteolysis tagged
<a name="l00078"></a>00078 %   monomer in the cytosol.
<a name="l00079"></a>00079 %
<a name="l00080"></a>00080 %   The process contains intermediate representation of protein complex
<a name="l00081"></a>00081 %   degradation, but not of protein monomer degradation, misfolding, or
<a name="l00082"></a>00082 %   refolding. Degrading complexes are broken up into the constituent RNAs and
<a name="l00083"></a>00083 %   monomers, and these components are marked as &quot;damaged&quot; and recognized by RNA
<a name="l00084"></a>00084 %   and protein monomer degradation as molecules with 0 half life. Protein
<a name="l00085"></a>00085 %   monomer degradation is treated as an all-or-nothing event that either
<a name="l00086"></a>00086 %   proceeds to complete with a time step or doesn&#39;t progress at all.
<a name="l00087"></a>00087 %
<a name="l00088"></a>00088 %   proteinMisfoldingRate represents the rate at which every protein misfolds in
<a name="l00089"></a>00089 %   seconds.
<a name="l00090"></a>00090 %
<a name="l00091"></a>00091 %   monomerDecayRates and complexDecayRates represent the decay rate of each
<a name="l00092"></a>00092 %   protein monomer and complex species in seconds. misfolded and damaged
<a name="l00093"></a>00093 %   proteins, signal sequences, and proteolysis tagged polypeptides have 0 half
<a name="l00094"></a>00094 %   lives. All other proteins have positive half lives.
<a name="l00095"></a>00095 %
<a name="l00096"></a>00096 %   monomerDecayReactions and complexDecayReactions represent the metabolites
<a name="l00097"></a>00097 %   required and released by hydrolytic cleavage of protein monomers and the
<a name="l00098"></a>00098 %   break down of macromolecular complexes into their constituent subunits and
<a name="l00099"></a>00099 %   sequestered prosthetic groups. monomerDecayReactions and
<a name="l00100"></a>00100 %   complexDecayReactions are computed by the knowledge base ProteinMonomer and
<a name="l00101"></a>00101 %   ProteinComplex classes.
<a name="l00102"></a>00102 %
<a name="l00103"></a>00103 %   Initialization
<a name="l00104"></a>00104 %   ===============
<a name="l00105"></a>00105 %   All protein monomers and complexs are initialized to the mature state. This
<a name="l00106"></a>00106 %   is accomplished by the simulation class initializeState method.
<a name="l00107"></a>00107 %
<a name="l00108"></a>00108 %   Simulation
<a name="l00109"></a>00109 %   ===============
<a name="l00110"></a>00110 %   This process models misfolding, refolding, and protein monomer and complex
<a name="l00111"></a>00111 %   degradation as an enyzme-dependent (excepct misfolding) poisson processes
<a name="l00112"></a>00112 %   with rate parameter:
<a name="l00113"></a>00113 %      lambda = proteins .* decayRates * stepSizeSec
<a name="l00114"></a>00114 %
<a name="l00115"></a>00115 %   Algorithm
<a name="l00116"></a>00116 %   +++++++++++++++
<a name="l00117"></a>00117 %   Each of misfolding, refolding, protein complex degradation, proteolysis
<a name="l00118"></a>00118 %   tagged monomer degradation, and protein monomer degradation use a variant of
<a name="l00119"></a>00119 %   the algorithm:
<a name="l00120"></a>00120 %
<a name="l00121"></a>00121 %   1. Stochastically select proteins to misfold/refolding/degrade based on
<a name="l00122"></a>00122 %      poission distribution with
<a name="l00123"></a>00123 %         lambda = proteins .* rates * stepSizeSec
<a name="l00124"></a>00124 %   2. Limit refolding/degradation by availability of water and energy
<a name="l00125"></a>00125 %   3. Limit protein refolding/degradation by available enzyme activity
<a name="l00126"></a>00126 %      a. Refolding requires ClpB
<a name="l00127"></a>00127 %      b. Protein monomer degradation requires Lon protease and 6 peptidases
<a name="l00128"></a>00128 %      c. Proteolysis tagged polypeptide degradation requires FtsH protease and
<a name="l00129"></a>00129 %         6 peptidases
<a name="l00130"></a>00130 %   4. Update counts of proteins
<a name="l00131"></a>00131 %   5. Update counts of metabolic reactants and byproducts of protein
<a name="l00132"></a>00132 %      refolding/degradation
<a name="l00133"></a>00133 %
<a name="l00134"></a>00134 %   Compartments
<a name="l00135"></a>00135 %   +++++++++++++++
<a name="l00136"></a>00136 %   - Misfolding: occurs in all compartments
<a name="l00137"></a>00137 %   - Refolding: occurs only in compartments where ClpB is present, that is the
<a name="l00138"></a>00138 %     cytosol (and terminal organelle cytosol)
<a name="l00139"></a>00139 %   - Complex degradation: occurs only in compartments where the protein decay
<a name="l00140"></a>00140 %     machinery (Lon protease and peptidases) are presents, that is the cytosol
<a name="l00141"></a>00141 %     (and terminal organelle cytosol)
<a name="l00142"></a>00142 %   - Proteolysis tagged polypeptide degradation: only exist in cytosol, and are
<a name="l00143"></a>00143 %     only degraded their (assumes they are accessible from within the cytosol
<a name="l00144"></a>00144 %     to the integral membrane FtsH protease)
<a name="l00145"></a>00145 %   - Protein monomer degradation:  occurs only in compartments where the Lon
<a name="l00146"></a>00146 %     protease and peptidases are present, that is the cytosol.
<a name="l00147"></a>00147 %
<a name="l00148"></a>00148 % Author: Jonathan Karr, jkarr@stanford.edu
<a name="l00149"></a>00149 % Affilitation: Covert Lab, Department of Bioengineering, Stanford University
<a name="l00150"></a>00150 % Last updated: 8/4/2010
<a name="l00151"></a>00151 
<a name="l00152"></a>00152 classdef ProteinDecay &lt; edu.stanford.covert.cell.sim.Process
<a name="l00153"></a>00153     %property annotations
<a name="l00154"></a>00154     properties (Constant)
<a name="l00155"></a>00155         optionNames__              = {}; %names of option <a class="code" href="json_parse_8m.html#a6e4256f69b751d3d7ec4ea52ff5a8dee">properties</a>
<a name="l00156"></a>00156         fixedConstantNames__       = {   %names of fixed constant <a class="code" href="json_parse_8m.html#a6e4256f69b751d3d7ec4ea52ff5a8dee">properties</a>
<a name="l00157"></a>00157                     <span class="stringliteral">&#39;lonProteaseSpecificRate&#39;</span>;
<a name="l00158"></a>00158             <span class="stringliteral">&#39;lonProteaseEnergyCost&#39;</span>;
<a name="l00159"></a>00159             <span class="stringliteral">&#39;lonProteaseFragmentLength&#39;</span>;
<a name="l00160"></a>00160             <span class="stringliteral">&#39;ftsHProteaseSpecificRate&#39;</span>;
<a name="l00161"></a>00161             <span class="stringliteral">&#39;ftsHProteaseEnergyCost&#39;</span>;
<a name="l00162"></a>00162             <span class="stringliteral">&#39;ftsHProteaseFragmentLength&#39;</span>;
<a name="l00163"></a>00163             <span class="stringliteral">&#39;oligoendopeptidaseFSpecificRate&#39;</span>;
<a name="l00164"></a>00164             <span class="stringliteral">&#39;proteinMisfoldingRate&#39;</span>;
<a name="l00165"></a>00165             <span class="stringliteral">&#39;proteinComplexRNAComposition&#39;</span>;
<a name="l00166"></a>00166             <span class="stringliteral">&#39;proteinComplexMonomerComposition&#39;</span>;
<a name="l00167"></a>00167             <span class="stringliteral">&#39;monomerLonProteaseCleavages&#39;</span>;
<a name="l00168"></a>00168             <span class="stringliteral">&#39;monomerDecayReactions&#39;</span>;
<a name="l00169"></a>00169             <span class="stringliteral">&#39;complexDecayReactions&#39;</span>;
<a name="l00170"></a>00170                     };
<a name="l00171"></a>00171         fittedConstantNames__      = {}; %names of fitted constant <a class="code" href="json_parse_8m.html#a6e4256f69b751d3d7ec4ea52ff5a8dee">properties</a>
<a name="l00172"></a>00172         localStateNames__          = {   %names of simulation state <a class="code" href="json_parse_8m.html#a6e4256f69b751d3d7ec4ea52ff5a8dee">properties</a> redundant with timecourses in <span class="keyword">this</span> or other processes or the simulation
<a name="l00173"></a>00173             <span class="stringliteral">&#39;monomers&#39;</span>;
<a name="l00174"></a>00174             <span class="stringliteral">&#39;complexs&#39;</span>;
<a name="l00175"></a>00175             <span class="stringliteral">&#39;RNAs&#39;</span>};
<a name="l00176"></a>00176     end
<a name="l00177"></a>00177 
<a name="l00178"></a>00178     %IDs, names, and local indices
<a name="l00179"></a>00179     <a class="code" href="json_parse_8m.html#a6e4256f69b751d3d7ec4ea52ff5a8dee">properties</a>
<a name="l00180"></a>00180         stimuliWholeCellModelIDs = {};   %whole cell model IDs of stimuli
<a name="l00181"></a>00181 
<a name="l00182"></a>00182         substrateWholeCellModelIDs = {}; %whole cell model IDs of substrates
<a name="l00183"></a>00183         substrateIndexs_atp              %index within substrates of ATP
<a name="l00184"></a>00184         substrateIndexs_adp              %index within substrates of ADP
<a name="l00185"></a>00185         substrateIndexs_phosphate        %index within substrates of inorganic phosphate
<a name="l00186"></a>00186         substrateIndexs_hydrogen         %index within substrates of hydrogen
<a name="l00187"></a>00187         substrateIndexs_water            %index within substrates of water
<a name="l00188"></a>00188         substrateIndexs_aminoAcids       %indices within substrates of amino acids
<a name="l00189"></a>00189         substrateIndexs_methionine       %index within substrates of methionine
<a name="l00190"></a>00190         substrateIndexs_fmethionine      %index within substrates of formylmethionine
<a name="l00191"></a>00191         substrateIndexs_glutamate        %index within substrates of glutamate
<a name="l00192"></a>00192         substrateIndexs_glutamine        %index within substrates of glutamine
<a name="l00193"></a>00193         substrateIndexs_formate          %index within substrates of formate
<a name="l00194"></a>00194         substrateIndexs_ammonia          %index within substrates of ammonia
<a name="l00195"></a>00195 
<a name="l00196"></a>00196         enzymeWholeCellModelIDs = { %enzyme whole cell model ids
<a name="l00197"></a>00197             <span class="stringliteral">&#39;MG_355_HEXAMER&#39;</span>;    %protease clpB            cytoplasmic
<a name="l00198"></a>00198             <span class="stringliteral">&#39;MG_239_HEXAMER&#39;</span>;    %protease La              cytoplasmic
<a name="l00199"></a>00199             <span class="stringliteral">&#39;MG_457_HEXAMER&#39;</span>;    %metalloprotease FtsH     integral membrane
<a name="l00200"></a>00200             <span class="stringliteral">&#39;MG_183_MONOMER&#39;</span>;    %oligoendopeptidase F     cytoplasmic
<a name="l00201"></a>00201             <span class="stringliteral">&#39;MG_324_MONOMER&#39;</span>;    %aminopeptidase           cytoplasmic
<a name="l00202"></a>00202             <span class="stringliteral">&#39;MG_391_HEXAMER&#39;</span>;    %cytosol aminopeptidase   cytoplasmic
<a name="l00203"></a>00203             <span class="stringliteral">&#39;MG_208_DIMER&#39;</span>;      %glycoprotease            cytoplasmic
<a name="l00204"></a>00204             <span class="stringliteral">&#39;MG_046_DIMER&#39;</span>;      %metalloendopeptidase     cytoplasmic
<a name="l00205"></a>00205             <span class="stringliteral">&#39;MG_020_MONOMER&#39;</span>};   %proline iminopeptidase   cytoplasmic
<a name="l00206"></a>00206         enzymeIndexs_clpBProtease          = 1;      %index within enzymes of protease clpB
<a name="l00207"></a>00207         enzymeIndexs_lonProtease           = 2;      %index within enzymes of protease La
<a name="l00208"></a>00208         enzymeIndexs_ftsHProtease          = 3;      %index within enzymes of metalloprotease FtsH
<a name="l00209"></a>00209         enzymeIndexs_oligoendopeptidaseF   = 4;      %index within enzymes of oligoendopeptidase F
<a name="l00210"></a>00210         enzymeIndexs_aminopeptidase        = 5;      %index within enzymes of aminopeptidase
<a name="l00211"></a>00211         enzymeIndexs_cytosolAminopeptidase = 6;      %index within enzymes of cytosol aminopeptidase
<a name="l00212"></a>00212         enzymeIndexs_glycoprotease         = 7;      %index within enzymes of glycoprotease
<a name="l00213"></a>00213         enzymeIndexs_metalloendopeptidase  = 8;      %index within enzymes of metalloendopeptidase
<a name="l00214"></a>00214         enzymeIndexs_prolineIminopeptidase = 9;      %index within enzymes of proline iminopeptidase
<a name="l00215"></a>00215         enzymeIndexs_peptidases            = (4:9)<span class="stringliteral">&#39;; %index within enzymes of peptidases</span>
<a name="l00216"></a>00216 <span class="stringliteral">        </span>
<a name="l00217"></a>00217 <span class="stringliteral">        rnaWholeCellModelIDs                        %whole cell model IDs of RNA subunits</span>
<a name="l00218"></a>00218 <span class="stringliteral">        damagedRNAGlobalIndexs                      %indices of damaged rRNA, sRNA within simulation.RNAs</span>
<a name="l00219"></a>00219 <span class="stringliteral">        </span>
<a name="l00220"></a>00220 <span class="stringliteral">        complexIndexs_modified                      %indices of modified protein complexes within this.complexs</span>
<a name="l00221"></a>00221 <span class="stringliteral">        complexIndexs_unmodified                    %indices of unmodified protein complexes within this.complexs</span>
<a name="l00222"></a>00222 <span class="stringliteral">    end</span>
<a name="l00223"></a>00223 <span class="stringliteral"></span>
<a name="l00224"></a>00224 <span class="stringliteral">    %fixed biological constants</span>
<a name="l00225"></a>00225 <span class="stringliteral">    properties</span>
<a name="l00226"></a>00226 <span class="stringliteral">        lonProteaseSpecificRate           %lon protease kinetic rate (1.667) [PUB_0029]</span>
<a name="l00227"></a>00227 <span class="stringliteral">        lonProteaseEnergyCost                 %lon protease energy requirement (6) [PUB_0029]</span>
<a name="l00228"></a>00228 <span class="stringliteral">        lonProteaseFragmentLength         %length of peptide fragments after cleavage by lon protease (20) [PUB_0029]</span>
<a name="l00229"></a>00229 <span class="stringliteral">        ftsHProteaseSpecificRate          %ftsH protease kinetic rate (0.03) [PUB_0031]</span>
<a name="l00230"></a>00230 <span class="stringliteral">        ftsHProteaseEnergyCost            %ftsH protease energy requirement (8.3) [PUB_0031]</span>
<a name="l00231"></a>00231 <span class="stringliteral">        ftsHProteaseFragmentLength        %length of peptide fragments after cleavage by ftsH protease (15) [PUB_0030]</span>
<a name="l00232"></a>00232 <span class="stringliteral">        oligoendopeptidaseFSpecificRate   %oligoendopeptidase F kinetic rate (27.474) [PUB_0035]</span>
<a name="l00233"></a>00233 <span class="stringliteral">        proteinMisfoldingRate             %expected number of misfolds per protein per time step (1e-6)</span>
<a name="l00234"></a>00234 <span class="stringliteral"></span>
<a name="l00235"></a>00235 <span class="stringliteral">        proteinComplexRNAComposition      %RNA subunit composition of protein complexes (RNAs X protein complexes X compartments)</span>
<a name="l00236"></a>00236 <span class="stringliteral">        proteinComplexMonomerComposition  %protein monomer subunit composition of protein complexes (protein monomers X protein complexes X compartments)</span>
<a name="l00237"></a>00237 <span class="stringliteral"></span>
<a name="l00238"></a>00238 <span class="stringliteral">        monomerLonProteaseCleavages       %number of lon protease cleavages required to degrade each protein monomer</span>
<a name="l00239"></a>00239 <span class="stringliteral">        monomerDecayReactions             %substrates consumed and produced by decay of each protein monomer</span>
<a name="l00240"></a>00240 <span class="stringliteral">        complexDecayReactions             %substrates consumed and produced by decay of each protein complex</span>
<a name="l00241"></a>00241 <span class="stringliteral">    end</span>
<a name="l00242"></a>00242 <span class="stringliteral"></span>
<a name="l00243"></a>00243 <span class="stringliteral">    %global state (stored locally for convenience)</span>
<a name="l00244"></a>00244 <span class="stringliteral">    properties</span>
<a name="l00245"></a>00245 <span class="stringliteral">        RNAs                              %counts of RNAs</span>
<a name="l00246"></a>00246 <span class="stringliteral">        monomers                          %counts of protein monomers</span>
<a name="l00247"></a>00247 <span class="stringliteral">        complexs                          %counts of protein complexes</span>
<a name="l00248"></a>00248 <span class="stringliteral">    end</span>
<a name="l00249"></a>00249 <span class="stringliteral"></span>
<a name="l00250"></a>00250 <span class="stringliteral">    %global state (referenced locally for convenience)</span>
<a name="l00251"></a>00251 <span class="stringliteral">    properties</span>
<a name="l00252"></a>00252 <span class="stringliteral">        polypeptide                       %New polypeptides state class</span>
<a name="l00253"></a>00253 <span class="stringliteral">    end</span>
<a name="l00254"></a>00254 <span class="stringliteral">    </span>
<a name="l00255"></a>00255 <span class="stringliteral">    %constructor</span>
<a name="l00256"></a>00256 <span class="stringliteral">    methods</span>
<a name="l00257"></a>00257 <span class="stringliteral">        function this = ProteinDecay(wholeCellModelID, name)</span>
<a name="l00258"></a>00258 <span class="stringliteral">            this = this@edu.stanford.covert.cell.sim.Process(wholeCellModelID, name);</span>
<a name="l00259"></a>00259 <span class="stringliteral">        end</span>
<a name="l00260"></a>00260 <span class="stringliteral">    end</span>
<a name="l00261"></a>00261 <span class="stringliteral"></span>
<a name="l00262"></a>00262 <span class="stringliteral">    %communication between process/simulation</span>
<a name="l00263"></a>00263 <span class="stringliteral">    methods</span>
<a name="l00264"></a>00264 <span class="stringliteral">        %set references to state objects</span>
<a name="l00265"></a>00265 <span class="stringliteral">        function storeObjectReferences(this, simulation)</span>
<a name="l00266"></a>00266 <span class="stringliteral">            this.storeObjectReferences@edu.stanford.covert.cell.sim.Process(simulation);</span>
<a name="l00267"></a>00267 <span class="stringliteral">            </span>
<a name="l00268"></a>00268 <span class="stringliteral">            this.polypeptide = simulation.state(&#39;</span>Polypeptide<span class="stringliteral">&#39;);</span>
<a name="l00269"></a>00269 <span class="stringliteral">            this.states = [this.states; {this.polypeptide}];</span>
<a name="l00270"></a>00270 <span class="stringliteral">        end</span>
<a name="l00271"></a>00271 <span class="stringliteral">        </span>
<a name="l00272"></a>00272 <span class="stringliteral">        %initialize constants</span>
<a name="l00273"></a>00273 <span class="stringliteral">        function initializeConstants(this, knowledgeBase, simulation, varargin)            </span>
<a name="l00274"></a>00274 <span class="stringliteral">            %include all necessary metabolites</span>
<a name="l00275"></a>00275 <span class="stringliteral">            decayReaction_nascentMonomers     = reshape([knowledgeBase.proteinMonomers.decayReaction],                    [], knowledgeBase.numProteinMonomers);</span>
<a name="l00276"></a>00276 <span class="stringliteral">            decayReaction_processedIMonomers  = reshape([knowledgeBase.proteinMonomers.processedISequenceDecayReaction],  [], knowledgeBase.numProteinMonomers);</span>
<a name="l00277"></a>00277 <span class="stringliteral">            decayReaction_processedIIMonomers = reshape([knowledgeBase.proteinMonomers.processedIISequenceDecayReaction], [], knowledgeBase.numProteinMonomers);</span>
<a name="l00278"></a>00278 <span class="stringliteral">            decayReaction_signalSequence      = reshape([knowledgeBase.proteinMonomers.signalSequenceDecayReaction],      [], knowledgeBase.numProteinMonomers);</span>
<a name="l00279"></a>00279 <span class="stringliteral">            decayReaction_foldedMonomers      = reshape([knowledgeBase.proteinMonomers.foldedSequenceDecayReaction],      [], knowledgeBase.numProteinMonomers);</span>
<a name="l00280"></a>00280 <span class="stringliteral">            decayReaction_matureMonomers      = reshape([knowledgeBase.proteinMonomers.matureSequenceDecayReaction],      [], knowledgeBase.numProteinMonomers);</span>
<a name="l00281"></a>00281 <span class="stringliteral">            decayReaction_nascentComplexs     = reshape([knowledgeBase.proteinComplexs.decayReaction],                    [], knowledgeBase.numProteinComplexs);</span>
<a name="l00282"></a>00282 <span class="stringliteral">            decayReaction_matureComplexs      = reshape([knowledgeBase.proteinComplexs.matureDecayReaction],              [], knowledgeBase.numProteinComplexs);</span>
<a name="l00283"></a>00283 <span class="stringliteral">            </span>
<a name="l00284"></a>00284 <span class="stringliteral">            m = this.metabolite;</span>
<a name="l00285"></a>00285 <span class="stringliteral">            gtpIdxs = m.getIndexs({&#39;</span>GTP<span class="stringliteral">&#39;; &#39;</span>H2O<span class="stringliteral">&#39;; &#39;</span>PI<span class="stringliteral">&#39;; &#39;</span>H<span class="stringliteral">&#39;});</span>
<a name="l00286"></a>00286 <span class="stringliteral">            gdpIdxs = m.getIndexs(&#39;</span>GDP<span class="stringliteral">&#39;);</span>
<a name="l00287"></a>00287 <span class="stringliteral">            for i = 1:size(decayReaction_nascentComplexs, 2)</span>
<a name="l00288"></a>00288 <span class="stringliteral">                tmp = decayReaction_nascentComplexs(gtpIdxs, i) ./ [1; 1; -1; -1];</span>
<a name="l00289"></a>00289 <span class="stringliteral">                if tmp(1) &amp;&amp; all(tmp == tmp(1))</span>
<a name="l00290"></a>00290 <span class="stringliteral">                    decayReaction_nascentComplexs(gdpIdxs, i) = tmp(1);</span>
<a name="l00291"></a>00291 <span class="stringliteral">                    decayReaction_nascentComplexs(gtpIdxs, i) = 0;</span>
<a name="l00292"></a>00292 <span class="stringliteral">                end</span>
<a name="l00293"></a>00293 <span class="stringliteral">                </span>
<a name="l00294"></a>00294 <span class="stringliteral">                tmp = decayReaction_matureComplexs(gtpIdxs, i) ./ [1; 1 ;-1; -1];</span>
<a name="l00295"></a>00295 <span class="stringliteral">                if tmp(1) &amp;&amp; all(tmp == tmp(1))</span>
<a name="l00296"></a>00296 <span class="stringliteral">                    decayReaction_matureComplexs(gdpIdxs, i) = tmp(1);</span>
<a name="l00297"></a>00297 <span class="stringliteral">                    decayReaction_matureComplexs(gtpIdxs, i) = 0;</span>
<a name="l00298"></a>00298 <span class="stringliteral">                end</span>
<a name="l00299"></a>00299 <span class="stringliteral">            end</span>
<a name="l00300"></a>00300 <span class="stringliteral">            </span>
<a name="l00301"></a>00301 <span class="stringliteral">            this.substrateWholeCellModelIDs = unique([</span>
<a name="l00302"></a>00302 <span class="stringliteral">                this.metabolite.wholeCellModelIDs(...</span>
<a name="l00303"></a>00303 <span class="stringliteral">                any(decayReaction_nascentMonomers,     2) | ...</span>
<a name="l00304"></a>00304 <span class="stringliteral">                any(decayReaction_processedIMonomers,  2) | ...</span>
<a name="l00305"></a>00305 <span class="stringliteral">                any(decayReaction_processedIIMonomers, 2) | ...</span>
<a name="l00306"></a>00306 <span class="stringliteral">                any(decayReaction_signalSequence,      2) | ...</span>
<a name="l00307"></a>00307 <span class="stringliteral">                any(decayReaction_foldedMonomers,      2) | ...</span>
<a name="l00308"></a>00308 <span class="stringliteral">                any(decayReaction_matureMonomers,      2) | ...</span>
<a name="l00309"></a>00309 <span class="stringliteral">                any(decayReaction_nascentComplexs,     2) | ...</span>
<a name="l00310"></a>00310 <span class="stringliteral">                any(decayReaction_matureComplexs,      2));</span>
<a name="l00311"></a>00311 <span class="stringliteral">                &#39;</span>ATP<span class="stringliteral">&#39;; &#39;</span>ADP<span class="stringliteral">&#39;; &#39;</span>PI<span class="stringliteral">&#39;; &#39;</span>H<span class="stringliteral">&#39;; &#39;</span>H2O<span class="stringliteral">&#39;; &#39;</span>NH3<span class="stringliteral">&#39;; &#39;</span>FOR<span class="stringliteral">&#39;]);</span>
<a name="l00312"></a>00312 <span class="stringliteral">            this.substrateIndexs_atp         = this.substrateIndexs({&#39;</span>ATP<span class="stringliteral">&#39;});</span>
<a name="l00313"></a>00313 <span class="stringliteral">            this.substrateIndexs_adp         = this.substrateIndexs({&#39;</span>ADP<span class="stringliteral">&#39;});</span>
<a name="l00314"></a>00314 <span class="stringliteral">            this.substrateIndexs_phosphate   = this.substrateIndexs({&#39;</span>PI<span class="stringliteral">&#39;});</span>
<a name="l00315"></a>00315 <span class="stringliteral">            this.substrateIndexs_hydrogen    = this.substrateIndexs({&#39;</span>H<span class="stringliteral">&#39;});</span>
<a name="l00316"></a>00316 <span class="stringliteral">            this.substrateIndexs_water       = this.substrateIndexs({&#39;</span>H2O<span class="stringliteral">&#39;});</span>
<a name="l00317"></a>00317 <span class="stringliteral">            this.substrateIndexs_aminoAcids  = this.substrateIndexs({&#39;</span>ALA<span class="stringliteral">&#39;; &#39;</span>ARG<span class="stringliteral">&#39;; &#39;</span>ASN<span class="stringliteral">&#39;; &#39;</span>ASP<span class="stringliteral">&#39;; &#39;</span>CYS<span class="stringliteral">&#39;; &#39;</span>GLN<span class="stringliteral">&#39;; &#39;</span>GLU<span class="stringliteral">&#39;; &#39;</span>GLY<span class="stringliteral">&#39;; &#39;</span>HIS<span class="stringliteral">&#39;; &#39;</span>ILE<span class="stringliteral">&#39;; &#39;</span>LEU<span class="stringliteral">&#39;; &#39;</span>LYS<span class="stringliteral">&#39;; &#39;</span>MET<span class="stringliteral">&#39;; &#39;</span>PHE<span class="stringliteral">&#39;; &#39;</span>PRO<span class="stringliteral">&#39;; &#39;</span>SER<span class="stringliteral">&#39;; &#39;</span>THR<span class="stringliteral">&#39;; &#39;</span>TRP<span class="stringliteral">&#39;; &#39;</span>TYR<span class="stringliteral">&#39;; &#39;</span>VAL<span class="stringliteral">&#39;; &#39;</span>FMET<span class="stringliteral">&#39;});</span>
<a name="l00318"></a>00318 <span class="stringliteral">            this.substrateIndexs_methionine  = this.substrateIndexs({&#39;</span>MET<span class="stringliteral">&#39;});</span>
<a name="l00319"></a>00319 <span class="stringliteral">            this.substrateIndexs_fmethionine = this.substrateIndexs({&#39;</span>FMET<span class="stringliteral">&#39;});</span>
<a name="l00320"></a>00320 <span class="stringliteral">            this.substrateIndexs_glutamate   = this.substrateIndexs({&#39;</span>GLU<span class="stringliteral">&#39;});</span>
<a name="l00321"></a>00321 <span class="stringliteral">            this.substrateIndexs_glutamine   = this.substrateIndexs({&#39;</span>GLN<span class="stringliteral">&#39;});</span>
<a name="l00322"></a>00322 <span class="stringliteral">            this.substrateIndexs_ammonia     = this.substrateIndexs({&#39;</span>NH3<span class="stringliteral">&#39;});</span>
<a name="l00323"></a>00323 <span class="stringliteral">            this.substrateIndexs_formate     = this.substrateIndexs({&#39;</span>FOR<span class="stringliteral">&#39;});</span>
<a name="l00324"></a>00324 <span class="stringliteral"></span>
<a name="l00325"></a>00325 <span class="stringliteral">            %call super class method</span>
<a name="l00326"></a>00326 <span class="stringliteral">            this.initializeConstants@edu.stanford.covert.cell.sim.Process(...</span>
<a name="l00327"></a>00327 <span class="stringliteral">                knowledgeBase, simulation, varargin{:});</span>
<a name="l00328"></a>00328 <span class="stringliteral"></span>
<a name="l00329"></a>00329 <span class="stringliteral">            %RNA, monomer, complex whole cell model IDs</span>
<a name="l00330"></a>00330 <span class="stringliteral">            this.rnaWholeCellModelIDs = this.rna.wholeCellModelIDs(this.rna.damagedIndexs([this.rna.matureRRNAIndexs; this.rna.matureSRNAIndexs]));</span>
<a name="l00331"></a>00331 <span class="stringliteral">            </span>
<a name="l00332"></a>00332 <span class="stringliteral">            %RNA indices</span>
<a name="l00333"></a>00333 <span class="stringliteral">            this.damagedRNAGlobalIndexs = this.rna.damagedIndexs([this.rna.matureRRNAIndexs; this.rna.matureSRNAIndexs]);</span>
<a name="l00334"></a>00334 <span class="stringliteral"></span>
<a name="l00335"></a>00335 <span class="stringliteral">            %complex composition</span>
<a name="l00336"></a>00336 <span class="stringliteral">            cplxRNAComp     = knowledgeBase.proteinComplexAllRNAComposition([this.gene.rRNAIndexs; this.gene.sRNAIndexs], :, this.compartment.cytosolIndexs);</span>
<a name="l00337"></a>00337 <span class="stringliteral">            cplxMonomerComp = knowledgeBase.proteinComplexAllMonomerComposition(:, :, this.compartment.cytosolIndexs);</span>
<a name="l00338"></a>00338 <span class="stringliteral"></span>
<a name="l00339"></a>00339 <span class="stringliteral">            this.proteinComplexRNAComposition = zeros(size(cplxRNAComp,1), numel(this.complex.wholeCellModelIDs));</span>
<a name="l00340"></a>00340 <span class="stringliteral">            this.proteinComplexRNAComposition(:, this.complex.nascentIndexs)     = cplxRNAComp;</span>
<a name="l00341"></a>00341 <span class="stringliteral">            this.proteinComplexRNAComposition(:, this.complex.matureIndexs)      = cplxRNAComp;</span>
<a name="l00342"></a>00342 <span class="stringliteral">            this.proteinComplexRNAComposition(:, this.complex.inactivatedIndexs) = cplxRNAComp;</span>
<a name="l00343"></a>00343 <span class="stringliteral">            this.proteinComplexRNAComposition(:, this.complex.boundIndexs)       = cplxRNAComp;</span>
<a name="l00344"></a>00344 <span class="stringliteral">            this.proteinComplexRNAComposition(:, this.complex.misfoldedIndexs)   = cplxRNAComp;</span>
<a name="l00345"></a>00345 <span class="stringliteral">            this.proteinComplexRNAComposition(:, this.complex.damagedIndexs)     = cplxRNAComp;</span>
<a name="l00346"></a>00346 <span class="stringliteral"></span>
<a name="l00347"></a>00347 <span class="stringliteral">            this.proteinComplexMonomerComposition = zeros(size(cplxMonomerComp,1), numel(this.complex.wholeCellModelIDs));</span>
<a name="l00348"></a>00348 <span class="stringliteral">            this.proteinComplexMonomerComposition(:, this.complex.nascentIndexs)     = cplxMonomerComp;</span>
<a name="l00349"></a>00349 <span class="stringliteral">            this.proteinComplexMonomerComposition(:, this.complex.matureIndexs)      = cplxMonomerComp;</span>
<a name="l00350"></a>00350 <span class="stringliteral">            this.proteinComplexMonomerComposition(:, this.complex.inactivatedIndexs) = cplxMonomerComp;</span>
<a name="l00351"></a>00351 <span class="stringliteral">            this.proteinComplexMonomerComposition(:, this.complex.boundIndexs)       = cplxMonomerComp;</span>
<a name="l00352"></a>00352 <span class="stringliteral">            this.proteinComplexMonomerComposition(:, this.complex.misfoldedIndexs)   = cplxMonomerComp;</span>
<a name="l00353"></a>00353 <span class="stringliteral">            this.proteinComplexMonomerComposition(:, this.complex.damagedIndexs)     = cplxMonomerComp;</span>
<a name="l00354"></a>00354 <span class="stringliteral"></span>
<a name="l00355"></a>00355 <span class="stringliteral">            %decay reactions           </span>
<a name="l00356"></a>00356 <span class="stringliteral">            this.monomerDecayReactions = zeros(numel(this.substrateWholeCellModelIDs), numel(this.monomer.wholeCellModelIDs));</span>
<a name="l00357"></a>00357 <span class="stringliteral">            this.complexDecayReactions = zeros(numel(this.substrateWholeCellModelIDs), numel(this.complex.wholeCellModelIDs));</span>
<a name="l00358"></a>00358 <span class="stringliteral">            </span>
<a name="l00359"></a>00359 <span class="stringliteral">            this.monomerDecayReactions(:, this.monomer.nascentIndexs)        = decayReaction_nascentMonomers(    this.substrateMetaboliteGlobalIndexs, :);</span>
<a name="l00360"></a>00360 <span class="stringliteral">            this.monomerDecayReactions(:, this.monomer.processedIIndexs)     = decayReaction_processedIMonomers( this.substrateMetaboliteGlobalIndexs, :);</span>
<a name="l00361"></a>00361 <span class="stringliteral">            this.monomerDecayReactions(:, this.monomer.processedIIIndexs)    = decayReaction_processedIIMonomers(this.substrateMetaboliteGlobalIndexs, :);</span>
<a name="l00362"></a>00362 <span class="stringliteral">            this.monomerDecayReactions(:, this.monomer.signalSequenceIndexs) = decayReaction_signalSequence(     this.substrateMetaboliteGlobalIndexs, :);</span>
<a name="l00363"></a>00363 <span class="stringliteral">            this.monomerDecayReactions(:, this.monomer.foldedIndexs)         = decayReaction_foldedMonomers(     this.substrateMetaboliteGlobalIndexs, :);</span>
<a name="l00364"></a>00364 <span class="stringliteral">            this.monomerDecayReactions(:, this.monomer.matureIndexs)         = decayReaction_matureMonomers(     this.substrateMetaboliteGlobalIndexs, :);</span>
<a name="l00365"></a>00365 <span class="stringliteral">            this.monomerDecayReactions(:, this.monomer.inactivatedIndexs)    = decayReaction_matureMonomers(     this.substrateMetaboliteGlobalIndexs, :);</span>
<a name="l00366"></a>00366 <span class="stringliteral">            this.monomerDecayReactions(:, this.monomer.boundIndexs)          = decayReaction_matureMonomers(     this.substrateMetaboliteGlobalIndexs, :);</span>
<a name="l00367"></a>00367 <span class="stringliteral">            this.monomerDecayReactions(:, this.monomer.misfoldedIndexs)      = decayReaction_matureMonomers(     this.substrateMetaboliteGlobalIndexs, :);</span>
<a name="l00368"></a>00368 <span class="stringliteral">            this.monomerDecayReactions(:, this.monomer.damagedIndexs)        = decayReaction_matureMonomers(     this.substrateMetaboliteGlobalIndexs, :);</span>
<a name="l00369"></a>00369 <span class="stringliteral">            </span>
<a name="l00370"></a>00370 <span class="stringliteral">            this.complexDecayReactions(:, this.complex.nascentIndexs)        = decayReaction_nascentComplexs(    this.substrateMetaboliteGlobalIndexs, :);</span>
<a name="l00371"></a>00371 <span class="stringliteral">            this.complexDecayReactions(:, this.complex.matureIndexs)         = decayReaction_matureComplexs(     this.substrateMetaboliteGlobalIndexs, :);</span>
<a name="l00372"></a>00372 <span class="stringliteral">            this.complexDecayReactions(:, this.complex.inactivatedIndexs)    = decayReaction_matureComplexs(     this.substrateMetaboliteGlobalIndexs, :);</span>
<a name="l00373"></a>00373 <span class="stringliteral">            this.complexDecayReactions(:, this.complex.boundIndexs)          = decayReaction_matureComplexs(     this.substrateMetaboliteGlobalIndexs, :);</span>
<a name="l00374"></a>00374 <span class="stringliteral">            this.complexDecayReactions(:, this.complex.misfoldedIndexs)      = decayReaction_matureComplexs(     this.substrateMetaboliteGlobalIndexs, :);</span>
<a name="l00375"></a>00375 <span class="stringliteral">            this.complexDecayReactions(:, this.complex.damagedIndexs)        = decayReaction_matureComplexs(     this.substrateMetaboliteGlobalIndexs, :);</span>
<a name="l00376"></a>00376 <span class="stringliteral">            </span>
<a name="l00377"></a>00377 <span class="stringliteral">            this.complexIndexs_modified = find(any(this.complexDecayReactions &gt; 0, 1)&#39;</span>);
<a name="l00378"></a>00378             this.complexIndexs_unmodified = setdiff((1:numel(this.complex.wholeCellModelIDs))<span class="stringliteral">&#39;, this.complexIndexs_modified);</span>
<a name="l00379"></a>00379 <span class="stringliteral">            </span>
<a name="l00380"></a>00380 <span class="stringliteral">            this.monomerLonProteaseCleavages = max(0, ceil(this.monomer.lengths/this.lonProteaseFragmentLength)-1);</span>
<a name="l00381"></a>00381 <span class="stringliteral">        end</span>
<a name="l00382"></a>00382 <span class="stringliteral"></span>
<a name="l00383"></a>00383 <span class="stringliteral">        %retrieve state from simulation</span>
<a name="l00384"></a>00384 <span class="stringliteral">        function copyFromState(this)</span>
<a name="l00385"></a>00385 <span class="stringliteral">            this.copyFromState@edu.stanford.covert.cell.sim.Process();</span>
<a name="l00386"></a>00386 <span class="stringliteral">            </span>
<a name="l00387"></a>00387 <span class="stringliteral">            this.monomers = this.monomer.counts;</span>
<a name="l00388"></a>00388 <span class="stringliteral">            this.complexs = this.complex.counts;</span>
<a name="l00389"></a>00389 <span class="stringliteral">            this.RNAs = this.rna.counts(this.damagedRNAGlobalIndexs, :, :);</span>
<a name="l00390"></a>00390 <span class="stringliteral">        end</span>
<a name="l00391"></a>00391 <span class="stringliteral"></span>
<a name="l00392"></a>00392 <span class="stringliteral">        %send state to simulation</span>
<a name="l00393"></a>00393 <span class="stringliteral">        function copyToState(this)</span>
<a name="l00394"></a>00394 <span class="stringliteral">            this.copyToState@edu.stanford.covert.cell.sim.Process();</span>
<a name="l00395"></a>00395 <span class="stringliteral"></span>
<a name="l00396"></a>00396 <span class="stringliteral">            this.monomer.counts = this.monomers;</span>
<a name="l00397"></a>00397 <span class="stringliteral">            this.complex.counts = this.complexs;</span>
<a name="l00398"></a>00398 <span class="stringliteral">            this.rna.counts(this.damagedRNAGlobalIndexs, :, :) = this.RNAs;</span>
<a name="l00399"></a>00399 <span class="stringliteral">        end</span>
<a name="l00400"></a>00400 <span class="stringliteral">    end</span>
<a name="l00401"></a>00401 <span class="stringliteral"></span>
<a name="l00402"></a>00402 <span class="stringliteral">    %allocate memory for state</span>
<a name="l00403"></a>00403 <span class="stringliteral">    methods</span>
<a name="l00404"></a>00404 <span class="stringliteral">        function allocateMemoryForState(this, numTimePoints)</span>
<a name="l00405"></a>00405 <span class="stringliteral">            this.allocateMemoryForState@edu.stanford.covert.cell.sim.Process(numTimePoints);</span>
<a name="l00406"></a>00406 <span class="stringliteral"></span>
<a name="l00407"></a>00407 <span class="stringliteral">            numCompartments = this.compartment.count;</span>
<a name="l00408"></a>00408 <span class="stringliteral"></span>
<a name="l00409"></a>00409 <span class="stringliteral">            this.RNAs     = zeros(size(this.rnaWholeCellModelIDs,1),      numCompartments, numTimePoints);</span>
<a name="l00410"></a>00410 <span class="stringliteral">            this.monomers = zeros(size(this.monomer.wholeCellModelIDs,1), numCompartments, numTimePoints);</span>
<a name="l00411"></a>00411 <span class="stringliteral">            this.complexs = zeros(size(this.complex.wholeCellModelIDs,1), numCompartments, numTimePoints);</span>
<a name="l00412"></a>00412 <span class="stringliteral">        end</span>
<a name="l00413"></a>00413 <span class="stringliteral">    end</span>
<a name="l00414"></a>00414 <span class="stringliteral"></span>
<a name="l00415"></a>00415 <span class="stringliteral">    %model</span>
<a name="l00416"></a>00416 <span class="stringliteral">    methods</span>
<a name="l00417"></a>00417 <span class="stringliteral">        %Calculate</span>
<a name="l00418"></a>00418 <span class="stringliteral">        %- contribution to FBA objective</span>
<a name="l00419"></a>00419 <span class="stringliteral">        %- minimum expression consistent with cell cycle length</span>
<a name="l00420"></a>00420 <span class="stringliteral">        function [bmProd, byProd, minEnzExp, maxEnzExp] = calcResourceRequirements_LifeCycle(this, ~, states)</span>
<a name="l00421"></a>00421 <span class="stringliteral">            %% initialize</span>
<a name="l00422"></a>00422 <span class="stringliteral">            bmProd = zeros(size(this.substrateWholeCellModelIDs));</span>
<a name="l00423"></a>00423 <span class="stringliteral">            byProd = zeros(size(this.substrateWholeCellModelIDs));</span>
<a name="l00424"></a>00424 <span class="stringliteral">            minEnzExp = zeros(size(this.enzymeWholeCellModelIDs));</span>
<a name="l00425"></a>00425 <span class="stringliteral">            maxEnzExp = Inf(size(this.enzymeWholeCellModelIDs));</span>
<a name="l00426"></a>00426 <span class="stringliteral">            </span>
<a name="l00427"></a>00427 <span class="stringliteral">            %% substrate and byproducts</span>
<a name="l00428"></a>00428 <span class="stringliteral">            %refolding</span>
<a name="l00429"></a>00429 <span class="stringliteral">            nRefolding = (sum(states.monomers) + sum(states.complexs)) * this.proteinMisfoldingRate;</span>
<a name="l00430"></a>00430 <span class="stringliteral">            </span>
<a name="l00431"></a>00431 <span class="stringliteral">            bmProd(this.substrateIndexs_atp)       = bmProd(this.substrateIndexs_atp)       + nRefolding;</span>
<a name="l00432"></a>00432 <span class="stringliteral">            bmProd(this.substrateIndexs_water)     = bmProd(this.substrateIndexs_water)     + nRefolding;</span>
<a name="l00433"></a>00433 <span class="stringliteral">            byProd(this.substrateIndexs_adp)       = byProd(this.substrateIndexs_adp)       + nRefolding;</span>
<a name="l00434"></a>00434 <span class="stringliteral">            byProd(this.substrateIndexs_phosphate) = byProd(this.substrateIndexs_phosphate) + nRefolding;</span>
<a name="l00435"></a>00435 <span class="stringliteral">            byProd(this.substrateIndexs_hydrogen)  = byProd(this.substrateIndexs_hydrogen)  + nRefolding;</span>
<a name="l00436"></a>00436 <span class="stringliteral">            </span>
<a name="l00437"></a>00437 <span class="stringliteral">            %complex degradation</span>
<a name="l00438"></a>00438 <span class="stringliteral">            bmProd = bmProd + max(0, -this.complexDecayReactions(:, this.complex.matureIndexs)) * states.complexDecays;</span>
<a name="l00439"></a>00439 <span class="stringliteral">            byProd = byProd + max(0,  this.complexDecayReactions(:, this.complex.matureIndexs)) * states.complexDecays;</span>
<a name="l00440"></a>00440 <span class="stringliteral">            </span>
<a name="l00441"></a>00441 <span class="stringliteral">            %mature monomer degradation</span>
<a name="l00442"></a>00442 <span class="stringliteral">            bmProd = bmProd + max(0, -this.monomerDecayReactions(:, this.monomer.matureIndexs)) * states.monomerDecays;</span>
<a name="l00443"></a>00443 <span class="stringliteral">            byProd = byProd + max(0,  this.monomerDecayReactions(:, this.monomer.matureIndexs)) * states.monomerDecays;</span>
<a name="l00444"></a>00444 <span class="stringliteral">            </span>
<a name="l00445"></a>00445 <span class="stringliteral">            lonCleavages = this.lonProteaseEnergyCost * states.monomerDecays&#39;</span> * <span class="keyword">this</span>.monomerLonProteaseCleavages(<span class="keyword">this</span>.monomer.matureIndexs);
<a name="l00446"></a>00446             bmProd(this.substrateIndexs_atp)       = bmProd(this.substrateIndexs_atp)       + lonCleavages;
<a name="l00447"></a>00447             bmProd(this.substrateIndexs_water)     = bmProd(this.substrateIndexs_water)     + lonCleavages;
<a name="l00448"></a>00448             byProd(this.substrateIndexs_adp)       = byProd(this.substrateIndexs_adp)       + lonCleavages;
<a name="l00449"></a>00449             byProd(this.substrateIndexs_phosphate) = byProd(this.substrateIndexs_phosphate) + lonCleavages;
<a name="l00450"></a>00450             byProd(this.substrateIndexs_hydrogen)  = byProd(this.substrateIndexs_hydrogen)  + lonCleavages;
<a name="l00451"></a>00451             
<a name="l00452"></a>00452             %signal sequence degradation
<a name="l00453"></a>00453             bmProd = bmProd + max(0, -this.monomerDecayReactions(:, this.monomer.signalSequenceIndexs)) * states.monomerProductions;
<a name="l00454"></a>00454             byProd = byProd + max(0,  this.monomerDecayReactions(:, this.monomer.signalSequenceIndexs)) * states.monomerProductions;
<a name="l00455"></a>00455             
<a name="l00456"></a>00456             lonCleavages = this.lonProteaseEnergyCost * states.monomerProductions<span class="stringliteral">&#39; * ...</span>
<a name="l00457"></a>00457 <span class="stringliteral">                (this.monomerLonProteaseCleavages(this.monomer.signalSequenceIndexs) .* ...</span>
<a name="l00458"></a>00458 <span class="stringliteral">                any(this.monomerDecayReactions(:, this.monomer.signalSequenceIndexs), 1)&#39;</span>);
<a name="l00459"></a>00459             bmProd(this.substrateIndexs_atp)       = bmProd(this.substrateIndexs_atp)       + lonCleavages;
<a name="l00460"></a>00460             bmProd(this.substrateIndexs_water)     = bmProd(this.substrateIndexs_water)     + lonCleavages;
<a name="l00461"></a>00461             byProd(this.substrateIndexs_adp)       = byProd(this.substrateIndexs_adp)       + lonCleavages;
<a name="l00462"></a>00462             byProd(this.substrateIndexs_phosphate) = byProd(this.substrateIndexs_phosphate) + lonCleavages;
<a name="l00463"></a>00463             byProd(this.substrateIndexs_hydrogen)  = byProd(this.substrateIndexs_hydrogen)  + lonCleavages;
<a name="l00464"></a>00464             
<a name="l00465"></a>00465             %stalled polypeptide degradation
<a name="l00466"></a>00466             ftsHCleavages = this.ftsHProteaseEnergyCost * 0;
<a name="l00467"></a>00467             bmProd(this.substrateIndexs_atp)       = bmProd(this.substrateIndexs_atp)       + ftsHCleavages;
<a name="l00468"></a>00468             bmProd(this.substrateIndexs_water)     = bmProd(this.substrateIndexs_water)     + ftsHCleavages;
<a name="l00469"></a>00469             byProd(this.substrateIndexs_adp)       = byProd(this.substrateIndexs_adp)       + ftsHCleavages;
<a name="l00470"></a>00470             byProd(this.substrateIndexs_phosphate) = byProd(this.substrateIndexs_phosphate) + ftsHCleavages;
<a name="l00471"></a>00471             byProd(this.substrateIndexs_hydrogen)  = byProd(this.substrateIndexs_hydrogen)  + ftsHCleavages;
<a name="l00472"></a>00472             
<a name="l00473"></a>00473             %% enzymes
<a name="l00474"></a>00474             minEnzExp(this.enzymeIndexs_lonProtease) = ...
<a name="l00475"></a>00475                 2 * states.monomerDecays0<span class="stringliteral">&#39; * this.monomerLonProteaseCleavages(this.monomer.matureIndexs) / ...</span>
<a name="l00476"></a>00476 <span class="stringliteral">                this.lonProteaseSpecificRate;</span>
<a name="l00477"></a>00477 <span class="stringliteral">            minEnzExp(this.enzymeIndexs_clpBProtease) = 2;</span>
<a name="l00478"></a>00478 <span class="stringliteral">            minEnzExp(this.enzymeIndexs_ftsHProtease) = 70; %to keep pace with degrading aborted polypeptides</span>
<a name="l00479"></a>00479 <span class="stringliteral">            minEnzExp(this.enzymeIndexs_peptidases)   = 2;</span>
<a name="l00480"></a>00480 <span class="stringliteral">        end</span>
<a name="l00481"></a>00481 <span class="stringliteral"></span>
<a name="l00482"></a>00482 <span class="stringliteral">        %initialization: proteins initialized to mature/bound/inactivated state</span>
<a name="l00483"></a>00483 <span class="stringliteral">        %by simulation initializeState method</span>
<a name="l00484"></a>00484 <span class="stringliteral">        function initializeState(~)</span>
<a name="l00485"></a>00485 <span class="stringliteral">        end</span>
<a name="l00486"></a>00486 <span class="stringliteral">        </span>
<a name="l00487"></a>00487 <span class="stringliteral">        %resource requirements</span>
<a name="l00488"></a>00488 <span class="stringliteral">        function result = calcResourceRequirements_Current(this)</span>
<a name="l00489"></a>00489 <span class="stringliteral">            result = zeros(size(this.substrates));</span>
<a name="l00490"></a>00490 <span class="stringliteral">            </span>
<a name="l00491"></a>00491 <span class="stringliteral">            mnmers = sum(this.monomers(:, [this.compartment.cytosolIndexs this.compartment.terminalOrganelleCytosolIndexs]), 2);</span>
<a name="l00492"></a>00492 <span class="stringliteral">            cmplxs = sum(this.complexs(:, [this.compartment.cytosolIndexs this.compartment.terminalOrganelleCytosolIndexs]), 2);</span>
<a name="l00493"></a>00493 <span class="stringliteral">            </span>
<a name="l00494"></a>00494 <span class="stringliteral">            %refolding</span>
<a name="l00495"></a>00495 <span class="stringliteral">            nRefolding = (this.enzymes(this.enzymeIndexs_clpBProtease) &gt; 0) * ( ...</span>
<a name="l00496"></a>00496 <span class="stringliteral">                sum(mnmers(this.monomer.misfoldedIndexs)) + ...</span>
<a name="l00497"></a>00497 <span class="stringliteral">                sum(cmplxs(this.complex.misfoldedIndexs)));</span>
<a name="l00498"></a>00498 <span class="stringliteral">            </span>
<a name="l00499"></a>00499 <span class="stringliteral">            if nRefolding &gt; 0 &amp;&amp; this.enzymes(this.enzymeIndexs_clpBProtease)</span>
<a name="l00500"></a>00500 <span class="stringliteral">                result(this.substrateIndexs_atp)   = result(this.substrateIndexs_atp)   + nRefolding;</span>
<a name="l00501"></a>00501 <span class="stringliteral">                result(this.substrateIndexs_water) = result(this.substrateIndexs_water) + nRefolding;</span>
<a name="l00502"></a>00502 <span class="stringliteral">            end</span>
<a name="l00503"></a>00503 <span class="stringliteral">            </span>
<a name="l00504"></a>00504 <span class="stringliteral">            %complex degradation</span>
<a name="l00505"></a>00505 <span class="stringliteral">            cmplxDecays = cmplxs .* min(1, this.complex.decayRates);</span>
<a name="l00506"></a>00506 <span class="stringliteral">            result = result + max(0, -this.complexDecayReactions) * cmplxDecays;</span>
<a name="l00507"></a>00507 <span class="stringliteral">            </span>
<a name="l00508"></a>00508 <span class="stringliteral">            %monomer degradation</span>
<a name="l00509"></a>00509 <span class="stringliteral">            mnmerDecays = mnmers .* min(1, this.monomer.decayRates);</span>
<a name="l00510"></a>00510 <span class="stringliteral">            result = result + max(0, -this.monomerDecayReactions) * mnmerDecays;</span>
<a name="l00511"></a>00511 <span class="stringliteral">            result([this.substrateIndexs_atp; this.substrateIndexs_water]) = ...</span>
<a name="l00512"></a>00512 <span class="stringliteral">                + result([this.substrateIndexs_atp; this.substrateIndexs_water]) ...</span>
<a name="l00513"></a>00513 <span class="stringliteral">                + all(this.enzymes(this.enzymeIndexs_peptidases)) * this.lonProteaseEnergyCost * min(...</span>
<a name="l00514"></a>00514 <span class="stringliteral">                    this.enzymes(this.enzymeIndexs_lonProtease) * this.lonProteaseSpecificRate, ...</span>
<a name="l00515"></a>00515 <span class="stringliteral">                    mnmerDecays&#39;</span> * this.monomerLonProteaseCleavages);
<a name="l00516"></a>00516             
<a name="l00517"></a>00517             %stalled polypeptide degradation
<a name="l00518"></a>00518             <span class="keywordflow">if</span> ~isempty(this.polypeptide.abortedPolypeptides)
<a name="l00519"></a>00519                 result([this.substrateIndexs_atp; this.substrateIndexs_water]) = ...
<a name="l00520"></a>00520                     result([this.substrateIndexs_atp; this.substrateIndexs_water]) + ...
<a name="l00521"></a>00521                     ceil(this.ftsHProteaseEnergyCost * min(...
<a name="l00522"></a>00522                         ceil(this.enzymes(this.enzymeIndexs_ftsHProtease) * this.ftsHProteaseSpecificRate), ...
<a name="l00523"></a>00523                         sum(ceil(this.polypeptide.abortedSequenceLengths / this.ftsHProteaseFragmentLength) - 1)));
<a name="l00524"></a>00524                         
<a name="l00525"></a>00525                 result(this.substrateIndexs_water) = ...
<a name="l00526"></a>00526                     result(this.substrateIndexs_water) + ...
<a name="l00527"></a>00527                     sum(this.polypeptide.abortedSequenceLengths - 1);
<a name="l00528"></a>00528             end
<a name="l00529"></a>00529         end
<a name="l00530"></a>00530 
<a name="l00531"></a>00531         %simulation
<a name="l00532"></a>00532         function evolveState(this)
<a name="l00533"></a>00533             this.evolveState_MisfoldProteins();   %misfold proteins
<a name="l00534"></a>00534             this.evolveState_RefoldProteins();    %refold proteins
<a name="l00535"></a>00535             this.evolveState_DegradeComplexes();  %degrade complexes to damaged monomers and RNAs
<a name="l00536"></a>00536             this.evolveState_DegradeAbortedPolypeptides(); %degrade proteolysis tagged monomers to amino acids
<a name="l00537"></a>00537             this.evolveState_DegradeMonomers();   %degrade monomers to amino acids (including monomers marked as damaged by evolveState_DegradeComplexes)
<a name="l00538"></a>00538         end
<a name="l00539"></a>00539 
<a name="l00540"></a>00540         %Model protein misfolding (of mature; inactive; and inactivated forms)
<a name="l00541"></a>00541         %as a passive poisson process with rate proteinMisfoldingRate.
<a name="l00542"></a>00542         %
<a name="l00543"></a>00543         %<a class="code" href="class_protein.html">Protein</a> misfolding does not require metabolites or enzymes.
<a name="l00544"></a>00544         function evolveState_MisfoldProteins(this)
<a name="l00545"></a>00545             %% monomers
<a name="l00546"></a>00546             misfolding = this.randStream.stochasticRound(...
<a name="l00547"></a>00547                 this.monomers([this.monomer.matureIndexs; this.monomer.boundIndexs; this.monomer.inactivatedIndexs], :) ...
<a name="l00548"></a>00548                 * min(1, this.proteinMisfoldingRate * this.stepSizeSec));
<a name="l00549"></a>00549             misfolding(:, this.compartment.membraneIndexs) = 0;
<a name="l00550"></a>00550             misfolding(:, this.compartment.terminalOrganelleMembraneIndexs) = 0;
<a name="l00551"></a>00551             
<a name="l00552"></a>00552             if any(misfolding(:))
<a name="l00553"></a>00553                 tmp = zeros(size(this.monomers));
<a name="l00554"></a>00554                 tmp([this.monomer.matureIndexs; this.monomer.boundIndexs; this.monomer.inactivatedIndexs], :) = -misfolding;
<a name="l00555"></a>00555                 tmp(this.monomer.misfoldedIndexs, :) = sum(permute(reshape(misfolding&#39;, [size(this.monomers, 2) numel(this.monomer.misfoldedIndexs) 3]), [2 1 3]), 3);
<a name="l00556"></a>00556                 notUpdatingProteins = this.monomer.updateExternalState(tmp, true);
<a name="l00557"></a>00557                 misfolding = misfolding - notUpdatingProteins([this.monomer.matureIndexs; this.monomer.boundIndexs; this.monomer.inactivatedIndexs], :);
<a name="l00558"></a>00558                 
<a name="l00559"></a>00559                 this.monomers([this.monomer.matureIndexs; this.monomer.boundIndexs; this.monomer.inactivatedIndexs], :) = ...
<a name="l00560"></a>00560                     this.monomers([this.monomer.matureIndexs; this.monomer.boundIndexs; this.monomer.inactivatedIndexs], :) - ...
<a name="l00561"></a>00561                     misfolding;
<a name="l00562"></a>00562                 this.monomers(this.monomer.misfoldedIndexs, :) = ...
<a name="l00563"></a>00563                     this.monomers(this.monomer.misfoldedIndexs, :) + ...
<a name="l00564"></a>00564                     sum(permute(reshape(misfolding&#39;, [size(this.monomers, 2) numel(this.monomer.misfoldedIndexs) 3]), [2 1 3]), 3);
<a name="l00565"></a>00565             end
<a name="l00566"></a>00566             
<a name="l00567"></a>00567             %% complexs
<a name="l00568"></a>00568             misfolding = this.randStream.stochasticRound(...
<a name="l00569"></a>00569                 this.complexs([this.complex.matureIndexs; this.complex.boundIndexs; this.complex.inactivatedIndexs], :) * ...
<a name="l00570"></a>00570                 min(1, this.proteinMisfoldingRate * this.stepSizeSec));
<a name="l00571"></a>00571             misfolding(:, this.compartment.membraneIndexs) = 0;
<a name="l00572"></a>00572             misfolding(:, this.compartment.terminalOrganelleMembraneIndexs) = 0;
<a name="l00573"></a>00573             
<a name="l00574"></a>00574             if any(misfolding(:))
<a name="l00575"></a>00575                 tmp = zeros(size(this.complex.counts));
<a name="l00576"></a>00576                 tmp([this.complex.matureIndexs; this.complex.boundIndexs; this.complex.inactivatedIndexs], :) = -misfolding;
<a name="l00577"></a>00577                 tmp(this.complex.misfoldedIndexs, :) = sum(permute(reshape(misfolding&#39;, [size(this.complexs, 2) numel(this.complex.misfoldedIndexs) 3]), [2 1 3]), 3);
<a name="l00578"></a>00578                 notUpdatingProteins = this.complex.updateExternalState(tmp, true);
<a name="l00579"></a>00579                 misfolding = misfolding - notUpdatingProteins([this.complex.matureIndexs; this.complex.boundIndexs; this.complex.inactivatedIndexs], :);
<a name="l00580"></a>00580                 
<a name="l00581"></a>00581                 this.complexs([this.complex.matureIndexs; this.complex.boundIndexs; this.complex.inactivatedIndexs], :) = ...
<a name="l00582"></a>00582                     this.complexs([this.complex.matureIndexs;this.complex.boundIndexs; this.complex.inactivatedIndexs], :) - ...
<a name="l00583"></a>00583                     misfolding;
<a name="l00584"></a>00584                 this.complexs(this.complex.misfoldedIndexs, :) = ...
<a name="l00585"></a>00585                     this.complexs(this.complex.misfoldedIndexs, :) + ...
<a name="l00586"></a>00586                     sum(permute(reshape(misfolding&#39;, [size(this.complexs, 2) numel(this.complex.misfoldedIndexs) 3]), [2 1 3]), 3);
<a name="l00587"></a>00587             end
<a name="l00588"></a>00588         end
<a name="l00589"></a>00589 
<a name="l00590"></a>00590         %Model protein refolding as a binary metabolite-limited process:
<a name="l00591"></a>00591         %If active ClpB is present in the cytosol and ATP is present, refold
<a name="l00592"></a>00592         %all misfolded cytoplasmic proteins up to the amount of available ATP.
<a name="l00593"></a>00593         %
<a name="l00594"></a>00594         %Because the ATP cost of ClpB-mediated protein refolding/disagggregation
<a name="l00595"></a>00595         %is unknown we assume that 1 ATP is required per refolding.
<a name="l00596"></a>00596         %
<a name="l00597"></a>00597         %Refolding proteins transition to the active state.
<a name="l00598"></a>00598         function evolveState_RefoldProteins(this)
<a name="l00599"></a>00599             %cumulative sum of misfolded monomers and complexes
<a name="l00600"></a>00600             if ...
<a name="l00601"></a>00601                     ~any(this.monomers(this.monomer.misfoldedIndexs, this.compartment.cytosolIndexs))                  &amp;&amp; ...
<a name="l00602"></a>00602                     ~any(this.monomers(this.monomer.misfoldedIndexs, this.compartment.terminalOrganelleCytosolIndexs)) &amp;&amp; ...
<a name="l00603"></a>00603                     ~any(this.complexs(this.complex.misfoldedIndexs, this.compartment.cytosolIndexs))                  &amp;&amp; ...
<a name="l00604"></a>00604                     ~any(this.complexs(this.complex.misfoldedIndexs, this.compartment.terminalOrganelleCytosolIndexs))
<a name="l00605"></a>00605                 return;
<a name="l00606"></a>00606             end
<a name="l00607"></a>00607             
<a name="l00608"></a>00608             %indices
<a name="l00609"></a>00609             cIdxs = [this.compartment.cytosolIndexs; this.compartment.terminalOrganelleCytosolIndexs];
<a name="l00610"></a>00610             reactantIdxs = [this.substrateIndexs_atp; this.substrateIndexs_water];
<a name="l00611"></a>00611             productIdxs  = [this.substrateIndexs_adp; this.substrateIndexs_phosphate; this.substrateIndexs_hydrogen];
<a name="l00612"></a>00612             
<a name="l00613"></a>00613             %stochastically select proteins to refold, up to limit of available
<a name="l00614"></a>00614             %reactants
<a name="l00615"></a>00615             cumsumMisfoldedProteins = cumsum([0;
<a name="l00616"></a>00616                 this.monomers(this.monomer.misfoldedIndexs, this.compartment.cytosolIndexs);
<a name="l00617"></a>00617                 this.monomers(this.monomer.misfoldedIndexs, this.compartment.terminalOrganelleCytosolIndexs);
<a name="l00618"></a>00618                 this.complexs(this.complex.misfoldedIndexs, this.compartment.cytosolIndexs);
<a name="l00619"></a>00619                 this.complexs(this.complex.misfoldedIndexs, this.compartment.terminalOrganelleCytosolIndexs)]);
<a name="l00620"></a>00620             numFoldingProteins = (this.enzymes(this.enzymeIndexs_clpBProtease) &gt; 0) * min([
<a name="l00621"></a>00621                 this.substrates(reactantIdxs)
<a name="l00622"></a>00622                 cumsumMisfoldedProteins(end)]);
<a name="l00623"></a>00623             if numFoldingProteins == 0
<a name="l00624"></a>00624                 return;
<a name="l00625"></a>00625             end
<a name="l00626"></a>00626             
<a name="l00627"></a>00627             foldingProteinIdxs = this.randStream.randperm(cumsumMisfoldedProteins(end));
<a name="l00628"></a>00628             
<a name="l00629"></a>00629             foldingMonomers = zeros(size(this.monomer.matureIndexs,1), 2);
<a name="l00630"></a>00630             foldingComplexs = zeros(size(this.complex.matureIndexs,1), 2);
<a name="l00631"></a>00631             
<a name="l00632"></a>00632             for i = 1:numFoldingProteins
<a name="l00633"></a>00633                 j = find(foldingProteinIdxs(i) &lt;= cumsumMisfoldedProteins, 1, &#39;first&#39;) - 1;
<a name="l00634"></a>00634                 if j &lt;= numel(this.monomer.matureIndexs)
<a name="l00635"></a>00635                     foldingMonomers(j,1) = foldingMonomers(j,1)+1;
<a name="l00636"></a>00636                 elseif j &lt;= 2*numel(this.monomer.matureIndexs)
<a name="l00637"></a>00637                     foldingMonomers(j - numel(this.monomer.matureIndexs), 2) = foldingMonomers(j - numel(this.monomer.matureIndexs), 2) + 1;
<a name="l00638"></a>00638                 elseif j &lt;= 2*numel(this.monomer.matureIndexs)+numel(this.complex.matureIndexs)
<a name="l00639"></a>00639                     foldingComplexs(j - 2*numel(this.monomer.matureIndexs), 1) = foldingComplexs(j - 2*numel(this.monomer.matureIndexs), 1) + 1;
<a name="l00640"></a>00640                 else
<a name="l00641"></a>00641                     foldingComplexs(j - 2*numel(this.monomer.matureIndexs) - numel(this.complex.matureIndexs), 2) = foldingComplexs(j - 2*numel(this.monomer.matureIndexs)-numel(this.complex.matureIndexs), 2) + 1;
<a name="l00642"></a>00642                 end
<a name="l00643"></a>00643             end
<a name="l00644"></a>00644             
<a name="l00645"></a>00645             %update metabolites
<a name="l00646"></a>00646             this.substrates(reactantIdxs) = this.substrates(reactantIdxs) - numFoldingProteins;
<a name="l00647"></a>00647             this.substrates(productIdxs)  = this.substrates(productIdxs)  + numFoldingProteins;
<a name="l00648"></a>00648             
<a name="l00649"></a>00649             %update monomers
<a name="l00650"></a>00650             this.monomers(this.monomer.matureIndexs, cIdxs) = ...
<a name="l00651"></a>00651                 this.monomers(this.monomer.matureIndexs, cIdxs) + ...
<a name="l00652"></a>00652                 foldingMonomers;
<a name="l00653"></a>00653             this.monomers(this.monomer.misfoldedIndexs, cIdxs) = ...
<a name="l00654"></a>00654                 this.monomers(this.monomer.misfoldedIndexs, cIdxs) - ...
<a name="l00655"></a>00655                 foldingMonomers;
<a name="l00656"></a>00656             
<a name="l00657"></a>00657             %update complexs
<a name="l00658"></a>00658             this.complexs(this.complex.matureIndexs, cIdxs) = ...
<a name="l00659"></a>00659                 this.complexs(this.complex.matureIndexs, cIdxs) + ...
<a name="l00660"></a>00660                 foldingComplexs;
<a name="l00661"></a>00661             this.complexs(this.complex.misfoldedIndexs, cIdxs) = ...
<a name="l00662"></a>00662                 this.complexs(this.complex.misfoldedIndexs, cIdxs) - ...
<a name="l00663"></a>00663                 foldingComplexs;
<a name="l00664"></a>00664         end
<a name="l00665"></a>00665 
<a name="l00666"></a>00666         %Model complex degradation as a stochastic enzyme-dependent poisson
<a name="l00667"></a>00667         %process. To simplify the modeling of complex degradation, this method
<a name="l00668"></a>00668         %first stochastically selects cytosolic complexes to degrade, and then
<a name="l00669"></a>00669         %breaks up the selected complexes into their constituents parts:
<a name="l00670"></a>00670         %prosthetic groups, protein monomers, and RNAs, and finally relies on
<a name="l00671"></a>00671         %the protein  monomer and <a class="code" href="class_r_n_a.html">RNA</a> degradation models
<a name="l00672"></a>00672         %(evolveState_DegradeMonomers and <a class="code" href="class_r_n_a.html">RNA</a> Decay process, respectively) to
<a name="l00673"></a>00673         %degrade the individual components. The method marks the component
<a name="l00674"></a>00674         %protein monomers and RNAs and selected complexes as &quot;damaged&quot; (that is
<a name="l00675"></a>00675         %in a state with 0 half life), so that the the protein monomer and <a class="code" href="class_r_n_a.html">RNA</a>
<a name="l00676"></a>00676         %machinery know to degrade these components.
<a name="l00677"></a>00677         %
<a name="l00678"></a>00678         %<a class="code" href="class_note.html">Note</a>: only cytosolic complexes are degraded because the M. genitalium
<a name="l00679"></a>00679         %protein degradation machinery (protease La) is a cytosolic protein, except that
<a name="l00680"></a>00680                 %all &quot;knocked&quot; out proteins are degraded.
<a name="l00681"></a>00681         function evolveState_DegradeComplexes(this)
<a name="l00682"></a>00682             %compartment indices
<a name="l00683"></a>00683             iC = this.compartment.cytosolIndexs;
<a name="l00684"></a>00684             iTC = this.compartment.terminalOrganelleCytosolIndexs;
<a name="l00685"></a>00685             
<a name="l00686"></a>00686             %stochastically choose complexes to decay
<a name="l00687"></a>00687             decayingRates = this.complexs .* ...
<a name="l00688"></a>00688                 min(1e6, this.complex.decayRates(:, ones(1, this.compartment.count))) * this.stepSizeSec;
<a name="l00689"></a>00689                         decayingRates(this.complex.decayRates &lt; 1e6, [1:iC-1  iC+1:iTC-1  iTC+1:end]) = 0;
<a name="l00690"></a>00690             decayingProteins = min(this.complexs, this.randStream.stochasticRound(decayingRates));
<a name="l00691"></a>00691             decayingRates(decayingProteins == 0) = 0;
<a name="l00692"></a>00692             if ~any(decayingProteins(:))
<a name="l00693"></a>00693                 return;
<a name="l00694"></a>00694             end
<a name="l00695"></a>00695             decayedProteins = zeros(size(decayingProteins));
<a name="l00696"></a>00696             
<a name="l00697"></a>00697             %% decay unmodified complexes
<a name="l00698"></a>00698             decayingSimpleProteins = decayingProteins(this.complexIndexs_unmodified, :);
<a name="l00699"></a>00699             
<a name="l00700"></a>00700             %decrement unmodified complexes
<a name="l00701"></a>00701             if any(decayingSimpleProteins(:))
<a name="l00702"></a>00702                 decayedProteins(this.complexIndexs_unmodified, :) = decayingSimpleProteins;
<a name="l00703"></a>00703             end
<a name="l00704"></a>00704             
<a name="l00705"></a>00705             %% decay modified complexes up to limit of metabolites
<a name="l00706"></a>00706             %(complex decay requires metabolites to reverse the affects of any
<a name="l00707"></a>00707             %modifications)
<a name="l00708"></a>00708             decayingModifiedRates = decayingRates(this.complexIndexs_modified, :);
<a name="l00709"></a>00709             decayingModifiedProteins = decayingProteins(this.complexIndexs_modified, :);
<a name="l00710"></a>00710             decayedModifiedProteins = zeros(size(decayingModifiedProteins));
<a name="l00711"></a>00711             substrates = this.substrates;
<a name="l00712"></a>00712             <a class="code" href="json_parse_8m.html#ad2fc293e8080616c8c1e2ffbd7b86742">while</a> any(decayingModifiedProteins(:))
<a name="l00713"></a>00713                 %randomly select complex, weighted by copy number * decay rate
<a name="l00714"></a>00714                 idx = this.randStream.randsample(numel(decayingModifiedRates), 1, true, decayingModifiedRates(:));
<a name="l00715"></a>00715                 iProtein = this.complexIndexs_modified(mod(idx - 1, numel(this.complexIndexs_modified)) + 1);
<a name="l00716"></a>00716                 
<a name="l00717"></a>00717                 %check if sufficient resources available
<a name="l00718"></a>00718                 if any(substrates &lt; max(0, -this.complexDecayReactions(:, iProtein)))
<a name="l00719"></a>00719                     break;
<a name="l00720"></a>00720                 end
<a name="l00721"></a>00721                 
<a name="l00722"></a>00722                 %decrement complexes
<a name="l00723"></a>00723                 decayingModifiedProteins(idx) = decayingModifiedProteins(idx) - 1;
<a name="l00724"></a>00724                 decayedModifiedProteins(idx) = decayedModifiedProteins(idx) + 1;
<a name="l00725"></a>00725                 if decayingModifiedProteins(idx) == 0
<a name="l00726"></a>00726                     decayingModifiedRates(idx) = 0;
<a name="l00727"></a>00727                 else
<a name="l00728"></a>00728                     decayingModifiedRates(idx) = max(0, decayingModifiedRates(idx) - min(1e6, this.complex.decayRates(iProtein)) * this.stepSizeSec);
<a name="l00729"></a>00729                 end
<a name="l00730"></a>00730                 
<a name="l00731"></a>00731                 %release prosthetic groups,
<a name="l00732"></a>00732                 %invert any modifications
<a name="l00733"></a>00733                 substrates = substrates + this.complexDecayReactions(:, iProtein);
<a name="l00734"></a>00734             end
<a name="l00735"></a>00735             decayedProteins(this.complexIndexs_modified, :) = decayedModifiedProteins;
<a name="l00736"></a>00736             
<a name="l00737"></a>00737             tmp = -decayedProteins;
<a name="l00738"></a>00738             notUpdatingProteins = this.complex.updateExternalState(tmp, true);
<a name="l00739"></a>00739             decayedProteins = decayedProteins - notUpdatingProteins;
<a name="l00740"></a>00740             
<a name="l00741"></a>00741             %decrease complexes. mark protein and <a class="code" href="class_r_n_a.html">RNA</a> monomers for decay
<a name="l00742"></a>00742             %(that is put them in a state with 0 half life) by
<a name="l00743"></a>00743             %evolveState_DegradeMonomers and RNADecay
<a name="l00744"></a>00744             this.complexs = this.complexs - decayedProteins;
<a name="l00745"></a>00745             this.monomers(this.monomer.damagedIndexs, iC) = ...
<a name="l00746"></a>00746                 this.monomers(this.monomer.damagedIndexs, iC) + ...
<a name="l00747"></a>00747                 this.proteinComplexMonomerComposition * sum(decayedProteins, 2);
<a name="l00748"></a>00748             this.RNAs(:, iC) = ...
<a name="l00749"></a>00749                 this.RNAs(:, iC) + ...
<a name="l00750"></a>00750                 this.proteinComplexRNAComposition * sum(decayedProteins, 2);
<a name="l00751"></a>00751             this.substrates = ...
<a name="l00752"></a>00752                 this.substrates + ...
<a name="l00753"></a>00753                 this.complexDecayReactions * sum(decayedProteins, 2);
<a name="l00754"></a>00754         end
<a name="l00755"></a>00755 
<a name="l00756"></a>00756         %Stochastically degrade proteolysis tagged monomers up to limit of
<a name="l00757"></a>00757         %available enzymes (metalloprotease FtsH and 6 peptidases) and their
<a name="l00758"></a>00758         %kinetics (only of FtsH; peptidase kinetics haven&#39;t been characterized;
<a name="l00759"></a>00759         %in this case we model that degradation can occur as only as at least 1
<a name="l00760"></a>00760         %of each peptidase species is present) and metabolites (hydrolytic
<a name="l00761"></a>00761         %cleavage requires water, and FtsH requires ATP). Because the specific
<a name="l00762"></a>00762         %function and mechanism of each peptidase hasn&#39;t been well characterized
<a name="l00763"></a>00763         %we assume that degradation requires all peptidases, and they don&#39;t
<a name="l00764"></a>00764         %require ATP. Proteolysis tagged polypeptides are assumed to be degraded
<a name="l00765"></a>00765         %immediately, that is they have a half life of 0.
<a name="l00766"></a>00766         function evolveState_DegradeAbortedPolypeptides(this)
<a name="l00767"></a>00767             %terminate early if no proteolysis tagged monomers
<a name="l00768"></a>00768             if isempty(this.polypeptide.abortedPolypeptides)
<a name="l00769"></a>00769                 return;
<a name="l00770"></a>00770             end
<a name="l00771"></a>00771             
<a name="l00772"></a>00772             % import classes
<a name="l00773"></a>00773             import edu.stanford.covert.cell.kb.<a class="code" href="class_protein_monomer.html">ProteinMonomer</a>;
<a name="l00774"></a>00774             
<a name="l00775"></a>00775             %indices
<a name="l00776"></a>00776             aaWaterIdxs = [
<a name="l00777"></a>00777                 this.substrateIndexs_aminoAcids;
<a name="l00778"></a>00778                 this.substrateIndexs_water];
<a name="l00779"></a>00779             energyIdxs = [
<a name="l00780"></a>00780                 this.substrateIndexs_atp;
<a name="l00781"></a>00781                 this.substrateIndexs_adp;
<a name="l00782"></a>00782                 this.substrateIndexs_phosphate;
<a name="l00783"></a>00783                 this.substrateIndexs_water;
<a name="l00784"></a>00784                 this.substrateIndexs_hydrogen];
<a name="l00785"></a>00785 
<a name="l00786"></a>00786             %randomly order proteolysis tagged monomer
<a name="l00787"></a>00787             if numel(this.polypeptide.abortedPolypeptides) &gt; 1
<a name="l00788"></a>00788                 this.polypeptide.abortedPolypeptides = ...
<a name="l00789"></a>00789                     this.polypeptide.abortedPolypeptides(this.randStream.randperm(size(this.polypeptide.abortedPolypeptides, 1)), :);
<a name="l00790"></a>00790             end
<a name="l00791"></a>00791             
<a name="l00792"></a>00792             %determine number of monomers that can be cleaved
<a name="l00793"></a>00793             cleavages = ceil(this.polypeptide.abortedSequenceLengths / this.ftsHProteaseFragmentLength) - 1;
<a name="l00794"></a>00794             peptidaseCosts = (1:numel(cleavages))&#39;;
<a name="l00795"></a>00795             if any(cleavages)
<a name="l00796"></a>00796                 ftsHCosts = this.randStream.stochasticRound((this.enzymes(this.enzymeIndexs_ftsHProtease) * this.stepSizeSec) ./ cumsum(cleavages / this.ftsHProteaseSpecificRate));
<a name="l00797"></a>00797                 ftsHCosts(find(ftsHCosts == 0, 1, &#39;first&#39;):end) = 0;
<a name="l00798"></a>00798             else
<a name="l00799"></a>00799                 ftsHCosts = ones(size(cleavages));
<a name="l00800"></a>00800             end
<a name="l00801"></a>00801             energyCosts = cumsum(this.randStream.stochasticRound(cleavages * this.ftsHProteaseEnergyCost));
<a name="l00802"></a>00802             waterCosts = energyCosts + cumsum(this.polypeptide.abortedSequenceLengths - 1);
<a name="l00803"></a>00803             numReactions = find(...
<a name="l00804"></a>00804                 [0; peptidaseCosts] &lt;= min(this.enzymes(this.enzymeIndexs_peptidases)) * this.stepSizeSec &amp; ...
<a name="l00805"></a>00805                 [1; ftsHCosts]      &gt;= 1 &amp; ...
<a name="l00806"></a>00806                 [0; energyCosts]    &lt;= this.substrates(this.substrateIndexs_atp) &amp; ...
<a name="l00807"></a>00807                 [0; waterCosts]     &lt;= this.substrates(this.substrateIndexs_water), ...
<a name="l00808"></a>00808                 1, &#39;last&#39;) - 1;
<a name="l00809"></a>00809             
<a name="l00810"></a>00810             %stop if no degradation
<a name="l00811"></a>00811             if isempty(numReactions) || numReactions == 0
<a name="l00812"></a>00812                 return;
<a name="l00813"></a>00813             end
<a name="l00814"></a>00814             
<a name="l00815"></a>00815             %loop over tagged monomers
<a name="l00816"></a>00816             decayReactions = zeros(22, 1);
<a name="l00817"></a>00817             abortedSeqs = this.polypeptide.abortedSequences;
<a name="l00818"></a>00818             for i = 1:numReactions
<a name="l00819"></a>00819                 decayReactions = ...
<a name="l00820"></a>00820                     + decayReactions ...
<a name="l00821"></a>00821                     + ProteinMonomer.computeDecayReaction(...
<a name="l00822"></a>00822                     ProteinMonomer.computeBaseCount(abortedSeqs{i}, 22, 1:21, <span class="keyword">true</span>),...
<a name="l00823"></a>00823                     length(abortedSeqs{i}), 22)<span class="stringliteral">&#39;;</span>
<a name="l00824"></a>00824 <span class="stringliteral">            end</span>
<a name="l00825"></a>00825 <span class="stringliteral"></span>
<a name="l00826"></a>00826 <span class="stringliteral">            %% update state</span>
<a name="l00827"></a>00827 <span class="stringliteral">            %metabolites</span>
<a name="l00828"></a>00828 <span class="stringliteral">            this.substrates(aaWaterIdxs) = this.substrates(aaWaterIdxs) + decayReactions;</span>
<a name="l00829"></a>00829 <span class="stringliteral">            this.substrates(energyIdxs)  = this.substrates(energyIdxs) + [-1; 1; 1; -1; 1] * energyCosts(numReactions);</span>
<a name="l00830"></a>00830 <span class="stringliteral"></span>
<a name="l00831"></a>00831 <span class="stringliteral">            %proteolysis tagged monomers</span>
<a name="l00832"></a>00832 <span class="stringliteral">            this.polypeptide.abortedPolypeptides = this.polypeptide.abortedPolypeptides(numReactions+1:end, :);</span>
<a name="l00833"></a>00833 <span class="stringliteral">        end</span>
<a name="l00834"></a>00834 <span class="stringliteral"></span>
<a name="l00835"></a>00835 <span class="stringliteral">        %Stochastically degrade monomers weighted by copy number and decay rate</span>
<a name="l00836"></a>00836 <span class="stringliteral">        %up to limit of available enzymes (monomers require protease La and 6</span>
<a name="l00837"></a>00837 <span class="stringliteral">        %peptidases) and their kinetics (except for the peptidases where</span>
<a name="l00838"></a>00838 <span class="stringliteral">        %their kinetics haven&#39;</span>t been characterized; in <span class="keyword">this</span> <span class="keywordflow">case</span> we model that
<a name="l00839"></a>00839         %degradation can occur as only as at least 1 of each peptidase species
<a name="l00840"></a>00840         %is present) and metabolites (hydrolytic cleavage requires water, and
<a name="l00841"></a>00841         %Lon requires ATP). Because the specific <span class="keyword">function</span> and  mechanism
<a name="l00842"></a>00842         %of each peptidase hasn<span class="stringliteral">&#39;t been well characterized we assume that</span>
<a name="l00843"></a>00843 <span class="stringliteral">        %degradation requires all peptidases, and they don&#39;</span>t require ATP.
<a name="l00844"></a>00844         <span class="keyword">function</span> evolveState_DegradeMonomers(<span class="keyword">this</span>)
<a name="l00845"></a>00845             %indices
<a name="l00846"></a>00846             iC = this.compartment.cytosolIndexs;
<a name="l00847"></a>00847             iTC = this.compartment.terminalOrganelleCytosolIndexs;
<a name="l00848"></a>00848             
<a name="l00849"></a>00849             %decaying monomers
<a name="l00850"></a>00850             decayingRates = this.monomers .* ...
<a name="l00851"></a>00851                 min(1e6, this.monomer.decayRates(:, ones(1, this.compartment.count))) * this.stepSizeSec;
<a name="l00852"></a>00852                         decayingRates(this.monomer.decayRates &lt; 1e6, [1:iC-1  iC+1:iTC-1  iTC+1:end]) = 0;
<a name="l00853"></a>00853             decayingProteins = min(this.monomers, this.randStream.stochasticRound(decayingRates));
<a name="l00854"></a>00854             decayingRates(decayingProteins == 0) = 0;
<a name="l00855"></a>00855             <span class="keywordflow">if</span> ~any(decayingProteins(:))
<a name="l00856"></a>00856                 return;
<a name="l00857"></a>00857             end
<a name="l00858"></a>00858             decayedProteins = zeros(size(decayingProteins));
<a name="l00859"></a>00859             
<a name="l00860"></a>00860             iEnergy = [
<a name="l00861"></a>00861                 this.substrateIndexs_atp
<a name="l00862"></a>00862                 this.substrateIndexs_adp
<a name="l00863"></a>00863                 this.substrateIndexs_phosphate
<a name="l00864"></a>00864                 this.substrateIndexs_water
<a name="l00865"></a>00865                 this.substrateIndexs_hydrogen
<a name="l00866"></a>00866                 ];
<a name="l00867"></a>00867             
<a name="l00868"></a>00868             protease  = this.enzymes(this.enzymeIndexs_lonProtease) * this.stepSizeSec * this.lonProteaseSpecificRate;
<a name="l00869"></a>00869             peptidase = this.randStream.stochasticRound(min(this.enzymes(this.enzymeIndexs_peptidases)) * this.stepSizeSec);
<a name="l00870"></a>00870             
<a name="l00871"></a>00871             %decay monomers up to limit of metabolites and enzymes
<a name="l00872"></a>00872             substrates = this.substrates;
<a name="l00873"></a>00873             <a class="code" href="json_parse_8m.html#ad2fc293e8080616c8c1e2ffbd7b86742">while</a> any(decayingProteins(:))
<a name="l00874"></a>00874                 %randomly select monomer, weighted by copy number * decay rate
<a name="l00875"></a>00875                 idx = this.randStream.randsample(numel(decayingRates), 1, true, decayingRates(:));
<a name="l00876"></a>00876                 iProtein = mod(idx - 1, size(this.monomers, 1)) + 1;
<a name="l00877"></a>00877                 
<a name="l00878"></a>00878                 %check if sufficient resources available
<a name="l00879"></a>00879                 substrateCost = -this.monomerDecayReactions(:, iProtein);
<a name="l00880"></a>00880                 substrateCost(iEnergy) = substrateCost(iEnergy) + ...
<a name="l00881"></a>00881                     [1; -1; -1; 1; -1] * this.lonProteaseEnergyCost * this.monomerLonProteaseCleavages(iProtein);
<a name="l00882"></a>00882                 proteaseCost = this.monomerLonProteaseCleavages(iProtein);
<a name="l00883"></a>00883                 
<a name="l00884"></a>00884                 if ...
<a name="l00885"></a>00885                         any(substrates(iEnergy) &lt; max(0, substrateCost(iEnergy))) || ...
<a name="l00886"></a>00886                         this.randStream.stochasticRound(protease / proteaseCost) &lt; 1 || ...
<a name="l00887"></a>00887                         peptidase &lt; 1
<a name="l00888"></a>00888                     break;
<a name="l00889"></a>00889                 end
<a name="l00890"></a>00890                 
<a name="l00891"></a>00891                 %update state
<a name="l00892"></a>00892                 decayingProteins(idx) = decayingProteins(idx) - 1;
<a name="l00893"></a>00893                 decayedProteins(idx) = decayedProteins(idx) + 1;
<a name="l00894"></a>00894                 if decayingProteins(idx) &lt;= 0
<a name="l00895"></a>00895                     decayingRates(idx) = 0;
<a name="l00896"></a>00896                 else
<a name="l00897"></a>00897                     decayingRates(idx) = max(0, decayingRates(idx) - min(1e6, this.monomer.decayRates(iProtein)) * this.stepSizeSec);
<a name="l00898"></a>00898                 end
<a name="l00899"></a>00899                 
<a name="l00900"></a>00900                 substrates = substrates - substrateCost;
<a name="l00901"></a>00901                 protease = max(0, protease - proteaseCost);
<a name="l00902"></a>00902             end
<a name="l00903"></a>00903             
<a name="l00904"></a>00904             tmp = -decayedProteins;
<a name="l00905"></a>00905             notUpdatingProteins = this.monomer.updateExternalState(tmp, true);
<a name="l00906"></a>00906             decayedProteins = decayedProteins - notUpdatingProteins;
<a name="l00907"></a>00907             
<a name="l00908"></a>00908             this.monomers = this.monomers - decayedProteins;
<a name="l00909"></a>00909             this.substrates = ...
<a name="l00910"></a>00910                 this.substrates + ...
<a name="l00911"></a>00911                 this.monomerDecayReactions * sum(decayedProteins, 2);
<a name="l00912"></a>00912             this.substrates(iEnergy) = ...
<a name="l00913"></a>00913                 this.substrates(iEnergy) - ...
<a name="l00914"></a>00914                 [1; -1; -1; 1; -1] * this.lonProteaseEnergyCost * (this.monomerLonProteaseCleavages&#39; * sum(decayedProteins, 2));
<a name="l00915"></a>00915         end
<a name="l00916"></a>00916     end
<a name="l00917"></a>00917 
<a name="l00918"></a>00918     %get methods of dependent local state
<a name="l00919"></a>00919     methods
<a name="l00920"></a>00920         function value = getDryWeight(this)
<a name="l00921"></a>00921             rnaMolecularWeights = this.rna.molecularWeights(this.rna.damagedIndexs([this.rna.matureRRNAIndexs; this.rna.matureSRNAIndexs]));
<a name="l00922"></a>00922             
<a name="l00923"></a>00923             if size(this.RNAs, 3) == 1
<a name="l00924"></a>00924                 value = this.getDryWeight@edu.stanford.covert.cell.sim.<a class="code" href="class_process.html">Process</a>() + (...
<a name="l00925"></a>00925                     rnaMolecularWeights&#39;           * sum(this.RNAs,     2) + ...
<a name="l00926"></a>00926                     this.monomer.molecularWeights&#39; * sum(this.monomers, 2) + ...
<a name="l00927"></a>00927                     this.complex.molecularWeights&#39; * sum(this.complexs, 2)) ...
<a name="l00928"></a>00928                     / edu.stanford.covert.util.<a class="code" href="class_constant_util.html">ConstantUtil</a>.nAvogadro;
<a name="l00929"></a>00929             else
<a name="l00930"></a>00930                 value = this.getDryWeight@edu.stanford.covert.cell.sim.<a class="code" href="class_process.html">Process</a>() + (...
<a name="l00931"></a>00931                     permute(rnaMolecularWeights&#39;           * permute(sum(this.RNAs,     2), [1 3 2]), [1 3 2]) + ...
<a name="l00932"></a>00932                     permute(this.monomer.molecularWeights&#39; * permute(sum(this.monomers, 2), [1 3 2]), [1 3 2]) + ...
<a name="l00933"></a>00933                     permute(this.complex.molecularWeights&#39; * permute(sum(this.complexs, 2), [1 3 2]), [1 3 2])) ...
<a name="l00934"></a>00934                     / edu.stanford.covert.util.<a class="code" href="class_constant_util.html">ConstantUtil</a>.nAvogadro;
<a name="l00935"></a>00935             end
<a name="l00936"></a>00936         end
<a name="l00937"></a>00937     end
<a name="l00938"></a>00938 end
</pre></div></div><!-- contents -->
</div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
<a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(0)"><span class="SelectionMark">&#160;</span>All</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(1)"><span class="SelectionMark">&#160;</span>Classes</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(2)"><span class="SelectionMark">&#160;</span>Files</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(3)"><span class="SelectionMark">&#160;</span>Functions</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(4)"><span class="SelectionMark">&#160;</span>Variables</a></div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

  <div id="nav-path" class="navpath">
    <ul>
      <li class="navelem"><a class="el" href="_protein_decay_8m.html">ProteinDecay.m</a>      </li>

    <li class="footer">Generated on Wed Feb 8 2012 21:42:32 for Whole Cell Simulation by
    <a href="http://www.doxygen.org/index.html">
    <img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.7.6.1 </li>
   </ul>
 </div>


</body>
</html>
