<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<title>Whole Cell Simulation: src/+edu/+stanford/+covert/+cell/+sim/+process/ProteinFolding.m Source File</title>

<link href="tabs.css" rel="stylesheet" type="text/css"/>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
<link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="navtree.js"></script>
<script type="text/javascript">
  $(document).ready(initResizable);
</script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
  $(document).ready(function() { searchBox.OnSelectItem(0); });
</script>

</head>
<body>
<div id="top"><!-- do not remove this div! -->


<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  
  
  <td style="padding-left: 0.5em;">
   <div id="projectname">Whole Cell Simulation
   
   </div>
   
  </td>
  
  
  
 </tr>
 </tbody>
</table>
</div>

<!-- Generated by Doxygen 1.7.6.1 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "search",false,'Search');
</script>
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&#160;Page</span></a></li>
      <li><a href="annotated.html"><span>Classes</span></a></li>
      <li class="current"><a href="files.html"><span>Files</span></a></li>
      <li>
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
  <div id="navrow2" class="tabs2">
    <ul class="tablist">
      <li><a href="files.html"><span>File&#160;List</span></a></li>
      <li><a href="globals.html"><span>File&#160;Members</span></a></li>
    </ul>
  </div>
</div>
<div id="side-nav" class="ui-resizable side-nav-resizable">
  <div id="nav-tree">
    <div id="nav-tree-contents">
    </div>
  </div>
  <div id="splitbar" style="-moz-user-select:none;" 
       class="ui-resizable-handle">
  </div>
</div>
<script type="text/javascript">
  initNavTree('_protein_folding_8m.html','');
</script>
<div id="doc-content">
<div class="header">
  <div class="headertitle">
<div class="title">src/+edu/+stanford/+covert/+cell/+sim/+process/ProteinFolding.m</div>  </div>
</div><!--header-->
<div class="contents">
<a href="_protein_folding_8m.html">Go to the documentation of this file.</a><div class="fragment"><pre class="fragment"><a name="l00001"></a>00001 %<a class="code" href="class_protein.html">Protein</a> Folding
<a name="l00002"></a>00002 %
<a name="l00003"></a>00003 % @wholeCellModelID Process_ProteinFolding
<a name="l00004"></a>00004 % @name             <a class="code" href="class_protein.html">Protein</a> Folding
<a name="l00005"></a>00005 % @description
<a name="l00006"></a>00006 %   Biology
<a name="l00007"></a>00007 %   ==============
<a name="l00008"></a>00008 %   <a class="code" href="class_protein.html">Protein</a> folding is the process whereby proteins which are produced as linear
<a name="l00009"></a>00009 %   amino acid polymers relax to their most energetically favorable, and often
<a name="l00010"></a>00010 %   more compact and catalytically active configuration. Some proteins relax
<a name="l00011"></a>00011 %   quickly and spontaneously to their free energy minimum, <span class="keywordflow">while</span> other proteins
<a name="l00012"></a>00012 %   require chaperones to help them achieve their folded configuration more
<a name="l00013"></a>00013 %   quickly. Additionally, some proteins coordinate prosthetic groups such as
<a name="l00014"></a>00014 %   metal ions and other small molecules during <span class="keyword">this</span> protein rearrangement
<a name="l00015"></a>00015 %   process. Prosthetic groups may help stabilize the protein<span class="stringliteral">&#39;s catalytically</span>
<a name="l00016"></a>00016 <span class="stringliteral">%   active configuration, or may perform folding process.</span>
<a name="l00017"></a>00017 <span class="stringliteral">%</span>
<a name="l00018"></a>00018 <span class="stringliteral">%   In addition to protein chaperones, membrane protein folding may be assisted</span>
<a name="l00019"></a>00019 <span class="stringliteral">%   by phosphatidyl ethanolamine (PE) and phosphatidyl glycerol (PG) [PUB_0646].</span>
<a name="l00020"></a>00020 <span class="stringliteral">%   This effect is not currently modeled by this process.</span>
<a name="l00021"></a>00021 <span class="stringliteral">%</span>
<a name="l00022"></a>00022 <span class="stringliteral">%   This process simulates</span>
<a name="l00023"></a>00023 <span class="stringliteral">%   1. Prosthetic group complexation</span>
<a name="l00024"></a>00024 <span class="stringliteral">%      - Inorganic ions: Fe2, Fe3, K, Mg, Mn, Na, Zn</span>
<a name="l00025"></a>00025 <span class="stringliteral">%   2. Protein folding, where required mediated by ATP-dependent chaperone(s)</span>
<a name="l00026"></a>00026 <span class="stringliteral">%      [PUB_0014, PUB_0644]:</span>
<a name="l00027"></a>00027 <span class="stringliteral">%      a. Trigger factor (Tig, MG_238_MONOMER) – Highly expressed protein which</span>
<a name="l00028"></a>00028 <span class="stringliteral">%         possibly interacts with the L23 subunit of every ribosome,</span>
<a name="l00029"></a>00029 <span class="stringliteral">%         co-translationally assisting in the early folding every protein</span>
<a name="l00030"></a>00030 <span class="stringliteral">%         [PUB_0014, PUB_0644]. Binds vacant ribosomes with half-life of 11 s</span>
<a name="l00031"></a>00031 <span class="stringliteral">%         [PUB_0005]; binds ribosome-peptide complexes with half-life of 15-50 s</span>
<a name="l00032"></a>00032 <span class="stringliteral">%         [PUB_0005]; remains associated with the peptide for a up to more</span>
<a name="l00033"></a>00033 <span class="stringliteral">%         seconds after peptide release from the ribosome [PUB_0005]. Trigger</span>
<a name="l00034"></a>00034 <span class="stringliteral">%         factor-50S ribosome affinity: KD=1 ?M [PUB_0005]. We model trigger as</span>
<a name="l00035"></a>00035 <span class="stringliteral">%         required for the proper folding of all proteins.</span>
<a name="l00036"></a>00036 <span class="stringliteral">%      b. Chaperone (DnaK, MG_305_MONOMER) – Folds 5-18% of proteins [PUB_0014,</span>
<a name="l00037"></a>00037 <span class="stringliteral">%         PUB_0644]. Typically folds proteins &gt; 30 kDa, typically in less than 2</span>
<a name="l00038"></a>00038 <span class="stringliteral">%         min [PUB_0014, PUB_0644]. Binds backbone of short, linear, unfolded,</span>
<a name="l00039"></a>00039 <span class="stringliteral">%         hydrophobic peptide segments [PUB_0014, PUB_0644]. ATP hydrolysis and</span>
<a name="l00040"></a>00040 <span class="stringliteral">%         peptide release catalyzed by GrpE (MG_201_DIMER) [PUB_0014, PUB_0644].</span>
<a name="l00041"></a>00041 <span class="stringliteral">%         Regulated by co-chaperone DnaJ (MG_019_DIMER) which interacts with the</span>
<a name="l00042"></a>00042 <span class="stringliteral">%         peptide side chains [PUB_0014, PUB_0644].</span>
<a name="l00043"></a>00043 <span class="stringliteral">%      c. Chaperonin (GroEL, GroES, MG_392_393_21MER) – Folds 10-15% proteins by</span>
<a name="l00044"></a>00044 <span class="stringliteral">%         ATP-dependent mechanism [PUB_0014, PUB_0644]. Typically folds proteins</span>
<a name="l00045"></a>00045 <span class="stringliteral">%         20-60 kDa in size [PUB_0014, PUB_0644] with half-life &gt; 10 min</span>
<a name="l00046"></a>00046 <span class="stringliteral">%         [PUB_0014, PUB_0644], 30-60 s [PUB_0389].</span>
<a name="l00047"></a>00047 <span class="stringliteral">%</span>
<a name="l00048"></a>00048 <span class="stringliteral">%   Knowledge Base</span>
<a name="l00049"></a>00049 <span class="stringliteral">%   ==============</span>
<a name="l00050"></a>00050 <span class="stringliteral">%   The chaperones (except Tig, which all protein monomers require) and</span>
<a name="l00051"></a>00051 <span class="stringliteral">%   prosthetic groups required to fold each protein were reconstructed from the</span>
<a name="l00052"></a>00052 <span class="stringliteral">%   literature (see following sections), organized into the knowledge base, and</span>
<a name="l00053"></a>00053 <span class="stringliteral">%   are retrieved from the knowledge base in initializeConstants. Where known</span>
<a name="l00054"></a>00054 <span class="stringliteral">%   the stoichiometry of each prosthetic group was stored in the knowledge base;</span>
<a name="l00055"></a>00055 <span class="stringliteral">%   prosthetic groups with unreported stoichiometry are indicated with</span>
<a name="l00056"></a>00056 <span class="stringliteral">%   stoichiometry values of -1.</span>
<a name="l00057"></a>00057 <span class="stringliteral">%</span>
<a name="l00058"></a>00058 <span class="stringliteral">%   Chaperones</span>
<a name="l00059"></a>00059 <span class="stringliteral">%   ++++++++++++++</span>
<a name="l00060"></a>00060 <span class="stringliteral">%   The chaperone requirements for the folding of each protein were compiled</span>
<a name="l00061"></a>00061 <span class="stringliteral">%   for other bacteria (see sources below), and mapped to M. genitalium by</span>
<a name="l00062"></a>00062 <span class="stringliteral">%   homology.</span>
<a name="l00063"></a>00063 <span class="stringliteral">%</span>
<a name="l00064"></a>00064 <span class="stringliteral">%       Tig    may interact with all nascent peptides at the ribosome</span>
<a name="l00065"></a>00065 <span class="stringliteral">%              polypeptide exit site [PUB_0005, PUB_0009, PUB_0388] and assist</span>
<a name="l00066"></a>00066 <span class="stringliteral">%              in early folding. Binds vacant ribosomes with half-life of 11 s</span>
<a name="l00067"></a>00067 <span class="stringliteral">%              [PUB_0005]; binds ribosome-peptide complexes with half-life of</span>
<a name="l00068"></a>00068 <span class="stringliteral">%              15-50 s [PUB_0005]; remains associated with the peptide for a</span>
<a name="l00069"></a>00069 <span class="stringliteral">%              up to more seconds after peptide release from the ribosome</span>
<a name="l00070"></a>00070 <span class="stringliteral">%              [PUB_0005]. Trigger factor-50S ribosome affinity: KD=1 ?M</span>
<a name="l00071"></a>00071 <span class="stringliteral">%              [PUB_0005]. We model trigger as required for the proper folding</span>
<a name="l00072"></a>00072 <span class="stringliteral">%              of all proteins.</span>
<a name="l00073"></a>00073 <span class="stringliteral">%       DnaK   Deuerling et al performed a proteome-scale search for DnaK</span>
<a name="l00074"></a>00074 <span class="stringliteral">%              substrates in E. coli [PUB_0388]</span>
<a name="l00075"></a>00075 <span class="stringliteral">%       GroEL  Kerner et al performed a proteome-scale search for GroEL</span>
<a name="l00076"></a>00076 <span class="stringliteral">%              substrates in E. coli [PUB_0389] and Endo and Kurusu performed</span>
<a name="l00077"></a>00077 <span class="stringliteral">%              a proteome-scale search for GroEL substrates in B. subtilis</span>
<a name="l00078"></a>00078 <span class="stringliteral">%              [PUB_0391]</span>
<a name="l00079"></a>00079 <span class="stringliteral">%       FtsH   may act as molecule chaperone for membrane proteins</span>
<a name="l00080"></a>00080 <span class="stringliteral">%              [PUB_0014]. Several other functions have also been associated</span>
<a name="l00081"></a>00081 <span class="stringliteral">%              with FtsH. To date no proteome-scale studies of FtsH activity</span>
<a name="l00082"></a>00082 <span class="stringliteral">%              has been performed, and FtsH&#39;</span>s chaperone substrates are not
<a name="l00083"></a>00083 %              well characterized. Conequently we chose not to model FtsH as
<a name="l00084"></a>00084 %              a molecular chaperone, but rather a protease.
<a name="l00085"></a>00085 %
<a name="l00086"></a>00086 %   Substrates of SecB, a molecular chaperone not present in M. genitalium, have
<a name="l00087"></a>00087 %   also been identified on proteome-scale in E. coli [PUB_0390].
<a name="l00088"></a>00088 %
<a name="l00089"></a>00089 %   The reconstruction found that complexes typically require no chaperones to
<a name="l00090"></a>00090 %   fold, with the notable exception of MG_392_393_21MER (which itself is a
<a name="l00091"></a>00091 %   chaperone), which requires four.
<a name="l00092"></a>00092 %
<a name="l00093"></a>00093 %   Prosthetic groups
<a name="l00094"></a>00094 %   ++++++++++++++
<a name="l00095"></a>00095 %   <a class="code" href="class_reaction.html">Reaction</a> coenzymes and prosthetic groups were reconstructed from several
<a name="l00096"></a>00096 %   sources including the databases BioCyc [PUB_0006], BRENDA [PUB_0570],
<a name="l00097"></a>00097 %   GenoBase [PUB_0386], Kinetikon [PUB_0571], Metal-MaCiE [PUB_0387], and
<a name="l00098"></a>00098 %   UniProt [PUB_0096] and the primary literature [PUB_0131]. These sources
<a name="l00099"></a>00099 %   frequently report protein prosthetic groups as a list of chemical species
<a name="l00100"></a>00100 %   (eg. metal ions) which can each fill the same prosthetic group role. We
<a name="l00101"></a>00101 %   simplified these ambiguous lists of possible prosthetic groups to a single
<a name="l00102"></a>00102 %   chemical species by choosing the most common chemical species according to
<a name="l00103"></a>00103 %   cell composition (Na+, K+ &gt; Mg2+, Cl- &gt; Fe2+, Fe3+ &gt; Ca2+ &gt; Mn2+ &gt; Cu2+ &gt;
<a name="l00104"></a>00104 %   Mo6+, Zn2, Co2, Ni2 [PUB_0393, PUB_0394, PUB_0395]), or the species with the
<a name="l00105"></a>00105 %   highest protein affinity (Irving-Williams Series, Mn2+ &lt; Fe2+, Fe3+ &lt; Co2+ &lt;
<a name="l00106"></a>00106 %   Ni2+ &lt; Cu2+ &gt; Zn2+ [PUB_0404]).
<a name="l00107"></a>00107 %
<a name="l00108"></a>00108 %   Roughly 20% of the monomers and 1% of the complexes bind at least one
<a name="l00109"></a>00109 %   prosthetic group.
<a name="l00110"></a>00110 %
<a name="l00111"></a>00111 %   Representation
<a name="l00112"></a>00112 %   ==============
<a name="l00113"></a>00113 %   substrates, enzymes, unfoldedMonomers, unfoldedComplexs, foldedMonomers, and
<a name="l00114"></a>00114 %   foldedComplexs represent the counts of free metabolites, chaperones,
<a name="l00115"></a>00115 %   unfolded protein monomers, unfolded protein complexes, folded protein
<a name="l00116"></a>00116 %   monomers, and folded protein complexes. Each of these component types have
<a name="l00117"></a>00117 %   been mapped from the several compartments of the simulation class to one
<a name="l00118"></a>00118 %   pseudo compartment in this process. That is this process only &quot;sees&quot; the
<a name="l00119"></a>00119 %   counts of metabolites, chaperones, and protein monomers and complexes that
<a name="l00120"></a>00120 %   are relevant to protein folding. The process contains no intermediate
<a name="l00121"></a>00121 %   representation of protein folding, that is protein folding is simulated as a
<a name="l00122"></a>00122 %   all-or-nothing process that either does not occur or proceeds to completion
<a name="l00123"></a>00123 %   within a single time step.
<a name="l00124"></a>00124 %
<a name="l00125"></a>00125 %   The proteinChaperoneMatrix and proteinProstheticGroupMatrix are adjacency
<a name="l00126"></a>00126 %   matrices which represent the chaperone and prosthetic groups each protein
<a name="l00127"></a>00127 %   monomer and complex requires to achieve its catalytically active
<a name="l00128"></a>00128 %   configuration. Where the stoichiometry of a prosthetic group is unknown and
<a name="l00129"></a>00129 %   indicated in the knowledge base with the value -1, here we assume a
<a name="l00130"></a>00130 %   stoichiometry of 1 and set the element of proteinProstheticGroupMatrix
<a name="l00131"></a>00131 %   accordingly.
<a name="l00132"></a>00132 %
<a name="l00133"></a>00133 %   Initialization
<a name="l00134"></a>00134 %   ==============
<a name="l00135"></a>00135 %   All protein monomers and complexes are initialized to the mature state. This
<a name="l00136"></a>00136 %   is accomplished by the simulation class initializeState method.
<a name="l00137"></a>00137 %
<a name="l00138"></a>00138 %   Simulation
<a name="l00139"></a>00139 %   ==============
<a name="l00140"></a>00140 %   As with the other protein maturation processes, protein folding is modeled as
<a name="l00141"></a>00141 %   as all-or-nothing process at the single time-step level. A protein only
<a name="l00142"></a>00142 %   proceeds through this phase if its chaperones (if any) and any prosthetic
<a name="l00143"></a>00143 %   group ions are all present in the same time step. Chaperone kinetics are not
<a name="l00144"></a>00144 %   well characterized, and are not currrently modeled. Proteins that require
<a name="l00145"></a>00145 %   neither prosthetic groups nor chaperones are considered folded after one
<a name="l00146"></a>00146 %   time step.
<a name="l00147"></a>00147 %
<a name="l00148"></a>00148 %   Algorithm
<a name="l00149"></a>00149 %   ++++++++++++++
<a name="l00150"></a>00150 %   While(true)
<a name="l00151"></a>00151 %     1. Calculate numbers of proteins that can fold based on prosthetic group,
<a name="l00152"></a>00152 %        chaperone, and unfolded protein availability.
<a name="l00153"></a>00153 %     2. Randomly select proteins to fold weighted by limits calculated in step
<a name="l00154"></a>00154 %        (1).
<a name="l00155"></a>00155 %     3. Update counts of prosthetic groups, chaperones, and unfolded and folded
<a name="l00156"></a>00156 %        proteins.
<a name="l00157"></a>00157 %     4. Repeat until insufficient resources to further fold proteins
<a name="l00158"></a>00158 %   End
<a name="l00159"></a>00159 %
<a name="l00160"></a>00160 % Author: Jonathan Karr, jkarr@stanford.edu
<a name="l00161"></a>00161 % Affilitation: Covert Lab, Department of Bioengineering, Stanford University
<a name="l00162"></a>00162 % Last updated: 1/4/2010
<a name="l00163"></a>00163 classdef ProteinFolding &lt; edu.stanford.covert.cell.sim.Process
<a name="l00164"></a>00164     %property annotations
<a name="l00165"></a>00165     properties (Constant)
<a name="l00166"></a>00166         optionNames__              = {}; %names of option <a class="code" href="json_parse_8m.html#a6e4256f69b751d3d7ec4ea52ff5a8dee">properties</a>
<a name="l00167"></a>00167         fixedConstantNames__       = {   %names of fixed constant <a class="code" href="json_parse_8m.html#a6e4256f69b751d3d7ec4ea52ff5a8dee">properties</a>
<a name="l00168"></a>00168                     <span class="stringliteral">&#39;proteinProstheticGroupMatrix&#39;</span>;
<a name="l00169"></a>00169             <span class="stringliteral">&#39;proteinChaperoneMatrix&#39;</span>;
<a name="l00170"></a>00170             <span class="stringliteral">&#39;monomerCompartments&#39;</span>;
<a name="l00171"></a>00171             <span class="stringliteral">&#39;complexCompartments&#39;</span>;
<a name="l00172"></a>00172                     };
<a name="l00173"></a>00173         fittedConstantNames__      = {}; %names of fitted constant <a class="code" href="json_parse_8m.html#a6e4256f69b751d3d7ec4ea52ff5a8dee">properties</a>
<a name="l00174"></a>00174         localStateNames__          = {   %names of simulation state <a class="code" href="json_parse_8m.html#a6e4256f69b751d3d7ec4ea52ff5a8dee">properties</a> redundant with timecourses in <span class="keyword">this</span> or other processes or the simulation
<a name="l00175"></a>00175             <span class="stringliteral">&#39;unfoldedMonomers&#39;</span>;
<a name="l00176"></a>00176             <span class="stringliteral">&#39;unfoldedComplexs&#39;</span>;
<a name="l00177"></a>00177             <span class="stringliteral">&#39;foldedMonomers&#39;</span>;
<a name="l00178"></a>00178             <span class="stringliteral">&#39;foldedComplexs&#39;</span>};
<a name="l00179"></a>00179     end
<a name="l00180"></a>00180 
<a name="l00181"></a>00181     %IDs, names, and local indices
<a name="l00182"></a>00182     <a class="code" href="json_parse_8m.html#a6e4256f69b751d3d7ec4ea52ff5a8dee">properties</a>
<a name="l00183"></a>00183         stimuliWholeCellModelIDs = {}; %whole cell model IDs of stimuli
<a name="l00184"></a>00184 
<a name="l00185"></a>00185         substrateWholeCellModelIDs = { %whole cell model IDs of substrates
<a name="l00186"></a>00186             <span class="stringliteral">&#39;ATP&#39;</span>;<span class="stringliteral">&#39;ADP&#39;</span>;<span class="stringliteral">&#39;PI&#39;</span>;<span class="stringliteral">&#39;H2O&#39;</span>;<span class="charliteral">&#39;H&#39;</span>};
<a name="l00187"></a>00187         substrateIndexs_adp       %index within substrates of ADP
<a name="l00188"></a>00188         substrateIndexs_atp       %index within substrates of ATP
<a name="l00189"></a>00189         substrateIndexs_fe2       %index within substrates of Fe+2 ion
<a name="l00190"></a>00190         substrateIndexs_hydrogen  %index within substrates of hydrogen
<a name="l00191"></a>00191         substrateIndexs_water     %index within substrates of water
<a name="l00192"></a>00192         substrateIndexs_k         %index within substrates of K+ ion
<a name="l00193"></a>00193         substrateIndexs_mg        %index within substrates of Mg2+ ion
<a name="l00194"></a>00194         substrateIndexs_mn        %index within substrates of Mn2+ ion
<a name="l00195"></a>00195         substrateIndexs_sodium    %index within substrates of Na+ ion
<a name="l00196"></a>00196         substrateIndexs_phosphate %index within substrates of inorganic phosphate
<a name="l00197"></a>00197         substrateIndexs_zinc      %index within substrates of Zn2+ ion
<a name="l00198"></a>00198 
<a name="l00199"></a>00199         enzymeWholeCellModelIDs = {     %enzyme whole cell model ids
<a name="l00200"></a>00200             <span class="stringliteral">&#39;MG_238_MONOMER&#39;</span>;           %trigger factor
<a name="l00201"></a>00201             <span class="stringliteral">&#39;MG_305_MONOMER&#39;</span>;           %chaperone
<a name="l00202"></a>00202             <span class="stringliteral">&#39;MG_019_DIMER&#39;</span>;             %chaperone protein DnaJ
<a name="l00203"></a>00203             <span class="stringliteral">&#39;MG_201_DIMER&#39;</span>;             %co-chaperone GrpE
<a name="l00204"></a>00204             &#39;MG_392_393_21MER&#39;};        %GroEL-GroES chaperonin complex
<a name="l00205"></a>00205         enzymeIndexs_triggerFactor      %index of trigger factor within enzymeWholeCellModelIDs
<a name="l00206"></a>00206         enzymeIndexs_dnaK               %index of DnaK within enzymeWholeCellModelIDs
<a name="l00207"></a>00207         enzymeIndexs_dnaJ               %index of DnaJ within enzymeWholeCellModelIDs
<a name="l00208"></a>00208         enzymeIndexs_grpE               %index of GrpE within enzymeWholeCellModelIDs
<a name="l00209"></a>00209         enzymeIndexs_groELES            %index of GroELES within enzymeWholeCellModelIDs
<a name="l00210"></a>00210 
<a name="l00211"></a>00211         unfoldedMonomerWholeCellModelIDs %whole cell model IDs of unfolded monomers
<a name="l00212"></a>00212         unfoldedComplexWholeCellModelIDs %whole cell model IDs of unfolded complexes
<a name="l00213"></a>00213         foldedMonomerWholeCellModelIDs   %whole cell model IDs of folded monomers
<a name="l00214"></a>00214         foldedComplexWholeCellModelIDs   %whole cell model IDs of folded complexes
<a name="l00215"></a>00215         
<a name="l00216"></a>00216         unfoldedMonomerCompartmentIndexs %indices of unfolded monomers within this.monomer.counts
<a name="l00217"></a>00217         unfoldedComplexCompartmentIndexs %indices of unfolded complexs within this.complex.counts
<a name="l00218"></a>00218         foldedMonomerCompartmentIndexs   %indices of folded monomers within this.monomer.counts
<a name="l00219"></a>00219         foldedComplexCompartmentIndexs   %indices of folded complexs within this.complex.counts
<a name="l00220"></a>00220                 
<a name="l00221"></a>00221         monomerIndexs_folding            %indices of monomers that actively fold within unfoldedMonomers
<a name="l00222"></a>00222         complexIndexs_folding            %indices of complexs that actively fold within unfoldedComplexs
<a name="l00223"></a>00223         monomerIndexs_notFolding         %indices of monomers that don&#39;t actively fold within unfoldedMonomers
<a name="l00224"></a>00224         complexIndexs_notFolding         %indices of complexs that don&#39;t actively fold within unfoldedComplexs
<a name="l00225"></a>00225         
<a name="l00226"></a>00226         monomerComplexIndexs_folded      %indices of monomers and complexs that actively fold within proteinProstheticGroupMatrix
<a name="l00227"></a>00227         
<a name="l00228"></a>00228         speciesIndexs_monomers           %indices of monomers that actively fold within speciesReactantByproductMatrix
<a name="l00229"></a>00229         speciesIndexs_complexs           %indices of complexs that actively fold within speciesReactantByproductMatrix
<a name="l00230"></a>00230     end
<a name="l00231"></a>00231     
<a name="l00232"></a>00232     %fixed biological constants
<a name="l00233"></a>00233     <a class="code" href="json_parse_8m.html#a6e4256f69b751d3d7ec4ea52ff5a8dee">properties</a>
<a name="l00234"></a>00234         proteinProstheticGroupMatrix     %proteins X metabolites X compartments
<a name="l00235"></a>00235         proteinChaperoneMatrix           %proteins X proteins (chaperone) X compartments
<a name="l00236"></a>00236 
<a name="l00237"></a>00237         monomerCompartments              %compartment indices with simulation.compartmentWholeCellModelIDs of monomers
<a name="l00238"></a>00238         complexCompartments              %compartment indices with simulation.compartmentWholeCellModelIDs of complexes
<a name="l00239"></a>00239                             
<a name="l00240"></a>00240         speciesReactantByproductMatrix   %stoichiometry of susbtrates, enzymes, monomers and complexs in reactions necessary to fold each protein: [Protein foldings] X [subtrates; enzymes; unfolded monomers; unfolded complexs]
<a name="l00241"></a>00241         speciesReactantMatrix            %reactant stoichiometry of susbtrates, enzymes, monomers and complexs in reactions necessary to fold each protein: [<a class="code" href="class_protein.html">Protein</a> foldings] X [subtrates; enzymes; unfolded monomers; unfolded complexs]
<a name="l00242"></a>00242     end
<a name="l00243"></a>00243 
<a name="l00244"></a>00244     %global state (stored locally <span class="keywordflow">for</span> convenience)
<a name="l00245"></a>00245     <a class="code" href="json_parse_8m.html#a6e4256f69b751d3d7ec4ea52ff5a8dee">properties</a>
<a name="l00246"></a>00246         unfoldedMonomers                 %amount of unfolded monomers
<a name="l00247"></a>00247         unfoldedComplexs                 %amount of unfolded complexes
<a name="l00248"></a>00248         foldedMonomers                   %amount of folded monomers
<a name="l00249"></a>00249         foldedComplexs                   %amount of folded complexes
<a name="l00250"></a>00250     end
<a name="l00251"></a>00251 
<a name="l00252"></a>00252     %constructor
<a name="l00253"></a>00253     methods
<a name="l00254"></a>00254         <span class="keyword">function</span> <span class="keyword">this</span> = ProteinFolding(wholeCellModelID, name)
<a name="l00255"></a>00255             this = this@edu.stanford.covert.cell.sim.<a class="code" href="class_process.html">Process</a>(wholeCellModelID, name);
<a name="l00256"></a>00256         end
<a name="l00257"></a>00257     end
<a name="l00258"></a>00258 
<a name="l00259"></a>00259     %communication between process/simulation
<a name="l00260"></a>00260     methods
<a name="l00261"></a>00261         %initialize constants
<a name="l00262"></a>00262         function initializeConstants(this, knowledgeBase, simulation, varargin)
<a name="l00263"></a>00263             %prosthetic groups
<a name="l00264"></a>00264             this.proteinProstheticGroupMatrix = knowledgeBase.proteinProstheticGroupMatrix(:, :, this.compartment.cytosolIndexs);
<a name="l00265"></a>00265             this.proteinProstheticGroupMatrix(this.proteinProstheticGroupMatrix==-1)=1;
<a name="l00266"></a>00266 
<a name="l00267"></a>00267             this.substrateWholeCellModelIDs = unique([...
<a name="l00268"></a>00268                 this.substrateWholeCellModelIDs;
<a name="l00269"></a>00269                 this.metabolite.wholeCellModelIDs(sum(this.proteinProstheticGroupMatrix)&gt;0)]);
<a name="l00270"></a>00270 
<a name="l00271"></a>00271             this.substrateIndexs_adp       = this.substrateIndexs({<span class="stringliteral">&#39;ADP&#39;</span>});
<a name="l00272"></a>00272             this.substrateIndexs_atp       = this.substrateIndexs({<span class="stringliteral">&#39;ATP&#39;</span>});
<a name="l00273"></a>00273             this.substrateIndexs_fe2       = this.substrateIndexs({<span class="stringliteral">&#39;FE2&#39;</span>});
<a name="l00274"></a>00274             this.substrateIndexs_hydrogen  = this.substrateIndexs({<span class="charliteral">&#39;H&#39;</span>});
<a name="l00275"></a>00275             this.substrateIndexs_water     = this.substrateIndexs({<span class="stringliteral">&#39;H2O&#39;</span>});
<a name="l00276"></a>00276             this.substrateIndexs_k         = this.substrateIndexs({<span class="charliteral">&#39;K&#39;</span>});
<a name="l00277"></a>00277             this.substrateIndexs_mg        = this.substrateIndexs({<span class="stringliteral">&#39;MG&#39;</span>});
<a name="l00278"></a>00278             this.substrateIndexs_mn        = this.substrateIndexs({<span class="stringliteral">&#39;MN&#39;</span>});
<a name="l00279"></a>00279             this.substrateIndexs_sodium    = this.substrateIndexs({<span class="stringliteral">&#39;NA&#39;</span>});
<a name="l00280"></a>00280             this.substrateIndexs_phosphate = this.substrateIndexs({<span class="stringliteral">&#39;PI&#39;</span>});
<a name="l00281"></a>00281             this.substrateIndexs_zinc      = this.substrateIndexs({<span class="stringliteral">&#39;ZN&#39;</span>});
<a name="l00282"></a>00282 
<a name="l00283"></a>00283             %chaperones
<a name="l00284"></a>00284             numMonomers = length(this.monomer.matureIndexs);
<a name="l00285"></a>00285             proteinChaperone = knowledgeBase.proteinChaperoneMatrix;
<a name="l00286"></a>00286             proteinMonomerChaperoneIndexs = logical(sum(sum(proteinChaperone(1:numMonomers,:,:),3),2));
<a name="l00287"></a>00287             proteinComplexChaperoneIndexs = logical(sum(sum(proteinChaperone(numMonomers+1:end,:,:),3),2));
<a name="l00288"></a>00288             this.enzymeWholeCellModelIDs = unique([
<a name="l00289"></a>00289                 this.enzymeWholeCellModelIDs
<a name="l00290"></a>00290                 this.monomer.wholeCellModelIDs{<span class="keyword">this</span>.monomer.matureIndexs(proteinMonomerChaperoneIndexs)}
<a name="l00291"></a>00291                 this.complex.wholeCellModelIDs{this.complex.matureIndexs(proteinComplexChaperoneIndexs)}
<a name="l00292"></a>00292                 ]);
<a name="l00293"></a>00293 
<a name="l00294"></a>00294             this.enzymeIndexs_triggerFactor = this.enzymeIndexs({<span class="stringliteral">&#39;MG_238_MONOMER&#39;</span>});
<a name="l00295"></a>00295             this.enzymeIndexs_dnaK          = this.enzymeIndexs({<span class="stringliteral">&#39;MG_305_MONOMER&#39;</span>});
<a name="l00296"></a>00296             this.enzymeIndexs_dnaJ          = this.enzymeIndexs({<span class="stringliteral">&#39;MG_019_DIMER&#39;</span>});
<a name="l00297"></a>00297             this.enzymeIndexs_grpE          = this.enzymeIndexs({<span class="stringliteral">&#39;MG_201_DIMER&#39;</span>});
<a name="l00298"></a>00298             this.enzymeIndexs_groELES       = this.enzymeIndexs({<span class="stringliteral">&#39;MG_392_393_21MER&#39;</span>});
<a name="l00299"></a>00299 
<a name="l00300"></a>00300             %call super <span class="keyword">class </span>method to get metabolite, enzyme indexs
<a name="l00301"></a>00301             this.initializeConstants@edu.stanford.covert.cell.sim.Process(...
<a name="l00302"></a>00302                 knowledgeBase, simulation, varargin{:});
<a name="l00303"></a>00303 
<a name="l00304"></a>00304             %prosthetic groups
<a name="l00305"></a>00305             this.proteinProstheticGroupMatrix = ...
<a name="l00306"></a>00306                 this.proteinProstheticGroupMatrix(:,this.substrateMetaboliteGlobalIndexs);
<a name="l00307"></a>00307 
<a name="l00308"></a>00308             %chaperones
<a name="l00309"></a>00309             numEnzymes    = length(this.enzymeWholeCellModelIDs);
<a name="l00310"></a>00310             numSubstrates = size(proteinChaperone,2);            
<a name="l00311"></a>00311 
<a name="l00312"></a>00312             enzymeCompartmentIndexs = zeros(size(this.enzymeWholeCellModelIDs));
<a name="l00313"></a>00313             enzymeCompartmentIndexs(this.enzymeMonomerLocalIndexs) = this.enzymeMonomerCompartmentIndexs;
<a name="l00314"></a>00314             enzymeCompartmentIndexs(this.enzymeComplexLocalIndexs) = this.enzymeComplexCompartmentIndexs;
<a name="l00315"></a>00315             
<a name="l00316"></a>00316             enzymeGlobalIndexs = this.enzymeGlobalIndexs;
<a name="l00317"></a>00317             enzymeGlobalIndexs(this.enzymeComplexLocalIndexs) = ...
<a name="l00318"></a>00318                 enzymeGlobalIndexs(this.enzymeComplexLocalIndexs) + ...
<a name="l00319"></a>00319                 numMonomers;
<a name="l00320"></a>00320             this.proteinChaperoneMatrix = proteinChaperone(sub2ind(...
<a name="l00321"></a>00321                 size(proteinChaperone),...
<a name="l00322"></a>00322                 repmat(enzymeGlobalIndexs<span class="stringliteral">&#39;,numSubstrates,1), ...</span>
<a name="l00323"></a>00323 <span class="stringliteral">                repmat((1:numSubstrates)&#39;</span>,[1 numEnzymes]),...
<a name="l00324"></a>00324                 repmat(enzymeCompartmentIndexs<span class="stringliteral">&#39;,numSubstrates,1)));</span>
<a name="l00325"></a>00325 <span class="stringliteral"></span>
<a name="l00326"></a>00326 <span class="stringliteral">            this.proteinChaperoneMatrix(1:numMonomers,this.enzymeIndexs_triggerFactor) = 1;</span>
<a name="l00327"></a>00327 <span class="stringliteral">            this.proteinChaperoneMatrix(:,this.enzymeIndexs_dnaJ) = this.proteinChaperoneMatrix(:,this.enzymeIndexs_dnaK);</span>
<a name="l00328"></a>00328 <span class="stringliteral">            this.proteinChaperoneMatrix(:,this.enzymeIndexs_grpE) = this.proteinChaperoneMatrix(:,this.enzymeIndexs_dnaK);</span>
<a name="l00329"></a>00329 <span class="stringliteral"></span>
<a name="l00330"></a>00330 <span class="stringliteral">            %monomers</span>
<a name="l00331"></a>00331 <span class="stringliteral">            this.unfoldedMonomerWholeCellModelIDs = this.monomer.wholeCellModelIDs(this.monomer.processedIIIndexs);</span>
<a name="l00332"></a>00332 <span class="stringliteral">            this.foldedMonomerWholeCellModelIDs   = this.monomer.wholeCellModelIDs(this.monomer.foldedIndexs);</span>
<a name="l00333"></a>00333 <span class="stringliteral">            this.monomerCompartments = double(knowledgeBase.proteinMonomerCompartments);</span>
<a name="l00334"></a>00334 <span class="stringliteral">            this.monomerCompartments(this.monomerCompartments == this.compartment.terminalOrganelleCytosolIndexs)  = this.compartment.cytosolIndexs;</span>
<a name="l00335"></a>00335 <span class="stringliteral">            this.monomerCompartments(this.monomerCompartments == this.compartment.terminalOrganelleMembraneIndexs) = this.compartment.membraneIndexs;</span>
<a name="l00336"></a>00336 <span class="stringliteral">            this.unfoldedMonomerCompartmentIndexs = sub2ind(...</span>
<a name="l00337"></a>00337 <span class="stringliteral">                [numel(this.monomer.wholeCellModelIDs) this.compartment.count], ...</span>
<a name="l00338"></a>00338 <span class="stringliteral">                this.monomer.processedIIIndexs, ...</span>
<a name="l00339"></a>00339 <span class="stringliteral">                this.monomerCompartments);</span>
<a name="l00340"></a>00340 <span class="stringliteral">            this.foldedMonomerCompartmentIndexs = sub2ind(...</span>
<a name="l00341"></a>00341 <span class="stringliteral">                [numel(this.monomer.wholeCellModelIDs) this.compartment.count], ...</span>
<a name="l00342"></a>00342 <span class="stringliteral">                this.monomer.foldedIndexs, ...</span>
<a name="l00343"></a>00343 <span class="stringliteral">                this.monomerCompartments);</span>
<a name="l00344"></a>00344 <span class="stringliteral"></span>
<a name="l00345"></a>00345 <span class="stringliteral">            %complexs</span>
<a name="l00346"></a>00346 <span class="stringliteral">            this.unfoldedComplexWholeCellModelIDs = this.complex.wholeCellModelIDs(this.complex.nascentIndexs);</span>
<a name="l00347"></a>00347 <span class="stringliteral">            this.foldedComplexWholeCellModelIDs   = this.complex.wholeCellModelIDs(this.complex.matureIndexs);</span>
<a name="l00348"></a>00348 <span class="stringliteral">            this.complexCompartments = double(knowledgeBase.proteinComplexCompartments);</span>
<a name="l00349"></a>00349 <span class="stringliteral">            this.unfoldedComplexCompartmentIndexs = sub2ind(...</span>
<a name="l00350"></a>00350 <span class="stringliteral">                [numel(this.complex.wholeCellModelIDs) this.compartment.count], ...</span>
<a name="l00351"></a>00351 <span class="stringliteral">                this.complex.nascentIndexs, ...</span>
<a name="l00352"></a>00352 <span class="stringliteral">                this.complexCompartments);</span>
<a name="l00353"></a>00353 <span class="stringliteral">            this.foldedComplexCompartmentIndexs = sub2ind(...</span>
<a name="l00354"></a>00354 <span class="stringliteral">                [numel(this.complex.wholeCellModelIDs) this.compartment.count], ...</span>
<a name="l00355"></a>00355 <span class="stringliteral">                this.complex.matureIndexs, ...</span>
<a name="l00356"></a>00356 <span class="stringliteral">                this.complexCompartments);</span>
<a name="l00357"></a>00357 <span class="stringliteral">            </span>
<a name="l00358"></a>00358 <span class="stringliteral">            this.initializeSpeciesNetwork();</span>
<a name="l00359"></a>00359 <span class="stringliteral">        end</span>
<a name="l00360"></a>00360 <span class="stringliteral">        </span>
<a name="l00361"></a>00361 <span class="stringliteral">        function initializeSpeciesNetwork(this)</span>
<a name="l00362"></a>00362 <span class="stringliteral">            this.monomerComplexIndexs_folded = find(sum(this.proteinProstheticGroupMatrix,2)&gt; 0 | sum(this.proteinChaperoneMatrix,2)&gt; 0);</span>
<a name="l00363"></a>00363 <span class="stringliteral">            </span>
<a name="l00364"></a>00364 <span class="stringliteral">            numMonomers   = length(this.unfoldedMonomerWholeCellModelIDs);</span>
<a name="l00365"></a>00365 <span class="stringliteral">            numComplexs   = length(this.unfoldedComplexWholeCellModelIDs);            </span>
<a name="l00366"></a>00366 <span class="stringliteral">            </span>
<a name="l00367"></a>00367 <span class="stringliteral">            [~, this.monomerIndexs_folding, this.speciesIndexs_monomers] = intersect(1:numMonomers, this.monomerComplexIndexs_folded); %use of intersect is safe because the inputs are already sorted</span>
<a name="l00368"></a>00368 <span class="stringliteral">            [~, this.complexIndexs_folding, this.speciesIndexs_complexs] = intersect((1:numComplexs) + numMonomers, this.monomerComplexIndexs_folded);</span>
<a name="l00369"></a>00369 <span class="stringliteral">            [~, this.monomerIndexs_notFolding] = setdiff(1:numMonomers,                 this.monomerComplexIndexs_folded);</span>
<a name="l00370"></a>00370 <span class="stringliteral">            [~, this.complexIndexs_notFolding] = setdiff((1:numComplexs) + numMonomers, this.monomerComplexIndexs_folded);</span>
<a name="l00371"></a>00371 <span class="stringliteral">            </span>
<a name="l00372"></a>00372 <span class="stringliteral">            this.speciesReactantByproductMatrix = [</span>
<a name="l00373"></a>00373 <span class="stringliteral">                this.proteinProstheticGroupMatrix(this.monomerComplexIndexs_folded,:)...</span>
<a name="l00374"></a>00374 <span class="stringliteral">                this.proteinChaperoneMatrix(this.monomerComplexIndexs_folded,:)...</span>
<a name="l00375"></a>00375 <span class="stringliteral">                eye(length(this.monomerComplexIndexs_folded))];</span>
<a name="l00376"></a>00376 <span class="stringliteral">            this.speciesReactantMatrix = max(0, this.speciesReactantByproductMatrix);</span>
<a name="l00377"></a>00377 <span class="stringliteral">        end</span>
<a name="l00378"></a>00378 <span class="stringliteral"></span>
<a name="l00379"></a>00379 <span class="stringliteral">        %retrieve state from simulation</span>
<a name="l00380"></a>00380 <span class="stringliteral">        function copyFromState(this)</span>
<a name="l00381"></a>00381 <span class="stringliteral">            this.copyFromState@edu.stanford.covert.cell.sim.Process();</span>
<a name="l00382"></a>00382 <span class="stringliteral"></span>
<a name="l00383"></a>00383 <span class="stringliteral">            numTimePoints = size(this.unfoldedMonomers, 3);</span>
<a name="l00384"></a>00384 <span class="stringliteral">            </span>
<a name="l00385"></a>00385 <span class="stringliteral">            if numTimePoints == 1</span>
<a name="l00386"></a>00386 <span class="stringliteral">                this.unfoldedMonomers = this.monomer.counts(this.unfoldedMonomerCompartmentIndexs);</span>
<a name="l00387"></a>00387 <span class="stringliteral">                this.unfoldedComplexs = this.complex.counts(this.unfoldedComplexCompartmentIndexs);</span>
<a name="l00388"></a>00388 <span class="stringliteral">                this.foldedMonomers = this.monomer.counts(this.foldedMonomerCompartmentIndexs);</span>
<a name="l00389"></a>00389 <span class="stringliteral">                this.foldedComplexs = this.complex.counts(this.foldedComplexCompartmentIndexs);</span>
<a name="l00390"></a>00390 <span class="stringliteral">            else</span>
<a name="l00391"></a>00391 <span class="stringliteral">                this.unfoldedMonomers = this.monomer.counts(sub2ind(...</span>
<a name="l00392"></a>00392 <span class="stringliteral">                    size(this.monomer.counts), ...</span>
<a name="l00393"></a>00393 <span class="stringliteral">                    repmat(this.monomer.processedIIIndexs, [1 1 numTimePoints]), ...</span>
<a name="l00394"></a>00394 <span class="stringliteral">                    repmat(this.monomerCompartments,[1 1 numTimePoints]),...</span>
<a name="l00395"></a>00395 <span class="stringliteral">                    permute(repmat((1:numTimePoints)&#39;</span>,[1 size(this.foldedMonomers,1) 1]),[2 3 1])));
<a name="l00396"></a>00396                 this.unfoldedComplexs = this.complex.counts(sub2ind(...
<a name="l00397"></a>00397                     size(this.complex.counts), ...
<a name="l00398"></a>00398                     repmat(this.complex.nascentIndexs, [1 1 numTimePoints]), ...
<a name="l00399"></a>00399                     repmat(this.complexCompartments,[1 1 numTimePoints]),...
<a name="l00400"></a>00400                     permute(repmat((1:numTimePoints)<span class="stringliteral">&#39;,[1 size(this.foldedComplexs,1) 1]),[2 3 1])));</span>
<a name="l00401"></a>00401 <span class="stringliteral">                this.foldedMonomers = this.monomer.counts(sub2ind(...</span>
<a name="l00402"></a>00402 <span class="stringliteral">                    size(this.monomer.counts), ...</span>
<a name="l00403"></a>00403 <span class="stringliteral">                    repmat(this.monomer.foldedIndexs, [1 1 numTimePoints]), ...</span>
<a name="l00404"></a>00404 <span class="stringliteral">                    repmat(this.monomerCompartments,[1 1 numTimePoints]),...</span>
<a name="l00405"></a>00405 <span class="stringliteral">                    permute(repmat((1:numTimePoints)&#39;</span>,[1 size(this.foldedMonomers,1) 1]),[2 3 1])));
<a name="l00406"></a>00406                 this.foldedComplexs = this.complex.counts(sub2ind(...
<a name="l00407"></a>00407                     size(this.complex.counts), ...
<a name="l00408"></a>00408                     repmat(this.complex.matureIndexs, [1 1 numTimePoints]), ...
<a name="l00409"></a>00409                     repmat(this.complexCompartments,[1 1 numTimePoints]),...
<a name="l00410"></a>00410                     permute(repmat((1:numTimePoints)<span class="stringliteral">&#39;,[1 size(this.foldedComplexs,1) 1]),[2 3 1])));</span>
<a name="l00411"></a>00411 <span class="stringliteral">            end</span>
<a name="l00412"></a>00412 <span class="stringliteral">        end</span>
<a name="l00413"></a>00413 <span class="stringliteral"></span>
<a name="l00414"></a>00414 <span class="stringliteral">        %send state to simulation</span>
<a name="l00415"></a>00415 <span class="stringliteral">        function copyToState(this)</span>
<a name="l00416"></a>00416 <span class="stringliteral">            this.copyToState@edu.stanford.covert.cell.sim.Process();</span>
<a name="l00417"></a>00417 <span class="stringliteral">            </span>
<a name="l00418"></a>00418 <span class="stringliteral">            numTimePoints = size(this.unfoldedMonomers, 3);</span>
<a name="l00419"></a>00419 <span class="stringliteral">            </span>
<a name="l00420"></a>00420 <span class="stringliteral">            if numTimePoints == 1</span>
<a name="l00421"></a>00421 <span class="stringliteral">                this.monomer.counts(this.unfoldedMonomerCompartmentIndexs) = this.unfoldedMonomers;</span>
<a name="l00422"></a>00422 <span class="stringliteral">                this.complex.counts(this.unfoldedComplexCompartmentIndexs) = this.unfoldedComplexs;</span>
<a name="l00423"></a>00423 <span class="stringliteral">                this.monomer.counts(this.foldedMonomerCompartmentIndexs) = this.foldedMonomers;</span>
<a name="l00424"></a>00424 <span class="stringliteral">                this.complex.counts(this.foldedComplexCompartmentIndexs) = this.foldedComplexs;</span>
<a name="l00425"></a>00425 <span class="stringliteral">            else                </span>
<a name="l00426"></a>00426 <span class="stringliteral">                this.monomer.counts(sub2ind(...</span>
<a name="l00427"></a>00427 <span class="stringliteral">                    size(this.monomer.counts), ...</span>
<a name="l00428"></a>00428 <span class="stringliteral">                    repmat(this.monomer.processedIIIndexs, [1 1 numTimePoints]), ...</span>
<a name="l00429"></a>00429 <span class="stringliteral">                    repmat(this.monomerCompartments,[1 1 numTimePoints]),...</span>
<a name="l00430"></a>00430 <span class="stringliteral">                    permute(repmat((1:numTimePoints)&#39;</span>,[1 size(this.foldedMonomers,1) 1]),[2 3 1]))) = ...
<a name="l00431"></a>00431                     this.unfoldedMonomers;
<a name="l00432"></a>00432                 this.complex.counts(sub2ind(...
<a name="l00433"></a>00433                     size(this.complex.counts), ...
<a name="l00434"></a>00434                     repmat(this.complex.nascentIndexs, [1 1 numTimePoints]), ...
<a name="l00435"></a>00435                     repmat(this.complexCompartments,[1 1 numTimePoints]),...
<a name="l00436"></a>00436                     permute(repmat((1:numTimePoints)<span class="stringliteral">&#39;,[1 size(this.foldedComplexs,1) 1]),[2 3 1]))) = ...</span>
<a name="l00437"></a>00437 <span class="stringliteral">                    this.unfoldedComplexs;</span>
<a name="l00438"></a>00438 <span class="stringliteral">                this.monomer.counts(sub2ind(...</span>
<a name="l00439"></a>00439 <span class="stringliteral">                    size(this.monomer.counts), ...</span>
<a name="l00440"></a>00440 <span class="stringliteral">                    repmat(this.monomer.foldedIndexs, [1 1 numTimePoints]), ...</span>
<a name="l00441"></a>00441 <span class="stringliteral">                    repmat(this.monomerCompartments,[1 1 numTimePoints]),...</span>
<a name="l00442"></a>00442 <span class="stringliteral">                    permute(repmat((1:numTimePoints)&#39;</span>,[1 size(this.foldedMonomers,1) 1]),[2 3 1]))) = ...
<a name="l00443"></a>00443                     this.foldedMonomers;
<a name="l00444"></a>00444                 this.complex.counts(sub2ind(...
<a name="l00445"></a>00445                     size(this.complex.counts), ...
<a name="l00446"></a>00446                     repmat(this.complex.matureIndexs, [1 1 numTimePoints]), ...
<a name="l00447"></a>00447                     repmat(this.complexCompartments,[1 1 numTimePoints]),...
<a name="l00448"></a>00448                     permute(repmat((1:numTimePoints)<span class="stringliteral">&#39;,[1 size(this.foldedComplexs,1) 1]),[2 3 1]))) = ...</span>
<a name="l00449"></a>00449 <span class="stringliteral">                    this.foldedComplexs;</span>
<a name="l00450"></a>00450 <span class="stringliteral">            end</span>
<a name="l00451"></a>00451 <span class="stringliteral">        end</span>
<a name="l00452"></a>00452 <span class="stringliteral">    end</span>
<a name="l00453"></a>00453 <span class="stringliteral"></span>
<a name="l00454"></a>00454 <span class="stringliteral">    %allocate memory for state</span>
<a name="l00455"></a>00455 <span class="stringliteral">    methods</span>
<a name="l00456"></a>00456 <span class="stringliteral">        function allocateMemoryForState(this, numTimePoints)</span>
<a name="l00457"></a>00457 <span class="stringliteral">            this.allocateMemoryForState@edu.stanford.covert.cell.sim.Process(numTimePoints);</span>
<a name="l00458"></a>00458 <span class="stringliteral"></span>
<a name="l00459"></a>00459 <span class="stringliteral">            this.unfoldedMonomers = zeros(length(this.unfoldedMonomerWholeCellModelIDs), 1, numTimePoints);</span>
<a name="l00460"></a>00460 <span class="stringliteral">            this.unfoldedComplexs = zeros(length(this.unfoldedComplexWholeCellModelIDs), 1, numTimePoints);</span>
<a name="l00461"></a>00461 <span class="stringliteral">            this.foldedMonomers   = zeros(length(this.foldedMonomerWholeCellModelIDs),   1, numTimePoints);</span>
<a name="l00462"></a>00462 <span class="stringliteral">            this.foldedComplexs   = zeros(length(this.foldedComplexWholeCellModelIDs),   1, numTimePoints);</span>
<a name="l00463"></a>00463 <span class="stringliteral">        end</span>
<a name="l00464"></a>00464 <span class="stringliteral">    end</span>
<a name="l00465"></a>00465 <span class="stringliteral"></span>
<a name="l00466"></a>00466 <span class="stringliteral">    %model</span>
<a name="l00467"></a>00467 <span class="stringliteral">    methods</span>
<a name="l00468"></a>00468 <span class="stringliteral">        %Calculate</span>
<a name="l00469"></a>00469 <span class="stringliteral">        %- contribution to FBA objective</span>
<a name="l00470"></a>00470 <span class="stringliteral">        %- minimum expression consistent with cell cycle length</span>
<a name="l00471"></a>00471 <span class="stringliteral">        function [bmProd, byProd, minEnzExp, maxEnzExp] = calcResourceRequirements_LifeCycle(this, ~, states)</span>
<a name="l00472"></a>00472 <span class="stringliteral">            %% substrate and byproducts</span>
<a name="l00473"></a>00473 <span class="stringliteral">            %prosthetic groups that complex with proteins</span>
<a name="l00474"></a>00474 <span class="stringliteral">            bmProd = ...</span>
<a name="l00475"></a>00475 <span class="stringliteral">                this.proteinProstheticGroupMatrix&#39;</span> * ...
<a name="l00476"></a>00476                 [states.monomerProductions; states.complexProductions];
<a name="l00477"></a>00477             
<a name="l00478"></a>00478             %no byproducts
<a name="l00479"></a>00479             byProd = zeros(size(this.substrateWholeCellModelIDs));
<a name="l00480"></a>00480             
<a name="l00481"></a>00481             %% enzymes: at least one of each enzyme
<a name="l00482"></a>00482             minEnzExp = zeros(size(this.enzymeWholeCellModelIDs));
<a name="l00483"></a>00483             minEnzExp([
<a name="l00484"></a>00484                 this.enzymeIndexs_triggerFactor;
<a name="l00485"></a>00485                 this.enzymeIndexs_dnaK;
<a name="l00486"></a>00486                 this.enzymeIndexs_dnaJ;
<a name="l00487"></a>00487                 this.enzymeIndexs_grpE;
<a name="l00488"></a>00488                 this.enzymeIndexs_groELES]) = 2;
<a name="l00489"></a>00489             maxEnzExp = Inf(size(this.enzymeWholeCellModelIDs));
<a name="l00490"></a>00490         end
<a name="l00491"></a>00491         
<a name="l00492"></a>00492         %initialization
<a name="l00493"></a>00493         %- monomers initialized to mature/bound/inactivated state
<a name="l00494"></a>00494         %  by simulation initializeState method
<a name="l00495"></a>00495         %- complexs initialized by
<a name="l00496"></a>00496         %  1. Macromolecular complexation
<a name="l00497"></a>00497         %  2. Ribosome assembly
<a name="l00498"></a>00498         %  3. <a class="code" href="class_protein.html">Protein</a> folding
<a name="l00499"></a>00499         %  4. <a class="code" href="class_protein.html">Protein</a> activation
<a name="l00500"></a>00500         %  Here we push complexs to folded state
<a name="l00501"></a>00501         function initializeState(this)
<a name="l00502"></a>00502             this.foldedComplexs = this.foldedComplexs + this.unfoldedComplexs;
<a name="l00503"></a>00503             this.unfoldedComplexs(:) = 0;
<a name="l00504"></a>00504         end
<a name="l00505"></a>00505 
<a name="l00506"></a>00506         %resource requirements
<a name="l00507"></a>00507         <span class="keyword">function</span> result = calcResourceRequirements_Current(<span class="keyword">this</span>)
<a name="l00508"></a>00508             if all(this.enzymes)
<a name="l00509"></a>00509                 result = this.proteinProstheticGroupMatrix&#39; * ...
<a name="l00510"></a>00510                     [this.unfoldedMonomers; this.unfoldedComplexs];
<a name="l00511"></a>00511             else
<a name="l00512"></a>00512                 result = this.proteinProstheticGroupMatrix&#39; * ...
<a name="l00513"></a>00513                     ((min(this.enzymes(:, ones(size(this.proteinChaperoneMatrix, 1), 1))&#39; ./ this.proteinChaperoneMatrix, [], 2) &gt; 0) .* ...
<a name="l00514"></a>00514                     [this.unfoldedMonomers; this.unfoldedComplexs]);
<a name="l00515"></a>00515             end
<a name="l00516"></a>00516         end
<a name="l00517"></a>00517 
<a name="l00518"></a>00518         %simulation
<a name="l00519"></a>00519         function evolveState(this)
<a name="l00520"></a>00520             %update monomers that don&#39;t require chaperones or prosthetic groups
<a name="l00521"></a>00521             this.foldedComplexs(this.complexIndexs_notFolding) = ...
<a name="l00522"></a>00522                 this.foldedComplexs(this.complexIndexs_notFolding) + ...
<a name="l00523"></a>00523                 this.unfoldedComplexs(this.complexIndexs_notFolding);
<a name="l00524"></a>00524             this.unfoldedComplexs(this.complexIndexs_notFolding) = 0;
<a name="l00525"></a>00525             
<a name="l00526"></a>00526             %stop early if no proteins to fold
<a name="l00527"></a>00527             if ~any(this.unfoldedMonomers) &amp;&amp; ~any(this.unfoldedComplexs)
<a name="l00528"></a>00528                 return;
<a name="l00529"></a>00529             end
<a name="l00530"></a>00530                         
<a name="l00531"></a>00531             numSubstrates = length(this.monomerComplexIndexs_folded);
<a name="l00532"></a>00532             
<a name="l00533"></a>00533             species = max(0, [
<a name="l00534"></a>00534                 this.substrates;
<a name="l00535"></a>00535                 this.enzymes * Inf;
<a name="l00536"></a>00536                 this.unfoldedMonomers;
<a name="l00537"></a>00537                 this.unfoldedComplexs(this.complexIndexs_folding)]&#39;);
<a name="l00538"></a>00538 
<a name="l00539"></a>00539             fluxes = zeros(size(this.speciesReactantByproductMatrix, 1),1);
<a name="l00540"></a>00540             anyFlux = false;
<a name="l00541"></a>00541 
<a name="l00542"></a>00542             % simulate modification
<a name="l00543"></a>00543             <a class="code" href="json_parse_8m.html#ad2fc293e8080616c8c1e2ffbd7b86742">while</a> true
<a name="l00544"></a>00544                 %compute maximum number of each species that can be modified
<a name="l00545"></a>00545                 substrateLimits = species(ones(numSubstrates,1), :) ./ this.speciesReactantMatrix;
<a name="l00546"></a>00546                 substrateLimits(:, this.substrateIndexs_water) = NaN;
<a name="l00547"></a>00547                 substrateLimits(:, this.substrateIndexs_hydrogen) = NaN;
<a name="l00548"></a>00548                 substrateLimits = min(substrateLimits, [], 2)&#39;;
<a name="l00549"></a>00549                 substrateLimits(isinf(substrateLimits) | isnan(substrateLimits) | substrateLimits&lt;1)=0;
<a name="l00550"></a>00550 
<a name="l00551"></a>00551                 %stop if no more substrates can be modified
<a name="l00552"></a>00552                 if ~any(substrateLimits); break; end;
<a name="l00553"></a>00553                 
<a name="l00554"></a>00554                 anyFlux = true;
<a name="l00555"></a>00555 
<a name="l00556"></a>00556                 %pick substrate, reaction, enzymes
<a name="l00557"></a>00557                 selectedSubstrate = this.randStream.randsample(numel(substrateLimits), 1, true, substrateLimits);
<a name="l00558"></a>00558                 fluxes(selectedSubstrate) = fluxes(selectedSubstrate) + 1;
<a name="l00559"></a>00559 
<a name="l00560"></a>00560                 %decrement metabolites, enzymes, unmodified substrates; increment modified substrates
<a name="l00561"></a>00561                 species = species - this.speciesReactantByproductMatrix(selectedSubstrate, :);
<a name="l00562"></a>00562             end
<a name="l00563"></a>00563             
<a name="l00564"></a>00564             %stop if no reactions can proceed
<a name="l00565"></a>00565             if ~anyFlux
<a name="l00566"></a>00566                 return;
<a name="l00567"></a>00567             end
<a name="l00568"></a>00568 
<a name="l00569"></a>00569             % compute results
<a name="l00570"></a>00570             this.substrates = this.substrates - this.proteinProstheticGroupMatrix(this.monomerComplexIndexs_folded, :)&#39; * fluxes;
<a name="l00571"></a>00571 
<a name="l00572"></a>00572             this.unfoldedMonomers = this.unfoldedMonomers - fluxes(this.speciesIndexs_monomers);
<a name="l00573"></a>00573             this.foldedMonomers   = this.foldedMonomers   + fluxes(this.speciesIndexs_monomers);
<a name="l00574"></a>00574 
<a name="l00575"></a>00575             this.unfoldedComplexs(this.complexIndexs_folding) = ...
<a name="l00576"></a>00576                 this.unfoldedComplexs(this.complexIndexs_folding) - ...
<a name="l00577"></a>00577                 fluxes(this.speciesIndexs_complexs);
<a name="l00578"></a>00578             this.foldedComplexs(this.complexIndexs_folding) = ...
<a name="l00579"></a>00579                 this.foldedComplexs(this.complexIndexs_folding) + ...
<a name="l00580"></a>00580                 fluxes(this.speciesIndexs_complexs);
<a name="l00581"></a>00581         end
<a name="l00582"></a>00582     end
<a name="l00583"></a>00583 
<a name="l00584"></a>00584     %get methods of dependent local state
<a name="l00585"></a>00585     methods
<a name="l00586"></a>00586         function value = getDryWeight(this)
<a name="l00587"></a>00587             if size(this.unfoldedComplexs, 3) == 1
<a name="l00588"></a>00588                 value = this.getDryWeight@edu.stanford.covert.cell.sim.<a class="code" href="class_process.html">Process</a>() + (...
<a name="l00589"></a>00589                     this.monomer.molecularWeights(this.monomer.processedIIIndexs)&#39; * this.unfoldedMonomers + ...
<a name="l00590"></a>00590                     this.monomer.molecularWeights(this.monomer.foldedIndexs)&#39;      * this.foldedMonomers + ...                
<a name="l00591"></a>00591                     this.complex.molecularWeights(this.complex.nascentIndexs)&#39;     * this.unfoldedComplexs + ...
<a name="l00592"></a>00592                     this.complex.molecularWeights(this.complex.matureIndexs)&#39;      * this.foldedComplexs) ...
<a name="l00593"></a>00593                     / edu.stanford.covert.util.<a class="code" href="class_constant_util.html">ConstantUtil</a>.nAvogadro;
<a name="l00594"></a>00594             else
<a name="l00595"></a>00595                 value = this.getDryWeight@edu.stanford.covert.cell.sim.<a class="code" href="class_process.html">Process</a>() + (...
<a name="l00596"></a>00596                     permute(this.monomer.molecularWeights(this.monomer.processedIIIndexs)&#39; * permute(this.unfoldedMonomers,[1 3 2]),[1 3 2]) + ...
<a name="l00597"></a>00597                     permute(this.monomer.molecularWeights(this.monomer.foldedIndexs)&#39;      * permute(this.foldedMonomers,  [1 3 2]),[1 3 2]) + ...      
<a name="l00598"></a>00598                     permute(this.complex.molecularWeights(this.complex.nascentIndexs)&#39;     * permute(this.unfoldedComplexs,[1 3 2]),[1 3 2]) + ...
<a name="l00599"></a>00599                     permute(this.complex.molecularWeights(this.complex.matureIndexs)&#39;      * permute(this.foldedComplexs  ,[1 3 2]),[1 3 2])) ...
<a name="l00600"></a>00600                     / edu.stanford.covert.util.<a class="code" href="class_constant_util.html">ConstantUtil</a>.nAvogadro;
<a name="l00601"></a>00601             end
<a name="l00602"></a>00602         end
<a name="l00603"></a>00603     end
<a name="l00604"></a>00604 end
</pre></div></div><!-- contents -->
</div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
<a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(0)"><span class="SelectionMark">&#160;</span>All</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(1)"><span class="SelectionMark">&#160;</span>Classes</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(2)"><span class="SelectionMark">&#160;</span>Files</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(3)"><span class="SelectionMark">&#160;</span>Functions</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(4)"><span class="SelectionMark">&#160;</span>Variables</a></div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

  <div id="nav-path" class="navpath">
    <ul>
      <li class="navelem"><a class="el" href="_protein_folding_8m.html">ProteinFolding.m</a>      </li>

    <li class="footer">Generated on Wed Feb 8 2012 21:42:32 for Whole Cell Simulation by
    <a href="http://www.doxygen.org/index.html">
    <img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.7.6.1 </li>
   </ul>
 </div>


</body>
</html>
