<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<title>Whole Cell Simulation: src/+edu/+stanford/+covert/+cell/+sim/+process/RNAProcessing.m Source File</title>

<link href="tabs.css" rel="stylesheet" type="text/css"/>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
<link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="navtree.js"></script>
<script type="text/javascript">
  $(document).ready(initResizable);
</script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
  $(document).ready(function() { searchBox.OnSelectItem(0); });
</script>

</head>
<body>
<div id="top"><!-- do not remove this div! -->


<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  
  
  <td style="padding-left: 0.5em;">
   <div id="projectname">Whole Cell Simulation
   
   </div>
   
  </td>
  
  
  
 </tr>
 </tbody>
</table>
</div>

<!-- Generated by Doxygen 1.7.6.1 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "search",false,'Search');
</script>
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&#160;Page</span></a></li>
      <li><a href="annotated.html"><span>Classes</span></a></li>
      <li class="current"><a href="files.html"><span>Files</span></a></li>
      <li>
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
  <div id="navrow2" class="tabs2">
    <ul class="tablist">
      <li><a href="files.html"><span>File&#160;List</span></a></li>
      <li><a href="globals.html"><span>File&#160;Members</span></a></li>
    </ul>
  </div>
</div>
<div id="side-nav" class="ui-resizable side-nav-resizable">
  <div id="nav-tree">
    <div id="nav-tree-contents">
    </div>
  </div>
  <div id="splitbar" style="-moz-user-select:none;" 
       class="ui-resizable-handle">
  </div>
</div>
<script type="text/javascript">
  initNavTree('_r_n_a_processing_8m.html','');
</script>
<div id="doc-content">
<div class="header">
  <div class="headertitle">
<div class="title">src/+edu/+stanford/+covert/+cell/+sim/+process/RNAProcessing.m</div>  </div>
</div><!--header-->
<div class="contents">
<a href="_r_n_a_processing_8m.html">Go to the documentation of this file.</a><div class="fragment"><pre class="fragment"><a name="l00001"></a>00001 %RNAProcessing
<a name="l00002"></a>00002 %
<a name="l00003"></a>00003 % @wholeCellModelID Process_RNAProcessing
<a name="l00004"></a>00004 % @name             <a class="code" href="class_r_n_a.html">RNA</a> Processing
<a name="l00005"></a>00005 % @description
<a name="l00006"></a>00006 %   Biology
<a name="l00007"></a>00007 %   ==================
<a name="l00008"></a>00008 %   <a class="code" href="class_transcription.html">Transcription</a> produces mono- as well as polycistronic RNAs.
<a name="l00009"></a>00009 %   Following transcription polycistronic <a class="code" href="classr_r_n_a.html">rRNA</a>, <a class="code" href="classt_r_n_a.html">tRNA</a>, and tmRNA transcripts are
<a name="l00010"></a>00010 %   cleaved, modified, and aminoacylated. First, 45S <a class="code" href="classr_r_n_a.html">rRNA</a> transcripts are
<a name="l00011"></a>00011 %   cleaved by ribonuclease III (MG_367_DIMER) into 5S, 16S, and 23S <a class="code" href="classr_r_n_a.html">rRNA</a>
<a name="l00012"></a>00012 %   precursors [PUB_0038, PUB_0039]. Next these precursors are cleaved at their
<a name="l00013"></a>00013 %   3&#39; and 5&#39; ends by ribonuclease J and RsgA [PUB_0038]. Similarly the 5&#39; ends
<a name="l00014"></a>00014 %   of each <a class="code" href="classt_r_n_a.html">tRNA</a> and tmRNA species are cleaved by ribonuclease P [PUB_0039,
<a name="l00015"></a>00015 %   PUB_0649]. Ribonuclease III also cleaves scRNA at two sites [PUB_0039].
<a name="l00016"></a>00016 %   Second, 13 enzymes formylate, lysidinate, methylate, pseudouridylate, and
<a name="l00017"></a>00017 %   thiolate 86 specific <a class="code" href="classr_r_n_a.html">rRNA</a> and <a class="code" href="classt_r_n_a.html">tRNA</a> bases. Third 19 <a class="code" href="classt_r_n_a.html">tRNA</a> synthetases
<a name="l00018"></a>00018 %   conjugate 19 amino acids to 37 <a class="code" href="classt_r_n_a.html">tRNA</a> and tmRNA species, and 2 <a class="code" href="classt_r_n_a.html">tRNA</a>
<a name="l00019"></a>00019 %   transferases complete the aminoacylation of the formylmethionine and
<a name="l00020"></a>00020 %   glutamine tRNAs.
<a name="l00021"></a>00021 %
<a name="l00022"></a>00022 %   <a class="code" href="classr_r_n_a.html">rRNA</a>, <a class="code" href="classt_r_n_a.html">tRNA</a>, and tmRNA cleavage, modification, and aminoacylation are modeled
<a name="l00023"></a>00023 %   motivated by mass-action kinetics. First, the maximum rate of maturation of
<a name="l00024"></a>00024 %   each <a class="code" href="class_r_n_a.html">RNA</a> species is calculated based on (1) the abundance of immature
<a name="l00025"></a>00025 %   transcripts, (2) the abundance of substrates and enzymes required to cleave
<a name="l00026"></a>00026 %   or modify each <a class="code" href="class_r_n_a.html">RNA</a> species, and (3) the experimentally measured kinetic rate
<a name="l00027"></a>00027 %   of each enzyme required to cleave or modify each <a class="code" href="class_r_n_a.html">RNA</a> species. Second, <a class="code" href="class_r_n_a.html">RNA</a>
<a name="l00028"></a>00028 %   maturations are randomly selected according to the calculated maximum rates.
<a name="l00029"></a>00029 %   Finally, steps (1) and (2) are repeated until insufficient resources are
<a name="l00030"></a>00030 %   available to mature additional RNAs.
<a name="l00031"></a>00031 %
<a name="l00032"></a>00032 %   This process simulates the cleavage of polycistronic r/s/t <a class="code" href="class_r_n_a.html">RNA</a> transcripts
<a name="l00033"></a>00033 %   into individual genes:
<a name="l00034"></a>00034 %   - 30S RNA        -&gt; pre 16/17S RNA, p23S RNA, 9S RNA    RNAseIII (MG_367)
<a name="l00035"></a>00035 %   - pre 16/17S RNA -&gt; 16S RNA                             5&#39; end:rnjA (MG_139); 3<span class="stringliteral">&#39; end:rsgA (MG_110)</span>
<a name="l00036"></a>00036 <span class="stringliteral">%   - p23S RNA       -&gt; 23S RNA                             5&#39;</span> end:rnjA (MG_139), deaD (MG_425)
<a name="l00037"></a>00037 %   - 9S <a class="code" href="class_r_n_a.html">RNA</a>         -&gt;  5S <a class="code" href="class_r_n_a.html">RNA</a>
<a name="l00038"></a>00038 %   - 3&#39; end <a class="code" href="classt_r_n_a.html">tRNA</a> precursors                                RNAseIII (MG_367)
<a name="l00039"></a>00039 %   - 5&#39; end <a class="code" href="classt_r_n_a.html">tRNA</a> precursors                                RNaseP (MG_0003, MG_465)
<a name="l00040"></a>00040 %   - scRNA (MG_0001) precusors                             RNAseIII (MG_367)
<a name="l00041"></a>00041 %   - 5&#39; end tmRNA precursors                               RNaseP (MG_0003, MG_465)
<a name="l00042"></a>00042 %
<a name="l00043"></a>00043 %   Reactions
<a name="l00044"></a>00044 %   ++++++++++++++++++
<a name="l00045"></a>00045 %   DeaD           phosphorolytic                      3.6.4.13
<a name="l00046"></a>00046 %   RsgA           phosphorolytic                      3.6.5.2, 3.6.1.-
<a name="l00047"></a>00047 %   RNAaseIII      hydrolytic       endoribonuclease   3.1.26.3
<a name="l00048"></a>00048 %   RNAseP         hydrolytic       endoribonuclease   3.1.26.5
<a name="l00049"></a>00049 %   RNAseJ         hydrolytic       endoribonuclease   3.1.27.2
<a name="l00050"></a>00050 %
<a name="l00051"></a>00051 %   Cofactors
<a name="l00052"></a>00052 %   ++++++++++++++++++
<a name="l00053"></a>00053 %   DeaD           Mg2+                [PUB_0071]
<a name="l00054"></a>00054 %   RsgA       (1) Zn2+                [PUB_0096]
<a name="l00055"></a>00055 %   RNAaseIII      Mg2+                [PUB_0039]
<a name="l00056"></a>00056 %   RNAseJ     (2) Zn2+                [PUB_0096]
<a name="l00057"></a>00057 %   RNAseP     (3) Mg2+ or (3) Mn2+    [PUB_0012, PUB_0013, PUB_0043]
<a name="l00058"></a>00058 %
<a name="l00059"></a>00059 %   Kinetics:
<a name="l00060"></a>00060 %   ++++++++++++++++++
<a name="l00061"></a>00061 %   DeaD       1.48   1/s   B. subtilis  [PUB_0688]
<a name="l00062"></a>00062 %   RsgA       0.2917 1/s   E. coli      [PUB_0103]
<a name="l00063"></a>00063 %   RNAseP     0.027  1/s   B. subtilis  [PUB_0013]
<a name="l00064"></a>00064 %   RNAseP     6      1/s   B. subtilis  [PUB_0690]
<a name="l00065"></a>00065 %   RNAseIII   7.7    1/s   E. coli      [PUB_0691]
<a name="l00066"></a>00066 %   RNAseJ     0.37   1/s   B. subtilis  [PUB_0692]
<a name="l00067"></a>00067 %
<a name="l00068"></a>00068 %   Energy-Dependence
<a name="l00069"></a>00069 %   ++++++++++++++++++
<a name="l00070"></a>00070 %   DeaD       2.37  ATP  1/rxn   B. subtilis  [PUB_0069] (210 ATP/min B. subtilis [PUB_0069])/(1.48 rxn/s B. subtilis  [PUB_0688])
<a name="l00071"></a>00071 %   RsgA       1     GTP  1/rxn   E. coli      [PUB_0103]
<a name="l00072"></a>00072 %
<a name="l00073"></a>00073 %   Knowledge Base
<a name="l00074"></a>00074 %   ==================
<a name="l00075"></a>00075 %   The transcription unit organization was predicted by mapping the
<a name="l00076"></a>00076 %   experimentally determined genome organization of M. pnemoniae by Güell et al
<a name="l00077"></a>00077 %   [Table S5 &quot;Suboperons&quot;, PUB_0418] onto the M. genitalium genome by homology
<a name="l00078"></a>00078 %   with 4 modifications:
<a name="l00079"></a>00079 %   - r/s/tRNAs were organized into transcription units according to their
<a name="l00080"></a>00080 %     &quot;reference operons&quot; (table s4)
<a name="l00081"></a>00081 %   - All mRNAs Güell Serrano et al did not assign to a suboperon because of
<a name="l00082"></a>00082 %     insufficient evidence (low quality expression data), were assigned to
<a name="l00083"></a>00083 %     their own transcription units. None of these mRNAs are located within
<a name="l00084"></a>00084 %     suboperons.
<a name="l00085"></a>00085 %   - All 5 mRNA genes which don&#39;t have homologs in M. pneumoniae were assigned
<a name="l00086"></a>00086 %     to their own transcription units. These occur outside other transcription
<a name="l00087"></a>00087 %     units.
<a name="l00088"></a>00088 %   - Genes which have undergone rearrangements between M. pnuemoniae and M.
<a name="l00089"></a>00089 %     genitalium were assigned to their own transcription units
<a name="l00090"></a>00090 %
<a name="l00091"></a>00091 %   The promoter (-35 and -10 boxes and TSS) for each transcription unit was set
<a name="l00092"></a>00092 %   predicted for its 3&#39; gene by Weiner et al [PUB_0411].
<a name="l00093"></a>00093 %
<a name="l00094"></a>00094 %   The composition of each transcription unit, and the location of its -35 and
<a name="l00095"></a>00095 %   -10 boxes and TSS relative to the start coordinate of the 3&#39; gene was stored
<a name="l00096"></a>00096 %   in the knowledge base, and it used by the TranscriptionUnit and Gene classes
<a name="l00097"></a>00097 %   to compute the sequences of transcription unit, gene, and intergenic
<a name="l00098"></a>00098 %   segment.
<a name="l00099"></a>00099 %
<a name="l00100"></a>00100 %   The kinetics and energy requirement of each RNA processing enzyme are
<a name="l00101"></a>00101 %   organized as parameters in the knowledge base. These values were curated
<a name="l00102"></a>00102 %   several papers [PUB_0069, PUB_0103, PUB_0688, PUB_0690, PUB_0691, PUB_0692].
<a name="l00103"></a>00103 %
<a name="l00104"></a>00104 %   Representation
<a name="l00105"></a>00105 %   ==================
<a name="l00106"></a>00106 %   substrates and enzymes represent the counts of metabolites and RNA
<a name="l00107"></a>00107 %   processing enzymes. unprocessedRNAs, processedRNAs, and intergenicRNAs
<a name="l00108"></a>00108 %   represent counts of RNAs. unprocessedRNAs represents the counts of nascent
<a name="l00109"></a>00109 %   RNAs produced by the RNA polymerase. processedRNA represents the counts of
<a name="l00110"></a>00110 %   r/s/tRNA precursors produced from cleavage of nascent transcripts, and mRNAs
<a name="l00111"></a>00111 %   that remain joined as transcription units. intergenic RNAs represents the
<a name="l00112"></a>00112 %   counts of RNA segments between r/s/tRNA genes in transcripts that are
<a name="l00113"></a>00113 %   released during the processing of these transcripts.
<a name="l00114"></a>00114 %
<a name="l00115"></a>00115 %   enzymeSpecificRate_* are the kcats of the five RNA processing enzymes. We
<a name="l00116"></a>00116 %   assume that each enzyme has the same kcat for all reactions it catalyzes.
<a name="l00117"></a>00117 %   enzymeEnergyCost_* are the amount of ATP required by the phosphorolytic RNA
<a name="l00118"></a>00118 %   processing enzymes (DeaD and Rsga) per cleavage reaction.
<a name="l00119"></a>00119 %
<a name="l00120"></a>00120 %   rna.nascentRNAMatureRNAComposition, rna.intergenicRNAMatrix,
<a name="l00121"></a>00121 %   reactantByproductMatrix, and catalysisMatrix are adjacency matrices.
<a name="l00122"></a>00122 %   nascentRNAMatureRNAComposition represents the the processed RNAs that arise
<a name="l00123"></a>00123 %   from each unprocessed RNA (eg. the r/s/t RNAs that arise from transcripts
<a name="l00124"></a>00124 %   containing multiple genes). intergenicRNAMatrix represents the intergenic
<a name="l00125"></a>00125 %   RNA segments that arise from the cleavage of r/s/tRNA transcripts into their
<a name="l00126"></a>00126 %   individual genes. reactantByproductMatrix represents the amounts of
<a name="l00127"></a>00127 %   metabolites required to cleave each polycistronic transcript, and the
<a name="l00128"></a>00128 %   metabolic byproducts of their cleavage. catalysisMatrix represents the
<a name="l00129"></a>00129 %   amount of enzyme required to cleave each polycistronic transcript.
<a name="l00130"></a>00130 %
<a name="l00131"></a>00131 %   Initialization
<a name="l00132"></a>00132 %   ==================
<a name="l00133"></a>00133 %   All RNAs are initialized to the mature state. This is accomplished by the
<a name="l00134"></a>00134 %   simulation class initializeState method.
<a name="l00135"></a>00135 %
<a name="l00136"></a>00136 %   Simulation
<a name="l00137"></a>00137 %   ==================
<a name="l00138"></a>00138 %   For each RNA class (<a class="code" href="classm_r_n_a.html">mRNA</a>, <a class="code" href="classr_r_n_a.html">rRNA</a>, <a class="code" href="classs_r_n_a.html">sRNA</a>, <a class="code" href="classt_r_n_a.html">tRNA</a>) in a random order:
<a name="l00139"></a>00139 %   1. Determine maximum number of RNAs that can be processed based on available
<a name="l00140"></a>00140 %      - RNA processing enzymes and their kinetics
<a name="l00141"></a>00141 %      - free metabolites
<a name="l00142"></a>00142 %      - RNAs that need processing
<a name="l00143"></a>00143 %   2. Randomly select among unprocessed RNAs proportional to count
<a name="l00144"></a>00144 %   3. Update state:
<a name="l00145"></a>00145 %      a. Decrement unprocessed RNAs
<a name="l00146"></a>00146 %      b. Increment processed RNAs, intergenic RNAs
<a name="l00147"></a>00147 %      c. Decrement metabolic reactants
<a name="l00148"></a>00148 %      d. Increment metabolic byproducts%
<a name="l00149"></a>00149 %
<a name="l00150"></a>00150 %   References
<a name="l00151"></a>00151 %   ==================
<a name="l00152"></a>00152 %   1. Güell M, van Noort V, Yus E, Chen WH, Leigh-Bell J, Michalodimitrakis K,
<a name="l00153"></a>00153 %      Yamada T, Arumugam M, Doerks T, Kühner S, Rode M, Suyama M, Schmidt S,
<a name="l00154"></a>00154 %      Gavin AC, Bork P, Serrano L (2009). Transcriptome complexity in a
<a name="l00155"></a>00155 %      genome-reduced bacterium. Science. 326(5957): 1268-71. [PUB_0418]
<a name="l00156"></a>00156 %   2. Weiner J 3rd, Herrmann R, Browning GF (2000). Transcription in Mycoplasma
<a name="l00157"></a>00157 %      pneumoniae. Nucleic Acids Res. 28: 4488-96. [PUB_0411].
<a name="l00158"></a>00158 %   3. Cordin O, Banroques J, Tanner NK, Linder P (2006). The DEAD-box protein
<a name="l00159"></a>00159 %      family of RNA helicases. Gene. 367: 17-37. [PUB_0069]
<a name="l00160"></a>00160 %   4. Himeno H, Hanawa-Suetsugu K, Kimura T, Takagi K, Sugiyama W, Shirata S,
<a name="l00161"></a>00161 %      Mikami T, Odagiri F, Osanai Y, Watanabe D, Goto S, Kalachnyuk L, Ushida
<a name="l00162"></a>00162 %      C, Muto A (2004). A  novel GTPase activated by the small subunit of
<a name="l00163"></a>00163 %      ribosome. Nucleic Acids Res. 32 (17): 5303-9. [PUB_0103]
<a name="l00164"></a>00164 %   5. Theissen B, Karow AR, Köhler J, Gubaev A, Klostermeier D (2008).
<a name="l00165"></a>00165 %      Cooperative binding of ATP and RNA induces a closed conformation in a
<a name="l00166"></a>00166 %      DEAD box RNA helicase. Proc Natl Acad Sci U S A. 105(2): 548-53.
<a name="l00167"></a>00167 %      [PUB_0688]
<a name="l00168"></a>00168 %   6. Xiao S, Scott F, Fierke CA, Engelke DR (2002). Eukaryotic ribonuclease P:
<a name="l00169"></a>00169 %      a plurality of ribonucleoprotein enzymes. Annu Rev Biochem. 71: 165-89.
<a name="l00170"></a>00170 %      [PUB_0690]
<a name="l00171"></a>00171 %   7. Amarasinghe AK, Calin-Jageman I, Harmouch A, Sun W, Nicholson AW (2001).
<a name="l00172"></a>00172 %      Escherichia coli ribonuclease III: affinity purification of
<a name="l00173"></a>00173 %      hexahistidine-tagged enzyme and assays for substrate binding and
<a name="l00174"></a>00174 %      cleavage. Methods Enzymol. 342: 143-58. [PUB_0691]
<a name="l00175"></a>00175 %   8. Niranjanakumari S, Day-Storms JJ, Ahmed M, Hsieh J, Zahler NH, Venters
<a name="l00176"></a>00176 %      RA, Fierke CA (2007). Probing the architecture of the B. subtilis RNase P
<a name="l00177"></a>00177 %      holoenzyme active site by cross-linking and affinity cleavage. RNA.
<a name="l00178"></a>00178 %      13(4): 521-35. [PUB_0692]
<a name="l00179"></a>00179 %
<a name="l00180"></a>00180 % Author: Jonathan Karr, jkarr@stanford.edu
<a name="l00181"></a>00181 % Affilitation: Covert Lab, Department of Bioengineering, Stanford University
<a name="l00182"></a>00182 % Last updated: 1/4/2010
<a name="l00183"></a>00183 classdef RNAProcessing &lt; edu.stanford.covert.cell.sim.Process
<a name="l00184"></a>00184     %property annotations
<a name="l00185"></a>00185     properties (Constant)
<a name="l00186"></a>00186         optionNames__              = {}; %names of option <a class="code" href="json_parse_8m.html#a6e4256f69b751d3d7ec4ea52ff5a8dee">properties</a>
<a name="l00187"></a>00187         fixedConstantNames__       = {   %names of fixed constant <a class="code" href="json_parse_8m.html#a6e4256f69b751d3d7ec4ea52ff5a8dee">properties</a>
<a name="l00188"></a>00188             <span class="stringliteral">&#39;enzymeSpecificRate_DeaD&#39;</span>;
<a name="l00189"></a>00189             <span class="stringliteral">&#39;enzymeSpecificRate_RsgA&#39;</span>;
<a name="l00190"></a>00190             <span class="stringliteral">&#39;enzymeSpecificRate_RNAseP&#39;</span>;
<a name="l00191"></a>00191             <span class="stringliteral">&#39;enzymeSpecificRate_RNAseIII&#39;</span>;
<a name="l00192"></a>00192             <span class="stringliteral">&#39;enzymeSpecificRate_RNAseJ&#39;</span>;
<a name="l00193"></a>00193             <span class="stringliteral">&#39;enzymeEnergyCost_DeaD&#39;</span>;
<a name="l00194"></a>00194             <span class="stringliteral">&#39;enzymeEnergyCost_RsgA&#39;</span>;            
<a name="l00195"></a>00195             <span class="stringliteral">&#39;reactantByproductMatrix&#39;</span>;
<a name="l00196"></a>00196             <span class="stringliteral">&#39;catalysisMatrix&#39;</span>;
<a name="l00197"></a>00197             };                  
<a name="l00198"></a>00198         fittedConstantNames__      = {}; %names of fitted constant <a class="code" href="json_parse_8m.html#a6e4256f69b751d3d7ec4ea52ff5a8dee">properties</a>
<a name="l00199"></a>00199         localStateNames__          = {   %names of simulation state <a class="code" href="json_parse_8m.html#a6e4256f69b751d3d7ec4ea52ff5a8dee">properties</a> redundant with timecourses in <span class="keyword">this</span> or other processes or the simulation
<a name="l00200"></a>00200             <span class="stringliteral">&#39;unprocessedRNAs&#39;</span>;
<a name="l00201"></a>00201             <span class="stringliteral">&#39;processedRNAs&#39;</span>;
<a name="l00202"></a>00202             <span class="stringliteral">&#39;intergenicRNAs&#39;</span>};
<a name="l00203"></a>00203     end
<a name="l00204"></a>00204     
<a name="l00205"></a>00205     %IDs, names, and local indices
<a name="l00206"></a>00206     <a class="code" href="json_parse_8m.html#a6e4256f69b751d3d7ec4ea52ff5a8dee">properties</a>
<a name="l00207"></a>00207         stimuliWholeCellModelIDs   = {}; %whole cell model IDs of stimuli
<a name="l00208"></a>00208         substrateWholeCellModelIDs = {  %whole cell model IDs of substrates
<a name="l00209"></a>00209             <span class="stringliteral">&#39;ATP&#39;</span>; <span class="stringliteral">&#39;GTP&#39;</span>;
<a name="l00210"></a>00210             <span class="stringliteral">&#39;ADP&#39;</span>; <span class="stringliteral">&#39;GDP&#39;</span>;
<a name="l00211"></a>00211             <span class="stringliteral">&#39;PI&#39;</span>; <span class="stringliteral">&#39;H2O&#39;</span>; <span class="charliteral">&#39;H&#39;</span>};
<a name="l00212"></a>00212         substrateIndexs_NTPs = (1:2)<span class="stringliteral">&#39;;</span>
<a name="l00213"></a>00213 <span class="stringliteral">        substrateIndexs_ATP  = 1;</span>
<a name="l00214"></a>00214 <span class="stringliteral">        substrateIndexs_GTP  = 2;</span>
<a name="l00215"></a>00215 <span class="stringliteral">        substrateIndexs_NDPs = (3:4)&#39;</span>;
<a name="l00216"></a>00216         substrateIndexs_ADP  = 3;
<a name="l00217"></a>00217         substrateIndexs_GDP  = 4;
<a name="l00218"></a>00218         substrateIndexs_PI   = 5;
<a name="l00219"></a>00219         substrateIndexs_H2O  = 6;
<a name="l00220"></a>00220         substrateIndexs_H    = 7;
<a name="l00221"></a>00221         
<a name="l00222"></a>00222         enzymeWholeCellModelIDs = {     %whole cell model IDs of enzymes
<a name="l00223"></a>00223             <span class="stringliteral">&#39;MG_367_DIMER&#39;</span>;             %ribonuclease III
<a name="l00224"></a>00224             <span class="stringliteral">&#39;MG_139_DIMER&#39;</span>;             %ribonuclease J
<a name="l00225"></a>00225             <span class="stringliteral">&#39;MG_0003_465&#39;</span>;              %ribonuclease P
<a name="l00226"></a>00226             <span class="stringliteral">&#39;MG_110_MONOMER&#39;</span>;           %ribosome small subunit-dependent GTPase A
<a name="l00227"></a>00227             &#39;MG_425_DIMER&#39;};            %ATP-dependent <a class="code" href="class_r_n_a.html">RNA</a> helicase
<a name="l00228"></a>00228         enzymeIndexs_RNAseIII = 1;      %index within enzymes of ribonuclease III
<a name="l00229"></a>00229         enzymeIndexs_RNAseJ   = 2;      %index within enzymes of ribonuclease J
<a name="l00230"></a>00230         enzymeIndexs_RNAseP   = 3;      %index within enzymes of ribonuclease P
<a name="l00231"></a>00231         enzymeIndexs_RsgA     = 4;      %index within enzymes of ribosome small subunit-dependent GTPase A
<a name="l00232"></a>00232         enzymeIndexs_DeaD     = 5;      %index within enzymes of ATP-dependent <a class="code" href="class_r_n_a.html">RNA</a> helicase
<a name="l00233"></a>00233         
<a name="l00234"></a>00234         unprocessedRNAWholeCellModelIDs %whole cell model IDs of unprocessed RNAs
<a name="l00235"></a>00235         processedRNAWholeCellModelIDs   %whole cell model IDs of processed RNAs
<a name="l00236"></a>00236         intergenicRNAWholeCellModelIDs  %whole cell model IDs of intergenic RNAs
<a name="l00237"></a>00237         
<a name="l00238"></a>00238         unprocessedRNAIndexs_mRNA       %indices of unprocessed mRNAs within unprocessedRNAs
<a name="l00239"></a>00239         unprocessedRNAIndexs_rRNA       %indices of unprocessed rRNAs within unprocessedRNAs
<a name="l00240"></a>00240         unprocessedRNAIndexs_sRNA       %indices of unprocessed sRNAs within unprocessedRNAs
<a name="l00241"></a>00241         unprocessedRNAIndexs_tRNA       %indices of unprocessed tRNAs within unprocessedRNAs
<a name="l00242"></a>00242         unprocessedRNAIndexs_scRNA      %indices of unprocessed scRNAs within unprocessedRNAs
<a name="l00243"></a>00243         unprocessedRNAIndexs_tmRNA      %indices of unprocessed tmRNAs within unprocessedRNAs
<a name="l00244"></a>00244         
<a name="l00245"></a>00245         processedRNAIndexs_mRNA         %indices of unprocessed mRNAs within processedRNAs
<a name="l00246"></a>00246         processedRNAIndexs_rRNA         %indices of unprocessed rRNAs within processedRNAs
<a name="l00247"></a>00247         processedRNAIndexs_sRNA         %indices of unprocessed sRNAs within processedRNAs
<a name="l00248"></a>00248         processedRNAIndexs_tRNA         %indices of unprocessed tRNAs within processedRNAs
<a name="l00249"></a>00249         processedRNAIndexs_scRNA        %indices of unprocessed scRNAs within processedRNAs
<a name="l00250"></a>00250         processedRNAIndexs_tmRNA        %indices of unprocessed tmRNAs within processedRNAs
<a name="l00251"></a>00251     end
<a name="l00252"></a>00252     
<a name="l00253"></a>00253     %fixed biological constants
<a name="l00254"></a>00254     <a class="code" href="json_parse_8m.html#a6e4256f69b751d3d7ec4ea52ff5a8dee">properties</a>
<a name="l00255"></a>00255         %enzyme kcats (reactions catalyzed per second per enzyme)
<a name="l00256"></a>00256         enzymeSpecificRate_DeaD         % 1.48   1/s   B. subtilis  [PUB_0688]
<a name="l00257"></a>00257         enzymeSpecificRate_RsgA         % 0.2917 1/s   E. coli      [PUB_0103]
<a name="l00258"></a>00258         enzymeSpecificRate_RNAseP       % 6      1/s   B. subtilis  [PUB_0690]
<a name="l00259"></a>00259         enzymeSpecificRate_RNAseIII     % 7.7    1/s   E. coli      [PUB_0691]
<a name="l00260"></a>00260         enzymeSpecificRate_RNAseJ       % 0.37   1/s   B. subtilis  [PUB_0692]
<a name="l00261"></a>00261         
<a name="l00262"></a>00262         %NTP hydrolyzed per reaction
<a name="l00263"></a>00263         enzymeEnergyCost_DeaD           % 2.37 ATP/reaction   (210 ATP/min B. subtilis [PUB_0069])/(1.48 rxn/s B. subtilis  [PUB_0688])
<a name="l00264"></a>00264         enzymeEnergyCost_RsgA           % 1    GTP/reaction   [PUB_0103]
<a name="l00265"></a>00265                 
<a name="l00266"></a>00266         %adjacency matrices
<a name="l00267"></a>00267         reactantByproductMatrix         %metabolites required to cleavage each transcription <a class="code" href="unit_8m.html#a94c091bed34da4edc337e42e23029000">unit</a>, and metabolic byproducts of cleavage [substrates X unprocessed RNAs]
<a name="l00268"></a>00268         catalysisMatrix                 %enzymes required to cleave each transcription <a class="code" href="unit_8m.html#a94c091bed34da4edc337e42e23029000">unit</a>, including the effect of their kcat [enzymes X unprocessed RNAs]
<a name="l00269"></a>00269     end
<a name="l00270"></a>00270     
<a name="l00271"></a>00271     %global state (copied locally for convenience)
<a name="l00272"></a>00272     <a class="code" href="json_parse_8m.html#a6e4256f69b751d3d7ec4ea52ff5a8dee">properties</a>
<a name="l00273"></a>00273         unprocessedRNAs                 %counts of unprocessed RNAs
<a name="l00274"></a>00274         processedRNAs                   %counts of processed RNAs
<a name="l00275"></a>00275         intergenicRNAs                  %counts of intergenic RNAs
<a name="l00276"></a>00276     end
<a name="l00277"></a>00277     
<a name="l00278"></a>00278     %constructor
<a name="l00279"></a>00279     methods
<a name="l00280"></a>00280         function this = RNAProcessing(wholeCellModelID, name)
<a name="l00281"></a>00281             this = this@edu.stanford.covert.cell.sim.<a class="code" href="class_process.html">Process</a>(wholeCellModelID, name);
<a name="l00282"></a>00282         end
<a name="l00283"></a>00283     end
<a name="l00284"></a>00284     
<a name="l00285"></a>00285     %communication between process/simulation
<a name="l00286"></a>00286     methods
<a name="l00287"></a>00287         <span class="keyword">function</span> initializeConstants(<span class="keyword">this</span>, knowledgeBase, simulation, varargin)
<a name="l00288"></a>00288             this.initializeConstants@edu.stanford.covert.cell.sim.<a class="code" href="class_process.html">Process</a>(knowledgeBase, simulation, varargin{:});
<a name="l00289"></a>00289 
<a name="l00290"></a>00290             %% RNAs
<a name="l00291"></a>00291             s = this.rna;
<a name="l00292"></a>00292             this.unprocessedRNAWholeCellModelIDs = s.wholeCellModelIDs(s.nascentIndexs);
<a name="l00293"></a>00293             this.processedRNAWholeCellModelIDs   = s.wholeCellModelIDs(s.processedIndexs);
<a name="l00294"></a>00294             this.intergenicRNAWholeCellModelIDs  = s.wholeCellModelIDs(s.intergenicIndexs);
<a name="l00295"></a>00295             
<a name="l00296"></a>00296             this.processedRNAIndexs_mRNA = s.matureMRNAIndexs;
<a name="l00297"></a>00297             this.processedRNAIndexs_rRNA = s.matureRRNAIndexs;
<a name="l00298"></a>00298             this.processedRNAIndexs_sRNA = s.matureSRNAIndexs;
<a name="l00299"></a>00299             this.processedRNAIndexs_tRNA = s.matureTRNAIndexs;
<a name="l00300"></a>00300             this.processedRNAIndexs_scRNA = find(strcmp(<span class="stringliteral">&#39;MG_0001&#39;</span>, this.processedRNAWholeCellModelIDs));
<a name="l00301"></a>00301             this.processedRNAIndexs_tmRNA = find(strcmp(<span class="stringliteral">&#39;MG_0004&#39;</span>, this.processedRNAWholeCellModelIDs));                    
<a name="l00302"></a>00302             
<a name="l00303"></a>00303             %% processing pathways
<a name="l00304"></a>00304             %unprocessed RNAs that are precursors to processed RNAs           
<a name="l00305"></a>00305             this.unprocessedRNAIndexs_mRNA = find(sum(this.rna.nascentRNAMatureRNAComposition(<span class="keyword">this</span>.processedRNAIndexs_mRNA, :),1));
<a name="l00306"></a>00306             this.unprocessedRNAIndexs_rRNA = find(sum(this.rna.nascentRNAMatureRNAComposition(<span class="keyword">this</span>.processedRNAIndexs_rRNA, :),1));
<a name="l00307"></a>00307             this.unprocessedRNAIndexs_sRNA = find(sum(this.rna.nascentRNAMatureRNAComposition(<span class="keyword">this</span>.processedRNAIndexs_sRNA, :),1));
<a name="l00308"></a>00308             this.unprocessedRNAIndexs_tRNA = find(sum(this.rna.nascentRNAMatureRNAComposition(<span class="keyword">this</span>.processedRNAIndexs_tRNA, :),1));
<a name="l00309"></a>00309             this.unprocessedRNAIndexs_scRNA = find(sum(this.rna.nascentRNAMatureRNAComposition(<span class="keyword">this</span>.processedRNAIndexs_scRNA, :),1));
<a name="l00310"></a>00310             this.unprocessedRNAIndexs_tmRNA = find(sum(this.rna.nascentRNAMatureRNAComposition(<span class="keyword">this</span>.processedRNAIndexs_tmRNA, :),1));
<a name="l00311"></a>00311             
<a name="l00312"></a>00312             %metabolites required to process each <a class="code" href="class_r_n_a.html">RNA</a>, and released as
<a name="l00313"></a>00313             %byproducts
<a name="l00314"></a>00314             this.reactantByproductMatrix = zeros(numel(this.substrateWholeCellModelIDs), numel(this.unprocessedRNAWholeCellModelIDs));
<a name="l00315"></a>00315             
<a name="l00316"></a>00316             this.reactantByproductMatrix(this.substrateIndexs_H2O, :) = -(sum(this.rna.nascentRNAMatureRNAComposition, 1) + sum(this.rna.intergenicRNAMatrix, 1) - 1);
<a name="l00317"></a>00317             this.reactantByproductMatrix(this.substrateIndexs_H,   :) =  (sum(this.rna.nascentRNAMatureRNAComposition, 1) + sum(this.rna.intergenicRNAMatrix, 1) - 1);
<a name="l00318"></a>00318             
<a name="l00319"></a>00319             energyIdxs_DeaD = [this.substrateIndexs_ATP; this.substrateIndexs_ADP; this.substrateIndexs_PI; this.substrateIndexs_H2O; this.substrateIndexs_H];
<a name="l00320"></a>00320             this.reactantByproductMatrix(energyIdxs_DeaD, this.unprocessedRNAIndexs_rRNA) = ...
<a name="l00321"></a>00321                 this.reactantByproductMatrix(energyIdxs_DeaD, this.unprocessedRNAIndexs_rRNA) + ...
<a name="l00322"></a>00322                 round(this.enzymeEnergyCost_RsgA) * repmat([-1; 1; 1; -1; 1], 1, numel(this.unprocessedRNAIndexs_rRNA));
<a name="l00323"></a>00323             
<a name="l00324"></a>00324             energyIdxs_RsgA = [this.substrateIndexs_GTP; this.substrateIndexs_GDP; this.substrateIndexs_PI; this.substrateIndexs_H2O; this.substrateIndexs_H];
<a name="l00325"></a>00325             this.reactantByproductMatrix(energyIdxs_RsgA, this.unprocessedRNAIndexs_rRNA) = ...
<a name="l00326"></a>00326                 this.reactantByproductMatrix(energyIdxs_RsgA, this.unprocessedRNAIndexs_rRNA) + ...
<a name="l00327"></a>00327                 round(this.enzymeEnergyCost_DeaD) * repmat([-1; 1; 1; -1; 1], 1, numel(this.unprocessedRNAIndexs_rRNA));
<a name="l00328"></a>00328             
<a name="l00329"></a>00329             %enzymes required to process each RNA (including affect of kcats)
<a name="l00330"></a>00330             catalysis = zeros(numel(this.enzymeWholeCellModelIDs), numel(this.unprocessedRNAWholeCellModelIDs));
<a name="l00331"></a>00331             catalysis(this.enzymeIndexs_DeaD,     this.unprocessedRNAIndexs_rRNA) = 1;
<a name="l00332"></a>00332             catalysis(this.enzymeIndexs_RNAseJ,   this.unprocessedRNAIndexs_rRNA) = 2;
<a name="l00333"></a>00333             catalysis(this.enzymeIndexs_RNAseIII, this.unprocessedRNAIndexs_rRNA) = 2;
<a name="l00334"></a>00334             catalysis(this.enzymeIndexs_RsgA,     this.unprocessedRNAIndexs_rRNA) = 1;
<a name="l00335"></a>00335             
<a name="l00336"></a>00336             catalysis(this.enzymeIndexs_RNAseIII, this.unprocessedRNAIndexs_scRNA) = 2;
<a name="l00337"></a>00337             catalysis(this.enzymeIndexs_RNAseP,   this.unprocessedRNAIndexs_tmRNA) = 1;
<a name="l00338"></a>00338             
<a name="l00339"></a>00339             catalysis(this.enzymeIndexs_RNAseIII, this.unprocessedRNAIndexs_tRNA) = 1;
<a name="l00340"></a>00340             catalysis(this.enzymeIndexs_RNAseP,   this.unprocessedRNAIndexs_tRNA) = 1;
<a name="l00341"></a>00341             
<a name="l00342"></a>00342             enzymeSpecificRates = zeros(5,1);
<a name="l00343"></a>00343             enzymeSpecificRates(this.enzymeIndexs_DeaD)     = this.enzymeSpecificRate_DeaD;
<a name="l00344"></a>00344             enzymeSpecificRates(this.enzymeIndexs_RsgA)     = this.enzymeSpecificRate_RsgA;
<a name="l00345"></a>00345             enzymeSpecificRates(this.enzymeIndexs_RNAseP)   = this.enzymeSpecificRate_RNAseP;
<a name="l00346"></a>00346             enzymeSpecificRates(this.enzymeIndexs_RNAseIII) = this.enzymeSpecificRate_RNAseIII;
<a name="l00347"></a>00347             enzymeSpecificRates(this.enzymeIndexs_RNAseJ)   = this.enzymeSpecificRate_RNAseJ;
<a name="l00348"></a>00348             
<a name="l00349"></a>00349             this.catalysisMatrix = catalysis ./ repmat(enzymeSpecificRates, 1, numel(this.unprocessedRNAWholeCellModelIDs));
<a name="l00350"></a>00350         end
<a name="l00351"></a>00351         
<a name="l00352"></a>00352         %retrieve state from simulation
<a name="l00353"></a>00353         <span class="keyword">function</span> copyFromState(<span class="keyword">this</span>)
<a name="l00354"></a>00354             this.copyFromState@edu.stanford.covert.cell.sim.<a class="code" href="class_process.html">Process</a>();
<a name="l00355"></a>00355             
<a name="l00356"></a>00356             this.unprocessedRNAs = this.rna.counts(this.rna.nascentIndexs,    this.compartment.cytosolIndexs, :);
<a name="l00357"></a>00357             this.processedRNAs   = this.rna.counts(this.rna.processedIndexs,  this.compartment.cytosolIndexs, :);
<a name="l00358"></a>00358             this.intergenicRNAs  = this.rna.counts(this.rna.intergenicIndexs, this.compartment.cytosolIndexs, :);
<a name="l00359"></a>00359         end
<a name="l00360"></a>00360         
<a name="l00361"></a>00361         %send state to simulation
<a name="l00362"></a>00362         function copyToState(this)
<a name="l00363"></a>00363             this.copyToState@edu.stanford.covert.cell.sim.Process();
<a name="l00364"></a>00364 
<a name="l00365"></a>00365             this.rna.counts(this.rna.nascentIndexs,    this.compartment.cytosolIndexs, :) = this.unprocessedRNAs;
<a name="l00366"></a>00366             this.rna.counts(this.rna.processedIndexs,  this.compartment.cytosolIndexs, :) = this.processedRNAs;
<a name="l00367"></a>00367             this.rna.counts(this.rna.intergenicIndexs, this.compartment.cytosolIndexs, :) = this.intergenicRNAs;
<a name="l00368"></a>00368         end
<a name="l00369"></a>00369     end
<a name="l00370"></a>00370     
<a name="l00371"></a>00371     %allocate memory for state
<a name="l00372"></a>00372     methods
<a name="l00373"></a>00373         function allocateMemoryForState(this, numTimePoints)
<a name="l00374"></a>00374             this.allocateMemoryForState@edu.stanford.covert.cell.sim.Process(numTimePoints);
<a name="l00375"></a>00375 
<a name="l00376"></a>00376             this.unprocessedRNAs = zeros(length(this.unprocessedRNAWholeCellModelIDs), 1, numTimePoints);
<a name="l00377"></a>00377             this.processedRNAs   = zeros(length(this.processedRNAWholeCellModelIDs),   1, numTimePoints);
<a name="l00378"></a>00378             this.intergenicRNAs  = zeros(length(this.intergenicRNAWholeCellModelIDs),  1, numTimePoints);
<a name="l00379"></a>00379         end
<a name="l00380"></a>00380     end
<a name="l00381"></a>00381     
<a name="l00382"></a>00382     %model
<a name="l00383"></a>00383     methods
<a name="l00384"></a>00384         %Calculate
<a name="l00385"></a>00385         %- contribution to <a class="code" href="class_f_b_a.html">FBA</a> objective
<a name="l00386"></a>00386         %- minimum expression consistent with cell cycle length
<a name="l00387"></a>00387         function [bmProd, byProd, minEnzExp, maxEnzExp] = calcResourceRequirements_LifeCycle(this, ~, states)
<a name="l00388"></a>00388             import edu.stanford.covert.util.<a class="code" href="class_computation_util.html">ComputationUtil</a>;
<a name="l00389"></a>00389             invMat = ComputationUtil.invertCompositionMatrix(this.rna.nascentRNAMatureRNAComposition);
<a name="l00390"></a>00390             
<a name="l00391"></a>00391             %substrate and byproducts
<a name="l00392"></a>00392             bmProd = max(0, -this.reactantByproductMatrix) * invMat * states.rnaProductions;
<a name="l00393"></a>00393             byProd = max(0,  this.reactantByproductMatrix) * invMat * states.rnaProductions;
<a name="l00394"></a>00394             
<a name="l00395"></a>00395             %enzymes
<a name="l00396"></a>00396             minEnzExp = 2 * this.catalysisMatrix * invMat * states.rnaProductions0;
<a name="l00397"></a>00397             maxEnzExp = Inf(size(this.enzymeWholeCellModelIDs));
<a name="l00398"></a>00398         end
<a name="l00399"></a>00399         
<a name="l00400"></a>00400         %initialization: RNAs initialized to the mature/aminoacylated state by
<a name="l00401"></a>00401         %simulation initializeState method
<a name="l00402"></a>00402         function initializeState(~)
<a name="l00403"></a>00403         end
<a name="l00404"></a>00404         
<a name="l00405"></a>00405         %resource requirements
<a name="l00406"></a>00406         function result = calcResourceRequirements_Current(this)
<a name="l00407"></a>00407             result = max(0, -this.reactantByproductMatrix) * min( ...
<a name="l00408"></a>00408                 ceil(min(this.enzymes(:, ones(size(this.catalysisMatrix, 2), 1)) ./ this.catalysisMatrix * this.stepSizeSec, [], 1))&#39;, ...
<a name="l00409"></a>00409                 this.unprocessedRNAs); %<span class="preprocessor">#ok&lt;UDIM&gt;</span>
<a name="l00410"></a>00410 <span class="preprocessor"></span>        end
<a name="l00411"></a>00411 
<a name="l00412"></a>00412         %simulation
<a name="l00413"></a>00413         <span class="keyword">function</span> evolveState(<span class="keyword">this</span>)
<a name="l00414"></a>00414             rnaTypes = {<span class="stringliteral">&#39;mRNA&#39;</span>; <span class="stringliteral">&#39;rRNA&#39;</span>; <span class="stringliteral">&#39;sRNA&#39;</span>; <span class="stringliteral">&#39;tRNA&#39;</span>};
<a name="l00415"></a>00415             
<a name="l00416"></a>00416             order = this.randStream.randperm(4);
<a name="l00417"></a>00417             <span class="keywordflow">for</span> i = 1:4
<a name="l00418"></a>00418                 this.evolveState_Helper(rnaTypes{order(i)});
<a name="l00419"></a>00419             end
<a name="l00420"></a>00420         end
<a name="l00421"></a>00421         
<a name="l00422"></a>00422         <span class="keyword">function</span> evolveState_Helper(<span class="keyword">this</span>, rnaType)
<a name="l00423"></a>00423             %indices of unprocessed RNAs of selected type
<a name="l00424"></a>00424             unprocessedRNAIndexs = this.([&#39;unprocessedRNAIndexs_&#39; rnaType]);
<a name="l00425"></a>00425             
<a name="l00426"></a>00426             %limit RNA processing by availability of
<a name="l00427"></a>00427             %- unprocessed RNAs
<a name="l00428"></a>00428             %- enzymes
<a name="l00429"></a>00429             %- free reactants
<a name="l00430"></a>00430             %
<a name="l00431"></a>00431             %Assumes that each RNA of this type uses same amount of substrates
<a name="l00432"></a>00432             %(except water and hydrogen) and enzymes. There is an assertion in
<a name="l00433"></a>00433             %the test class to check that this is true. For water and hydrogen
<a name="l00434"></a>00434             %this upper bound calculation uses that most amount of water and
<a name="l00435"></a>00435             %hydrogen required across all RNAs of the current type (<a class="code" href="class_note.html">Note</a>: the
<a name="l00436"></a>00436             %correct stoichiometry of water and hydrogen are used to update
<a name="l00437"></a>00437             %state).
<a name="l00438"></a>00438             totRNAs = sum(this.unprocessedRNAs(unprocessedRNAIndexs));
<a name="l00439"></a>00439             <span class="keywordflow">if</span> totRNAs == 0
<a name="l00440"></a>00440                 <span class="keywordflow">return</span>;
<a name="l00441"></a>00441             end
<a name="l00442"></a>00442             
<a name="l00443"></a>00443             substrates = this.substrates ./ max(0, -min(this.reactantByproductMatrix(:, unprocessedRNAIndexs), [], 2));
<a name="l00444"></a>00444             numReactions = floor(min([
<a name="l00445"></a>00445                 totRNAs;
<a name="l00446"></a>00446                 this.randStream.stochasticRound((this.enzymes * this.stepSizeSec) ./ this.catalysisMatrix(:, unprocessedRNAIndexs(1)));
<a name="l00447"></a>00447                 substrates ...
<a name="l00448"></a>00448                 ]));
<a name="l00449"></a>00449             <span class="keywordflow">if</span> numReactions &lt;= 0
<a name="l00450"></a>00450                 <span class="keywordflow">return</span>;
<a name="l00451"></a>00451             end
<a name="l00452"></a>00452             
<a name="l00453"></a>00453             %randomly select RNAs to process
<a name="l00454"></a>00454             processingRNAs = this.randStream.randCounts(this.unprocessedRNAs(unprocessedRNAIndexs), numReactions);
<a name="l00455"></a>00455             
<a name="l00456"></a>00456             %increment processed RNAs
<a name="l00457"></a>00457             this.processedRNAs = this.processedRNAs + ...
<a name="l00458"></a>00458                 this.rna.nascentRNAMatureRNAComposition(:, unprocessedRNAIndexs) * processingRNAs;
<a name="l00459"></a>00459             
<a name="l00460"></a>00460             %increment intergenic RNAs
<a name="l00461"></a>00461             this.intergenicRNAs = this.intergenicRNAs + ...
<a name="l00462"></a>00462                 this.rna.intergenicRNAMatrix(:, unprocessedRNAIndexs) * processingRNAs;
<a name="l00463"></a>00463             
<a name="l00464"></a>00464             %decrement unprocessed RNAs
<a name="l00465"></a>00465             this.unprocessedRNAs(unprocessedRNAIndexs) = ...
<a name="l00466"></a>00466                 this.unprocessedRNAs(unprocessedRNAIndexs) - processingRNAs;
<a name="l00467"></a>00467             
<a name="l00468"></a>00468             %decrement reactants, increment byproducts
<a name="l00469"></a>00469             this.substrates = this.substrates + ...
<a name="l00470"></a>00470                 this.reactantByproductMatrix(:, unprocessedRNAIndexs) * processingRNAs;
<a name="l00471"></a>00471         end
<a name="l00472"></a>00472     end
<a name="l00473"></a>00473     
<a name="l00474"></a>00474     %<span class="keyword">get</span> methods of dependent local state
<a name="l00475"></a>00475     methods
<a name="l00476"></a>00476         <span class="keyword">function</span> value = getDryWeight(<span class="keyword">this</span>)
<a name="l00477"></a>00477             if size(this.unprocessedRNAs, 3) == 1
<a name="l00478"></a>00478                 value = this.getDryWeight@edu.stanford.covert.cell.sim.Process() + (...
<a name="l00479"></a>00479                     this.rna.molecularWeights(this.rna.nascentIndexs)&#39;    * this.unprocessedRNAs + ...
<a name="l00480"></a>00480                     this.rna.molecularWeights(this.rna.processedIndexs)&#39;  * this.processedRNAs   + ...
<a name="l00481"></a>00481                     this.rna.molecularWeights(this.rna.intergenicIndexs)&#39; * this.intergenicRNAs) ...
<a name="l00482"></a>00482                     / edu.stanford.covert.util.<a class="code" href="class_constant_util.html">ConstantUtil</a>.nAvogadro;
<a name="l00483"></a>00483             else
<a name="l00484"></a>00484                 value = this.getDryWeight@edu.stanford.covert.cell.sim.Process() + (...
<a name="l00485"></a>00485                     permute(this.rna.molecularWeights(this.rna.nascentIndexs)&#39;    * permute(this.unprocessedRNAs,[1 3 2]),[1 3 2]) + ...
<a name="l00486"></a>00486                     permute(this.rna.molecularWeights(this.rna.processedIndexs)&#39;  * permute(this.processedRNAs,  [1 3 2]),[1 3 2]) + ...
<a name="l00487"></a>00487                     permute(this.rna.molecularWeights(this.rna.intergenicIndexs)&#39; * permute(this.intergenicRNAs, [1 3 2]),[1 3 2])) ...
<a name="l00488"></a>00488                     / edu.stanford.covert.util.<a class="code" href="class_constant_util.html">ConstantUtil</a>.nAvogadro;
<a name="l00489"></a>00489             end
<a name="l00490"></a>00490         end
<a name="l00491"></a>00491     end
<a name="l00492"></a>00492 end
</pre></div></div><!-- contents -->
</div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
<a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(0)"><span class="SelectionMark">&#160;</span>All</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(1)"><span class="SelectionMark">&#160;</span>Classes</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(2)"><span class="SelectionMark">&#160;</span>Files</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(3)"><span class="SelectionMark">&#160;</span>Functions</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(4)"><span class="SelectionMark">&#160;</span>Variables</a></div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

  <div id="nav-path" class="navpath">
    <ul>
      <li class="navelem"><a class="el" href="_r_n_a_processing_8m.html">RNAProcessing.m</a>      </li>

    <li class="footer">Generated on Wed Feb 8 2012 21:42:32 for Whole Cell Simulation by
    <a href="http://www.doxygen.org/index.html">
    <img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.7.6.1 </li>
   </ul>
 </div>


</body>
</html>
