<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<title>Whole Cell Simulation: src/+edu/+stanford/+covert/+cell/+sim/+analysis/SimulationStructure.m Source File</title>

<link href="tabs.css" rel="stylesheet" type="text/css"/>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
<link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="navtree.js"></script>
<script type="text/javascript">
  $(document).ready(initResizable);
</script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
  $(document).ready(function() { searchBox.OnSelectItem(0); });
</script>

</head>
<body>
<div id="top"><!-- do not remove this div! -->


<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  
  
  <td style="padding-left: 0.5em;">
   <div id="projectname">Whole Cell Simulation
   
   </div>
   
  </td>
  
  
  
 </tr>
 </tbody>
</table>
</div>

<!-- Generated by Doxygen 1.7.6.1 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "search",false,'Search');
</script>
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&#160;Page</span></a></li>
      <li><a href="annotated.html"><span>Classes</span></a></li>
      <li class="current"><a href="files.html"><span>Files</span></a></li>
      <li>
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
  <div id="navrow2" class="tabs2">
    <ul class="tablist">
      <li><a href="files.html"><span>File&#160;List</span></a></li>
      <li><a href="globals.html"><span>File&#160;Members</span></a></li>
    </ul>
  </div>
</div>
<div id="side-nav" class="ui-resizable side-nav-resizable">
  <div id="nav-tree">
    <div id="nav-tree-contents">
    </div>
  </div>
  <div id="splitbar" style="-moz-user-select:none;" 
       class="ui-resizable-handle">
  </div>
</div>
<script type="text/javascript">
  initNavTree('_simulation_structure_8m.html','');
</script>
<div id="doc-content">
<div class="header">
  <div class="headertitle">
<div class="title">src/+edu/+stanford/+covert/+cell/+sim/+analysis/SimulationStructure.m</div>  </div>
</div><!--header-->
<div class="contents">
<a href="_simulation_structure_8m.html">Go to the documentation of this file.</a><div class="fragment"><pre class="fragment"><a name="l00001"></a>00001 %<a class="code" href="class_simulation_structure.html">SimulationStructure</a>
<a name="l00002"></a>00002 % Outputs analysis of the structure of the simulation:
<a name="l00003"></a>00003 % - which processes use which metabolites / gene products / enzymes
<a name="l00004"></a>00004 % - which metabolites / gene products / enzymes aren&#39;t used by any process
<a name="l00005"></a>00005 %
<a name="l00006"></a>00006 % Author: Jonathan Karr, jkarr@stanford.edu
<a name="l00007"></a>00007 % Affiliation: Covert Lab, Department of Bioengineering, Stanford University
<a name="l00008"></a>00008 % Last Updated: 9/12/2010
<a name="l00009"></a><a class="code" href="class_simulation_structure.html">00009</a> classdef SimulationStructure
<a name="l00010"></a>00010     methods (Static = true)        
<a name="l00011"></a>00011         function run(sim, fileName)
<a name="l00012"></a>00012             import edu.stanford.covert.cell.sim.analysis.SimulationStructure;
<a name="l00013"></a>00013             <span class="keyword">import</span> edu.stanford.covert.cell.sim.util.PrintUtil;
<a name="l00014"></a>00014             
<a name="l00015"></a>00015             % excel file
<a name="l00016"></a>00016             [content, colLabels] = <a class="code" href="class_simulation_structure.html">SimulationStructure</a>.<a class="code" href="class_simulation_structure.html#a81eadcd51ae837bb9c8551c8b5c88d47">processMetabolites</a>(sim);
<a name="l00017"></a>00017             <span class="keywordflow">if</span> nargin == 1
<a name="l00018"></a>00018                 PrintUtil.printToStdIO(content, colLabels);
<a name="l00019"></a>00019             <span class="keywordflow">else</span>
<a name="l00020"></a>00020                 PrintUtil.printToFile(content, colLabels, [fileName <span class="stringliteral">&#39;.xls&#39;</span>], <span class="stringliteral">&#39;Process-Metabolites&#39;</span>);
<a name="l00021"></a>00021             end
<a name="l00022"></a>00022             
<a name="l00023"></a>00023             [content, colLabels] = <a class="code" href="class_simulation_structure.html">SimulationStructure</a>.<a class="code" href="class_simulation_structure.html#a90b1a23a4b899b21761fb7ff1a2f10dc">processGeneProducts</a>(sim);
<a name="l00024"></a>00024             <span class="keywordflow">if</span> nargin == 1
<a name="l00025"></a>00025                 PrintUtil.printToStdIO(content, colLabels);
<a name="l00026"></a>00026             <span class="keywordflow">else</span>
<a name="l00027"></a>00027                 PrintUtil.printToFile(content, colLabels, [fileName <span class="stringliteral">&#39;.xls&#39;</span>], <span class="stringliteral">&#39;Process-Gene Product&#39;</span>);
<a name="l00028"></a>00028             end
<a name="l00029"></a>00029             
<a name="l00030"></a>00030             [content, colLabels] = <a class="code" href="class_simulation_structure.html">SimulationStructure</a>.<a class="code" href="class_simulation_structure.html#a5b0c337cc5d9dd42c45ba4954c87896e">unincludedMetabolites</a>(sim);
<a name="l00031"></a>00031             <span class="keywordflow">if</span> nargin == 1
<a name="l00032"></a>00032                 PrintUtil.printToStdIO(content, colLabels);
<a name="l00033"></a>00033             <span class="keywordflow">else</span>
<a name="l00034"></a>00034                 PrintUtil.printToFile(content, colLabels, [fileName <span class="stringliteral">&#39;.xls&#39;</span>], <span class="stringliteral">&#39;Unincluded Metabolites&#39;</span>);
<a name="l00035"></a>00035             end
<a name="l00036"></a>00036             
<a name="l00037"></a>00037             [content, colLabels] = <a class="code" href="class_simulation_structure.html">SimulationStructure</a>.<a class="code" href="class_simulation_structure.html#a9d0f644afe3991ffa176ca9c223547cb">unincludedGenes</a>(sim);
<a name="l00038"></a>00038             <span class="keywordflow">if</span> nargin == 1
<a name="l00039"></a>00039                 PrintUtil.printToStdIO(content, colLabels);
<a name="l00040"></a>00040             <span class="keywordflow">else</span>
<a name="l00041"></a>00041                 PrintUtil.printToFile(content, colLabels, [fileName <span class="stringliteral">&#39;.xls&#39;</span>], <span class="stringliteral">&#39;Unincluded Genes&#39;</span>);
<a name="l00042"></a>00042             end
<a name="l00043"></a>00043             
<a name="l00044"></a>00044             [content, colLabels] = <a class="code" href="class_simulation_structure.html">SimulationStructure</a>.<a class="code" href="class_simulation_structure.html#a974a211686398659a120c064c7cf8a4b">unincludedProteins</a>(sim);
<a name="l00045"></a>00045             <span class="keywordflow">if</span> nargin == 1
<a name="l00046"></a>00046                 PrintUtil.printToStdIO(content, colLabels);
<a name="l00047"></a>00047             <span class="keywordflow">else</span>
<a name="l00048"></a>00048                 PrintUtil.printToFile(content, colLabels, [fileName <span class="stringliteral">&#39;.xls&#39;</span>], <span class="stringliteral">&#39;Unincluded Proteins&#39;</span>);
<a name="l00049"></a>00049             end
<a name="l00050"></a>00050             
<a name="l00051"></a>00051             % plots
<a name="l00052"></a>00052             <span class="keywordflow">if</span> nargin == 1
<a name="l00053"></a>00053                 <a class="code" href="class_simulation_structure.html">SimulationStructure</a>.<a class="code" href="class_simulation_structure.html#a6b5a93f145fb80e89bff13055381e8a9">plotProcessMetaboliteSharing</a>(sim);
<a name="l00054"></a>00054                 <a class="code" href="class_simulation_structure.html">SimulationStructure</a>.<a class="code" href="class_simulation_structure.html#ab20ac4b4a3d413c31344012765100aa1">plotProcessMetabolites</a>(sim);
<a name="l00055"></a>00055                 <a class="code" href="class_simulation_structure.html">SimulationStructure</a>.<a class="code" href="class_simulation_structure.html#a63432ec9f96e4c4bd8e76840a9b49864">plotProcessSharedMetabolites</a>(sim);
<a name="l00056"></a>00056                 <a class="code" href="class_simulation_structure.html">SimulationStructure</a>.<a class="code" href="class_simulation_structure.html#acebbb38b71c45277a6800498946c8d10">plotProcessGeneProducts</a>(sim);
<a name="l00057"></a>00057             <span class="keywordflow">else</span>
<a name="l00058"></a>00058                 [~, figHandle] = edu.stanford.covert.cell.sim.util.PlotUtil.newAxesHandle();
<a name="l00059"></a>00059                 
<a name="l00060"></a>00060                 clf(figHandle);
<a name="l00061"></a>00061                 <a class="code" href="class_simulation_structure.html">SimulationStructure</a>.<a class="code" href="class_simulation_structure.html#a6b5a93f145fb80e89bff13055381e8a9">plotProcessMetaboliteSharing</a>(sim, figHandle);
<a name="l00062"></a>00062                 saveas(figHandle, [fileName <span class="stringliteral">&#39;-ProcessMetaboliteSharing.pdf&#39;</span>]);
<a name="l00063"></a>00063                 
<a name="l00064"></a>00064                 clf(figHandle);
<a name="l00065"></a>00065                 <a class="code" href="class_simulation_structure.html">SimulationStructure</a>.<a class="code" href="class_simulation_structure.html#ab20ac4b4a3d413c31344012765100aa1">plotProcessMetabolites</a>(sim, figHandle);
<a name="l00066"></a>00066                 saveas(figHandle, [fileName <span class="stringliteral">&#39;-ProcessMetabolites.pdf&#39;</span>]);
<a name="l00067"></a>00067                 
<a name="l00068"></a>00068                 clf(figHandle);
<a name="l00069"></a>00069                 <a class="code" href="class_simulation_structure.html">SimulationStructure</a>.<a class="code" href="class_simulation_structure.html#a63432ec9f96e4c4bd8e76840a9b49864">plotProcessSharedMetabolites</a>(sim, figHandle);
<a name="l00070"></a>00070                 saveas(figHandle, [fileName <span class="stringliteral">&#39;-ProcessSharedMetabolites.pdf&#39;</span>]);
<a name="l00071"></a>00071                 
<a name="l00072"></a>00072                 clf(figHandle);
<a name="l00073"></a>00073                 <a class="code" href="class_simulation_structure.html">SimulationStructure</a>.<a class="code" href="class_simulation_structure.html#acebbb38b71c45277a6800498946c8d10">plotProcessGeneProducts</a>(sim, figHandle);
<a name="l00074"></a>00074                 saveas(figHandle, [fileName <span class="stringliteral">&#39;-ProcessGeneProducts.pdf&#39;</span>]);
<a name="l00075"></a>00075                 
<a name="l00076"></a>00076                 close(figHandle);
<a name="l00077"></a>00077             end
<a name="l00078"></a>00078         end
<a name="l00079"></a>00079     end
<a name="l00080"></a>00080     
<a name="l00081"></a>00081     %analysis
<a name="l00082"></a>00082     methods (Static)
<a name="l00083"></a>00083         %which processes use which metabolites
<a name="l00084"></a>00084         <span class="keyword">function</span> [content, colLabels] = processMetabolites(sim)
<a name="l00085"></a>00085             colLabels = {<span class="stringliteral">&#39;Process ID&#39;</span>, <span class="stringliteral">&#39;Process Name&#39;</span>, <span class="stringliteral">&#39;Metabolite ID&#39;</span>, <span class="stringliteral">&#39;Metabolite Name&#39;</span>};
<a name="l00086"></a>00086             content = cell(0, numel(colLabels));
<a name="l00087"></a>00087             
<a name="l00088"></a>00088             <span class="keywordflow">for</span> i = 1:length(sim.processes)
<a name="l00089"></a>00089                 m = sim.processes{i};
<a name="l00090"></a>00090                 
<a name="l00091"></a>00091                 idxs = m.substrateMetaboliteLocalIndexs;
<a name="l00092"></a>00092                 
<a name="l00093"></a>00093                 content = [content;
<a name="l00094"></a>00094                     repmat({m.wholeCellModelID m.name}, numel(idxs), 1) ...
<a name="l00095"></a>00095                     m.substrateWholeCellModelIDs(idxs) ...
<a name="l00096"></a>00096                     m.substrateNames(idxs)]; %#ok&lt;AGROW&gt;
<a name="l00097"></a>00097             end
<a name="l00098"></a>00098         end
<a name="l00099"></a>00099         
<a name="l00100"></a>00100         %which metabolites aren<span class="stringliteral">&#39;t used in any processes</span>
<a name="l00101"></a>00101 <span class="stringliteral">        function [content, colLabels] = unincludedMetabolites(sim)</span>
<a name="l00102"></a>00102 <span class="stringliteral">            colLabels = {&#39;</span><a class="code" href="class_metabolite.html">Metabolite</a> ID<span class="stringliteral">&#39;, &#39;</span><a class="code" href="class_metabolite.html">Metabolite</a> Name<span class="stringliteral">&#39;};</span>
<a name="l00103"></a>00103 <span class="stringliteral">            </span>
<a name="l00104"></a>00104 <span class="stringliteral">            included = edu.stanford.covert.cell.sim.analysis.SimulationStructure.processMetabolites(sim);</span>
<a name="l00105"></a>00105 <span class="stringliteral">            </span>
<a name="l00106"></a>00106 <span class="stringliteral">            [~, idxs] = setdiff(sim.state(&#39;</span><a class="code" href="class_metabolite.html">Metabolite</a><span class="stringliteral">&#39;).wholeCellModelIDs, unique(included(:,3)));</span>
<a name="l00107"></a>00107 <span class="stringliteral">            content = [</span>
<a name="l00108"></a>00108 <span class="stringliteral">                sim.state(&#39;</span><a class="code" href="class_metabolite.html">Metabolite</a><span class="stringliteral">&#39;).wholeCellModelIDs(idxs) ...</span>
<a name="l00109"></a>00109 <span class="stringliteral">                sim.state(&#39;</span><a class="code" href="class_metabolite.html">Metabolite</a><span class="stringliteral">&#39;).names(idxs)];</span>
<a name="l00110"></a>00110 <span class="stringliteral">        end</span>
<a name="l00111"></a>00111 <span class="stringliteral">        </span>
<a name="l00112"></a>00112 <span class="stringliteral">        %which processes use which gene products, and in which compartments, as enzymes</span>
<a name="l00113"></a>00113 <span class="stringliteral">        function [content, colLabels] = processGeneProducts(sim)</span>
<a name="l00114"></a>00114 <span class="stringliteral">            colLabels = {&#39;</span><a class="code" href="class_process.html">Process</a> ID<span class="stringliteral">&#39;, &#39;</span><a class="code" href="class_process.html">Process</a> Name<span class="stringliteral">&#39;, &#39;</span><a class="code" href="class_gene.html">Gene</a> ID<span class="stringliteral">&#39;, &#39;</span><a class="code" href="class_gene.html">Gene</a> Name<span class="stringliteral">&#39;, &#39;</span>Product ID<span class="stringliteral">&#39;, &#39;</span>Product Name<span class="stringliteral">&#39;, &#39;</span>Product <a class="code" href="class_compartment.html">Compartment</a> ID<span class="stringliteral">&#39;, &#39;</span>Product <a class="code" href="class_compartment.html">Compartment</a> Name<span class="stringliteral">&#39;, &#39;</span>Implemented<span class="stringliteral">&#39;};</span>
<a name="l00115"></a>00115 <span class="stringliteral">            content = cell(0, numel(colLabels));</span>
<a name="l00116"></a>00116 <span class="stringliteral">            </span>
<a name="l00117"></a>00117 <span class="stringliteral">            for i = 1:length(sim.processes)</span>
<a name="l00118"></a>00118 <span class="stringliteral">                m = sim.processes{i};</span>
<a name="l00119"></a>00119 <span class="stringliteral">                </span>
<a name="l00120"></a>00120 <span class="stringliteral">                %stimuli</span>
<a name="l00121"></a>00121 <span class="stringliteral">                stimuliGeneComposition = m.stimuliGeneComposition();</span>
<a name="l00122"></a>00122 <span class="stringliteral">                for j = 1:length(sim.gene.wholeCellModelIDs)</span>
<a name="l00123"></a>00123 <span class="stringliteral">                    for k = 1:length(m.stimuliWholeCellModelIDs)</span>
<a name="l00124"></a>00124 <span class="stringliteral">                        if ~stimuliGeneComposition(j, k); continue; end;</span>
<a name="l00125"></a>00125 <span class="stringliteral">                        </span>
<a name="l00126"></a>00126 <span class="stringliteral">                        if size(m.stimuliCompartments, 2) == 1</span>
<a name="l00127"></a>00127 <span class="stringliteral">                            compartmentWholeCellModelID = sim.compartment.wholeCellModelIDs{m.stimuliCompartments(k)};</span>
<a name="l00128"></a>00128 <span class="stringliteral">                            compartmentName = sim.compartment.names{m.stimuliCompartments(k)};</span>
<a name="l00129"></a>00129 <span class="stringliteral">                        else</span>
<a name="l00130"></a>00130 <span class="stringliteral">                            compartmentWholeCellModelID = cell(1,1);</span>
<a name="l00131"></a>00131 <span class="stringliteral">                            compartmentName = cell(1,1);</span>
<a name="l00132"></a>00132 <span class="stringliteral">                        end</span>
<a name="l00133"></a>00133 <span class="stringliteral">                        </span>
<a name="l00134"></a>00134 <span class="stringliteral">                        content = [content;</span>
<a name="l00135"></a>00135 <span class="stringliteral">                            m.wholeCellModelID            m.name            ...</span>
<a name="l00136"></a>00136 <span class="stringliteral">                            sim.gene.wholeCellModelIDs(j) sim.gene.names(j)  ...</span>
<a name="l00137"></a>00137 <span class="stringliteral">                            m.stimuliWholeCellModelIDs(k) m.stimuliNames(k) ...</span>
<a name="l00138"></a>00138 <span class="stringliteral">                            compartmentWholeCellModelID   compartmentName   ...</span>
<a name="l00139"></a>00139 <span class="stringliteral">                            &#39;</span>Stimulus<span class="stringliteral">&#39;]; %#ok&lt;AGROW&gt;</span>
<a name="l00140"></a>00140 <span class="stringliteral">                    end</span>
<a name="l00141"></a>00141 <span class="stringliteral">                end</span>
<a name="l00142"></a>00142 <span class="stringliteral">                </span>
<a name="l00143"></a>00143 <span class="stringliteral">                %substrates</span>
<a name="l00144"></a>00144 <span class="stringliteral">                substrateGeneComposition = m.substrateGeneComposition();</span>
<a name="l00145"></a>00145 <span class="stringliteral">                for j = 1:length(sim.gene.wholeCellModelIDs)</span>
<a name="l00146"></a>00146 <span class="stringliteral">                    for k = 1:length(m.substrateWholeCellModelIDs)</span>
<a name="l00147"></a>00147 <span class="stringliteral">                        if ~substrateGeneComposition(j, k); continue; end;</span>
<a name="l00148"></a>00148 <span class="stringliteral">                        </span>
<a name="l00149"></a>00149 <span class="stringliteral">                        if size(m.substrateCompartments, 2) == 1</span>
<a name="l00150"></a>00150 <span class="stringliteral">                            compartmentWholeCellModelID = sim.compartment.wholeCellModelIDs{m.substrateCompartments(k)};</span>
<a name="l00151"></a>00151 <span class="stringliteral">                            compartmentName = sim.compartment.names{m.substrateCompartments(k)};</span>
<a name="l00152"></a>00152 <span class="stringliteral">                        else</span>
<a name="l00153"></a>00153 <span class="stringliteral">                            compartmentWholeCellModelID = cell(1,1);</span>
<a name="l00154"></a>00154 <span class="stringliteral">                            compartmentName = cell(1,1);</span>
<a name="l00155"></a>00155 <span class="stringliteral">                        end</span>
<a name="l00156"></a>00156 <span class="stringliteral">                        </span>
<a name="l00157"></a>00157 <span class="stringliteral">                        content = [content;</span>
<a name="l00158"></a>00158 <span class="stringliteral">                            m.wholeCellModelID              m.name            ...</span>
<a name="l00159"></a>00159 <span class="stringliteral">                            sim.gene.wholeCellModelIDs(j)   sim.gene.names(j)  ...</span>
<a name="l00160"></a>00160 <span class="stringliteral">                            m.substrateWholeCellModelIDs(k) m.substrateNames(k) ...</span>
<a name="l00161"></a>00161 <span class="stringliteral">                            compartmentWholeCellModelID     compartmentName   ...</span>
<a name="l00162"></a>00162 <span class="stringliteral">                            &#39;</span>Substrate<span class="stringliteral">&#39;]; %#ok&lt;AGROW&gt;</span>
<a name="l00163"></a>00163 <span class="stringliteral">                    end</span>
<a name="l00164"></a>00164 <span class="stringliteral">                end</span>
<a name="l00165"></a>00165 <span class="stringliteral">                </span>
<a name="l00166"></a>00166 <span class="stringliteral">                %enzymes</span>
<a name="l00167"></a>00167 <span class="stringliteral">                enzymeGeneComposition = m.enzymeGeneComposition();</span>
<a name="l00168"></a>00168 <span class="stringliteral">                for j = 1:length(sim.gene.wholeCellModelIDs)</span>
<a name="l00169"></a>00169 <span class="stringliteral">                    for k = 1:length(m.enzymeWholeCellModelIDs)</span>
<a name="l00170"></a>00170 <span class="stringliteral">                        if ~enzymeGeneComposition(j, k); continue; end;</span>
<a name="l00171"></a>00171 <span class="stringliteral">                        </span>
<a name="l00172"></a>00172 <span class="stringliteral">                        if size(m.enzymeCompartments, 2) == 1</span>
<a name="l00173"></a>00173 <span class="stringliteral">                            compartmentWholeCellModelID = sim.compartment.wholeCellModelIDs{m.enzymeCompartments(k)};</span>
<a name="l00174"></a>00174 <span class="stringliteral">                            compartmentName = sim.compartment.names{m.enzymeCompartments(k)};</span>
<a name="l00175"></a>00175 <span class="stringliteral">                        else</span>
<a name="l00176"></a>00176 <span class="stringliteral">                            compartmentWholeCellModelID = cell(1,1);</span>
<a name="l00177"></a>00177 <span class="stringliteral">                            compartmentName = cell(1,1);</span>
<a name="l00178"></a>00178 <span class="stringliteral">                        end</span>
<a name="l00179"></a>00179 <span class="stringliteral">                        </span>
<a name="l00180"></a>00180 <span class="stringliteral">                        content = [content;</span>
<a name="l00181"></a>00181 <span class="stringliteral">                            m.wholeCellModelID            m.name           ...</span>
<a name="l00182"></a>00182 <span class="stringliteral">                            sim.gene.wholeCellModelIDs(j) sim.gene.names(j) ...</span>
<a name="l00183"></a>00183 <span class="stringliteral">                            m.enzymeWholeCellModelIDs(k)  m.enzymeNames(k) ...</span>
<a name="l00184"></a>00184 <span class="stringliteral">                            compartmentWholeCellModelID   compartmentName  ...</span>
<a name="l00185"></a>00185 <span class="stringliteral">                            &#39;</span><a class="code" href="class_enzyme.html">Enzyme</a><span class="stringliteral">&#39;]; %#ok&lt;AGROW&gt;</span>
<a name="l00186"></a>00186 <span class="stringliteral">                    end</span>
<a name="l00187"></a>00187 <span class="stringliteral">                end</span>
<a name="l00188"></a>00188 <span class="stringliteral">            end</span>
<a name="l00189"></a>00189 <span class="stringliteral">        end</span>
<a name="l00190"></a>00190 <span class="stringliteral">        </span>
<a name="l00191"></a>00191 <span class="stringliteral">        %gene whose products aren&#39;</span>t functionally used in any process
<a name="l00192"></a>00192         <span class="keyword">function</span> [content, colLabels] = unincludedGenes(sim)
<a name="l00193"></a>00193             colLabels = {<span class="stringliteral">&#39;Gene ID&#39;</span>,<span class="stringliteral">&#39;Gene Name&#39;</span>};
<a name="l00194"></a>00194             
<a name="l00195"></a>00195             included = edu.stanford.covert.cell.sim.analysis.SimulationStructure.processGeneProducts(sim);
<a name="l00196"></a>00196             
<a name="l00197"></a>00197             [~, idxs] = setdiff(sim.gene.wholeCellModelIDs, unique(included(:,3)));
<a name="l00198"></a>00198             content = [
<a name="l00199"></a>00199                 sim.gene.wholeCellModelIDs(idxs) ...
<a name="l00200"></a>00200                 sim.gene.names(idxs)];
<a name="l00201"></a>00201         end
<a name="l00202"></a>00202         
<a name="l00203"></a>00203         %protein monomers and complexes which aren<span class="stringliteral">&#39;t used in any process</span>
<a name="l00204"></a>00204 <span class="stringliteral">        function [content, colLabels] = unincludedProteins(sim)</span>
<a name="l00205"></a>00205 <span class="stringliteral">            pm = sim.state(&#39;</span><a class="code" href="class_protein_monomer.html">ProteinMonomer</a><span class="stringliteral">&#39;);</span>
<a name="l00206"></a>00206 <span class="stringliteral">            pc = sim.state(&#39;</span><a class="code" href="class_protein_complex.html">ProteinComplex</a><span class="stringliteral">&#39;);</span>
<a name="l00207"></a>00207 <span class="stringliteral">            </span>
<a name="l00208"></a>00208 <span class="stringliteral">            colLabels = {&#39;</span><a class="code" href="class_protein.html">Protein</a> ID<span class="stringliteral">&#39;, &#39;</span><a class="code" href="class_protein.html">Protein</a> Name<span class="stringliteral">&#39;};</span>
<a name="l00209"></a>00209 <span class="stringliteral">            </span>
<a name="l00210"></a>00210 <span class="stringliteral">            includedMonomers = {};</span>
<a name="l00211"></a>00211 <span class="stringliteral">            includedComplexs = {};</span>
<a name="l00212"></a>00212 <span class="stringliteral">            for i = 1:length(sim.processes)</span>
<a name="l00213"></a>00213 <span class="stringliteral">                m = sim.processes{i};</span>
<a name="l00214"></a>00214 <span class="stringliteral">                </span>
<a name="l00215"></a>00215 <span class="stringliteral">                includedMonomers = [includedMonomers; m.stimuliWholeCellModelIDs(m.stimulusMonomerLocalIndexs)]; %#ok&lt;AGROW&gt;</span>
<a name="l00216"></a>00216 <span class="stringliteral">                includedMonomers = [includedMonomers; m.substrateWholeCellModelIDs(m.substrateMonomerLocalIndexs)]; %#ok&lt;AGROW&gt;</span>
<a name="l00217"></a>00217 <span class="stringliteral">                includedMonomers = [includedMonomers; m.enzymeWholeCellModelIDs(m.enzymeMonomerLocalIndexs)]; %#ok&lt;AGROW&gt;</span>
<a name="l00218"></a>00218 <span class="stringliteral">                </span>
<a name="l00219"></a>00219 <span class="stringliteral">                includedComplexs = [includedComplexs; m.stimuliWholeCellModelIDs(m.stimulusComplexLocalIndexs)]; %#ok&lt;AGROW&gt;</span>
<a name="l00220"></a>00220 <span class="stringliteral">                includedComplexs = [includedComplexs; m.substrateWholeCellModelIDs(m.substrateComplexLocalIndexs)]; %#ok&lt;AGROW&gt;</span>
<a name="l00221"></a>00221 <span class="stringliteral">                includedComplexs = [includedComplexs; m.enzymeWholeCellModelIDs(m.enzymeComplexLocalIndexs)]; %#ok&lt;AGROW&gt;</span>
<a name="l00222"></a>00222 <span class="stringliteral">            end</span>
<a name="l00223"></a>00223 <span class="stringliteral">            </span>
<a name="l00224"></a>00224 <span class="stringliteral">            [~, unincludedMonomerIdxs] = setdiff(pm.wholeCellModelIDs(pm.matureIndexs), unique(includedMonomers));</span>
<a name="l00225"></a>00225 <span class="stringliteral">            [~, unincludedComplexIdxs] = setdiff(pc.wholeCellModelIDs(pc.matureIndexs), unique(includedComplexs));</span>
<a name="l00226"></a>00226 <span class="stringliteral">            </span>
<a name="l00227"></a>00227 <span class="stringliteral">            unincludedMonomerIdxs(any(any(pc.proteinComplexComposition(sim.gene.mRNAIndexs(unincludedMonomerIdxs), setdiff(1:end, unincludedComplexIdxs), :),3),2)) = [];</span>
<a name="l00228"></a>00228 <span class="stringliteral">            </span>
<a name="l00229"></a>00229 <span class="stringliteral">            content = [</span>
<a name="l00230"></a>00230 <span class="stringliteral">                pm.wholeCellModelIDs(pm.matureIndexs(unincludedMonomerIdxs)) ...</span>
<a name="l00231"></a>00231 <span class="stringliteral">                pm.names(pm.matureIndexs(unincludedMonomerIdxs));</span>
<a name="l00232"></a>00232 <span class="stringliteral">                pc.wholeCellModelIDs(pc.matureIndexs(unincludedComplexIdxs)) ...</span>
<a name="l00233"></a>00233 <span class="stringliteral">                pc.names(pc.matureIndexs(unincludedComplexIdxs))];</span>
<a name="l00234"></a>00234 <span class="stringliteral">        end</span>
<a name="l00235"></a>00235 <span class="stringliteral">    end</span>
<a name="l00236"></a>00236 <span class="stringliteral">    </span>
<a name="l00237"></a>00237 <span class="stringliteral">    %plots</span>
<a name="l00238"></a>00238 <span class="stringliteral">    methods (Static)</span>
<a name="l00239"></a>00239 <span class="stringliteral">        function plotProcessMetaboliteSharing(sim, figHandle)</span>
<a name="l00240"></a>00240 <span class="stringliteral">            import edu.stanford.covert.cell.sim.analysis.SimulationStructure;</span>
<a name="l00241"></a>00241 <span class="stringliteral">            </span>
<a name="l00242"></a>00242 <span class="stringliteral">            content = SimulationStructure.processMetabolites(sim);</span>
<a name="l00243"></a>00243 <span class="stringliteral">            [~, modIdxs] = ismember(content(:, 1), sim.processWholeCellModelIDs);</span>
<a name="l00244"></a>00244 <span class="stringliteral">            [~, metIdxs] = ismember(content(:, 3), sim.state(&#39;</span><a class="code" href="class_metabolite.html">Metabolite</a><span class="stringliteral">&#39;).wholeCellModelIDs);</span>
<a name="l00245"></a>00245 <span class="stringliteral">            mat = zeros(numel(sim.processWholeCellModelIDs), numel(sim.state(&#39;</span><a class="code" href="class_metabolite.html">Metabolite</a><span class="stringliteral">&#39;).wholeCellModelIDs));</span>
<a name="l00246"></a>00246 <span class="stringliteral">            mat(sub2ind(size(mat), modIdxs, metIdxs)) = 1;</span>
<a name="l00247"></a>00247 <span class="stringliteral">            </span>
<a name="l00248"></a>00248 <span class="stringliteral">            mat2 = ones(numel(sim.processWholeCellModelIDs));</span>
<a name="l00249"></a>00249 <span class="stringliteral">            for i = 1:size(mat,1)</span>
<a name="l00250"></a>00250 <span class="stringliteral">                for j = 1:size(mat,1)</span>
<a name="l00251"></a>00251 <span class="stringliteral">                    mat2(i,j) = 1 + double(any(mat(i, :) &amp; mat(j, :)));</span>
<a name="l00252"></a>00252 <span class="stringliteral">                end</span>
<a name="l00253"></a>00253 <span class="stringliteral">            end</span>
<a name="l00254"></a>00254 <span class="stringliteral">            mat2(sub2ind(size(mat2), 1:size(mat2,1), 1:size(mat2,1))) = ...</span>
<a name="l00255"></a>00255 <span class="stringliteral">                double(mat2(sub2ind(size(mat2), 1:size(mat2,1), 1:size(mat2,1)))==1) + ...</span>
<a name="l00256"></a>00256 <span class="stringliteral">                double(mat2(sub2ind(size(mat2), 1:size(mat2,1), 1:size(mat2,1)))~=1)*3;</span>
<a name="l00257"></a>00257 <span class="stringliteral">            </span>
<a name="l00258"></a>00258 <span class="stringliteral">            dist = pdist(mat2, &#39;</span>cityblock<span class="stringliteral">&#39;);</span>
<a name="l00259"></a>00259 <span class="stringliteral">            tree = linkage(dist, &#39;</span>average<span class="stringliteral">&#39;);</span>
<a name="l00260"></a>00260 <span class="stringliteral">            modOrder = optimalleaforder(tree, dist);</span>
<a name="l00261"></a>00261 <span class="stringliteral">            </span>
<a name="l00262"></a>00262 <span class="stringliteral">            %plot</span>
<a name="l00263"></a>00263 <span class="stringliteral">            if ~exist(&#39;</span>figHandle<span class="stringliteral">&#39;, &#39;</span>var<span class="stringliteral">&#39;)</span>
<a name="l00264"></a>00264 <span class="stringliteral">                axesHandle = edu.stanford.covert.cell.sim.util.PlotUtil.newAxesHandle();</span>
<a name="l00265"></a>00265 <span class="stringliteral">            else</span>
<a name="l00266"></a>00266 <span class="stringliteral">                axesHandle = subplot(1, 1, 1, &#39;</span>Parent<span class="stringliteral">&#39;, figHandle);</span>
<a name="l00267"></a>00267 <span class="stringliteral">            end</span>
<a name="l00268"></a>00268 <span class="stringliteral">            </span>
<a name="l00269"></a>00269 <span class="stringliteral">            image(mat2(modOrder, modOrder), &#39;</span>Parent<span class="stringliteral">&#39;, axesHandle);</span>
<a name="l00270"></a>00270 <span class="stringliteral">            colormap(axesHandle, [</span>
<a name="l00271"></a>00271 <span class="stringliteral">                1 1 1;</span>
<a name="l00272"></a>00272 <span class="stringliteral">                0 0 0;</span>
<a name="l00273"></a>00273 <span class="stringliteral">                0.5 0.5 0.5;</span>
<a name="l00274"></a>00274 <span class="stringliteral">                ]);</span>
<a name="l00275"></a>00275 <span class="stringliteral">            </span>
<a name="l00276"></a>00276 <span class="stringliteral">            processNames = SimulationStructure.getShortProcessNames(sim);</span>
<a name="l00277"></a>00277 <span class="stringliteral">            </span>
<a name="l00278"></a>00278 <span class="stringliteral">            for i=1:numel(sim.processWholeCellModelIDs)</span>
<a name="l00279"></a>00279 <span class="stringliteral">                text(0, i, processNames{modOrder(i)}, &#39;</span>Parent<span class="stringliteral">&#39;, axesHandle, &#39;</span>FontSize<span class="stringliteral">&#39;, 10, &#39;</span>Rotation<span class="stringliteral">&#39;, 0,   &#39;</span>HorizontalAlignment<span class="stringliteral">&#39;, &#39;</span>right<span class="stringliteral">&#39;, &#39;</span>VerticalAlignment<span class="charliteral">&#39;,&#39;</span>middle<span class="stringliteral">&#39;, &#39;</span>units<span class="stringliteral">&#39;, &#39;</span>data<span class="stringliteral">&#39;);</span>
<a name="l00280"></a>00280 <span class="stringliteral">                text(i, 0, processNames{modOrder(i)}, &#39;</span>Parent<span class="stringliteral">&#39;, axesHandle, &#39;</span>FontSize<span class="stringliteral">&#39;, 10, &#39;</span>Rotation<span class="stringliteral">&#39;, 270, &#39;</span>HorizontalAlignment<span class="stringliteral">&#39;, &#39;</span>right<span class="stringliteral">&#39;, &#39;</span>VerticalAlignment<span class="charliteral">&#39;,&#39;</span>middle<span class="stringliteral">&#39;, &#39;</span>units<span class="stringliteral">&#39;, &#39;</span>data<span class="stringliteral">&#39;);</span>
<a name="l00281"></a>00281 <span class="stringliteral">            end</span>
<a name="l00282"></a>00282 <span class="stringliteral">            </span>
<a name="l00283"></a>00283 <span class="stringliteral">            set(axesHandle, &#39;</span>XTick<span class="stringliteral">&#39;, 1:numel(processNames));</span>
<a name="l00284"></a>00284 <span class="stringliteral">            set(axesHandle, &#39;</span>YTick<span class="stringliteral">&#39;, 1:numel(processNames));</span>
<a name="l00285"></a>00285 <span class="stringliteral">            set(axesHandle, &#39;</span>XTickLabel<span class="stringliteral">&#39;, cell(size(processNames)));</span>
<a name="l00286"></a>00286 <span class="stringliteral">            set(axesHandle, &#39;</span>YTickLabel<span class="stringliteral">&#39;, cell(size(processNames)));</span>
<a name="l00287"></a>00287 <span class="stringliteral">            </span>
<a name="l00288"></a>00288 <span class="stringliteral">            ylabel(&#39;</span><a class="code" href="class_process.html">Process</a><span class="stringliteral">&#39;, &#39;</span>FontSize<span class="stringliteral">&#39;, 18);</span>
<a name="l00289"></a>00289 <span class="stringliteral">            xlabel(&#39;</span><a class="code" href="class_process.html">Process</a><span class="stringliteral">&#39;, &#39;</span>FontSize<span class="stringliteral">&#39;, 18);</span>
<a name="l00290"></a>00290 <span class="stringliteral">            set(axesHandle, &#39;</span>XAxisLocation<span class="charliteral">&#39;,&#39;</span>top<span class="stringliteral">&#39;)</span>
<a name="l00291"></a>00291 <span class="stringliteral">            axesPos = get(axesHandle&#39;</span>, <span class="stringliteral">&#39;Position&#39;</span>);
<a name="l00292"></a>00292             <span class="keyword">set</span>(axesHandle, <span class="stringliteral">&#39;Position&#39;</span>, axesPos + [.11 -.1 -.1 -.1]);
<a name="l00293"></a>00293             xLabelPos = <span class="keyword">get</span>(<span class="keyword">get</span>(axesHandle,<span class="stringliteral">&#39;XLabel&#39;</span>), <span class="stringliteral">&#39;Position&#39;</span>);
<a name="l00294"></a>00294             <span class="keyword">set</span>(<span class="keyword">get</span>(axesHandle,<span class="stringliteral">&#39;XLabel&#39;</span>), <span class="stringliteral">&#39;Position&#39;</span>, xLabelPos + [0 -7 0]);
<a name="l00295"></a>00295             yLabelPos = <span class="keyword">get</span>(<span class="keyword">get</span>(axesHandle,<span class="stringliteral">&#39;YLabel&#39;</span>), <span class="stringliteral">&#39;Position&#39;</span>);
<a name="l00296"></a>00296             <span class="keyword">set</span>(<span class="keyword">get</span>(axesHandle,<span class="stringliteral">&#39;YLabel&#39;</span>), <span class="stringliteral">&#39;Position&#39;</span>, yLabelPos + [-6 0 0]);
<a name="l00297"></a>00297         end
<a name="l00298"></a>00298         
<a name="l00299"></a>00299         <span class="keyword">function</span> plotProcessMetabolites(sim, figHandle)
<a name="l00300"></a>00300             import edu.stanford.covert.cell.sim.analysis.<a class="code" href="class_simulation_structure.html">SimulationStructure</a>;
<a name="l00301"></a>00301             
<a name="l00302"></a>00302             content = SimulationStructure.processMetabolites(sim);
<a name="l00303"></a>00303             [~, modIdxs] = ismember(content(:, 1), sim.processWholeCellModelIDs);
<a name="l00304"></a>00304             [~, metIdxs] = ismember(content(:, 3), sim.state(&#39;<a class="code" href="class_metabolite.html">Metabolite</a>&#39;).wholeCellModelIDs);
<a name="l00305"></a>00305             mat = zeros(numel(sim.processWholeCellModelIDs), numel(sim.state(&#39;<a class="code" href="class_metabolite.html">Metabolite</a>&#39;).wholeCellModelIDs));
<a name="l00306"></a>00306             mat(sub2ind(size(mat), modIdxs, metIdxs)) = 1;
<a name="l00307"></a>00307             
<a name="l00308"></a>00308             [mDNAIdxs, mRNAIdxs, mProtIdxs, mOtherIdxs] = SimulationStructure.processIndexs(sim);
<a name="l00309"></a>00309             order = [mDNAIdxs; mRNAIdxs; mProtIdxs; mOtherIdxs];
<a name="l00310"></a>00310             
<a name="l00311"></a>00311             processWholeCellModelIDs = {sim.processWholeCellModelIDs{order}}<span class="stringliteral">&#39;; %#ok&lt;CCAT1&gt;</span>
<a name="l00312"></a>00312 <span class="stringliteral">            processNames = SimulationStructure.getShortProcessNames(sim);</span>
<a name="l00313"></a>00313 <span class="stringliteral">            processNames = processNames(order(end:-1:1));</span>
<a name="l00314"></a>00314 <span class="stringliteral">            mat = mat(order, :);</span>
<a name="l00315"></a>00315 <span class="stringliteral">            </span>
<a name="l00316"></a>00316 <span class="stringliteral">            %plot</span>
<a name="l00317"></a>00317 <span class="stringliteral">            if ~exist(&#39;</span>figHandle<span class="stringliteral">&#39;, &#39;</span>var<span class="stringliteral">&#39;)</span>
<a name="l00318"></a>00318 <span class="stringliteral">                [axesHandle, figHandle] = edu.stanford.covert.cell.sim.util.PlotUtil.newAxesHandle();</span>
<a name="l00319"></a>00319 <span class="stringliteral">            else</span>
<a name="l00320"></a>00320 <span class="stringliteral">                axesHandle = subplot(1, 1, 1, &#39;</span>Parent<span class="stringliteral">&#39;, figHandle);</span>
<a name="l00321"></a>00321 <span class="stringliteral">            end</span>
<a name="l00322"></a>00322 <span class="stringliteral">            </span>
<a name="l00323"></a>00323 <span class="stringliteral">            image(mat+1, &#39;</span>Parent<span class="stringliteral">&#39;, axesHandle);</span>
<a name="l00324"></a>00324 <span class="stringliteral">            colormap(axesHandle, [</span>
<a name="l00325"></a>00325 <span class="stringliteral">                1 1 1;</span>
<a name="l00326"></a>00326 <span class="stringliteral">                0 0 0]);</span>
<a name="l00327"></a>00327 <span class="stringliteral">            </span>
<a name="l00328"></a>00328 <span class="stringliteral">            set(axesHandle, &#39;</span>XTick<span class="stringliteral">&#39;, []);</span>
<a name="l00329"></a>00329 <span class="stringliteral">            set(axesHandle, &#39;</span>YTick<span class="stringliteral">&#39;, 1:numel(processWholeCellModelIDs))</span>
<a name="l00330"></a>00330 <span class="stringliteral">            set(axesHandle, &#39;</span>YTickLabel<span class="stringliteral">&#39;, cell(size(processNames)));</span>
<a name="l00331"></a>00331 <span class="stringliteral">            </span>
<a name="l00332"></a>00332 <span class="stringliteral">            ylabel(&#39;</span><a class="code" href="class_process.html">Process</a><span class="stringliteral">&#39;, &#39;</span>FontSize<span class="stringliteral">&#39;, 28, &#39;</span>Units<span class="charliteral">&#39;,&#39;</span>normalized<span class="stringliteral">&#39;);</span>
<a name="l00333"></a>00333 <span class="stringliteral">            xlabel(&#39;</span><a class="code" href="class_metabolite.html">Metabolite</a><span class="stringliteral">&#39;, &#39;</span>FontSize<span class="stringliteral">&#39;, 28);</span>
<a name="l00334"></a>00334 <span class="stringliteral">            set(axesHandle, &#39;</span>XAxisLocation<span class="charliteral">&#39;,&#39;</span>top<span class="stringliteral">&#39;)</span>
<a name="l00335"></a>00335 <span class="stringliteral">            axesPos = get(axesHandle&#39;</span>, <span class="stringliteral">&#39;Position&#39;</span>);
<a name="l00336"></a>00336             <span class="keyword">set</span>(axesHandle, <span class="stringliteral">&#39;Position&#39;</span>, axesPos + [.15 -.10 -.08 -.08]);
<a name="l00337"></a>00337             xLabelPos = <span class="keyword">get</span>(<span class="keyword">get</span>(axesHandle,<span class="stringliteral">&#39;XLabel&#39;</span>), <span class="stringliteral">&#39;Position&#39;</span>);
<a name="l00338"></a>00338             <span class="keyword">set</span>(<span class="keyword">get</span>(axesHandle,<span class="stringliteral">&#39;XLabel&#39;</span>), <span class="stringliteral">&#39;Position&#39;</span>, xLabelPos + [0 -5.5 0]);
<a name="l00339"></a>00339             yLabelPos = <span class="keyword">get</span>(<span class="keyword">get</span>(axesHandle,<span class="stringliteral">&#39;YLabel&#39;</span>), <span class="stringliteral">&#39;Position&#39;</span>);
<a name="l00340"></a>00340             <span class="keyword">set</span>(<span class="keyword">get</span>(axesHandle,<span class="stringliteral">&#39;YLabel&#39;</span>), <span class="stringliteral">&#39;Position&#39;</span>, yLabelPos + [-.27 0 0]);
<a name="l00341"></a>00341             
<a name="l00342"></a>00342             %process labels
<a name="l00343"></a>00343             axisPos = <span class="keyword">get</span>(axesHandle,<span class="stringliteral">&#39;position&#39;</span>);
<a name="l00344"></a>00344             props = {<span class="stringliteral">&#39;FontSize&#39;</span>, 10, <span class="stringliteral">&#39;HorizontalAlignment&#39;</span>, <span class="stringliteral">&#39;right&#39;</span>, <span class="stringliteral">&#39;VerticalAlignment&#39;</span>,<span class="stringliteral">&#39;middle&#39;</span>, <span class="stringliteral">&#39;units&#39;</span>, <span class="stringliteral">&#39;normalized&#39;</span>, <span class="stringliteral">&#39;Parent&#39;</span>, axesHandle};
<a name="l00345"></a>00345             <span class="keywordflow">for</span> i = 1:numel(processWholeCellModelIDs)
<a name="l00346"></a>00346                 text(-axisPos(3)/100, (i+0.1)/(numel(order)+1), processNames{i}, props{:});
<a name="l00347"></a>00347             end
<a name="l00348"></a>00348             
<a name="l00349"></a>00349             %process  category labels
<a name="l00350"></a>00350             props = {<span class="stringliteral">&#39;Rotation&#39;</span>, 90, <span class="stringliteral">&#39;FontSize&#39;</span>, 14, <span class="stringliteral">&#39;HorizontalAlignment&#39;</span>, <span class="stringliteral">&#39;center&#39;</span>, <span class="stringliteral">&#39;VerticalAlignment&#39;</span>,<span class="stringliteral">&#39;middle&#39;</span>, <span class="stringliteral">&#39;units&#39;</span>, <span class="stringliteral">&#39;normalized&#39;</span>, <span class="stringliteral">&#39;Parent&#39;</span>, axesHandle};
<a name="l00351"></a>00351             text(-0.27, 1/(numel(order)+1)*(1/2 + numel(mOtherIdxs)/2), <span class="stringliteral">&#39;Other&#39;</span>, props{:});
<a name="l00352"></a>00352             text(-0.27, 1/(numel(order)+1)*(1/2 + numel(mOtherIdxs) + numel(mProtIdxs)/2), <span class="stringliteral">&#39;Protein&#39;</span>, props{:});
<a name="l00353"></a>00353             text(-0.27, 1/(numel(order)+1)*(1/2 + numel(mOtherIdxs) + numel(mProtIdxs) + numel(mRNAIdxs)/2), <span class="stringliteral">&#39;RNA&#39;</span>, props{:});
<a name="l00354"></a>00354             text(-0.27, 1/(numel(order)+1)*(1/2 + numel(mOtherIdxs) + numel(mProtIdxs) + numel(mRNAIdxs) + numel(mDNAIdxs)/2), <span class="stringliteral">&#39;DNA&#39;</span>, props{:});
<a name="l00355"></a>00355             
<a name="l00356"></a>00356             %process category brackets
<a name="l00357"></a>00357             props = {<span class="stringliteral">&#39;LineWidth&#39;</span>, 1};
<a name="l00358"></a>00358             margin = 0.2;
<a name="l00359"></a>00359             tickW = 0.015;
<a name="l00360"></a>00360             yTickW = -0.27+0.03;
<a name="l00361"></a>00361             
<a name="l00362"></a>00362             annotation(figHandle, <span class="stringliteral">&#39;line&#39;</span>, axisPos(1) + axisPos(3)*(yTickW + [0; 0]),     axisPos(2) + axisPos(4)/(numel(order)+1)*(1/2 + [margin; numel(mOtherIdxs)-margin]), props{:});
<a name="l00363"></a>00363             annotation(figHandle, <span class="stringliteral">&#39;line&#39;</span>, axisPos(1) + axisPos(3)*(yTickW + [0; tickW]), axisPos(2) + axisPos(4)/(numel(order)+1)*(1/2 + [margin; margin]), props{:});
<a name="l00364"></a>00364             annotation(figHandle, <span class="stringliteral">&#39;line&#39;</span>, axisPos(1) + axisPos(3)*(yTickW + [0; tickW]), axisPos(2) + axisPos(4)/(numel(order)+1)*(1/2 + [numel(mOtherIdxs)-margin; numel(mOtherIdxs)-margin]), props{:});
<a name="l00365"></a>00365             
<a name="l00366"></a>00366             annotation(figHandle, <span class="stringliteral">&#39;line&#39;</span>, axisPos(1) + axisPos(3)*(yTickW + [0; 0]),     axisPos(2) + axisPos(4)/(numel(order)+1)*(1/2 + numel(mOtherIdxs) + [margin; numel(mProtIdxs)-margin]), props{:});
<a name="l00367"></a>00367             annotation(figHandle, <span class="stringliteral">&#39;line&#39;</span>, axisPos(1) + axisPos(3)*(yTickW + [0; tickW]), axisPos(2) + axisPos(4)/(numel(order)+1)*(1/2 + numel(mOtherIdxs) + [margin; margin]), props{:});
<a name="l00368"></a>00368             annotation(figHandle, <span class="stringliteral">&#39;line&#39;</span>, axisPos(1) + axisPos(3)*(yTickW + [0; tickW]), axisPos(2) + axisPos(4)/(numel(order)+1)*(1/2 + numel(mOtherIdxs) + [numel(mProtIdxs)-margin; numel(mProtIdxs)-margin]), props{:});
<a name="l00369"></a>00369             
<a name="l00370"></a>00370             annotation(figHandle, <span class="stringliteral">&#39;line&#39;</span>, axisPos(1) + axisPos(3)*(yTickW + [0; 0]),     axisPos(2) + axisPos(4)/(numel(order)+1)*(1/2 + numel(mOtherIdxs) + numel(mProtIdxs) + [margin; numel(mRNAIdxs)-margin]), props{:});
<a name="l00371"></a>00371             annotation(figHandle, <span class="stringliteral">&#39;line&#39;</span>, axisPos(1) + axisPos(3)*(yTickW + [0; tickW]), axisPos(2) + axisPos(4)/(numel(order)+1)*(1/2 + numel(mOtherIdxs) + numel(mProtIdxs) + [margin; margin]), props{:});
<a name="l00372"></a>00372             annotation(figHandle, <span class="stringliteral">&#39;line&#39;</span>, axisPos(1) + axisPos(3)*(yTickW + [0; tickW]), axisPos(2) + axisPos(4)/(numel(order)+1)*(1/2 + numel(mOtherIdxs) + numel(mProtIdxs) + [numel(mRNAIdxs)-margin; numel(mRNAIdxs)-margin]), props{:});
<a name="l00373"></a>00373             
<a name="l00374"></a>00374             annotation(figHandle, <span class="stringliteral">&#39;line&#39;</span>, axisPos(1) + axisPos(3)*(yTickW + [0; 0]),     axisPos(2) + axisPos(4)/(numel(order)+1)*(1/2 + numel(mOtherIdxs) + numel(mProtIdxs) + numel(mRNAIdxs) + [margin; numel(mDNAIdxs)-margin]), props{:});
<a name="l00375"></a>00375             annotation(figHandle, <span class="stringliteral">&#39;line&#39;</span>, axisPos(1) + axisPos(3)*(yTickW + [0; tickW]), axisPos(2) + axisPos(4)/(numel(order)+1)*(1/2 + numel(mOtherIdxs) + numel(mProtIdxs) + numel(mRNAIdxs) + [margin; margin]), props{:});
<a name="l00376"></a>00376             annotation(figHandle, <span class="stringliteral">&#39;line&#39;</span>, axisPos(1) + axisPos(3)*(yTickW + [0; tickW]), axisPos(2) + axisPos(4)/(numel(order)+1)*(1/2 + numel(mOtherIdxs) + numel(mProtIdxs) + numel(mRNAIdxs) + [numel(mDNAIdxs)-margin; numel(mDNAIdxs)-margin]), props{:});
<a name="l00377"></a>00377         end
<a name="l00378"></a>00378         
<a name="l00379"></a>00379         <span class="keyword">function</span> plotProcessSharedMetabolites(sim, figHandle)
<a name="l00380"></a>00380             import edu.stanford.covert.cell.sim.analysis.<a class="code" href="class_simulation_structure.html">SimulationStructure</a>;
<a name="l00381"></a>00381             
<a name="l00382"></a>00382             content = SimulationStructure.processMetabolites(sim);
<a name="l00383"></a>00383             [~, modIdxs] = ismember(content(:, 1), sim.processWholeCellModelIDs);
<a name="l00384"></a>00384             [~, metIdxs] = ismember(content(:, 3), sim.state(&#39;<a class="code" href="class_metabolite.html">Metabolite</a>&#39;).wholeCellModelIDs);
<a name="l00385"></a>00385             mat = zeros(numel(sim.processWholeCellModelIDs), numel(sim.state(&#39;<a class="code" href="class_metabolite.html">Metabolite</a>&#39;).wholeCellModelIDs));
<a name="l00386"></a>00386             mat(sub2ind(size(mat), modIdxs, metIdxs)) = 1;
<a name="l00387"></a>00387             
<a name="l00388"></a>00388             idxs = find(sum(mat, 1) &gt; 4);
<a name="l00389"></a>00389             metaboliteWholeCellModelIDs = sim.state(&#39;<a class="code" href="class_metabolite.html">Metabolite</a>&#39;).wholeCellModelIDs(idxs);
<a name="l00390"></a>00390             mat = mat(:, idxs);
<a name="l00391"></a>00391             
<a name="l00392"></a>00392             [~,idxs] = ismember({
<a name="l00393"></a>00393                 <span class="stringliteral">&#39;ATP&#39;</span>; <span class="stringliteral">&#39;ADP&#39;</span>; <span class="stringliteral">&#39;AMP&#39;</span>; <span class="stringliteral">&#39;GTP&#39;</span>; <span class="stringliteral">&#39;GDP&#39;</span>;
<a name="l00394"></a>00394                 <span class="stringliteral">&#39;PI&#39;</span>; <span class="stringliteral">&#39;PPI&#39;</span>;
<a name="l00395"></a>00395                 <span class="charliteral">&#39;H&#39;</span>; <span class="stringliteral">&#39;H2O&#39;</span>;
<a name="l00396"></a>00396                 <span class="stringliteral">&#39;ALA&#39;</span>; <span class="stringliteral">&#39;ARG&#39;</span>; <span class="stringliteral">&#39;ASN&#39;</span>; <span class="stringliteral">&#39;ASP&#39;</span>; <span class="stringliteral">&#39;CYS&#39;</span>; <span class="stringliteral">&#39;FMET&#39;</span>; <span class="stringliteral">&#39;GLN&#39;</span>; <span class="stringliteral">&#39;GLU&#39;</span>; <span class="stringliteral">&#39;GLY&#39;</span>; <span class="stringliteral">&#39;HIS&#39;</span>; <span class="stringliteral">&#39;ILE&#39;</span>; <span class="stringliteral">&#39;LEU&#39;</span>; <span class="stringliteral">&#39;LYS&#39;</span>; <span class="stringliteral">&#39;MET&#39;</span>; <span class="stringliteral">&#39;PHE&#39;</span>; <span class="stringliteral">&#39;PRO&#39;</span>; <span class="stringliteral">&#39;SER&#39;</span>; <span class="stringliteral">&#39;THR&#39;</span>; <span class="stringliteral">&#39;TRP&#39;</span>; <span class="stringliteral">&#39;TYR&#39;</span>; <span class="stringliteral">&#39;VAL&#39;</span>}, ...
<a name="l00397"></a>00397                 metaboliteWholeCellModelIDs);
<a name="l00398"></a>00398             metaboliteWholeCellModelIDs = metaboliteWholeCellModelIDs(idxs);
<a name="l00399"></a>00399             mat = mat(:, idxs);
<a name="l00400"></a>00400             
<a name="l00401"></a>00401             nucleicAcidIndexs = (1:5)<span class="stringliteral">&#39;;</span>
<a name="l00402"></a>00402 <span class="stringliteral">            phosphateIndexs = (6:7)&#39;</span>;
<a name="l00403"></a>00403             waterIndexs = (8:9)<span class="stringliteral">&#39;;</span>
<a name="l00404"></a>00404 <span class="stringliteral">            aminoAcidIndexs = (10:30)&#39;</span>;
<a name="l00405"></a>00405             
<a name="l00406"></a>00406             [mDNAIdxs, mRNAIdxs, mProtIdxs, mOtherIdxs] = <a class="code" href="class_simulation_structure.html">SimulationStructure</a>.<a class="code" href="class_simulation_structure.html#af4001e459efc820915ecb97ca1322c82">processIndexs</a>(sim);
<a name="l00407"></a>00407             order = [mDNAIdxs; mRNAIdxs; mProtIdxs; mOtherIdxs];
<a name="l00408"></a>00408             
<a name="l00409"></a>00409             processWholeCellModelIDs = {sim.processWholeCellModelIDs{order}}<span class="stringliteral">&#39;; %#ok&lt;CCAT1&gt;</span>
<a name="l00410"></a>00410 <span class="stringliteral">            processNames = SimulationStructure.getShortProcessNames(sim);</span>
<a name="l00411"></a>00411 <span class="stringliteral">            processNames = processNames(order(end:-1:1));</span>
<a name="l00412"></a>00412 <span class="stringliteral">            mat = mat(order, :);</span>
<a name="l00413"></a>00413 <span class="stringliteral">            </span>
<a name="l00414"></a>00414 <span class="stringliteral">            %plot</span>
<a name="l00415"></a>00415 <span class="stringliteral">            if ~exist(&#39;</span>figHandle<span class="stringliteral">&#39;, &#39;</span>var<span class="stringliteral">&#39;)</span>
<a name="l00416"></a>00416 <span class="stringliteral">                [axesHandle, figHandle] = edu.stanford.covert.cell.sim.util.PlotUtil.newAxesHandle();</span>
<a name="l00417"></a>00417 <span class="stringliteral">            else</span>
<a name="l00418"></a>00418 <span class="stringliteral">                axesHandle = subplot(1, 1, 1, &#39;</span>Parent<span class="stringliteral">&#39;, figHandle);</span>
<a name="l00419"></a>00419 <span class="stringliteral">            end</span>
<a name="l00420"></a>00420 <span class="stringliteral">            </span>
<a name="l00421"></a>00421 <span class="stringliteral">            image(mat+1, &#39;</span>Parent<span class="stringliteral">&#39;, axesHandle);</span>
<a name="l00422"></a>00422 <span class="stringliteral">            colormap(axesHandle, [</span>
<a name="l00423"></a>00423 <span class="stringliteral">                1 1 1;</span>
<a name="l00424"></a>00424 <span class="stringliteral">                0 0 0]);</span>
<a name="l00425"></a>00425 <span class="stringliteral">            </span>
<a name="l00426"></a>00426 <span class="stringliteral">            set(axesHandle, &#39;</span>XTick<span class="stringliteral">&#39;, []);</span>
<a name="l00427"></a>00427 <span class="stringliteral">            set(axesHandle, &#39;</span>YTick<span class="stringliteral">&#39;, 1:numel(processWholeCellModelIDs))</span>
<a name="l00428"></a>00428 <span class="stringliteral">            set(axesHandle, &#39;</span>YTickLabel<span class="stringliteral">&#39;, cell(size(processNames)));</span>
<a name="l00429"></a>00429 <span class="stringliteral">            </span>
<a name="l00430"></a>00430 <span class="stringliteral">            ylabel(&#39;</span><a class="code" href="class_process.html">Process</a><span class="stringliteral">&#39;, &#39;</span>FontSize<span class="stringliteral">&#39;, 28, &#39;</span>Units<span class="charliteral">&#39;,&#39;</span>normalized<span class="stringliteral">&#39;);</span>
<a name="l00431"></a>00431 <span class="stringliteral">            xlabel(&#39;</span><a class="code" href="class_metabolite.html">Metabolite</a><span class="stringliteral">&#39;, &#39;</span>FontSize<span class="stringliteral">&#39;, 28);</span>
<a name="l00432"></a>00432 <span class="stringliteral">            set(axesHandle, &#39;</span>XAxisLocation<span class="charliteral">&#39;,&#39;</span>top<span class="stringliteral">&#39;)</span>
<a name="l00433"></a>00433 <span class="stringliteral">            axesPos = get(axesHandle&#39;</span>, <span class="stringliteral">&#39;Position&#39;</span>);
<a name="l00434"></a>00434             <span class="keyword">set</span>(axesHandle, <span class="stringliteral">&#39;Position&#39;</span>, axesPos + [.15 -.10 -.08 -.08]);
<a name="l00435"></a>00435             xLabelPos = <span class="keyword">get</span>(<span class="keyword">get</span>(axesHandle,<span class="stringliteral">&#39;XLabel&#39;</span>), <span class="stringliteral">&#39;Position&#39;</span>);
<a name="l00436"></a>00436             <span class="keyword">set</span>(<span class="keyword">get</span>(axesHandle,<span class="stringliteral">&#39;XLabel&#39;</span>), <span class="stringliteral">&#39;Position&#39;</span>, xLabelPos + [0 -5.5 0]);
<a name="l00437"></a>00437             yLabelPos = <span class="keyword">get</span>(<span class="keyword">get</span>(axesHandle,<span class="stringliteral">&#39;YLabel&#39;</span>), <span class="stringliteral">&#39;Position&#39;</span>);
<a name="l00438"></a>00438             <span class="keyword">set</span>(<span class="keyword">get</span>(axesHandle,<span class="stringliteral">&#39;YLabel&#39;</span>), <span class="stringliteral">&#39;Position&#39;</span>, yLabelPos + [-.27 0 0]);
<a name="l00439"></a>00439             
<a name="l00440"></a>00440             %metabolite labels
<a name="l00441"></a>00441             metIDs = metaboliteWholeCellModelIDs(sum(mat, 1) &gt; 4);
<a name="l00442"></a>00442             <span class="keywordflow">for</span> i = 1:numel(metIDs)
<a name="l00443"></a>00443                 x = 1/2 + find(strcmp(metaboliteWholeCellModelIDs, metIDs{i}));
<a name="l00444"></a>00444                 text(x, 0, metaboliteWholeCellModelIDs{i}, <span class="stringliteral">&#39;Parent&#39;</span>, axesHandle, <span class="stringliteral">&#39;FontSize&#39;</span>, 10, <span class="stringliteral">&#39;Rotation&#39;</span>, 270, <span class="stringliteral">&#39;HorizontalAlignment&#39;</span>, <span class="stringliteral">&#39;right&#39;</span>, <span class="stringliteral">&#39;VerticalAlignment&#39;</span>,<span class="stringliteral">&#39;middle&#39;</span>, <span class="stringliteral">&#39;units&#39;</span>, <span class="stringliteral">&#39;data&#39;</span>);
<a name="l00445"></a>00445             end
<a name="l00446"></a>00446             
<a name="l00447"></a>00447             %metabolite  category labels
<a name="l00448"></a>00448             props = {<span class="stringliteral">&#39;FontSize&#39;</span>, 14, <span class="stringliteral">&#39;HorizontalAlignment&#39;</span>, <span class="stringliteral">&#39;center&#39;</span>, <span class="stringliteral">&#39;VerticalAlignment&#39;</span>,<span class="stringliteral">&#39;middle&#39;</span>, <span class="stringliteral">&#39;units&#39;</span>, <span class="stringliteral">&#39;normalized&#39;</span>, <span class="stringliteral">&#39;Parent&#39;</span>, axesHandle};
<a name="l00449"></a>00449             text(1/(numel(metaboliteWholeCellModelIDs)+1)*(1/2 + numel(nucleicAcidIndexs)/2),1.19, <span class="stringliteral">&#39;NXP&#39;</span>, props{:});
<a name="l00450"></a>00450             text(1/(numel(metaboliteWholeCellModelIDs)+1)*(1/2 + numel(nucleicAcidIndexs) + numel(phosphateIndexs)/2), 1.19, <span class="stringliteral">&#39;Pi&#39;</span>, props{:});
<a name="l00451"></a>00451             text(1/(numel(metaboliteWholeCellModelIDs)+1)*(1/2 + numel(nucleicAcidIndexs) + numel(phosphateIndexs) + numel(waterIndexs)/2), 1.19, <span class="charliteral">&#39;H&#39;</span>, props{:});
<a name="l00452"></a>00452             text(1/(numel(metaboliteWholeCellModelIDs)+1)*(1/2 + numel(nucleicAcidIndexs) + numel(phosphateIndexs) + numel(waterIndexs) + numel(aminoAcidIndexs)/2), 1.19, <span class="stringliteral">&#39;Amino Acid&#39;</span>, props{:});
<a name="l00453"></a>00453             
<a name="l00454"></a>00454             %metabolite category brackets
<a name="l00455"></a>00455             props = {<span class="stringliteral">&#39;LineWidth&#39;</span>, 1};
<a name="l00456"></a>00456             margin = 0.2;
<a name="l00457"></a>00457             tickW = -0.015;
<a name="l00458"></a>00458             yTickW = 1.15;
<a name="l00459"></a>00459             axisPos = <span class="keyword">get</span>(axesHandle,<span class="stringliteral">&#39;position&#39;</span>);
<a name="l00460"></a>00460             
<a name="l00461"></a>00461             annotation(figHandle, <span class="stringliteral">&#39;line&#39;</span>, axisPos(1) + axisPos(3)/(numel(metaboliteWholeCellModelIDs)+1)*(1/2 + [margin; numel(nucleicAcidIndexs)-margin]), axisPos(2) + axisPos(4)*(yTickW + [0; 0]), props{:});
<a name="l00462"></a>00462             annotation(figHandle, <span class="stringliteral">&#39;line&#39;</span>, axisPos(1) + axisPos(3)/(numel(metaboliteWholeCellModelIDs)+1)*(1/2 + [margin; margin]), axisPos(2) + axisPos(4)*(yTickW + [0; tickW]), props{:});
<a name="l00463"></a>00463             annotation(figHandle, <span class="stringliteral">&#39;line&#39;</span>, axisPos(1) + axisPos(3)/(numel(metaboliteWholeCellModelIDs)+1)*(1/2 + [numel(nucleicAcidIndexs)-margin; numel(nucleicAcidIndexs)-margin]), axisPos(2) + axisPos(4)*(yTickW + [0; tickW]), props{:});
<a name="l00464"></a>00464             
<a name="l00465"></a>00465             annotation(figHandle, <span class="stringliteral">&#39;line&#39;</span>, axisPos(1) + axisPos(3)/(numel(metaboliteWholeCellModelIDs)+1)*(1/2 + numel(nucleicAcidIndexs) + [margin; numel(phosphateIndexs)-margin]), axisPos(2) + axisPos(4)*(yTickW + [0; 0]), props{:});
<a name="l00466"></a>00466             annotation(figHandle, <span class="stringliteral">&#39;line&#39;</span>, axisPos(1) + axisPos(3)/(numel(metaboliteWholeCellModelIDs)+1)*(1/2 + numel(nucleicAcidIndexs) + [margin; margin]), axisPos(2) + axisPos(4)*(yTickW + [0; tickW]), props{:});
<a name="l00467"></a>00467             annotation(figHandle, <span class="stringliteral">&#39;line&#39;</span>, axisPos(1) + axisPos(3)/(numel(metaboliteWholeCellModelIDs)+1)*(1/2 + numel(nucleicAcidIndexs) + [numel(phosphateIndexs)-margin; numel(phosphateIndexs)-margin]), axisPos(2) + axisPos(4)*(yTickW + [0; tickW]), props{:});
<a name="l00468"></a>00468             
<a name="l00469"></a>00469             annotation(figHandle, <span class="stringliteral">&#39;line&#39;</span>, axisPos(1) + axisPos(3)/(numel(metaboliteWholeCellModelIDs)+1)*(1/2 + numel(nucleicAcidIndexs) + numel(phosphateIndexs) + [margin; numel(waterIndexs)-margin]), axisPos(2) + axisPos(4)*(yTickW + [0; 0]), props{:});
<a name="l00470"></a>00470             annotation(figHandle, <span class="stringliteral">&#39;line&#39;</span>, axisPos(1) + axisPos(3)/(numel(metaboliteWholeCellModelIDs)+1)*(1/2 + numel(nucleicAcidIndexs) + numel(phosphateIndexs) + [margin; margin]), axisPos(2) + axisPos(4)*(yTickW + [0; tickW]), props{:});
<a name="l00471"></a>00471             annotation(figHandle, <span class="stringliteral">&#39;line&#39;</span>, axisPos(1) + axisPos(3)/(numel(metaboliteWholeCellModelIDs)+1)*(1/2 + numel(nucleicAcidIndexs) + numel(phosphateIndexs) + [numel(waterIndexs)-margin; numel(waterIndexs)-margin]), axisPos(2) + axisPos(4)*(yTickW + [0; tickW]), props{:});
<a name="l00472"></a>00472             
<a name="l00473"></a>00473             annotation(figHandle, <span class="stringliteral">&#39;line&#39;</span>, axisPos(1) + axisPos(3)/(numel(metaboliteWholeCellModelIDs)+1)*(1/2 + numel(nucleicAcidIndexs) + numel(phosphateIndexs) + numel(waterIndexs) + [margin; numel(aminoAcidIndexs)-margin]), axisPos(2) + axisPos(4)*(yTickW + [0; 0]), props{:});
<a name="l00474"></a>00474             annotation(figHandle, <span class="stringliteral">&#39;line&#39;</span>, axisPos(1) + axisPos(3)/(numel(metaboliteWholeCellModelIDs)+1)*(1/2 + numel(nucleicAcidIndexs) + numel(phosphateIndexs) + numel(waterIndexs) + [margin; margin]), axisPos(2) + axisPos(4)*(yTickW + [0; tickW]), props{:});
<a name="l00475"></a>00475             annotation(figHandle, <span class="stringliteral">&#39;line&#39;</span>, axisPos(1) + axisPos(3)/(numel(metaboliteWholeCellModelIDs)+1)*(1/2 + numel(nucleicAcidIndexs) + numel(phosphateIndexs) + numel(waterIndexs) + [numel(aminoAcidIndexs)-margin; numel(aminoAcidIndexs)-margin]), axisPos(2) + axisPos(4)*(yTickW + [0; tickW]), props{:});
<a name="l00476"></a>00476             
<a name="l00477"></a>00477             %process labels
<a name="l00478"></a>00478             props = {<span class="stringliteral">&#39;FontSize&#39;</span>, 10, <span class="stringliteral">&#39;HorizontalAlignment&#39;</span>, <span class="stringliteral">&#39;right&#39;</span>, <span class="stringliteral">&#39;VerticalAlignment&#39;</span>,<span class="stringliteral">&#39;middle&#39;</span>, <span class="stringliteral">&#39;units&#39;</span>, <span class="stringliteral">&#39;normalized&#39;</span>, <span class="stringliteral">&#39;Parent&#39;</span>, axesHandle};
<a name="l00479"></a>00479             <span class="keywordflow">for</span> i = 1:numel(processWholeCellModelIDs)
<a name="l00480"></a>00480                 text(-axisPos(3)/100, (i+0.1)/(numel(order)+1), processNames{i}, props{:});
<a name="l00481"></a>00481             end
<a name="l00482"></a>00482             
<a name="l00483"></a>00483             %process  category labels
<a name="l00484"></a>00484             props = {<span class="stringliteral">&#39;Rotation&#39;</span>, 90, <span class="stringliteral">&#39;FontSize&#39;</span>, 14, <span class="stringliteral">&#39;HorizontalAlignment&#39;</span>, <span class="stringliteral">&#39;center&#39;</span>, <span class="stringliteral">&#39;VerticalAlignment&#39;</span>,<span class="stringliteral">&#39;middle&#39;</span>, <span class="stringliteral">&#39;units&#39;</span>, <span class="stringliteral">&#39;normalized&#39;</span>, <span class="stringliteral">&#39;Parent&#39;</span>, axesHandle};
<a name="l00485"></a>00485             text(-0.27, 1/(numel(order)+1)*(1/2 + numel(mOtherIdxs)/2), <span class="stringliteral">&#39;Other&#39;</span>, props{:});
<a name="l00486"></a>00486             text(-0.27, 1/(numel(order)+1)*(1/2 + numel(mOtherIdxs) + numel(mProtIdxs)/2), <span class="stringliteral">&#39;Protein&#39;</span>, props{:});
<a name="l00487"></a>00487             text(-0.27, 1/(numel(order)+1)*(1/2 + numel(mOtherIdxs) + numel(mProtIdxs) + numel(mRNAIdxs)/2), <span class="stringliteral">&#39;RNA&#39;</span>, props{:});
<a name="l00488"></a>00488             text(-0.27, 1/(numel(order)+1)*(1/2 + numel(mOtherIdxs) + numel(mProtIdxs) + numel(mRNAIdxs) + numel(mDNAIdxs)/2), <span class="stringliteral">&#39;DNA&#39;</span>, props{:});
<a name="l00489"></a>00489             
<a name="l00490"></a>00490             %process category brackets
<a name="l00491"></a>00491             props = {<span class="stringliteral">&#39;LineWidth&#39;</span>, 1};
<a name="l00492"></a>00492             margin = 0.2;
<a name="l00493"></a>00493             tickW = 0.015;
<a name="l00494"></a>00494             yTickW = -0.27+0.03;
<a name="l00495"></a>00495             
<a name="l00496"></a>00496             annotation(figHandle, <span class="stringliteral">&#39;line&#39;</span>, axisPos(1) + axisPos(3)*(yTickW + [0; 0]),     axisPos(2) + axisPos(4)/(numel(order)+1)*(1/2 + [margin; numel(mOtherIdxs)-margin]), props{:});
<a name="l00497"></a>00497             annotation(figHandle, <span class="stringliteral">&#39;line&#39;</span>, axisPos(1) + axisPos(3)*(yTickW + [0; tickW]), axisPos(2) + axisPos(4)/(numel(order)+1)*(1/2 + [margin; margin]), props{:});
<a name="l00498"></a>00498             annotation(figHandle, <span class="stringliteral">&#39;line&#39;</span>, axisPos(1) + axisPos(3)*(yTickW + [0; tickW]), axisPos(2) + axisPos(4)/(numel(order)+1)*(1/2 + [numel(mOtherIdxs)-margin; numel(mOtherIdxs)-margin]), props{:});
<a name="l00499"></a>00499             
<a name="l00500"></a>00500             annotation(figHandle, <span class="stringliteral">&#39;line&#39;</span>, axisPos(1) + axisPos(3)*(yTickW + [0; 0]),     axisPos(2) + axisPos(4)/(numel(order)+1)*(1/2 + numel(mOtherIdxs) + [margin; numel(mProtIdxs)-margin]), props{:});
<a name="l00501"></a>00501             annotation(figHandle, <span class="stringliteral">&#39;line&#39;</span>, axisPos(1) + axisPos(3)*(yTickW + [0; tickW]), axisPos(2) + axisPos(4)/(numel(order)+1)*(1/2 + numel(mOtherIdxs) + [margin; margin]), props{:});
<a name="l00502"></a>00502             annotation(figHandle, <span class="stringliteral">&#39;line&#39;</span>, axisPos(1) + axisPos(3)*(yTickW + [0; tickW]), axisPos(2) + axisPos(4)/(numel(order)+1)*(1/2 + numel(mOtherIdxs) + [numel(mProtIdxs)-margin; numel(mProtIdxs)-margin]), props{:});
<a name="l00503"></a>00503             
<a name="l00504"></a>00504             annotation(figHandle, <span class="stringliteral">&#39;line&#39;</span>, axisPos(1) + axisPos(3)*(yTickW + [0; 0]),     axisPos(2) + axisPos(4)/(numel(order)+1)*(1/2 + numel(mOtherIdxs) + numel(mProtIdxs) + [margin; numel(mRNAIdxs)-margin]), props{:});
<a name="l00505"></a>00505             annotation(figHandle, <span class="stringliteral">&#39;line&#39;</span>, axisPos(1) + axisPos(3)*(yTickW + [0; tickW]), axisPos(2) + axisPos(4)/(numel(order)+1)*(1/2 + numel(mOtherIdxs) + numel(mProtIdxs) + [margin; margin]), props{:});
<a name="l00506"></a>00506             annotation(figHandle, <span class="stringliteral">&#39;line&#39;</span>, axisPos(1) + axisPos(3)*(yTickW + [0; tickW]), axisPos(2) + axisPos(4)/(numel(order)+1)*(1/2 + numel(mOtherIdxs) + numel(mProtIdxs) + [numel(mRNAIdxs)-margin; numel(mRNAIdxs)-margin]), props{:});
<a name="l00507"></a>00507             
<a name="l00508"></a>00508             annotation(figHandle, <span class="stringliteral">&#39;line&#39;</span>, axisPos(1) + axisPos(3)*(yTickW + [0; 0]),     axisPos(2) + axisPos(4)/(numel(order)+1)*(1/2 + numel(mOtherIdxs) + numel(mProtIdxs) + numel(mRNAIdxs) + [margin; numel(mDNAIdxs)-margin]), props{:});
<a name="l00509"></a>00509             annotation(figHandle, <span class="stringliteral">&#39;line&#39;</span>, axisPos(1) + axisPos(3)*(yTickW + [0; tickW]), axisPos(2) + axisPos(4)/(numel(order)+1)*(1/2 + numel(mOtherIdxs) + numel(mProtIdxs) + numel(mRNAIdxs) + [margin; margin]), props{:});
<a name="l00510"></a>00510             annotation(figHandle, <span class="stringliteral">&#39;line&#39;</span>, axisPos(1) + axisPos(3)*(yTickW + [0; tickW]), axisPos(2) + axisPos(4)/(numel(order)+1)*(1/2 + numel(mOtherIdxs) + numel(mProtIdxs) + numel(mRNAIdxs) + [numel(mDNAIdxs)-margin; numel(mDNAIdxs)-margin]), props{:});
<a name="l00511"></a>00511         end
<a name="l00512"></a>00512         
<a name="l00513"></a>00513         <span class="keyword">function</span> plotProcessGeneProducts(sim, figHandle)
<a name="l00514"></a>00514             import edu.stanford.covert.cell.sim.analysis.<a class="code" href="class_simulation_structure.html">SimulationStructure</a>;
<a name="l00515"></a>00515             
<a name="l00516"></a>00516             %process names
<a name="l00517"></a>00517             processNames = SimulationStructure.getShortProcessNames(sim);
<a name="l00518"></a>00518             
<a name="l00519"></a>00519             %order processes
<a name="l00520"></a>00520             [mDNAIdxs, mRNAIdxs, mProtIdxs, mOtherIdxs] = SimulationStructure.processIndexs(sim);
<a name="l00521"></a>00521             
<a name="l00522"></a>00522             order = [mDNAIdxs; mRNAIdxs; mProtIdxs; mOtherIdxs];
<a name="l00523"></a>00523             order = order(end:-1:1);
<a name="l00524"></a>00524             if numel(order) &lt; numel(sim.processWholeCellModelIDs)
<a name="l00525"></a>00525                 throw(MException(&#39;SimulationStructure:error&#39;, &#39;Not all processes included&#39;));
<a name="l00526"></a>00526             end
<a name="l00527"></a>00527             
<a name="l00528"></a>00528             %count numbers of genes used in each process
<a name="l00529"></a>00529             content = SimulationStructure.processGeneProducts(sim);
<a name="l00530"></a>00530             [~, mIdxs] = ismember(content(:,1), sim.processWholeCellModelIDs);
<a name="l00531"></a>00531             mCnts = histc(mIdxs, 1:numel(sim.processWholeCellModelIDs));
<a name="l00532"></a>00532             
<a name="l00533"></a>00533             %bar graph
<a name="l00534"></a>00534             if ~exist(&#39;figHandle&#39;, &#39;var&#39;)
<a name="l00535"></a>00535                 [axesHandle, figHandle] = edu.stanford.covert.cell.sim.util.PlotUtil.newAxesHandle();
<a name="l00536"></a>00536             else
<a name="l00537"></a>00537                 axesHandle = subplot(1, 1, 1, &#39;Parent&#39;, figHandle);
<a name="l00538"></a>00538             end
<a name="l00539"></a>00539             
<a name="l00540"></a>00540             hold on;
<a name="l00541"></a>00541             colors = SimulationStructure.barColors();
<a name="l00542"></a>00542             %barh((1:numel(sim.processWholeCellModelIDs))&#39;, mCnts(order));
<a name="l00543"></a>00543             barh((1:numel(mOtherIdxs))&#39;, mCnts(mOtherIdxs(end:-1:1)), &#39;facecolor&#39;, colors(1, :), &#39;Parent&#39;, axesHandle);
<a name="l00544"></a>00544             barh((1:numel(mProtIdxs))&#39; + numel(mOtherIdxs), mCnts(mProtIdxs(end:-1:1)), &#39;facecolor&#39;, colors(2, :), &#39;Parent&#39;, axesHandle);
<a name="l00545"></a>00545             barh((1:numel(mRNAIdxs))&#39; + numel(mOtherIdxs) + numel(mProtIdxs), mCnts(mRNAIdxs(end:-1:1)), &#39;facecolor&#39;, colors(3, :), &#39;Parent&#39;, axesHandle);
<a name="l00546"></a>00546             barh((1:numel(mDNAIdxs))&#39; + numel(mOtherIdxs) + numel(mProtIdxs) + numel(mRNAIdxs), mCnts(mDNAIdxs(end:-1:1)), &#39;facecolor&#39;, colors(4, :), &#39;Parent&#39;, axesHandle);
<a name="l00547"></a>00547             
<a name="l00548"></a>00548             ylim([0 numel(sim.processWholeCellModelIDs)+1]);
<a name="l00549"></a>00549             xlim([1 max(mCnts)*1.1]);
<a name="l00550"></a>00550             set(axesHandle, &#39;YTick&#39;, []);
<a name="l00551"></a>00551             set(axesHandle, &#39;YTickLabel&#39;, []);
<a name="l00552"></a>00552             set(axesHandle, &#39;XScale&#39;, &#39;log&#39;);
<a name="l00553"></a>00553             set(axesHandle, &#39;box&#39;, &#39;on&#39;);
<a name="l00554"></a>00554             
<a name="l00555"></a>00555             %axis labels
<a name="l00556"></a>00556             ylabel(&#39;<a class="code" href="class_process.html">Process</a>&#39;, &#39;FontSize&#39;, 18);
<a name="l00557"></a>00557             xlabel(&#39;Genes&#39;, &#39;FontSize&#39;, 18);
<a name="l00558"></a>00558             yLabelMargin  = 0.10;
<a name="l00559"></a>00559             
<a name="l00560"></a>00560             axisPos = get(axesHandle, &#39;Position&#39;);
<a name="l00561"></a>00561             set(axesHandle, &#39;Position&#39;, axisPos + [1.7*yLabelMargin 0 -yLabelMargin 0]);
<a name="l00562"></a>00562             axisPos = get(axesHandle, &#39;Position&#39;);
<a name="l00563"></a>00563             
<a name="l00564"></a>00564             set(get(axesHandle, &#39;YLabel&#39;), &#39;Units&#39;, &#39;normalized&#39;);
<a name="l00565"></a>00565             yLabelPos = get(get(axesHandle,&#39;YLabel&#39;), &#39;Position&#39;);
<a name="l00566"></a>00566             set(get(axesHandle, &#39;YLabel&#39;), &#39;Position&#39;, yLabelPos + [-2.1*yLabelMargin/axisPos(3) 0 0]);
<a name="l00567"></a>00567             
<a name="l00568"></a>00568             %tick labels
<a name="l00569"></a>00569             props = {<span class="stringliteral">&#39;FontSize&#39;</span>, 10, <span class="stringliteral">&#39;HorizontalAlignment&#39;</span>, <span class="stringliteral">&#39;right&#39;</span>, <span class="stringliteral">&#39;VerticalAlignment&#39;</span>,<span class="stringliteral">&#39;middle&#39;</span>, <span class="stringliteral">&#39;units&#39;</span>, <span class="stringliteral">&#39;normalized&#39;</span>, <span class="stringliteral">&#39;Parent&#39;</span>, axesHandle};
<a name="l00570"></a>00570             <span class="keywordflow">for</span> i = 1:numel(sim.processWholeCellModelIDs)
<a name="l00571"></a>00571                 if ismember(order(i), mOtherIdxs)
<a name="l00572"></a>00572                     color = colors(1, :);
<a name="l00573"></a>00573                 elseif ismember(order(i), mProtIdxs)
<a name="l00574"></a>00574                     color = colors(2, :);
<a name="l00575"></a>00575                 elseif ismember(order(i), mRNAIdxs)
<a name="l00576"></a>00576                     color = colors(3, :);
<a name="l00577"></a>00577                 else
<a name="l00578"></a>00578                     color = colors(4, :);
<a name="l00579"></a>00579                 end
<a name="l00580"></a>00580                 text(-axisPos(3)/100, (i+0.1)/(numel(order)+1), processNames{order(i)}, props{:}, <span class="stringliteral">&#39;Color&#39;</span>, color);
<a name="l00581"></a>00581             end
<a name="l00582"></a>00582             
<a name="l00583"></a>00583             %category labels
<a name="l00584"></a>00584             props = {<span class="stringliteral">&#39;Rotation&#39;</span>, 90, <span class="stringliteral">&#39;FontSize&#39;</span>, 14, <span class="stringliteral">&#39;HorizontalAlignment&#39;</span>, <span class="stringliteral">&#39;center&#39;</span>, <span class="stringliteral">&#39;VerticalAlignment&#39;</span>,<span class="stringliteral">&#39;middle&#39;</span>, <span class="stringliteral">&#39;units&#39;</span>, <span class="stringliteral">&#39;normalized&#39;</span>, <span class="stringliteral">&#39;Parent&#39;</span>, axesHandle};
<a name="l00585"></a>00585             yLabelPos = <span class="keyword">get</span>(<span class="keyword">get</span>(axesHandle,<span class="stringliteral">&#39;ylabel&#39;</span>), <span class="stringliteral">&#39;position&#39;</span>);
<a name="l00586"></a>00586             text(yLabelPos(1)+yLabelMargin - 0.06, 1/(numel(order)+1)*(1/2 + numel(mOtherIdxs)/2), <span class="stringliteral">&#39;Other&#39;</span>, props{:}, <span class="stringliteral">&#39;Color&#39;</span>, colors(1, :));
<a name="l00587"></a>00587             text(yLabelPos(1)+yLabelMargin - 0.06, 1/(numel(order)+1)*(1/2 + numel(mOtherIdxs) + numel(mProtIdxs)/2), <span class="stringliteral">&#39;Protein&#39;</span>, props{:}, <span class="stringliteral">&#39;Color&#39;</span>, colors(2, :));
<a name="l00588"></a>00588             text(yLabelPos(1)+yLabelMargin - 0.06, 1/(numel(order)+1)*(1/2 + numel(mOtherIdxs) + numel(mProtIdxs) + numel(mRNAIdxs)/2), <span class="stringliteral">&#39;RNA&#39;</span>, props{:}, <span class="stringliteral">&#39;Color&#39;</span>, colors(3, :));
<a name="l00589"></a>00589             text(yLabelPos(1)+yLabelMargin - 0.06, 1/(numel(order)+1)*(1/2 + numel(mOtherIdxs) + numel(mProtIdxs) + numel(mRNAIdxs) + numel(mDNAIdxs)/2), <span class="stringliteral">&#39;DNA&#39;</span>, props{:}, <span class="stringliteral">&#39;Color&#39;</span>, colors(4, :));
<a name="l00590"></a>00590             
<a name="l00591"></a>00591             %category brackets
<a name="l00592"></a>00592             props = {<span class="stringliteral">&#39;LineWidth&#39;</span>, 1};
<a name="l00593"></a>00593             margin = 0.2;
<a name="l00594"></a>00594             tickW = 0.015;
<a name="l00595"></a>00595             yTickW = yLabelPos(1) + yLabelMargin + 0.03 - 0.05;
<a name="l00596"></a>00596             
<a name="l00597"></a>00597             annotation(figHandle, <span class="stringliteral">&#39;line&#39;</span>, axisPos(1) + axisPos(3)*(yTickW + [0; 0]),     axisPos(2) + axisPos(4)/(numel(order)+1)*(1/2 + [margin; numel(mOtherIdxs)-margin]), props{:}, <span class="stringliteral">&#39;Color&#39;</span>, colors(1, :));
<a name="l00598"></a>00598             annotation(figHandle, <span class="stringliteral">&#39;line&#39;</span>, axisPos(1) + axisPos(3)*(yTickW + [0; tickW]), axisPos(2) + axisPos(4)/(numel(order)+1)*(1/2 + [margin; margin]), props{:}, <span class="stringliteral">&#39;Color&#39;</span>, colors(1, :));
<a name="l00599"></a>00599             annotation(figHandle, <span class="stringliteral">&#39;line&#39;</span>, axisPos(1) + axisPos(3)*(yTickW + [0; tickW]), axisPos(2) + axisPos(4)/(numel(order)+1)*(1/2 + [numel(mOtherIdxs)-margin; numel(mOtherIdxs)-margin]), props{:}, <span class="stringliteral">&#39;Color&#39;</span>, colors(1, :));
<a name="l00600"></a>00600             
<a name="l00601"></a>00601             annotation(figHandle, <span class="stringliteral">&#39;line&#39;</span>, axisPos(1) + axisPos(3)*(yTickW + [0; 0]),     axisPos(2) + axisPos(4)/(numel(order)+1)*(1/2 + numel(mOtherIdxs) + [margin; numel(mProtIdxs)-margin]), props{:}, <span class="stringliteral">&#39;Color&#39;</span>, colors(2, :));
<a name="l00602"></a>00602             annotation(figHandle, <span class="stringliteral">&#39;line&#39;</span>, axisPos(1) + axisPos(3)*(yTickW + [0; tickW]), axisPos(2) + axisPos(4)/(numel(order)+1)*(1/2 + numel(mOtherIdxs) + [margin; margin]), props{:}, <span class="stringliteral">&#39;Color&#39;</span>, colors(2, :));
<a name="l00603"></a>00603             annotation(figHandle, <span class="stringliteral">&#39;line&#39;</span>, axisPos(1) + axisPos(3)*(yTickW + [0; tickW]), axisPos(2) + axisPos(4)/(numel(order)+1)*(1/2 + numel(mOtherIdxs) + [numel(mProtIdxs)-margin; numel(mProtIdxs)-margin]), props{:}, <span class="stringliteral">&#39;Color&#39;</span>, colors(2, :));
<a name="l00604"></a>00604             
<a name="l00605"></a>00605             annotation(figHandle, <span class="stringliteral">&#39;line&#39;</span>, axisPos(1) + axisPos(3)*(yTickW + [0; 0]),     axisPos(2) + axisPos(4)/(numel(order)+1)*(1/2 + numel(mOtherIdxs) + numel(mProtIdxs) + [margin; numel(mRNAIdxs)-margin]), props{:}, <span class="stringliteral">&#39;Color&#39;</span>, colors(3, :));
<a name="l00606"></a>00606             annotation(figHandle, <span class="stringliteral">&#39;line&#39;</span>, axisPos(1) + axisPos(3)*(yTickW + [0; tickW]), axisPos(2) + axisPos(4)/(numel(order)+1)*(1/2 + numel(mOtherIdxs) + numel(mProtIdxs) + [margin; margin]), props{:}, <span class="stringliteral">&#39;Color&#39;</span>, colors(3, :));
<a name="l00607"></a>00607             annotation(figHandle, <span class="stringliteral">&#39;line&#39;</span>, axisPos(1) + axisPos(3)*(yTickW + [0; tickW]), axisPos(2) + axisPos(4)/(numel(order)+1)*(1/2 + numel(mOtherIdxs) + numel(mProtIdxs) + [numel(mRNAIdxs)-margin; numel(mRNAIdxs)-margin]), props{:}, <span class="stringliteral">&#39;Color&#39;</span>, colors(3, :));
<a name="l00608"></a>00608             
<a name="l00609"></a>00609             annotation(figHandle, <span class="stringliteral">&#39;line&#39;</span>, axisPos(1) + axisPos(3)*(yTickW + [0; 0]),     axisPos(2) + axisPos(4)/(numel(order)+1)*(1/2 + numel(mOtherIdxs) + numel(mProtIdxs) + numel(mRNAIdxs) + [margin; numel(mDNAIdxs)-margin]), props{:}, <span class="stringliteral">&#39;Color&#39;</span>, colors(4, :));
<a name="l00610"></a>00610             annotation(figHandle, <span class="stringliteral">&#39;line&#39;</span>, axisPos(1) + axisPos(3)*(yTickW + [0; tickW]), axisPos(2) + axisPos(4)/(numel(order)+1)*(1/2 + numel(mOtherIdxs) + numel(mProtIdxs) + numel(mRNAIdxs) + [margin; margin]), props{:}, <span class="stringliteral">&#39;Color&#39;</span>, colors(4, :));
<a name="l00611"></a>00611             annotation(figHandle, <span class="stringliteral">&#39;line&#39;</span>, axisPos(1) + axisPos(3)*(yTickW + [0; tickW]), axisPos(2) + axisPos(4)/(numel(order)+1)*(1/2 + numel(mOtherIdxs) + numel(mProtIdxs) + numel(mRNAIdxs) + [numel(mDNAIdxs)-margin; numel(mDNAIdxs)-margin]), props{:}, <span class="stringliteral">&#39;Color&#39;</span>, colors(4, :));
<a name="l00612"></a>00612         end
<a name="l00613"></a>00613         
<a name="l00614"></a>00614         <span class="keyword">function</span> processNames = getShortProcessNames(sim)
<a name="l00615"></a>00615             processNames = cellfun(@(x) x.name, sim.processes, &#39;UniformOutput&#39;, false);
<a name="l00616"></a>00616             processNames{strcmp(sim.processWholeCellModelIDs, <span class="stringliteral">&#39;Process_ChromosomeCondensation&#39;</span>)}     = <span class="stringliteral">&#39;Condensation&#39;</span>;
<a name="l00617"></a>00617             processNames{strcmp(sim.processWholeCellModelIDs, <span class="stringliteral">&#39;Process_ChromosomeSegregation&#39;</span>)}      = <span class="stringliteral">&#39;Segregation&#39;</span>;
<a name="l00618"></a>00618             processNames{strcmp(sim.processWholeCellModelIDs, <span class="stringliteral">&#39;Process_Cytokinesis&#39;</span>)}                = <span class="stringliteral">&#39;Cytokinesis&#39;</span>;
<a name="l00619"></a>00619             processNames{strcmp(sim.processWholeCellModelIDs, <span class="stringliteral">&#39;Process_DNADamage&#39;</span>)}                  = <span class="stringliteral">&#39;Damage&#39;</span>;
<a name="l00620"></a>00620             processNames{strcmp(sim.processWholeCellModelIDs, <span class="stringliteral">&#39;Process_DNARepair&#39;</span>)}                  = <span class="stringliteral">&#39;Repair&#39;</span>;
<a name="l00621"></a>00621             processNames{strcmp(sim.processWholeCellModelIDs, <span class="stringliteral">&#39;Process_FtsZPolymerization&#39;</span>)}         = <span class="stringliteral">&#39;FtsZ&#39;</span>;
<a name="l00622"></a>00622             processNames{strcmp(sim.processWholeCellModelIDs, <span class="stringliteral">&#39;Process_HostInteraction&#39;</span>)}            = <span class="stringliteral">&#39;Host Interaction&#39;</span>;
<a name="l00623"></a>00623             processNames{strcmp(sim.processWholeCellModelIDs, <span class="stringliteral">&#39;Process_MacromolecularComplexation&#39;</span>)} = <span class="stringliteral">&#39;Complexation&#39;</span>;
<a name="l00624"></a>00624             processNames{strcmp(sim.processWholeCellModelIDs, <span class="stringliteral">&#39;Process_Metabolism&#39;</span>)}                 = <span class="stringliteral">&#39;Metabolism&#39;</span>;
<a name="l00625"></a>00625             processNames{strcmp(sim.processWholeCellModelIDs, <span class="stringliteral">&#39;Process_ProteinActivation&#39;</span>)}          = <span class="stringliteral">&#39;Activation&#39;</span>;
<a name="l00626"></a>00626             processNames{strcmp(sim.processWholeCellModelIDs, <span class="stringliteral">&#39;Process_ProteinDecay&#39;</span>)}               = <span class="stringliteral">&#39;Degradation&#39;</span>;
<a name="l00627"></a>00627             processNames{strcmp(sim.processWholeCellModelIDs, <span class="stringliteral">&#39;Process_ProteinFolding&#39;</span>)}             = <span class="stringliteral">&#39;Folding&#39;</span>;
<a name="l00628"></a>00628             processNames{strcmp(sim.processWholeCellModelIDs, <span class="stringliteral">&#39;Process_ProteinModification&#39;</span>)}        = <span class="stringliteral">&#39;Modification&#39;</span>;
<a name="l00629"></a>00629             processNames{strcmp(sim.processWholeCellModelIDs, <span class="stringliteral">&#39;Process_ProteinProcessingI&#39;</span>)}         = <span class="stringliteral">&#39;Processing I&#39;</span>;
<a name="l00630"></a>00630             processNames{strcmp(sim.processWholeCellModelIDs, <span class="stringliteral">&#39;Process_ProteinProcessingII&#39;</span>)}        = <span class="stringliteral">&#39;Processing II&#39;</span>;
<a name="l00631"></a>00631             processNames{strcmp(sim.processWholeCellModelIDs, <span class="stringliteral">&#39;Process_ProteinTranslocation&#39;</span>)}       = <span class="stringliteral">&#39;Translocation&#39;</span>;
<a name="l00632"></a>00632             processNames{strcmp(sim.processWholeCellModelIDs, <span class="stringliteral">&#39;Process_RNADecay&#39;</span>)}                   = <span class="stringliteral">&#39;Degradation&#39;</span>;
<a name="l00633"></a>00633             processNames{strcmp(sim.processWholeCellModelIDs, <span class="stringliteral">&#39;Process_RNAModification&#39;</span>)}            = <span class="stringliteral">&#39;Modification&#39;</span>;
<a name="l00634"></a>00634             processNames{strcmp(sim.processWholeCellModelIDs, <span class="stringliteral">&#39;Process_RNAProcessing&#39;</span>)}              = <span class="stringliteral">&#39;Processing&#39;</span>;
<a name="l00635"></a>00635             processNames{strcmp(sim.processWholeCellModelIDs, <span class="stringliteral">&#39;Process_Replication&#39;</span>)}                = <span class="stringliteral">&#39;Replication&#39;</span>;
<a name="l00636"></a>00636             processNames{strcmp(sim.processWholeCellModelIDs, <span class="stringliteral">&#39;Process_ReplicationInitiation&#39;</span>)}      = <span class="stringliteral">&#39;Rep Init&#39;</span>;
<a name="l00637"></a>00637             processNames{strcmp(sim.processWholeCellModelIDs, <span class="stringliteral">&#39;Process_RibosomeAssembly&#39;</span>)}           = <span class="stringliteral">&#39;Ribosome&#39;</span>;
<a name="l00638"></a>00638             processNames{strcmp(sim.processWholeCellModelIDs, <span class="stringliteral">&#39;Process_TerminalOrganelleAssembly&#39;</span>)}  = <span class="stringliteral">&#39;Term Org&#39;</span>;
<a name="l00639"></a>00639             processNames{strcmp(sim.processWholeCellModelIDs, <span class="stringliteral">&#39;Process_Transcription&#39;</span>)}              = <span class="stringliteral">&#39;Transcription&#39;</span>;
<a name="l00640"></a>00640             processNames{strcmp(sim.processWholeCellModelIDs, <span class="stringliteral">&#39;Process_TranscriptionalRegulation&#39;</span>)}  = <span class="stringliteral">&#39;Trans Reg&#39;</span>;
<a name="l00641"></a>00641             processNames{strcmp(sim.processWholeCellModelIDs, <span class="stringliteral">&#39;Process_Translation&#39;</span>)}                = <span class="stringliteral">&#39;Translation&#39;</span>;
<a name="l00642"></a>00642             processNames{strcmp(sim.processWholeCellModelIDs, <span class="stringliteral">&#39;Process_tRNAAminoacylation&#39;</span>)}         = <span class="stringliteral">&#39;Aminoacylation&#39;</span>;
<a name="l00643"></a>00643         end
<a name="l00644"></a>00644         
<a name="l00645"></a>00645         <span class="keyword">function</span> [mDNAIdxs, mRNAIdxs, mProtIdxs, mOtherIdxs] = processIndexs(sim)
<a name="l00646"></a>00646             mDNAIdxs = find(ismember(sim.processWholeCellModelIDs, {
<a name="l00647"></a>00647                 <span class="stringliteral">&#39;Process_ChromosomeCondensation&#39;</span>;
<a name="l00648"></a>00648                 <span class="stringliteral">&#39;Process_ChromosomeSegregation&#39;</span>;
<a name="l00649"></a>00649                 <span class="stringliteral">&#39;Process_DNADamage&#39;</span>;
<a name="l00650"></a>00650                 <span class="stringliteral">&#39;Process_DNARepair&#39;</span>;
<a name="l00651"></a>00651                 <span class="stringliteral">&#39;Process_Replication&#39;</span>;
<a name="l00652"></a>00652                 <span class="stringliteral">&#39;Process_ReplicationInitiation&#39;</span>;
<a name="l00653"></a>00653                 <span class="stringliteral">&#39;Process_DNASupercoiling&#39;</span>;
<a name="l00654"></a>00654                 <span class="stringliteral">&#39;Process_TranscriptionalRegulation&#39;</span>}));
<a name="l00655"></a>00655             mRNAIdxs = find(ismember(sim.processWholeCellModelIDs, {
<a name="l00656"></a>00656                 <span class="stringliteral">&#39;Process_RNADecay&#39;</span>;
<a name="l00657"></a>00657                 <span class="stringliteral">&#39;Process_RNAModification&#39;</span>;
<a name="l00658"></a>00658                 <span class="stringliteral">&#39;Process_RNAProcessing&#39;</span>;
<a name="l00659"></a>00659                 <span class="stringliteral">&#39;Process_Transcription&#39;</span>;
<a name="l00660"></a>00660                 <span class="stringliteral">&#39;Process_tRNAAminoacylation&#39;</span>}));
<a name="l00661"></a>00661             mProtIdxs = find(ismember(sim.processWholeCellModelIDs, {
<a name="l00662"></a>00662                 <span class="stringliteral">&#39;Process_MacromolecularComplexation&#39;</span>;
<a name="l00663"></a>00663                 <span class="stringliteral">&#39;Process_ProteinActivation&#39;</span>;
<a name="l00664"></a>00664                 <span class="stringliteral">&#39;Process_ProteinDecay&#39;</span>;
<a name="l00665"></a>00665                 <span class="stringliteral">&#39;Process_ProteinFolding&#39;</span>;
<a name="l00666"></a>00666                 <span class="stringliteral">&#39;Process_ProteinModification&#39;</span>;
<a name="l00667"></a>00667                 <span class="stringliteral">&#39;Process_ProteinProcessingI&#39;</span>;
<a name="l00668"></a>00668                 <span class="stringliteral">&#39;Process_ProteinProcessingII&#39;</span>;
<a name="l00669"></a>00669                 <span class="stringliteral">&#39;Process_ProteinTranslocation&#39;</span>;
<a name="l00670"></a>00670                 <span class="stringliteral">&#39;Process_RibosomeAssembly&#39;</span>;
<a name="l00671"></a>00671                 <span class="stringliteral">&#39;Process_TerminalOrganelleAssembly&#39;</span>;
<a name="l00672"></a>00672                 <span class="stringliteral">&#39;Process_Translation&#39;</span>}));
<a name="l00673"></a>00673             mOtherIdxs = find(ismember(sim.processWholeCellModelIDs, {
<a name="l00674"></a>00674                 <span class="stringliteral">&#39;Process_Cytokinesis&#39;</span>;
<a name="l00675"></a>00675                 <span class="stringliteral">&#39;Process_FtsZPolymerization&#39;</span>;
<a name="l00676"></a>00676                 <span class="stringliteral">&#39;Process_HostInteraction&#39;</span>;
<a name="l00677"></a>00677                 <span class="stringliteral">&#39;Process_Metabolism&#39;</span>}));
<a name="l00678"></a>00678         end
<a name="l00679"></a>00679         
<a name="l00680"></a>00680         <span class="keyword">function</span> colors = barColors()
<a name="l00681"></a>00681             colors1 = [
<a name="l00682"></a>00682                 179 162 199;
<a name="l00683"></a>00683                 217 150 148;
<a name="l00684"></a>00684                 195 214 155;
<a name="l00685"></a>00685                 149 179 215;
<a name="l00686"></a>00686                 250 192 144;]/255;
<a name="l00687"></a>00687             colors2 = [
<a name="l00688"></a>00688                 96 74 123;
<a name="l00689"></a>00689                 149 55 53;
<a name="l00690"></a>00690                 119 147 60;
<a name="l00691"></a>00691                 55 96 146;
<a name="l00692"></a>00692                 228 108 10;]/255;
<a name="l00693"></a>00693             colors = (colors1+colors2)/2;
<a name="l00694"></a>00694         end
<a name="l00695"></a>00695     end
<a name="l00696"></a>00696 end
</pre></div></div><!-- contents -->
</div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
<a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(0)"><span class="SelectionMark">&#160;</span>All</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(1)"><span class="SelectionMark">&#160;</span>Classes</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(2)"><span class="SelectionMark">&#160;</span>Files</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(3)"><span class="SelectionMark">&#160;</span>Functions</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(4)"><span class="SelectionMark">&#160;</span>Variables</a></div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

  <div id="nav-path" class="navpath">
    <ul>
      <li class="navelem"><a class="el" href="_simulation_structure_8m.html">SimulationStructure.m</a>      </li>

    <li class="footer">Generated on Wed Feb 8 2012 21:42:32 for Whole Cell Simulation by
    <a href="http://www.doxygen.org/index.html">
    <img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.7.6.1 </li>
   </ul>
 </div>


</body>
</html>
