<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<title>Whole Cell Simulation: src/+edu/+stanford/+covert/+cell/+sim/@Simulation/initializeState.m Source File</title>

<link href="tabs.css" rel="stylesheet" type="text/css"/>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
<link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="navtree.js"></script>
<script type="text/javascript">
  $(document).ready(initResizable);
</script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
  $(document).ready(function() { searchBox.OnSelectItem(0); });
</script>

</head>
<body>
<div id="top"><!-- do not remove this div! -->


<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  
  
  <td style="padding-left: 0.5em;">
   <div id="projectname">Whole Cell Simulation
   
   </div>
   
  </td>
  
  
  
 </tr>
 </tbody>
</table>
</div>

<!-- Generated by Doxygen 1.7.6.1 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "search",false,'Search');
</script>
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&#160;Page</span></a></li>
      <li><a href="annotated.html"><span>Classes</span></a></li>
      <li class="current"><a href="files.html"><span>Files</span></a></li>
      <li>
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
  <div id="navrow2" class="tabs2">
    <ul class="tablist">
      <li><a href="files.html"><span>File&#160;List</span></a></li>
      <li><a href="globals.html"><span>File&#160;Members</span></a></li>
    </ul>
  </div>
</div>
<div id="side-nav" class="ui-resizable side-nav-resizable">
  <div id="nav-tree">
    <div id="nav-tree-contents">
    </div>
  </div>
  <div id="splitbar" style="-moz-user-select:none;" 
       class="ui-resizable-handle">
  </div>
</div>
<script type="text/javascript">
  initNavTree('initialize_state_8m.html','');
</script>
<div id="doc-content">
<div class="header">
  <div class="headertitle">
<div class="title">src/+edu/+stanford/+covert/+cell/+sim/@Simulation/initializeState.m</div>  </div>
</div><!--header-->
<div class="contents">
<a href="initialize_state_8m.html">Go to the documentation of this file.</a><div class="fragment"><pre class="fragment"><a name="l00001"></a>00001 <span class="keyword">function</span> <span class="keyword">this</span> = initializeState(<span class="keyword">this</span>)
<a name="l00002"></a>00002 % Initialize state of organism to expectation values.
<a name="l00003"></a>00003 % - Time
<a name="l00004"></a>00004 % - Numbers of metabolites, RNAs, protein monomers, protein complexes
<a name="l00005"></a>00005 % - Resource allocation
<a name="l00006"></a>00006 %
<a name="l00007"></a>00007 % Author: Jonathan Karr, jkarr@stanford.edu
<a name="l00008"></a>00008 % Affiliation: Covert Lab, Department of Bioengineering, Stanford University
<a name="l00009"></a>00009 % Last updated: 7/17/2009
<a name="l00010"></a>00010 
<a name="l00011"></a>00011 %% import classes
<a name="l00012"></a>00012 import edu.stanford.covert.util.ComputationUtil;
<a name="l00013"></a>00013 <span class="keyword">import</span> edu.stanford.covert.util.ConstantUtil;
<a name="l00014"></a>00014 <span class="keyword">import</span> edu.stanford.covert.cell.sim.constant.Condition;
<a name="l00015"></a>00015 <span class="keyword">import</span> edu.stanford.covert.cell.sim.Simulation;
<a name="l00016"></a>00016 
<a name="l00017"></a>00017 %% data
<a name="l00018"></a>00018 %states
<a name="l00019"></a>00019 g = this.gene;
<a name="l00020"></a>00020 t = this.state(<span class="stringliteral">&#39;Time&#39;</span>);
<a name="l00021"></a>00021 s = this.state(<span class="stringliteral">&#39;Stimulus&#39;</span>);
<a name="l00022"></a>00022 m = this.state(<span class="stringliteral">&#39;Metabolite&#39;</span>);
<a name="l00023"></a>00023 c = this.state(<span class="stringliteral">&#39;Chromosome&#39;</span>);
<a name="l00024"></a>00024 r = this.state(<span class="stringliteral">&#39;Rna&#39;</span>);
<a name="l00025"></a>00025 pm = this.state(<span class="stringliteral">&#39;ProteinMonomer&#39;</span>);
<a name="l00026"></a>00026 pc = this.state(<span class="stringliteral">&#39;ProteinComplex&#39;</span>);
<a name="l00027"></a>00027 mass = this.state(<span class="stringliteral">&#39;Mass&#39;</span>);
<a name="l00028"></a>00028 mr = this.state(<span class="stringliteral">&#39;MetabolicReaction&#39;</span>);
<a name="l00029"></a>00029 geom = this.state(<span class="stringliteral">&#39;Geometry&#39;</span>);
<a name="l00030"></a>00030 ftsZRing = this.state(<span class="stringliteral">&#39;FtsZRing&#39;</span>);
<a name="l00031"></a>00031 transcript = this.state(<span class="stringliteral">&#39;Transcript&#39;</span>);
<a name="l00032"></a>00032 polypeptide = this.state(<span class="stringliteral">&#39;Polypeptide&#39;</span>);
<a name="l00033"></a>00033 dnaRepair = this.process(<span class="stringliteral">&#39;DNARepair&#39;</span>);
<a name="l00034"></a>00034 
<a name="l00035"></a>00035 %indices
<a name="l00036"></a>00036 cIdx = this.compartment.cytosolIndexs;
<a name="l00037"></a>00037 
<a name="l00038"></a>00038 %create local variables <span class="keywordflow">for</span> convenience
<a name="l00039"></a>00039 rnaMWs = 1 / <a class="code" href="class_constant_util.html">ConstantUtil</a>.<a class="code" href="class_constant_util.html#a9da93ee8ba55dd66d9406905f4c63a4b">nAvogadro</a> * r.molecularWeights;
<a name="l00040"></a>00040 monMWs = 1 / <a class="code" href="class_constant_util.html">ConstantUtil</a>.<a class="code" href="class_constant_util.html#a9da93ee8ba55dd66d9406905f4c63a4b">nAvogadro</a> * pm.molecularWeights;
<a name="l00041"></a>00041 
<a name="l00042"></a>00042 chrComp = [getBaseCounts(c.sequence); 0];
<a name="l00043"></a>00043 <span class="keywordflow">if</span> ~isempty(dnaRepair)
<a name="l00044"></a>00044     m6dAMPComp = 2 * size(dnaRepair.RM_MunI_RecognitionSites, 1);
<a name="l00045"></a>00045     chrComp(5) = m6dAMPComp;
<a name="l00046"></a>00046     chrComp(1) = chrComp(1) - m6dAMPComp;
<a name="l00047"></a>00047 end
<a name="l00048"></a>00048 rnaComp = r.baseCounts;
<a name="l00049"></a>00049 monComp = pm.baseCounts;
<a name="l00050"></a>00050 
<a name="l00051"></a>00051 rnaExp = r.expression;
<a name="l00052"></a>00052 
<a name="l00053"></a>00053 %% seed random number generator
<a name="l00054"></a>00054 this.seedRandStream();
<a name="l00055"></a>00055 
<a name="l00056"></a>00056 %% clear simulated <a class="code" href="json_parse_8m.html#a6e4256f69b751d3d7ec4ea52ff5a8dee">properties</a>
<a name="l00057"></a>00057 this.allocateMemoryForState(1);
<a name="l00058"></a>00058 
<a name="l00059"></a>00059 %% time
<a name="l00060"></a>00060 t.initialize();
<a name="l00061"></a>00061 
<a name="l00062"></a>00062 %% total cell mass
<a name="l00063"></a>00063 totCellWt = this.randStream.random(&#39;norm&#39;, 1, mass.cellInitialMassVariation);
<a name="l00064"></a>00064 
<a name="l00065"></a>00065 bComp = totCellWt * m.biomassComposition;
<a name="l00066"></a>00066 rnaWt = totCellWt * mass.cellInitialDryWeight * mass.dryWeightFractionRNA;
<a name="l00067"></a>00067 prtWt = totCellWt * mass.cellInitialDryWeight * mass.dryWeightFractionProtein;
<a name="l00068"></a>00068 
<a name="l00069"></a>00069 %% metabolites
<a name="l00070"></a>00070 m.counts = max(0, this.randStream.stochasticRound(bComp));
<a name="l00071"></a>00071 
<a name="l00072"></a>00072 %% DNA
<a name="l00073"></a>00073 %Initialize the free dNTP pool. <a class="code" href="class_chromosome.html">Chromosome</a> is initialized below in
<a name="l00074"></a>00074 %initialize states section.
<a name="l00075"></a>00075 c.initialize();
<a name="l00076"></a>00076 m.counts([m.dnmpIndexs; m.getIndexs(&#39;m6dAMP&#39;)], cIdx) = ...
<a name="l00077"></a>00077     m.counts([m.dnmpIndexs; m.getIndexs(&#39;m6dAMP&#39;)], cIdx) - chrComp;
<a name="l00078"></a>00078 
<a name="l00079"></a>00079 %% RNAs
<a name="l00080"></a>00080 %initialize RNAs based on expected expression
<a name="l00081"></a>00081 switch this.macromoleculeStateInitialization
<a name="l00082"></a>00082     case &#39;expected&#39;
<a name="l00083"></a>00083         r.counts(:, cIdx) = this.randStream.stochasticRound(...
<a name="l00084"></a>00084             mass.initialFractionNTPsInRNAs * rnaWt * rnaExp / (rnaMWs&#39; * rnaExp));
<a name="l00085"></a>00085     case &#39;multinomial&#39;
<a name="l00086"></a>00086         prob = rnaExp;
<a name="l00087"></a>00087         idxs = this.randStream.randsample(numel(prob), ceil(2 * rnaWt / (rnaMWs&#39; * prob)), true, prob);
<a name="l00088"></a>00088         idxs = idxs(1:find(cumsum(rnaMWs(idxs)) &gt; mass.initialFractionNTPsInRNAs * rnaWt, 1, &#39;first&#39;) - 1);
<a name="l00089"></a>00089         r.counts(:, cIdx) = this.randStream.stochasticRound(histc(idxs, 1:numel(prob)));
<a name="l00090"></a>00090     otherwise, throw(MException(&#39;Simulation:error&#39;, &#39;Unsupported macromoleculeStateInitialization %s&#39;, this.macromoleculeStateInitialization));
<a name="l00091"></a>00091 end
<a name="l00092"></a>00092 m.counts = m.counts - rnaComp&#39; * r.counts;
<a name="l00093"></a>00093 
<a name="l00094"></a>00094 %% protein monomers
<a name="l00095"></a>00095 %1. initialize protein monomers based on expected expression
<a name="l00096"></a>00096 %2. sort to appropriate compartment
<a name="l00097"></a>00097 switch this.macromoleculeStateInitialization
<a name="l00098"></a>00098     case &#39;expected&#39;
<a name="l00099"></a>00099         monExp = zeros(size(pm.wholeCellModelIDs));
<a name="l00100"></a>00100         monExp(pm.matureIndexs) = (r.matureRNAGeneComposition(g.mRNAIndexs, :) * r.expression(r.matureIndexs)) ./ ...
<a name="l00101"></a>00101             (log(2) / t.cellCycleLength + pm.decayRates(pm.matureIndexs));
<a name="l00102"></a>00102         monExp = monExp / sum(monExp);
<a name="l00103"></a>00103         
<a name="l00104"></a>00104         pm.counts(sub2ind(size(pm.counts), (1:size(pm.counts,1))&#39;, pm.compartments)) = ...
<a name="l00105"></a>00105             this.randStream.stochasticRound(mass.initialFractionAAsInMonomers * prtWt * monExp / (monMWs&#39; * monExp));
<a name="l00106"></a>00106     case &#39;multinomial&#39;
<a name="l00107"></a>00107         n = pm.macromoleculeStateInitializationVariation;
<a name="l00108"></a>00108         
<a name="l00109"></a>00109         if rnaWt &gt; 0
<a name="l00110"></a>00110             mRNAExp = rnaExp(r.matureIndexs(r.matureMRNAIndexs));
<a name="l00111"></a>00111             mRNAExp = mRNAExp / sum(mRNAExp);
<a name="l00112"></a>00112             mRNAProd = mRNAExp .* r.decayRates(r.matureIndexs(r.matureMRNAIndexs));
<a name="l00113"></a>00113             mRNAProd = mRNAProd / sum(mRNAProd);
<a name="l00114"></a>00114             mRNACnt = mRNAExp * rnaWt * r.weightFractionMRNA * <a class="code" href="class_constant_util.html">ConstantUtil</a>.nAvogadro / ...
<a name="l00115"></a>00115                 (mRNAExp&#39; * r.molecularWeights(r.matureIndexs(r.matureMRNAIndexs)));
<a name="l00116"></a>00116             totMRNAProd = (sum(mRNACnt) * log(2) / t.cellCycleLength + r.decayRates(r.matureIndexs(r.matureMRNAIndexs))&#39; * mRNACnt) * t.cellCycleLength / log(2);
<a name="l00117"></a>00117             
<a name="l00118"></a>00118             sample_mRNAProd = this.randStream.mnrnd(round(n * totMRNAProd), mRNAProd)&#39;;
<a name="l00119"></a>00119             sample_mRNAExp = zeros(size(sample_mRNAProd));
<a name="l00120"></a>00120             for i = 1:numel(sample_mRNAExp)
<a name="l00121"></a>00121                 if sample_mRNAProd(i) &gt; 0
<a name="l00122"></a>00122                     sample_mRNAExp(i) = sample_mRNAProd(i) * ...
<a name="l00123"></a>00123                         mean(this.randStream.random(&#39;exponential&#39;, ...
<a name="l00124"></a>00124                         1 / r.decayRates(r.matureIndexs(r.matureMRNAIndexs(i))), ...
<a name="l00125"></a>00125                         [sample_mRNAProd(i) 1]));
<a name="l00126"></a>00126                 end
<a name="l00127"></a>00127             end
<a name="l00128"></a>00128         else
<a name="l00129"></a>00129             sample_mRNAExp = r.expression(r.matureIndexs(r.matureMRNAIndexs));
<a name="l00130"></a>00130             sample_mRNAExp = sample_mRNAExp / sum(sample_mRNAExp);
<a name="l00131"></a>00131         end
<a name="l00132"></a>00132         
<a name="l00133"></a>00133         if prtWt &gt; 0
<a name="l00134"></a>00134             monProd = r.matureRNAGeneComposition(g.mRNAIndexs, r.matureMRNAIndexs) * sample_mRNAExp;
<a name="l00135"></a>00135             monProd = monProd / sum(monProd);
<a name="l00136"></a>00136             monExp = monProd ./ (log(2) / t.cellCycleLength + pm.decayRates(pm.matureIndexs));
<a name="l00137"></a>00137             monExp = monExp / sum(monExp);
<a name="l00138"></a>00138             monCnt = prtWt * <a class="code" href="class_constant_util.html">ConstantUtil</a>.nAvogadro / (monExp&#39; * pm.molecularWeights(pm.matureIndexs)) * monExp;
<a name="l00139"></a>00139             totMonProd = monCnt&#39; * (1 + t.cellCycleLength / log(2) * pm.decayRates(pm.matureIndexs));
<a name="l00140"></a>00140             
<a name="l00141"></a>00141             sample_monExp = this.randStream.mnrnd(round(n * totMonProd), monExp)&#39;;
<a name="l00142"></a>00142             sample_monExp = this.randStream.stochasticRound(n * sample_monExp * prtWt * mass.initialFractionAAsInMonomers * ...
<a name="l00143"></a>00143                 <a class="code" href="class_constant_util.html">ConstantUtil</a>.nAvogadro / (sample_monExp&#39; * pm.molecularWeights(pm.matureIndexs)));
<a name="l00144"></a>00144             sample_monExp2 = zeros(size(sample_monExp));
<a name="l00145"></a>00145             tmp = cumsum(sample_monExp);
<a name="l00146"></a>00146             order = this.randStream.randperm(tmp(end));
<a name="l00147"></a>00147             for i = 1:tmp(end)/2
<a name="l00148"></a>00148                 idx = find(order(i) &lt;= tmp, 1, &#39;first&#39;);
<a name="l00149"></a>00149                 sample_monExp2(idx) = sample_monExp2(idx) + 1;
<a name="l00150"></a>00150             end
<a name="l00151"></a>00151             pm.counts(sub2ind(size(pm.counts), pm.matureIndexs, pm.compartments(pm.matureIndexs))) = ...
<a name="l00152"></a>00152                 this.randStream.stochasticRound(2/n * sample_monExp2);
<a name="l00153"></a>00153         end
<a name="l00154"></a>00154     otherwise, throw(MException(&#39;Simulation:error&#39;, &#39;Unsupported macromoleculeStateInitialization %s&#39;, this.macromoleculeStateInitialization));
<a name="l00155"></a>00155 end
<a name="l00156"></a>00156 m.counts(:, cIdx) = m.counts(:, cIdx) - monComp&#39; * sum(pm.counts, 2);
<a name="l00157"></a>00157 
<a name="l00158"></a>00158 if strcmp(this.macromoleculeStateInitialization, &#39;expected&#39;)
<a name="l00159"></a>00159     pcComp = sum(pc.proteinComplexComposition, 3);
<a name="l00160"></a>00160     nonFormCpxIdxs = find(pc.formationProcesses(pc.matureIndexs) ~= this.processIndex(&#39;MacromolecularComplexation&#39;));
<a name="l00161"></a>00161     
<a name="l00162"></a>00162     subunits = zeros(size(g.wholeCellModelIDs));
<a name="l00163"></a>00163     subunits(g.mRNAIndexs) = sum(pm.counts(pm.matureIndexs, :), 2);
<a name="l00164"></a>00164     subunits(g.rRNAIndexs) = sum(r.counts(r.matureRRNAIndexs, :), 2);
<a name="l00165"></a>00165     subunits(g.sRNAIndexs) = sum(r.counts(r.matureSRNAIndexs, :), 2);
<a name="l00166"></a>00166     subunits(g.tRNAIndexs) = sum(r.counts(r.matureTRNAIndexs, :), 2);
<a name="l00167"></a>00167     cpxs = floor(min(repmat(subunits, 1, numel(pc.matureIndexs)) ./ pcComp, [], 1))&#39;;
<a name="l00168"></a>00168     cpxs(nonFormCpxIdxs) = 0;
<a name="l00169"></a>00169     
<a name="l00170"></a>00170     [subs2Nets, cpxs2Nets, nets] = edu.stanford.covert.util.<a class="code" href="find_non_interacting_rows_and_columns_8m.html#a35f7ca7ac9040917387bac96a24f6b30">findNonInteractingRowsAndColumns</a>(pcComp);
<a name="l00171"></a>00171     for i = 1:numel(nets)
<a name="l00172"></a>00172         if size(nets{i}, 2) &lt;= 1
<a name="l00173"></a>00173             <span class="keywordflow">continue</span>;
<a name="l00174"></a>00174         end
<a name="l00175"></a>00175         
<a name="l00176"></a>00176         tmpSubIdxs = find(subs2Nets == i);
<a name="l00177"></a>00177         tmpCpxIdxs = find(cpxs2Nets == i);
<a name="l00178"></a>00178         tmpPcComp = pcComp(tmpSubIdxs, tmpCpxIdxs);
<a name="l00179"></a>00179         tmpSubs = subunits(tmpSubIdxs);
<a name="l00180"></a>00180         tmpCpxs = zeros(size(tmpCpxIdxs));
<a name="l00181"></a>00181         <span class="keywordflow">while</span> <span class="keyword">true</span>
<a name="l00182"></a>00182             <span class="keywordflow">if</span> all(sum(tmpCpxIdxs, 1) == sum(tmpCpxIdxs(:, 1)))
<a name="l00183"></a>00183                 tmpRates = prod(repmat(tmpSubs / mean(tmpSubs), 1, numel(tmpCpxIdxs)) .^ tmpPcComp, 1)&#39;;
<a name="l00184"></a>00184             else
<a name="l00185"></a>00185                 tmpRates = prod(repmat(tmpSubs / sum(totMons), 1, numel(tmpCpxIdxs)) .^ tmpPcComp, 1)&#39;;
<a name="l00186"></a>00186             end
<a name="l00187"></a>00187             tmpRates(ismember(tmpCpxIdxs, nonFormCpxIdxs)) = 0;
<a name="l00188"></a>00188             if ~any(tmpRates) || max(min(repmat(tmpSubs, 1, numel(tmpCpxIdxs)) ./ tmpPcComp, [], 1)) &lt; 1<a class="code" href="json_parse_8m.html#ac14d26badced2970320ee4d772ad0ae9">e</a>-3
<a name="l00189"></a>00189                 break;
<a name="l00190"></a>00190             end
<a name="l00191"></a>00191             tmpCpxs = tmpCpxs + tmpRates * min(tmpSubs ./ (tmpPcComp * tmpRates));
<a name="l00192"></a>00192             tmpSubs = subunits(tmpSubIdxs) - tmpPcComp * tmpCpxs;
<a name="l00193"></a>00193         end
<a name="l00194"></a>00194         cpxs(tmpCpxIdxs) = floor(tmpCpxs);
<a name="l00195"></a>00195     end
<a name="l00196"></a>00196     
<a name="l00197"></a>00197     pc.counts(sub2ind(...
<a name="l00198"></a>00198         size(pc.counts), ...
<a name="l00199"></a>00199         pc.matureIndexs, ...
<a name="l00200"></a>00200         pc.compartments(pc.matureIndexs))) = ...
<a name="l00201"></a>00201         cpxs;
<a name="l00202"></a>00202     
<a name="l00203"></a>00203     r.counts(r.matureIndexs(setdiff(1:end, r.matureMRNAIndexs)), :) = ...
<a name="l00204"></a>00204         r.counts(r.matureIndexs(setdiff(1:end, r.matureMRNAIndexs)), :) - ...
<a name="l00205"></a>00205         sum(permute(pc.proteinComplexComposition(setdiff(1:end, g.mRNAIndexs), :, :), [1 3 2]) .* ...
<a name="l00206"></a>00206         repmat(permute(cpxs, [2 3 1]), [numel(g.rRNAIndexs)+numel(g.sRNAIndexs)+numel(g.tRNAIndexs) size(pc.proteinComplexComposition, 3) 1]), 3);
<a name="l00207"></a>00207     pm.counts(pm.matureIndexs, :) = ...
<a name="l00208"></a>00208         pm.counts(pm.matureIndexs, :) - ...
<a name="l00209"></a>00209         sum(permute(pc.proteinComplexComposition(g.mRNAIndexs, :, :), [1 3 2]) .* ...
<a name="l00210"></a>00210         repmat(permute(cpxs, [2 3 1]), [numel(g.mRNAIndexs) size(pc.proteinComplexComposition, 3) 1]), 3);
<a name="l00211"></a>00211 end
<a name="l00212"></a>00212 
<a name="l00213"></a>00213 %% nucleotides
<a name="l00214"></a>00214 V = totCellWt * mass.cellInitialDryWeight / (1 - mass.fractionWetWeight) / geom.density; %volume (L)
<a name="l00215"></a>00215 m.counts(m.ntpIndexs([1 3]), cIdx)  = round(1<a class="code" href="json_parse_8m.html#ac14d26badced2970320ee4d772ad0ae9">e</a>-3 * <a class="code" href="class_constant_util.html">ConstantUtil</a>.nAvogadro * V * m.meanNTPConcentration);
<a name="l00216"></a>00216 m.counts(m.ntpIndexs([2 4]), cIdx)  = 0;
<a name="l00217"></a>00217 m.counts(m.ndpIndexs([1 3]), cIdx)  = round(1<a class="code" href="json_parse_8m.html#ac14d26badced2970320ee4d772ad0ae9">e</a>-3 * <a class="code" href="class_constant_util.html">ConstantUtil</a>.nAvogadro * V * m.meanNDPConcentration);
<a name="l00218"></a>00218 m.counts(m.ndpIndexs([2 4]), cIdx)  = 0;
<a name="l00219"></a>00219 m.counts(m.nmpIndexs, cIdx)         = round(1<a class="code" href="json_parse_8m.html#ac14d26badced2970320ee4d772ad0ae9">e</a>-3 * <a class="code" href="class_constant_util.html">ConstantUtil</a>.nAvogadro * V * m.meanNMPConcentration);
<a name="l00220"></a>00220 m.counts(m.phosphateIndexs, cIdx)   = sum(m.counts(m.ndpIndexs, cIdx));
<a name="l00221"></a>00221 m.counts(m.diphosphateIndexs, cIdx) = sum(m.counts(m.nmpIndexs, cIdx));
<a name="l00222"></a>00222 m.counts(m.hydrogenIndexs, cIdx)    = sum(m.counts(m.ndpIndexs, cIdx)) + sum(m.counts(m.nmpIndexs, cIdx));
<a name="l00223"></a>00223 
<a name="l00224"></a>00224 %% protein complexes: initialized by processes (see initialize processes section below)
<a name="l00225"></a>00225 %1. Macromolecular complexation
<a name="l00226"></a>00226 %2. Ribosome assembly
<a name="l00227"></a>00227 %3. <a class="code" href="class_protein.html">Protein</a> folding
<a name="l00228"></a>00228 %4. <a class="code" href="class_protein.html">Protein</a> activation
<a name="l00229"></a>00229 
<a name="l00230"></a>00230 %% initialize states
<a name="l00231"></a>00231 mass.initialize();
<a name="l00232"></a>00232 geom.initialize();
<a name="l00233"></a>00233 ftsZRing.initialize();
<a name="l00234"></a>00234 transcript.initialize();
<a name="l00235"></a>00235 polypeptide.initialize();
<a name="l00236"></a>00236 
<a name="l00237"></a>00237 %% apply media and stimulation conditions
<a name="l00238"></a>00238 m.counts = <a class="code" href="class_condition.html">Condition</a>.applyConditions(m.counts, m.setCounts, t.values); %media
<a name="l00239"></a>00239 s.values = <a class="code" href="class_condition.html">Condition</a>.applyConditions(s.values, s.setValues, t.values); %stimulation
<a name="l00240"></a>00240 
<a name="l00241"></a>00241 %% Assert metabolites non-negative
<a name="l00242"></a>00242 negIdxs = find(any(m.counts &lt; min(0, -bComp) * 15<a class="code" href="json_parse_8m.html#ac14d26badced2970320ee4d772ad0ae9">e</a>-2, 2));
<a name="l00243"></a>00243 if ~isempty(negIdxs)
<a name="l00244"></a>00244     warnStr = cellfun(@(x, y) sprintf(&#39;- %s: %.3e&#39;, x, y), m.wholeCellModelIDs(negIdxs), num2cell(m.counts(negIdxs)), &#39;UniformOutput&#39;, false);
<a name="l00245"></a>00245     warning(&#39;WholeCell:warning:initializeNegativeCounts&#39;, ...
<a name="l00246"></a>00246         [&#39;Metabolites should be non-negative:\n&#39; repmat(&#39;%s\n&#39;, 1, numel(negIdxs))], warnStr{:});
<a name="l00247"></a>00247 end
<a name="l00248"></a>00248 m.counts = max(0, m.counts);
<a name="l00249"></a>00249 
<a name="l00250"></a>00250 %% initialize processes
<a name="l00251"></a>00251 %- Initialize &quot;own&quot; state <a class="code" href="json_parse_8m.html#ad2fc293e8080616c8c1e2ffbd7b86742">while</a> preserving the cell mass
<a name="l00252"></a>00252 %- Processes &quot;own&quot; state includes, for example,
<a name="l00253"></a>00253 %  - <a class="code" href="class_r_n_a.html">RNA</a> polymerases in transcription
<a name="l00254"></a>00254 %  - Ribosomes in translation
<a name="l00255"></a>00255 %- In particular, when processes incorporate free metabolites into the cell&#39;s mass
<a name="l00256"></a>00256 %  they must alter the counts of simulation&#39;s free metabolites
<a name="l00257"></a>00257 %- Processes should assume that energy is not limited, and should not account for
<a name="l00258"></a>00258 %  energy usage. Specifically, processes should not alter the simulation&#39;s
<a name="l00259"></a>00259 %  metabolite counts to account for energy usage.
<a name="l00260"></a>00260 processes = this.processesInInitOrder;
<a name="l00261"></a>00261 <span class="keywordflow">for</span> i = 1:length(processes)
<a name="l00262"></a>00262     process = processes{i};
<a name="l00263"></a>00263     process.simulationStateSideEffects = [];
<a name="l00264"></a>00264     process.copyFromState();
<a name="l00265"></a>00265     process.initializeState();
<a name="l00266"></a>00266     process.copyToState();
<a name="l00267"></a>00267     <span class="keywordflow">if</span> ~isempty(process.simulationStateSideEffects)
<a name="l00268"></a>00268         process.simulationStateSideEffects.updateSimulationState(this);
<a name="l00269"></a>00269     end
<a name="l00270"></a>00270 end
<a name="l00271"></a>00271 this.state(&#39;Polypeptide&#39;).abortedPolypeptides = this.state(&#39;Polypeptide&#39;).abortedPolypeptides([], :);
<a name="l00272"></a>00272 this.state(&#39;Transcript&#39;).abortedTranscripts = this.state(&#39;Transcript&#39;).abortedTranscripts([], :);
<a name="l00273"></a>00273 
<a name="l00274"></a>00274 %% Assert metabolites non-negative
<a name="l00275"></a>00275 negIdxs = find(any(m.counts &lt; min(0, -bComp) * 15<a class="code" href="json_parse_8m.html#ac14d26badced2970320ee4d772ad0ae9">e</a>-2, 2));
<a name="l00276"></a>00276 if ~isempty(negIdxs)
<a name="l00277"></a>00277     warnStr = cellfun(@(x, y) sprintf(&#39;- %s: %.3e&#39;, x, y), m.wholeCellModelIDs(negIdxs), num2cell(m.counts(negIdxs)), &#39;UniformOutput&#39;, false);
<a name="l00278"></a>00278     warning(&#39;WholeCell:warning&#39;, ...
<a name="l00279"></a>00279         [&#39;Metabolites should be non-negative:\n&#39; repmat(&#39;%s\n&#39;, 1, numel(negIdxs))], warnStr{:});
<a name="l00280"></a>00280 end
<a name="l00281"></a>00281 m.counts = max(0, m.counts);
<a name="l00282"></a>00282 
<a name="l00283"></a>00283 %% update cell volume
<a name="l00284"></a>00284 mass.calcMass();
<a name="l00285"></a>00285 geom.calculateVolume();
<a name="l00286"></a>00286 
<a name="l00287"></a>00287 %% apply media and stimulation conditions
<a name="l00288"></a>00288 m.counts = <a class="code" href="class_condition.html">Condition</a>.<a class="code" href="class_condition.html#ab837d627d83db887962990db244930e4">applyConditions</a>(m.counts, m.setCounts, t.values); %media
<a name="l00289"></a>00289 s.values = <a class="code" href="class_condition.html">Condition</a>.<a class="code" href="class_condition.html#ab837d627d83db887962990db244930e4">applyConditions</a>(s.values, s.setValues, t.values); %stimulation
<a name="l00290"></a>00290 
<a name="l00291"></a>00291 %% Synchronize process states with simulation
<a name="l00292"></a>00292 <span class="keywordflow">for</span> i = 1:length(this.processes)
<a name="l00293"></a>00293     this.processes{i}.copyFromState();
<a name="l00294"></a>00294 end
<a name="l00295"></a>00295 
<a name="l00296"></a>00296 %% If cell is dead, rerun initialize state
<a name="l00297"></a>00297 <span class="keywordflow">if</span> abs(mr.growth - mr.meanInitialGrowthRate) / mr.meanInitialGrowthRate &gt; mr.initialGrowthFilterWidth &amp;&amp; ~isempty(this.seed)
<a name="l00298"></a>00298     this.seed = this.randStream.randi([0 2^32-1], 1);
<a name="l00299"></a>00299     this.initializeState();
<a name="l00300"></a>00300 end
</pre></div></div><!-- contents -->
</div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
<a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(0)"><span class="SelectionMark">&#160;</span>All</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(1)"><span class="SelectionMark">&#160;</span>Classes</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(2)"><span class="SelectionMark">&#160;</span>Files</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(3)"><span class="SelectionMark">&#160;</span>Functions</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(4)"><span class="SelectionMark">&#160;</span>Variables</a></div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

  <div id="nav-path" class="navpath">
    <ul>
      <li class="navelem"><a class="el" href="initialize_state_8m.html">initializeState.m</a>      </li>

    <li class="footer">Generated on Wed Feb 8 2012 21:42:32 for Whole Cell Simulation by
    <a href="http://www.doxygen.org/index.html">
    <img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.7.6.1 </li>
   </ul>
 </div>


</body>
</html>
