<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN"
                "http://www.w3.org/TR/REC-html40/loose.dtd">
<html>
<head>
  <title>Description of Figures34</title>
  <meta name="keywords" content="Figures34">
  <meta name="description" content="Figures 3-4">
  <meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1">
  <meta name="generator" content="m2html v1.5 &copy; 2003-2005 Guillaume Flandin">
  <meta name="robots" content="index, follow">
  <link type="text/css" rel="stylesheet" href="../../../../../../../../m2html.css">
</head>
<body>
<a name="_top"></a>
<div><a href="../../../../../../../../index.html">Home</a> &gt;  <a href="../../../../../../../index.html">simulation</a> &gt; <a href="#">src</a> &gt; <a href="#">+edu</a> &gt; <a href="#">+stanford</a> &gt; <a href="#">+covert</a> &gt; <a href="#">+cell</a> &gt; <a href="../index.html">+sim</a> &gt; <a href="index.html">+analysis</a> &gt; Figures34.m</div>

<!--<table width="100%"><tr><td align="left"><a href="../../../../../../../../index.html"><img alt="<" border="0" src="../../../../../../../../left.png">&nbsp;Master index</a></td>
<td align="right"><a href="index.html">Index for simulation/src/+edu/+stanford/+covert/+cell/+sim/+analysis&nbsp;<img alt=">" border="0" src="../../../../../../../../right.png"></a></td></tr></table>-->

<h1>Figures34
</h1>

<h2><a name="_name"></a>PURPOSE <a href="#_top"><img alt="^" border="0" src="../../../../../../../../up.png"></a></h2>
<div class="box"><strong>Figures 3-4</strong></div>

<h2><a name="_synopsis"></a>SYNOPSIS <a href="#_top"><img alt="^" border="0" src="../../../../../../../../up.png"></a></h2>
<div class="box"><strong>This is a script file. </strong></div>

<h2><a name="_description"></a>DESCRIPTION <a href="#_top"><img alt="^" border="0" src="../../../../../../../../up.png"></a></h2>
<div class="fragment"><pre class="comment"> Figures 3-4

 Author: Derek Macklin, macklin@stanford.edu
 Author: Jonathan Karr, jkarr@stanford.edu
 Affiliation: Covert Lab, Department of Bioengineering, Stanford University
 Created: 1/12/2012</pre></div>

<!-- crossreference -->
<h2><a name="_cross"></a>CROSS-REFERENCE INFORMATION <a href="#_top"><img alt="^" border="0" src="../../../../../../../../up.png"></a></h2>
This function calls:
<ul style="list-style-image:url(../../../../../../../../matlabicon.gif)">
<li><a href="Figures34.html" class="code" title="">Figures34</a>	Figures 3-4</li></ul>
This function is called by:
<ul style="list-style-image:url(../../../../../../../../matlabicon.gif)">
<li><a href="Figures34.html" class="code" title="">Figures34</a>	Figures 3-4</li></ul>
<!-- crossreference -->

<h2><a name="_subfunctions"></a>SUBFUNCTIONS <a href="#_top"><img alt="^" border="0" src="../../../../../../../../up.png"></a></h2>
<ul style="list-style-image:url(../../../../../../../../matlabicon.gif)">
<li><a href="#_sub1" class="code">function run(outDirectory)</a></li><li><a href="#_sub2" class="code">function figure2(simBatchDir, selectedSim, outDirectory)</a></li><li><a href="#_sub3" class="code">function figure3(simBatchDir, selectedSim, outDirectory)</a></li><li><a href="#_sub4" class="code">function figure4(simBatchDir, selectedSim, outDirectory)</a></li><li><a href="#_sub5" class="code">function figure5(simBatchDir, selectedSim, outDirectory, plotUsageRate)</a></li><li><a href="#_sub6" class="code">function growthSensitivity(simBatchDir, selectedSim, outDirectory)</a></li><li><a href="#_sub7" class="code">function [optGrowths, optEnzymes, incEnzymes] = calcOptimalEnzymeExpressionChanges(fbaObj, fbaSMat, fbaRHS,</a></li><li><a href="#_sub8" class="code">function RNAvProtein(simBatchDir, outDirectory, figHandle, position)</a></li><li><a href="#_sub9" class="code">function cellShape(outDirectory, simBatchDir, selectedSim)</a></li><li><a href="#_sub10" class="code">function normCellMass(simBatchDir, selectedSim, outDirectory, figHandle, position)</a></li><li><a href="#_sub11" class="code">function populationMass(simBatchDir, selectedSim, figHandle, position, outDirectory)</a></li><li><a href="#_sub12" class="code">function growthMeasurement(figHandle, position, outDirectory)</a></li><li><a href="#_sub13" class="code">function doublingTime(outDirectory)</a></li><li><a href="#_sub14" class="code">function dryBiomassComposition(simBatchDir, figHandle, position, outDirectory)</a></li><li><a href="#_sub15" class="code">function cellCycleLength(simBatchDir, selectedSim, figHandle, position)</a></li><li><a href="#_sub16" class="code">function replicationInitiationVsReplicationDuration(simBatchDir, selectedSim, figHandle, position)</a></li><li><a href="#_sub17" class="code">function replicationInitiationVsReplicationDurationFit(simBatchDir, selectedSim, figHandle, position)</a></li><li><a href="#_sub18" class="code">function proteinBursting(simBatchDir, selectedSim, outDirectory, figHandle, position)</a></li><li><a href="#_sub19" class="code">function geneToPhenotype(simBatchDir, selectedSim, outDirectory, figHandle, position)</a></li><li><a href="#_sub20" class="code">function spaceTimePlot(simBatchDir, selectedSim, outDirectory, figHandle, position)</a></li><li><a href="#_sub21" class="code">function metaboliteConcentrations(simBatchDir, selectedSim, outDirectory, figHandle, position)</a></li><li><a href="#_sub22" class="code">function [exptGeneExpr, modelGeneExp] = geneExpression(simBatchDir, selectedSim, figHandle, position, logScale)</a></li><li><a href="#_sub23" class="code">function geneExpressionRatios(exptGeneExpr, modelGeneExpr, axesHandle)</a></li><li><a href="#_sub24" class="code">function circosPlot(simBatchDir, selectedSim, outDirectory, figHandle, position)</a></li><li><a href="#_sub25" class="code">function dnaBoundProteinDensityVsDisplacement(simBatchDir, selectedSim, outDirectory, figHandle, position)</a></li><li><a href="#_sub26" class="code">function [posFreq, protDensity] = analyzeDNABoundProteinDisplacements(simBatchDir, outDirectory, figHandle, position)</a></li><li><a href="#_sub27" class="code">function [mat, allIds, ids, names, shortNames] = groupProteinDisplacements(protByProtFreq, simEndTimes, sim)</a></li><li><a href="#_sub28" class="code">function plotCollisionProteinDistribution(figHandle, position, position2, mat, shortNames)</a></li><li><a href="#_sub29" class="code">function totalDensityMatrix = calculateDNABoundProteinDensity(outDirectory, simBatchDir, selectedSim)</a></li><li><a href="#_sub30" class="code">function displacements = calculateDNABoundProteinDisplacements(outDirectory, simBatchDir, selectedSim, iJob, nJobs)</a></li><li><a href="#_sub31" class="code">function metabolicMap(simBatchDir, selectedSim, outDirectory, figHandle, position, showAllLabels)</a></li><li><a href="#_sub32" class="code">function chromosomeSampling(simBatchDir, selectedSim, outDirectory, figHandle, position)</a></li><li><a href="#_sub33" class="code">function plotRNAPolSampling(simBatchDir, selectedSim, outDirectory, figHandle, position)</a></li><li><a href="#_sub34" class="code">function timeBound = calcChromosomeSampling(startPos, endPos, times, pos1, pos2, L)</a></li><li><a href="#_sub35" class="code">function rnaSampling(simBatchDir, selectedSim, outDirectory, figHandle, position)</a></li><li><a href="#_sub36" class="code">function mrnaSampling(simBatchDir, selectedSim, outDirectory, figHandle, position)</a></li><li><a href="#_sub37" class="code">function rgbStr = colorFunc(normVal)</a></li><li><a href="#_sub38" class="code">function svg = drawCell(scale, width, cylindricalLength, septumLength, maxTotalLength, maxWidth)</a></li></ul>

<h2><a name="_source"></a>SOURCE CODE <a href="#_top"><img alt="^" border="0" src="../../../../../../../../up.png"></a></h2>
<div class="fragment"><pre>0001 <span class="comment">% Figures 3-4</span>
0002 <span class="comment">%</span>
0003 <span class="comment">% Author: Derek Macklin, macklin@stanford.edu</span>
0004 <span class="comment">% Author: Jonathan Karr, jkarr@stanford.edu</span>
0005 <span class="comment">% Affiliation: Covert Lab, Department of Bioengineering, Stanford University</span>
0006 <span class="comment">% Created: 1/12/2012</span>
0007 classdef <a href="Figures34.html" class="code" title="">Figures34</a>
0008     methods (Static = true)
0009         <a name="_sub0" href="#_subfunctions" class="code">function run(outDirectory)</a>
0010             import edu.stanford.covert.cell.sim.analysis.Figures34;
0011             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
0012             
0013             <span class="comment">%output directory</span>
0014             <span class="keyword">if</span> nargin &lt; 1
0015                 outDirectory = <span class="string">'documentation/paper/figures'</span>;
0016             <span class="keyword">end</span>
0017             
0018             <span class="comment">%select simulation</span>
0019             simBatchDir = [SimulationDiskUtil.getBaseDir() filesep <span class="string">'2011_10_25_01_45_44'</span> filesep];
0020             selectedSim = 10;
0021             
0022             <span class="comment">%setup linux path</span>
0023             <span class="keyword">if</span> isunix()
0024                 a = getenv(<span class="string">'LD_LIBRARY_PATH'</span>);
0025                 a = [<span class="string">'/usr/lib:'</span> a];
0026                 setenv(<span class="string">'LD_LIBRARY_PATH'</span>, a);
0027             <span class="keyword">end</span>
0028             
0029             <span class="comment">% figure 2: fitting, validation</span>
0030             Figures34.figure2(simBatchDir, selectedSim, outDirectory)
0031             
0032             <span class="comment">% figure 3: chromosome</span>
0033             Figures34.figure3(simBatchDir, selectedSim, outDirectory)
0034             
0035             <span class="comment">% figure 4: cell cycle</span>
0036             Figures34.figure4(simBatchDir, selectedSim, outDirectory)
0037             
0038             <span class="comment">% figure 5: energy</span>
0039             Figures34.figure5(simBatchDir, selectedSim, outDirectory)
0040             
0041             <span class="comment">%growth sensitivity</span>
0042             Figures34.growthSensitivity(simBatchDir, selectedSim, outDirectory)
0043         <span class="keyword">end</span>
0044         
0045         <span class="comment">%fitting, validation</span>
0046         <a name="_sub1" href="#_subfunctions" class="code">function figure2(simBatchDir, selectedSim, outDirectory)</a>
0047             import edu.stanford.covert.cell.sim.analysis.Figures34;
0048             import edu.stanford.covert.cell.sim.util.PlotUtil;
0049             
0050             outDir = [outDirectory filesep <span class="string">'fittingAndValidation'</span> filesep];
0051             
0052             <span class="comment">%% figure</span>
0053             w = 11.4; h = 17; <span class="comment">% Cell 1.5 column</span>
0054             [~, figHandle] = PlotUtil.newAxesHandle();
0055             clf(figHandle);
0056             set(figHandle, <span class="string">'PaperUnits'</span>, <span class="string">'centimeters'</span>);
0057             set(figHandle, <span class="string">'PaperSize'</span>, [w h]);
0058             set(figHandle, <span class="string">'PaperPositionMode'</span>, <span class="string">'manual'</span>);
0059             set(figHandle, <span class="string">'PaperPosition'</span>, [0 0 get(figHandle, <span class="string">'PaperSize'</span>)]);
0060             set(figHandle, <span class="string">'Units'</span>, <span class="string">'normalized'</span>);
0061             
0062             <span class="comment">%A: growth measurement</span>
0063             axesPosA = [0.10 0.82 0.415 0.165];
0064             PlotUtil.labelSubFigure(<span class="string">'A'</span>, [0.128 0.992 -0.0286 -0.0208], figHandle);
0065             Figures34.growthMeasurement(figHandle, axesPosA, outDir);
0066             
0067             <span class="comment">%B: simulated growth</span>
0068             PlotUtil.labelSubFigure(<span class="string">'B'</span>, [0.607 0.992 -0.0286 -0.0208], figHandle);
0069             Figures34.populationMass(simBatchDir, selectedSim, figHandle, [0.585 0.82 0.415 0.165], outDir);
0070             
0071             <span class="comment">%C: biomass composition</span>
0072             axesPosC = [0.10 0.595 0.415 0.165];
0073             PlotUtil.labelSubFigure(<span class="string">'C'</span>, [0.128 0.772 -0.0286 -0.0208], figHandle);
0074             Figures34.dryBiomassComposition(simBatchDir, figHandle, axesPosC, outDir);
0075             
0076             <span class="comment">%D: mass fraction dynamics</span>
0077             PlotUtil.labelSubFigure(<span class="string">'D'</span>, [0.607 0.772 -0.0286 -0.0208], figHandle);
0078             Figures34.normCellMass(simBatchDir, selectedSim, outDir, figHandle, [0.577 0.595 0.423 0.165]);
0079             
0080             <span class="comment">%E: metabolic map</span>
0081             axesPosE = [0.07 0.29 0.50 0.215];
0082             PlotUtil.labelSubFigure(<span class="string">'E'</span>, [0.135 0.527 -0.0286 -0.0208], figHandle);
0083             Figures34.metabolicMap(simBatchDir, selectedSim, outDir, figHandle, axesPosE);
0084             
0085             <span class="comment">%F: ratio of predicted to observed metabolite concentrations</span>
0086             axesPosF = [0.11 0.057 0.58 0.18];
0087             PlotUtil.labelSubFigure(<span class="string">'F'</span>, [0.135 0.255 -0.0286 -0.0208], figHandle);
0088             Figures34.metaboliteConcentrations(simBatchDir, selectedSim, outDir, figHandle, axesPosF)
0089             
0090             <span class="comment">%G: mRNA, protein expression dynamics</span>
0091             PlotUtil.labelSubFigure(<span class="string">'G'</span>, [0.684 0.527 -0.0286 -0.0208], figHandle);
0092             Figures34.proteinBursting(simBatchDir, selectedSim, outDir, figHandle, [0.645 0.29 0.35 0.215]);
0093             
0094             <span class="comment">%H: scatter plot of RNA vs protein expression with marginals</span>
0095             PlotUtil.labelSubFigure(<span class="string">'H'</span>, [0.799 0.255 -0.0286 -0.0208], figHandle);
0096             Figures34.RNAvProtein(simBatchDir, outDir, figHandle, [0.765 0.057 0.23 0.18]);
0097             
0098             <span class="comment">% area labels</span>
0099             axesHandleA = findobj(figHandle, <span class="string">'position'</span>, axesPosA);
0100             axesHandleE = findobj(figHandle, <span class="string">'position'</span>, axesPosE);
0101             text(-0.212, 0.5 * (axesPosC(2)+axesPosC(4)-axesPosA(2)) / axesPosA(4), <span class="string">'Training Data'</span>, <span class="keyword">...</span>
0102                 <span class="string">'FontSize'</span>, 10, <span class="string">'rotation'</span>, 90, <span class="string">'Parent'</span>, axesHandleA, <span class="keyword">...</span>
0103                 <span class="string">'units'</span>, <span class="string">'normalized'</span>, <span class="string">'HorizontalAlign'</span>, <span class="string">'center'</span>, <span class="string">'VerticalAlign'</span>, <span class="string">'middle'</span>, <span class="keyword">...</span>
0104                 <span class="string">'FontWeight'</span>, <span class="string">'bold'</span>);
0105             text((-0.212 - (axesPosE(1) - axesPosA(1))/axesPosA(3)) * axesPosA(3) / axesPosE(3), <span class="keyword">...</span>
0106                 0.5 * (axesPosF(2)+axesPosF(4)-axesPosE(2)) / axesPosE(4), <span class="string">'Independent Validation'</span>, <span class="keyword">...</span>
0107                 <span class="string">'FontSize'</span>, 10, <span class="string">'rotation'</span>, 90, <span class="string">'Parent'</span>, axesHandleE, <span class="keyword">...</span>
0108                 <span class="string">'units'</span>, <span class="string">'normalized'</span>, <span class="string">'HorizontalAlign'</span>, <span class="string">'center'</span>, <span class="string">'VerticalAlign'</span>, <span class="string">'middle'</span>, <span class="keyword">...</span>
0109                 <span class="string">'FontWeight'</span>, <span class="string">'bold'</span>);
0110             annotation(figHandle, <span class="string">'line'</span>, 0.035 * [1 1], [0.570 0.997], <span class="string">'Color'</span>, <span class="string">'k'</span>, <span class="string">'LineWidth'</span>, 1)
0111             annotation(figHandle, <span class="string">'line'</span>, 0.035 * [1 1], [0.003 0.530], <span class="string">'Color'</span>, <span class="string">'k'</span>, <span class="string">'LineWidth'</span>, 1)
0112             annotation(figHandle, <span class="string">'line'</span>, [0.035 0.05], 0.570 * [1 1], <span class="string">'Color'</span>, <span class="string">'k'</span>, <span class="string">'LineWidth'</span>, 1)
0113             annotation(figHandle, <span class="string">'line'</span>, [0.035 0.05], 0.997 * [1 1], <span class="string">'Color'</span>, <span class="string">'k'</span>, <span class="string">'LineWidth'</span>, 1)
0114             annotation(figHandle, <span class="string">'line'</span>, [0.035 0.05], 0.530 * [1 1], <span class="string">'Color'</span>, <span class="string">'k'</span>, <span class="string">'LineWidth'</span>, 1)
0115             annotation(figHandle, <span class="string">'line'</span>, [0.035 0.05], 0.003 * [1 1], <span class="string">'Color'</span>, <span class="string">'k'</span>, <span class="string">'LineWidth'</span>, 1)
0116             
0117             <span class="comment">%standarize tick lengths</span>
0118             axesHandles = findobj(figHandle, <span class="string">'type'</span>, <span class="string">'axes'</span>, <span class="string">'-and'</span>, <span class="string">'tag'</span>, <span class="string">''</span>);
0119             axesPos = cell2mat(get(axesHandles, <span class="string">'position'</span>));
0120             tickLen = 0.01 * max(max(axesPos(:, 3:4), [], 1) .* [w h]);
0121             <span class="keyword">for</span> i = 1:numel(axesHandles)
0122                 maxVal = max(axesPos(i, 3:4) .* [w h]);
0123                 set(axesHandles(i), <span class="string">'TickLen'</span>, [tickLen / maxVal  0.0250]);
0124             <span class="keyword">end</span>
0125             
0126             <span class="comment">%save</span>
0127             print(figHandle, [outDirectory filesep <span class="string">'figure2.pdf'</span>], <span class="string">'-dpdf'</span>, <span class="string">'-rgb'</span>);
0128             close(figHandle);
0129             <span class="keyword">return</span>
0130             
0131             <span class="comment">%% supplemental figures</span>
0132             
0133             <span class="comment">%gene expression</span>
0134             w = 11.4; h = w * (1-0.09)/(1-0.075);
0135             [~, figHandle] = PlotUtil.newAxesHandle();
0136             set(figHandle, <span class="string">'PaperUnits'</span>, <span class="string">'centimeters'</span>);
0137             set(figHandle, <span class="string">'PaperSize'</span>, [w h]);
0138             set(figHandle, <span class="string">'PaperPositionMode'</span>, <span class="string">'manual'</span>);
0139             set(figHandle, <span class="string">'PaperPosition'</span>, [0 0 get(figHandle, <span class="string">'PaperSize'</span>)]);
0140             set(figHandle, <span class="string">'Units'</span>, <span class="string">'normalized'</span>);
0141             figPos = get(figHandle, <span class="string">'Position'</span>);
0142             set(figHandle, <span class="string">'Position'</span>, [figPos(1:2) figPos(3) figPos(3) * h / w]);
0143             
0144             clf(figHandle);
0145             [exptGeneExpr, modelGeneExpr] = Figures34.geneExpression(simBatchDir, selectedSim, figHandle, [0.09 0.075 1-0.09 1-0.075], true);
0146             print(figHandle, [outDirectory filesep <span class="string">'../supplement/figures/rnaExpressionComparison_log.pdf'</span>], <span class="string">'-dpdf'</span>, <span class="string">'-rgb'</span>);
0147             
0148             clf(figHandle);
0149             [exptGeneExpr, modelGeneExpr] = Figures34.geneExpression(simBatchDir, selectedSim, figHandle, [0.082 0.065 1-0.082 1-0.065], false);
0150             print(figHandle, [outDirectory filesep <span class="string">'figureS1a.pdf'</span>], <span class="string">'-dpdf'</span>, <span class="string">'-rgb'</span>);
0151             
0152             close(figHandle);
0153             
0154             <span class="comment">%gene expression ratios</span>
0155             [axesHandle, figHandle] = edu.stanford.covert.cell.sim.util.PlotUtil.newAxesHandle();
0156             Figures34.geneExpressionRatios(exptGeneExpr, modelGeneExpr, axesHandle);
0157             print(figHandle, [outDir <span class="string">'RNAExprRatios.pdf'</span>], <span class="string">'-dpdf'</span>, <span class="string">'-rgb'</span>);
0158             close(figHandle);
0159             
0160             <span class="comment">%dynamics of gene, RNA, protein copy number; reaction flux, growth rate</span>
0161             [~, figHandle] = PlotUtil.newAxesHandle();
0162             Figures34.geneToPhenotype(simBatchDir, selectedSim, outDir, figHandle, [0.1 0.1 0.8 0.8]);
0163             print(figHandle, [outDir <span class="string">'geneToPhenotype.pdf'</span>], <span class="string">'-dpdf'</span>, <span class="string">'-rgb'</span>);
0164             close(figHandle);
0165             
0166             <span class="comment">%metabolic map</span>
0167             [~, figHandle] = PlotUtil.newAxesHandle();
0168             Figures34.metabolicMap(simBatchDir, selectedSim, outDir, figHandle, [0.1 0.2 0.8 0.6], true);
0169             print(figHandle, [outDirectory filesep <span class="string">'figureS1b.pdf'</span>], <span class="string">'-dpdf'</span>, <span class="string">'-rgb'</span>);
0170             close(figHandle);
0171         <span class="keyword">end</span>
0172         
0173         <span class="comment">%chromosome</span>
0174         <a name="_sub2" href="#_subfunctions" class="code">function figure3(simBatchDir, selectedSim, outDirectory)</a>
0175             import edu.stanford.covert.cell.sim.analysis.Figures34;
0176             import edu.stanford.covert.cell.sim.util.PlotUtil;
0177             
0178             outDir = [outDirectory filesep <span class="string">'chromosome'</span> filesep];
0179             
0180             w = 11.4; h = 16.1; <span class="comment">% Cell 1.5 column</span>
0181             h2 = 0.2025;
0182             [~, figHandle] = PlotUtil.newAxesHandle([w h]);
0183             clf(figHandle);
0184             
0185             <span class="comment">%A: DNA-bound protein density</span>
0186             PlotUtil.labelSubFigure(<span class="string">'A'</span>, [0.080 0.997 -0.0286 -0.0208], figHandle);
0187             Figures34.circosPlot(simBatchDir, selectedSim, outDir, figHandle, [0.00 0.66 0.56 0.3]);
0188             
0189             <span class="comment">%B: time to DNA occupancy</span>
0190             PlotUtil.labelSubFigure(<span class="string">'B'</span>, [0.605 0.997 -0.0286 -0.0208], figHandle);
0191             Figures34.chromosomeSampling(simBatchDir, selectedSim, outDir, figHandle, [0.595 0.86 0.40 0.13]);
0192             
0193             <span class="comment">%C: time to RNA expression</span>
0194             PlotUtil.labelSubFigure(<span class="string">'C'</span>, [0.605 0.817 -0.0286 -0.0208], figHandle);
0195             Figures34.rnaSampling(simBatchDir, selectedSim, outDir, figHandle, [0.595 0.68 0.40 0.13]);
0196             
0197             <span class="comment">%D: space-time plot of DNA-bound protein</span>
0198             PlotUtil.labelSubFigure(<span class="string">'D'</span>, [0.080 0.647 -0.0286 -0.0208], figHandle);
0199             Figures34.spaceTimePlot(simBatchDir, selectedSim, outDir, figHandle, [0.06 0.36 0.925 0.28]);
0200             
0201             <span class="comment">%E-G: collisions</span>
0202             PlotUtil.labelSubFigure(<span class="string">'E'</span>, [0.080 0.325 -0.0286 -0.0208], figHandle);
0203             PlotUtil.labelSubFigure(<span class="string">'F'</span>, [0.633 0.325 -0.0286 -0.0208], figHandle);
0204             [posFreq, protDensity] = Figures34.analyzeDNABoundProteinDisplacements(simBatchDir, outDir, figHandle, <span class="keyword">...</span>
0205                 [0.095 0.07 0.94 0.242]);
0206             
0207             <span class="comment">%standarize tick lengths</span>
0208             axesHandles = findobj(figHandle, <span class="string">'type'</span>, <span class="string">'axes'</span>, <span class="string">'-and'</span>, <span class="string">'tag'</span>, <span class="string">''</span>);
0209             axesPos = cell2mat(get(axesHandles, <span class="string">'position'</span>));
0210             tickLen = 0.0075 * max(max(axesPos(:, 3:4), [], 1) .* [w h]);
0211             <span class="keyword">for</span> i = 1:numel(axesHandles)
0212                 maxVal = max(axesPos(i, 3:4) .* [w h]);
0213                 set(axesHandles(i), <span class="string">'TickLen'</span>, [tickLen / maxVal  0.0250]);
0214             <span class="keyword">end</span>
0215             
0216             <span class="comment">%save</span>
0217             print(figHandle, [outDirectory filesep <span class="string">'figure3.pdf'</span>], <span class="string">'-dpdf'</span>, <span class="string">'-rgb'</span>);
0218             close(figHandle);
0219             
0220             <span class="comment">%% supplement</span>
0221             <span class="comment">%fit</span>
0222             [~, figHandle] = edu.stanford.covert.cell.sim.util.PlotUtil.newAxesHandle();
0223             clf(figHandle);
0224             Figures34.replicationInitiationVsReplicationDurationFit(simBatchDir, selectedSim, figHandle, [0.1 0.1 0.8 0.8]);
0225             print(figHandle, [outDir filesep <span class="string">'ReplicationInitiationVsReplicationDurationFit.pdf'</span>], <span class="string">'-dpdf'</span>, <span class="string">'-rgb'</span>);
0226             close(figHandle);
0227             
0228             <span class="comment">%analysis of RNA polymerase exploration</span>
0229             [~, figHandle] = edu.stanford.covert.cell.sim.util.PlotUtil.newAxesHandle();
0230             clf(figHandle);
0231             Figures34.plotRNAPolSampling(simBatchDir, selectedSim, outDir, figHandle, [0.1 0.1 0.8 0.8]);
0232             print(figHandle, [outDir filesep <span class="string">'RNAPolSampling.pdf'</span>], <span class="string">'-dpdf'</span>, <span class="string">'-rgb'</span>);
0233             close(figHandle);
0234             
0235             <span class="comment">%F: scatter plot of density vs collisions</span>
0236             PlotUtil.labelSubFigure(<span class="string">'F'</span>, [0.715 0.80 -0.0286 -0.0690], figHandle);
0237             Figures34.dnaBoundProteinDensityVsDisplacement(simBatchDir, selectedSim, outDir, figHandle, [0.71 0.755-h2 0.29 h2]);
0238             
0239             <span class="comment">%more supplemental material</span>
0240             Figures34.cellShape(outDir, simBatchDir, selectedSim);
0241             Figures34.mrnaSampling(simBatchDir, selectedSim, outDir, figHandle, [0.71 0.755-0.12 0.29 0.12]);
0242         <span class="keyword">end</span>
0243         
0244         <span class="comment">%cell cycle</span>
0245         <a name="_sub3" href="#_subfunctions" class="code">function figure4(simBatchDir, selectedSim, outDirectory)</a>
0246             import edu.stanford.covert.cell.sim.analysis.Figures34;
0247             import edu.stanford.covert.cell.sim.util.PlotUtil;
0248             
0249             outDir = [outDirectory filesep <span class="string">'cellCycle'</span> filesep];
0250             
0251             w = 11.4; h = 18; <span class="comment">% Cell 1.5 column</span>
0252             h2 = 0.2025;
0253             [~, figHandle] = PlotUtil.newAxesHandle([w h]);
0254             clf(figHandle);
0255             
0256             <span class="comment">%F: histograms of cell cycle phase lengths</span>
0257             PlotUtil.labelSubFigure(<span class="string">'F'</span>, [0.420 0.36 -0.0286 -0.0208], figHandle, 8, <span class="string">'normalized'</span>);
0258             Figures34.cellCycleLength(simBatchDir, selectedSim, figHandle, [0.385 0.16 0.27 0.18]);
0259             
0260             <span class="comment">%G: scatter plot of replication vs. replication initiation durations</span>
0261             PlotUtil.labelSubFigure(<span class="string">'G'</span>, [0.750 0.36 -0.0286 -0.0208], figHandle, 8, <span class="string">'normalized'</span>);
0262             Figures34.replicationInitiationVsReplicationDuration(simBatchDir, selectedSim, figHandle, [0.72 0.16 0.27 0.18]);
0263             
0264             <span class="comment">%standarize tick lengths</span>
0265             axesHandles = findobj(figHandle, <span class="string">'type'</span>, <span class="string">'axes'</span>, <span class="string">'-and'</span>, <span class="string">'tag'</span>, <span class="string">''</span>);
0266             axesPos = cell2mat(get(axesHandles, <span class="string">'position'</span>));
0267             tickLen = 0.0075 * max(max(axesPos(:, 3:4), [], 1) .* [w h]);
0268             <span class="keyword">for</span> i = 1:numel(axesHandles)
0269                 maxVal = max(axesPos(i, 3:4) .* [w h]);
0270                 set(axesHandles(i), <span class="string">'TickLen'</span>, [tickLen / maxVal  0.0250]);
0271             <span class="keyword">end</span>
0272             
0273             <span class="comment">%save</span>
0274             print(figHandle, [outDirectory filesep <span class="string">'figure4.pdf'</span>], <span class="string">'-dpdf'</span>, <span class="string">'-rgb'</span>);
0275             close(figHandle);
0276         <span class="keyword">end</span>
0277         
0278         <span class="comment">%energy</span>
0279         <a name="_sub4" href="#_subfunctions" class="code">function figure5(simBatchDir, selectedSim, outDirectory, plotUsageRate)</a>
0280             import edu.stanford.covert.cell.sim.analysis.Figures34;
0281             import edu.stanford.covert.cell.sim.util.PlotUtil;
0282             import edu.stanford.covert.cell.sim.util.SimulationEnsemble;
0283             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
0284             import edu.stanford.covert.cell.sim.util.SummaryLogger;
0285             import edu.stanford.covert.util.ConstantUtil;
0286             
0287             <span class="keyword">if</span> nargin &lt; 4
0288                 plotUsageRate = false;
0289             <span class="keyword">end</span>
0290             
0291             outDir = [outDirectory filesep <span class="string">'energy'</span> filesep];
0292             
0293             <span class="comment">%% get data</span>
0294             [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep <span class="string">'1'</span>]);
0295             c = sim.compartment;
0296             met = sim.state(<span class="string">'Metabolite'</span>);
0297             metab = sim.process(<span class="string">'Metabolism'</span>);
0298             
0299             nSims = SimulationDiskUtil.getNumSimulations(simBatchDir);
0300             
0301             <span class="keyword">if</span> ~exist([outDir <span class="string">'data.mat'</span>], <span class="string">'file'</span>)
0302                 stateNames = {
0303                     <span class="string">'atpUsage'</span> <span class="string">'ATP'</span>
0304                     <span class="string">'gtpUsage'</span> <span class="string">'GTP'</span>
0305                     };
0306                 ensemble = SimulationEnsemble(simBatchDir, stateNames, [], 1:nSims);
0307                 simEndTimes = ensemble.stateData.simulationEndTimes;
0308                 
0309                 totalEnergyUsagePerCell = permute(nansum([
0310                     ensemble.stateData.values(ensemble.getPropertyIndices(<span class="string">'atpUsage'</span>), :, :, :)
0311                     ensemble.stateData.values(ensemble.getPropertyIndices(<span class="string">'gtpUsage'</span>), :, :, :)
0312                     ], 3), [1 4 2 3]) / ConstantUtil.nAvogadro * 1e3;
0313                 
0314                 metIdxs = sub2ind([numel(met.wholeCellModelIDs) numel(c.wholeCellModelIDs)], <span class="keyword">...</span>
0315                     met.getIndexs({<span class="string">'ATP'</span>; <span class="string">'GTP'</span>}), c.cytosolIndexs([1 1])');
0316                 stateNames = {
0317                     <span class="string">'Time'</span>       <span class="string">'values'</span>  <span class="string">':'</span>  <span class="string">':'</span>
0318                     <span class="string">'Geometry'</span>   <span class="string">'volume'</span>  <span class="string">':'</span>  <span class="string">':'</span>
0319                     <span class="string">'Metabolite'</span> <span class="string">'counts'</span>  met.getIndexs({<span class="string">'ATP'</span>; <span class="string">'GTP'</span>; <span class="string">'FAD'</span>; <span class="string">'NAD'</span>; <span class="string">'NADP'</span>; <span class="string">'FADH2'</span>; <span class="string">'NADH'</span>; <span class="string">'NADPH'</span>})  c.cytosolIndexs
0320                     <span class="string">'Metabolite'</span> <span class="string">'processUsages'</span> metIdxs <span class="string">':'</span>
0321                     };
0322                 states = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, <span class="string">'extract'</span>, selectedSim);
0323                 
0324                 simEndTime = states.Time.values(end);
0325                 
0326                 energyCarrierDynamics = full(permute(states.Metabolite.counts, [1 3 2]));
0327                 energyCarrierDynamics(3:5, :) = energyCarrierDynamics(3:5, :) + energyCarrierDynamics(6:8, :);
0328                 energyCarrierDynamics = energyCarrierDynamics(1:5, :) ./ repmat(permute(states.Geometry.volume, [1 3 2]), [5 1]) / ConstantUtil.nAvogadro * 1e3;
0329                 
0330                 processEnergyUsageDynamics = full(states.Metabolite.processUsages);
0331                 
0332                 stateNames = {
0333                     <span class="string">'Metabolite'</span> <span class="string">'processUsages'</span> [metIdxs(1) metIdxs(2)] <span class="string">':'</span>
0334                     };
0335                 processEnergyUsages = zeros(numel(sim.processes), 1);
0336                 <span class="keyword">for</span> i = 1:nSims
0337                     states = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, <span class="string">'extract'</span>, i);
0338                     processEnergyUsages = processEnergyUsages + full(sum(states.Metabolite.processUsages, 3)');
0339                 <span class="keyword">end</span>
0340                 
0341                 clear ensemble states;
0342                                 
0343                 metIdxs = sub2ind([numel(met.wholeCellModelIDs) numel(c.wholeCellModelIDs)], <span class="keyword">...</span>
0344                     met.getIndexs({<span class="string">'ATP'</span>; <span class="string">'CTP'</span>; <span class="string">'GTP'</span>; <span class="string">'UTP'</span>; <span class="string">'AMP'</span>; <span class="string">'CMP'</span>; <span class="string">'GMP'</span>; <span class="string">'UMP'</span>; <span class="string">'FAD'</span>; <span class="string">'NAD'</span>; <span class="string">'NADP'</span>; <span class="string">'FADH2'</span>; <span class="string">'NADH'</span>; <span class="string">'NADPH'</span>}), c.cytosolIndexs(ones(14, 1)));
0345                 stateNames = {
0346                     <span class="string">'MetabolicReaction'</span> <span class="string">'growth'</span> <span class="string">':'</span> <span class="string">':'</span>
0347                     <span class="string">'Metabolite'</span> <span class="string">'counts'</span> met.ntpIndexs' c.cytosolIndexs
0348                     <span class="string">'Metabolite'</span> <span class="string">'processUsages'</span> metIdxs [sim.processIndex(<span class="string">'Metabolism'</span>); sim.processIndex(<span class="string">'Transcription'</span>); sim.processIndex(<span class="string">'RNADecay'</span>)]
0349                     };
0350                 states = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, <span class="string">'extract'</span>, selectedSim);
0351                 growthDynamics = full(permute(states.MetabolicReaction.growth, [1 3 2]));
0352                 energyCarrierSynthesis = full(permute(states.Metabolite.processUsages([1 3 end-5:end], 1, :), [1 3 2]));
0353                 energyCarrierSynthesis(3:5, :) = energyCarrierSynthesis(3:5, :) + energyCarrierSynthesis(6:8, :);
0354                 energyCarrierSynthesis = energyCarrierSynthesis(1:5, :);
0355                 energyCarrierSynthesis(1, :) = energyCarrierSynthesis(1, :) - metab.unaccountedEnergyConsumption * growthDynamics;
0356                 
0357                 ntpDynamics = full(permute(states.Metabolite.counts, [1 3 2]));
0358                 
0359                 transcriptionDynamics = full(permute(sum(states.Metabolite.processUsages(1:4, 2, :), 1), [1 3 2]));
0360                 rnaDecayDynamics = -full(permute(sum(states.Metabolite.processUsages(5:8, 3, :), 1), [1 3 2]));
0361                 
0362                 stateNames = {
0363                     <span class="string">'mass'</span> <span class="string">'Mass'</span>
0364                     <span class="string">'ntps'</span> <span class="string">'NTP'</span>
0365                     <span class="string">'superhelicity'</span> <span class="string">'superhelicity'</span>
0366                     };
0367                 ensemble = SimulationEnsemble(simBatchDir, stateNames, [], 1:nSims);
0368                 ntpCounts = permute(nanmean(<span class="keyword">...</span>
0369                     ensemble.stateData.values(ensemble.getPropertyIndices(<span class="string">'ntps'</span>), :, :, :), <span class="keyword">...</span>
0370                     3), [1 4 2 3]) / ConstantUtil.nAvogadro * 1e3;
0371                 superhelicity = permute(nanmean(<span class="keyword">...</span>
0372                     ensemble.stateData.values(ensemble.getPropertyIndices(<span class="string">'superhelicity'</span>), :, :, :), <span class="keyword">...</span>
0373                     3), [1 4 2 3]);
0374                 
0375                 stateNames = {
0376                     <span class="string">'Mass'</span> <span class="string">'rnaWt'</span>
0377                     <span class="string">'Mass'</span> <span class="string">'proteinWt'</span>
0378                     <span class="string">'RNAPolymerase'</span> <span class="string">'nActive'</span>
0379                     <span class="string">'Ribosome'</span> <span class="string">'nStalled'</span>
0380                     <span class="string">'Geometry'</span> <span class="string">'volume'</span>
0381                     };
0382                 states = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, <span class="string">'extract'</span>, 1:nSims);
0383                 rnaWt = permute(nanmean(full(states.Mass.rnaWt(:, c.cytosolIndexs, :, :)), 3), [1 4 3 2]);
0384                 proteinWt = permute(nanmean(full(states.Mass.proteinWt(:, c.cytosolIndexs, :, :)), 3), [1 4 3 2]);
0385                 activeRNAPols =  permute(nanmean(full(states.RNAPolymerase.nActive), 3), [1 4 3 2]);
0386                 stalledRibs = permute(nanmean(full(states.Ribosome.nStalled), 3), [1 4 3 2]);
0387                 
0388                 ntpConcs = permute(nanmean(<span class="keyword">...</span>
0389                     ensemble.stateData.values(ensemble.getPropertyIndices(<span class="string">'ntps'</span>), :, 2:<span class="keyword">end</span>, :) ./ <span class="keyword">...</span>
0390                     states.Geometry.volume, <span class="keyword">...</span>
0391                     3), [1 4 2 3]) / ConstantUtil.nAvogadro * 1e3;
0392                 
0393                 simStats = SummaryLogger.getSimulationStatistics(simBatchDir);
0394                 repInitDuration = simStats(:, SummaryLogger.SIM_STATUS_INDEX_REP_INIT_TIME);
0395                 repDuration = diff(simStats(:, [SummaryLogger.SIM_STATUS_INDEX_REP_INIT_TIME SummaryLogger.SIM_STATUS_INDEX_REP_TIME]), 1, 2);
0396                 
0397                 save([outDir <span class="string">'data.mat'</span>], <span class="keyword">...</span>
0398                     <span class="string">'simEndTimes'</span>, <span class="string">'totalEnergyUsagePerCell'</span>, <span class="string">'energyCarrierDynamics'</span>, <span class="keyword">...</span>
0399                     <span class="string">'processEnergyUsageDynamics'</span>, <span class="string">'processEnergyUsages'</span>, <span class="keyword">...</span>
0400                     <span class="string">'growthDynamics'</span>, <span class="string">'energyCarrierSynthesis'</span>, <span class="string">'ntpDynamics'</span>, <span class="string">'transcriptionDynamics'</span>, <span class="string">'rnaDecayDynamics'</span>, <span class="keyword">...</span>
0401                     <span class="string">'ntpCounts'</span>, <span class="string">'superhelicity'</span>, <span class="keyword">...</span>
0402                     <span class="string">'rnaWt'</span>, <span class="string">'proteinWt'</span>, <span class="string">'activeRNAPols'</span>, <span class="string">'stalledRibs'</span>, <span class="string">'ntpConcs'</span>, <span class="keyword">...</span>
0403                     <span class="string">'repInitDuration'</span>, <span class="string">'repDuration'</span>);
0404             <span class="keyword">else</span>
0405                 load([outDir <span class="string">'data.mat'</span>]);
0406             <span class="keyword">end</span>            
0407             
0408             <span class="comment">%% figure</span>
0409             figW = 8.5; figH = 10; <span class="comment">% Cell 1 column</span>
0410             [~, figHandle] = PlotUtil.newAxesHandle([figW figH]);
0411             clf(figHandle);
0412             
0413             leftColX = 0.085;
0414             leftColW = 0.42;
0415             leftColLabelX = 0.105;
0416             
0417             window = 300;
0418             
0419             <span class="comment">%% A: dynamics of energy carrier synthesis</span>
0420             <span class="keyword">for</span> i = 1:size(energyCarrierSynthesis, 1)
0421                 energyCarrierSynthesis(i, :) = conv(energyCarrierSynthesis(i, :), ones(window, 1) / window, <span class="string">'same'</span>);
0422             <span class="keyword">end</span>
0423             
0424             PlotUtil.labelSubFigure(<span class="string">'A'</span>, [leftColLabelX 0.993 -0.0286 -0.0208], figHandle, 8, <span class="string">'normalized'</span>);
0425             axesHandle = subplot(<span class="string">'Position'</span>, [leftColX 0.745 leftColW 0.23], <span class="string">'Parent'</span>, figHandle);
0426             h = plot(axesHandle, (window:simEndTimes(selectedSim)-window+1) / 3600, -energyCarrierSynthesis(:, window:end-window+1) / ConstantUtil.nAvogadro * 1e21);
0427             set(h(1), <span class="string">'Color'</span>, <span class="string">'b'</span>);
0428             set(h(2), <span class="string">'Color'</span>, <span class="string">'g'</span>);
0429             set(h(3), <span class="string">'Color'</span>, <span class="string">'r'</span>)
0430             set(h(4), <span class="string">'Color'</span>, <span class="string">'c'</span>)
0431             set(h(5), <span class="string">'Color'</span>, [1 0.5 0])
0432             
0433             text(4.1, 20.0, <span class="string">'ATP'</span>,      <span class="string">'FontSize'</span>, 6, <span class="string">'Color'</span>, get(h(1), <span class="string">'Color'</span>), <span class="string">'HorizontalAlign'</span>, <span class="string">'center'</span>);
0434             text(4.1, 0.50, <span class="string">'GTP'</span>,      <span class="string">'FontSize'</span>, 6, <span class="string">'Color'</span>, get(h(2), <span class="string">'Color'</span>), <span class="string">'HorizontalAlign'</span>, <span class="string">'center'</span>);
0435             text(4.1, 2e-5, <span class="string">'FAD(H_2)'</span>, <span class="string">'FontSize'</span>, 6, <span class="string">'Color'</span>, get(h(3), <span class="string">'Color'</span>), <span class="string">'HorizontalAlign'</span>, <span class="string">'center'</span>);
0436             text(2.5, 5e-3, <span class="string">'NAD(H)'</span>,   <span class="string">'FontSize'</span>, 6, <span class="string">'Color'</span>, get(h(4), <span class="string">'Color'</span>), <span class="string">'HorizontalAlign'</span>, <span class="string">'center'</span>);
0437             text(5.5, 5e-3, <span class="string">'NADP(H)'</span>,  <span class="string">'FontSize'</span>, 6, <span class="string">'Color'</span>, get(h(5), <span class="string">'Color'</span>), <span class="string">'HorizontalAlign'</span>, <span class="string">'center'</span>);
0438             
0439             xlim(axesHandle, [0 simEndTimes(selectedSim) / 3600]);
0440             xlabel(axesHandle, <span class="string">'Time (h)'</span>, <span class="string">'FontSize'</span>, 7);
0441             ylabel(axesHandle, <span class="string">'Synthesis (10^{-21} mol s^{-1})'</span>, <span class="string">'FontSize'</span>, 7);
0442             set(axesHandle, <span class="string">'FontSize'</span>, 5, <span class="string">'Box'</span>, <span class="string">'off'</span>, <span class="string">'tickdir'</span>, <span class="string">'out'</span>, <span class="keyword">...</span>
0443                 <span class="string">'XTick'</span>, 0:4:8, <span class="string">'yscale'</span>, <span class="string">'log'</span>, <span class="string">'ytick'</span>, 10.^(-6:2:2))
0444             
0445             tick2text(axesHandle, <span class="string">'axis'</span>, <span class="string">'xy'</span>, <span class="string">'xtickoffset'</span>, 1e-7, <span class="string">'ytickoffset'</span>, 0.018);
0446             xTicks = getappdata(axesHandle, <span class="string">'XTickText'</span>);
0447             yTicks = getappdata(axesHandle, <span class="string">'YTickText'</span>);
0448             <span class="keyword">for</span> i = 1:numel(xTicks)
0449                 xTickPos = get(xTicks(i), <span class="string">'Position'</span>);
0450                 set(xTicks(i), <span class="string">'Position'</span>, [xTickPos(1) 3e-7 xTickPos(3:end)]);
0451             <span class="keyword">end</span>
0452             set(xTicks, <span class="string">'FontSize'</span>, 5);
0453             set(yTicks, <span class="string">'FontSize'</span>, 5, <span class="string">'HorizontalAlign'</span>, <span class="string">'right'</span>);
0454             set(yTicks(1), <span class="string">'String'</span>, <span class="string">'10^{-6}'</span>)
0455             set(yTicks(2), <span class="string">'String'</span>, <span class="string">'10^{-4}'</span>)
0456             set(yTicks(3), <span class="string">'String'</span>, <span class="string">'10^{-2}'</span>)
0457             set(yTicks(4), <span class="string">'String'</span>, <span class="string">'10^{0}'</span>)
0458             set(yTicks(5), <span class="string">'String'</span>, <span class="string">'10^{2}'</span>)
0459             
0460             xLabelPos = get(get(axesHandle, <span class="string">'xlabel'</span>), <span class="string">'position'</span>);
0461             yLabelPos = get(get(axesHandle, <span class="string">'ylabel'</span>), <span class="string">'position'</span>);
0462             set(get(axesHandle, <span class="string">'xlabel'</span>), <span class="string">'position'</span>, [xLabelPos(1) 1e-7 xLabelPos(end)]);
0463             set(get(axesHandle, <span class="string">'ylabel'</span>), <span class="string">'position'</span>, [-1.0 yLabelPos(2:end)]);
0464             
0465             <span class="comment">%% C energy dynamics of processes</span>
0466             [~, pIdxs] = sort(processEnergyUsages);
0467             pIdxs = pIdxs(end:-1:end-14);
0468             
0469             <span class="keyword">for</span> i = 1:size(processEnergyUsageDynamics, 1)
0470                 <span class="keyword">for</span> j = 1:size(processEnergyUsageDynamics, 2)
0471                     processEnergyUsageDynamics(i, j, :) = <span class="keyword">...</span>
0472                         conv(squeeze(processEnergyUsageDynamics(i, j, :)), ones(window, 1) / window, <span class="string">'same'</span>);
0473                 <span class="keyword">end</span>
0474             <span class="keyword">end</span>
0475             processEnergyUsageDynamics = processEnergyUsageDynamics / ConstantUtil.nAvogadro * 1e24;
0476             
0477             totEnergy = sum(max(0, processEnergyUsages)) + sim.process(<span class="string">'Metabolism'</span>).unaccountedEnergyConsumption * nSims;
0478             
0479             nCols = 1;
0480             nRows = ceil(numel(pIdxs) / nCols);
0481             
0482             W = 0.27;
0483             H = 0.98;
0484             
0485             x = 0.65;
0486             y = 0.065;
0487             xMargin = 0.05;
0488             yMargin = 0.03;
0489             xAxisOffset = 0.01;
0490             yAxisOffset = xAxisOffset * figW / figH;
0491             w = (W - (nCols - 1) * xMargin) / nCols;
0492             h = (H - y - (nRows - 1) * yMargin) / nRows;
0493             
0494             PlotUtil.labelSubFigure(<span class="string">'C'</span>, [x-0.01 0.993 -0.0286 -0.0208], figHandle, 8, <span class="string">'normalized'</span>);
0495             
0496             <span class="keyword">for</span> i = 1:numel(pIdxs)
0497                 row = ceil(i / nCols);
0498                 col = mod(i-1, nCols) + 1;
0499                 axesHandle = subplot(<span class="string">'Position'</span>, [x+(col-1)*(w+xMargin)  y+(nRows-row)*(h+yMargin) w  h], <span class="string">'Parent'</span>, figHandle);
0500                 handle = plot(axesHandle, (window:simEndTimes(selectedSim)-window+1) / 3600, <span class="keyword">...</span>
0501                     permute(processEnergyUsageDynamics(:, pIdxs(i), window:end-window+1), [1 3 2]));
0502                 set(handle(1), <span class="string">'Color'</span>, <span class="string">'b'</span>);
0503                 set(handle(2), <span class="string">'Color'</span>, <span class="string">'g'</span>);
0504                 set(axesHandle, <span class="string">'FontSize'</span>, 5, <span class="string">'visible'</span>, <span class="string">'off'</span>, <span class="string">'yscale'</span>, <span class="string">'linear'</span>);
0505                 xlim(axesHandle, [0 simEndTimes(selectedSim) / 3600]);
0506                 ylim(axesHandle, [0 max(max(processEnergyUsageDynamics(:, pIdxs(i), :)))]);
0507                 annotation(figHandle, <span class="string">'textbox'</span>, [x+(col-1)*(w+xMargin)  y+(nRows-row)*(h+yMargin)+h-0.02 w 1e-6], <span class="keyword">...</span>
0508                     <span class="string">'String'</span>, sim.processMetadata.names{pIdxs(i)}, <span class="keyword">...</span>
0509                     <span class="string">'EdgeColor'</span>, <span class="string">'none'</span>, <span class="string">'HorizontalAlignment'</span>, <span class="string">'center'</span>, <span class="string">'VerticalAlignment'</span>, <span class="string">'bottom'</span>, <span class="keyword">...</span>
0510                     <span class="string">'FontSize'</span>, 6);
0511                 
0512                 <span class="comment">%y axis</span>
0513                 yAxesHandle = subplot(<span class="string">'Position'</span>, [x-xAxisOffset y+(nRows-row)*(h+yMargin) 1e-6 h], <span class="string">'Parent'</span>, figHandle);
0514                 ylim(yAxesHandle, ylim(axesHandle));
0515                 set(yAxesHandle, <span class="string">'FontSize'</span>, 5, <span class="string">'yscale'</span>, <span class="string">'linear'</span>, <span class="string">'box'</span>, <span class="string">'off'</span>, <span class="string">'tickdir'</span>, <span class="string">'out'</span>);
0516                 <span class="keyword">if</span> row == ceil(nRows / 2) &amp;&amp; col == 1
0517                     ylabel(yAxesHandle, {<span class="string">'NTP Use (10^{-24} mol s^{-1})'</span>}, <span class="string">'FontSize'</span>, 7);
0518                     yLabelPos = get(get(yAxesHandle, <span class="string">'ylabel'</span>), <span class="string">'position'</span>);
0519                     set(get(yAxesHandle, <span class="string">'ylabel'</span>), <span class="string">'position'</span>, [-140 yLabelPos(2:end)]);
0520                 <span class="keyword">end</span>
0521                 
0522                 tick2text(yAxesHandle, <span class="string">'axis'</span>, <span class="string">'y'</span>, <span class="string">'ytickoffset'</span>, 20);
0523                 yTicks = getappdata(yAxesHandle, <span class="string">'YTickText'</span>);
0524                 set(yTicks, <span class="string">'FontSize'</span>, 5, <span class="string">'HorizontalAlign'</span>, <span class="string">'right'</span>);
0525                 
0526                 delete(yTicks(2:end-1));
0527                 
0528                 <span class="comment">%pie</span>
0529                 val = processEnergyUsages(pIdxs(i)) / totEnergy;
0530                 
0531                 scale = 1.4;
0532                 axesHandle = subplot(<span class="string">'Position'</span>, [1-scale*h*figH/figW-0.005  y+(nRows-row)*(h+yMargin)+h/2-scale*h/2  scale*h*figH/figW  scale*h], <span class="string">'Parent'</span>, figHandle);                                
0533                 pieHandle = pie(axesHandle, 100 * [val 1-val]);
0534                 set(pieHandle(1), <span class="string">'EdgeColor'</span>, <span class="string">'k'</span>, <span class="string">'FaceColor'</span>, <span class="string">'r'</span>);
0535                 set(pieHandle(3), <span class="string">'EdgeColor'</span>, <span class="string">'k'</span>, <span class="string">'FaceColor'</span>, 0.75 * [1 1 1]);
0536                 set(pieHandle(2:2:end), <span class="string">'visible'</span>, <span class="string">'off'</span>);
0537                 set(axesHandle, <span class="string">'visible'</span>, <span class="string">'off'</span>, <span class="string">'FontSize'</span>, 5);
0538                 axis(axesHandle, <span class="string">'equal'</span>);
0539                 xlim(axesHandle, [-1.05 1.05])
0540                 ylim(axesHandle, [-1.05 1.05])
0541             <span class="keyword">end</span>
0542             
0543             <span class="comment">%x axis</span>
0544             xAxesHandle = subplot(<span class="string">'Position'</span>, [x y-yAxisOffset w 1e-6], <span class="string">'Parent'</span>, figHandle);
0545             xlim(xAxesHandle, [0 simEndTimes(selectedSim) / 3600]);
0546             xlabel(xAxesHandle, <span class="string">'Time (h)'</span>, <span class="string">'FontSize'</span>, 7);
0547             set(xAxesHandle, <span class="string">'FontSize'</span>, 5, <span class="string">'XTick'</span>, 0:4:8, <span class="string">'box'</span>, <span class="string">'off'</span>, <span class="string">'tickdir'</span>, <span class="string">'out'</span>);
0548             
0549             tick2text(xAxesHandle, <span class="string">'axis'</span>, <span class="string">'x'</span>, <span class="string">'xtickoffset'</span>, 50);
0550             xTicks = getappdata(xAxesHandle, <span class="string">'XTickText'</span>);
0551             set(xTicks, <span class="string">'FontSize'</span>, 5);
0552             
0553             xLabelPos = get(get(xAxesHandle, <span class="string">'xlabel'</span>), <span class="string">'position'</span>);
0554             set(get(xAxesHandle, <span class="string">'xlabel'</span>), <span class="string">'position'</span>, [xLabelPos(1) -100 xLabelPos(3:end)]);
0555             
0556             <span class="comment">%legend</span>
0557             annotation(figHandle, <span class="string">'ellipse'</span>, [1-0.06  0.015-0.01*figW/figH+0.02  0.01  0.01*figW/figH], <span class="keyword">...</span>
0558                 <span class="string">'EdgeColor'</span>, <span class="string">'none'</span>, <span class="string">'FaceColor'</span>, <span class="string">'b'</span>);
0559             annotation(figHandle, <span class="string">'ellipse'</span>, [1-0.06  0.015-0.01*figW/figH  0.01  0.01*figW/figH], <span class="keyword">...</span>
0560                 <span class="string">'EdgeColor'</span>, <span class="string">'none'</span>, <span class="string">'FaceColor'</span>, <span class="string">'g'</span>);
0561             annotation(figHandle, <span class="string">'textbox'</span>, [1 - 0.065, 0.005+0.02 1e-6 1e-6], <span class="string">'String'</span>, <span class="string">'ATP'</span>, <span class="keyword">...</span>
0562                 <span class="string">'FontSize'</span>, 5, <span class="string">'HorizontalAlignment'</span>, <span class="string">'left'</span>, <span class="string">'VerticalAlignment'</span>, <span class="string">'middle'</span>, <span class="string">'EdgeColor'</span>, <span class="string">'none'</span>);
0563             annotation(figHandle, <span class="string">'textbox'</span>, [1 - 0.065, 0.005 1e-6 1e-6], <span class="string">'String'</span>, <span class="string">'GTP'</span>, <span class="keyword">...</span>
0564                 <span class="string">'FontSize'</span>, 5, <span class="string">'HorizontalAlignment'</span>, <span class="string">'left'</span>, <span class="string">'VerticalAlignment'</span>, <span class="string">'middle'</span>, <span class="string">'EdgeColor'</span>, <span class="string">'none'</span>);
0565             
0566             <span class="comment">%% B: total energy usage vs cell cycle length</span>
0567             PlotUtil.labelSubFigure(<span class="string">'B'</span>, [leftColLabelX 0.673 -0.0286 -0.0208], figHandle, 8, <span class="string">'normalized'</span>);
0568             axesHandle = subplot(<span class="string">'Position'</span>, [leftColX 0.42 leftColW 0.23], <span class="string">'Parent'</span>, figHandle);
0569             
0570             hold(axesHandle, <span class="string">'on'</span>);
0571             <span class="keyword">if</span> plotUsageRate
0572                 ydata = totalEnergyUsagePerCell * 1e15 ./ simEndTimes(:, [1 1])' * 3600;
0573                 ylim(axesHandle, [4.8 10.8]);
0574                 ylabel(axesHandle, <span class="string">'NTP Use (10^{-18} mol h^{-1})'</span>, <span class="string">'FontSize'</span>, 7);
0575                 set(axesHandle, <span class="string">'YTick'</span>, 6:2:10);
0576                 
0577                 plot(axesHandle, 8, 10.5, <span class="string">'b.'</span>, <span class="string">'MarkerSize'</span>, 10);
0578                 plot(axesHandle, 8, 9.8, <span class="string">'g.'</span>, <span class="string">'MarkerSize'</span>, 10);
0579                 text(8.2, 10.5, <span class="string">'ATP'</span>, <span class="string">'FontSize'</span>, 6, <span class="string">'VerticalAlignment'</span>, <span class="string">'middle'</span>);
0580                 text(8.2, 9.8, <span class="string">'GTP'</span>, <span class="string">'FontSize'</span>, 6, <span class="string">'VerticalAlignment'</span>, <span class="string">'middle'</span>);
0581             <span class="keyword">else</span>
0582                 ydata = totalEnergyUsagePerCell * 1e15;
0583                 ylim(axesHandle, [50 max(totalEnergyUsagePerCell(:)) * 1e15]);
0584                 ylabel(axesHandle, <span class="string">'Tot NTP Use (10^{-18} mol)'</span>, <span class="string">'FontSize'</span>, 7);
0585                 set(axesHandle, <span class="string">'YTick'</span>, 50:25:125);
0586                 
0587                 plot(axesHandle, 8, 125, <span class="string">'b.'</span>, <span class="string">'MarkerSize'</span>, 10);
0588                 plot(axesHandle, 8, 115, <span class="string">'g.'</span>, <span class="string">'MarkerSize'</span>, 10);
0589                 text(8.2, 125, <span class="string">'ATP'</span>, <span class="string">'FontSize'</span>, 6, <span class="string">'VerticalAlignment'</span>, <span class="string">'middle'</span>);
0590                 text(8.2, 115, <span class="string">'GTP'</span>, <span class="string">'FontSize'</span>, 6, <span class="string">'VerticalAlignment'</span>, <span class="string">'middle'</span>);
0591             <span class="keyword">end</span>
0592             h = plot(axesHandle, simEndTimes / 3600, ydata, <span class="string">'.'</span>);
0593             set(h(1), <span class="string">'Color'</span>, <span class="string">'b'</span>);
0594             set(h(2), <span class="string">'Color'</span>, <span class="string">'g'</span>);
0595             xlim(axesHandle, [7.7 14]);
0596             xlabel(axesHandle, <span class="string">'Cell Cycle Length (h)'</span>, <span class="string">'FontSize'</span>, 7);
0597             set(axesHandle, <span class="string">'FontSize'</span>, 5, <span class="string">'Box'</span>, <span class="string">'off'</span>, <span class="string">'TickDir'</span>, <span class="string">'out'</span>, <span class="keyword">...</span>
0598                 <span class="string">'XTick'</span>, 8:2:14)
0599             
0600             tick2text(axesHandle, <span class="string">'axis'</span>, <span class="string">'xy'</span>, <span class="string">'xtickoffset'</span>, 0.06, <span class="string">'ytickoffset'</span>, 0.015);
0601             xTicks = getappdata(axesHandle, <span class="string">'XTickText'</span>);
0602             yTicks = getappdata(axesHandle, <span class="string">'YTickText'</span>);
0603             set(xTicks, <span class="string">'FontSize'</span>, 5);
0604             set(yTicks, <span class="string">'FontSize'</span>, 5, <span class="string">'HorizontalAlign'</span>, <span class="string">'right'</span>);
0605             
0606             xLabelPos = get(get(axesHandle, <span class="string">'xlabel'</span>), <span class="string">'position'</span>);
0607             yLabelPos = get(get(axesHandle, <span class="string">'ylabel'</span>), <span class="string">'position'</span>);
0608             <span class="keyword">if</span> plotUsageRate
0609                 set(get(axesHandle, <span class="string">'xlabel'</span>), <span class="string">'position'</span>, [xLabelPos(1) 4 xLabelPos(end)]);
0610                 set(get(axesHandle, <span class="string">'ylabel'</span>), <span class="string">'position'</span>, [7.1 yLabelPos(2:end)]);
0611             <span class="keyword">else</span>
0612                 set(get(axesHandle, <span class="string">'xlabel'</span>), <span class="string">'position'</span>, [xLabelPos(1) 41 xLabelPos(end)]);
0613                 set(get(axesHandle, <span class="string">'ylabel'</span>), <span class="string">'position'</span>, [7.02 yLabelPos(2:end)]);
0614             <span class="keyword">end</span>
0615             
0616             <span class="comment">%% D: energy allocation among processes</span>
0617             processNames = sim.processMetadata.names;
0618             processNames(sim.processIndex(<span class="string">'DNASupercoiling'</span>)) = {<span class="string">'Supercoil'</span>};
0619             processNames(sim.processIndex(<span class="string">'ProteinDecay'</span>)) = {<span class="string">'Prot Dcy'</span>};
0620             processNames(sim.processIndex(<span class="string">'tRNAAminoacylation'</span>)) = {<span class="string">'tRNA Acyl'</span>};
0621             processNames(sim.processIndex(<span class="string">'Transcription'</span>)) = {<span class="string">'Transcription'</span>};
0622             processNames(sim.processIndex(<span class="string">'Translation'</span>)) = {<span class="string">'Translation'</span>};
0623             processNames(sim.processIndex(<span class="string">'Replication'</span>)) = {<span class="string">'DNA Rep'</span>};
0624             
0625             PlotUtil.labelSubFigure(<span class="string">'D'</span>, [leftColLabelX 0.32 -0.0286 -0.0208], figHandle, 8, <span class="string">'normalized'</span>);
0626             axesHandle = subplot(<span class="string">'Position'</span>, [leftColLabelX-0.02 -0.085 leftColW 0.50], <span class="string">'Parent'</span>, figHandle);
0627             [~, pIdxs] = sort(processEnergyUsages);
0628             pIdxs = pIdxs(end:-1:end-2);
0629             vals = [
0630                 sum(max(0, processEnergyUsages(setdiff(1:<span class="keyword">end</span>, pIdxs))))
0631                 sim.process(<span class="string">'Metabolism'</span>).unaccountedEnergyConsumption * nSims
0632                 processEnergyUsages(pIdxs)                
0633                 ];
0634             totEnergy = sum(vals);
0635             vals = vals / totEnergy * 100;
0636             labels = [<span class="string">'Other'</span>; <span class="string">'Unaccounted'</span>; processNames(pIdxs)];
0637             
0638             colors = [                
0639                 0 0 1
0640                 1 0 0
0641                 0 1 0
0642                 0 1 1
0643                 1 0.5 0
0644                 ];
0645             explode = [1 0.5 0.5 0.7 0.55];
0646             shiftX = [0.25 0 -0.13 0.05 0];
0647             shiftY = [0 0 0 0 0];
0648             <span class="keyword">for</span> i = 1:numel(vals)
0649                 ptch = patch(<span class="keyword">...</span>
0650                     [0 cos((linspace(sum(vals(1:i-1)), sum(vals(1:i)), 50)-vals(1)/2)/100 * 2 * pi)], <span class="keyword">...</span>
0651                     [0 sin((linspace(sum(vals(1:i-1)), sum(vals(1:i)), 50)-vals(1)/2)/100 * 2 * pi)], <span class="keyword">...</span>
0652                     1, <span class="keyword">...</span>
0653                     <span class="string">'Parent'</span>, axesHandle, <span class="keyword">...</span>
0654                     <span class="string">'EdgeColor'</span>, <span class="string">'k'</span>, <span class="keyword">...</span>
0655                     <span class="string">'FaceColor'</span>, colors(i, :), <span class="keyword">...</span>
0656                     <span class="string">'FaceAlpha'</span>, 1, <span class="string">'EdgeAlpha'</span>, 1);
0657             <span class="keyword">end</span>
0658             <span class="keyword">for</span> i = 1:numel(vals)
0659                 text(<span class="keyword">...</span>
0660                     explode(i) * cos((sum(vals(1:i-1))+vals(i)/2-vals(1)/2)/100 * 2 * pi) + shiftX(i), <span class="keyword">...</span>
0661                     explode(i) * sin((sum(vals(1:i-1))+vals(i)/2-vals(1)/2)/100 * 2 * pi) + shiftY(i) + 0.07, <span class="keyword">...</span>
0662                     labels{i}, <span class="string">'FontSize'</span>, 6, <span class="keyword">...</span>
0663                     <span class="string">'HorizontalAlign'</span>, <span class="string">'center'</span>, <span class="string">'VerticalAlign'</span>, <span class="string">'middle'</span>);
0664                 text(<span class="keyword">...</span>
0665                     explode(i) * cos((sum(vals(1:i-1))+vals(i)/2-vals(1)/2)/100 * 2 * pi) + shiftX(i), <span class="keyword">...</span>
0666                     explode(i) * sin((sum(vals(1:i-1))+vals(i)/2-vals(1)/2)/100 * 2 * pi) + shiftY(i) - 0.07, <span class="keyword">...</span>
0667                     sprintf(<span class="string">'(%0.1f%%)'</span>, vals(i)), <span class="string">'FontSize'</span>, 6, <span class="keyword">...</span>
0668                     <span class="string">'HorizontalAlign'</span>, <span class="string">'center'</span>, <span class="string">'VerticalAlign'</span>, <span class="string">'middle'</span>);
0669             <span class="keyword">end</span>
0670             
0671             xlim(axesHandle, [-1.05 1.05]);
0672             ylim(axesHandle, [-1.05 1.05]);
0673             axis(axesHandle, <span class="string">'equal'</span>);
0674             set(axesHandle, <span class="string">'FontSize'</span>, 5, <span class="string">'visible'</span>, <span class="string">'off'</span>);
0675             
0676             <span class="comment">%% standarize tick lengths</span>
0677             axesHandles = findobj(figHandle, <span class="string">'type'</span>, <span class="string">'axes'</span>, <span class="string">'-and'</span>, <span class="string">'tag'</span>, <span class="string">''</span>);
0678             axesPos = cell2mat(get(axesHandles, <span class="string">'position'</span>));
0679             tickLen = 0.01 * max(max(axesPos(:, 3:4), [], 1) .* [figW figH]);
0680             <span class="keyword">for</span> i = 1:numel(axesHandles)
0681                 maxVal = max(axesPos(i, 3:4) .* [figW figH]);
0682                 set(axesHandles(i), <span class="string">'TickLen'</span>, [tickLen / maxVal  0.0250]);
0683             <span class="keyword">end</span>
0684             
0685             <span class="comment">%% save</span>
0686             print(figHandle, [outDirectory filesep <span class="string">'figure5.pdf'</span>], <span class="string">'-dpdf'</span>, <span class="string">'-rgb'</span>);
0687             print(figHandle, [outDirectory filesep <span class="string">'figure5.eps'</span>], <span class="string">'-deps'</span>, <span class="string">'-rgb'</span>);
0688             close(figHandle);
0689             
0690             <span class="comment">%% supplemental figures</span>
0691             
0692             <span class="comment">%energy carrier production</span>
0693             <span class="keyword">for</span> i = 1:size(energyCarrierSynthesis, 1)
0694                 energyCarrierSynthesis(i, :) = conv(energyCarrierSynthesis(i, :), ones(window, 1) / window, <span class="string">'same'</span>);
0695             <span class="keyword">end</span>
0696             [axesHandle, figHandle] = PlotUtil.newAxesHandle([8.5 8.5]);            
0697             h = plot(axesHandle, (window:simEndTimes(selectedSim)-window+1) / 3600, -energyCarrierSynthesis(:, window:end-window+1));
0698             xlim(axesHandle, [0 simEndTimes(selectedSim) / 3600]);
0699             xlabel(axesHandle, <span class="string">'Time (h)'</span>, <span class="string">'FontSize'</span>, 7);
0700             ylabel(axesHandle, <span class="string">'Synthesis (molecule s^{-1})'</span>, <span class="string">'FontSize'</span>, 7);
0701             set(axesHandle, <span class="string">'box'</span>, <span class="string">'off'</span>, <span class="string">'tickdir'</span>, <span class="string">'out'</span>, <span class="string">'FontSize'</span>, 5, <span class="string">'yscale'</span>, <span class="string">'log'</span>, <span class="string">'XTick'</span>, 0:2:8, <span class="string">'YTick'</span>, 10.^(-2:2:4));
0702             legend(h, {<span class="string">'ATP'</span>; <span class="string">'GTP'</span>; <span class="string">'FAD(H_2)'</span>; <span class="string">'NAD(H)'</span>; <span class="string">'NADP(H)'</span>}, <span class="string">'Location'</span>, <span class="string">'NorthWest'</span>)
0703             saveas(figHandle, [outDir <span class="string">'energyCarrierSynthesis.pdf'</span>]);
0704             close(figHandle);            
0705             
0706             <span class="comment">%RNA decay vs transcription dynamics</span>
0707             [axesHandle, figHandle] = PlotUtil.newAxesHandle([8.5 4]);
0708             clf(figHandle);
0709             
0710             axesHandle = subplot(<span class="string">'Position'</span>, [0.11 0.14 0.36 0.85], <span class="string">'Parent'</span>, figHandle);
0711             x = -15:15;
0712             y = xcorr(transcriptionDynamics, ntpDynamics, x(end), <span class="string">'coeff'</span>);
0713             plot(axesHandle, x, y);
0714             [~, idx] = max(y);
0715             xlim(axesHandle, [-15 15]);
0716             ylim(axesHandle, [min(y) max(y)])
0717             line(x(idx) * [1 1], ylim(axesHandle), <span class="string">'Color'</span>, <span class="string">'r'</span>, <span class="string">'LineStyle'</span>, <span class="string">':'</span>, <span class="string">'Parent'</span>, axesHandle)
0718             text(2.75, min(y) + 0.95 * range(y), sprintf(<span class="string">'%d s'</span>, x(idx)), <span class="string">'Color'</span>, <span class="string">'r'</span>, <span class="string">'FontSize'</span>, 6, <span class="string">'Parent'</span>, axesHandle);
0719             xlabel(axesHandle, <span class="string">'Lag (s)'</span>, <span class="string">'FontSize'</span>, 7);
0720             ylabel(axesHandle, {<span class="string">'NTP-Transcription X-Corr'</span>}, <span class="string">'FontSize'</span>, 7);
0721             set(axesHandle, <span class="string">'box'</span>, <span class="string">'off'</span>, <span class="string">'tickdir'</span>, <span class="string">'out'</span>, <span class="string">'FontSize'</span>, 5, <span class="string">'XTick'</span>, -10:10:10, <span class="string">'YTick'</span>, 0.600:0.002:0.606);
0722             tick2text(axesHandle, <span class="string">'axis'</span>, <span class="string">'xy'</span>, <span class="string">'xtickoffset'</span>, 0.04, <span class="string">'ytickoffset'</span>, 0.03);
0723             set(getappdata(axesHandle, <span class="string">'XTickText'</span>), <span class="string">'FontSize'</span>, 5);
0724             set(getappdata(axesHandle, <span class="string">'YTickText'</span>), <span class="string">'FontSize'</span>, 5, <span class="string">'HorizontalAlign'</span>, <span class="string">'right'</span>);
0725             set(get(axesHandle, <span class="string">'ylabel'</span>), <span class="string">'Position'</span>, [-21 mean(ylim(axesHandle)) 0]);
0726             
0727             axesHandle = subplot(<span class="string">'Position'</span>, [0.64 0.14 0.36 0.85], <span class="string">'Parent'</span>, figHandle);
0728             x = -15:15;
0729             y = xcorr(transcriptionDynamics, rnaDecayDynamics, x(end), <span class="string">'coeff'</span>);
0730             plot(axesHandle, x, y);
0731             [~, idx] = max(y);
0732             xlim(axesHandle, [-15 15]);
0733             ylim(axesHandle, [min(y) max(y)])
0734             line(x(idx) * [1 1], ylim(axesHandle), <span class="string">'Color'</span>, <span class="string">'r'</span>, <span class="string">'LineStyle'</span>, <span class="string">':'</span>, <span class="string">'Parent'</span>, axesHandle)
0735             text(3.75, min(y) + 0.95 * range(y), sprintf(<span class="string">'%d s'</span>, x(idx)), <span class="string">'Color'</span>, <span class="string">'r'</span>, <span class="string">'FontSize'</span>, 6, <span class="string">'Parent'</span>, axesHandle);
0736             xlabel(axesHandle, <span class="string">'Lag (s)'</span>, <span class="string">'FontSize'</span>, 7);
0737             ylabel(axesHandle, <span class="string">'Transcription-Decay X-Corr'</span>, <span class="string">'FontSize'</span>, 7);
0738             set(axesHandle, <span class="string">'box'</span>, <span class="string">'off'</span>, <span class="string">'tickdir'</span>, <span class="string">'out'</span>, <span class="string">'FontSize'</span>, 5, <span class="string">'XTick'</span>, -10:10:10, <span class="string">'YTick'</span>, 0.4:0.2:0.8);
0739             tick2text(axesHandle, <span class="string">'axis'</span>, <span class="string">'xy'</span>, <span class="string">'xtickoffset'</span>, 0.04, <span class="string">'ytickoffset'</span>, 0.03);
0740             set(getappdata(axesHandle, <span class="string">'XTickText'</span>), <span class="string">'FontSize'</span>, 5);
0741             set(getappdata(axesHandle, <span class="string">'YTickText'</span>), <span class="string">'FontSize'</span>, 5, <span class="string">'HorizontalAlign'</span>, <span class="string">'right'</span>)
0742             set(get(axesHandle, <span class="string">'ylabel'</span>), <span class="string">'Position'</span>, [-19 mean(ylim(axesHandle)) 0]);
0743             
0744             saveas(figHandle, [outDir <span class="string">'transcriptionVsRNADecayDynamics.pdf'</span>]);
0745             close(figHandle);
0746             
0747             <span class="comment">%correlations with NTP concentrations</span>
0748             [axesHandle, figHandle] = PlotUtil.newAxesHandle([17.4 17.4]);
0749             clf(figHandle);
0750             
0751             axesHandle = subplot(3, 3, 1, <span class="string">'Parent'</span>, figHandle);
0752             plot(axesHandle, ntpCounts * 1e15, simEndTimes / 3600, <span class="string">'.'</span>);
0753             xlabel(axesHandle, <span class="string">'Mean NTP Count (amol)'</span>, <span class="string">'FontSize'</span>, 7);
0754             ylabel(axesHandle, <span class="string">'Cell Cycle Length (h)'</span>, <span class="string">'FontSize'</span>, 7);
0755             set(axesHandle, <span class="string">'FontSize'</span>, 5, <span class="string">'box'</span>, <span class="string">'off'</span>, <span class="string">'tickDir'</span>, <span class="string">'out'</span>)
0756             
0757             axesHandle = subplot(3, 3, 2, <span class="string">'Parent'</span>, figHandle);
0758             plot(axesHandle, ntpConcs, simEndTimes / 3600, <span class="string">'.'</span>);
0759             xlabel(axesHandle, <span class="string">'Mean NTP Conc (mM)'</span>, <span class="string">'FontSize'</span>, 7);
0760             ylabel(axesHandle, <span class="string">'Cell Cycle Length (h)'</span>, <span class="string">'FontSize'</span>, 7);
0761             set(axesHandle, <span class="string">'FontSize'</span>, 5, <span class="string">'box'</span>, <span class="string">'off'</span>, <span class="string">'tickDir'</span>, <span class="string">'out'</span>)
0762             
0763             axesHandle = subplot(3, 3, 3, <span class="string">'Parent'</span>, figHandle);
0764             plot(axesHandle, ntpConcs, repInitDuration / 3600, <span class="string">'.'</span>);
0765             xlabel(axesHandle, <span class="string">'Mean NTP Conc (mM)'</span>, <span class="string">'FontSize'</span>, 7);
0766             ylabel(axesHandle, <span class="string">'Rep Init Duration (h)'</span>, <span class="string">'FontSize'</span>, 7);
0767             set(axesHandle, <span class="string">'FontSize'</span>, 5, <span class="string">'box'</span>, <span class="string">'off'</span>, <span class="string">'tickDir'</span>, <span class="string">'out'</span>)
0768             
0769             axesHandle = subplot(3, 3, 4, <span class="string">'Parent'</span>, figHandle);
0770             plot(axesHandle, ntpConcs, repDuration / 3600, <span class="string">'.'</span>);
0771             xlabel(axesHandle, <span class="string">'Mean NTP Conc (mM)'</span>, <span class="string">'FontSize'</span>, 7);
0772             ylabel(axesHandle, <span class="string">'Replication Duration (h)'</span>, <span class="string">'FontSize'</span>, 7);
0773             set(axesHandle, <span class="string">'FontSize'</span>, 5, <span class="string">'box'</span>, <span class="string">'off'</span>, <span class="string">'tickDir'</span>, <span class="string">'out'</span>)
0774             
0775             axesHandle = subplot(3, 3, 5, <span class="string">'Parent'</span>, figHandle);
0776             plot(axesHandle, ntpConcs, rnaWt * 1e15, <span class="string">'.'</span>);
0777             xlabel(axesHandle, <span class="string">'Mean NTP Conc (mM)'</span>, <span class="string">'FontSize'</span>, 7);
0778             ylabel(axesHandle, <span class="string">'RNA (fg)'</span>, <span class="string">'FontSize'</span>, 7);
0779             set(axesHandle, <span class="string">'FontSize'</span>, 5, <span class="string">'box'</span>, <span class="string">'off'</span>, <span class="string">'tickDir'</span>, <span class="string">'out'</span>)
0780             
0781             axesHandle = subplot(3, 3, 6, <span class="string">'Parent'</span>, figHandle);
0782             plot(axesHandle, ntpConcs, proteinWt * 1e15, <span class="string">'.'</span>);
0783             xlabel(axesHandle, <span class="string">'Mean NTP Conc (mM)'</span>, <span class="string">'FontSize'</span>, 7);
0784             ylabel(axesHandle, <span class="string">'Protein (fg)'</span>, <span class="string">'FontSize'</span>, 7);
0785             set(axesHandle, <span class="string">'FontSize'</span>, 5, <span class="string">'box'</span>, <span class="string">'off'</span>, <span class="string">'tickDir'</span>, <span class="string">'out'</span>)
0786             
0787             axesHandle = subplot(3, 3, 7, <span class="string">'Parent'</span>, figHandle);
0788             plot(axesHandle, ntpConcs, superhelicity, <span class="string">'.'</span>);
0789             xlabel(axesHandle, <span class="string">'Mean NTP Conc (mM)'</span>, <span class="string">'FontSize'</span>, 7);
0790             ylabel(axesHandle, <span class="string">'Superhelicity'</span>, <span class="string">'FontSize'</span>, 7);
0791             set(axesHandle, <span class="string">'FontSize'</span>, 5, <span class="string">'box'</span>, <span class="string">'off'</span>, <span class="string">'tickDir'</span>, <span class="string">'out'</span>)
0792             
0793             axesHandle = subplot(3, 3, 8, <span class="string">'Parent'</span>, figHandle);
0794             plot(axesHandle, ntpConcs, activeRNAPols, <span class="string">'.'</span>);
0795             xlabel(axesHandle, <span class="string">'Mean NTP Conc (mM)'</span>, <span class="string">'FontSize'</span>, 7);
0796             ylabel(axesHandle, <span class="string">'Active RNA Pols'</span>, <span class="string">'FontSize'</span>, 7);
0797             set(axesHandle, <span class="string">'FontSize'</span>, 5, <span class="string">'box'</span>, <span class="string">'off'</span>, <span class="string">'tickDir'</span>, <span class="string">'out'</span>)
0798             
0799             axesHandle = subplot(3, 3, 9, <span class="string">'Parent'</span>, figHandle);
0800             plot(axesHandle, ntpConcs, stalledRibs, <span class="string">'.'</span>);
0801             xlabel(axesHandle, <span class="string">'Mean NTP Conc (mM)'</span>, <span class="string">'FontSize'</span>, 7);
0802             ylabel(axesHandle, <span class="string">'Stalled Ribosomes'</span>, <span class="string">'FontSize'</span>, 7);
0803             set(axesHandle, <span class="string">'FontSize'</span>, 5, <span class="string">'box'</span>, <span class="string">'off'</span>, <span class="string">'tickDir'</span>, <span class="string">'out'</span>)
0804             
0805             saveas(figHandle, [outDir <span class="string">'ntpCorrelations.pdf'</span>]);
0806             close(figHandle);
0807         <span class="keyword">end</span>
0808         
0809         <a name="_sub5" href="#_subfunctions" class="code">function growthSensitivity(simBatchDir, selectedSim, outDirectory)</a>
0810             import edu.stanford.covert.cell.sim.analysis.Figures34;
0811             import edu.stanford.covert.cell.sim.util.FitConstants;
0812             import edu.stanford.covert.cell.sim.util.PlotUtil;
0813             import edu.stanford.covert.cell.sim.util.SimulationEnsemble;
0814             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;            
0815             import edu.stanford.covert.util.ComputationUtil;
0816             
0817             <span class="keyword">if</span> ~exist([outDirectory filesep <span class="string">'growthSensitivity'</span>], <span class="string">'dir'</span>)
0818                 mkdir([outDirectory filesep <span class="string">'growthSensitivity'</span>]);
0819             <span class="keyword">end</span>            
0820             outDir = [outDirectory filesep <span class="string">'growthSensitivity'</span> filesep];
0821             
0822             <span class="comment">%% get data</span>
0823             [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep <span class="string">'1'</span>]);
0824             
0825 <span class="comment">%             if ~exist([outDirectory filesep 'growthSensitivity.mat'], 'file')</span>
0826 <span class="comment">%                 save([outDirectory filesep 'growthSensitivity.mat']);</span>
0827 <span class="comment">%             else</span>
0828 <span class="comment">%                 load([outDirectory filesep 'growthSensitivity.mat']);</span>
0829 <span class="comment">%             end</span>
0830 
0831             g = sim.gene;
0832             r = sim.state(<span class="string">'Rna'</span>);
0833             mr = sim.state(<span class="string">'MetabolicReaction'</span>);
0834             met = sim.process(<span class="string">'Metabolism'</span>);
0835             
0836             <span class="comment">%% path to optimal growth</span>
0837             <span class="comment">%get metabolic network</span>
0838             fitter = FitConstants(sim);
0839             [~, freeMons, freeCpxs] = fitter.calcMacromolecularCounts(fitter.constructParameterVectorFromSimulation());
0840             
0841             growth0 = mr.growth0 * met.macromoleculeStateInitializationGrowthFactor;
0842             
0843             substrates = zeros(size(met.substrates));
0844             substrates(met.substrateMetaboliteLocalIndexs, :) = met.substrates(met.substrateMetaboliteLocalIndexs, :);
0845             substrates(met.substrateIndexs_internalExchangedLimitedMetabolites) = <span class="keyword">...</span>
0846                 -growth0 *  met.metabolismRecyclingProduction(met.substrateIndexs_internalExchangedLimitedMetabolites);
0847             substrates(met.substrateMonomerLocalIndexs, :) = repmat(freeMons(met.substrateMonomerGlobalIndexs(:, 1)), 1, size(substrates, 2));
0848             substrates(met.substrateComplexLocalIndexs, :) = repmat(freeCpxs(met.substrateComplexGlobalIndexs(:, 1)), 1, size(substrates, 2));
0849             
0850             enzymes = zeros(size(met.enzymeWholeCellModelIDs));
0851             enzymes(met.enzymeMonomerLocalIndexs) = freeMons(met.enzymeMonomerGlobalIndexs);
0852             enzymes(met.enzymeComplexLocalIndexs) = freeCpxs(met.enzymeComplexGlobalIndexs);
0853             
0854             [fbaObj, fbaSMat, fbaRxnBounds] = met.calcEffectiveFBANetwork();
0855             fluxBounds = met.calcFluxBounds(substrates, enzymes, fbaRxnBounds, met.fbaEnzymeBounds);
0856             [initFbaReactionFluxs, ~, ~, errFlag, errMsg] = ComputationUtil.linearProgramming(<span class="keyword">...</span>
0857                 <span class="string">'maximize'</span>, fbaObj, fbaSMat, <span class="keyword">...</span>
0858                 met.fbaRightHandSide, fluxBounds(:, 1), fluxBounds(:, 2), <span class="keyword">...</span>
0859                 <span class="string">'S'</span>, <span class="string">'C'</span>, met.linearProgrammingOptions);
0860             initGrowth = initFbaReactionFluxs(met.fbaReactionIndexs_biomassProduction);
0861             
0862             fbaEnzBounds = met.fbaEnzymeBounds;
0863             fbaRHS = met.fbaRightHandSide;
0864             fbaRxnCatMat = met.fbaReactionCatalysisMatrix;
0865             bmRxnTfs = false(size(fbaSMat, 2), 1);
0866             bmRxnTfs(met.fbaReactionIndexs_biomassProduction) = true;
0867             nonKineticFluxBounds = met.calcFluxBounds(<span class="keyword">...</span>
0868                 substrates, enzymes, fbaRxnBounds, met.fbaEnzymeBounds, <span class="keyword">...</span>
0869                 false, true, true, false, true, true);
0870             
0871             kineticLimitedEnzymes = any(met.fbaReactionCatalysisMatrix(any(isfinite(met.fbaEnzymeBounds), 2), :), 1)';
0872             kineticUnlimitedEnzymes = any(met.fbaReactionCatalysisMatrix(any(isinf(met.fbaEnzymeBounds), 2), :), 1)';
0873             
0874             <span class="comment">%split reversible reactions into positive and negative</span>
0875             <span class="comment">%components</span>
0876             idxs = find(fluxBounds(:, 1) &lt; 0);
0877             fluxBounds = [max(fluxBounds, 0); max(-fliplr(fluxBounds(idxs, :)), 0)];
0878             nonKineticFluxBounds = [max(nonKineticFluxBounds, 0); max(-fliplr(nonKineticFluxBounds(idxs, :)), 0)];
0879             fbaEnzBounds = [max(fbaEnzBounds, 0); max(-fliplr(fbaEnzBounds(idxs, :)), 0)];
0880             fbaSMat = [fbaSMat -fbaSMat(:, idxs)];
0881             fbaObj = [fbaObj; -fbaObj(idxs)];
0882             fbaRxnCatMat = [fbaRxnCatMat; fbaRxnCatMat(idxs, :)];
0883             bmRxnTfs = [bmRxnTfs; bmRxnTfs(idxs)];
0884             
0885             idxs = find(fluxBounds(:, 2) ~= 0);
0886             fluxBounds = fluxBounds(idxs, :);
0887             nonKineticFluxBounds = nonKineticFluxBounds(idxs, :);
0888             fbaEnzBounds = fbaEnzBounds(idxs, :);
0889             fbaSMat = fbaSMat(:, idxs);
0890             fbaObj = fbaObj(idxs);
0891             fbaRxnCatMat = fbaRxnCatMat(idxs, :);
0892             bmRxnTfs = bmRxnTfs(idxs);            
0893             
0894             nMet = size(fbaSMat, 1);
0895             nRxn = size(fbaSMat, 2);
0896             nEnz = size(fbaRxnCatMat, 2);
0897             nKinEnz = sum(kineticLimitedEnzymes);
0898             bmRxnIdx = find(bmRxnTfs);
0899             uFbaEnzBounds = unique(fbaEnzBounds(:, 2));
0900             kineticConstraints = (1 ./ min(uFbaEnzBounds(end-1), fbaEnzBounds(:, 2 * ones(nKinEnz, 1))))';
0901             kineticConstraints(fbaRxnCatMat(:, kineticLimitedEnzymes)' == 0) = 0;
0902             enzymeConstraints = zeros(nKinEnz, nEnz);
0903             enzymeConstraints(sub2ind(size(enzymeConstraints), 1:nKinEnz, find(kineticLimitedEnzymes)')) = 1;
0904             
0905             enzGeneComp = met.enzymeGeneComposition;
0906             
0907             <span class="comment">%check that growth not affected by splitting</span>
0908             [initFbaReactionFluxs_flux, ~, ~, errFlag, errMsg] = ComputationUtil.linearProgramming(<span class="keyword">...</span>
0909                 <span class="string">'maximize'</span>, fbaObj, fbaSMat, <span class="keyword">...</span>
0910                 fbaRHS, <span class="keyword">...</span>
0911                 fluxBounds(:, 1), fluxBounds(:, 2), <span class="keyword">...</span>
0912                 <span class="string">'S'</span>, <span class="string">'C'</span>, met.linearProgrammingOptions);
0913             <span class="keyword">if</span> errFlag
0914                 throw(MException(<span class="string">'Metabolism:error'</span>, <span class="string">'Unable to optimize growth: %s'</span>, errMsg));
0915             <span class="keyword">end</span>
0916             assertElementsAlmostEqual(initGrowth, initFbaReactionFluxs_flux(bmRxnIdx));
0917             
0918             <span class="comment">%check that growth not affected additional constraints on</span>
0919             <span class="comment">%enzyme mass</span>
0920             [initFbaReactionFluxs_enzyme, ~, ~, errFlag, errMsg] = ComputationUtil.linearProgramming(<span class="keyword">...</span>
0921                 <span class="string">'maximize'</span>, [fbaObj; zeros(nEnz, 1)], [fbaSMat zeros(nMet, nEnz); zeros(1, nRxn) met.enzymeMolecularWeights'; kineticConstraints -enzymeConstraints], <span class="keyword">...</span>
0922                 [fbaRHS; enzymes' * met.enzymeMolecularWeights; zeros(nKinEnz, 1)], <span class="keyword">...</span>
0923                 [nonKineticFluxBounds(:, 1); zeros(nEnz, 1) + enzymes], [nonKineticFluxBounds(:, 2); inf(nEnz, 1)], <span class="keyword">...</span>
0924                 [repmat(<span class="string">'S'</span>, size(fbaRHS)); <span class="string">'S'</span>; repmat(<span class="string">'U'</span>, nKinEnz, 1)], <span class="string">'C'</span>, met.linearProgrammingOptions);
0925             <span class="keyword">if</span> errFlag
0926                 throw(MException(<span class="string">'Metabolism:error'</span>, <span class="string">'Unable to optimize growth: %s'</span>, errMsg));
0927             <span class="keyword">end</span>
0928             assertElementsAlmostEqual(initGrowth, initFbaReactionFluxs_enzyme(bmRxnIdx));
0929             
0930             <span class="comment">%find maximal growth: enzyme method</span>
0931             [result, ~, ~, errFlag, errMsg] = ComputationUtil.linearProgramming(<span class="keyword">...</span>
0932                 <span class="string">'maximize'</span>, [fbaObj; zeros(nEnz, 1)], [fbaSMat zeros(nMet, nEnz); zeros(1, nRxn) met.enzymeMolecularWeights'; kineticConstraints -enzymeConstraints], <span class="keyword">...</span>
0933                 [fbaRHS; enzymes' * met.enzymeMolecularWeights; zeros(nKinEnz, 1)], <span class="keyword">...</span>
0934                 [nonKineticFluxBounds(:, 1); (kineticLimitedEnzymes &amp; kineticUnlimitedEnzymes) + (~kineticLimitedEnzymes) .* enzymes], [nonKineticFluxBounds(:, 2); inf(nEnz, 1)], <span class="keyword">...</span>
0935                 [repmat(<span class="string">'S'</span>, size(fbaRHS)); <span class="string">'S'</span>; repmat(<span class="string">'U'</span>, nKinEnz, 1)], <span class="string">'C'</span>, met.linearProgrammingOptions);
0936             <span class="keyword">if</span> errFlag
0937                 throw(MException(<span class="string">'Metabolism:error'</span>, <span class="string">'Unable to optimize growth: %s'</span>, errMsg));
0938             <span class="keyword">end</span>
0939             maxGrowth_enzyme = result(bmRxnIdx);
0940             assertElementsAlmostEqual(maxGrowth_enzyme, <span class="keyword">...</span>
0941                 met.calcGrowthRate(met.calcFluxBounds(substrates, result(end-nEnz+1:end), met.fbaReactionBounds, met.fbaEnzymeBounds)));
0942             
0943             <span class="comment">%find maximal growth: flux method</span>
0944             totalEnzymeConstriant = 1 ./ fbaEnzBounds(:, 2) .* (fbaRxnCatMat * met.enzymeMolecularWeights) * 1e-7;
0945             totalEnzymeConstriant(isinf(fbaEnzBounds(:, 2)) | fbaEnzBounds(:, 2) == 0) = 0;
0946             [result, ~, ~, errFlag, errMsg] = ComputationUtil.linearProgramming(<span class="keyword">...</span>
0947                 <span class="string">'maximize'</span>, fbaObj, [fbaSMat; totalEnzymeConstriant'], <span class="keyword">...</span>
0948                 [fbaRHS; enzymes(kineticLimitedEnzymes)' * met.enzymeMolecularWeights(kineticLimitedEnzymes) * 1e-7], <span class="keyword">...</span>
0949                 nonKineticFluxBounds(:, 1), nonKineticFluxBounds(:, 2), <span class="keyword">...</span>
0950                 [repmat(<span class="string">'S'</span>, size(fbaRHS)); <span class="string">'U'</span>], <span class="string">'C'</span>, met.linearProgrammingOptions);
0951             <span class="keyword">if</span> errFlag
0952                 throw(MException(<span class="string">'Metabolism:error'</span>, <span class="string">'Unable to optimize growth: %s'</span>, errMsg));
0953             <span class="keyword">end</span>
0954             maxGrowth_flux = result(bmRxnIdx); 
0955             
0956             <span class="comment">%find intermediate growths</span>
0957             [optGrowths, optEnzymes, optEnzymeOrder] = Figures34.calcOptimalEnzymeExpressionChanges(fbaObj, fbaSMat, fbaRHS, <span class="keyword">...</span>
0958                 kineticConstraints, enzymeConstraints, nonKineticFluxBounds, <span class="keyword">...</span>
0959                 enzymes, kineticLimitedEnzymes, kineticUnlimitedEnzymes, bmRxnIdx, sim);
0960             
0961             <span class="comment">%find intermediate growths</span>
0962             nSteps = 40;
0963             growths_enzyme = [
0964                 linspace(0, initGrowth, 20) <span class="keyword">...</span>
0965                 linspace(initGrowth, maxGrowth_enzyme, 20)
0966                 ];
0967             growths_flux = [
0968                 linspace(0, initGrowth, 20) <span class="keyword">...</span>
0969                 linspace(initGrowth, maxGrowth_flux, 20)
0970                 ];
0971             distsL1 = zeros(nSteps, 2);
0972             distsL2 = zeros(nSteps, 2);
0973             distsInf = zeros(nSteps, 2);
0974             distsL1Reg = zeros(nSteps, 2);
0975             cplexopts = cplexoptimset(<span class="string">'Display'</span>, <span class="string">'off'</span>);
0976             <span class="keyword">for</span> i = 1:nSteps
0977                 <span class="comment">%enzyme methods</span>
0978                 tmpNonKineticFluxBounds = nonKineticFluxBounds;
0979                 tmpNonKineticFluxBounds(bmRxnIdx, :) = growths_enzyme(i);
0980                 [~, ~, distsL1(i, 1), errFlag, errMsg] = ComputationUtil.linearProgramming(<span class="keyword">...</span>
0981                     <span class="string">'minimize'</span>, [
0982                     zeros(nRxn, 1)
0983                     zeros(nEnz, 1)
0984                     ones(nEnz, 1)
0985                     ], [<span class="keyword">...</span>
0986                     fbaSMat zeros(nMet, nEnz) zeros(nMet, nEnz)
0987                     zeros(1, nRxn) met.enzymeMolecularWeights' zeros(1, nEnz)
0988                     kineticConstraints -enzymeConstraints zeros(nKinEnz, nEnz)
0989                     zeros(nEnz, nRxn)  eye(nEnz) -eye(nEnz)
0990                     zeros(nEnz, nRxn) -eye(nEnz) -eye(nEnz)
0991                     ], [
0992                     fbaRHS
0993                     enzymes' * met.enzymeMolecularWeights
0994                     zeros(nKinEnz, 1)
0995                     initFbaReactionFluxs_enzyme(end-nEnz+1:end)
0996                     -initFbaReactionFluxs_enzyme(end-nEnz+1:end)
0997                     ], [
0998                     tmpNonKineticFluxBounds(:, 1)
0999                     (kineticLimitedEnzymes &amp; kineticUnlimitedEnzymes) + (~kineticLimitedEnzymes) .* enzymes
1000                     -inf(nEnz, 1)
1001                     ], [
1002                     tmpNonKineticFluxBounds(:, 2)
1003                     inf(nEnz, 1)
1004                     inf(nEnz, 1)
1005                     ], <span class="keyword">...</span>
1006                     [
1007                     repmat(<span class="string">'S'</span>, size(fbaRHS))
1008                     <span class="string">'S'</span>
1009                     repmat(<span class="string">'U'</span>, nKinEnz, 1)
1010                     repmat(<span class="string">'U'</span>, 2 * nEnz, 1)
1011                     ], <span class="string">'C'</span>, met.linearProgrammingOptions);
1012                 <span class="keyword">if</span> errFlag
1013                     throw(MException(<span class="string">'Metabolism:error'</span>, <span class="string">'Unable to optimize growth: %s'</span>, errMsg));
1014                 <span class="keyword">end</span>
1015                 
1016                 [~, distsL2(i, 1), errFlag, output] = cplexqp(<span class="keyword">...</span>
1017                     2 * diag([zeros(nRxn, 1); ones(nEnz, 1)]), -2 * ([zeros(nRxn, 1); ones(nEnz, 1)]) .* initFbaReactionFluxs_enzyme, <span class="keyword">...</span>
1018                     [kineticConstraints -enzymeConstraints], zeros(nKinEnz, 1), <span class="keyword">...</span>
1019                     [fbaSMat zeros(nMet, nEnz); zeros(1, nRxn) met.enzymeMolecularWeights' * 1e-7], [fbaRHS; enzymes' * met.enzymeMolecularWeights * 1e-7], <span class="keyword">...</span>
1020                     [tmpNonKineticFluxBounds(:, 1); (kineticLimitedEnzymes &amp; kineticUnlimitedEnzymes) + (~kineticLimitedEnzymes) .* enzymes], [tmpNonKineticFluxBounds(:, 2); inf(nEnz, 1)], <span class="keyword">...</span>
1021                     [], cplexopts);
1022                 <span class="keyword">if</span> errFlag ~= 1
1023                     throw(MException(<span class="string">'Metabolism:error'</span>, <span class="string">'Unable to optimize growth: %s'</span>, output.message));
1024                 <span class="keyword">end</span>
1025                 
1026                 [~, ~,  distsInf(i, 1), errFlag, errMsg] = ComputationUtil.linearProgramming(<span class="keyword">...</span>
1027                     <span class="string">'minimize'</span>, [zeros(nRxn, 1); zeros(nEnz, 1); 1], [
1028                     fbaSMat zeros(nMet, nEnz) zeros(nMet, 1)
1029                     zeros(1, nRxn) met.enzymeMolecularWeights' 0
1030                     kineticConstraints -enzymeConstraints zeros(nKinEnz, 1)
1031                     zeros(nEnz, nRxn)  eye(nEnz) -ones(nEnz, 1)
1032                     zeros(nEnz, nRxn) -eye(nEnz) -ones(nEnz, 1)
1033                     ], [
1034                     fbaRHS
1035                     enzymes' * met.enzymeMolecularWeights
1036                     zeros(nKinEnz, 1)
1037                     initFbaReactionFluxs_enzyme(end-nEnz+1:end)
1038                     -initFbaReactionFluxs_enzyme(end-nEnz+1:end)
1039                     ], <span class="keyword">...</span>
1040                     [tmpNonKineticFluxBounds(:, 1); (kineticLimitedEnzymes &amp; kineticUnlimitedEnzymes) + (~kineticLimitedEnzymes) .* enzymes; -inf], <span class="keyword">...</span>
1041                     [tmpNonKineticFluxBounds(:, 2); inf(nEnz, 1); inf], <span class="keyword">...</span>
1042                     [repmat(<span class="string">'S'</span>, size(fbaRHS)); <span class="string">'S'</span>; repmat(<span class="string">'U'</span>, nKinEnz, 1); repmat(<span class="string">'U'</span>, 2 * nEnz, 1)], <span class="string">'C'</span>, met.linearProgrammingOptions);
1043                 <span class="keyword">if</span> errFlag
1044                     throw(MException(<span class="string">'Metabolism:error'</span>, <span class="string">'Unable to optimize growth: %s'</span>, errMsg));
1045                 <span class="keyword">end</span>
1046                 
1047                 <span class="comment">%flux methods</span>
1048                 tmpNonKineticFluxBounds = nonKineticFluxBounds;
1049                 tmpNonKineticFluxBounds(bmRxnIdx, :) = growths_flux(i);
1050                 
1051                 [~, ~, distsL1(i, 2), errFlag, errMsg] = ComputationUtil.linearProgramming(<span class="keyword">...</span>
1052                     <span class="string">'minimize'</span>, [zeros(nRxn, 1); ones(nRxn, 1)], <span class="keyword">...</span>
1053                     [[fbaSMat; totalEnzymeConstriant'] zeros(nMet+1, nRxn); eye(nRxn) -eye(nRxn); -eye(nRxn) -eye(nRxn)], <span class="keyword">...</span>
1054                     [fbaRHS; enzymes(kineticLimitedEnzymes)' * met.enzymeMolecularWeights(kineticLimitedEnzymes) * 1e-7; initFbaReactionFluxs_flux; -initFbaReactionFluxs_flux], <span class="keyword">...</span>
1055                     [tmpNonKineticFluxBounds(:, 1); -inf(nRxn, 1)], [tmpNonKineticFluxBounds(:, 2); inf(nRxn, 1)], <span class="keyword">...</span>
1056                     [repmat(<span class="string">'S'</span>, nMet, 1); <span class="string">'U'</span>; repmat(<span class="string">'U'</span>, 2 * nRxn, 1)], <span class="string">'C'</span>, met.linearProgrammingOptions);
1057                 <span class="keyword">if</span> errFlag
1058                     throw(MException(<span class="string">'Metabolism:error'</span>, <span class="string">'Unable to optimize growth: %s'</span>, errMsg));
1059                 <span class="keyword">end</span>
1060                 
1061                 [~, distsL2(i, 2), errFlag, output] = cplexqp(<span class="keyword">...</span>
1062                     2 * eye(nRxn), -2 * initFbaReactionFluxs_flux, <span class="keyword">...</span>
1063                     totalEnzymeConstriant', enzymes(kineticLimitedEnzymes)' * met.enzymeMolecularWeights(kineticLimitedEnzymes) * 1e-7, <span class="keyword">...</span>
1064                     fbaSMat, fbaRHS, <span class="keyword">...</span>
1065                     tmpNonKineticFluxBounds(:, 1), tmpNonKineticFluxBounds(:, 2), <span class="keyword">...</span>
1066                     [], cplexopts);
1067                 <span class="keyword">if</span> errFlag ~= 1
1068                     throw(MException(<span class="string">'Metabolism:error'</span>, <span class="string">'Unable to optimize growth: %s'</span>, output.message));
1069                 <span class="keyword">end</span>
1070                 
1071                 [~, ~, distsInf(i, 2), errFlag, errMsg] = ComputationUtil.linearProgramming(<span class="keyword">...</span>
1072                     <span class="string">'minimize'</span>, [zeros(nRxn, 1); 1], <span class="keyword">...</span>
1073                     [[fbaSMat; totalEnzymeConstriant'] zeros(nMet+1, 1); eye(nRxn) -ones(nRxn, 1); -eye(nRxn) -ones(nRxn, 1)], <span class="keyword">...</span>
1074                     [fbaRHS; enzymes(kineticLimitedEnzymes)' * met.enzymeMolecularWeights(kineticLimitedEnzymes) * 1e-7; initFbaReactionFluxs_flux; -initFbaReactionFluxs_flux], <span class="keyword">...</span>
1075                     [tmpNonKineticFluxBounds(:, 1); -inf], [tmpNonKineticFluxBounds(:, 2); inf], <span class="keyword">...</span>
1076                     [repmat(<span class="string">'S'</span>, nMet, 1); <span class="string">'U'</span>; repmat(<span class="string">'U'</span>, 2 * nRxn, 1)], <span class="string">'C'</span>, met.linearProgrammingOptions);
1077                 <span class="keyword">if</span> errFlag
1078                     throw(MException(<span class="string">'Metabolism:error'</span>, <span class="string">'Unable to optimize growth: %s'</span>, errMsg));
1079                 <span class="keyword">end</span>
1080             <span class="keyword">end</span>
1081             
1082             <span class="comment">% plot</span>
1083             figHandle = figure;
1084             ylims = [initGrowth max(maxGrowth_enzyme, maxGrowth_flux)];
1085             
1086             axesHandle = subplot(1, 1, 1);
1087             h = plot(0:size(optGrowths, 1) - 1, optGrowths);
1088             set(axesHandle, <span class="string">'FontSize'</span>, 5, <span class="string">'TickDir'</span>, <span class="string">'out'</span>, <span class="string">'box'</span>, <span class="string">'off'</span>);
1089             xlabel(<span class="string">'Regularized Enzyme distance'</span>, <span class="string">'FontSize'</span>, 7);
1090             ylabel(<span class="string">'Growth (cell s^{-1})'</span>, <span class="string">'FontSize'</span>, 7);
1091             ylim(ylims);
1092             xlim([-0.5 numel(optGrowths)-0.5]);
1093             <span class="keyword">for</span> i = 1:size(optGrowths, 1)
1094                 text(i-1, optGrowths(i), sprintf(<span class="string">'%d '</span>, optEnzymeOrder{i}), <span class="keyword">...</span>
1095                     <span class="string">'FontSize'</span>, 5, <span class="string">'HorizontalAlign'</span>, <span class="string">'left'</span>, <span class="string">'VerticalAlign'</span>, <span class="string">'bottom'</span>, <span class="string">'interpreter'</span>, <span class="string">'none'</span>, <span class="keyword">...</span>
1096                     <span class="string">'Color'</span>, get(h(2), <span class="string">'Color'</span>));
1097             <span class="keyword">end</span>
1098             
1099             saveas(figHandle, [outDir <span class="string">'growthEvolution-1 gene at a time.pdf'</span>]);
1100             close(figHandle);
1101             
1102             <span class="comment">% plot all metrics</span>
1103             figHandle = figure;
1104             ylims = [0 max(maxGrowth_enzyme, maxGrowth_flux)];
1105             
1106             axesHandle = subplot(2, 4, 1);
1107             h = plot(0:size(optGrowths, 1) - 1, optGrowths);
1108             set(axesHandle, <span class="string">'FontSize'</span>, 5, <span class="string">'TickDir'</span>, <span class="string">'out'</span>, <span class="string">'box'</span>, <span class="string">'off'</span>);
1109             xlabel(<span class="string">'Regularized Enzyme distance'</span>, <span class="string">'FontSize'</span>, 7);
1110             ylabel(<span class="string">'Growth (cell s^{-1})'</span>, <span class="string">'FontSize'</span>, 7);
1111             ylim(ylims);
1112             xlim([-0.5 numel(optGrowths)-0.5]);
1113             <span class="keyword">for</span> i = 1:size(optGrowths, 1)
1114                 text(i-1, optGrowths(i), sprintf(<span class="string">'%d '</span>, optEnzymeOrder{i}), <span class="keyword">...</span>
1115                     <span class="string">'FontSize'</span>, 5, <span class="string">'HorizontalAlign'</span>, <span class="string">'left'</span>, <span class="string">'VerticalAlign'</span>, <span class="string">'bottom'</span>, <span class="string">'interpreter'</span>, <span class="string">'none'</span>, <span class="keyword">...</span>
1116                     <span class="string">'Color'</span>, get(h(2), <span class="string">'Color'</span>));
1117             <span class="keyword">end</span>
1118             
1119             axesHandle = subplot(2, 4, 2);
1120             plot(distsL1(:, 1), growths_enzyme)
1121             xlabel(<span class="string">'L_1 Enzyme distance'</span>, <span class="string">'FontSize'</span>, 7)
1122             ylim(ylims);
1123             set(axesHandle, <span class="string">'FontSize'</span>, 5, <span class="string">'TickDir'</span>, <span class="string">'out'</span>, <span class="string">'box'</span>, <span class="string">'off'</span>);
1124             
1125             axesHandle = subplot(2, 4, 3);
1126             plot(distsL2(:, 1), growths_enzyme)
1127             xlabel(<span class="string">'L_2 Enzyme distance'</span>, <span class="string">'FontSize'</span>, 7)
1128             ylim(ylims);
1129             set(axesHandle, <span class="string">'FontSize'</span>, 5, <span class="string">'TickDir'</span>, <span class="string">'out'</span>, <span class="string">'box'</span>, <span class="string">'off'</span>);
1130             
1131             axesHandle = subplot(2, 4, 4);
1132             plot(distsInf(:, 1), growths_enzyme)
1133             xlabel(<span class="string">'L_{\infty} Enzyme distance'</span>, <span class="string">'FontSize'</span>, 7)
1134             ylim(ylims);
1135             set(axesHandle, <span class="string">'FontSize'</span>, 5, <span class="string">'TickDir'</span>, <span class="string">'out'</span>, <span class="string">'box'</span>, <span class="string">'off'</span>);
1136             
1137             axesHandle = subplot(2, 4, 6);
1138             plot(distsL1(:, 2), growths_flux)
1139             xlabel(<span class="string">'L_1 Flux distance'</span>, <span class="string">'FontSize'</span>, 7)
1140             ylim(ylims);
1141             set(axesHandle, <span class="string">'FontSize'</span>, 5, <span class="string">'TickDir'</span>, <span class="string">'out'</span>, <span class="string">'box'</span>, <span class="string">'off'</span>);
1142             
1143             axesHandle = subplot(2, 4, 7);
1144             plot(distsL2(:, 2), growths_flux)
1145             xlabel(<span class="string">'L_2 Flux distance'</span>, <span class="string">'FontSize'</span>, 7)
1146             ylim(ylims);
1147             set(axesHandle, <span class="string">'FontSize'</span>, 5, <span class="string">'TickDir'</span>, <span class="string">'out'</span>, <span class="string">'box'</span>, <span class="string">'off'</span>);
1148             
1149             axesHandle = subplot(2, 4, 8);
1150             plot(distsInf(:, 2), growths_flux)
1151             xlabel(<span class="string">'L_{\infty} Flux distance'</span>, <span class="string">'FontSize'</span>, 7)
1152             ylim(ylims);
1153             set(axesHandle, <span class="string">'FontSize'</span>, 5, <span class="string">'TickDir'</span>, <span class="string">'out'</span>, <span class="string">'box'</span>, <span class="string">'off'</span>);
1154             
1155             saveas(figHandle, [outDir <span class="string">'growthEvolution-several metrics.pdf'</span>]);
1156             close(figHandle);
1157             
1158             <span class="comment">%% compare optimal to suboptimal growth</span>
1159             paramVec0 = fitter.constructParameterVectorFromSimulation();
1160             [rnaExp, nmpComp, aaComp, rnaWtFracs, rnaDecayRates, <span class="keyword">...</span>
1161                 biomassComposition, biomassProduction, byproducts, <span class="keyword">...</span>
1162                 unaccountedEnergyConsumption] = fitter.extractParameterVector(paramVec0);
1163             
1164             import edu.stanford.covert.util.ConstantUtil;
1165             pm = sim.state(<span class="string">'ProteinMonomer'</span>);
1166             t = sim.state(<span class="string">'Time'</span>);
1167             mass = sim.state(<span class="string">'Mass'</span>);
1168             rnaMWs = r.molecularWeights(r.matureIndexs) / ConstantUtil.nAvogadro;
1169             monMWs = pm.molecularWeights(pm.matureIndexs) / ConstantUtil.nAvogadro;
1170             
1171             rnaCnt = zeros(size(r.matureIndexs));
1172             rnaCnt(r.matureMRNAIndexs) = <span class="keyword">...</span>
1173                 rnaExp(r.matureMRNAIndexs) * <span class="keyword">...</span>
1174                 mass.cellInitialDryWeight * mass.initialFractionNTPsInRNAs * mass.dryWeightFractionRNA * <span class="keyword">...</span>
1175                 rnaWtFracs(r.mRNAWeightFractionIndexs) / (rnaMWs(r.matureMRNAIndexs)' * rnaExp(r.matureMRNAIndexs));
1176             rnaCnt(r.matureRRNAIndexs) = <span class="keyword">...</span>
1177                 rnaExp(r.matureRRNAIndexs) * <span class="keyword">...</span>
1178                 mass.cellInitialDryWeight * mass.initialFractionNTPsInRNAs * mass.dryWeightFractionRNA * <span class="keyword">...</span>
1179                 sum(rnaWtFracs(r.rRNAWeightFractionIndexs)) / (rnaMWs(r.matureRRNAIndexs)' * rnaExp(r.matureRRNAIndexs));
1180             rnaCnt(r.matureSRNAIndexs) = <span class="keyword">...</span>
1181                 rnaExp(r.matureSRNAIndexs) * <span class="keyword">...</span>
1182                 mass.cellInitialDryWeight * mass.initialFractionNTPsInRNAs * mass.dryWeightFractionRNA * <span class="keyword">...</span>
1183                 rnaWtFracs(r.sRNAWeightFractionIndexs) / (rnaMWs(r.matureSRNAIndexs)' * rnaExp(r.matureSRNAIndexs));
1184             rnaCnt(r.matureTRNAIndexs) = <span class="keyword">...</span>
1185                 rnaExp(r.matureTRNAIndexs) * <span class="keyword">...</span>
1186                 mass.cellInitialDryWeight * mass.initialFractionNTPsInRNAs * mass.dryWeightFractionRNA * <span class="keyword">...</span>
1187                 rnaWtFracs(r.tRNAWeightFractionIndexs) / (rnaMWs(r.matureTRNAIndexs)' * rnaExp(r.matureTRNAIndexs));
1188             
1189             monCnt = (r.matureRNAGeneComposition(g.mRNAIndexs, r.matureMRNAIndexs) * rnaCnt(r.matureMRNAIndexs)) ./ <span class="keyword">...</span>
1190                 (log(2) / t.cellCycleLength + pm.decayRates(pm.matureIndexs));
1191             monCnt = mass.cellInitialDryWeight * mass.initialFractionAAsInMonomers * mass.dryWeightFractionProtein * monCnt / (monMWs' * monCnt);
1192             
1193             idx = find(diff(optGrowths(:, 1)) /  maxGrowth_enzyme &gt; 1e-6, 1, <span class="string">'last'</span>) + 1;
1194             metMonCnt = r.matureRNAGeneComposition(g.mRNAIndexs, :) * max(r.matureRNAGeneComposition .* repmat(enzGeneComp * squeeze(optEnzymes(idx, 1, :)), 1, size(r.matureRNAGeneComposition, 2)), [], 1)';
1195             tfs = metMonCnt ~= 0;
1196             monCnt1 = max(monCnt, metMonCnt);
1197             monCnt1(~tfs) = monCnt1(~tfs) * (monCnt' * monMWs - monCnt1(tfs)' * monMWs(tfs)) / (monCnt1(~tfs)' * monMWs(~tfs));            
1198             monCnt = monCnt1;
1199             
1200             mrnaExp = ComputationUtil.invertCompositionMatrix(r.matureRNAGeneComposition(g.mRNAIndexs, r.matureMRNAIndexs)) * <span class="keyword">...</span>
1201                 (monCnt .* (log(2) / t.cellCycleLength + pm.decayRates(pm.matureIndexs)));
1202             rnaCnt(r.matureMRNAIndexs) = mrnaExp * <span class="keyword">...</span>
1203                 mass.cellInitialDryWeight * mass.initialFractionNTPsInRNAs * mass.dryWeightFractionRNA * <span class="keyword">...</span>
1204                 rnaWtFracs(r.mRNAWeightFractionIndexs) / (rnaMWs(r.matureMRNAIndexs)' * mrnaExp);
1205             rnaExp = rnaCnt / sum(rnaCnt);
1206             
1207             paramVec = fitter.constructParameterVector(<span class="keyword">...</span>
1208                 rnaExp, nmpComp, aaComp, rnaWtFracs, rnaDecayRates, <span class="keyword">...</span>
1209                 biomassComposition, biomassProduction, byproducts, unaccountedEnergyConsumption);
1210             [~, freeMons, freeCpxs] = fitter.calcMacromolecularCounts(paramVec);
1211             tmpEnzymes = zeros(size(met.enzymeWholeCellModelIDs));
1212             tmpEnzymes(met.enzymeMonomerLocalIndexs) = freeMons(met.enzymeMonomerGlobalIndexs);
1213             tmpEnzymes(met.enzymeComplexLocalIndexs) = freeCpxs(met.enzymeComplexGlobalIndexs);
1214             assertElementsAlmostEqual(maxGrowth_enzyme, <span class="keyword">...</span>
1215                 met.calcGrowthRate(met.calcFluxBounds(substrates, tmpEnzymes, met.fbaReactionBounds, met.fbaEnzymeBounds)), <span class="keyword">...</span>
1216                 <span class="string">'relative'</span>, 1e-2);
1217             
1218 <span class="comment">%             %simulate growth rate distribution</span>
1219 <span class="comment">%             paramVecs = {paramVec0; paramVec};</span>
1220 <span class="comment">%             meanGrowthRates = mr.meanInitialGrowthRate * [1 maxGrowth_enzyme / initGrowth];</span>
1221 <span class="comment">%             mr.initialGrowthFilterWidth = inf;</span>
1222 <span class="comment">%             growths = zeros(numel(paramVecs), 100);</span>
1223 <span class="comment">%             for i = 1:numel(paramVecs)</span>
1224 <span class="comment">%                 fitter.applyParameterVectorToSimulation(paramVecs{i});</span>
1225 <span class="comment">%                 mr.meanInitialGrowthRate = meanGrowthRates(i);</span>
1226 <span class="comment">%                 for j = 1:100</span>
1227 <span class="comment">%                     sim.applyOptions('seed', 100 * i + j);</span>
1228 <span class="comment">%                     sim.initializeState();</span>
1229 <span class="comment">%                     growths(i, j) = mr.growth;</span>
1230 <span class="comment">%                 end</span>
1231 <span class="comment">%             end</span>
1232 <span class="comment">%</span>
1233 <span class="comment">%             %compare distribution</span>
1234 <span class="comment">%             figHandle = figure();</span>
1235 <span class="comment">%</span>
1236 <span class="comment">%             axesHandle = subplot(1, 1, 1);</span>
1237 <span class="comment">%             edges = linspace(quantile(growths(:), 0.02), quantile(growths(:), 0.98), 10);</span>
1238 <span class="comment">%             cnts = [</span>
1239 <span class="comment">%                 histc(growths(1, :), edges)</span>
1240 <span class="comment">%                 histc(growths(2, :), edges)</span>
1241 <span class="comment">%                 ];</span>
1242 <span class="comment">%             h = plot(edges, cnts);</span>
1243 <span class="comment">%             legend(h, {'Baseline', 'Optimum'}, 'Location', 'NorthWest');</span>
1244 <span class="comment">%             ylabel('Freq', 'FontSize', 7)</span>
1245 <span class="comment">%             xlabel('Growth (cell s^{-1})', 'FontSize', 7);</span>
1246 <span class="comment">%             set(axesHandle, 'FontSize', 5, 'tickDir', 'out', 'box', 'off');</span>
1247 <span class="comment">%</span>
1248 <span class="comment">%             saveas(figHandle, [outDir 'growthDistributions.pdf']);</span>
1249 <span class="comment">%             close(figHandle);</span>
1250             
1251             <span class="comment">%% metabolic / non-metabolic tradeoff</span>
1252             <span class="comment">%max growth supported by various total metabolic enzyme masses</span>
1253             totMetEnzMass = linspace(<span class="keyword">...</span>
1254                 ((kineticLimitedEnzymes &amp; kineticUnlimitedEnzymes) + (~kineticLimitedEnzymes) .* enzymes)' * met.enzymeMolecularWeights, <span class="keyword">...</span>
1255                 mass.cellInitialDryWeight * mass.dryWeightFractionProtein * ConstantUtil.nAvogadro, <span class="keyword">...</span>
1256                 20);
1257             metEnzGrowths = zeros(numel(totMetEnzMass), 1);            
1258             <span class="keyword">for</span> i = 1:numel(totMetEnzMass)
1259                 [result, ~, ~, errFlag, errMsg] = ComputationUtil.linearProgramming(<span class="keyword">...</span>
1260                     <span class="string">'maximize'</span>, [fbaObj; zeros(nEnz, 1)], [fbaSMat zeros(nMet, nEnz); zeros(1, nRxn) met.enzymeMolecularWeights'; kineticConstraints -enzymeConstraints], <span class="keyword">...</span>
1261                     [fbaRHS; totMetEnzMass(i); zeros(nKinEnz, 1)], <span class="keyword">...</span>
1262                     [nonKineticFluxBounds(:, 1); (kineticLimitedEnzymes &amp; kineticUnlimitedEnzymes) + (~kineticLimitedEnzymes) .* enzymes], <span class="keyword">...</span>
1263                     [nonKineticFluxBounds(:, 2); inf(nEnz, 1)], <span class="keyword">...</span>
1264                     [repmat(<span class="string">'S'</span>, size(fbaRHS)); <span class="string">'S'</span>; repmat(<span class="string">'U'</span>, nKinEnz, 1)], <span class="string">'C'</span>, met.linearProgrammingOptions);
1265                 <span class="keyword">if</span> errFlag
1266                     metEnzGrowths(i, 1) = NaN;
1267                 <span class="keyword">else</span>
1268                     metEnzGrowths(i, 1) = result(bmRxnIdx);
1269                 <span class="keyword">end</span>
1270             <span class="keyword">end</span>
1271             
1272             <span class="comment">%max growth supported by various total non-metabolic enzyme masses</span>
1273             nonMetProts = ~any(enzGeneComp(g.mRNAIndexs, :), 2);
1274             [~, ~, ~, ~, ~, minMonCnt, ~, ~] = fitter.calcResourceRequirements(fitter.constructParameterVectorFromSimulation());
1275             totNonMetProtMass = minMonCnt(nonMetProts)' * pm.molecularWeights(pm.matureIndexs(nonMetProts));
1276             
1277             <span class="comment">%max growth supported by media</span>
1278             tmpFluxBounds = met.calcFluxBounds(<span class="keyword">...</span>
1279                 substrates, enzymes, met.fbaReactionBounds, met.fbaEnzymeBounds, <span class="keyword">...</span>
1280                 false, true, true, true, true, true);
1281             [result, ~, ~, errFlag, errMsg] = ComputationUtil.linearProgramming(<span class="keyword">...</span>
1282                 <span class="string">'maximize'</span>, met.fbaObjective, met.fbaReactionStoichiometryMatrix, met.fbaRightHandSide, <span class="keyword">...</span>
1283                 tmpFluxBounds(:, 1), tmpFluxBounds(:, 2), <span class="keyword">...</span>
1284                 <span class="string">'S'</span>, <span class="string">'C'</span>, met.linearProgrammingOptions);
1285             <span class="keyword">if</span> errFlag
1286                 throw(MException(<span class="string">'Metabolism:error'</span>, <span class="string">'Unable to optimize growth: %s'</span>, errMsg));
1287             <span class="keyword">end</span>
1288             mediaGrowth = result(met.fbaReactionIndexs_biomassProduction);
1289             
1290             <span class="comment">%plot</span>
1291             figHandle = figure();
1292             
1293             axesHandle = subplot(1, 1, 1);
1294             h  = [
1295                 plot(totMetEnzMass, metEnzGrowths, <span class="string">'Color'</span>, <span class="string">'b'</span>)
1296                 line([0 totNonMetProtMass totNonMetProtMass totMetEnzMass(end)], [0 0 metEnzGrowths(end) metEnzGrowths(end)], <span class="string">'Color'</span>, <span class="string">'g'</span>)
1297                 line(xlim(axesHandle), mediaGrowth * [1 1], <span class="string">'Color'</span>, <span class="string">'r'</span>, <span class="string">'LineStyle'</span>, <span class="string">':'</span>, <span class="string">'Parent'</span>, axesHandle)
1298                 ];
1299             xlabel(axesHandle, <span class="string">'Tot Metabolic Enzyme Mass (Da)'</span>, <span class="string">'FontSize'</span>, 7);
1300             ylabel(axesHandle, <span class="string">'Growth (cell s^{-1}'</span>, <span class="string">'FontSize'</span>, 7);
1301             xlim(axesHandle, [0 totMetEnzMass(end)]);
1302             ylim(axesHandle, [0 metEnzGrowths(end)]);
1303             legend(h, {<span class="string">'Metabolism'</span>, <span class="string">'Non-metabolism'</span>, <span class="string">'Media'</span>});
1304             set(axesHandle, <span class="string">'FontSize'</span>, 5, <span class="string">'tickdir'</span>, <span class="string">'out'</span>, <span class="string">'box'</span>, <span class="string">'off'</span>);
1305             
1306             saveas(figHandle, [outDir <span class="string">'metabolism-nonmetabolism tradeoff.pdf'</span>]);
1307             close(figHandle);
1308             
1309             <span class="comment">%% two gene tradeoff</span>
1310             <span class="comment">%calculate</span>
1311             essMetGenes = g.getIndexs({
1312                 <span class="string">'MG_006'</span>; <span class="string">'MG_013'</span>; <span class="string">'MG_023'</span>; <span class="string">'MG_034'</span>; <span class="string">'MG_037'</span>; <span class="string">'MG_038'</span>; <span class="string">'MG_041'</span>; 
1313                 <span class="string">'MG_042'</span>; <span class="string">'MG_043'</span>; <span class="string">'MG_044'</span>; <span class="string">'MG_045'</span>; <span class="string">'MG_047'</span>;
1314                 <span class="string">'MG_053'</span>; <span class="string">'MG_058'</span>; <span class="string">'MG_066'</span>; <span class="string">'MG_069'</span>; <span class="string">'MG_071'</span>; 
1315                 <span class="string">'MG_077'</span>; <span class="string">'MG_078'</span>; <span class="string">'MG_079'</span>; <span class="string">'MG_080'</span>; <span class="string">'MG_102'</span>; <span class="string">'MG_107'</span>; 
1316                 <span class="string">'MG_111'</span>; <span class="string">'MG_112'</span>; <span class="string">'MG_114'</span>; <span class="string">'MG_118'</span>; <span class="string">'MG_124'</span>; <span class="string">'MG_128'</span>; 
1317                 <span class="string">'MG_137'</span>; <span class="string">'MG_145'</span>; <span class="string">'MG_171'</span>; <span class="string">'MG_179'</span>; <span class="string">'MG_180'</span>; <span class="string">'MG_181'</span>; 
1318                 <span class="string">'MG_212'</span>; <span class="string">'MG_215'</span>; <span class="string">'MG_216'</span>; <span class="string">'MG_228'</span>; <span class="string">'MG_229'</span>; <span class="string">'MG_230'</span>; 
1319                 <span class="string">'MG_231'</span>; <span class="string">'MG_245'</span>; <span class="string">'MG_270'</span>; <span class="string">'MG_271'</span>; <span class="string">'MG_272'</span>; <span class="string">'MG_273'</span>; 
1320                 <span class="string">'MG_274'</span>; <span class="string">'MG_275'</span>; <span class="string">'MG_276'</span>; <span class="string">'MG_278'</span>; <span class="string">'MG_287'</span>; <span class="string">'MG_299'</span>; 
1321                 <span class="string">'MG_300'</span>; <span class="string">'MG_301'</span>; <span class="string">'MG_302'</span>; <span class="string">'MG_303'</span>; <span class="string">'MG_304'</span>; <span class="string">'MG_321'</span>; 
1322                 <span class="string">'MG_322'</span>; <span class="string">'MG_323'</span>; <span class="string">'MG_330'</span>; <span class="string">'MG_342'</span>; <span class="string">'MG_351'</span>; <span class="string">'MG_357'</span>; 
1323                 <span class="string">'MG_368'</span>; <span class="string">'MG_382'</span>; <span class="string">'MG_383'</span>; <span class="string">'MG_394'</span>; <span class="string">'MG_396'</span>; <span class="string">'MG_407'</span>; 
1324                 <span class="string">'MG_429'</span>; <span class="string">'MG_430'</span>; <span class="string">'MG_431'</span>; <span class="string">'MG_434'</span>; <span class="string">'MG_437'</span>; <span class="string">'MG_453'</span>; 
1325                 <span class="string">'MG_458'</span>; <span class="string">'MG_517'</span>});
1326             possEnz = find(any(enzGeneComp(essMetGenes, :), 1)' &amp; kineticLimitedEnzymes);
1327             
1328             selectedEnzIdxs = find(any(enzGeneComp(g.getIndexs({<span class="string">'MG_107'</span>; <span class="string">'MG_111'</span>}), :), 1));
1329             assert(all(ismember(selectedEnzIdxs, possEnz)))
1330             selectedEnz = false(nEnz, 1);
1331             selectedEnz(selectedEnzIdxs) = true;
1332             selectedEnz1 = false(nEnz, 1);
1333             selectedEnz1(selectedEnzIdxs(1)) = true;
1334             
1335             nSteps = 20;
1336             totEnzMass = linspace(0, 0.2 * enzymes(selectedEnz)' * met.enzymeMolecularWeights(selectedEnz), nSteps);
1337             optEnzExp = zeros(nSteps, 2);
1338             optGrowth = zeros(nSteps, 2);
1339             enzExpConstantGrowth = zeros(nSteps, 2, nSteps);
1340             <span class="keyword">for</span> i = 1:nSteps
1341                 [result, ~, ~, errFlag, errMsg] = ComputationUtil.linearProgramming(<span class="keyword">...</span>
1342                     <span class="string">'maximize'</span>, [fbaObj; zeros(nEnz, 1)], <span class="keyword">...</span>
1343                     [fbaSMat zeros(nMet, nEnz); zeros(1, nRxn) selectedEnz'.*met.enzymeMolecularWeights'; kineticConstraints -enzymeConstraints], <span class="keyword">...</span>
1344                     [fbaRHS; totEnzMass(i); zeros(nKinEnz, 1)], <span class="keyword">...</span>
1345                     [nonKineticFluxBounds(:, 1); ~selectedEnz .* enzymes], <span class="keyword">...</span>
1346                     [nonKineticFluxBounds(:, 2); max(~selectedEnz .* enzymes, selectedEnz .* inf(nEnz, 1))], <span class="keyword">...</span>
1347                     [repmat(<span class="string">'S'</span>, size(fbaRHS)); <span class="string">'S'</span>; repmat(<span class="string">'U'</span>, nKinEnz, 1)], <span class="string">'C'</span>, met.linearProgrammingOptions);
1348                 <span class="keyword">if</span> errFlag
1349                     throw(MException(<span class="string">'Metabolism:error'</span>, <span class="string">'Unable to optimize growth: %s'</span>, errMsg));
1350                 <span class="keyword">end</span>
1351                 optEnzExp(i, :) = result(end - nEnz + selectedEnzIdxs);
1352                 optGrowth(i, 1) = result(bmRxnIdx);
1353 
1354                 tmpNonKineticFluxBounds = nonKineticFluxBounds;
1355                 tmpNonKineticFluxBounds(bmRxnIdx, :) = optGrowth(i, 1);
1356                 totEnzMass2 = linspace(totEnzMass(i), 2 * totEnzMass(end), nSteps);
1357                 <span class="keyword">for</span> j = 1:nSteps/2
1358                     [result, ~, ~, errFlag, errMsg] = ComputationUtil.linearProgramming(<span class="keyword">...</span>
1359                         <span class="string">'minimize'</span>, [zeros(nRxn, 1); selectedEnz1], <span class="keyword">...</span>
1360                         [fbaSMat zeros(nMet, nEnz); zeros(1, nRxn) selectedEnz'.*met.enzymeMolecularWeights'; kineticConstraints -enzymeConstraints], <span class="keyword">...</span>
1361                         [fbaRHS; totEnzMass2(j); zeros(nKinEnz, 1)], <span class="keyword">...</span>
1362                         [tmpNonKineticFluxBounds(:, 1); ~selectedEnz .* enzymes], <span class="keyword">...</span>
1363                         [tmpNonKineticFluxBounds(:, 2); max(~selectedEnz .* enzymes, selectedEnz .* inf(nEnz, 1))], <span class="keyword">...</span>
1364                         [repmat(<span class="string">'S'</span>, size(fbaRHS)); <span class="string">'S'</span>; repmat(<span class="string">'U'</span>, nKinEnz, 1)], <span class="string">'C'</span>, met.linearProgrammingOptions);
1365                     <span class="keyword">if</span> errFlag
1366                         throw(MException(<span class="string">'Metabolism:error'</span>, <span class="string">'Unable to optimize growth: %s'</span>, errMsg));
1367                     <span class="keyword">end</span>
1368                     enzExpConstantGrowth(i, :, nSteps/2 - j + 1) = result(end - nEnz + selectedEnzIdxs);
1369 
1370                     [result, ~, ~, errFlag, errMsg] = ComputationUtil.linearProgramming(<span class="keyword">...</span>
1371                         <span class="string">'maximize'</span>, [zeros(nRxn, 1); selectedEnz1], <span class="keyword">...</span>
1372                         [fbaSMat zeros(nMet, nEnz); zeros(1, nRxn) selectedEnz'.*met.enzymeMolecularWeights'; kineticConstraints -enzymeConstraints], <span class="keyword">...</span>
1373                         [fbaRHS; totEnzMass2(j); zeros(nKinEnz, 1)], <span class="keyword">...</span>
1374                         [tmpNonKineticFluxBounds(:, 1); ~selectedEnz .* enzymes], <span class="keyword">...</span>
1375                         [tmpNonKineticFluxBounds(:, 2); max(~selectedEnz .* enzymes, selectedEnz .* inf(nEnz, 1))], <span class="keyword">...</span>
1376                         [repmat(<span class="string">'S'</span>, size(fbaRHS)); <span class="string">'S'</span>; repmat(<span class="string">'U'</span>, nKinEnz, 1)], <span class="string">'C'</span>, met.linearProgrammingOptions);
1377                     <span class="keyword">if</span> errFlag
1378                         throw(MException(<span class="string">'Metabolism:error'</span>, <span class="string">'Unable to optimize growth: %s'</span>, errMsg));
1379                     <span class="keyword">end</span>
1380                     enzExpConstantGrowth(i, :, nSteps/2 + j) = result(end - nEnz + selectedEnzIdxs);
1381                 <span class="keyword">end</span>
1382             <span class="keyword">end</span>
1383             
1384             <span class="comment">%plot</span>
1385             figHandle = figure();
1386             
1387             axesHandle = subplot(1, 1, 1);
1388             hold(axesHandle, <span class="string">'on'</span>);
1389             h  = zeros(3, 1);
1390             h(1) = plot(axesHandle, optEnzExp(:, 1), optEnzExp(:, 2), <span class="string">'r'</span>);
1391             h(2) = plot(axesHandle, <span class="keyword">...</span>
1392                 reshape([permute(enzExpConstantGrowth(:, 1, :), [1 3 2]) NaN(nSteps, 1)]', 1, []), <span class="keyword">...</span>
1393                 reshape([permute(enzExpConstantGrowth(:, 2, :), [1 3 2]) NaN(nSteps, 1)]', 1, []), <span class="keyword">...</span>
1394                 <span class="string">'Color'</span>, <span class="string">'b'</span>);
1395             h(3) = plot(axesHandle, <span class="keyword">...</span>
1396                 reshape([zeros(nSteps, 1)  2 * totEnzMass' / met.enzymeMolecularWeights(selectedEnzIdxs(1))  NaN(nSteps, 1)]', 1, []), <span class="keyword">...</span>
1397                 reshape([2 * totEnzMass' / met.enzymeMolecularWeights(selectedEnzIdxs(2))  zeros(nSteps, 1)  NaN(nSteps, 1)]', 1, []), <span class="keyword">...</span>
1398                 <span class="string">'Color'</span>, [0.75 0.75 0.75]);
1399             uistack(h(2), <span class="string">'top'</span>);
1400             uistack(h(1), <span class="string">'top'</span>);
1401             xlabel(axesHandle, sprintf(<span class="string">'Expression %s'</span>, met.enzymeNames{selectedEnzIdxs(1)}), <span class="string">'FontSize'</span>, 7);
1402             ylabel(axesHandle, sprintf(<span class="string">'Expression %s'</span>, met.enzymeNames{selectedEnzIdxs(2)}), <span class="string">'FontSize'</span>, 7);
1403             xlim(axesHandle, [0 totEnzMass(end) / met.enzymeMolecularWeights(selectedEnzIdxs(1))]);
1404             ylim(axesHandle, [0 totEnzMass(end) / met.enzymeMolecularWeights(selectedEnzIdxs(2))]);
1405             legend(h, {<span class="string">'Optima'</span>, <span class="string">'Constant Growth'</span>, <span class="string">'Constant Mass'</span>});
1406             set(axesHandle, <span class="string">'FontSize'</span>, 5, <span class="string">'tickdir'</span>, <span class="string">'out'</span>, <span class="string">'box'</span>, <span class="string">'off'</span>);
1407             
1408             saveas(figHandle, [outDir <span class="string">'enzyme A-B tradeoff.pdf'</span>]);
1409             close(figHandle);                        
1410             
1411             <span class="comment">%% figure</span>
1412             figW = 8.5; figH = 10; <span class="comment">% Cell 1 column</span>
1413             [~, figHandle] = PlotUtil.newAxesHandle([figW figH]);
1414             clf(figHandle);
1415             
1416             <span class="comment">%%A</span>
1417             PlotUtil.labelSubFigure(<span class="string">'A'</span>, [leftColLabelX 0.993 -0.0286 -0.0208], figHandle, 8, <span class="string">'normalized'</span>);
1418             axesHandle = subplot(<span class="string">'Position'</span>, [leftColX 0.745 leftColW 0.23], <span class="string">'Parent'</span>, figHandle);
1419             
1420             <span class="comment">%% standarize tick lengths</span>
1421             axesHandles = findobj(figHandle, <span class="string">'type'</span>, <span class="string">'axes'</span>, <span class="string">'-and'</span>, <span class="string">'tag'</span>, <span class="string">''</span>);
1422             axesPos = cell2mat(get(axesHandles, <span class="string">'position'</span>));
1423             tickLen = 0.01 * max(max(axesPos(:, 3:4), [], 1) .* [figW figH]);
1424             <span class="keyword">for</span> i = 1:numel(axesHandles)
1425                 maxVal = max(axesPos(i, 3:4) .* [figW figH]);
1426                 set(axesHandles(i), <span class="string">'TickLen'</span>, [tickLen / maxVal  0.0250]);
1427             <span class="keyword">end</span>
1428             
1429             <span class="comment">%% save</span>
1430             print(figHandle, [outDirectory filesep <span class="string">'growthSensitivity.pdf'</span>], <span class="string">'-dpdf'</span>, <span class="string">'-rgb'</span>);
1431             close(figHandle);
1432         <span class="keyword">end</span>
1433         
1434         <a name="_sub6" href="#_subfunctions" class="code">function [optGrowths, optEnzymes, incEnzymes] = calcOptimalEnzymeExpressionChanges(fbaObj, fbaSMat, fbaRHS, </a><span class="keyword">...</span>
1435                 kineticConstraints, enzymeConstraints, nonKineticFluxBounds, <span class="keyword">...</span>
1436                 enzymes, kineticLimitedEnzymes, kineticUnlimitedEnzymes, bmRxnIdx, sim)
1437             import edu.stanford.covert.util.ComputationUtil;
1438             
1439             met = sim.process(<span class="string">'Metabolism'</span>);
1440             nMet = size(fbaSMat, 1);
1441             nRxn = size(fbaSMat, 2);
1442             nEnz = numel(kineticLimitedEnzymes);
1443             nKinEnz = sum(kineticLimitedEnzymes);
1444             Emax = (enzymes' * met.enzymeMolecularWeights) ./ met.enzymeMolecularWeights;
1445             
1446             [result, ~, ~, errFlag, errMsg] = ComputationUtil.linearProgramming(<span class="keyword">...</span>
1447                 <span class="string">'maximize'</span>, [fbaObj; zeros(nEnz, 1)], [fbaSMat zeros(nMet, nEnz); zeros(1, nRxn) met.enzymeMolecularWeights'; kineticConstraints -enzymeConstraints], <span class="keyword">...</span>
1448                 [fbaRHS; enzymes' * met.enzymeMolecularWeights; zeros(nKinEnz, 1)], <span class="keyword">...</span>
1449                 [nonKineticFluxBounds(:, 1); enzymes], [nonKineticFluxBounds(:, 2); inf(nEnz, 1)], <span class="keyword">...</span>
1450                 [repmat(<span class="string">'S'</span>, size(fbaRHS)); <span class="string">'S'</span>; repmat(<span class="string">'U'</span>, nKinEnz, 1)], <span class="string">'C'</span>, met.linearProgrammingOptions);
1451             <span class="keyword">if</span> errFlag
1452                 throw(MException(<span class="string">'Metabolism:error'</span>, <span class="string">'Unable to optimize growth: %s'</span>, errMsg));
1453             <span class="keyword">end</span>
1454             initGrowth = result(bmRxnIdx);
1455             
1456             [result, ~, ~, errFlag, errMsg] = ComputationUtil.linearProgramming(<span class="keyword">...</span>
1457                 <span class="string">'maximize'</span>, [fbaObj; zeros(nEnz, 1)], [fbaSMat zeros(nMet, nEnz); zeros(1, nRxn) met.enzymeMolecularWeights'; kineticConstraints -enzymeConstraints], <span class="keyword">...</span>
1458                 [fbaRHS; enzymes' * met.enzymeMolecularWeights; zeros(nKinEnz, 1)], <span class="keyword">...</span>
1459                 [nonKineticFluxBounds(:, 1); (kineticLimitedEnzymes &amp; kineticUnlimitedEnzymes) + (~kineticLimitedEnzymes) .* enzymes], [nonKineticFluxBounds(:, 2); inf(nEnz, 1)], <span class="keyword">...</span>
1460                 [repmat(<span class="string">'S'</span>, size(fbaRHS)); <span class="string">'S'</span>; repmat(<span class="string">'U'</span>, nKinEnz, 1)], <span class="string">'C'</span>, met.linearProgrammingOptions);
1461             <span class="keyword">if</span> errFlag
1462                 throw(MException(<span class="string">'Metabolism:error'</span>, <span class="string">'Unable to optimize growth: %s'</span>, errMsg));
1463             <span class="keyword">end</span>
1464             maxGrowth = result(bmRxnIdx);
1465             
1466             tmpNonKineticFluxBounds = nonKineticFluxBounds;
1467             tmpNonKineticFluxBounds(bmRxnIdx) = maxGrowth;
1468             [result, ~, ~, errFlag, errMsg] = ComputationUtil.linearProgramming(<span class="keyword">...</span>
1469                 <span class="string">'minimize'</span>, [
1470                 zeros(nRxn, 1)
1471                 zeros(nEnz, 1)
1472                 ones(nEnz, 1)
1473                 ], [
1474                 fbaSMat             zeros(nMet, nEnz)             zeros(nMet, nEnz)
1475                 zeros(1, nRxn)      met.enzymeMolecularWeights'   zeros(1, nEnz)
1476                 kineticConstraints  -enzymeConstraints            zeros(nKinEnz, nEnz)
1477                 zeros(nEnz, nRxn)   eye(nEnz)                     -diag(Emax - enzymes)
1478                 zeros(1, nRxn)      zeros(1, nEnz)                ones(1, nEnz)
1479                 ], [
1480                 fbaRHS
1481                 enzymes' * met.enzymeMolecularWeights
1482                 zeros(nKinEnz, 1)
1483                 enzymes
1484                 nEnz
1485                 ], [
1486                 tmpNonKineticFluxBounds(:, 1)
1487                 (kineticLimitedEnzymes &amp; kineticUnlimitedEnzymes) + (~kineticLimitedEnzymes) .* enzymes
1488                 zeros(nEnz, 1)
1489                 ], [
1490                 tmpNonKineticFluxBounds(:, 2)
1491                 inf(nEnz, 1)
1492                 ones(nEnz, 1)
1493                 ], [
1494                 repmat(<span class="string">'S'</span>, size(fbaRHS))
1495                 <span class="string">'S'</span>
1496                 repmat(<span class="string">'U'</span>, nKinEnz, 1)
1497                 repmat(<span class="string">'U'</span>, nEnz, 1)
1498                 <span class="string">'U'</span>
1499                 ], [
1500                 repmat(<span class="string">'C'</span>, nRxn, 1)
1501                 repmat(<span class="string">'C'</span>, nEnz, 1)
1502                 repmat(<span class="string">'I'</span>, nEnz, 1)
1503                 ], met.linearProgrammingOptions);
1504             <span class="keyword">if</span> errFlag
1505                 throw(MException(<span class="string">'Metabolism:error'</span>, <span class="string">'Unable to optimize growth: %s'</span>, errMsg));
1506             <span class="keyword">end</span>
1507             assertElementsAlmostEqual(maxGrowth, result(bmRxnIdx));
1508             optEnzyme = result((end-nEnz+1:end) - nEnz);
1509             
1510             maxModEnz = sum(result(end-nEnz+1:end));
1511             optGrowths = zeros(nEnz + 1, 2, 1);
1512             optEnzymes = zeros(nEnz + 1, 2, nEnz);
1513             incEnzymes = cell(nEnz + 1, 2);
1514             optGrowths(1, :) = initGrowth;
1515             optEnzymes(1, 1, :) = enzymes;
1516             optEnzymes(1, 2, :) = enzymes;
1517             nIncreases = 1;
1518             <span class="keyword">for</span> i = 1:nEnz
1519                 [result, ~, ~, errFlag, errMsg] = ComputationUtil.linearProgramming(<span class="keyword">...</span>
1520                     <span class="string">'maximize'</span>, [
1521                     fbaObj
1522                     zeros(nEnz, 1)
1523                     zeros(nEnz, 1)
1524                     ], [
1525                     fbaSMat             zeros(nMet, nEnz)             zeros(nMet, nEnz)
1526                     zeros(1, nRxn)      met.enzymeMolecularWeights'   zeros(1, nEnz)
1527                     kineticConstraints  -enzymeConstraints            zeros(nKinEnz, nEnz)
1528                     zeros(nEnz, nRxn)   eye(nEnz)                     -diag(Emax - enzymes)
1529                     zeros(1, nRxn)      zeros(1, nEnz)                ones(1, nEnz)
1530                     ], [
1531                     fbaRHS
1532                     enzymes' * met.enzymeMolecularWeights
1533                     zeros(nKinEnz, 1)
1534                     enzymes
1535                     i
1536                     ], [
1537                     nonKineticFluxBounds(:, 1)
1538                     (kineticLimitedEnzymes &amp; kineticUnlimitedEnzymes) + (~kineticLimitedEnzymes) .* enzymes
1539                     zeros(nEnz, 1)
1540                     ], [
1541                     nonKineticFluxBounds(:, 2)
1542                     inf(nEnz, 1)
1543                     ones(nEnz, 1)
1544                     ], [
1545                     repmat(<span class="string">'S'</span>, size(fbaRHS))
1546                     <span class="string">'S'</span>
1547                     repmat(<span class="string">'U'</span>, nKinEnz, 1)
1548                     repmat(<span class="string">'U'</span>, nEnz, 1)
1549                     <span class="string">'U'</span>
1550                     ], [
1551                     repmat(<span class="string">'C'</span>, nRxn, 1)
1552                     repmat(<span class="string">'C'</span>, nEnz, 1)
1553                     repmat(<span class="string">'I'</span>, nEnz, 1)
1554                     ], met.linearProgrammingOptions);
1555                 <span class="keyword">if</span> errFlag
1556                     throw(MException(<span class="string">'Metabolism:error'</span>, <span class="string">'Unable to optimize growth: %s'</span>, errMsg));
1557                 <span class="keyword">end</span>
1558                 optGrowths(i+1, 1) = result(bmRxnIdx);
1559                 optEnzymes(i+1, 1, :) = result((end-nEnz+1:end) - nEnz);
1560                 incEnzymes{i+1, 1} = find(squeeze(optEnzymes(i+1, 1, :)) &gt; enzymes);
1561                 
1562                 [result, ~, ~, errFlag, errMsg] = ComputationUtil.linearProgramming(<span class="keyword">...</span>
1563                     <span class="string">'maximize'</span>, [
1564                     fbaObj
1565                     zeros(nEnz, 1)
1566                     zeros(nEnz, 1)
1567                     ], [
1568                     fbaSMat             zeros(nMet, nEnz)             zeros(nMet, nEnz)
1569                     zeros(1, nRxn)      met.enzymeMolecularWeights'   zeros(1, nEnz)
1570                     kineticConstraints  -enzymeConstraints            zeros(nKinEnz, nEnz)
1571                     zeros(nEnz, nRxn)   eye(nEnz)                     -diag(Emax - squeeze(optEnzymes(i, 2, :)))
1572                     zeros(1, nRxn)      zeros(1, nEnz)                ones(1, nEnz)
1573                     ], [
1574                     fbaRHS
1575                     enzymes' * met.enzymeMolecularWeights
1576                     zeros(nKinEnz, 1)
1577                     squeeze(optEnzymes(i, 2, :))
1578                     nIncreases
1579                     ], [
1580                     nonKineticFluxBounds(:, 1)
1581                     (kineticLimitedEnzymes &amp; kineticUnlimitedEnzymes) + (~kineticLimitedEnzymes) .* enzymes
1582                     zeros(nEnz, 1)
1583                     ], [
1584                     nonKineticFluxBounds(:, 2)
1585                     inf(nEnz, 1)
1586                     ones(nEnz, 1)
1587                     ], [
1588                     repmat(<span class="string">'S'</span>, size(fbaRHS))
1589                     <span class="string">'S'</span>
1590                     repmat(<span class="string">'U'</span>, nKinEnz, 1)
1591                     repmat(<span class="string">'U'</span>, nEnz, 1)
1592                     <span class="string">'U'</span>
1593                     ], [
1594                     repmat(<span class="string">'C'</span>, nRxn, 1)
1595                     repmat(<span class="string">'C'</span>, nEnz, 1)
1596                     repmat(<span class="string">'I'</span>, nEnz, 1)
1597                     ], met.linearProgrammingOptions);
1598                 <span class="keyword">if</span> errFlag
1599                     throw(MException(<span class="string">'Metabolism:error'</span>, <span class="string">'Unable to optimize growth: %s'</span>, errMsg));
1600                 <span class="keyword">end</span>
1601                 <span class="keyword">if</span> result(bmRxnIdx) - optGrowths(i, 2) &gt; 1e-6 * maxGrowth
1602                     optGrowths(i+1, 2) = result(bmRxnIdx);
1603                     optEnzymes(i+1, 2, :) = result((end-nEnz+1:end) - nEnz);
1604                     incEnzymes{i+1, 2} = find(optEnzymes(i+1, 2, :) &gt; optEnzymes(i, 2, :));
1605                     nIncreases = 1;
1606                 <span class="keyword">else</span>
1607                     optGrowths(i+1, 2) = optGrowths(i, 2);
1608                     optEnzymes(i+1, 2, :) = optEnzymes(i, 2, :);
1609                     incEnzymes{i+1, 2} = [];
1610                     nIncreases = nIncreases + 1;
1611                 <span class="keyword">end</span>
1612             <span class="keyword">end</span>
1613             optGrowths(maxModEnz+1:<span class="keyword">end</span>, 1) = maxGrowth;
1614             optEnzymes(maxModEnz+1:<span class="keyword">end</span>, 1, :) = repmat(permute(optEnzyme, [3 2 1]), nEnz - maxModEnz + 1, 1);
1615             
1616             idx = find(any(diff(optGrowths, 1, 1) &gt; 1e-6 * maxGrowth, 2), 1, <span class="string">'last'</span>) + 1;
1617             optGrowths = optGrowths(1:idx, :);
1618             optEnzymes = optEnzymes(1:idx, :, :);
1619             incEnzymes = incEnzymes(1:idx, :);
1620         <span class="keyword">end</span>
1621         
1622         <a name="_sub7" href="#_subfunctions" class="code">function RNAvProtein(simBatchDir, outDirectory, figHandle, position)</a>
1623             import edu.stanford.covert.cell.sim.util.SimulationEnsemble;
1624             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
1625             import edu.stanford.covert.cell.sim.util.PlotUtil;
1626             import edu.stanford.covert.cell.sim.util.MacromoleculeUtil;
1627             
1628             <span class="keyword">if</span> ~exist(<span class="string">'simBatchDir'</span>, <span class="string">'var'</span>) || isempty(simBatchDir)
1629                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
1630             <span class="keyword">end</span>
1631             selectedSimulations = 1:SimulationDiskUtil.getNumSimulations(simBatchDir);
1632             
1633             <span class="comment">%% get data</span>
1634             <span class="keyword">if</span> exist([outDirectory <span class="string">'RNAvProtein.mat'</span>], <span class="string">'file'</span>)
1635                 load([outDirectory <span class="string">'RNAvProtein.mat'</span>]);
1636             <span class="keyword">else</span>
1637                 [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep <span class="string">'1'</span>]);
1638                 g = sim.gene;
1639                 pm = sim.state(<span class="string">'ProteinMonomer'</span>);
1640                 pc = sim.state(<span class="string">'ProteinComplex'</span>);
1641                 rna = sim.state(<span class="string">'Rna'</span>);
1642                 
1643                 geneId = <span class="string">'MG_218'</span>;
1644                 [~, ~, ~, ~, ~, <span class="keyword">...</span>
1645                     ~, ~, ~, ~, <span class="keyword">...</span>
1646                     ~, ~, ~, ~, <span class="keyword">...</span>
1647                     matureRnaIdxs, ~, ~, ~, <span class="keyword">...</span>
1648                     monomerIdxs, ~, ~, ~, <span class="keyword">...</span>
1649                     ~, ~, ~, ~] = <span class="keyword">...</span>
1650                     MacromoleculeUtil.getMacromoleculeIndexsIDsNames(geneId, sim);
1651                 
1652                 stateNames = {
1653                     <span class="string">'Rna'</span>               <span class="string">'counts'</span> rna.matureIndexs(matureRnaIdxs)         <span class="string">'-sum'</span>
1654                     <span class="string">'ProteinMonomer'</span>    <span class="string">'counts'</span> pm.matureIndexs(monomerIdxs)            <span class="string">'-sum'</span>
1655                     };
1656                 cr = zeros(numel(selectedSimulations), 1);
1657                 cp = zeros(numel(selectedSimulations), 1);
1658                 <span class="keyword">for</span> i = 1:numel(selectedSimulations)
1659                     states = SimulationEnsemble.load(simBatchDir, {<span class="string">'Time'</span> <span class="string">'values'</span>}, [], [], 1, <span class="string">'extract'</span>, selectedSimulations(i));
1660                     
1661                     tIdx = randi(numel(states.Time.values), 1);
1662                     states = SimulationEnsemble.load(simBatchDir, stateNames, tIdx, tIdx, 1, <span class="string">'extract'</span>, selectedSimulations(i));
1663                     cr(i) = states.Rna.counts;
1664                     cp(i) = states.ProteinMonomer.counts;
1665                 <span class="keyword">end</span>
1666                 
1667                 save([outDirectory <span class="string">'RNAvProtein.mat'</span>], <span class="string">'cp'</span>, <span class="string">'cr'</span>);
1668             <span class="keyword">end</span>
1669             
1670             <span class="comment">%% plot</span>
1671             <span class="keyword">if</span> nargin &lt; 4
1672                 [~, figHandle] = edu.stanford.covert.cell.sim.util.PlotUtil.newAxesHandle();
1673                 position = [0.1 0.1 0.8 0.8];
1674             <span class="keyword">end</span>
1675             x = position(1);
1676             y = position(2);
1677             w = position(3);
1678             h = position(4);
1679             figPos = get(figHandle, <span class="string">'papersize'</span>);
1680             figW = figPos(1);
1681             figH = figPos(2);
1682             axesHandles = [
1683                 subplot(<span class="string">'Position'</span>, [x, y + h-0.15*w*figW/figH, 0.8*w, 0.15*w*figW/figH], <span class="string">'Parent'</span>, figHandle)
1684                 subplot(<span class="string">'Position'</span>, [x + 0.85*w, y, 0.15*w, h-0.2*w*figW/figH], <span class="string">'Parent'</span>, figHandle)
1685                 subplot(<span class="string">'Position'</span>, [x, y, 0.8*w, h-0.2*w*figW/figH], <span class="string">'Parent'</span>, figHandle)
1686                 ];
1687             
1688             xlims = [min(cr) - 0.5, max(cr) + 0.5];
1689             ylims = [0 60];
1690             
1691             <span class="comment">%RNA marginal</span>
1692             parmhat = gamfit(cr);
1693             
1694             axesHandle = axesHandles(1);
1695             hold(axesHandle, <span class="string">'on'</span>);
1696             [n, edges] = hist(axesHandle, cr, min(cr):max(cr));
1697             n = n / sum(n) * 100;
1698             bar(axesHandle, edges, n, <span class="string">'b'</span>, <span class="string">'EdgeColor'</span>, <span class="string">'none'</span>);
1699             x = linspace(min(cr), max(cr), 20);
1700             plot(axesHandle, x, gampdf(x, parmhat(1), parmhat(2)) * 100, <span class="string">'r'</span>);
1701             ylim(axesHandle, [0 max(n)]);
1702             xlim(axesHandle, xlims);
1703             set(axesHandle, <span class="string">'Box'</span>, <span class="string">'off'</span>);
1704             set(axesHandle, <span class="string">'TickDir'</span>, <span class="string">'out'</span>);
1705             set(axesHandle, <span class="string">'XTickLabel'</span>, <span class="string">''</span>);
1706             set(axesHandle, <span class="string">'XTick'</span>, []);
1707             set(axesHandle, <span class="string">'YTick'</span>, 0:25:50);
1708             
1709             tick2text(axesHandle, <span class="string">'axis'</span>, <span class="string">'y'</span>, <span class="string">'ytickoffset'</span>, 0.04);
1710             yTicks = getappdata(axesHandle, <span class="string">'YTickText'</span>);
1711             set(yTicks, <span class="string">'FontSize'</span>, 5, <span class="string">'HorizontalAlign'</span>, <span class="string">'right'</span>);
1712             delete(yTicks(2));
1713             
1714             ylabel(axesHandle, <span class="string">'Freq'</span>, <span class="string">'FontSize'</span>, 7);
1715             ylabelPos = get(get(axesHandle, <span class="string">'ylabel'</span>), <span class="string">'position'</span>);
1716             set(get(axesHandle, <span class="string">'ylabel'</span>), <span class="string">'position'</span>, [-0.83 25 ylabelPos(3:end)]);
1717             
1718             <span class="comment">%protein marginal</span>
1719             parmhat = gamfit(cp);
1720             
1721             axesHandle = axesHandles(2);
1722             hold(axesHandle, <span class="string">'on'</span>);
1723             [n, xout] = hist(cp, 50);
1724             n = n / sum(n) * 100;
1725             barh(axesHandle, xout, n, <span class="string">'b'</span>, <span class="string">'EdgeColor'</span>, <span class="string">'none'</span>);
1726             x = linspace(0, 60, 100);
1727             plot(axesHandle, gampdf(x, parmhat(1), parmhat(2)) * 100, x, <span class="string">'r'</span>);
1728             xlim(axesHandle, [0 max(n)]);
1729             ylim(axesHandle, ylims);
1730             set(axesHandle, <span class="string">'Box'</span>, <span class="string">'off'</span>);
1731             set(axesHandle, <span class="string">'TickDir'</span>, <span class="string">'out'</span>);
1732             set(axesHandle, <span class="string">'YTickLabel'</span>, []);
1733             set(axesHandle, <span class="string">'YTick'</span>, []);
1734             set(axesHandle, <span class="string">'XTick'</span>, 0:2:6);
1735             
1736             tick2text(axesHandle, <span class="string">'axis'</span>, <span class="string">'x'</span>, <span class="string">'xtickoffset'</span>, 0.062);
1737             xTicks = getappdata(axesHandle, <span class="string">'XTickText'</span>);
1738             set(xTicks, <span class="string">'FontSize'</span>, 5);
1739             delete(xTicks(2:3));
1740             
1741             xlabel(axesHandle, <span class="string">'Freq'</span>, <span class="string">'FontSize'</span>, 7);
1742             xlabelPos = get(get(axesHandle, <span class="string">'xlabel'</span>), <span class="string">'position'</span>);
1743             set(get(axesHandle, <span class="string">'xlabel'</span>), <span class="string">'position'</span>, [xlabelPos(1)-0.3 -6 xlabelPos(3:end)]);
1744             
1745             <span class="comment">%RNA vs protein</span>
1746             axesHandle = axesHandles(3);
1747             plot(axesHandle, cr, cp, <span class="string">'b.'</span>);
1748             xlim(axesHandle, xlims);
1749             ylim(axesHandle, ylims);
1750             set(axesHandle, <span class="string">'XTick'</span>, [0 1 2], <span class="string">'YTick'</span>, 0:25:50);
1751             set(axesHandle, <span class="string">'TickDir'</span>, <span class="string">'out'</span>, <span class="string">'box'</span>, <span class="string">'off'</span>);
1752             
1753             tick2text(axesHandle, <span class="string">'axis'</span>, <span class="string">'xy'</span>, <span class="string">'xtickoffset'</span>, 0.062, <span class="string">'ytickoffset'</span>, 0.04);
1754             xTicks = getappdata(axesHandle, <span class="string">'XTickText'</span>);
1755             yTicks = getappdata(axesHandle, <span class="string">'YTickText'</span>);
1756             set(xTicks, <span class="string">'FontSize'</span>, 5);
1757             set(yTicks, <span class="string">'FontSize'</span>, 5, <span class="string">'HorizontalAlign'</span>, <span class="string">'right'</span>);
1758             delete(yTicks(2));
1759             
1760             xlabel(axesHandle, <span class="string">'mRNA count'</span>, <span class="string">'FontSize'</span>, 7);
1761             ylabel(axesHandle, <span class="string">'Protein count'</span>, <span class="string">'FontSize'</span>, 7);
1762             xlabelPos = get(get(axesHandle, <span class="string">'xlabel'</span>), <span class="string">'position'</span>);
1763             ylabelPos = get(get(axesHandle, <span class="string">'ylabel'</span>), <span class="string">'position'</span>);
1764             set(get(axesHandle, <span class="string">'xlabel'</span>), <span class="string">'position'</span>, [xlabelPos(1) -6 xlabelPos(3:end)]);
1765             set(get(axesHandle, <span class="string">'ylabel'</span>), <span class="string">'position'</span>, [-0.83 ylabelPos(2:end)]);
1766         <span class="keyword">end</span>
1767         
1768         <a name="_sub8" href="#_subfunctions" class="code">function cellShape(outDirectory, simBatchDir, selectedSim)</a>
1769             import edu.stanford.covert.cell.sim.util.SimulationEnsemble;
1770             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
1771             import edu.stanford.covert.cell.sim.analysis.Figures34;
1772             import edu.stanford.covert.cell.sim.state.CellGeometry;
1773             
1774             <span class="keyword">if</span> ~exist(<span class="string">'simBatchDir'</span>, <span class="string">'var'</span>) || isempty(simBatchDir)
1775                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
1776             <span class="keyword">end</span>
1777             <span class="keyword">if</span> ~exist(<span class="string">'selectedSim'</span>, <span class="string">'var'</span>) || isempty(selectedSim)
1778                 selectedSim = 1:SimulationDiskUtil.getNumSimulations(simBatchDir);
1779             <span class="keyword">end</span>
1780             
1781             [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep <span class="string">'1'</span>]);
1782             
1783             stateNames={<span class="string">'Geometry'</span> <span class="string">'width'</span>
1784                 <span class="string">'Geometry'</span> <span class="string">'pinchedDiameter'</span>
1785                 <span class="string">'Chromosome'</span> <span class="string">'polymerizedRegions'</span>
1786                 <span class="string">'Chromosome'</span> <span class="string">'monomerBoundSites'</span>
1787                 <span class="string">'Chromosome'</span> <span class="string">'complexBoundSites'</span>
1788                 <span class="string">'Time'</span> <span class="string">'values'</span>};
1789             
1790             states = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, <span class="string">'extract'</span>, selectedSim);
1791             
1792             log = load([simBatchDir filesep num2str(selectedSim) filesep <span class="string">'summary.mat'</span>]);
1793             
1794             cm = sim.state(<span class="string">'Mass'</span>);
1795             geom = sim.state(<span class="string">'Geometry'</span>);
1796             density = geom.density;
1797             mass = (log.mass(2:end) * cm.cellInitialDryWeight)/(1-cm.fractionWetWeight);
1798             width = squeeze(states.Geometry.width);
1799             pinchedDiameter = squeeze(states.Geometry.pinchedDiameter);
1800             t = squeeze(states.Time.values);
1801             septumLength = (width - pinchedDiameter)/2;
1802             
1803             cylindricalLength = zeros(size(t));
1804             maxTotalLengths = zeros(size(t));
1805             
1806             <span class="keyword">for</span> i=1:length(cylindricalLength)
1807                 <span class="keyword">if</span>(pinchedDiameter(i) &gt; 0)
1808                     [cylindricalLength(i), ~, maxTotalLengths(i)] = CellGeometry.calculateGeometry(<span class="keyword">...</span>
1809                         width(i), <span class="keyword">...</span>
1810                         pinchedDiameter(i), <span class="keyword">...</span>
1811                         mass(i)/density);
1812                 <span class="keyword">end</span>
1813             <span class="keyword">end</span>
1814             
1815             tidxs = [round(1:t(end)/4:t(end)) t(end)-1];
1816             
1817             scale = 1e-9;
1818             
1819             <span class="keyword">for</span> i = 1:numel(tidxs)
1820                 
1821                 svg_width = maxTotalLengths(tidxs(i)) / scale + 10;
1822                 svg_height = max(width) / scale + 10;
1823                 
1824                 fileNameSVG = sprintf(<span class="string">'%s%sCellShape_%d.svg'</span>, outDirectory, filesep, i);
1825                 fileNamePDF = sprintf(<span class="string">'%s%sCellShape_%d.pdf'</span>, outDirectory, filesep, i);
1826                 
1827                 <span class="comment">%open svg</span>
1828                 fid = fopen(fileNameSVG, <span class="string">'w'</span>);
1829                 fprintf(fid, <span class="string">'&lt;?xml version=&quot;1.0&quot; standalone=&quot;no&quot;?&gt;\n'</span>);
1830                 fprintf(fid, <span class="string">'&lt;!DOCTYPE svg PUBLIC &quot;-//W3C//DTD SVG 1.1//EN&quot; &quot;http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd&quot;&gt;\n'</span>);
1831                 fprintf(fid, <span class="string">'&lt;svg width=&quot;%d&quot; height=&quot;%d&quot; version=&quot;1.1&quot; xmlns=&quot;http://www.w3.org/2000/svg&quot; viewBox=&quot;%d %d %d %d&quot;&gt;\n'</span>, svg_width, svg_height, 0, 0, svg_width, svg_height);
1832                 fprintf(fid, <span class="string">'&lt;defs&gt;\n'</span>);
1833                 fprintf(fid, <span class="string">'&lt;linearGradient id=&quot;cellBg&quot; x1=&quot;0%%&quot; y1=&quot;0%%&quot; x2=&quot;0%%&quot; y2=&quot;100%%&quot;&gt;\n'</span>);
1834                 fprintf(fid, <span class="string">'&lt;stop style=&quot;stop-color:#97b0fb;stop-opacity:0.05&quot; offset=&quot;0&quot;/&gt;\n'</span>);
1835                 fprintf(fid, <span class="string">'&lt;stop style=&quot;stop-color:#89b8d6;stop-opacity:1&quot; offset=&quot;1&quot;/&gt;\n'</span>);
1836                 fprintf(fid, <span class="string">'&lt;/linearGradient&gt;'</span>);
1837                 fprintf(fid, <span class="string">'&lt;/defs&gt;\n'</span>);
1838                 
1839                 <span class="comment">% White background</span>
1840                 fprintf(fid, <span class="string">'&lt;rect x=&quot;%d&quot; y=&quot;%d&quot; width=&quot;%d&quot; height=&quot;%d&quot; style=&quot;fill:white; stroke:none&quot;/&gt;\n'</span>, <span class="keyword">...</span>
1841                     0, 0, svg_width, svg_height);
1842                 
1843                 fprintf(fid, Figures34.drawCell(scale,width(tidxs(i)),cylindricalLength(tidxs(i)),septumLength(tidxs(i)),max(maxTotalLengths),max(width)));
1844                 
1845                 <span class="comment">%close svg</span>
1846                 fprintf(fid, <span class="string">'&lt;/svg&gt;\n'</span>);
1847                 fclose(fid);
1848                 
1849                 [status, result] = system(sprintf(<span class="string">'inkscape &quot;%s&quot; --export-pdf=&quot;%s&quot;'</span>, fileNameSVG, fileNamePDF));
1850                 <span class="keyword">if</span> status ~= 0
1851                     throw(MException(<span class="string">'Figures34:error'</span>, <span class="string">'Failed to convert svg to pdf: %s'</span>, result));
1852                 <span class="keyword">end</span>
1853             <span class="keyword">end</span>
1854         <span class="keyword">end</span>
1855         
1856         <a name="_sub9" href="#_subfunctions" class="code">function normCellMass(simBatchDir, selectedSim, outDirectory, figHandle, position)</a>
1857             import edu.stanford.covert.cell.sim.util.SimulationEnsemble;
1858             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
1859             
1860             <span class="keyword">if</span> ~exist(<span class="string">'simBatchDir'</span>, <span class="string">'var'</span>) || isempty(simBatchDir)
1861                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
1862             <span class="keyword">end</span>
1863             <span class="keyword">if</span> ~exist(<span class="string">'selectedSim'</span>, <span class="string">'var'</span>) || isempty(selectedSim)
1864                 selectedSim = 1:SimulationDiskUtil.getNumSimulations(simBatchDir);
1865             <span class="keyword">end</span>
1866             
1867             <span class="comment">%% get constants</span>
1868             [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep <span class="string">'1'</span>]);
1869             comp = sim.compartment;
1870             
1871             <span class="comment">%% get data</span>
1872             <span class="keyword">if</span> exist([outDirectory <span class="string">'normCellMass.mat'</span>], <span class="string">'file'</span>)
1873                 load([outDirectory <span class="string">'normCellMass.mat'</span>])
1874             <span class="keyword">else</span>
1875                 stateNames = {
1876                     <span class="string">'Time'</span> <span class="string">'values'</span>       <span class="string">':'</span> <span class="string">':'</span>
1877                     <span class="string">'Mass'</span> <span class="string">'dnaWt'</span>        <span class="string">':'</span> <span class="string">'-sum'</span>
1878                     <span class="string">'Mass'</span> <span class="string">'rnaWt'</span>        <span class="string">':'</span> <span class="string">'-sum'</span>
1879                     <span class="string">'Mass'</span> <span class="string">'proteinWt'</span>    <span class="string">':'</span> <span class="string">'-sum'</span>
1880                     <span class="string">'Mass'</span> <span class="string">'metaboliteWt'</span> <span class="string">':'</span> [comp.cytosolIndexs; comp.membraneIndexs]
1881                     <span class="string">'Mass'</span> <span class="string">'cellDry'</span>         <span class="string">':'</span> <span class="string">'-sum'</span>
1882                     };
1883                 states = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, <span class="string">'extract'</span>, selectedSim);
1884                 
1885                 proteinWt = permute(states.Mass.proteinWt, [1 3 2]);
1886                 dnaWt = permute(states.Mass.dnaWt, [1 3 2]);
1887                 membraneWt = permute(states.Mass.metaboliteWt(:, 2, :), [1 3 2]);
1888                 rnaWt = permute(states.Mass.rnaWt, [1 3 2]);
1889                 cellWt = permute(states.Mass.cellDry, [1 3 2]);
1890                 time = permute(states.Time.values, [1 3 2]) / 3600;
1891                 
1892                 proteinWt = proteinWt / proteinWt(1);
1893                 dnaWt = dnaWt / dnaWt(1);
1894                 membraneWt = membraneWt / membraneWt(1);
1895                 rnaWt = rnaWt / rnaWt(1);
1896                 cellWt = cellWt / cellWt(1);
1897                 
1898                 save([outDirectory <span class="string">'normCellMass.mat'</span>], <span class="string">'proteinWt'</span>, <span class="string">'dnaWt'</span>, <span class="string">'membraneWt'</span>, <span class="string">'rnaWt'</span>, <span class="string">'cellWt'</span>, <span class="string">'time'</span>);
1899             <span class="keyword">end</span>
1900             
1901             <span class="comment">%% plot</span>
1902             <span class="keyword">if</span> nargin &gt;= 4
1903                 axesHandle = subplot(<span class="string">'position'</span>, position, <span class="string">'parent'</span>, figHandle);
1904             <span class="keyword">else</span>
1905                 axesHandle = edu.stanford.covert.cell.sim.util.PlotUtil.newAxesHandle();
1906             <span class="keyword">end</span>
1907             
1908             lh = plot(axesHandle, time, [cellWt; dnaWt; rnaWt; proteinWt; membraneWt]);
1909             set(lh(2), <span class="string">'LineWidth'</span>, 1, <span class="string">'Color'</span>, <span class="string">'g'</span>);
1910             uistack(lh(4), <span class="string">'top'</span>);
1911             uistack(lh(1), <span class="string">'top'</span>);
1912             uistack(lh(2), <span class="string">'bottom'</span>);
1913             uistack(lh(3), <span class="string">'bottom'</span>);
1914             
1915             xlim(axesHandle, [0 time(end)]);
1916             ylim(axesHandle, [0.92 2.05]);
1917             
1918             set(axesHandle, <span class="string">'Box'</span>, <span class="string">'off'</span>);
1919             set(axesHandle, <span class="string">'TickDir'</span>, <span class="string">'out'</span>);
1920             set(axesHandle, <span class="string">'YTick'</span>, 1:0.5:2);
1921             set(axesHandle, <span class="string">'XTick'</span>, 0:2:8);
1922             set(axesHandle, <span class="string">'FontSize'</span>, 5);
1923             
1924             tick2text(axesHandle, <span class="string">'axis'</span>, <span class="string">'xy'</span>, <span class="string">'xtickoffset'</span>, 0.05, <span class="string">'ytickoffset'</span>, 0.015);
1925             xTicks = getappdata(axesHandle, <span class="string">'XTickText'</span>);
1926             yTicks = getappdata(axesHandle, <span class="string">'YTickText'</span>);
1927             set(xTicks, <span class="string">'FontSize'</span>, 5);
1928             set(yTicks, <span class="string">'FontSize'</span>, 5, <span class="string">'HorizontalAlign'</span>, <span class="string">'right'</span>);
1929             delete(yTicks(2));
1930             delete(xTicks(2:2:end));
1931             
1932             xlabel(axesHandle, <span class="string">'Time (h)'</span>, <span class="string">'FontSize'</span>, 7);
1933             ylabel(axesHandle, <span class="string">'Mass (norm)'</span>, <span class="string">'FontSize'</span>, 7);
1934             xlabelPos = get(get(axesHandle, <span class="string">'xlabel'</span>), <span class="string">'position'</span>);
1935             ylabelPos = get(get(axesHandle, <span class="string">'ylabel'</span>), <span class="string">'position'</span>);
1936             set(get(axesHandle, <span class="string">'xlabel'</span>), <span class="string">'position'</span>, [xlabelPos(1) 0.82 xlabelPos(3:end)]);
1937             set(get(axesHandle, <span class="string">'ylabel'</span>), <span class="string">'position'</span>, [-0.50 ylabelPos(2:end)]);
1938             
1939             legendHandle = legend(lh, {<span class="string">'Total'</span>, <span class="string">'DNA'</span>, <span class="string">'RNA'</span>, <span class="string">'Protein'</span>, <span class="string">'Membrane'</span>}, <span class="string">'Location'</span>, <span class="string">'NorthWest'</span>, <span class="string">'FontSize'</span>, 7);
1940             legend(axesHandle, <span class="string">'boxoff'</span>);
1941             legendPos = get(legendHandle, <span class="string">'Position'</span>);
1942             set(legendHandle, <span class="string">'Position'</span>, [0.45 1.05 legendPos(3:end)])
1943             set(legendHandle, <span class="string">'units'</span>, <span class="string">'pixels'</span>);
1944             legendPos = get(legendHandle, <span class="string">'Position'</span>);
1945             set(legendHandle, <span class="string">'Position'</span>, [legendPos(1:3) 49]);
1946             legendChildren = get(legendHandle, <span class="string">'children'</span>);
1947             set(legendChildren(1:3:end), <span class="string">'visible'</span>, <span class="string">'off'</span>);
1948             <span class="keyword">for</span> i = 1:3:numel(legendChildren)
1949                 set(legendChildren(i+1), <span class="string">'xdata'</span>, [0.03 0.10], <span class="string">'LineWidth'</span>, 1)
1950                 
1951                 pos = get(legendChildren(i+2), <span class="string">'Position'</span>);
1952                 set(legendChildren(i+2), <span class="string">'Position'</span>, [0.13 pos(2) 0])
1953             <span class="keyword">end</span>
1954         <span class="keyword">end</span>
1955         
1956         <a name="_sub10" href="#_subfunctions" class="code">function populationMass(simBatchDir, selectedSim, figHandle, position, outDirectory)</a>
1957             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
1958             import edu.stanford.covert.cell.sim.util.PlotUtil;
1959             
1960             <span class="keyword">if</span> ~exist(<span class="string">'simBatchDir'</span>, <span class="string">'var'</span>) || isempty(simBatchDir)
1961                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
1962             <span class="keyword">end</span>
1963             
1964             <span class="keyword">if</span> exist([outDirectory <span class="string">'populationMassDistribution.mat'</span>], <span class="string">'file'</span>)
1965                 load([outDirectory <span class="string">'populationMassDistribution.mat'</span>]);
1966             <span class="keyword">else</span>
1967                 data = load([SimulationDiskUtil.getSimulationBatchDir([simBatchDir filesep num2str(selectedSim)]) filesep <span class="string">'population-MassDistribution.mat'</span>], <span class="string">'mass'</span>, <span class="string">'time'</span>);
1968                 mass = data.mass;
1969                 time = data.time;
1970                 save([outDirectory <span class="string">'populationMassDistribution.mat'</span>], <span class="string">'mass'</span>, <span class="string">'time'</span>);
1971             <span class="keyword">end</span>
1972             
1973             <span class="comment">% Determine division times</span>
1974             simEndTimes = zeros(1, size(mass, 2));
1975             <span class="keyword">for</span> i = 1:size(mass, 2)
1976                 et = find(isnan(mass(:, i)), 1);
1977                 <span class="keyword">if</span> isempty(et)
1978                     et = NaN;
1979                 <span class="keyword">end</span>
1980                 simEndTimes(i) = et - 1;
1981             <span class="keyword">end</span>
1982             
1983             finalMass = zeros(1, size(mass, 2));
1984             <span class="keyword">for</span> i = 1:size(finalMass, 2)
1985                 <span class="keyword">if</span> isnan(simEndTimes(i))
1986                     finalMass(i) = mass(<span class="keyword">end</span>, i);
1987                 <span class="keyword">else</span>
1988                     finalMass(i) = mass(simEndTimes(i), i);
1989                 <span class="keyword">end</span>
1990             <span class="keyword">end</span>
1991             
1992             <span class="comment">%markIdx = find(simEndTimes == floor(nanmedian(simEndTimes)));</span>
1993             markIdx = 10; <span class="comment">%52;</span>
1994             
1995             <span class="comment">%pick every other cell</span>
1996             simEndTimes = simEndTimes(2:2:end);
1997             mass = mass(:, 2:2:end);
1998             markIdx = markIdx / 2;
1999             
2000             <span class="comment">%% plot</span>
2001             axesSizes = [50; 5];
2002             
2003             options = struct;
2004             <span class="keyword">if</span> nargin &gt;= 4
2005                 options.position = position;
2006             <span class="keyword">else</span>
2007                 [~, figHandle] = edu.stanford.covert.cell.sim.util.PlotUtil.newAxesHandle();
2008                 clf(figHandle);
2009             <span class="keyword">end</span>
2010             [axesHandles, xAxisHandle] = PlotUtil.multiElementPlot(figHandle, axesSizes, time([1 end])', options);
2011             xlim(xAxisHandle, [0 max(time)]);
2012             set(get(xAxisHandle, <span class="string">'xlabel'</span>), <span class="string">'FontSize'</span>, 7);
2013             set(xAxisHandle, <span class="string">'FontSize'</span>, 5);
2014             set(xAxisHandle, <span class="string">'XTick'</span>, 0:5:10);
2015             tick2text(xAxisHandle, <span class="string">'axis'</span>, <span class="string">'x'</span>, <span class="string">'xtickoffset'</span>, 50);
2016             xTicks = getappdata(xAxisHandle, <span class="string">'XTickText'</span>);
2017             set(xTicks, <span class="string">'FontSize'</span>, 5);
2018             xlabelPos = get(get(xAxisHandle, <span class="string">'XLabel'</span>), <span class="string">'Position'</span>);
2019             set(get(xAxisHandle, <span class="string">'XLabel'</span>), <span class="string">'Position'</span>, [xlabelPos(1) -120 xlabelPos(3:end)]);
2020             
2021             <span class="comment">% Downsample</span>
2022             [~, simOrder] = sort(mass(1, :));
2023             timeds = time(1:100:end);
2024             massds = mass(1:100:<span class="keyword">end</span>, :);
2025             
2026             axesHandle = axesHandles(1);
2027             cla(axesHandle)
2028             hold(axesHandle, <span class="string">'on'</span>);
2029             n1 = floor(numel(simOrder) / 2);
2030             n2 = ceil(numel(simOrder) / 2);
2031             color1 = 0.50 * [1 1 1];
2032             color2 = 0.25 * [1 1 1];
2033             color3 = 0.00 * [1 1 1];
2034             colorOrder = zeros(numel(simOrder), 3);
2035             colorOrder(simOrder, :) = [
2036                 repmat(linspace(1, 0, n1)', 1, 3) .* repmat(color1, n1, 1)  +  repmat(linspace(0, 1, n1)', 1, 3) .* repmat(color2, n1, 1)
2037                 repmat(linspace(1, 0, n2)', 1, 3) .* repmat(color2, n2, 1)  +  repmat(linspace(0, 1, n2)', 1, 3) .* repmat(color3, n2, 1)
2038                 ];
2039             set(axesHandle, <span class="string">'ColorOrder'</span>, colorOrder);
2040             h = plot(axesHandle, timeds, massds, <span class="string">'LineWidth'</span>, 0.25);
2041             set(h(markIdx), <span class="string">'Color'</span>, <span class="string">'r'</span>, <span class="string">'LineWidth'</span>, 2);
2042             uistack(h(markIdx), <span class="string">'top'</span>);
2043             set(axesHandle, <span class="string">'XLim'</span>, [0 max(time)], <span class="string">'FontSize'</span>, 5);
2044             set(axesHandle, <span class="string">'YTick'</span>, 20:10:30);
2045             tick2text(axesHandle, <span class="string">'axis'</span>, <span class="string">'y'</span>, <span class="string">'ytickoffset'</span>, 0.035);
2046             yTicks = getappdata(axesHandle, <span class="string">'YTickText'</span>);
2047             set(yTicks, <span class="string">'FontSize'</span>, 5, <span class="string">'HorizontalAlign'</span>, <span class="string">'right'</span>);
2048             ylabel(axesHandle, <span class="string">'Mass (fg)'</span>, <span class="string">'FontSize'</span>, 7);
2049             ylabelPos = get(get(axesHandle, <span class="string">'ylabel'</span>), <span class="string">'position'</span>);
2050             set(get(axesHandle, <span class="string">'ylabel'</span>), <span class="string">'position'</span>, [-1.05 ylabelPos(2:end)]);
2051             ylim(axesHandle, [min(massds(:)), max(massds(:))]);
2052             
2053             axesHandle = axesHandles(2);
2054             hold(axesHandle, <span class="string">'on'</span>);
2055             
2056             edges = linspace(time(1), time(end), 100);
2057             freq = histc(simEndTimes / 3600, edges) / numel(simEndTimes) * 100;
2058             freq = smooth(freq, 9, <span class="string">'lowess'</span>);
2059             line(edges, freq, <span class="string">'Color'</span>, <span class="string">'k'</span>, <span class="string">'Parent'</span>, axesHandle)
2060             
2061             idx = find(edges &lt;= simEndTimes(markIdx) / 3600, 1, <span class="string">'last'</span>);
2062             y = freq(idx) + diff(freq(idx:idx+1)) / diff(edges(idx:idx+1)) * (simEndTimes(markIdx) / 3600 - edges(idx));
2063             
2064             axesPos = get(axesHandle, <span class="string">'Position'</span>);
2065             arrowX = axesPos(1) + simEndTimes(markIdx) / 3600 / time(end) * axesPos(3);
2066             arrowY1 = axesPos(2);
2067             axesPos = get(axesHandles(1), <span class="string">'Position'</span>);
2068             arrowY2 = axesPos(2) + axesPos(4)*(massds(floor(simEndTimes(markIdx)/100), markIdx) - min(massds(:))) / range(massds(:));
2069             y = linspace(arrowY1, arrowY2-0.01, 81);
2070             <span class="keyword">for</span> i = 1:3:numel(y)
2071                 <span class="keyword">if</span> i &gt;= 30 &amp;&amp; i &lt;= 50
2072                     <span class="keyword">continue</span>;
2073                 <span class="keyword">end</span>
2074                 h = annotation(get(axesHandle, <span class="string">'Parent'</span>), <span class="string">'line'</span>, arrowX * [1 1], y([i+1 i]), <span class="keyword">...</span>
2075                     <span class="string">'Color'</span>, <span class="string">'r'</span>);
2076                 uistack(h, <span class="string">'bottom'</span>);
2077             <span class="keyword">end</span>
2078             
2079             annotation(<span class="string">'textbox'</span>, [arrowX-0.06 y(26)+0.001 0.12 0.05], <span class="string">'String'</span>, <span class="string">'Median cell'</span>, <span class="keyword">...</span>
2080                 <span class="string">'HorizontalAlign'</span>, <span class="string">'center'</span>, <span class="string">'verticalAlign'</span>, <span class="string">'middle'</span>, <span class="keyword">...</span><span class="comment">.</span>
2081                 <span class="string">'FontSize'</span>, 7, <span class="string">'EdgeColor'</span>, <span class="string">'none'</span>, <span class="string">'BackgroundColor'</span>, <span class="string">'none'</span>, <span class="string">'Color'</span>, <span class="string">'r'</span>);
2082             annotation(<span class="string">'textbox'</span>, [arrowX-0.06 y(26)-0.014 0.12 0.05], <span class="string">'String'</span>, sprintf(<span class="string">'\\tau = %0.1f h'</span>, 8.9), <span class="keyword">...</span>
2083                 <span class="string">'HorizontalAlign'</span>, <span class="string">'center'</span>, <span class="string">'verticalAlign'</span>, <span class="string">'middle'</span>, <span class="keyword">...</span><span class="comment">.</span>
2084                 <span class="string">'FontSize'</span>, 7, <span class="string">'EdgeColor'</span>, <span class="string">'none'</span>, <span class="string">'BackgroundColor'</span>, <span class="string">'none'</span>, <span class="string">'Color'</span>, <span class="string">'r'</span>);
2085             
2086             ylim(axesHandle, [0 max(freq)]);
2087             xlim(axesHandle, [time(1) time(end)]);
2088             
2089             ylabel(axesHandle, {<span class="string">'% Cell div'</span>}, <span class="string">'FontSize'</span>, 7);
2090             ylabelPos = get(get(axesHandle, <span class="string">'ylabel'</span>), <span class="string">'position'</span>);
2091             set(get(axesHandle, <span class="string">'ylabel'</span>), <span class="string">'position'</span>, [-1.05 ylabelPos(2:end)]);
2092             
2093             set(axesHandle, <span class="string">'YTick'</span>, 0:4:8);
2094             set(axesHandle, <span class="string">'XTickLabel'</span>,<span class="string">''</span>);
2095             set(axesHandle, <span class="string">'XTick'</span>, []);
2096             set(axesHandle, <span class="string">'Box'</span>, <span class="string">'off'</span>);
2097             set(axesHandle, <span class="string">'XLim'</span>, [0 max(time)]);
2098             
2099             tick2text(axesHandle, <span class="string">'axis'</span>, <span class="string">'y'</span>, <span class="string">'ytickoffset'</span>, 0.035);
2100             yTicks = getappdata(axesHandle, <span class="string">'YTickText'</span>);
2101             set(yTicks, <span class="string">'FontSize'</span>, 5, <span class="string">'HorizontalAlign'</span>, <span class="string">'right'</span>);
2102             delete(yTicks(2));
2103             
2104             <span class="comment">%% offset axes</span>
2105             PlotUtil.offsetYAxes(axesHandles, 0.02);
2106         <span class="keyword">end</span>
2107         
2108         <a name="_sub11" href="#_subfunctions" class="code">function growthMeasurement(figHandle, position, outDirectory)</a>
2109             <span class="comment">%% parameters</span>
2110             [r, host] = system(<span class="string">'hostname'</span>);
2111             
2112             <span class="keyword">if</span> r ~= 0
2113                 warning(<span class="string">'WholeCell:warning:network'</span>, <span class="string">'Could not determine host name'</span>);
2114                 value = <span class="string">''</span>;
2115                 <span class="keyword">return</span>;
2116             <span class="keyword">end</span>
2117             host = deblank(host);
2118             <span class="keyword">switch</span> host
2119                 <span class="comment">% Edit as appropriate (add jayodita and markus)</span>
2120                 <span class="keyword">case</span> <span class="string">'hydra'</span>
2121                     baseDir = <span class="string">'/covertlab/home/share/Mycoplasma genitalium/MGenGrowthCurves'</span>;
2122                 <span class="keyword">case</span> <span class="string">'silico'</span>
2123                     baseDir = <span class="string">'/covertlab/home/share/Mycoplasma genitalium/MGenGrowthCurves'</span>;
2124                 <span class="keyword">case</span> <span class="string">'covertlab-jkarr'</span>
2125                     baseDir = <span class="string">'Z:\share\Mycoplasma genitalium\MGenGrowthCurves'</span>;
2126                 <span class="keyword">otherwise</span>
2127                     <span class="keyword">if</span> strcmp(computer, <span class="string">'GLNXA64'</span>)
2128                         <span class="comment">% Probably the cluster</span>
2129                         baseDir = <span class="string">'/home/share/Mycoplasma genitalium/MGenGrowthCurves'</span>;
2130                     <span class="keyword">else</span>
2131                         <span class="comment">% Assume a PC with the correct Z drive mapping</span>
2132                         baseDir = <span class="string">'Z:\share\Mycoplasma genitalium\MGenGrowthCurves'</span>;
2133                         warning(<span class="string">'WholeCell:warning:network'</span>, <span class="string">'Assuming M. genitalium growth curves are in %s'</span>, baseDir);
2134                     <span class="keyword">end</span>
2135             <span class="keyword">end</span>
2136             
2137             plate = 93;
2138             condition = 1;
2139             replicate = 2;
2140             blankCondition = 2;
2141             blankReplicate = 1;
2142             nDilutions = 6;
2143             
2144             nReplicates = 3;
2145             nColumns = 10;
2146             excelRows = 3:8;
2147             <span class="comment">%4: 1X</span>
2148             <span class="comment">%5: 1/5X</span>
2149             <span class="comment">%6: 1/25X</span>
2150             excelCols = 3:12;
2151             <span class="comment">%3: Cond 1, Rep 1;</span>
2152             <span class="comment">%4: Cond 1, Rep 2;</span>
2153             <span class="comment">%5: Cond 1, Rep 3;</span>
2154             <span class="comment">%6: Cond 2, Rep 1;</span>
2155             <span class="comment">%7: Cond 2, Rep 2;</span>
2156             <span class="comment">%8: Cond 2, Rep 3;</span>
2157             <span class="comment">%9: Cond 3, Rep 1;</span>
2158             <span class="comment">%10: Cond 3, Rep 2;</span>
2159             <span class="comment">%11: Cond 3, Rep 3;</span>
2160             <span class="comment">%12: Blank</span>
2161             
2162             <span class="comment">%% get file name and times</span>
2163             files = dir(sprintf(<span class="string">'%s%sPlate%d'</span>, baseDir, filesep, plate));
2164             files = files(3:end);
2165             [~, order] = sort(cellfun(@(x) str2double(x(end-5:end-4)), {files.name}));
2166             files = files(order);
2167             times = cellfun(@(x) datenum(<span class="keyword">...</span>
2168                 2000 + str2double(x(end-9:end-8)), <span class="keyword">...</span><span class="comment">   %year</span>
2169                 str2double(x(end-15:end-14)), <span class="keyword">...</span><span class="comment"> %month</span>
2170                 str2double(x(end-12:end-11)), <span class="keyword">...</span><span class="comment"> %day</span>
2171                 str2double(x(end-21:end-20)), <span class="keyword">...</span><span class="comment"> %hour</span>
2172                 str2double(x(end-18:end-17)), <span class="keyword">...</span><span class="comment"> %minute</span>
2173                 0 <span class="keyword">...</span><span class="comment"> %second</span>
2174                 ), {files.name});
2175             times = times - times(1);
2176             
2177             <span class="comment">%% get data</span>
2178             <span class="comment">%from plate reader files</span>
2179             <span class="keyword">if</span> exist([outDirectory <span class="string">'growthMeasurement.mat'</span>], <span class="string">'file'</span>)
2180                 load([outDirectory <span class="string">'growthMeasurement.mat'</span>])
2181             <span class="keyword">else</span>
2182                 data = zeros(nDilutions, nColumns, numel(files));
2183                 <span class="keyword">for</span> i = 1:numel(files)
2184                     [~, ~, raw] = xlsread(sprintf(<span class="string">'%s%sPlate%d%s%s'</span>, baseDir, filesep, plate, filesep, files(i).name));
2185                     data(:, :, i) = cell2mat(raw(excelRows, excelCols));
2186                 <span class="keyword">end</span>
2187                 save([outDirectory <span class="string">'growthMeasurement.mat'</span>], <span class="string">'data'</span>);
2188             <span class="keyword">end</span>
2189             
2190             <span class="comment">% from Jayodita's excel sheet</span>
2191             [~, ~, raw] = xlsread([outDirectory <span class="string">'growthMeasurement.xls'</span>]);
2192             tfRows = [false; false; ~isnan(cell2mat(raw(3:<span class="keyword">end</span>, 1)))];
2193             times = cell2mat(raw(tfRows, 1))';
2194             times = (times - times(1)) / 24;
2195             data = zeros(nDilutions, nColumns, numel(times));
2196             data(3:5, (condition-1) * nReplicates + replicate, :) = permute(cell2mat(raw(tfRows, 2:4)), [2 3 1]);
2197             data(2:6, (blankCondition-1) * nReplicates + blankReplicate, :) = repmat(permute(cell2mat(raw(tfRows, 5)), [2 3 1]), [5 1 1]);
2198             
2199             <span class="comment">%% fit data</span>
2200             t = NaN(nDilutions, 1);
2201             <span class="keyword">for</span> i = 1:nDilutions
2202                 tmpData = permute(data(i, (condition-1) * nReplicates + replicate, :), [1 3 2]);
2203                 idx1 = find(tmpData &gt; 0.35, 1, <span class="string">'last'</span>);
2204                 idx2 = find(tmpData &lt; 0.2, 1, <span class="string">'first'</span>);
2205                 <span class="keyword">if</span> ~isempty(idx1) &amp;&amp; ~isempty(idx2)
2206                     t(i) = interp1(tmpData(:, idx1:idx2), times(:, idx1:idx2) * 24, 0.28, <span class="string">'linear'</span>);
2207                 <span class="keyword">end</span>
2208             <span class="keyword">end</span>
2209             deltaT = diff(t);
2210             tau = log(2) ./ (log(5) ./ deltaT);
2211             
2212             <span class="comment">%% plot</span>
2213             <span class="keyword">if</span> nargin &gt;= 2
2214                 axesHandle = subplot(<span class="string">'position'</span>, position, <span class="string">'Parent'</span>, figHandle);
2215             <span class="keyword">else</span>
2216                 axesHandle = edu.stanford.covert.cell.sim.util.PlotUtil.newAxesHandle();
2217             <span class="keyword">end</span>
2218             hold(axesHandle, <span class="string">'on'</span>);
2219             
2220             colors = 1 / 255 * [
2221                 25 25 112
2222                 0 0 255
2223                 30 144 255
2224                 139 137 137
2225                 ];
2226             
2227             h = plot(axesHandle, times, permute(data(3:5, (condition-1) * nReplicates + replicate, :), [1 3 2]), <span class="string">'LineWidth'</span>, 1); <span class="comment">%dilutions</span>
2228             plot(axesHandle, times, permute(mean(data(2:6, (blankCondition-1) * nReplicates + blankReplicate, :), 1), [1 3 2]), <span class="string">'Color'</span>, colors(4, :), <span class="string">'LineWidth'</span>, 1); <span class="comment">%blank</span>
2229             set(h(1), <span class="string">'Color'</span>, colors(1, :)); <span class="comment">%1x</span>
2230             set(h(2), <span class="string">'Color'</span>, colors(2, :)); <span class="comment">%1/5 X</span>
2231             set(h(3), <span class="string">'Color'</span>, colors(3, :)); <span class="comment">%1/25 X</span>
2232             xlabel(axesHandle, <span class="string">'Time (d)'</span>, <span class="string">'FontSize'</span>, 7);
2233             ylabel(axesHandle, <span class="string">'OD550'</span>, <span class="string">'FontSize'</span>, 7);
2234             
2235             set(axesHandle, <span class="string">'FontSize'</span>, 5, <span class="string">'TickDir'</span>, <span class="string">'out'</span>, <span class="string">'YDir'</span>, <span class="string">'Reverse'</span>, <span class="string">'XTick'</span>, 0:5:15, <span class="string">'YTick'</span>, 0:.2:0.6);
2236             box(axesHandle, <span class="string">'off'</span>);
2237             xlim(axesHandle, [0 17]);
2238             ylim(axesHandle, [0 0.65]);
2239             
2240             text(0.3, 0.048, <span class="string">'\tau ='</span>, <span class="string">'FontSize'</span>, 6, <span class="string">'Interpreter'</span>, <span class="string">'tex'</span>)
2241             text(4.2, 0.0125, <span class="string">'ln(2) {\Delta}{\it{t}}'</span>, <span class="string">'FontSize'</span>, 6, <span class="string">'Interpreter'</span>, <span class="string">'tex'</span>, <span class="string">'HorizontalAlign'</span>, <span class="string">'center'</span>)
2242             text(4.2, 0.0785, <span class="string">'ln(dilution factor)'</span>, <span class="string">'FontSize'</span>, 6, <span class="string">'Interpreter'</span>, <span class="string">'tex'</span>, <span class="string">'HorizontalAlign'</span>, <span class="string">'center'</span>)
2243             line([1.4 7], 0.051 * [1 1], <span class="string">'Parent'</span>, axesHandle, <span class="string">'Color'</span>, <span class="string">'k'</span>);
2244             
2245             text(0.3, .35, <span class="string">'1X dilution'</span>, <span class="string">'Parent'</span>, axesHandle, <span class="string">'FontSize'</span>, 6, <span class="string">'VerticalAlign'</span>, <span class="string">'middle'</span>, <span class="string">'Color'</span>, colors(1, :))
2246             text(0.3, .40, <span class="string">'5X dilution'</span>, <span class="string">'Parent'</span>, axesHandle, <span class="string">'FontSize'</span>, 6, <span class="string">'VerticalAlign'</span>, <span class="string">'middle'</span>, <span class="string">'Color'</span>, colors(2, :))
2247             text(0.3, .45, <span class="string">'25X dilution'</span>, <span class="string">'Parent'</span>, axesHandle, <span class="string">'FontSize'</span>, 6, <span class="string">'VerticalAlign'</span>, <span class="string">'middle'</span>, <span class="string">'Color'</span>, colors(3, :))
2248             text(0.3, .50, <span class="string">'Blank'</span>, <span class="string">'Parent'</span>, axesHandle, <span class="string">'FontSize'</span>, 6, <span class="string">'VerticalAlign'</span>, <span class="string">'middle'</span>, <span class="string">'Color'</span>, colors(4, :))
2249             
2250             axesPos = get(axesHandle, <span class="string">'Position'</span>);
2251             
2252             y = 0.28;
2253             x = t(3)/24-0.3;
2254             text(x, y - 0.025, sprintf(<span class="string">'{\\Delta}{\\it{t}} = %0.1f h'</span>, deltaT(3)), <span class="string">'Parent'</span>, axesHandle, <span class="string">'FontSize'</span>, 6, <span class="string">'HorizontalAlign'</span>, <span class="string">'right'</span>, <span class="string">'verticalAlign'</span>, <span class="string">'middle'</span>, <span class="string">'Interpreter'</span>, <span class="string">'tex'</span>)
2255             text(x-0.42, y + 0.025, sprintf(<span class="string">'\\tau = %0.1f h'</span>, tau(3)), <span class="string">'Parent'</span>, axesHandle, <span class="string">'FontSize'</span>, 6, <span class="string">'HorizontalAlign'</span>, <span class="string">'right'</span>, <span class="string">'verticalAlign'</span>, <span class="string">'middle'</span>, <span class="string">'Interpreter'</span>, <span class="string">'tex'</span>)
2256             annotation(<span class="string">'doublearrow'</span>, axesPos(1) + axesPos(3) * t([3 4])/24 / max(xlim(axesHandle)), <span class="keyword">...</span>
2257                 axesPos(2) + axesPos(4) * (max(ylim(axesHandle))-y) / range(ylim(axesHandle)) * [1 1], <span class="keyword">...</span>
2258                 <span class="string">'HeadLength'</span>, 1.5, <span class="string">'HeadWidth'</span>, 2, <span class="string">'HeadStyle'</span>, <span class="string">'vback1'</span>)
2259             
2260             y = 0.28;
2261             x = t(5)/24+0.3;
2262             text(x, y - 0.025, sprintf(<span class="string">'{\\Delta}{\\it{t}} = %0.1f h'</span>, deltaT(4)), <span class="string">'Parent'</span>, axesHandle, <span class="string">'FontSize'</span>, 6, <span class="string">'HorizontalAlign'</span>, <span class="string">'left'</span>, <span class="string">'verticalAlign'</span>, <span class="string">'middle'</span>, <span class="string">'Interpreter'</span>, <span class="string">'tex'</span>)
2263             text(x + 0.39, y + 0.025, sprintf(<span class="string">'\\tau = %0.1f h'</span>, tau(4)), <span class="string">'Parent'</span>, axesHandle, <span class="string">'FontSize'</span>, 6, <span class="string">'HorizontalAlign'</span>, <span class="string">'left'</span>, <span class="string">'verticalAlign'</span>, <span class="string">'middle'</span>, <span class="string">'Interpreter'</span>, <span class="string">'tex'</span>)
2264             annotation(<span class="string">'doublearrow'</span>, axesPos(1) + axesPos(3) * t([4 5])/24 / max(xlim(axesHandle)), <span class="keyword">...</span>
2265                 axesPos(2) + axesPos(4) * (max(ylim(axesHandle))-y) / range(ylim(axesHandle)) * [1 1], <span class="keyword">...</span>
2266                 <span class="string">'HeadLength'</span>, 1.5, <span class="string">'HeadWidth'</span>, 2, <span class="string">'HeadStyle'</span>, <span class="string">'vback1'</span>)
2267             
2268             annotation(<span class="string">'textbox'</span>, [0.37 0.8525+0.001 0.12 0.05], <span class="string">'String'</span>, <span class="string">'Mean'</span>, <span class="keyword">...</span>
2269                 <span class="string">'HorizontalAlign'</span>, <span class="string">'center'</span>, <span class="string">'verticalAlign'</span>, <span class="string">'middle'</span>, <span class="keyword">...</span><span class="comment">.</span>
2270                 <span class="string">'FontSize'</span>, 7, <span class="string">'EdgeColor'</span>, <span class="string">'none'</span>, <span class="string">'BackgroundColor'</span>, <span class="string">'none'</span>, <span class="string">'Color'</span>, <span class="string">'r'</span>);
2271             annotation(<span class="string">'textbox'</span>, [0.37 0.8525-0.014 0.12 0.05], <span class="string">'String'</span>, sprintf(<span class="string">'\\tau = %0.1f h'</span>, 9.03), <span class="keyword">...</span><span class="comment"> %Jayodita: where do this come from?</span>
2272                 <span class="string">'HorizontalAlign'</span>, <span class="string">'center'</span>, <span class="string">'verticalAlign'</span>, <span class="string">'middle'</span>, <span class="keyword">...</span><span class="comment">.</span>
2273                 <span class="string">'FontSize'</span>, 7, <span class="string">'EdgeColor'</span>, <span class="string">'none'</span>, <span class="string">'BackgroundColor'</span>, <span class="string">'none'</span>, <span class="string">'Color'</span>, <span class="string">'r'</span>);
2274             
2275             tick2text(axesHandle, <span class="string">'axis'</span>, <span class="string">'xy'</span>, <span class="string">'xtickoffset'</span>, -1.05, <span class="string">'ytickoffset'</span>, 0.015);
2276             xlabelPos = get(get(axesHandle, <span class="string">'xlabel'</span>), <span class="string">'position'</span>);
2277             ylabelPos = get(get(axesHandle, <span class="string">'ylabel'</span>), <span class="string">'position'</span>);
2278             xTicks = getappdata(axesHandle, <span class="string">'XTickText'</span>);
2279             yTicks = getappdata(axesHandle, <span class="string">'YTickText'</span>);
2280             set(xTicks, <span class="string">'FontSize'</span>, 5);
2281             set(yTicks, <span class="string">'FontSize'</span>, 5, <span class="string">'horizontalAlign'</span>, <span class="string">'right'</span>);
2282             set(get(axesHandle, <span class="string">'xlabel'</span>), <span class="string">'position'</span>, [xlabelPos(1) 0.71 xlabelPos(3:end)]);
2283             set(get(axesHandle, <span class="string">'ylabel'</span>), <span class="string">'position'</span>, [-1.15 ylabelPos(2:end)]);
2284         <span class="keyword">end</span>
2285         
2286         <a name="_sub12" href="#_subfunctions" class="code">function doublingTime(outDirectory)</a>
2287             <span class="comment">% From</span>
2288             <span class="comment">% /home/projects/WholeCell/simulation/output/runSimulation/singleGeneDeletions/summary.xls</span>
2289             modelMassDTMean = 8.92200510884003; <span class="comment">% Col W</span>
2290             modelMassDTStddev = 0.807282508239662; <span class="comment">% Col X</span>
2291             expMassDTMean = 9.03411809133663; <span class="comment">% Col AX</span>
2292             expMassDTStderr = 0.889827929955399; <span class="comment">% Col AY</span>
2293             
2294             [axesHandle, figHandle] = edu.stanford.covert.cell.sim.util.PlotUtil.newAxesHandle();
2295             hold(<span class="string">'on'</span>);
2296             bar(1, modelMassDTMean);
2297             bar(2, expMassDTMean, <span class="string">'k'</span>);
2298             eh1 = errorbar([modelMassDTMean nan], [modelMassDTStddev nan], <span class="string">'.g'</span>);
2299             eh2 = errorbar([nan expMassDTMean], [nan expMassDTStderr], <span class="string">'.r'</span>);
2300             hold(<span class="string">'off'</span>);
2301             legend([eh1 eh2], <span class="string">'Standard Deviation'</span>, <span class="string">'Standard Error'</span>, <span class="string">'Location'</span>, <span class="string">'BestOutside'</span>);
2302             ylabel(<span class="string">'Mass Doubling Time (h)'</span>);
2303             set(axesHandle, <span class="string">'XTick'</span>, [1 2]);
2304             set(axesHandle, <span class="string">'XTickLabel'</span>, {<span class="string">'Model'</span> <span class="string">'Experiment'</span>});
2305             set(axesHandle, <span class="string">'YTick'</span>, [0 2 4 6 8 10]);
2306             set(axesHandle, <span class="string">'Box'</span>, <span class="string">'off'</span>);
2307             set(axesHandle, <span class="string">'TickDir'</span>, <span class="string">'out'</span>);
2308             
2309             set(figHandle, <span class="string">'Renderer'</span>, <span class="string">'Painters'</span>)
2310             print(figHandle, [outDirectory filesep <span class="string">'MassDoublingTime.eps'</span>], <span class="string">'-depsc'</span>, <span class="string">'-rgb'</span>);
2311         <span class="keyword">end</span>
2312         
2313         <a name="_sub13" href="#_subfunctions" class="code">function dryBiomassComposition(simBatchDir, figHandle, position, outDirectory)</a>
2314             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
2315             import edu.stanford.covert.cell.sim.util.SimulationEnsemble;
2316             
2317             <span class="keyword">if</span> ~exist(<span class="string">'simBatchDir'</span>, <span class="string">'var'</span>) || isempty(simBatchDir)
2318                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
2319             <span class="keyword">end</span>
2320             selectedSimulations = 1:SimulationDiskUtil.getNumSimulations(simBatchDir);
2321             
2322             <span class="comment">%% get data</span>
2323             <span class="keyword">if</span> exist([outDirectory <span class="string">'dryBiomassComposition.mat'</span>], <span class="string">'file'</span>)
2324                 load([outDirectory <span class="string">'dryBiomassComposition.mat'</span>])
2325             <span class="keyword">else</span>
2326                 [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep <span class="string">'1'</span>]);
2327                 comp = sim.compartment;
2328                 
2329                 stateNames = {
2330                     <span class="string">'Time'</span> <span class="string">'values'</span>       <span class="string">':'</span> <span class="string">':'</span>
2331                     <span class="string">'Mass'</span> <span class="string">'dnaWt'</span>        <span class="string">':'</span> <span class="string">'-sum'</span>
2332                     <span class="string">'Mass'</span> <span class="string">'rnaWt'</span>        <span class="string">':'</span> <span class="string">'-sum'</span>
2333                     <span class="string">'Mass'</span> <span class="string">'proteinWt'</span>    <span class="string">':'</span> <span class="string">'-sum'</span>
2334                     <span class="string">'Mass'</span> <span class="string">'metaboliteWt'</span> <span class="string">':'</span> [comp.membraneIndexs]
2335                     <span class="string">'Mass'</span> <span class="string">'cellDry'</span>         <span class="string">':'</span> <span class="string">'-sum'</span>
2336                     };
2337                 states = SimulationEnsemble.load(simBatchDir, stateNames, 3600, 3600, 1, <span class="string">'extract'</span>, selectedSimulations);
2338                 
2339                 dnaWt = permute(states.Mass.dnaWt, [4 3 2 1]) ./ permute(states.Mass.cellDry, [4 3 2 1]) * 100;
2340                 rnaWt = permute(states.Mass.rnaWt, [4 3 2 1]) ./ permute(states.Mass.cellDry, [4 3 2 1]) * 100;
2341                 proteinWt = permute(states.Mass.proteinWt, [4 3 2 1]) ./ permute(states.Mass.cellDry, [4 3 2 1]) * 100;
2342                 lipidWt = permute(states.Mass.metaboliteWt, [4 3 2 1]) ./ permute(states.Mass.cellDry, [4 3 2 1]) * 100;
2343                 
2344                 modelDNA = mean(dnaWt);
2345                 modelRNA = mean(rnaWt);
2346                 modelProtein = mean(proteinWt);
2347                 modelLipid = mean(lipidWt);
2348                 
2349                 stdDNA = std(dnaWt);
2350                 stdRNA = std(rnaWt);
2351                 stdProtein = std(proteinWt);
2352                 stdLipid = std(lipidWt);
2353                 
2354                 save([outDirectory <span class="string">'dryBiomassComposition.mat'</span>], <span class="keyword">...</span>
2355                     <span class="string">'modelDNA'</span>, <span class="string">'modelRNA'</span>, <span class="string">'modelProtein'</span>, <span class="string">'modelLipid'</span>, <span class="keyword">...</span>
2356                     <span class="string">'stdDNA'</span>, <span class="string">'stdRNA'</span>, <span class="string">'stdProtein'</span>, <span class="string">'stdLipid'</span>);
2357             <span class="keyword">end</span>
2358             
2359             <span class="comment">% From Jonathan</span>
2360             expDNA = 4;
2361             expLipid = 11;
2362             expProtein = 80;
2363             expRNA = 8;
2364             
2365             <span class="keyword">if</span> nargin &gt;= 4
2366                 axesHandle = subplot(<span class="string">'Position'</span>, position, <span class="string">'Parent'</span>, figHandle);
2367             <span class="keyword">else</span>
2368                 axesHandle = edu.stanford.covert.cell.sim.util.PlotUtil.newAxesHandle();
2369             <span class="keyword">end</span>
2370             hold(axesHandle, <span class="string">'on'</span>);
2371             
2372             h = [
2373                 bar(axesHandle, (0:3)*2.5-0.5, [modelDNA modelLipid modelProtein modelRNA], 0.37, <span class="string">'b'</span>)
2374                 bar(axesHandle, (0:3)*2.5+0.5, [expDNA expLipid expProtein expRNA], 0.37, <span class="string">'k'</span>)
2375                 ];
2376             errorbar((0:3)*2.5-0.5, [modelDNA modelLipid modelProtein modelRNA], [stdDNA stdLipid stdProtein stdRNA], <span class="keyword">...</span>
2377                 <span class="string">'r'</span>, <span class="string">'Parent'</span>, axesHandle, <span class="string">'LineStyle'</span>, <span class="string">'none'</span>);
2378             set(h, <span class="string">'EdgeColor'</span>, <span class="string">'none'</span>)
2379             
2380             set(axesHandle, <span class="string">'XTick'</span>, (0:3)*2.5, <span class="string">'Xlim'</span>, [-1.3 3*2.5+1.3]);
2381             set(axesHandle, <span class="string">'XTickLabel'</span>, {<span class="string">'DNA'</span> <span class="string">'Lipid'</span> <span class="string">'Protein'</span> <span class="string">'RNA'</span>});
2382             set(axesHandle, <span class="string">'YTick'</span>, 0:25:75);
2383             tick2text(axesHandle, <span class="string">'axis'</span>, <span class="string">'y'</span>, <span class="string">'ytickoffset'</span>, 0.015);
2384             yTicks = getappdata(axesHandle, <span class="string">'YTickText'</span>);
2385             set(yTicks, <span class="string">'FontSize'</span>, 5, <span class="string">'horizontalAlign'</span>, <span class="string">'right'</span>);
2386             
2387             set(axesHandle, <span class="string">'Box'</span>, <span class="string">'off'</span>);
2388             set(axesHandle, <span class="string">'TickDir'</span>, <span class="string">'out'</span>);
2389             set(axesHandle, <span class="string">'FontSize'</span>, 5);
2390             
2391             legendHandle = legend({<span class="string">'Model'</span>, <span class="string">'Experiment'</span>}, <span class="string">'Location'</span>, <span class="string">'NorthWest'</span>, <span class="string">'FontSize'</span>, 7);
2392             legend(axesHandle, <span class="string">'boxoff'</span>);
2393             legendPos = get(legendHandle, <span class="string">'Position'</span>);
2394             set(legendHandle, <span class="string">'Position'</span>, [0.99 1.025 legendPos(3:end)])
2395             set(legendHandle, <span class="string">'Units'</span>, <span class="string">'pixels'</span>);
2396             legendPos = get(legendHandle, <span class="string">'Position'</span>);
2397             set(legendHandle, <span class="string">'Position'</span>, [legendPos(1:3) 24])
2398             legendChildren = get(legendHandle, <span class="string">'children'</span>);
2399             <span class="keyword">for</span> i = 1:2:numel(legendChildren)
2400                 ptch = get(legendChildren(i), <span class="string">'children'</span>);
2401                 set(ptch, <span class="string">'xdata'</span>, [0.05 0.05 0.12 0.12 0.05])
2402                 
2403                 pos = get(legendChildren(i+1), <span class="string">'Position'</span>);
2404                 set(legendChildren(i+1), <span class="string">'Position'</span>, [0.15 pos(2) 0])
2405             <span class="keyword">end</span>
2406             
2407             ylabel(axesHandle, <span class="string">'Percent dry mass'</span>, <span class="string">'FontSize'</span>, 7);
2408             ylabelPos = get(get(axesHandle, <span class="string">'ylabel'</span>), <span class="string">'position'</span>);
2409             set(get(axesHandle, <span class="string">'ylabel'</span>), <span class="string">'position'</span>, [-1.95 ylabelPos(2:end)]);
2410             
2411             tick2text(axesHandle, <span class="string">'axis'</span>, <span class="string">'x'</span>, <span class="string">'xtickoffset'</span>, 0.05);
2412             xTicks = getappdata(axesHandle, <span class="string">'XTickText'</span>);
2413             set(xTicks, <span class="string">'FontSize'</span>, 7);
2414             set(xTicks(1), <span class="string">'String'</span>, <span class="string">'DNA'</span>);
2415             set(xTicks(2), <span class="string">'String'</span>, <span class="string">'Lipid'</span>);
2416             set(xTicks(3), <span class="string">'String'</span>, <span class="string">'Protein'</span>);
2417             set(xTicks(4), <span class="string">'String'</span>, <span class="string">'RNA'</span>);
2418         <span class="keyword">end</span>
2419         
2420         <a name="_sub14" href="#_subfunctions" class="code">function cellCycleLength(simBatchDir, selectedSim, figHandle, position)</a>
2421             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
2422             import edu.stanford.covert.cell.sim.util.PlotUtil;
2423             import edu.stanford.covert.cell.sim.util.SummaryLogger;
2424             
2425             <span class="keyword">if</span> ~exist(<span class="string">'simBatchDir'</span>, <span class="string">'var'</span>) || isempty(simBatchDir)
2426                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
2427             <span class="keyword">end</span>
2428             
2429             <span class="comment">%% get data</span>
2430             data = load([SimulationDiskUtil.getSimulationBatchDir([simBatchDir filesep num2str(selectedSim)]) filesep <span class="string">'population-CellCyclePhases.mat'</span>]);
2431             stats = data.stats;
2432             maxTime = data.maxTime;
2433             
2434             edges = linspace(0, maxTime, 50);
2435             n1 = histc(stats(:, SummaryLogger.SIM_STATUS_INDEX_CYTOKINESIS_TIME), edges);
2436             n2 = histc(stats(:, SummaryLogger.SIM_STATUS_INDEX_REP_INIT_TIME), edges);
2437             n3 = histc(stats(:, SummaryLogger.SIM_STATUS_INDEX_REP_TIME) - stats(:, SummaryLogger.SIM_STATUS_INDEX_REP_INIT_TIME), edges);
2438             n4 = histc(stats(:, SummaryLogger.SIM_STATUS_INDEX_CYTOKINESIS_TIME) - stats(:, SummaryLogger.SIM_STATUS_INDEX_REP_TIME), edges);
2439             
2440             lowIdx = find(n1 &gt; 0, 1, <span class="string">'first'</span>) - 2;
2441             highIdx = find(n1 &gt; 0, 1, <span class="string">'last'</span>) + 2;
2442             n1([1:lowIdx highIdx:end]) = NaN;
2443             lowIdx = find(n2 &gt; 0, 1, <span class="string">'first'</span>) - 2;
2444             highIdx = find(n2 &gt; 0, 1, <span class="string">'last'</span>) + 2;
2445             n2([1:lowIdx highIdx:end]) = NaN;
2446             lowIdx = find(n3 &gt; 0, 1, <span class="string">'first'</span>) - 2;
2447             highIdx = find(n3 &gt; 0, 1, <span class="string">'last'</span>) + 2;
2448             n3([1:lowIdx highIdx:end]) = NaN;
2449             lowIdx = find(n4 &gt; 0, 1, <span class="string">'first'</span>) - 2;
2450             highIdx = find(n4 &gt; 0, 1, <span class="string">'last'</span>) + 2;
2451             n4([1:lowIdx highIdx:end]) = NaN;
2452             
2453             <span class="comment">%% plot data</span>
2454             <span class="keyword">if</span> nargin &gt;= 4
2455                 axesHandle = subplot(<span class="string">'position'</span>, position, <span class="string">'parent'</span>, figHandle);
2456             <span class="keyword">else</span>
2457                 axesHandle = PlotUtil.newAxesHandle();
2458             <span class="keyword">end</span>
2459             
2460             n1 = smooth(n1, <span class="string">'lowess'</span>, 13) / size(stats, 1) * 100;
2461             n2 = smooth(n2, <span class="string">'lowess'</span>, 13) / size(stats, 1) * 100;
2462             n3 = smooth(n3, <span class="string">'lowess'</span>, 13) / size(stats, 1) * 100;
2463             n4 = smooth(n4, <span class="string">'lowess'</span>, 13) / size(stats, 1) * 100;
2464             h = plot(axesHandle, edges / 3600, [n1 n2 n3 n4], <span class="string">'LineWidth'</span>, 1);
2465             set(h(2), <span class="string">'Color'</span>, <span class="string">'g'</span>);
2466             
2467             ylim(axesHandle, [0 max(max([n1 n2 n3 n4]))]);
2468             set(axesHandle, <span class="string">'XTick'</span>, 0:5:10);
2469             set(axesHandle, <span class="string">'YTick'</span>, 0:10:40);
2470             set(axesHandle, <span class="string">'Box'</span>, <span class="string">'off'</span>);
2471             set(axesHandle, <span class="string">'TickDir'</span>, <span class="string">'out'</span>);
2472             set(axesHandle, <span class="string">'FontSize'</span>, 5);
2473             
2474             xlabel(axesHandle, <span class="string">'Duration (h)'</span>, <span class="string">'FontSize'</span>, 7);
2475             ylabel(axesHandle, <span class="string">'% Cells'</span>, <span class="string">'FontSize'</span>, 7);
2476             xlabelPos = get(get(axesHandle, <span class="string">'xlabel'</span>), <span class="string">'position'</span>);
2477             ylabelPos = get(get(axesHandle, <span class="string">'ylabel'</span>), <span class="string">'position'</span>);
2478             set(get(axesHandle, <span class="string">'xlabel'</span>), <span class="string">'position'</span>, [xlabelPos(1) -3.7 xlabelPos(3:end)]);
2479             set(get(axesHandle, <span class="string">'ylabel'</span>), <span class="string">'position'</span>, [-1.35 ylabelPos(2:end)]);
2480             
2481             tick2text(axesHandle, <span class="string">'axis'</span>, <span class="string">'xy'</span>, <span class="string">'xtickoffset'</span>, 0.04, <span class="string">'ytickoffset'</span>, 0.015);
2482             xTicks = getappdata(axesHandle, <span class="string">'XTickText'</span>);
2483             yTicks = getappdata(axesHandle, <span class="string">'YTickText'</span>);
2484             set(xTicks, <span class="string">'FontSize'</span>, 5);
2485             set(yTicks, <span class="string">'FontSize'</span>, 5, <span class="string">'HorizontalAlign'</span>, <span class="string">'right'</span>);
2486             delete(xTicks(2));
2487             delete(yTicks(2:2:end));
2488             
2489             legendHandle = legend(h, {<span class="string">'Cell cycle'</span>, <span class="string">'Replication initiation'</span>, <span class="string">'Replication'</span>, <span class="string">'Cytokinesis'</span>}, <span class="string">'Location'</span>, <span class="string">'NorthEast'</span>, <span class="string">'FontSize'</span>, 7);
2490             legend(axesHandle, <span class="string">'boxoff'</span>);
2491             legendPos = get(legendHandle, <span class="string">'Position'</span>);
2492             set(legendHandle, <span class="string">'Position'</span>, [0.295 0.43 legendPos(3:end)])
2493             set(legendHandle, <span class="string">'units'</span>, <span class="string">'pixels'</span>);
2494             legendPos = get(legendHandle, <span class="string">'Position'</span>);
2495             set(legendHandle, <span class="string">'Position'</span>, [legendPos(1:3) 40]);
2496             legendChildren = get(legendHandle, <span class="string">'children'</span>);
2497             set(legendChildren(1:3:end), <span class="string">'visible'</span>, <span class="string">'off'</span>);
2498             <span class="keyword">for</span> i = 1:3:numel(legendChildren)
2499                 set(legendChildren(i+1), <span class="string">'xdata'</span>, [0.03 0.08])
2500                 
2501                 pos = get(legendChildren(i+2), <span class="string">'Position'</span>);
2502                 set(legendChildren(i+2), <span class="string">'Position'</span>, [0.11 pos(2) 0])
2503             <span class="keyword">end</span>
2504         <span class="keyword">end</span>
2505         
2506         <a name="_sub15" href="#_subfunctions" class="code">function replicationInitiationVsReplicationDuration(simBatchDir, selectedSim, figHandle, position)</a>
2507             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
2508             import edu.stanford.covert.cell.sim.util.PlotUtil;
2509             import edu.stanford.covert.cell.sim.util.SummaryLogger;
2510             
2511             <span class="keyword">if</span> ~exist(<span class="string">'simBatchDir'</span>, <span class="string">'var'</span>) || isempty(simBatchDir)
2512                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
2513             <span class="keyword">end</span>
2514             
2515             <span class="comment">%% get data</span>
2516             data = load([SimulationDiskUtil.getSimulationBatchDir([simBatchDir filesep num2str(selectedSim)]) filesep <span class="string">'population-CellCyclePhases.mat'</span>]);
2517             repInitDur = data.stats(:, SummaryLogger.SIM_STATUS_INDEX_REP_INIT_TIME) / 3600;
2518             repDur = (data.stats(:, SummaryLogger.SIM_STATUS_INDEX_REP_TIME) - data.stats(:, SummaryLogger.SIM_STATUS_INDEX_REP_INIT_TIME)) / 3600;
2519             
2520             <span class="comment">%% plot data</span>
2521             <span class="keyword">if</span> nargin &gt;= 4
2522                 axesHandle = subplot(<span class="string">'position'</span>, position, <span class="string">'parent'</span>, figHandle);
2523             <span class="keyword">else</span>
2524                 axesHandle = PlotUtil.newAxesHandle();
2525             <span class="keyword">end</span>
2526             
2527             plot(axesHandle, repInitDur, repDur, <span class="string">'b.'</span>)
2528             
2529             set(axesHandle, <span class="string">'XTick'</span>, 0:5:10);
2530             set(axesHandle, <span class="string">'YTick'</span>, 0:4:8);
2531             set(axesHandle, <span class="string">'Box'</span>, <span class="string">'off'</span>);
2532             set(axesHandle, <span class="string">'TickDir'</span>, <span class="string">'out'</span>);
2533             set(axesHandle, <span class="string">'FontSize'</span>, 5);
2534             
2535             xlabel(axesHandle, {<span class="string">'Replication initiation'</span> <span class="string">'duration (h)'</span>}, <span class="string">'FontSize'</span>, 7);
2536             ylabel(axesHandle, {<span class="string">'Replication'</span> <span class="string">'duration (h)'</span>}, <span class="string">'FontSize'</span>, 7);
2537             xlabelPos = get(get(axesHandle, <span class="string">'xlabel'</span>), <span class="string">'position'</span>);
2538             ylabelPos = get(get(axesHandle, <span class="string">'ylabel'</span>), <span class="string">'position'</span>);
2539             set(get(axesHandle, <span class="string">'xlabel'</span>), <span class="string">'position'</span>, [xlabelPos(1)*1.25 -0.7 xlabelPos(3:end)]);
2540             set(get(axesHandle, <span class="string">'ylabel'</span>), <span class="string">'position'</span>, [-0.4 4 ylabelPos(3:end)]);
2541             
2542             tick2text(axesHandle, <span class="string">'axis'</span>, <span class="string">'xy'</span>, <span class="string">'xtickoffset'</span>, 0.04, <span class="string">'ytickoffset'</span>, 0.015);
2543             xTicks = getappdata(axesHandle, <span class="string">'XTickText'</span>);
2544             yTicks = getappdata(axesHandle, <span class="string">'YTickText'</span>);
2545             set(xTicks, <span class="string">'FontSize'</span>, 5);
2546             set(yTicks, <span class="string">'FontSize'</span>, 5, <span class="string">'HorizontalAlign'</span>, <span class="string">'right'</span>);
2547             delete(xTicks(2));
2548             delete(yTicks(2:2:end));
2549         <span class="keyword">end</span>
2550         
2551         <a name="_sub16" href="#_subfunctions" class="code">function replicationInitiationVsReplicationDurationFit(simBatchDir, selectedSim, figHandle, position)</a>
2552             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
2553             import edu.stanford.covert.cell.sim.util.PlotUtil;
2554             import edu.stanford.covert.cell.sim.util.SummaryLogger;
2555             import edu.stanford.covert.util.ComputationUtil;
2556             
2557             <span class="keyword">if</span> ~exist(<span class="string">'simBatchDir'</span>, <span class="string">'var'</span>) || isempty(simBatchDir)
2558                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
2559             <span class="keyword">end</span>
2560             
2561             <span class="comment">%% get data</span>
2562             data = load([SimulationDiskUtil.getSimulationBatchDir([simBatchDir filesep num2str(selectedSim)]) filesep <span class="string">'population-CellCyclePhases.mat'</span>]);
2563             repInitDur = data.stats(:, SummaryLogger.SIM_STATUS_INDEX_REP_INIT_TIME) / 3600;
2564             repDur = (data.stats(:, SummaryLogger.SIM_STATUS_INDEX_REP_TIME) - data.stats(:, SummaryLogger.SIM_STATUS_INDEX_REP_INIT_TIME)) / 3600;
2565             
2566             <span class="comment">%% fit</span>
2567             [a, b, cutX, gof] = ComputationUtil.bilinearFit(repInitDur, repDur);
2568             
2569             <span class="comment">%% plot</span>
2570             <span class="keyword">if</span> nargin &gt;= 4
2571                 axesHandle = subplot(<span class="string">'position'</span>, position, <span class="string">'parent'</span>, figHandle);
2572             <span class="keyword">else</span>
2573                 axesHandle = PlotUtil.newAxesHandle();
2574             <span class="keyword">end</span>
2575             hold(axesHandle, <span class="string">'on'</span>);
2576             x1 = [min(repInitDur) cutX];
2577             x2 = [cutX max(repInitDur)];
2578             h = [
2579                 plot(axesHandle, repInitDur, repDur, <span class="string">'b.'</span>)
2580                 plot(axesHandle, x1, a(1) + b(1) * x1, <span class="string">'r'</span>)
2581                 plot(axesHandle, x2, a(2) + b(2) * x2, <span class="string">'r'</span>)
2582                 ];
2583             legend(h(1:2), {<span class="string">'Data'</span>, <span class="string">'Fit'</span>}, <span class="string">'Location'</span>, <span class="string">'NorthEast'</span>);
2584             xlabel(axesHandle, <span class="string">'Replication Initiation Duration (h)'</span>)
2585             ylabel(axesHandle, <span class="string">'Replication Duration (h)'</span>)
2586             
2587             text(10, 6.5, sprintf(<span class="string">'R^2_{adj} = %0.2f'</span>, gof.adjrsquare(1)), <span class="string">'Parent'</span>, axesHandle);
2588             text(0.7, 4, sprintf(<span class="string">'R^2_{adj} = %0.2f'</span>, gof.adjrsquare(2)), <span class="string">'Parent'</span>, axesHandle);
2589             text(8, 3, sprintf(<span class="string">'R^2_{adj} = %0.2f'</span>, gof.adjrsquare(3)), <span class="string">'Parent'</span>, axesHandle);
2590         <span class="keyword">end</span>
2591         
2592         <a name="_sub17" href="#_subfunctions" class="code">function proteinBursting(simBatchDir, selectedSim, outDirectory, figHandle, position)</a>
2593             import edu.stanford.covert.cell.sim.util.MacromoleculeUtil;
2594             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
2595             import edu.stanford.covert.cell.sim.util.SimulationEnsemble;
2596             import edu.stanford.covert.cell.sim.util.PlotUtil;
2597             
2598             <span class="keyword">if</span> ~exist(<span class="string">'simBatchDir'</span>, <span class="string">'var'</span>) || isempty(simBatchDir)
2599                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
2600             <span class="keyword">end</span>
2601             
2602             <span class="comment">%% get data</span>
2603             <span class="keyword">if</span> exist([outDirectory <span class="string">'proteinBursting.mat'</span>], <span class="string">'file'</span>)
2604                 load([outDirectory <span class="string">'proteinBursting.mat'</span>]);
2605             <span class="keyword">else</span>
2606                 [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep num2str(selectedSim)]);
2607                 
2608                 comp = sim.compartment;
2609                 rna = sim.state(<span class="string">'Rna'</span>);
2610                 pm = sim.state(<span class="string">'ProteinMonomer'</span>);
2611                 
2612                 [~, ~, ~, ~, ~, <span class="keyword">...</span>
2613                     ~, ~, ~, ~, <span class="keyword">...</span>
2614                     ~, ~, ~, ~, <span class="keyword">...</span>
2615                     matureRnaIdxs, ~, ~, ~, <span class="keyword">...</span>
2616                     monomerIdxs, ~, ~, ~, <span class="keyword">...</span>
2617                     complexIdxs, ~, ~, ~] = <span class="keyword">...</span>
2618                     MacromoleculeUtil.getMacromoleculeIndexsIDsNames(<span class="string">'MG_218'</span>, sim);
2619                 
2620                 stateNames = {
2621                     <span class="string">'Rna'</span>               <span class="string">'counts'</span> rna.matureIndexs(matureRnaIdxs)    comp.cytosolIndexs
2622                     <span class="string">'ProteinMonomer'</span>    <span class="string">'counts'</span> pm.matureIndexs(monomerIdxs)       <span class="string">'-sum'</span>
2623                     
2624                     };
2625                 states = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, <span class="string">'extract'</span>, selectedSim);
2626                 
2627                 
2628                 enzymeRNA = permute(states.Rna.counts, [1 3 2]);
2629                 enzymeMonomer = permute(states.ProteinMonomer.counts, [1 3 2]);
2630                 time = (1:numel(enzymeRNA)) / 3600;
2631                 
2632                 save([outDirectory <span class="string">'proteinBursting.mat'</span>], <span class="keyword">...</span>
2633                     <span class="string">'time'</span>, <span class="string">'enzymeRNA'</span>, <span class="string">'enzymeMonomer'</span>);
2634             <span class="keyword">end</span>
2635             
2636             
2637             <span class="comment">%% plot</span>
2638             options = struct;
2639             <span class="keyword">if</span> nargin &gt;= 4
2640                 options.position = position;
2641             <span class="keyword">else</span>
2642                 [~, figHandle] = edu.stanford.covert.cell.sim.util.PlotUtil.newAxesHandle();
2643                 clf(figHandle);
2644             <span class="keyword">end</span>
2645             
2646             options.xdata = time;
2647             options.ydata = {
2648                 enzymeMonomer
2649                 enzymeRNA
2650                 };
2651             options.ylabelStr = {
2652                 {<span class="string">'Protein (cnt)'</span>}
2653                 {<span class="string">'mRNA (cnt)'</span>}
2654                 };
2655             options.offsetYAxes = -0.02;
2656             [axesHandles, xAxisHandle, ~, offsetAxesHandles] = PlotUtil.multiElementPlot(figHandle, 8 * ones(2, 1), [0 time(end)], options);
2657             set(xAxisHandle, <span class="string">'XTick'</span>, 0:2:8, <span class="string">'FontSize'</span>, 5);
2658             set(get(xAxisHandle, <span class="string">'XLabel'</span>), <span class="string">'FontSize'</span>, 7);
2659             xlabelPos = get(get(xAxisHandle, <span class="string">'xlabel'</span>), <span class="string">'position'</span>);
2660             set(get(xAxisHandle, <span class="string">'xlabel'</span>), <span class="string">'position'</span>, [xlabelPos(1) -151 xlabelPos(3:end)]);
2661             
2662             tick2text(xAxisHandle, <span class="string">'axis'</span>, <span class="string">'x'</span>, <span class="string">'xtickoffset'</span>, 50);
2663             xTicks = getappdata(xAxisHandle, <span class="string">'XTickText'</span>);
2664             set(xTicks, <span class="string">'FontSize'</span>, 5);
2665             delete(xTicks(2:4));
2666             
2667             <span class="keyword">for</span> i = 1:numel(options.ydata)
2668                 set(get(offsetAxesHandles(i), <span class="string">'ylabel'</span>), <span class="string">'FontSize'</span>, 7)
2669                 set(offsetAxesHandles(i), <span class="string">'FontSize'</span>, 5)
2670                 set(axesHandles(i), <span class="string">'FontSize'</span>, 5)
2671                 
2672                 yTicks = getappdata(offsetAxesHandles(i), <span class="string">'YTickText'</span>);
2673                 <span class="keyword">if</span> isempty(yTicks)
2674                     tick2text(offsetAxesHandles(i), <span class="string">'axis'</span>, <span class="string">'y'</span>, <span class="string">'ytickoffset'</span>, 0.14)
2675                     yTicks = getappdata(offsetAxesHandles(i), <span class="string">'YTickText'</span>);
2676                 <span class="keyword">end</span>
2677                 
2678                 set(yTicks, <span class="string">'FontSize'</span>, 5);
2679                 <span class="keyword">for</span> j = 1:numel(yTicks)
2680                     pos = get(yTicks(j), <span class="string">'position'</span>);
2681                     set(yTicks(j), <span class="string">'position'</span>, [-0.18 pos(2:end)], <span class="string">'horizontalalign'</span>, <span class="string">'right'</span>);
2682                 <span class="keyword">end</span>
2683                 
2684                 ylabelPos = get(get(offsetAxesHandles(i), <span class="string">'ylabel'</span>), <span class="string">'position'</span>);
2685                 set(get(offsetAxesHandles(i), <span class="string">'ylabel'</span>), <span class="string">'position'</span>, [-0.6 ylabelPos(2:end)], <span class="string">'FontSize'</span>, 7, <span class="string">'VerticalAlign'</span>, <span class="string">'bottom'</span>);
2686             <span class="keyword">end</span>
2687             
2688             burstIdxs = find(diff(enzymeRNA) &gt; 0);
2689             axesPos1 = get(axesHandles(1), <span class="string">'position'</span>);
2690             axesPos2 = get(axesHandles(2), <span class="string">'position'</span>);
2691             monYLims = ylim(axesHandles(1));
2692             rnaYLims = ylim(axesHandles(2));
2693             <span class="keyword">for</span> i = 1:numel(burstIdxs)
2694                 y = (axesPos2(2):0.0025:(axesPos1(2) + axesPos1(4) * (enzymeMonomer(burstIdxs(i)) - monYLims(1)) / range(monYLims)));
2695                 x = (axesPos2(1) + axesPos2(3) * time(burstIdxs(i))/time(end)) * [1 1];
2696                 <span class="keyword">for</span> j = 1:3:3 * floor(numel(y) / 3)
2697                     <span class="comment">%if any(...</span>
2698                     <span class="comment">%        (y(j:j+1)-axesPos2(2))/axesPos2(4)*range(rnaYLims) + rnaYLims(1) &gt;= enzymeRNA(burstIdxs(i)) &amp; ...</span>
2699                     <span class="comment">%        (y(j:j+1)-axesPos2(2))/axesPos2(4)*range(rnaYLims) + rnaYLims(1) &lt;= enzymeRNA(burstIdxs(i)+1))</span>
2700                     <span class="comment">%    continue;</span>
2701                     <span class="comment">%end</span>
2702                     annotation(figHandle, <span class="string">'line'</span>, x, y(j:j+1), <span class="string">'Color'</span>, [1 0 0]);
2703                 <span class="keyword">end</span>
2704             <span class="keyword">end</span>
2705         <span class="keyword">end</span>
2706         
2707         <a name="_sub18" href="#_subfunctions" class="code">function geneToPhenotype(simBatchDir, selectedSim, outDirectory, figHandle, position)</a>
2708             import edu.stanford.covert.cell.sim.analysis.SingleCell;
2709             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
2710             import edu.stanford.covert.cell.sim.util.PlotUtil;
2711             
2712             <span class="keyword">if</span> ~exist(<span class="string">'simBatchDir'</span>, <span class="string">'var'</span>) || isempty(simBatchDir)
2713                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
2714             <span class="keyword">end</span>
2715             
2716             <span class="comment">%% get data</span>
2717             <span class="keyword">if</span> exist([outDirectory <span class="string">'geneToPhenotype.mat'</span>], <span class="string">'file'</span>)
2718                 load([outDirectory <span class="string">'geneToPhenotype.mat'</span>]);
2719             <span class="keyword">else</span>
2720                 [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep num2str(selectedSim)]);
2721                 
2722                 [time, ~, ~, ~, ~, ~, ~, growth, reactionFlux, <span class="keyword">...</span>
2723                     enzymeComplex, enzymeMonomer, enzymeRNA, enzymeGene] = <span class="keyword">...</span>
2724                     SingleCell.calcGrowthData(sim, simBatchDir, selectedSim, <span class="string">'GalE'</span>);
2725                 enzymeGene = permute(enzymeGene, [1 3 2]);
2726                 enzymeRNA = permute(enzymeRNA, [1 3 2]);
2727                 enzymeMonomer = permute(enzymeMonomer, [1 3 2]);
2728                 enzymeComplex = permute(enzymeComplex, [1 3 2]);
2729                 reactionFlux = permute(reactionFlux, [1 3 2]);
2730                 growth = permute(growth, [1 3 2]);
2731                 
2732                 save([outDirectory <span class="string">'geneToPhenotype.mat'</span>], <span class="keyword">...</span>
2733                     <span class="string">'time'</span>, <span class="string">'enzymeGene'</span>, <span class="string">'enzymeRNA'</span>, <span class="string">'enzymeMonomer'</span>, <span class="string">'enzymeComplex'</span>, <span class="string">'reactionFlux'</span>, <span class="string">'growth'</span>);
2734             <span class="keyword">end</span>
2735             reactionFlux = -reactionFlux * 1000;
2736             
2737             <span class="comment">%% plot</span>
2738             options = struct;
2739             <span class="keyword">if</span> nargin &gt;= 4
2740                 options.position = position;
2741             <span class="keyword">else</span>
2742                 [~, figHandle] = edu.stanford.covert.cell.sim.util.PlotUtil.newAxesHandle();
2743                 clf(figHandle);
2744             <span class="keyword">end</span>
2745             
2746             options.xdata = time;
2747             options.ydata = {
2748                 growth
2749                 reactionFlux
2750                 enzymeComplex
2751                 enzymeMonomer
2752                 enzymeRNA
2753                 enzymeGene
2754                 };
2755             options.ylabelStr = {
2756                 {<span class="string">'Growth'</span> <span class="string">'(fg/h)'</span>}
2757                 {<span class="string">'Flux'</span> <span class="string">'(10^3 rxn/h)'</span>}
2758                 {<span class="string">'Cpx'</span> <span class="string">'(cnt)'</span>}
2759                 {<span class="string">'Mon'</span> <span class="string">'(cnt)'</span>}
2760                 {<span class="string">'mRNA'</span> <span class="string">'(cnt)'</span>}
2761                 {<span class="string">'Gene'</span> <span class="string">'(cnt)'</span>}
2762                 };
2763             options.offsetYAxes = false;
2764             [axesHandles, xAxisHandle, ~, offsetAxesHandles] = PlotUtil.multiElementPlot(figHandle, 2 * ones(6, 1), time([1 end])', options);
2765             set(xAxisHandle, <span class="string">'XTick'</span>, 0:2:8, <span class="string">'FontSize'</span>, 5);
2766             set(get(xAxisHandle, <span class="string">'XLabel'</span>), <span class="string">'FontSize'</span>, 7);
2767             xlabelPos = get(get(xAxisHandle, <span class="string">'xlabel'</span>), <span class="string">'position'</span>);
2768             set(get(xAxisHandle, <span class="string">'xlabel'</span>), <span class="string">'position'</span>, [xlabelPos(1) -80 xlabelPos(3:end)]);
2769             
2770             tick2text(xAxisHandle, <span class="string">'axis'</span>, <span class="string">'x'</span>, <span class="string">'xtickoffset'</span>, 60);
2771             xTicks = getappdata(xAxisHandle, <span class="string">'XTickText'</span>);
2772             set(xTicks, <span class="string">'FontSize'</span>, 5);
2773             delete(xTicks(2:4));
2774             
2775             <span class="keyword">for</span> i = 1:numel(axesHandles)
2776                 set(get(axesHandles(i), <span class="string">'ylabel'</span>), <span class="string">'FontSize'</span>, 7)
2777                 set(axesHandles(i), <span class="string">'FontSize'</span>, 5)
2778                 
2779                 yTicks = getappdata(axesHandles(i), <span class="string">'YTickText'</span>);
2780                 <span class="keyword">if</span> isempty(yTicks)
2781                     tick2text(axesHandles(i), <span class="string">'axis'</span>, <span class="string">'y'</span>, <span class="string">'ytickoffset'</span>, 0.06)
2782                     yTicks = getappdata(axesHandles(i), <span class="string">'YTickText'</span>);
2783                 <span class="keyword">end</span>
2784                 
2785                 set(yTicks, <span class="string">'FontSize'</span>, 5);
2786                 <span class="keyword">for</span> j = 1:numel(yTicks)
2787                     pos = get(yTicks(j), <span class="string">'position'</span>);
2788                     set(yTicks(j), <span class="string">'position'</span>, [-0.15 pos(2:end)], <span class="string">'horizontalalign'</span>, <span class="string">'right'</span>);
2789                 <span class="keyword">end</span>
2790                 
2791                 ylabelPos = get(get(axesHandles(i), <span class="string">'ylabel'</span>), <span class="string">'position'</span>);
2792                 set(get(axesHandles(i), <span class="string">'ylabel'</span>), <span class="string">'position'</span>, [-0.8 ylabelPos(2:end)], <span class="string">'FontSize'</span>, 5, <span class="string">'VerticalAlign'</span>, <span class="string">'bottom'</span>);
2793             <span class="keyword">end</span>
2794             
2795             ylabelPos = get(get(axesHandles(1), <span class="string">'ylabel'</span>), <span class="string">'position'</span>);
2796             set(get(axesHandles(1), <span class="string">'ylabel'</span>), <span class="string">'position'</span>, [ylabelPos(1) 1.65 ylabelPos(3)]);
2797         <span class="keyword">end</span>
2798         
2799         <a name="_sub19" href="#_subfunctions" class="code">function spaceTimePlot(simBatchDir, selectedSim, outDirectory, figHandle, position)</a>
2800             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
2801             import edu.stanford.covert.cell.sim.util.DiskLogger;
2802             import edu.stanford.covert.cell.sim.util.PlotUtil;
2803             import edu.stanford.covert.cell.sim.analysis.ChromosomeSpaceTimePlot;
2804             
2805             <span class="keyword">if</span> ~exist(<span class="string">'simBatchDir'</span>, <span class="string">'var'</span>) || isempty(simBatchDir)
2806                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
2807             <span class="keyword">end</span>
2808             <span class="keyword">if</span> ~exist(<span class="string">'selectedSim'</span>, <span class="string">'var'</span>) || isempty(selectedSim)
2809                 selectedSim = 1;
2810             <span class="keyword">end</span>
2811             
2812             [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep num2str(selectedSim)]);
2813             c = sim.state(<span class="string">'Chromosome'</span>);
2814             
2815             <span class="comment">%% get data</span>
2816             <span class="keyword">if</span> exist([outDirectory <span class="string">'spaceTimePlot.mat'</span>], <span class="string">'file'</span>)
2817                 load([outDirectory <span class="string">'spaceTimePlot.mat'</span>])
2818             <span class="keyword">else</span>
2819                 [axesHandle, figHandle] = edu.stanford.covert.cell.sim.util.PlotUtil.newAxesHandle();
2820                 [ensemble, states, dnaABoxBound] = ChromosomeSpaceTimePlot.plotSpaceTime(<span class="keyword">...</span>
2821                     axesHandle, simBatchDir, selectedSim, [], false, [], false, true); <span class="comment">%#ok&lt;*NASGU,*ASGLU&gt;</span>
2822                 save([outDirectory <span class="string">'spaceTimePlot.mat'</span>], <span class="string">'ensemble'</span>, <span class="string">'states'</span>, <span class="string">'dnaABoxBound'</span>)
2823                 close(figHandle);
2824             <span class="keyword">end</span>
2825             
2826             <span class="comment">%% space-time density plots</span>
2827             <span class="comment">%create axes</span>
2828             <span class="keyword">if</span> nargin &lt; 4
2829                 [~, figHandle] = edu.stanford.covert.cell.sim.util.PlotUtil.newAxesHandle();
2830                 position = [0.1 0.1 0.8 0.8];
2831             <span class="keyword">end</span>
2832             
2833             x = position(1);
2834             y = position(2);
2835             w = position(3);
2836             h = position(4);
2837             spcg = 0.08;
2838             figW = 1.45 * (0.5 - x);
2839             
2840             axesHandles = [
2841                 subplot(<span class="string">'position'</span>, [x            y  figW  h], <span class="string">'parent'</span>, figHandle)
2842                 subplot(<span class="string">'position'</span>, [x+figW+spcg  y  w-figW-spcg  h], <span class="string">'parent'</span>, figHandle)
2843                 ];
2844             
2845             zoomTime = [1.34 1.41];
2846             zoomPos = [267000 273000];
2847             time = ensemble.stateData.time / 3600;
2848             
2849             zoomBoxHandle = [
2850                 annotation(figHandle, <span class="string">'line'</span>, x + [figW*zoomTime(2)/time(end) figW+spcg], y + [h*zoomPos(end)/c.sequenceLen h])
2851                 annotation(figHandle, <span class="string">'line'</span>, x + [figW*zoomTime(2)/time(end) figW+spcg], y + [h*zoomPos(1)/c.sequenceLen 0])
2852                 annotation(figHandle, <span class="string">'rectangle'</span>, [
2853                 x + figW*zoomTime(1)/time(end)
2854                 y + h*zoomPos(1)/c.sequenceLen
2855                 figW*diff(zoomTime)/time(end)
2856                 h*diff(zoomPos)/c.sequenceLen
2857                 ]')
2858                 ];
2859             set(zoomBoxHandle, <span class="string">'LineWidth'</span>, 1);
2860             
2861             <span class="comment">%% plot 1</span>
2862             axesHandle = axesHandles(1);
2863             
2864             <span class="comment">%plot</span>
2865             
2866             [~, ~, ~, legendHandle] = ChromosomeSpaceTimePlot.plotSpaceTime(<span class="keyword">...</span>
2867                 axesHandle, simBatchDir, selectedSim, [], false, [], <span class="keyword">...</span>
2868                 false, true, ensemble, states, dnaABoxBound, 2, 0.25, false);
2869             
2870             <span class="comment">%style</span>
2871             xlim(axesHandle, [0 time(end)]);
2872             ylim(axesHandle, [1 c.sequenceLen]);
2873             set(axesHandle, <span class="string">'TickDir'</span>, <span class="string">'out'</span>, <span class="string">'XTick'</span>, 0:2:8, <span class="string">'FontSize'</span>, 5);
2874             
2875             tick2text(axesHandle, <span class="string">'axis'</span>, <span class="string">'xy'</span>, <span class="string">'xtickoffset'</span>, 0.035, <span class="string">'ytickoffset'</span>, 0.008);
2876             xTicks = getappdata(axesHandle, <span class="string">'XTickText'</span>);
2877             yTicks = getappdata(axesHandle, <span class="string">'YTickText'</span>);
2878             set(xTicks, <span class="string">'FontSize'</span>, 5);
2879             set(yTicks, <span class="string">'FontSize'</span>, 5, <span class="string">'HorizontalAlign'</span>, <span class="string">'right'</span>);
2880             delete(xTicks(2:4));
2881             set(yTicks([1 3]), <span class="string">'String'</span>, <span class="string">'terC'</span>);
2882             set(yTicks(2), <span class="string">'String'</span>, <span class="string">'oriC'</span>);
2883             
2884             xlabel(axesHandle, <span class="string">'Time (h)'</span>, <span class="string">'FontSize'</span>, 7);
2885             ylabel(axesHandle, <span class="string">'Position'</span>, <span class="string">'FontSize'</span>, 7);
2886             xLabelPos = get(get(axesHandle, <span class="string">'xlabel'</span>), <span class="string">'position'</span>);
2887             yLabelPos = get(get(axesHandle, <span class="string">'ylabel'</span>), <span class="string">'position'</span>);
2888             set(get(axesHandle, <span class="string">'xlabel'</span>), <span class="string">'position'</span>, [xLabelPos(1) -25000 xLabelPos(3)])
2889             set(get(axesHandle, <span class="string">'ylabel'</span>), <span class="string">'position'</span>, [-0.5 yLabelPos(2:3)])
2890             
2891             <span class="comment">%legend</span>
2892             legendChildren = get(legendHandle, <span class="string">'children'</span>);
2893             set(legendHandle, <span class="string">'FontSize'</span>, 7);
2894             legendPos = get(legendHandle, <span class="string">'Position'</span>);
2895             set(legendHandle, <span class="string">'Position'</span>, [0.037 0.346 legendPos(3:4)]);
2896             set(legendHandle, <span class="string">'units'</span>, <span class="string">'pixels'</span>);
2897             legendPos = get(legendHandle, <span class="string">'Position'</span>);
2898             set(legendHandle, <span class="string">'Position'</span>, [legendPos(1:2) 180 30]);
2899             tmp = findobj(legendChildren, <span class="string">'type'</span>, <span class="string">'line'</span>);
2900             set(tmp(1:2:end), <span class="string">'visible'</span>, <span class="string">'off'</span>);
2901             set(tmp(2:2:end), <span class="string">'visible'</span>, <span class="string">'on'</span>, <span class="string">'LineWidth'</span>, 4)
2902             set(tmp(2:2:end), <span class="string">'XData'</span>, [0.07 0.18])
2903             tmp = findobj(legendChildren, <span class="string">'type'</span>, <span class="string">'text'</span>);
2904             <span class="keyword">for</span> i = 1:numel(tmp)
2905                 pos = get(tmp(i), <span class="string">'Position'</span>);
2906                 set(tmp(i), <span class="string">'Position'</span>, [0.24 pos(2) 0])
2907             <span class="keyword">end</span>
2908             tmp = findobj(legendChildren, <span class="string">'type'</span>, <span class="string">'patch'</span>);
2909             line([0.07 0.18], [min(get(tmp, <span class="string">'ydata'</span>)) max(get(tmp, <span class="string">'ydata'</span>))], <span class="string">'LineWidth'</span>, 4, <span class="string">'Parent'</span>, legendHandle, <span class="string">'Color'</span>, <span class="string">'r'</span>);
2910             set(tmp, <span class="string">'visible'</span>, <span class="string">'off'</span>);
2911             
2912             set(legendHandle, <span class="string">'units'</span>, <span class="string">'pixels'</span>);
2913             legendPos = get(legendHandle, <span class="string">'position'</span>);
2914             set(legendHandle, <span class="string">'position'</span>, [[1.4 1.53] .* legendPos(1:2) 0.47*legendPos(3) legendPos(4)]);
2915             
2916             <span class="comment">%% plot 2</span>
2917             axesHandle = axesHandles(2);
2918             
2919             <span class="comment">%plot</span>
2920             ChromosomeSpaceTimePlot.plotSpaceTime(axesHandle, simBatchDir, selectedSim, [], <span class="keyword">...</span>
2921                 false, [], false, false, ensemble, states, dnaABoxBound, 0.5, 0.5, true); <span class="comment">%#ok&lt;*NODEF&gt;</span>
2922             
2923             xlim(axesHandle, zoomTime);
2924             ylim(axesHandle, zoomPos);
2925             
2926             <span class="comment">%style</span>
2927             set(axesHandle, <span class="string">'TickDir'</span>, <span class="string">'out'</span>, <span class="keyword">...</span>
2928                 <span class="string">'XTick'</span>, linspace(zoomTime(1), zoomTime(2), 8), <span class="keyword">...</span>
2929                 <span class="string">'YTick'</span>, linspace(zoomPos(1), zoomPos(2), 7)-38, <span class="keyword">...</span>
2930                 <span class="string">'FontSize'</span>, 5);
2931             
2932             tick2text(axesHandle, <span class="string">'axis'</span>, <span class="string">'xy'</span>, <span class="string">'xtickoffset'</span>, 0.035, <span class="string">'ytickoffset'</span>, 0.015);
2933             xTicks = getappdata(axesHandle, <span class="string">'XTickText'</span>);
2934             yTicks = getappdata(axesHandle, <span class="string">'YTickText'</span>);
2935             set(xTicks, <span class="string">'FontSize'</span>, 5);
2936             set(yTicks, <span class="string">'FontSize'</span>, 5, <span class="string">'HorizontalAlign'</span>, <span class="string">'right'</span>);
2937             set(yTicks(2), <span class="string">'String'</span>, <span class="string">'558000'</span>)
2938             set(yTicks(4), <span class="string">'String'</span>, <span class="string">'560000'</span>)
2939             set(yTicks(6), <span class="string">'String'</span>, <span class="string">'562000'</span>)
2940             delete(xTicks(setdiff(1:<span class="keyword">end</span>, [2 7])));
2941             delete(yTicks(1:2:end));
2942             
2943             xlabel(axesHandle, <span class="string">'Time (h)'</span>, <span class="string">'FontSize'</span>, 7);
2944             ylabel(axesHandle, <span class="string">'Position (nt)'</span>, <span class="string">'FontSize'</span>, 7);
2945             xLabelPos = get(get(axesHandle, <span class="string">'xlabel'</span>), <span class="string">'position'</span>);
2946             yLabelPos = get(get(axesHandle, <span class="string">'ylabel'</span>), <span class="string">'position'</span>);
2947             set(get(axesHandle, <span class="string">'xlabel'</span>), <span class="string">'position'</span>, [xLabelPos(1) 266750 xLabelPos(3)])
2948             set(get(axesHandle, <span class="string">'yLabel'</span>), <span class="string">'position'</span>, [1.321 yLabelPos(2:end)])
2949             
2950             figPos = get(figHandle, <span class="string">'PaperSize'</span>);
2951             figW = figPos(1);
2952             figH = figPos(2);
2953             axisPos = get(axesHandle, <span class="string">'position'</span>);
2954             x = axisPos(1);
2955             y = axisPos(2);
2956             w = axisPos(3);
2957             h = axisPos(4);
2958             annotation(figHandle, <span class="string">'ellipse'</span>, [x+0.595*w  y+0.645*h  0.01 0.01*figW/figH], <span class="string">'Color'</span>, <span class="string">'k'</span>)
2959             annotation(figHandle, <span class="string">'ellipse'</span>, [x+0.775*w  y+0.245*h  0.01 0.01*figW/figH], <span class="string">'Color'</span>, <span class="string">'k'</span>)
2960             annotation(figHandle, <span class="string">'ellipse'</span>, [x+0.48*w  y+0.875*h  0.01 0.01*figW/figH], <span class="string">'Color'</span>, <span class="string">'k'</span>)
2961             annotation(figHandle, <span class="string">'ellipse'</span>, [x+0.645*w  y+0.84*h  0.01 0.01*figW/figH], <span class="string">'Color'</span>, <span class="string">'k'</span>)
2962             annotation(figHandle, <span class="string">'ellipse'</span>, [x+0.655*w  y+0.86*h  0.01 0.01*figW/figH], <span class="string">'Color'</span>, <span class="string">'k'</span>)
2963             annotation(figHandle, <span class="string">'arrow'</span>, [x+0.535*w  x+0.60*w], [y+0.57*h  y+0.63*h], <span class="string">'Color'</span>, <span class="string">'k'</span>, <span class="keyword">...</span>
2964                 <span class="string">'HeadLength'</span>, 3, <span class="string">'HeadWidth'</span>, 5, <span class="string">'HeadStyle'</span>, <span class="string">'vback1'</span>)
2965             text(1.364, 270100, {<span class="string">'DNA pol-RNA'</span> <span class="string">'pol collision'</span>}, <span class="string">'parent'</span>, axesHandle, <span class="keyword">...</span>
2966                 <span class="string">'FontSize'</span>, 6, <span class="string">'VerticalAlign'</span>, <span class="string">'middle'</span>, <span class="string">'HorizontalAlign'</span>, <span class="string">'center'</span>, <span class="string">'Color'</span>, <span class="string">'k'</span>);
2967             text(1.3860, 267400, {<span class="string">'Leading'</span> <span class="string">'DNA pol'</span>}, <span class="string">'parent'</span>, axesHandle, <span class="keyword">...</span>
2968                 <span class="string">'FontSize'</span>, 6, <span class="string">'VerticalAlign'</span>, <span class="string">'middle'</span>, <span class="string">'HorizontalAlign'</span>, <span class="string">'center'</span>, <span class="string">'Color'</span>, <span class="string">'g'</span>);
2969             text(1.4020, 272500, {<span class="string">'Lagging'</span> <span class="string">'DNA pol'</span>}, <span class="string">'parent'</span>, axesHandle, <span class="keyword">...</span>
2970                 <span class="string">'FontSize'</span>, 6, <span class="string">'VerticalAlign'</span>, <span class="string">'middle'</span>, <span class="string">'HorizontalAlign'</span>, <span class="string">'center'</span>, <span class="string">'Color'</span>, <span class="string">'g'</span>);
2971             
2972             <span class="comment">%% zoom box</span>
2973             uistack(zoomBoxHandle, <span class="string">'top'</span>);
2974         <span class="keyword">end</span>
2975         
2976         <a name="_sub20" href="#_subfunctions" class="code">function metaboliteConcentrations(simBatchDir, selectedSim, outDirectory, figHandle, position)</a>
2977             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
2978             import edu.stanford.covert.cell.sim.util.PlotUtil;
2979             import edu.stanford.covert.cell.sim.util.PrintUtil;
2980             import edu.stanford.covert.cell.sim.util.SummaryLogger;
2981             
2982             <span class="keyword">if</span> ~exist(<span class="string">'simBatchDir'</span>, <span class="string">'var'</span>) || isempty(simBatchDir)
2983                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
2984             <span class="keyword">end</span>
2985             
2986             <span class="comment">%% get constants</span>
2987             [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep num2str(selectedSim)]);
2988             m = sim.state(<span class="string">'Metabolite'</span>);
2989             metIdxs = [m.ntpIndexs; m.ndpIndexs; m.nmpIndexs; m.dntpIndexs; m.aminoAcidIndexs; m.phosphateIndexs; m.diphosphateIndexs; m.hydrogenIndexs];
2990             
2991             metIDs = m.wholeCellModelIDs;
2992             metIDs(m.dntpIndexs) = {<span class="string">'dATP'</span>, <span class="string">'dCTP'</span>, <span class="string">'dGTP'</span>, <span class="string">'dTTP'</span>};
2993             metIDs(m.aminoAcidIndexs) = {
2994                 <span class="string">'Ala'</span>
2995                 <span class="string">'Arg'</span>
2996                 <span class="string">'Asn'</span>
2997                 <span class="string">'Asp'</span>
2998                 <span class="string">'Cys'</span>
2999                 <span class="string">'Gln'</span>
3000                 <span class="string">'Glu'</span>
3001                 <span class="string">'Gly'</span>
3002                 <span class="string">'His'</span>
3003                 <span class="string">'Ile'</span>
3004                 <span class="string">'Leu'</span>
3005                 <span class="string">'Lys'</span>
3006                 <span class="string">'Met'</span>
3007                 <span class="string">'Phe'</span>
3008                 <span class="string">'Pro'</span>
3009                 <span class="string">'Ser'</span>
3010                 <span class="string">'Thr'</span>
3011                 <span class="string">'Trp'</span>
3012                 <span class="string">'Tyr'</span>
3013                 <span class="string">'Val'</span>
3014                 <span class="string">'fMet'</span>
3015                 };
3016             metIDs{m.phosphateIndexs} = <span class="string">'Pi'</span>;
3017             metIDs{m.diphosphateIndexs} = <span class="string">'PPi'</span>;
3018             metIDs{m.hydrogenIndexs} = <span class="string">'H^+'</span>;
3019             
3020             <span class="comment">%% get data</span>
3021             tmp = load([SimulationDiskUtil.getSimulationBatchDir([simBatchDir filesep <span class="string">'1'</span>]) filesep <span class="string">'population-MetaboliteConcentrations.mat'</span>]);
3022             tmp2 = [tmp.meanConcs tmp.stdConcs];
3023             modelConcs = zeros(numel(m.wholeCellModelIDs), 2);
3024             modelConcs(m.aminoAcidIndexs, :) = tmp2(17:37, :);
3025             modelConcs(m.ntpIndexs, :) = tmp2(1:4, :);
3026             modelConcs(m.ndpIndexs, :) = tmp2(5:8, :);
3027             modelConcs(m.nmpIndexs, :) = tmp2(9:12, :);
3028             modelConcs(m.dntpIndexs, :) = tmp2(13:16, :);
3029             modelConcs(m.phosphateIndexs, :) = tmp2(38, :);
3030             modelConcs(m.diphosphateIndexs, :) = tmp2(39, :);
3031             modelConcs(m.hydrogenIndexs, :) = tmp2(40, :);
3032             
3033             [~, ~, data] = xlsread(<span class="string">'documentation/reconstruction/Metabolite Concentrations.xls'</span>);
3034             tfs = ~cellfun(@(x) any(isnan(x)), data(:, 1));
3035             tfs(1:2) = false;
3036             data = data(tfs, :);
3037             
3038             [~, idxs] = ismember(data(:, 1), m.wholeCellModelIDs);
3039             expConcs = NaN(size(modelConcs, 1), 3);
3040             cyberCellConc = NaN(size(modelConcs, 1), 1);
3041             expConcs(idxs, :) = cell2mat(data(:, 2:4));
3042             cyberCellConc(idxs, :) = cell2mat(data(:, 5));
3043             
3044             sortedIdxs = [
3045                 m.aminoAcidIndexs(1:end-1)
3046                 m.ntpIndexs
3047                 m.ndpIndexs
3048                 m.nmpIndexs
3049                 m.dntpIndexs
3050                 m.phosphateIndexs
3051                 m.diphosphateIndexs
3052                 m.hydrogenIndexs
3053                 ];
3054             allSortedIdxs = [
3055                 m.dntpIndexs
3056                 m.ndpIndexs
3057                 m.nmpIndexs
3058                 m.ntpIndexs
3059                 m.aminoAcidIndexs
3060                 m.phosphateIndexs
3061                 m.diphosphateIndexs
3062                 m.hydrogenIndexs
3063                 ];
3064             
3065             <span class="comment">%% plot</span>
3066             <span class="keyword">if</span> nargin &gt;= 4
3067                 axesHandle = subplot(<span class="string">'position'</span>, position, <span class="string">'parent'</span>, figHandle);
3068             <span class="keyword">else</span>
3069                 axesHandle = edu.stanford.covert.cell.sim.util.PlotUtil.newAxesHandle();
3070             <span class="keyword">end</span>
3071             
3072             hold(axesHandle, <span class="string">'on'</span>);
3073             set(axesHandle, <span class="keyword">...</span>
3074                 <span class="string">'YScale'</span>, <span class="string">'log'</span>, <span class="keyword">...</span>
3075                 <span class="string">'YMinorTick'</span>, <span class="string">'on'</span>, <span class="keyword">...</span>
3076                 <span class="string">'XTick'</span>, 1:numel(sortedIdxs), <span class="keyword">...</span>
3077                 <span class="string">'XTickLabel'</span>, m.wholeCellModelIDs(sortedIdxs), <span class="keyword">...</span>
3078                 <span class="string">'YTick'</span>, logspace(-2, 2, 5), <span class="keyword">...</span>
3079                 <span class="string">'FontSize'</span>, 5, <span class="keyword">...</span>
3080                 <span class="string">'TickDir'</span>, <span class="string">'out'</span>);
3081             xlim(axesHandle, [0.5 numel(sortedIdxs)+0.5]);
3082             ylim(axesHandle, [1e-2/1.5 1.5e2]);
3083             
3084             plot(axesHandle, 1:numel(sortedIdxs), expConcs(sortedIdxs, 1) ./ modelConcs(sortedIdxs, 1), <span class="keyword">...</span>
3085                 <span class="string">'rd'</span>, <span class="string">'MarkerSize'</span>, 3, <span class="string">'MarkerFaceColor'</span>, <span class="string">'r'</span>, <span class="string">'MarkerEdgeColor'</span>, <span class="string">'r'</span>);
3086             plot(axesHandle, 1:numel(sortedIdxs), cyberCellConc(sortedIdxs) ./ modelConcs(sortedIdxs, 1), <span class="keyword">...</span>
3087                 <span class="string">'gs'</span>, <span class="string">'MarkerSize'</span>, 3, <span class="string">'MarkerFaceColor'</span>, <span class="string">'g'</span>, <span class="string">'MarkerEdgeColor'</span>, <span class="string">'g'</span>);
3088             nModel = numel(sortedIdxs);
3089             line([1:nModel; 1:nModel], max([ 1 + (modelConcs(sortedIdxs, 2) ./ modelConcs(sortedIdxs, 1))'; 1 - (modelConcs(sortedIdxs, 2) ./ modelConcs(sortedIdxs, 1))'], eps), <span class="string">'Color'</span>, <span class="string">'b'</span>);
3090             
3091             tick2text(axesHandle, <span class="string">'axis'</span>, <span class="string">'xy'</span>, <span class="string">'xtickoffset'</span>, 0.13, <span class="string">'ytickoffset'</span>, 0.013);
3092             xTicks = getappdata(axesHandle, <span class="string">'XTickText'</span>);
3093             yTicks = getappdata(axesHandle, <span class="string">'YTickText'</span>);
3094             set(xTicks, <span class="string">'FontSize'</span>, 5, <span class="string">'HorizontalAlign'</span>, <span class="string">'left'</span>, <span class="string">'VerticalAlign'</span>, <span class="string">'middle'</span>, <span class="string">'Interpreter'</span>, <span class="string">'tex'</span>, <span class="string">'rotation'</span>, 270);
3095             set(yTicks, <span class="string">'FontSize'</span>, 5, <span class="string">'HorizontalAlign'</span>, <span class="string">'right'</span>, <span class="string">'Interpreter'</span>, <span class="string">'tex'</span>);
3096             <span class="keyword">for</span> i = 1:numel(xTicks)
3097                 set(xTicks(i), <span class="string">'String'</span>, metIDs{sortedIdxs(i)}, <span class="string">'position'</span>, [i 5e-3 -1], <span class="string">'Margin'</span>, 1e-6);
3098             <span class="keyword">end</span>
3099             <span class="keyword">for</span> i = 1:numel(yTicks)
3100                 set(yTicks(i), <span class="string">'String'</span>, sprintf(<span class="string">'10^{%d}'</span>, log10(str2double(get(yTicks(i), <span class="string">'string'</span>)))));
3101             <span class="keyword">end</span>
3102             
3103             ylabel(axesHandle, <span class="string">'Expt/model concentration'</span>, <span class="string">'FontSize'</span>, 7);
3104             yLabelPos = get(get(axesHandle, <span class="string">'ylabel'</span>), <span class="string">'position'</span>);
3105             set(get(axesHandle, <span class="string">'ylabel'</span>), <span class="string">'position'</span>, [-1.6 0.8 yLabelPos(3)])
3106             
3107             legendHandle = legend({<span class="string">'Bennett {\it{et al.}}, 2009'</span>, <span class="string">'Literature (CCDB)'</span>, <span class="string">'Model s.d.'</span>}, <span class="string">'Location'</span>, <span class="string">'SouthEast'</span>, <span class="string">'FontSize'</span>, 6);
3108             legend(axesHandle, <span class="string">'boxoff'</span>);
3109             legendPos = get(legendHandle, <span class="string">'Position'</span>);
3110             set(legendHandle, <span class="string">'Position'</span>, [0.088 0.092 legendPos(3:end)])
3111             set(legendHandle, <span class="string">'units'</span>, <span class="string">'pixels'</span>);
3112             legendPos = get(legendHandle, <span class="string">'Position'</span>);
3113             set(legendHandle, <span class="string">'Position'</span>, [legendPos(1:2) 82 28]);
3114             legendChildren = get(legendHandle, <span class="string">'children'</span>);
3115             <span class="keyword">for</span> i = 1:3:numel(legendChildren)
3116                 <span class="keyword">if</span> strcmp(get(legendChildren(i), <span class="string">'marker'</span>), <span class="string">'none'</span>)
3117                     set(legendChildren(i), <span class="string">'visible'</span>, <span class="string">'off'</span>);
3118                     set(legendChildren(i+1), <span class="string">'xdata'</span>, [0.03 0.10])
3119                 <span class="keyword">else</span>
3120                     set(legendChildren(i+1), <span class="string">'visible'</span>, <span class="string">'off'</span>);
3121                     set(legendChildren(i), <span class="string">'xdata'</span>, 0.06)
3122                 <span class="keyword">end</span>
3123                 
3124                 pos = get(legendChildren(i+2), <span class="string">'Position'</span>);
3125                 set(legendChildren(i+2), <span class="string">'Position'</span>, [0.13 pos(2) 0])
3126             <span class="keyword">end</span>
3127             
3128             axisPos = get(axesHandle, <span class="string">'Position'</span>);
3129             x = axisPos(1);
3130             w = axisPos(3);
3131             y = 0.021;
3132             annotation(figHandle, <span class="string">'line'</span>, x + ([ 1 20]-0.5 + [-0.2 0.2])/39*w, y * [1 1])
3133             annotation(figHandle, <span class="string">'line'</span>, x + ([21 24]-0.5 + [-0.2 0.2])/39*w, y * [1 1])
3134             annotation(figHandle, <span class="string">'line'</span>, x + ([25 28]-0.5 + [-0.2 0.2])/39*w, y * [1 1])
3135             annotation(figHandle, <span class="string">'line'</span>, x + ([29 32]-0.5 + [-0.2 0.2])/39*w, y * [1 1])
3136             annotation(figHandle, <span class="string">'line'</span>, x + ([33 36]-0.5 + [-0.2 0.2])/39*w, y * [1 1])
3137             annotation(figHandle, <span class="string">'line'</span>, x + ([37 39]-0.5 + [-0.2 0.2])/39*w, y * [1 1])
3138             
3139             y = 0.5e-3;
3140             text(mean([ 1 20]), y, <span class="string">'Amino acid'</span>, <span class="string">'FontSize'</span>, 7, <span class="string">'HorizontalAlign'</span>, <span class="string">'center'</span>, <span class="string">'Parent'</span>, axesHandle)
3141             text(mean([21 24]), y, <span class="string">'NTP'</span>,  <span class="string">'FontSize'</span>, 7, <span class="string">'HorizontalAlign'</span>, <span class="string">'center'</span>, <span class="string">'Parent'</span>, axesHandle)
3142             text(mean([25 28]), y, <span class="string">'NDP'</span>,  <span class="string">'FontSize'</span>, 7, <span class="string">'HorizontalAlign'</span>, <span class="string">'center'</span>, <span class="string">'Parent'</span>, axesHandle)
3143             text(mean([29 32]), y, <span class="string">'NMP'</span>,  <span class="string">'FontSize'</span>, 7, <span class="string">'HorizontalAlign'</span>, <span class="string">'center'</span>, <span class="string">'Parent'</span>, axesHandle)
3144             text(mean([33 36]), y, <span class="string">'dNTP'</span>, <span class="string">'FontSize'</span>, 7, <span class="string">'HorizontalAlign'</span>, <span class="string">'center'</span>, <span class="string">'Parent'</span>, axesHandle)
3145             text(mean([37 39]), y, <span class="string">'Ion'</span>,  <span class="string">'FontSize'</span>, 7, <span class="string">'HorizontalAlign'</span>, <span class="string">'center'</span>, <span class="string">'Parent'</span>, axesHandle)
3146             
3147             ptch = patch([0.5 numel(sortedIdxs)+0.5 numel(sortedIdxs)+0.5 0.5], <span class="keyword">...</span>
3148                 [1e1 1e1 1e-1 1e-1], 1, <span class="keyword">...</span>
3149                 <span class="string">'Parent'</span>, axesHandle, <span class="string">'EdgeColor'</span>, <span class="string">'none'</span>);
3150             set(ptch, <span class="string">'FaceColor'</span>, 0.925 * [1 1 1], <span class="string">'FaceAlpha'</span>, 1, <span class="string">'EdgeAlpha'</span>, 1);
3151             h = line([0.5 0.5], [1/2*1e-1 2*1e1], <span class="string">'LineWidth'</span>, 0.5, <span class="string">'Color'</span>, <span class="string">'k'</span>, <span class="string">'Parent'</span>, axesHandle);
3152             uistack(ptch, <span class="string">'bottom'</span>);
3153             uistack(h, <span class="string">'top'</span>);
3154             
3155             <span class="comment">%% table</span>
3156             colLabels = {<span class="string">'Metabolite ID'</span> <span class="string">'Name'</span>, <span class="string">'Whole Cell Model Mean'</span> <span class="string">'Whole Cell Model Std'</span> <span class="string">'Bennet et al. 2009 Mean'</span> <span class="string">'Bennet et al. 2009 Lower Bound'</span> <span class="string">'Bennet et al. 2009 Upper Bound'</span> <span class="string">'CyberCell'</span>};
3157             content = [m.wholeCellModelIDs  m.names  <span class="keyword">...</span>
3158                 num2cell([modelConcs expConcs cyberCellConc])];
3159             content = content(allSortedIdxs, :);
3160             
3161             PrintUtil.printToFile(content, colLabels, [outDirectory <span class="string">'metaboliteConcentrations.xls'</span>], <span class="string">'Metabolite Concentrations'</span>);
3162         <span class="keyword">end</span>
3163         
3164         <a name="_sub21" href="#_subfunctions" class="code">function [exptGeneExpr, modelGeneExp] = geneExpression(simBatchDir, selectedSim, figHandle, position, logScale)</a>
3165             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
3166             import edu.stanford.covert.cell.sim.util.PlotUtil;
3167             
3168             <span class="keyword">if</span> ~exist(<span class="string">'simBatchDir'</span>, <span class="string">'var'</span>) || isempty(simBatchDir)
3169                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
3170             <span class="keyword">end</span>
3171             <span class="keyword">if</span> ~exist(<span class="string">'selectedSim'</span>, <span class="string">'var'</span>) || isempty(selectedSim)
3172                 selectedSim = 1;
3173             <span class="keyword">end</span>
3174             
3175             <span class="comment">%% get constants</span>
3176             [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep num2str(selectedSim)]);
3177             rna = sim.state(<span class="string">'Rna'</span>);
3178             g = sim.gene;
3179             
3180             <span class="comment">%% get data</span>
3181             exptGeneExpr = rna.expectedGeneExpression(:, 1);
3182             exptGeneExpr(exptGeneExpr == 0) = NaN;
3183             exptGeneExpr(g.mRNAIndexs) = exptGeneExpr(g.mRNAIndexs) / nansum(exptGeneExpr(g.mRNAIndexs));
3184             exptGeneExpr(setdiff(1:<span class="keyword">end</span>, g.mRNAIndexs)) = exptGeneExpr(setdiff(1:<span class="keyword">end</span>, g.mRNAIndexs)) / nansum(exptGeneExpr(setdiff(1:<span class="keyword">end</span>, g.mRNAIndexs)));
3185             
3186             modelGeneExp = rna.matureRNAGeneComposition * rna.expression(rna.matureIndexs);
3187             modelGeneExp(g.mRNAIndexs) = modelGeneExp(g.mRNAIndexs) / sum(modelGeneExp(g.mRNAIndexs));
3188             modelGeneExp(setdiff(1:<span class="keyword">end</span>, g.mRNAIndexs)) = modelGeneExp(setdiff(1:<span class="keyword">end</span>, g.mRNAIndexs)) / sum(modelGeneExp(setdiff(1:<span class="keyword">end</span>, g.mRNAIndexs)));
3189             
3190             <span class="comment">%% plot</span>
3191             <span class="keyword">if</span> nargin &gt;= 4
3192                 axesHandle = subplot(<span class="string">'Position'</span>, position, <span class="string">'Parent'</span>, figHandle);
3193             <span class="keyword">else</span>
3194                 axesHandle = edu.stanford.covert.cell.sim.util.PlotUtil.newAxesHandle();
3195             <span class="keyword">end</span>
3196             
3197             <span class="keyword">if</span> logScale
3198                 set(axesHandle, <span class="string">'XScale'</span>, <span class="string">'log'</span>);
3199                 set(axesHandle, <span class="string">'YScale'</span>, <span class="string">'log'</span>);
3200             <span class="keyword">end</span>
3201             set(axesHandle, <span class="string">'TickDir'</span>, <span class="string">'out'</span>);
3202             set(axesHandle, <span class="string">'FontSize'</span>, 5);
3203             hold(axesHandle, <span class="string">'on'</span>);
3204             
3205             plot(axesHandle, exptGeneExpr(g.mRNAIndexs), modelGeneExp(g.mRNAIndexs, 1), <span class="string">'.'</span>, <span class="string">'Color'</span>, [0 1 1])
3206             plot(axesHandle, exptGeneExpr(g.rRNAIndexs), modelGeneExp(g.rRNAIndexs, 1), <span class="string">'.'</span>, <span class="string">'Color'</span>, [0 0 1])
3207             plot(axesHandle, exptGeneExpr(g.sRNAIndexs), modelGeneExp(g.sRNAIndexs, 1), <span class="string">'.'</span>, <span class="string">'Color'</span>, [1 0 0])
3208             plot(axesHandle, exptGeneExpr(g.tRNAIndexs), modelGeneExp(g.tRNAIndexs, 1), <span class="string">'.'</span>, <span class="string">'Color'</span>, [0 1 0])
3209             
3210             <span class="keyword">if</span> logScale
3211                 xlim(axesHandle, [6.5e-5 8e-2]);
3212             <span class="keyword">else</span>
3213                 xlim(axesHandle, [0 7.2e-2])
3214                 set(axesHandle, <span class="string">'XTick'</span>, 0:0.02:0.06, <span class="string">'YTick'</span>, 0:0.02:0.06);
3215             <span class="keyword">end</span>
3216             ylim(axesHandle, xlim(axesHandle));
3217             
3218             tick2text(axesHandle, <span class="string">'axis'</span>, <span class="string">'xy'</span>, <span class="string">'xtickoffset'</span>, 0.02, <span class="string">'ytickoffset'</span>, 0.03);
3219             xTicks = getappdata(axesHandle, <span class="string">'XTickText'</span>);
3220             yTicks = getappdata(axesHandle, <span class="string">'YTickText'</span>);
3221             set(xTicks, <span class="string">'FontSize'</span>, 5);
3222             set(yTicks, <span class="string">'FontSize'</span>, 5);
3223             <span class="keyword">if</span> logScale
3224                 <span class="keyword">for</span> i = 1:numel(xTicks)
3225                     xTickPos = get(xTicks(i), <span class="string">'position'</span>);
3226                     set(xTicks(i), <span class="string">'position'</span>, [xTickPos(1) 5.5e-5 xTickPos(3)], <span class="keyword">...</span>
3227                         <span class="string">'String'</span>, sprintf(<span class="string">'10^{%d}'</span>, log10(str2double(get(xTicks(i), <span class="string">'string'</span>)))), <span class="keyword">...</span>
3228                         <span class="string">'Interpreter'</span>, <span class="string">'tex'</span>);
3229                     
3230                     yTickPos = get(yTicks(i), <span class="string">'position'</span>);
3231                     set(yTicks(i), <span class="string">'position'</span>, [5e-5 yTickPos(2:3)], <span class="keyword">...</span>
3232                         <span class="string">'String'</span>, sprintf(<span class="string">'10^{%d}'</span>, log10(str2double(get(yTicks(i), <span class="string">'string'</span>)))), <span class="keyword">...</span>
3233                         <span class="string">'Interpreter'</span>, <span class="string">'tex'</span>);
3234                 <span class="keyword">end</span>
3235             <span class="keyword">end</span>
3236             
3237             xlabel(axesHandle, {<span class="string">'Observed gene expression (norm)'</span>}, <span class="string">'FontSize'</span>, 7);
3238             ylabel(axesHandle, {<span class="string">'Predicted gene expression (norm)'</span>}, <span class="string">'FontSize'</span>, 7);
3239             xLabelPos = get(get(axesHandle, <span class="string">'xlabel'</span>), <span class="string">'position'</span>);
3240             yLabelPos = get(get(axesHandle, <span class="string">'ylabel'</span>), <span class="string">'position'</span>);
3241             <span class="keyword">if</span> logScale
3242                 set(get(axesHandle, <span class="string">'xlabel'</span>), <span class="string">'position'</span>, [xLabelPos(1) 4.75e-5 xLabelPos(3)])
3243                 set(get(axesHandle, <span class="string">'ylabel'</span>), <span class="string">'position'</span>, [4.1e-5 yLabelPos(2:3)])
3244             <span class="keyword">else</span>
3245                 set(get(axesHandle, <span class="string">'xlabel'</span>), <span class="string">'position'</span>, [xLabelPos(1) -0.0032 xLabelPos(3)])
3246                 set(get(axesHandle, <span class="string">'ylabel'</span>), <span class="string">'position'</span>, [-0.0047 yLabelPos(2:3)])
3247             <span class="keyword">end</span>
3248             
3249             <span class="keyword">if</span> logScale
3250                 text(3e-3, 3e-4, <span class="string">'mRNA'</span>, <span class="string">'Color'</span>, [0 1 1], <span class="string">'Parent'</span>, axesHandle, <span class="string">'FontSize'</span>, 7)
3251                 text(7.8e-2, 2.5e-2, <span class="string">'rRNA'</span>, <span class="string">'Color'</span>, [0 0 1], <span class="string">'Parent'</span>, axesHandle, <span class="string">'FontSize'</span>, 7, <span class="string">'HorizontalAlign'</span>, <span class="string">'right'</span>)
3252                 text(1.35e-2, 0.77e-2, <span class="string">'sRNA'</span>, <span class="string">'Color'</span>, [1 0 0], <span class="string">'Parent'</span>, axesHandle, <span class="string">'FontSize'</span>, 7)
3253                 text(1.7e-2, 6e-2, <span class="string">'tRNA'</span>, <span class="string">'Color'</span>, [0 1 0], <span class="string">'Parent'</span>, axesHandle, <span class="string">'FontSize'</span>, 7, <span class="string">'HorizontalAlign'</span>, <span class="string">'right'</span>)
3254             <span class="keyword">else</span>
3255                 text(0.012, 0.003, <span class="string">'mRNA'</span>, <span class="string">'Color'</span>, [0 1 1], <span class="string">'Parent'</span>, axesHandle, <span class="string">'FontSize'</span>, 7)
3256                 text(7e-2, 3.1e-2, <span class="string">'rRNA'</span>, <span class="string">'Color'</span>, [0 0 1], <span class="string">'Parent'</span>, axesHandle, <span class="string">'FontSize'</span>, 7, <span class="string">'HorizontalAlign'</span>, <span class="string">'right'</span>)
3257                 text(1.35e-2, 0.77e-2, <span class="string">'sRNA'</span>, <span class="string">'Color'</span>, [1 0 0], <span class="string">'Parent'</span>, axesHandle, <span class="string">'FontSize'</span>, 7)
3258                 text(4.5e-2, 2.7e-2, <span class="string">'tRNA'</span>, <span class="string">'Color'</span>, [0 1 0], <span class="string">'Parent'</span>, axesHandle, <span class="string">'FontSize'</span>, 7, <span class="string">'HorizontalAlign'</span>, <span class="string">'right'</span>)
3259             <span class="keyword">end</span>
3260         <span class="keyword">end</span>
3261         
3262         <a name="_sub22" href="#_subfunctions" class="code">function geneExpressionRatios(exptGeneExpr, modelGeneExpr, axesHandle)</a>
3263             import edu.stanford.covert.cell.sim.util.CachedSimulationObjectUtil;
3264             import edu.stanford.covert.cell.sim.util.PlotUtil;
3265             
3266             <span class="comment">%% get constants</span>
3267             sim = CachedSimulationObjectUtil.load();
3268             g = sim.gene;
3269             
3270             x1 = modelGeneExpr;
3271             x2 = exptGeneExpr;
3272             y1 = modelGeneExpr(g.mRNAIndexs);
3273             y2 = exptGeneExpr(g.mRNAIndexs);
3274             z1 = modelGeneExpr(setdiff(1:<span class="keyword">end</span>, g.mRNAIndexs));
3275             z2 = exptGeneExpr(setdiff(1:<span class="keyword">end</span>, g.mRNAIndexs));
3276             
3277             corr(x1(~isnan(x1) &amp; ~isnan(x2)), x2(~isnan(x1) &amp; ~isnan(x2)))
3278             corr(y1(~isnan(y1) &amp; ~isnan(y2)), y2(~isnan(y1) &amp; ~isnan(y2)))
3279             corr(z1(~isnan(z1) &amp; ~isnan(z2)), z2(~isnan(z1) &amp; ~isnan(z2)))
3280             
3281             <span class="comment">%% plot</span>
3282             ratio = modelGeneExpr ./ exptGeneExpr;
3283             
3284             spcg = 50;
3285             xPos = zeros(size(ratio));
3286             xPos(g.mRNAIndexs) = 1:numel(g.mRNAIndexs);
3287             xPos(g.rRNAIndexs) = (1:numel(g.rRNAIndexs)) + spcg + numel(g.mRNAIndexs);
3288             xPos(g.sRNAIndexs) = (1:numel(g.sRNAIndexs)) + 2*spcg + numel(g.mRNAIndexs) + numel(g.rRNAIndexs);
3289             xPos(g.tRNAIndexs) = (1:numel(g.tRNAIndexs)) + 3*spcg + numel(g.mRNAIndexs) + numel(g.rRNAIndexs) + numel(g.sRNAIndexs);
3290             
3291             hold(axesHandle, <span class="string">'on'</span>);
3292             plot(axesHandle, xPos(g.mRNAIndexs), ratio(g.mRNAIndexs), <span class="string">'r.'</span>);
3293             plot(axesHandle, xPos(g.rRNAIndexs), ratio(g.rRNAIndexs), <span class="string">'g.'</span>);
3294             plot(axesHandle, xPos(g.sRNAIndexs), ratio(g.sRNAIndexs), <span class="string">'b.'</span>);
3295             plot(axesHandle, xPos(g.tRNAIndexs), ratio(g.tRNAIndexs), <span class="string">'c.'</span>);
3296             
3297             xlim(axesHandle, [-spcg numel(ratio)+4*spcg]);
3298             ylim(axesHandle, [1/2*1e-1 2*1e1]);
3299             ylabel(axesHandle, <span class="string">'Model / Expt Gene Expression'</span>);
3300             xlabel(axesHandle, <span class="string">'Gene'</span>);
3301             set(axesHandle, <span class="string">'XTick'</span>, [], <span class="string">'YScale'</span>, <span class="string">'log'</span>, <span class="string">'TickDir'</span>, <span class="string">'out'</span>);
3302             box(axesHandle, <span class="string">'off'</span>);
3303             
3304             lineY = 0.88 * max(ylim(axesHandle));
3305             textY = 0.99 * max(ylim(axesHandle));
3306             
3307             line(xPos(g.mRNAIndexs([1 end])) + [-10; 10], lineY * [1 1], <span class="string">'Parent'</span>, axesHandle, <span class="string">'Color'</span>, <span class="string">'r'</span>)
3308             line(xPos(g.rRNAIndexs([1 end])) + [-10; 10], lineY * [1 1], <span class="string">'Parent'</span>, axesHandle, <span class="string">'Color'</span>, <span class="string">'g'</span>)
3309             line(xPos(g.sRNAIndexs([1 end])) + [-10; 10], lineY * [1 1], <span class="string">'Parent'</span>, axesHandle, <span class="string">'Color'</span>, <span class="string">'b'</span>)
3310             line(xPos(g.tRNAIndexs([1 end])) + [-10; 10], lineY * [1 1], <span class="string">'Parent'</span>, axesHandle, <span class="string">'Color'</span>, <span class="string">'c'</span>)
3311             text(mean(xPos(g.mRNAIndexs([1 end]))), textY, <span class="string">'mRNA'</span>, <span class="string">'HorizontalAlign'</span>, <span class="string">'center'</span>, <span class="string">'Color'</span>, <span class="string">'r'</span>);
3312             text(mean(xPos(g.rRNAIndexs([1 end]))), textY, <span class="string">'rRNA'</span>, <span class="string">'HorizontalAlign'</span>, <span class="string">'center'</span>, <span class="string">'Color'</span>, <span class="string">'g'</span>);
3313             text(mean(xPos(g.sRNAIndexs([1 end]))), textY, <span class="string">'sRNA'</span>, <span class="string">'HorizontalAlign'</span>, <span class="string">'center'</span>, <span class="string">'Color'</span>, <span class="string">'b'</span>);
3314             text(mean(xPos(g.tRNAIndexs([1 end]))), textY, <span class="string">'tRNA'</span>, <span class="string">'HorizontalAlign'</span>, <span class="string">'center'</span>, <span class="string">'Color'</span>, <span class="string">'c'</span>);
3315         <span class="keyword">end</span>
3316         
3317         <a name="_sub23" href="#_subfunctions" class="code">function circosPlot(simBatchDir, selectedSim, outDirectory, figHandle, position)</a>
3318             import edu.stanford.covert.cell.sim.util.SimulationEnsemble;
3319             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
3320             import edu.stanford.covert.cell.sim.util.PlotUtil;
3321             import edu.stanford.covert.cell.sim.analysis.ChromosomePositionHistogram;
3322             import edu.stanford.covert.cell.sim.analysis.ChromosomeSpaceTimePlot;
3323             import edu.stanford.covert.cell.sim.analysis.Figures34;
3324             
3325             <span class="keyword">if</span> ~exist(<span class="string">'simBatchDir'</span>, <span class="string">'var'</span>) || isempty(simBatchDir)
3326                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
3327             <span class="keyword">end</span>
3328             <span class="keyword">if</span> ~exist(<span class="string">'selectedSim'</span>, <span class="string">'var'</span>) || isempty(selectedSim)
3329                 selectedSim = 1;
3330             <span class="keyword">end</span>
3331             
3332             [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep num2str(selectedSim)]);
3333             ch = sim.state(<span class="string">'Chromosome'</span>);
3334             pc = sim.state(<span class="string">'ProteinComplex'</span>);
3335             
3336             <span class="comment">%% get data</span>
3337             <span class="keyword">if</span> exist([outDirectory <span class="string">'circosPlot.mat'</span>], <span class="string">'file'</span>)
3338                 load([outDirectory <span class="string">'circosPlot.mat'</span>]);
3339             <span class="keyword">else</span>
3340                 stateNames = {
3341                     <span class="string">'Time'</span>       <span class="string">'values'</span>                   <span class="string">':'</span> <span class="string">':'</span>
3342                     <span class="string">'Chromosome'</span> <span class="string">'complexBoundSites'</span>        <span class="string">':'</span> <span class="string">':'</span>
3343                     };
3344                 states = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, <span class="string">'extract'</span>, selectedSim);
3345                 
3346                 chromProteins = {<span class="keyword">...</span>
3347                     <span class="comment">%whole cell ids                strands  number mers</span>
3348                     {<span class="string">'MG_094_HEXAMER'</span>},            1:4      1              <span class="comment">%DNA Pol</span>
3349                     {<span class="string">'MG_469_1MER_ADP'</span> <span class="keyword">...</span><span class="comment">                                 %DnaA</span>
3350                     <span class="string">'MG_469_1MER_ATP'</span> <span class="keyword">...</span>
3351                     <span class="string">'MG_469_2MER_ATP'</span> <span class="keyword">...</span>
3352                     <span class="string">'MG_469_3MER_ATP'</span> <span class="keyword">...</span>
3353                     <span class="string">'MG_469_4MER_ATP'</span> <span class="keyword">...</span>
3354                     <span class="string">'MG_469_5MER_ATP'</span> <span class="keyword">...</span>
3355                     <span class="string">'MG_469_6MER_ATP'</span> <span class="keyword">...</span>
3356                     <span class="string">'MG_469_7MER_ATP'</span>},           1:2      [1 1:7]
3357                     {<span class="string">'RNA_POLYMERASE'</span>                                      <span class="comment">%RNA Pol</span>
3358                     <span class="string">'RNA_POLYMERASE_HOLOENZYME'</span>}, 1:2      [1 1]
3359                     };
3360                 
3361                 densityBEMatrix = zeros(1000, size(chromProteins, 1));
3362                 <span class="keyword">for</span> i = 1:size(chromProteins, 1)
3363                     <span class="keyword">for</span> j = 1:numel(chromProteins{i, 1})
3364                         protIdx = pc.getIndexs(chromProteins{i, 1}{j});
3365                         
3366                         <span class="comment">%calculate frequency binding each position</span>
3367                         tmp = full(sum(sum(states.Chromosome.complexBoundSites(:, chromProteins{i, 2}, :) == protIdx, 3), 2));
3368                         
3369                         <span class="comment">%circularly convolve with footprint</span>
3370                         tmp = cconv(tmp, chromProteins{i, 3}(j) * ones(ch.complexDNAFootprints(protIdx), 1), ch.sequenceLen);
3371                         
3372                         <span class="comment">%bin to calculate density</span>
3373                         densityBEMatrix(:, i) = <span class="keyword">...</span>
3374                             + densityBEMatrix(:, i) <span class="keyword">...</span>
3375                             + sum(reshape([tmp; tmp(1:1000 - mod(ch.sequenceLen, 1000), :)], [], 1000), 1)';
3376                     <span class="keyword">end</span>
3377                 <span class="keyword">end</span>
3378                 densityBEMatrix = densityBEMatrix / ceil(ch.sequenceLen/1000) / numel(states.Time.values); <span class="comment">%proteins/(nt h)</span>
3379                 
3380                 nBins = ch.sequenceLen - 1;
3381                 densityBEMatrixOS = zeros(nBins, size(chromProteins, 1));
3382                 <span class="keyword">for</span> i = 1:size(chromProteins, 1)
3383                     <span class="keyword">for</span> j = 1:numel(chromProteins{i, 1})
3384                         protIdx = pc.getIndexs(chromProteins{i, 1}{j});
3385                         
3386                         <span class="comment">%calculate frequency binding each position</span>
3387                         tmp = full(sum(sum(states.Chromosome.complexBoundSites(:, chromProteins{i, 2}, :) == protIdx, 3), 2));
3388                         
3389                         <span class="comment">%circularly convolve with footprint</span>
3390                         tmp = cconv(tmp, chromProteins{i, 3}(j) * ones(ch.complexDNAFootprints(protIdx), 1), ch.sequenceLen);
3391                         
3392                         <span class="comment">%bin to calculate density</span>
3393                         densityBEMatrixOS(:, i) = <span class="keyword">...</span>
3394                             + densityBEMatrixOS(:, i) <span class="keyword">...</span>
3395                             + sum(reshape([tmp; tmp(1:nBins - mod(ch.sequenceLen, nBins), :)], [], nBins), 1)';
3396                     <span class="keyword">end</span>
3397                 <span class="keyword">end</span>
3398                 
3399                 densityBEMatrixOS = densityBEMatrixOS / ceil(ch.sequenceLen/nBins) / numel(states.Time.values); <span class="comment">%proteins/(nt h)</span>
3400                 
3401                 
3402                 <span class="comment">%save</span>
3403                 save([outDirectory <span class="string">'circosPlot.mat'</span>], <span class="string">'densityBEMatrix'</span>, <span class="string">'densityBEMatrixOS'</span>);
3404             <span class="keyword">end</span>
3405             
3406             totalDensityMatrix = Figures34.calculateDNABoundProteinDensity(outDirectory, simBatchDir, selectedSim); <span class="comment">%1/(nt s)</span>
3407             totalDensityMatrix = sum(reshape([totalDensityMatrix; totalDensityMatrix(1:1000 - mod(numel(totalDensityMatrix), 1000), :)], [], 1000), 1)' <span class="keyword">...</span>
3408                 / ceil(ch.sequenceLen/1000); <span class="comment">%1/(nt s)</span>
3409             
3410             densityBEMatrix = [densityBEMatrix totalDensityMatrix];
3411             <span class="comment">%             densityBEMatrix = [densityBEMatrix [densityBEMatrixOS(579491:579560, 2); zeros(860, 1); densityBEMatrixOS(579421:579490, 2)]];</span>
3412             <span class="comment">%             clear('densityBEMatrixOS');</span>
3413             
3414             
3415             <span class="comment">%% plot</span>
3416             <span class="keyword">if</span> nargin &lt; 5
3417                 [~, figHandle] = edu.stanford.covert.cell.sim.util.PlotUtil.newAxesHandle();
3418                 clf(figHandle);
3419                 position = [0.1 0.1 0.8 0.8];
3420             <span class="keyword">end</span>
3421             x = position(1);
3422             y = position(2);
3423             w = position(3);
3424             h = position(4);
3425             
3426             figPos = get(figHandle, <span class="string">'papersize'</span>);
3427             figW = figPos(1);
3428             figH = figPos(2);
3429             
3430             y2 = y+(h-0.75*w*figW/figH)/2 + 0.81*w*figW/figH/2 - 0.5*h;
3431             
3432             colorbarHandles = [
3433                 subplot(<span class="string">'position'</span>, [x+0.727*w  y2            0.03*w  0.2*h], <span class="string">'parent'</span>, figHandle)
3434                 subplot(<span class="string">'position'</span>, [x+0.727*w  y2+0.8*1/3*h  0.03*w  0.2*h], <span class="string">'parent'</span>, figHandle)
3435                 subplot(<span class="string">'position'</span>, [x+0.727*w  y2+0.8*2/3*h  0.03*w  0.2*h], <span class="string">'parent'</span>, figHandle)
3436                 subplot(<span class="string">'position'</span>, [x+0.727*w  y2+0.8*h      0.03*w  0.2*h], <span class="string">'parent'</span>, figHandle)
3437                 ];
3438             axesHandle = subplot(<span class="string">'position'</span>, [x  y+(h-0.77*w*figW/figH)/2  0.70*w  0.81*w*figW/figH], <span class="string">'parent'</span>, figHandle);
3439             
3440             cla(axesHandle);
3441             hold(axesHandle, <span class="string">'on'</span>);
3442             set(axesHandle, <span class="string">'visible'</span>, <span class="string">'off'</span>);
3443             
3444             cmap = cell(4, 1);
3445             cmapLims = zeros(4, 2);
3446             cmapTicks = cell(4, 1);
3447             cmapFuncs = {
3448                 [0 1 0]
3449                 [1 0 0]
3450                 [0 0 1]
3451                 [0 1 1]
3452                 };
3453             totcmap = zeros(0, 3);
3454             
3455             nTheta = size(densityBEMatrix, 1);
3456             <span class="keyword">for</span> i = 1:size(densityBEMatrix, 2)
3457                 x = zeros(10, size(densityBEMatrix, 1));
3458                 y = zeros(10, size(densityBEMatrix, 1));
3459                 c = zeros(size(densityBEMatrix, 1), 1);
3460                 
3461                 data = densityBEMatrix(:, i);
3462                 
3463                 cmapLims(i, :) = [
3464                     max(-6, floor(log10(min(data(data &gt; 0)))))
3465                     ceil(log10(max(data)))
3466                     ];
3467                 
3468                 <span class="keyword">if</span> i == 2
3469                     cmapLims(i, :) = [-5 0];
3470                 <span class="keyword">end</span>
3471                 
3472                 nBins = 2 + 10 * diff(cmapLims(i, :));
3473                 cmap{i} = ones(nBins, 3) - repmat([1 1 1] - cmapFuncs{i}, nBins, 1) .* repmat(linspace(0, 1, nBins)', 1, 3);
3474                 cmapTicks{i} = [0 logspace(cmapLims(i, 1), cmapLims(i, 2), 1 + 10 * diff(cmapLims(i, :)))];
3475                 
3476                 <span class="keyword">for</span> j = 1:nTheta
3477                     x(:, j) = [
3478                         (i+1)*cos(linspace(2*pi/nTheta*j, 2*pi/nTheta*(j+1), 5)+pi/2) <span class="keyword">...</span>
3479                         (i+2)*cos(linspace(2*pi/nTheta*(j+1), 2*pi/nTheta*j, 5)+pi/2) <span class="keyword">...</span>
3480                         ];
3481                     y(:, j) = [
3482                         (i+1)*sin(linspace(2*pi/nTheta*j, 2*pi/nTheta*(j+1), 5)+pi/2) <span class="keyword">...</span>
3483                         (i+2)*sin(linspace(2*pi/nTheta*(j+1), 2*pi/nTheta*j, 5)+pi/2) <span class="keyword">...</span>
3484                         ];
3485                     c(j) = find(cmapTicks{i} &gt;= data(j), 1, <span class="string">'first'</span>);
3486                 <span class="keyword">end</span>
3487                 
3488                 c = c + size(totcmap, 1);
3489                 totcmap = [totcmap; cmap{i}]; <span class="comment">%#ok&lt;AGROW&gt;</span>
3490                 
3491                 ptch = patch(x, y, ones(1, nTheta), <span class="string">'EdgeColor'</span>, <span class="string">'none'</span>, <span class="string">'Parent'</span>, axesHandle);
3492                 set(ptch, <span class="string">'FaceColor'</span>, <span class="string">'flat'</span>, <span class="string">'FaceVertexCData'</span>, c, <span class="string">'CDataMapping'</span>, <span class="string">'direct'</span>, <span class="string">'FaceAlpha'</span>, 1, <span class="string">'EdgeAlpha'</span>, 1);
3493             <span class="keyword">end</span>
3494             colormap(axesHandle, totcmap);
3495             
3496             xlims = [-7.8 6.2];
3497             xlim(axesHandle, xlims);
3498             ylim(axesHandle, range(xlims)*0.81/0.7/2 * [-1 1]);
3499             
3500             <span class="comment">%oriC label</span>
3501             line([0 0], [2 6.2], <span class="string">'Color'</span>, <span class="string">'k'</span>, <span class="string">'Parent'</span>, axesHandle);
3502             text(0, 6.18, <span class="string">'oriC'</span>, <span class="keyword">...</span>
3503                 <span class="string">'Parent'</span>, axesHandle, <span class="keyword">...</span>
3504                 <span class="string">'VerticalAlign'</span>, <span class="string">'bottom'</span>, <span class="keyword">...</span>
3505                 <span class="string">'HorizontalAlign'</span>, <span class="string">'center'</span>, <span class="keyword">...</span>
3506                 <span class="string">'FontSize'</span>, 5);
3507             
3508             <span class="comment">% rRNA label</span>
3509             
3510             <span class="comment">% &quot;Box&quot;</span>
3511             s1 = 0.85;
3512             s2 = 1 + (1 - s1);
3513             theta1 = (170007)/580076 * (2*pi);
3514             theta2 = (170007+4786)/580076 * (2*pi);
3515             nDivs = 500;
3516             x = zeros(nDivs, 1);
3517             y = zeros(nDivs, 1);
3518             z = ones(nDivs, 1);
3519             <span class="keyword">for</span> i = 1:nDivs
3520                 x(:, i) =  [
3521                     6.25 * cos(linspace(s1 * theta1, s2 * theta2, nDivs/2) + pi/2)<span class="keyword">...</span>
3522                     7.65 * cos(linspace(s2 * theta2, s1 * theta1, nDivs/2) + pi/2)<span class="keyword">...</span>
3523                     ];
3524                 y(:, i) =  [
3525                     6.25 * sin(linspace(s1 * theta1, s2 * theta2, nDivs/2) + pi/2)<span class="keyword">...</span>
3526                     7.65 * sin(linspace(s2 * theta2, s1 * theta1, nDivs/2) + pi/2)<span class="keyword">...</span>
3527                     ];
3528             <span class="keyword">end</span>
3529             p = patch(x, y, z, <span class="string">'EdgeColor'</span>, <span class="string">'k'</span>, <span class="string">'Parent'</span>, axesHandle);
3530             set(p, <span class="string">'FaceColor'</span>, <span class="string">'flat'</span>, <span class="string">'CDataMapping'</span>, <span class="string">'direct'</span>, <span class="string">'FaceAlpha'</span>, 1);
3531             
3532             line([(4.9 * (cos(theta1 + pi/2))) (6.25 * (cos(s1 * theta1 + pi/2)))], <span class="keyword">...</span>
3533                 [(4.9 * (sin(theta1 + pi/2))) (6.25 * (sin(s1 * theta1 + pi/2)))], <span class="keyword">...</span>
3534                 <span class="string">'Color'</span>, <span class="string">'k'</span>, <span class="string">'Parent'</span>, axesHandle);
3535             line([(4.9 * (cos(theta2 + pi/2))) (6.25 * (cos(s2 * theta2 + pi/2)))], <span class="keyword">...</span>
3536                 [(4.9 * (sin(theta2 + pi/2))) (6.25 * (sin(s2 * theta2 + pi/2)))], <span class="keyword">...</span>
3537                 <span class="string">'Color'</span>, <span class="string">'k'</span>, <span class="string">'Parent'</span>, axesHandle);
3538             
3539             theta = mean([theta1 theta2]);
3540             text((6.3 * (cos(theta + pi/2))), (6.25 * (sin(theta + pi/2))), <span class="keyword">...</span>
3541                 {<span class="string">'rRNA'</span> <span class="string">'5S, 16S, 23S'</span>}, <span class="keyword">...</span>
3542                 <span class="string">'Parent'</span>, axesHandle, <span class="keyword">...</span>
3543                 <span class="string">'VerticalAlign'</span>, <span class="string">'bottom'</span>, <span class="keyword">...</span>
3544                 <span class="string">'HorizontalAlign'</span>, <span class="string">'center'</span>, <span class="keyword">...</span>
3545                 <span class="string">'FontSize'</span>, 5, <span class="keyword">...</span>
3546                 <span class="string">'Rotation'</span>, theta * 180/pi);
3547             
3548             <span class="comment">% DnaA label</span>
3549             
3550             <span class="comment">% &quot;Box&quot;</span>
3551             s1 = 0.97;
3552             s2 = 1 + (1 - s1);
3553             theta1 = 997/1000 * (2*pi);
3554             theta2 = 997/1000 * (2*pi);
3555             nDivs = 500;
3556             off = pi / 10;
3557             x = zeros(nDivs, 1);
3558             y = zeros(nDivs, 1);
3559             z = ones(nDivs, 1);
3560             <span class="keyword">for</span> i = 1:nDivs
3561                 x(:, i) =  [
3562                     6.25 * cos(linspace(s1 * theta1, s2 * theta2, nDivs/2) + pi/2 - off)<span class="keyword">...</span>
3563                     7.65 * cos(linspace(s2 * theta2, s1 * theta1, nDivs/2) + pi/2 - off)<span class="keyword">...</span>
3564                     ];
3565                 y(:, i) =  [
3566                     6.25 * sin(linspace(s1 * theta1, s2 * theta2, nDivs/2) + pi/2 - off)<span class="keyword">...</span>
3567                     7.65 * sin(linspace(s2 * theta2, s1 * theta1, nDivs/2) + pi/2 - off)<span class="keyword">...</span>
3568                     ];
3569             <span class="keyword">end</span>
3570             p = patch(x, y, z, <span class="string">'EdgeColor'</span>, <span class="string">'k'</span>, <span class="string">'Parent'</span>, axesHandle);
3571             set(p, <span class="string">'FaceColor'</span>, <span class="string">'flat'</span>, <span class="string">'CDataMapping'</span>, <span class="string">'direct'</span>, <span class="string">'FaceAlpha'</span>, 1);
3572             
3573             line([(3.7 * (cos(theta1 + pi/2))) (6.25 * (cos(s1 * theta1 + pi/2 - off)))], <span class="keyword">...</span>
3574                 [(3.7 * (sin(theta1 + pi/2))) (6.25 * (sin(s1 * theta1 + pi/2 - off)))], <span class="keyword">...</span>
3575                 <span class="string">'Color'</span>, <span class="string">'k'</span>, <span class="string">'Parent'</span>, axesHandle);
3576             line([(3.7 * (cos(theta2 + pi/2))) (6.25 * (cos(s2 * theta2 + pi/2 - off)))], <span class="keyword">...</span>
3577                 [(3.7 * (sin(theta2 + pi/2))) (6.25 * (sin(s2 * theta2 + pi/2 - off)))], <span class="keyword">...</span>
3578                 <span class="string">'Color'</span>, <span class="string">'k'</span>, <span class="string">'Parent'</span>, axesHandle);
3579             
3580             data = densityBEMatrixOS(579434:579550, 2);
3581             nTheta = numel(data) + 8;
3582             x = zeros(10, numel(data));
3583             y = zeros(10, numel(data));
3584             c = zeros(numel(data), 1);
3585             
3586             arcLen = s2 * theta2 - s1 * theta1;
3587             off = off + arcLen / 2;
3588             <span class="keyword">for</span> j = 1:numel(data)
3589                 x(:, j) = [
3590                     (6.25)*cos(linspace(arcLen/nTheta*j, arcLen/nTheta*(j+1), 5)+pi/2-off) <span class="keyword">...</span>
3591                     (7.65)*cos(linspace(arcLen/nTheta*(j+1), arcLen/nTheta*j, 5)+pi/2-off) <span class="keyword">...</span>
3592                     ];
3593                 y(:, j) = [
3594                     (6.25)*sin(linspace(arcLen/nTheta*j, arcLen/nTheta*(j+1), 5)+pi/2-off) <span class="keyword">...</span>
3595                     (7.65)*sin(linspace(arcLen/nTheta*(j+1), arcLen/nTheta*j, 5)+pi/2-off) <span class="keyword">...</span>
3596                     ];
3597                 c(j) = find(cmapTicks{2} &gt;= data(j), 1, <span class="string">'first'</span>);
3598             <span class="keyword">end</span>
3599             
3600             c = c + size(cmap{1},1);
3601             
3602             ptch = patch(x, y, ones(1, numel(data)), <span class="string">'EdgeColor'</span>, <span class="string">'none'</span>, <span class="string">'Parent'</span>, axesHandle);
3603             set(ptch, <span class="string">'FaceColor'</span>, <span class="string">'flat'</span>, <span class="string">'FaceVertexCData'</span>, c, <span class="string">'CDataMapping'</span>, <span class="string">'direct'</span>, <span class="string">'FaceAlpha'</span>, 1, <span class="string">'EdgeAlpha'</span>, 1);
3604             
3605             <span class="comment">%             text(5.4, 6.6, ...</span>
3606             <span class="comment">%                 {'DnaA complex'}, ...</span>
3607             <span class="comment">%                 'Parent', axesHandle, ...</span>
3608             <span class="comment">%                 'VerticalAlign', 'bottom', ...</span>
3609             <span class="comment">%                 'HorizontalAlign', 'center', ...</span>
3610             <span class="comment">%                 'FontSize', 5, ...</span>
3611             <span class="comment">%                 'Rotation', 0);</span>
3612             
3613             theta = mean([(s1 * theta1) (s2 * theta2)]) - off + arcLen/2;
3614             text(2.9, 7.80, <span class="keyword">...</span>
3615                 {<span class="string">'DnaA complex'</span>}, <span class="keyword">...</span>
3616                 <span class="string">'Parent'</span>, axesHandle, <span class="keyword">...</span>
3617                 <span class="string">'VerticalAlign'</span>, <span class="string">'bottom'</span>, <span class="keyword">...</span>
3618                 <span class="string">'HorizontalAlign'</span>, <span class="string">'center'</span>, <span class="keyword">...</span>
3619                 <span class="string">'FontSize'</span>, 5, <span class="keyword">...</span>
3620                 <span class="string">'Rotation'</span>, theta * 180/pi + 2);
3621             
3622             text(1.34, 7.42, <span class="string">'*'</span>, <span class="string">'Parent'</span>, axesHandle, <span class="string">'VerticalAlign'</span>, <span class="string">'bottom'</span>, <span class="keyword">...</span>
3623                 <span class="string">'HorizontalAlign'</span>, <span class="string">'center'</span>, <span class="string">'FontSize'</span>, 5, <span class="string">'Rotation'</span>, 0);
3624             text(2.30, 7.19, <span class="string">'*'</span>, <span class="string">'Parent'</span>, axesHandle, <span class="string">'VerticalAlign'</span>, <span class="string">'bottom'</span>, <span class="keyword">...</span>
3625                 <span class="string">'HorizontalAlign'</span>, <span class="string">'center'</span>, <span class="string">'FontSize'</span>, 5, <span class="string">'Rotation'</span>, 0);
3626             text(2.64, 7.07, <span class="string">'*'</span>, <span class="string">'Parent'</span>, axesHandle, <span class="string">'VerticalAlign'</span>, <span class="string">'bottom'</span>, <span class="keyword">...</span>
3627                 <span class="string">'HorizontalAlign'</span>, <span class="string">'center'</span>, <span class="string">'FontSize'</span>, 5, <span class="string">'Rotation'</span>, 0);
3628             text(3.75, 6.58, <span class="string">'*'</span>, <span class="string">'Parent'</span>, axesHandle, <span class="string">'VerticalAlign'</span>, <span class="string">'bottom'</span>, <span class="keyword">...</span>
3629                 <span class="string">'HorizontalAlign'</span>, <span class="string">'center'</span>, <span class="string">'FontSize'</span>, 5, <span class="string">'Rotation'</span>, 0);
3630             
3631             <span class="comment">%color bars</span>
3632             ylabels = {
3633                 {<span class="string">'DNA pol'</span> <span class="string">'prob. bound'</span>}
3634                 {<span class="string">'DnaA'</span> <span class="string">'prob. bound'</span>}
3635                 {<span class="string">'RNA pol'</span> <span class="string">'prob. bound'</span>}
3636                 {<span class="string">'All proteins'</span> <span class="string">'prob. bound'</span>}
3637                 };
3638             <span class="keyword">for</span> i = 1:numel(colorbarHandles)
3639                 image(permute(cmap{i}, [1 3 2]), <span class="string">'Parent'</span>, colorbarHandles(i));
3640                 
3641                 set(colorbarHandles(i), <span class="keyword">...</span>
3642                     <span class="string">'TickDir'</span>, <span class="string">'out'</span>, <span class="keyword">...</span>
3643                     <span class="string">'box'</span>, <span class="string">'off'</span>, <span class="keyword">...</span>
3644                     <span class="string">'FontSize'</span>, 5, <span class="keyword">...</span>
3645                     <span class="string">'YAxisLocation'</span>, <span class="string">'right'</span>, <span class="keyword">...</span>
3646                     <span class="string">'XTick'</span>, [], <span class="keyword">...</span>
3647                     <span class="string">'YTick'</span>, 2:10:size(cmap{i}, 1), <span class="keyword">...</span>
3648                     <span class="string">'XLim'</span>, [0.5 1.5], <span class="keyword">...</span>
3649                     <span class="string">'visible'</span>, <span class="string">'on'</span>, <span class="keyword">...</span>
3650                     <span class="string">'YDir'</span>, <span class="string">'normal'</span>);
3651                 ylim(colorbarHandles(i), [0 size(cmap{i}, 1)]+0.5);
3652                 
3653                 line([min(xlim(colorbarHandles(i))) * [1 1]   max(xlim(colorbarHandles(i)))], <span class="keyword">...</span>
3654                     [ylim(colorbarHandles(i)) max(ylim(colorbarHandles(i)))], <span class="keyword">...</span>
3655                     <span class="string">'parent'</span>, colorbarHandles(i), <span class="string">'color'</span>, <span class="string">'k'</span>);
3656                 
3657                 tick2text(colorbarHandles(i), <span class="string">'axis'</span>, <span class="string">'y'</span>, <span class="string">'ytickoffset'</span>, -1.25);
3658                 yTicks = getappdata(colorbarHandles(i), <span class="string">'YTickText'</span>);
3659                 set(yTicks, <span class="string">'FontSize'</span>, 5, <span class="string">'HorizontalAlign'</span>, <span class="string">'left'</span>, <span class="string">'Interpreter'</span>, <span class="string">'tex'</span>);
3660                 
3661                 <span class="keyword">for</span> j = 1:numel(yTicks)
3662                     set(yTicks(j), <span class="string">'String'</span>, sprintf(<span class="string">'10^{%d}'</span>, cmapLims(i, 1)+j-1));
3663                 <span class="keyword">end</span>
3664                 <span class="keyword">if</span> numel(yTicks) == 6
3665                     delete(yTicks(1:2:5));
3666                 <span class="keyword">end</span>
3667                 
3668                 ylabel(colorbarHandles(i), ylabels{i}, <span class="string">'FontSize'</span>, 5);
3669                 ylabelpos = get(get(colorbarHandles(i), <span class="string">'ylabel'</span>), <span class="string">'position'</span>);
3670                 set(get(colorbarHandles(i), <span class="string">'ylabel'</span>), <span class="string">'position'</span>, [3.75 ylabelpos(2:end)]);
3671             <span class="keyword">end</span>
3672         <span class="keyword">end</span>
3673         
3674         <a name="_sub24" href="#_subfunctions" class="code">function dnaBoundProteinDensityVsDisplacement(simBatchDir, selectedSim, outDirectory, figHandle, position)</a>
3675             import edu.stanford.covert.cell.sim.analysis.Figures34;
3676             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
3677             
3678             <span class="keyword">if</span> ~exist(<span class="string">'simBatchDir'</span>, <span class="string">'var'</span>) || isempty(simBatchDir)
3679                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
3680             <span class="keyword">end</span>
3681             <span class="keyword">if</span> ~exist(<span class="string">'selectedSim'</span>, <span class="string">'var'</span>) || isempty(selectedSim)
3682                 selectedSim = 1;
3683             <span class="keyword">end</span>
3684             
3685             [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep num2str(selectedSim)]);
3686             c = sim.state(<span class="string">'Chromosome'</span>);
3687             
3688             <span class="comment">%% get data</span>
3689             dnaBoundProteinDensity = Figures34.calculateDNABoundProteinDensity(outDirectory, simBatchDir, selectedSim);
3690             
3691             dnaBoundProteinDisplacements = histc(<span class="keyword">...</span>
3692                 Figures34.calculateDNABoundProteinDisplacements(outDirectory, simBatchDir, selectedSim), <span class="keyword">...</span>
3693                 1:c.sequenceLen);
3694             dnaBoundProteinDisplacements = conv(dnaBoundProteinDisplacements, ones(1000, 1), <span class="string">'same'</span>);
3695             
3696             <span class="comment">%% plot</span>
3697             <span class="keyword">if</span> nargin &gt;= 5
3698                 x = position(1);
3699                 y = position(2);
3700                 w = position(3);
3701                 h = position(4);
3702                 axesHandle = subplot(<span class="string">'position'</span>, [x y 0.82*w h], <span class="string">'parent'</span>, figHandle);
3703                 colorBarHandle = subplot(<span class="string">'position'</span>, [x+0.85*w y 0.05*w h], <span class="string">'parent'</span>, figHandle);
3704             <span class="keyword">else</span>
3705                 axesHandle = edu.stanford.covert.cell.sim.util.PlotUtil.newAxesHandle();
3706                 colorBarHandle = colorbar();
3707             <span class="keyword">end</span>
3708             
3709             <span class="comment">% data</span>
3710             [cnts, centers] = hist3([dnaBoundProteinDensity dnaBoundProteinDisplacements], [100 100]);
3711             cnts = cnts / c.sequenceLen;
3712             edges = logspace(-6, -1, 51);
3713             colormap = jet(numel(edges) + 1);
3714             colors = zeros(100, 100, 3);
3715             
3716             [i, j] = find(cnts &lt; edges(1));
3717             colors(sub2ind(size(colors), i, j, ones(size(i)))) = colormap(1, 1);
3718             colors(sub2ind(size(colors), i, j, 2*ones(size(i)))) = colormap(1, 2);
3719             colors(sub2ind(size(colors), i, j, 3*ones(size(i)))) = colormap(1, 3);
3720             
3721             [i, j] = find(cnts &gt; edges(end-1));
3722             colors(sub2ind(size(colors), i, j, ones(size(i)))) = colormap(<span class="keyword">end</span>, 1);
3723             colors(sub2ind(size(colors), i, j, 2*ones(size(i)))) = colormap(<span class="keyword">end</span>, 2);
3724             colors(sub2ind(size(colors), i, j, 3*ones(size(i)))) = colormap(<span class="keyword">end</span>, 3);
3725             
3726             <span class="keyword">for</span> idx = 1:numel(edges)-1
3727                 [i, j] = find(cnts &gt;= edges(idx) &amp; cnts &lt;= edges(idx+1));
3728                 colors(sub2ind(size(colors), i, j, ones(size(i)))) = colormap(idx+1, 1);
3729                 colors(sub2ind(size(colors), i, j, 2*ones(size(i)))) = colormap(idx+1, 2);
3730                 colors(sub2ind(size(colors), i, j, 3*ones(size(i)))) = colormap(idx+1, 3);
3731             <span class="keyword">end</span>
3732             
3733             image(permute(colors, [2 1 3]), <span class="string">'Parent'</span>, axesHandle);
3734             set(axesHandle, <span class="string">'XDir'</span>, <span class="string">'normal'</span>);
3735             set(axesHandle, <span class="string">'YDir'</span>, <span class="string">'normal'</span>);
3736             
3737             set(axesHandle, <span class="string">'Box'</span>, <span class="string">'on'</span>, <span class="string">'FontSize'</span>, 5, <span class="string">'TickDir'</span>, <span class="string">'out'</span>);
3738             xlim(axesHandle, [0.5 0.4/centers{1}(end)*100])
3739             ylim(axesHandle, [0.5 1500/centers{2}(end)*100])
3740             
3741             set(axesHandle, <span class="string">'XTick'</span>, 0.5 + (0:0.1:0.4)/centers{1}(end)*100);
3742             set(axesHandle, <span class="string">'YTick'</span>, 0.5 + (0:500:1500)/centers{2}(end)*100);
3743             tick2text(axesHandle, <span class="string">'axis'</span>, <span class="string">'xy'</span>, <span class="string">'xtickoffset'</span>, 0.06, <span class="string">'ytickoffset'</span>, 0.015);
3744             xTicks = getappdata(axesHandle, <span class="string">'XTickText'</span>);
3745             yTicks = getappdata(axesHandle, <span class="string">'YTickText'</span>);
3746             set(xTicks, <span class="string">'FontSize'</span>, 5);
3747             set(yTicks, <span class="string">'FontSize'</span>, 5, <span class="string">'HorizontalAlign'</span>, <span class="string">'right'</span>);
3748             set(xTicks(1), <span class="string">'String'</span>, <span class="string">'0'</span>)
3749             set(xTicks(end), <span class="string">'String'</span>, <span class="string">'0.4'</span>)
3750             set(yTicks(1), <span class="string">'String'</span>, <span class="string">'0'</span>)
3751             set(yTicks(end), <span class="string">'String'</span>, <span class="string">'1500'</span>)
3752             delete(yTicks(2:3));
3753             delete(xTicks(2:4));
3754             
3755             xlabel(axesHandle, <span class="string">'Density'</span>, <span class="string">'FontSize'</span>, 7);
3756             ylabel(axesHandle, <span class="string">'Displacement'</span>, <span class="string">'FontSize'</span>, 7);
3757             xlabelPos = get(get(axesHandle, <span class="string">'xlabel'</span>), <span class="string">'position'</span>);
3758             ylabelPos = get(get(axesHandle, <span class="string">'ylabel'</span>), <span class="string">'position'</span>);
3759             set(get(axesHandle, <span class="string">'xlabel'</span>), <span class="string">'position'</span>, [xlabelPos(1) -3 xlabelPos(3:end)]);
3760             set(get(axesHandle, <span class="string">'ylabel'</span>), <span class="string">'position'</span>, [-3 ylabelPos(2:end)]);
3761             
3762             <span class="comment">% color bar</span>
3763             image(permute(colormap, [1 3 2]), <span class="string">'Parent'</span>, colorBarHandle);
3764             
3765             box(colorBarHandle, <span class="string">'on'</span>);
3766             set(colorBarHandle, <span class="string">'XTick'</span>, []);
3767             set(colorBarHandle, <span class="string">'TickDir'</span>, <span class="string">'out'</span>);
3768             set(colorBarHandle, <span class="string">'YAxisLocation'</span>, <span class="string">'right'</span>);
3769             set(colorBarHandle, <span class="string">'YDir'</span>, <span class="string">'normal'</span>)
3770             set(colorBarHandle, <span class="string">'FontSize'</span>, 5);
3771             
3772             set(colorBarHandle, <span class="string">'YTick'</span>, 2:10:52);
3773             tick2text(colorBarHandle, <span class="string">'axis'</span>, <span class="string">'y'</span>, <span class="string">'ytickoffset'</span>, -1.15);
3774             yTicks = getappdata(colorBarHandle, <span class="string">'YTickText'</span>);
3775             set(yTicks, <span class="string">'FontSize'</span>, 5, <span class="string">'HorizontalAlign'</span>, <span class="string">'left'</span>, <span class="string">'Interpreter'</span>, <span class="string">'tex'</span>);
3776             <span class="keyword">for</span> i = 1:6
3777                 set(yTicks(i), <span class="string">'String'</span>, sprintf(<span class="string">'10^{%d}'</span>, -7+i));
3778             <span class="keyword">end</span>
3779             
3780             ylabel(colorBarHandle, <span class="string">'% Bases'</span>, <span class="string">'FontSize'</span>, 7);
3781             ylabelPos = get(get(colorBarHandle, <span class="string">'ylabel'</span>), <span class="string">'position'</span>);
3782             set(get(colorBarHandle, <span class="string">'ylabel'</span>), <span class="string">'position'</span>, [3.0 ylabelPos(2:end)]);
3783         <span class="keyword">end</span>
3784         
3785         <a name="_sub25" href="#_subfunctions" class="code">function [posFreq, protDensity] = analyzeDNABoundProteinDisplacements(simBatchDir, outDirectory, figHandle, position)</a>
3786             import edu.stanford.covert.cell.sim.analysis.Figures34;
3787             import edu.stanford.covert.cell.sim.util.CachedSimulationObjectUtil;
3788             import edu.stanford.covert.cell.sim.util.PlotUtil;
3789             import edu.stanford.covert.cell.sim.util.PrintUtil;
3790             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
3791             import edu.stanford.covert.cell.sim.util.SimulationEnsemble;
3792             import edu.stanford.covert.cell.sim.util.SummaryLogger;
3793             
3794             <span class="comment">%% constants</span>
3795             sim = CachedSimulationObjectUtil.load();
3796             
3797             g = sim.gene;
3798             c = sim.state(<span class="string">'Chromosome'</span>);
3799             transcript = sim.state(<span class="string">'Transcript'</span>);
3800             pm = sim.state(<span class="string">'ProteinMonomer'</span>);
3801             pc = sim.state(<span class="string">'ProteinComplex'</span>);
3802             rna = sim.state(<span class="string">'Rna'</span>);
3803             transcription = sim.process(<span class="string">'Transcription'</span>);
3804             
3805             <span class="comment">%% get data</span>
3806             nMon = numel(pm.matureIndexs);
3807             nCpx = numel(pc.matureIndexs);
3808             nProt = nMon + nCpx;            
3809             
3810             <span class="keyword">if</span> ~exist([outDirectory filesep <span class="string">'dnaBoundProteinDisplacement.mat'</span>], <span class="string">'file'</span>)
3811                 nSims = SimulationDiskUtil.getNumSimulations(simBatchDir);
3812                 
3813                 ensemble = SimulationEnsemble(simBatchDir, cell(0, 2), [], 1:nSims);
3814                 simEndTimes = ensemble.stateData.simulationEndTimes;
3815                 
3816                 nDisplacements = 0;
3817                 allDisplacements = zeros(0, 5);
3818                 protByProtFreq = zeros(nProt);
3819                 timeFreq = zeros(1000, 1);
3820                 posFreq = zeros(c.sequenceLen, 2);
3821                 <span class="keyword">for</span> i = 1:nSims
3822                     dnaBoundProteinDisplacements = Figures34.calculateDNABoundProteinDisplacements(outDirectory, simBatchDir, i);
3823                     
3824                     nDisplacements = nDisplacements + size(dnaBoundProteinDisplacements, 1);
3825                     allDisplacements = [allDisplacements; dnaBoundProteinDisplacements]; <span class="comment">%#ok&lt;AGROW&gt;</span>
3826                     
3827                     tmp = dnaBoundProteinDisplacements(:, 1:2);
3828                     tmp(tmp &lt; 0) = -tmp(tmp &lt; 0) + nMon;
3829                     tmpCnt = hist3(tmp, <span class="string">'Edges'</span>, {1:nProt 1:nProt});
3830                     assert(sum(tmpCnt(:)) == size(dnaBoundProteinDisplacements, 1));
3831                     protByProtFreq = protByProtFreq + tmpCnt;
3832                     
3833                     timeFreq = timeFreq + histc(dnaBoundProteinDisplacements(:, 5), linspace(1, simEndTimes(i), 1000));
3834                     posFreq(:, 1) = posFreq(:, 1) + histc(dnaBoundProteinDisplacements(:, 3), 1:c.sequenceLen);
3835                     posFreq(:, 2) = posFreq(:, 2) + histc(dnaBoundProteinDisplacements(:, 4), 1:c.sequenceLen);
3836                 <span class="keyword">end</span>
3837                 
3838                 simStats = SummaryLogger.getSimulationStatistics(simBatchDir);
3839                 repInitDuration = simStats(:, SummaryLogger.SIM_STATUS_INDEX_REP_INIT_TIME);
3840                 repDuration = diff(simStats(:, [SummaryLogger.SIM_STATUS_INDEX_REP_INIT_TIME SummaryLogger.SIM_STATUS_INDEX_REP_TIME]), 1, 2);
3841                 cytokinesisDuration = diff(simStats(:, [SummaryLogger.SIM_STATUS_INDEX_REP_TIME SummaryLogger.SIM_STATUS_INDEX_CYTOKINESIS_TIME]), 1, 2);
3842                 
3843                 save([outDirectory filesep <span class="string">'dnaBoundProteinDisplacement.mat'</span>], <span class="keyword">...</span>
3844                     <span class="string">'nSims'</span>, <span class="string">'simEndTimes'</span>, <span class="string">'allDisplacements'</span>, <span class="string">'nDisplacements'</span>, <span class="keyword">...</span>
3845                     <span class="string">'protByProtFreq'</span>, <span class="string">'timeFreq'</span>, <span class="string">'posFreq'</span>, <span class="keyword">...</span>
3846                     <span class="string">'repInitDuration'</span>, <span class="string">'repDuration'</span>, <span class="string">'cytokinesisDuration'</span>)
3847             <span class="keyword">else</span>
3848                 load([outDirectory filesep <span class="string">'dnaBoundProteinDisplacement.mat'</span>]);
3849             <span class="keyword">end</span>
3850             
3851             [mat, allIds, ids, names, shortNames] = Figures34.groupProteinDisplacements(protByProtFreq, simEndTimes, sim);
3852             
3853             monIds = intersect(allIds, pm.wholeCellModelIDs);
3854             cpxIds = intersect(allIds, pm.wholeCellModelIDs);
3855             
3856             [monTfs, monIdxs] = ismember(allIds, pm.wholeCellModelIDs(pm.matureIndexs));
3857             monIdxs = monIdxs(monTfs);
3858             
3859             [cpxTfs, cpxIdxs] = ismember(allIds, pc.wholeCellModelIDs(pc.matureIndexs));
3860             cpxIdxs = cpxIdxs(cpxTfs);
3861             
3862             <span class="keyword">if</span> ~exist([outDirectory filesep <span class="string">'proteinDensity.mat'</span>], <span class="string">'file'</span>)
3863                 protDensity = zeros(numel(allIds), c.sequenceLen);
3864                 <span class="keyword">for</span> i = 1:nSims
3865                     states = SimulationEnsemble.load(simBatchDir, {
3866                         <span class="string">'Chromosome'</span> <span class="string">'monomerBoundSites'</span>
3867                         <span class="string">'Chromosome'</span> <span class="string">'complexBoundSites'</span>
3868                         }, [], [], 1, <span class="string">'extract'</span>, i);
3869                     
3870                     [subs, vals] = find(states.Chromosome.monomerBoundSites);
3871                     tfs = ismember(vals, monIdxs);
3872                     protDensity(monTfs, :) = protDensity(monTfs, :) + hist3([vals(tfs) subs(tfs, 1)], {monIdxs 1:c.sequenceLen});
3873                     
3874                     [subs, vals] = find(states.Chromosome.complexBoundSites);
3875                     tfs = ismember(vals, cpxIdxs);
3876                     protDensity(cpxTfs, :) = protDensity(cpxTfs, :) + hist3([vals(tfs) subs(tfs, 1)], {cpxIdxs 1:c.sequenceLen});
3877                     
3878                     clear states;
3879                 <span class="keyword">end</span>
3880                 
3881                 save([outDirectory filesep <span class="string">'proteinDensity.mat'</span>], <span class="string">'protDensity'</span>);
3882             <span class="keyword">else</span>
3883                 load([outDirectory filesep <span class="string">'proteinDensity.mat'</span>]);
3884             <span class="keyword">end</span>
3885             
3886             <span class="comment">%%figure</span>
3887             x = position(1);
3888             y = position(2);
3889             W = position(3);
3890             H = position(4);
3891             paperSize = get(figHandle, <span class="string">'PaperSize'</span>);
3892             
3893             <span class="comment">%A</span>
3894             w = H*paperSize(2)/paperSize(1);
3895             Figures34.plotCollisionProteinDistribution(figHandle, [x y w H], [x+w+0.02 y 0.02 H], mat, shortNames);
3896             axesHandle = findobj(figHandle, <span class="string">'Position'</span>, [x y w H]);
3897             xlabelPos = get(get(axesHandle, <span class="string">'xlabel'</span>), <span class="string">'position'</span>);
3898             ylabelPos = get(get(axesHandle, <span class="string">'ylabel'</span>), <span class="string">'position'</span>);
3899             set(get(axesHandle, <span class="string">'xlabel'</span>), <span class="string">'FontSize'</span>, 7, <span class="string">'position'</span>, [xlabelPos(1) 16.4 xlabelPos(3:end)]);
3900             set(get(axesHandle, <span class="string">'ylabel'</span>), <span class="string">'FontSize'</span>, 7, <span class="string">'position'</span>, [-2.3 ylabelPos(2:end)]);
3901             xTicks = getappdata(axesHandle, <span class="string">'XTickText'</span>);
3902             yTicks = getappdata(axesHandle, <span class="string">'YTickText'</span>);
3903             <span class="keyword">for</span> i = 1:numel(xTicks)
3904                 set(xTicks(i), <span class="string">'position'</span>, [i 13.75 0])
3905                 set(yTicks(i), <span class="string">'position'</span>, [0.25 i 0])
3906             <span class="keyword">end</span>
3907             
3908             <span class="comment">%B</span>
3909             x = x+w+0.18;
3910             axesHandle = subplot(<span class="string">'Position'</span>, [x y 1-x H], <span class="string">'Parent'</span>, figHandle);
3911             tmpX = sum(reshape(sum(protDensity(:, 1:580000), 1), 1000, []) / sum(simEndTimes) / 1000 * 1e3, 1);
3912             tmpY = sum(reshape(posFreq(1:580000, 1)', 1000, []) / sum(simEndTimes) / 1000 * 3600 * 1e3, 1);
3913             plot(axesHandle, tmpX, tmpY, <span class="string">'.'</span>);
3914             xlim(axesHandle, [min(tmpX) 2.05] + range(tmpX)*0.025*[-1 0]);
3915             ylim(axesHandle, [min(tmpY) 14.2] + range(tmpY)*0.025*[-1 1]);
3916             set(axesHandle, <span class="string">'FontSize'</span>, 5, <span class="string">'box'</span>, <span class="string">'off'</span>, <span class="string">'tickDir'</span>, <span class="string">'out'</span>, <span class="string">'XTick'</span>, 0.5:0.5:3.5);
3917             xlabel(axesHandle, <span class="string">'Density (knt^{-1})'</span>, <span class="string">'FontSize'</span>, 7);
3918             ylabel(axesHandle, <span class="string">'Collisions (knt^{-1} h^{-1})'</span>, <span class="string">'FontSize'</span>, 7);
3919             xlabelPos = get(get(axesHandle, <span class="string">'xlabel'</span>), <span class="string">'position'</span>);
3920             ylabelPos = get(get(axesHandle, <span class="string">'ylabel'</span>), <span class="string">'position'</span>);
3921             set(get(axesHandle, <span class="string">'xlabel'</span>), <span class="string">'FontSize'</span>, 7, <span class="string">'position'</span>, [xlabelPos(1) -1.2 xlabelPos(3:end)]);
3922             set(get(axesHandle, <span class="string">'ylabel'</span>), <span class="string">'FontSize'</span>, 7, <span class="string">'position'</span>, [0.03 ylabelPos(2:end)]);
3923             tick2text(axesHandle, <span class="string">'axis'</span>, <span class="string">'xy'</span>, <span class="string">'xtickoffset'</span>, 0.05, <span class="string">'ytickoffset'</span>, 0.03);
3924             xTicks = getappdata(axesHandle, <span class="string">'XTickText'</span>);
3925             yTicks = getappdata(axesHandle, <span class="string">'YTickText'</span>);            
3926             set(xTicks, <span class="string">'FontSize'</span>, 5);
3927             set(yTicks, <span class="string">'FontSize'</span>, 5, <span class="string">'HorizontalAlign'</span>, <span class="string">'right'</span>);
3928             delete(xTicks(1:2:end));
3929             
3930             <span class="comment">%% supplement</span>
3931             fprintf(<span class="string">'Displacement rate: %0.2f (1/s)\n'</span>, nDisplacements / sum(simEndTimes));
3932             
3933             tmp = num2cell(mat);
3934             tmp(mat == 0) = {<span class="string">''</span>};
3935             content = [
3936                 [cell(3, 3) [ids names shortNames]']
3937                 [ids names shortNames tmp]
3938                 ];
3939             PrintUtil.printToFile(content, {}, [outDirectory filesep <span class="string">'dnaBoundProteinDisplacement.xls'</span>], <span class="string">'DNA Bound Protein Displacement'</span>);
3940             
3941             monIdxs = setdiff(unique([
3942                 find(any(c.reactionMonomerCatalysisMatrix, 1))'
3943                 c.reactionBoundMonomer
3944                 ]), 0);
3945             cpxIdxs = setdiff(unique([
3946                 find(any(c.reactionComplexCatalysisMatrix, 1))'
3947                 c.reactionBoundComplex
3948                 ]), 0);
3949             content = [[
3950                 pm.wholeCellModelIDs(pm.matureIndexs(monIdxs))
3951                 pc.wholeCellModelIDs(pc.matureIndexs(cpxIdxs))
3952                 ] [
3953                 pm.names(pm.matureIndexs(monIdxs))
3954                 pc.names(pc.matureIndexs(cpxIdxs))
3955                 ] cell(numel(monIdxs) + numel(cpxIdxs), 1)];
3956             tfs = ismember(content(:, 1), allIds);
3957             content(tfs, 3) = {<span class="string">'Y'</span>};
3958             PrintUtil.printToFile(content, {<span class="string">'ID'</span> <span class="string">'Name'</span> <span class="string">'Displaced/Displacing'</span>}, [outDirectory filesep <span class="string">'dnaBoundProteinDisplacement.xls'</span>], <span class="string">'Displac(ed|ing) Proteins'</span>);
3959                         
3960             <span class="comment">%protein distribution</span>
3961             w = 11.4; h = 9.7; <span class="comment">% Cell 1.5 column</span>
3962             [~, figHandle] = PlotUtil.newAxesHandle([w h]);
3963             clf(figHandle);
3964             
3965             Figures34.plotCollisionProteinDistribution(figHandle, [0.12 0.125 0.74 0.74*w/h], [0.90 0.125 0.03 0.74*w/h], mat, shortNames);
3966             
3967             saveas(figHandle, [outDirectory filesep <span class="string">'dnaBoundProteinDisplacement-proteins.pdf'</span>]);
3968             close(figHandle);
3969             
3970             <span class="comment">%time, spatial distribution</span>
3971             w = 11.4; h = 5; <span class="comment">% Cell 1.5 column</span>
3972             [~, figHandle] = PlotUtil.newAxesHandle([w h]);
3973             clf(figHandle);
3974             
3975             
3976             axesHandle = subplot(<span class="string">'Position'</span>, [0.07 0.12 0.40 0.835], <span class="string">'Parent'</span>, figHandle);
3977             x = linspace(1, 100, 1000);
3978             plot(axesHandle, x(2:end-1), timeFreq(2:end-1) / nSims / (sim.state(<span class="string">'Time'</span>).cellCycleLength / 1000));
3979             xlabel(axesHandle, <span class="string">'Time (% Cell Cycle)'</span>, <span class="string">'FontSize'</span>, 7);
3980             ylabel(axesHandle, {<span class="string">'Displacements (s^{-1})'</span>}, <span class="string">'FontSize'</span>, 7);
3981             xlim(axesHandle, [0 100])
3982             ylim(axesHandle, [0.65 1.3]);
3983             set(axesHandle, <span class="string">'box'</span>, <span class="string">'off'</span>, <span class="string">'tickdir'</span>, <span class="string">'out'</span>, <span class="string">'FontSize'</span>, 5, <span class="string">'XTick'</span>, 0:25:100, <span class="string">'YTick'</span>, 0.75:0.25:1.25);
3984             tick2text(axesHandle, <span class="string">'axis'</span>, <span class="string">'xy'</span>, <span class="string">'xtickoffset'</span>, 0.035, <span class="string">'ytickoffset'</span>, 0.02);
3985             xTicks = getappdata(axesHandle, <span class="string">'XTickText'</span>);
3986             yTicks = getappdata(axesHandle, <span class="string">'YTickText'</span>);
3987             set(xTicks, <span class="string">'FontSize'</span>, 5);
3988             set(yTicks, <span class="string">'FontSize'</span>, 5, <span class="string">'HorizontalAlign'</span>, <span class="string">'right'</span>);
3989             ylabelPos = get(get(axesHandle, <span class="string">'ylabel'</span>), <span class="string">'position'</span>);
3990             set(get(axesHandle, <span class="string">'ylabel'</span>), <span class="string">'position'</span>, [-10 ylabelPos(2:end)]);
3991             
3992             <a href="#_sub15" class="code" title="subfunction cellCycleLength(simBatchDir, selectedSim, figHandle, position)">cellCycleLength</a> = nanmean(repInitDuration + repDuration + cytokinesisDuration);
3993             y = [ylim(axesHandle) fliplr(ylim(axesHandle))];
3994             x = [0 0 nanmean(repInitDuration) * [1 1]] / <a href="#_sub15" class="code" title="subfunction cellCycleLength(simBatchDir, selectedSim, figHandle, position)">cellCycleLength</a> * 100;
3995             ptch = patch(x, y, 1, <span class="string">'Parent'</span>, axesHandle, <span class="string">'EdgeColor'</span>, <span class="string">'none'</span>, <span class="string">'FaceColor'</span>, 1 - [0.1 0.1 0.1], <span class="string">'FaceAlpha'</span>, 1, <span class="string">'EdgeAlpha'</span>, 1);
3996             uistack(ptch, <span class="string">'bottom'</span>);
3997             x = [nanmean(repInitDuration) * [1 1]  nanmean(repInitDuration+repDuration) * [1 1]] / <a href="#_sub15" class="code" title="subfunction cellCycleLength(simBatchDir, selectedSim, figHandle, position)">cellCycleLength</a> * 100;
3998             ptch = patch(x, y, 1, <span class="string">'Parent'</span>, axesHandle, <span class="string">'EdgeColor'</span>, <span class="string">'none'</span>, <span class="string">'FaceColor'</span>, 1 - [0.15 0.15 0.15], <span class="string">'FaceAlpha'</span>, 1, <span class="string">'EdgeAlpha'</span>, 1);
3999             uistack(ptch, <span class="string">'bottom'</span>);
4000             x = [nanmean(repInitDuration+repDuration) / <a href="#_sub15" class="code" title="subfunction cellCycleLength(simBatchDir, selectedSim, figHandle, position)">cellCycleLength</a> * [1 1] 1 1] * 100;
4001             ptch = patch(x, y, 1, <span class="string">'Parent'</span>, axesHandle, <span class="string">'EdgeColor'</span>, <span class="string">'none'</span>, <span class="string">'FaceColor'</span>, 1 - [0.1 0.1 0.1], <span class="string">'FaceAlpha'</span>, 1, <span class="string">'EdgeAlpha'</span>, 1);
4002             uistack(ptch, <span class="string">'bottom'</span>);
4003             line(xlim(axesHandle), min(ylim(axesHandle)) * [1 1], <span class="string">'Color'</span>, <span class="string">'k'</span>, <span class="string">'Parent'</span>, axesHandle);
4004             line(min(xlim(axesHandle)) * [1 1], ylim(axesHandle), <span class="string">'Color'</span>, <span class="string">'k'</span>, <span class="string">'Parent'</span>, axesHandle);
4005             text(nanmean(0.5 * repInitDuration)/ <a href="#_sub15" class="code" title="subfunction cellCycleLength(simBatchDir, selectedSim, figHandle, position)">cellCycleLength</a> * 100, min(ylim(axesHandle)) + 1.03 * range(ylim(axesHandle)), <span class="keyword">...</span>
4006                 <span class="string">'Replication Initiation'</span>, <span class="keyword">...</span>
4007                 <span class="string">'FontSize'</span>, 6, <span class="string">'HorizontalAlign'</span>, <span class="string">'center'</span>);
4008             text(nanmean(repInitDuration + 0.5 * repDuration)/ <a href="#_sub15" class="code" title="subfunction cellCycleLength(simBatchDir, selectedSim, figHandle, position)">cellCycleLength</a> * 100, min(ylim(axesHandle)) + 1.03 * range(ylim(axesHandle)), <span class="keyword">...</span>
4009                 <span class="string">'Replication'</span>, <span class="keyword">...</span>
4010                 <span class="string">'FontSize'</span>, 6, <span class="string">'HorizontalAlign'</span>, <span class="string">'center'</span>);
4011             text(nanmean(repInitDuration + repDuration + 0.5 * cytokinesisDuration)/ <a href="#_sub15" class="code" title="subfunction cellCycleLength(simBatchDir, selectedSim, figHandle, position)">cellCycleLength</a> * 100, min(ylim(axesHandle)) + 1.03 * range(ylim(axesHandle)), <span class="keyword">...</span>
4012                 <span class="string">'Cytokinesis'</span>, <span class="keyword">...</span>
4013                 <span class="string">'FontSize'</span>, 6, <span class="string">'HorizontalAlign'</span>, <span class="string">'center'</span>);
4014             
4015             
4016             axesHandle = subplot(<span class="string">'Position'</span>, [0.60 0.12 0.40 0.845], <span class="string">'Parent'</span>, figHandle);
4017             tmp = [
4018                 cconv(posFreq(:, 1) / nSims, ones(1000, 1) / 1000, c.sequenceLen) <span class="keyword">...</span>
4019                 cconv(posFreq(:, 2) / nSims, ones(1000, 1) / 1000, c.sequenceLen)
4020                 ] / sum(simEndTimes) * 1e9;
4021             h = plot(axesHandle, (1:c.sequenceLen)' / 1e3, tmp);
4022             xlabel(axesHandle, <span class="string">'Position (kb)'</span>, <span class="string">'FontSize'</span>, 7);
4023             ylabel(axesHandle, {<span class="string">'Displacements'</span> <span class="string">'(10^{-9} s^{-1})'</span>}, <span class="string">'FontSize'</span>, 7);
4024             xlim(axesHandle, [1 c.sequenceLen] / 1e3)
4025             ylim(axesHandle, [5 45]);
4026             set(axesHandle, <span class="string">'box'</span>, <span class="string">'off'</span>, <span class="string">'tickdir'</span>, <span class="string">'out'</span>, <span class="string">'FontSize'</span>, 5, <span class="string">'YTick'</span>, 10:10:40);            
4027             tick2text(axesHandle, <span class="string">'axis'</span>, <span class="string">'xy'</span>, <span class="string">'xtickoffset'</span>, 0.035, <span class="string">'ytickoffset'</span>, 0.02);
4028             xTicks = getappdata(axesHandle, <span class="string">'XTickText'</span>);
4029             yTicks = getappdata(axesHandle, <span class="string">'YTickText'</span>);
4030             set(xTicks, <span class="string">'FontSize'</span>, 5);
4031             set(yTicks, <span class="string">'FontSize'</span>, 5, <span class="string">'HorizontalAlign'</span>, <span class="string">'right'</span>);            
4032             legend(h, {<span class="string">'Unbinding'</span>; <span class="string">'Binding'</span>}, <span class="string">'Location'</span>, <span class="string">'NorthEast'</span>)
4033             legend(axesHandle, <span class="string">'boxoff'</span>);
4034             
4035             [peaks, locs] = findpeaks(tmp(:, 1), <span class="string">'NPEAKS'</span>, 5, <span class="string">'SORTSTR'</span>, <span class="string">'descend'</span>, <span class="string">'MINPEAKDISTANCE'</span>, 5000);
4036             
4037             g = sim.gene;
4038             <span class="keyword">for</span> i = 1:numel(locs)
4039                 gIdx = find(g.startCoordinates &lt;= locs(i) &amp; g.startCoordinates + g.lengths - 1 &gt;= locs(i));
4040                 text(locs(i) / 1e3, peaks(i), sprintf(<span class="string">'%s\n(%s)'</span>, g.names{gIdx}, g.wholeCellModelIDs{gIdx}), <span class="keyword">...</span>
4041                     <span class="string">'FontSize'</span>, 5, <span class="string">'HorizontalAlign'</span>, <span class="string">'center'</span>, <span class="string">'VerticalAlign'</span>, <span class="string">'bottom'</span>, <span class="string">'Interpreter'</span>, <span class="string">'none'</span>);
4042             <span class="keyword">end</span>
4043             
4044             saveas(figHandle, [outDirectory filesep <span class="string">'dnaBoundProteinDisplacement-time-space.pdf'</span>]);
4045             close(figHandle);
4046             
4047             <span class="comment">%% get data</span>
4048             dnaBoundProteinDisplacements = histc(allDisplacements(:, 3), 1:c.sequenceLen);
4049             dnaBoundProteinDisplacements = cconv(dnaBoundProteinDisplacements / nSims, ones(1000, 1) / 1000, c.sequenceLen) / sum(simEndTimes) * 1e9;
4050             
4051             <span class="comment">%% find meaning of peaks</span>
4052             proteinIdxs = zeros(size(ids));
4053             [~, tmp] = ismember(ids, pm.wholeCellModelIDs(pm.matureIndexs));
4054             proteinIdxs(tmp ~= 0) = tmp(tmp ~= 0);
4055             [~, tmp] = ismember(ids, pc.wholeCellModelIDs(pc.matureIndexs));
4056             proteinIdxs(tmp ~= 0) = -tmp(tmp ~= 0);
4057                         
4058             [~, peaks] = findpeaks(dnaBoundProteinDisplacements, <span class="string">'MINPEAKDISTANCE'</span>, 5000, <span class="string">'NPEAKS'</span>, 10, <span class="string">'SORTSTR'</span>, <span class="string">'descend'</span>);
4059             tuStarts = transcript.transcriptionUnitFivePrimeCoordinates - (transcript.transcriptionUnitLengths - 1) .* (transcript.transcriptionUnitDirections == 0);
4060             tuEnds = transcript.transcriptionUnitFivePrimeCoordinates + (transcript.transcriptionUnitLengths - 1) .* (transcript.transcriptionUnitDirections == 1);
4061             peakGenes = NaN(size(peaks));
4062             peakTUs = NaN(size(peaks));
4063             <span class="keyword">for</span> i = 1:numel(peaks)
4064                 tmp = find(g.startCoordinates &lt;= peaks(i) &amp; g.startCoordinates + g.lengths - 1 &gt;= peaks(i));
4065                 <span class="keyword">if</span> ~isempty(tmp)
4066                     peakGenes(i) = tmp;
4067                 <span class="keyword">end</span>
4068                 tmp = find(tuStarts &lt;= peaks(i) &amp; tuEnds &gt;= peaks(i));
4069                 <span class="keyword">if</span> ~isempty(tmp)
4070                     peakTUs(i) = tmp;
4071                 <span class="keyword">end</span>
4072                 
4073                 <span class="keyword">if</span> isnan(peakGenes(i))
4074                     <span class="keyword">continue</span>;
4075                 <span class="keyword">end</span>
4076                 
4077                 tfs = <span class="keyword">...</span>
4078                     allDisplacements(:, 3) &gt;= g.startCoordinates(peakGenes(i)) - 500  &amp; <span class="keyword">...</span>
4079                     allDisplacements(:, 3) &lt;= g.startCoordinates(peakGenes(i)) + g.lengths(peakGenes(i)) - 1 + 500;
4080                 idxs = allDisplacements(tfs, 1:2);
4081                 idxs(idxs &lt; 0) = -idxs(idxs &lt; 0) + nMon;
4082                 protByProtFreq = hist3(idxs, <span class="string">'Edges'</span>, {1:nMon+nCpx  1:nMon+nCpx});
4083                 [mat, allIds, ids, names, shortNames] = Figures34.groupProteinDisplacements(protByProtFreq, simEndTimes, sim);
4084                 
4085                 w = 11.4; h = 9.7; <span class="comment">% Cell 1.5 column</span>
4086                 [~, figHandle] = PlotUtil.newAxesHandle([w h]);
4087                 clf(figHandle);
4088                 Figures34.plotCollisionProteinDistribution(figHandle, [0.12 0.125 0.74 0.74*w/h], [0.90 0.125 0.03 0.74*w/h], mat, shortNames);
4089                 saveas(figHandle, [outDirectory filesep sprintf(<span class="string">'dnaBoundProteinDisplacement-proteins-%s.pdf'</span>, g.wholeCellModelIDs{peakGenes(i)})]);
4090                 close(figHandle);
4091             <span class="keyword">end</span>
4092             [g.wholeCellModelIDs(peakGenes(~isnan(peakGenes))) <span class="keyword">...</span><span class="comment">  %most displaced genes</span>
4093                 g.names(peakGenes(~isnan(peakGenes))) <span class="keyword">...</span>
4094                 num2cell(dnaBoundProteinDisplacements(peaks(~isnan(peakGenes)))) <span class="keyword">...</span>
4095                 num2cell(rna.nascentRNAGeneComposition(peakGenes(~isnan(peakGenes)), :) * transcription.transcriptionUnitBindingProbabilities)
4096                 ] <span class="comment">%#ok&lt;NOPRT&gt;</span>
4097             [transcript.transcriptionUnitWholeCellModelIDs(peakTUs(~isnan(peakTUs))) <span class="keyword">...</span><span class="comment"> %most displaced transcription units</span>
4098                 num2cell(dnaBoundProteinDisplacements(peaks(~isnan(peakTUs)))) <span class="keyword">...</span>
4099                 num2cell(transcription.transcriptionUnitBindingProbabilities(peakTUs(~isnan(peakTUs))))] <span class="comment">%#ok&lt;NOPRT&gt;</span>
4100             
4101             <span class="comment">%% correlate displacement spatial frequency with SMC density</span>
4102             [monTfs, monIdxs] = ismember(allIds, pm.wholeCellModelIDs(pm.matureIndexs));
4103             [cpxTfs, cpxIdxs] = ismember(allIds, pc.wholeCellModelIDs(pc.matureIndexs));
4104             tmp = zeros(size(protDensity));
4105             <span class="keyword">for</span> i = 1:size(protDensity, 1)
4106                 <span class="keyword">if</span> monTfs(i)
4107                     tmp(i, :) = cconv(protDensity(i, :)', ones(c.monomerDNAFootprints(monIdxs(i)), 1), c.sequenceLen);
4108                 <span class="keyword">else</span>
4109                     tmp(i, :) = cconv(protDensity(i, :)', ones(c.complexDNAFootprints(cpxIdxs(i)), 1), c.sequenceLen);
4110                 <span class="keyword">end</span>
4111             <span class="keyword">end</span>
4112             
4113             figHandle = figure();
4114             plot(sum(reshape(sum(protDensity(:, 1:580000), 1), 1000, []) / sum(simEndTimes) / 1000, 1), <span class="keyword">...</span>
4115                 sum(reshape(posFreq(1:580000, 1)', 1000, []) / sum(simEndTimes) / 1000, 1), <span class="keyword">...</span>
4116                 <span class="string">'.'</span>);
4117             xlabel(<span class="string">'Protein Density (nt^{-1})'</span>);
4118             ylabel(<span class="string">'Collisions Density (nt^{-1} s^{-1})'</span>);
4119             saveas(figHandle, [outDirectory filesep <span class="string">'proteinVsCollisionDensity.pdf'</span>]);
4120             close(figHandle);
4121             
4122             figHandle = figure();
4123             plot(sum(reshape(protDensity(strcmp(<span class="string">'MG_213_214_298_6MER_ADP'</span>, allIds), 1:580000), 1000, []) / sum(simEndTimes) / 1000, 1), <span class="keyword">...</span>
4124                 sum(reshape(posFreq(1:580000, 1)', 1000, []) / sum(simEndTimes) / 1000, 1), <span class="keyword">...</span>
4125                 <span class="string">'.'</span>);
4126             xlabel(<span class="string">'SMC Density'</span>);
4127             ylabel(<span class="string">'Collisions Density'</span>);
4128             saveas(figHandle, [outDirectory filesep <span class="string">'smcVsCollisionDensity.pdf'</span>]);
4129             close(figHandle);
4130         <span class="keyword">end</span>
4131         
4132         <a name="_sub26" href="#_subfunctions" class="code">function [mat, allIds, ids, names, shortNames] = groupProteinDisplacements(protByProtFreq, simEndTimes, sim)</a>
4133             pm = sim.state(<span class="string">'ProteinMonomer'</span>);
4134             pc = sim.state(<span class="string">'ProteinComplex'</span>);
4135             nMon = numel(pm.matureIndexs);
4136             
4137             [~, idxs] = ismember({
4138                 <span class="string">'MG_101_MONOMER'</span>
4139                 <span class="string">'MG_236_MONOMER'</span>
4140                 <span class="string">'DNA_GYRASE'</span>
4141                 <span class="string">'DNA_POLYMERASE_2CORE_BETA_CLAMP_GAMMA_COMPLEX_PRIMASE'</span>
4142                 <span class="string">'DNA_POLYMERASE_CORE_BETA_CLAMP_GAMMA_COMPLEX'</span>
4143                 <span class="string">'DNA_POLYMERASE_CORE_BETA_CLAMP_PRIMASE'</span>
4144                 <span class="string">'MG_001_DIMER'</span>
4145                 <span class="string">'MG_091_OCTAMER'</span>
4146                 <span class="string">'MG_094_HEXAMER'</span>
4147                 <span class="string">'MG_203_204_TETRAMER'</span>
4148                 <span class="string">'MG_205_DIMER'</span>
4149                 <span class="string">'MG_213_214_298_6MER_ADP'</span>
4150                 <span class="string">'MG_428_DIMER'</span>
4151                 <span class="string">'MG_469_1MER_ATP'</span>
4152                 <span class="string">'MG_469_1MER_ADP'</span>
4153                 <span class="string">'MG_469_7MER_ATP'</span>
4154                 <span class="string">'RNA_POLYMERASE'</span>
4155                 <span class="string">'RNA_POLYMERASE_HOLOENZYME'</span>
4156                 }, [pm.wholeCellModelIDs(pm.matureIndexs); pc.wholeCellModelIDs(pc.matureIndexs)]);            
4157             idxs = unique([idxs; find(any(protByProtFreq, 2) | any(protByProtFreq, 1)')]);
4158             
4159             mat = protByProtFreq(idxs, idxs) / sum(simEndTimes) * 3600;
4160             allIds = [
4161                 pm.wholeCellModelIDs(pm.matureIndexs(idxs(idxs &lt;= nMon)))
4162                 pc.wholeCellModelIDs(pc.matureIndexs(idxs(idxs &gt; nMon) - nMon))
4163                 ];
4164             ids = allIds;
4165             names = [
4166                 pm.names(pm.matureIndexs(idxs(idxs &lt;= nMon)))
4167                 pc.names(pc.matureIndexs(idxs(idxs &gt; nMon) - nMon))
4168                 ];
4169             groups = {
4170                 {<span class="string">'RNA_POLYMERASE'</span>, <span class="string">'RNA_POLYMERASE_HOLOENZYME'</span>}
4171                 {<span class="string">'DNA_POLYMERASE_2CORE_BETA_CLAMP_GAMMA_COMPLEX_PRIMASE'</span>, <span class="string">'DNA_POLYMERASE_CORE_BETA_CLAMP_GAMMA_COMPLEX'</span>, <span class="string">'DNA_POLYMERASE_CORE_BETA_CLAMP_PRIMASE'</span>}
4172                 {<span class="string">'MG_469_1MER_ATP'</span>, <span class="string">'MG_469_1MER_ADP'</span> <span class="string">'MG_469_7MER_ATP'</span>}
4173                 };
4174             <span class="keyword">for</span> i = 1:numel(groups)
4175                 [~, idxs] = ismember(groups{i}, ids);
4176                 idxs = idxs(idxs &gt; 0);
4177                 ids = ids(setdiff(1:<span class="keyword">end</span>, idxs(2:end)));
4178                 names = names(setdiff(1:<span class="keyword">end</span>, idxs(2:end)));
4179                 mat(:, idxs(1)) = mat(:, idxs(1)) + sum(mat(:, idxs(2:end)), 2);
4180                 mat(idxs(1), :) = mat(idxs(1), :) + sum(mat(idxs(2:end), :), 1);
4181                 mat = mat(setdiff(1:<span class="keyword">end</span>, idxs(2:end)), setdiff(1:<span class="keyword">end</span>, idxs(2:end)));
4182             <span class="keyword">end</span>
4183             
4184             shortNames = cell(size(ids));
4185             shortNames(strcmp(ids, <span class="string">'MG_101_MONOMER'</span>)) = {<span class="string">'GntR'</span>};
4186             shortNames(strcmp(ids, <span class="string">'MG_236_MONOMER'</span>)) = {<span class="string">'Fur'</span>};
4187             shortNames(strcmp(ids, <span class="string">'DNA_GYRASE'</span>)) = {<span class="string">'GyrAB'</span>};
4188             shortNames(strcmp(ids, <span class="string">'DNA_POLYMERASE_2CORE_BETA_CLAMP_GAMMA_COMPLEX_PRIMASE'</span>)) = {<span class="string">'DNA Pol'</span>};
4189             shortNames(strcmp(ids, <span class="string">'DNA_POLYMERASE_CORE_BETA_CLAMP_GAMMA_COMPLEX'</span>)) = {<span class="string">'DNA Pol'</span>};
4190             shortNames(strcmp(ids, <span class="string">'DNA_POLYMERASE_CORE_BETA_CLAMP_PRIMASE'</span>)) = {<span class="string">'DNA Pol'</span>};
4191             shortNames(strcmp(ids, <span class="string">'MG_001_DIMER'</span>)) = {<span class="string">'DnaN'</span>};
4192             shortNames(strcmp(ids, <span class="string">'MG_091_OCTAMER'</span>)) = {<span class="string">'SSB'</span>};
4193             shortNames(strcmp(ids, <span class="string">'MG_094_HEXAMER'</span>)) = {<span class="string">'DnaB'</span>};
4194             shortNames(strcmp(ids, <span class="string">'MG_203_204_TETRAMER'</span>)) = {<span class="string">'Topo IV'</span>};
4195             shortNames(strcmp(ids, <span class="string">'MG_205_DIMER'</span>)) = {<span class="string">'HrcA'</span>};
4196             shortNames(strcmp(ids, <span class="string">'MG_213_214_298_6MER_ADP'</span>)) = {<span class="string">'SMC'</span>};
4197             shortNames(strcmp(ids, <span class="string">'MG_428_DIMER'</span>)) = {<span class="string">'LuxR'</span>};
4198             shortNames(strcmp(ids, <span class="string">'MG_469_1MER_ATP'</span>)) = {<span class="string">'DnaA'</span>};
4199             shortNames(strcmp(ids, <span class="string">'MG_469_1MER_ADP'</span>)) = {<span class="string">'DnaA'</span>};
4200             shortNames(strcmp(ids, <span class="string">'MG_469_7MER_ATP'</span>)) = {<span class="string">'DnaA'</span>};
4201             shortNames(strcmp(ids, <span class="string">'RNA_POLYMERASE'</span>)) = {<span class="string">'RNA Pol'</span>};
4202             shortNames(strcmp(ids, <span class="string">'RNA_POLYMERASE_HOLOENZYME'</span>)) = {<span class="string">'RNA Pol'</span>};
4203                         
4204             [~, order] = sort(shortNames);
4205             ids = ids(order);
4206             names = names(order);
4207             shortNames = shortNames(order);
4208             mat = mat(order, order);
4209         <span class="keyword">end</span>
4210         
4211         <a name="_sub27" href="#_subfunctions" class="code">function plotCollisionProteinDistribution(figHandle, position, position2, mat, shortNames)</a>
4212             paperSize = get(figHandle, <span class="string">'PaperSize'</span>);
4213             w = paperSize(1);
4214             h = paperSize(2);
4215             
4216             axesHandle = subplot(<span class="string">'Position'</span>, position, <span class="string">'Parent'</span>, figHandle);
4217             
4218             tmp = (log10(mat) - -3.4) / (3.4 - -3.4);
4219             colorsB = 0.85 * (1 - tmp);
4220             colorsG = 0.85 * (1 - tmp);
4221             colorsR = ones(size(tmp));
4222             colorsR(mat == 0) = 1;
4223             colorsG(mat == 0) = 1;
4224             colorsB(mat == 0) = 1;
4225             colors = cat(3, colorsR, colorsG, colorsB);
4226             <span class="keyword">for</span> i = 1:size(colors, 1)
4227                 <span class="keyword">for</span> j = 1:size(colors, 2)
4228                     <span class="keyword">if</span> mat(i, j)
4229                         patch([j-0.5 j-0.5 j+0.5 j+0.5], [i-0.5 i+0.5 i+0.5 i-0.5], 1, <span class="keyword">...</span>
4230                             <span class="string">'FaceColor'</span>, permute(colors(i, j, :), [1 3 2]), <span class="keyword">...</span>
4231                             <span class="string">'EdgeColor'</span>, <span class="string">'none'</span>, <span class="string">'FaceAlpha'</span>, 1, <span class="string">'EdgeAlpha'</span>, 1, <span class="keyword">...</span>
4232                             <span class="string">'Parent'</span>, axesHandle);
4233                     <span class="keyword">end</span>
4234                 <span class="keyword">end</span>
4235             <span class="keyword">end</span>
4236             set(axesHandle, <span class="keyword">...</span>
4237                 <span class="string">'FontSize'</span>, 5, <span class="string">'TickDir'</span>, <span class="string">'out'</span>, <span class="string">'box'</span>, <span class="string">'off'</span>, <span class="keyword">...</span>
4238                 <span class="string">'GridLineStyle'</span>, <span class="string">'none'</span>, <span class="keyword">...</span>
4239                 <span class="string">'MinorGridLineStyle'</span>, <span class="string">'-'</span>, <span class="keyword">...</span>
4240                 <span class="string">'XTick'</span>, 1:size(mat, 1), <span class="string">'XAxisLocation'</span>, <span class="string">'bottom'</span>, <span class="keyword">...</span>
4241                 <span class="string">'YTick'</span>, 1:size(mat, 1), <span class="string">'YDir'</span>, <span class="string">'reverse'</span>, <span class="string">'visible'</span>, <span class="string">'on'</span>);
4242             tick2text(axesHandle, <span class="string">'axis'</span>, <span class="string">'xy'</span>, <span class="string">'xtickoffset'</span>, -1.02, <span class="string">'ytickoffset'</span>, 0.02);
4243             xTicks = getappdata(axesHandle, <span class="string">'XTickText'</span>);
4244             yTicks = getappdata(axesHandle, <span class="string">'YTickText'</span>);
4245             set(xTicks, <span class="string">'FontSize'</span>, 5, <span class="string">'HorizontalAlign'</span>, <span class="string">'right'</span>, <span class="string">'interpreter'</span>, <span class="string">'none'</span>, <span class="string">'rotation'</span>, 90);
4246             set(yTicks, <span class="string">'FontSize'</span>, 5, <span class="string">'HorizontalAlign'</span>, <span class="string">'right'</span>, <span class="string">'interpreter'</span>, <span class="string">'none'</span>);
4247             <span class="keyword">for</span> i = 1:numel(shortNames)
4248                 set(xTicks(i), <span class="string">'String'</span>, shortNames{i}(1:min(<span class="keyword">end</span>, 24)))
4249                 set(yTicks(i), <span class="string">'String'</span>, shortNames{i}(1:min(<span class="keyword">end</span>, 24)))
4250             <span class="keyword">end</span>
4251             xlabel(axesHandle, <span class="string">'Binding'</span>, <span class="string">'FontSize'</span>, 7);
4252             ylabel(axesHandle, <span class="string">'Unbinding'</span>, <span class="string">'FontSize'</span>, 7);
4253             xlabelPos = get(get(axesHandle, <span class="string">'xlabel'</span>), <span class="string">'position'</span>);
4254             ylabelPos = get(get(axesHandle, <span class="string">'ylabel'</span>), <span class="string">'position'</span>);
4255             set(get(axesHandle, <span class="string">'xlabel'</span>), <span class="string">'position'</span>, [xlabelPos(1) 14.8 xlabelPos(3:end)]);
4256             set(get(axesHandle, <span class="string">'ylabel'</span>), <span class="string">'position'</span>, [-1 ylabelPos(2:end)]);
4257             xlim(axesHandle, [0.5 size(mat, 1)+0.5]);
4258             ylim(axesHandle, [0.5 size(mat, 1)+0.5]);
4259             <span class="keyword">for</span> i = 0.5:size(mat, 1)+0.5
4260                 line([0.5 size(mat, 1)+0.5], [i i], <span class="string">'Parent'</span>, axesHandle, <span class="string">'Color'</span>, <span class="string">'k'</span>);
4261                 line([i i], [0.5 size(mat, 1)+0.5], <span class="string">'Parent'</span>, axesHandle, <span class="string">'Color'</span>, <span class="string">'k'</span>);
4262             <span class="keyword">end</span>
4263             axis(axesHandle, <span class="string">'square'</span>);
4264             
4265             axesHandle = subplot(<span class="string">'Position'</span>, position2, <span class="string">'Parent'</span>, figHandle);            
4266             image(cat(3, <span class="keyword">...</span>
4267                 ones(69, 1), <span class="keyword">...</span>
4268                 0.85 * (1 - linspace(0, 1, 69))', <span class="keyword">...</span>
4269                 0.85 * (1 - linspace(0, 1, 69))'), <span class="keyword">...</span>
4270                 <span class="string">'Parent'</span>, axesHandle);
4271             set(axesHandle, <span class="keyword">...</span>
4272                 <span class="string">'FontSize'</span>, 5, <span class="string">'TickDir'</span>, <span class="string">'out'</span>, <span class="string">'box'</span>, <span class="string">'off'</span>, <span class="keyword">...</span>
4273                 <span class="string">'XTick'</span>, [], <span class="keyword">...</span>
4274                 <span class="string">'YTick'</span>, 15:20:55, <span class="keyword">...</span>
4275                 <span class="string">'YAxisLocation'</span>, <span class="string">'right'</span>, <span class="string">'YDir'</span>, <span class="string">'normal'</span>);
4276             tick2text(axesHandle, <span class="string">'axis'</span>, <span class="string">'y'</span>, <span class="string">'ytickoffset'</span>, -1.2);
4277             yTicks = getappdata(axesHandle, <span class="string">'YTickText'</span>);
4278             set(yTicks, <span class="string">'FontSize'</span>, 5, <span class="string">'HorizontalAlign'</span>, <span class="string">'left'</span>, <span class="string">'interpreter'</span>, <span class="string">'tex'</span>);
4279             yTickLabels = {<span class="string">'10^{-2}'</span>, <span class="string">'10^{0}'</span>, <span class="string">'10^{2}'</span>};
4280             <span class="keyword">for</span> i = 1:numel(yTicks)
4281                 set(yTicks(i), <span class="string">'String'</span>, yTickLabels{i});
4282             <span class="keyword">end</span>
4283             xlim(axesHandle, [0.5 1.5]);
4284             ylim(axesHandle, [0.5 69.5]);
4285             ylabel(axesHandle, <span class="string">'Freq (h^{-1})'</span>, <span class="string">'FontSize'</span>, 7);
4286             line([0.5 0.5], [0.5 69.5], <span class="string">'Parent'</span>, axesHandle, <span class="string">'Color'</span>, <span class="string">'k'</span>)
4287             line([0.5 1.5], [69.5 69.5], <span class="string">'Parent'</span>, axesHandle, <span class="string">'Color'</span>, <span class="string">'k'</span>)
4288         <span class="keyword">end</span>
4289         
4290         <a name="_sub28" href="#_subfunctions" class="code">function totalDensityMatrix = calculateDNABoundProteinDensity(outDirectory, simBatchDir, selectedSim)</a>
4291             import edu.stanford.covert.cell.sim.analysis.ChromosomePositionHistogram;
4292             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
4293             import edu.stanford.covert.cell.sim.util.SimulationEnsemble;
4294             
4295             <span class="keyword">if</span> exist([outDirectory <span class="string">'DNABoundProteinDensity.mat'</span>], <span class="string">'file'</span>)
4296                 load([outDirectory <span class="string">'DNABoundProteinDensity.mat'</span>])
4297             <span class="keyword">else</span>
4298                 <span class="comment">%% get constants</span>
4299                 [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep num2str(selectedSim)]);
4300                 
4301                 stateNames = {
4302                     <span class="string">'Chromosome'</span> <span class="string">'complexBoundSites'</span>        <span class="string">':'</span> <span class="string">':'</span>
4303                     <span class="string">'Chromosome'</span> <span class="string">'monomerBoundSites'</span>        <span class="string">':'</span> <span class="string">':'</span>
4304                     };
4305                 states = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, <span class="string">'extract'</span>, selectedSim);
4306                 
4307                 complexIdxs = full(unique(states.Chromosome.complexBoundSites));
4308                 complexIdxs = complexIdxs(complexIdxs &gt; 0);
4309                 
4310                 monomerIdxs = full(unique(states.Chromosome.monomerBoundSites));
4311                 monomerIdxs = monomerIdxs(monomerIdxs &gt; 0);
4312                 
4313                 boundIndices = struct(<span class="string">'complex'</span>, {num2cell(complexIdxs')}, <span class="string">'monomer'</span>, {num2cell(monomerIdxs')});
4314                 
4315                 [cDensityMatrix, mDensityMatrix] = ChromosomePositionHistogram.makeDensityMatrix(states, boundIndices, sim);
4316                 totalDensityMatrix = sum(cDensityMatrix, 2) + sum(mDensityMatrix, 2);
4317                 
4318                 save([outDirectory <span class="string">'DNABoundProteinDensity.mat'</span>], <span class="string">'totalDensityMatrix'</span>)
4319             <span class="keyword">end</span>
4320         <span class="keyword">end</span>
4321         
4322         <a name="_sub29" href="#_subfunctions" class="code">function displacements = calculateDNABoundProteinDisplacements(outDirectory, simBatchDir, selectedSim, iJob, nJobs)</a>
4323             import edu.stanford.covert.cell.sim.util.SimulationEnsemble;
4324             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
4325             
4326             <span class="keyword">if</span> nargin &lt; 5
4327                 nJobs = 1;
4328             <span class="keyword">end</span>
4329             
4330             relSimBatchDir = simBatchDir;
4331             <span class="comment">% Make simBatchDir a relative path</span>
4332             <span class="keyword">while</span> relSimBatchDir(end) == filesep
4333                 relSimBatchDir(end) = [];
4334             <span class="keyword">end</span>
4335             relSimBatchDir = [relSimBatchDir filesep];
4336             p = find(relSimBatchDir == filesep, 2, <span class="string">'last'</span>);
4337             relSimBatchDir = relSimBatchDir(p(1) + 1:p(2) - 1);
4338             
4339             outDirectory = [outDirectory filesep];
4340             
4341             <span class="comment">%% name of file where results will be cached</span>
4342             <span class="keyword">if</span> ~exist(sprintf(<span class="string">'%sdnaBoundProteinDisplacement'</span>, outDirectory), <span class="string">'dir'</span>)
4343                 mkdir(sprintf(<span class="string">'%sdnaBoundProteinDisplacement'</span>, outDirectory));
4344             <span class="keyword">end</span>
4345             <span class="keyword">if</span> nJobs == 1 || nargin &lt; 4 || isempty(iJob)
4346                 fileName = sprintf(<span class="string">'%sdnaBoundProteinDisplacement%s%s-%d.mat'</span>, outDirectory, filesep, relSimBatchDir, selectedSim);
4347             <span class="keyword">else</span>
4348                 fileName = sprintf(<span class="string">'%sdnaBoundProteinDisplacement%s%s-%d-%d-%d.mat'</span>, outDirectory, filesep, relSimBatchDir, selectedSim, iJob, nJobs);
4349             <span class="keyword">end</span>
4350             
4351             <span class="comment">%% get parameters</span>
4352             [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep num2str(selectedSim)]);
4353             c = sim.state(<span class="string">'Chromosome'</span>);
4354             pm = sim.state(<span class="string">'ProteinMonomer'</span>);
4355             pc = sim.state(<span class="string">'ProteinComplex'</span>);
4356             
4357             <span class="comment">%% if data already computed, just load data</span>
4358             <span class="keyword">if</span> exist(fileName, <span class="string">'file'</span>)
4359                 load(fileName);
4360                 <span class="keyword">return</span>;
4361             <span class="keyword">end</span>
4362             
4363             <span class="comment">%% stitch together jobs</span>
4364             <span class="keyword">if</span> nargin &gt;= 5 &amp;&amp; isempty(iJob)
4365                 displacements = zeros(0, 5);
4366                 <span class="keyword">for</span> iJob = 1:nJobs
4367                     tmp = load(sprintf(<span class="string">'%sdnaBoundProteinDisplacement-%s-%d-%d-%d.mat'</span>, outDirectory, simBatchDir, selectedSim, iJob, nJobs));
4368                     displacements = [displacements; tmp.displacements]; <span class="comment">%#ok&lt;AGROW&gt;</span>
4369                 <span class="keyword">end</span>
4370                 save(fileName, <span class="string">'displacements'</span>);
4371                 <span class="keyword">return</span>;
4372             <span class="keyword">end</span>
4373             
4374             <span class="comment">%% get data</span>
4375             stateNames = {
4376                 <span class="string">'Chromosome'</span> <span class="string">'monomerBoundSites'</span>
4377                 <span class="string">'Chromosome'</span> <span class="string">'complexBoundSites'</span>
4378                 };
4379             displacements = zeros(0, 5);
4380             
4381             states = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, <span class="string">'extract'</span>, selectedSim);
4382             
4383             monBndSites = permute(max(states.Chromosome.monomerBoundSites(:, 1:2, :), [], 2), [1 3 2]);
4384             cpxBndSites = permute(max(states.Chromosome.complexBoundSites(:, 1:2, :), [], 2), [1 3 2]);
4385             
4386             monMonDisplacement = zeros(numel(pm.matureIndexs), numel(pm.matureIndexs));
4387             monCpxDisplacement = zeros(numel(pm.matureIndexs), numel(pc.matureIndexs));
4388             cpxMonDisplacement = zeros(numel(pc.matureIndexs), numel(pm.matureIndexs));
4389             cpxCpxDisplacement = zeros(numel(pc.matureIndexs), numel(pc.matureIndexs));
4390             
4391             [i, j] = find(c.reactionMonomerCatalysisMatrix);
4392             tfs = c.reactionBoundMonomer(i) ~= 0;
4393             monMonDisplacement(sub2ind(size(monMonDisplacement), c.reactionBoundMonomer(i(tfs)), j(tfs))) = 1;
4394             cpxMonDisplacement(sub2ind(size(cpxMonDisplacement), c.reactionBoundComplex(i(~tfs)), j(~tfs))) = 1;
4395             
4396             [i, j] = find(c.reactionComplexCatalysisMatrix);
4397             tfs = c.reactionBoundMonomer(i) ~= 0;
4398             monCpxDisplacement(sub2ind(size(monCpxDisplacement), c.reactionBoundMonomer(i(tfs)), j(tfs))) = 1;
4399             cpxCpxDisplacement(sub2ind(size(cpxCpxDisplacement), c.reactionBoundComplex(i(~tfs)), j(~tfs))) = 1;
4400             
4401             monDNAFootprints = c.monomerDNAFootprints;
4402             cpxDNAFootprints = c.complexDNAFootprints;
4403             cpxDNAFootprints(pc.rnaPolymeraseIndexs) = max(cpxDNAFootprints(pc.rnaPolymeraseIndexs), sim.process(<span class="string">'Transcription'</span>).rnaPolymeraseElongationRate);
4404             cpxDNAFootprints(pc.dnaPolymeraseIndexs) = max(cpxDNAFootprints(pc.dnaPolymeraseIndexs), sim.process(<span class="string">'Replication'</span>).dnaPolymeraseElongationRate);
4405             
4406             monDNANegFootprints = zeros(size(c.monomerDNAFootprints));
4407             cpxDNANegFootprints = zeros(size(c.complexDNAFootprints));
4408             cpxDNANegFootprints(pc.rnaPolymeraseIndexs) = sim.process(<span class="string">'Transcription'</span>).rnaPolymeraseElongationRate - c.complexDNAFootprints(pc.rnaPolymeraseIndexs);
4409             cpxDNANegFootprints(pc.dnaPolymeraseIndexs) = sim.process(<span class="string">'Replication'</span>).dnaPolymeraseElongationRate - c.complexDNAFootprints(pc.dnaPolymeraseIndexs);
4410             cpxDNANegFootprints = max(0, cpxDNANegFootprints);
4411             
4412             <span class="keyword">for</span> i = iJob:nJobs:size(states.Chromosome.monomerBoundSites, 3) - 1
4413                 monBndSites1 = monBndSites(:, i);
4414                 monBndSites2 = monBndSites(:, i+1);
4415                 [subs1, vals1] = find(monBndSites1);
4416                 [subs2, vals2] = find(monBndSites2);
4417                 monBndSites1(subs2(monBndSites1(subs2) == vals2, :)) = 0;
4418                 monBndSites2(subs1(monBndSites2(subs1) == vals1, :)) = 0;
4419                 [monPos1, monIdxs1] = find(monBndSites1);
4420                 [monPos2, monIdxs2] = find(monBndSites2);
4421                 monStarts1 = monPos1(:, 1) - monDNANegFootprints(monIdxs1);
4422                 monStarts2 = monPos2(:, 1) - monDNANegFootprints(monIdxs2);
4423                 monEnds1 = monPos1(:, 1) + monDNAFootprints(monIdxs1);
4424                 monEnds2 = monPos2(:, 1) + monDNAFootprints(monIdxs2);
4425                 
4426                 cpxBndSites1 = cpxBndSites(:, i);
4427                 cpxBndSites2 = cpxBndSites(:, i+1);
4428                 [subs1, vals1] = find(cpxBndSites1);
4429                 [subs2, vals2] = find(cpxBndSites2);
4430                 cpxBndSites1(subs2(cpxBndSites1(subs2) == vals2, :)) = 0;
4431                 cpxBndSites2(subs1(cpxBndSites2(subs1) == vals1, :)) = 0;
4432                 [cpxPos1, cpxIdxs1] = find(cpxBndSites1);
4433                 [cpxPos2, cpxIdxs2] = find(cpxBndSites2);
4434                 cpxStarts1 = cpxPos1(:, 1) - cpxDNANegFootprints(cpxIdxs1);
4435                 cpxStarts2 = cpxPos2(:, 1) - cpxDNANegFootprints(cpxIdxs2);
4436                 cpxEnds1 = cpxPos1(:, 1) + cpxDNAFootprints(cpxIdxs1);
4437                 cpxEnds2 = cpxPos2(:, 1) + cpxDNAFootprints(cpxIdxs2);
4438                 
4439                 <span class="keyword">if</span> any(any(monMonDisplacement(monIdxs1, monIdxs2)))
4440                     <span class="keyword">for</span> j = 1:size(monPos1, 1)
4441                         tfs = <span class="keyword">...</span>
4442                             (monStarts1(j) &gt;= monStarts2 &amp;  monStarts1(j) &lt;= monEnds2) | <span class="keyword">...</span>
4443                             (monEnds1(j) &gt;= monStarts2  &amp;  monEnds1(j) &lt;= monEnds2) | <span class="keyword">...</span>
4444                             (monStarts1(j) &lt;= monStarts2  &amp;  monEnds1(j) &gt;= monEnds2);
4445                         <span class="keyword">if</span> any(monMonDisplacement(monIdxs1(j), monIdxs2(tfs)))
4446                             displacements = [displacements;
4447                                 monIdxs1(j(ones(sum(tfs), 1))) monIdxs2(tfs) monStarts1(j(ones(sum(tfs), 1))) monStarts2(tfs) i(ones(sum(tfs), 1))]; <span class="comment">%#ok&lt;AGROW&gt;</span>
4448                         <span class="keyword">end</span>
4449                     <span class="keyword">end</span>
4450                 <span class="keyword">end</span>
4451                 
4452                 <span class="keyword">if</span> any(any(monCpxDisplacement(monIdxs1, cpxIdxs2)))
4453                     <span class="keyword">for</span> j = 1:size(monPos1, 1)
4454                         tfs = <span class="keyword">...</span>
4455                             (monStarts1(j) &gt;= cpxStarts2 &amp;  monStarts1(j) &lt;= cpxEnds2) | <span class="keyword">...</span>
4456                             (monEnds1(j) &gt;= cpxStarts2  &amp;  monEnds1(j) &lt;= cpxEnds2) | <span class="keyword">...</span>
4457                             (monStarts1(j) &lt;= cpxStarts2  &amp;  monEnds1(j) &gt;= cpxEnds2);
4458                         <span class="keyword">if</span> any(monCpxDisplacement(monIdxs1(j), cpxIdxs2(tfs)))
4459                             displacements = [displacements;
4460                                 monIdxs1(j(ones(sum(tfs), 1))) -cpxIdxs2(tfs) monStarts1(j(ones(sum(tfs), 1))) cpxStarts2(tfs) i(ones(sum(tfs), 1))]; <span class="comment">%#ok&lt;AGROW&gt;</span>
4461                         <span class="keyword">end</span>
4462                     <span class="keyword">end</span>
4463                 <span class="keyword">end</span>
4464                 
4465                 <span class="keyword">if</span> any(any(cpxMonDisplacement(cpxIdxs1, monIdxs2)))
4466                     <span class="keyword">for</span> j = 1:size(monPos2, 1)
4467                         tfs = <span class="keyword">...</span>
4468                             (cpxStarts1 &gt;= monStarts2(j) &amp;  cpxStarts1 &lt;= monEnds2(j)) | <span class="keyword">...</span>
4469                             (cpxEnds1 &gt;= monStarts2(j)  &amp;  cpxEnds1 &lt;= monEnds2(j)) | <span class="keyword">...</span>
4470                             (cpxStarts1 &lt;= monStarts2(j)  &amp;  cpxEnds1 &gt;= monEnds2(j));
4471                         <span class="keyword">if</span> any(cpxMonDisplacement(cpxIdxs1(tfs), monIdxs2(j)))
4472                             displacements = [displacements;
4473                                 -cpxIdxs1(tfs) monIdxs2(j(ones(sum(tfs), 1))) cpxStarts1(tfs) monStarts2(j(ones(sum(tfs), 1))) i(ones(sum(tfs), 1))]; <span class="comment">%#ok&lt;AGROW&gt;</span>
4474                         <span class="keyword">end</span>
4475                     <span class="keyword">end</span>
4476                 <span class="keyword">end</span>
4477                 
4478                 <span class="keyword">if</span> any(any(cpxCpxDisplacement(cpxIdxs1, cpxIdxs2)))
4479                     <span class="keyword">for</span> j = 1:size(cpxPos1, 1)
4480                         tfs = <span class="keyword">...</span>
4481                             (cpxStarts1(j) &gt;= cpxStarts2 &amp;  cpxStarts1(j) &lt;= cpxEnds2) | <span class="keyword">...</span>
4482                             (cpxEnds1(j) &gt;= cpxStarts2  &amp;  cpxEnds1(j) &lt;= cpxEnds2) | <span class="keyword">...</span>
4483                             (cpxStarts1(j) &lt;= cpxStarts2  &amp;  cpxEnds1(j) &gt;= cpxEnds2);
4484                         <span class="keyword">if</span> any(cpxCpxDisplacement(cpxIdxs1(j), cpxIdxs2(tfs)))
4485                             displacements = [displacements;
4486                                 -cpxIdxs1(j(ones(sum(tfs), 1))) -cpxIdxs2(tfs) cpxStarts1(j(ones(sum(tfs), 1))) cpxStarts2(tfs) i(ones(sum(tfs), 1))]; <span class="comment">%#ok&lt;AGROW&gt;</span>
4487                         <span class="keyword">end</span>
4488                     <span class="keyword">end</span>
4489                 <span class="keyword">end</span>
4490             <span class="keyword">end</span>
4491             
4492             <span class="comment">%% save</span>
4493             save(fileName, <span class="string">'displacements'</span>);
4494         <span class="keyword">end</span>
4495         
4496         <a name="_sub30" href="#_subfunctions" class="code">function metabolicMap(simBatchDir, selectedSim, outDirectory, figHandle, position, showAllLabels)</a>
4497             import edu.stanford.covert.cell.kb.KnowledgeBaseUtil;
4498             import edu.stanford.covert.cell.sim.util.SimulationEnsemble;
4499             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
4500             import edu.stanford.covert.cell.sim.util.PlotUtil;
4501             import edu.stanford.covert.cell.sim.util.PrintUtil;
4502             import edu.stanford.covert.cell.sim.analysis.Figures34;
4503             import edu.stanford.covert.db.MySQLDatabase;
4504             
4505             <span class="keyword">if</span> ~exist(<span class="string">'simBatchDir'</span>, <span class="string">'var'</span>) || isempty(simBatchDir)
4506                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
4507             <span class="keyword">end</span>
4508             <span class="keyword">if</span> ~exist(<span class="string">'selectedSim'</span>, <span class="string">'var'</span>) || isempty(selectedSim)
4509                 selectedSim = 1;
4510             <span class="keyword">end</span>
4511             <span class="keyword">if</span> nargin &lt; 6
4512                 showAllLabels = false;
4513             <span class="keyword">end</span>
4514             
4515             <span class="comment">%% get constants</span>
4516             [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep num2str(selectedSim)]);
4517             comp = sim.compartment;
4518             met = sim.state(<span class="string">'Metabolite'</span>);
4519             mr = sim.state(<span class="string">'MetabolicReaction'</span>);
4520             metabolism = sim.process(<span class="string">'Metabolism'</span>);
4521             
4522             <span class="comment">%% get data</span>
4523             <span class="keyword">if</span> exist([outDirectory <span class="string">'metabolicMap.mat'</span>], <span class="string">'file'</span>)
4524                 load([outDirectory <span class="string">'metabolicMap.mat'</span>])
4525             <span class="keyword">else</span>
4526                 dbConnectionParameters = config();
4527                 database = MySQLDatabase(dbConnectionParameters);
4528                 database.setNullValue(0);
4529                 
4530                 kbWID = KnowledgeBaseUtil.selectLatestKnowledgeBase(database);
4531                 
4532                 database.prepareStatement(<span class="string">'CALL get_metabolicmapmetabolites(&quot;{Si}&quot;, null)'</span>, kbWID);
4533                 data = database.query();
4534                 mapMets = struct;
4535                 [~, mapMets.idxs] = ismember(data.Metabolite, met.wholeCellModelIDs);
4536                 [~, mapMets.compIdxs] = ismember(data.Compartment, comp.wholeCellModelIDs);
4537                 [~, mapMets.uidxs] = ismember(mapMets.idxs, unique(mapMets.idxs));
4538                 mapMets.x = data.X;
4539                 mapMets.y = cellfun(@str2double, data.Y);
4540                 
4541                 database.prepareStatement(<span class="string">'CALL get_metabolicmapreactions(&quot;{Si}&quot;, null)'</span>, kbWID);
4542                 data = database.query();
4543                 mapRxns = struct;
4544                 [~, mapRxns.idxs] = ismember(data.Reaction, mr.reactionWholeCellModelIDs);
4545                 [~, mapRxns.uidxs] = ismember(mapRxns.idxs, unique(mapRxns.idxs));
4546                 mapRxns.path = data.Path;
4547                 mapRxns.labelX = data.LabelX;
4548                 mapRxns.labelY = data.LabelY;
4549                 mapRxns.valueX = data.ValueX;
4550                 mapRxns.valueY = data.ValueY;
4551                 
4552                 <span class="comment">%% get data</span>
4553                 stateNames = {
4554                     <span class="string">'MetabolicReaction'</span>  <span class="string">'fluxs'</span>  <span class="string">':'</span>  <span class="string">':'</span>
4555                     };
4556                 states = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, <span class="string">'extract'</span>, selectedSim);
4557                 
4558                 states.MetabolicReaction.fluxs = full(states.MetabolicReaction.fluxs);
4559                 
4560                 <span class="comment">%% table</span>
4561                 colLabels = {<span class="string">'Reaction ID'</span>, <span class="string">'Reaction Name'</span>, <span class="string">'Flux Mean (rxn/s)'</span>, <span class="string">'Flux Std (rxn/s)'</span>};
4562                 content = [
4563                     metabolism.reactionWholeCellModelIDs  <span class="keyword">...</span>
4564                     metabolism.reactionNames  <span class="keyword">...</span>
4565                     num2cell(nanmean(states.MetabolicReaction.fluxs, 3)) <span class="keyword">...</span>
4566                     num2cell(nanstd(states.MetabolicReaction.fluxs, [], 3))
4567                     ];
4568                 PrintUtil.printToFile(content, colLabels, [outDirectory <span class="string">'reactionFluxes.xls'</span>], <span class="string">'Reaction Fluxes'</span>);
4569                 
4570                 <span class="comment">%% iso fluxs</span>
4571                 isoFluxIDs = {
4572                     {<span class="string">'Adk1'</span>, <span class="string">'Adk2'</span>, <span class="string">'Adk3'</span>, <span class="string">'Adk4'</span>}
4573                     {<span class="string">'PfkA1'</span>, <span class="string">'PfkA2'</span>, <span class="string">'PfkA3'</span>, <span class="string">'PfkA4'</span>, <span class="string">'PfkA5'</span>}
4574                     {<span class="string">'Pgk1'</span>, <span class="string">'Pgk2'</span>, <span class="string">'Pgk3'</span>, <span class="string">'Pgk4'</span>, <span class="string">'Pgk5'</span>}
4575                     {<span class="string">'Pyk_ADP'</span>, <span class="string">'Pyk_CDP'</span>, <span class="string">'Pyk_DADP'</span>, <span class="string">'Pyk_DCDP'</span>, <span class="string">'Pyk_DGDP'</span>, <span class="string">'Pyk_DTDP'</span>, <span class="string">'Pyk_DUDP'</span>, <span class="string">'Pyk_GDP'</span>, <span class="string">'Pyk_IDP'</span>, <span class="string">'Pyk_TDP'</span>, <span class="string">'Pyk_UDP'</span>}
4576                     };
4577                 isoFluxs = false(size(mr.reactionWholeCellModelIDs));
4578                 <span class="keyword">for</span> i = 1:numel(isoFluxIDs)
4579                     tfs2 = ismember(mr.reactionWholeCellModelIDs, isoFluxIDs{i}(2:end));
4580                     isoFluxs = isoFluxs | tfs2;
4581                     
4582                     [~, idxs1] = ismember(isoFluxIDs{i}(1), mr.reactionWholeCellModelIDs);
4583                     [~, idxs2] = ismember(isoFluxIDs{i}(2:end), mr.reactionWholeCellModelIDs);
4584                     states.MetabolicReaction.fluxs(idxs1, :, :) = <span class="keyword">...</span>
4585                         + states.MetabolicReaction.fluxs(idxs1, :, :) <span class="keyword">...</span>
4586                         + sum(states.MetabolicReaction.fluxs(idxs2, :, :), 1);
4587                 <span class="keyword">end</span>
4588                 
4589                 fluxMeans = nanmean(states.MetabolicReaction.fluxs(mapRxns.idxs, :, :), 3);
4590                 fluxStds = nanstd(states.MetabolicReaction.fluxs(mapRxns.idxs, :, :), [], 3) ./ abs(fluxMeans); <span class="comment">% (Standard Deviation)/|Mean|</span>
4591                 
4592                 <span class="comment">%% save</span>
4593                 save([outDirectory <span class="string">'metabolicMap.mat'</span>], <span class="string">'mapMets'</span>, <span class="string">'mapRxns'</span>, <span class="string">'fluxMeans'</span>, <span class="string">'fluxStds'</span>, <span class="string">'isoFluxs'</span>);
4594             <span class="keyword">end</span>
4595             
4596             <span class="comment">%% plot map</span>
4597             H = 600;
4598             W = 800;
4599             metRadius = 3;
4600             arrowRadius = 14;
4601             barW = 20;
4602             reactionLims = [0 2 4];
4603             strokes = [0.25 1.125 1];
4604             
4605             <span class="comment">%align</span>
4606             mapRxns.path{mapRxns.idxs == metabolism.reactionIndexs(<span class="string">'PfkA1'</span>)} = <span class="string">'M 230,145 L 230,178'</span>;
4607             mapRxns.path{mapRxns.idxs == metabolism.reactionIndexs(<span class="string">'Pgk1'</span>)} = <span class="string">'M 270,265 L 270,293'</span>;
4608             mapRxns.path{mapRxns.idxs == metabolism.reactionIndexs(<span class="string">'GpmA'</span>)} = <span class="string">'M 270,305 L 270,331'</span>;
4609             mapRxns.path{mapRxns.idxs == metabolism.reactionIndexs(<span class="string">'Pyk_ADP'</span>)} = <span class="string">'M 270,385 L 270,418'</span>;
4610             mapRxns.path{mapRxns.idxs == metabolism.reactionIndexs(<span class="string">'Adk1'</span>)} = <span class="string">'M 761,80 L 815,80'</span>;
4611             
4612             <span class="comment">%scale</span>
4613             minX = min(mapMets.x)-metRadius;
4614             minY = min(mapMets.y)+metRadius;
4615             scaleX = W / (range(mapMets.x) + 2*metRadius);
4616             scaleY = H / (range(mapMets.y) + 2*metRadius);
4617             mapMets.x = scaleX * (mapMets.x - minX);
4618             mapMets.y = scaleY * (mapMets.y - minY);
4619             <span class="keyword">for</span> i = 1:numel(mapRxns.path)
4620                 [starts, ends, tokens] = regexp(mapRxns.path{i}, <span class="string">'(-*\d*\.*\d*),(-*\d*\.*\d*)'</span>, <span class="string">'start'</span>, <span class="string">'end'</span>, <span class="string">'tokens'</span>);
4621                 <span class="keyword">for</span> j = 1:numel(starts)
4622                     valX = num2str(scaleX * (str2double(tokens{j}{1}) - minX));
4623                     valY = num2str(scaleY * (str2double(tokens{j}{2}) - minY));
4624                     mapRxns.path{i} = [<span class="keyword">...</span>
4625                         mapRxns.path{i}(1:starts(j)-1) <span class="keyword">...</span>
4626                         valX <span class="keyword">...</span>
4627                         <span class="string">','</span> <span class="keyword">...</span>
4628                         valY <span class="keyword">...</span>
4629                         mapRxns.path{i}(ends(j)+1:end) <span class="keyword">...</span>
4630                         ];
4631                     starts = starts + numel(valX) + numel(valY) - numel(tokens{j}{1}) - numel(tokens{j}{2});
4632                     ends = ends + numel(valX) + numel(valY) - numel(tokens{j}{1}) - numel(tokens{j}{2});
4633                 <span class="keyword">end</span>
4634             <span class="keyword">end</span>
4635             mapRxns.labelX = scaleX * (mapRxns.labelX - minX);
4636             mapRxns.labelY = scaleY * (mapRxns.labelY - minY);
4637             mapRxns.valueX = scaleX * (mapRxns.valueX - minX);
4638             mapRxns.valueY = scaleY * (mapRxns.valueY - minY);
4639             
4640             <span class="keyword">if</span> nargin &gt;= 5
4641                 axesHandle = subplot(<span class="string">'position'</span>, position, <span class="string">'parent'</span>, figHandle);
4642             <span class="keyword">else</span>
4643                 axesHandle = edu.stanford.covert.cell.sim.util.PlotUtil.newAxesHandle();
4644             <span class="keyword">end</span>
4645             hold(axesHandle, <span class="string">'on'</span>);
4646             set(axesHandle, <span class="string">'YDir'</span>, <span class="string">'reverse'</span>);
4647             xlim(axesHandle, [0 W+10]);
4648             ylim(axesHandle, [-8 H-10]);
4649             set(axesHandle, <span class="string">'visible'</span>, <span class="string">'off'</span>)
4650             
4651             <span class="comment">%metabolites</span>
4652             tfs = true(size(mapMets.x));
4653             tfs([59:63 14 39 85 53 67 18 69 35 40 33 44 37 70 83]) = false;
4654             plot(mapMets.x(tfs), mapMets.y(tfs), <span class="string">'o'</span>, <span class="string">'MarkerEdgeColor'</span>, <span class="string">'k'</span>, <span class="string">'MarkerFaceColor'</span>, 0.2 * [1 1 1], <span class="string">'MarkerSize'</span>, metRadius);
4655             <span class="keyword">if</span> showAllLabels
4656                 <span class="keyword">for</span> i = 1:numel(mapMets.x)
4657                     <span class="keyword">if</span> ~tfs(i)
4658                         <span class="keyword">continue</span>;
4659                     <span class="keyword">end</span>
4660                     text(mapMets.x(i)+7, mapMets.y(i)+7, met.wholeCellModelIDs{mapMets.idxs(i)}, <span class="keyword">...</span>
4661                         <span class="string">'FontSize'</span>, 5, <span class="string">'Color'</span>, <span class="string">'k'</span>, <span class="string">'Interpreter'</span>, <span class="string">'none'</span>);
4662                 <span class="keyword">end</span>
4663             <span class="keyword">end</span>
4664             
4665             <span class="comment">%reactions</span>
4666             relMean = log10(abs(fluxMeans));
4667             relMean(isinf(relMean)) = NaN;
4668             relMean = max(0, min(4, relMean)) / 4;
4669             strokeWidth = 0.25 * ones(size(relMean));
4670             color1 = 0.2 * [0 0 1] + 0.8 * (0.9 * [1 1 1]);
4671             color2 = [0 0 1];
4672             colors = <span class="keyword">...</span>
4673                 + repmat(color1, numel(relMean), 1) .* repmat(1-relMean, 1, 3) <span class="keyword">...</span>
4674                 + repmat(color2, numel(relMean), 1) .* repmat(relMean, 1, 3);
4675             
4676             [~, order] = sort(relMean);
4677             relMean = relMean(order, :);
4678             fluxMeans = fluxMeans(order, :);
4679             colors = colors(order, :);
4680             strokeWidth = strokeWidth(order, :);
4681             mapRxns.idxs = mapRxns.idxs(order);
4682             mapRxns.path = mapRxns.path(order);
4683             mapRxns.labelX = mapRxns.labelX(order);
4684             mapRxns.labelY = mapRxns.labelY(order);
4685             mapRxns.valueX = mapRxns.valueX(order);
4686             mapRxns.valueY = mapRxns.valueY(order);
4687             
4688             arrowRadii = arrowRadius * ones(size(mapRxns.idxs));
4689             arrowRadii(mapRxns.idxs == metabolism.reactionIndexs(<span class="string">'AspC1'</span>)) = 0.6 * arrowRadius;
4690             
4691             [~, hideFluxIdxs] = ismember({<span class="string">'GlpQ1'</span>, <span class="string">'GlpQ2'</span>, <span class="string">'GlpQ3'</span>, <span class="string">'GlpQ4'</span>, <span class="string">'GlpQ5'</span>, <span class="string">'DeoD'</span>, <span class="string">'Rpe'</span>, <span class="string">'RpiB'</span>}, mr.reactionWholeCellModelIDs);
4692             <span class="keyword">for</span> i = 1:numel(mapRxns.idxs)
4693                 <span class="keyword">if</span> isoFluxs(mapRxns.idxs(i)) || any(mapRxns.idxs(i) == hideFluxIdxs) || fluxMeans(i) == 0
4694                     <span class="keyword">continue</span>;
4695                 <span class="keyword">end</span>
4696                 
4697                 endX = 0;
4698                 endY = 0;
4699                 endAngle = NaN;
4700                 tokens = regexp(mapRxns.path{i}, <span class="string">'([A-Za-z])( (-*\d*\.*\d*),(-*\d*\.*\d*))+'</span>, <span class="string">'tokens'</span>);
4701                 <span class="keyword">for</span> j = 1:numel(tokens)
4702                     tokens2 = regexp(tokens{j}{2}, <span class="string">'(-*\d*\.*\d*),(-*\d*\.*\d*)'</span>, <span class="string">'tokens'</span>);
4703                     <span class="keyword">switch</span> tokens{j}{1}
4704                         <span class="keyword">case</span> <span class="string">'M'</span>
4705                             endX = str2double(tokens2{1}{1});
4706                             endY = str2double(tokens2{1}{2});
4707                         <span class="keyword">case</span> <span class="string">'L'</span>
4708                             startX = endX;
4709                             startY = endY;
4710                             endX2 = str2double(tokens2{1}{1});
4711                             endY2 = str2double(tokens2{1}{2});
4712                             line([endX endX2], [endY endY2], <span class="keyword">...</span>
4713                                 <span class="string">'Parent'</span>, axesHandle, <span class="keyword">...</span>
4714                                 <span class="string">'Color'</span>, colors(i, :) * (~showAllLabels), <span class="keyword">...</span>
4715                                 <span class="string">'lineWidth'</span>, strokeWidth(i));
4716                             startAngle = atan2(endY-endY2, endX-endX2);
4717                             endAngle = atan2(endY2 - endY, endX2 - endX);
4718                             endX = endX2;
4719                             endY = endY2;
4720                         <span class="keyword">case</span> <span class="string">'C'</span>
4721                             startX = endX;
4722                             startY = endY;
4723                             endX2 = str2double(tokens2{1}{1});
4724                             endY2 = str2double(tokens2{1}{2});
4725                             endX3 = str2double(tokens2{2}{1});
4726                             endY3 = str2double(tokens2{2}{2});
4727                             endX4 = str2double(tokens2{3}{1});
4728                             endY4 = str2double(tokens2{3}{2});
4729                             [h, qx, qy] = Funct_Bezier([endX endX2 endX3 endX4], [endY endY2 endY3 endY4], 50, axesHandle);
4730                             set(h, <span class="string">'color'</span>, colors(i, :) * (~showAllLabels), <span class="string">'lineWidth'</span>, strokeWidth(i));
4731                             startAngle = atan2(qy(1)-qy(2), qx(1)-qx(2));
4732                             endAngle = atan2(qy(end)-qy(end-1), qx(end)-qx(end-1));
4733                             endX = endX4;
4734                             endY = endY4;
4735                         <span class="keyword">otherwise</span>
4736                             throw(MException(<span class="string">'Figures34:error'</span>, <span class="string">'undefined command %s'</span>, tokens{j}{1}))
4737                     <span class="keyword">end</span>
4738                     <span class="keyword">if</span> j == 2 &amp;&amp; fluxMeans(i) &lt; 0
4739                         x = startX + 2*cos(startAngle) + [0  arrowRadii(i)*cos(startAngle+pi/3+pi/2)  arrowRadii(i)*cos(startAngle+pi*2/3+pi/2)];
4740                         y = startY + 2*sin(startAngle) + [0  arrowRadii(i)*sin(startAngle+pi/3+pi/2)  arrowRadii(i)*sin(startAngle+pi*2/3+pi/2)];
4741                         
4742                         arrowPtch = patch(x, y, 1, <span class="string">'Parent'</span>, axesHandle, <span class="string">'EdgeColor'</span>, <span class="string">'none'</span>);
4743                         set(arrowPtch, <span class="string">'FaceColor'</span>, colors(i, :) * (~showAllLabels), <span class="string">'FaceAlpha'</span>, 1, <span class="string">'EdgeAlpha'</span>, 1);
4744                     <span class="keyword">end</span>
4745                 <span class="keyword">end</span>
4746                 <span class="keyword">if</span> ~isnan(endAngle) &amp;&amp; fluxMeans(i) &gt; 0
4747                     x = endX + 2*cos(endAngle) + [0  arrowRadii(i)*cos(endAngle+pi/3+pi/2)  arrowRadii(i)*cos(endAngle+pi*2/3+pi/2)];
4748                     y = endY + 2*sin(endAngle) + [0  arrowRadii(i)*sin(endAngle+pi/3+pi/2)  arrowRadii(i)*sin(endAngle+pi*2/3+pi/2)];
4749                     
4750                     arrowPtch = patch(x, y, 1, <span class="string">'Parent'</span>, axesHandle, <span class="string">'EdgeColor'</span>, <span class="string">'none'</span>);
4751                     set(arrowPtch, <span class="string">'FaceColor'</span>, colors(i, :) * (~showAllLabels), <span class="string">'FaceAlpha'</span>, 1, <span class="string">'EdgeAlpha'</span>, 1);
4752                 <span class="keyword">end</span>
4753                 
4754                 <span class="keyword">if</span> showAllLabels
4755                     text(mapRxns.labelX(i), mapRxns.labelY(i), <span class="keyword">...</span>
4756                         metabolism.reactionWholeCellModelIDs(mapRxns.idxs(i)), <span class="keyword">...</span>
4757                         <span class="string">'FontSize'</span>, 5, <span class="string">'Interpreter'</span>, <span class="string">'none'</span>);
4758                 <span class="keyword">end</span>
4759             <span class="keyword">end</span>
4760             
4761             <span class="comment">%labels</span>
4762             labels = {
4763                 [90 90 210 210]            [60 400 400 60]          [0 1 0]  [105 335] <span class="string">'Glycolysis'</span>                 0.2  0
4764                 [230 230 370 370]          [55 230 230 55]          [1 0 0]  [205 190] {<span class="string">'Pentose'</span> <span class="string">'phosphate'</span>}      0.2  90
4765                 [265 265 425 425]          [230 340 340 230]        [0 0 1]  [375 310] {<span class="string">'ATP'</span> <span class="string">'synthesis'</span>}          0.2  0
4766                 [450 680 680 820 820 450]  [20 20 320 320 580 580]  [0 1 1]  [635 305] <span class="string">'Nucleotide metabolism'</span>     0.1  0
4767                 [170 390 390 170]          [420 420 810 810]        [1 0.5 0]  [280 470] {<span class="string">'Pyruvate'</span> <span class="string">'metabolism'</span>} 0.3  0
4768                 };
4769             h = zeros(size(labels, 1), 1);
4770             <span class="keyword">for</span> i = 1:size(labels, 1)
4771                 <span class="keyword">if</span> showAllLabels
4772                     ptch = patch(labels{i, 1}, labels{i, 2}, 1, <span class="string">'Parent'</span>, axesHandle, <span class="string">'EdgeColor'</span>, <span class="string">'none'</span>);
4773                     set(ptch, <span class="string">'FaceColor'</span>, labels{i,6}*labels{i, 3} + (1-labels{i,6})*[1 1 1], <span class="string">'FaceAlpha'</span>, 1, <span class="string">'EdgeAlpha'</span>, 1);
4774                     uistack(ptch, <span class="string">'bottom'</span>);
4775                     
4776                     h(i) = text(labels{i, 4}(1), labels{i, 4}(2), labels{i, 5}, <span class="keyword">...</span>
4777                         <span class="string">'HorizontalAlign'</span>, <span class="string">'center'</span>,  <span class="keyword">...</span>
4778                         <span class="string">'VerticalAlign'</span>, <span class="string">'middle'</span>,  <span class="keyword">...</span>
4779                         <span class="string">'FontSize'</span>, 5, <span class="keyword">...</span>
4780                         <span class="string">'Color'</span>, labels{i, 3});
4781                 <span class="keyword">else</span>
4782                     h(i) = text(labels{i, 4}(1), labels{i, 4}(2), labels{i, 5}, <span class="keyword">...</span>
4783                         <span class="string">'HorizontalAlign'</span>, <span class="string">'center'</span>,  <span class="keyword">...</span>
4784                         <span class="string">'VerticalAlign'</span>, <span class="string">'middle'</span>,  <span class="keyword">...</span>
4785                         <span class="string">'FontSize'</span>, 5, <span class="keyword">...</span>
4786                         <span class="string">'Color'</span>, <span class="string">'k'</span>, <span class="string">'rotation'</span>, labels{i, 7});
4787                 <span class="keyword">end</span>
4788             <span class="keyword">end</span>
4789             set(h(2:3), <span class="string">'HorizontalAlign'</span>, <span class="string">'left'</span>)
4790             
4791             <span class="comment">%colormap</span>
4792             <span class="keyword">if</span> ~showAllLabels
4793                 x1 = W-100;
4794                 x2 = x1 + 20;
4795                 h = 6.5;
4796                 y = 0 + h*40;
4797                 colorMap = <span class="keyword">...</span>
4798                     + repmat(color1, 40, 1) .* repmat(1 - linspace(0, 1, 40)', 1, 3) <span class="keyword">...</span>
4799                     + repmat(color2, 40, 1) .* repmat(    linspace(0, 1, 40)', 1, 3);
4800                 <span class="keyword">for</span> i = 1:40
4801                     ptch = patch([x1 x2 x2 x1], y - h*[i-1 i-1 i i], 1, <span class="string">'EdgeColor'</span>, <span class="string">'none'</span>, <span class="string">'Parent'</span>, axesHandle);
4802                     set(ptch, <span class="string">'FaceColor'</span>, colorMap(i, :), <span class="string">'FaceAlpha'</span>, 1, <span class="string">'EdgeAlpha'</span>, 1);
4803                 <span class="keyword">end</span>
4804                 patch([x1 x2 x2 x1], y - [0 0 40*h 40*h], 1, <span class="string">'EdgeColor'</span>, <span class="string">'k'</span>, <span class="string">'FaceColor'</span>, <span class="string">'none'</span>);
4805                 
4806                 line([x2 x2+5], y + [0 0], <span class="string">'Color'</span>, <span class="string">'k'</span>, <span class="string">'parent'</span>, axesHandle)
4807                 line([x2 x2+5], y-20*h + [0 0], <span class="string">'Color'</span>, <span class="string">'k'</span>, <span class="string">'parent'</span>, axesHandle)
4808                 line([x2 x2+5], y-40*h + [0 0], <span class="string">'Color'</span>, <span class="string">'k'</span>, <span class="string">'parent'</span>, axesHandle)
4809                 
4810                 text(x2+75, y-20*h, <span class="string">'Flux (rxn s^{-1})'</span>, <span class="keyword">...</span>
4811                     <span class="string">'Color'</span>, <span class="string">'k'</span>, <span class="string">'parent'</span>, axesHandle, <span class="string">'Interpreter'</span>, <span class="string">'tex'</span>, <span class="keyword">...</span>
4812                     <span class="string">'HorizontalAlign'</span>, <span class="string">'center'</span>, <span class="string">'VerticalAlign'</span>, <span class="string">'middle'</span>, <span class="string">'FontSize'</span>, 7, <span class="string">'rotation'</span>, 270);
4813                 text(x2+7, y, sprintf(<span class="string">'10^{%d}'</span>, 0), <span class="keyword">...</span>
4814                     <span class="string">'Color'</span>, <span class="string">'k'</span>, <span class="string">'parent'</span>, axesHandle, <span class="string">'Interpreter'</span>, <span class="string">'tex'</span>, <span class="keyword">...</span>
4815                     <span class="string">'HorizontalAlign'</span>, <span class="string">'left'</span>, <span class="string">'VerticalAlign'</span>, <span class="string">'middle'</span>, <span class="string">'FontSize'</span>, 5);
4816                 text(x2+7, y-20*h, sprintf(<span class="string">'10^{%d}'</span>, 2), <span class="keyword">...</span>
4817                     <span class="string">'Color'</span>, <span class="string">'k'</span>, <span class="string">'parent'</span>, axesHandle, <span class="string">'Interpreter'</span>, <span class="string">'tex'</span>, <span class="keyword">...</span>
4818                     <span class="string">'HorizontalAlign'</span>, <span class="string">'left'</span>, <span class="string">'VerticalAlign'</span>, <span class="string">'middle'</span>, <span class="string">'FontSize'</span>, 5);
4819                 text(x2+7, y-40*h, sprintf(<span class="string">'10^{%d}'</span>, 4), <span class="keyword">...</span>
4820                     <span class="string">'Color'</span>, <span class="string">'k'</span>, <span class="string">'parent'</span>, axesHandle, <span class="string">'Interpreter'</span>, <span class="string">'tex'</span>, <span class="keyword">...</span>
4821                     <span class="string">'HorizontalAlign'</span>, <span class="string">'left'</span>, <span class="string">'VerticalAlign'</span>, <span class="string">'middle'</span>, <span class="string">'FontSize'</span>, 5);
4822             <span class="keyword">end</span>
4823             
4824             <span class="comment">%% labels</span>
4825             <span class="keyword">if</span> ~showAllLabels
4826                 idxGpsA = find(mapRxns.idxs == sim.process(<span class="string">'Metabolism'</span>).reactionIndexs(<span class="string">'GpsA'</span>), 1, <span class="string">'first'</span>);
4827                 idxTalA = find(mapRxns.idxs == sim.process(<span class="string">'Metabolism'</span>).reactionIndexs(<span class="string">'TalA'</span>), 1, <span class="string">'first'</span>);
4828                 idxPgi = find(mapRxns.idxs == sim.process(<span class="string">'Metabolism'</span>).reactionIndexs(<span class="string">'Pgi'</span>), 1, <span class="string">'first'</span>);
4829                 
4830                 ptch = patch([335 335 450 450], [90 190 190 90], 1, <span class="string">'Parent'</span>, axesHandle, <span class="string">'EdgeColor'</span>, <span class="string">'none'</span>);
4831                 set(ptch, <span class="string">'FaceColor'</span>, 0.2*[1 0 0] + (1-0.2)*[1 1 1], <span class="string">'FaceAlpha'</span>, 1, <span class="string">'EdgeAlpha'</span>, 1);
4832                 uistack(ptch, <span class="string">'bottom'</span>);
4833                 
4834                 ptch = patch([0 0 107 107], [190 285 285 190], 1, <span class="string">'Parent'</span>, axesHandle, <span class="string">'EdgeColor'</span>, <span class="string">'none'</span>);
4835                 set(ptch, <span class="string">'FaceColor'</span>, 0.2*[1 0 0] + (1-0.2)*[1 1 1], <span class="string">'FaceAlpha'</span>, 1, <span class="string">'EdgeAlpha'</span>, 1);
4836                 uistack(ptch, <span class="string">'bottom'</span>);
4837                 
4838                 text(95, 220, <span class="string">'GpsA'</span>, <span class="keyword">...</span>
4839                     <span class="string">'FontSize'</span>, 7, <span class="keyword">...</span>
4840                     <span class="string">'HorizontalAlign'</span>, <span class="string">'right'</span>)
4841                 text(95, 260, sprintf(<span class="string">'(%.2f%%)'</span>, abs(fluxMeans(idxGpsA)/fluxMeans(idxPgi)*100)), <span class="keyword">...</span>
4842                     <span class="string">'FontSize'</span>, 5, <span class="keyword">...</span>
4843                     <span class="string">'HorizontalAlign'</span>, <span class="string">'right'</span>)
4844                 text(390, 120, <span class="string">'TalA'</span>, <span class="keyword">...</span>
4845                     <span class="string">'FontSize'</span>, 7, <span class="keyword">...</span>
4846                     <span class="string">'HorizontalAlign'</span>, <span class="string">'center'</span>)
4847                 text(390, 160, sprintf(<span class="string">'(%.2f%%)'</span>, abs(fluxMeans(idxTalA)/fluxMeans(idxPgi)*100)), <span class="keyword">...</span>
4848                     <span class="string">'FontSize'</span>, 5, <span class="keyword">...</span>
4849                     <span class="string">'HorizontalAlign'</span>, <span class="string">'center'</span>)
4850             <span class="keyword">end</span>
4851         <span class="keyword">end</span>
4852         
4853         <a name="_sub31" href="#_subfunctions" class="code">function chromosomeSampling(simBatchDir, selectedSim, outDirectory, figHandle, position)</a>
4854             import edu.stanford.covert.cell.sim.analysis.Figures34;
4855             import edu.stanford.covert.cell.sim.util.PlotUtil;
4856             import edu.stanford.covert.cell.sim.util.SimulationEnsemble;
4857             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
4858             
4859             <span class="keyword">if</span> ~exist(<span class="string">'simBatchDir'</span>, <span class="string">'var'</span>) || isempty(simBatchDir)
4860                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
4861             <span class="keyword">end</span>
4862             <span class="keyword">if</span> ~exist(<span class="string">'selectedSim'</span>, <span class="string">'var'</span>) || isempty(selectedSim)
4863                 selectedSim = 1;
4864             <span class="keyword">end</span>
4865             
4866             <span class="comment">%% get data</span>
4867             <span class="keyword">if</span> exist([outDirectory <span class="string">'chromosomeSampling.mat'</span>], <span class="string">'file'</span>)
4868                 load([outDirectory <span class="string">'chromosomeSampling.mat'</span>]);
4869             <span class="keyword">else</span>
4870                 [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep num2str(selectedSim)]);
4871                 ch = sim.state(<span class="string">'Chromosome'</span>);
4872                 pc = sim.state(<span class="string">'ProteinComplex'</span>);
4873                 rnaPol = sim.state(<span class="string">'RNAPolymerase'</span>);
4874                 
4875                 stateNames = {
4876                     <span class="string">'Time'</span>        <span class="string">'values'</span>             <span class="string">':'</span>  <span class="string">':'</span>
4877                     <span class="string">'Chromosome'</span>  <span class="string">'monomerBoundSites'</span>  <span class="string">':'</span>  <span class="string">':'</span>
4878                     <span class="string">'Chromosome'</span>  <span class="string">'complexBoundSites'</span>  <span class="string">':'</span>  <span class="string">':'</span>
4879                     <span class="string">'RNAPolymerase'</span> <span class="string">'positionStrands'</span>  <span class="string">':'</span>  <span class="string">':'</span>
4880                     <span class="string">'RNAPolymerase'</span> <span class="string">'states'</span>           <span class="string">':'</span>  <span class="string">':'</span>
4881                     };
4882                 states = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, <span class="string">'extract'</span>, selectedSim);
4883                 time = permute(states.Time.values, [3 1 2]) / 3600;
4884                 
4885                 <span class="comment">%monomers</span>
4886                 [subs, vals] = find(states.Chromosome.monomerBoundSites);
4887                 times = subs(:, 3);
4888                 startPos = subs(:, 1);
4889                 endPos = subs(:, 1) + ch.monomerDNAFootprints(vals) - 1;
4890                 idxs = find(endPos &gt; ch.sequenceLen);
4891                 vals = [vals; vals(idxs)];
4892                 times = [times; times(idxs)];
4893                 startPos = [startPos; ones(size(idxs))];
4894                 endPos = [endPos; endPos(idxs) - ch.sequenceLen];
4895                 endPos(idxs) = ch.sequenceLen;
4896                 monTimeBound = Figures34.calcChromosomeSampling(startPos, endPos, times, 1, ch.sequenceLen, ch.sequenceLen);
4897                 
4898                 <span class="comment">%complexes</span>
4899                 complexDNAFootprints = ch.complexDNAFootprints;
4900                 complexDNAFootprints(pc.dnaPolymeraseIndexs) = max(complexDNAFootprints(pc.dnaPolymeraseIndexs), sim.process(<span class="string">'Replication'</span>).dnaPolymeraseElongationRate);
4901                 complexDNAFootprints(pc.getIndexs(<span class="string">'MG_094_HEXAMER'</span>)) = max(complexDNAFootprints(pc.getIndexs(<span class="string">'MG_094_HEXAMER'</span>)), sim.process(<span class="string">'Replication'</span>).dnaPolymeraseElongationRate);
4902                 
4903                 [subs, vals] = find(states.Chromosome.complexBoundSites);
4904                 subs(ismember(vals, pc.rnaPolymeraseIndexs), :) = [];
4905                 vals(ismember(vals, pc.rnaPolymeraseIndexs), :) = [];
4906                 times = subs(:, 3);
4907                 startPos = subs(:, 1);
4908                 endPos = subs(:, 1) + complexDNAFootprints(vals) - 1;
4909                 
4910                 tmp = [reshape(states.RNAPolymerase.positionStrands(:, 1, :), [], 1)  <span class="keyword">...</span>
4911                     reshape(repmat(permute(1:size(states.RNAPolymerase.positionStrands, 3), [1 3 2]), [size(states.RNAPolymerase.positionStrands, 1), 1 1]), [], 1) <span class="keyword">...</span>
4912                     reshape(states.RNAPolymerase.states(:, 1, :), [], 1)];
4913                 tmp(tmp(:, 3) == rnaPol.freeValue | tmp(:, 3) == rnaPol.notExistValue, :) = [];
4914                 times = [times; tmp(:, 2)];
4915                 vals = [vals; pc.rnaPolymeraseIndexs(ones(size(tmp, 1), 1))];
4916                 startPos = [startPos; tmp(:, 1)];
4917                 endPos = [endPos;
4918                     tmp(:, 1) <span class="keyword">...</span>
4919                     + ch.complexDNAFootprints(pc.rnaPolymeraseIndexs(1)) * (tmp(:, 1) &lt; rnaPol.activelyTranscribingValue) <span class="keyword">...</span>
4920                     + sim.process(<span class="string">'Transcription'</span>).rnaPolymeraseElongationRate * (tmp(:, 1) &gt;= rnaPol.activelyTranscribingValue)
4921                     ];
4922                 
4923                 idxs = find(endPos &gt; ch.sequenceLen);
4924                 vals = [vals; vals(idxs)];
4925                 times = [times; times(idxs)];
4926                 startPos = [startPos; ones(size(idxs))];
4927                 endPos = [endPos; endPos(idxs) - ch.sequenceLen];
4928                 endPos(idxs) = ch.sequenceLen;
4929                 cpxTimeBound = Figures34.calcChromosomeSampling(startPos, endPos, times, 1, ch.sequenceLen, ch.sequenceLen);
4930                 
4931                 <span class="comment">%fraction sampled</span>
4932                 timeBound = min(monTimeBound, cpxTimeBound);
4933                 fractionSampled = cumsum(histc(timeBound, (1:size(states.Time.values, 3))')) / ch.sequenceLen;
4934                 
4935                 <span class="comment">%fraction sampled - individual proteins</span>
4936                 cpxIdxs = pc.getIndexs({<span class="string">'MG_213_214_298_6MER_ADP'</span>, <span class="string">'RNA_POLYMERASE'</span>, <span class="string">'MG_203_204_TETRAMER'</span>, <span class="string">'MG_094_HEXAMER'</span>});
4937                 fractionSampled_IndivProtein = zeros(size(states.Time.values, 3), numel(cpxIdxs));
4938                 <span class="keyword">for</span> i = 1:numel(cpxIdxs)
4939                     timeBound = Figures34.calcChromosomeSampling(startPos(vals == cpxIdxs(i)), endPos(vals == cpxIdxs(i)), times(vals == cpxIdxs(i)), 1, ch.sequenceLen, ch.sequenceLen);
4940                     fractionSampled_IndivProtein(:, i) = cumsum(histc(timeBound, (1:size(states.Time.values, 3))')) / ch.sequenceLen;
4941                 <span class="keyword">end</span>
4942                 
4943                 <span class="comment">%fraction sampled - RNA Pol</span>
4944                 nTime = size(states.RNAPolymerase.positionStrands, 3);
4945                 nPols = size(states.RNAPolymerase.positionStrands, 1);
4946                 rnaPolRate = sim.process(<span class="string">'Transcription'</span>).rnaPolymeraseElongationRate;
4947                 rnaPolFtpt = ch.complexDNAFootprints(pc.rnaPolymeraseIndexs(1));
4948                 fractionSampled_RNAPol = zeros(nTime, 3);
4949                 
4950                 tmpPosTime = [reshape(permute(states.RNAPolymerase.positionStrands(:, 1, :), [1 3 2]), [], 1)  reshape(repmat(1:nTime, nPols, 1), [], 1)];
4951                 tmpStates = reshape(permute(states.RNAPolymerase.states, [1 3 2]), [], 1);
4952                 
4953                 tfs = tmpStates &gt;= rnaPol.activelyTranscribingValue;
4954                 timeBound = Figures34.calcChromosomeSampling(tmpPosTime(tfs, 1), tmpPosTime(tfs, 1)+rnaPolRate, tmpPosTime(tfs, 2), 1, ch.sequenceLen, ch.sequenceLen);
4955                 fractionSampled_RNAPol(:, 1) = cumsum(histc(timeBound, (1:nTime)')) / ch.sequenceLen;
4956                 
4957                 tfs = false(ch.sequenceLen, 1);
4958                 <span class="keyword">for</span> j = 1:numel(rnaPol.transcripts.transcriptionUnitFivePrimeCoordinates)
4959                     <span class="keyword">if</span> rnaPol.transcripts.transcriptionUnitDirections(j) == 1
4960                         tfs(rnaPol.transcripts.transcriptionUnitFivePrimeCoordinates(j)+(1:rnaPol.transcripts.transcriptionUnitLengths(j))-1) = true;
4961                     <span class="keyword">else</span>
4962                         tfs(rnaPol.transcripts.transcriptionUnitFivePrimeCoordinates(j)-(1:rnaPol.transcripts.transcriptionUnitLengths(j))+1) = true;
4963                     <span class="keyword">end</span>
4964                 <span class="keyword">end</span>
4965                 timeBound = timeBound(tfs);
4966                 fractionSampled_RNAPol(:, 3) = cumsum(histc(timeBound, (1:nTime)')) / numel(timeBound);
4967                 
4968                 tfs = tmpStates == rnaPol.nonSpecificallyBoundValue | tmpStates == rnaPol.specificallyBoundValue;
4969                 timeBound = Figures34.calcChromosomeSampling(tmpPosTime(tfs, 1), tmpPosTime(tfs, 1)+rnaPolFtpt, tmpPosTime(tfs, 2), 1, ch.sequenceLen, ch.sequenceLen);
4970                 fractionSampled_RNAPol(:, 2) = cumsum(histc(timeBound, (1:nTime)')) / ch.sequenceLen;
4971                 
4972                 <span class="comment">%fit</span>
4973                 expFittype = fittype(<span class="string">'1 - a * exp(-log(2) * t / tau)'</span>, <span class="keyword">...</span>
4974                     <span class="string">'independent'</span>, <span class="string">'t'</span>, <span class="keyword">...</span>
4975                     <span class="string">'coefficients'</span>, {<span class="string">'a'</span>, <span class="string">'tau'</span>});
4976                 expFitOptions = fitoptions(<span class="keyword">...</span>
4977                     <span class="string">'Method'</span>, <span class="string">'NonlinearLeastSquares'</span>, <span class="keyword">...</span>
4978                     <span class="string">'Startpoint'</span>, [.25 2], <span class="keyword">...</span>
4979                     <span class="string">'Lower'</span>, [0 0], <span class="keyword">...</span>
4980                     <span class="string">'Upper'</span>, [1 10]);
4981                 
4982                 [fitResult, gof] = fit(time, fractionSampled, expFittype, expFitOptions);
4983                 
4984                 tau = zeros(size(fractionSampled_IndivProtein, 2), 1);
4985                 <span class="keyword">for</span> i = 1:size(fractionSampled_IndivProtein, 2)
4986                     tmp = fit(time, fractionSampled_IndivProtein(:, i), expFittype, expFitOptions);
4987                     tau(i) = tmp.tau;
4988                 <span class="keyword">end</span>
4989                 
4990                 tau_RNAPol = zeros(size(fractionSampled_RNAPol, 2), 1);
4991                 <span class="keyword">for</span> i = 1:size(fractionSampled_RNAPol, 2)
4992                     tmp = fit(time, fractionSampled_RNAPol(:, i), expFittype, expFitOptions);
4993                     tau_RNAPol(i) = tmp.tau;
4994                 <span class="keyword">end</span>
4995                 
4996                 save([outDirectory <span class="string">'chromosomeSampling.mat'</span>], <span class="string">'time'</span>, <span class="string">'fractionSampled'</span>, <span class="keyword">...</span>
4997                     <span class="string">'fractionSampled_IndivProtein'</span>, <span class="string">'fractionSampled_RNAPol'</span>, <span class="keyword">...</span>
4998                     <span class="string">'fitResult'</span>, <span class="string">'gof'</span>, <span class="string">'tau'</span>, <span class="string">'tau_RNAPol'</span>);
4999             <span class="keyword">end</span>
5000             
5001             <span class="comment">%% plot</span>
5002             <span class="keyword">if</span> nargin &gt;= 5
5003                 axesHandle = subplot(<span class="string">'position'</span>, position, <span class="string">'parent'</span>, figHandle);
5004             <span class="keyword">else</span>
5005                 axesHandle = edu.stanford.covert.cell.sim.util.PlotUtil.newAxesHandle();
5006             <span class="keyword">end</span>
5007             
5008             labels = {<span class="string">'All'</span>, <span class="string">'DNA pol'</span>, <span class="string">'Gyrase'</span>, <span class="string">'RNA pol'</span>, <span class="string">'SMC'</span>};
5009             colors = [
5010                 0 1 1
5011                 0 1 0
5012                 1 0 0
5013                 0 0 1
5014                 1 0 1
5015                 ];
5016             data = [fractionSampled  fractionSampled_IndivProtein(:, end:-1:1)];
5017             tau = [fitResult.tau; tau(end:-1:1, 1)] * 60 * log(2);
5018             
5019             show = true(size(labels));
5020             <span class="comment">%show([3 5]) = false;</span>
5021             labels = labels(show);
5022             colors = colors(show, :);
5023             data = data(:, show);
5024             tau = tau(show, :);
5025             
5026             h = plot(axesHandle, time, 100 * data, <span class="string">'LineWidth'</span>, 1);
5027             
5028             [~, order] = sort(tau);
5029             labels = labels(order);
5030             tau = tau(order);
5031             h = h(order);
5032             colors = colors(order, :);
5033             data = data(:, order);
5034             
5035             tableFont = 6;
5036             rowSep = 12;
5037             text(7.4, 88, <span class="string">'Protein'</span>, <span class="string">'FontSize'</span>, tableFont, <span class="string">'HorizontalAlign'</span>, <span class="string">'left'</span>, <span class="string">'VerticalAlign'</span>, <span class="string">'middle'</span>, <span class="string">'Parent'</span>, axesHandle);
5038             line(7.4 + [0 1.4], 82 + [0 0], <span class="string">'Parent'</span>, axesHandle, <span class="string">'Color'</span>, <span class="string">'k'</span>)
5039             <span class="keyword">for</span> i = 1:numel(labels)
5040                 set(h(i), <span class="string">'Color'</span>, colors(i, :));
5041                 line(6.9 + [0 0.3], 87-i*rowSep + [0 0], <span class="string">'Parent'</span>, axesHandle, <span class="string">'Color'</span>, colors(i, :), <span class="string">'LineWidth'</span>, 2)
5042                 text(7.4, 87-i*rowSep, labels{i}, <span class="string">'FontSize'</span>, tableFont, <span class="string">'HorizontalAlign'</span>, <span class="string">'left'</span>, <span class="string">'VerticalAlign'</span>, <span class="string">'middle'</span>, <span class="string">'Parent'</span>, axesHandle);
5043             <span class="keyword">end</span>
5044             
5045             xlim(axesHandle, [0 time(end)]);
5046             ylim(axesHandle, [0 100]);
5047             
5048             set(axesHandle, <span class="string">'FontSize'</span>, 5, <span class="string">'TickDir'</span>, <span class="string">'out'</span>, <span class="string">'XTick'</span>, 0:2:8, <span class="string">'YTick'</span>, 0:25:100, <span class="string">'box'</span>, <span class="string">'off'</span>);
5049             
5050             tick2text(axesHandle, <span class="string">'axis'</span>, <span class="string">'xy'</span>, <span class="string">'xtickoffset'</span>, 0.08, <span class="string">'ytickoffset'</span>, 0.025);
5051             xTicks = getappdata(axesHandle, <span class="string">'XTickText'</span>);
5052             yTicks = getappdata(axesHandle, <span class="string">'YTickText'</span>);
5053             set(xTicks, <span class="string">'FontSize'</span>, 5);
5054             set(yTicks, <span class="string">'FontSize'</span>, 5, <span class="string">'HorizontalAlign'</span>, <span class="string">'right'</span>);
5055             delete(xTicks(2:4));
5056             delete(yTicks(2:4));
5057             
5058             xlabel(axesHandle, <span class="string">'Time (h)'</span>, <span class="string">'FontSize'</span>, 7);
5059             ylabel(axesHandle, {<span class="string">'% Chromosome'</span> <span class="string">'explored'</span>}, <span class="string">'FontSize'</span>, 7);
5060             xlabelPos = get(get(axesHandle, <span class="string">'xlabel'</span>), <span class="string">'position'</span>);
5061             ylabelPos = get(get(axesHandle, <span class="string">'ylabel'</span>), <span class="string">'position'</span>);
5062             set(get(axesHandle, <span class="string">'xlabel'</span>), <span class="string">'position'</span>, [xlabelPos(1) -11 xlabelPos(3:end)]);
5063             set(get(axesHandle, <span class="string">'ylabel'</span>), <span class="string">'position'</span>, [-0.55 ylabelPos(2:end)]);
5064         <span class="keyword">end</span>
5065         
5066         <a name="_sub32" href="#_subfunctions" class="code">function plotRNAPolSampling(simBatchDir, selectedSim, outDirectory, figHandle, position)</a>
5067             import edu.stanford.covert.cell.sim.util.PlotUtil;
5068             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
5069             
5070             <span class="comment">%% lood constants</span>
5071             [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep num2str(selectedSim)]);
5072             ch = sim.state(<span class="string">'Chromosome'</span>);
5073             pc = sim.state(<span class="string">'ProteinComplex'</span>);
5074             rnaPol = sim.state(<span class="string">'RNAPolymerase'</span>);
5075             
5076             <span class="comment">%% load data</span>
5077             load([outDirectory <span class="string">'chromosomeSampling.mat'</span>]);
5078             load([outDirectory <span class="string">'rnaSampling.mat'</span>]);
5079             
5080             <span class="comment">%% plot</span>
5081             <span class="keyword">if</span> nargin &gt;= 5
5082                 axesHandle = subplot(<span class="string">'position'</span>, position, <span class="string">'parent'</span>, figHandle);
5083             <span class="keyword">else</span>
5084                 axesHandle = edu.stanford.covert.cell.sim.util.PlotUtil.newAxesHandle();
5085             <span class="keyword">end</span>
5086             
5087             h = plot(axesHandle, time, 100 * [fractionSampled_IndivProtein(:, 2) fractionSampled_RNAPol(:, [1 3 2]) fracRNASampled]); <span class="comment">%#ok&lt;NODEF&gt;</span>
5088             legend(h, {
5089                 sprintf(<span class="string">'All RNA Polymerase (\\tau = %0.1f h)'</span>, tau(2))
5090                 sprintf(<span class="string">'Active/SB - of whole Chromosome (\\tau = %0.1f h)'</span>, tau_RNAPol(1))
5091                 sprintf(<span class="string">'Active/SB - of TUs (\\tau = %0.1f h)'</span>, tau_RNAPol(3))
5092                 sprintf(<span class="string">'Non-SB (\\tau = %0.1f h)'</span>, tau_RNAPol(2))
5093                 sprintf(<span class="string">'Expressed RNA (\\tau = %0.1f h)'</span>, rnaFitResult.tau)
5094                 }, <span class="keyword">...</span>
5095                 <span class="string">'Location'</span>, <span class="string">'SouthEast'</span>, <span class="string">'FontSize'</span>, 8);
5096             xlabel(axesHandle, <span class="string">'Time (h)'</span>)
5097             ylabel(axesHandle, <span class="string">'% Explored / expressed'</span>)
5098             xlim(axesHandle, [0 max(time)])
5099             ylim(axesHandle, [0 100])
5100         <span class="keyword">end</span>
5101         
5102         <a name="_sub33" href="#_subfunctions" class="code">function timeBound = calcChromosomeSampling(startPos, endPos, times, pos1, pos2, L)</a>
5103             import edu.stanford.covert.cell.sim.analysis.Figures34;
5104             
5105             <span class="keyword">if</span> pos2-pos1 &gt; 500
5106                 pos21 = floor((pos1+pos2)/2);
5107                 pos12 = pos21+1;
5108                 tfs1 = startPos &lt;= pos21;
5109                 tfs2 = endPos &gt;= pos12;
5110                 timeBound = min(<span class="keyword">...</span>
5111                     Figures34.calcChromosomeSampling(startPos(tfs1), endPos(tfs1), times(tfs1), pos1, pos21, L), <span class="keyword">...</span>
5112                     Figures34.calcChromosomeSampling(startPos(tfs2), endPos(tfs2), times(tfs2), pos12, pos2, L));
5113             <span class="keyword">else</span>
5114                 timeBound = NaN(L, 1);
5115                 <span class="keyword">for</span> i = pos1:pos2
5116                     tf = startPos &lt;= i &amp; endPos &gt;= i;
5117                     <span class="keyword">if</span> ~any(tf)
5118                         <span class="keyword">continue</span>;
5119                     <span class="keyword">end</span>
5120                     timeBound(i) = min(times(tf));
5121                 <span class="keyword">end</span>
5122             <span class="keyword">end</span>
5123         <span class="keyword">end</span>
5124         
5125         <a name="_sub34" href="#_subfunctions" class="code">function rnaSampling(simBatchDir, selectedSim, outDirectory, figHandle, position)</a>
5126             import edu.stanford.covert.cell.sim.util.PlotUtil;
5127             import edu.stanford.covert.cell.sim.util.SimulationEnsemble;
5128             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
5129             
5130             <span class="keyword">if</span> ~exist(<span class="string">'simBatchDir'</span>, <span class="string">'var'</span>) || isempty(simBatchDir)
5131                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
5132             <span class="keyword">end</span>
5133             <span class="keyword">if</span> ~exist(<span class="string">'selectedSim'</span>, <span class="string">'var'</span>) || isempty(selectedSim)
5134                 selectedSim = 1;
5135             <span class="keyword">end</span>
5136             
5137             <span class="comment">%% data</span>
5138             <span class="keyword">if</span> exist([outDirectory <span class="string">'rnaSampling.mat'</span>], <span class="string">'file'</span>)
5139                 load([outDirectory <span class="string">'rnaSampling.mat'</span>]);
5140             <span class="keyword">else</span>
5141                 [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep num2str(selectedSim)]);
5142                 comp = sim.compartment;
5143                 rna = sim.state(<span class="string">'Rna'</span>);
5144                 
5145                 stateNames = {
5146                     <span class="string">'Time'</span>       <span class="string">'values'</span>        <span class="string">':'</span>  <span class="string">':'</span>
5147                     <span class="string">'Rna'</span>        <span class="string">'counts'</span>        rna.matureIndexs  comp.cytosolIndexs
5148                     };
5149                 states = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, <span class="string">'extract'</span>, selectedSim);
5150                 time = permute(states.Time.values, [3 1 2]) / 3600;
5151                 
5152                 [i, j] = find(permute(states.Rna.counts, [1 3 2]));
5153                 timeFirstExpressed = NaN(size(rna.matureIndexs));
5154                 timeFirstExpressed(i(end:-1:1)) = j(end:-1:1);
5155                 fracRNASampled = cumsum(histc(timeFirstExpressed, (1:numel(time)))) / numel(rna.matureIndexs);
5156                 
5157                 expFittype = fittype(<span class="string">'1 - a * exp(-log(2) * t / tau)'</span>, <span class="keyword">...</span>
5158                     <span class="string">'independent'</span>, <span class="string">'t'</span>, <span class="keyword">...</span>
5159                     <span class="string">'coefficients'</span>, {<span class="string">'a'</span>, <span class="string">'tau'</span>});
5160                 expFitOptions = fitoptions(<span class="keyword">...</span>
5161                     <span class="string">'Method'</span>, <span class="string">'NonlinearLeastSquares'</span>, <span class="keyword">...</span>
5162                     <span class="string">'Startpoint'</span>, [.25 2], <span class="keyword">...</span>
5163                     <span class="string">'Lower'</span>, [0 0], <span class="keyword">...</span>
5164                     <span class="string">'Upper'</span>, [1 10]);
5165                 [rnaFitResult, rnaGof] = fit(time, fracRNASampled, expFittype, expFitOptions);
5166                 
5167                 save([outDirectory <span class="string">'rnaSampling.mat'</span>], <span class="string">'time'</span>, <span class="string">'fracRNASampled'</span>, <span class="string">'rnaFitResult'</span>, <span class="string">'rnaGof'</span>)
5168             <span class="keyword">end</span>
5169             
5170             <span class="comment">%% plot</span>
5171             <span class="keyword">if</span> nargin &gt;= 5
5172                 axesHandle = subplot(<span class="string">'position'</span>, position, <span class="string">'parent'</span>, figHandle);
5173             <span class="keyword">else</span>
5174                 axesHandle = edu.stanford.covert.cell.sim.util.PlotUtil.newAxesHandle();
5175             <span class="keyword">end</span>
5176             
5177             plot(axesHandle, time, 100 * fracRNASampled, <span class="string">'Color'</span>, <span class="string">'b'</span>, <span class="string">'LineWidth'</span>, 1);
5178             
5179             xlim(axesHandle, [0 time(end)]);
5180             ylim(axesHandle, [0 100]);
5181             
5182             set(axesHandle, <span class="string">'FontSize'</span>, 5, <span class="string">'TickDir'</span>, <span class="string">'out'</span>, <span class="string">'XTick'</span>, 0:2:8, <span class="string">'YTick'</span>, 0:25:100, <span class="string">'box'</span>, <span class="string">'off'</span>);
5183             
5184             t_50 = min(time(fracRNASampled &gt;= 0.50)) * 60;
5185             text(8.84, 98, sprintf(<span class="string">'t_{50} = %.0f min'</span>, t_50), <span class="keyword">...</span>
5186                 <span class="string">'Parent'</span>, axesHandle, <span class="keyword">...</span>
5187                 <span class="string">'FontSize'</span>, 6, <span class="keyword">...</span>
5188                 <span class="string">'horizontalalign'</span>, <span class="string">'right'</span>, <span class="keyword">...</span>
5189                 <span class="string">'verticalalign'</span>, <span class="string">'top'</span>);
5190             
5191             tick2text(axesHandle, <span class="string">'axis'</span>, <span class="string">'xy'</span>, <span class="string">'xtickoffset'</span>, 0.08, <span class="string">'ytickoffset'</span>, 0.025);
5192             xTicks = getappdata(axesHandle, <span class="string">'XTickText'</span>);
5193             yTicks = getappdata(axesHandle, <span class="string">'YTickText'</span>);
5194             set(xTicks, <span class="string">'FontSize'</span>, 5);
5195             set(yTicks, <span class="string">'FontSize'</span>, 5, <span class="string">'HorizontalAlign'</span>, <span class="string">'right'</span>);
5196             delete(xTicks(2:4));
5197             delete(yTicks(2:4));
5198             
5199             xlabel(axesHandle, <span class="string">'Time (h)'</span>, <span class="string">'FontSize'</span>, 7);
5200             ylabel(axesHandle, {<span class="string">'% RNA'</span> <span class="string">'expressed'</span>}, <span class="string">'FontSize'</span>, 7);
5201             xlabelPos = get(get(axesHandle, <span class="string">'xlabel'</span>), <span class="string">'position'</span>);
5202             ylabelPos = get(get(axesHandle, <span class="string">'ylabel'</span>), <span class="string">'position'</span>);
5203             set(get(axesHandle, <span class="string">'xlabel'</span>), <span class="string">'position'</span>, [xlabelPos(1) -11 xlabelPos(3:end)]);
5204             set(get(axesHandle, <span class="string">'ylabel'</span>), <span class="string">'position'</span>, [-0.55 ylabelPos(2:end)]);
5205         <span class="keyword">end</span>
5206         
5207         <a name="_sub35" href="#_subfunctions" class="code">function mrnaSampling(simBatchDir, selectedSim, outDirectory, figHandle, position)</a>
5208             import edu.stanford.covert.cell.sim.util.PlotUtil;
5209             import edu.stanford.covert.cell.sim.util.SimulationEnsemble;
5210             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
5211             
5212             <span class="keyword">if</span> ~exist(<span class="string">'simBatchDir'</span>, <span class="string">'var'</span>) || isempty(simBatchDir)
5213                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
5214             <span class="keyword">end</span>
5215             <span class="keyword">if</span> ~exist(<span class="string">'selectedSim'</span>, <span class="string">'var'</span>) || isempty(selectedSim)
5216                 selectedSim = 1;
5217             <span class="keyword">end</span>
5218             
5219             <span class="comment">%% data</span>
5220             <span class="keyword">if</span> exist([outDirectory <span class="string">'mrnaSampling.mat'</span>], <span class="string">'file'</span>)
5221                 load([outDirectory <span class="string">'mrnaSampling.mat'</span>]);
5222             <span class="keyword">else</span>
5223                 [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep num2str(selectedSim)]);
5224                 comp = sim.compartment;
5225                 rna = sim.state(<span class="string">'Rna'</span>);
5226                 
5227                 stateNames = {
5228                     <span class="string">'Time'</span>       <span class="string">'values'</span>        <span class="string">':'</span>  <span class="string">':'</span>
5229                     <span class="string">'Rna'</span>        <span class="string">'counts'</span>        rna.matureIndexs  comp.cytosolIndexs
5230                     };
5231                 states = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, <span class="string">'extract'</span>, selectedSim);
5232                 time = permute(states.Time.values, [3 1 2]) / 3600;
5233                 
5234                 [i, j] = find(permute(states.Rna.counts, [1 3 2]));
5235                 timeFirstExpressed = NaN(size(rna.matureIndexs));
5236                 timeFirstExpressed(i(end:-1:1)) = j(end:-1:1);
5237                 fracMRNASampled = cumsum(histc(timeFirstExpressed(rna.matureMRNAIndexs), (1:numel(time)))) / numel(rna.matureMRNAIndexs);
5238                 
5239                 expFittype = fittype(<span class="string">'1 - a * exp(-log(2) * t / tau)'</span>, <span class="keyword">...</span>
5240                     <span class="string">'independent'</span>, <span class="string">'t'</span>, <span class="keyword">...</span>
5241                     <span class="string">'coefficients'</span>, {<span class="string">'a'</span>, <span class="string">'tau'</span>});
5242                 expFitOptions = fitoptions(<span class="keyword">...</span>
5243                     <span class="string">'Method'</span>, <span class="string">'NonlinearLeastSquares'</span>, <span class="keyword">...</span>
5244                     <span class="string">'Startpoint'</span>, [.25 2], <span class="keyword">...</span>
5245                     <span class="string">'Lower'</span>, [0 0], <span class="keyword">...</span>
5246                     <span class="string">'Upper'</span>, [1 10]);
5247                 [mrnaFitResult, mrnaGof] = fit(time, fracMRNASampled, expFittype, expFitOptions);
5248                 
5249                 save([outDirectory <span class="string">'mrnaSampling.mat'</span>], <span class="string">'time'</span>, <span class="string">'fracMRNASampled'</span>, <span class="string">'mrnaFitResult'</span>, <span class="string">'mrnaGof'</span>)
5250             <span class="keyword">end</span>
5251             
5252             <span class="comment">%% plot mRNA</span>
5253             <span class="keyword">if</span> nargin &gt;= 5
5254                 axesHandle = subplot(<span class="string">'position'</span>, position, <span class="string">'parent'</span>, figHandle);
5255             <span class="keyword">else</span>
5256                 axesHandle = edu.stanford.covert.cell.sim.util.PlotUtil.newAxesHandle();
5257             <span class="keyword">end</span>
5258             
5259             h = plot(axesHandle, time, 100 * [fracMRNASampled   1-mrnaFitResult.a*exp(-log(2)*time/mrnaFitResult.tau)]);
5260             set(h(1), <span class="string">'Color'</span>, <span class="string">'b'</span>);
5261             set(h(2), <span class="string">'Color'</span>, <span class="string">'k'</span>);
5262             uistack(h(1), <span class="string">'top'</span>);
5263             
5264             xlim(axesHandle, time([1 end]));
5265             ylim(axesHandle, [0 100]);
5266             
5267             xlabel(axesHandle, <span class="string">'Time (h)'</span>, <span class="string">'FontSize'</span>, 7);
5268             ylabel(axesHandle, <span class="string">'% mRNA expressed'</span>, <span class="string">'FontSize'</span>, 7);
5269             set(axesHandle, <span class="string">'FontSize'</span>, 5, <span class="string">'TickDir'</span>, <span class="string">'out'</span>, <span class="string">'XTick'</span>, [0 5], <span class="string">'YTick'</span>, [0 50 100], <span class="string">'box'</span>, <span class="string">'off'</span>);
5270             
5271             text(time(end), 98.5, sprintf(<span class="string">'\\gamma = %.0f min'</span>, mrnaFitResult.tau * 60 * log(2)), <span class="keyword">...</span>
5272                 <span class="string">'Parent'</span>, axesHandle, <span class="keyword">...</span>
5273                 <span class="string">'FontSize'</span>, 7, <span class="keyword">...</span>
5274                 <span class="string">'horizontalalign'</span>, <span class="string">'right'</span>, <span class="keyword">...</span>
5275                 <span class="string">'verticalalign'</span>, <span class="string">'top'</span>);
5276             text(time(end), 93.5, sprintf(<span class="string">'R^2_{adj} = %.2f'</span>, mrnaGof.adjrsquare), <span class="keyword">...</span>
5277                 <span class="string">'Parent'</span>, axesHandle, <span class="keyword">...</span>
5278                 <span class="string">'FontSize'</span>, 7, <span class="keyword">...</span>
5279                 <span class="string">'horizontalalign'</span>, <span class="string">'right'</span>, <span class="keyword">...</span>
5280                 <span class="string">'verticalalign'</span>, <span class="string">'top'</span>);
5281         <span class="keyword">end</span>
5282     <span class="keyword">end</span>
5283     
5284     methods (Static = true);
5285         <a name="_sub36" href="#_subfunctions" class="code">function rgbStr = colorFunc(normVal)</a>
5286             r = 0; g = 0; b = 0;
5287             
5288             <span class="keyword">if</span> normVal &lt;= 0.5
5289                 b = 255 * (1 - 2 * normVal);
5290                 g = 255 * (2 * normVal);
5291             <span class="keyword">else</span>
5292                 r = 255 * (2 * normVal - 1);
5293                 g = 255 * (2 - 2 * normVal);
5294             <span class="keyword">end</span>
5295             
5296             rgbStr = sprintf(<span class="string">'rgb(%d,%d,%d)'</span>, uint8(r), uint8(g), uint8(b));
5297         <span class="keyword">end</span>
5298         
5299         <a name="_sub37" href="#_subfunctions" class="code">function svg = drawCell(scale, width, cylindricalLength, septumLength, maxTotalLength, maxWidth) </a><span class="comment">%#ok&lt;INUSD&gt;</span>
5300             width = width / scale;
5301             cylindricalLength = cylindricalLength / scale;
5302             septumLength = septumLength / scale;
5303             H = maxTotalLength / scale;
5304             W = width;
5305             
5306             svg = [<span class="keyword">...</span>
5307                 sprintf(<span class="string">'  &lt;path d=&quot;\n'</span>) <span class="keyword">...</span>
5308                 sprintf(<span class="string">'    M%0.4f,%0.4f a%0.4f,%0.4f 0 1 0 0,%0.4f\n'</span>, W/2+5, 5, width/2, width/2, width), <span class="keyword">...</span>
5309                 sprintf(<span class="string">'    l%0.4f,0\n'</span>, cylindricalLength/2) <span class="keyword">...</span>
5310                 sprintf(<span class="string">'    a%0.4f,%0.4f 0 0 0 %0.4f,-%0.4f\n'</span>, septumLength, septumLength, septumLength, septumLength) <span class="keyword">...</span>
5311                 sprintf(<span class="string">'    a%0.4f,%0.4f 0 0 0 %0.4f,%0.4f\n'</span>, septumLength, septumLength, septumLength, septumLength) <span class="keyword">...</span>
5312                 sprintf(<span class="string">'    l%0.4f,0\n'</span>, cylindricalLength/2) <span class="keyword">...</span>
5313                 sprintf(<span class="string">'    a%0.4f,%0.4f 0 1 0 0,-%0.4f\n'</span>, width/2,width/2, width) <span class="keyword">...</span>
5314                 sprintf(<span class="string">'    l-%0.4f,0\n'</span>, cylindricalLength/2) <span class="keyword">...</span>
5315                 sprintf(<span class="string">'    a%0.4f,%0.4f 0 0 0 -%0.4f,%0.4f\n'</span>, septumLength, septumLength, septumLength, septumLength) <span class="keyword">...</span>
5316                 sprintf(<span class="string">'    a%0.4f,%0.4f 0 0 0 -%0.4f,-%0.4f\n'</span>, septumLength, septumLength, septumLength, septumLength) <span class="keyword">...</span>
5317                 sprintf(<span class="string">'    z\n'</span>) <span class="keyword">...</span>
5318                 <span class="string">'    &quot; style=&quot;stroke:darkblue;stroke-width:2;stroke-linecap:round;stroke-linejoin:miter;fill:url(#cellBg)&quot;/&gt;\n'</span>];
5319         <span class="keyword">end</span>
5320     <span class="keyword">end</span>
5321 <span class="keyword">end</span></pre></div>
<hr><address>Generated on Thu 19-Jul-2012 18:32:13 by <strong><a href="http://www.artefact.tk/software/matlab/m2html/" title="Matlab Documentation in HTML">m2html</a></strong> &copy; 2005</address>
</body>
</html>