<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN"
                "http://www.w3.org/TR/REC-html40/loose.dtd">
<html>
<head>
  <title>Description of Population</title>
  <meta name="keywords" content="Population">
  <meta name="description" content="Population">
  <meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1">
  <meta name="generator" content="m2html v1.5 &copy; 2003-2005 Guillaume Flandin">
  <meta name="robots" content="index, follow">
  <link type="text/css" rel="stylesheet" href="../../../../../../../../m2html.css">
</head>
<body>
<a name="_top"></a>
<div><a href="../../../../../../../../index.html">Home</a> &gt;  <a href="../../../../../../../index.html">simulation</a> &gt; <a href="#">src</a> &gt; <a href="#">+edu</a> &gt; <a href="#">+stanford</a> &gt; <a href="#">+covert</a> &gt; <a href="#">+cell</a> &gt; <a href="../index.html">+sim</a> &gt; <a href="index.html">+analysis</a> &gt; Population.m</div>

<!--<table width="100%"><tr><td align="left"><a href="../../../../../../../../index.html"><img alt="<" border="0" src="../../../../../../../../left.png">&nbsp;Master index</a></td>
<td align="right"><a href="index.html">Index for simulation/src/+edu/+stanford/+covert/+cell/+sim/+analysis&nbsp;<img alt=">" border="0" src="../../../../../../../../right.png"></a></td></tr></table>-->

<h1>Population
</h1>

<h2><a name="_name"></a>PURPOSE <a href="#_top"><img alt="^" border="0" src="../../../../../../../../up.png"></a></h2>
<div class="box"><strong>Population</strong></div>

<h2><a name="_synopsis"></a>SYNOPSIS <a href="#_top"><img alt="^" border="0" src="../../../../../../../../up.png"></a></h2>
<div class="box"><strong>This is a script file. </strong></div>

<h2><a name="_description"></a>DESCRIPTION <a href="#_top"><img alt="^" border="0" src="../../../../../../../../up.png"></a></h2>
<div class="fragment"><pre class="comment">Population

 Author: Jonathan Karr, jkarr@stanford.edu
 Author: Derek Macklin, macklin@stanford.edu
 Affiliation: Covert Lab, Department of Bioengineering, Stanford University
 Last Updated: 7/6/2011</pre></div>

<!-- crossreference -->
<h2><a name="_cross"></a>CROSS-REFERENCE INFORMATION <a href="#_top"><img alt="^" border="0" src="../../../../../../../../up.png"></a></h2>
This function calls:
<ul style="list-style-image:url(../../../../../../../../matlabicon.gif)">
<li><a href="Population.html" class="code" title="">Population</a>	Population</li></ul>
This function is called by:
<ul style="list-style-image:url(../../../../../../../../matlabicon.gif)">
<li><a href="Population.html" class="code" title="">Population</a>	Population</li></ul>
<!-- crossreference -->

<h2><a name="_subfunctions"></a>SUBFUNCTIONS <a href="#_top"><img alt="^" border="0" src="../../../../../../../../up.png"></a></h2>
<ul style="list-style-image:url(../../../../../../../../matlabicon.gif)">
<li><a href="#_sub1" class="code">function run(simBatchDir, fileName, selectedSimulations, methodsToRun)</a></li><li><a href="#_sub2" class="code">function figData = growth(figHandle, simBatchDir, selectedSimulations)</a></li><li><a href="#_sub3" class="code">function figData = energyProduction(figHandle, simBatchDir, selectedSimulations)</a></li><li><a href="#_sub4" class="code">function figData = cellMassDistribution(figHandle, simBatchDir, selectedSimulations)</a></li><li><a href="#_sub5" class="code">function figData = massDistribution(figHandle, simBatchDir, selectedSimulations)</a></li><li><a href="#_sub6" class="code">function figData = cellCyclePhases(figHandle, simBatchDir, selectedSimulations)</a></li><li><a href="#_sub7" class="code">function [figData, tabContent, tabColLabels] = nucleotideDistribution(figHandle, simBatchDir, selectedSimulations)</a></li><li><a href="#_sub8" class="code">function figData = aminoAcidCounts(figHandle, simBatchDir, selectedSimulations)</a></li><li><a href="#_sub9" class="code">function [figData, tabContent, tabColLabels] = metabolism(figHandle, simBatchDir, selectedSimulations)</a></li><li><a href="#_sub10" class="code">function figData = metaboliteConcentrations(figHandle, simBatchDir, selectedSimulations)</a></li><li><a href="#_sub11" class="code">function figData = translation(figHandle, simBatchDir, selectedSimulations)</a></li><li><a href="#_sub12" class="code">function figData = proteinSynthesis(figHandle, simBatchDir, selectedSimulations)</a></li><li><a href="#_sub13" class="code">function figData = rnaSynthesis(figHandle, simBatchDir, selectedSimulations)</a></li><li><a href="#_sub14" class="code">function figData = cellShape(figHandle, simBatchDir, selectedSimulations)</a></li><li><a href="#_sub15" class="code">function figData = rnaPolymerases(figHandle, simBatchDir, selectedSimulations)</a></li><li><a href="#_sub16" class="code">function figData = ribosomes(figHandle, simBatchDir, selectedSimulations)</a></li><li><a href="#_sub17" class="code">function figData = macromoleculeExpression(figHandle, simBatchDir, selectedSimulations)</a></li><li><a href="#_sub18" class="code">function figData = dnaBoundProteinDisplacement(figHandle, simBatchDir, selectedSimulations)</a></li><li><a href="#_sub19" class="code">function figData = macromoleculeHalfLives(figHandle, simBatchDir, selectedSimulations)</a></li><li><a href="#_sub20" class="code">function figData = rnaSynthesisDuration(figHandle, simBatchDir, selectedSimulations)</a></li><li><a href="#_sub21" class="code">function figData = proteinSynthesisDuration(figHandle, simBatchDir, selectedSimulations)</a></li><li><a href="#_sub22" class="code">function figData = macromolecularComplexes(figHandle, simBatchDir, selectedSimulations)</a></li><li><a href="#_sub23" class="code">function figData = replication(figHandle, simBatchDir, selectedSimulations)</a></li><li><a href="#_sub24" class="code">function figData = SSBs(figHandle, simBatchDir, selectedSimulations)</a></li><li><a href="#_sub25" class="code">function figData = dnaRepair(figHandle, simBatchDir, selectedSimulations)</a></li><li><a href="#_sub26" class="code">function figData = immuneActivation(figHandle, simBatchDir, selectedSimulations)</a></li><li><a href="#_sub27" class="code">function figData = smcSpacing(figHandle, simBatchDir, selectedSimulations)</a></li><li><a href="#_sub28" class="code">function figData = geneCopyNumber(figHandle, simBatchDir, selectedSimulations)</a></li><li><a href="#_sub29" class="code">function figData = unsynchronizedPopulation(figHandle, simBatchDir, selectedSimulations)</a></li><li><a href="#_sub30" class="code">function figData = blockedDecayEvents(figHandle, simBatchDir, selectedSimulations)</a></li><li><a href="#_sub31" class="code">function figData = secondaryReplicationInitiation(figHandle, simBatchDir, selectedSimulations)</a></li><li><a href="#_sub32" class="code">function [figData, tabContent, tabColLabels] = warnings(figHandle, simBatchDir, selectedSimulations)</a></li><li><a href="#_sub33" class="code">function [unsyncTime, unsyncData, cellCount, simOrder] = desynchronizePopulation(sim, syncTime, syncData, simEndTimes)</a></li></ul>

<h2><a name="_source"></a>SOURCE CODE <a href="#_top"><img alt="^" border="0" src="../../../../../../../../up.png"></a></h2>
<div class="fragment"><pre>0001 <span class="comment">%Population</span>
0002 <span class="comment">%</span>
0003 <span class="comment">% Author: Jonathan Karr, jkarr@stanford.edu</span>
0004 <span class="comment">% Author: Derek Macklin, macklin@stanford.edu</span>
0005 <span class="comment">% Affiliation: Covert Lab, Department of Bioengineering, Stanford University</span>
0006 <span class="comment">% Last Updated: 7/6/2011</span>
0007 classdef <a href="Population.html" class="code" title="">Population</a>
0008     methods (Static = true)
0009         <a name="_sub0" href="#_subfunctions" class="code">function run(simBatchDir, fileName, selectedSimulations, methodsToRun)</a>
0010             import edu.stanford.covert.cell.sim.analysis.Population;
0011             import edu.stanford.covert.cell.sim.util.PlotUtil;
0012             import edu.stanford.covert.cell.sim.util.PrintUtil;
0013             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
0014             
0015             <span class="keyword">if</span> nargin &lt; 1 || isempty(simBatchDir)
0016                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
0017             <span class="keyword">end</span>
0018             <span class="keyword">if</span> nargin &lt; 3 || isempty(selectedSimulations)
0019                 selectedSimulations = SimulationDiskUtil.getCompleteSimulations(simBatchDir);
0020             <span class="keyword">end</span>
0021             
0022             <span class="comment">%evaluate each method and optionally save at the result as</span>
0023             <span class="comment">%pdf/xls files</span>
0024             metaData = meta.class.fromName(<span class="string">'edu.stanford.covert.cell.sim.analysis.Population'</span>);
0025             <span class="keyword">for</span> i = 1:numel(metaData.Methods)
0026                 <span class="comment">%only plot methods matching the signature</span>
0027                 <span class="comment">%  plotFcn(figHandle, simBatchDir, selectedSimulations)</span>
0028                 <span class="keyword">if</span> <span class="keyword">...</span>
0029                         numel(metaData.Methods{i}.InputNames) &lt; 3 || ~isequal(metaData.Methods{i}.InputNames(1:3), {<span class="string">'figHandle'</span>; <span class="string">'simBatchDir'</span>; <span class="string">'selectedSimulations'</span>}) || <span class="keyword">...</span>
0030                         numel(metaData.Methods{i}.OutputNames) &lt; 1 || ~isequal(metaData.Methods{i}.OutputNames(1), {<span class="string">'figData'</span>})
0031                     <span class="keyword">continue</span>;
0032                 <span class="keyword">end</span>
0033                 
0034                 <span class="comment">%get method name</span>
0035                 methodName = metaData.Methods{i}.Name;
0036                 MethodName = [upper(methodName(1)) methodName(2:end)];
0037                 <span class="keyword">if</span> nargin &gt;= 4 &amp;&amp; ~ismember(metaData.Methods{i}.Name, methodsToRun)
0038                     <span class="keyword">continue</span>;
0039                 <span class="keyword">end</span>
0040                 
0041                 <span class="comment">%run plot method</span>
0042                 [~, figHandle] = edu.stanford.covert.cell.sim.util.PlotUtil.newAxesHandle();
0043                 <span class="keyword">try</span>
0044                     argout = cell(size(metaData.Methods{i}.OutputNames));
0045                     [argout{:}] = Population.(methodName)(figHandle, simBatchDir, selectedSimulations);
0046                     
0047                     <span class="comment">% save plot</span>
0048                     <span class="keyword">if</span> nargin &gt;= 2
0049                         figData = argout{1}; <span class="comment">%#ok&lt;NASGU&gt;</span>
0050                         save([fileName <span class="string">'-'</span> MethodName <span class="string">'.mat'</span>], <span class="string">'-struct'</span>, <span class="string">'figData'</span>);
0051                         
0052                         orient(figHandle, <span class="string">'portrait'</span>);
0053                         saveas(figHandle, [fileName <span class="string">'-'</span> MethodName <span class="string">'.pdf'</span>]);
0054                     <span class="keyword">end</span>
0055                     
0056                     <span class="comment">% save table</span>
0057                     <span class="keyword">if</span> numel(argout) &gt; 1
0058                         <span class="keyword">if</span> nargin &gt;= 2
0059                             PrintUtil.printToFile(argout{2}, argout{3}, [fileName <span class="string">'-'</span> MethodName <span class="string">'.xls'</span>], MethodName);
0060                         <span class="keyword">else</span>
0061                             PrintUtil.printToStdIO(argout{2}, argout{3});
0062                         <span class="keyword">end</span>
0063                     <span class="keyword">end</span>
0064                 <span class="keyword">catch</span> exception
0065                     warning(<span class="string">'WholeCell:warning'</span>, <span class="string">'Unable to make %s plot:\n%s'</span>, methodName, exception.getReport());
0066                 <span class="keyword">end</span>
0067                 
0068                 <span class="keyword">if</span> nargin &gt;= 2
0069                     close(figHandle);
0070                     clear figHandle;
0071                 <span class="keyword">end</span>
0072                 clear argout figData;
0073             <span class="keyword">end</span>
0074         <span class="keyword">end</span>
0075         
0076         <a name="_sub1" href="#_subfunctions" class="code">function figData = growth(figHandle, simBatchDir, selectedSimulations)</a>
0077             import edu.stanford.covert.cell.sim.util.MacromoleculeUtil;
0078             import edu.stanford.covert.cell.sim.util.SimulationEnsemble;
0079             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
0080             import edu.stanford.covert.cell.sim.util.PlotUtil;
0081             import edu.stanford.covert.cell.sim.analysis.CellOverview;
0082             import edu.stanford.covert.util.ConstantUtil;
0083             
0084             <span class="keyword">if</span> nargin &lt; 2 || isempty(simBatchDir)
0085                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
0086             <span class="keyword">end</span>
0087             <span class="keyword">if</span> nargin &lt; 3 || isempty(selectedSimulations)
0088                 selectedSimulations = SimulationDiskUtil.getCompleteSimulations(simBatchDir);
0089             <span class="keyword">end</span>
0090             
0091             [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep <span class="string">'1'</span>]);
0092             comp = sim.compartment;
0093             g = sim.gene;
0094             r = sim.state(<span class="string">'Rna'</span>);
0095             pm = sim.state(<span class="string">'ProteinMonomer'</span>);
0096             pc = sim.state(<span class="string">'ProteinComplex'</span>);
0097             
0098             <span class="comment">%% get data</span>
0099             stateNames = {
0100                 <span class="string">'mass'</span>              <span class="string">'Mass'</span>
0101                 <span class="string">'ploidy'</span>            {<span class="string">'Chromosome'</span> <span class="string">'Copy Number'</span>}
0102                 <span class="string">'rnas'</span>              <span class="string">'RNA'</span>
0103                 <span class="string">'proteins'</span>          <span class="string">'Proteins'</span>
0104                 <span class="string">'lipids'</span>            <span class="string">'Lipids'</span>
0105                 <span class="string">'growth_rate'</span>       <span class="string">'Growth'</span>
0106                 <span class="string">'ribosomes'</span>         <span class="string">'Ribosomes'</span>
0107                 };
0108             ensemble = SimulationEnsemble(simBatchDir, stateNames, [], selectedSimulations);
0109             simEndTimes = ensemble.stateData.simulationEndTimes;
0110             
0111             time = ensemble.stateData.time / 3600;
0112             mass = ensemble.stateData.values(ensemble.getPropertyIndices(<span class="string">'mass'</span>), :, :, :) * sim.state(<span class="string">'Mass'</span>).cellInitialDryWeight / (1-sim.state(<span class="string">'Mass'</span>).fractionWetWeight) * 1e15;
0113             <a href="#_sub2" class="code" title="subfunction figData = growth(figHandle, simBatchDir, selectedSimulations)">growth</a> = ensemble.stateData.values(ensemble.getPropertyIndices(<span class="string">'growth_rate'</span>), :, :, :) * 3600 * sim.state(<span class="string">'Mass'</span>).cellInitialDryWeight / (1 - sim.state(<span class="string">'Mass'</span>).fractionWetWeight) * 1e15;
0114             ploidy = ensemble.stateData.values(ensemble.getPropertyIndices(<span class="string">'ploidy'</span>), :, :, :);
0115             rnas = ensemble.stateData.values(ensemble.getPropertyIndices(<span class="string">'rnas'</span>), :, :, :);
0116             proteins = ensemble.stateData.values(ensemble.getPropertyIndices(<span class="string">'proteins'</span>), :, :, :);
0117             lipids = ensemble.stateData.values(ensemble.getPropertyIndices(<span class="string">'lipids'</span>), :, :, :);
0118             <a href="#_sub16" class="code" title="subfunction figData = ribosomes(figHandle, simBatchDir, selectedSimulations)">ribosomes</a> = ensemble.stateData.values(ensemble.getPropertyIndices(<span class="string">'ribosomes'</span>), :, :, :);
0119             
0120             [~, simOrder] = sort(mass(:, 1, min(ensemble.stateData.simulationEndTimes), :));
0121             simOrder = simOrder(:);
0122             
0123             clear ensemble;
0124             
0125             ribMonomerIdxs = find(any(pc.proteinComplexComposition(g.mRNAIndexs, pc.ribosome70SIndexs, :), 3));
0126             [~, tmp] = min(r.matureRNAGeneComposition(g.mRNAIndexs(ribMonomerIdxs), :) * r.expression(r.matureIndexs));
0127             [~, ~, ~, ~, ~, <span class="keyword">...</span>
0128                 ~, ~, ~, ~, <span class="keyword">...</span>
0129                 ~, ~, ~, ~, <span class="keyword">...</span>
0130                 ribMatureRnaIdxs, ~, ~, ~, <span class="keyword">...</span>
0131                 ribMonomerIdxs, ~, ~, ribMonomerNames] = <span class="keyword">...</span>
0132                 MacromoleculeUtil.getMacromoleculeIndexsIDsNames(pm.wholeCellModelIDs{pm.matureIndexs(g.mRNAIndexs(ribMonomerIdxs(tmp)))}, sim);
0133             ribMonomerNames = strrep(ribMonomerNames, <span class="string">'ribosomal protein '</span>, <span class="string">''</span>);
0134             stateNames = {
0135                 <span class="string">'Rna'</span>               <span class="string">'counts'</span>  r.matureIndexs(ribMatureRnaIdxs)   comp.cytosolIndexs
0136                 <span class="string">'ProteinMonomer'</span>    <span class="string">'counts'</span>  pm.matureIndexs(ribMonomerIdxs)    comp.cytosolIndexs
0137                 <span class="string">'Mass'</span>              <span class="string">'proteinWt'</span> <span class="string">':'</span> <span class="string">':'</span>
0138                 };
0139             states = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, <span class="string">'extract'</span>, selectedSimulations);
0140             ribRNAs = states.Rna.counts;
0141             ribMonomers = states.ProteinMonomer.counts;
0142             proteinWt = states.Mass.proteinWt;
0143             
0144             <span class="keyword">for</span> i = 1:numel(simEndTimes)
0145                 ribRNAs(:, :, simEndTimes(i)+1:<span class="keyword">end</span>, i) = NaN;
0146                 ribMonomers(:, :, simEndTimes(i)+1:<span class="keyword">end</span>, i) = NaN;
0147                 proteinWt(:, :, simEndTimes(i)+1:<span class="keyword">end</span>, i) = NaN;
0148             <span class="keyword">end</span>
0149             
0150             clear states;
0151             
0152             <span class="comment">%% clear figure</span>
0153             clf(figHandle);
0154             
0155             <span class="comment">%% left panel</span>
0156             
0157             <span class="comment">%layout plots</span>
0158             axesHandlesL = PlotUtil.multiElementPlot(figHandle, 3.5 * ones(5, 1), [0 time(end)], struct(<span class="keyword">...</span>
0159                 <span class="string">'titleStr'</span>, <span class="string">'Composition'</span>, <span class="keyword">...</span>
0160                 <span class="string">'position'</span>, [0.12 0.52 0.21 0.4], <span class="keyword">...</span>
0161                 <span class="string">'colorOrder'</span>, PlotUtil.getRedGreenColorOrder(simOrder)));
0162             
0163             <span class="comment">%plot data</span>
0164             PlotUtil.plotLine(axesHandlesL(1), time, permute(mass, [4 3 2 1]), false, true, false);
0165             ylabel(axesHandlesL(1), {<span class="string">'Mass'</span> <span class="string">'(fg)'</span>});
0166             
0167             PlotUtil.plotLine(axesHandlesL(2), time, permute(ploidy, [4 3 2 1]), false, true, false);
0168             ylabel(axesHandlesL(2), {<span class="string">'Chrosome'</span> <span class="string">'Copy'</span> <span class="string">'Number'</span>});
0169             
0170             PlotUtil.plotLine(axesHandlesL(3), time, permute(rnas, [4 3 2 1]), false, true, false);
0171             ylabel(axesHandlesL(3), <span class="string">'RNAs'</span>);
0172             
0173             PlotUtil.plotLine(axesHandlesL(4), time, permute(proteins, [4 3 2 1]), false, true, false);
0174             ylabel(axesHandlesL(4), <span class="string">'Proteins'</span>);
0175             
0176             PlotUtil.plotLine(axesHandlesL(5), time, permute(lipids, [4 3 2 1]), false, true, false);
0177             ylabel(axesHandlesL(5), <span class="string">'Lipids'</span>);
0178             
0179             <span class="comment">% Format Y-axis</span>
0180             PlotUtil.alignYAxesLabels(axesHandlesL);
0181             PlotUtil.offsetYAxes(axesHandlesL, 0.04);
0182             PlotUtil.labelSubplots(axesHandlesL(1), <span class="string">'A'</span>, -0.42, 1.2);
0183             
0184             <span class="comment">%% right panel</span>
0185             
0186             <span class="comment">%layout plots</span>
0187             axesHandlesR = PlotUtil.multiElementPlot(figHandle, 3.5 * ones(5, 1), [0 time(end)], struct(<span class="keyword">...</span>
0188                 <span class="string">'titleStr'</span>, <span class="string">'Growth'</span>, <span class="keyword">...</span>
0189                 <span class="string">'position'</span>, [0.45 0.52 0.21 0.4], <span class="keyword">...</span>
0190                 <span class="string">'colorOrder'</span>, PlotUtil.getRedGreenColorOrder(simOrder)));
0191             
0192             <span class="comment">%plot data</span>
0193             PlotUtil.plotLine(axesHandlesR(1), time, permute(<a href="#_sub2" class="code" title="subfunction figData = growth(figHandle, simBatchDir, selectedSimulations)">growth</a>, [4 3 2 1]), false, true, false);
0194             ylabel(axesHandlesR(1), {<span class="string">'Growth'</span> <span class="string">'(fg h^{-1})'</span>});
0195             
0196             PlotUtil.plotLine(axesHandlesR(2), time(3:end), diff(permute(sum(proteinWt, 2), [4 3 2 1]), 1, 2), false, true, false);
0197             ylabel(axesHandlesR(2), {<span class="string">'Protein'</span> <span class="string">'Synthesis'</span>});
0198             
0199             PlotUtil.plotLine(axesHandlesR(3), time, permute(<a href="#_sub16" class="code" title="subfunction figData = ribosomes(figHandle, simBatchDir, selectedSimulations)">ribosomes</a>, [4 3 2 1]), false, true, false);
0200             ylabel(axesHandlesR(3), <span class="string">'Ribosomes'</span>);
0201             
0202             PlotUtil.plotLine(axesHandlesR(4), time(2:end), permute(ribRNAs, [4 3 2 1]), false, true, false);
0203             ylabel(axesHandlesR(4), [ribMonomerNames <span class="string">'Protein'</span>]);
0204             
0205             PlotUtil.plotLine(axesHandlesR(5), time(2:end), permute(ribMonomers, [4 3 2 1]), false, true, false);
0206             ylabel(axesHandlesR(5), [ribMonomerNames <span class="string">'mRNA'</span>]);
0207             
0208             <span class="comment">% Format Y-axis</span>
0209             PlotUtil.alignYAxesLabels(axesHandlesR);
0210             PlotUtil.offsetYAxes(axesHandlesR, 0.04);
0211             PlotUtil.labelSubplots(axesHandlesR(1), <span class="string">'B'</span>, -0.34, 1.2);
0212             
0213             <span class="comment">%% organize figure data</span>
0214             figData = struct;
0215             figData.simOrder = simOrder;
0216             figData.time = time;
0217             figData.mass = mass;
0218             figData.ploidy = ploidy;
0219             figData.rnas = rnas;
0220             figData.proteins = proteins;
0221             figData.lipids = lipids;
0222             figData.growth = <a href="#_sub2" class="code" title="subfunction figData = growth(figHandle, simBatchDir, selectedSimulations)">growth</a>;
0223             figData.proteinWt = proteinWt;
0224             figData.ribosomes = <a href="#_sub16" class="code" title="subfunction figData = ribosomes(figHandle, simBatchDir, selectedSimulations)">ribosomes</a>;
0225             figData.ribRNAs = ribRNAs;
0226             figData.ribMonomers = ribMonomers;
0227             figData.ribMonomerNames = ribMonomerNames;
0228         <span class="keyword">end</span>
0229         
0230         <a name="_sub2" href="#_subfunctions" class="code">function figData = energyProduction(figHandle, simBatchDir, selectedSimulations)</a>
0231             import edu.stanford.covert.cell.sim.util.PlotUtil;
0232             import edu.stanford.covert.cell.sim.util.SimulationEnsemble;
0233             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
0234             
0235             <span class="keyword">if</span> nargin &lt; 2 || isempty(simBatchDir)
0236                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
0237             <span class="keyword">end</span>
0238             <span class="keyword">if</span> nargin &lt; 3 || isempty(selectedSimulations)
0239                 selectedSimulations = SimulationDiskUtil.getCompleteSimulations(simBatchDir);
0240             <span class="keyword">end</span>
0241             
0242             <span class="comment">%% constants</span>
0243             [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep <span class="string">'1'</span>]);
0244             comp = sim.compartment;
0245             m = sim.state(<span class="string">'Metabolite'</span>);
0246             met = sim.process(<span class="string">'Metabolism'</span>);
0247             mass = sim.state(<span class="string">'Mass'</span>);
0248             
0249             metIdxs = sub2ind([numel(m.wholeCellModelIDs) comp.count], <span class="keyword">...</span>
0250                 [m.ntpIndexs; m.ndpIndexs; m.nmpIndexs], repmat(comp.cytosolIndexs, 12, 1));
0251             metProcessIdx = sim.processIndex(met.wholeCellModelID);
0252             
0253             <span class="comment">%% get data</span>
0254             stateNames = {
0255                 <span class="string">'Time'</span>               <span class="string">'values'</span>         <span class="string">':'</span>     <span class="string">':'</span>
0256                 <span class="string">'MetabolicReaction'</span>  <span class="string">'growth'</span>         <span class="string">':'</span>     <span class="string">':'</span>
0257                 <span class="string">'Mass'</span>               <span class="string">'cell'</span>           <span class="string">':'</span>     <span class="string">'-sum'</span>
0258                 <span class="string">'Metabolite'</span>         <span class="string">'processUsages'</span>  metIdxs  metProcessIdx
0259                 };
0260             states = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, <span class="string">'extract'</span>, selectedSimulations);
0261             simEndTimes = permute(max(states.Time.values, [], 3), [4 1 2 3]);
0262             
0263             timeIdx = min(simEndTimes);
0264             [~, simOrder] = sort(states.Mass.cell(:, :, timeIdx, :));
0265             simOrder = simOrder(:);
0266             
0267             time = permute(max(states.Time.values, [], 4), [3 4 1 2]) / 3600;
0268             states.Metabolite.processUsages = -full(states.Metabolite.processUsages);
0269             <span class="keyword">for</span> i = 1:numel(simEndTimes)
0270                 states.MetabolicReaction.growth(:, :, simEndTimes(i)+1:<span class="keyword">end</span>, i) = NaN;
0271                 states.Metabolite.processUsages(:, :, simEndTimes(i)+1:<span class="keyword">end</span>, i) = NaN;
0272             <span class="keyword">end</span>
0273             
0274             <span class="comment">%% plot data</span>
0275             clf(figHandle);
0276             options = struct;
0277             options.position = [0.02 0.32 0.95 0.4];
0278             options.titleStr = {<span class="string">'Growth'</span>; <span class="string">'A'</span>; <span class="string">'C'</span>; <span class="string">'G'</span>; <span class="string">'U'</span>};
0279             options.colorOrder = PlotUtil.getRedGreenColorOrder(simOrder);
0280             options.xdata = time;
0281             options.ydata = {
0282                 {permute(states.MetabolicReaction.growth * 3600 * mass.cellInitialDryWeight / (1 - mass.fractionWetWeight) * 1e15, [3 4 1 2])}
0283                 permute(mat2cell(permute(states.Metabolite.processUsages(1:4:<span class="keyword">end</span>, :, :, :), [3 4 1 2]), numel(time), numel(simEndTimes), ones(3, 1)), [3 1 2])
0284                 permute(mat2cell(permute(states.Metabolite.processUsages(2:4:<span class="keyword">end</span>, :, :, :), [3 4 1 2]), numel(time), numel(simEndTimes), ones(3, 1)), [3 1 2])
0285                 permute(mat2cell(permute(states.Metabolite.processUsages(3:4:<span class="keyword">end</span>, :, :, :), [3 4 1 2]), numel(time), numel(simEndTimes), ones(3, 1)), [3 1 2])
0286                 permute(mat2cell(permute(states.Metabolite.processUsages(4:4:<span class="keyword">end</span>, :, :, :), [3 4 1 2]), numel(time), numel(simEndTimes), ones(3, 1)), [3 1 2])
0287                 };
0288             options.ylabelStr = {
0289                 {<span class="string">'Growth (fg h^{-1})'</span>}
0290                 {<span class="string">'ATP'</span>; <span class="string">'ADP'</span>; <span class="string">'AMP'</span>}
0291                 {<span class="string">'CTP'</span>; <span class="string">'CDP'</span>; <span class="string">'CMP'</span>}
0292                 {<span class="string">'GTP'</span>; <span class="string">'GDP'</span>; <span class="string">'GMP'</span>}
0293                 {<span class="string">'UTP'</span>; <span class="string">'UDP'</span>; <span class="string">'UMP'</span>}
0294                 };
0295             PlotUtil.multiElementPlot(figHandle, {30 [10; 10; 10] [10; 10; 10] [10; 10; 10] [10; 10; 10]}, [0 time(end)], options);
0296             
0297             clear options;
0298             
0299             <span class="comment">%% organize figure data</span>
0300             figData = struct;
0301             figData.simOrder = simOrder;
0302             figData.time = time;
0303             figData.growth = states.MetabolicReaction.growth;
0304             figData.processUsages = states.Metabolite.processUsages;
0305         <span class="keyword">end</span>
0306         
0307         <a name="_sub3" href="#_subfunctions" class="code">function figData = cellMassDistribution(figHandle, simBatchDir, selectedSimulations)</a>
0308             import edu.stanford.covert.cell.sim.util.PlotUtil;
0309             import edu.stanford.covert.cell.sim.util.SimulationEnsemble;
0310             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
0311             import edu.stanford.covert.util.ConstantUtil;
0312             
0313             <span class="keyword">if</span> nargin &lt; 2 || isempty(simBatchDir)
0314                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
0315             <span class="keyword">end</span>
0316             <span class="keyword">if</span> nargin &lt; 3 || isempty(selectedSimulations)
0317                 selectedSimulations = SimulationDiskUtil.getCompleteSimulations(simBatchDir);
0318             <span class="keyword">end</span>
0319             
0320             <span class="comment">%% get data</span>
0321             stateNames = {
0322                 <span class="string">'mass'</span> <span class="string">'Mass'</span>
0323                 };
0324             ensemble = SimulationEnsemble(simBatchDir, stateNames, [], selectedSimulations);
0325             mass = ensemble.stateData.values(ensemble.getPropertyIndices(<span class="string">'mass'</span>), :, :, :);
0326             
0327             simEndTimes = ensemble.stateData.simulationEndTimes;
0328             
0329             initMass = permute(mass(:, :, 1, :), [4 3 1 2]);
0330             finMass = mass(sub2ind(size(mass), <span class="keyword">...</span>
0331                 ones(size(simEndTimes)), <span class="keyword">...</span>
0332                 ones(size(simEndTimes)), <span class="keyword">...</span>
0333                 simEndTimes, <span class="keyword">...</span>
0334                 (1:numel(simEndTimes))'));
0335             
0336             <span class="comment">%% plot data</span>
0337             clf(figHandle);
0338             
0339             mass = linspace(min([initMass; finMass - 1]), max([initMass; finMass - 1]), 20);
0340             initMassFreq = histc(initMass, mass);
0341             finMassFreq = histc(finMass - 1, mass);
0342             
0343             axesHandle = subplot(2, 1, 1, <span class="string">'Parent'</span>, figHandle);
0344             bar(axesHandle, mass, initMassFreq);
0345             xlabel(axesHandle, <span class="string">'Initial Mass (cell dry weight)'</span>);
0346             ylabel(axesHandle, <span class="string">'Frequency'</span>);
0347             
0348             axesHandle = subplot(2, 1, 2, <span class="string">'Parent'</span>, figHandle);
0349             bar(axesHandle, mass + 1, finMassFreq);
0350             xlabel(axesHandle, <span class="string">'Final Mass (cell dry weight)'</span>);
0351             ylabel(axesHandle, <span class="string">'Frequency'</span>);
0352             
0353             <span class="comment">%% figure data</span>
0354             figData = struct;
0355             figData.initMass = initMass;
0356             figData.finMass = finMass;
0357         <span class="keyword">end</span>
0358         
0359         <a name="_sub4" href="#_subfunctions" class="code">function figData = massDistribution(figHandle, simBatchDir, selectedSimulations)</a>
0360             import edu.stanford.covert.cell.sim.util.PlotUtil;
0361             import edu.stanford.covert.cell.sim.util.SimulationEnsemble;
0362             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
0363             import edu.stanford.covert.util.ConstantUtil;
0364             
0365             <span class="keyword">if</span> nargin &lt; 2 || isempty(simBatchDir)
0366                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
0367             <span class="keyword">end</span>
0368             <span class="keyword">if</span> nargin &lt; 3 || isempty(selectedSimulations)
0369                 selectedSimulations = SimulationDiskUtil.getCompleteSimulations(simBatchDir);
0370             <span class="keyword">end</span>
0371             
0372             [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep <span class="string">'1'</span>]);
0373             comp = sim.compartment;
0374             
0375             <span class="comment">%% get data</span>
0376             stateNames = {
0377                 <span class="string">'Time'</span>      <span class="string">'values'</span>             <span class="string">':'</span> <span class="string">':'</span>
0378                 <span class="string">'Mass'</span>      <span class="string">'cell'</span>               <span class="string">':'</span> <span class="string">'-sum'</span>
0379                 <span class="string">'Mass'</span>      <span class="string">'dnaWt'</span>              <span class="string">':'</span> comp.cytosolIndexs
0380                 <span class="string">'Mass'</span>      <span class="string">'rnaWt'</span>              <span class="string">':'</span> comp.cytosolIndexs
0381                 <span class="string">'Mass'</span>      <span class="string">'proteinWt'</span>          <span class="string">':'</span> <span class="string">'-sum'</span>
0382                 <span class="string">'Mass'</span>      <span class="string">'metaboliteWt'</span>       <span class="string">':'</span> [comp.cytosolIndexs; comp.membraneIndexs]
0383                 };
0384             states = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, <span class="string">'extract'</span>, selectedSimulations);
0385             
0386             time = permute(max(states.Time.values, [], 4), [3 4 1 2]) / 3600;
0387             mass = permute(states.Mass.cell, [3 4 1 2]) * 1e15;
0388             dnaWt = permute(states.Mass.dnaWt, [3 4 1 2]) * 1e15;
0389             rnaWt = permute(states.Mass.rnaWt, [3 4 1 2]) * 1e15;
0390             proteinWt = permute(states.Mass.proteinWt, [3 4 1 2]) * 1e15;
0391             membraneWt = permute(states.Mass.metaboliteWt(:, 2, :, :), [3 4 1 2]) * 1e15;
0392             metaboliteWt = permute(states.Mass.metaboliteWt(:, 1, :, :), [3 4 1 2]) * 1e15;
0393             
0394             dnaWt(mass == 0) = NaN;
0395             rnaWt(mass == 0) = NaN;
0396             proteinWt(mass == 0) = NaN;
0397             membraneWt(mass == 0) = NaN;
0398             metaboliteWt(mass == 0) = NaN;
0399             mass(mass == 0) = NaN;
0400             
0401             minSimEndTime = min(max(states.Time.values, [], 4));
0402             [~, simOrder] = sort(mass(minSimEndTime, :));
0403             
0404             clear states;
0405             
0406             <span class="comment">%% layout figure</span>
0407             clf(figHandle);
0408             
0409             options = struct();
0410             options.titleStr = <span class="string">'Cell Mass'</span>;
0411             options.colorOrder = PlotUtil.getRedGreenColorOrder(simOrder);
0412             axesHandles = PlotUtil.multiElementPlot(figHandle, 3 * ones(6, 1), [0 time(end)], options);
0413             
0414             <span class="comment">%% plot data</span>
0415             
0416             PlotUtil.plotLine(axesHandles(1), time, mass, false, true, false);
0417             ylabel(axesHandles(1), <span class="string">'Mass (fg)'</span>);
0418             
0419             PlotUtil.plotLine(axesHandles(2), time, dnaWt, false, true, false);
0420             ylabel(axesHandles(2), <span class="string">'DNA (fg)'</span>);
0421             
0422             PlotUtil.plotLine(axesHandles(3), time, rnaWt, false, true, false);
0423             ylabel(axesHandles(3), <span class="string">'RNA (fg)'</span>);
0424             
0425             PlotUtil.plotLine(axesHandles(4), time, proteinWt, false, true, false);
0426             ylabel(axesHandles(4), <span class="string">'Protein (fg)'</span>);
0427             
0428             PlotUtil.plotLine(axesHandles(5), time, membraneWt, false, true, false);
0429             ylabel(axesHandles(5), <span class="string">'Membrane (fg)'</span>);
0430             
0431             PlotUtil.plotLine(axesHandles(6), time, metaboliteWt, false, true, false);
0432             ylabel(axesHandles(6), <span class="string">'Metabolites (fg)'</span>);
0433             
0434             <span class="comment">%% Format Y-axis</span>
0435             PlotUtil.alignYAxesLabels(axesHandles);
0436             PlotUtil.offsetYAxes(axesHandles, 0.03);
0437             
0438             <span class="comment">%% organize figure data</span>
0439             figData = struct;
0440             figData.simOrder = simOrder;
0441             figData.time = time;
0442             figData.mass = mass;
0443             figData.dnaWt = dnaWt;
0444             figData.rnaWt = rnaWt;
0445             figData.proteinWt = proteinWt;
0446             figData.membraneWt = membraneWt;
0447             figData.metaboliteWt = metaboliteWt;
0448         <span class="keyword">end</span>
0449         
0450         <a name="_sub5" href="#_subfunctions" class="code">function figData = cellCyclePhases(figHandle, simBatchDir, selectedSimulations)</a>
0451             import edu.stanford.covert.cell.sim.util.SummaryLogger;
0452             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
0453             import edu.stanford.covert.cell.sim.util.SimulationEnsemble;
0454             import edu.stanford.covert.cell.sim.util.PlotUtil;
0455             
0456             <span class="keyword">if</span> nargin &lt; 2 || isempty(simBatchDir)
0457                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
0458             <span class="keyword">end</span>
0459             <span class="keyword">if</span> nargin &lt; 3 || isempty(selectedSimulations)
0460                 selectedSimulations = SimulationDiskUtil.getCompleteSimulations(simBatchDir);
0461             <span class="keyword">end</span>
0462             
0463             <span class="comment">%% get data</span>
0464             [simDir, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep <span class="string">'1'</span>]);
0465             simBatchDir = simDir(1:find(simDir == filesep, 1, <span class="string">'last'</span>) - 1);
0466             mass = sim.state(<span class="string">'Mass'</span>);
0467             stats = SummaryLogger.getSimulationStatistics(simBatchDir);
0468             nSims = size(stats, 1);
0469             
0470             ensemble = SimulationEnsemble(simBatchDir, cell(0, 2), [], selectedSimulations);
0471             maxTime = ensemble.stateData.time(end);
0472             
0473             clear ensemble;
0474             
0475             <span class="comment">%% clear figure</span>
0476             clf(figHandle);
0477             
0478             <span class="comment">%% top panel</span>
0479             
0480             <span class="comment">%layout plots</span>
0481             edges = 15 * 60 * (0 : ceil(maxTime / (15 * 60)));
0482             n1 = histc(stats(:, SummaryLogger.SIM_STATUS_INDEX_CYTOKINESIS_TIME), edges);
0483             n2 = histc(stats(:, SummaryLogger.SIM_STATUS_INDEX_REP_INIT_TIME), edges);
0484             n3 = histc(stats(:, SummaryLogger.SIM_STATUS_INDEX_REP_TIME) - stats(:, SummaryLogger.SIM_STATUS_INDEX_REP_INIT_TIME), edges);
0485             n4 = histc(stats(:, SummaryLogger.SIM_STATUS_INDEX_CYTOKINESIS_TIME) - stats(:, SummaryLogger.SIM_STATUS_INDEX_REP_TIME), edges);
0486             
0487             [axesHandlesT, xAxesHandleT] = PlotUtil.multiElementPlot(figHandle, 3 * ones(4, 1), edges([1 end]) / 3600', <span class="keyword">...</span>
0488                 struct(<span class="string">'position'</span>, [0.12 0.68 0.21 0.24], <span class="string">'xlabelStr'</span>, <span class="string">'Cell Cycle Phase Duration (h)'</span>)); <span class="comment">%#ok&lt;NASGU&gt;</span>
0489             
0490             <span class="comment">%plot data</span>
0491             bar(axesHandlesT(1), edges / 3600, n1 / nSims * 100, <span class="string">'histc'</span>);
0492             ylabel(axesHandlesT(1), {<span class="string">'Cell'</span> <span class="string">'Cycle'</span>});
0493             ytick = get(axesHandlesT(1), <span class="string">'YTick'</span>);
0494             set(axesHandlesT(1), <span class="string">'YTick'</span>, ytick([1 end]));
0495             ylim = get(axesHandlesT(1), <span class="string">'YLim'</span>);
0496             set(axesHandlesT(1), <span class="string">'YLim'</span>, ylim);
0497             
0498             bar(axesHandlesT(2), edges / 3600, n2 / nSims * 100, <span class="string">'histc'</span>);
0499             ylabel(axesHandlesT(2), {<span class="string">'Rep'</span> <span class="string">'Init'</span>});
0500             ytick = get(axesHandlesT(2), <span class="string">'YTick'</span>);
0501             set(axesHandlesT(2), <span class="string">'YTick'</span>, ytick([1 end]));
0502             ylim = get(axesHandlesT(2), <span class="string">'YLim'</span>);
0503             set(axesHandlesT(2), <span class="string">'YLim'</span>, ylim);
0504             
0505             bar(axesHandlesT(3), edges / 3600, n3 / nSims * 100, <span class="string">'histc'</span>);
0506             ylabel(axesHandlesT(3), {<span class="string">'Repli-'</span> <span class="string">'cation'</span>});
0507             ytick = get(axesHandlesT(3), <span class="string">'YTick'</span>);
0508             set(axesHandlesT(3), <span class="string">'YTick'</span>, ytick([1 end]));
0509             ylim = get(axesHandlesT(3), <span class="string">'YLim'</span>);
0510             set(axesHandlesT(3), <span class="string">'YLim'</span>, ylim);
0511             
0512             bar(axesHandlesT(4), edges / 3600, n4 / nSims * 100, <span class="string">'histc'</span>);
0513             ylabel(axesHandlesT(4), {<span class="string">'Cyto-'</span> <span class="string">'kinesis'</span>});
0514             ytick = get(axesHandlesT(4), <span class="string">'YTick'</span>);
0515             set(axesHandlesT(4), <span class="string">'YTick'</span>, ytick([1 end]));
0516             ylim = get(axesHandlesT(4), <span class="string">'YLim'</span>);
0517             set(axesHandlesT(4), <span class="string">'YLim'</span>, ylim);
0518             
0519             <span class="comment">%xlabelPos = get(get(xAxesHandleT, 'XLabel'), 'Position');</span>
0520             <span class="comment">%set(get(xAxesHandleT, 'XLabel'), 'Position', xlabelPos+[0 17 0]);</span>
0521             
0522             <span class="comment">%% bottom panel</span>
0523             edges = linspace(min(stats(:, SummaryLogger.SIM_STATUS_INDEX_FIN_MASS)), max(stats(:, SummaryLogger.SIM_STATUS_INDEX_FIN_MASS)), 50);
0524             n = histc(stats(:, SummaryLogger.SIM_STATUS_INDEX_FIN_MASS), edges);
0525             
0526             [axesHandlesB, xAxesHandleB] = PlotUtil.multiElementPlot(figHandle, 3, <span class="keyword">...</span>
0527                 edges([1 end]) * mass.cellInitialDryWeight / (1-mass.fractionWetWeight) * 1e15, <span class="keyword">...</span>
0528                 struct(<span class="string">'position'</span>, [0.12 0.52 0.21 0.06], <span class="string">'xlabelStr'</span>, <span class="string">'Cell Mass (fg)'</span>)); <span class="comment">%#ok&lt;NASGU&gt;</span>
0529             
0530             bar(axesHandlesB(1), edges * mass.cellInitialDryWeight / (1-mass.fractionWetWeight) * 1e15, n / nSims * 100, <span class="string">'histc'</span>);
0531             ylabel(axesHandlesB(1), <span class="string">'Freq'</span>);
0532             ytick = get(axesHandlesB(1), <span class="string">'YTick'</span>);
0533             set(axesHandlesB(1), <span class="string">'YTick'</span>, ytick([1 end]));
0534             ylim = get(axesHandlesB(1), <span class="string">'YLim'</span>);
0535             set(axesHandlesB(1), <span class="string">'YLim'</span>, ylim);
0536             
0537             <span class="comment">%xlabelPos = get(get(xAxesHandleB, 'XLabel'), 'Position');</span>
0538             <span class="comment">%set(get(xAxesHandleB, 'XLabel'), 'Position', xlabelPos+[0 15 0]);</span>
0539             
0540             <span class="comment">%% Format Y-axis</span>
0541             PlotUtil.alignYAxesLabels(axesHandlesT);
0542             PlotUtil.offsetYAxes([axesHandlesT; axesHandlesB], 0.04);
0543             PlotUtil.labelSubplots(axesHandlesT(1), <span class="string">'A'</span>, -0.25, 1.2);
0544             PlotUtil.labelSubplots(axesHandlesB(1), <span class="string">'B'</span>, -0.25, 1.2);
0545             
0546             <span class="comment">%% organize figure data</span>
0547             figData = struct;
0548             figData.nSims = nSims;
0549             figData.maxTime = maxTime;
0550             figData.mass = mass;
0551             figData.stats = stats;
0552         <span class="keyword">end</span>
0553         
0554         <a name="_sub6" href="#_subfunctions" class="code">function [figData, tabContent, tabColLabels] = nucleotideDistribution(figHandle, simBatchDir, selectedSimulations)</a>
0555             import edu.stanford.covert.cell.sim.util.MacromoleculeUtil;
0556             import edu.stanford.covert.cell.sim.util.SimulationEnsemble;
0557             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
0558             import edu.stanford.covert.cell.sim.util.PlotUtil;
0559             import edu.stanford.covert.cell.sim.analysis.CellOverview;
0560             import edu.stanford.covert.util.ConstantUtil;
0561             
0562             <span class="keyword">if</span> nargin &lt; 2 || isempty(simBatchDir)
0563                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
0564             <span class="keyword">end</span>
0565             <span class="keyword">if</span> nargin &lt; 3 || isempty(selectedSimulations)
0566                 selectedSimulations = SimulationDiskUtil.getCompleteSimulations(simBatchDir);
0567             <span class="keyword">end</span>
0568             
0569             [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep <span class="string">'1'</span>]);
0570             comp = sim.compartment;
0571             m = sim.state(<span class="string">'Metabolite'</span>);
0572             
0573             <span class="comment">%% get data</span>
0574             stateNames = {
0575                 <span class="string">'mass'</span> <span class="string">'Mass'</span>
0576                 };
0577             ensemble = SimulationEnsemble(simBatchDir, stateNames, [], selectedSimulations);            
0578             mass = ensemble.stateData.values(ensemble.getPropertyIndices(<span class="string">'mass'</span>), :, :, :) * sim.state(<span class="string">'Mass'</span>).cellInitialDryWeight / (1-sim.state(<span class="string">'Mass'</span>).fractionWetWeight) * 1e15;
0579             
0580             simEndTimes = ensemble.stateData.simulationEndTimes;
0581             timeIdx = ceil(min(simEndTimes) / ensemble.stateData.downsampleStepSec);
0582             [~, simOrder] = sort(mass(:, 1, timeIdx, :));
0583             simOrder = simOrder(:);
0584             
0585             clear ensemble mass;
0586             
0587             stateNames = {
0588                 <span class="string">'Time'</span>       <span class="string">'values'</span> <span class="string">':'</span>  <span class="string">':'</span>
0589                 <span class="string">'Metabolite'</span> <span class="string">'counts'</span> [m.ntpIndexs; m.ndpIndexs; m.nmpIndexs; m.aminoAcidIndexs] comp.cytosolIndexs
0590                 };
0591             states = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, <span class="string">'extract'</span>, selectedSimulations);
0592             time = max(full(states.Time.values), [], 4) / 3600;
0593             ntps = full(states.Metabolite.counts(1:4, :, :, :));
0594             ndps = full(states.Metabolite.counts(5:8, :, :, :));
0595             nmps = full(states.Metabolite.counts(9:12, :, :, :));
0596             aas = full(states.Metabolite.counts(13:<span class="keyword">end</span>, :, :, :));
0597             
0598             <span class="keyword">for</span> i = 1:numel(simEndTimes)
0599                 ntps(:, :, simEndTimes(i)+1:<span class="keyword">end</span>, i) = NaN;
0600                 ndps(:, :, simEndTimes(i)+1:<span class="keyword">end</span>, i) = NaN;
0601                 nmps(:, :, simEndTimes(i)+1:<span class="keyword">end</span>, i) = NaN;
0602                 aas(:, :, simEndTimes(i)+1:<span class="keyword">end</span>, i) = NaN;
0603             <span class="keyword">end</span>
0604             
0605             clear states;
0606             
0607             <span class="comment">%% clear figure</span>
0608             clf(figHandle);
0609             
0610             <span class="comment">%% left panel</span>
0611             
0612             <span class="comment">%layout plots</span>
0613             options = struct();
0614             options.titleStr = {<span class="string">'Nucleotides'</span>; <span class="string">'AXP'</span>; <span class="string">'CXP'</span>; <span class="string">'GXP'</span>; <span class="string">'UXP'</span>};
0615             options.xdata = permute(time, [4 3 2 1]);
0616             options.ydata = {{
0617                 permute(sum(aas, 1), [4 3 2 1])
0618                 permute(sum(ntps, 1), [4 3 2 1])
0619                 permute(sum(ndps, 1), [4 3 2 1])
0620                 permute(sum(nmps, 1), [4 3 2 1])
0621                 };{
0622                 []
0623                 permute(ntps(1, :, :, :), [4 3 2 1])
0624                 permute(ndps(1, :, :, :), [4 3 2 1])
0625                 permute(nmps(1, :, :, :), [4 3 2 1])
0626                 };{
0627                 []
0628                 permute(ntps(2, :, :, :), [4 3 2 1])
0629                 []
0630                 permute(nmps(2, :, :, :), [4 3 2 1])
0631                 };{
0632                 []
0633                 permute(ntps(3, :, :, :), [4 3 2 1])
0634                 permute(ndps(3, :, :, :), [4 3 2 1])
0635                 permute(nmps(3, :, :, :), [4 3 2 1])
0636                 }; {
0637                 []
0638                 permute(ntps(4, :, :, :), [4 3 2 1])
0639                 []
0640                 permute(nmps(4, :, :, :), [4 3 2 1])
0641                 }};
0642             options.ylabelStr = {
0643                 {<span class="string">'Amino Acid'</span>; <span class="string">'NTP'</span>; <span class="string">'NDP'</span>; <span class="string">'NMP'</span>};
0644                 {}; {}; {}; {}
0645                 };
0646             options.colorOrder = PlotUtil.getRedGreenColorOrder(simOrder);
0647             [axesHandles, ~, ~, offsetAxesHandles] = PlotUtil.multiElementPlot(figHandle, repmat({3.5 * ones(4, 1)}, 5, 1), [0 time(end)], options);
0648             
0649             set(cellfun(@(x) x(1), axesHandles(2:end)), <span class="string">'visible'</span>, <span class="string">'off'</span>);
0650             set(cellfun(@(x) x(1), offsetAxesHandles(2:end)), <span class="string">'visible'</span>, <span class="string">'off'</span>);
0651             set(cellfun(@(x) x(3), axesHandles([3 5])), <span class="string">'visible'</span>, <span class="string">'off'</span>);
0652             set(cellfun(@(x) x(3), offsetAxesHandles([3 5])), <span class="string">'visible'</span>, <span class="string">'off'</span>);
0653             
0654             clear options;
0655             
0656             <span class="comment">%% table</span>
0657             aaSums = sum(aas, 1);
0658             medianNtps = [
0659                 median(ntps(sub2ind(size(ntps), 1 * ones(size(simEndTimes)), ones(size(simEndTimes)), simEndTimes, (1:numel(simEndTimes))')))
0660                 median(ntps(sub2ind(size(ntps), 2 * ones(size(simEndTimes)), ones(size(simEndTimes)), simEndTimes, (1:numel(simEndTimes))')))
0661                 median(ntps(sub2ind(size(ntps), 3 * ones(size(simEndTimes)), ones(size(simEndTimes)), simEndTimes, (1:numel(simEndTimes))')))
0662                 median(ntps(sub2ind(size(ntps), 4 * ones(size(simEndTimes)), ones(size(simEndTimes)), simEndTimes, (1:numel(simEndTimes))')))
0663                 ];
0664             medianNdps = [
0665                 median(ndps(sub2ind(size(ndps), 1 * ones(size(simEndTimes)), ones(size(simEndTimes)), simEndTimes, (1:numel(simEndTimes))')))
0666                 median(ndps(sub2ind(size(ndps), 2 * ones(size(simEndTimes)), ones(size(simEndTimes)), simEndTimes, (1:numel(simEndTimes))')))
0667                 median(ndps(sub2ind(size(ndps), 3 * ones(size(simEndTimes)), ones(size(simEndTimes)), simEndTimes, (1:numel(simEndTimes))')))
0668                 median(ndps(sub2ind(size(ndps), 4 * ones(size(simEndTimes)), ones(size(simEndTimes)), simEndTimes, (1:numel(simEndTimes))')))
0669                 ];
0670             medianNmps = [
0671                 median(nmps(sub2ind(size(nmps), 1 * ones(size(simEndTimes)), ones(size(simEndTimes)), simEndTimes, (1:numel(simEndTimes))')))
0672                 median(nmps(sub2ind(size(nmps), 2 * ones(size(simEndTimes)), ones(size(simEndTimes)), simEndTimes, (1:numel(simEndTimes))')))
0673                 median(nmps(sub2ind(size(nmps), 3 * ones(size(simEndTimes)), ones(size(simEndTimes)), simEndTimes, (1:numel(simEndTimes))')))
0674                 median(nmps(sub2ind(size(nmps), 4 * ones(size(simEndTimes)), ones(size(simEndTimes)), simEndTimes, (1:numel(simEndTimes))')))
0675                 ];
0676             medianAAs = median(aaSums(sub2ind(size(aaSums), ones(size(simEndTimes)), ones(size(simEndTimes)), simEndTimes, (1:numel(simEndTimes))')));
0677             nxps = [
0678                 median(ntps(:, :, 1, :), 4)' medianNtps'
0679                 median(ndps(:, :, 1, :), 4)' medianNdps'
0680                 median(nmps(:, :, 1, :), 4)' medianNmps'
0681                 median(aaSums(:, :, 1, :)) zeros(1, 3) medianAAs zeros(1, 3)
0682                 ];
0683             tabColLabels = {[] <span class="string">'t = 0'</span> [] [] [] [<span class="string">'t = '</span> num2str(median(simEndTimes))] [] [] []};
0684             tabContent = [{[] <span class="string">'A'</span> <span class="string">'C'</span> <span class="string">'G'</span> <span class="string">'U'</span> <span class="string">'A'</span> <span class="string">'C'</span> <span class="string">'G'</span> <span class="string">'U'</span>;}
0685                 {<span class="string">'NTP'</span>; <span class="string">'NDP'</span>; <span class="string">'NMP'</span>; <span class="string">'AAs'</span>} num2cell(nxps)];
0686             tabContent(<span class="keyword">end</span>, [3:5 7:9]) = {[]};
0687             
0688             <span class="comment">%% organize figure data</span>
0689             figData = struct;
0690             figData.simEndTimes = simEndTimes;
0691             figData.simOrder = simOrder;
0692             figData.timeIdx = timeIdx;
0693             figData.time = time;
0694             figData.aas = aas;
0695             figData.ntps = ntps;
0696             figData.ndps = ndps;
0697             figData.nmps = nmps;
0698         <span class="keyword">end</span>
0699         
0700         <a name="_sub7" href="#_subfunctions" class="code">function figData = aminoAcidCounts(figHandle, simBatchDir, selectedSimulations)</a>
0701             import edu.stanford.covert.cell.sim.util.SimulationEnsemble;
0702             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
0703             import edu.stanford.covert.cell.sim.util.PlotUtil;
0704             import edu.stanford.covert.util.ConstantUtil;
0705             
0706             <span class="keyword">if</span> nargin &lt; 2 || isempty(simBatchDir)
0707                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
0708             <span class="keyword">end</span>
0709             <span class="keyword">if</span> nargin &lt; 3 || isempty(selectedSimulations)
0710                 selectedSimulations = SimulationDiskUtil.getCompleteSimulations(simBatchDir);
0711             <span class="keyword">end</span>
0712             
0713             [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep <span class="string">'1'</span>]);
0714             comp = sim.compartment;
0715             m = sim.state(<span class="string">'Metabolite'</span>);
0716             
0717             <span class="comment">%% get data</span>
0718             stateNames = {
0719                 <span class="string">'Time'</span>        <span class="string">'values'</span>  <span class="string">':'</span>                <span class="string">':'</span>
0720                 <span class="string">'Mass'</span>        <span class="string">'cell'</span>    <span class="string">':'</span>                <span class="string">'-sum'</span>
0721                 <span class="string">'Metabolite'</span>  <span class="string">'counts'</span>  m.aminoAcidIndexs  comp.cytosolIndexs
0722                 };
0723             states = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, <span class="string">'extract'</span>, selectedSimulations);
0724             
0725             time = permute(max(states.Time.values, [], 4), [4 3 1 2]) / 3600;
0726             simEndTimes = squeeze(max(states.Time.values, [], 3));
0727             [~, simOrder] = sort(states.Mass.cell(sub2ind(size(states.Mass.cell), ones(size(simEndTimes)), ones(size(simEndTimes)), simEndTimes, (1:numel(simEndTimes))')));
0728             
0729             aminoAcids = full(permute(states.Metabolite.counts, [4 3 1 2]));
0730             <span class="keyword">for</span> i = 1:numel(simEndTimes)
0731                 aminoAcids(i, simEndTimes(i) + 1:<span class="keyword">end</span>, :, :) = NaN;
0732             <span class="keyword">end</span>
0733             
0734             <span class="comment">%% plot data</span>
0735             clf(figHandle);
0736             
0737             options = struct;
0738             options.colorOrder = PlotUtil.getRedGreenColorOrder(simOrder);
0739             options.xdata = time;
0740             options.ylabelStr = {
0741                 m.wholeCellModelIDs(m.aminoAcidIndexs(1:6));
0742                 cell(6, 1);
0743                 m.wholeCellModelIDs(m.aminoAcidIndexs(7:12));
0744                 cell(6, 1);
0745                 m.wholeCellModelIDs(m.aminoAcidIndexs(13:18));
0746                 cell(6, 1);
0747                 [m.wholeCellModelIDs(m.aminoAcidIndexs(19:21)); {[]; []; []}]
0748                 cell(6, 1);
0749                 };
0750             options.ydata = {
0751                 permute(mat2cell(aminoAcids(:, :, 1:6), size(aminoAcids, 1), size(aminoAcids, 2), ones(1, 6)), [3 1 2])
0752                 repmat({[]}, 6, 1)
0753                 permute(mat2cell(aminoAcids(:, :, 7:12), size(aminoAcids, 1), size(aminoAcids, 2), ones(1, 6)), [3 1 2])
0754                 repmat({[]}, 6, 1)
0755                 permute(mat2cell(aminoAcids(:, :, 13:18), size(aminoAcids, 1), size(aminoAcids, 2), ones(1, 6)), [3 1 2])
0756                 repmat({[]}, 6, 1)
0757                 [permute(mat2cell(aminoAcids(:, :, 19:21), size(aminoAcids, 1), size(aminoAcids, 2), ones(1, 3)), [3 1 2]); {[]; []; []}]
0758                 repmat({[]}, 6, 1)
0759                 };
0760             options.colWidths = repmat([4; 1], 4, 1);
0761             options.yAxesLabelWidths = repmat([0.1; 0.01], 4, 1);
0762             [axesHandles, xAxesHandles, ~, offsetAxesHandles] = PlotUtil.multiElementPlot(figHandle, repmat({10 * ones(6, 1)}, 8, 1), [0 time(end)], options);
0763             
0764             clear options;
0765             
0766             delete(axesHandles{7}(4:6));
0767             delete(axesHandles{8}(4:6));
0768             delete(offsetAxesHandles{7}(4:6));
0769             delete(offsetAxesHandles{8}(4:6));
0770             delete(cell2mat(xAxesHandles(2:2:end)));
0771             
0772             <span class="keyword">for</span> i = 1:numel(m.aminoAcidIndexs)
0773                 row = mod(i - 1, 6) + 1;
0774                 col = 2 * ceil(i / 6);
0775                 
0776                 cnts = aminoAcids(sub2ind(size(aminoAcids), (1:numel(simEndTimes))', simEndTimes, i * ones(size(simEndTimes))));
0777                 edges = linspace(min(min(aminoAcids(:, :, i))), max(max(aminoAcids(:, :, i))), 10);
0778                 edges = edges(:);
0779                 freq = histc(cnts, edges);
0780                 
0781                 axesHandle = axesHandles{col}(row);
0782                 offsetAxesHandle = offsetAxesHandles{col}(row);
0783                 
0784                 cla(axesHandle)
0785                 cla(offsetAxesHandle)
0786                 
0787                 barh(axesHandle, edges, freq)
0788                 xlim(axesHandle, [0 max(freq(:))]);
0789                 ylim(axesHandle, [min(min(aminoAcids(:, :, i))) max(max(aminoAcids(:, :, i)))]);
0790                 set(axesHandle, <span class="string">'YTick'</span>, [])
0791                 set(axesHandle, <span class="string">'XTick'</span>, [])
0792                 set(offsetAxesHandle, <span class="string">'YTick'</span>, [])
0793                 set(offsetAxesHandle, <span class="string">'XTick'</span>, [])
0794             <span class="keyword">end</span>
0795             
0796             <span class="comment">%% data</span>
0797             figData = struct;
0798             figData.simOrder = simOrder;
0799             figData.time = time;
0800             figData.aminoAcids = aminoAcids;
0801         <span class="keyword">end</span>
0802         
0803         <a name="_sub8" href="#_subfunctions" class="code">function [figData, tabContent, tabColLabels] = metabolism(figHandle, simBatchDir, selectedSimulations)</a>
0804             import edu.stanford.covert.cell.sim.util.SimulationEnsemble;
0805             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
0806             import edu.stanford.covert.cell.sim.util.PlotUtil;
0807             
0808             <span class="keyword">if</span> nargin &lt; 2 || isempty(simBatchDir)
0809                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
0810             <span class="keyword">end</span>
0811             <span class="keyword">if</span> nargin &lt; 3 || isempty(selectedSimulations)
0812                 selectedSimulations = SimulationDiskUtil.getCompleteSimulations(simBatchDir);
0813             <span class="keyword">end</span>
0814             
0815             [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep <span class="string">'1'</span>]);
0816             pm = sim.state(<span class="string">'ProteinMonomer'</span>);
0817             pc = sim.state(<span class="string">'ProteinComplex'</span>);
0818             met = sim.process(<span class="string">'Metabolism'</span>);
0819             
0820             <span class="comment">%% get data</span>
0821             corrEnzymeFlux = zeros(numel(met.reactionWholeCellModelIDs), numel(selectedSimulations));
0822             corrEnzymeGrowth = zeros(numel(met.enzymeWholeCellModelIDs), numel(selectedSimulations));
0823             corrFluxGrowth = zeros(numel(met.reactionWholeCellModelIDs), numel(selectedSimulations));
0824             stateNames = {
0825                 <span class="string">'MetabolicReaction'</span>  <span class="string">'fluxs'</span>   <span class="string">':'</span>                                            <span class="string">':'</span>
0826                 <span class="string">'MetabolicReaction'</span>  <span class="string">'growth'</span>  <span class="string">':'</span>                                            <span class="string">':'</span>
0827                 <span class="string">'ProteinMonomer'</span>     <span class="string">'counts'</span>  pm.matureIndexs(met.enzymeMonomerGlobalIndexs) <span class="string">'-sum'</span>
0828                 <span class="string">'ProteinComplex'</span>     <span class="string">'counts'</span>  pc.matureIndexs(met.enzymeComplexGlobalIndexs) <span class="string">'-sum'</span>
0829                 };
0830             <span class="keyword">for</span> i = 1:numel(selectedSimulations)
0831                 states = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, <span class="string">'extract'</span>, selectedSimulations(i));
0832                 <a href="#_sub2" class="code" title="subfunction figData = growth(figHandle, simBatchDir, selectedSimulations)">growth</a> = permute(states.MetabolicReaction.growth, [1 3 2]);
0833                 enzymes = zeros(numel(met.enzymeWholeCellModelIDs), size(states.ProteinMonomer.counts, 3));
0834                 enzymes(met.enzymeMonomerLocalIndexs, :, :) = full(permute(states.ProteinMonomer.counts, [1 3 2]));
0835                 enzymes(met.enzymeComplexLocalIndexs, :, :) = full(permute(states.ProteinComplex.counts, [1 3 2]));
0836                 fluxs = permute(states.MetabolicReaction.fluxs, [1 3 2]);
0837                 
0838                 <span class="keyword">for</span> j = 1:numel(met.reactionIndexs_fba)
0839                     k = met.reactionIndexs_fba(j);
0840                     corrEnzymeFlux(k, i) = corr(fluxs(k, :)', (met.reactionCatalysisMatrix(k, :) * enzymes)');
0841                 <span class="keyword">end</span>
0842                 corrEnzymeGrowth(:, i) = corr(enzymes', <a href="#_sub2" class="code" title="subfunction figData = growth(figHandle, simBatchDir, selectedSimulations)">growth</a>');
0843                 corrFluxGrowth(:, i) = corr(fluxs', <a href="#_sub2" class="code" title="subfunction figData = growth(figHandle, simBatchDir, selectedSimulations)">growth</a>');
0844                 
0845                 clear states;
0846             <span class="keyword">end</span>
0847             
0848             corrEnzymeFlux(isnan(corrEnzymeFlux)) = 0;
0849             corrEnzymeGrowth(isnan(corrEnzymeGrowth)) = 0;
0850             corrFluxGrowth(isnan(corrFluxGrowth)) = 0;
0851             
0852             corrEnzymeFluxMean = nanmean(corrEnzymeFlux, 2);
0853             corrEnzymeGrowthMean = nanmean(corrEnzymeGrowth, 2);
0854             corrFluxGrowthMean = nanmean(corrFluxGrowth, 2);
0855             corrEnzymeFluxStd = nanstd(corrEnzymeFlux, [], 2);
0856             corrEnzymeGrowthStd = nanstd(corrEnzymeGrowth, [], 2);
0857             corrFluxGrowthStd = nanstd(corrFluxGrowth, [], 2);
0858             
0859             [~, enzIdx] = max(corrEnzymeGrowthMean);
0860             [~, simOrder] = sort(corrEnzymeGrowth(enzIdx, :));
0861             
0862             clear corrEnzymeFlux corrEnzymeGrowth corrFluxGrowth;
0863             
0864             <span class="comment">%% get temporal data</span>
0865             rxnIdx = find(met.reactionCatalysisMatrix(:, enzIdx));
0866             monIdx = pm.matureIndexs(met.enzymeMonomerGlobalIndexs(met.enzymeMonomerLocalIndexs == enzIdx));
0867             cpxIdx = pc.matureIndexs(met.enzymeComplexGlobalIndexs(met.enzymeComplexLocalIndexs == enzIdx));
0868             
0869             stateNames = {
0870                 <span class="string">'Time'</span>               <span class="string">'values'</span>  <span class="string">':'</span>     <span class="string">':'</span>
0871                 <span class="string">'MetabolicReaction'</span>  <span class="string">'fluxs'</span>   rxnIdx  <span class="string">':'</span>
0872                 };
0873             <span class="keyword">if</span> ~isempty(monIdx)
0874                 stateNames = [stateNames; {<span class="string">'ProteinMonomer'</span> <span class="string">'counts'</span> monIdx <span class="string">'-sum'</span>}];
0875                 states = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, <span class="string">'extract'</span>, selectedSimulations);
0876                 enzymes = full(permute(states.ProteinMonomer.counts, [3 4 1 2]));
0877             <span class="keyword">else</span>
0878                 stateNames = [stateNames; {<span class="string">'ProteinComplex'</span> <span class="string">'counts'</span> cpxIdx <span class="string">'-sum'</span>}];
0879                 states = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, <span class="string">'extract'</span>, selectedSimulations);
0880                 enzymes = full(permute(states.ProteinComplex.counts, [3 4 1 2]));
0881             <span class="keyword">end</span>
0882             time = permute(max(states.Time.values, [], 4), [3 4 1 2]) / 3600;
0883             simEndTimes = squeeze(max(states.Time.values, [], 3));
0884             fluxs = full(permute(states.MetabolicReaction.fluxs, [3 4 1 2]));
0885             
0886             clear states;
0887             
0888             <span class="keyword">for</span> i = 1:numel(simEndTimes)
0889                 enzymes(simEndTimes(i)+1:<span class="keyword">end</span>, i) = NaN;
0890                 fluxs(simEndTimes(i)+1:<span class="keyword">end</span>, i) = NaN;
0891             <span class="keyword">end</span>
0892             
0893             colorOrder = PlotUtil.getRedGreenColorOrder(simOrder);
0894             
0895             <span class="comment">%% plot</span>
0896             clf(figHandle);
0897             
0898             axesHandles = repmat({zeros(4, 1)}, 2, 1);
0899             
0900             <span class="comment">%enzyme-fluxs</span>
0901             axesHandle = subplot(4, 2, 1);
0902             axesHandles{1}(1) = axesHandle;
0903             [~, order] = sort(corrEnzymeFluxMean);
0904             errorbar((1:numel(corrEnzymeFluxMean))', corrEnzymeFluxMean(order), corrEnzymeFluxStd(order), <span class="keyword">...</span>
0905                 <span class="string">'Parent'</span>, axesHandle, <span class="string">'LineStyle'</span>, <span class="string">'none'</span>, <span class="string">'Marker'</span>, <span class="string">'.'</span>);
0906             xlabel(axesHandle, <span class="string">'Flux'</span>, <span class="string">'FontSize'</span>, 8);
0907             ylabel(axesHandle, {<span class="string">'Corr'</span> <span class="string">'(Flux, Enzyme)'</span>}, <span class="string">'FontSize'</span>, 8);
0908             xlim(axesHandle, [0.5 numel(corrEnzymeFluxMean) + 0.5]);
0909             ylim(axesHandle, [min(corrEnzymeFluxMean - corrEnzymeFluxStd) max(corrEnzymeFluxMean - corrEnzymeFluxStd)]);
0910             
0911             axesHandle = subplot(4, 2, 2);
0912             axesHandles{2}(1) = axesHandle;
0913             hist(axesHandle, corrEnzymeFluxMean);
0914             xlabel(axesHandle, <span class="string">'Corr (Flux, Enzyme)'</span>, <span class="string">'FontSize'</span>, 8);
0915             ylabel(axesHandle, <span class="string">'Freq'</span>, <span class="string">'FontSize'</span>, 8);
0916             
0917             <span class="comment">%fluxes-growth</span>
0918             axesHandle = subplot(4, 2, 3);
0919             axesHandles{1}(2) = axesHandle;
0920             [~, order] = sort(corrFluxGrowthMean);
0921             errorbar((1:numel(corrFluxGrowthMean))', corrFluxGrowthMean(order), corrFluxGrowthStd(order), <span class="keyword">...</span>
0922                 <span class="string">'Parent'</span>, axesHandle, <span class="string">'LineStyle'</span>, <span class="string">'none'</span>, <span class="string">'Marker'</span>, <span class="string">'.'</span>);
0923             xlabel(axesHandle, <span class="string">'Flux'</span>, <span class="string">'FontSize'</span>, 8);
0924             ylabel(axesHandle, {<span class="string">'Corr'</span> <span class="string">'(Flux, Growth)'</span>}, <span class="string">'FontSize'</span>, 8);
0925             xlim(axesHandle, [0.5 numel(corrFluxGrowthMean) + 0.5]);
0926             ylim(axesHandle, [min(corrFluxGrowthMean - corrFluxGrowthStd) max(corrFluxGrowthMean - corrFluxGrowthStd)]);
0927             
0928             axesHandle = subplot(4, 2, 4);
0929             axesHandles{2}(2) = axesHandle;
0930             hist(axesHandle, corrFluxGrowthMean);
0931             xlabel(axesHandle, <span class="string">'Corr (Flux, Growth)'</span>, <span class="string">'FontSize'</span>, 8);
0932             ylabel(axesHandle, <span class="string">'Freq'</span>, <span class="string">'FontSize'</span>, 8);
0933             
0934             <span class="comment">%enzymes-growth</span>
0935             axesHandle = subplot(4, 2, 5);
0936             axesHandles{1}(3) = axesHandle;
0937             [~, order] = sort(corrEnzymeGrowthMean);
0938             errorbar((1:numel(corrEnzymeGrowthMean))', corrEnzymeGrowthMean(order), corrEnzymeGrowthStd(order), <span class="keyword">...</span>
0939                 <span class="string">'Parent'</span>, axesHandle, <span class="string">'LineStyle'</span>, <span class="string">'none'</span>, <span class="string">'Marker'</span>, <span class="string">'.'</span>);
0940             xlabel(axesHandle, <span class="string">'Flux'</span>, <span class="string">'FontSize'</span>, 8);
0941             ylabel(axesHandle, {<span class="string">'Corr'</span> <span class="string">'(Growth, Enzyme)'</span>}, <span class="string">'FontSize'</span>, 8);
0942             xlim(axesHandle, [0.5 numel(corrEnzymeGrowthMean) + 0.5]);
0943             ylim(axesHandle, [min(corrEnzymeGrowthMean - corrEnzymeGrowthStd) max(corrEnzymeGrowthMean - corrEnzymeGrowthStd)]);
0944             
0945             axesHandle = subplot(4, 2, 6);
0946             axesHandles{2}(3) = axesHandle;
0947             hist(axesHandle, corrEnzymeGrowthMean);
0948             xlabel(axesHandle, <span class="string">'Corr (Growth, Enzyme)'</span>, <span class="string">'FontSize'</span>, 8);
0949             ylabel(axesHandle, <span class="string">'Freq'</span>, <span class="string">'FontSize'</span>, 8);
0950             
0951             <span class="comment">%flux-time</span>
0952             axesHandle = subplot(4, 2, 7);
0953             axesHandles{1}(4) = axesHandle;
0954             hold(axesHandle, <span class="string">'on'</span>);
0955             set(axesHandle, <span class="string">'ColorOrder'</span>, colorOrder);
0956             PlotUtil.plotLine(axesHandle, time, fluxs, true, true);
0957             xlabel(axesHandle, <span class="string">'Time (h)'</span>, <span class="string">'FontSize'</span>, 8);
0958             ylabel(axesHandle, [met.reactionWholeCellModelIDs(rxnIdx) <span class="string">'Flux'</span>], <span class="keyword">...</span>
0959                 <span class="string">'FontSize'</span>, 8, <span class="string">'Interpreter'</span>, <span class="string">'none'</span>);
0960             
0961             <span class="comment">%enzyme-time</span>
0962             axesHandle = subplot(4, 2, 8);
0963             axesHandles{2}(4) = axesHandle;
0964             hold(axesHandle, <span class="string">'on'</span>);
0965             set(axesHandle, <span class="string">'ColorOrder'</span>, colorOrder);
0966             PlotUtil.plotLine(axesHandle, time, enzymes, true, true);
0967             xlabel(axesHandle, <span class="string">'Time (h)'</span>, <span class="string">'FontSize'</span>, 8);
0968             ylabel(axesHandle, [met.enzymeWholeCellModelIDs(enzIdx) <span class="string">'Enzyme'</span>], <span class="keyword">...</span>
0969                 <span class="string">'FontSize'</span>, 8, <span class="string">'Interpreter'</span>, <span class="string">'none'</span>);
0970             
0971             <span class="comment">%% format plots</span>
0972             <span class="keyword">for</span> i = 1:numel(axesHandles)
0973                 <span class="keyword">for</span> j = 1:numel(axesHandles{i})
0974                     axesHandle = axesHandles{i}(j);
0975                     set(axesHandle, <span class="string">'FontSize'</span>, 6);
0976                     set(get(axesHandle, <span class="string">'YLabel'</span>), <span class="string">'Units'</span>, <span class="string">'normalized'</span>)
0977                 <span class="keyword">end</span>
0978             <span class="keyword">end</span>
0979             PlotUtil.alignYAxesLabels(axesHandles);
0980             
0981             <span class="comment">%% layout table</span>
0982             tabColLabels = {
0983                 <span class="string">'Reaction'</span> <span class="string">'Mean Enzyme Correlation'</span> <span class="string">'Std Enzyme Correlation'</span> <span class="keyword">...</span>
0984                 [] <span class="keyword">...</span>
0985                 <span class="string">'Reaction'</span> <span class="string">'Mean Growth Correlation'</span> <span class="string">'Std Growth Correlation'</span> <span class="keyword">...</span>
0986                 [] <span class="keyword">...</span>
0987                 <span class="string">'Enzyme'</span>   <span class="string">'Mean Growth Correlation'</span> <span class="string">'Std Growth Correlation'</span> <span class="keyword">...</span>
0988                 };
0989             tabContent = [
0990                 met.reactionWholeCellModelIDs num2cell(corrEnzymeFluxMean) num2cell(corrEnzymeFluxStd) <span class="keyword">...</span>
0991                 cell(size(met.reactionWholeCellModelIDs)) <span class="keyword">...</span>
0992                 met.reactionWholeCellModelIDs num2cell(corrFluxGrowthMean) num2cell(corrFluxGrowthStd) <span class="keyword">...</span>
0993                 cell(size(met.reactionWholeCellModelIDs)) <span class="keyword">...</span>
0994                 [met.enzymeWholeCellModelIDs num2cell(corrEnzymeGrowthMean) num2cell(corrEnzymeGrowthStd); cell(numel(met.reactionWholeCellModelIDs) - numel(met.enzymeWholeCellModelIDs), 3)] <span class="keyword">...</span>
0995                 ];
0996             
0997             <span class="comment">%% organize figure data</span>
0998             figData = struct;
0999             figData.simOrder = simOrder;
1000             figData.time = time;
1001             figData.enzymes = enzymes;
1002             figData.fluxs = fluxs;
1003         <span class="keyword">end</span>
1004         
1005         <a name="_sub9" href="#_subfunctions" class="code">function figData = metaboliteConcentrations(figHandle, simBatchDir, selectedSimulations)</a>
1006             import edu.stanford.covert.cell.sim.util.SimulationEnsemble;
1007             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
1008             import edu.stanford.covert.cell.sim.util.PlotUtil;
1009             import edu.stanford.covert.util.ConstantUtil;
1010             
1011             <span class="keyword">if</span> nargin &lt; 2 || isempty(simBatchDir)
1012                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
1013             <span class="keyword">end</span>
1014             <span class="keyword">if</span> nargin &lt; 3 || isempty(selectedSimulations)
1015                 selectedSimulations = SimulationDiskUtil.getCompleteSimulations(simBatchDir);
1016             <span class="keyword">end</span>
1017             
1018             [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep <span class="string">'1'</span>]);
1019             comp = sim.compartment;
1020             m = sim.state(<span class="string">'Metabolite'</span>);
1021             
1022             <span class="comment">%% get data</span>
1023             metIdxs = [m.ntpIndexs; m.ndpIndexs; m.nmpIndexs; m.dntpIndexs; m.aminoAcidIndexs; m.phosphateIndexs; m.diphosphateIndexs; m.hydrogenIndexs];
1024             stateNames = {
1025                 <span class="string">'Geometry'</span>    <span class="string">'volume'</span>  <span class="string">':'</span>      <span class="string">':'</span>
1026                 <span class="string">'Metabolite'</span>  <span class="string">'counts'</span>  metIdxs  comp.cytosolIndexs
1027                 };
1028             meanMetConcs = zeros(numel(metIdxs), 1, 1, numel(selectedSimulations));
1029             <span class="keyword">for</span> i = 1:numel(selectedSimulations)
1030                 states = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, <span class="string">'extract'</span>, selectedSimulations(i));
1031                 meanMetConcs(:, :, :, i) = mean(states.Metabolite.counts ./ states.Geometry.volume(ones(size(metIdxs)), :, :), 3) / ConstantUtil.nAvogadro * 1000;
1032                 
1033                 clear states;
1034             <span class="keyword">end</span>
1035             
1036             meanConcs = mean(meanMetConcs, 4);
1037             stdConcs = std(meanMetConcs, [], 4);
1038             
1039             clear meanMetConcs;
1040             
1041             <span class="comment">%% plot</span>
1042             clf(figHandle);
1043             
1044             axesHandle = subplot(1, 1, 1, <span class="string">'Position'</span>, [0.1300    0.1500    0.7750    0.8150]);
1045             errorbar((1:numel(metIdxs))', meanConcs, meanConcs - max(1e-3, meanConcs - stdConcs), stdConcs, <span class="keyword">...</span>
1046                 <span class="string">'Parent'</span>, axesHandle, <span class="string">'LineStyle'</span>, <span class="string">'none'</span>, <span class="string">'Marker'</span>, <span class="string">'.'</span>);
1047             set(axesHandle, <span class="string">'YScale'</span>, <span class="string">'log'</span>);
1048             xlim(axesHandle, [0.5  numel(metIdxs) + 0.5]);
1049             set(axesHandle, <span class="keyword">...</span>
1050                 <span class="string">'XTick'</span>, 1:numel(metIdxs), <span class="keyword">...</span>
1051                 <span class="string">'XTickLabel'</span>, [], <span class="keyword">...</span>
1052                 <span class="string">'FontSize'</span>, 8, <span class="keyword">...</span>
1053                 <span class="string">'TickDir'</span>, <span class="string">'out'</span>);
1054             tick2text(axesHandle, <span class="string">'axis'</span>, <span class="string">'x'</span>);
1055             <span class="keyword">for</span> i = 1:numel(metIdxs)
1056                 text((i-0.5) / numel(metIdxs), -0.02, m.wholeCellModelIDs{metIdxs(i)}, <span class="keyword">...</span>
1057                     <span class="string">'Parent'</span>, axesHandle, <span class="keyword">...</span>
1058                     <span class="string">'FontSize'</span>, 6, <span class="keyword">...</span>
1059                     <span class="string">'Rotation'</span>, 270, <span class="keyword">...</span>
1060                     <span class="string">'VerticalAlignment'</span>, <span class="string">'middle'</span>, <span class="keyword">...</span>
1061                     <span class="string">'HorizontalAlignment'</span>, <span class="string">'left'</span>, <span class="keyword">...</span>
1062                     <span class="string">'Units'</span>, <span class="string">'normalized'</span>);
1063             <span class="keyword">end</span>
1064             axesPos = get(axesHandle, <span class="string">'Position'</span>);
1065             
1066             annotation(figHandle, <span class="string">'line'</span>, axesPos(1) + axesPos(3) * ((-0.5 + [1 4] + [-0.25 0.35]) / numel(metIdxs)), axesPos(2) + axesPos(4) * (-0.09 * [1 1]), <span class="keyword">...</span>
1067                 <span class="string">'Units'</span>, <span class="string">'normalized'</span>);
1068             annotation(figHandle, <span class="string">'line'</span>, axesPos(1) + axesPos(3) * ((-0.5 + [5 8] + [-0.25 0.35]) / numel(metIdxs)), axesPos(2) + axesPos(4) * (-0.09 * [1 1]), <span class="keyword">...</span>
1069                 <span class="string">'Units'</span>, <span class="string">'normalized'</span>);
1070             annotation(figHandle, <span class="string">'line'</span>, axesPos(1) + axesPos(3) * ((-0.5 + [9 12] + [-0.25 0.35]) / numel(metIdxs)), axesPos(2) + axesPos(4) * (-0.09 * [1 1]), <span class="keyword">...</span>
1071                 <span class="string">'Units'</span>, <span class="string">'normalized'</span>);
1072             annotation(figHandle, <span class="string">'line'</span>, axesPos(1) + axesPos(3) * ((-0.5 + [13 16] + [-0.25 0.35]) / numel(metIdxs)), axesPos(2) + axesPos(4) * (-0.09 * [1 1]), <span class="keyword">...</span>
1073                 <span class="string">'Units'</span>, <span class="string">'normalized'</span>);
1074             annotation(figHandle, <span class="string">'line'</span>, axesPos(1) + axesPos(3) * ((-0.5 + [17 37] + [-0.25 0.35]) / numel(metIdxs)), axesPos(2) + axesPos(4) * (-0.09 * [1 1]), <span class="keyword">...</span>
1075                 <span class="string">'Units'</span>, <span class="string">'normalized'</span>);
1076             annotation(figHandle, <span class="string">'line'</span>, axesPos(1) + axesPos(3) * ((-0.5 + [38 40] + [-0.25 0.35]) / numel(metIdxs)), axesPos(2) + axesPos(4) * (-0.09 * [1 1]), <span class="keyword">...</span>
1077                 <span class="string">'Units'</span>, <span class="string">'normalized'</span>);
1078             
1079             text(2 / numel(metIdxs), -0.092, <span class="string">'NTPs'</span>, <span class="keyword">...</span>
1080                 <span class="string">'Parent'</span>, axesHandle, <span class="keyword">...</span>
1081                 <span class="string">'FontSize'</span>, 6, <span class="keyword">...</span>
1082                 <span class="string">'VerticalAlignment'</span>, <span class="string">'top'</span>, <span class="keyword">...</span>
1083                 <span class="string">'HorizontalAlignment'</span>, <span class="string">'center'</span>, <span class="keyword">...</span>
1084                 <span class="string">'Units'</span>, <span class="string">'normalized'</span>);
1085             text(6 / numel(metIdxs), -0.092, <span class="string">'NDPs'</span>, <span class="keyword">...</span>
1086                 <span class="string">'Parent'</span>, axesHandle, <span class="keyword">...</span>
1087                 <span class="string">'FontSize'</span>, 6, <span class="keyword">...</span>
1088                 <span class="string">'VerticalAlignment'</span>, <span class="string">'top'</span>, <span class="keyword">...</span>
1089                 <span class="string">'HorizontalAlignment'</span>, <span class="string">'center'</span>, <span class="keyword">...</span>
1090                 <span class="string">'Units'</span>, <span class="string">'normalized'</span>);
1091             text(10 / numel(metIdxs), -0.092, <span class="string">'NMPs'</span>, <span class="keyword">...</span>
1092                 <span class="string">'Parent'</span>, axesHandle, <span class="keyword">...</span>
1093                 <span class="string">'FontSize'</span>, 6, <span class="keyword">...</span>
1094                 <span class="string">'VerticalAlignment'</span>, <span class="string">'top'</span>, <span class="keyword">...</span>
1095                 <span class="string">'HorizontalAlignment'</span>, <span class="string">'center'</span>, <span class="keyword">...</span>
1096                 <span class="string">'Units'</span>, <span class="string">'normalized'</span>);
1097             text(14 / numel(metIdxs), -0.092, <span class="string">'dNTPs'</span>, <span class="keyword">...</span>
1098                 <span class="string">'Parent'</span>, axesHandle, <span class="keyword">...</span>
1099                 <span class="string">'FontSize'</span>, 6, <span class="keyword">...</span>
1100                 <span class="string">'VerticalAlignment'</span>, <span class="string">'top'</span>, <span class="keyword">...</span>
1101                 <span class="string">'HorizontalAlignment'</span>, <span class="string">'center'</span>, <span class="keyword">...</span>
1102                 <span class="string">'Units'</span>, <span class="string">'normalized'</span>);
1103             text(26.5 / numel(metIdxs), -0.092, <span class="string">'AAs'</span>, <span class="keyword">...</span>
1104                 <span class="string">'Parent'</span>, axesHandle, <span class="keyword">...</span>
1105                 <span class="string">'FontSize'</span>, 6, <span class="keyword">...</span>
1106                 <span class="string">'VerticalAlignment'</span>, <span class="string">'top'</span>, <span class="keyword">...</span>
1107                 <span class="string">'HorizontalAlignment'</span>, <span class="string">'center'</span>, <span class="keyword">...</span>
1108                 <span class="string">'Units'</span>, <span class="string">'normalized'</span>);
1109             text(38.5 / numel(metIdxs), -0.092, <span class="string">'Other'</span>, <span class="keyword">...</span>
1110                 <span class="string">'Parent'</span>, axesHandle, <span class="keyword">...</span>
1111                 <span class="string">'FontSize'</span>, 6, <span class="keyword">...</span>
1112                 <span class="string">'VerticalAlignment'</span>, <span class="string">'top'</span>, <span class="keyword">...</span>
1113                 <span class="string">'HorizontalAlignment'</span>, <span class="string">'center'</span>, <span class="keyword">...</span>
1114                 <span class="string">'Units'</span>, <span class="string">'normalized'</span>);
1115             
1116             xlabel(axesHandle, <span class="string">'Metabolite'</span>, <span class="string">'FontSize'</span>, 12);
1117             set(get(axesHandle, <span class="string">'xlabel'</span>), <span class="string">'Position'</span>, [0.5 -0.12 1], <span class="string">'Units'</span>, <span class="string">'normalized'</span>, <span class="string">'HorizontalAlignment'</span>, <span class="string">'center'</span>, <span class="string">'VerticalAlignment'</span>, <span class="string">'top'</span>);
1118             ylabel(axesHandle, <span class="string">'Concentration (mM)'</span>, <span class="string">'FontSize'</span>, 12);
1119             
1120             <span class="comment">%% organize figure data</span>
1121             figData = struct;
1122             figData.meanConcs = meanConcs;
1123             figData.stdConcs = stdConcs;
1124         <span class="keyword">end</span>
1125         
1126         <a name="_sub10" href="#_subfunctions" class="code">function figData = translation(figHandle, simBatchDir, selectedSimulations)</a>
1127             import edu.stanford.covert.cell.sim.util.PlotUtil;
1128             import edu.stanford.covert.cell.sim.util.SimulationEnsemble;
1129             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
1130             import edu.stanford.covert.util.ConstantUtil;
1131             
1132             <span class="keyword">if</span> nargin &lt; 2 || isempty(simBatchDir)
1133                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
1134             <span class="keyword">end</span>
1135             <span class="keyword">if</span> nargin &lt; 3 || isempty(selectedSimulations)
1136                 selectedSimulations = SimulationDiskUtil.getCompleteSimulations(simBatchDir);
1137             <span class="keyword">end</span>
1138             
1139             [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep <span class="string">'1'</span>]);
1140             comp = sim.compartment;
1141             g = sim.gene;
1142             
1143             t = sim.state(<span class="string">'Time'</span>);
1144             mass = sim.state(<span class="string">'Mass'</span>);
1145             m = sim.state(<span class="string">'Metabolite'</span>);
1146             rna = sim.state(<span class="string">'Rna'</span>);
1147             pm = sim.state(<span class="string">'ProteinMonomer'</span>);
1148             pc = sim.state(<span class="string">'ProteinComplex'</span>);
1149             rib = sim.state(<span class="string">'Ribosome'</span>);
1150             pol = sim.state(<span class="string">'Polypeptide'</span>);
1151             
1152             tl = sim.process(<span class="string">'Translation'</span>);
1153             ta = sim.process(<span class="string">'tRNAAminoacylation'</span>);
1154             
1155             expTotRnas = mass.initialFractionNTPsInRNAs * mass.dryWeightFractionRNA * mass.cellInitialDryWeight * rna.expression(rna.matureIndexs) / <span class="keyword">...</span>
1156                 (rna.expression(rna.matureIndexs)' * rna.molecularWeights(rna.matureIndexs) / ConstantUtil.nAvogadro);
1157             monExp = (rna.matureRNAGeneComposition(g.mRNAIndexs, :) * rna.expression(rna.matureIndexs)) ./ <span class="keyword">...</span>
1158                 (log(2) / t.cellCycleLength + pm.decayRates(pm.matureIndexs));
1159             expTotMons = mass.initialFractionAAsInMonomers * mass.dryWeightFractionProtein * mass.cellInitialDryWeight * <span class="keyword">...</span>
1160                 monExp / (monExp' * pm.molecularWeights(pm.matureIndexs) / ConstantUtil.nAvogadro);
1161             expTotCpxs = min(repmat(expTotMons, [1 numel(pc.matureIndexs)]) ./ sum(pc.proteinComplexComposition(g.mRNAIndexs, :, :), 3), [], 1);
1162             
1163             <span class="comment">%% get data</span>
1164             <span class="comment">%mass, energy</span>
1165             ensemble = SimulationEnsemble(simBatchDir, {<span class="string">'mass'</span> <span class="string">'mass'</span>; <span class="string">'atp'</span> <span class="string">'atp'</span>}, [], selectedSimulations);
1166             time = ensemble.stateData.time / 3600;
1167             mass = ensemble.stateData.values(ensemble.getPropertyIndices(<span class="string">'mass'</span>), :, :, :) * sim.state(<span class="string">'Mass'</span>).cellInitialDryWeight / (1-sim.state(<span class="string">'Mass'</span>).fractionWetWeight) * 1e15;
1168             simEndTimes = ensemble.stateData.simulationEndTimes;
1169             
1170             [~, simOrder] = sort(mass(:, 1, min(simEndTimes), :));
1171             simOrder = simOrder(:);
1172             
1173             <span class="comment">%cleanup</span>
1174             clear ensemble mass;
1175             
1176             <span class="comment">%metabolites, rna, proteins</span>
1177             monIdxs = [tl.enzymeMonomerGlobalIndexs; ta.enzymeMonomerGlobalIndexs];
1178             cpxIdxs = [tl.enzymeComplexGlobalIndexs; ta.enzymeComplexGlobalIndexs];
1179             stateNames = {
1180                 <span class="string">'Metabolite'</span>     <span class="string">'counts'</span> [m.ntpIndexs([1 3]); m.ndpIndexs([1 3]); m.nmpIndexs(1); m.aminoAcidIndexs; m.getIndexs(<span class="string">'FTHF10'</span>)] comp.cytosolIndexs
1181                 <span class="string">'Rna'</span>            <span class="string">'counts'</span> [rna.matureIndexs(rna.matureTRNAIndexs); rna.aminoacylatedIndexs(rna.matureTRNAIndexs)] comp.cytosolIndexs
1182                 <span class="string">'ProteinMonomer'</span> <span class="string">'counts'</span> [pm.matureIndexs(monIdxs) pm.boundIndexs(monIdxs)] comp.cytosolIndexs
1183                 <span class="string">'ProteinComplex'</span> <span class="string">'counts'</span> [pc.matureIndexs(cpxIdxs) pc.boundIndexs(cpxIdxs)] comp.cytosolIndexs
1184                 };
1185             states = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, <span class="string">'extract'</span>, selectedSimulations);
1186             
1187             tmp = full(states.Metabolite.counts);
1188             atp = permute(tmp(1, :, :, :), [4 3 1 2]);
1189             gtp = permute(tmp(2, :, :, :), [4 3 1 2]);
1190             adp = permute(tmp(3, :, :, :), [4 3 1 2]);
1191             gdp = permute(tmp(4, :, :, :), [4 3 1 2]);
1192             amp = permute(tmp(5, :, :, :), [4 3 1 2]);
1193             aas = permute(sum(tmp(6:26, :, :, :), 1), [4 3 1 2]);
1194             fthf10 = permute(sum(tmp(27:<span class="keyword">end</span>, :, :, :), 1), [4 3 1 2]);
1195             
1196             monCnts = full(states.ProteinMonomer.counts);
1197             cpxCnts = full(states.ProteinComplex.counts);
1198             
1199             enzProcesses = [
1200                 repmat({<span class="string">'TL'</span>}, numel(tl.enzymeWholeCellModelIDs), 1)
1201                 repmat({<span class="string">'TA'</span>}, numel(ta.enzymeWholeCellModelIDs)-numel(ta.enzymeRNALocalIndexs), 1)
1202                 ];
1203             enzProcessIdxs = [
1204                 (1:numel(tl.enzymeWholeCellModelIDs))'
1205                 (1:numel(ta.enzymeWholeCellModelIDs)-numel(ta.enzymeRNALocalIndexs))'
1206                 ];
1207             
1208             enzymes = zeros(0, 1, size(monCnts, 3), size(monCnts, 4));
1209             expEnzymes = zeros(0, 1);
1210             
1211             tmp = zeros(numel(tl.enzymes), 1, size(monCnts, 3), size(monCnts, 4));
1212             tmp(tl.enzymeMonomerLocalIndexs, :, :, :) = monCnts(1:numel(tl.enzymeMonomerGlobalIndexs), :, :, :);
1213             tmp(tl.enzymeComplexLocalIndexs, :, :, :) = cpxCnts(1:numel(tl.enzymeComplexGlobalIndexs), :, :, :);
1214             tmp2 = zeros(numel(tl.enzymes), 1);
1215             tmp2(tl.enzymeMonomerLocalIndexs) = expTotMons(tl.enzymeMonomerGlobalIndexs);
1216             tmp2(tl.enzymeComplexLocalIndexs) = expTotCpxs(tl.enzymeComplexGlobalIndexs);
1217             enzymes = [enzymes; tmp];
1218             expEnzymes = [expEnzymes; tmp2];
1219             
1220             tmp = zeros(numel(ta.enzymes), 1, size(monCnts, 3), size(monCnts, 4));
1221             tmp(ta.enzymeMonomerLocalIndexs, :, :, :) = monCnts(end-numel(ta.enzymeMonomerGlobalIndexs)+1:<span class="keyword">end</span>, :, :, :);
1222             tmp(ta.enzymeComplexLocalIndexs, :, :, :) = cpxCnts(end-numel(ta.enzymeComplexGlobalIndexs)+1:<span class="keyword">end</span>, :, :, :);
1223             tmp2 = zeros(numel(ta.enzymes), 1);
1224             tmp2(ta.enzymeMonomerLocalIndexs) = expTotMons(ta.enzymeMonomerGlobalIndexs);
1225             tmp2(ta.enzymeComplexLocalIndexs) = expTotCpxs(ta.enzymeComplexGlobalIndexs);
1226             tmp(ta.enzymeRNALocalIndexs, :, :, :) = [];
1227             tmp2(ta.enzymeRNALocalIndexs) = [];
1228             enzymes = [enzymes; tmp];
1229             expEnzymes = [expEnzymes; tmp2];
1230             
1231             expEnzymes = repmat(expEnzymes, [1 1 numel(time)-1 1]) .* exp(log(2) * repmat(permute(time(2:end),[1 3 2]), size(expEnzymes, 1), 1) / (t.cellCycleLength/3600));
1232             
1233             trnas = states.Rna.counts(1:end/2, :, :, :);
1234             aatrnas = states.Rna.counts(end/2+1:<span class="keyword">end</span>, :, :, :);
1235             expTRNAs = repmat(expTotRnas(rna.matureTRNAIndexs), [1 1 numel(time)-1 1]) .* exp(log(2) * repmat(permute(time(2:end),[1 3 2]), size(rna.matureTRNAIndexs, 1), 1) / (t.cellCycleLength/3600));
1236             
1237             <span class="keyword">for</span> i = 1:numel(simEndTimes)
1238                 atp(i, simEndTimes(i)+1:end) = NaN;
1239                 gtp(i, simEndTimes(i)+1:end) = NaN;
1240                 adp(i, simEndTimes(i)+1:end) = NaN;
1241                 gdp(i, simEndTimes(i)+1:end) = NaN;
1242                 amp(i, simEndTimes(i)+1:end) = NaN;
1243                 aas(i, simEndTimes(i)+1:end) = NaN;
1244                 fthf10(i, simEndTimes(i)+1:end) = NaN;
1245                 enzymes(:, :, simEndTimes(i)+1:<span class="keyword">end</span>, i) = NaN;
1246                 trnas(:, :, simEndTimes(i)+1:<span class="keyword">end</span>, i) = NaN;
1247                 aatrnas(:, :, simEndTimes(i)+1:<span class="keyword">end</span>, i) = NaN;
1248             <span class="keyword">end</span>
1249             
1250             <span class="comment">%cleanup</span>
1251             clear states rnaCnts monCnts cpxCnts;
1252             
1253             <span class="comment">%ribosomes</span>
1254             pols = NaN(4, 1, numel(time)-1, numel(selectedSimulations));
1255             stateNames = {
1256                 <span class="string">'Ribosome'</span>  <span class="string">'states'</span>          <span class="string">':'</span> <span class="string">':'</span>
1257                 <span class="string">'Ribosome'</span>  <span class="string">'boundMRNAs'</span>      <span class="string">':'</span> <span class="string">':'</span>
1258                 <span class="string">'Ribosome'</span>  <span class="string">'mRNAPositions'</span>   <span class="string">':'</span> <span class="string">':'</span>
1259                 <span class="string">'Ribosome'</span>  <span class="string">'tmRNAPositions'</span>  <span class="string">':'</span> <span class="string">':'</span>
1260                 };
1261             <span class="keyword">for</span> j = 1:numel(selectedSimulations)
1262                 states = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, <span class="string">'extract'</span>, selectedSimulations(j));
1263                 
1264                 <span class="keyword">for</span> i = 1:size(states.Ribosome.states, 3)
1265                     tfs = states.Ribosome.states(:, :, i) == rib.activeValue | states.Ribosome.states(:, :, i) == rib.stalledValue;
1266                     boundMRNAs = states.Ribosome.boundMRNAs(tfs, :, i);
1267                     lengths1 = states.Ribosome.mRNAPositions(tfs, :, i);
1268                     lengths2 = states.Ribosome.tmRNAPositions(tfs, :, i);
1269                     
1270                     pols(1, 1, i, j) = sum(pm.lengths(pm.nascentIndexs(boundMRNAs)) == lengths1 | lengths2 == pol.proteolysisTagLength);
1271                     pols(2, 1, i, j) = sum(lengths1 == 0 &amp; lengths2 == 0);
1272                     pols(4, 1, i, j) = sum(states.Ribosome.states(tfs, :, i) == rib.stalledValue &amp; lengths2 ~= pol.proteolysisTagLength);
1273                     pols(3, 1, i, j) = sum(tfs) - sum(pols([1 2 4], 1, i, j));
1274                 <span class="keyword">end</span>
1275                 
1276                 clear states tfs boundMRNAs lengths1 lengths2;
1277             <span class="keyword">end</span>
1278             
1279             <span class="comment">%% clear figure</span>
1280             clf(figHandle);
1281             C = 70;
1282             
1283             <span class="comment">%% left panel</span>
1284             
1285             <span class="comment">%layout plots</span>
1286             h = (C - 10 + 1) / 10;
1287             axesHandlesL = PlotUtil.multiElementPlot(figHandle, h * ones(11, 1), [0 time(end)], struct(<span class="keyword">...</span>
1288                 <span class="string">'titleStr'</span>, <span class="string">'Metabolites'</span>, <span class="keyword">...</span>
1289                 <span class="string">'position'</span>, [0.10 0.10 0.20 0.8], <span class="keyword">...</span>
1290                 <span class="string">'colorOrder'</span>, PlotUtil.getRedGreenColorOrder(simOrder)));
1291             
1292             <span class="comment">%plot data</span>
1293             PlotUtil.plotLine(axesHandlesL(1), time(2:end), atp, false, true, false);
1294             ylabel(axesHandlesL(1), <span class="string">'ATP'</span>);
1295             
1296             PlotUtil.plotLine(axesHandlesL(2), time(2:end), gtp, false, true, false);
1297             ylabel(axesHandlesL(2), <span class="string">'GTP'</span>);
1298             
1299             PlotUtil.plotLine(axesHandlesL(3), time(2:end), adp, false, true, false);
1300             ylabel(axesHandlesL(3), <span class="string">'ADP'</span>);
1301             
1302             PlotUtil.plotLine(axesHandlesL(4), time(2:end), gdp, false, true, false);
1303             ylabel(axesHandlesL(4), <span class="string">'GDP'</span>);
1304             
1305             PlotUtil.plotLine(axesHandlesL(5), time(2:end), amp, false, true, false);
1306             ylabel(axesHandlesL(5), <span class="string">'AMP'</span>);
1307             
1308             PlotUtil.plotLine(axesHandlesL(6), time(2:end), aas, false, true, false);
1309             ylabel(axesHandlesL(6), <span class="string">'AAs'</span>);
1310             
1311             PlotUtil.plotLine(axesHandlesL(7), time(2:end), fthf10, false, true, false);
1312             ylabel(axesHandlesL(7), <span class="string">'FTHF10'</span>);
1313             
1314             PlotUtil.plotLine(axesHandlesL(8), time(2:end), permute(pols(2, :, :, :), [4 3 1 2]), false, true, false);
1315             ylabel(axesHandlesL(8), {<span class="string">'Init'</span> <span class="string">'Ribs'</span>});
1316             
1317             PlotUtil.plotLine(axesHandlesL(9), time(2:end), permute(pols(3, :, :, :), [4 3 1 2]), false, true, false);
1318             ylabel(axesHandlesL(9), {<span class="string">'Elng'</span> <span class="string">'Ribs'</span>});
1319             
1320             PlotUtil.plotLine(axesHandlesL(10), time(2:end), permute(pols(1, :, :, :), [4 3 1 2]), false, true, false);
1321             ylabel(axesHandlesL(10), {<span class="string">'Term'</span> <span class="string">'Ribs'</span>});
1322             
1323             PlotUtil.plotLine(axesHandlesL(11), time(2:end), permute(pols(4, :, :, :), [4 3 1 2]), false, true, false);
1324             ylabel(axesHandlesL(11), {<span class="string">'Stalled'</span> <span class="string">'Ribs'</span>});
1325             
1326             <span class="comment">% Format Y-axis</span>
1327             PlotUtil.alignYAxesLabels(axesHandlesL);
1328             PlotUtil.offsetYAxes(axesHandlesL, 0.04);
1329             PlotUtil.labelSubplots(axesHandlesL(1), <span class="string">'A'</span>, -0.42, 1.2);
1330             
1331             <span class="comment">%% middle panels</span>
1332             
1333             nPlots1 = ceil(size(enzymes, 1) / 2);
1334             nPlots2 = floor(size(enzymes, 1) / 2);
1335             
1336             <span class="comment">%layout plots</span>
1337             colorOrder = [PlotUtil.getRedGreenColorOrder(simOrder); 0.5 0.5 0.5];
1338             h = (C - ceil(size(enzymes, 1)/2) + 1) / ceil(size(enzymes, 1) / 2);
1339             axesHandlesM1 = PlotUtil.multiElementPlot(figHandle, h * ones(nPlots1, 1), [0 time(end)], struct(<span class="keyword">...</span>
1340                 <span class="string">'titleStr'</span>, <span class="string">'Enzymes'</span>, <span class="keyword">...</span>
1341                 <span class="string">'position'</span>, [0.36 0.10 0.10 0.8], <span class="keyword">...</span>
1342                 <span class="string">'colorOrder'</span>, colorOrder));
1343             axesHandlesM2 = PlotUtil.multiElementPlot(figHandle, h * ones(nPlots1, 1), [0 time(end)], struct(<span class="keyword">...</span>
1344                 <span class="string">'position'</span>, [0.53 0.10 0.10 0.8], <span class="keyword">...</span>
1345                 <span class="string">'colorOrder'</span>, colorOrder));
1346             
1347             <span class="comment">%plot data</span>
1348             <span class="keyword">for</span> i = 1:nPlots1
1349                 PlotUtil.plotLine(axesHandlesM1(i), time(2:end), permute(cat(4, enzymes(i, :, :, :), expEnzymes(i, :, :, :)), [4 3 1 2]), false, true, false);
1350                 ylabel(axesHandlesM1(i), [enzProcesses(i) num2str(enzProcessIdxs(i))]);
1351             <span class="keyword">end</span>
1352             <span class="keyword">for</span> i = 1:nPlots2
1353                 PlotUtil.plotLine(axesHandlesM2(i), time(2:end), permute(cat(4, enzymes(nPlots1+i, :, :, :), expEnzymes(nPlots1+i, :, :, :)), [4 3 1 2]), false, true, false);
1354                 ylabel(axesHandlesM2(i), [enzProcesses(nPlots1+i) num2str(enzProcessIdxs(nPlots1+i))]);
1355             <span class="keyword">end</span>
1356             
1357             <span class="comment">% Format Y-axis</span>
1358             PlotUtil.alignYAxesLabels(axesHandlesM1);
1359             PlotUtil.alignYAxesLabels(axesHandlesM2);
1360             axesHandlesM12 = PlotUtil.offsetYAxes(axesHandlesM1, 0.04);
1361             axesHandlesM22 = PlotUtil.offsetYAxes(axesHandlesM2, 0.04);
1362             PlotUtil.labelSubplots(axesHandlesM1(1), <span class="string">'B'</span>, -0.42, 1.2);
1363             
1364             set(axesHandlesM1(nPlots1+1:end), <span class="string">'visible'</span>, <span class="string">'off'</span>)
1365             set(axesHandlesM2(nPlots2+1:end), <span class="string">'visible'</span>, <span class="string">'off'</span>)
1366             set(axesHandlesM12(nPlots1+1:end), <span class="string">'visible'</span>, <span class="string">'off'</span>)
1367             set(axesHandlesM22(nPlots2+1:end), <span class="string">'visible'</span>, <span class="string">'off'</span>)
1368             
1369             <span class="comment">%% right panels</span>
1370             nPlots1 = ceil(size(trnas, 1) / 2);
1371             nPlots2 = floor(size(trnas, 1) / 2);
1372             
1373             <span class="comment">%layout plots</span>
1374             colorOrder = [PlotUtil.getRedGreenColorOrder(simOrder); PlotUtil.getRedGreenColorOrder(simOrder); 0.5 0.5 0.5];
1375             h = (C - ceil(size(trnas, 1)/2) + 1) / ceil(size(trnas, 1) / 2);
1376             axesHandlesR1 = PlotUtil.multiElementPlot(figHandle, h * ones(nPlots1, 1), [0 time(end)], struct(<span class="keyword">...</span>
1377                 <span class="string">'titleStr'</span>, <span class="string">'tRNA'</span>, <span class="keyword">...</span>
1378                 <span class="string">'position'</span>, [0.70 0.10 0.10 0.8], <span class="keyword">...</span>
1379                 <span class="string">'colorOrder'</span>, colorOrder));
1380             axesHandlesR2 = PlotUtil.multiElementPlot(figHandle, h * ones(nPlots1, 1), [0 time(end)], struct(<span class="keyword">...</span>
1381                 <span class="string">'position'</span>, [0.87 0.10 0.10 0.8], <span class="keyword">...</span>
1382                 <span class="string">'colorOrder'</span>, colorOrder));
1383             
1384             <span class="comment">%plot data</span>
1385             <span class="keyword">for</span> i = 1:nPlots1
1386                 PlotUtil.plotLine(axesHandlesR1(i), time(2:end), permute(cat(4, trnas(i, :, :, :), aatrnas(i, :, :, :), expTRNAs(i, :, :, :)), [4 3 1 2]), false, true, false);
1387                 ylabel(axesHandlesR1(i), num2str(i));
1388             <span class="keyword">end</span>
1389             <span class="keyword">for</span> i = 1:nPlots2
1390                 PlotUtil.plotLine(axesHandlesR2(i), time(2:end), permute(cat(4, trnas(nPlots1+i, :, :, :), aatrnas(nPlots1+i, :, :, :), expTRNAs(nPlots1+i, :, :, :)), [4 3 1 2]), false, true, false);
1391                 ylabel(axesHandlesR2(i), num2str(nPlots1+i));
1392             <span class="keyword">end</span>
1393             
1394             <span class="comment">% Format Y-axis</span>
1395             PlotUtil.alignYAxesLabels(axesHandlesR1);
1396             PlotUtil.alignYAxesLabels(axesHandlesR2);
1397             axesHandlesR12 = PlotUtil.offsetYAxes(axesHandlesR1, 0.04);
1398             axesHandlesR22 = PlotUtil.offsetYAxes(axesHandlesR2, 0.04);
1399             PlotUtil.labelSubplots(axesHandlesR1(1), <span class="string">'C'</span>, -0.42, 1.2);
1400             
1401             set(axesHandlesR1(nPlots1+1:end), <span class="string">'visible'</span>, <span class="string">'off'</span>)
1402             set(axesHandlesR2(nPlots2+1:end), <span class="string">'visible'</span>, <span class="string">'off'</span>)
1403             set(axesHandlesR12(nPlots1+1:end), <span class="string">'visible'</span>, <span class="string">'off'</span>)
1404             set(axesHandlesR22(nPlots2+1:end), <span class="string">'visible'</span>, <span class="string">'off'</span>)
1405             
1406             <span class="comment">%% organize figure data</span>
1407             figData = struct;
1408             figData.selectedSimulations = selectedSimulations;
1409             figData.simOrder = simOrder;
1410             figData.time = time;
1411             figData.atp = atp;
1412             figData.gtp = gtp;
1413             figData.adp = adp;
1414             figData.gdp = gdp;
1415             figData.amp = amp;
1416             figData.aas = aas;
1417             figData.fthf10 = fthf10;
1418             figData.enzProcesses = enzProcesses;
1419             figData.enzProcessIdxs = enzProcessIdxs;
1420             figData.enzymes = enzymes;
1421             figData.expEnzymes = expEnzymes;
1422             figData.trnas = trnas;
1423             figData.aatrnas = aatrnas;
1424             figData.expTRNAs = expTRNAs;
1425             figData.pols = pols;
1426         <span class="keyword">end</span>
1427         
1428         <a name="_sub11" href="#_subfunctions" class="code">function figData = proteinSynthesis(figHandle, simBatchDir, selectedSimulations)</a>
1429             import edu.stanford.covert.cell.sim.util.PlotUtil;
1430             import edu.stanford.covert.cell.sim.util.SimulationEnsemble;
1431             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
1432             
1433             <span class="keyword">if</span> nargin &lt; 2 || isempty(simBatchDir)
1434                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
1435             <span class="keyword">end</span>
1436             <span class="keyword">if</span> nargin &lt; 3 || isempty(selectedSimulations)
1437                 selectedSimulations = SimulationDiskUtil.getCompleteSimulations(simBatchDir);
1438             <span class="keyword">end</span>
1439             
1440             [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep <span class="string">'1'</span>]);
1441             comp = sim.compartment;
1442             pm = sim.state(<span class="string">'ProteinMonomer'</span>);
1443             pc = sim.state(<span class="string">'ProteinComplex'</span>);
1444             
1445             ppI = sim.process(<span class="string">'ProteinProcessingI'</span>);
1446             ppII = sim.process(<span class="string">'ProteinProcessingII'</span>);
1447             pt = sim.process(<span class="string">'ProteinTranslocation'</span>);
1448             pf = sim.process(<span class="string">'ProteinFolding'</span>);
1449             pmod = sim.process(<span class="string">'ProteinModification'</span>);
1450             pd = sim.process(<span class="string">'ProteinDecay'</span>);
1451             
1452             enzProcesses = [
1453                 repmat({<span class="string">'PP-I'</span>}, size(ppI.enzymeWholeCellModelIDs))
1454                 repmat({<span class="string">'PT'</span>}, size(pt.enzymeWholeCellModelIDs))
1455                 repmat({<span class="string">'PP-II'</span>}, size(ppII.enzymeWholeCellModelIDs))
1456                 repmat({<span class="string">'PF'</span>}, size(pf.enzymeWholeCellModelIDs))
1457                 repmat({<span class="string">'PMod'</span>}, size(pmod.enzymeWholeCellModelIDs))
1458                 repmat({<span class="string">'PD'</span>}, size(pd.enzymeWholeCellModelIDs))
1459                 ];
1460             enzProcessIdxs = [
1461                 (1:size(ppI.enzymeWholeCellModelIDs))'
1462                 (1:size(pt.enzymeWholeCellModelIDs))'
1463                 (1:size(ppII.enzymeWholeCellModelIDs))'
1464                 (1:size(pf.enzymeWholeCellModelIDs))'
1465                 (1:size(pmod.enzymeWholeCellModelIDs))'
1466                 (1:size(pd.enzymeWholeCellModelIDs))'
1467                 ];
1468             
1469             <span class="comment">%% get data</span>
1470             ensemble = SimulationEnsemble(simBatchDir, {<span class="string">'mass'</span> <span class="string">'mass'</span>; <span class="string">'immatureMonomers'</span> <span class="string">'immatureMonomers'</span>}, [], selectedSimulations);
1471             time = ensemble.stateData.time / 3600;
1472             mass = ensemble.stateData.values(ensemble.getPropertyIndices(<span class="string">'mass'</span>), :, :, :) * sim.state(<span class="string">'Mass'</span>).cellInitialDryWeight / (1-sim.state(<span class="string">'Mass'</span>).fractionWetWeight) * 1e15;
1473             simEndTimes = ensemble.stateData.simulationEndTimes;
1474             
1475             [~, simOrder] = sort(mass(:, 1, min(simEndTimes), :));
1476             simOrder = simOrder(:);
1477             
1478             clear ensemble;
1479             
1480             stateNames = {
1481                 <span class="string">'Polypeptide'</span>    <span class="string">'abortedPolypeptides'</span>  <span class="string">'-nnz'</span>           1
1482                 <span class="string">'ProteinMonomer'</span> <span class="string">'counts'</span>               <span class="string">':'</span>              <span class="string">':'</span>
1483                 <span class="string">'ProteinComplex'</span> <span class="string">'counts'</span>               pc.matureIndexs  <span class="string">'-sum'</span>
1484                 };
1485             nscnt = NaN(numel(selectedSimulations), max(simEndTimes));
1486             procsI = NaN(numel(selectedSimulations), max(simEndTimes));
1487             trans = NaN(numel(selectedSimulations), max(simEndTimes));
1488             procsII = NaN(numel(selectedSimulations), max(simEndTimes));
1489             folds = NaN(numel(selectedSimulations), max(simEndTimes));
1490             signals = NaN(numel(selectedSimulations), max(simEndTimes));
1491             inacts = NaN(numel(selectedSimulations), max(simEndTimes));
1492             misfolds = NaN(numel(selectedSimulations), max(simEndTimes));
1493             damages = NaN(numel(selectedSimulations), max(simEndTimes));
1494             aborts = NaN(numel(selectedSimulations), max(simEndTimes));
1495             enzymes = NaN(numel(enzProcesses), 1, max(simEndTimes), numel(selectedSimulations));
1496             <span class="keyword">for</span> i = 1:numel(selectedSimulations)
1497                 states = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, <span class="string">'extract'</span>, selectedSimulations(i));
1498                 
1499                 monCnts = full(sum(states.ProteinMonomer.counts, 2));
1500                 nscnt(i, 1:simEndTimes(i)) = permute(sum(monCnts(pm.nascentIndexs, :, :, :), 1), [4 3 1 2]);
1501                 procsI(i, 1:simEndTimes(i)) = full(permute(sum(states.ProteinMonomer.counts(<span class="keyword">...</span>
1502                     pm.processedIIndexs(pm.compartments(pm.processedIIIndexs) ~= comp.cytosolIndexs), <span class="keyword">...</span>
1503                     comp.cytosolIndexs, :, :), 1), [4 3 1 2]));
1504                 trans(i, 1:simEndTimes(i)) = permute(sum(monCnts(pm.processedIIndexs, :, :, :), 1), [4 3 1 2]) - procsI(i, 1:simEndTimes(i));
1505                 procsII(i, 1:simEndTimes(i)) = permute(sum(monCnts(pm.processedIIIndexs, :, :, :), 1), [4 3 1 2]);
1506                 folds(i, 1:simEndTimes(i)) = permute(sum(monCnts(pm.foldedIndexs, :, :, :), 1), [4 3 1 2]);
1507                 signals(i, 1:simEndTimes(i)) = permute(sum(monCnts(pm.signalSequenceIndexs, :, :, :), 1), [4 3 1 2]);
1508                 inacts(i, 1:simEndTimes(i)) = permute(sum(monCnts(pm.inactivatedIndexs, :, :, :), 1), [4 3 1 2]);
1509                 misfolds(i, 1:simEndTimes(i)) = permute(sum(monCnts(pm.misfoldedIndexs, :, :, :), 1), [4 3 1 2]);
1510                 damages(i, 1:simEndTimes(i)) = permute(sum(monCnts(pm.damagedIndexs, :, :, :), 1), [4 3 1 2]);
1511                 aborts(i, 1:simEndTimes(i)) = permute(full(states.Polypeptide.abortedPolypeptides), [4 3 1 2]);
1512                 
1513                 monCnts = monCnts(pm.matureIndexs, :, :, :);
1514                 cpxCnts = full(states.ProteinComplex.counts);
1515                 
1516                 tmpEnzymes = zeros(0, 1, size(monCnts, 3), size(monCnts, 4));
1517                 
1518                 tmp = zeros(numel(ppI.enzymes), 1, size(monCnts, 3), size(monCnts, 4));
1519                 tmp(ppI.enzymeMonomerLocalIndexs, :, :, :) = monCnts(ppI.enzymeMonomerGlobalIndexs, :, :, :);
1520                 tmp(ppI.enzymeComplexLocalIndexs, :, :, :) = cpxCnts(ppI.enzymeComplexGlobalIndexs, :, :, :);
1521                 tmpEnzymes = [tmpEnzymes; tmp]; <span class="comment">%#ok&lt;AGROW&gt;</span>
1522                 
1523                 tmp = zeros(numel(pt.enzymes), 1, size(monCnts, 3), size(monCnts, 4));
1524                 tmp(pt.enzymeMonomerLocalIndexs, :, :, :) = monCnts(pt.enzymeMonomerGlobalIndexs, :, :, :);
1525                 tmp(pt.enzymeComplexLocalIndexs, :, :, :) = cpxCnts(pt.enzymeComplexGlobalIndexs, :, :, :);
1526                 tmpEnzymes = [tmpEnzymes; tmp]; <span class="comment">%#ok&lt;AGROW&gt;</span>
1527                 
1528                 tmp = zeros(numel(ppII.enzymes), 1, size(monCnts, 3), size(monCnts, 4));
1529                 tmp(ppII.enzymeMonomerLocalIndexs, :, :, :) = monCnts(ppII.enzymeMonomerGlobalIndexs, :, :, :);
1530                 tmp(ppII.enzymeComplexLocalIndexs, :, :, :) = cpxCnts(ppII.enzymeComplexGlobalIndexs, :, :, :);
1531                 tmpEnzymes = [tmpEnzymes; tmp]; <span class="comment">%#ok&lt;AGROW&gt;</span>
1532                 
1533                 tmp = zeros(numel(pf.enzymes), 1, size(monCnts, 3), size(monCnts, 4));
1534                 tmp(pf.enzymeMonomerLocalIndexs, :, :, :) = monCnts(pf.enzymeMonomerGlobalIndexs, :, :, :);
1535                 tmp(pf.enzymeComplexLocalIndexs, :, :, :) = cpxCnts(pf.enzymeComplexGlobalIndexs, :, :, :);
1536                 tmpEnzymes = [tmpEnzymes; tmp]; <span class="comment">%#ok&lt;AGROW&gt;</span>
1537                 
1538                 tmp = zeros(numel(pmod.enzymes), 1, size(monCnts, 3), size(monCnts, 4));
1539                 tmp(pmod.enzymeMonomerLocalIndexs, :, :, :) = monCnts(pmod.enzymeMonomerGlobalIndexs, :, :, :);
1540                 tmp(pmod.enzymeComplexLocalIndexs, :, :, :) = cpxCnts(pmod.enzymeComplexGlobalIndexs, :, :, :);
1541                 tmpEnzymes = [tmpEnzymes; tmp]; <span class="comment">%#ok&lt;AGROW&gt;</span>
1542                 
1543                 tmp = zeros(numel(pd.enzymes), 1, size(monCnts, 3), size(monCnts, 4));
1544                 tmp(pd.enzymeMonomerLocalIndexs, :, :, :) = monCnts(pd.enzymeMonomerGlobalIndexs, :, :, :);
1545                 tmp(pd.enzymeComplexLocalIndexs, :, :, :) = cpxCnts(pd.enzymeComplexGlobalIndexs, :, :, :);
1546                 tmpEnzymes = [tmpEnzymes; tmp]; <span class="comment">%#ok&lt;AGROW&gt;</span>
1547                 
1548                 enzymes(:, :, 1:simEndTimes(i), i) = tmpEnzymes;
1549                 
1550                 clear states monCnts cpxCnts tmpEnzymes tmp;
1551             <span class="keyword">end</span>
1552             
1553             <span class="comment">%% clear figure</span>
1554             clf(figHandle);
1555             
1556             <span class="comment">%% left panel</span>
1557             
1558             <span class="comment">%layout plots</span>
1559             axesHandlesL = PlotUtil.multiElementPlot(figHandle, 3.5 * ones(10, 1), [0 time(end)], struct(<span class="keyword">...</span>
1560                 <span class="string">'titleStr'</span>, <span class="string">'Monomers'</span>, <span class="keyword">...</span>
1561                 <span class="string">'position'</span>, [0.12 0.10 0.21 0.8], <span class="keyword">...</span>
1562                 <span class="string">'colorOrder'</span>, PlotUtil.getRedGreenColorOrder(simOrder)));
1563             
1564             <span class="comment">%plot data</span>
1565             PlotUtil.plotLine(axesHandlesL(1), time(2:end), nscnt, false, true, false);
1566             ylabel(axesHandlesL(1), <span class="string">'Nascent'</span>);
1567             
1568             PlotUtil.plotLine(axesHandlesL(2), time(2:end), procsI, false, true, false);
1569             ylabel(axesHandlesL(2), <span class="string">'Processed-I'</span>);
1570             
1571             PlotUtil.plotLine(axesHandlesL(3), time(2:end), trans, false, true, false);
1572             ylabel(axesHandlesL(3), <span class="string">'Translocated'</span>);
1573             
1574             PlotUtil.plotLine(axesHandlesL(4), time(2:end), procsII, false, true, false);
1575             ylabel(axesHandlesL(4), <span class="string">'Processed-II'</span>);
1576             
1577             PlotUtil.plotLine(axesHandlesL(5), time(2:end), folds, false, true, false);
1578             ylabel(axesHandlesL(5), <span class="string">'Folded'</span>);
1579             
1580             PlotUtil.plotLine(axesHandlesL(6), time(2:end), signals, false, true, false);
1581             ylabel(axesHandlesL(6), {<span class="string">'Signal'</span> <span class="string">'Sequences'</span>});
1582             
1583             PlotUtil.plotLine(axesHandlesL(7), time(2:end), inacts, false, true, false);
1584             ylabel(axesHandlesL(7), <span class="string">'Inactivated'</span>);
1585             
1586             PlotUtil.plotLine(axesHandlesL(8), time(2:end), misfolds, false, true, false);
1587             ylabel(axesHandlesL(8), <span class="string">'Misfolded'</span>);
1588             
1589             PlotUtil.plotLine(axesHandlesL(9), time(2:end), damages, false, true, false);
1590             ylabel(axesHandlesL(9), <span class="string">'Damaged'</span>);
1591             
1592             PlotUtil.plotLine(axesHandlesL(10), time(2:end), aborts, false, true, false);
1593             ylabel(axesHandlesL(10), <span class="string">'Aborted'</span>);
1594             
1595             <span class="comment">% Format Y-axis</span>
1596             PlotUtil.alignYAxesLabels(axesHandlesL);
1597             PlotUtil.offsetYAxes(axesHandlesL, 0.04);
1598             PlotUtil.labelSubplots(axesHandlesL(1), <span class="string">'A'</span>, -0.42, 1.2);
1599             
1600             <span class="comment">%% right panels</span>
1601             
1602             nPlots1 = ceil(size(enzymes, 1) / 2);
1603             nPlots2 = floor(size(enzymes, 1) / 2);
1604             
1605             <span class="comment">%layout plots</span>
1606             axesHandlesM = PlotUtil.multiElementPlot(figHandle, 3.5 * ones(nPlots1, 1), [0 time(end)], struct(<span class="keyword">...</span>
1607                 <span class="string">'titleStr'</span>, <span class="string">'Enzymes'</span>, <span class="keyword">...</span>
1608                 <span class="string">'position'</span>, [0.45 0.10 0.21 0.8], <span class="keyword">...</span>
1609                 <span class="string">'colorOrder'</span>, PlotUtil.getRedGreenColorOrder(simOrder)));
1610             axesHandlesR = PlotUtil.multiElementPlot(figHandle, 3.5 * ones(nPlots2, 1), [0 time(end)], struct(<span class="keyword">...</span>
1611                 <span class="string">'position'</span>, [0.78 0.10 0.21 0.8], <span class="keyword">...</span>
1612                 <span class="string">'colorOrder'</span>, PlotUtil.getRedGreenColorOrder(simOrder)));
1613             
1614             <span class="comment">%plot data</span>
1615             <span class="keyword">for</span> i = 1:nPlots1
1616                 PlotUtil.plotLine(axesHandlesM(i), time(2:end), permute(enzymes(i, :, :, :), [4 3 1 2]), false, true, false);
1617                 ylabel(axesHandlesM(i), [enzProcesses(i) num2str(enzProcessIdxs(i))]);
1618             <span class="keyword">end</span>
1619             <span class="keyword">for</span> i = 1:nPlots2
1620                 PlotUtil.plotLine(axesHandlesR(i), time(2:end), permute(enzymes(nPlots1+i, :, :, :), [4 3 1 2]), false, true, false);
1621                 ylabel(axesHandlesR(i), [enzProcesses(nPlots1+i) num2str(enzProcessIdxs(nPlots1+i))]);
1622             <span class="keyword">end</span>
1623             
1624             <span class="comment">% Format Y-axis</span>
1625             PlotUtil.alignYAxesLabels(axesHandlesM);
1626             PlotUtil.alignYAxesLabels(axesHandlesR);
1627             PlotUtil.offsetYAxes(axesHandlesM, 0.04);
1628             PlotUtil.offsetYAxes(axesHandlesR, 0.04);
1629             PlotUtil.labelSubplots(axesHandlesM(1), <span class="string">'B'</span>, -0.42, 1.2);
1630             
1631             <span class="comment">%% organize figure data</span>
1632             figData = struct;
1633             figData.simOrder = simOrder;
1634             figData.time = time;
1635             figData.nscnt = nscnt;
1636             figData.procsI = procsI;
1637             figData.trans = trans;
1638             figData.procsII = procsII;
1639             figData.folds = folds;
1640             figData.signals = signals;
1641             figData.inacts = inacts;
1642             figData.misfolds = misfolds;
1643             figData.damages = damages;
1644             figData.aborts = aborts;
1645             figData.enzymes = enzymes;
1646             figData.enzProcesses = enzProcesses;
1647         <span class="keyword">end</span>
1648         
1649         <a name="_sub12" href="#_subfunctions" class="code">function figData = rnaSynthesis(figHandle, simBatchDir, selectedSimulations)</a>
1650             import edu.stanford.covert.cell.sim.util.PlotUtil;
1651             import edu.stanford.covert.cell.sim.util.SimulationEnsemble;
1652             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
1653             import edu.stanford.covert.util.ConstantUtil;
1654             
1655             <span class="keyword">if</span> nargin &lt; 2 || isempty(simBatchDir)
1656                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
1657             <span class="keyword">end</span>
1658             <span class="keyword">if</span> nargin &lt; 3 || isempty(selectedSimulations)
1659                 selectedSimulations = SimulationDiskUtil.getCompleteSimulations(simBatchDir);
1660             <span class="keyword">end</span>
1661             
1662             [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep <span class="string">'1'</span>]);
1663             g = sim.gene;
1664             
1665             t = sim.state(<span class="string">'Time'</span>);
1666             mass = sim.state(<span class="string">'Mass'</span>);
1667             rna = sim.state(<span class="string">'Rna'</span>);
1668             pm = sim.state(<span class="string">'ProteinMonomer'</span>);
1669             pc = sim.state(<span class="string">'ProteinComplex'</span>);
1670             
1671             rp = sim.process(<span class="string">'RNAProcessing'</span>);
1672             rmod = sim.process(<span class="string">'RNAModification'</span>);
1673             rd = sim.process(<span class="string">'RNADecay'</span>);
1674             
1675             monExp = (rna.matureRNAGeneComposition(g.mRNAIndexs, :) * rna.expression(rna.matureIndexs)) ./ <span class="keyword">...</span>
1676                 (log(2) / t.cellCycleLength + pm.decayRates(pm.matureIndexs));
1677             expTotMons = mass.initialFractionAAsInMonomers * mass.dryWeightFractionProtein * mass.cellInitialDryWeight * <span class="keyword">...</span>
1678                 monExp / (monExp' * pm.molecularWeights(pm.matureIndexs) / ConstantUtil.nAvogadro);
1679             expTotCpxs = min(repmat(expTotMons, [1 numel(pc.matureIndexs)]) ./ sum(pc.proteinComplexComposition(g.mRNAIndexs, :, :), 3), [], 1);
1680             
1681             enzProcesses = [
1682                 repmat({<span class="string">'RP'</span>}, size(rp.enzymeWholeCellModelIDs))
1683                 repmat({<span class="string">'RMod'</span>}, size(rmod.enzymeWholeCellModelIDs))
1684                 repmat({<span class="string">'RD'</span>}, size(rd.enzymeWholeCellModelIDs))
1685                 ];
1686             enzProcessIdxs = [
1687                 (1:size(rp.enzymeWholeCellModelIDs))'
1688                 (1:size(rmod.enzymeWholeCellModelIDs))'
1689                 (1:size(rd.enzymeWholeCellModelIDs))'
1690                 ];
1691             
1692             <span class="comment">%% get data</span>
1693             ensemble = SimulationEnsemble(simBatchDir, {<span class="string">'mass'</span> <span class="string">'mass'</span>; <span class="string">'immatureRnas'</span> <span class="string">'immatureRnas'</span>}, [], selectedSimulations);
1694             time = ensemble.stateData.time / 3600;
1695             mass = ensemble.stateData.values(ensemble.getPropertyIndices(<span class="string">'mass'</span>), :, :, :) * sim.state(<span class="string">'Mass'</span>).cellInitialDryWeight / (1-sim.state(<span class="string">'Mass'</span>).fractionWetWeight) * 1e15;
1696             simEndTimes = ensemble.stateData.simulationEndTimes();
1697             
1698             [~, simOrder] = sort(mass(:, 1, min(simEndTimes), :));
1699             simOrder = simOrder(:);
1700             
1701             clear ensemble;
1702             
1703             stateNames = {
1704                 <span class="string">'Rna'</span>            <span class="string">'counts'</span>  <span class="string">':'</span>               <span class="string">'-sum'</span>
1705                 <span class="string">'Transcript'</span>     <span class="string">'abortedTranscripts'</span> <span class="string">'-nnz'</span>  1
1706                 <span class="string">'ProteinMonomer'</span> <span class="string">'counts'</span>  pm.matureIndexs  <span class="string">'-sum'</span>
1707                 <span class="string">'ProteinComplex'</span> <span class="string">'counts'</span>  pc.matureIndexs  <span class="string">'-sum'</span>
1708                 };
1709             nscnt = NaN(numel(selectedSimulations), max(simEndTimes));
1710             procs = NaN(numel(selectedSimulations), max(simEndTimes));
1711             misfolds = NaN(numel(selectedSimulations), max(simEndTimes));
1712             damages = NaN(numel(selectedSimulations), max(simEndTimes));
1713             aborts = NaN(numel(selectedSimulations), max(simEndTimes));
1714             enzymes = NaN(numel(enzProcessIdxs), 1, max(simEndTimes), numel(selectedSimulations));
1715             <span class="keyword">for</span> i = 1:numel(selectedSimulations)
1716                 states = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, <span class="string">'extract'</span>, selectedSimulations(i));
1717                 
1718                 rnaCnts = full(states.Rna.counts);
1719                 monCnts = full(states.ProteinMonomer.counts);
1720                 cpxCnts = full(states.ProteinComplex.counts);
1721                 nscnt(i, 1:simEndTimes(i)) = permute(sum(rnaCnts(rna.nascentIndexs, :, :, :), 1), [4 3 1 2]);
1722                 procs(i, 1:simEndTimes(i)) = permute(sum(rnaCnts(rna.processedIndexs, :, :, :), 1), [4 3 1 2]);
1723                 misfolds (i, 1:simEndTimes(i))= permute(sum(rnaCnts(rna.misfoldedIndexs, :, :, :), 1), [4 3 1 2]);
1724                 damages(i, 1:simEndTimes(i)) = permute(sum(rnaCnts(rna.damagedIndexs, :, :, :), 1), [4 3 1 2]);
1725                 aborts(i, 1:simEndTimes(i)) = permute(full(states.Transcript.abortedTranscripts), [4 3 1 2]);
1726                 
1727                 tmpEnzymes = zeros(0, 1, size(monCnts, 3), size(monCnts, 4));
1728                 
1729                 tmp = zeros(numel(rp.enzymes), 1, size(monCnts, 3), size(monCnts, 4));
1730                 tmp(rp.enzymeRNALocalIndexs, :, :, :) = rnaCnts(numel(rna.nascentIndexs)+numel(rna.processedIndexs) + rp.enzymeRNAGlobalIndexs, :, :, :);
1731                 tmp(rp.enzymeMonomerLocalIndexs, :, :, :) = monCnts(rp.enzymeMonomerGlobalIndexs, :, :, :);
1732                 tmp(rp.enzymeComplexLocalIndexs, :, :, :) = cpxCnts(rp.enzymeComplexGlobalIndexs, :, :, :);
1733                 tmpEnzymes = [tmpEnzymes; tmp]; <span class="comment">%#ok&lt;AGROW&gt;</span>
1734                 
1735                 tmp = zeros(numel(rmod.enzymes), 1, size(monCnts, 3), size(monCnts, 4));
1736                 tmp(rmod.enzymeRNALocalIndexs, :, :, :) = rnaCnts(numel(rna.nascentIndexs)+numel(rna.processedIndexs) + rmod.enzymeRNAGlobalIndexs, :, :, :);
1737                 tmp(rmod.enzymeMonomerLocalIndexs, :, :, :) = monCnts(rmod.enzymeMonomerGlobalIndexs, :, :, :);
1738                 tmp(rmod.enzymeComplexLocalIndexs, :, :, :) = cpxCnts(rmod.enzymeComplexGlobalIndexs, :, :, :);
1739                 tmpEnzymes = [tmpEnzymes; tmp]; <span class="comment">%#ok&lt;AGROW&gt;</span>
1740                 
1741                 tmp = zeros(numel(rd.enzymes), 1, size(monCnts, 3), size(monCnts, 4));
1742                 tmp(rd.enzymeRNALocalIndexs, :, :, :) = rnaCnts(numel(rna.nascentIndexs)+numel(rna.processedIndexs) + rd.enzymeRNAGlobalIndexs, :, :, :);
1743                 tmp(rd.enzymeMonomerLocalIndexs, :, :, :) = monCnts(rd.enzymeMonomerGlobalIndexs, :, :, :);
1744                 tmp(rd.enzymeComplexLocalIndexs, :, :, :) = cpxCnts(rd.enzymeComplexGlobalIndexs, :, :, :);
1745                 tmpEnzymes = [tmpEnzymes; tmp]; <span class="comment">%#ok&lt;AGROW&gt;</span>
1746                 
1747                 enzymes(:, :, 1:simEndTimes(i), i) = tmpEnzymes;
1748                 
1749                 clear states rnaCnts monCnts cpxCnts tmpEnzymes tmp;
1750             <span class="keyword">end</span>
1751             
1752             expEnzymes = zeros(0, 1);
1753             
1754             tmp2 = zeros(numel(rp.enzymes), 1);
1755             tmp2(rp.enzymeMonomerLocalIndexs) = expTotMons(rp.enzymeMonomerGlobalIndexs);
1756             tmp2(rp.enzymeComplexLocalIndexs) = expTotCpxs(rp.enzymeComplexGlobalIndexs);
1757             expEnzymes = [expEnzymes; tmp2];
1758             
1759             tmp2 = zeros(numel(rmod.enzymes), 1);
1760             tmp2(rmod.enzymeMonomerLocalIndexs) = expTotMons(rmod.enzymeMonomerGlobalIndexs);
1761             tmp2(rmod.enzymeComplexLocalIndexs) = expTotCpxs(rmod.enzymeComplexGlobalIndexs);
1762             expEnzymes = [expEnzymes; tmp2];
1763             
1764             tmp2 = zeros(numel(rd.enzymes), 1);
1765             tmp2(rd.enzymeMonomerLocalIndexs) = expTotMons(rd.enzymeMonomerGlobalIndexs);
1766             tmp2(rd.enzymeComplexLocalIndexs) = expTotCpxs(rd.enzymeComplexGlobalIndexs);
1767             expEnzymes = [expEnzymes; tmp2];
1768             
1769             expEnzymes = repmat(expEnzymes, [1 1 numel(time)-1 1]) .* exp(log(2) * repmat(permute(time(2:end),[1 3 2]), size(expEnzymes, 1), 1) / (t.cellCycleLength/3600));
1770             
1771             <span class="comment">%% clear figure</span>
1772             clf(figHandle);
1773             C = 70;
1774             
1775             <span class="comment">%% left panel</span>
1776             
1777             <span class="comment">%layout plots</span>
1778             h = (C - 2 + 1) / 2;
1779             axesHandlesL = PlotUtil.multiElementPlot(figHandle, h* ones(5, 1), [0 time(end)], struct(<span class="keyword">...</span>
1780                 <span class="string">'titleStr'</span>, <span class="string">'RNAs'</span>, <span class="keyword">...</span>
1781                 <span class="string">'position'</span>, [0.12 0.10 0.21 0.8], <span class="keyword">...</span>
1782                 <span class="string">'colorOrder'</span>, PlotUtil.getRedGreenColorOrder(simOrder)));
1783             
1784             <span class="comment">%plot data</span>
1785             PlotUtil.plotLine(axesHandlesL(1), time(2:end), nscnt, false, true, false);
1786             ylabel(axesHandlesL(1), <span class="string">'Nascent'</span>);
1787             
1788             PlotUtil.plotLine(axesHandlesL(2), time(2:end), procs, false, true, false);
1789             ylabel(axesHandlesL(2), <span class="string">'Processed'</span>);
1790             
1791             PlotUtil.plotLine(axesHandlesL(3), time(2:end), misfolds, false, true, false);
1792             ylabel(axesHandlesL(3), <span class="string">'Misfolded'</span>);
1793             
1794             PlotUtil.plotLine(axesHandlesL(4), time(2:end), damages, false, true, false);
1795             ylabel(axesHandlesL(4), <span class="string">'Damaged'</span>);
1796             
1797             PlotUtil.plotLine(axesHandlesL(5), time(2:end), aborts, false, true, false);
1798             ylabel(axesHandlesL(5), <span class="string">'Aborted'</span>);
1799             
1800             <span class="comment">% Format Y-axis</span>
1801             PlotUtil.alignYAxesLabels(axesHandlesL);
1802             PlotUtil.offsetYAxes(axesHandlesL, 0.04);
1803             PlotUtil.labelSubplots(axesHandlesL(1), <span class="string">'A'</span>, -0.42, 1.2);
1804             
1805             <span class="comment">%% right panels</span>
1806             
1807             nPlots1 = ceil(size(enzymes, 1) / 2);
1808             nPlots2 = floor(size(enzymes, 1) / 2);
1809             
1810             <span class="comment">%layout plots</span>
1811             colorOrder = [PlotUtil.getRedGreenColorOrder(simOrder); 0.5 0.5 0.5];
1812             h = (C - ceil(size(enzymes, 1)/2) + 1) / ceil(size(enzymes, 1) / 2);
1813             axesHandlesM = PlotUtil.multiElementPlot(figHandle, h * ones(nPlots1, 1), [0 time(end)], struct(<span class="keyword">...</span>
1814                 <span class="string">'titleStr'</span>, <span class="string">'Enzymes'</span>, <span class="keyword">...</span>
1815                 <span class="string">'position'</span>, [0.45 0.10 0.21 0.8], <span class="keyword">...</span>
1816                 <span class="string">'colorOrder'</span>, colorOrder));
1817             axesHandlesR = PlotUtil.multiElementPlot(figHandle, h * ones(nPlots2, 1), [0 time(end)], struct(<span class="keyword">...</span>
1818                 <span class="string">'position'</span>, [0.78 0.10 0.21 0.8], <span class="keyword">...</span>
1819                 <span class="string">'colorOrder'</span>, colorOrder));
1820             
1821             <span class="comment">%plot data</span>
1822             <span class="keyword">for</span> i = 1:nPlots1
1823                 PlotUtil.plotLine(axesHandlesM(i), time(2:end), permute(cat(4, enzymes(i, :, :, :), expEnzymes(i, :, :, :)), [4 3 1 2]), false, true, false);
1824                 ylabel(axesHandlesM(i), [enzProcesses(i) num2str(enzProcessIdxs(i))]);
1825             <span class="keyword">end</span>
1826             <span class="keyword">for</span> i = 1:nPlots2
1827                 PlotUtil.plotLine(axesHandlesR(i), time(2:end), permute(cat(4, enzymes(nPlots1+i, :, :, :), expEnzymes(nPlots1+i, :, :, :)), [4 3 1 2]), false, true, false);
1828                 ylabel(axesHandlesR(i), [enzProcesses(nPlots1+i) num2str(enzProcessIdxs(nPlots1+i))]);
1829             <span class="keyword">end</span>
1830             
1831             <span class="comment">% Format Y-axis</span>
1832             PlotUtil.alignYAxesLabels(axesHandlesM);
1833             PlotUtil.alignYAxesLabels(axesHandlesR);
1834             PlotUtil.offsetYAxes(axesHandlesM, 0.04);
1835             PlotUtil.offsetYAxes(axesHandlesR, 0.04);
1836             PlotUtil.labelSubplots(axesHandlesM(1), <span class="string">'B'</span>, -0.42, 1.2);
1837             
1838             <span class="comment">%% organize figure data</span>
1839             figData = struct;
1840             figData.simOrder = simOrder;
1841             figData.time = time;
1842             figData.nscnt = nscnt;
1843             figData.procs = procs;
1844             figData.misfolds = misfolds;
1845             figData.damages = damages;
1846             figData.aborts = aborts;
1847             figData.enzymes = enzymes;
1848             figData.expEnzymes = expEnzymes;
1849             figData.enzProcessIdxs = enzProcessIdxs;
1850         <span class="keyword">end</span>
1851         
1852         <a name="_sub13" href="#_subfunctions" class="code">function figData = cellShape(figHandle, simBatchDir, selectedSimulations)</a>
1853             import edu.stanford.covert.cell.sim.util.PlotUtil;
1854             import edu.stanford.covert.cell.sim.util.SimulationEnsemble;
1855             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
1856             import edu.stanford.covert.util.ConstantUtil;
1857             
1858             <span class="keyword">if</span> nargin &lt; 2 || isempty(simBatchDir)
1859                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
1860             <span class="keyword">end</span>
1861             <span class="keyword">if</span> nargin &lt; 3 || isempty(selectedSimulations)
1862                 selectedSimulations = SimulationDiskUtil.getCompleteSimulations(simBatchDir);
1863             <span class="keyword">end</span>
1864             
1865             <span class="comment">%% get data</span>
1866             stateNames = {
1867                 <span class="string">'Time'</span>      <span class="string">'values'</span>             <span class="string">':'</span> <span class="string">':'</span>
1868                 <span class="string">'Mass'</span>      <span class="string">'cell'</span>               <span class="string">':'</span> <span class="string">'-sum'</span>
1869                 <span class="string">'Geometry'</span>  <span class="string">'volume'</span>             <span class="string">':'</span> <span class="string">':'</span>
1870                 <span class="string">'Geometry'</span>  <span class="string">'width'</span>              <span class="string">':'</span> <span class="string">':'</span>
1871                 <span class="string">'Geometry'</span>  <span class="string">'totalLength'</span>        <span class="string">':'</span> <span class="string">':'</span>
1872                 <span class="string">'Geometry'</span>  <span class="string">'surfaceArea'</span>        <span class="string">':'</span> <span class="string">':'</span>
1873                 <span class="string">'Geometry'</span>  <span class="string">'pinchedDiameter'</span>    <span class="string">':'</span> <span class="string">':'</span>
1874                 };
1875             states = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, <span class="string">'extract'</span>, selectedSimulations);
1876             simEndTimes = permute(max(states.Time.values, [], 3), [4 3 1 2]);
1877             
1878             states.Geometry.width(states.Geometry.width == -1) = NaN;
1879             states.Geometry.totalLength(states.Geometry.totalLength == -1) = NaN;
1880             states.Geometry.surfaceArea(states.Geometry.surfaceArea == -1) = NaN;
1881             
1882             time = permute(max(states.Time.values, [], 4), [3 4 1 2]) / 3600;
1883             mass = permute(states.Mass.cell, [3 4 1 2]) * 1e15;
1884             volume = permute(states.Geometry.volume, [3 4 1 2]) * 1e15;
1885             width = permute(states.Geometry.width, [3 4 1 2]) * 1e6;
1886             totalLength = permute(states.Geometry.totalLength, [3 4 1 2]) * 1e6;
1887             surfaceArea = permute(states.Geometry.surfaceArea, [3 4 1 2]) * 1e12;
1888             pinchedDiameter = permute(states.Geometry.pinchedDiameter, [3 4 1 2]) * 1e6;
1889             
1890             minSimEndTime = min(max(states.Time.values, [], 4));
1891             [~, simOrder] = sort(mass(minSimEndTime, :));
1892             
1893             <span class="keyword">for</span> i = 1:numel(simEndTimes)
1894                 mass(simEndTimes(i)+1:<span class="keyword">end</span>, i) = NaN;
1895                 volume(simEndTimes(i)+1:<span class="keyword">end</span>, i) = NaN;
1896                 width(simEndTimes(i)+1:<span class="keyword">end</span>, i) = NaN;
1897                 totalLength(simEndTimes(i)+1:<span class="keyword">end</span>, i) = NaN;
1898                 surfaceArea(simEndTimes(i)+1:<span class="keyword">end</span>, i) = NaN;
1899                 pinchedDiameter(simEndTimes(i)+1:<span class="keyword">end</span>, i) = NaN;
1900             <span class="keyword">end</span>
1901             
1902             clear states;
1903             
1904             <span class="comment">%% plot data</span>
1905             clf(figHandle);
1906             
1907             options = struct();
1908             options.titleStr = <span class="string">'Cell Shape'</span>;
1909             options.colorOrder = PlotUtil.getRedGreenColorOrder(simOrder);
1910             options.xdata = time;
1911             options.ylabelStr = {
1912                 <span class="string">'Mass (fg)'</span>
1913                 <span class="string">'Volume (fL)'</span>
1914                 {<span class="string">'Surface Area'</span> <span class="string">'({\mu}m^2)'</span>}
1915                 <span class="string">'Length ({\mu}m)'</span>
1916                 <span class="string">'Width ({\mu}m)'</span>
1917                 {<span class="string">'Pinched'</span> <span class="string">'Diameter ({\mu}m)'</span>}
1918                 };
1919             options.ydata = {
1920                 mass
1921                 volume
1922                 surfaceArea
1923                 totalLength
1924                 width
1925                 pinchedDiameter
1926                 };
1927             PlotUtil.multiElementPlot(figHandle, 3 * ones(6, 1), [0 time(end)], options);
1928             
1929             clear options;
1930             
1931             <span class="comment">%% organize figure data</span>
1932             figData = struct;
1933             figData.simOrder = simOrder;
1934             figData.time = time;
1935             figData.mass = mass;
1936             figData.volume = volume;
1937             figData.surfaceArea = surfaceArea;
1938             figData.totalLength = totalLength;
1939             figData.width = width;
1940             figData.pinchedDiameter = pinchedDiameter;
1941         <span class="keyword">end</span>
1942         
1943         <a name="_sub14" href="#_subfunctions" class="code">function figData = rnaPolymerases(figHandle, simBatchDir, selectedSimulations)</a>
1944             import edu.stanford.covert.cell.sim.util.PlotUtil;
1945             import edu.stanford.covert.cell.sim.util.SimulationEnsemble;
1946             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
1947             
1948             <span class="keyword">if</span> nargin &lt; 2 || isempty(simBatchDir)
1949                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
1950             <span class="keyword">end</span>
1951             <span class="keyword">if</span> nargin &lt; 3 || isempty(selectedSimulations)
1952                 selectedSimulations = SimulationDiskUtil.getCompleteSimulations(simBatchDir);
1953             <span class="keyword">end</span>
1954             
1955             <span class="comment">%% constants</span>
1956             [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep <span class="string">'1'</span>]);
1957             g = sim.gene;
1958             time = sim.state(<span class="string">'Time'</span>);
1959             rna = sim.state(<span class="string">'Rna'</span>);
1960             pc = sim.state(<span class="string">'ProteinComplex'</span>);
1961             rnaPol = sim.state(<span class="string">'RNAPolymerase'</span>);
1962             tr = sim.process(<span class="string">'TranscriptionalRegulation'</span>);
1963             sc = sim.process(<span class="string">'DNASupercoiling'</span>);
1964             pcComp = sum(pc.proteinComplexComposition(setdiff(1:<span class="keyword">end</span>, g.mRNAIndexs), :, :), 3);
1965             
1966             fcRNAIdxs = unique([tr.tuIndexs(:); sc.tuIndexs(:)]);
1967             
1968             expectedRNASyn = rna.expression(rna.matureIndexs);
1969             expectedRNASyn = expectedRNASyn / sum(expectedRNASyn);
1970             
1971             expectedRNAExp = expectedRNASyn ./ (log(2) / time.cellCycleLength + rna.decayRates(rna.matureIndexs));
1972             expectedRNAExp = expectedRNAExp / sum(expectedRNAExp);
1973             
1974             <span class="comment">%% get data</span>
1975             <span class="comment">%RNA polymerases</span>
1976             stateNames = {
1977                 <span class="string">'Time'</span>           <span class="string">'values'</span>                                    <span class="string">':'</span> <span class="string">':'</span>
1978                 <span class="string">'Mass'</span>           <span class="string">'cell'</span>                                      <span class="string">':'</span> <span class="string">'-sum'</span>
1979                 <span class="string">'RNAPolymerase'</span>  <span class="string">'nActive'</span>                                   <span class="string">':'</span> <span class="string">':'</span>
1980                 <span class="string">'RNAPolymerase'</span>  <span class="string">'nSpecificallyBound'</span>                        <span class="string">':'</span> <span class="string">':'</span>
1981                 <span class="string">'RNAPolymerase'</span>  <span class="string">'nNonSpecificallyBound'</span>                     <span class="string">':'</span> <span class="string">':'</span>
1982                 <span class="string">'RNAPolymerase'</span>  <span class="string">'nFree'</span>                                     <span class="string">':'</span> <span class="string">':'</span>
1983                 <span class="string">'RNAPolymerase'</span>  <span class="string">'transcriptionFactorBindingProbFoldChange'</span>  fcRNAIdxs 1
1984                 <span class="string">'RNAPolymerase'</span>  <span class="string">'supercoilingBindingProbFoldChange'</span>         fcRNAIdxs 1
1985                 };
1986             states = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, <span class="string">'extract'</span>, selectedSimulations);
1987             time = permute(max(states.Time.values, [], 4), [3 4 1 2]) / 3600;
1988             mass = permute(states.Mass.cell, [3 4 1 2]);
1989             
1990             simEndTimes = permute(max(states.Time.values, [], 3), [4 3 1 2]);
1991             minSimEndTime = min(simEndTimes);
1992             [~, simOrder] = sort(mass(minSimEndTime, :));
1993             
1994             foldChanges = <span class="keyword">...</span>
1995                 states.RNAPolymerase.transcriptionFactorBindingProbFoldChange .* <span class="keyword">...</span>
1996                 states.RNAPolymerase.supercoilingBindingProbFoldChange;
1997             <span class="keyword">for</span> i = 1:numel(simEndTimes)
1998                 foldChanges(:, :, simEndTimes(i)+1:<span class="keyword">end</span>, i) = NaN;
1999             <span class="keyword">end</span>
2000             
2001             actPols = states.RNAPolymerase.nActive;
2002             sbPols = states.RNAPolymerase.nSpecificallyBound;
2003             nsbPols = states.RNAPolymerase.nNonSpecificallyBound;
2004             freePols = states.RNAPolymerase.nFree;
2005             tmp = (actPols + sbPols + nsbPols + freePols);
2006             actPols = actPols ./ tmp * 100;
2007             sbPols = sbPols ./ tmp * 100;
2008             nsbPols = nsbPols ./ tmp * 100;
2009             freePols = freePols ./ tmp * 100;
2010             
2011             <span class="keyword">for</span> i = 1:numel(simEndTimes)
2012                 actPols(:, :, simEndTimes(i)+1:<span class="keyword">end</span>, i) = NaN;
2013                 sbPols(:, :, simEndTimes(i)+1:<span class="keyword">end</span>, i) = NaN;
2014                 nsbPols(:, :, simEndTimes(i)+1:<span class="keyword">end</span>, i) = NaN;
2015                 freePols(:, :, simEndTimes(i)+1:<span class="keyword">end</span>, i) = NaN;
2016             <span class="keyword">end</span>
2017             
2018             <span class="comment">%cleanup</span>
2019             clear clear states mass minSimEndTime tmp;
2020             
2021             <span class="comment">%transcription units</span>
2022             stateNames = {
2023                 <span class="string">'RNAPolymerase'</span>  <span class="string">'states'</span>                   <span class="string">':'</span>  <span class="string">':'</span>
2024                 <span class="string">'Transcript'</span>     <span class="string">'boundTranscriptionUnits'</span>  <span class="string">':'</span>  <span class="string">':'</span>
2025                 <span class="string">'Transcript'</span>     <span class="string">'boundTranscriptProgress'</span>  <span class="string">':'</span>  <span class="string">':'</span>
2026                 <span class="string">'Rna'</span>            <span class="string">'counts'</span>                   <span class="string">':'</span>  <span class="string">'-sum'</span>
2027                 <span class="string">'ProteinComplex'</span> <span class="string">'counts'</span>                   <span class="string">':'</span>  <span class="string">'-sum'</span>
2028                 };
2029             rnaSyn = zeros(numel(rna.matureIndexs), 1, 1, numel(selectedSimulations));
2030             rnaExp = zeros(numel(rna.matureIndexs), 1, 1, numel(selectedSimulations));
2031             
2032             rnaPolAbsPosDensity = zeros(numel(rna.nascentIndexs), 1 + max(rna.lengths(rna.nascentIndexs)));
2033             
2034             <span class="keyword">for</span> i = 1:numel(selectedSimulations)
2035                 states = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, <span class="string">'extract'</span>, selectedSimulations(i));
2036                 
2037                 tfs = cat(3, diff(states.Transcript.boundTranscriptProgress, [], 3) &lt; 0, states.Transcript.boundTranscriptionUnits(:, :, end) ~= rnaPol.notExistValue);
2038                 rnaSyn(:, :, :, i) = histc(states.Transcript.boundTranscriptionUnits(tfs), (1:numel(rna.matureIndexs)));
2039                 
2040                 tmp = sum(states.Rna.counts, 3);
2041                 rnaExp(:, :, :, i) = full(<span class="keyword">...</span>
2042                     + tmp(rna.processedIndexs, :) <span class="keyword">...</span>
2043                     + tmp(rna.matureIndexs, :) <span class="keyword">...</span>
2044                     + tmp(rna.aminoacylatedIndexs, :) <span class="keyword">...</span>
2045                     + tmp(rna.boundIndexs, :) <span class="keyword">...</span>
2046                     + tmp(rna.misfoldedIndexs, :) <span class="keyword">...</span>
2047                     + tmp(rna.damagedIndexs, :));
2048                 rnaExp(setdiff(1:<span class="keyword">end</span>, rna.matureMRNAIndexs), :, :, i) = <span class="keyword">...</span>
2049                     + rnaExp(setdiff(1:<span class="keyword">end</span>, rna.matureMRNAIndexs), :, :, i) <span class="keyword">...</span>
2050                     + pcComp * full(sum(reshape(sum(states.ProteinComplex.counts, 3), <span class="keyword">...</span>
2051                     [numel(pc.matureIndexs)  numel(pc.wholeCellModelIDs)/numel(pc.matureIndexs)]), 2));
2052                 
2053                 states.RNAPolymerase.states = full(states.RNAPolymerase.states);
2054                 states.Transcript.boundTranscriptionUnits = full(states.Transcript.boundTranscriptionUnits);
2055                 states.Transcript.boundTranscriptProgress = full(states.Transcript.boundTranscriptProgress);
2056                 states.RNAPolymerase.states = states.RNAPolymerase.states(:);
2057                 states.Transcript.boundTranscriptionUnits = states.Transcript.boundTranscriptionUnits(:);
2058                 states.Transcript.boundTranscriptProgress = states.Transcript.boundTranscriptProgress(:);
2059                 tfs = states.RNAPolymerase.states &gt;= rnaPol.activelyTranscribingValue;
2060                 rnaPolAbsPosDensity = <span class="keyword">...</span>
2061                     + rnaPolAbsPosDensity <span class="keyword">...</span>
2062                     + accumarray([states.Transcript.boundTranscriptionUnits(tfs)  states.Transcript.boundTranscriptProgress(tfs) + 1], ones(sum(tfs), 1), size(rnaPolAbsPosDensity));
2063                 
2064                 clear states tfs tmp;
2065             <span class="keyword">end</span>
2066             
2067             totRNASyn = sum(rnaSyn, 1);
2068             rnaSyn = rnaSyn ./ totRNASyn(ones(size(rnaSyn, 1), 1), :, :, :);
2069             meanRNASyn = mean(rnaSyn, 4);
2070             stdRNASyn = std(rnaSyn, [], 4);
2071             
2072             totRNAExp = sum(rnaExp, 1);
2073             rnaExp = rnaExp ./ totRNAExp(ones(size(rnaExp, 1), 1), :, :, :);
2074             meanRNAExp = mean(rnaExp, 4);
2075             stdRNAExp = std(rnaExp, [], 4);
2076             
2077             rnaPolRelPosDensity = zeros(numel(rna.nascentIndexs), 100);
2078             <span class="keyword">for</span> i = 1:numel(rna.nascentIndexs)
2079                 rnaPolRelPosDensity(i, :) = accumarray(<span class="keyword">...</span>
2080                     [ones(rna.lengths(rna.nascentIndexs(i)) + 1, 1)  ceil((1:rna.lengths(rna.nascentIndexs(i)) + 1)' / (rna.lengths(rna.nascentIndexs(i)) + 1) * 100)], <span class="keyword">...</span>
2081                     rnaPolAbsPosDensity(i, 1:rna.lengths(rna.nascentIndexs(i)) + 1)', [1 100]);
2082             <span class="keyword">end</span>
2083             
2084             <span class="comment">%cleanup</span>
2085             clear rnaSyn rnaExp totRNASyn totRNAExp;
2086             
2087             <span class="comment">%% plot data</span>
2088             clf(figHandle);
2089             
2090             options = struct();
2091             options.titleStr = {<span class="string">'Transcriptional Regulation'</span>; <span class="string">'RNA Polymerase'</span>; <span class="string">'RNA'</span>};
2092             options.colorOrder = PlotUtil.getRedGreenColorOrder(simOrder);
2093             axesHandles = PlotUtil.multiElementPlot(figHandle, {10 * ones(size(fcRNAIdxs)) [10 10 10 10] [10 10 10 10]}, [0 time(end)], options);
2094             
2095             <span class="comment">%left</span>
2096             <span class="keyword">for</span> i = 1:numel(fcRNAIdxs)
2097                 axesHandle = axesHandles{1}(i);
2098                 PlotUtil.plotLine(axesHandle, time, permute(foldChanges(i, :, :, :), [3 4 1 2]));
2099                 xlim(axesHandle, [0 time(end)]);
2100                 ylabel(axesHandle, rna.wholeCellModelIDs{rna.nascentIndexs(fcRNAIdxs(i))});
2101             <span class="keyword">end</span>
2102             
2103             <span class="comment">%middle</span>
2104             axesHandle = axesHandles{2}(1);
2105             plot(axesHandle, time, permute(actPols, [3 4 1 2]));
2106             xlim(axesHandle, [0 time(end)])
2107             ylim(axesHandle, [0 100])
2108             set(axesHandle, <span class="string">'YTick'</span>, [0 50 100]);
2109             ylabel(axesHandle, <span class="string">'Active'</span>);
2110             
2111             axesHandle = axesHandles{2}(2);
2112             plot(axesHandle, time, permute(sbPols, [3 4 1 2]));
2113             xlim(axesHandle, [0 time(end)])
2114             ylim(axesHandle, [0 100])
2115             set(axesHandle, <span class="string">'YTick'</span>, [0 50 100]);
2116             ylabel(axesHandle, {<span class="string">'Specifically'</span> <span class="string">'bound'</span>});
2117             
2118             axesHandle = axesHandles{2}(3);
2119             plot(axesHandle, time, permute(nsbPols, [3 4 1 2]));
2120             xlim(axesHandle, [0 time(end)])
2121             ylim(axesHandle, [0 100])
2122             set(axesHandle, <span class="string">'YTick'</span>, [0 50 100]);
2123             ylabel(axesHandle, {<span class="string">'Non-specifically'</span> <span class="string">'bound'</span>});
2124             
2125             axesHandle = axesHandles{2}(4);
2126             plot(axesHandle, time, permute(freePols, [3 4 1 2]));
2127             xlim(axesHandle, [0 time(end)])
2128             ylim(axesHandle, [0 100])
2129             set(axesHandle, <span class="string">'YTick'</span>, [0 50 100]);
2130             ylabel(axesHandle, <span class="string">'Free'</span>);
2131             
2132             <span class="comment">%right</span>
2133             axesHandle = axesHandles{3}(1);
2134             cla(axesHandle);
2135             density = smooth(sum(rnaPolAbsPosDensity, 1) / sum(simEndTimes) * 3600, 11, <span class="string">'moving'</span>);
2136             plot(axesHandle, 0:max(rna.lengths(rna.nascentIndexs)), density, <span class="string">'Color'</span>, <span class="string">'r'</span>);
2137             xlim(axesHandle, [-0.5 max(rna.lengths(rna.nascentIndexs)) + 0.5]);
2138             ylim(axesHandle, [0 max(density)]);
2139             set(axesHandle, <span class="string">'XTick'</span>, [0 max(rna.lengths(rna.nascentIndexs))]);
2140             set(axesHandle, <span class="string">'YTickMode'</span>, <span class="string">'auto'</span>);
2141             set(axesHandle, <span class="string">'XColor'</span>, <span class="string">'k'</span>);
2142             set(axesHandle, <span class="string">'YScale'</span>, <span class="string">'log'</span>)
2143             xlabel(axesHandle, <span class="string">'Absolute Position'</span>);
2144             ylabel(axesHandle, {<span class="string">'Density'</span>  <span class="string">'(h^{-1})'</span>});
2145             
2146             axesHandle = axesHandles{3}(2);
2147             density = smooth(sum(rnaPolRelPosDensity, 1) / sum(simEndTimes) * 3600, 11, <span class="string">'moving'</span>);
2148             plot(axesHandle, (1:100) - 0.5, density, <span class="string">'Color'</span>, <span class="string">'r'</span>);
2149             xlim(axesHandle, [0 100]);
2150             ylim(axesHandle, [0 max(density)]);
2151             set(axesHandle, <span class="string">'XTick'</span>, [0 100]);
2152             set(axesHandle, <span class="string">'YTickMode'</span>, <span class="string">'auto'</span>);
2153             set(axesHandle, <span class="string">'XColor'</span>, <span class="string">'k'</span>);
2154             set(axesHandle, <span class="string">'YScale'</span>, <span class="string">'log'</span>)
2155             xlabel(axesHandle, <span class="string">'Relative Position (%)'</span>);
2156             ylabel(axesHandle, {<span class="string">'Density'</span>  <span class="string">'(h^{-1})'</span>});
2157             
2158             axesHandle = axesHandles{3}(3);
2159             [~, order] = sort(expectedRNASyn);
2160             errorbar((1:numel(rna.matureIndexs))', meanRNASyn(order), stdRNASyn(order), <span class="string">'Parent'</span>, axesHandle, <span class="string">'Color'</span>, <span class="string">'g'</span>);
2161             plot(axesHandle, (1:numel(rna.matureIndexs))', expectedRNASyn(order), <span class="string">'Color'</span>, <span class="string">'b'</span>);
2162             xlabel(axesHandle, <span class="string">'Transcription Units'</span>);
2163             xlim(axesHandle, [0.5 numel(rna.matureIndexs)+0.5])
2164             ylabel(axesHandle, <span class="string">'Synthesis'</span>);
2165             ylim(axesHandle, [0 0.05]);
2166             set(axesHandle, <span class="string">'YTick'</span>, 0:0.05:0.05);
2167             
2168             axesHandle = axesHandles{3}(4);
2169             [~, order] = sort(expectedRNAExp);
2170             errorbar((1:numel(rna.matureIndexs))', meanRNAExp(order), stdRNAExp(order), <span class="string">'Parent'</span>, axesHandle, <span class="string">'Color'</span>, <span class="string">'g'</span>);
2171             plot(axesHandle, (1:numel(rna.matureIndexs))', expectedRNAExp(order), <span class="string">'Color'</span>, <span class="string">'b'</span>);
2172             xlabel(axesHandle, <span class="string">'Transcription Units'</span>);
2173             xlim(axesHandle, [0.5 numel(rna.matureIndexs)+0.5])
2174             ylabel(axesHandle, <span class="string">'Expression'</span>);
2175             ylim(axesHandle, [0 0.05]);
2176             set(axesHandle, <span class="string">'YTick'</span>, 0:0.05:0.05);
2177             
2178             <span class="comment">%y axes</span>
2179             PlotUtil.alignYAxesLabels(axesHandles{1});
2180             PlotUtil.alignYAxesLabels(axesHandles{2});
2181             PlotUtil.alignYAxesLabels(axesHandles{3});
2182             PlotUtil.offsetYAxes(axesHandles{1}, 0.03);
2183             PlotUtil.offsetYAxes(axesHandles{2}, 0.03);
2184             PlotUtil.offsetYAxes(axesHandles{3}, 0.03);
2185             
2186             <span class="comment">%% organize figure data</span>
2187             figData = struct;
2188             figData.simEndTimes = simEndTimes;
2189             figData.simOrder = simOrder;
2190             figData.time = time;
2191             figData.foldChanges = foldChanges;
2192             figData.actPols = actPols;
2193             figData.sbPols = sbPols;
2194             figData.nsbPols = nsbPols;
2195             figData.freePols = freePols;
2196             figData.meanRNASyn = meanRNASyn;
2197             figData.stdRNASyn = stdRNASyn;
2198             figData.meanRNAExp = meanRNAExp;
2199             figData.stdRNAExp = stdRNAExp;
2200             figData.expectedRNASyn = expectedRNASyn;
2201             figData.expectedRNAExp = expectedRNAExp;
2202             figData.rnaPolAbsPosDensity = rnaPolAbsPosDensity;
2203             figData.rnaPolRelPosDensity = rnaPolRelPosDensity;
2204         <span class="keyword">end</span>
2205         
2206         <a name="_sub15" href="#_subfunctions" class="code">function figData = ribosomes(figHandle, simBatchDir, selectedSimulations)</a>
2207             import edu.stanford.covert.cell.sim.util.PlotUtil;
2208             import edu.stanford.covert.cell.sim.util.SimulationEnsemble;
2209             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
2210             
2211             <span class="keyword">if</span> nargin &lt; 2 || isempty(simBatchDir)
2212                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
2213             <span class="keyword">end</span>
2214             <span class="keyword">if</span> nargin &lt; 3 || isempty(selectedSimulations)
2215                 selectedSimulations = SimulationDiskUtil.getCompleteSimulations(simBatchDir);
2216             <span class="keyword">end</span>
2217             
2218             <span class="comment">%% constants</span>
2219             [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep <span class="string">'1'</span>]);
2220             g = sim.gene;
2221             time = sim.state(<span class="string">'Time'</span>);
2222             rna = sim.state(<span class="string">'Rna'</span>);
2223             pm = sim.state(<span class="string">'ProteinMonomer'</span>);
2224             pc = sim.state(<span class="string">'ProteinComplex'</span>);
2225             rib = sim.state(<span class="string">'Ribosome'</span>);
2226             pcComp = sum(pc.proteinComplexComposition(g.mRNAIndexs, :, :), 3);
2227             
2228             expectedMonSyn = rna.matureRNAGeneComposition(g.mRNAIndexs, :) * rna.expression(rna.matureIndexs);
2229             expectedMonSyn = expectedMonSyn / sum(expectedMonSyn);
2230             
2231             expectedMonExp = expectedMonSyn ./ (log(2) / time.cellCycleLength + pm.decayRates(pm.matureIndexs));
2232             expectedMonExp = expectedMonExp / sum(expectedMonExp);
2233             
2234             <span class="comment">%% get data</span>
2235             <span class="comment">%ribosome states</span>
2236             stateNames = {
2237                 <span class="string">'Time'</span>           <span class="string">'values'</span>          <span class="string">':'</span> <span class="string">':'</span>
2238                 <span class="string">'Mass'</span>           <span class="string">'cell'</span>            <span class="string">':'</span> <span class="string">'-sum'</span>
2239                 <span class="string">'Ribosome'</span>       <span class="string">'nActive'</span>         <span class="string">':'</span> <span class="string">':'</span>
2240                 <span class="string">'Ribosome'</span>       <span class="string">'nStalled'</span>        <span class="string">':'</span> <span class="string">':'</span>
2241                 };
2242             states = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, <span class="string">'extract'</span>, selectedSimulations);
2243             time = permute(max(states.Time.values, [], 4), [3 4 1 2]) / 3600;
2244             mass = permute(states.Mass.cell, [3 4 1 2]);
2245             
2246             simEndTimes = permute(max(states.Time.values, [], 3), [4 3 1 2]);
2247             minSimEndTime = min(simEndTimes);
2248             [~, simOrder] = sort(mass(minSimEndTime, :));
2249             
2250             actRibs = states.Ribosome.nActive;
2251             stalledRibs = states.Ribosome.nStalled;
2252             tmp = (actRibs + stalledRibs);
2253             actRibs = actRibs ./ tmp * 100;
2254             stalledRibs = stalledRibs ./ tmp * 100;
2255             
2256             <span class="keyword">for</span> i = 1:numel(simEndTimes)
2257                 actRibs(:, :, simEndTimes(i)+1:<span class="keyword">end</span>, i) = NaN;
2258                 stalledRibs(:, :, simEndTimes(i)+1:<span class="keyword">end</span>, i) = NaN;
2259             <span class="keyword">end</span>
2260             
2261             <span class="comment">%cleanup</span>
2262             clear states mass minSimEndTime tmp;
2263             
2264             <span class="comment">%synthesis, expression</span>
2265             stateNames = {
2266                 <span class="string">'Ribosome'</span>       <span class="string">'states'</span>          <span class="string">':'</span>  <span class="string">':'</span>
2267                 <span class="string">'Ribosome'</span>       <span class="string">'boundMRNAs'</span>      <span class="string">':'</span>  <span class="string">':'</span>
2268                 <span class="string">'Ribosome'</span>       <span class="string">'mRNAPositions'</span>   <span class="string">':'</span>  <span class="string">':'</span>
2269                 <span class="string">'ProteinMonomer'</span> <span class="string">'counts'</span>          <span class="string">':'</span>  <span class="string">'-sum'</span>
2270                 <span class="string">'ProteinComplex'</span> <span class="string">'counts'</span>          <span class="string">':'</span>  <span class="string">'-sum'</span>
2271                 };
2272             monSyn = zeros(numel(pm.matureIndexs), 1, 1, numel(selectedSimulations));
2273             monExp = zeros(numel(pm.matureIndexs), 1, 1, numel(selectedSimulations));
2274             cpxExp = zeros(numel(pc.matureIndexs), 1, 1, numel(selectedSimulations));
2275             ribAbsPosDensity = zeros(numel(pm.nascentIndexs), 1 + max(pm.lengths(pm.nascentIndexs)));
2276             <span class="keyword">for</span> i = 1:numel(selectedSimulations)
2277                 states = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, <span class="string">'extract'</span>, selectedSimulations(i));
2278                 
2279                 boundMRNAs = states.Ribosome.boundMRNAs;
2280                 mRNAPositions = states.Ribosome.mRNAPositions;
2281                 tfs = cat(3, diff(mRNAPositions, [], 3) &lt; 0, boundMRNAs(:, :, end) ~= rib.notExistValue);
2282                 monSyn(:, :, :, i) = histc(boundMRNAs(tfs), (1:numel(pm.matureIndexs)));
2283                 
2284                 monExp(:, :, :, i) = full(sum(reshape(sum(states.ProteinMonomer.counts, 3), <span class="keyword">...</span>
2285                     [numel(pm.matureIndexs)  numel(pm.wholeCellModelIDs)/numel(pm.matureIndexs)]), 2));
2286                 cpxExp(:, :, :, i) = full(sum(reshape(sum(states.ProteinComplex.counts, 3), <span class="keyword">...</span>
2287                     [numel(pc.matureIndexs)  numel(pc.wholeCellModelIDs)/numel(pc.matureIndexs)]), 2));
2288                 
2289                 states.Ribosome.states = full(states.Ribosome.states);
2290                 states.Ribosome.boundMRNAs = full(states.Ribosome.boundMRNAs);
2291                 states.Ribosome.mRNAPositions = full(states.Ribosome.mRNAPositions);
2292                 states.Ribosome.states = states.Ribosome.states(:);
2293                 states.Ribosome.boundMRNAs = states.Ribosome.boundMRNAs(:);
2294                 states.Ribosome.mRNAPositions = states.Ribosome.mRNAPositions(:);
2295                 tfs = states.Ribosome.states == rib.activeValue;
2296                 ribAbsPosDensity = <span class="keyword">...</span>
2297                     + ribAbsPosDensity <span class="keyword">...</span>
2298                     + accumarray([states.Ribosome.boundMRNAs(tfs)  states.Ribosome.mRNAPositions(tfs) + 1], ones(sum(tfs), 1), size(ribAbsPosDensity));
2299                 
2300                 clear states boundMRNAs mRNAPositions tfs;
2301             <span class="keyword">end</span>
2302             
2303             totMonSyn = sum(monSyn, 1);
2304             monSyn = monSyn ./ totMonSyn(ones(size(monSyn, 1), 1), :, :, :);
2305             meanMonSyn = mean(monSyn, 4);
2306             stdMonSyn = std(monSyn, [], 4);
2307             
2308             <span class="keyword">for</span> j = 1:size(monExp, 4)
2309                 monExp(:, :, :, j) = <span class="keyword">...</span>
2310                     + monExp(:, :, :, j) <span class="keyword">...</span>
2311                     + pcComp * cpxExp(:, :, :, j);
2312             <span class="keyword">end</span>
2313             totMonExp = sum(monExp, 1);
2314             monExp = monExp ./ totMonExp(ones(size(monExp, 1), 1), :, :, :);
2315             meanMonExp = mean(monExp, 4);
2316             stdMonExp = std(monExp, [], 4);
2317             
2318             ribRelPosDensity = zeros(numel(pm.nascentIndexs), 100);
2319             <span class="keyword">for</span> i = 1:numel(pm.nascentIndexs)
2320                 ribRelPosDensity(i, :) = accumarray(<span class="keyword">...</span>
2321                     [ones(pm.lengths(pm.nascentIndexs(i)) + 1, 1)  ceil((1:pm.lengths(pm.nascentIndexs(i)) + 1)' / (pm.lengths(pm.nascentIndexs(i)) + 1) * 100)], <span class="keyword">...</span>
2322                     ribAbsPosDensity(i, 1:pm.lengths(pm.nascentIndexs(i)) + 1)', [1 100]);
2323             <span class="keyword">end</span>
2324             
2325             <span class="comment">%cleanup</span>
2326             clear monExp monSyn cpxExp totMonSyn totMonExp;
2327             
2328             <span class="comment">%% plot data</span>
2329             clf(figHandle);
2330             
2331             options = struct();
2332             options.titleStr = {<span class="string">'States'</span>; <span class="string">'Monomers'</span>};
2333             options.colorOrder = PlotUtil.getRedGreenColorOrder(simOrder);
2334             axesHandles = PlotUtil.multiElementPlot(figHandle, {[10 10] [10 10 10 10]}, [0 time(end)], options);
2335             
2336             <span class="comment">%left</span>
2337             axesHandle = axesHandles{1}(1);
2338             plot(axesHandle, time, permute(actRibs, [3 4 1 2]));
2339             xlim(axesHandle, [0 time(end)])
2340             ylim(axesHandle, [0 100])
2341             set(axesHandle, <span class="string">'YTick'</span>, [0 50 100]);
2342             ylabel(axesHandle, <span class="string">'Active'</span>);
2343             
2344             axesHandle = axesHandles{1}(2);
2345             plot(axesHandle, time, permute(stalledRibs, [3 4 1 2]));
2346             xlim(axesHandle, [0 time(end)])
2347             ylim(axesHandle, [0 100])
2348             set(axesHandle, <span class="string">'YTick'</span>, [0 50 100]);
2349             ylabel(axesHandle, <span class="string">'Stalled'</span>);
2350             
2351             <span class="comment">%right</span>
2352             axesHandle = axesHandles{2}(1);
2353             density = smooth(sum(ribAbsPosDensity, 1) / sum(simEndTimes) * 3600, 11, <span class="string">'moving'</span>);
2354             plot(axesHandle, 0:max(pm.lengths(pm.nascentIndexs)), density, <span class="string">'Color'</span>, <span class="string">'r'</span>);
2355             xlim(axesHandle, [-0.5 max(pm.lengths(pm.nascentIndexs)) + 0.5]);
2356             ylim(axesHandle, [0 max(density)]);
2357             set(axesHandle, <span class="string">'XTick'</span>, [0 max(pm.lengths(pm.nascentIndexs))]);
2358             set(axesHandle, <span class="string">'YTickMode'</span>, <span class="string">'auto'</span>, <span class="string">'YScale'</span>, <span class="string">'log'</span>);
2359             set(axesHandle, <span class="string">'XColor'</span>, <span class="string">'k'</span>);
2360             xlabel(axesHandle, <span class="string">'Absolute Position'</span>);
2361             ylabel(axesHandle, {<span class="string">'Density'</span>  <span class="string">'(h^{-1})'</span>});
2362             
2363             axesHandle = axesHandles{2}(2);
2364             density = smooth(sum(ribRelPosDensity, 1) / sum(simEndTimes) * 3600, 11, <span class="string">'moving'</span>);
2365             plot(axesHandle, (1:100) - 0.5, density, <span class="string">'Color'</span>, <span class="string">'r'</span>);
2366             xlim(axesHandle, [0 100]);
2367             ylim(axesHandle, [0 max(density)]);
2368             set(axesHandle, <span class="string">'XTick'</span>, [0 100]);
2369             set(axesHandle, <span class="string">'YTickMode'</span>, <span class="string">'auto'</span>, <span class="string">'YScale'</span>, <span class="string">'log'</span>);
2370             set(axesHandle, <span class="string">'XColor'</span>, <span class="string">'k'</span>);
2371             xlabel(axesHandle, <span class="string">'Relative Position (%)'</span>);
2372             ylabel(axesHandle, {<span class="string">'Density'</span>  <span class="string">'(h^{-1})'</span>});
2373             
2374             axesHandle = axesHandles{2}(3);
2375             [~, order] = sort(expectedMonSyn);
2376             errorbar((1:numel(pm.matureIndexs))', meanMonSyn(order), stdMonSyn(order), <span class="string">'Parent'</span>, axesHandle, <span class="string">'Color'</span>, <span class="string">'g'</span>);
2377             plot(axesHandle, (1:numel(pm.matureIndexs))', expectedMonSyn(order), <span class="string">'Color'</span>, <span class="string">'b'</span>);
2378             xlabel(axesHandle, <span class="string">'Protein-Coding Genes'</span>);
2379             xlim(axesHandle, [0.5 numel(pm.matureIndexs)+0.5])
2380             ylabel(axesHandle, <span class="string">'Synthesis'</span>);
2381             ylim(axesHandle, [min(meanMonSyn(order) - stdMonSyn(order)) max(meanMonSyn(order) + stdMonSyn(order))]);
2382             set(axesHandle, <span class="string">'YTick'</span>, 0:0.01:0.02);
2383             
2384             axesHandle = axesHandles{2}(4);
2385             [~, order] = sort(expectedMonExp);
2386             errorbar((1:numel(pm.matureIndexs))', meanMonExp(order), stdMonExp(order), <span class="string">'Parent'</span>, axesHandle, <span class="string">'Color'</span>, <span class="string">'g'</span>);
2387             plot(axesHandle, (1:numel(pm.matureIndexs))', expectedMonExp(order), <span class="string">'Color'</span>, <span class="string">'b'</span>);
2388             xlabel(axesHandle, <span class="string">'Protein-Coding Genes'</span>);
2389             xlim(axesHandle, [0.5 numel(pm.matureIndexs)+0.5])
2390             ylabel(axesHandle, <span class="string">'Expression'</span>);
2391             ylim(axesHandle, [min(meanMonExp(order) - stdMonExp(order)) max(meanMonExp(order) + stdMonExp(order))]);
2392             set(axesHandle, <span class="string">'YTick'</span>, 0:0.01:0.02);
2393             
2394             <span class="comment">%y axes</span>
2395             PlotUtil.alignYAxesLabels(axesHandles{1});
2396             PlotUtil.alignYAxesLabels(axesHandles{2});
2397             PlotUtil.offsetYAxes(axesHandles{1}, 0.03);
2398             PlotUtil.offsetYAxes(axesHandles{2}, 0.03);
2399             
2400             <span class="comment">%% organize figure data</span>
2401             figData = struct;
2402             figData.simEndTimes = simEndTimes;
2403             figData.simOrder = simOrder;
2404             figData.time = time;
2405             figData.actRibs = actRibs;
2406             figData.stalledRibs = stalledRibs;
2407             figData.expectedMonSyn = expectedMonSyn;
2408             figData.expectedMonExp = expectedMonExp;
2409             figData.meanMonSyn = meanMonSyn;
2410             figData.stdMonSyn = stdMonSyn;
2411             figData.meanMonExp = meanMonExp;
2412             figData.stdMonExp = stdMonExp;
2413             figData.ribAbsPosDensity = ribAbsPosDensity;
2414             figData.ribRelPosDensity = ribRelPosDensity;
2415         <span class="keyword">end</span>
2416         
2417         <a name="_sub16" href="#_subfunctions" class="code">function figData = macromoleculeExpression(figHandle, simBatchDir, selectedSimulations)</a>
2418             import edu.stanford.covert.cell.sim.util.PlotUtil;
2419             import edu.stanford.covert.cell.sim.util.SimulationEnsemble;
2420             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
2421             
2422             <span class="comment">%% get data</span>
2423             <span class="keyword">if</span> nargin &lt; 2 || isempty(simBatchDir)
2424                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
2425             <span class="keyword">end</span>
2426             <span class="keyword">if</span> nargin &lt; 3 || isempty(selectedSimulations)
2427                 selectedSimulations = SimulationDiskUtil.getCompleteSimulations(simBatchDir);
2428             <span class="keyword">end</span>
2429             
2430             <span class="comment">%% constants</span>
2431             [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep <span class="string">'1'</span>]);
2432             g = sim.gene;
2433             m = sim.state(<span class="string">'Metabolite'</span>);
2434             rna = sim.state(<span class="string">'Rna'</span>);
2435             mon = sim.state(<span class="string">'ProteinMonomer'</span>);
2436             cpx = sim.state(<span class="string">'ProteinComplex'</span>);
2437             
2438             rnaPcComp = sum(cpx.proteinComplexComposition(setdiff(1:<span class="keyword">end</span>, g.mRNAIndexs), :, :), 3);
2439             monPcComp = sum(cpx.proteinComplexComposition(g.mRNAIndexs, :, :), 3);
2440             
2441             essGene = strcmp(g.essential, <span class="string">'Y'</span>);
2442             
2443             <span class="comment">%% get data</span>
2444             stateNames = {
2445                 <span class="string">'Rna'</span>             <span class="string">'counts'</span>
2446                 <span class="string">'ProteinMonomer'</span>  <span class="string">'counts'</span>
2447                 <span class="string">'ProteinComplex'</span>  <span class="string">'counts'</span>
2448                 };
2449             rnaFreq = zeros(1000, numel(rna.matureIndexs));
2450             monFreq = zeros(10000, numel(mon.matureIndexs));
2451             <span class="keyword">for</span> i = 1:numel(selectedSimulations)
2452                 states = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, <span class="string">'extract'</span>, selectedSimulations(i));
2453                 
2454                 tmp = permute(sum(states.Rna.counts, 2), [1 3 2]);
2455                 rnaCnts = <span class="keyword">...</span>
2456                     + rna.nascentRNAMatureRNAComposition * full(tmp(rna.nascentIndexs, :)) <span class="keyword">...</span>
2457                     + full(<span class="keyword">...</span>
2458                     + tmp(rna.processedIndexs, :) <span class="keyword">...</span>
2459                     + tmp(rna.matureIndexs, :) <span class="keyword">...</span>
2460                     + tmp(rna.boundIndexs, :) <span class="keyword">...</span>
2461                     + tmp(rna.misfoldedIndexs, :) <span class="keyword">...</span>
2462                     + tmp(rna.damagedIndexs, :) <span class="keyword">...</span>
2463                     + tmp(rna.aminoacylatedIndexs, :) <span class="keyword">...</span>
2464                     );
2465                 clear tmp;
2466                 
2467                 monCnts = full(permute(sum(reshape(sum(states.ProteinMonomer.counts, 2), [numel(mon.matureIndexs)  size(states.ProteinMonomer.counts, 1) / numel(mon.matureIndexs) size(states.ProteinMonomer.counts, 3)]), 2), [1 3 2]));
2468                 cpxCnts = full(permute(sum(reshape(sum(states.ProteinComplex.counts, 2), [numel(cpx.matureIndexs)  size(states.ProteinComplex.counts, 1) / numel(cpx.matureIndexs) size(states.ProteinComplex.counts, 3)]), 2), [1 3 2]));
2469                 rnaCnts(setdiff(1:<span class="keyword">end</span>, rna.matureMRNAIndexs), :) = <span class="keyword">...</span>
2470                     + rnaCnts(setdiff(1:<span class="keyword">end</span>, rna.matureMRNAIndexs), :) <span class="keyword">...</span>
2471                     + rnaPcComp * cpxCnts;
2472                 monCnts = <span class="keyword">...</span>
2473                     + monCnts <span class="keyword">...</span>
2474                     + monPcComp * cpxCnts;
2475                 
2476                 idxs = (1:numel(rna.matureIndexs))';
2477                 idxs = idxs(:, ones(size(rnaCnts, 2), 1));
2478                 rnaFreq = <span class="keyword">...</span>
2479                     + rnaFreq <span class="keyword">...</span>
2480                     + accumarray([rnaCnts(:)+1 idxs(:)], ones(numel(rnaCnts), 1), size(rnaFreq));
2481                 
2482                 idxs = (1:numel(mon.matureIndexs))';
2483                 idxs = idxs(:, ones(size(monCnts, 2), 1));
2484                 monFreq = <span class="keyword">...</span>
2485                     + monFreq <span class="keyword">...</span>
2486                     + accumarray([monCnts(:)+1 idxs(:)], ones(numel(monCnts), 1), size(monFreq));
2487                 
2488                 clear states rnaCnts monCnts cpxCnts idxs;
2489             <span class="keyword">end</span>
2490             
2491             rnaFreq = rnaFreq(1:find(any(rnaFreq, 2), 1, <span class="string">'last'</span>), :);
2492             monFreq = monFreq(1:find(any(monFreq, 2), 1, <span class="string">'last'</span>), :);
2493             
2494             meanRnaCnt = ((0:size(rnaFreq, 1)-1) * rnaFreq) ./ sum(rnaFreq, 1);
2495             stdRnaCnt = sqrt((((1:size(rnaFreq, 1)).^2) * rnaFreq) ./ sum(rnaFreq, 1) - meanRnaCnt.^2);
2496             meanMRnaCnt = meanRnaCnt * rna.matureRNAGeneComposition(g.mRNAIndexs, :)';
2497             stdMRnaCnt = stdRnaCnt * rna.matureRNAGeneComposition(g.mRNAIndexs, :)';
2498             meanMonCnt = ((0:size(monFreq, 1)-1) * monFreq) ./ sum(monFreq, 1);
2499             stdMonCnt = sqrt((((1:size(monFreq, 1)).^2) * monFreq) ./ sum(monFreq, 1) - meanMonCnt.^2);
2500             
2501             rnaLognParam = NaN(numel(rna.matureIndexs), 2);
2502             rnaLognParamCI = NaN(numel(rna.matureIndexs), 2, 2);
2503             rnaNonNormPValue = zeros(numel(rna.matureIndexs), 1);
2504             <span class="keyword">for</span> i = 1:numel(rna.matureIndexs)
2505                 cnts = rude(rnaFreq(:, i), 0:size(rnaFreq, 1) - 1);
2506                 <span class="keyword">if</span> ~rnaFreq(1, i)
2507                     [rnaLognParam(i, :), rnaLognParamCI(i, :, :)] = lognfit(cnts);
2508                 <span class="keyword">end</span>
2509                 <span class="keyword">if</span> cnts &gt; 5e3
2510                     cnts = cnts(randsample(numel(cnts), 5e3, false));
2511                 <span class="keyword">end</span>
2512                 [~, rnaNonNormPValue(i)] = swtest(log10(cnts), 0.05, 0);
2513             <span class="keyword">end</span>
2514             
2515             monGamParam = zeros(numel(mon.matureIndexs), 2);
2516             monGamParamCI = zeros(numel(mon.matureIndexs), 2, 2);
2517             monNonNormPValue = zeros(numel(mon.matureIndexs), 1);
2518             <span class="keyword">for</span> i = 1:numel(mon.matureIndexs)
2519                 cnts = rude(monFreq(:, i), 0:size(monFreq, 1) - 1);
2520                 [monGamParam(i, :), monGamParamCI(i, :, :)] = gamfit(cnts);
2521                 <span class="keyword">if</span> cnts &gt; 5e3
2522                     cnts = cnts(randsample(numel(cnts), 5e3, false));
2523                 <span class="keyword">end</span>
2524                 [~, monNonNormPValue(i)] = swtest(cnts, 0.05, 0);
2525             <span class="keyword">end</span>
2526             
2527             <span class="comment">%% plot data</span>
2528             clf(figHandle);
2529             
2530             [axesHandles, xAxesHandles, ~, offsetAxesHandles] = PlotUtil.multiElementPlot(figHandle, {1 * ones(3, 1); 1 * ones(2,1); 3 * ones(3, 1); 3 * ones(4, 1)}, [0 1]);
2531             delete(cell2mat(xAxesHandles));
2532             
2533             set(cell2mat(offsetAxesHandles), <span class="string">'visible'</span>, <span class="string">'on'</span>);
2534             set(cell2mat(axesHandles), <span class="string">'xcolor'</span>, <span class="string">'k'</span>, <span class="string">'xtickmode'</span>, <span class="string">'auto'</span>, <span class="string">'xlimmode'</span>, <span class="string">'auto'</span>, <span class="string">'ylimmode'</span>, <span class="string">'auto'</span>);
2535             
2536             <span class="comment">%mRNA count distribution for particular gene</span>
2537             [~, idx] = max(meanMRnaCnt);
2538             tmp = rnaFreq(:, rna.matureMRNAIndexs(idx)) / sum(rnaFreq(:, rna.matureMRNAIndexs(idx)));
2539             axesHandle = axesHandles{1}(1);
2540             cla(axesHandle);
2541             bar(axesHandle, (1:size(tmp, 1))'-1, tmp, <span class="string">'EdgeColor'</span>, <span class="string">'none'</span>, <span class="string">'BarWidth'</span>, 0.9);
2542             minX = find(tmp, 1, <span class="string">'first'</span>) - 1.5;
2543             maxX = find(tmp, 1, <span class="string">'last'</span>) - 0.5;
2544             xlim(axesHandle, [minX maxX]);
2545             ylim(axesHandle, [0 max(tmp)]);
2546             xlabel(axesHandle, <span class="string">'mRNA Count'</span>);
2547             ylabel(axesHandle, <span class="string">'Freq'</span>);
2548             
2549             <span class="comment">%rRNA count distribution for particular gene</span>
2550             x = rnaFreq(:, rna.matureRRNAIndexs(1));
2551             cnts = [];
2552             <span class="keyword">for</span> j = 0:numel(x) - 1
2553                 cnts = [cnts; j(ones(x(j+1), 1), 1)]; <span class="comment">%#ok&lt;AGROW&gt;</span>
2554             <span class="keyword">end</span>
2555             axesHandle = axesHandles{1}(2);
2556             cla(axesHandle);
2557             hold(axesHandle, <span class="string">'on'</span>);
2558             
2559             edges = logspace(log10(min(cnts)), log10(max(cnts)), 200);
2560             freq = histc(cnts, edges);
2561             freq = freq / sum(freq);
2562             bar(axesHandle, log10(edges), freq, <span class="string">'EdgeColor'</span>, <span class="string">'none'</span>, <span class="string">'BarWidth'</span>, 2);
2563             set(axesHandle, <span class="string">'XScale'</span>, <span class="string">'linear'</span>);
2564             xlim(axesHandle, log10(edges([1 end])))
2565             set(axesHandle, <span class="string">'XTickMode'</span>, <span class="string">'auto'</span>);
2566             set(axesHandle, <span class="string">'XTick'</span>, 10.^get(axesHandle, <span class="string">'XTick'</span>));
2567             ylim(axesHandle, [0 max(freq)])
2568             
2569             y = lognpdf(edges, rnaLognParam(rna.matureRRNAIndexs(2), 1), rnaLognParam(rna.matureRRNAIndexs(2), 2));
2570             plot(axesHandle, log10(edges), y, <span class="string">'r'</span>);
2571             
2572             xlabel(axesHandle, <span class="string">'rRNA Count'</span>);
2573             ylabel(axesHandle, <span class="string">'Freq'</span>);
2574             
2575             <span class="comment">%monomer count distribution for particular gene</span>
2576             tmp = monFreq(:, 1) ./ sum(monFreq(:, 1));
2577             axesHandle = axesHandles{1}(3);
2578             cla(axesHandle);
2579             bar(axesHandle, (1:size(tmp, 1))'-1, tmp, <span class="string">'EdgeColor'</span>, <span class="string">'none'</span>, <span class="string">'BarWidth'</span>, 2);
2580             minX = find(tmp, 1, <span class="string">'first'</span>) - 1.5;
2581             maxX = find(tmp, 1, <span class="string">'last'</span>) - 0.5;
2582             xlim(axesHandle, [minX maxX])
2583             ylim([0 max(tmp)]);
2584             
2585             x = linspace(max(0, minX), maxX, 100);
2586             y = gampdf(x, monGamParam(1, 1), monGamParam(1, 2));
2587             plot(axesHandle, x, y, <span class="string">'r'</span>);
2588             
2589             xlabel(axesHandle, <span class="string">'Protein Count'</span>);
2590             ylabel(axesHandle, <span class="string">'Freq'</span>);
2591             
2592             <span class="comment">%freq of strains with mean protein number</span>
2593             axesHandle = axesHandles{2}(1);
2594             cla(axesHandle);
2595             edges = logspace(log10(min(meanMonCnt)), log10(max(meanMonCnt)), 10);
2596             cnts = [
2597                 histc(meanMonCnt, edges)
2598                 histc(meanMonCnt(essGene(g.mRNAIndexs)), edges)
2599                 ];
2600             h = bar(axesHandle, log10(edges'), cnts', <span class="string">'grouped'</span>);
2601             set(h, <span class="string">'BarWidth'</span>, 2);
2602             set(h(1), <span class="string">'EdgeColor'</span>, <span class="string">'none'</span>)
2603             set(h(2), <span class="string">'EdgeColor'</span>, <span class="string">'none'</span>)
2604             set(axesHandle, <span class="string">'XScale'</span>, <span class="string">'linear'</span>);
2605             xlim(axesHandle, log10(edges([1 end])))
2606             set(axesHandle, <span class="string">'XTickMode'</span>, <span class="string">'auto'</span>);
2607             set(axesHandle, <span class="string">'XTick'</span>, 10.^get(axesHandle, <span class="string">'XTick'</span>));
2608             xlabel(axesHandle, <span class="string">'Prot Exp'</span>);
2609             ylabel(axesHandle, <span class="string">'Freq'</span>);
2610             
2611             <span class="comment">%noise vs mean protein number</span>
2612             axesHandle = axesHandles{2}(2);
2613             cla(axesHandle);
2614             hold(axesHandle, <span class="string">'on'</span>);
2615             noise = (stdMonCnt ./ meanMonCnt).^2;
2616             x = quantile(meanMonCnt, [0.01 0.99])';
2617             y = 5./quantile(meanMonCnt, [0.01 0.99])';
2618             plot(axesHandle, meanMonCnt, noise, <span class="string">'k.'</span>);
2619             plot(axesHandle, x, y, <span class="string">'r-'</span>);
2620             plot(axesHandle, [min(meanMonCnt) max(meanMonCnt)], quantile(noise, 0.01) * [1 1], <span class="string">'b-'</span>);
2621             ylim(axesHandle, [min(noise) max(noise)])
2622             xlim(axesHandle, [min(meanMonCnt) max(meanMonCnt)])
2623             set(axesHandle, <span class="string">'XScale'</span>, <span class="string">'log'</span>);
2624             set(axesHandle, <span class="string">'YScale'</span>, <span class="string">'log'</span>);
2625             xlabel(axesHandle, <span class="string">'Mean Protein Number (\mu)'</span>);
2626             ylabel(axesHandle, <span class="string">'Protein Noise (\sigma^2 / \mu^2)'</span>);
2627             
2628             <span class="comment">%Protein vs RNA expression</span>
2629             axesHandle = axesHandles{3}(1);
2630             cla(axesHandle);
2631             plot(axesHandle, meanMRnaCnt, meanMonCnt, <span class="string">'.'</span>);
2632             xlim(axesHandle, [min(meanMRnaCnt) max(meanMRnaCnt)])
2633             ylim(axesHandle, [min(meanMonCnt) max(meanMonCnt)])
2634             xlabel(axesHandle, <span class="string">'Mean mRNA Level'</span>);
2635             ylabel(axesHandle, <span class="string">'Mean Protein Number'</span>);
2636             set(axesHandle, <span class="string">'XScale'</span>, <span class="string">'log'</span>);
2637             set(axesHandle, <span class="string">'YScale'</span>, <span class="string">'log'</span>);
2638             
2639             <span class="comment">%RNA expression vs noise</span>
2640             axesHandle = axesHandles{3}(2);
2641             cla(axesHandle);
2642             noise = (stdMRnaCnt ./ meanMRnaCnt) .^2;
2643             plot(axesHandle, meanMRnaCnt, noise, <span class="string">'.'</span>);
2644             xlim(axesHandle, [min(meanMRnaCnt) max(meanMRnaCnt)])
2645             ylim(axesHandle, [min(noise) max(noise)])
2646             xlabel(axesHandle, <span class="string">'Mean MRNA Number'</span>);
2647             ylabel(axesHandle, <span class="string">'mRNA Noise'</span>);
2648             set(axesHandle, <span class="string">'XScale'</span>, <span class="string">'log'</span>);
2649             set(axesHandle, <span class="string">'YScale'</span>, <span class="string">'log'</span>);
2650             
2651             <span class="comment">%mRNA fano factor frequence</span>
2652             axesHandle = axesHandles{3}(3);
2653             cla(axesHandle);
2654             tfs = ~isnan(stdMRnaCnt) &amp; meanMRnaCnt &gt; 0;
2655             hist(axesHandle, (stdMRnaCnt(tfs) .^2) ./ meanMRnaCnt(tfs), 200);
2656             h = findobj(axesHandle, <span class="string">'Type'</span>, <span class="string">'patch'</span>);
2657             set(h, <span class="string">'EdgeColor'</span>, <span class="string">'none'</span>);
2658             xlim(axesHandle, [2 20]);
2659             xlabel(axesHandle, <span class="string">'mRNA Fano Factor'</span>);
2660             ylabel(axesHandle, <span class="string">'Number of Strains'</span>);
2661             
2662             <span class="comment">%mRNA gamma a vs b</span>
2663             axesHandle = axesHandles{4}(1);
2664             cla(axesHandle)
2665             plot(axesHandle, monGamParam(:, 1), monGamParam(:, 2), <span class="string">'.'</span>);
2666             xlim(axesHandle, [min(monGamParam(:, 1)) max(monGamParam(:, 1))])
2667             ylim(axesHandle, [min(monGamParam(:, 2)) max(monGamParam(:, 2))])
2668             xlabel(axesHandle, <span class="string">'a Value'</span>);
2669             ylabel(axesHandle, <span class="string">'b Value'</span>);
2670             set(axesHandle, <span class="string">'XScale'</span>, <span class="string">'log'</span>);
2671             set(axesHandle, <span class="string">'XScale'</span>, <span class="string">'log'</span>);
2672             
2673             axesHandle = axesHandles{4}(2);
2674             cla(axesHandle)
2675             plot(axesHandle, meanMonCnt, monGamParam(:, 2), <span class="string">'.'</span>);
2676             xlim(axesHandle, [min(meanMonCnt) max(meanMonCnt)])
2677             ylim(axesHandle, [min(monGamParam(:, 2)) max(monGamParam(:, 2))])
2678             xlabel(axesHandle, <span class="string">'Mean Protein Number'</span>);
2679             ylabel(axesHandle, <span class="string">'b Value'</span>);
2680             set(axesHandle, <span class="string">'XScale'</span>, <span class="string">'log'</span>);
2681             set(axesHandle, <span class="string">'XScale'</span>, <span class="string">'log'</span>);
2682             
2683             <span class="comment">%protein expression vs gc content</span>
2684             axesHandle = axesHandles{4}(3);
2685             cla(axesHandle)
2686             tmp = rna.baseCounts(rna.matureIndexs, m.nmpIndexs);
2687             gcContent = rna.matureRNAGeneComposition(g.mRNAIndexs, :) * (sum(tmp(:, [2 3]), 2) ./ sum(tmp, 2));
2688             plot(axesHandle, meanMonCnt, gcContent, <span class="string">'.'</span>);
2689             xlim(axesHandle, [min(meanMonCnt) max(meanMonCnt)])
2690             ylim(axesHandle, [min(gcContent) max(gcContent)])
2691             xlabel(<span class="string">'Mean Protein Number'</span>);
2692             ylabel(<span class="string">'GC Content'</span>);
2693             set(axesHandle, <span class="string">'XScale'</span>, <span class="string">'log'</span>);
2694             
2695             <span class="comment">%protein expression vs mRNA half life</span>
2696             axesHandle = axesHandles{4}(4);
2697             cla(axesHandle)
2698             plot(axesHandle, meanMonCnt, rna.matureRNAGeneComposition(g.mRNAIndexs, :) * rna.halfLives(rna.matureIndexs), <span class="string">'.'</span>);
2699             xlim(axesHandle, [min(meanMonCnt) max(meanMonCnt)])
2700             ylim(axesHandle, [min(rna.halfLives(rna.matureIndexs(rna.matureMRNAIndexs))) max(rna.halfLives(rna.matureIndexs(rna.matureMRNAIndexs)))])
2701             xlabel(<span class="string">'Mean Protein Number'</span>);
2702             ylabel(<span class="string">'mRNA Life Time (min)'</span>);
2703             set(axesHandle, <span class="string">'XScale'</span>, <span class="string">'log'</span>);
2704             set(axesHandle, <span class="string">'YScale'</span>, <span class="string">'log'</span>);
2705             
2706             <span class="comment">%% data</span>
2707             figData = struct;
2708             figData.rnaFreq = rnaFreq;
2709             figData.monFreq = monFreq;
2710             figData.meanRnaCnt = meanRnaCnt;
2711             figData.stdRnaCnt = stdRnaCnt;
2712             figData.meanMRnaCnt = meanMRnaCnt;
2713             figData.stdMRnaCnt = stdMRnaCnt;
2714             figData.meanMonCnt = meanMonCnt;
2715             figData.stdMonCnt = stdMonCnt;
2716             figData.rnaLognParam = rnaLognParam;
2717             figData.rnaLognParamCI = rnaLognParamCI;
2718             figData.rnaNonNormPValue = rnaNonNormPValue;
2719             figData.monGamParam = monGamParam;
2720             figData.monGamParamCI = monGamParamCI;
2721             figData.monNonNormPValue = monNonNormPValue;
2722         <span class="keyword">end</span>
2723         
2724         <a name="_sub17" href="#_subfunctions" class="code">function figData = dnaBoundProteinDisplacement(figHandle, simBatchDir, selectedSimulations)</a>
2725             import edu.stanford.covert.cell.sim.util.PlotUtil;
2726             import edu.stanford.covert.cell.sim.util.SimulationEnsemble;
2727             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
2728             
2729             <span class="comment">%% get data</span>
2730             <span class="keyword">if</span> nargin &lt; 2 || isempty(simBatchDir)
2731                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
2732             <span class="keyword">end</span>
2733             <span class="keyword">if</span> nargin &lt; 3 || isempty(selectedSimulations)
2734                 selectedSimulations = SimulationDiskUtil.getCompleteSimulations(simBatchDir);
2735             <span class="keyword">end</span>
2736             
2737             <span class="comment">%% constants</span>
2738             nSims = numel(selectedSimulations);
2739             [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep <span class="string">'1'</span>]);
2740             c = sim.state(<span class="string">'Chromosome'</span>);
2741             pc = sim.state(<span class="string">'ProteinComplex'</span>);
2742             
2743             <span class="comment">%% get data</span>
2744             stateNames = {
2745                 <span class="string">'Chromosome'</span> <span class="string">'monomerBoundSites'</span>
2746                 <span class="string">'Chromosome'</span> <span class="string">'complexBoundSites'</span>
2747                 };
2748             displacements = zeros(numel(c.reactionBoundMonomer), nSims);
2749             simEndTimes = zeros(1, nSims);
2750             <span class="keyword">for</span> i = 1:nSims
2751                 states = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, <span class="string">'extract'</span>, selectedSimulations(i));
2752                 
2753                 simEndTimes(i) = size(states.Chromosome.monomerBoundSites, 3);
2754                 [monPosStrndTimes, monIdxs] = find(states.Chromosome.monomerBoundSites);
2755                 [cpxPosStrndTimes, cpxIdxs] = find(states.Chromosome.complexBoundSites);
2756                 
2757                 rxnIdxs = find(<span class="keyword">...</span>
2758                     (ismember(c.reactionBoundComplex, [pc.dnaPolymeraseIndexs; pc.rnaPolymeraseIndexs]) | <span class="keyword">...</span>
2759                     any(c.reactionComplexCatalysisMatrix(:, [pc.dnaPolymeraseIndexs; pc.rnaPolymeraseIndexs]), 2)) &amp; <span class="keyword">...</span>
2760                     (ismember(c.reactionBoundMonomer, monIdxs) | ismember(c.reactionBoundComplex, cpxIdxs)) &amp; <span class="keyword">...</span>
2761                     (any(c.reactionMonomerCatalysisMatrix(:, unique(monIdxs)), 2) | any(c.reactionComplexCatalysisMatrix(:, unique(cpxIdxs)), 2)) <span class="keyword">...</span>
2762                     );
2763                 <span class="keyword">for</span> j = 1:numel(rxnIdxs)
2764                     <span class="keyword">if</span> c.reactionBoundMonomer(rxnIdxs(j))
2765                         tfs = monIdxs == c.reactionBoundMonomer(rxnIdxs(j));
2766                         boundPosStrdTimes = monPosStrndTimes(tfs, :);
2767                         boundFtpt = c.monomerDNAFootprints(c.reactionBoundMonomer(rxnIdxs(j)));
2768                     <span class="keyword">else</span>
2769                         tfs = cpxIdxs == c.reactionBoundComplex(rxnIdxs(j));
2770                         boundPosStrdTimes = cpxPosStrndTimes(tfs, :);
2771                         boundFtpt = c.complexDNAFootprints(c.reactionBoundComplex(rxnIdxs(j)));
2772                     <span class="keyword">end</span>
2773                     <span class="keyword">if</span> isempty(boundPosStrdTimes)
2774                         <span class="keyword">continue</span>;
2775                     <span class="keyword">end</span>
2776                     
2777                     <span class="keyword">if</span> any(c.reactionMonomerCatalysisMatrix(rxnIdxs(j), :))
2778                         tfs = monIdxs == find(c.reactionMonomerCatalysisMatrix(rxnIdxs(j), :));
2779                         bindingPosStrdTimes = monPosStrndTimes(tfs, :);
2780                         bindingFtpt = c.monomerDNAFootprints(c.reactionMonomerCatalysisMatrix(rxnIdxs(j), :) ~= 0);
2781                     <span class="keyword">else</span>
2782                         tfs = cpxIdxs == find(c.reactionComplexCatalysisMatrix(rxnIdxs(j), :));
2783                         bindingPosStrdTimes = cpxPosStrndTimes(tfs, :);
2784                         bindingFtpt = c.complexDNAFootprints(c.reactionComplexCatalysisMatrix(rxnIdxs(j), :) ~= 0);
2785                     <span class="keyword">end</span>
2786                     <span class="keyword">if</span> isempty(bindingPosStrdTimes)
2787                         <span class="keyword">continue</span>;
2788                     <span class="keyword">end</span>
2789                     
2790                     times = intersect(unique(boundPosStrdTimes(:, 3)), unique(bindingPosStrdTimes(:, 3) - 1));
2791                     <span class="keyword">for</span> k = 1:numel(times)
2792                         tmpboundPosStrdTimes = boundPosStrdTimes(boundPosStrdTimes(:, 3) == times(k), 1:2);
2793                         tmpbindingPosStrdTimes = bindingPosStrdTimes(bindingPosStrdTimes(:, 3) == times(k) + 1, 1:2);
2794                         
2795                         <span class="keyword">if</span> numel(tmpboundPosStrdTimes) &lt;= numel(tmpbindingPosStrdTimes)
2796                             <span class="keyword">for</span> l = 1:size(tmpboundPosStrdTimes, 1)
2797                                 <span class="keyword">if</span> <span class="keyword">...</span>
2798                                         any(tmpbindingPosStrdTimes(tmpbindingPosStrdTimes(:, 1) &gt;= tmpboundPosStrdTimes(l, 1), 1) - tmpboundPosStrdTimes(l, 1) &lt; boundFtpt) || <span class="keyword">...</span>
2799                                         any(tmpbindingPosStrdTimes(:, 1) + c.sequenceLen - tmpboundPosStrdTimes(l, 1) &lt; boundFtpt) || <span class="keyword">...</span>
2800                                         any(tmpboundPosStrdTimes(l, 1) - tmpbindingPosStrdTimes(tmpbindingPosStrdTimes(:, 1) &lt; tmpboundPosStrdTimes(l, 1), 1) &lt; bindingFtpt) || <span class="keyword">...</span>
2801                                         any(tmpboundPosStrdTimes(l, 1) + c.sequenceLen - tmpbindingPosStrdTimes(:, 1) &lt; boundFtpt)
2802                                     displacements(rxnIdxs(j), i) = <span class="keyword">...</span>
2803                                         displacements(rxnIdxs(j), i) + 1;
2804                                 <span class="keyword">end</span>
2805                             <span class="keyword">end</span>
2806                         <span class="keyword">else</span>
2807                             <span class="keyword">for</span> l = 1:size(tmpbindingPosStrdTimes, 1)
2808                                 <span class="keyword">if</span> <span class="keyword">...</span>
2809                                         any(tmpboundPosStrdTimes(tmpboundPosStrdTimes(:, 1) &gt;= tmpbindingPosStrdTimes(l, 1), 1) - tmpbindingPosStrdTimes(l, 1) &lt; bindingFtpt) || <span class="keyword">...</span>
2810                                         any(tmpboundPosStrdTimes(:, 1) + c.sequenceLen - tmpbindingPosStrdTimes(l, 1) &lt; bindingFtpt) || <span class="keyword">...</span>
2811                                         any(tmpbindingPosStrdTimes(l, 1) - tmpboundPosStrdTimes(tmpboundPosStrdTimes(:, 1) &lt; tmpbindingPosStrdTimes(l, 1), 1) &lt; boundFtpt) || <span class="keyword">...</span>
2812                                         any(tmpbindingPosStrdTimes(l, 1) + c.sequenceLen - tmpboundPosStrdTimes(:, 1) &lt; boundFtpt)
2813                                     displacements(rxnIdxs(j), i) = <span class="keyword">...</span>
2814                                         displacements(rxnIdxs(j), i) + 1;
2815                                 <span class="keyword">end</span>
2816                             <span class="keyword">end</span>
2817                         <span class="keyword">end</span>
2818                     <span class="keyword">end</span>
2819                     
2820                     clear tfs boundPosStrdTimes boundFtpt bindingPosStrdTimes bindingFtpt tmpboundPosStrdTimes tmpbindingPosStrdTimes;
2821                 <span class="keyword">end</span>
2822                 
2823                 clear states monPosStrndTimes cpxPosStrndTimes monIdxs cpxIdxs;
2824             <span class="keyword">end</span>
2825             
2826             <span class="comment">%% plot data</span>
2827             clf(figHandle);
2828             
2829             meanDisplacements = mean(displacements ./ simEndTimes(ones(size(displacements, 1), 1), :), 2) * 3600;
2830             stdDisplacements = std(displacements ./ simEndTimes(ones(size(displacements, 1), 1), :), [], 2) * 3600;
2831             axesHandles = zeros(2, 2);
2832             
2833             axesHandle = subplot(2, 2, 1);
2834             axesHandles(1, 1) = axesHandle;
2835             hold(axesHandle, <span class="string">'on'</span>);
2836             means = meanDisplacements(any(c.reactionComplexCatalysisMatrix(:, pc.dnaPolymeraseIndexs), 2));
2837             stds = stdDisplacements(any(c.reactionComplexCatalysisMatrix(:, pc.dnaPolymeraseIndexs), 2));
2838             bar(axesHandle, (1:numel(means))', means);
2839             errorbar((1:numel(means))', means, stds, <span class="string">'Parent'</span>, axesHandle, <span class="string">'LineStyle'</span>, <span class="string">'none'</span>, <span class="string">'Marker'</span>, <span class="string">'none'</span>);
2840             title(axesHandle, <span class="string">'DNA Pol-displaced Proteins'</span>, <span class="string">'FontSize'</span>, 8);
2841             xlabel(axesHandle, <span class="string">'Displaced Protein'</span>, <span class="string">'FontSize'</span>, 8);
2842             ylabel(axesHandle, <span class="string">'Collisions h^{-1}'</span>, <span class="string">'FontSize'</span>, 8);
2843             set(axesHandle, <span class="string">'XTick'</span>, []);
2844             ylim(axesHandle, [0 max(means + stds)])
2845             
2846             axesHandle = subplot(2, 2, 2);
2847             axesHandles(1, 2) = axesHandle;
2848             hold(axesHandle, <span class="string">'on'</span>);
2849             means = meanDisplacements(any(c.reactionComplexCatalysisMatrix(:, pc.rnaPolymeraseIndexs), 2));
2850             stds = stdDisplacements(any(c.reactionComplexCatalysisMatrix(:, pc.rnaPolymeraseIndexs), 2));
2851             bar(axesHandle, (1:numel(means))', means);
2852             errorbar((1:numel(means))', means, stds, <span class="string">'Parent'</span>, axesHandle, <span class="string">'LineStyle'</span>, <span class="string">'none'</span>, <span class="string">'Marker'</span>, <span class="string">'none'</span>);
2853             title(axesHandle, <span class="string">'RNA Pol-displaced Proteins'</span>, <span class="string">'FontSize'</span>, 8);
2854             xlabel(axesHandle, <span class="string">'Displaced Protein'</span>, <span class="string">'FontSize'</span>, 8);
2855             ylabel(axesHandle, <span class="string">'Collisions h^{-1}'</span>, <span class="string">'FontSize'</span>, 8);
2856             set(axesHandle, <span class="string">'XTick'</span>, []);
2857             ylim(axesHandle, [0 max(means + stds)])
2858             
2859             axesHandle = subplot(2, 2, 3);
2860             axesHandles(2, 1) = axesHandle;
2861             hold(axesHandle, <span class="string">'on'</span>);
2862             means = meanDisplacements(ismember(c.reactionBoundComplex, pc.rnaPolymeraseIndexs));
2863             stds = stdDisplacements(ismember(c.reactionBoundComplex, pc.rnaPolymeraseIndexs));
2864             bar(axesHandle, (1:numel(means))', means);
2865             errorbar((1:numel(means))', means, stds, <span class="string">'Parent'</span>, axesHandle, <span class="string">'LineStyle'</span>, <span class="string">'none'</span>, <span class="string">'Marker'</span>, <span class="string">'none'</span>);
2866             title(axesHandle, <span class="string">'RNA Pol Displacements'</span>, <span class="string">'FontSize'</span>, 8);
2867             xlabel(axesHandle, <span class="string">'Binding Protein'</span>, <span class="string">'FontSize'</span>, 8);
2868             ylabel(axesHandle, <span class="string">'Collisions h^{-1}'</span>, <span class="string">'FontSize'</span>, 8);
2869             set(axesHandle, <span class="string">'XTick'</span>, []);
2870             ylim(axesHandle, [0 max(means + stds)])
2871             
2872             axesHandle = subplot(2, 2, 4);
2873             axesHandles(2, 2) = axesHandle;
2874             hold(axesHandle, <span class="string">'on'</span>);
2875             tmp = sum(displacements(ismember(c.reactionBoundComplex, pc.rnaPolymeraseIndexs) &amp; any(c.reactionComplexCatalysisMatrix(:, pc.dnaPolymeraseIndexs), 2), :), 1) ./ simEndTimes * 3600;
2876             bar(axesHandle, tmp);
2877             title(axesHandle, <span class="string">'RNA Pol Displacements by DNA Pol'</span>, <span class="string">'FontSize'</span>, 8);
2878             xlabel(axesHandle, <span class="string">'Collisions h^{-1}'</span>, <span class="string">'FontSize'</span>, 8);
2879             ylabel(axesHandle, <span class="string">'Frequency'</span>, <span class="string">'FontSize'</span>, 8);
2880             
2881             <span class="comment">%y-axes</span>
2882             set(axesHandles, <span class="string">'FontSize'</span>, 6);
2883             set(cell2mat(get(axesHandles, <span class="string">'Ylabel'</span>)), <span class="string">'Units'</span>, <span class="string">'normalized'</span>)
2884             PlotUtil.alignYAxesLabels(axesHandles(:, 1));
2885             PlotUtil.alignYAxesLabels(axesHandles(:, 2));
2886             
2887             <span class="comment">%% organize figure data</span>
2888             figData = struct;
2889             figData.simEndTimes = simEndTimes;
2890             figData.displacements = displacements;
2891         <span class="keyword">end</span>
2892         
2893         <a name="_sub18" href="#_subfunctions" class="code">function figData = macromoleculeHalfLives(figHandle, simBatchDir, selectedSimulations)</a>
2894             import edu.stanford.covert.cell.sim.util.PlotUtil;
2895             import edu.stanford.covert.cell.sim.util.SimulationEnsemble;
2896             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
2897             
2898             <span class="keyword">if</span> nargin &lt; 2 || isempty(simBatchDir)
2899                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
2900             <span class="keyword">end</span>
2901             <span class="keyword">if</span> nargin &lt; 3 || isempty(selectedSimulations)
2902                 selectedSimulations = SimulationDiskUtil.getCompleteSimulations(simBatchDir);
2903             <span class="keyword">end</span>
2904             
2905             <span class="comment">%% constants</span>
2906             nSims = numel(selectedSimulations);
2907             
2908             [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep <span class="string">'1'</span>]);
2909             g = sim.gene;
2910             rna = sim.state(<span class="string">'Rna'</span>);
2911             pm = sim.state(<span class="string">'ProteinMonomer'</span>);
2912             pc = sim.state(<span class="string">'ProteinComplex'</span>);
2913             
2914             pcComp = sum(pc.proteinComplexComposition, 3);
2915             
2916             expRNAHalfLives = rna.halfLives(rna.matureIndexs) / 60;
2917             expMonHalfLives = pm.halfLives(pm.matureIndexs) / 3600;
2918             expCpxHalfLives = pc.halfLives(pc.matureIndexs) / 3600;
2919             
2920             expRNAHalfLives(isinf(expRNAHalfLives)) = NaN;
2921             expMonHalfLives(isinf(expMonHalfLives)) = NaN;
2922             expCpxHalfLives(isinf(expCpxHalfLives)) = NaN;
2923             
2924             <span class="comment">%% get data</span>
2925             rnaCounts = zeros(numel(rna.matureIndexs), nSims);
2926             monCounts = zeros(numel(pm.matureIndexs),  nSims);
2927             cpxCounts = zeros(numel(pc.matureIndexs),  nSims);
2928             rnaDecays = zeros(numel(rna.matureIndexs), nSims);
2929             monDecays = zeros(numel(pm.matureIndexs),  nSims);
2930             cpxDecays = zeros(numel(pc.matureIndexs),  nSims);
2931             
2932             stateNames = {
2933                 <span class="string">'Rna'</span>             <span class="string">'counts'</span>  <span class="string">':'</span>  <span class="string">'-sum'</span>
2934                 <span class="string">'ProteinMonomer'</span>  <span class="string">'counts'</span>  <span class="string">':'</span>  <span class="string">'-sum'</span>
2935                 <span class="string">'ProteinComplex'</span>  <span class="string">'counts'</span>  <span class="string">':'</span>  <span class="string">'-sum'</span>
2936                 };
2937             <span class="keyword">for</span> i = 1:nSims
2938                 states = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, <span class="string">'extract'</span>, selectedSimulations(i));
2939                 
2940                 <span class="keyword">if</span> isa(states.Rna.counts, <span class="string">'edu.stanford.covert.util.SparseMat'</span>)
2941                     [subs, vals] = find(states.Rna.counts);
2942                     tfs = <span class="keyword">...</span>
2943                         (subs(:, 1) &gt;= rna.nascentIndexs(1) &amp; subs(:, 1) &lt;= rna.nascentIndexs(end)) | <span class="keyword">...</span>
2944                         (subs(:, 1) &gt;= rna.intergenicIndexs(1) &amp; subs(:, 1) &lt;= rna.intergenicIndexs(end));
2945                     subs = subs(~tfs, :);
2946                     vals = vals(~tfs, :);
2947                     
2948                     subs(subs(:, 1) &gt; rna.intergenicIndexs(end)) = subs(subs(:, 1) &gt; rna.intergenicIndexs(end)) - numel(rna.intergenicIndexs);
2949                     subs(subs(:, 1) &gt; rna.nascentIndexs(end)) = subs(subs(:, 1) &gt; rna.nascentIndexs(end)) - numel(rna.nascentIndexs);
2950                     
2951                     tmp = edu.stanford.covert.util.SparseMat(subs, vals, [6 * numel(rna.matureIndexs) 1 size(states.Rna.counts, 3)]);
2952                     states.Rna.counts = full(permute(sum(reshape(tmp, [numel(rna.matureIndexs) 6 size(states.Rna.counts, 3)]), 2), [1 3 2 4]));
2953                     
2954                     clear tmp tfs subs vals;
2955                 <span class="keyword">else</span>
2956                     states.Rna.counts = permute(<span class="keyword">...</span>
2957                         + states.Rna.counts(rna.processedIndexs, :, :) <span class="keyword">...</span>
2958                         + states.Rna.counts(rna.matureIndexs, :, :) <span class="keyword">...</span>
2959                         + states.Rna.counts(rna.aminoacylatedIndexs, :, :) <span class="keyword">...</span>
2960                         + states.Rna.counts(rna.boundIndexs, :, :) <span class="keyword">...</span>
2961                         + states.Rna.counts(rna.misfoldedIndexs, :, :) <span class="keyword">...</span>
2962                         + states.Rna.counts(rna.damagedIndexs, :, :)<span class="keyword">...</span>
2963                         , [1 3 2]);
2964                 <span class="keyword">end</span>
2965                 
2966                 states.ProteinMonomer.counts = full(permute(sum(reshape(states.ProteinMonomer.counts, <span class="keyword">...</span>
2967                     [numel(pm.matureIndexs)  numel(pm.wholeCellModelIDs)/numel(pm.matureIndexs)  size(states.ProteinMonomer.counts, 3)]), 2), [1 3 2]));
2968                 states.ProteinComplex.counts = full(permute(sum(reshape(states.ProteinComplex.counts, <span class="keyword">...</span>
2969                     [numel(pc.matureIndexs)  numel(pc.wholeCellModelIDs)/numel(pc.matureIndexs)  size(states.ProteinComplex.counts, 3)]), 2), [1 3 2]));
2970                 
2971                 tmp = states.Rna.counts;
2972                 tmp(setdiff(1:<span class="keyword">end</span>, rna.matureMRNAIndexs), :) = <span class="keyword">...</span>
2973                     + tmp(setdiff(1:<span class="keyword">end</span>, rna.matureMRNAIndexs), :) <span class="keyword">...</span>
2974                     + pcComp(setdiff(1:<span class="keyword">end</span>, g.mRNAIndexs), :) * states.ProteinComplex.counts;
2975                 rnaCounts(:, i) = sum(tmp(:, 1:end-1), 2);
2976                 rnaDecays(:, i) = -sum(min(0, diff(tmp, 1, 2)), 2);
2977                 
2978                 tmp = <span class="keyword">...</span>
2979                     + states.ProteinMonomer.counts <span class="keyword">...</span>
2980                     + pcComp(g.mRNAIndexs, :) * states.ProteinComplex.counts;
2981                 monCounts(:, i) = sum(tmp(:, 1:end-1), 2);
2982                 monDecays(:, i) = -sum(min(0, diff(tmp, 1, 2)), 2);
2983                 
2984                 tmp = states.ProteinComplex.counts;
2985                 cpxCounts(:, i) = sum(tmp(:, 1:end-1), 2);
2986                 cpxDecays(:, i) = -sum(min(0, diff(tmp, 1, 2)), 2);
2987                 
2988                 clear states simEndTimes tmp;
2989             <span class="keyword">end</span>
2990             
2991             avgRnaHalfLives = 1 / 60   * log(2) ./ (sum(rnaDecays, 2) ./ sum(rnaCounts, 2));
2992             avgMonHalfLives = 1 / 3600 * log(2) ./ (sum(monDecays, 2) ./ sum(monCounts, 2));
2993             avgCpxHalfLives = 1 / 3600 * log(2) ./ (sum(cpxDecays, 2) ./ sum(cpxCounts, 2));
2994             stdRnaHalfLives = std(1 / 60   * log(2) ./ (rnaDecays ./ rnaCounts), [], 2);
2995             stdMonHalfLives = std(1 / 3600 * log(2) ./ (monDecays ./ monCounts), [], 2);
2996             stdCpxHalfLives = std(1 / 3600 * log(2) ./ (cpxDecays ./ cpxCounts), [], 2);
2997             
2998             avgRnaHalfLives(isinf(avgRnaHalfLives) | isnan(expRNAHalfLives)) = NaN;
2999             avgMonHalfLives(isinf(avgMonHalfLives) | isnan(expMonHalfLives)) = NaN;
3000             avgCpxHalfLives(isinf(avgCpxHalfLives) | isnan(expCpxHalfLives)) = NaN;
3001             stdRnaHalfLives(isinf(stdRnaHalfLives) | isnan(expRNAHalfLives)) = NaN;
3002             stdMonHalfLives(isinf(stdMonHalfLives) | isnan(expMonHalfLives)) = NaN;
3003             stdCpxHalfLives(isinf(stdCpxHalfLives) | isnan(expCpxHalfLives)) = NaN;
3004             
3005             avgMonHalfLive = nanmean(avgMonHalfLives);
3006             avgCpxHalfLive = nanmean(avgCpxHalfLives);
3007             
3008             clear rnaCounts monCounts cpxCounts rnaDecays monDecays cpxDecays;
3009             
3010             <span class="comment">%% plot data</span>
3011             clf(figHandle);
3012             [axesHandles, xAxesHandles] = PlotUtil.multiElementPlot(figHandle, {100 100 100}, [0 1], struct(<span class="keyword">...</span>
3013                 <span class="string">'position'</span>, [0.02 0.62 0.95 0.2417]));
3014             
3015             <span class="comment">%RNA</span>
3016             axesHandle = axesHandles{1};
3017             xAxesHandle = xAxesHandles{1};
3018             hold(axesHandle, <span class="string">'on'</span>);
3019             h = zeros(4, 1);
3020             h(1) = errorbar(expRNAHalfLives(rna.matureMRNAIndexs), avgRnaHalfLives(rna.matureMRNAIndexs), stdRnaHalfLives(rna.matureMRNAIndexs), <span class="keyword">...</span>
3021                 <span class="string">'Parent'</span>, axesHandle, <span class="string">'LineStyle'</span>, <span class="string">'none'</span>, <span class="string">'Color'</span>, <span class="string">'r'</span>, <span class="string">'Marker'</span>, <span class="string">'.'</span>);
3022             h(2) = errorbar(expRNAHalfLives(rna.matureRRNAIndexs), avgRnaHalfLives(rna.matureRRNAIndexs), stdRnaHalfLives(rna.matureRRNAIndexs), <span class="keyword">...</span>
3023                 <span class="string">'Parent'</span>, axesHandle, <span class="string">'LineStyle'</span>, <span class="string">'none'</span>, <span class="string">'Color'</span>, <span class="string">'g'</span>, <span class="string">'Marker'</span>, <span class="string">'.'</span>);
3024             h(3) = errorbar(expRNAHalfLives(rna.matureSRNAIndexs), avgRnaHalfLives(rna.matureSRNAIndexs), stdRnaHalfLives(rna.matureSRNAIndexs), <span class="keyword">...</span>
3025                 <span class="string">'Parent'</span>, axesHandle, <span class="string">'LineStyle'</span>, <span class="string">'none'</span>, <span class="string">'Color'</span>, <span class="string">'b'</span>, <span class="string">'Marker'</span>, <span class="string">'.'</span>);
3026             h(4) = errorbar(expRNAHalfLives(rna.matureTRNAIndexs), avgRnaHalfLives(rna.matureTRNAIndexs), stdRnaHalfLives(rna.matureTRNAIndexs), <span class="keyword">...</span>
3027                 <span class="string">'Parent'</span>, axesHandle, <span class="string">'LineStyle'</span>, <span class="string">'none'</span>, <span class="string">'Color'</span>, <span class="string">'c'</span>, <span class="string">'Marker'</span>, <span class="string">'.'</span>);
3028             minVal = min(min(avgRnaHalfLives - stdRnaHalfLives, expRNAHalfLives - 1));
3029             maxVal = max(max(avgRnaHalfLives - stdRnaHalfLives, expRNAHalfLives + 1));
3030             xlim(xAxesHandle, [minVal maxVal]);
3031             xlim(axesHandle, [minVal maxVal]);
3032             ylim(axesHandle, [minVal maxVal]);
3033             title(axesHandle, <span class="string">'RNAs'</span>, <span class="string">'FontSize'</span>, 12);
3034             xlabel(xAxesHandle, <span class="string">'ExpHalf Life (min)'</span>, <span class="string">'FontSize'</span>, 8);
3035             ylabel(axesHandle, <span class="string">'Sim Half Life (min)'</span>, <span class="string">'FontSize'</span>, 8);
3036             set(axesHandle, <span class="string">'FontSize'</span>, 6);
3037             set(axesHandle, <span class="string">'YTick'</span>, get(xAxesHandle, <span class="string">'XTick'</span>));
3038             legend(h, {<span class="string">'m'</span>, <span class="string">'r'</span>, <span class="string">'s'</span>, <span class="string">'t'</span>}, <span class="string">'Location'</span>, <span class="string">'SouthEast'</span>);
3039             
3040             <span class="comment">%Protein Monomers</span>
3041             axesHandle = axesHandles{2};
3042             xAxesHandle = xAxesHandles{2};
3043             [~, order] = sort(avgMonHalfLives);
3044             avgMonHalfLives = avgMonHalfLives(order);
3045             stdMonHalfLives = stdMonHalfLives(order);
3046             expMonHalfLives = expMonHalfLives(order);
3047             hold(axesHandle, <span class="string">'on'</span>);
3048             line([0.5 numel(avgMonHalfLives)+0.5], expMonHalfLives([1 1]), <span class="string">'Parent'</span>, axesHandle, <span class="string">'Color'</span>, <span class="string">'r'</span>);
3049             line([0.5 numel(avgMonHalfLives)+0.5], avgMonHalfLive([1 1]), <span class="string">'Parent'</span>, axesHandle, <span class="string">'Color'</span>, <span class="string">'g'</span>);
3050             idxs = find(~isnan(stdMonHalfLives));
3051             errorbar(idxs, avgMonHalfLives(idxs), stdMonHalfLives(idxs), <span class="keyword">...</span>
3052                 <span class="string">'Parent'</span>, axesHandle, <span class="keyword">...</span>
3053                 <span class="string">'LineStyle'</span>, <span class="string">'none'</span>, <span class="keyword">...</span>
3054                 <span class="string">'Marker'</span>, <span class="string">'.'</span>, <span class="keyword">...</span>
3055                 <span class="string">'Color'</span>, <span class="string">'b'</span>);
3056             idxs = find(isnan(stdMonHalfLives) &amp; ~isnan(avgMonHalfLives));
3057             plot(axesHandle, idxs, avgMonHalfLives(idxs), <span class="string">'b.'</span>);
3058             minVal = max(0, min(min(avgMonHalfLives - stdMonHalfLives, expMonHalfLives - 1)));
3059             maxVal = max(max(avgMonHalfLives - stdMonHalfLives, expMonHalfLives + 1));
3060             xlim(xAxesHandle, [0.5 numel(avgMonHalfLives)+0.5]);
3061             xlim(axesHandle, [0.5 numel(avgMonHalfLives)+0.5]);
3062             ylim(axesHandle, [minVal maxVal]);
3063             title(axesHandle, <span class="string">'Monomers'</span>, <span class="string">'FontSize'</span>, 12);
3064             xlabel(xAxesHandle, <span class="string">'Monomer'</span>, <span class="string">'FontSize'</span>, 8);
3065             ylabel(axesHandle, <span class="string">'Sim Half Life (h)'</span>, <span class="string">'FontSize'</span>, 8);
3066             set(axesHandle, <span class="string">'FontSize'</span>, 6);
3067             
3068             <span class="comment">%Protein Complexes</span>
3069             axesHandle = axesHandles{3};
3070             xAxesHandle = xAxesHandles{3};
3071             [~, order] = sort(avgCpxHalfLives);
3072             avgCpxHalfLives = avgCpxHalfLives(order);
3073             stdCpxHalfLives = stdCpxHalfLives(order);
3074             expCpxHalfLives = expCpxHalfLives(order);
3075             hold(axesHandle, <span class="string">'on'</span>);
3076             line([0.5 numel(avgCpxHalfLives)+0.5], expCpxHalfLives([1 1]), <span class="string">'Parent'</span>, axesHandle, <span class="string">'Color'</span>, <span class="string">'r'</span>);
3077             line([0.5 numel(avgCpxHalfLives)+0.5], avgCpxHalfLive([1 1]), <span class="string">'Parent'</span>, axesHandle, <span class="string">'Color'</span>, <span class="string">'g'</span>);
3078             idxs = find(~isnan(stdCpxHalfLives));
3079             errorbar(idxs, avgCpxHalfLives(idxs), stdCpxHalfLives(idxs), <span class="keyword">...</span>
3080                 <span class="string">'Parent'</span>, axesHandle, <span class="keyword">...</span>
3081                 <span class="string">'LineStyle'</span>, <span class="string">'none'</span>, <span class="keyword">...</span>
3082                 <span class="string">'Marker'</span>, <span class="string">'.'</span>, <span class="keyword">...</span>
3083                 <span class="string">'Color'</span>, <span class="string">'b'</span>);
3084             idxs = find(isnan(stdCpxHalfLives) &amp; ~isnan(avgCpxHalfLives));
3085             plot(axesHandle, idxs, avgCpxHalfLives(idxs), <span class="string">'b.'</span>);
3086             minVal = max(0, min(min(avgCpxHalfLives - stdCpxHalfLives, expCpxHalfLives - 1)));
3087             maxVal = max(max(avgCpxHalfLives - stdCpxHalfLives, expCpxHalfLives + 1));
3088             xlim(xAxesHandle, [0.5 numel(avgCpxHalfLives)+0.5]);
3089             xlim(axesHandle, [0.5 numel(avgCpxHalfLives)+0.5]);
3090             ylim(axesHandle, [minVal maxVal]);
3091             title(axesHandle, <span class="string">'Complexes'</span>, <span class="string">'FontSize'</span>, 12);
3092             xlabel(xAxesHandle, <span class="string">'Complex'</span>, <span class="string">'FontSize'</span>, 8);
3093             ylabel(axesHandle, <span class="string">'Sim Half Life (h)'</span>, <span class="string">'FontSize'</span>, 8);
3094             set(axesHandle, <span class="string">'FontSize'</span>, 6);
3095             
3096             <span class="comment">%align</span>
3097             PlotUtil.offsetYAxes(axesHandles, 0.04);
3098             PlotUtil.labelSubplots(cell2mat(axesHandles));
3099             
3100             <span class="comment">%% organize figure data</span>
3101             figData = struct;
3102             figData.expRNAHalfLives = expRNAHalfLives;
3103             figData.avgRnaHalfLives = avgRnaHalfLives;
3104             figData.stdRnaHalfLives = stdRnaHalfLives;
3105             figData.avgMonHalfLives = avgMonHalfLives;
3106             figData.stdMonHalfLives = stdMonHalfLives;
3107             figData.expMonHalfLives = expMonHalfLives;
3108             figData.avgCpxHalfLives = avgCpxHalfLives;
3109             figData.stdCpxHalfLives = stdCpxHalfLives;
3110             figData.expCpxHalfLives = expCpxHalfLives;
3111         <span class="keyword">end</span>
3112         
3113         <a name="_sub19" href="#_subfunctions" class="code">function figData = rnaSynthesisDuration(figHandle, simBatchDir, selectedSimulations)</a>
3114             import edu.stanford.covert.cell.sim.util.PlotUtil;
3115             import edu.stanford.covert.cell.sim.util.SimulationEnsemble;
3116             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
3117             
3118             <span class="keyword">if</span> nargin &lt; 2 || isempty(simBatchDir)
3119                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
3120             <span class="keyword">end</span>
3121             <span class="keyword">if</span> nargin &lt; 3 || isempty(selectedSimulations)
3122                 selectedSimulations = SimulationDiskUtil.getCompleteSimulations(simBatchDir);
3123             <span class="keyword">end</span>
3124             
3125             <span class="comment">%% constants</span>
3126             [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep <span class="string">'1'</span>]);
3127             g = sim.gene;
3128             comp = sim.compartment;
3129             c = sim.state(<span class="string">'Chromosome'</span>);
3130             r = sim.state(<span class="string">'Rna'</span>);
3131             pc = sim.state(<span class="string">'ProteinComplex'</span>);
3132             rnaPol = sim.state(<span class="string">'RNAPolymerase'</span>);
3133             transcript = sim.state(<span class="string">'Transcript'</span>);
3134             
3135             nSims = numel(selectedSimulations);
3136             pcComp = sum(pc.proteinComplexComposition(setdiff(1:<span class="keyword">end</span>, g.mRNAIndexs), :, :), 3);
3137             
3138             <span class="comment">%% get data</span>
3139             stateNames = {
3140                 <span class="string">'Rna'</span>             <span class="string">'counts'</span>                   <span class="string">':'</span>  comp.cytosolIndexs
3141                 <span class="string">'Transcript'</span>      <span class="string">'boundTranscriptionUnits'</span>  <span class="string">':'</span>  <span class="string">':'</span>
3142                 <span class="string">'Transcript'</span>      <span class="string">'boundTranscriptProgress'</span>  <span class="string">':'</span>  <span class="string">':'</span>
3143                 <span class="string">'RNAPolymerase'</span>   <span class="string">'positionStrands'</span>          <span class="string">':'</span>  <span class="string">':'</span>
3144                 <span class="string">'RNAPolymerase'</span>   <span class="string">'states'</span>                   <span class="string">':'</span>  <span class="string">':'</span>
3145                 <span class="string">'Ribosome'</span>        <span class="string">'boundMRNAs'</span>               <span class="string">':'</span>  <span class="string">':'</span>
3146                 <span class="string">'ProteinComplex'</span>  <span class="string">'counts'</span>                   <span class="string">':'</span>  <span class="string">'-sum'</span>
3147                 };
3148             rnaSynthesisTimes = zeros(numel(r.matureIndexs), 3, nSims);
3149             mrnaFracBounds = zeros(numel(g.mRNAIndexs), 1, nSims);
3150             rnaFracAminoacylated = zeros(numel(r.matureIndexs), 1, nSims);
3151             rnaPauses = zeros(4, 1000);
3152             <span class="keyword">for</span> i = 1:nSims
3153                 states = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, <span class="string">'extract'</span>, selectedSimulations(i));
3154                 
3155                 states.RNAPolymerase.states = permute(states.RNAPolymerase.states, [1 3 2]);
3156                 states.RNAPolymerase.positionStrands = permute(states.RNAPolymerase.positionStrands, [1 3 2]);
3157                 states.Transcript.boundTranscriptionUnits = permute(states.Transcript.boundTranscriptionUnits, [1 3 2]);
3158                 states.Transcript.boundTranscriptProgress = permute(states.Transcript.boundTranscriptProgress, [1 3 2]);
3159                 states.Ribosome.boundMRNAs = permute(states.Ribosome.boundMRNAs, [1 3 2]);
3160                 
3161                 states.Rna.counts = permute(states.Rna.counts, [1 3 2]);
3162                 
3163                 nascs = full(states.Rna.counts(r.nascentIndexs, :));
3164                 procs = full(states.Rna.counts(r.processedIndexs, :));
3165                 matrs = full(<span class="keyword">...</span>
3166                     + states.Rna.counts(r.matureIndexs, :) <span class="keyword">...</span>
3167                     + states.Rna.counts(r.boundIndexs, :) <span class="keyword">...</span>
3168                     + states.Rna.counts(r.misfoldedIndexs, :) <span class="keyword">...</span>
3169                     + states.Rna.counts(r.damagedIndexs, :) <span class="keyword">...</span>
3170                     + states.Rna.counts(r.aminoacylatedIndexs, :));
3171                 matrs(setdiff(1:<span class="keyword">end</span>, r.matureMRNAIndexs), :) = <span class="keyword">...</span>
3172                     + matrs(setdiff(1:<span class="keyword">end</span>, r.matureMRNAIndexs), :) <span class="keyword">...</span>
3173                     + pcComp * full(permute(sum(reshape(states.ProteinComplex.counts, <span class="keyword">...</span>
3174                     [numel(pc.matureIndexs)  numel(pc.wholeCellModelIDs)/numel(pc.matureIndexs)  size(states.ProteinComplex.counts, 3)]), 2), [1 3 2]));
3175                 amincs = full(sum(states.Rna.counts(r.aminoacylatedIndexs, :), 2));
3176                 
3177                 <span class="comment">%transcription times</span>
3178                 [startTimes, startPols] = find(cat(2, <span class="keyword">...</span>
3179                     states.RNAPolymerase.states(:, 1) &gt;= rnaPol.activelyTranscribingValue, <span class="keyword">...</span>
3180                     states.RNAPolymerase.states(:, 2:end)   &gt;= rnaPol.activelyTranscribingValue &amp; <span class="keyword">...</span>
3181                     states.RNAPolymerase.states(:, 1:end-1) &lt; rnaPol.activelyTranscribingValue)');
3182                 [endTimes, endPols] = find(cat(2, <span class="keyword">...</span>
3183                     states.RNAPolymerase.states(:, 1:end-1) &gt;= rnaPol.activelyTranscribingValue &amp; <span class="keyword">...</span>
3184                     states.RNAPolymerase.states(:, 2:end)   &lt; rnaPol.activelyTranscribingValue, <span class="keyword">...</span>
3185                     states.RNAPolymerase.states(:, end) &gt;= rnaPol.activelyTranscribingValue)');
3186                 assert(all(startPols == endPols));
3187                 assert(all(endTimes &gt;= startTimes));
3188                 boundTUs = states.Transcript.boundTranscriptionUnits(sub2ind(<span class="keyword">...</span>
3189                     size(states.Transcript.boundTranscriptionUnits), <span class="keyword">...</span>
3190                     startPols, startTimes));
3191                 rnaSynthesisTimes(1:numel(r.nascentIndexs), 1, i) = <span class="keyword">...</span>
3192                     accumarray(boundTUs, endTimes - startTimes + 1, [numel(r.nascentIndexs) 1]) ./ histc(boundTUs, (1:numel(r.nascentIndexs))');
3193                 
3194                 <span class="comment">%transcription pauses</span>
3195                 pause = states.RNAPolymerase.states(:, 1:end-1) &gt;= rnaPol.activelyTranscribingValue &amp; diff(states.Transcript.boundTranscriptProgress, 1, 2) == 0;
3196                 [startTimes, startPols] = find([pause(:, 1)  diff(pause, 1, 2) &gt; 0]');
3197                 [endTimes, endPols] = find([diff(pause, 1, 2) &lt; 0  pause(:, end)]');
3198                 assert(all(startPols == endPols));
3199                 assert(all(endTimes &gt;= startTimes));
3200                 
3201                 [~, bases] = ismember(c.sequence.subsequence(states.RNAPolymerase.positionStrands(sub2ind(<span class="keyword">...</span>
3202                     size(states.RNAPolymerase.positionStrands), <span class="keyword">...</span>
3203                     [startPols startPols], <span class="keyword">...</span>
3204                     [startTimes startTimes], <span class="keyword">...</span>
3205                     [ones(size(startPols, 1), 1) 2 * ones(size(startPols, 1), 1)]))), <span class="string">'ACGT'</span>);
3206                 
3207                 <span class="keyword">if</span> max(endTimes - startTimes + 1) &gt; size(rnaPauses, 2)
3208                     rnaPauses = [rnaPauses zeros(4, max(endTimes - startTimes + 1) - size(rnaPauses, 2))]; <span class="comment">%#ok&lt;AGROW&gt;</span>
3209                 <span class="keyword">end</span>
3210                 rnaPauses = <span class="keyword">...</span>
3211                     + rnaPauses<span class="keyword">...</span>
3212                     + accumarray([bases  endTimes - startTimes + 1], ones(size(startTimes)), [4 size(rnaPauses, 2)]);
3213                 
3214                 <span class="comment">%processing time</span>
3215                 rnaSynthesisTimes(1:numel(r.nascentIndexs), 2, i) = log(2) ./ (sum(max(0, -diff(nascs, 1, 2)), 2) ./ sum(nascs(:, 1:end-1), 2));
3216                 
3217                 <span class="comment">%modification time</span>
3218                 rnaSynthesisTimes(:, 3, i) = log(2) ./ (sum(max(0, -diff(procs, 1, 2)), 2) ./ sum(procs(:, 1:end-1), 2));
3219                 
3220                 <span class="comment">%fraction time bound</span>
3221                 [~, j, val] = find(states.Ribosome.boundMRNAs);
3222                 tmp = zeros(numel(g.mRNAIndexs), size(states.Ribosome.boundMRNAs, 2));
3223                 tmp(sub2ind(size(tmp), val, j)) = 1;
3224                 mrnaFracBounds(:, 1, i) = sum(tmp, 2) ./ sum(r.matureRNAGeneComposition(g.mRNAIndexs, :) * matrs &gt; 0, 2);
3225                 
3226                 <span class="comment">%fraction aminoacylated bound</span>
3227                 rnaFracAminoacylated(:, 1, i) = amincs ./ sum(matrs, 2);
3228                 
3229                 clear states nascs procs matrs amincs startTimes startPols endTimes endPols boundTUs pause j val tmp;
3230             <span class="keyword">end</span>
3231             
3232             rnaSynthesisTimes(isinf(rnaSynthesisTimes)) = NaN;
3233             mrnaFracBounds(isinf(mrnaFracBounds)) = NaN;
3234             rnaFracAminoacylated(isinf(rnaFracAminoacylated)) = NaN;
3235             
3236             <span class="comment">%% plot</span>
3237             clf(figHandle);
3238             
3239             <span class="comment">%synthesis times</span>
3240             axesHandle = subplot(3, 2, 1, <span class="string">'Parent'</span>, figHandle);
3241             axesHandles(1, 1) = axesHandle;
3242             tmp = [zeros(numel(r.matureIndexs), 2)  nanmean(rnaSynthesisTimes(:, 3, :), 3)];
3243             <span class="keyword">for</span> i = 1:numel(r.nascentIndexs)
3244                 tmpTfs = r.nascentRNAMatureRNAComposition(:, i) == 1;
3245                 tmp(tmpTfs, 1) = nanmean(rnaSynthesisTimes(i, 1, :), 3);
3246                 tmp(tmpTfs, 2) = nanmean(rnaSynthesisTimes(i, 2, :), 3);
3247             <span class="keyword">end</span>
3248             [~, order] = sort(nansum(tmp, 2));
3249             h = bar(axesHandle, 1:numel(r.matureIndexs),  tmp(order, :), <span class="string">'stacked'</span>);
3250             xlim(axesHandle, [0.5 numel(r.matureIndexs) + 0.5]);
3251             ylim(axesHandle, [0 max(nansum(tmp, 2))]);
3252             legend(h, {<span class="string">'Transcription'</span>, <span class="string">'Processing'</span>, <span class="string">'Modification'</span>}, <span class="string">'Location'</span>, <span class="string">'NorthWest'</span>);
3253             xlabel(axesHandle, <span class="string">'RNA'</span>, <span class="string">'FontSize'</span>, 8);
3254             ylabel(axesHandle, {<span class="string">'Synthesis'</span> <span class="string">'Duration (s)'</span>}, <span class="string">'FontSize'</span>, 8);
3255             
3256             <span class="comment">%transcription time vs length</span>
3257             axesHandle = subplot(3, 2, 3, <span class="string">'Parent'</span>, figHandle);
3258             axesHandles(2, 1) = axesHandle;
3259             meanVal = nanmean(rnaSynthesisTimes(1:numel(r.nascentIndexs), 1, :), 3);
3260             stdVal = nanstd(rnaSynthesisTimes(1:numel(r.nascentIndexs), 1, :), [], 3);
3261             errorbar(r.lengths(r.nascentIndexs), meanVal, stdVal, <span class="keyword">...</span>
3262                 <span class="string">'LineStyle'</span>, <span class="string">'none'</span>, <span class="string">'Parent'</span>, axesHandle);
3263             xlim(axesHandle, [min(r.lengths(r.nascentIndexs)) max(r.lengths(r.nascentIndexs))]);
3264             ylim(axesHandle, [min(meanVal - stdVal) max(meanVal + stdVal)]);
3265             xlabel(axesHandle, <span class="string">'Length'</span>, <span class="string">'FontSize'</span>, 8);
3266             ylabel(axesHandle, {<span class="string">'Transcription'</span> <span class="string">'Duration (s)'</span>}, <span class="string">'FontSize'</span>, 8)
3267             
3268             <span class="comment">%transcription time vs G/C content</span>
3269             axesHandle = subplot(3, 2, 5, <span class="string">'Parent'</span>, figHandle);
3270             axesHandles(3, 1) = axesHandle;
3271             gcContent = cellfun(@(seq) sum(ismember(seq, <span class="string">'GC'</span>)) / numel(seq), transcript.transcriptionUnitSequences(r.nascentIndexs));
3272             meanVal = nanmean(rnaSynthesisTimes(1:numel(r.nascentIndexs), 1, :), 3);
3273             stdVal = nanstd(rnaSynthesisTimes(1:numel(r.nascentIndexs), 1, :), [], 3);
3274             errorbar(gcContent, <span class="keyword">...</span>
3275                 nanmean(rnaSynthesisTimes(1:numel(r.nascentIndexs), 1, :), 3), <span class="keyword">...</span>
3276                 nanstd(rnaSynthesisTimes(1:numel(r.nascentIndexs), 1, :), [], 3), <span class="keyword">...</span>
3277                 <span class="string">'LineStyle'</span>, <span class="string">'none'</span>, <span class="string">'Parent'</span>, axesHandle);
3278             xlim(axesHandle, [min(gcContent) max(gcContent)]);
3279             ylim(axesHandle, [min(meanVal - stdVal) max(meanVal + stdVal)]);
3280             xlabel(axesHandle, <span class="string">'G/C Content'</span>, <span class="string">'FontSize'</span>, 8);
3281             ylabel(axesHandle, {<span class="string">'Transcription'</span> <span class="string">'Duration'</span>}, <span class="string">'FontSize'</span>, 8)
3282             
3283             <span class="comment">%pauses</span>
3284             axesHandle = subplot(3, 2, 2, <span class="string">'Parent'</span>, figHandle);
3285             axesHandles(1, 2) = axesHandle;
3286             tmp = rnaPauses ./ repmat(sum(rnaPauses, 2), 1, size(rnaPauses, 2));
3287             h = bar(axesHandle, 1:size(rnaPauses, 2), tmp', <span class="string">'group'</span>);
3288             xlim(axesHandle, [0.5 find(any(tmp &gt; 0.01, 1), 1, <span class="string">'last'</span>) + 0.5]);
3289             legend(h, {<span class="string">'A'</span>, <span class="string">'C'</span>, <span class="string">'G'</span>, <span class="string">'T'</span>});
3290             ylim(axesHandle, [0 max(tmp(:))]);
3291             xlabel(axesHandle, <span class="string">'Pause Duration (s)'</span>, <span class="string">'FontSize'</span>, 8);
3292             ylabel(axesHandle, <span class="string">'Freq'</span>, <span class="string">'FontSize'</span>, 8)
3293             
3294             <span class="comment">%fraction bound</span>
3295             axesHandle = subplot(3, 2, 4, <span class="string">'Parent'</span>, figHandle);
3296             axesHandles(2, 2) = axesHandle;
3297             [~, order] = sort(nanmean(mrnaFracBounds, 3));
3298             errorbar((1:numel(g.mRNAIndexs))', <span class="keyword">...</span>
3299                 nanmean(mrnaFracBounds(order, :, :), 3), <span class="keyword">...</span>
3300                 nanstd(mrnaFracBounds(order, :, :), [], 3), <span class="keyword">...</span>
3301                 <span class="string">'LineStyle'</span>, <span class="string">'none'</span>, <span class="string">'Parent'</span>, axesHandle);
3302             xlim(axesHandle, [0.5 numel(g.mRNAIndexs) + 0.5])
3303             ylim(axesHandle, [
3304                 min(nanmean(mrnaFracBounds, 3) - nanstd(mrnaFracBounds, [], 3))  <span class="keyword">...</span>
3305                 max(nanmean(mrnaFracBounds, 3) + nanstd(mrnaFracBounds, [], 3))
3306                 ]);
3307             xlabel(axesHandle, <span class="string">'mRNA'</span>, <span class="string">'FontSize'</span>, 8)
3308             ylabel(axesHandle, {<span class="string">'Frac'</span> <span class="string">'Bound'</span>}, <span class="string">'FontSize'</span>, 8);
3309             
3310             <span class="comment">%fraction aminoacylated</span>
3311             axesHandle = subplot(3, 2, 6, <span class="string">'Parent'</span>, figHandle);
3312             axesHandles(3, 2) = axesHandle;
3313             expr = r.expression(r.matureIndexs(r.matureTRNAIndexs));
3314             expr = expr / sum(expr);
3315             errorbar(expr, <span class="keyword">...</span>
3316                 nanmean(rnaFracAminoacylated(r.matureTRNAIndexs, :, :), 3), <span class="keyword">...</span>
3317                 nanstd(rnaFracAminoacylated(r.matureTRNAIndexs, :, :), [], 3), <span class="keyword">...</span>
3318                 <span class="string">'LineStyle'</span>, <span class="string">'none'</span>, <span class="string">'Parent'</span>, axesHandle);
3319             xlim(axesHandle, [min(expr) max(expr)]);
3320             ylim(axesHandle, [
3321                 min(nanmean(rnaFracAminoacylated(r.matureTRNAIndexs, :, :), 3) - nanstd(rnaFracAminoacylated(r.matureTRNAIndexs, :, :), [], 3)) <span class="keyword">...</span>
3322                 max(nanmean(rnaFracAminoacylated(r.matureTRNAIndexs, :, :), 3) + nanstd(rnaFracAminoacylated(r.matureTRNAIndexs, :, :), [], 3))
3323                 ]);
3324             xlabel(axesHandle, <span class="string">'tRNA Exp'</span>, <span class="string">'FontSize'</span>, 8);
3325             ylabel(axesHandle, {<span class="string">'Frac'</span> <span class="string">'Aminoacylated'</span>}, <span class="string">'FontSize'</span>, 8);
3326             
3327             <span class="comment">%align axes labels</span>
3328             set(axesHandles, <span class="string">'FontSize'</span>, 6);
3329             PlotUtil.alignYAxesLabels(axesHandles(:, 1));
3330             PlotUtil.alignYAxesLabels(axesHandles(:, 2));
3331             
3332             <span class="comment">%% organize figure data</span>
3333             figData = struct;
3334             figData.rnaSynthesisTimes = rnaSynthesisTimes;
3335             figData.mrnaFracBounds = mrnaFracBounds;
3336             figData.rnaFracAminoacylated = rnaFracAminoacylated;
3337             figData.rnaPauses = rnaPauses;
3338         <span class="keyword">end</span>
3339         
3340         <a name="_sub20" href="#_subfunctions" class="code">function figData = proteinSynthesisDuration(figHandle, simBatchDir, selectedSimulations)</a>
3341             import edu.stanford.covert.cell.sim.util.PlotUtil;
3342             import edu.stanford.covert.cell.sim.util.SimulationEnsemble;
3343             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
3344             
3345             <span class="keyword">if</span> nargin &lt; 2 || isempty(simBatchDir)
3346                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
3347             <span class="keyword">end</span>
3348             <span class="keyword">if</span> nargin &lt; 3 || isempty(selectedSimulations)
3349                 selectedSimulations = SimulationDiskUtil.getCompleteSimulations(simBatchDir);
3350             <span class="keyword">end</span>
3351             
3352             <span class="comment">%% constants</span>
3353             [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep <span class="string">'1'</span>]);
3354             g = sim.gene;
3355             comp = sim.compartment;
3356             m = sim.state(<span class="string">'Metabolite'</span>);
3357             pm = sim.state(<span class="string">'ProteinMonomer'</span>);
3358             pc = sim.state(<span class="string">'ProteinComplex'</span>);
3359             rib = sim.state(<span class="string">'Ribosome'</span>);
3360             pol = sim.state(<span class="string">'Polypeptide'</span>);
3361             
3362             nSims = numel(selectedSimulations);
3363             pcComp = sum(pc.proteinComplexComposition(g.mRNAIndexs, :, :), 3);
3364             
3365             <span class="comment">%% get data</span>
3366             stateNames = {
3367                 <span class="string">'ProteinMonomer'</span>  <span class="string">'counts'</span>                   <span class="string">':'</span>  <span class="string">':'</span>
3368                 <span class="string">'ProteinComplex'</span>  <span class="string">'counts'</span>                   <span class="string">':'</span>  <span class="string">'-sum'</span>
3369                 <span class="string">'Ribosome'</span>        <span class="string">'states'</span>                   <span class="string">':'</span>  <span class="string">':'</span>
3370                 <span class="string">'Ribosome'</span>        <span class="string">'boundMRNAs'</span>               <span class="string">':'</span>  <span class="string">':'</span>
3371                 <span class="string">'Ribosome'</span>        <span class="string">'mRNAPositions'</span>            <span class="string">':'</span>  <span class="string">':'</span>
3372                 <span class="string">'Ribosome'</span>        <span class="string">'tmRNAPositions'</span>           <span class="string">':'</span>  <span class="string">':'</span>
3373                 };
3374             synthesisTimes = zeros(numel(pm.matureIndexs), 6, nSims);
3375             fracBounds = zeros(numel(pm.matureIndexs), 1, nSims);
3376             trnaPauses = zeros(36, 1000);
3377             aaPauses = zeros(21, 1000);
3378             <span class="keyword">for</span> i = 1:nSims
3379                 states = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, <span class="string">'extract'</span>, selectedSimulations(i));
3380                 states.Ribosome.states = permute(states.Ribosome.states, [1 3 2]);
3381                 states.Ribosome.boundMRNAs = permute(states.Ribosome.boundMRNAs, [1 3 2]);
3382                 states.Ribosome.mRNAPositions = permute(states.Ribosome.mRNAPositions, [1 3 2]);
3383                 states.Ribosome.tmRNAPositions = permute(states.Ribosome.tmRNAPositions, [1 3 2]);
3384                 
3385                 monCounts = permute(sum(states.ProteinMonomer.counts, 2), [1 3 2]);
3386                 cpxCounts = full(permute(sum(reshape(states.ProteinComplex.counts, <span class="keyword">...</span>
3387                     [numel(pc.matureIndexs)  numel(pc.wholeCellModelIDs)/numel(pc.matureIndexs)  size(states.ProteinComplex.counts, 3)]), 2), [1 3 2]));
3388                 nascs = full(monCounts(pm.nascentIndexs, :));
3389                 procsI = zeros(numel(pm.matureIndexs), size(states.ProteinMonomer.counts, 3));
3390                 procsI(pm.compartments(pm.processedIIIndexs) ~= comp.cytosolIndexs, :) = <span class="keyword">...</span>
3391                     full(permute(sum(states.ProteinMonomer.counts(<span class="keyword">...</span>
3392                     pm.processedIIndexs(pm.compartments(pm.processedIIIndexs) ~= comp.cytosolIndexs), <span class="keyword">...</span>
3393                     comp.cytosolIndexs, :), 2), [1 3 2]));
3394                 trans = full(monCounts(pm.processedIIndexs, :)) - procsI;
3395                 procsII = full(monCounts(pm.processedIIIndexs, :));
3396                 folds = full(monCounts(pm.foldedIndexs, :));
3397                 signals = full(monCounts(pm.signalSequenceIndexs, :));
3398                 matrs = full(permute(sum(reshape(permute(monCounts, [1 3 2]), <span class="keyword">...</span>
3399                     [numel(pm.matureIndexs)  numel(pm.wholeCellModelIDs)/numel(pm.matureIndexs)  size(states.ProteinMonomer.counts, 3)]), 2), [1 3 2])) <span class="keyword">...</span>
3400                     - nascs <span class="keyword">...</span>
3401                     - procsI <span class="keyword">...</span>
3402                     - trans <span class="keyword">...</span>
3403                     - procsII <span class="keyword">...</span>
3404                     - folds <span class="keyword">...</span>
3405                     - signals <span class="keyword">...</span>
3406                     + pcComp * cpxCounts;
3407                 bounds = <span class="keyword">...</span>
3408                     + monCounts(pm.boundIndexs, :) <span class="keyword">...</span>
3409                     + pcComp * full(permute(states.ProteinComplex.counts(pc.boundIndexs, :, :), [1 3 2]));
3410                 
3411                 <span class="comment">%synthesis times</span>
3412                 [startTimes, startRibs] = find(cat(2, <span class="keyword">...</span>
3413                     states.Ribosome.mRNAPositions(:, 1) &gt; 0 | <span class="keyword">...</span>
3414                     states.Ribosome.tmRNAPositions(:, 1) &gt; 0, <span class="keyword">...</span>
3415                     states.Ribosome.mRNAPositions(:, 1:end-1) == 0 &amp; <span class="keyword">...</span>
3416                     states.Ribosome.tmRNAPositions(:, 1:end-1) == 0 &amp; <span class="keyword">...</span>
3417                     (states.Ribosome.mRNAPositions(:, 2:end) &gt; 0 | <span class="keyword">...</span>
3418                     states.Ribosome.tmRNAPositions(:, 2:end) &gt; 0))');
3419                 [endTimes, endRibs] = find(cat(2, <span class="keyword">...</span>
3420                     (states.Ribosome.mRNAPositions(:, 1:end-1) &gt; 0 | <span class="keyword">...</span>
3421                     states.Ribosome.tmRNAPositions(:, 1:end-1) &gt; 0) &amp; <span class="keyword">...</span>
3422                     states.Ribosome.mRNAPositions(:, 2:end) == 0 &amp; <span class="keyword">...</span>
3423                     states.Ribosome.tmRNAPositions(:, 2:end) == 0, <span class="keyword">...</span>
3424                     states.Ribosome.mRNAPositions(:, end) &gt; 0 | <span class="keyword">...</span>
3425                     states.Ribosome.tmRNAPositions(:, end) &gt; 0)');
3426                 assert(all(startRibs == endRibs));
3427                 assert(all(endTimes &gt;= startTimes));
3428                 boundMRNAs = states.Ribosome.boundMRNAs(sub2ind(<span class="keyword">...</span>
3429                     size(states.Ribosome.boundMRNAs), <span class="keyword">...</span>
3430                     startRibs, startTimes));
3431                 synthesisTimes(:, 1, i) = <span class="keyword">...</span>
3432                     accumarray(boundMRNAs, endTimes - startTimes + 1, <span class="keyword">...</span>
3433                     [numel(pm.matureIndexs) 1]) ./ histc(boundMRNAs, (1:numel(pm.matureIndexs))');
3434                 
3435                 <span class="comment">%translation pauses</span>
3436                 pause = <span class="keyword">...</span>
3437                     states.Ribosome.states(:, 1:end-1) ~= rib.notExistValue &amp; <span class="keyword">...</span>
3438                     diff(states.Ribosome.mRNAPositions + states.Ribosome.tmRNAPositions, 1, 2) == 0;
3439                 [startTimes, startRibs] = find([pause(:, 1)  diff(pause, 1, 2) &gt; 0]');
3440                 [endTimes, endRibs] = find([diff(pause, 1, 2) &lt; 0  pause(:, end)]');
3441                 assert(all(startRibs == endRibs));
3442                 assert(all(endTimes &gt;= startTimes));
3443                 
3444                 <span class="keyword">for</span> j = 1:numel(startRibs)
3445                     <span class="keyword">if</span> <span class="keyword">...</span>
3446                             states.Ribosome.mRNAPositions(startRibs(j),  startTimes(j)) == pol.monomerLengths(states.Ribosome.boundMRNAs(startRibs(j),  startTimes(j))) || <span class="keyword">...</span>
3447                             states.Ribosome.tmRNAPositions(startRibs(j), startTimes(j)) == pol.proteolysisTagLength
3448                         <span class="keyword">continue</span>;
3449                     <span class="keyword">end</span>
3450                     
3451                     <span class="keyword">if</span> states.Ribosome.tmRNAPositions(startRibs(j), startTimes(j)) == 0
3452                         aa = pol.monomerAASequences{states.Ribosome.boundMRNAs(startRibs(j), startTimes(j))}(states.Ribosome.mRNAPositions(startRibs(j),  startTimes(j))+1);
3453                         tnra = pol.monomerTRNASequences{states.Ribosome.boundMRNAs(startRibs(j), startTimes(j))}(states.Ribosome.mRNAPositions(startRibs(j),  startTimes(j))+1);
3454                     <span class="keyword">else</span>
3455                         aa = pol.proteolysisTagAASequence(states.Ribosome.tmRNAPositions(startRibs(j), startTimes(j))+1);
3456                         tnra = pol.proteolysisTagTRNASequence(states.Ribosome.tmRNAPositions(startRibs(j), startTimes(j))+1);
3457                     <span class="keyword">end</span>
3458                     <span class="keyword">if</span> states.Ribosome.mRNAPositions(startRibs(j), startTimes(j)) == 0
3459                         aaIdx = 21;
3460                     <span class="keyword">else</span>
3461                         aaIdx = find(aa == edu.stanford.covert.cell.kb.ProteinMonomer.bases);
3462                     <span class="keyword">end</span>
3463                     <span class="keyword">if</span> endTimes(j) - startTimes(j) + 1 &gt; size(aaPauses, 2)
3464                         aaPauses = [aaPauses zeros(size(aaPauses, 1), (endTimes(j) - startTimes(j) + 1) - size(aaPauses, 2))]; <span class="comment">%#ok&lt;AGROW&gt;</span>
3465                         trnaPauses = [trnaPauses zeros(size(trnaPauses, 1), (endTimes(j) - startTimes(j) + 1) - size(trnaPauses, 2))]; <span class="comment">%#ok&lt;AGROW&gt;</span>
3466                     <span class="keyword">end</span>
3467                     aaPauses(aaIdx, endTimes(j) - startTimes(j) + 1) = <span class="keyword">...</span>
3468                         aaPauses(aaIdx, endTimes(j) - startTimes(j) + 1) + 1;
3469                     trnaPauses(tnra, endTimes(j) - startTimes(j) + 1) = <span class="keyword">...</span>
3470                         trnaPauses(tnra, endTimes(j) - startTimes(j) + 1) + 1;
3471                 <span class="keyword">end</span>
3472                 
3473                 <span class="comment">%processing I time</span>
3474                 synthesisTimes(:, 2, i) = log(2) ./ (sum(max(0, -diff(nascs, 1, 2)), 2) ./ sum(nascs(:, 1:end-1), 2));
3475                 
3476                 <span class="comment">%translocation time</span>
3477                 synthesisTimes(:, 3, i) = log(2) ./ (sum(max(0, -diff(procsI, 1, 2)), 2) ./ sum(procsI(:, 1:end-1), 2));
3478                 
3479                 <span class="comment">%processing II time</span>
3480                 synthesisTimes(:, 4, i) = log(2) ./ (sum(max(0, -diff(trans, 1, 2)), 2) ./ sum(trans(:, 1:end-1), 2));
3481                 
3482                 <span class="comment">%folding time</span>
3483                 synthesisTimes(:, 5, i) = log(2) ./ (sum(max(0, -diff(procsII, 1, 2)), 2) ./ sum(procsII(:, 1:end-1), 2));
3484                 
3485                 <span class="comment">%modification time</span>
3486                 synthesisTimes(:, 6, i) = log(2) ./ (sum(max(0, -diff(folds, 1, 2)), 2) ./ sum(folds(:, 1:end-1), 2));
3487                 
3488                 <span class="comment">%fraction bound</span>
3489                 fracBounds(:, 1, i) = sum(bounds, 2) ./ sum(matrs, 2);
3490                 
3491                 clear states monCounts cpxCounts nascs procsI trans procsII folds signals matrs bounds;
3492                 clear boundMRNAs startTimes endTimes startRibs endRibs pause;
3493             <span class="keyword">end</span>
3494             
3495             synthesisTimes(isinf(synthesisTimes)) = NaN;
3496             fracBounds(isinf(fracBounds)) = NaN;
3497             
3498             <span class="comment">%% plot</span>
3499             clf(figHandle);
3500             axesHandles = zeros(3, 2);
3501             
3502             <span class="comment">%synthesis times</span>
3503             axesHandle = subplot(3, 2, 1, <span class="string">'Parent'</span>, figHandle);
3504             axesHandles(1, 1) = axesHandle;
3505             tmp = nanmean(synthesisTimes, 3);
3506             [~, order] = sort(nansum(tmp, 2));
3507             h = bar(axesHandle, 1:numel(pm.matureIndexs),  tmp(order, :), <span class="string">'stacked'</span>);
3508             xlim(axesHandle, [0.5 numel(pm.matureIndexs) + 0.5]);
3509             ylim(axesHandle, [0 max(nansum(tmp, 2))]);
3510             legend(h, {<span class="string">'Translation'</span>, <span class="string">'Processing-I'</span>, <span class="string">'Translocation'</span>, <span class="string">'Processing-II'</span>, <span class="string">'Folding'</span>, <span class="string">'Modification'</span>}, <span class="keyword">...</span>
3511                 <span class="string">'Location'</span>, <span class="string">'NorthWest'</span>, <span class="string">'FontSize'</span>, 4);
3512             xlabel(axesHandle, <span class="string">'Monomer'</span>, <span class="string">'FontSize'</span>, 8);
3513             ylabel(axesHandle, {<span class="string">'Synthesis'</span> <span class="string">'Duration (s)'</span>}, <span class="string">'FontSize'</span>, 8);
3514             
3515             <span class="comment">%transcription time vs length</span>
3516             axesHandle = subplot(3, 2, 3, <span class="string">'Parent'</span>, figHandle);
3517             axesHandles(2, 1) = axesHandle;
3518             meanVal = nanmean(synthesisTimes(:, 1, :), 3);
3519             stdVal = nanstd(synthesisTimes(:, 1, :), [], 3);
3520             errorbar(pm.lengths(pm.nascentIndexs), meanVal, stdVal, <span class="keyword">...</span>
3521                 <span class="string">'LineStyle'</span>, <span class="string">'none'</span>, <span class="string">'Parent'</span>, axesHandle);
3522             xlim(axesHandle, [min(pm.lengths(pm.nascentIndexs)) max(pm.lengths(pm.nascentIndexs))]);
3523             ylim(axesHandle, [min(meanVal - stdVal) max(meanVal + stdVal)]);
3524             xlabel(axesHandle, <span class="string">'Length'</span>, <span class="string">'FontSize'</span>, 8);
3525             ylabel(axesHandle, {<span class="string">'Translation'</span> <span class="string">'Duration (s)'</span>}, <span class="string">'FontSize'</span>, 8)
3526             
3527             <span class="comment">%transcription time vs G/C content</span>
3528             axesHandle = subplot(3, 2, 5, <span class="string">'Parent'</span>, figHandle);
3529             axesHandles(3, 1) = axesHandle;
3530             [~, idx] = min(m.biomassComposition(m.aminoAcidIndexs(1:20)));
3531             aaContent = pm.baseCounts(pm.nascentIndexs, m.aminoAcidIndexs(idx)) ./ pm.lengths(pm.nascentIndexs);
3532             meanVal = nanmean(synthesisTimes(:, 1, :), 3);
3533             stdVal = nanstd(synthesisTimes(:, 1, :), [], 3);
3534             errorbar(aaContent, meanVal, stdVal, <span class="keyword">...</span>
3535                 <span class="string">'LineStyle'</span>, <span class="string">'none'</span>, <span class="string">'Parent'</span>, axesHandle);
3536             xlim(axesHandle, [min(aaContent) max(aaContent)]);
3537             ylim(axesHandle, [min(meanVal - stdVal) max(meanVal + stdVal)]);
3538             xlabel(axesHandle, [m.names{m.aminoAcidIndexs(idx)} <span class="string">' Content'</span>], <span class="string">'FontSize'</span>, 8);
3539             ylabel(axesHandle, {<span class="string">'Translation'</span> <span class="string">'Duration'</span>}, <span class="string">'FontSize'</span>, 8)
3540             
3541             <span class="comment">%pauses -- tRNA</span>
3542             axesHandle = subplot(3, 2, 2, <span class="string">'Parent'</span>, figHandle);
3543             axesHandles(1, 2) = axesHandle;
3544             tmp = trnaPauses ./ repmat(sum(trnaPauses, 2), 1, size(trnaPauses, 2));
3545             bar(axesHandle, 1:size(trnaPauses, 2), tmp', <span class="string">'group'</span>);
3546             xlim(axesHandle, [0.5 find(any(tmp &gt; 0.01, 1), 1, <span class="string">'last'</span>)+0.5]);
3547             ylim(axesHandle, [0 max(tmp(:))]);
3548             xlabel(axesHandle, <span class="string">'Pause Duration (s)'</span>, <span class="string">'FontSize'</span>, 8);
3549             ylabel(axesHandle, <span class="string">'Freq'</span>, <span class="string">'FontSize'</span>, 8)
3550             
3551             <span class="comment">%pauses -- AA</span>
3552             axesHandle = subplot(3, 2, 4, <span class="string">'Parent'</span>, figHandle);
3553             axesHandles(2, 2) = axesHandle;
3554             tmp = aaPauses ./ repmat(sum(aaPauses, 2), 1, size(aaPauses, 2));
3555             bar(axesHandle, 1:size(aaPauses, 2), tmp', <span class="string">'group'</span>);
3556             xlim(axesHandle, [0.5 find(any(tmp &gt; 0.01, 1), 1, <span class="string">'last'</span>) + 0.5]);
3557             ylim(axesHandle, [0 max(tmp(:))]);
3558             xlabel(axesHandle, <span class="string">'Pause Duration (s)'</span>, <span class="string">'FontSize'</span>, 8);
3559             ylabel(axesHandle, <span class="string">'Freq'</span>, <span class="string">'FontSize'</span>, 8)
3560             
3561             <span class="comment">%fraction bound</span>
3562             axesHandle = subplot(3, 2, 6, <span class="string">'Parent'</span>, figHandle);
3563             axesHandles(3, 2) = axesHandle;
3564             [~, order] = sort(nanmean(fracBounds, 3));
3565             order = order(any(fracBounds(order, :) &gt; 0, 2));
3566             errorbar((1:numel(order))', <span class="keyword">...</span>
3567                 nanmean(fracBounds(order, :, :), 3), <span class="keyword">...</span>
3568                 nanstd(fracBounds(order, :, :), [], 3), <span class="keyword">...</span>
3569                 <span class="string">'LineStyle'</span>, <span class="string">'none'</span>, <span class="string">'Parent'</span>, axesHandle);
3570             xlim(axesHandle, [0.5 numel(order) + 0.5])
3571             ylim(axesHandle, [
3572                 min(nanmean(fracBounds, 3) - nanstd(fracBounds, [], 3))  <span class="keyword">...</span>
3573                 max(nanmean(fracBounds, 3) + nanstd(fracBounds, [], 3))
3574                 ]);
3575             xlabel(axesHandle, <span class="string">'Monomer'</span>, <span class="string">'FontSize'</span>, 8)
3576             ylabel(axesHandle, {<span class="string">'Frac'</span> <span class="string">'Bound'</span>}, <span class="string">'FontSize'</span>, 8);
3577             
3578             <span class="comment">%align axes labels</span>
3579             set(axesHandles, <span class="string">'FontSize'</span>, 6);
3580             PlotUtil.alignYAxesLabels(axesHandles(:, 1));
3581             PlotUtil.alignYAxesLabels(axesHandles(:, 2));
3582             
3583             <span class="comment">%% organize figure data</span>
3584             figData = struct;
3585             figData.synthesisTimes = synthesisTimes;
3586             figData.fracBounds = fracBounds;
3587             figData.trnaPauses = trnaPauses;
3588             figData.aaPauses = aaPauses;
3589         <span class="keyword">end</span>
3590         
3591         <a name="_sub21" href="#_subfunctions" class="code">function figData = macromolecularComplexes(figHandle, simBatchDir, selectedSimulations)</a>
3592             import edu.stanford.covert.cell.sim.util.PlotUtil;
3593             import edu.stanford.covert.cell.sim.util.SimulationEnsemble;
3594             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
3595             import edu.stanford.covert.util.ConstantUtil;
3596             
3597             <span class="keyword">if</span> nargin &lt; 2 || isempty(simBatchDir)
3598                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
3599             <span class="keyword">end</span>
3600             <span class="keyword">if</span> nargin &lt; 3 || isempty(selectedSimulations)
3601                 selectedSimulations = SimulationDiskUtil.getCompleteSimulations(simBatchDir);
3602             <span class="keyword">end</span>
3603             
3604             <span class="comment">%% constants</span>
3605             [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep <span class="string">'1'</span>]);
3606             g = sim.gene;
3607             pm = sim.state(<span class="string">'ProteinMonomer'</span>);
3608             pc = sim.state(<span class="string">'ProteinComplex'</span>);
3609             
3610             pcComp = sum(pc.proteinComplexComposition(g.mRNAIndexs, :, :), 3);
3611             monIdxs = find(any(pcComp, 2));
3612             pcComp = pcComp(monIdxs, :);
3613             
3614             <span class="comment">%% load data</span>
3615             stateNames = {
3616                 <span class="string">'Geometry'</span>       <span class="string">'volume'</span>  <span class="string">':'</span>  <span class="string">':'</span>
3617                 <span class="string">'ProteinMonomer'</span> <span class="string">'counts'</span>  <span class="string">':'</span>  <span class="string">'-sum'</span>
3618                 <span class="string">'ProteinComplex'</span> <span class="string">'counts'</span>  <span class="string">':'</span>  <span class="string">'-sum'</span>
3619                 };
3620             monConc = zeros(numel(monIdxs), numel(selectedSimulations));
3621             cpxConc = zeros(numel(pc.matureIndexs), numel(selectedSimulations));
3622             <span class="keyword">for</span> i = 1:numel(selectedSimulations)
3623                 states = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, <span class="string">'extract'</span>, selectedSimulations(i));
3624                 states.ProteinMonomer.counts = full(sum(reshape(states.ProteinMonomer.counts, <span class="keyword">...</span>
3625                     [numel(pm.matureIndexs)  numel(pm.wholeCellModelIDs)/numel(pm.matureIndexs)  size(states.ProteinMonomer.counts, 3)]), 2));
3626                 states.ProteinMonomer.counts = states.ProteinMonomer.counts(monIdxs, :, :);
3627                 states.ProteinComplex.counts = full(sum(reshape(states.ProteinComplex.counts, <span class="keyword">...</span>
3628                     [numel(pc.matureIndexs)  numel(pc.wholeCellModelIDs)/numel(pc.matureIndexs)  size(states.ProteinComplex.counts, 3)]), 2));
3629                 
3630                 monConc(:, i) = mean((<span class="keyword">...</span>
3631                     + permute(states.ProteinMonomer.counts, [1 3 2]) <span class="keyword">...</span>
3632                     + pcComp * permute(states.ProteinComplex.counts, [1 3 2])) <span class="keyword">...</span>
3633                     ./ permute(states.Geometry.volume(ones(numel(monIdxs), 1), :, :) / ConstantUtil.nAvogadro * 1e6, [1 3 2]), <span class="keyword">...</span>
3634                     2);
3635                 cpxConc(:, i) = mean((<span class="keyword">...</span>
3636                     permute(states.ProteinComplex.counts, [1 3 2])) <span class="keyword">...</span>
3637                     ./ permute(states.Geometry.volume(ones(numel(pc.matureIndexs), 1), :, :) / ConstantUtil.nAvogadro * 1e6, [1 3 2]), <span class="keyword">...</span>
3638                     2);
3639                 
3640                 clear states;
3641             <span class="keyword">end</span>
3642             
3643             meanMonConc = mean(monConc, 2);
3644             stdMonConc = std(monConc, [], 2);
3645             meanCpxConc = pcComp * mean(cpxConc, 2);
3646             stdCpxConc = pcComp * std(cpxConc, [], 2);
3647             
3648             clear monConc cpxConc;
3649             
3650             <span class="comment">%% plot</span>
3651             clf(figHandle);
3652             [axesHandle, xAxesHandle] = PlotUtil.multiElementPlot(figHandle, 50, [min(meanCpxConc-stdCpxConc) max(meanCpxConc+stdCpxConc)], <span class="keyword">...</span>
3653                 struct(<span class="string">'xlabelStr'</span>, <span class="string">'Complex Concentration ({\mu}M)'</span>));
3654             xlabel(xAxesHandle, <span class="string">'Complex Concentration ({\mu}M)'</span>, <span class="string">'FontSize'</span>, 12, <span class="string">'Units'</span>, <span class="string">'normalized'</span>);
3655             
3656             PlotUtil.plotBubbles(axesHandle, [meanCpxConc stdCpxConc/3], [meanMonConc stdMonConc/3]);
3657             
3658             set(axesHandle, <span class="string">'YTick'</span>, [0 100]);
3659             ylim(axesHandle, [min(meanMonConc-stdMonConc) max(meanMonConc+stdMonConc)]);
3660             ylabel(axesHandle, <span class="string">'Monomer Concentration ({\mu}M)'</span>, <span class="string">'FontSize'</span>, 12);
3661             
3662             PlotUtil.offsetYAxes(axesHandle, 0.02);
3663             
3664             <span class="comment">%% organize figure data</span>
3665             figData = struct;
3666             figData.meanMonConc = meanMonConc;
3667             figData.stdMonConc = stdMonConc;
3668             figData.meanCpxConc = meanCpxConc;
3669             figData.stdCpxConc = stdCpxConc;
3670         <span class="keyword">end</span>
3671         
3672         <a name="_sub22" href="#_subfunctions" class="code">function figData = replication(figHandle, simBatchDir, selectedSimulations)</a>
3673             import edu.stanford.covert.cell.sim.util.PlotUtil;
3674             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
3675             import edu.stanford.covert.cell.sim.util.SimulationEnsemble;
3676             
3677             <span class="keyword">if</span> nargin &lt; 2 || isempty(simBatchDir)
3678                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
3679             <span class="keyword">end</span>
3680             <span class="keyword">if</span> nargin &lt; 3 || isempty(selectedSimulations)
3681                 selectedSimulations = SimulationDiskUtil.getCompleteSimulations(simBatchDir);
3682             <span class="keyword">end</span>
3683             
3684             <span class="comment">%% constants</span>
3685             [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep <span class="string">'1'</span>]);
3686             comp = sim.compartment;
3687             c = sim.state(<span class="string">'Chromosome'</span>);
3688             pm = sim.state(<span class="string">'ProteinMonomer'</span>);
3689             pc = sim.state(<span class="string">'ProteinComplex'</span>);
3690             rep = sim.process(<span class="string">'Replication'</span>);
3691             
3692             cpxIdxs = [
3693                 pc.matureIndexs(rep.enzymeGlobalIndexs([rep.enzymeIndexs_ssb4mer; rep.enzymeIndexs_ssb8mer]))
3694                 pc.boundIndexs(rep.enzymeGlobalIndexs([rep.enzymeIndexs_ssb4mer; rep.enzymeIndexs_ssb8mer]))
3695                 ]';
3696             
3697             nSims = numel(selectedSimulations);
3698             
3699             <span class="comment">%% get data</span>
3700             stateNames = {
3701                 <span class="string">'mass'</span>         <span class="string">'Mass (fg)'</span>
3702                 <span class="string">'growth_rate'</span>  <span class="string">'Growth (fg h^{-1})'</span>
3703                 <span class="string">'leadingPol1'</span>  <span class="string">'Leading Pol 1'</span>
3704                 <span class="string">'leadingPol2'</span>  <span class="string">'Leading Pol 2'</span>
3705                 <span class="string">'laggingPol1'</span>  <span class="string">'Lagging Pol 1'</span>
3706                 <span class="string">'laggingPol2'</span>  <span class="string">'Lagging Pol 2'</span>
3707                 };
3708             ensemble = SimulationEnsemble(simBatchDir, stateNames, [], selectedSimulations);
3709             mass = permute(ensemble.stateData.values(ensemble.getPropertyIndices(<span class="string">'mass'</span>), :, :, :) * sim.state(<span class="string">'Mass'</span>).cellInitialDryWeight / (1-sim.state(<span class="string">'Mass'</span>).fractionWetWeight) * 1e15, [4 3 1 2]);
3710             <a href="#_sub2" class="code" title="subfunction figData = growth(figHandle, simBatchDir, selectedSimulations)">growth</a> = permute(ensemble.stateData.values(ensemble.getPropertyIndices(<span class="string">'growth_rate'</span>), :, :, :) * 3600 * sim.state(<span class="string">'Mass'</span>).cellInitialDryWeight / (1 - sim.state(<span class="string">'Mass'</span>).fractionWetWeight) * 1e15, [4 3 1 2]);
3711             ledPol1 = permute(ensemble.stateData.values(ensemble.getPropertyIndices(<span class="string">'leadingPol1'</span>), :, :, :), [4 3 1 2]);
3712             ledPol2 = permute(ensemble.stateData.values(ensemble.getPropertyIndices(<span class="string">'leadingPol2'</span>), :, :, :), [4 3 1 2]);
3713             lagPol1 = permute(ensemble.stateData.values(ensemble.getPropertyIndices(<span class="string">'laggingPol1'</span>), :, :, :), [4 3 1 2]);
3714             lagPol2 = permute(ensemble.stateData.values(ensemble.getPropertyIndices(<span class="string">'laggingPol2'</span>), :, :, :), [4 3 1 2]);
3715             
3716             stateNames = {
3717                 <span class="string">'ProteinMonomer'</span> <span class="string">'counts'</span> pm.matureIndexs(rep.enzymeGlobalIndexs(rep.enzymeIndexs_ligase)) comp.cytosolIndexs
3718                 <span class="string">'ProteinComplex'</span> <span class="string">'counts'</span> cpxIdxs comp.cytosolIndexs
3719                 <span class="string">'Chromosome'</span>     <span class="string">'strandBreaks'</span> <span class="string">':'</span> <span class="string">':'</span>
3720                 };
3721             states = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, <span class="string">'extract'</span>, selectedSimulations);
3722             ssbs = permute(states.ProteinComplex.counts, [4 3 1 2]);
3723             [~, ligationTimes] = find(sum(states.Chromosome.strandBreaks, 3));
3724             
3725             repStarts = NaN(nSims, 1);
3726             repEnds = NaN(nSims, 1);
3727             kineticallyLimitedEnds = NaN(nSims, 1);
3728             repStartMass = NaN(nSims, 1);
3729             repStartGrowth = NaN(nSims, 1);
3730             repStartSSbs = NaN(nSims, 1);
3731             leadLeadCorr = NaN(nSims, 1);
3732             okFragSynTimes = [];
3733             <span class="keyword">for</span> i = 1:nSims
3734                 <span class="keyword">if</span> ~all(isnan(ledPol1(i, :)))
3735                     repStarts(i) = find(~isnan(ledPol1(i, :)), 1, <span class="string">'first'</span>);
3736                     repEnds(i) = find(~isnan(ledPol1(i, :)), 1, <span class="string">'last'</span>);
3737                     repStartMass(i) = mass(i, repStarts(i));
3738                     repStartGrowth(i) = <a href="#_sub2" class="code" title="subfunction figData = growth(figHandle, simBatchDir, selectedSimulations)">growth</a>(i, repStarts(i));
3739                     repStartSSbs(i) = ssbs(i, repStarts(i) - 1);
3740                     
3741                     tmp = find(conv(diff(ledPol2(i, :), 1, 2), ones(1, 100), <span class="string">'same'</span>) &gt; 0.75 * 100 * rep.dnaPolymeraseElongationRate, 1, <span class="string">'last'</span>) + 1;
3742                     <span class="keyword">if</span> ~isempty(tmp)
3743                         kineticallyLimitedEnds(i) = tmp;
3744                     <span class="keyword">else</span>
3745                         kineticallyLimitedEnds(i) = repEnds(i);
3746                     <span class="keyword">end</span>
3747                     
3748                     leadLeadCorr(i) = -corr(ledPol1(i, ~isnan(ledPol1(i, :)))', ledPol2(i, ~isnan(ledPol1(i, :)))');
3749                     
3750                     okFragSynTimes = [okFragSynTimes diff(find(diff(lagPol1(i, :), 1, 2) &lt; 0)) diff(find(diff(lagPol2(i, :), 1, 2) &gt; 0))]; <span class="comment">%#ok&lt;AGROW&gt;</span>
3751                 <span class="keyword">end</span>
3752             <span class="keyword">end</span>
3753             
3754             clear ensemble states mass <a href="#_sub2" class="code" title="subfunction figData = growth(figHandle, simBatchDir, selectedSimulations)">growth</a> ssbs;
3755             
3756             <span class="comment">%% plot data</span>
3757             clf(figHandle);
3758             axesHandles = zeros(5, 2);
3759             
3760             axesHandle = subplot(5, 2, 1, <span class="string">'Parent'</span>, figHandle);
3761             axesHandles(1, 1) = axesHandle;
3762             h = plot(axesHandle, repStartMass, [(repEnds - repStarts + 1)  (kineticallyLimitedEnds - repStarts + 1)  (repEnds - kineticallyLimitedEnds + 1)] / 3600, <span class="string">'.'</span>);
3763             <span class="keyword">if</span> any(repStartMass) &amp;&amp; range(repStartMass)
3764                 xlim(axesHandle, [min(repStartMass) max(repStartMass)])
3765             <span class="keyword">end</span>
3766             <span class="keyword">if</span> any(repEnds - repStarts) &amp;&amp; range(repEnds - repStarts)
3767                 ylim(axesHandle, [min(repEnds - repStarts + 1) max(repEnds - repStarts + 1)] / 3600)
3768             <span class="keyword">end</span>
3769             legend(h, {<span class="string">'Replication'</span>, <span class="string">'Kin Lim'</span>, <span class="string">'Met Lim'</span>}, <span class="string">'FontSize'</span>, 6)
3770             xlabel(axesHandle, <span class="string">'Mass (fg)'</span>, <span class="string">'FontSize'</span>, 8);
3771             ylabel(axesHandle, <span class="string">'Duration (h)'</span>, <span class="string">'FontSize'</span>, 8);
3772             
3773             axesHandle = subplot(5, 2, 3, <span class="string">'Parent'</span>, figHandle);
3774             axesHandles(2, 1) = axesHandle;
3775             plot(axesHandle, repStartGrowth, [(repEnds - repStarts + 1)  (kineticallyLimitedEnds - repStarts + 1)  (repEnds - kineticallyLimitedEnds + 1)] / 3600, <span class="string">'.'</span>);
3776             <span class="keyword">if</span> any(repStartGrowth) &amp;&amp; range(repStartGrowth)
3777                 xlim(axesHandle, [min(repStartGrowth) max(repStartGrowth)])
3778             <span class="keyword">end</span>
3779             <span class="keyword">if</span> any(repEnds - repStarts) &amp;&amp; range(repEnds - repStarts)
3780                 ylim(axesHandle, [min(repEnds - repStarts + 1) max(repEnds - repStarts + 1)] / 3600)
3781             <span class="keyword">end</span>
3782             xlabel(axesHandle, <span class="string">'Growth (fg h^{-1})'</span>, <span class="string">'FontSize'</span>, 8);
3783             ylabel(axesHandle, <span class="string">'Duration (h)'</span>, <span class="string">'FontSize'</span>, 8);
3784             
3785             axesHandle = subplot(5, 2, 5, <span class="string">'Parent'</span>, figHandle);
3786             axesHandles(3, 1) = axesHandle;
3787             plot(axesHandle, repStartSSbs, [(repEnds - repStarts + 1)  (kineticallyLimitedEnds - repStarts + 1)  (repEnds - kineticallyLimitedEnds + 1)] / 3600, <span class="string">'.'</span>);
3788             <span class="keyword">if</span> any(repStartSSbs) &amp;&amp; range(repStartSSbs)
3789                 xlim(axesHandle, [min(repStartSSbs) max(repStartSSbs)])
3790             <span class="keyword">end</span>
3791             <span class="keyword">if</span> any(repEnds - repStarts) &amp;&amp; range(repEnds - repStarts)
3792                 ylim(axesHandle, [min(repEnds - repStarts + 1) max(repEnds - repStarts + 1)] / 3600)
3793             <span class="keyword">end</span>
3794             xlabel(axesHandle, <span class="string">'SSBs'</span>, <span class="string">'FontSize'</span>, 8);
3795             ylabel(axesHandle, <span class="string">'Duration (h)'</span>, <span class="string">'FontSize'</span>, 8);
3796             
3797             axesHandle = subplot(5, 2, 2, <span class="string">'Parent'</span>, figHandle);
3798             axesHandles(1, 2) = axesHandle;
3799             tmp = [abs(ledPol1(:) - lagPol1(:)); abs(ledPol2(:) - lagPol2(:))] * 1e-3;
3800             hist(tmp)
3801             <span class="keyword">if</span> any(tmp)
3802                 xlim(axesHandle, [0.5 max(tmp)+0.5])
3803             <span class="keyword">end</span>
3804             ylabel(axesHandle, <span class="string">'Freq'</span>, <span class="string">'FontSize'</span>, 8);
3805             xlabel(axesHandle, {<span class="string">'Lead-Lag Distance (knt)'</span>}, <span class="string">'FontSize'</span>, 8);
3806             
3807             axesHandle = subplot(5, 2, 4, <span class="string">'Parent'</span>, figHandle);
3808             axesHandles(2, 2) = axesHandle;
3809             tmp = abs((c.sequenceLen - ledPol1(:)) - ledPol2(:)) * 1e-3;
3810             hist(tmp)
3811             <span class="keyword">if</span> any(tmp)
3812                 xlim(axesHandle, [0.5 max(tmp)+0.5])
3813             <span class="keyword">end</span>
3814             ylabel(axesHandle, <span class="string">'Freq'</span>, <span class="string">'FontSize'</span>, 8);
3815             xlabel(axesHandle, {<span class="string">'Lead-Lead Distance (knt)'</span>}, <span class="string">'FontSize'</span>, 8);
3816             
3817             axesHandle = subplot(5, 2, 6, <span class="string">'Parent'</span>, figHandle);
3818             axesHandles(3, 2) = axesHandle;
3819             hist(leadLeadCorr)
3820             <span class="keyword">if</span> any(leadLeadCorr) &amp;&amp; range(leadLeadCorr)
3821                 xlim(axesHandle, [min(leadLeadCorr) max(leadLeadCorr)])
3822             <span class="keyword">end</span>
3823             ylabel(axesHandle, <span class="string">'Freq'</span>, <span class="string">'FontSize'</span>, 8);
3824             xlabel(axesHandle, {<span class="string">'Lead-Lead Corr'</span>}, <span class="string">'FontSize'</span>, 8);
3825             
3826             axesHandle = subplot(5, 2, 8, <span class="string">'Parent'</span>, figHandle);
3827             axesHandles(4, 2) = axesHandle;
3828             hist(okFragSynTimes)
3829             <span class="keyword">if</span> ~isempty(okFragSynTimes) &amp;&amp; range(okFragSynTimes)
3830                 xlim(axesHandle, [min(okFragSynTimes) max(okFragSynTimes)])
3831             <span class="keyword">end</span>
3832             ylabel(axesHandle, <span class="string">'Freq'</span>, <span class="string">'FontSize'</span>, 8);
3833             xlabel(axesHandle, {<span class="string">'Okazaki Synthesis Time (s)'</span>}, <span class="string">'FontSize'</span>, 8);
3834             
3835             axesHandle = subplot(5, 2, 10, <span class="string">'Parent'</span>, figHandle);
3836             axesHandles(5, 2) = axesHandle;
3837             hist(ligationTimes)
3838             <span class="keyword">if</span> ~isempty(ligationTimes) &amp;&amp; range(ligationTimes)
3839                 xlim(axesHandle, [min(ligationTimes) max(ligationTimes)])
3840             <span class="keyword">end</span>
3841             ylabel(axesHandle, <span class="string">'Freq'</span>, <span class="string">'FontSize'</span>, 8);
3842             xlabel(axesHandle, {<span class="string">'Okazaki Synthesis Time (s)'</span>}, <span class="string">'FontSize'</span>, 8);
3843             
3844             <span class="comment">%format axes</span>
3845             <span class="keyword">for</span> i = 1:numel(axesHandles)
3846                 <span class="keyword">if</span> ~axesHandles(i)
3847                     <span class="keyword">continue</span>;
3848                 <span class="keyword">end</span>
3849                 set(axesHandles(i), <span class="string">'FontSize'</span>, 6);
3850             <span class="keyword">end</span>
3851             
3852             <span class="comment">%% organize figure data</span>
3853             figData = struct;
3854             figData.ledPol1 = ledPol1;
3855             figData.ledPol2 = ledPol2;
3856             figData.lagPol1 = lagPol1;
3857             figData.lagPol2 = lagPol2;
3858             figData.ligationTimes = ligationTimes;
3859             figData.repStarts = repStarts;
3860             figData.repEnds = repEnds;
3861             figData.kineticallyLimitedEnds = kineticallyLimitedEnds;
3862             figData.repStartMass = repStartMass;
3863             figData.repStartGrowth = repStartGrowth;
3864             figData.repStartSSbs = repStartSSbs;
3865             figData.leadLeadCorr = leadLeadCorr;
3866             figData.okFragSynTimes = okFragSynTimes;
3867         <span class="keyword">end</span>
3868         
3869         <a name="_sub23" href="#_subfunctions" class="code">function figData = SSBs(figHandle, simBatchDir, selectedSimulations)</a>
3870             import edu.stanford.covert.cell.sim.util.PlotUtil;
3871             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
3872             import edu.stanford.covert.cell.sim.util.SimulationEnsemble;
3873             
3874             <span class="keyword">if</span> nargin &lt; 2 || isempty(simBatchDir)
3875                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
3876             <span class="keyword">end</span>
3877             <span class="keyword">if</span> nargin &lt; 3 || isempty(selectedSimulations)
3878                 selectedSimulations = SimulationDiskUtil.getCompleteSimulations(simBatchDir);
3879             <span class="keyword">end</span>
3880             
3881             <span class="comment">%% constants</span>
3882             [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep <span class="string">'1'</span>]);
3883             comp = sim.compartment;
3884             pc = sim.state(<span class="string">'ProteinComplex'</span>);
3885             
3886             cytosolIndexs = comp.cytosolIndexs;
3887             cpxIdxs = [
3888                 pc.matureIndexs(pc.getIndexs(<span class="string">'MG_091_TETRAMER'</span>))
3889                 pc.matureIndexs(pc.getIndexs(<span class="string">'MG_091_OCTAMER'</span>))
3890                 pc.boundIndexs(pc.getIndexs(<span class="string">'MG_091_OCTAMER'</span>))
3891                 ];
3892             
3893             <span class="comment">%% get data</span>
3894             stateNames = {
3895                 <span class="string">'Time'</span>            <span class="string">'values'</span>  <span class="string">':'</span>      <span class="string">':'</span>
3896                 <span class="string">'Mass'</span>            <span class="string">'cell'</span>    <span class="string">':'</span>      <span class="string">'-sum'</span>
3897                 <span class="string">'ProteinComplex'</span>  <span class="string">'counts'</span>  cpxIdxs  cytosolIndexs
3898                 };
3899             tmpStates = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, <span class="string">'extract'</span>, selectedSimulations);
3900             
3901             mass = permute(tmpStates.Mass.cell, [3 4 1 2]);
3902             simEndTimes = permute(max(tmpStates.Time.values, [], 3), [4 3 1 2]);
3903             minSimEndTime = min(simEndTimes);
3904             [~, simOrder] = sort(mass(minSimEndTime, :));
3905             
3906             states.ProteinComplex.free4mer = permute(tmpStates.ProteinComplex.counts(1, :, :, :), [3 4 1 2]);
3907             states.ProteinComplex.free8mer = permute(tmpStates.ProteinComplex.counts(2, :, :, :), [3 4 1 2]);
3908             states.ProteinComplex.bound8mer = permute(tmpStates.ProteinComplex.counts(3, :, :, :), [3 4 1 2]);
3909             
3910             time = (1:size(states.ProteinComplex.bound8mer, 1))' / 3600;
3911             
3912             clear tmpStates mass;
3913             
3914             <span class="comment">%% plot</span>
3915             clf(figHandle);
3916             options = struct;
3917             options.colorOrder = PlotUtil.getRedGreenColorOrder(simOrder);
3918             options.xdata = time;
3919             options.ylabelStr = {
3920                 {<span class="string">'Bound'</span> <span class="string">'8mer'</span>}
3921                 {<span class="string">'Free'</span> <span class="string">'8mer'</span>}
3922                 {<span class="string">'Free'</span> <span class="string">'4mer'</span>}
3923                 };
3924             options.ydata = {
3925                 states.ProteinComplex.bound8mer
3926                 states.ProteinComplex.free8mer
3927                 states.ProteinComplex.free4mer
3928                 };
3929             PlotUtil.multiElementPlot(figHandle, [10; 10; 10], [0 time(end)], options);
3930             
3931             clear options;
3932             
3933             <span class="comment">%% organize figure data</span>
3934             figData = struct;
3935             figData.time = time;
3936             figData.free4mer = states.ProteinComplex.free4mer;
3937             figData.free8mer = states.ProteinComplex.free8mer;
3938             figData.bound8mer = states.ProteinComplex.bound8mer;
3939         <span class="keyword">end</span>
3940         
3941         <a name="_sub24" href="#_subfunctions" class="code">function figData = dnaRepair(figHandle, simBatchDir, selectedSimulations)</a>
3942             import edu.stanford.covert.cell.sim.util.PlotUtil;
3943             import edu.stanford.covert.cell.sim.util.SimulationEnsemble;
3944             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
3945             import edu.stanford.covert.util.CircularSparseMat;
3946             import edu.stanford.covert.util.ConstantUtil;
3947             
3948             <span class="keyword">if</span> nargin &lt; 2 || isempty(simBatchDir)
3949                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
3950             <span class="keyword">end</span>
3951             <span class="keyword">if</span> nargin &lt; 3 || isempty(selectedSimulations)
3952                 selectedSimulations = SimulationDiskUtil.getCompleteSimulations(simBatchDir);
3953             <span class="keyword">end</span>
3954             
3955             [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep <span class="string">'1'</span>]);
3956             pm = sim.state(<span class="string">'ProteinMonomer'</span>);
3957             pc = sim.state(<span class="string">'ProteinComplex'</span>);
3958             <a href="#_sub25" class="code" title="subfunction figData = dnaRepair(figHandle, simBatchDir, selectedSimulations)">dnaRepair</a> = sim.process(<span class="string">'DNARepair'</span>);
3959             
3960             monIdxs = pm.matureIndexs(dnaRepair.enzymeMonomerGlobalIndexs);
3961             cpxIdxs = pc.matureIndexs(dnaRepair.enzymeComplexGlobalIndexs);
3962             
3963             <span class="comment">%% get data</span>
3964             stateNames = {
3965                 <span class="string">'Time'</span>            <span class="string">'values'</span>                  <span class="string">':'</span>      <span class="string">':'</span>
3966                 <span class="string">'Chromosome'</span>      <span class="string">'abasicSites'</span>             <span class="string">'-sum'</span>   <span class="string">'-sum'</span>
3967                 <span class="string">'Chromosome'</span>      <span class="string">'gapSites'</span>                <span class="string">'-sum'</span>   <span class="string">'-sum'</span>
3968                 <span class="string">'Chromosome'</span>      <span class="string">'damagedSugarPhosphates'</span>  <span class="string">'-sum'</span>   <span class="string">'-sum'</span>
3969                 <span class="string">'Chromosome'</span>      <span class="string">'intrastrandCrossLinks'</span>   <span class="string">'-sum'</span>   <span class="string">'-sum'</span>
3970                 <span class="string">'Chromosome'</span>      <span class="string">'strandBreaks'</span>            <span class="string">'-sum'</span>   <span class="string">'-sum'</span>
3971                 <span class="string">'Chromosome'</span>      <span class="string">'hollidayJunctions'</span>       <span class="string">'-sum'</span>   <span class="string">'-sum'</span>
3972                 <span class="string">'Mass'</span>            <span class="string">'cell'</span>                    <span class="string">':'</span>      <span class="string">'-sum'</span>
3973                 <span class="string">'ProteinMonomer'</span>  <span class="string">'counts'</span>                  monIdxs  <span class="string">'-sum'</span>
3974                 <span class="string">'ProteinComplex'</span>  <span class="string">'counts'</span>                  cpxIdxs  <span class="string">'-sum'</span>
3975                 };
3976             states = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, <span class="string">'extract'</span>, selectedSimulations);
3977             
3978             time = permute(max(states.Time.values, [], 4), [3 4 1 2]) / 3600;
3979             mass = permute(states.Mass.cell, [3 4 1 2]);
3980             
3981             simEndTimes = permute(max(states.Time.values, [], 3), [4 3 1 2]);
3982             minSimEndTime = min(simEndTimes);
3983             [~, simOrder] = sort(mass(minSimEndTime, :));
3984             
3985             clear mass;
3986             
3987             <span class="comment">%damage</span>
3988             gapSites = full(states.Chromosome.gapSites);
3989             abasicSites = full(states.Chromosome.abasicSites);
3990             damagedSugarPhosphates = full(states.Chromosome.damagedSugarPhosphates);
3991             intrastrandCrossLinks = full(states.Chromosome.intrastrandCrossLinks);
3992             strandBreaks = full(states.Chromosome.strandBreaks);
3993             hollidayJunctions = full(states.Chromosome.hollidayJunctions);
3994             
3995             <span class="comment">%enzymes</span>
3996             enzymes(dnaRepair.enzymeMonomerLocalIndexs, :, :, :) = states.ProteinMonomer.counts;
3997             enzymes(dnaRepair.enzymeComplexLocalIndexs, :, :, :) = states.ProteinComplex.counts;
3998             
3999             <span class="comment">%cleanup</span>
4000             clear states;
4001             
4002             <span class="comment">%damaged bases -- calculate separately to minimize memory profile</span>
4003             damagedBases = zeros(size(gapSites));
4004             <span class="keyword">for</span> i = 1:numel(selectedSimulations)
4005                 tmpStates = SimulationEnsemble.load(simBatchDir, {<span class="string">'Chromosome'</span> <span class="string">'damagedBases'</span>}, [], [], 1, <span class="string">'extract'</span>, selectedSimulations(i));
4006                 [posStrnds, vals] = find(tmpStates.Chromosome.damagedBases);
4007                 tfs = ~ismember(vals, dnaRepair.substrateGlobalIndexs(dnaRepair.substrateIndexs_m6AD));
4008                 damagedBases(:, :, 1:size(tmpStates.Chromosome.damagedBases, 3), i) = <span class="keyword">...</span>
4009                     full(sum(sum(CircularSparseMat(posStrnds(tfs, :), ones(sum(tfs), 1), size(tmpStates.Chromosome.damagedBases), 1), 2), 1));
4010                 clear tmpStates posStrnds vals tfs;
4011             <span class="keyword">end</span>
4012             
4013             <span class="comment">%NaN</span>
4014             <span class="keyword">for</span> i = 1:numel(simEndTimes)
4015                 gapSites(:, :, simEndTimes(i)+1:<span class="keyword">end</span>, i) = NaN;
4016                 abasicSites(:, :, simEndTimes(i)+1:<span class="keyword">end</span>, i) = NaN;
4017                 damagedSugarPhosphates(:, :, simEndTimes(i)+1:<span class="keyword">end</span>, i) = NaN;
4018                 intrastrandCrossLinks(:, :, simEndTimes(i)+1:<span class="keyword">end</span>, i) = NaN;
4019                 strandBreaks(:, :, simEndTimes(i)+1:<span class="keyword">end</span>, i) = NaN;
4020                 hollidayJunctions(:, :, simEndTimes(i)+1:<span class="keyword">end</span>, i) = NaN;
4021                 enzymes(:, :, simEndTimes(i)+1:<span class="keyword">end</span>, i) = NaN;
4022             <span class="keyword">end</span>
4023             
4024             <span class="comment">%% plot data</span>
4025             clf(figHandle);
4026             
4027             options = struct();
4028             options.titleStr = {<span class="string">'Damage'</span>; <span class="string">'Repair'</span>};
4029             options.colorOrder = PlotUtil.getRedGreenColorOrder(simOrder);
4030             options.xdata = time;
4031             options.ylabelStr = {
4032                 {
4033                 {<span class="string">'Gap'</span> <span class="string">'Sites'</span>}
4034                 {<span class="string">'Abasic'</span> <span class="string">'Sites'</span>}
4035                 {<span class="string">'Damaged'</span> <span class="string">'Sugar-Phosphates'</span>}
4036                 {<span class="string">'Damaged'</span> <span class="string">'Bases'</span>}
4037                 {<span class="string">'Intrastrand'</span> <span class="string">'Crosslinks'</span>}
4038                 {<span class="string">'Strand'</span> <span class="string">'Breaks'</span>}
4039                 {<span class="string">'Holliday'</span> <span class="string">'Junctions'</span>}
4040                 }
4041                 dnaRepair.enzymeWholeCellModelIDs
4042                 };
4043             options.ydata = {
4044                 {
4045                 permute(gapSites, [3 4 1 2])
4046                 permute(abasicSites, [3 4 1 2])
4047                 permute(damagedSugarPhosphates, [3 4 1 2])
4048                 permute(damagedBases, [3 4 1 2])
4049                 permute(intrastrandCrossLinks, [3 4 1 2])
4050                 permute(strandBreaks, [3 4 1 2])
4051                 permute(hollidayJunctions, [3 4 1 2])
4052                 }
4053                 permute(mat2cell(permute(enzymes, [3 4 1 2]), size(enzymes, 3), size(enzymes, 4), ones(size(enzymes, 1), 1)), [3 1 2])
4054                 };
4055             PlotUtil.multiElementPlot(figHandle, cellfun(@(x) 3 * ones(size(x)), options.ylabelStr, <span class="string">'UniformOutput'</span>, false), [0 time(end)], options);
4056             
4057             clear options;
4058             
4059             <span class="comment">%% organize figure data</span>
4060             figData = struct;
4061             figData.simOrder = simOrder;
4062             figData.time = time;
4063             figData.gapSites = gapSites;
4064             figData.abasicSites = abasicSites;
4065             figData.damagedSugarPhosphates = damagedSugarPhosphates;
4066             figData.damagedBases = damagedBases;
4067             figData.intrastrandCrossLinks = intrastrandCrossLinks;
4068             figData.strandBreaks = strandBreaks;
4069             figData.hollidayJunctions = hollidayJunctions;
4070             figData.enzymes = enzymes;
4071         <span class="keyword">end</span>
4072         
4073         <a name="_sub25" href="#_subfunctions" class="code">function figData = immuneActivation(figHandle, simBatchDir, selectedSimulations)</a>
4074             import edu.stanford.covert.cell.sim.util.PlotUtil;
4075             import edu.stanford.covert.cell.sim.util.SimulationEnsemble;
4076             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
4077             import edu.stanford.covert.util.ConstantUtil;
4078             
4079             <span class="keyword">if</span> nargin &lt; 2 || isempty(simBatchDir)
4080                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
4081             <span class="keyword">end</span>
4082             <span class="keyword">if</span> nargin &lt; 3 || isempty(selectedSimulations)
4083                 selectedSimulations = SimulationDiskUtil.getCompleteSimulations(simBatchDir);
4084             <span class="keyword">end</span>
4085             
4086             [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep <span class="string">'1'</span>]);
4087             comp = sim.compartment;
4088             pm = sim.state(<span class="string">'ProteinMonomer'</span>);
4089             host = sim.state(<span class="string">'Host'</span>);
4090             
4091             <span class="comment">%% get data</span>
4092             monIdxs = pm.matureIndexs(ismember(pm.compartments(pm.matureIndexs), [comp.terminalOrganelleCytosolIndexs comp.terminalOrganelleMembraneIndexs]));
4093             stateNames = {
4094                 <span class="string">'Time'</span>            <span class="string">'values'</span>                           <span class="string">':'</span> <span class="string">':'</span>
4095                 <span class="string">'Mass'</span>            <span class="string">'cell'</span>                             <span class="string">':'</span> <span class="string">'-sum'</span>
4096                 <span class="string">'Host'</span>            <span class="string">'isBacteriumAdherent'</span>              <span class="string">':'</span> <span class="string">':'</span>
4097                 <span class="string">'Host'</span>            <span class="string">'isTLRActivated'</span>                   <span class="string">':'</span> <span class="string">':'</span>
4098                 <span class="string">'Host'</span>            <span class="string">'isNFkBActivated'</span>                  <span class="string">':'</span> <span class="string">':'</span>
4099                 <span class="string">'Host'</span>            <span class="string">'isInflammatoryResponseActivated'</span>  <span class="string">':'</span> <span class="string">':'</span>
4100                 <span class="string">'ProteinMonomer'</span>  <span class="string">'counts'</span>                           monIdxs [comp.terminalOrganelleCytosolIndexs comp.terminalOrganelleMembraneIndexs]
4101                 };
4102             states = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, <span class="string">'extract'</span>, selectedSimulations);
4103             
4104             time = permute(max(states.Time.values, [], 4), [3 4 1 2]) / 3600;
4105             mass = permute(states.Mass.cell, [3 4 1 2]);
4106             adherence = double(permute(states.Host.isBacteriumAdherent, [3 4 1 2]));
4107             tlr = double(permute(states.Host.isTLRActivated, [3 4 1 2]));
4108             nfkb = double(permute(states.Host.isNFkBActivated, [3 4 1 2]));
4109             inflammation = double(permute(states.Host.isInflammatoryResponseActivated, [3 4 1 2]));
4110             proteins = permute(states.ProteinMonomer.counts, [3 4 1 2]);
4111             
4112             simEndTimes = permute(max(states.Time.values, [], 3), [4 3 1 2]);
4113             minSimEndTime = min(simEndTimes);
4114             [~, simOrder] = sort(mass(minSimEndTime, :));
4115             
4116             <span class="keyword">for</span> i = 1:numel(simEndTimes)
4117                 adherence(simEndTimes(i)+1:<span class="keyword">end</span>, i, :) = NaN;
4118                 tlr(simEndTimes(i)+1:<span class="keyword">end</span>, i, :) = NaN;
4119                 nfkb(simEndTimes(i)+1:<span class="keyword">end</span>, i, :) = NaN;
4120                 inflammation(simEndTimes(i)+1:<span class="keyword">end</span>, i, :) = NaN;
4121                 proteins(simEndTimes(i)+1:<span class="keyword">end</span>, i, :) = NaN;
4122             <span class="keyword">end</span>
4123             
4124             clear states mass;
4125             
4126             <span class="comment">%% plot data</span>
4127             clf(figHandle);
4128             
4129             options = struct();
4130             options.titleStr = {<span class="string">'{\it M. genitalium}'</span> <span class="string">'Host Response'</span>};
4131             options.colorOrder = PlotUtil.getRedGreenColorOrder(simOrder);
4132             options.xdata = time;
4133             options.ylabelStr = {
4134                 cellfun(@(x) strrep(x, <span class="string">'_'</span>, <span class="string">'\_'</span>), pm.wholeCellModelIDs(monIdxs), <span class="string">'UniformOutput'</span>, false)
4135                 {
4136                 <span class="string">'Adherence'</span>
4137                 <span class="string">'TLR 1'</span>
4138                 <span class="string">'TLR 2'</span>
4139                 <span class="string">'TLR 6'</span>
4140                 <span class="string">'NF-{\kappa}B'</span>
4141                 <span class="string">'Inflammation'</span>
4142                 }
4143                 };
4144             options.ydata = {
4145                 mat2cell(proteins, size(proteins, 1), size(proteins, 2), ones(size(proteins, 3), 1))
4146                 {
4147                 adherence
4148                 tlr(:, :, host.tlrIndexs_1)
4149                 tlr(:, :, host.tlrIndexs_2)
4150                 tlr(:, :, host.tlrIndexs_6)
4151                 nfkb
4152                 inflammation
4153                 }};
4154             PlotUtil.multiElementPlot(figHandle, {3 * ones(size(monIdxs)) 3 * ones(6, 1)}, [0 time(end)], options);
4155             
4156             clear options;
4157             
4158             <span class="comment">%% organize figure data</span>
4159             figData = struct;
4160             figData.simOrder = simOrder;
4161             figData.time = time;
4162             figData.proteins = proteins;
4163             figData.adherence = adherence;
4164             figData.tlr = tlr;
4165             figData.nfkb = nfkb;
4166             figData.inflammation = inflammation;
4167         <span class="keyword">end</span>
4168         
4169         <a name="_sub26" href="#_subfunctions" class="code">function figData = smcSpacing(figHandle, simBatchDir, selectedSimulations)</a>
4170             import edu.stanford.covert.cell.sim.util.DiskLogger;
4171             import edu.stanford.covert.cell.sim.util.SimulationEnsemble;
4172             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
4173             import edu.stanford.covert.cell.sim.util.PlotUtil;
4174             
4175             <span class="keyword">if</span> nargin &lt; 2 || isempty(simBatchDir)
4176                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
4177             <span class="keyword">end</span>
4178             <span class="keyword">if</span> nargin &lt; 3 || isempty(selectedSimulations)
4179                 selectedSimulations = SimulationDiskUtil.getCompleteSimulations(simBatchDir);
4180             <span class="keyword">end</span>
4181             
4182             [simDir, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep <span class="string">'1'</span>]);
4183             simBatchDir = SimulationDiskUtil.getSimulationBatchDir(simDir);
4184             pc = sim.state(<span class="string">'ProteinComplex'</span>);
4185             ch = sim.state(<span class="string">'Chromosome'</span>);
4186             
4187             stateNames = {<span class="keyword">...</span>
4188                 <span class="string">'Mass'</span>       <span class="string">'cell'</span>                     <span class="string">':'</span>     <span class="string">'-sum'</span>
4189                 <span class="string">'Time'</span>       <span class="string">'values'</span>                   <span class="string">':'</span>     <span class="string">':'</span>
4190                 };
4191             states = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, <span class="string">'extract'</span>, selectedSimulations);
4192             simEndTimes = permute(max(states.Time.values, [], 3), [4 1 2 3]);
4193             timeIdx = min(simEndTimes);
4194             [~, simOrder] = sort(states.Mass.cell(:, :, timeIdx, :));
4195             simOrder = simOrder(:);
4196             clear(<span class="string">'states'</span>);
4197             
4198             maxTime = 0;
4199             <span class="keyword">for</span> i = 1:length(selectedSimulations)
4200                 summary = load([simBatchDir filesep num2str(selectedSimulations(i)) filesep <span class="string">'summary.mat'</span>], <span class="string">'time'</span>);
4201                 maxTime = max(maxTime, sum(summary.time &gt;= 0));
4202             <span class="keyword">end</span>
4203             
4204             time = 1:maxTime;
4205             <a href="#_sub27" class="code" title="subfunction figData = smcSpacing(figHandle, simBatchDir, selectedSimulations)">smcSpacing</a> = zeros(numel(selectedSimulations), numel(time));
4206             
4207             stateNames = {<span class="keyword">...</span>
4208                 <span class="string">'Chromosome'</span> <span class="string">'complexBoundSites'</span>        <span class="string">':'</span>     <span class="string">':'</span>
4209                 };
4210             
4211             <span class="keyword">for</span> selSimIdx = 1:numel(selectedSimulations)
4212                 <span class="comment">% Run out of memory when using SimulationEnsemble</span>
4213                 <span class="comment">% Instead, load and analyze each simulation one at a time</span>
4214                 states = DiskLogger.load([simBatchDir filesep num2str(selectedSimulations(selSimIdx))], stateNames, [], [], 1, <span class="string">'extract'</span>);
4215                 
4216                 cbs = states.Chromosome.complexBoundSites;
4217                 subs = find(cbs == pc.getIndexs(<span class="string">'MG_213_214_298_6MER_ADP'</span>));
4218                 subs = subs(subs(:, 2) &lt;= 2, :);
4219                 
4220                 <a href="#_sub27" class="code" title="subfunction figData = smcSpacing(figHandle, simBatchDir, selectedSimulations)">smcSpacing</a>(selSimIdx, :) = ch.sequenceLen ./  histc(subs(:, 3), time);
4221                 
4222             <span class="keyword">end</span>
4223             
4224             time = time / 3600;
4225             <a href="#_sub27" class="code" title="subfunction figData = smcSpacing(figHandle, simBatchDir, selectedSimulations)">smcSpacing</a>(<a href="#_sub27" class="code" title="subfunction figData = smcSpacing(figHandle, simBatchDir, selectedSimulations)">smcSpacing</a> == Inf) = NaN;
4226             
4227             <span class="comment">%% plot data</span>
4228             clf(figHandle);
4229             options = struct;
4230             options.titleStr = {<span class="string">'SMC Spacing'</span>};
4231             options.colorOrder = PlotUtil.getRedGreenColorOrder(simOrder);
4232             options.xdata = time;
4233             options.ydata = {
4234                 <a href="#_sub27" class="code" title="subfunction figData = smcSpacing(figHandle, simBatchDir, selectedSimulations)">smcSpacing</a>
4235                 };
4236             options.ylabelStr = {
4237                 {<span class="string">'Average Distance (bp)'</span>}
4238                 };
4239             PlotUtil.multiElementPlot(figHandle, {30}, [0 time(end)], options);
4240             
4241             clear options;
4242             
4243             <span class="comment">%% organize figure data</span>
4244             figData = struct;
4245             figData.simOrder = simOrder;
4246             figData.time = time;
4247             figData.smcSpacing = <a href="#_sub27" class="code" title="subfunction figData = smcSpacing(figHandle, simBatchDir, selectedSimulations)">smcSpacing</a>;
4248         <span class="keyword">end</span>
4249         
4250         <a name="_sub27" href="#_subfunctions" class="code">function figData = geneCopyNumber(figHandle, simBatchDir, selectedSimulations)</a>
4251             import edu.stanford.covert.cell.sim.util.SimulationEnsemble;
4252             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
4253             import edu.stanford.covert.cell.sim.util.PlotUtil;
4254             import edu.stanford.covert.cell.sim.util.DiskLogger;
4255             <span class="keyword">if</span> ~exist(<span class="string">'simBatchDir'</span>, <span class="string">'var'</span>) || isempty(simBatchDir)
4256                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
4257             <span class="keyword">end</span>
4258             <span class="keyword">if</span> ~exist(<span class="string">'selectedSimulations'</span>, <span class="string">'var'</span>) || isempty(selectedSimulations)
4259                 selectedSimulations = SimulationDiskUtil.getCompleteSimulations(simBatchDir);
4260             <span class="keyword">end</span>
4261             
4262             [simDir, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep <span class="string">'1'</span>]);
4263             simBatchDir = SimulationDiskUtil.getSimulationBatchDir(simDir);
4264             ch = sim.state(<span class="string">'Chromosome'</span>);
4265             
4266             <span class="comment">%% get data</span>
4267             stateNames = {<span class="keyword">...</span>
4268                 <span class="string">'Mass'</span>       <span class="string">'cell'</span>                     <span class="string">':'</span>     <span class="string">'-sum'</span>
4269                 <span class="string">'Time'</span>       <span class="string">'values'</span>                   <span class="string">':'</span>     <span class="string">':'</span>
4270                 };
4271             states = SimulationEnsemble.load(simBatchDir, stateNames, [], [], 1, <span class="string">'extract'</span>, selectedSimulations);
4272             simEndTimes = permute(max(states.Time.values, [], 3), [4 1 2 3]);
4273             timeIdx = min(simEndTimes);
4274             [~, simOrder] = sort(states.Mass.cell(:, :, timeIdx, :));
4275             simOrder = simOrder(:);
4276             clear(<span class="string">'states'</span>);
4277             
4278             stateNames = {<span class="keyword">...</span>
4279                 <span class="string">'Chromosome'</span> <span class="string">'polymerizedRegions'</span>        <span class="string">':'</span>     <span class="string">':'</span>
4280                 };
4281             
4282             copyNumber = zeros(numel(selectedSimulations), ch.sequenceLen);
4283             
4284             <span class="keyword">for</span> i = 1:numel(selectedSimulations)
4285                 <span class="comment">% Function is ~5x faster if I do this rather than</span>
4286                 <span class="comment">% SimulationEnsemble.load() outside of the loop due to</span>
4287                 <span class="comment">% avoiding SparseMat concatenation</span>
4288                 states = DiskLogger.load([simBatchDir filesep num2str(selectedSimulations(i))], stateNames, [], [], 1, <span class="string">'extract'</span>);
4289                 
4290                 [subs vals] = find(states.Chromosome.polymerizedRegions);
4291                 
4292                 startpos = subs(:, 1);
4293                 endpos = startpos + vals(:, 1) - 1;
4294                 
4295                 copyNumber(i, :) = cumsum(histc(startpos, 1:ch.sequenceLen) - histc(endpos + 1, 1:ch.sequenceLen)) / simEndTimes(i) / 2; <span class="comment">% 2 strands per chromosome</span>
4296             <span class="keyword">end</span>
4297             
4298             <span class="comment">%% plot data</span>
4299             clf(figHandle);
4300             options = struct;
4301             options.titleStr = {<span class="string">'Copy Number Over Cell Cycle'</span>};
4302             options.colorOrder = PlotUtil.getRedGreenColorOrder(simOrder);
4303             options.xdata = 1:ch.sequenceLen;
4304             options.xlabelStr = {<span class="string">'Position on Chromosome'</span>};
4305             options.ydata = {
4306                 copyNumber
4307                 };
4308             options.ylabelStr = {
4309                 {<span class="string">'Average Copy Number'</span>}
4310                 };
4311             [~, xAxesHandle] = PlotUtil.multiElementPlot(figHandle, {30}, [1 ch.sequenceLen], options);
4312             
4313             set(xAxesHandle{1}, <span class="string">'XTick'</span>, [1 ch.terCPosition ch.sequenceLen], <span class="string">'XTickLabel'</span>, {<span class="string">'oriC'</span>, <span class="string">'terC'</span>, <span class="string">'oriC'</span>});
4314             
4315             clear options;
4316             
4317             <span class="comment">%% organize figure data</span>
4318             figData = struct;
4319             figData.simOrder = simOrder;
4320             figData.copyNumber = copyNumber;
4321         <span class="keyword">end</span>
4322         
4323         <a name="_sub28" href="#_subfunctions" class="code">function figData = unsynchronizedPopulation(figHandle, simBatchDir, selectedSimulations)</a>
4324             import edu.stanford.covert.cell.sim.analysis.Population;
4325             import edu.stanford.covert.cell.sim.util.SimulationEnsemble;
4326             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
4327             import edu.stanford.covert.cell.sim.util.PlotUtil;
4328             import edu.stanford.covert.cell.sim.util.DiskLogger;
4329             
4330             <span class="keyword">if</span> ~exist(<span class="string">'simBatchDir'</span>, <span class="string">'var'</span>) || isempty(simBatchDir)
4331                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
4332             <span class="keyword">end</span>
4333             <span class="keyword">if</span> ~exist(<span class="string">'selectedSimulations'</span>, <span class="string">'var'</span>) || isempty(selectedSimulations)
4334                 selectedSimulations = SimulationDiskUtil.getCompleteSimulations(simBatchDir);
4335             <span class="keyword">end</span>
4336             
4337             <span class="comment">%% get constants</span>
4338             [~, ~, sim] = SimulationDiskUtil.getSimulation([simBatchDir filesep <span class="string">'1'</span>]);
4339             
4340             <span class="comment">%% get data</span>
4341             stateNames = {
4342                 <span class="string">'growth_rate'</span>       <span class="string">'Growth'</span>
4343                 <span class="string">'mass'</span>              <span class="string">'Mass'</span>
4344                 <span class="string">'ploidy'</span>            {<span class="string">'Chromosome'</span> <span class="string">'Copy Number'</span>}
4345                 <span class="string">'ftsZRing1st'</span>       <span class="string">'ftsZRing'</span>
4346                 <span class="string">'ftsZRing2st'</span>       <span class="string">'ftsZRing'</span>
4347                 <span class="string">'ftsZRing2bt'</span>       <span class="string">'ftsZRing'</span>
4348                 <span class="string">'ftsZRingRbt'</span>       <span class="string">'ftsZRing'</span>
4349                 <span class="string">'pinchedDiameter'</span>   <span class="string">'pinchedDiameter'</span>
4350                 };
4351             ensemble = SimulationEnsemble(simBatchDir, stateNames, [], selectedSimulations);
4352             simEndTimes = ensemble.stateData.simulationEndTimes;
4353             
4354             [time, values, cellCount, simOrder] = Population.desynchronizePopulation(sim, ensemble.stateData.time, ensemble.stateData.values, simEndTimes);
4355             time = time / 3600;
4356             <a href="#_sub2" class="code" title="subfunction figData = growth(figHandle, simBatchDir, selectedSimulations)">growth</a> = permute(values(ensemble.getPropertyIndices(<span class="string">'growth_rate'</span>), :, :, :), [4 3 1 2]) * 3600 * sim.state(<span class="string">'Mass'</span>).cellInitialDryWeight / (1 - sim.state(<span class="string">'Mass'</span>).fractionWetWeight);
4357             mass = permute(values(ensemble.getPropertyIndices(<span class="string">'mass'</span>), :, :, :), [4 3 1 2]) * 1e15;
4358             ploidy = permute(values(ensemble.getPropertyIndices(<span class="string">'ploidy'</span>), :, :, :), [4 3 1 2]);
4359             ftsZRing1st = permute(values(ensemble.getPropertyIndices(<span class="string">'ftsZRing1st'</span>), :, :, :), [4 3 1 2]);
4360             ftsZRing2st = permute(values(ensemble.getPropertyIndices(<span class="string">'ftsZRing2st'</span>), :, :, :), [4 3 1 2]);
4361             ftsZRing2bt = permute(values(ensemble.getPropertyIndices(<span class="string">'ftsZRing2bt'</span>), :, :, :), [4 3 1 2]);
4362             ftsZRingRbt = permute(values(ensemble.getPropertyIndices(<span class="string">'ftsZRingRbt'</span>), :, :, :), [4 3 1 2]);
4363             pinchedDiameter = permute(values(ensemble.getPropertyIndices(<span class="string">'pinchedDiameter'</span>), :, :, :), [4 3 1 2]);
4364             cellCount = permute(cellCount, [4 3 1 2]);
4365             
4366             <span class="comment">%% plot data</span>
4367             clf(figHandle);
4368             options = struct;
4369             options.colorOrder = PlotUtil.getRedGreenColorOrder(simOrder);
4370             options.xdata = time;
4371             options.ydata = {
4372                 <a href="#_sub2" class="code" title="subfunction figData = growth(figHandle, simBatchDir, selectedSimulations)">growth</a>
4373                 mass
4374                 ploidy
4375                 ftsZRing1st
4376                 ftsZRing2st
4377                 ftsZRing2bt
4378                 ftsZRingRbt
4379                 pinchedDiameter
4380                 cellCount
4381                 };
4382             options.ylabelStr = {
4383                 {<span class="string">'Growth'</span> <span class="string">'(fg h^{-1})'</span>}
4384                 {<span class="string">'Mass'</span> <span class="string">'(fg)'</span>}
4385                 {<span class="string">'Chr Copy'</span> <span class="string">'Number'</span>}
4386                 {<span class="string">'FtsZ'</span> <span class="string">'Ring'</span> <span class="string">'1-Strt'</span>}
4387                 {<span class="string">'FtsZ'</span> <span class="string">'Ring'</span> <span class="string">'2-Strt'</span>}
4388                 {<span class="string">'FtsZ'</span> <span class="string">'Ring'</span> <span class="string">'2-Bent'</span>}
4389                 {<span class="string">'FtsZ'</span> <span class="string">'Ring'</span> <span class="string">'R-Bent'</span>}
4390                 {<span class="string">'Pinched'</span> <span class="string">'Diameter'</span>}
4391                 {<span class="string">'Cell'</span> <span class="string">'Count'</span>}
4392                 };
4393             PlotUtil.multiElementPlot(figHandle, 10 * ones(size(options.ydata)), [0 time(end)], options);
4394             
4395             clear options;
4396             
4397             <span class="comment">%% organize figure data</span>
4398             figData = struct;
4399             figData.simOrder = simOrder;
4400             figData.time = time;
4401             figData.growth = <a href="#_sub2" class="code" title="subfunction figData = growth(figHandle, simBatchDir, selectedSimulations)">growth</a>;
4402             figData.mass = mass;
4403             figData.ploidy = ploidy;
4404             figData.cellCount = cellCount;
4405         <span class="keyword">end</span>
4406         
4407         <a name="_sub29" href="#_subfunctions" class="code">function figData = blockedDecayEvents(figHandle, simBatchDir, selectedSimulations)</a>
4408             import edu.stanford.covert.cell.sim.util.SimulationEnsemble;
4409             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
4410             import edu.stanford.covert.cell.sim.util.PlotUtil;
4411             
4412             <span class="keyword">if</span> ~exist(<span class="string">'simBatchDir'</span>, <span class="string">'var'</span>) || isempty(simBatchDir)
4413                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
4414             <span class="keyword">end</span>
4415             <span class="keyword">if</span> ~exist(<span class="string">'selectedSimulations'</span>, <span class="string">'var'</span>) || isempty(selectedSimulations)
4416                 selectedSimulations = SimulationDiskUtil.getCompleteSimulations(simBatchDir);
4417             <span class="keyword">end</span>
4418             
4419             <span class="comment">%% get data</span>
4420             nonDecays_DnaA = zeros(0, 1);
4421             nonDecays_FtsZ = zeros(0, 1);
4422             nonDecays_Replisome = zeros(0, 1);
4423             nonDecayCnts_DnaA = zeros(size(selectedSimulations));
4424             nonDecayCnts_FtsZ = zeros(size(selectedSimulations));
4425             nonDecayCnts_Replisome = zeros(size(selectedSimulations));
4426             <span class="keyword">for</span> i = 1:numel(selectedSimulations)
4427                 <a href="#_sub32" class="code" title="subfunction [figData, tabContent, tabColLabels] = warnings(figHandle, simBatchDir, selectedSimulations)">warnings</a> = SimulationDiskUtil.getSimulationWarnings(simBatchDir, i);
4428                 
4429                 idxs = find(strcmp({warnings.message}, <span class="string">'DnaA complex not decayed'</span>));
4430                 <span class="keyword">for</span> j = 1:numel(idxs)
4431                     nonDecays_DnaA = [nonDecays_DnaA; <a href="#_sub32" class="code" title="subfunction [figData, tabContent, tabColLabels] = warnings(figHandle, simBatchDir, selectedSimulations)">warnings</a>(idxs(j)).times]; <span class="comment">%#ok&lt;AGROW&gt;</span>
4432                     nonDecayCnts_DnaA(i) = nonDecayCnts_DnaA(i) + numel(<a href="#_sub32" class="code" title="subfunction [figData, tabContent, tabColLabels] = warnings(figHandle, simBatchDir, selectedSimulations)">warnings</a>(idxs(j)).times);
4433                 <span class="keyword">end</span>
4434                 
4435                 idxs = find(strcmp({warnings.message}, <span class="string">'FtsZ not decayed'</span>));
4436                 <span class="keyword">for</span> j = 1:numel(idxs)
4437                     nonDecays_FtsZ = [nonDecays_FtsZ; <a href="#_sub32" class="code" title="subfunction [figData, tabContent, tabColLabels] = warnings(figHandle, simBatchDir, selectedSimulations)">warnings</a>(idxs(j)).times]; <span class="comment">%#ok&lt;AGROW&gt;</span>
4438                     nonDecayCnts_FtsZ(i) = nonDecayCnts_FtsZ(i) + numel(<a href="#_sub32" class="code" title="subfunction [figData, tabContent, tabColLabels] = warnings(figHandle, simBatchDir, selectedSimulations)">warnings</a>(idxs(j)).times);
4439                 <span class="keyword">end</span>
4440                 
4441                 idxs = find(strcmp({warnings.message}, <span class="string">'DNA polymerase not decayed'</span>));
4442                 <span class="keyword">for</span> j = 1:numel(idxs)
4443                     nonDecays_Replisome = [nonDecays_Replisome; <a href="#_sub32" class="code" title="subfunction [figData, tabContent, tabColLabels] = warnings(figHandle, simBatchDir, selectedSimulations)">warnings</a>(idxs(j)).times]; <span class="comment">%#ok&lt;AGROW&gt;</span>
4444                     nonDecayCnts_Replisome(i) = nonDecayCnts_Replisome(i) + numel(<a href="#_sub32" class="code" title="subfunction [figData, tabContent, tabColLabels] = warnings(figHandle, simBatchDir, selectedSimulations)">warnings</a>(idxs(j)).times);
4445                 <span class="keyword">end</span>
4446             <span class="keyword">end</span>
4447             
4448             nonDecays_DnaA = nonDecays_DnaA / 3600;
4449             nonDecays_FtsZ = nonDecays_FtsZ / 3600;
4450             nonDecays_Replisome = nonDecays_Replisome / 3600;
4451             
4452             <span class="comment">%% plot data</span>
4453             clf(figHandle);
4454             options = struct();
4455             options.titleStr = {
4456                 <span class="string">'Non Decay Event Timing'</span>
4457                 <span class="string">'Non Decay Events Per Simulation'</span>
4458                 };
4459             [axesHandles, xAxesHandles] = PlotUtil.multiElementPlot(figHandle, repmat({10 * ones(3, 1)}, 2, 1), [0 max([nonDecays_DnaA; nonDecays_FtsZ; nonDecays_Replisome])], options);
4460             delete(xAxesHandles{2});
4461             
4462             <span class="comment">%Non Decay Event Timing</span>
4463             edges = linspace(0, max([nonDecays_DnaA; nonDecays_FtsZ; nonDecays_Replisome]), 50);
4464             
4465             axesHandle = axesHandles{1}(1);
4466             hist(axesHandle, nonDecays_DnaA, edges);
4467             set(findobj(axesHandle, <span class="string">'type'</span>, <span class="string">'patch'</span>), <span class="string">'EdgeColor'</span>, <span class="string">'b'</span>, <span class="string">'FaceColor'</span>, <span class="string">'b'</span>);
4468             ylabel(axesHandle, <span class="string">'DnaA'</span>);
4469             
4470             axesHandle = axesHandles{1}(2);
4471             hist(axesHandle, nonDecays_FtsZ, edges);
4472             set(findobj(axesHandle, <span class="string">'type'</span>, <span class="string">'patch'</span>), <span class="string">'EdgeColor'</span>, <span class="string">'b'</span>, <span class="string">'FaceColor'</span>, <span class="string">'b'</span>);
4473             ylabel(axesHandle, <span class="string">'FtsZ'</span>);
4474             
4475             axesHandle = axesHandles{1}(3);
4476             hist(axesHandle, nonDecays_Replisome, edges);
4477             set(findobj(axesHandle, <span class="string">'type'</span>, <span class="string">'patch'</span>), <span class="string">'EdgeColor'</span>, <span class="string">'b'</span>, <span class="string">'FaceColor'</span>, <span class="string">'b'</span>);
4478             ylabel(axesHandle, <span class="string">'Replisome'</span>);
4479             
4480             <span class="comment">%Non Decay Events Per Simulation</span>
4481             axesHandle = axesHandles{2}(1);
4482             hist(axesHandle, nonDecayCnts_DnaA, 10);
4483             set(findobj(axesHandle, <span class="string">'type'</span>, <span class="string">'patch'</span>), <span class="string">'EdgeColor'</span>, <span class="string">'b'</span>, <span class="string">'FaceColor'</span>, <span class="string">'b'</span>);
4484             xlabel(axesHandle, <span class="string">'Non-Decays'</span>);
4485             ylabel(axesHandle, <span class="string">'DnaA'</span>);
4486             
4487             axesHandle = axesHandles{2}(2);
4488             hist(axesHandle, nonDecayCnts_DnaA, 10);
4489             set(findobj(axesHandle, <span class="string">'type'</span>, <span class="string">'patch'</span>), <span class="string">'EdgeColor'</span>, <span class="string">'b'</span>, <span class="string">'FaceColor'</span>, <span class="string">'b'</span>);
4490             xlabel(axesHandle, <span class="string">'Non-Decays'</span>);
4491             ylabel(axesHandle, <span class="string">'FtsZ'</span>);
4492             
4493             axesHandle = axesHandles{2}(3);
4494             hist(axesHandle, nonDecayCnts_DnaA, 10);
4495             set(findobj(axesHandle, <span class="string">'type'</span>, <span class="string">'patch'</span>), <span class="string">'EdgeColor'</span>, <span class="string">'b'</span>, <span class="string">'FaceColor'</span>, <span class="string">'b'</span>);
4496             xlabel(axesHandle, <span class="string">'Non-Decays'</span>);
4497             ylabel(axesHandle, <span class="string">'Replisome'</span>);
4498             
4499             PlotUtil.alignYAxesLabels(axesHandles);
4500             PlotUtil.offsetYAxes(axesHandles, 0.04);
4501             
4502             <span class="comment">%% data</span>
4503             figData = struct;
4504             figData.nonDecays_DnaA = nonDecays_DnaA;
4505             figData.nonDecays_FtsZ = nonDecays_FtsZ;
4506             figData.nonDecays_Replisome = nonDecays_Replisome;
4507             figData.nonDecayCnts_DnaA = nonDecayCnts_DnaA;
4508             figData.nonDecayCnts_FtsZ = nonDecayCnts_FtsZ;
4509             figData.nonDecayCnts_Replisome = nonDecayCnts_Replisome;
4510         <span class="keyword">end</span>
4511         
4512         <a name="_sub30" href="#_subfunctions" class="code">function figData = secondaryReplicationInitiation(figHandle, simBatchDir, selectedSimulations)</a>
4513             import edu.stanford.covert.cell.sim.util.SimulationEnsemble;
4514             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
4515             import edu.stanford.covert.cell.sim.util.SummaryLogger;
4516             import edu.stanford.covert.cell.sim.util.PlotUtil;
4517             
4518             <span class="keyword">if</span> ~exist(<span class="string">'simBatchDir'</span>, <span class="string">'var'</span>) || isempty(simBatchDir)
4519                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
4520             <span class="keyword">end</span>
4521             <span class="keyword">if</span> ~exist(<span class="string">'selectedSimulations'</span>, <span class="string">'var'</span>) || isempty(selectedSimulations)
4522                 selectedSimulations = SimulationDiskUtil.getCompleteSimulations(simBatchDir);
4523             <span class="keyword">end</span>
4524             
4525             <span class="comment">%% get data</span>
4526             secondaryInitiationTimes = [];
4527             secondaryInitiationInitTimes = NaN(numel(selectedSimulations), 1);
4528             secondaryInitiationCnts = zeros(numel(selectedSimulations), 1);
4529             <span class="keyword">for</span> i = 1:numel(selectedSimulations)
4530                 <a href="#_sub32" class="code" title="subfunction [figData, tabContent, tabColLabels] = warnings(figHandle, simBatchDir, selectedSimulations)">warnings</a> = SimulationDiskUtil.getSimulationWarnings(simBatchDir, i);
4531                 
4532                 idxs = find(strcmp({warnings.message}, <span class="string">'Unable to represent multiple replication initiation events per cell cycle'</span>));
4533                 <span class="keyword">for</span> j = 1:numel(idxs)
4534                     secondaryInitiationTimes = [secondaryInitiationTimes; <a href="#_sub32" class="code" title="subfunction [figData, tabContent, tabColLabels] = warnings(figHandle, simBatchDir, selectedSimulations)">warnings</a>(idxs(j)).times]; <span class="comment">%#ok&lt;AGROW&gt;</span>
4535                     secondaryInitiationInitTimes(i) = min(secondaryInitiationInitTimes(i), <a href="#_sub32" class="code" title="subfunction [figData, tabContent, tabColLabels] = warnings(figHandle, simBatchDir, selectedSimulations)">warnings</a>(idxs).times(1));
4536                     secondaryInitiationCnts(i) = secondaryInitiationCnts(i) + numel(<a href="#_sub32" class="code" title="subfunction [figData, tabContent, tabColLabels] = warnings(figHandle, simBatchDir, selectedSimulations)">warnings</a>(idxs(j)).times);
4537                 <span class="keyword">end</span>
4538             <span class="keyword">end</span>
4539             secondaryInitiationTimes = secondaryInitiationTimes / 3600;
4540             secondaryInitiationInitTimes = secondaryInitiationInitTimes / 3600;
4541             
4542             simStats = SummaryLogger.getSimulationStatistics([SimulationDiskUtil.getBaseDir() filesep simBatchDir], selectedSimulations);
4543             repInitTimes = simStats(:, SummaryLogger.SIM_STATUS_INDEX_REP_INIT_TIME) / 3600;
4544             cellCycleTimes = simStats(:, SummaryLogger.SIM_STATUS_INDEX_CYTOKINESIS_TIME) / 3600;
4545             
4546             clear simStats;
4547             
4548             stateNames = {
4549                 <span class="string">'dnaA_box1'</span>
4550                 <span class="string">'dnaA_box2'</span>
4551                 <span class="string">'dnaA_box3'</span>
4552                 <span class="string">'dnaA_box4'</span>
4553                 <span class="string">'dnaA_box5'</span>
4554                 };
4555             ensemble = SimulationEnsemble(simBatchDir, stateNames, [], selectedSimulations);
4556             tmp = permute(sum(ensemble.stateData.values, 1), [4 3 1 2]);
4557             dnaAComplexSizes = tmp(sub2ind(size(tmp), (1:numel(selectedSimulations))', <span class="keyword">...</span>
4558                 floor(ensemble.stateData.simulationEndTimes / ensemble.stateData.downsampleStepSec)));
4559             
4560             clear ensemble stateNames;
4561             
4562             <span class="comment">%% plot data</span>
4563             clf(figHandle);
4564             options = struct();
4565             options.xlabelStr = {<span class="string">'Secondary Replication Initiation Time (h)'</span>, <span class="string">'Frequency'</span>};
4566             [axesHandles, xAxesHandles] = PlotUtil.multiElementPlot(figHandle, {[25 25 25] [1 1]}, [0 max(cellCycleTimes)], options);
4567             
4568             axesHandle = axesHandles{1}(1);
4569             edges = reshape(linspace(0, max(cellCycleTimes), 50), [], 1);
4570             cla(axesHandle);
4571             hist(axesHandle, secondaryInitiationTimes, edges);
4572             set(findobj(axesHandle, <span class="string">'type'</span>, <span class="string">'patch'</span>), <span class="string">'EdgeColor'</span>, <span class="string">'b'</span>, <span class="string">'FaceColor'</span>, <span class="string">'b'</span>);
4573             set(axesHandle, <span class="string">'YTick'</span>, [0 max(get(axesHandle, <span class="string">'YTick'</span>))]);
4574             ylabel(axesHandle, {<span class="string">'Attempt Frequency'</span>});
4575             
4576             axesHandle = axesHandles{1}(2);
4577             cla(axesHandle)
4578             edges = reshape(linspace(0, max(cellCycleTimes), 10), [], 1);
4579             tmp = [histc(repInitTimes, edges)  histc(secondaryInitiationInitTimes, edges)];
4580             h = bar(axesHandle, edges, tmp);
4581             legend(h, {<span class="string">'1st'</span>, <span class="string">'2nd'</span>}, <span class="string">'Location'</span>, <span class="string">'NorthWest'</span>);
4582             set(h, <span class="string">'EdgeColor'</span>, <span class="string">'none'</span>);
4583             set(h(1), <span class="string">'FaceColor'</span>, <span class="string">'r'</span>);
4584             set(h(2), <span class="string">'FaceColor'</span>, <span class="string">'b'</span>);
4585             xlim(axesHandle, [0 max(cellCycleTimes)]);
4586             set(axesHandle, <span class="string">'YTick'</span>, [0 max(get(axesHandle, <span class="string">'YTick'</span>))]);
4587             ylabel(axesHandle, {<span class="string">'Frequency'</span>});
4588             
4589             axesHandle = axesHandles{1}(3);
4590             hold(axesHandle, <span class="string">'on'</span>);
4591             plot(axesHandle, secondaryInitiationInitTimes, repInitTimes, <span class="string">'.'</span>);
4592             line([0 max(cellCycleTimes)], [0 max(cellCycleTimes)], <span class="keyword">...</span>
4593                 <span class="string">'Color'</span>, <span class="string">'r'</span>, <span class="string">'LineStyle'</span>, <span class="string">':'</span>, <span class="string">'Parent'</span>, axesHandle);
4594             xlim(axesHandle, [0 max(cellCycleTimes)]);
4595             ylim(axesHandle, [0 max(cellCycleTimes)]);
4596             set(axesHandle, <span class="string">'YTick'</span>, get(xAxesHandles{1}, <span class="string">'XTick'</span>));
4597             ylabel(axesHandle, {<span class="string">'Replication Initiation Time'</span>});
4598             
4599             axesHandle = axesHandles{2}(1);
4600             cla(axesHandle);
4601             edges = linspace(min(secondaryInitiationCnts), max(max(secondaryInitiationCnts), 9), 10);
4602             freq1 = histc(secondaryInitiationCnts, edges);
4603             barh(axesHandle, edges, freq1, <span class="string">'EdgeColor'</span>, <span class="string">'b'</span>, <span class="string">'FaceColor'</span>, <span class="string">'b'</span>);
4604             ylabel(axesHandle, <span class="string">'Secondary Initiation Attempts'</span>);
4605             ylim(axesHandle, edges([1 end]) + [-0.5 0.5]);
4606             set(axesHandle, <span class="string">'YTick'</span>, [0 max(get(axesHandle, <span class="string">'YTick'</span>))]);
4607             
4608             axesHandle = axesHandles{2}(2);
4609             cla(axesHandle);
4610             edges = (0:29)';
4611             freq2 = histc(dnaAComplexSizes, edges);
4612             barh(axesHandle, edges, freq2, <span class="string">'EdgeColor'</span>, <span class="string">'b'</span>, <span class="string">'FaceColor'</span>, <span class="string">'b'</span>);
4613             ylabel(axesHandle, <span class="string">'DnaA Complex Size'</span>);
4614             ylim(axesHandle, edges([1 end]) + [-0.5; 0.5]);
4615             set(axesHandle, <span class="string">'YTick'</span>, 0:4:28);
4616             set(axesHandle, <span class="string">'YTickLabel'</span>, 0:7);
4617             
4618             xlim(xAxesHandles{2}, [min([freq1; freq2]) max([freq1; freq2])]);
4619             xlim(axesHandles{2}(1), [min([freq1; freq2]) max([freq1; freq2])]);
4620             xlim(axesHandles{2}(2), [min([freq1; freq2]) max([freq1; freq2])]);
4621             set(xAxesHandles{2}, <span class="string">'XTickMode'</span>, <span class="string">'auto'</span>);
4622             
4623             PlotUtil.offsetYAxes(axesHandles, 0.03);
4624             
4625             <span class="comment">%% data</span>
4626             figData = struct;
4627             figData.repInitTimes = repInitTimes;
4628             figData.secondaryInitiationTimes = secondaryInitiationTimes;
4629             figData.secondaryInitiationInitTimes = secondaryInitiationInitTimes;
4630             figData.secondaryInitiationCnts = secondaryInitiationCnts;
4631             figData.dnaAComplexSizes = dnaAComplexSizes;
4632         <span class="keyword">end</span>
4633         
4634         <a name="_sub31" href="#_subfunctions" class="code">function [figData, tabContent, tabColLabels] = warnings(figHandle, simBatchDir, selectedSimulations)</a>
4635             import edu.stanford.covert.cell.sim.util.SimulationEnsemble;
4636             import edu.stanford.covert.cell.sim.util.SimulationDiskUtil;
4637             import edu.stanford.covert.cell.sim.util.SummaryLogger;
4638             import edu.stanford.covert.cell.sim.util.PlotUtil;
4639             
4640             <span class="keyword">if</span> ~exist(<span class="string">'simBatchDir'</span>, <span class="string">'var'</span>) || isempty(simBatchDir)
4641                 simBatchDir = SimulationDiskUtil.getLatestSimulationGroup();
4642             <span class="keyword">end</span>
4643             <span class="keyword">if</span> ~exist(<span class="string">'selectedSimulations'</span>, <span class="string">'var'</span>) || isempty(selectedSimulations)
4644                 selectedSimulations = SimulationDiskUtil.getCompleteSimulations(simBatchDir);
4645             <span class="keyword">end</span>
4646             
4647             <span class="comment">%% get data</span>
4648             <a href="#_sub32" class="code" title="subfunction [figData, tabContent, tabColLabels] = warnings(figHandle, simBatchDir, selectedSimulations)">warnings</a> = [];
4649             freq = zeros(10, 1);
4650             <span class="keyword">for</span> i = 1:numel(selectedSimulations)
4651                 tmpWarnings = SimulationDiskUtil.getSimulationWarnings(simBatchDir, selectedSimulations(i));
4652                 
4653                 repeatedWarning = 0;
4654                 <span class="keyword">for</span> k = 1:numel(tmpWarnings)
4655                     <span class="keyword">if</span> strcmp(tmpWarnings(k).message, <span class="string">'Unable to represent multiple replication initiation events per cell cycle'</span>)
4656                         freq(1) = freq(1) + numel(tmpWarnings(k).times);
4657                     <span class="keyword">elseif</span> strcmp(tmpWarnings(k).message, <span class="string">'DnaA complex not decayed'</span>)
4658                         freq(2) = freq(2) + numel(tmpWarnings(k).times);
4659                     <span class="keyword">elseif</span> strcmp(tmpWarnings(k).message, <span class="string">'FtsZ not decayed'</span>)
4660                         freq(3) = freq(3) + numel(tmpWarnings(k).times);
4661                     <span class="keyword">elseif</span> strcmp(tmpWarnings(k).message, <span class="string">'DNA polymerase not decayed'</span>)
4662                         freq(4) = freq(4) + numel(tmpWarnings(k).times);
4663                     <span class="keyword">elseif</span> strcmp(tmpWarnings(k).message, <span class="string">'Cell has divided. Cell shape undefined'</span>)
4664                         freq(5) = freq(5) + numel(tmpWarnings(k).times);
4665                     <span class="keyword">elseif</span> length(tmpWarnings(k).message) &gt; 24 &amp;&amp; strcmp(tmpWarnings(k).message(1:24), <span class="string">'Linear programming error'</span>)
4666                         freq(6) = freq(6) + numel(tmpWarnings(k).times);
4667                     <span class="keyword">elseif</span> length(tmpWarnings(k).message) &gt; 50 &amp;&amp; strcmp(tmpWarnings(k).message(1:50), <span class="string">'Simulation state side effects should preserve mass'</span>)
4668                         freq(7) = freq(7) + numel(tmpWarnings(k).times);
4669                     <span class="keyword">elseif</span> length(tmpWarnings(k).message) &gt; 35 &amp;&amp; strcmp(tmpWarnings(k).message(1:35), <span class="string">'Summary figure could not be saved: '</span>)
4670                         freq(8) = freq(8) + numel(tmpWarnings(k).times);
4671                     <span class="keyword">elseif</span> regexp(tmpWarnings(k).message, <span class="string">'metabolites have negative counts'</span>)
4672                         freq(9) = freq(9) + numel(tmpWarnings(k).times);
4673                     <span class="keyword">else</span>
4674                         freq(10) = freq(10) + numel(tmpWarnings(k).times);
4675                     <span class="keyword">end</span>
4676                     
4677                     tmpWarnings(k).simulations = repmat(i, size(tmpWarnings(k).times));
4678                     
4679                     <span class="keyword">for</span> j = 1:numel(<a href="#_sub32" class="code" title="subfunction [figData, tabContent, tabColLabels] = warnings(figHandle, simBatchDir, selectedSimulations)">warnings</a>)
4680                         <span class="keyword">if</span> isequal(<a href="#_sub32" class="code" title="subfunction [figData, tabContent, tabColLabels] = warnings(figHandle, simBatchDir, selectedSimulations)">warnings</a>(j).message, tmpWarnings(k).message) &amp;&amp; <span class="keyword">...</span>
4681                                 isequal(<a href="#_sub32" class="code" title="subfunction [figData, tabContent, tabColLabels] = warnings(figHandle, simBatchDir, selectedSimulations)">warnings</a>(j).files, tmpWarnings(k).files) &amp;&amp; <span class="keyword">...</span>
4682                                 isequal(<a href="#_sub32" class="code" title="subfunction [figData, tabContent, tabColLabels] = warnings(figHandle, simBatchDir, selectedSimulations)">warnings</a>(j).lineNumbers, tmpWarnings(k).lineNumbers)
4683                             repeatedWarning = j;
4684                             <span class="keyword">break</span>;
4685                         <span class="keyword">end</span>
4686                     <span class="keyword">end</span>
4687                     
4688                     <span class="keyword">if</span> isempty(<a href="#_sub32" class="code" title="subfunction [figData, tabContent, tabColLabels] = warnings(figHandle, simBatchDir, selectedSimulations)">warnings</a>)
4689                         <a href="#_sub32" class="code" title="subfunction [figData, tabContent, tabColLabels] = warnings(figHandle, simBatchDir, selectedSimulations)">warnings</a> = tmpWarnings(k);
4690                     <span class="keyword">elseif</span> repeatedWarning
4691                         <a href="#_sub32" class="code" title="subfunction [figData, tabContent, tabColLabels] = warnings(figHandle, simBatchDir, selectedSimulations)">warnings</a>(repeatedWarning).times = <span class="keyword">...</span>
4692                             [<a href="#_sub32" class="code" title="subfunction [figData, tabContent, tabColLabels] = warnings(figHandle, simBatchDir, selectedSimulations)">warnings</a>(repeatedWarning).times; tmpWarnings(k).times]; <span class="comment">%#ok&lt;AGROW&gt;</span>
4693                         <a href="#_sub32" class="code" title="subfunction [figData, tabContent, tabColLabels] = warnings(figHandle, simBatchDir, selectedSimulations)">warnings</a>(repeatedWarning).simulations = <span class="keyword">...</span>
4694                             [<a href="#_sub32" class="code" title="subfunction [figData, tabContent, tabColLabels] = warnings(figHandle, simBatchDir, selectedSimulations)">warnings</a>(repeatedWarning).simulations; tmpWarnings(k).simulations]; <span class="comment">%#ok&lt;AGROW&gt;</span>
4695                     <span class="keyword">else</span>
4696                         <a href="#_sub32" class="code" title="subfunction [figData, tabContent, tabColLabels] = warnings(figHandle, simBatchDir, selectedSimulations)">warnings</a> = [<a href="#_sub32" class="code" title="subfunction [figData, tabContent, tabColLabels] = warnings(figHandle, simBatchDir, selectedSimulations)">warnings</a>; tmpWarnings(k)]; <span class="comment">%#ok&lt;AGROW&gt;</span>
4697                     <span class="keyword">end</span>
4698                 <span class="keyword">end</span>
4699             <span class="keyword">end</span>
4700             
4701             <span class="comment">%% plot data</span>
4702             clf(figHandle);
4703             axesHandle = subplot(1, 1, 1, <span class="string">'Parent'</span>, figHandle);
4704             barh(axesHandle, freq)
4705             set(axesHandle, <span class="string">'YTick'</span>, 1:numel(freq),  <span class="string">'YTickLabel'</span>, {<span class="string">'2nd Rep Init'</span>, <span class="string">'DnaA Decay'</span>, <span class="string">'FtsZ Decay'</span>, <span class="string">'Replisome Decay'</span>, <span class="string">'Shape Undef'</span>, <span class="string">'Lin Prog'</span>, <span class="string">'Neg Mets'</span>, <span class="string">'Sum Fig Save'</span>, <span class="string">'Mass Bal'</span>, <span class="string">'Other'</span>});
4706             ylabel(axesHandle, <span class="string">'Warning'</span>);
4707             xlabel(axesHandle, <span class="string">'Frequency'</span>);
4708             set(axesHandle, <span class="string">'XScale'</span>, <span class="string">'log'</span>);
4709             ylim(axesHandle, [0.5 numel(freq) + 0.5]);
4710             
4711             <span class="comment">%% table</span>
4712             <a href="#_sub32" class="code" title="subfunction [figData, tabContent, tabColLabels] = warnings(figHandle, simBatchDir, selectedSimulations)">warnings</a> = <a href="#_sub32" class="code" title="subfunction [figData, tabContent, tabColLabels] = warnings(figHandle, simBatchDir, selectedSimulations)">warnings</a>(~ismember({warnings.message}, {
4713                 <span class="string">'Unable to represent multiple replication initiation events per cell cycle'</span>
4714                 <span class="string">'DnaA complex not decayed'</span>
4715                 <span class="string">'FtsZ not decayed'</span>
4716                 <span class="string">'DNA polymerase not decayed'</span>
4717                 <span class="string">'Cell has divided. Cell shape undefined'</span>
4718                 }));
4719             tfs = true(size(<a href="#_sub32" class="code" title="subfunction [figData, tabContent, tabColLabels] = warnings(figHandle, simBatchDir, selectedSimulations)">warnings</a>));
4720             <span class="keyword">for</span> i = 1:numel(<a href="#_sub32" class="code" title="subfunction [figData, tabContent, tabColLabels] = warnings(figHandle, simBatchDir, selectedSimulations)">warnings</a>)
4721                 tfs(i) = length(<a href="#_sub32" class="code" title="subfunction [figData, tabContent, tabColLabels] = warnings(figHandle, simBatchDir, selectedSimulations)">warnings</a>(i).message) &lt; 35 || ~isequal(<a href="#_sub32" class="code" title="subfunction [figData, tabContent, tabColLabels] = warnings(figHandle, simBatchDir, selectedSimulations)">warnings</a>(i).message(1:35), <span class="string">'Summary figure could not be saved: '</span>);
4722             <span class="keyword">end</span>
4723             <a href="#_sub32" class="code" title="subfunction [figData, tabContent, tabColLabels] = warnings(figHandle, simBatchDir, selectedSimulations)">warnings</a> = <a href="#_sub32" class="code" title="subfunction [figData, tabContent, tabColLabels] = warnings(figHandle, simBatchDir, selectedSimulations)">warnings</a>(tfs);
4724             tabColLabels = {<span class="string">'Warning'</span>, <span class="string">'Stack Trace'</span>, <span class="string">'Simulation.Times'</span>};
4725             tabContent = cell(numel(<a href="#_sub32" class="code" title="subfunction [figData, tabContent, tabColLabels] = warnings(figHandle, simBatchDir, selectedSimulations)">warnings</a>), 3);
4726             <span class="keyword">for</span> i = 1:numel(<a href="#_sub32" class="code" title="subfunction [figData, tabContent, tabColLabels] = warnings(figHandle, simBatchDir, selectedSimulations)">warnings</a>)
4727                 tabContent{i, 1} = <a href="#_sub32" class="code" title="subfunction [figData, tabContent, tabColLabels] = warnings(figHandle, simBatchDir, selectedSimulations)">warnings</a>(i).message;
4728                 
4729                 <span class="keyword">for</span> j = 1:numel(<a href="#_sub32" class="code" title="subfunction [figData, tabContent, tabColLabels] = warnings(figHandle, simBatchDir, selectedSimulations)">warnings</a>(i).files)
4730                     tabContent{i, 2} = [tabContent{i, 2} <span class="keyword">...</span>
4731                         sprintf(<span class="string">'%s at %d\n'</span>, <a href="#_sub32" class="code" title="subfunction [figData, tabContent, tabColLabels] = warnings(figHandle, simBatchDir, selectedSimulations)">warnings</a>(i).files{j}, <a href="#_sub32" class="code" title="subfunction [figData, tabContent, tabColLabels] = warnings(figHandle, simBatchDir, selectedSimulations)">warnings</a>(i).lineNumbers(j))];
4732                 <span class="keyword">end</span>
4733                 tabContent{i, 2} = tabContent{i, 2}(1:end-1);
4734                 
4735                 <span class="keyword">for</span> j = 1:numel(<a href="#_sub32" class="code" title="subfunction [figData, tabContent, tabColLabels] = warnings(figHandle, simBatchDir, selectedSimulations)">warnings</a>(i).simulations)
4736                     tabContent{i, 3} = [tabContent{i, 3} <span class="keyword">...</span>
4737                         sprintf(<span class="string">'%d.%d\n'</span>, <a href="#_sub32" class="code" title="subfunction [figData, tabContent, tabColLabels] = warnings(figHandle, simBatchDir, selectedSimulations)">warnings</a>(i).simulations(j), <a href="#_sub32" class="code" title="subfunction [figData, tabContent, tabColLabels] = warnings(figHandle, simBatchDir, selectedSimulations)">warnings</a>(i).times(j))];
4738                 <span class="keyword">end</span>
4739                 tabContent{i, 3} = tabContent{i, 3}(1:end-1);
4740             <span class="keyword">end</span>
4741             
4742             <span class="comment">%% data</span>
4743             figData = struct;
4744             figData.freq = freq;
4745         <span class="keyword">end</span>
4746     <span class="keyword">end</span>
4747     
4748     <span class="comment">%helper methods</span>
4749     methods (Static = true)
4750         <a name="_sub32" href="#_subfunctions" class="code">function [unsyncTime, unsyncData, cellCount, simOrder] = desynchronizePopulation(sim, syncTime, syncData, simEndTimes)</a>
4751             finTimeIdx = min(simEndTimes);
4752             finTime = syncTime(finTimeIdx);
4753             
4754             cellCycleLength = sim.state(<span class="string">'Time'</span>).cellCycleLength;
4755             randStream = edu.stanford.covert.util.RandStream(<span class="string">'mcg16807'</span>);
4756             
4757             unsyncData = NaN(size(syncData, 1), size(syncData, 2), min(simEndTimes), 2 * size(syncData, 4));
4758             simStartTimes = zeros(size(syncData, 4), 1);
4759             <span class="keyword">for</span> i = 1:ceil(size(syncData, 4) * (1 - exp(-log(2) * finTimeIdx / cellCycleLength)))
4760                 simTimeIdx = ceil(randStream.rand * finTimeIdx);
4761                 unsyncData(:, :, 1:simTimeIdx, i) = syncData(:, :, simEndTimes(i)-simTimeIdx+1:simEndTimes(i), i);
4762                 unsyncData(:, :, simTimeIdx+1:<span class="keyword">end</span>, i) = syncData(:, :, 1:finTimeIdx - simTimeIdx, i);
4763             <span class="keyword">end</span>
4764             <span class="keyword">for</span> j = i+1:size(syncData, 4)
4765                 simStartTimes(j) = randStream.stochasticRound(log(1 + randStream.rand * (exp(log(2) * finTime / cellCycleLength) - 1)) * cellCycleLength / log(2));
4766                 simTimeIdx = find(syncTime == simStartTimes(j));
4767                 <span class="keyword">if</span> randStream.rand &gt; 0.5
4768                     unsyncData(:, :, simTimeIdx+1:<span class="keyword">end</span>, j) = syncData(:, :, 1:finTimeIdx - simTimeIdx, j);
4769                 <span class="keyword">else</span>
4770                     unsyncData(:, :, simTimeIdx+1:<span class="keyword">end</span>, j) = syncData(:, :, simEndTimes(j) - (finTimeIdx - simTimeIdx)+1:simEndTimes(j), j);
4771                 <span class="keyword">end</span>
4772             <span class="keyword">end</span>
4773             [~, simOrder] = sort(simStartTimes);
4774             
4775             cellCount = sum(~isnan(unsyncData(1, :, :, :)), 4);
4776             
4777             unsyncTime = syncTime(:, 1:finTimeIdx);
4778         <span class="keyword">end</span>
4779     <span class="keyword">end</span>
4780 <span class="keyword">end</span></pre></div>
<hr><address>Generated on Thu 19-Jul-2012 18:32:13 by <strong><a href="http://www.artefact.tk/software/matlab/m2html/" title="Matlab Documentation in HTML">m2html</a></strong> &copy; 2005</address>
</body>
</html>