<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN"
                "http://www.w3.org/TR/REC-html40/loose.dtd">
<html>
<head>
  <title>Description of DNADamage</title>
  <meta name="keywords" content="DNADamage">
  <meta name="description" content="DNA Damage">
  <meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1">
  <meta name="generator" content="m2html v1.5 &copy; 2003-2005 Guillaume Flandin">
  <meta name="robots" content="index, follow">
  <link type="text/css" rel="stylesheet" href="../../../../../../../../m2html.css">
</head>
<body>
<a name="_top"></a>
<div><a href="../../../../../../../../index.html">Home</a> &gt;  <a href="../../../../../../../index.html">simulation</a> &gt; <a href="#">src</a> &gt; <a href="#">+edu</a> &gt; <a href="#">+stanford</a> &gt; <a href="#">+covert</a> &gt; <a href="#">+cell</a> &gt; <a href="../index.html">+sim</a> &gt; <a href="index.html">+process</a> &gt; DNADamage.m</div>

<!--<table width="100%"><tr><td align="left"><a href="../../../../../../../../index.html"><img alt="<" border="0" src="../../../../../../../../left.png">&nbsp;Master index</a></td>
<td align="right"><a href="index.html">Index for simulation/src/+edu/+stanford/+covert/+cell/+sim/+process&nbsp;<img alt=">" border="0" src="../../../../../../../../right.png"></a></td></tr></table>-->

<h1>DNADamage
</h1>

<h2><a name="_name"></a>PURPOSE <a href="#_top"><img alt="^" border="0" src="../../../../../../../../up.png"></a></h2>
<div class="box"><strong>DNA Damage</strong></div>

<h2><a name="_synopsis"></a>SYNOPSIS <a href="#_top"><img alt="^" border="0" src="../../../../../../../../up.png"></a></h2>
<div class="box"><strong>This is a script file. </strong></div>

<h2><a name="_description"></a>DESCRIPTION <a href="#_top"><img alt="^" border="0" src="../../../../../../../../up.png"></a></h2>
<div class="fragment"><pre class="comment">DNA Damage

 @wholeCellModelID Process_DNADamage
 @name             DNA Damage
 @description
   Biology
   ======================
   Simulates a spectrum of DNA lesions repaired by base excision repair,
   nucleotide excision repair, and homologous recombination repair:
   - DNA strand breaks: caused by
     - hydrolysis and beta-elimination of abasic sites
     - ionizing radiation (alpha, beta particles, protons, electrons,
       gamma rays)
   - Spontaneous base loss
   - Base damage: caused by
     - Spontaneous deamination
     - oxidation -- oxygen radicals generated by metabolism, ionizing radiation, chemicals
     - alkylation -- carbon radicals generated by metabolism, ionizing radiation, chemicals
     - amination -- nitrogen radicals generated by metabolism, ionizing radiation, chemicals
   - Intrastrand cross links: caused by
     - UV-B induced cyclization

   The simulation does not consider the following types of DNA damage
   - interstrand cross links: caused by bifunctional akylating agents
   - psoralens
   - DNA-protein cross links
   - Sugar-phosphate damage other than strand breaks

   Additionally, the simulation does not consider mutiple damages per
   base or sugar-phosphate.

   Categorizes strand breaks in DNA segments (216 bases) according to
   scheme described in the track structure model literature [PUB_0486]:
   - NB     -- no strand break
   - SSB    -- single strand break
   - SSB+   -- multiple single strand breaks on segment all on same strand
   - 2SSB   -- single strand breaks on both strand separated by at least 10 bases on DNA segment
   - DSB    -- two single strand breaks on opposite strands within 10 bases
   - DSB+   -- double strand break with additional strand break on DNA segment
   - DSB++  -- two single strand breaks on same DNA segment

   Effect of DNA damage on physiology
   ++++++++++++++++++++++
   The exact effect of DNA damage on physiology is not well characterized.
   Certain kinds of damage are believed to arrest transcription
   [PUB_0780]. Tornaletti and Hanawalt provide the most extensive review
   of the effect of specific damages on transcription [PUB_0781, PUB_0782,
   PUB_0783]. For simplicity we we model all damages (except m6AD involved
   in R/M systems) as barriers to protein binding and function. This is
   implemented with the chromosome state class.

   Knowledge Base
   ======================
   As of 8/12/2010 the M. genitalium knowledge contained the stochiometry
   and kinetics of 128 DNA damage reactions involving 15 damaging agents
   and 9 spontaneous reactions. These reactions were compiled from
   several sources. The rates of the most frequent types of damage are
   listed below.

      Type                                            Reactions
      ============================================    =========
      base alkylation                                   7
      base amination                                    3
      base ethylation                                  39
      base glucosyl transfer                            2
      base methylation                                 42
      base reduction                                    1
      photooxidation                                    2
      radiation (gamma-ray) induced base oxidation     19
      spontaneous base deamination                      4
      spontaneous base loss                             4
      strand break                                      2
      UV-B photodimerization                           12

      Damaging agent                                  Reactions
      ============================================    =========
      butylene glycol                                   1
      CH3                                               4
      chloroethylene oxide                              5
      DES                                              13
      EMS                                              13
      ENU                                              13
      Gamma radiation                                   1
      hydrogen radical                                  1
      hydroxyl radical                                 19
      MMS                                               9
      MMSO4                                             7
      MNNG                                             10
      MNU                                              12
      NH3                                               1
      Spontaneous                                       9
      UVB radiation                                    14

      Damage                          Spec Rate             Rate (1/s)   No (1/Cycle)
      ============================    ===================   ==========   ============
      Spontaneous base deamination    1.1e-12 1/nt/s        1.4e-06      0.1
      Spontaneous base loss           7.2e-11 1/nt/s        8.4e-05      7.8
      UV-B Dimer                      1.2e-07 1/nt/(J/m^2)  6.0e-01      5.6e4
      \gamma-ray base oxidation       5.2e-07 1/nt/Gy       1.7e-11      1.6e-6
      \gamma-ray strand break         1.7e-07 1/nt/Gy       5.6e-12      5.2e-7
      base alkylation                 ?                     ?            ?
      base amination                  ?                     ?            ?
      base ethylation                 ?                     ?            ?
      base glucosyl transfer          ?                     ?            ?
      base methylation                ?                     ?            ?
      base reduction                  ?                     ?            ?
      photooxidation                  ?                     ?            ?

   Representation
   ======================
   7 properties are used to keep track of the damage state of DNA. The first
   dimension of each has length equal to that of the chromosome.
   - gapSites                  true/false    sugar gaps
   - abasicSites               true/false    abasic sites
   - damagedSugarPhosphates    enumeration   damaged sugar phosphates
   - damagedBases              enumeration   damaged bases
   - intrastrandCrossLinks     true/false    intrastand cross link
   - strandBreaks              true/false    strand break (missing sugar-phosphate bond)
   - hollidayJunctions         true/false    holliday junction

   substrates represents the amounts of free metabolites. initializeConstants
   synchronizes the substrates of this process with that of the DNA repair
   process to make it easy for these processes to communicate. That is
   initializeConstants sets the substrates of both processes equal to the union
   of the two, are orders the substrates the same in both processes. Thus
   substrate indices defined in process are valid on properties received from
   the other process, and it is unecessary to remap properties in
   copyFromState.

   stimuli represents the amounts of DNA damaging agents including several
   kinds of radiation.

   This process uses three properties rather than reactionStoichiometryMatrix
   to represent DNA damaging reactions and the DNA, damaging agents, and free
   metabolites that participate in them. These three matrices are a
   factorization of the super class reactionStoichiometryMatrix property.
   - reactionSmallMoleculeStoichiometryMatrix: the stochiometry of free
     metabolites in damage reactions
   - reactionDNAStoichiometryMatrix: the stoichiometry of DNA bases in
     damage reactions
   - reactionRadiationStoichiometryMatrix: the stoichiometry of radiation
     stimuli in damage reactions

   This process uses reactionVulnerableMotifs, rather the super class
   reactionModificationMatrix property to represent the affects of the DNA
   damage reaction on macromolecules. reactionVulnerableMotifs represents the DNA
   sequence(s) vulnerable to each damage reaction.

   Initialization
   ======================
   The DNA is initialized to the undamaged state, except that MunI
   restriction/modification sites are initialized to the methylated state (see
   DNA Repair process).

   Simulation
   ======================
   For each reaction
     1. Compute rate that individual vulnerable DNA regions succombs to damage
     2. Randomly select among vulnerable DNA regions at rate computed in step
        (1)
     3. Mark DNA selected in step (2) as damaged by updating gapSites,
        abasicSites, damagedSugarPhosphates, damagedBases,
        intrastrandCrossLinks, strandBreaks, and hollidayJunctions
     4. Decrement/increment free metabolites reaction reactants and products
        using reactionSmallMoleculeStoichiometryMatrix
   end

   References
   ======================
   1. Errol C Friedberg, Graham C Walker, and Wolfram Siede (1995). DNA
      Repair and Mutagenesis. ASM Press: Washington, DC. [PUB_0461]
   2. Nikjoo H, O'Neill P, Goodhead DT, and Terrissol M. Computational
      modelling of low-energy electron-induced DNA damage by  early physical
      and chemical events (1997). Int J Radiat Biol. 71(5): 467-83.
      [PUB_0486]
   3. Cline SD, Hanawalt PC (2003). Who's on first in the cellular
      response to DNA damage? Nat Rev Mol Cell Biol. 4(5):361-72.
      [PUB_0780]
   4. Tornaletti S (2005). Transcription arrest at DNA damage sites. Mutat
      Res. 577(1-2):131-45. [PUB_0781]
   5. Tornaletti S    (2009). Transcription-coupled DNA repair: directing
      your effort where it's most needed. Cell Mol Life Sci.
      66(6):1010-20. [PUB_0782]
   6. Tornaletti S, Hanawalt PC (1999). Effect of DNA lesions on
      transcription elongation. Biochimie. 81 (1-2):139-46. [PUB_0783] 

 Author: Jonathan Karr, jkarr@stanford.edu
 Affilitation: Covert Lab, Department of Bioengineering, Stanford University
 Last updated: 4/25/2010</pre></div>

<!-- crossreference -->
<h2><a name="_cross"></a>CROSS-REFERENCE INFORMATION <a href="#_top"><img alt="^" border="0" src="../../../../../../../../up.png"></a></h2>
This function calls:
<ul style="list-style-image:url(../../../../../../../../matlabicon.gif)">
<li><a href="DNADamage.html" class="code" title="">DNADamage</a>	DNA Damage</li></ul>
This function is called by:
<ul style="list-style-image:url(../../../../../../../../matlabicon.gif)">
<li><a href="DNADamage.html" class="code" title="">DNADamage</a>	DNA Damage</li></ul>
<!-- crossreference -->

<h2><a name="_subfunctions"></a>SUBFUNCTIONS <a href="#_top"><img alt="^" border="0" src="../../../../../../../../up.png"></a></h2>
<ul style="list-style-image:url(../../../../../../../../matlabicon.gif)">
<li><a href="#_sub1" class="code">function this = DNADamage(wholeCellModelID, name)</a></li><li><a href="#_sub2" class="code">function storeObjectReferences(this, simulation)</a></li><li><a href="#_sub3" class="code">function initializeConstants(this, knowledgeBase, simulation, varargin)</a></li><li><a href="#_sub4" class="code">function [bmProd, byProd, minEnzExp, maxEnzExp] = calcResourceRequirements_LifeCycle(this, constants, ~)</a></li><li><a href="#_sub5" class="code">function initializeState(~)</a></li><li><a href="#_sub6" class="code">function result = calcResourceRequirements_Current(this)</a></li><li><a href="#_sub7" class="code">function rates = calcExpectedReactionRates(this)</a></li><li><a href="#_sub8" class="code">function nVulnerableSites = calcNumberVulnerableSites(this)</a></li><li><a href="#_sub9" class="code">function evolveState(this)</a></li></ul>

<h2><a name="_source"></a>SOURCE CODE <a href="#_top"><img alt="^" border="0" src="../../../../../../../../up.png"></a></h2>
<div class="fragment"><pre>0001 <span class="comment">%DNA Damage</span>
0002 <span class="comment">%</span>
0003 <span class="comment">% @wholeCellModelID Process_DNADamage</span>
0004 <span class="comment">% @name             DNA Damage</span>
0005 <span class="comment">% @description</span>
0006 <span class="comment">%   Biology</span>
0007 <span class="comment">%   ======================</span>
0008 <span class="comment">%   Simulates a spectrum of DNA lesions repaired by base excision repair,</span>
0009 <span class="comment">%   nucleotide excision repair, and homologous recombination repair:</span>
0010 <span class="comment">%   - DNA strand breaks: caused by</span>
0011 <span class="comment">%     - hydrolysis and beta-elimination of abasic sites</span>
0012 <span class="comment">%     - ionizing radiation (alpha, beta particles, protons, electrons,</span>
0013 <span class="comment">%       gamma rays)</span>
0014 <span class="comment">%   - Spontaneous base loss</span>
0015 <span class="comment">%   - Base damage: caused by</span>
0016 <span class="comment">%     - Spontaneous deamination</span>
0017 <span class="comment">%     - oxidation -- oxygen radicals generated by metabolism, ionizing radiation, chemicals</span>
0018 <span class="comment">%     - alkylation -- carbon radicals generated by metabolism, ionizing radiation, chemicals</span>
0019 <span class="comment">%     - amination -- nitrogen radicals generated by metabolism, ionizing radiation, chemicals</span>
0020 <span class="comment">%   - Intrastrand cross links: caused by</span>
0021 <span class="comment">%     - UV-B induced cyclization</span>
0022 <span class="comment">%</span>
0023 <span class="comment">%   The simulation does not consider the following types of DNA damage</span>
0024 <span class="comment">%   - interstrand cross links: caused by bifunctional akylating agents</span>
0025 <span class="comment">%   - psoralens</span>
0026 <span class="comment">%   - DNA-protein cross links</span>
0027 <span class="comment">%   - Sugar-phosphate damage other than strand breaks</span>
0028 <span class="comment">%</span>
0029 <span class="comment">%   Additionally, the simulation does not consider mutiple damages per</span>
0030 <span class="comment">%   base or sugar-phosphate.</span>
0031 <span class="comment">%</span>
0032 <span class="comment">%   Categorizes strand breaks in DNA segments (216 bases) according to</span>
0033 <span class="comment">%   scheme described in the track structure model literature [PUB_0486]:</span>
0034 <span class="comment">%   - NB     -- no strand break</span>
0035 <span class="comment">%   - SSB    -- single strand break</span>
0036 <span class="comment">%   - SSB+   -- multiple single strand breaks on segment all on same strand</span>
0037 <span class="comment">%   - 2SSB   -- single strand breaks on both strand separated by at least 10 bases on DNA segment</span>
0038 <span class="comment">%   - DSB    -- two single strand breaks on opposite strands within 10 bases</span>
0039 <span class="comment">%   - DSB+   -- double strand break with additional strand break on DNA segment</span>
0040 <span class="comment">%   - DSB++  -- two single strand breaks on same DNA segment</span>
0041 <span class="comment">%</span>
0042 <span class="comment">%   Effect of DNA damage on physiology</span>
0043 <span class="comment">%   ++++++++++++++++++++++</span>
0044 <span class="comment">%   The exact effect of DNA damage on physiology is not well characterized.</span>
0045 <span class="comment">%   Certain kinds of damage are believed to arrest transcription</span>
0046 <span class="comment">%   [PUB_0780]. Tornaletti and Hanawalt provide the most extensive review</span>
0047 <span class="comment">%   of the effect of specific damages on transcription [PUB_0781, PUB_0782,</span>
0048 <span class="comment">%   PUB_0783]. For simplicity we we model all damages (except m6AD involved</span>
0049 <span class="comment">%   in R/M systems) as barriers to protein binding and function. This is</span>
0050 <span class="comment">%   implemented with the chromosome state class.</span>
0051 <span class="comment">%</span>
0052 <span class="comment">%   Knowledge Base</span>
0053 <span class="comment">%   ======================</span>
0054 <span class="comment">%   As of 8/12/2010 the M. genitalium knowledge contained the stochiometry</span>
0055 <span class="comment">%   and kinetics of 128 DNA damage reactions involving 15 damaging agents</span>
0056 <span class="comment">%   and 9 spontaneous reactions. These reactions were compiled from</span>
0057 <span class="comment">%   several sources. The rates of the most frequent types of damage are</span>
0058 <span class="comment">%   listed below.</span>
0059 <span class="comment">%</span>
0060 <span class="comment">%      Type                                            Reactions</span>
0061 <span class="comment">%      ============================================    =========</span>
0062 <span class="comment">%      base alkylation                                   7</span>
0063 <span class="comment">%      base amination                                    3</span>
0064 <span class="comment">%      base ethylation                                  39</span>
0065 <span class="comment">%      base glucosyl transfer                            2</span>
0066 <span class="comment">%      base methylation                                 42</span>
0067 <span class="comment">%      base reduction                                    1</span>
0068 <span class="comment">%      photooxidation                                    2</span>
0069 <span class="comment">%      radiation (gamma-ray) induced base oxidation     19</span>
0070 <span class="comment">%      spontaneous base deamination                      4</span>
0071 <span class="comment">%      spontaneous base loss                             4</span>
0072 <span class="comment">%      strand break                                      2</span>
0073 <span class="comment">%      UV-B photodimerization                           12</span>
0074 <span class="comment">%</span>
0075 <span class="comment">%      Damaging agent                                  Reactions</span>
0076 <span class="comment">%      ============================================    =========</span>
0077 <span class="comment">%      butylene glycol                                   1</span>
0078 <span class="comment">%      CH3                                               4</span>
0079 <span class="comment">%      chloroethylene oxide                              5</span>
0080 <span class="comment">%      DES                                              13</span>
0081 <span class="comment">%      EMS                                              13</span>
0082 <span class="comment">%      ENU                                              13</span>
0083 <span class="comment">%      Gamma radiation                                   1</span>
0084 <span class="comment">%      hydrogen radical                                  1</span>
0085 <span class="comment">%      hydroxyl radical                                 19</span>
0086 <span class="comment">%      MMS                                               9</span>
0087 <span class="comment">%      MMSO4                                             7</span>
0088 <span class="comment">%      MNNG                                             10</span>
0089 <span class="comment">%      MNU                                              12</span>
0090 <span class="comment">%      NH3                                               1</span>
0091 <span class="comment">%      Spontaneous                                       9</span>
0092 <span class="comment">%      UVB radiation                                    14</span>
0093 <span class="comment">%</span>
0094 <span class="comment">%      Damage                          Spec Rate             Rate (1/s)   No (1/Cycle)</span>
0095 <span class="comment">%      ============================    ===================   ==========   ============</span>
0096 <span class="comment">%      Spontaneous base deamination    1.1e-12 1/nt/s        1.4e-06      0.1</span>
0097 <span class="comment">%      Spontaneous base loss           7.2e-11 1/nt/s        8.4e-05      7.8</span>
0098 <span class="comment">%      UV-B Dimer                      1.2e-07 1/nt/(J/m^2)  6.0e-01      5.6e4</span>
0099 <span class="comment">%      \gamma-ray base oxidation       5.2e-07 1/nt/Gy       1.7e-11      1.6e-6</span>
0100 <span class="comment">%      \gamma-ray strand break         1.7e-07 1/nt/Gy       5.6e-12      5.2e-7</span>
0101 <span class="comment">%      base alkylation                 ?                     ?            ?</span>
0102 <span class="comment">%      base amination                  ?                     ?            ?</span>
0103 <span class="comment">%      base ethylation                 ?                     ?            ?</span>
0104 <span class="comment">%      base glucosyl transfer          ?                     ?            ?</span>
0105 <span class="comment">%      base methylation                ?                     ?            ?</span>
0106 <span class="comment">%      base reduction                  ?                     ?            ?</span>
0107 <span class="comment">%      photooxidation                  ?                     ?            ?</span>
0108 <span class="comment">%</span>
0109 <span class="comment">%   Representation</span>
0110 <span class="comment">%   ======================</span>
0111 <span class="comment">%   7 properties are used to keep track of the damage state of DNA. The first</span>
0112 <span class="comment">%   dimension of each has length equal to that of the chromosome.</span>
0113 <span class="comment">%   - gapSites                  true/false    sugar gaps</span>
0114 <span class="comment">%   - abasicSites               true/false    abasic sites</span>
0115 <span class="comment">%   - damagedSugarPhosphates    enumeration   damaged sugar phosphates</span>
0116 <span class="comment">%   - damagedBases              enumeration   damaged bases</span>
0117 <span class="comment">%   - intrastrandCrossLinks     true/false    intrastand cross link</span>
0118 <span class="comment">%   - strandBreaks              true/false    strand break (missing sugar-phosphate bond)</span>
0119 <span class="comment">%   - hollidayJunctions         true/false    holliday junction</span>
0120 <span class="comment">%</span>
0121 <span class="comment">%   substrates represents the amounts of free metabolites. initializeConstants</span>
0122 <span class="comment">%   synchronizes the substrates of this process with that of the DNA repair</span>
0123 <span class="comment">%   process to make it easy for these processes to communicate. That is</span>
0124 <span class="comment">%   initializeConstants sets the substrates of both processes equal to the union</span>
0125 <span class="comment">%   of the two, are orders the substrates the same in both processes. Thus</span>
0126 <span class="comment">%   substrate indices defined in process are valid on properties received from</span>
0127 <span class="comment">%   the other process, and it is unecessary to remap properties in</span>
0128 <span class="comment">%   copyFromState.</span>
0129 <span class="comment">%</span>
0130 <span class="comment">%   stimuli represents the amounts of DNA damaging agents including several</span>
0131 <span class="comment">%   kinds of radiation.</span>
0132 <span class="comment">%</span>
0133 <span class="comment">%   This process uses three properties rather than reactionStoichiometryMatrix</span>
0134 <span class="comment">%   to represent DNA damaging reactions and the DNA, damaging agents, and free</span>
0135 <span class="comment">%   metabolites that participate in them. These three matrices are a</span>
0136 <span class="comment">%   factorization of the super class reactionStoichiometryMatrix property.</span>
0137 <span class="comment">%   - reactionSmallMoleculeStoichiometryMatrix: the stochiometry of free</span>
0138 <span class="comment">%     metabolites in damage reactions</span>
0139 <span class="comment">%   - reactionDNAStoichiometryMatrix: the stoichiometry of DNA bases in</span>
0140 <span class="comment">%     damage reactions</span>
0141 <span class="comment">%   - reactionRadiationStoichiometryMatrix: the stoichiometry of radiation</span>
0142 <span class="comment">%     stimuli in damage reactions</span>
0143 <span class="comment">%</span>
0144 <span class="comment">%   This process uses reactionVulnerableMotifs, rather the super class</span>
0145 <span class="comment">%   reactionModificationMatrix property to represent the affects of the DNA</span>
0146 <span class="comment">%   damage reaction on macromolecules. reactionVulnerableMotifs represents the DNA</span>
0147 <span class="comment">%   sequence(s) vulnerable to each damage reaction.</span>
0148 <span class="comment">%</span>
0149 <span class="comment">%   Initialization</span>
0150 <span class="comment">%   ======================</span>
0151 <span class="comment">%   The DNA is initialized to the undamaged state, except that MunI</span>
0152 <span class="comment">%   restriction/modification sites are initialized to the methylated state (see</span>
0153 <span class="comment">%   DNA Repair process).</span>
0154 <span class="comment">%</span>
0155 <span class="comment">%   Simulation</span>
0156 <span class="comment">%   ======================</span>
0157 <span class="comment">%   For each reaction</span>
0158 <span class="comment">%     1. Compute rate that individual vulnerable DNA regions succombs to damage</span>
0159 <span class="comment">%     2. Randomly select among vulnerable DNA regions at rate computed in step</span>
0160 <span class="comment">%        (1)</span>
0161 <span class="comment">%     3. Mark DNA selected in step (2) as damaged by updating gapSites,</span>
0162 <span class="comment">%        abasicSites, damagedSugarPhosphates, damagedBases,</span>
0163 <span class="comment">%        intrastrandCrossLinks, strandBreaks, and hollidayJunctions</span>
0164 <span class="comment">%     4. Decrement/increment free metabolites reaction reactants and products</span>
0165 <span class="comment">%        using reactionSmallMoleculeStoichiometryMatrix</span>
0166 <span class="comment">%   end</span>
0167 <span class="comment">%</span>
0168 <span class="comment">%   References</span>
0169 <span class="comment">%   ======================</span>
0170 <span class="comment">%   1. Errol C Friedberg, Graham C Walker, and Wolfram Siede (1995). DNA</span>
0171 <span class="comment">%      Repair and Mutagenesis. ASM Press: Washington, DC. [PUB_0461]</span>
0172 <span class="comment">%   2. Nikjoo H, O'Neill P, Goodhead DT, and Terrissol M. Computational</span>
0173 <span class="comment">%      modelling of low-energy electron-induced DNA damage by  early physical</span>
0174 <span class="comment">%      and chemical events (1997). Int J Radiat Biol. 71(5): 467-83.</span>
0175 <span class="comment">%      [PUB_0486]</span>
0176 <span class="comment">%   3. Cline SD, Hanawalt PC (2003). Who's on first in the cellular</span>
0177 <span class="comment">%      response to DNA damage? Nat Rev Mol Cell Biol. 4(5):361-72.</span>
0178 <span class="comment">%      [PUB_0780]</span>
0179 <span class="comment">%   4. Tornaletti S (2005). Transcription arrest at DNA damage sites. Mutat</span>
0180 <span class="comment">%      Res. 577(1-2):131-45. [PUB_0781]</span>
0181 <span class="comment">%   5. Tornaletti S    (2009). Transcription-coupled DNA repair: directing</span>
0182 <span class="comment">%      your effort where it's most needed. Cell Mol Life Sci.</span>
0183 <span class="comment">%      66(6):1010-20. [PUB_0782]</span>
0184 <span class="comment">%   6. Tornaletti S, Hanawalt PC (1999). Effect of DNA lesions on</span>
0185 <span class="comment">%      transcription elongation. Biochimie. 81 (1-2):139-46. [PUB_0783]</span>
0186 <span class="comment">%</span>
0187 <span class="comment">% Author: Jonathan Karr, jkarr@stanford.edu</span>
0188 <span class="comment">% Affilitation: Covert Lab, Department of Bioengineering, Stanford University</span>
0189 <span class="comment">% Last updated: 4/25/2010</span>
0190 classdef  <a href="DNADamage.html" class="code" title="">DNADamage</a> &lt; edu.stanford.covert.cell.sim.ReactionProcess &amp; edu.stanford.covert.cell.sim.ChromosomeProcessAspect
0191     <span class="comment">%property annotations</span>
0192     properties (Constant)
0193         optionNames__              = {}; <span class="comment">%names of option properties</span>
0194         fixedConstantNames__       = {   <span class="comment">%names of fixed constant properties</span>
0195             <span class="string">'reactionVulnerableMotifs'</span>;
0196             <span class="string">'reactionVulnerableMotifTypes'</span>;
0197             <span class="string">'reactionDamageTypes'</span>;
0198             <span class="string">'reactionSmallMoleculeStoichiometryMatrix'</span>;
0199             <span class="string">'reactionDNAStoichiometryMatrix'</span>;
0200             <span class="string">'reactionRadiationStoichiometryMatrix'</span>;
0201             <span class="string">'reactionDNAProduct'</span>;
0202             <span class="string">'reactionRadiation'</span>;
0203             };
0204         fittedConstantNames__      = {}; <span class="comment">%names of fitted constant properties</span>
0205         localStateNames__          = {}; <span class="comment">%names of simulation state properties redundant with timecourses in this or other processes or the simulation</span>
0206     <span class="keyword">end</span>
0207     
0208     <span class="comment">%enumerations</span>
0209     properties (Constant)
0210         unmodifiedNucleobases = <span class="string">'ACGT'</span>;           <span class="comment">%symbols for unmodified bases</span>
0211     <span class="keyword">end</span>
0212     
0213     <span class="comment">%IDs, names, and local indices</span>
0214     properties
0215         substrateIndexs_radiation                 <span class="comment">%indices within substrates of radiation</span>
0216         substrateIndexs_hydrogen                  <span class="comment">%index within substrates of hydrogen</span>
0217         substrateIndexs_hydroxylRadical           <span class="comment">%index within substrates of hydroxyl radical</span>
0218         substrateIndexs_water                     <span class="comment">%index within substrates of water</span>
0219         substrateIndexs_DR5P                      <span class="comment">%index within substrates of deoxyribose-5-phosphate</span>
0220         substrateIndexs_dRibose5P_dRibose5P       <span class="comment">%index within substrates of (deoxyribose-5-phosphate)-(deoxyribose-5-phosphate)</span>
0221         substrateIndexs_unmodifiedNucleobases     <span class="comment">%indices within substrates of unmodified nucleobases</span>
0222         substrateIndexs_unmodifiedNucleosides     <span class="comment">%indices within substrates of unmodified deoxyribonucleosides</span>
0223         substrateIndexs_unmodifiedDNMPs           <span class="comment">%indices within substrates of unmodified dNMPs</span>
0224         substrateIndexs_modifiedNucleobases       <span class="comment">%indices within substrates of modified nucleobase</span>
0225         substrateIndexs_modifiedNucleosides       <span class="comment">%indices within substrates of modified deoxyribonucleoside</span>
0226         substrateIndexs_modifiedDNMPs             <span class="comment">%indices within substrates of modified dNMPs</span>
0227         
0228         reactionIndexs_spontaneousBaseLoss        <span class="comment">%index within reactions of spontaneous base loss</span>
0229         reactionIndexs_spontaneousBaseDeamination <span class="comment">%index within reactions of spontaneous base deamination</span>
0230         reactionIndexs_gammaRayBaseOxidation      <span class="comment">%index within reactions of gamma ray base oxidation</span>
0231         reactionIndexs_uvbDimerization            <span class="comment">%index within reactions of UVB dimerization</span>
0232         reactionIndexs_baseReduction              <span class="comment">%index within reactions of base reduction</span>
0233         reactionIndexs_baseAmination              <span class="comment">%index within reactions of base amination</span>
0234         reactionIndexs_baseGlucosylTransfer       <span class="comment">%index within reactions of base glucosyl transfer</span>
0235         reactionIndexs_baseAlkylation             <span class="comment">%index within reactions of base alkylation</span>
0236         reactionIndexs_baseEthylation             <span class="comment">%index within reactions of base ethylation</span>
0237         reactionIndexs_baseMethylation            <span class="comment">%index within reactions of base methylation</span>
0238         reactionIndexs_photooxidation             <span class="comment">%index within reactions of photooxidation</span>
0239         reactionIndexs_strandBreak                <span class="comment">%index within reactions of strand break</span>
0240     <span class="keyword">end</span>
0241     
0242     <span class="comment">%fixed biological constants</span>
0243     properties
0244         reactionVulnerableMotifs                 <span class="comment">%DNA base sequence / damaged bases / sugar-phosphate subject to each damage reaction</span>
0245         reactionVulnerableMotifTypes             <span class="comment">%type of each reactionVulnerableMotif (sequence, gapSites, abasicSites, damagedSugarPhosphates, damagedBases, intrastrandCrossLinks, strandBreaks, or hollidayJunctions)</span>
0246         reactionDamageTypes                      <span class="comment">%type of damage of reach reaction (gapSites, abasicSites, damagedSugarPhosphates, damagedBases, intrastrandCrossLinks, strandBreaks, or hollidayJunctions)</span>
0247         reactionSmallMoleculeStoichiometryMatrix <span class="comment">%small molecule stoichiometry of reactions</span>
0248         reactionDNAStoichiometryMatrix           <span class="comment">%DNA base stoichiometry of reactions</span>
0249         reactionRadiationStoichiometryMatrix     <span class="comment">%DNA radiation stoichiometry of reactions</span>
0250         reactionDNAProduct                       <span class="comment">%global index of resultant DNA product of each reaction</span>
0251         reactionRadiation                        <span class="comment">%local index of radiation causative of each reaction</span>
0252     <span class="keyword">end</span>
0253     
0254     <span class="comment">%constructor</span>
0255     methods
0256         <a name="_sub0" href="#_subfunctions" class="code">function this = DNADamage(wholeCellModelID, name)</a>
0257             this = this@edu.stanford.covert.cell.sim.ReactionProcess(wholeCellModelID, name);
0258         <span class="keyword">end</span>
0259     <span class="keyword">end</span>
0260     
0261     <span class="comment">%communication between process/simulation</span>
0262     methods
0263         <span class="comment">%set references to state objects</span>
0264         <a name="_sub1" href="#_subfunctions" class="code">function storeObjectReferences(this, simulation)</a>
0265             this.storeObjectReferences@edu.stanford.covert.cell.sim.Process(simulation);
0266             this.storeObjectReferences@edu.stanford.covert.cell.sim.ChromosomeProcessAspect(simulation);
0267         <span class="keyword">end</span>
0268         
0269         <span class="comment">%initialize constants</span>
0270         <a name="_sub2" href="#_subfunctions" class="code">function initializeConstants(this, knowledgeBase, simulation, varargin)</a>
0271             <span class="comment">%% call super class method to get reactions, substrates, and enzymes</span>
0272             this.initializeConstants@edu.stanford.covert.cell.sim.ReactionProcess(knowledgeBase, simulation, varargin{:});
0273             this.initializeConstants@edu.stanford.covert.cell.sim.ChromosomeProcessAspect(knowledgeBase, simulation, varargin{:});
0274             
0275             <span class="comment">%% remove reactions with zero rate</span>
0276             this.initializeConstants_overrideReactions(simulation, this.reactionWholeCellModelIDs(this.reactionBounds(:, 2) ~= 0));
0277             
0278             <span class="comment">%% additional substrates</span>
0279             substrateWholeCellModelIDs  = unique([
0280                 this.substrateWholeCellModelIDs;
0281                 <span class="string">'DA'</span>; <span class="string">'DC'</span>; <span class="string">'DG'</span>; <span class="string">'DT'</span>]);
0282             this.initializeConstants_overrideSubstrates(substrateWholeCellModelIDs);
0283             
0284             <span class="comment">%% substrate indices</span>
0285             this.substrateIndexs_radiation             = this.substrateIndexs({<span class="string">'UVB_radiation'</span>;<span class="string">'gamma_radiation'</span>});
0286             this.substrateIndexs_hydrogen              = this.substrateIndexs({<span class="string">'H'</span>});
0287             this.substrateIndexs_hydroxylRadical       = this.substrateIndexs({<span class="string">'hydroxyl_radical'</span>});
0288             this.substrateIndexs_water                 = this.substrateIndexs({<span class="string">'H2O'</span>});
0289             this.substrateIndexs_DR5P                  = this.substrateIndexs({<span class="string">'DR5P'</span>});
0290             this.substrateIndexs_dRibose5P_dRibose5P   = this.substrateIndexs({<span class="string">'dRibose5P_dRibose5P'</span>});
0291             this.substrateIndexs_unmodifiedNucleobases = this.substrateIndexs({<span class="string">'AD'</span>;<span class="string">'CSN'</span>;<span class="string">'GN'</span>;<span class="string">'THY'</span>});
0292             this.substrateIndexs_unmodifiedNucleosides = this.substrateIndexs({<span class="string">'DA'</span>;<span class="string">'DC'</span>;<span class="string">'DG'</span>;<span class="string">'DT'</span>});
0293             this.substrateIndexs_unmodifiedDNMPs       = this.substrateIndexs({<span class="string">'DAMP'</span>;<span class="string">'DCMP'</span>;<span class="string">'DGMP'</span>;<span class="string">'DTMP'</span>});
0294             this.substrateIndexs_modifiedNucleobases   = this.substrateMetaboliteLocalIndexs(<span class="keyword">...</span>
0295                 strcmp({knowledgeBase.metabolites(this.substrateMetaboliteGlobalIndexs).category},<span class="string">'modified nucleobase'</span>) | <span class="keyword">...</span>
0296                 strcmp({knowledgeBase.metabolites(this.substrateMetaboliteGlobalIndexs).wholeCellModelID},<span class="string">'URA'</span>));
0297             this.substrateIndexs_modifiedNucleosides   = this.substrateMetaboliteLocalIndexs(<span class="keyword">...</span>
0298                 strcmp({knowledgeBase.metabolites(this.substrateMetaboliteGlobalIndexs).category},<span class="string">'modified deoxyribonucleoside'</span>) | <span class="keyword">...</span>
0299                 strcmp({knowledgeBase.metabolites(this.substrateMetaboliteGlobalIndexs).wholeCellModelID},<span class="string">'URI'</span>));
0300             this.substrateIndexs_modifiedDNMPs         = this.substrateMetaboliteLocalIndexs(<span class="keyword">...</span>
0301                 strcmp({knowledgeBase.metabolites(this.substrateMetaboliteGlobalIndexs).category},<span class="string">'modified deoxyribonucleotide monophosphate'</span>) | <span class="keyword">...</span>
0302                 strcmp({knowledgeBase.metabolites(this.substrateMetaboliteGlobalIndexs).wholeCellModelID},<span class="string">'DUMP'</span>));
0303             
0304             <span class="comment">%% reaction indices</span>
0305             this.reactionIndexs_spontaneousBaseLoss        = find(strcmp(this.reactionTypes,<span class="string">'spontaneous base loss'</span>));
0306             this.reactionIndexs_spontaneousBaseDeamination = find(strcmp(this.reactionTypes,<span class="string">'spontaneous base deamination'</span>));
0307             this.reactionIndexs_gammaRayBaseOxidation      = find(strcmp(this.reactionTypes,<span class="string">'radiation (gamma-ray) induced base oxidation'</span>));
0308             this.reactionIndexs_uvbDimerization            = find(strcmp(this.reactionTypes,<span class="string">'UV-B photodimerization'</span>));
0309             this.reactionIndexs_strandBreak                = find(strcmp(this.reactionTypes,<span class="string">'strand break'</span>));
0310             this.reactionIndexs_baseReduction              = find(strcmp(this.reactionTypes,<span class="string">'base reduction'</span>));
0311             this.reactionIndexs_baseAmination              = find(strcmp(this.reactionTypes,<span class="string">'base amination'</span>));
0312             this.reactionIndexs_baseGlucosylTransfer       = find(strcmp(this.reactionTypes,<span class="string">'base glucosyl transfer'</span>));
0313             this.reactionIndexs_baseAlkylation             = find(strcmp(this.reactionTypes,<span class="string">'base alkylation'</span>));
0314             this.reactionIndexs_baseEthylation             = find(strcmp(this.reactionTypes,<span class="string">'base ethylation'</span>));
0315             this.reactionIndexs_baseMethylation            = find(strcmp(this.reactionTypes,<span class="string">'base methylation'</span>));
0316             this.reactionIndexs_photooxidation             = find(strcmp(this.reactionTypes,<span class="string">'photooxidation'</span>));
0317             
0318             <span class="comment">%% reaction types</span>
0319             this.reactionDamageTypes = cell(size(this.reactionWholeCellModelIDs));
0320             this.reactionDamageTypes(this.reactionIndexs_spontaneousBaseLoss)        = {<span class="string">'abasicSites'</span>};
0321             this.reactionDamageTypes(this.reactionIndexs_spontaneousBaseDeamination) = {<span class="string">'damagedBases'</span>};
0322             this.reactionDamageTypes(this.reactionIndexs_gammaRayBaseOxidation)      = {<span class="string">'damagedBases'</span>};
0323             this.reactionDamageTypes(this.reactionIndexs_uvbDimerization)            = {<span class="string">'intrastrandCrossLinks'</span>};
0324             this.reactionDamageTypes(this.reactionIndexs_strandBreak)                = {<span class="string">'strandBreaks'</span>};
0325             this.reactionDamageTypes(this.reactionIndexs_baseReduction)              = {<span class="string">'damagedBases'</span>};
0326             this.reactionDamageTypes(this.reactionIndexs_baseAmination)              = {<span class="string">'damagedBases'</span>};
0327             this.reactionDamageTypes(this.reactionIndexs_baseGlucosylTransfer)       = {<span class="string">'damagedBases'</span>};
0328             this.reactionDamageTypes(this.reactionIndexs_baseAlkylation)             = {<span class="string">'damagedBases'</span>};
0329             this.reactionDamageTypes(this.reactionIndexs_baseEthylation)             = {<span class="string">'damagedBases'</span>};
0330             this.reactionDamageTypes(this.reactionIndexs_baseMethylation)            = {<span class="string">'damagedBases'</span>};
0331             this.reactionDamageTypes(this.reactionIndexs_photooxidation)             = {<span class="string">'damagedBases'</span>};
0332             
0333             <span class="comment">%% Reaction stoichiometry matrices</span>
0334             <span class="comment">%Factor reactionStoichiometryMatrix property into 3 parts</span>
0335             <span class="comment">%- reactionSmallMoleculeStoichiometryMatrix: the stochiometry of</span>
0336             <span class="comment">%  free metabolites in damage reactions. Since we don't have an</span>
0337             <span class="comment">%  explicit model of the production and amount of hydroxyl</span>
0338             <span class="comment">%  radicals, and the rate of hydroxyl radical production from</span>
0339             <span class="comment">%  water using radiation is implicit in the rate of the damage</span>
0340             <span class="comment">%  reactions that use hydroxyl radicals, we replace hydroxyl</span>
0341             <span class="comment">%  radical with water and hydrogen in reaction stoichiometries.</span>
0342             <span class="comment">%- reactionDNAStoichiometryMatrix: the stoichiometry of DNA bases</span>
0343             <span class="comment">%  in damage reactions</span>
0344             <span class="comment">%- reactionRadiationStoichiometryMatrix: the stoichiometry of</span>
0345             <span class="comment">%  radiation stimuli in damage reactions</span>
0346             numReactions = length(this.reactionWholeCellModelIDs);
0347             
0348             <span class="comment">%indices of substrates in reactions that should be considered part</span>
0349             <span class="comment">%of DNA, rather than free metabolites</span>
0350             substrateIndexs_DNA = [
0351                 this.substrateIndexs_unmodifiedNucleobases;
0352                 this.substrateIndexs_unmodifiedNucleosides;
0353                 this.substrateIndexs_unmodifiedDNMPs;
0354                 this.substrateIndexs_modifiedNucleobases;
0355                 this.substrateIndexs_modifiedNucleosides;
0356                 this.substrateIndexs_modifiedDNMPs;
0357                 this.substrateIndexs_DR5P;
0358                 this.substrateIndexs_dRibose5P_dRibose5P];
0359             
0360             <span class="comment">%allocate memory</span>
0361             this.reactionSmallMoleculeStoichiometryMatrix = zeros(size(this.reactionStoichiometryMatrix));
0362             this.reactionDNAStoichiometryMatrix           = zeros(size(this.reactionStoichiometryMatrix));
0363             this.reactionRadiationStoichiometryMatrix     = zeros(size(this.reactionStoichiometryMatrix));
0364             
0365             <span class="comment">%DNA</span>
0366             this.reactionDNAStoichiometryMatrix(substrateIndexs_DNA, :) = <span class="keyword">...</span>
0367                 this.reactionStoichiometryMatrix(substrateIndexs_DNA, :);
0368             
0369             [i, j] = find(this.reactionDNAStoichiometryMatrix &gt; 0);
0370             this.reactionDNAProduct = zeros(numel(this.reactionWholeCellModelIDs), 1);
0371             this.reactionDNAProduct(j) = this.substrateGlobalIndexs(i);
0372             
0373             <span class="comment">%radiation</span>
0374             this.reactionRadiationStoichiometryMatrix(this.substrateIndexs_radiation,:) = <span class="keyword">...</span>
0375                 this.reactionStoichiometryMatrix(this.substrateIndexs_radiation, :);
0376             
0377             [i, j] = find(this.reactionRadiationStoichiometryMatrix);
0378             this.reactionRadiation = zeros(numel(this.reactionWholeCellModelIDs), 1);
0379             this.reactionRadiation(j) = i;
0380             
0381             <span class="comment">%free metabolites</span>
0382             this.reactionSmallMoleculeStoichiometryMatrix = <span class="keyword">...</span>
0383                 this.reactionStoichiometryMatrix - <span class="keyword">...</span>
0384                 this.reactionDNAStoichiometryMatrix - <span class="keyword">...</span>
0385                 this.reactionRadiationStoichiometryMatrix;
0386             
0387             <span class="comment">%radicals</span>
0388             this.reactionSmallMoleculeStoichiometryMatrix([this.substrateIndexs_water;this.substrateIndexs_hydrogen],:) = <span class="keyword">...</span>
0389                 this.reactionSmallMoleculeStoichiometryMatrix([this.substrateIndexs_water;this.substrateIndexs_hydrogen],:) + <span class="keyword">...</span>
0390                 repmat([1;-1], 1, numReactions).*repmat(this.reactionSmallMoleculeStoichiometryMatrix(this.substrateIndexs_hydroxylRadical,:),2,1);
0391             this.reactionSmallMoleculeStoichiometryMatrix(this.substrateIndexs_hydroxylRadical,:) = 0;
0392             
0393             <span class="comment">%reaction-specific corrections</span>
0394             this.reactionSmallMoleculeStoichiometryMatrix(this.substrateIndexs_unmodifiedNucleobases,this.reactionIndexs_spontaneousBaseLoss) = <span class="keyword">...</span>
0395                 this.reactionStoichiometryMatrix(this.substrateIndexs_unmodifiedNucleobases,this.reactionIndexs_spontaneousBaseLoss);
0396             this.reactionDNAStoichiometryMatrix(this.substrateIndexs_unmodifiedNucleobases,this.reactionIndexs_spontaneousBaseLoss) = 0;
0397             
0398             <span class="comment">%% reaction sequences</span>
0399             <span class="comment">% represents the DNA sequence(s) vulnerable to each damage reaction</span>
0400             this.reactionVulnerableMotifs = repmat({<span class="string">''</span>}, size(this.reactionStoichiometryMatrix,2), 1);
0401             this.reactionVulnerableMotifTypes = repmat({<span class="string">''</span>}, size(this.reactionStoichiometryMatrix,2), 1);
0402             
0403             <span class="keyword">for</span> i = 1:length(this.reactionWholeCellModelIDs)
0404                 dnaSubstrateLocalIdx = find(this.reactionDNAStoichiometryMatrix(:, i) &lt; 0);
0405                 <span class="keyword">if</span> this.reactionDNAStoichiometryMatrix(dnaSubstrateLocalIdx, i) ~= -1; <span class="keyword">continue</span>; <span class="keyword">end</span>;
0406                 
0407                 [~, baseIdx] = ismember(dnaSubstrateLocalIdx, [this.substrateIndexs_unmodifiedNucleobases; this.substrateIndexs_unmodifiedNucleosides; this.substrateIndexs_unmodifiedDNMPs]);
0408                 <span class="keyword">if</span> baseIdx
0409                     this.reactionVulnerableMotifs{i} = this.unmodifiedNucleobases(mod(baseIdx - 1,4) + 1);
0410                     this.reactionVulnerableMotifTypes{i} = <span class="string">'sequence'</span>;
0411                 <span class="keyword">else</span>
0412                     this.reactionVulnerableMotifs{i} = this.substrateGlobalIndexs(dnaSubstrateLocalIdx);
0413                     this.reactionVulnerableMotifTypes{i} = this.reactionDamageTypes{i};
0414                 <span class="keyword">end</span>
0415             <span class="keyword">end</span>
0416             
0417             this.reactionVulnerableMotifs(this.reactionIndexs_strandBreak) = {true};
0418             this.reactionVulnerableMotifTypes(this.reactionIndexs_strandBreak) = {<span class="string">'abasicSites'</span>};
0419             
0420             this.reactionVulnerableMotifTypes{this.reactionIndexs({<span class="string">'DNADamage_CSNCSN_cyclobutane_CSNCSN_UVB_radiation'</span>})}  = <span class="string">'sequence'</span>;
0421             this.reactionVulnerableMotifTypes{this.reactionIndexs({<span class="string">'DNADamage_THYTHY_cyclobutane_THYTHY_UVB_radiation'</span>})}  = <span class="string">'sequence'</span>;
0422             this.reactionVulnerableMotifTypes{this.reactionIndexs({<span class="string">'DNADamage_CSNTHY_cyclobutane_CSNTHY_UVB_radiation'</span>})}  = <span class="string">'sequence'</span>;
0423             this.reactionVulnerableMotifTypes{this.reactionIndexs({<span class="string">'DNADamage_THYCSN_cyclobutane_THYCSN_UVB_radiation'</span>})}  = <span class="string">'sequence'</span>;
0424             this.reactionVulnerableMotifTypes{this.reactionIndexs({<span class="string">'DNADamage_CSNTHY_CSN64THY_UVB_radiation'</span>})}            = <span class="string">'sequence'</span>;
0425             <span class="comment">%this.reactionVulnerableMotifTypes{this.reactionIndexs({'DNADamage_CSNCSN_CSN64CSN_UVB_radiation'})}            = 'sequence';</span>
0426             this.reactionVulnerableMotifTypes{this.reactionIndexs({<span class="string">'DNADamage_THYCSN_THY64CSN_UVB_radiation'</span>})}            = <span class="string">'sequence'</span>;
0427             this.reactionVulnerableMotifTypes{this.reactionIndexs({<span class="string">'DNADamage_THYTHY_THY64THY_UVB_radiation'</span>})}            = <span class="string">'sequence'</span>;
0428             
0429             this.reactionVulnerableMotifs{this.reactionIndexs({<span class="string">'DNADamage_CSNCSN_cyclobutane_CSNCSN_UVB_radiation'</span>})}  = <span class="string">'CC'</span>;
0430             this.reactionVulnerableMotifs{this.reactionIndexs({<span class="string">'DNADamage_THYTHY_cyclobutane_THYTHY_UVB_radiation'</span>})}  = <span class="string">'TT'</span>;
0431             this.reactionVulnerableMotifs{this.reactionIndexs({<span class="string">'DNADamage_CSNTHY_cyclobutane_CSNTHY_UVB_radiation'</span>})}  = <span class="string">'CT'</span>;
0432             this.reactionVulnerableMotifs{this.reactionIndexs({<span class="string">'DNADamage_THYCSN_cyclobutane_THYCSN_UVB_radiation'</span>})}  = <span class="string">'TC'</span>;
0433             this.reactionVulnerableMotifs{this.reactionIndexs({<span class="string">'DNADamage_CSNTHY_CSN64THY_UVB_radiation'</span>})}            = <span class="string">'CT'</span>;
0434             <span class="comment">%this.reactionVulnerableMotifs{this.reactionIndexs({'DNADamage_CSNCSN_CSN64CSN_UVB_radiation'})}            = 'CC';</span>
0435             this.reactionVulnerableMotifs{this.reactionIndexs({<span class="string">'DNADamage_THYCSN_THY64CSN_UVB_radiation'</span>})}            = <span class="string">'TC'</span>;
0436             this.reactionVulnerableMotifs{this.reactionIndexs({<span class="string">'DNADamage_THYTHY_THY64THY_UVB_radiation'</span>})}            = <span class="string">'TT'</span>;
0437             
0438             <span class="comment">%% reaction rates</span>
0439             this.reactionBounds(:, 1) = 0;
0440         <span class="keyword">end</span>
0441     <span class="keyword">end</span>
0442     
0443     <span class="comment">%model</span>
0444     methods
0445         <span class="comment">%Calculate</span>
0446         <span class="comment">%- contribution to FBA objective</span>
0447         <span class="comment">%- minimum expression consistent with cell cycle length</span>
0448         <a name="_sub3" href="#_subfunctions" class="code">function [bmProd, byProd, minEnzExp, maxEnzExp] = calcResourceRequirements_LifeCycle(this, constants, ~)</a>
0449             <span class="comment">%substrate and byproducts: metabolic effects of damage reactions</span>
0450             rates = this.calcExpectedReactionRates() * (<span class="keyword">...</span>
0451                 + constants.states.Time.cellCycleLength + <span class="keyword">...</span>
0452                 + 0.5 * constants.states.Time.replicationDuration <span class="keyword">...</span>
0453                 + constants.states.Time.cytokinesisDuration);
0454             bmProd = max(0, -this.reactionSmallMoleculeStoichiometryMatrix) * rates;
0455             byProd = max(0,  this.reactionSmallMoleculeStoichiometryMatrix) * rates;
0456             
0457             <span class="comment">%no enzymes involved in DNA damage</span>
0458             minEnzExp = zeros(size(this.enzymeWholeCellModelIDs));
0459             maxEnzExp = Inf(size(this.enzymeWholeCellModelIDs));
0460         <span class="keyword">end</span>
0461         
0462         <span class="comment">%initialization -- simulation initializeState initializes 1 undamaged</span>
0463         <span class="comment">%chromosome (except methylation at restriction/modification sites which</span>
0464         <span class="comment">%is initialized by DNA repair)</span>
0465         <a name="_sub4" href="#_subfunctions" class="code">function initializeState(~)</a>
0466         <span class="keyword">end</span>
0467         
0468         <span class="comment">%resource requirements</span>
0469         <a name="_sub5" href="#_subfunctions" class="code">function result = calcResourceRequirements_Current(this)</a>
0470             <span class="comment">%metabolic effects of reactions</span>
0471             result = max(0, -this.reactionSmallMoleculeStoichiometryMatrix * this.calcExpectedReactionRates());
0472         <span class="keyword">end</span>
0473         
0474         <a name="_sub6" href="#_subfunctions" class="code">function rates = calcExpectedReactionRates(this)</a>
0475             <span class="comment">%number of sites vulnerable to reaction</span>
0476             nVulnerableSites = this.calcNumberVulnerableSites();
0477             
0478             <span class="comment">%calculate rate, product of</span>
0479             <span class="comment">%- number of vulnerable site / chromosome</span>
0480             <span class="comment">%- integrated number of chromosomes over cell cycle</span>
0481             <span class="comment">%- reaction rate per vulnerable site (and per unit radiation</span>
0482             <span class="comment">%  for applicable reactions)</span>
0483             <span class="comment">%- amount of radition, for applicable reactions</span>
0484             rates = nVulnerableSites .* this.reactionBounds(:, 2);
0485             rates(this.reactionRadiation ~= 0) = <span class="keyword">...</span>
0486                 rates(this.reactionRadiation ~= 0) .* <span class="keyword">...</span>
0487                 this.substrates(this.reactionRadiation(this.reactionRadiation ~= 0));
0488         <span class="keyword">end</span>
0489         
0490         <a name="_sub7" href="#_subfunctions" class="code">function nVulnerableSites = calcNumberVulnerableSites(this)</a>
0491             c = this.chromosome;
0492             polDNA = collapse(c.polymerizedRegions);
0493             bases = <span class="string">'ACGT'</span>;
0494             dntpComposition = [
0495                 1 - c.sequenceGCContent;
0496                 c.sequenceGCContent;
0497                 c.sequenceGCContent;
0498                 1 - c.sequenceGCContent] / 2;
0499             [posStrnds, dmg] = find(c.damagedSites); <span class="comment">%safely ignores m6AD because this process doesn't produce or modify m6AD; this is asserted in testConstants in unit test</span>
0500             
0501             nVulnerableSites = zeros(size(this.reactionStoichiometryMatrix, 2), 1);
0502             <span class="keyword">for</span> i = 1:size(this.reactionStoichiometryMatrix, 2)
0503                 <span class="comment">%% damage would be to undamaged sequence</span>
0504                 <span class="keyword">if</span> ischar(this.reactionVulnerableMotifs{i})
0505                     nVulnerableSites(i) = polDNA * prod(dntpComposition(ismembc2(this.reactionVulnerableMotifs{i}, bases)));
0506                     <span class="keyword">continue</span>;
0507                 <span class="keyword">end</span>
0508                 
0509                 <span class="comment">%% damage would be to already damaged sequence</span>
0510                 <span class="keyword">if</span> isempty(posStrnds)
0511                     <span class="keyword">continue</span>;
0512                 <span class="keyword">end</span>
0513                 
0514                 [spcPosStrmds, spcDmg] = find(c.(this.reactionVulnerableMotifTypes{i}));
0515                 <span class="keyword">if</span> isempty(spcPosStrmds)
0516                     <span class="keyword">continue</span>;
0517                 <span class="keyword">end</span>
0518                 
0519                 spcIdxs = find(spcDmg == this.reactionVulnerableMotifs{i});
0520                 <span class="keyword">if</span> isempty(spcIdxs)
0521                     <span class="keyword">continue</span>;
0522                 <span class="keyword">end</span>
0523                 
0524                 idxs = find(dmg == this.reactionVulnerableMotifs{i});
0525                 <span class="keyword">if</span> isempty(idxs)
0526                     <span class="keyword">continue</span>;
0527                 <span class="keyword">end</span>
0528                 
0529                 nVulnerableSites(i) = sum(edu.stanford.covert.util.SparseMat.ismember_subs(<span class="keyword">...</span>
0530                     spcPosStrmds(spcIdxs, :), posStrnds(idxs, :), [c.sequenceLen c.nCompartments]));
0531             <span class="keyword">end</span>
0532         <span class="keyword">end</span>
0533         
0534         <span class="comment">%simulation</span>
0535         <a name="_sub8" href="#_subfunctions" class="code">function evolveState(this)</a>
0536             <span class="comment">%loop over reactions in random order</span>
0537             randomOrder = this.randStream.randperm(numel(this.reactionWholeCellModelIDs));
0538             <span class="keyword">for</span> i = 1:length(this.reactionWholeCellModelIDs)
0539                 j = randomOrder(i);
0540                 
0541                 <span class="comment">%maximum number of reactions, based on substrate availability</span>
0542                 maxReactions = floor(min(this.substrates ./ max(0, -this.reactionSmallMoleculeStoichiometryMatrix(:, j))));
0543                 <span class="keyword">if</span> maxReactions &lt;= 0
0544                     <span class="keyword">continue</span>;
0545                 <span class="keyword">end</span>
0546                 
0547                 <span class="comment">% probability that valid site undergoes reaction</span>
0548                 radiationLclIdx  = this.reactionRadiation(j);
0549                 <span class="keyword">if</span> radiationLclIdx ~= 0
0550                     selectionProbability = this.stepSizeSec * this.reactionBounds(j, 2) * this.substrates(radiationLclIdx);
0551                 <span class="keyword">else</span>
0552                     selectionProbability = this.stepSizeSec * this.reactionBounds(j, 2);
0553                 <span class="keyword">end</span>
0554                 <span class="keyword">if</span> selectionProbability == 0
0555                     <span class="keyword">continue</span>;
0556                 <span class="keyword">end</span>
0557                 
0558                 <span class="comment">%randomly damage sites</span>
0559                 [positionsStrands, sideEffects] = this.chromosome.setSiteDamaged(<span class="keyword">...</span>
0560                     this.reactionDamageTypes{j}, this.reactionDNAProduct(j), selectionProbability, <span class="keyword">...</span>
0561                     maxReactions, this.reactionVulnerableMotifs{j}, this.reactionVulnerableMotifTypes{j});
0562                 <span class="keyword">if</span> isempty(positionsStrands)
0563                     <span class="keyword">continue</span>;
0564                 <span class="keyword">end</span>
0565                 
0566                 <span class="comment">% use small molecule reactants, produce small molecule products</span>
0567                 this.substrates = this.substrates + <span class="keyword">...</span>
0568                     size(positionsStrands, 1) * this.reactionSmallMoleculeStoichiometryMatrix(:,j);
0569                 
0570                 <span class="comment">%side effects</span>
0571                 <span class="keyword">if</span> ~isempty(sideEffects)
0572                     this.simulationStateSideEffects = [this.simulationStateSideEffects; sideEffects];
0573                 <span class="keyword">end</span>
0574             <span class="keyword">end</span>
0575         <span class="keyword">end</span>
0576     <span class="keyword">end</span>
0577 <span class="keyword">end</span></pre></div>
<hr><address>Generated on Thu 19-Jul-2012 18:32:13 by <strong><a href="http://www.artefact.tk/software/matlab/m2html/" title="Matlab Documentation in HTML">m2html</a></strong> &copy; 2005</address>
</body>
</html>