<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN"
                "http://www.w3.org/TR/REC-html40/loose.dtd">
<html>
<head>
  <title>Description of Metabolism</title>
  <meta name="keywords" content="Metabolism">
  <meta name="description" content="Metabolism">
  <meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1">
  <meta name="generator" content="m2html v1.5 &copy; 2003-2005 Guillaume Flandin">
  <meta name="robots" content="index, follow">
  <link type="text/css" rel="stylesheet" href="../../../../../../../../m2html.css">
</head>
<body>
<a name="_top"></a>
<div><a href="../../../../../../../../index.html">Home</a> &gt;  <a href="../../../../../../../index.html">simulation</a> &gt; <a href="#">src</a> &gt; <a href="#">+edu</a> &gt; <a href="#">+stanford</a> &gt; <a href="#">+covert</a> &gt; <a href="#">+cell</a> &gt; <a href="../index.html">+sim</a> &gt; <a href="index.html">+process</a> &gt; Metabolism.m</div>

<!--<table width="100%"><tr><td align="left"><a href="../../../../../../../../index.html"><img alt="<" border="0" src="../../../../../../../../left.png">&nbsp;Master index</a></td>
<td align="right"><a href="index.html">Index for simulation/src/+edu/+stanford/+covert/+cell/+sim/+process&nbsp;<img alt=">" border="0" src="../../../../../../../../right.png"></a></td></tr></table>-->

<h1>Metabolism
</h1>

<h2><a name="_name"></a>PURPOSE <a href="#_top"><img alt="^" border="0" src="../../../../../../../../up.png"></a></h2>
<div class="box"><strong>Metabolism</strong></div>

<h2><a name="_synopsis"></a>SYNOPSIS <a href="#_top"><img alt="^" border="0" src="../../../../../../../../up.png"></a></h2>
<div class="box"><strong>This is a script file. </strong></div>

<h2><a name="_description"></a>DESCRIPTION <a href="#_top"><img alt="^" border="0" src="../../../../../../../../up.png"></a></h2>
<div class="fragment"><pre class="comment">Metabolism

 @wholeCellModelID Process_Metabolism
 @name             Metabolism
 @description
   Biology
   ===============
   To grow and replicate cells must uptake or produce their building blocks and
   intermediate energy stores, particularly nucleic and amino acids and lipids,
   as well as secrete metabolites such as modified nucleic acids which it
   cannot catabolize and extract usable material or energy from. Furthermore,
   the import/synthesis and export/breakdown of metabolites must be carefully
   matched to the metabolic demands of the cell to ensure that cellular
   processes aren't limited by too few metabolites or poisoned by too many
   metabolites.

   This process simulates
   - the uptake of nutrients from the external environment
   - the processing of imported nutrients to intermediate energy forms (ATP,
     GTP) and macromolecule building blocks (dNTPs, NTPs, and amino acids)
   - the assembly of lipids, their insertion into the membrane, and their
     maturation within the membrane
   - the catabolism of byproducts of macromolecule assembly and degradation
   - the export to the external environment of uncatabolizable chemicals such
     as modified nucleobases

   Knowledge Base
   ===============
   M. genitalium metabolism was reconstructed from a variety of sources,
   including flux-balance analysis metabolic models of other bacterial species
   and the reaction kinetics database SABIO-RK, and was organized into 641
   reactions in the knowledge base. These reactions are loaded into this process
   by the initializeConstants method.

      Object       No.
      ==========   ===
      substrates   580
      enzymes      100
      reactions    641
        chemical   433
        transport  208

   Representation
   ===============
   The properties substrates and enzymes represent the counts of metabolites
   and metabolic enzymes.

   fbaReactionStoichiometryMatrix represents the stoichiometry and compartments
   of metabolites and biomass in each of the 641 chemical/transport reactions,
   exchange pseudoreactions, and biomass production pseudoreaction.
   fbaReactionCatalysisMatrix represents the enzyme which catalyzes each
   reaction. fbaEnzymeBounds represents the foward and backward kcat of the
   catalyzing enzyme of each reaction. fbaReactionBounds represents the maximal
   import and export rates of each  metabolite. fbaObjective indicates which
   reaction represents the biomass production pseudoreaction. fbaRightHandSide
   is a vector of zeros representing the change in concentration over time of
   each metabolite and biomass. metabolismProduction is redundant with the
   biomass production reaction in fbaReactionStoichiometryMatrix.
   metabolismProduction is calculated by summing the metabolic demands of all
   the other processes over the entire cell cycle. The table below lists the
   units of several properties of this process.

      Property                       Units
      ===========================    ==============================
      fbaEnzymeBounds                molecules/enzyme/s
      fbaReactionBounds              molecules/(gram dry biomass)/s
      metabolites                    molecules
      enzymes                        molecules
      stepSizeSec                    s
      lowerBounds                    reactions/s
      upperBounds                    reactions/s
      growth                         cell/s
      biomassComposition             molecules/cell
      metabolismProduction           molecules/cell
      chamberVolume                  L
      setValues                      molecules/chamber
      growthAssociatedMaintanence

   Initialization
   ===============
   The simulation is initialized with 1 cell weight of macromolecules and
   water, and few free metabolites by the simulation class' initializeState
   method. In addition this process is initialized to a positive growth rate of
   approximately log(2)/cellCycleLength computing using the flux-balance
   analysis model implemented in evolveState.

   Simulation
   ===============
   Transport and metabolism are modeled using the constraint-based method
   flux-balance analysis (FBA). Briefly, FBA assumes that bacteria have evolved
   to maximize growth, and poses transport and metabolism as the optimization
   of cellular building block and energy production subject to available
   nutrients, and allowed chemical reactions. FBA models are typically
   constrained by experimentally measured transport and diffusion rates. In
   addition to these constraints, we constrain constrain our FBA process by our
   model's predicted enzyme abundances, and by experimentally measured kinetic
   parameters. These additional constraints yield a more accurate model of
   transport and metabolism. Furthermore, we use our other processes to obtain a
   more extensive and accurate metabolic objective than used in earlier FBA
   models. The optimization problem specific by FBA is posed as a linear
   optimization problem, and solving using one of several publically available
   linear programming packages.

   Algorithm
   ++++++++++++++++
   1. Compute reaction bounds based on
     - enzyme kinetics,
     - enzyme availability
     - maximal metabolite exchange rates
     - external metabolite availability
     - protein availability
   2. Computes optimal reaction fluxes which maximize biomass production
   3. Computes integer-valued production of biomass components, and export of
      byproducts
   4. Updates amounts of biomass components and byproducts

   References
   ===============
   1. Orth JD, Thiele I, Palsson BO (2010). What is flux balance analysis?
      Nat Biotechnol. 28(3):245-8 [PUB_0687].
   2. Thiele I, Palsson BO (2010). A protocol for generating a
      high-quality genome-scale metabolic reconstruction. Nat Protoc.
      5(1): 93-121. [PUB_0686].
   3. Covert MW, Xiao N, Chen TJ, Karr JR (2008). Integrating metabolic,
      transcriptional regulatory and signal transduction models in
      Escherichia coli. Bioinformatics. 24(18):2044-50. [PUB_0684].
   4. Covert MW, Knight EM, Reed JL, Herrgard MJ, Palsson BO (2004).
      Integrating high-throughput and computational data elucidates
      bacterial networks. Nature. 429 (6987): 92-6. [PUB_0618].
   5. Covert MW, Palsson BO (2003). Constraints-based models: regulation
      of gene expression reduces the steady-state solution space. J Theor
      Biol. 221(3): 309-25. [PUB_0685].

 Author: Jonathan Karr, jkarr@stanford.edu
 Author: Markus Covert, mcovert@stanford.edu
 Author: Jayodita Sanghvi, jayodita@stanford.edu
 Affilitation: Covert Lab, Department of Bioengineering, Stanford University
 Last updated: 3/22/2011</pre></div>

<!-- crossreference -->
<h2><a name="_cross"></a>CROSS-REFERENCE INFORMATION <a href="#_top"><img alt="^" border="0" src="../../../../../../../../up.png"></a></h2>
This function calls:
<ul style="list-style-image:url(../../../../../../../../matlabicon.gif)">
<li><a href="Metabolism.html" class="code" title="">Metabolism</a>	Metabolism</li></ul>
This function is called by:
<ul style="list-style-image:url(../../../../../../../../matlabicon.gif)">
<li><a href="Metabolism.html" class="code" title="">Metabolism</a>	Metabolism</li></ul>
<!-- crossreference -->

<h2><a name="_subfunctions"></a>SUBFUNCTIONS <a href="#_top"><img alt="^" border="0" src="../../../../../../../../up.png"></a></h2>
<ul style="list-style-image:url(../../../../../../../../matlabicon.gif)">
<li><a href="#_sub1" class="code">function this = Metabolism(wholeCellModelID, name)</a></li><li><a href="#_sub2" class="code">function storeObjectReferences(this, simulation)</a></li><li><a href="#_sub3" class="code">function initializeConstants(this, knowledgeBase, simulation, varargin)</a></li><li><a href="#_sub4" class="code">function [subCmpRemoved, rxnRemoved] = formulateFBA(this, metabolismProduction, unaccountedEnergyConsumption, simplifyNetwork)</a></li><li><a href="#_sub5" class="code">function [bmProd, byProd, minEnzExp, maxEnzExp] = calcResourceRequirements_LifeCycle(this, ~, states)</a></li><li><a href="#_sub6" class="code">function enzymes = fitEnzymes(this, growth0, substrates, enzymes)</a></li><li><a href="#_sub7" class="code">function enzymes = adjustEnzymeExpressionSparsely(this, growth0, substrates, enzymes, initEnzymes)</a></li><li><a href="#_sub8" class="code">function tfs = adjustEnzymeExpressionSparselyHelper(this, growth0, substrates, enzymes, initEnzymes, tfs)</a></li><li><a href="#_sub9" class="code">function [minEnzymes, maxEnzymes] = calcMinMaxEnzymes(this, substrates, enzymes)</a></li><li><a href="#_sub10" class="code">function [forIdxs, revIdxs] = calcEnzymeKineticallyLimitedReactions(this, substrates, enzymes, fbaObj, fbaSMat, fbaRxnBounds)</a></li><li><a href="#_sub11" class="code">function [fbaObj, fbaSMat, fbaRxnBounds] = calcEffectiveFBANetwork(this)</a></li><li><a href="#_sub12" class="code">function initializeState(this)</a></li><li><a href="#_sub13" class="code">function result = calcResourceRequirements_Current(this)</a></li><li><a href="#_sub14" class="code">function evolveState(this)</a></li><li><a href="#_sub15" class="code">function [growth, reactionFluxs, fbaReactionFluxs,</a></li><li><a href="#_sub16" class="code">function bounds = calcFluxBounds(this, substrates, enzymes, fbaReactionBounds, fbaEnzymeBounds,</a></li></ul>

<h2><a name="_source"></a>SOURCE CODE <a href="#_top"><img alt="^" border="0" src="../../../../../../../../up.png"></a></h2>
<div class="fragment"><pre>0001 <span class="comment">%Metabolism</span>
0002 <span class="comment">%</span>
0003 <span class="comment">% @wholeCellModelID Process_Metabolism</span>
0004 <span class="comment">% @name             Metabolism</span>
0005 <span class="comment">% @description</span>
0006 <span class="comment">%   Biology</span>
0007 <span class="comment">%   ===============</span>
0008 <span class="comment">%   To grow and replicate cells must uptake or produce their building blocks and</span>
0009 <span class="comment">%   intermediate energy stores, particularly nucleic and amino acids and lipids,</span>
0010 <span class="comment">%   as well as secrete metabolites such as modified nucleic acids which it</span>
0011 <span class="comment">%   cannot catabolize and extract usable material or energy from. Furthermore,</span>
0012 <span class="comment">%   the import/synthesis and export/breakdown of metabolites must be carefully</span>
0013 <span class="comment">%   matched to the metabolic demands of the cell to ensure that cellular</span>
0014 <span class="comment">%   processes aren't limited by too few metabolites or poisoned by too many</span>
0015 <span class="comment">%   metabolites.</span>
0016 <span class="comment">%</span>
0017 <span class="comment">%   This process simulates</span>
0018 <span class="comment">%   - the uptake of nutrients from the external environment</span>
0019 <span class="comment">%   - the processing of imported nutrients to intermediate energy forms (ATP,</span>
0020 <span class="comment">%     GTP) and macromolecule building blocks (dNTPs, NTPs, and amino acids)</span>
0021 <span class="comment">%   - the assembly of lipids, their insertion into the membrane, and their</span>
0022 <span class="comment">%     maturation within the membrane</span>
0023 <span class="comment">%   - the catabolism of byproducts of macromolecule assembly and degradation</span>
0024 <span class="comment">%   - the export to the external environment of uncatabolizable chemicals such</span>
0025 <span class="comment">%     as modified nucleobases</span>
0026 <span class="comment">%</span>
0027 <span class="comment">%   Knowledge Base</span>
0028 <span class="comment">%   ===============</span>
0029 <span class="comment">%   M. genitalium metabolism was reconstructed from a variety of sources,</span>
0030 <span class="comment">%   including flux-balance analysis metabolic models of other bacterial species</span>
0031 <span class="comment">%   and the reaction kinetics database SABIO-RK, and was organized into 641</span>
0032 <span class="comment">%   reactions in the knowledge base. These reactions are loaded into this process</span>
0033 <span class="comment">%   by the initializeConstants method.</span>
0034 <span class="comment">%</span>
0035 <span class="comment">%      Object       No.</span>
0036 <span class="comment">%      ==========   ===</span>
0037 <span class="comment">%      substrates   580</span>
0038 <span class="comment">%      enzymes      100</span>
0039 <span class="comment">%      reactions    641</span>
0040 <span class="comment">%        chemical   433</span>
0041 <span class="comment">%        transport  208</span>
0042 <span class="comment">%</span>
0043 <span class="comment">%   Representation</span>
0044 <span class="comment">%   ===============</span>
0045 <span class="comment">%   The properties substrates and enzymes represent the counts of metabolites</span>
0046 <span class="comment">%   and metabolic enzymes.</span>
0047 <span class="comment">%</span>
0048 <span class="comment">%   fbaReactionStoichiometryMatrix represents the stoichiometry and compartments</span>
0049 <span class="comment">%   of metabolites and biomass in each of the 641 chemical/transport reactions,</span>
0050 <span class="comment">%   exchange pseudoreactions, and biomass production pseudoreaction.</span>
0051 <span class="comment">%   fbaReactionCatalysisMatrix represents the enzyme which catalyzes each</span>
0052 <span class="comment">%   reaction. fbaEnzymeBounds represents the foward and backward kcat of the</span>
0053 <span class="comment">%   catalyzing enzyme of each reaction. fbaReactionBounds represents the maximal</span>
0054 <span class="comment">%   import and export rates of each  metabolite. fbaObjective indicates which</span>
0055 <span class="comment">%   reaction represents the biomass production pseudoreaction. fbaRightHandSide</span>
0056 <span class="comment">%   is a vector of zeros representing the change in concentration over time of</span>
0057 <span class="comment">%   each metabolite and biomass. metabolismProduction is redundant with the</span>
0058 <span class="comment">%   biomass production reaction in fbaReactionStoichiometryMatrix.</span>
0059 <span class="comment">%   metabolismProduction is calculated by summing the metabolic demands of all</span>
0060 <span class="comment">%   the other processes over the entire cell cycle. The table below lists the</span>
0061 <span class="comment">%   units of several properties of this process.</span>
0062 <span class="comment">%</span>
0063 <span class="comment">%      Property                       Units</span>
0064 <span class="comment">%      ===========================    ==============================</span>
0065 <span class="comment">%      fbaEnzymeBounds                molecules/enzyme/s</span>
0066 <span class="comment">%      fbaReactionBounds              molecules/(gram dry biomass)/s</span>
0067 <span class="comment">%      metabolites                    molecules</span>
0068 <span class="comment">%      enzymes                        molecules</span>
0069 <span class="comment">%      stepSizeSec                    s</span>
0070 <span class="comment">%      lowerBounds                    reactions/s</span>
0071 <span class="comment">%      upperBounds                    reactions/s</span>
0072 <span class="comment">%      growth                         cell/s</span>
0073 <span class="comment">%      biomassComposition             molecules/cell</span>
0074 <span class="comment">%      metabolismProduction           molecules/cell</span>
0075 <span class="comment">%      chamberVolume                  L</span>
0076 <span class="comment">%      setValues                      molecules/chamber</span>
0077 <span class="comment">%      growthAssociatedMaintanence</span>
0078 <span class="comment">%</span>
0079 <span class="comment">%   Initialization</span>
0080 <span class="comment">%   ===============</span>
0081 <span class="comment">%   The simulation is initialized with 1 cell weight of macromolecules and</span>
0082 <span class="comment">%   water, and few free metabolites by the simulation class' initializeState</span>
0083 <span class="comment">%   method. In addition this process is initialized to a positive growth rate of</span>
0084 <span class="comment">%   approximately log(2)/cellCycleLength computing using the flux-balance</span>
0085 <span class="comment">%   analysis model implemented in evolveState.</span>
0086 <span class="comment">%</span>
0087 <span class="comment">%   Simulation</span>
0088 <span class="comment">%   ===============</span>
0089 <span class="comment">%   Transport and metabolism are modeled using the constraint-based method</span>
0090 <span class="comment">%   flux-balance analysis (FBA). Briefly, FBA assumes that bacteria have evolved</span>
0091 <span class="comment">%   to maximize growth, and poses transport and metabolism as the optimization</span>
0092 <span class="comment">%   of cellular building block and energy production subject to available</span>
0093 <span class="comment">%   nutrients, and allowed chemical reactions. FBA models are typically</span>
0094 <span class="comment">%   constrained by experimentally measured transport and diffusion rates. In</span>
0095 <span class="comment">%   addition to these constraints, we constrain constrain our FBA process by our</span>
0096 <span class="comment">%   model's predicted enzyme abundances, and by experimentally measured kinetic</span>
0097 <span class="comment">%   parameters. These additional constraints yield a more accurate model of</span>
0098 <span class="comment">%   transport and metabolism. Furthermore, we use our other processes to obtain a</span>
0099 <span class="comment">%   more extensive and accurate metabolic objective than used in earlier FBA</span>
0100 <span class="comment">%   models. The optimization problem specific by FBA is posed as a linear</span>
0101 <span class="comment">%   optimization problem, and solving using one of several publically available</span>
0102 <span class="comment">%   linear programming packages.</span>
0103 <span class="comment">%</span>
0104 <span class="comment">%   Algorithm</span>
0105 <span class="comment">%   ++++++++++++++++</span>
0106 <span class="comment">%   1. Compute reaction bounds based on</span>
0107 <span class="comment">%     - enzyme kinetics,</span>
0108 <span class="comment">%     - enzyme availability</span>
0109 <span class="comment">%     - maximal metabolite exchange rates</span>
0110 <span class="comment">%     - external metabolite availability</span>
0111 <span class="comment">%     - protein availability</span>
0112 <span class="comment">%   2. Computes optimal reaction fluxes which maximize biomass production</span>
0113 <span class="comment">%   3. Computes integer-valued production of biomass components, and export of</span>
0114 <span class="comment">%      byproducts</span>
0115 <span class="comment">%   4. Updates amounts of biomass components and byproducts</span>
0116 <span class="comment">%</span>
0117 <span class="comment">%   References</span>
0118 <span class="comment">%   ===============</span>
0119 <span class="comment">%   1. Orth JD, Thiele I, Palsson BO (2010). What is flux balance analysis?</span>
0120 <span class="comment">%      Nat Biotechnol. 28(3):245-8 [PUB_0687].</span>
0121 <span class="comment">%   2. Thiele I, Palsson BO (2010). A protocol for generating a</span>
0122 <span class="comment">%      high-quality genome-scale metabolic reconstruction. Nat Protoc.</span>
0123 <span class="comment">%      5(1): 93-121. [PUB_0686].</span>
0124 <span class="comment">%   3. Covert MW, Xiao N, Chen TJ, Karr JR (2008). Integrating metabolic,</span>
0125 <span class="comment">%      transcriptional regulatory and signal transduction models in</span>
0126 <span class="comment">%      Escherichia coli. Bioinformatics. 24(18):2044-50. [PUB_0684].</span>
0127 <span class="comment">%   4. Covert MW, Knight EM, Reed JL, Herrgard MJ, Palsson BO (2004).</span>
0128 <span class="comment">%      Integrating high-throughput and computational data elucidates</span>
0129 <span class="comment">%      bacterial networks. Nature. 429 (6987): 92-6. [PUB_0618].</span>
0130 <span class="comment">%   5. Covert MW, Palsson BO (2003). Constraints-based models: regulation</span>
0131 <span class="comment">%      of gene expression reduces the steady-state solution space. J Theor</span>
0132 <span class="comment">%      Biol. 221(3): 309-25. [PUB_0685].</span>
0133 <span class="comment">%</span>
0134 <span class="comment">% Author: Jonathan Karr, jkarr@stanford.edu</span>
0135 <span class="comment">% Author: Markus Covert, mcovert@stanford.edu</span>
0136 <span class="comment">% Author: Jayodita Sanghvi, jayodita@stanford.edu</span>
0137 <span class="comment">% Affilitation: Covert Lab, Department of Bioengineering, Stanford University</span>
0138 <span class="comment">% Last updated: 3/22/2011</span>
0139 classdef  <a href="Metabolism.html" class="code" title="">Metabolism</a> &lt; edu.stanford.covert.cell.sim.ReactionProcess
0140     <span class="comment">%property annotations</span>
0141     properties (Constant)
0142         optionNames__              = {   <span class="comment">%names of option properties</span>
0143             <span class="string">'linearProgrammingOptions'</span>;
0144             <span class="string">'macromoleculeStateInitialization'</span>;
0145             <span class="string">'tolerance'</span>
0146             <span class="string">'realmax'</span>
0147             };
0148         fixedConstantNames__       = {   <span class="comment">%names of fixed constant properties</span>
0149             <span class="string">'growthAssociatedMaintenance'</span>;
0150             <span class="string">'nonGrowthAssociatedMaintenance'</span>;
0151             <span class="string">'cellCycleLength'</span>;
0152             <span class="string">'exchangeRateUpperBound_carbon'</span>;
0153             <span class="string">'exchangeRateUpperBound_noncarbon'</span>;
0154             <span class="string">'fbaObjective'</span>;
0155             <span class="string">'fbaRightHandSide'</span>;
0156             <span class="string">'fbaReactionStoichiometryMatrix'</span>;
0157             <span class="string">'fbaReactionCatalysisMatrix'</span>;
0158             <span class="string">'fbaReactionBounds'</span>;
0159             <span class="string">'fbaEnzymeBounds'</span>;
0160             <span class="string">'metabolismProduction'</span>;
0161             <span class="string">'monomerCompartments'</span>;
0162             };
0163         fittedConstantNames__      = {   <span class="comment">%names of fitted constant properties</span>
0164             <span class="string">'macromoleculeStateInitializationGrowthFactor'</span>
0165             <span class="string">'unaccountedEnergyConsumption'</span>;
0166             };
0167         localStateNames__          = {}; <span class="comment">%names of simulation state properties redundant with timecourses in this or other processes or the simulation</span>
0168     <span class="keyword">end</span>
0169     
0170     <span class="comment">%options</span>
0171     properties
0172         <span class="comment">%linear programming</span>
0173         linearProgrammingOptions = struct(<span class="keyword">...</span>
0174             <span class="string">'solver'</span>, <span class="string">'glpk'</span>,<span class="keyword">...</span>
0175             <span class="string">'solverOptions'</span>, struct(<span class="keyword">...</span>
0176                 <span class="string">'glpk'</span>, struct(<span class="string">'lpsolver'</span>, 1, <span class="string">'presol'</span>, 1, <span class="string">'scale'</span>, 1, <span class="string">'msglev'</span>, 0, <span class="string">'tolbnd'</span>, 10e-7),<span class="keyword">...</span>
0177                 <span class="string">'linprog'</span>, struct(<span class="string">'Display'</span>,<span class="string">'off'</span>),<span class="keyword">...</span>
0178                 <span class="string">'lp_solve'</span>, struct(<span class="string">'verbose'</span>, 0, <span class="string">'scaling'</span>, 3 + 64 + 128, <span class="string">'presolve'</span>, 0), <span class="keyword">...</span>
0179                 <span class="string">'qsopt'</span>, struct()));
0180         
0181         <span class="comment">%initial macromolecule state distribution</span>
0182         macromoleculeStateInitialization = <span class="string">'multinomial'</span>;
0183         
0184         <span class="comment">%fitting tolerance</span>
0185         tolerance = 1e-3;
0186         
0187         <span class="comment">%realmax</span>
0188         <span class="comment">% Set to several orders of magnitude higher than non-cycle highest</span>
0189         <span class="comment">% fluxs. In tests gplk will tolerates as high as 10^9. Higher than</span>
0190         <span class="comment">% 10^9 glpk and other solvers have trouble finding feasible</span>
0191         <span class="comment">% solutions.</span>
0192         realmax = 1e6;
0193     <span class="keyword">end</span>
0194     
0195     <span class="comment">%IDs, names, and local indices</span>
0196     properties
0197         substrateIndexs_amp                   <span class="comment">%index of AMP within reactionStoichiometryMatrix</span>
0198         substrateIndexs_adp                   <span class="comment">%index of ADP within reactionStoichiometryMatrix</span>
0199         substrateIndexs_atp                   <span class="comment">%index of ATP within reactionStoichiometryMatrix</span>
0200         substrateIndexs_diphosphate           <span class="comment">%index of PPi within reactionStoichiometryMatrix</span>
0201         substrateIndexs_phosphate             <span class="comment">%index of Pi  within reactionStoichiometryMatrix</span>
0202         substrateIndexs_water                 <span class="comment">%index of H2O within reactionStoichiometryMatrix</span>
0203         substrateIndexs_hydrogen              <span class="comment">%index of H   within reactionStoichiometryMatrix</span>
0204         substrateIndexs_atpHydrolysis         <span class="comment">%index of ATP, H2O, ADP, PI, H within reactionStoichiometryMatrix</span>
0205         substrateIndexs_energy                <span class="comment">%index of NxPs, PPI, PI, H within reactionStoichiometryMatrix</span>
0206         
0207         substrateIndexs_limitableProteins     <span class="comment">%indices of limitable proteins within reactionStoichiometryMatrix</span>
0208         substrateIndexs_fba                   <span class="comment">%indices of FBA substrates within reactionStoichiometryMatrix</span>
0209         substrateIndexs_externalExchangedMetabolites  <span class="comment">%indices of exchanged metabolites within reactionStoichiometryMatrix</span>
0210         substrateIndexs_internalExchangedMetabolites  <span class="comment">%indices of exchanged metabolites within reactionStoichiometryMatrix</span>
0211         substrateIndexs_internalExchangedLimitedMetabolites <span class="comment">%indices of limited, exchanged metabolites within reactionStoichiometryMatrix</span>
0212         
0213         fbaSubstrateIndexs_substrates         <span class="comment">%indices of substrates (metabolites, RNAs, proteins) within fbaReactionStoichiometryMatrix</span>
0214         fbaSubstrateIndexs_metaboliteInternalExchangeConstraints <span class="comment">%indices of substrates with are internally exchanged</span>
0215         fbaSubstrateIndexs_biomass            <span class="comment">%index of biomass within fbaReactionStoichiometryMatrix</span>
0216         
0217         reactionIndexs_chemical               <span class="comment">%indices of chemical reactions within reactionStoichiometryMatrix</span>
0218         reactionIndexs_transport              <span class="comment">%indices of transport reactions within reactionStoichiometryMatrix</span>
0219         reactionIndexs_fba                    <span class="comment">%indices of FBA reactions within reactionStoichiometryMatrix</span>
0220         
0221         fbaReactionIndexs_metabolicConversion <span class="comment">%indices of metabolic reactions (chemical, transport) within fbaReactionStoichiometryMatrix</span>
0222         fbaReactionIndexs_metaboliteExternalExchange <span class="comment">%indices of exchange reactions within fbaReactionStoichiometryMatrix</span>
0223         fbaReactionIndexs_metaboliteInternalExchange <span class="comment">%indices of exchange reactions within fbaReactionStoichiometryMatrix</span>
0224         fbaReactionIndexs_metaboliteInternalLimitedExchange <span class="comment">%indices of exchange reactions within fbaReactionStoichiometryMatrix</span>
0225         fbaReactionIndexs_metaboliteInternalUnlimitedExchange <span class="comment">%indices of exchange reactions within fbaReactionStoichiometryMatrix</span>
0226         fbaReactionIndexs_biomassProduction   <span class="comment">%index of biomass production reaction within fbaReactionStoichiometryMatrix</span>
0227         fbaReactionIndexs_biomassExchange     <span class="comment">%index of biomass efflux reaction within fbaReactionStoichiometryMatrix</span>
0228         
0229         compartmentIndexs_cytosol             <span class="comment">%index of cytosol compartment within reactionStoichiometryMatrix</span>
0230         compartmentIndexs_extracellular       <span class="comment">%index of extracellular compartment within reactionStoichiometryMatrix</span>
0231         compartmentIndexs_membrane            <span class="comment">%index of membrane compartment within reactionStoichiometryMatrix</span>
0232     <span class="keyword">end</span>
0233     
0234     <span class="comment">%fixed biological constants</span>
0235     properties
0236         macromoleculeStateInitializationGrowthFactor <span class="comment">%set by running fitting once and seeing what mean growth rate is predicted; see fitEnzymes</span>
0237         
0238         growthAssociatedMaintenance         <span class="comment">%mmol ATP/gDCW [59.810; PUB_0558]</span>
0239         nonGrowthAssociatedMaintenance      <span class="comment">%mmol ATP/gDCW [8.390; PUB_0558]</span>
0240         cellCycleLength                     <span class="comment">%cell cycle length (s)</span>
0241         exchangeRateUpperBound_carbon       <span class="comment">%exchange rate upper bound -- carbon-containing metabolites (mmol/gDCW/h) [12]</span>
0242         exchangeRateUpperBound_noncarbon    <span class="comment">%exchange rate upper bound -- non-carbon-containing metabolites (mmol/gDCW/h) [20]</span>
0243         substrateExternalExchangeBounds     <span class="comment">%substrate exchange bounds                                                          [substrates X 2]</span>
0244         substrateExchangeBounds             <span class="comment">%substrate exchange bounds                                                          [substrates X 2]</span>
0245         fbaObjective                        <span class="comment">%FBA objective function corresponding to maximize growth                            [reactions;growth;biomass efflux] X 1</span>
0246         fbaRightHandSide                    <span class="comment">%right-hand side of equation Sv=0                                                   [substrates;biomass] X 1</span>
0247         fbaReactionStoichiometryMatrix      <span class="comment">%FBA stoichiometry matrix                                                           [substrates;biomass] X [reactions;growth;biomass efflux] X compartments</span>
0248         fbaReactionCatalysisMatrix          <span class="comment">%Enzymes catalyzing each FBA reaction                                               [reactions;growth;biomass efflux] X enzymes</span>
0249         fbaReactionBounds                   <span class="comment">%Exchange bounds associated with each FBA reaction (molecules/gram dry biomass/s)   [reactions;growth;biomass efflux] X 2</span>
0250         fbaEnzymeBounds                     <span class="comment">%Kinetics bounds associated with each FBA reaction (molecules/enzyme/s)             [reactions;growth;biomass efflux] X 2</span>
0251         metabolismProduction                <span class="comment">%biomass composition + energy + byproducts (molecules/cell)                         substrates X compartments</span>
0252         metabolismNewProduction             <span class="comment">%metabolism output represented by biomass pseudoreaction</span>
0253         metabolismRecyclingProduction       <span class="comment">%metabolism output represented by internal exchange reactions</span>
0254         unaccountedEnergyConsumption        <span class="comment">%unaccounted energy consumption (ATP molecules / cell life cycle)                   ATP X 1</span>
0255         monomerCompartments                 <span class="comment">%Indexs of compartments of protein monomers</span>
0256         proteinLimitableProteinComposition  <span class="comment">%protein complex composition                                                        [monomers; complexs] X [monomers; limitable complexs]</span>
0257     <span class="keyword">end</span>
0258     
0259     <span class="comment">%global state (referenced locally for convenience)</span>
0260     properties
0261         mass                      <span class="comment">%cell mass</span>
0262         metabolicReaction         <span class="comment">%metabolic flux</span>
0263     <span class="keyword">end</span>
0264     
0265     <span class="comment">%constructor</span>
0266     methods
0267         <a name="_sub0" href="#_subfunctions" class="code">function this = Metabolism(wholeCellModelID, name)</a>
0268             this = this@edu.stanford.covert.cell.sim.ReactionProcess(wholeCellModelID, name);
0269         <span class="keyword">end</span>
0270     <span class="keyword">end</span>
0271     
0272     <span class="comment">%communication between process/simulation</span>
0273     methods
0274         <span class="comment">%set references to state objects</span>
0275         <a name="_sub1" href="#_subfunctions" class="code">function storeObjectReferences(this, simulation)</a>
0276             this.storeObjectReferences@edu.stanford.covert.cell.sim.ReactionProcess(simulation);
0277             
0278             this.mass = simulation.state(<span class="string">'Mass'</span>);
0279             this.metabolicReaction = simulation.state(<span class="string">'MetabolicReaction'</span>);
0280             this.states = [this.states; {this.mass; this.metabolicReaction}];
0281         <span class="keyword">end</span>
0282         
0283         <span class="comment">%initialize constants</span>
0284         <a name="_sub2" href="#_subfunctions" class="code">function initializeConstants(this, knowledgeBase, simulation, varargin)</a>
0285             import edu.stanford.covert.util.ConstantUtil;
0286             
0287             <span class="comment">%% options</span>
0288             options = struct(<span class="string">'retainSubstrateCompartments'</span>, true);
0289             
0290             <span class="comment">%% call super class method</span>
0291             this.initializeConstants@edu.stanford.covert.cell.sim.ReactionProcess(<span class="keyword">...</span>
0292                 knowledgeBase, simulation, options);
0293             
0294             <span class="comment">%% cell cycle length</span>
0295             this.cellCycleLength = simulation.state(<span class="string">'Time'</span>).cellCycleLength;
0296             
0297             <span class="comment">%% compartments</span>
0298             this.compartmentIndexs_cytosol       = 1;
0299             this.compartmentIndexs_extracellular = 2;
0300             this.compartmentIndexs_membrane      = 3;
0301             
0302             compartmentIndexs = [
0303                 this.compartment.cytosolIndexs;
0304                 this.compartment.extracellularIndexs;
0305                 this.compartment.membraneIndexs];
0306             
0307             this.substrateStimulusGlobalIndexs        = this.substrateStimulusGlobalIndexs(  :, compartmentIndexs);
0308             this.substrateMetaboliteGlobalIndexs      = this.substrateMetaboliteGlobalIndexs(:, compartmentIndexs);
0309             this.substrateRNAGlobalIndexs             = this.substrateRNAGlobalIndexs(       :, compartmentIndexs);
0310             this.substrateMonomerGlobalIndexs         = this.substrateMonomerGlobalIndexs(   :, compartmentIndexs);
0311             this.substrateComplexGlobalIndexs         = this.substrateComplexGlobalIndexs(   :, compartmentIndexs);
0312             
0313             this.substrateStimulusCompartmentIndexs   = this.substrateStimulusCompartmentIndexs(  :, compartmentIndexs);
0314             this.substrateMetaboliteCompartmentIndexs = this.substrateMetaboliteCompartmentIndexs(:, compartmentIndexs);
0315             this.substrateRNACompartmentIndexs        = this.substrateRNACompartmentIndexs(       :, compartmentIndexs);
0316             this.substrateMonomerCompartmentIndexs    = this.substrateMonomerCompartmentIndexs(   :, compartmentIndexs);
0317             this.substrateComplexCompartmentIndexs    = this.substrateComplexCompartmentIndexs(   :, compartmentIndexs);
0318             
0319             this.substrateStimulusGlobalCompartmentIndexs = sub2ind(<span class="keyword">...</span>
0320                 [numel(this.stimulus.wholeCellModelIDs) this.compartment.count],<span class="keyword">...</span>
0321                 this.substrateStimulusGlobalIndexs,<span class="keyword">...</span>
0322                 this.substrateStimulusCompartmentIndexs);
0323             this.substrateMetaboliteGlobalCompartmentIndexs = sub2ind(<span class="keyword">...</span>
0324                 [numel(this.metabolite.wholeCellModelIDs) this.compartment.count],<span class="keyword">...</span>
0325                 this.substrateMetaboliteGlobalIndexs,<span class="keyword">...</span>
0326                 this.substrateMetaboliteCompartmentIndexs);
0327             this.substrateRNAGlobalCompartmentIndexs = sub2ind(<span class="keyword">...</span>
0328                 [numel(this.rna.wholeCellModelIDs) this.compartment.count],<span class="keyword">...</span>
0329                 this.rna.matureIndexs(this.substrateRNAGlobalIndexs),<span class="keyword">...</span>
0330                 this.substrateRNACompartmentIndexs);
0331             this.substrateMonomerGlobalCompartmentIndexs = sub2ind(<span class="keyword">...</span>
0332                 [numel(this.monomer.wholeCellModelIDs) this.compartment.count],<span class="keyword">...</span>
0333                 this.monomer.matureIndexs(this.substrateMonomerGlobalIndexs),<span class="keyword">...</span>
0334                 this.substrateMonomerCompartmentIndexs);
0335             this.substrateComplexGlobalCompartmentIndexs = sub2ind(<span class="keyword">...</span>
0336                 [numel(this.complex.wholeCellModelIDs) this.compartment.count],<span class="keyword">...</span>
0337                 this.complex.matureIndexs(this.substrateComplexGlobalIndexs),<span class="keyword">...</span>
0338                 this.substrateComplexCompartmentIndexs);
0339             
0340             this.reactionStoichiometryMatrix = this.reactionStoichiometryMatrix(:, :, compartmentIndexs);
0341             this.reactionCoenzymeMatrix = this.reactionCoenzymeMatrix(:, :, compartmentIndexs);
0342             
0343             <span class="comment">%% metabolites</span>
0344             this.substrateIndexs_amp           = this.substrateIndexs({<span class="string">'AMP'</span>});
0345             this.substrateIndexs_adp           = this.substrateIndexs({<span class="string">'ADP'</span>});
0346             this.substrateIndexs_atp           = this.substrateIndexs({<span class="string">'ATP'</span>});
0347             this.substrateIndexs_diphosphate   = this.substrateIndexs({<span class="string">'PPI'</span>});
0348             this.substrateIndexs_phosphate     = this.substrateIndexs({<span class="string">'PI'</span>});
0349             this.substrateIndexs_water         = this.substrateIndexs({<span class="string">'H2O'</span>});
0350             this.substrateIndexs_hydrogen      = this.substrateIndexs({<span class="string">'H'</span>});
0351             this.substrateIndexs_atpHydrolysis = this.substrateIndexs({<span class="string">'ATP'</span>; <span class="string">'H2O'</span>; <span class="string">'ADP'</span>; <span class="string">'PI'</span>; <span class="string">'H'</span>});
0352             this.substrateIndexs_energy        = sort(this.substrateIndexs({<span class="string">'AMP'</span>, <span class="string">'CMP'</span>, <span class="string">'GMP'</span>, <span class="string">'UMP'</span>, <span class="string">'ADP'</span>, <span class="string">'CDP'</span>, <span class="string">'GDP'</span>, <span class="string">'UDP'</span>, <span class="string">'ATP'</span>, <span class="string">'CTP'</span>, <span class="string">'GTP'</span>, <span class="string">'UTP'</span>, <span class="string">'PI'</span>, <span class="string">'PPI'</span>, <span class="string">'H'</span>}));
0353             
0354             <span class="comment">%% reactions</span>
0355             this.reactionIndexs_chemical  = find(strcmp(this.reactionTypes, <span class="string">'chemical'</span>));
0356             this.reactionIndexs_transport = find(strcmp(this.reactionTypes, <span class="string">'transport'</span>));
0357             
0358             <span class="comment">%% FBA</span>
0359             <span class="comment">%comlexes</span>
0360             complexFormationProcesses = [knowledgeBase.proteinComplexs(this.substrateComplexGlobalIndexs(:, 1)).complexFormationProcess];
0361             this.substrateIndexs_limitableProteins = [<span class="keyword">...</span>
0362                 this.substrateMonomerLocalIndexs
0363                 this.substrateComplexLocalIndexs(strcmp({complexFormationProcesses.wholeCellModelID}, <span class="string">'Process_MacromolecularComplexation'</span>), :)];
0364             
0365             substratesByLimitableSubstrates = zeros(numel(this.substrateWholeCellModelIDs), numel(this.substrateWholeCellModelIDs));
0366             substratesByLimitableSubstrates(this.substrateComplexLocalIndexs, this.substrateMonomerLocalIndexs) = <span class="keyword">...</span>
0367                 any(knowledgeBase.proteinComplexMonomerComposition(this.gene.mRNAIndexs(this.substrateMonomerGlobalIndexs(:, 1)), this.substrateComplexGlobalIndexs(:, 1), :), 3)';
0368             substratesByLimitableSubstrates(this.substrateComplexLocalIndexs, this.substrateComplexLocalIndexs) =  <span class="keyword">...</span>
0369                 any(knowledgeBase.proteinComplexComplexComposition(this.substrateComplexGlobalIndexs(:, 1), this.substrateComplexGlobalIndexs(:, 1), :), 3)';
0370             substratesByLimitableSubstrates(:, setdiff(1:<span class="keyword">end</span>, this.substrateIndexs_limitableProteins)) = 0;
0371             substratesByLimitableSubstrates(this.substrateIndexs_limitableProteins, :) = 0;
0372             substratesByLimitableSubstrates(sub2ind(<span class="keyword">...</span>
0373                 size(substratesByLimitableSubstrates), <span class="keyword">...</span><span class="comment">.</span>
0374                 this.substrateIndexs_limitableProteins, <span class="keyword">...</span>
0375                 this.substrateIndexs_limitableProteins)) = 1;
0376             
0377             this.proteinLimitableProteinComposition = substratesByLimitableSubstrates([this.substrateMonomerLocalIndexs; this.substrateComplexLocalIndexs], this.substrateIndexs_limitableProteins);
0378             
0379             <span class="comment">%exchange bounds</span>
0380             mets = knowledgeBase.metabolites(this.substrateMetaboliteGlobalIndexs(:, 1));
0381             this.substrateExternalExchangeBounds = zeros(numel(this.substrateWholeCellModelIDs), 2);
0382             <span class="keyword">for</span> i = 1:numel(mets)
0383                 met = mets(i);
0384                 <span class="keyword">if</span> isfield(met.empiricalFormula, <span class="string">'C'</span>) &amp;&amp; met.empiricalFormula.C &gt; 0
0385                     this.substrateExternalExchangeBounds(this.substrateMetaboliteLocalIndexs(i), 1) = max(-this.exchangeRateUpperBound_carbon, met.exchangeLowerBound);
0386                     this.substrateExternalExchangeBounds(this.substrateMetaboliteLocalIndexs(i), 2) = min( this.exchangeRateUpperBound_carbon, met.exchangeUpperBound);
0387                 <span class="keyword">else</span>
0388                     this.substrateExternalExchangeBounds(this.substrateMetaboliteLocalIndexs(i), 1) = max(-this.exchangeRateUpperBound_noncarbon, met.exchangeLowerBound);
0389                     this.substrateExternalExchangeBounds(this.substrateMetaboliteLocalIndexs(i), 2) = min( this.exchangeRateUpperBound_noncarbon, met.exchangeUpperBound);
0390                 <span class="keyword">end</span>
0391             <span class="keyword">end</span>
0392             this.substrateExternalExchangeBounds = this.substrateExternalExchangeBounds / <span class="keyword">...</span>
0393                 ConstantUtil.secondsPerHour * (ConstantUtil.nAvogadro / 1000);
0394             
0395             <span class="comment">%% Monomers</span>
0396             this.monomerCompartments = knowledgeBase.proteinMonomerCompartments;
0397         <span class="keyword">end</span>
0398         
0399         <span class="comment">%Formulate FBA model for current biomass composition stored in</span>
0400         <span class="comment">%metabolite state and this module's reactions</span>
0401         <span class="comment">%- objective</span>
0402         <span class="comment">%- stoichiometry matrix</span>
0403         <span class="comment">%- bounds</span>
0404         <span class="comment">%</span>
0405         <span class="comment">%To decrease linear programming run-time and increase linear</span>
0406         <span class="comment">%programming accuracy this method automatically reduces the size of</span>
0407         <span class="comment">%the linear programming problem by removing</span>
0408         <span class="comment">%- reaction and substrates disconnected from biomass</span>
0409         <span class="comment">%- leaf metabolites (and reactions involving them) which cannot be</span>
0410         <span class="comment">%  balanced with non-zero fluxes</span>
0411         <span class="comment">%- metabolites (and reactions involving them) which can't move in</span>
0412         <span class="comment">%  both the forward and backward directions and thus cannot be</span>
0413         <span class="comment">%  balanced</span>
0414         <span class="comment">%- other reactions (and substrates unique to them) whose flux is</span>
0415         <span class="comment">%  zero in all solutions of the matrix problem Sv=0. That is</span>
0416         <span class="comment">%  reactions which have only zero components in the null space of</span>
0417         <span class="comment">%  S.</span>
0418         <a name="_sub3" href="#_subfunctions" class="code">function [subCmpRemoved, rxnRemoved] = formulateFBA(this, metabolismProduction, unaccountedEnergyConsumption, simplifyNetwork)</a>
0419             import edu.stanford.covert.util.ComputationUtil;
0420             import edu.stanford.covert.util.ConstantUtil;
0421             import edu.stanford.covert.util.findNonInteractingRowsAndColumns;
0422             
0423             <span class="comment">%% metabolism output</span>
0424             <span class="keyword">if</span> nargin &gt;= 2 &amp;&amp; ~isempty(metabolismProduction)
0425                 metabolismProduction(abs(metabolismProduction) &lt; 1e-4) = 0; <span class="comment">%zero out metabolites that almost zero up to rounding error</span>
0426                 this.metabolismProduction = zeros(numel(this.substrateWholeCellModelIDs), size(this.substrateMetaboliteGlobalIndexs, 2));
0427                 this.metabolismProduction(this.substrateMetaboliteLocalIndexs, :) = metabolismProduction(sub2ind(<span class="keyword">...</span>
0428                     size(metabolismProduction), <span class="keyword">...</span>
0429                     this.substrateMetaboliteGlobalIndexs, <span class="keyword">...</span>
0430                     this.substrateMetaboliteCompartmentIndexs));
0431             <span class="keyword">end</span>
0432             <span class="keyword">if</span> nargin &gt;= 3 &amp;&amp; ~isempty(unaccountedEnergyConsumption)
0433                 this.unaccountedEnergyConsumption = unaccountedEnergyConsumption;
0434             <span class="keyword">end</span>
0435             assert(~any(sum(this.metabolismProduction ~= 0, 2) &gt; 1));
0436             
0437             this.metabolismNewProduction = max(0, this.metabolismProduction);
0438             this.metabolismRecyclingProduction = min(0, this.metabolismProduction);
0439             
0440             this.metabolismRecyclingProduction(this.substrateIndexs({<span class="string">'ATP'</span>; <span class="string">'CTP'</span>; <span class="string">'GTP'</span>; <span class="string">'UTP'</span>}), :) = <span class="keyword">...</span>
0441                 + this.metabolismRecyclingProduction(this.substrateIndexs({<span class="string">'ATP'</span>; <span class="string">'CTP'</span>; <span class="string">'GTP'</span>; <span class="string">'UTP'</span>}), :) <span class="keyword">...</span>
0442                 - this.metabolismRecyclingProduction(this.substrateIndexs({<span class="string">'ADP'</span>; <span class="string">'CDP'</span>; <span class="string">'GDP'</span>; <span class="string">'UDP'</span>}), :) <span class="keyword">...</span>
0443                 - this.metabolismRecyclingProduction(this.substrateIndexs({<span class="string">'AMP'</span>; <span class="string">'CMP'</span>; <span class="string">'GMP'</span>; <span class="string">'UMP'</span>}), :);
0444             this.metabolismNewProduction(this.substrateIndexs({<span class="string">'ATP'</span>; <span class="string">'CTP'</span>; <span class="string">'GTP'</span>; <span class="string">'UTP'</span>}), :) = <span class="keyword">...</span>
0445                 + this.metabolismNewProduction(this.substrateIndexs({<span class="string">'ATP'</span>; <span class="string">'CTP'</span>; <span class="string">'GTP'</span>; <span class="string">'UTP'</span>}), :) <span class="keyword">...</span>
0446                 + this.metabolismRecyclingProduction(this.substrateIndexs({<span class="string">'ADP'</span>; <span class="string">'CDP'</span>; <span class="string">'GDP'</span>; <span class="string">'UDP'</span>}), :) <span class="keyword">...</span>
0447                 + this.metabolismRecyclingProduction(this.substrateIndexs({<span class="string">'AMP'</span>; <span class="string">'CMP'</span>; <span class="string">'GMP'</span>; <span class="string">'UMP'</span>}), :);
0448             
0449             this.metabolismRecyclingProduction(this.substrateIndexs({<span class="string">'H2O'</span>}), :) = <span class="keyword">...</span>
0450                 + this.metabolismRecyclingProduction(this.substrateIndexs({<span class="string">'H2O'</span>}), :) <span class="keyword">...</span>
0451                 - sum(this.metabolismRecyclingProduction(this.substrateIndexs({<span class="string">'ADP'</span>; <span class="string">'CDP'</span>; <span class="string">'GDP'</span>; <span class="string">'UDP'</span>}), :)) <span class="keyword">...</span>
0452                 - sum(this.metabolismRecyclingProduction(this.substrateIndexs({<span class="string">'AMP'</span>; <span class="string">'CMP'</span>; <span class="string">'GMP'</span>; <span class="string">'UMP'</span>}), :));
0453             this.metabolismNewProduction(this.substrateIndexs({<span class="string">'H2O'</span>}), :) = <span class="keyword">...</span>
0454                 + this.metabolismNewProduction(this.substrateIndexs({<span class="string">'H2O'</span>}), :) <span class="keyword">...</span>
0455                 + sum(this.metabolismRecyclingProduction(this.substrateIndexs({<span class="string">'ADP'</span>; <span class="string">'CDP'</span>; <span class="string">'GDP'</span>; <span class="string">'UDP'</span>}), :)) <span class="keyword">...</span>
0456                 + sum(this.metabolismRecyclingProduction(this.substrateIndexs({<span class="string">'AMP'</span>; <span class="string">'CMP'</span>; <span class="string">'GMP'</span>; <span class="string">'UMP'</span>}), :));
0457             
0458             this.metabolismRecyclingProduction(this.substrateIndexs({<span class="string">'ATP'</span>; <span class="string">'ADP'</span>; <span class="string">'PI'</span>; <span class="string">'H2O'</span>; <span class="string">'H'</span>}), this.compartmentIndexs_cytosol) = <span class="keyword">...</span>
0459                 + this.metabolismRecyclingProduction(this.substrateIndexs({<span class="string">'ATP'</span>; <span class="string">'ADP'</span>; <span class="string">'PI'</span>; <span class="string">'H2O'</span>; <span class="string">'H'</span>}), this.compartmentIndexs_cytosol) <span class="keyword">...</span>
0460                 + this.unaccountedEnergyConsumption * [-1; 1; 1; -1; 1];
0461             this.metabolismNewProduction(this.substrateIndexs({<span class="string">'ATP'</span>; <span class="string">'ADP'</span>; <span class="string">'PI'</span>; <span class="string">'H2O'</span>; <span class="string">'H'</span>}), this.compartmentIndexs_cytosol) = <span class="keyword">...</span>
0462                 + this.metabolismNewProduction(this.substrateIndexs({<span class="string">'ATP'</span>; <span class="string">'ADP'</span>; <span class="string">'PI'</span>; <span class="string">'H2O'</span>; <span class="string">'H'</span>}), this.compartmentIndexs_cytosol) <span class="keyword">...</span>
0463                 - this.unaccountedEnergyConsumption * [-1; 1; 1; -1; 1];
0464             
0465             this.metabolismNewProduction(this.substrateIndexs({<span class="string">'MET'</span>; <span class="string">'H'</span>; <span class="string">'H2O'</span>}), this.compartmentIndexs_cytosol) = <span class="keyword">...</span>
0466                 + this.metabolismNewProduction(this.substrateIndexs({<span class="string">'MET'</span>; <span class="string">'H'</span>; <span class="string">'H2O'</span>}), this.compartmentIndexs_cytosol) <span class="keyword">...</span>
0467                 + [1; 1; -1] .* this.metabolismRecyclingProduction(this.substrateIndexs({<span class="string">'FMET'</span>}), this.compartmentIndexs_cytosol);
0468             this.metabolismRecyclingProduction(this.substrateIndexs({<span class="string">'MET'</span>; <span class="string">'H'</span>; <span class="string">'H2O'</span>}), this.compartmentIndexs_cytosol) = <span class="keyword">...</span>
0469                 + this.metabolismRecyclingProduction(this.substrateIndexs({<span class="string">'MET'</span>; <span class="string">'H'</span>; <span class="string">'H2O'</span>}), this.compartmentIndexs_cytosol) <span class="keyword">...</span>
0470                 - [1; 1; -1] .* this.metabolismRecyclingProduction(this.substrateIndexs({<span class="string">'FMET'</span>}), this.compartmentIndexs_cytosol);
0471             
0472             tmp = <span class="keyword">...</span>
0473                 + this.metabolismNewProduction(this.substrateIndexs({<span class="string">'FTHF10'</span>; <span class="string">'FOR'</span>; <span class="string">'THF'</span>}), this.compartmentIndexs_cytosol) <span class="keyword">...</span>
0474                 + this.metabolismRecyclingProduction(this.substrateIndexs({<span class="string">'FTHF10'</span>; <span class="string">'FOR'</span>; <span class="string">'THF'</span>}), this.compartmentIndexs_cytosol);
0475             this.metabolismNewProduction(this.substrateIndexs({<span class="string">'FTHF10'</span>; <span class="string">'FOR'</span>; <span class="string">'THF'</span>}), this.compartmentIndexs_cytosol) = max(0, <span class="keyword">...</span>
0476                 + tmp <span class="keyword">...</span>
0477                 - [1; -1; -1] .* max(0, min(tmp ./ [1; -1; -1])));
0478             this.metabolismRecyclingProduction(this.substrateIndexs({<span class="string">'FTHF10'</span>; <span class="string">'FOR'</span>; <span class="string">'THF'</span>}), this.compartmentIndexs_cytosol) = <span class="keyword">...</span>
0479                 + tmp <span class="keyword">...</span>
0480                 - this.metabolismNewProduction(this.substrateIndexs({<span class="string">'FTHF10'</span>; <span class="string">'FOR'</span>; <span class="string">'THF'</span>}), this.compartmentIndexs_cytosol);
0481             
0482             nInternalExchangeReactions = nnz(this.metabolismRecyclingProduction);
0483             nInternalExchangeReactionConstraints = 7;
0484             
0485             <span class="comment">%% numbers and indices of components</span>
0486             nCompartments = size(this.reactionStoichiometryMatrix, 3);
0487             nSubstrates   = size(this.reactionStoichiometryMatrix, 1) + nInternalExchangeReactionConstraints + 1;
0488             nEnzymes      = length(this.enzymeWholeCellModelIDs);
0489             
0490             subIdxs_substrates = (1:size(this.reactionStoichiometryMatrix, 1))';
0491             subIdxs_metaboliteInternalExchangeConstraints = subIdxs_substrates(end) + (1:nInternalExchangeReactionConstraints)';
0492             subIdxs_biomass = nSubstrates;
0493             
0494             subCmpIdxs_metaboliteInternalExchangeConstraints = sub2ind([nSubstrates nCompartments], <span class="keyword">...</span>
0495                 subIdxs_metaboliteInternalExchangeConstraints, repmat(this.compartmentIndexs_cytosol, size(subIdxs_metaboliteInternalExchangeConstraints)));
0496             subCmpIdxs_biomass = sub2ind([nSubstrates nCompartments], subIdxs_biomass, this.compartmentIndexs_cytosol);
0497             
0498             nReactions = <span class="keyword">...</span>
0499                 + size(this.reactionStoichiometryMatrix, 2) <span class="keyword">...</span>
0500                 + length(subIdxs_substrates) <span class="keyword">...</span>
0501                 + nInternalExchangeReactions <span class="keyword">...</span>
0502                 + 2;
0503             
0504             rxnIdxs_metabolicConversion = (1:size(this.reactionStoichiometryMatrix, 2))';
0505             rxnIdxs_metaboliteExternalExchange = rxnIdxs_metabolicConversion(end) + (1:length(subIdxs_substrates))';
0506             rxnIdxs_metaboliteInternalExchange = rxnIdxs_metaboliteExternalExchange(end) + (1:nInternalExchangeReactions)';
0507             rxnIdxs_biomassProduction = nReactions - 1;
0508             rxnIdxs_biomassExchange = nReactions;
0509             
0510             <span class="comment">%% reactions</span>
0511             <span class="comment">%initialize</span>
0512             fbaSMat = zeros(nSubstrates, nReactions, nCompartments);
0513             
0514             <span class="comment">%chemical and transport reactions</span>
0515             fbaSMat(subIdxs_substrates, rxnIdxs_metabolicConversion, :) = <span class="keyword">...</span>
0516                 this.reactionStoichiometryMatrix;
0517             
0518             <span class="comment">%metabolite external exchange</span>
0519             fbaSMat(sub2ind(<span class="keyword">...</span>
0520                 [nSubstrates, nReactions, nCompartments],<span class="keyword">...</span>
0521                 subIdxs_substrates,<span class="keyword">...</span>
0522                 rxnIdxs_metaboliteExternalExchange,<span class="keyword">...</span>
0523                 repmat(this.compartmentIndexs_extracellular, length(subIdxs_substrates), 1))) = <span class="keyword">...</span>
0524                 1;
0525             
0526             <span class="comment">%metabolite internal exchange</span>
0527             [i, j] = find(this.metabolismRecyclingProduction);
0528             fbaSMat(sub2ind(<span class="keyword">...</span>
0529                 [nSubstrates, nReactions, nCompartments],<span class="keyword">...</span>
0530                 i, rxnIdxs_metaboliteInternalExchange, j)) = -1;
0531             rxnIdxs_metaboliteInternalLimitedExchange = <span class="keyword">...</span>
0532                 rxnIdxs_metaboliteInternalExchange(<span class="keyword">...</span>
0533                 this.metabolismRecyclingProduction(this.metabolismRecyclingProduction ~= 0) &lt; 0);
0534             rxnIdxs_metaboliteInternalUnlimitedExchange = <span class="keyword">...</span>
0535                 rxnIdxs_metaboliteInternalExchange(<span class="keyword">...</span>
0536                 this.metabolismRecyclingProduction(this.metabolismRecyclingProduction ~= 0) &gt; 0);
0537             
0538             [~, ntpIdxs] = ismember(this.substrateIndexs({<span class="string">'ATP'</span>; <span class="string">'CTP'</span>; <span class="string">'GTP'</span>; <span class="string">'UTP'</span>}), i);
0539             [~, ndpIdxs] = ismember(this.substrateIndexs({<span class="string">'ADP'</span>; <span class="string">'CDP'</span>; <span class="string">'GDP'</span>; <span class="string">'UDP'</span>}), i);
0540             [~, nmpIdxs] = ismember(this.substrateIndexs({<span class="string">'AMP'</span>; <span class="string">'CMP'</span>; <span class="string">'GMP'</span>; <span class="string">'UMP'</span>}), i);
0541             [~, fthf10Idxs] = ismember(this.substrateIndexs({<span class="string">'FTHF10'</span>; <span class="string">'THF'</span>; <span class="string">'FOR'</span>}), i);
0542             [~, metIdxs] = ismember(this.substrateIndexs({<span class="string">'MET'</span>; <span class="string">'FMET'</span>}), i);
0543             tmp = zeros(nInternalExchangeReactionConstraints, nInternalExchangeReactions);
0544             idxs = [ntpIdxs(1) ndpIdxs(1) nmpIdxs(1)]; tmp(1, idxs(idxs ~= 0)) = 1;
0545             idxs = [ntpIdxs(2) ndpIdxs(2) nmpIdxs(2)]; tmp(2, idxs(idxs ~= 0)) = 1;
0546             idxs = [ntpIdxs(3) ndpIdxs(3) nmpIdxs(3)]; tmp(3, idxs(idxs ~= 0)) = 1;
0547             idxs = [ntpIdxs(4) ndpIdxs(4) nmpIdxs(4)]; tmp(4, idxs(idxs ~= 0)) = 1;
0548             idxs = fthf10Idxs(1:2); tmp(5, idxs(idxs ~= 0)) = 1;
0549             idxs = [fthf10Idxs([1 3]); metIdxs(2)]; tmp(6, idxs(idxs ~= 0)) = 1;
0550             idxs = metIdxs; tmp(7, idxs(idxs ~= 0)) = 1;
0551             fbaSMat(subIdxs_metaboliteInternalExchangeConstraints, rxnIdxs_metaboliteInternalExchange, this.compartmentIndexs_cytosol) = <span class="keyword">...</span>
0552                 + fbaSMat(subIdxs_metaboliteInternalExchangeConstraints, rxnIdxs_metaboliteInternalExchange, this.compartmentIndexs_cytosol) <span class="keyword">...</span>
0553                 + tmp;
0554             
0555             <span class="comment">%biomass production</span>
0556             fbaSMat(subIdxs_substrates, rxnIdxs_biomassProduction, :) = <span class="keyword">...</span>
0557                 -permute(this.metabolismNewProduction, [1 3 2]);
0558             fbaSMat(<span class="keyword">...</span>
0559                 subIdxs_biomass, <span class="keyword">...</span>
0560                 rxnIdxs_biomassProduction, <span class="keyword">...</span>
0561                 this.compartmentIndexs_cytosol) = 1;
0562             
0563             <span class="comment">%biomass exchange</span>
0564             fbaSMat(<span class="keyword">...</span>
0565                 subIdxs_biomass,<span class="keyword">...</span>
0566                 rxnIdxs_biomassExchange,<span class="keyword">...</span>
0567                 this.compartmentIndexs_cytosol) = 1;
0568             
0569             <span class="comment">%reshape</span>
0570             fbaSMatFull = fbaSMat;
0571             fbaSMatFull(subIdxs_substrates, rxnIdxs_biomassProduction, :) = <span class="keyword">...</span>
0572                 -permute(this.metabolismProduction, [1 3 2]);
0573             fbaSMat = reshape(<span class="keyword">...</span>
0574                 permute(fbaSMat, [2 1 3]), <span class="keyword">...</span>
0575                 size(fbaSMat, 2), [])';
0576             fbaSMatFull = reshape(<span class="keyword">...</span>
0577                 permute(fbaSMatFull, [2 1 3]), <span class="keyword">...</span>
0578                 size(fbaSMatFull, 2), [])';
0579             
0580             <span class="comment">%% reaction catalysis</span>
0581             fbaCatMat = zeros(nReactions, nEnzymes);
0582             fbaCatMat(rxnIdxs_metabolicConversion, :) = this.reactionCatalysisMatrix;
0583             
0584             <span class="comment">%% reaction bounds</span>
0585             fbaRxnBnds = repmat([-Inf Inf], nReactions, 1);
0586             
0587             this.reactionBounds(this.reactionBounds &lt; 0) = -Inf;
0588             this.reactionBounds(this.reactionBounds &gt; 0) =  Inf;
0589             fbaRxnBnds(rxnIdxs_metabolicConversion, :) = this.reactionBounds;
0590             
0591             exchangeLBs = this.substrateExternalExchangeBounds(:, 1);
0592             exchangeUBs = this.substrateExternalExchangeBounds(:, 2);
0593             fbaRxnBnds(rxnIdxs_metaboliteExternalExchange(~isnan(exchangeLBs)), 1) = exchangeLBs(~isnan(exchangeLBs));
0594             fbaRxnBnds(rxnIdxs_metaboliteExternalExchange(~isnan(exchangeUBs)), 2) = exchangeUBs(~isnan(exchangeUBs));
0595             
0596             fbaRxnBnds(rxnIdxs_metaboliteInternalLimitedExchange, 1) = -Inf;
0597             fbaRxnBnds(rxnIdxs_metaboliteInternalLimitedExchange, 2) = 0;
0598             fbaRxnBnds(rxnIdxs_metaboliteInternalUnlimitedExchange, 1) = 0;
0599             fbaRxnBnds(rxnIdxs_metaboliteInternalUnlimitedExchange, 2) = Inf;
0600             
0601             fbaRxnBnds(rxnIdxs_biomassProduction, :) = [0 Inf];
0602             fbaRxnBnds(rxnIdxs_biomassExchange, :) = [-Inf 0];
0603             
0604             <span class="comment">%% enzyme kinetics</span>
0605             fbaEnzBnds = NaN(nReactions, 2);
0606             fbaEnzBnds(rxnIdxs_metabolicConversion, :) = this.enzymeBounds;
0607             
0608             <span class="comment">%% right hand side</span>
0609             fbaRHS = zeros(nSubstrates * nCompartments, 1);
0610             
0611             <span class="comment">%% objective</span>
0612             fbaObj = zeros(nReactions, 1);
0613             fbaObj(rxnIdxs_biomassProduction) = 1e3;
0614             fbaObj(rxnIdxs_metaboliteInternalLimitedExchange) = 1 / sum(min(0, this.metabolismNewProduction(:)));
0615             
0616             <span class="comment">%% simplify stoichiometry matrix</span>
0617             subCmpIdxs_fba = (1:size(fbaSMat, 1))';
0618             rxnIdxs_fba = (1:size(fbaSMat, 2))';
0619             
0620             subCmpRemoved = zeros(size(fbaSMat, 1), 3);
0621             rxnRemoved = zeros(size(fbaSMat, 2), 3);
0622             
0623             <span class="keyword">if</span> ~exist(<span class="string">'simplifyNetwork'</span>, <span class="string">'var'</span>) || simplifyNetwork
0624                 iter = 0;
0625                 <span class="keyword">while</span> true
0626                     iter = iter + 1;
0627                     
0628                     <span class="comment">%factor</span>
0629                     [subCmpBlockIdxs, rxnBlockIdxs] = findNonInteractingRowsAndColumns(fbaSMatFull(subCmpIdxs_fba, rxnIdxs_fba));
0630                     <span class="keyword">if</span> all(subCmpBlockIdxs == 1)
0631                         <span class="keyword">break</span>;
0632                     <span class="keyword">end</span>
0633                     
0634                     blockIdx = subCmpBlockIdxs(subCmpIdxs_fba == subCmpIdxs_biomass);
0635                     subCmpRemoved(subCmpIdxs_fba(subCmpBlockIdxs ~= blockIdx), 1) = iter;
0636                     rxnRemoved(rxnIdxs_fba(rxnBlockIdxs ~= blockIdx), 1) = iter;
0637                     subCmpRemoved(subCmpIdxs_fba(subCmpBlockIdxs ~= blockIdx), 3) = 1;
0638                     rxnRemoved(rxnIdxs_fba(rxnBlockIdxs ~= blockIdx), 3) = 1;
0639                     subCmpIdxs_fba = subCmpIdxs_fba(subCmpBlockIdxs == blockIdx);
0640                     rxnIdxs_fba = rxnIdxs_fba(rxnBlockIdxs == blockIdx);
0641                     
0642                     <span class="comment">%prune</span>
0643                     flags = true(3, 1);
0644                     iter2 = 0;
0645                     <span class="keyword">while</span> any(flags)
0646                         iter2 = iter2 + 1;
0647                         flags = true(3, 1);
0648                         
0649                         <span class="comment">%remov leaf metabolites (and reactions involving them)</span>
0650                         <span class="comment">%which cannot be balanced with non-zero fluxes</span>
0651                         <span class="keyword">while</span> flags(1)
0652                             tmpSubCmpTfs = sum(fbaSMatFull(subCmpIdxs_fba, rxnIdxs_fba) ~= 0, 2) == 1;
0653                             tmpRxnTFs = any(fbaSMatFull(subCmpIdxs_fba(tmpSubCmpTfs), rxnIdxs_fba), 1);
0654                             
0655                             subCmpRemoved(subCmpIdxs_fba(tmpSubCmpTfs), 1) = iter;
0656                             rxnRemoved(rxnIdxs_fba(tmpRxnTFs), 1) = iter;
0657                             subCmpRemoved(subCmpIdxs_fba(tmpSubCmpTfs), 2) = iter2;
0658                             rxnRemoved(rxnIdxs_fba(tmpRxnTFs), 2) = iter2;
0659                             subCmpRemoved(subCmpIdxs_fba(tmpSubCmpTfs), 3) = 2;
0660                             rxnRemoved(rxnIdxs_fba(tmpRxnTFs), 3) = 2;
0661                             
0662                             subCmpIdxs_fba = subCmpIdxs_fba(~tmpSubCmpTfs);
0663                             rxnIdxs_fba = rxnIdxs_fba(~tmpRxnTFs);
0664                             
0665                             flags(1) = any(tmpSubCmpTfs);
0666                         <span class="keyword">end</span>
0667                         
0668                         <span class="comment">%remove metabolites (and reactions involving them)</span>
0669                         <span class="comment">%which can't move in both the forward and backward</span>
0670                         <span class="comment">%directions and thus cannot be balanced</span>
0671                         tmpSubCmpTfs = <span class="keyword">...</span>
0672                             (any(fbaSMatFull(subCmpIdxs_fba, rxnIdxs_fba) .* repmat(fbaRxnBnds(rxnIdxs_fba, 1)', numel(subCmpIdxs_fba), 1) &lt; 0, 2) | <span class="keyword">...</span>
0673                             any(fbaSMatFull(subCmpIdxs_fba, rxnIdxs_fba) .* repmat(fbaRxnBnds(rxnIdxs_fba, 2)', numel(subCmpIdxs_fba), 1) &lt; 0, 2)) &amp; <span class="keyword">...</span>
0674                             (any(fbaSMatFull(subCmpIdxs_fba, rxnIdxs_fba) .* repmat(fbaRxnBnds(rxnIdxs_fba, 1)', numel(subCmpIdxs_fba), 1) &gt; 0, 2) | <span class="keyword">...</span>
0675                             any(fbaSMatFull(subCmpIdxs_fba, rxnIdxs_fba) .* repmat(fbaRxnBnds(rxnIdxs_fba, 2)', numel(subCmpIdxs_fba), 1) &gt; 0, 2));
0676                         tmpRxnTFs = ~any(fbaSMatFull(subCmpIdxs_fba(~tmpSubCmpTfs), rxnIdxs_fba), 1);
0677                         
0678                         subCmpRemoved(subCmpIdxs_fba(~tmpSubCmpTfs), 1) = iter;
0679                         rxnRemoved(rxnIdxs_fba(~tmpRxnTFs), 1) = iter;
0680                         subCmpRemoved(subCmpIdxs_fba(~tmpSubCmpTfs), 2) = iter2;
0681                         rxnRemoved(rxnIdxs_fba(~tmpRxnTFs), 2) = iter2;
0682                         subCmpRemoved(subCmpIdxs_fba(~tmpSubCmpTfs), 3) = 3;
0683                         rxnRemoved(rxnIdxs_fba(~tmpRxnTFs), 3) = 3;
0684                         
0685                         subCmpIdxs_fba = subCmpIdxs_fba(tmpSubCmpTfs);
0686                         rxnIdxs_fba = rxnIdxs_fba(tmpRxnTFs);
0687                         
0688                         flags(2) = ~all(tmpSubCmpTfs);
0689                         
0690                         <span class="comment">%remove other reactions (and substrates unique to them)</span>
0691                         <span class="comment">%whose flux is zero in all solutions of the matrix</span>
0692                         <span class="comment">%problem Sv=0. That is reactions which have only zero</span>
0693                         <span class="comment">%components in the null space of S.</span>
0694                         tmpRxnTFs = any(null(fbaSMatFull(subCmpIdxs_fba, rxnIdxs_fba), <span class="string">'r'</span>), 2);
0695                         tmpSubCmpTfs = any(fbaSMatFull(subCmpIdxs_fba, rxnIdxs_fba(tmpRxnTFs)), 2);
0696                         
0697                         subCmpRemoved(subCmpIdxs_fba(~tmpSubCmpTfs), 1) = iter;
0698                         rxnRemoved(rxnIdxs_fba(~tmpRxnTFs), 1) = iter;
0699                         subCmpRemoved(subCmpIdxs_fba(~tmpSubCmpTfs), 2) = iter2;
0700                         rxnRemoved(rxnIdxs_fba(~tmpRxnTFs), 2) = iter2;
0701                         subCmpRemoved(subCmpIdxs_fba(~tmpSubCmpTfs), 3) = 4;
0702                         rxnRemoved(rxnIdxs_fba(~tmpRxnTFs), 3) = 4;
0703                         
0704                         subCmpIdxs_fba = subCmpIdxs_fba(tmpSubCmpTfs);
0705                         rxnIdxs_fba = rxnIdxs_fba(tmpRxnTFs);
0706                         
0707                         flags(3) = ~all(tmpSubCmpTfs);
0708                     <span class="keyword">end</span>
0709                 <span class="keyword">end</span>
0710             <span class="keyword">end</span>
0711             
0712             subCmpIdxs_fba = unique([subCmpIdxs_fba; subCmpIdxs_metaboliteInternalExchangeConstraints]);
0713             subCmpRemoved(subCmpIdxs_metaboliteInternalExchangeConstraints, :) = 0;
0714             
0715             subCmpRemoved = subCmpRemoved(reshape(sub2ind([nSubstrates nCompartments], <span class="keyword">...</span>
0716                 repmat(subIdxs_substrates, 1, nCompartments), <span class="keyword">...</span>
0717                 repmat(1:nCompartments, numel(subIdxs_substrates), 1)), [], 1), :);
0718             rxnRemoved = rxnRemoved(rxnIdxs_metabolicConversion, :);
0719             
0720             <span class="comment">%substrate indices</span>
0721             [subIdxs_fba, cmpIdxs_fba] = ind2sub([nSubstrates, nCompartments], subCmpIdxs_fba);
0722             this.fbaSubstrateIndexs_substrates = find(ismember(subIdxs_fba, subIdxs_substrates));
0723             this.fbaSubstrateIndexs_metaboliteInternalExchangeConstraints = find(ismember(subCmpIdxs_fba, subCmpIdxs_metaboliteInternalExchangeConstraints));
0724             this.substrateIndexs_fba = sub2ind(<span class="keyword">...</span>
0725                 [numel(this.substrateWholeCellModelIDs) nCompartments], <span class="keyword">...</span>
0726                 subIdxs_fba(this.fbaSubstrateIndexs_substrates), <span class="keyword">...</span>
0727                 cmpIdxs_fba(this.fbaSubstrateIndexs_substrates));
0728             [tfs, idxs] = ismember(rxnIdxs_fba, rxnIdxs_metaboliteExternalExchange);
0729             this.substrateIndexs_externalExchangedMetabolites = idxs(tfs);
0730             this.substrateIndexs_internalExchangedMetabolites = <span class="keyword">...</span>
0731                 find(this.metabolismRecyclingProduction);
0732             this.substrateIndexs_internalExchangedLimitedMetabolites = <span class="keyword">...</span>
0733                 find(this.metabolismRecyclingProduction &lt; 0);
0734             
0735             this.fbaSubstrateIndexs_biomass = find(ismember(subCmpIdxs_fba, subCmpIdxs_biomass));
0736             
0737             <span class="comment">%reaction indices</span>
0738             assert(all(ismember(rxnIdxs_metaboliteInternalLimitedExchange, rxnIdxs_fba)))
0739             
0740             this.reactionIndexs_fba = find(ismember(rxnIdxs_metabolicConversion, rxnIdxs_fba));
0741             this.fbaReactionIndexs_metabolicConversion = find(ismember(rxnIdxs_fba, rxnIdxs_metabolicConversion));
0742             this.fbaReactionIndexs_metaboliteExternalExchange = find(ismember(rxnIdxs_fba, rxnIdxs_metaboliteExternalExchange));
0743             this.fbaReactionIndexs_metaboliteInternalExchange = find(ismember(rxnIdxs_fba, rxnIdxs_metaboliteInternalExchange));
0744             this.fbaReactionIndexs_metaboliteInternalLimitedExchange = find(ismember(rxnIdxs_fba, rxnIdxs_metaboliteInternalLimitedExchange));
0745             this.fbaReactionIndexs_metaboliteInternalUnlimitedExchange = find(ismember(rxnIdxs_fba, rxnIdxs_metaboliteInternalUnlimitedExchange));
0746             this.fbaReactionIndexs_biomassProduction = find(ismember(rxnIdxs_fba, rxnIdxs_biomassProduction));
0747             this.fbaReactionIndexs_biomassExchange = find(ismember(rxnIdxs_fba, rxnIdxs_biomassExchange));
0748             
0749             fbaSMat = fbaSMat(subCmpIdxs_fba, rxnIdxs_fba);
0750             fbaCatMat = fbaCatMat(rxnIdxs_fba, :);
0751             fbaRHS = fbaRHS(subCmpIdxs_fba, :);
0752             fbaRxnBnds = fbaRxnBnds(rxnIdxs_fba, :);
0753             fbaEnzBnds = fbaEnzBnds(rxnIdxs_fba, :);
0754             fbaObj = fbaObj(rxnIdxs_fba, :);
0755             
0756             <span class="comment">%% set properties</span>
0757             this.fbaReactionStoichiometryMatrix = fbaSMat;
0758             this.fbaReactionCatalysisMatrix = fbaCatMat;
0759             this.fbaRightHandSide = fbaRHS;
0760             this.fbaReactionBounds = fbaRxnBnds;
0761             this.fbaEnzymeBounds = fbaEnzBnds;
0762             this.fbaObjective = fbaObj;
0763         <span class="keyword">end</span>
0764     <span class="keyword">end</span>
0765     
0766     <span class="comment">%model</span>
0767     methods
0768         <span class="comment">%Calculate</span>
0769         <span class="comment">%- contribution to FBA objective</span>
0770         <span class="comment">%- minimum expression consistent with cell cycle length</span>
0771         <a name="_sub4" href="#_subfunctions" class="code">function [bmProd, byProd, minEnzExp, maxEnzExp] = calcResourceRequirements_LifeCycle(this, ~, states)</a>
0772             <span class="comment">%% import classes</span>
0773             import edu.stanford.covert.util.ComputationUtil;
0774             
0775             <span class="comment">%% substrate and byproducts: none used by metabolism</span>
0776             bmProd = zeros(size(this.substrateWholeCellModelIDs));
0777             byProd = zeros(size(this.substrateWholeCellModelIDs));
0778             
0779             <span class="comment">%% fit enzyme expression consistent with growth rate</span>
0780             <span class="keyword">if</span> strcmp(this.macromoleculeStateInitialization, <span class="string">'expected'</span>)
0781                 growth0 = this.metabolicReaction.growth0;
0782             <span class="keyword">else</span>
0783                 growth0 = this.metabolicReaction.growth0 * this.macromoleculeStateInitializationGrowthFactor;
0784             <span class="keyword">end</span>
0785             
0786             substrates = zeros(size(this.substrates));
0787             substrates(this.substrateMetaboliteLocalIndexs, :) = this.substrates(this.substrateMetaboliteLocalIndexs, :);
0788             substrates(this.substrateIndexs_internalExchangedLimitedMetabolites) = <span class="keyword">...</span>
0789                 -growth0 *  this.metabolismRecyclingProduction(this.substrateIndexs_internalExchangedLimitedMetabolites);
0790             substrates(this.substrateMonomerLocalIndexs, :) = repmat(states.monomers0(this.substrateMonomerGlobalIndexs(:, 1)), 1, size(substrates, 2));
0791             substrates(this.substrateComplexLocalIndexs, :) = repmat(states.complexs0(this.substrateComplexGlobalIndexs(:, 1)), 1, size(substrates, 2));
0792             
0793             enzymes = zeros(size(this.enzymeWholeCellModelIDs));
0794             enzymes(this.enzymeMonomerLocalIndexs) = states.monomers0(this.enzymeMonomerGlobalIndexs);
0795             enzymes(this.enzymeComplexLocalIndexs) = states.complexs0(this.enzymeComplexGlobalIndexs);
0796             
0797             enzymes = this.fitEnzymes(growth0, substrates, enzymes);
0798             
0799             <span class="comment">%% metabolic enzyme levels constistent with observed growth rate</span>
0800             [minEnzExp, maxEnzExp] = this.calcMinMaxEnzymes(substrates, enzymes);
0801         <span class="keyword">end</span>
0802         
0803         <span class="comment">%calculate enzyme expression consistent with growth rate</span>
0804         <span class="comment">%- by minimizing the fractional change in fluxes by linear</span>
0805         <span class="comment">%  optimization (similar to MOMA)</span>
0806         <span class="comment">%- if linear optimization fails, increase/decrease all flux</span>
0807         <span class="comment">%  bounds by a factor of growth/growth0 and check if this yields</span>
0808         <span class="comment">%  a growth rate of growth0</span>
0809         <a name="_sub5" href="#_subfunctions" class="code">function enzymes = fitEnzymes(this, growth0, substrates, enzymes)</a>
0810             <span class="comment">% import classes</span>
0811             import edu.stanford.covert.util.ComputationUtil;
0812             import edu.stanford.covert.util.ConstantUtil;
0813             
0814             <span class="comment">%constants</span>
0815             nEnzymes = numel(enzymes);
0816             enzMWs = this.enzymeMolecularWeights;
0817             
0818             minAvgExp = zeros(size(enzymes));
0819             minAvgExp(this.enzymeMonomerLocalIndexs) = this.monomer.minimumAverageExpression;
0820             minAvgExp(this.enzymeComplexLocalIndexs) = this.complex.minimumAverageExpression;
0821             
0822             loIdxs = find(enzymes &lt; minAvgExp &amp; any(this.fbaReactionCatalysisMatrix, 1)');
0823             hiIdxs = find(enzymes &gt; minAvgExp);
0824             <span class="keyword">if</span> (minAvgExp(loIdxs)-enzymes(loIdxs))' * enzMWs(loIdxs) &lt; (enzymes(hiIdxs) - minAvgExp(hiIdxs))' * enzMWs(hiIdxs)
0825                 enzymes(hiIdxs) = minAvgExp(hiIdxs) + (enzymes(hiIdxs) - minAvgExp(hiIdxs)) * <span class="keyword">...</span>
0826                     ((enzymes(hiIdxs)-minAvgExp(hiIdxs))' * enzMWs(hiIdxs) - <span class="keyword">...</span>
0827                     (minAvgExp(loIdxs)-enzymes(loIdxs))' * enzMWs(loIdxs)) / <span class="keyword">...</span>
0828                     ((enzymes(hiIdxs)-minAvgExp(hiIdxs))' * enzMWs(hiIdxs));
0829                 enzymes(loIdxs) = minAvgExp(loIdxs);
0830             <span class="keyword">else</span>
0831                 warning(<span class="string">'WholeCell:warning'</span>, <span class="string">'Cannot ensure minimum enzyme expression'</span>);
0832             <span class="keyword">end</span>
0833             
0834             initEnzymes = enzymes;
0835             
0836             <span class="comment">% calculate current growth and fluxes</span>
0837             [fbaObj, fbaSMat, fbaRxnBounds] = this.calcEffectiveFBANetwork();
0838             fluxBounds = this.calcFluxBounds(substrates, enzymes, fbaRxnBounds, this.fbaEnzymeBounds);
0839             [growth, ~, oldFbaReactionFluxs] = this.calcGrowthRate(fluxBounds, fbaObj, fbaSMat);
0840             
0841             <span class="keyword">if</span> abs(growth - growth0) / growth0 &lt; this.tolerance || abs(growth - growth0) &lt; 1e-12
0842                 <span class="keyword">return</span>;
0843             <span class="keyword">end</span>
0844             
0845             <span class="keyword">if</span> growth &lt; growth0 &amp;&amp; growth0 &gt; this.calcGrowthRate(this.calcFluxBounds(<span class="keyword">...</span>
0846                     substrates, enzymes, fbaRxnBounds, this.fbaEnzymeBounds, false), fbaObj, fbaSMat)
0847                 throw(MException(<span class="string">'Metabolism:error'</span>, <span class="string">'Cannot fit enzyme expression to match growth rate'</span>));
0848             <span class="keyword">end</span>
0849             
0850             <span class="comment">% reactions constrainted by enzyme expression</span>
0851             [forIdxs, revIdxs] = this.calcEnzymeKineticallyLimitedReactions(substrates, enzymes, fbaObj, fbaSMat, fbaRxnBounds);
0852             unconstrainedEnzIdxs = find(~any(this.fbaReactionCatalysisMatrix([forIdxs; revIdxs], :), 1)');
0853             
0854             <span class="comment">% minimize difference from current enzyme expression</span>
0855             objectiveFunc = zeros(size(fbaObj));
0856             objectiveFunc([forIdxs; revIdxs]) = 1 ./ oldFbaReactionFluxs([forIdxs; revIdxs]);
0857             
0858             <span class="comment">% minimum enzyme expression</span>
0859             loFluxBounds = fluxBounds(:, 1);
0860             upFluxBounds = fluxBounds(:, 2);
0861             
0862             <span class="keyword">if</span> growth &lt; growth0
0863                 tmpFluxBounds = this.calcFluxBounds(<span class="keyword">...</span>
0864                     substrates, enzymes, fbaRxnBounds, this.fbaEnzymeBounds, false);
0865                 loFluxBounds(forIdxs) = oldFbaReactionFluxs(forIdxs);
0866                 upFluxBounds(forIdxs) = min(tmpFluxBounds(forIdxs, 2), oldFbaReactionFluxs(forIdxs) * (1 + this.tolerance) * growth0 / growth);
0867                 loFluxBounds(revIdxs) = max(tmpFluxBounds(revIdxs, 1), oldFbaReactionFluxs(revIdxs) * (1 + this.tolerance) * growth0 / growth);
0868                 upFluxBounds(revIdxs) = oldFbaReactionFluxs(revIdxs);
0869                 
0870                 objective = <span class="string">'minimize'</span>;
0871             <span class="keyword">else</span>
0872                 loFluxBounds(forIdxs) = 0;
0873                 upFluxBounds(forIdxs) = oldFbaReactionFluxs(forIdxs);
0874                 loFluxBounds(revIdxs) = oldFbaReactionFluxs(revIdxs);
0875                 upFluxBounds(revIdxs) = 0;
0876                 
0877                 objective = <span class="string">'maximize'</span>;
0878             <span class="keyword">end</span>
0879             
0880             loFluxBounds(this.fbaReactionIndexs_biomassProduction) = growth0;
0881             upFluxBounds(this.fbaReactionIndexs_biomassProduction) = growth0;
0882             
0883             loFluxBounds = max(loFluxBounds, -this.realmax);
0884             upFluxBounds = min(upFluxBounds,  this.realmax);
0885             [fbaReactionFluxs, ~, ~, errFlag, errMsg] = ComputationUtil.linearProgramming(<span class="keyword">...</span>
0886                 objective, objectiveFunc, fbaSMat, <span class="keyword">...</span>
0887                 this.fbaRightHandSide, loFluxBounds, upFluxBounds, <span class="keyword">...</span>
0888                 <span class="string">'S'</span>, <span class="string">'C'</span>, this.linearProgrammingOptions);
0889             <span class="keyword">if</span> errFlag
0890                 throw(MException(<span class="string">'Metabolism:error'</span>, <span class="string">'Unable to optimize fluxes: %s'</span>, errMsg));
0891             <span class="keyword">end</span>
0892             
0893             <span class="comment">%adjust enzyme expression</span>
0894             fbaReactionFluxs = max(min(fbaReactionFluxs, upFluxBounds), loFluxBounds);
0895             
0896             reactionEnzymes = zeros(size(fbaReactionFluxs));
0897             reactionEnzymes(forIdxs) = fbaReactionFluxs(forIdxs) ./ this.fbaEnzymeBounds(forIdxs, 2);
0898             reactionEnzymes(revIdxs) = fbaReactionFluxs(revIdxs) ./ this.fbaEnzymeBounds(revIdxs, 1);
0899             
0900             enzymes = max(this.fbaReactionCatalysisMatrix([forIdxs; revIdxs], :) .* <span class="keyword">...</span>
0901                 repmat(reactionEnzymes([forIdxs; revIdxs]), 1, nEnzymes), [], 1)';
0902             enzymes(unconstrainedEnzIdxs) = initEnzymes(unconstrainedEnzIdxs);
0903             enzymes = this.adjustEnzymeExpressionSparsely(growth0, substrates, enzymes, initEnzymes);
0904             enzymes = max(minAvgExp, enzymes);
0905             
0906             <span class="keyword">if</span> growth &lt; growth0
0907                 incEnzymes = find(enzymes &gt; initEnzymes);
0908                 decEnzymes = setdiff(unique([find(enzymes &lt; initEnzymes); unconstrainedEnzIdxs]), incEnzymes);
0909                 enzymes = max(initEnzymes, enzymes);
0910                 enzymes(decEnzymes) = minAvgExp(decEnzymes) + (initEnzymes(decEnzymes) - minAvgExp(decEnzymes)) * <span class="keyword">...</span>
0911                     ((initEnzymes(decEnzymes) - minAvgExp(decEnzymes))' * enzMWs(decEnzymes) - <span class="keyword">...</span>
0912                     (enzymes(incEnzymes) - initEnzymes(incEnzymes))' * enzMWs(incEnzymes)) / <span class="keyword">...</span>
0913                     ((initEnzymes(decEnzymes) - minAvgExp(decEnzymes))' * enzMWs(decEnzymes));
0914             <span class="keyword">else</span>
0915                 incEnzymes = unique([find(enzymes &gt; initEnzymes); unconstrainedEnzIdxs]);
0916                 decEnzymes = setdiff(find(enzymes &lt; initEnzymes), incEnzymes);
0917                 enzymes = min(initEnzymes, enzymes);
0918                 enzymes(incEnzymes) = minAvgExp(incEnzymes) + (initEnzymes(incEnzymes) - minAvgExp(incEnzymes)) * <span class="keyword">...</span>
0919                     ((initEnzymes(incEnzymes) - minAvgExp(incEnzymes))' * enzMWs(incEnzymes) - <span class="keyword">...</span>
0920                     (enzymes(decEnzymes) - initEnzymes(decEnzymes))' * enzMWs(decEnzymes)) / <span class="keyword">...</span>
0921                     ((initEnzymes(incEnzymes) - minAvgExp(incEnzymes))' * enzMWs(incEnzymes));
0922             <span class="keyword">end</span>
0923             
0924             <span class="comment">%check growth rate fit</span>
0925             <span class="keyword">if</span> abs(this.calcGrowthRate(this.calcFluxBounds(<span class="keyword">...</span>
0926                     substrates, enzymes, fbaRxnBounds, this.fbaEnzymeBounds), fbaObj, fbaSMat) <span class="keyword">...</span>
0927                     - growth0) / growth0 &gt; this.tolerance
0928                 throw(MException(<span class="string">'Metabolism:error'</span>, <span class="string">'Growth rate should not change'</span>));
0929             <span class="keyword">end</span>
0930             <span class="keyword">if</span> abs((enzymes' * enzMWs - initEnzymes' * enzMWs) / (initEnzymes' * enzMWs)) &gt; this.tolerance
0931                 throw(MException(<span class="string">'Metabolism:error'</span>, <span class="string">'Protein mass should not change'</span>));
0932             <span class="keyword">end</span>
0933         <span class="keyword">end</span>
0934         
0935         <a name="_sub6" href="#_subfunctions" class="code">function enzymes = adjustEnzymeExpressionSparsely(this, growth0, substrates, enzymes, initEnzymes)</a>
0936             [fbaObj, fbaSMat, fbaRxnBounds] = this.calcEffectiveFBANetwork();
0937             
0938             tfs = this.adjustEnzymeExpressionSparselyHelper(growth0, substrates, enzymes, initEnzymes, true(size(enzymes)));
0939             enzymes(tfs) = initEnzymes(tfs);
0940             idxs = find(~tfs);
0941             
0942             <span class="comment">%one at a time</span>
0943             tf = false;
0944             <span class="keyword">for</span> i = 1:numel(idxs)
0945                 tmpEnzymes = enzymes;
0946                 tmpEnzymes(idxs(i)) = initEnzymes(idxs(i));
0947                 
0948                 <span class="keyword">if</span> abs(this.calcGrowthRate(this.calcFluxBounds(<span class="keyword">...</span>
0949                         substrates, tmpEnzymes, fbaRxnBounds, this.fbaEnzymeBounds), fbaObj, fbaSMat) <span class="keyword">...</span>
0950                         - growth0) / growth0 &lt;= this.tolerance
0951                     tf = true;
0952                     <span class="keyword">break</span>;
0953                 <span class="keyword">end</span>
0954             <span class="keyword">end</span>
0955             <span class="keyword">if</span> ~tf
0956                 <span class="keyword">return</span>;
0957             <span class="keyword">end</span>
0958             
0959             <span class="comment">%combinations</span>
0960             <span class="keyword">try</span>
0961                 tmpTfs = dec2bin(0:2^numel(idxs)-1) == <span class="string">'1'</span>;
0962             <span class="keyword">catch</span> <span class="comment">%#ok&lt;CTCH&gt;</span>
0963                 <span class="keyword">return</span>;
0964             <span class="keyword">end</span>
0965             [~, order] = sort(sum(tmpTfs, 2), 1, <span class="string">'descend'</span>);
0966             tmpTfs = tmpTfs(order, :);
0967             
0968             <span class="keyword">for</span> i = 1:2^numel(idxs)
0969                 tmpEnzymes = enzymes;
0970                 tmpEnzymes(idxs(tmpTfs(i, :))) = initEnzymes(idxs(tmpTfs(i, :)));
0971                 
0972                 <span class="keyword">if</span> abs(this.calcGrowthRate(this.calcFluxBounds(<span class="keyword">...</span>
0973                         substrates, tmpEnzymes, fbaRxnBounds, this.fbaEnzymeBounds), fbaObj, fbaSMat) <span class="keyword">...</span>
0974                         - growth0) / growth0 &lt;= this.tolerance
0975                     enzymes = tmpEnzymes;
0976                     <span class="keyword">return</span>;
0977                 <span class="keyword">end</span>
0978             <span class="keyword">end</span>
0979         <span class="keyword">end</span>
0980         
0981         <a name="_sub7" href="#_subfunctions" class="code">function tfs = adjustEnzymeExpressionSparselyHelper(this, growth0, substrates, enzymes, initEnzymes, tfs)</a>
0982             [fbaObj, fbaSMat, fbaRxnBounds] = this.calcEffectiveFBANetwork();
0983             
0984             tmpEnzymes = enzymes;
0985             tmpEnzymes(tfs) = initEnzymes(tfs);
0986             
0987             <span class="keyword">if</span> sum(tfs) == 0 || abs(this.calcGrowthRate(this.calcFluxBounds(<span class="keyword">...</span>
0988                     substrates, tmpEnzymes, fbaRxnBounds, this.fbaEnzymeBounds), fbaObj, fbaSMat) <span class="keyword">...</span>
0989                     - growth0) / growth0 &lt;= this.tolerance
0990                 <span class="keyword">return</span>;
0991             <span class="keyword">elseif</span> sum(tfs) == 1
0992                 tfs(:) = false;
0993                 <span class="keyword">return</span>;
0994             <span class="keyword">end</span>
0995             
0996             idxs = find(tfs);
0997             tfs1 = tfs;
0998             tfs2 = tfs;
0999             tfs1(idxs(1:ceil(end/2))) = false;
1000             tfs2(idxs(ceil(end/2))+1:end) = false;
1001             tfs = <span class="keyword">...</span>
1002                 this.adjustEnzymeExpressionSparselyHelper(growth0, substrates, enzymes, initEnzymes, tfs1) | <span class="keyword">...</span>
1003                 this.adjustEnzymeExpressionSparselyHelper(growth0, substrates, enzymes, initEnzymes, tfs2);
1004         <span class="keyword">end</span>
1005         
1006         <span class="comment">%calculate minimum and maximum enzyme weight consistent with growth</span>
1007         <span class="comment">%rate</span>
1008         <span class="comment">%- by minimizing/maximizing the total enzyme weight consistent with</span>
1009         <span class="comment">%  the current growth rate by linear optimization</span>
1010         <span class="comment">%- if linear optimization fails, minEnzymes/maxEnzymes is set to</span>
1011         <span class="comment">%  the current enyzme level</span>
1012         <a name="_sub8" href="#_subfunctions" class="code">function [minEnzymes, maxEnzymes] = calcMinMaxEnzymes(this, substrates, enzymes)</a>
1013             <span class="comment">% import classes</span>
1014             import edu.stanford.covert.util.ComputationUtil;
1015             import edu.stanford.covert.util.ConstantUtil;
1016             
1017             <span class="comment">%constants</span>
1018             nEnzymes = numel(enzymes);
1019             
1020             minAvgExp = zeros(size(enzymes));
1021             minAvgExp(this.enzymeMonomerLocalIndexs) = this.monomer.minimumAverageExpression;
1022             minAvgExp(this.enzymeComplexLocalIndexs) = this.complex.minimumAverageExpression;
1023             
1024             <span class="comment">% calculate current growth and fluxes</span>
1025             [fbaObj, fbaSMat, fbaRxnBounds] = this.calcEffectiveFBANetwork();
1026             fluxBounds = this.calcFluxBounds(<span class="keyword">...</span>
1027                 substrates, enzymes, fbaRxnBounds, this.fbaEnzymeBounds);
1028             [growth, ~, oldFbaReactionFluxs] = this.calcGrowthRate(fluxBounds, fbaObj, fbaSMat);
1029             
1030             <span class="comment">% reactions constrainted by enzyme expression</span>
1031             [forIdxs, revIdxs] = this.calcEnzymeKineticallyLimitedReactions(substrates, enzymes, fbaObj, fbaSMat, fbaRxnBounds);
1032             partlyUnconstrainedEnzIdxs = find(any(this.fbaReactionCatalysisMatrix(setdiff(1:<span class="keyword">end</span>, [forIdxs; revIdxs]), :), 1));
1033             unconstrainedEnzIdxs = find(~any(this.fbaReactionCatalysisMatrix([forIdxs; revIdxs], :), 1));
1034             
1035             <span class="keyword">if</span> isempty(forIdxs) &amp;&amp; isempty(revIdxs)
1036                 warning(<span class="string">'WholeCell:warning'</span>, <span class="string">'Unable to fit growth rate'</span>);
1037                 minEnzymes = enzymes;
1038                 maxEnzymes = enzymes;
1039                 <span class="keyword">return</span>;
1040             <span class="keyword">end</span>
1041             
1042             <span class="comment">% enzyme expression range</span>
1043             objectiveFunc = zeros(size(fbaObj));
1044             objectiveFunc(forIdxs) = (this.fbaReactionCatalysisMatrix(forIdxs, :) * this.enzymeMolecularWeights) ./ this.fbaEnzymeBounds(forIdxs, 2);
1045             objectiveFunc(revIdxs) = (this.fbaReactionCatalysisMatrix(revIdxs, :) * this.enzymeMolecularWeights) ./ this.fbaEnzymeBounds(revIdxs, 1);
1046             
1047             <span class="comment">% minimum enzyme expression</span>
1048             loFluxBounds = fluxBounds(:, 1);
1049             upFluxBounds = fluxBounds(:, 2);
1050             
1051             loFluxBounds(this.fbaReactionIndexs_biomassProduction) = growth;
1052             upFluxBounds(this.fbaReactionIndexs_biomassProduction) = growth;
1053             
1054             loFluxBounds(forIdxs) = 0;
1055             upFluxBounds(forIdxs) = oldFbaReactionFluxs(forIdxs);
1056             loFluxBounds(revIdxs) = oldFbaReactionFluxs(revIdxs);
1057             upFluxBounds(revIdxs) = 0;
1058             
1059             loFluxBounds = max(loFluxBounds, -this.realmax);
1060             upFluxBounds = min(upFluxBounds,  this.realmax);
1061             [fbaReactionFluxs, ~, ~, errFlag, errMsg] = ComputationUtil.linearProgramming(<span class="keyword">...</span>
1062                 <span class="string">'minimize'</span>, objectiveFunc, fbaSMat, <span class="keyword">...</span>
1063                 this.fbaRightHandSide, loFluxBounds, upFluxBounds, <span class="keyword">...</span>
1064                 <span class="string">'S'</span>, <span class="string">'C'</span>, this.linearProgrammingOptions);
1065             <span class="keyword">if</span> errFlag
1066                 warning(<span class="string">'WholeCell:warning'</span>, errMsg);
1067                 minEnzymes = enzymes;
1068             <span class="keyword">else</span>
1069                 fbaReactionFluxs = max(min(fbaReactionFluxs, upFluxBounds), loFluxBounds);
1070                 
1071                 reactionEnzymes = zeros(size(oldFbaReactionFluxs));
1072                 reactionEnzymes(forIdxs) = fbaReactionFluxs(forIdxs) ./ this.fbaEnzymeBounds(forIdxs, 2);
1073                 reactionEnzymes(revIdxs) = fbaReactionFluxs(revIdxs) ./ this.fbaEnzymeBounds(revIdxs, 1);
1074                 
1075                 minEnzymes = max(this.fbaReactionCatalysisMatrix([forIdxs; revIdxs], :) .* <span class="keyword">...</span>
1076                     repmat(reactionEnzymes([forIdxs; revIdxs]), 1, nEnzymes), [], 1)';
1077                 minEnzymes(unconstrainedEnzIdxs) = 1;
1078                 minEnzymes(partlyUnconstrainedEnzIdxs) = max(minEnzymes(partlyUnconstrainedEnzIdxs), 1);
1079                 minEnzymes = min(enzymes, minEnzymes);
1080                 minEnzymes = max(minAvgExp, minEnzymes);
1081             <span class="keyword">end</span>
1082             
1083             <span class="keyword">if</span> abs(this.calcGrowthRate(this.calcFluxBounds(<span class="keyword">...</span>
1084                     substrates, minEnzymes, fbaRxnBounds, this.fbaEnzymeBounds), fbaObj, fbaSMat) <span class="keyword">...</span>
1085                     - growth) / growth &gt; this.tolerance || <span class="keyword">...</span>
1086                     any(minEnzymes &gt; enzymes)
1087                 warning(<span class="string">'WholeCell:warning'</span>, <span class="string">'Unable to calculate minimum enzymes'</span>);
1088                 minEnzymes = enzymes;
1089             <span class="keyword">end</span>
1090             
1091             <span class="comment">% maximum enzyme expression</span>
1092             loFluxBounds = fluxBounds(:, 1);
1093             upFluxBounds = fluxBounds(:, 2);
1094             
1095             loFluxBounds(this.fbaReactionIndexs_biomassProduction) = growth;
1096             upFluxBounds(this.fbaReactionIndexs_biomassProduction) = growth;
1097             
1098             tmpFluxBounds = this.calcFluxBounds(<span class="keyword">...</span>
1099                 substrates, enzymes, fbaRxnBounds, this.fbaEnzymeBounds, false);
1100             loFluxBounds(forIdxs) = oldFbaReactionFluxs(forIdxs);
1101             upFluxBounds(forIdxs) = tmpFluxBounds(forIdxs, 2);
1102             loFluxBounds(revIdxs) = tmpFluxBounds(revIdxs, 1);
1103             upFluxBounds(revIdxs) = oldFbaReactionFluxs(revIdxs);
1104             
1105             loFluxBounds = max(loFluxBounds, -this.realmax);
1106             upFluxBounds = min(upFluxBounds,  this.realmax);
1107             [fbaReactionFluxs, ~, ~, errFlag, errMsg] = ComputationUtil.linearProgramming(<span class="keyword">...</span>
1108                 <span class="string">'maximize'</span>, objectiveFunc, fbaSMat, <span class="keyword">...</span>
1109                 this.fbaRightHandSide, loFluxBounds, upFluxBounds, <span class="keyword">...</span>
1110                 <span class="string">'S'</span>, <span class="string">'C'</span>, this.linearProgrammingOptions);
1111             <span class="keyword">if</span> errFlag
1112                 warning(<span class="string">'WholeCell:warning'</span>, errMsg);
1113                 maxEnzymes = enzymes;
1114             <span class="keyword">else</span>
1115                 fbaReactionFluxs = max(min(fbaReactionFluxs, upFluxBounds), loFluxBounds);
1116                 
1117                 reactionEnzymes = zeros(size(oldFbaReactionFluxs));
1118                 reactionEnzymes(forIdxs) = fbaReactionFluxs(forIdxs) ./ this.fbaEnzymeBounds(forIdxs, 2);
1119                 reactionEnzymes(revIdxs) = fbaReactionFluxs(revIdxs) ./ this.fbaEnzymeBounds(revIdxs, 1);
1120                 
1121                 tmpReactionCatalysisMatrix = this.fbaReactionCatalysisMatrix;
1122                 tmpReactionCatalysisMatrix(tmpReactionCatalysisMatrix == 0) = NaN;
1123                 maxEnzymes = 0*min(tmpReactionCatalysisMatrix([forIdxs; revIdxs], :) .* <span class="keyword">...</span>
1124                     repmat(reactionEnzymes([forIdxs; revIdxs]), 1, nEnzymes), [], 1)';
1125                 maxEnzymes(unconstrainedEnzIdxs) = realmax; <span class="comment">%#ok&lt;CPROP,PROP&gt;</span>
1126                 maxEnzymes(partlyUnconstrainedEnzIdxs) = max(maxEnzymes(partlyUnconstrainedEnzIdxs), 1);
1127                 maxEnzymes = max(enzymes, maxEnzymes);
1128             <span class="keyword">end</span>
1129             
1130             <span class="keyword">if</span> abs(this.calcGrowthRate(this.calcFluxBounds(<span class="keyword">...</span>
1131                     substrates, maxEnzymes, fbaRxnBounds, this.fbaEnzymeBounds), fbaObj, fbaSMat) <span class="keyword">...</span>
1132                     - growth) / growth &gt; this.tolerance || <span class="keyword">...</span>
1133                     any(maxEnzymes &lt; enzymes)
1134                 warning(<span class="string">'WholeCell:warning'</span>, <span class="string">'Unable to calculate maximum enzymes'</span>);
1135                 maxEnzymes = enzymes;
1136             <span class="keyword">end</span>
1137         <span class="keyword">end</span>
1138         
1139         <span class="comment">%calculate indices of reactions whose flux is actively limited by</span>
1140         <span class="comment">%the product of the expression of the catalyzing enzyme and its</span>
1141         <span class="comment">%reaction kinetics</span>
1142         <a name="_sub9" href="#_subfunctions" class="code">function [forIdxs, revIdxs] = calcEnzymeKineticallyLimitedReactions(this, substrates, enzymes, fbaObj, fbaSMat, fbaRxnBounds)</a>
1143             fluxBounds = this.calcFluxBounds(<span class="keyword">...</span>
1144                 substrates, enzymes, fbaRxnBounds, this.fbaEnzymeBounds);
1145             fluxBounds_nokinetics = this.calcFluxBounds(<span class="keyword">...</span>
1146                 substrates, enzymes, fbaRxnBounds, this.fbaEnzymeBounds, false);
1147             [~, ~, fbaReactionFluxs] = this.calcGrowthRate(fluxBounds, fbaObj, fbaSMat);
1148             
1149             rxnIdxs = this.fbaReactionIndexs_metabolicConversion;
1150             forTfs = <span class="keyword">...</span>
1151                 fbaReactionFluxs(rxnIdxs) &gt; 0 &amp; <span class="keyword">...</span>
1152                 fluxBounds(rxnIdxs, 2) &lt; fluxBounds_nokinetics(rxnIdxs, 2);
1153             revTfs = <span class="keyword">...</span>
1154                 fbaReactionFluxs(rxnIdxs) &lt; 0 &amp; <span class="keyword">...</span>
1155                 fluxBounds(rxnIdxs, 1) &gt; fluxBounds_nokinetics(rxnIdxs, 1);
1156             
1157             forIdxs = rxnIdxs(forTfs);
1158             revIdxs = rxnIdxs(revTfs);
1159         <span class="keyword">end</span>
1160         
1161         <a name="_sub10" href="#_subfunctions" class="code">function [fbaObj, fbaSMat, fbaRxnBounds] = calcEffectiveFBANetwork(this)</a>
1162             fbaObj = this.fbaObjective;
1163             fbaObj(:) = 0;
1164             fbaObj(this.fbaReactionIndexs_biomassProduction) = 1;
1165             
1166             fbaSMat = this.fbaReactionStoichiometryMatrix;
1167             fbaSMat(this.fbaSubstrateIndexs_substrates, this.fbaReactionIndexs_biomassProduction) = <span class="keyword">...</span>
1168                 - this.metabolismNewProduction(this.substrateIndexs_fba) <span class="keyword">...</span>
1169                 - 4 * this.metabolismRecyclingProduction(this.substrateIndexs_fba);
1170             
1171             fbaRxnBounds = this.fbaReactionBounds;
1172             fbaRxnBounds(this.fbaReactionIndexs_metaboliteInternalExchange, :) = 0;
1173         <span class="keyword">end</span>
1174         
1175         <span class="comment">%initialization</span>
1176         <span class="comment">%- Compute growth rate, reaction fluxs, reduced costs, duals</span>
1177         <span class="comment">%- Don't update metabolite counts</span>
1178         <a name="_sub11" href="#_subfunctions" class="code">function initializeState(this)</a>
1179             [this.metabolicReaction.growth, this.metabolicReaction.fluxs] = <span class="keyword">...</span>
1180                 this.calcGrowthRate(this.calcFluxBounds(this.substrates, this.enzymes, this.fbaReactionBounds, this.fbaEnzymeBounds));
1181         <span class="keyword">end</span>
1182         
1183         <span class="comment">%resource requirements</span>
1184         <a name="_sub12" href="#_subfunctions" class="code">function result = calcResourceRequirements_Current(this)</a>
1185             result = zeros(size(this.substrates));
1186             result(this.substrateIndexs_externalExchangedMetabolites, this.compartmentIndexs_extracellular) = <span class="keyword">...</span>
1187                 this.fbaReactionBounds(this.fbaReactionIndexs_metaboliteExternalExchange, 2) * sum(this.mass.cellDry) * this.stepSizeSec;
1188             result = <span class="keyword">...</span>
1189                 + result <span class="keyword">...</span>
1190                 - this.metabolismRecyclingProduction * 2 * max(<span class="keyword">...</span>
1191                     sum(this.mass.cellDry) / this.mass.cellInitialDryWeight * log(2) / this.cellCycleLength, <span class="keyword">...</span>
1192                     this.metabolicReaction.growth);
1193         <span class="keyword">end</span>
1194         
1195         <span class="comment">%simulation</span>
1196         <a name="_sub13" href="#_subfunctions" class="code">function evolveState(this)</a>
1197             <span class="comment">%% Calculate growth rate</span>
1198             [this.metabolicReaction.growth, this.metabolicReaction.fluxs, fbaReactionFluxs] = <span class="keyword">...</span>
1199                 this.calcGrowthRate(this.calcFluxBounds(this.substrates, this.enzymes, this.fbaReactionBounds, this.fbaEnzymeBounds));
1200             
1201             <span class="comment">%% Compute real-valued amount of nutrient imported, biomass (DNA,</span>
1202             <span class="comment">%RNA, protein, membrane, etc) precursors produced and consumed,</span>
1203             <span class="comment">%and monomers modified. Stochastically round to:</span>
1204             <span class="comment">%1. Instantaneously, maintain integer-valued amounts of metabolites</span>
1205             <span class="comment">%2. Over time, maintain experimentally determined ratios of biomass</span>
1206             <span class="comment">%   components (stored in this.metabolismProduction)</span>
1207             
1208             <span class="comment">%nutrient uptake</span>
1209             this.substrates(this.substrateIndexs_externalExchangedMetabolites, this.compartmentIndexs_extracellular) = <span class="keyword">...</span>
1210                 + this.substrates(this.substrateIndexs_externalExchangedMetabolites, this.compartmentIndexs_extracellular) <span class="keyword">...</span>
1211                 - this.randStream.stochasticRound(fbaReactionFluxs(this.fbaReactionIndexs_metaboliteExternalExchange) * this.stepSizeSec);
1212             
1213             <span class="comment">%recycled metabolites</span>
1214             this.substrates(this.substrateIndexs_internalExchangedMetabolites) = <span class="keyword">...</span>
1215                 + this.substrates(this.substrateIndexs_internalExchangedMetabolites) <span class="keyword">...</span>
1216                 + this.randStream.stochasticRound(fbaReactionFluxs(this.fbaReactionIndexs_metaboliteInternalExchange));
1217             
1218             <span class="comment">%new metabolites</span>
1219             this.substrates = <span class="keyword">...</span>
1220                 + this.substrates <span class="keyword">...</span>
1221                 + this.randStream.stochasticRound(this.metabolismNewProduction * this.metabolicReaction.growth * this.stepSizeSec);
1222             
1223             <span class="comment">%unaccounted energy consumption</span>
1224             this.substrates(this.substrateIndexs_atpHydrolysis) = <span class="keyword">...</span>
1225                 + this.substrates(this.substrateIndexs_atpHydrolysis) <span class="keyword">...</span>
1226                 + [-1; -1; 1; 1; 1] * this.randStream.stochasticRound(<span class="keyword">...</span>
1227                 this.unaccountedEnergyConsumption * this.metabolicReaction.growth * this.stepSizeSec);
1228             
1229             <span class="comment">%make metabolites counts positive (in case stochastic rounding</span>
1230             <span class="comment">%made them slightly negative eg -1) except H2O, H+</span>
1231             <span class="keyword">if</span> any(any(this.substrates(this.substrateMetaboliteLocalIndexs(:, 1), :) &lt; -1))
1232                 [i, j] = find(this.substrates(this.substrateMetaboliteLocalIndexs(:, 1), :) &lt; -1);
1233                 compIDs = cell(3, 1);
1234                 compIDs(this.compartmentIndexs_cytosol) = {<span class="string">'c'</span>};
1235                 compIDs(this.compartmentIndexs_extracellular) = {<span class="string">'e'</span>};
1236                 compIDs(this.compartmentIndexs_membrane) = {<span class="string">'m'</span>};
1237                 n = max(cellfun(@length, this.substrateWholeCellModelIDs(this.substrateMetaboliteLocalIndexs(i, 1))));
1238                 msg = cellfun(@(id, comp, val) sprintf([<span class="string">'%-'</span> num2str(n) <span class="string">'s  %-3s  %10d'</span>], id, comp, val), <span class="keyword">...</span>
1239                     this.substrateWholeCellModelIDs(this.substrateMetaboliteLocalIndexs(i, 1)), compIDs(j), <span class="keyword">...</span>
1240                     num2cell(this.substrates(sub2ind(size(this.substrates), this.substrateMetaboliteLocalIndexs(i, 1), j))), <span class="keyword">...</span>
1241                     <span class="string">'UniformOutput'</span>, false);
1242                 warning(<span class="string">'WholeCell:warning:negativeMetabolites'</span>, <span class="keyword">...</span>
1243                     [<span class="string">'%d metabolites have negative counts\n%-'</span> num2str(n) <span class="string">'s  %-3s  %10s\n%-'</span> num2str(n) <span class="string">'s  %-3s  %10s\n%s'</span>], <span class="keyword">...</span>
1244                     numel(i), <span class="keyword">...</span>
1245                     <span class="string">'Met'</span>, <span class="string">'Cmp'</span>, <span class="string">'Val'</span>, <span class="keyword">...</span>
1246                     repmat(<span class="string">'='</span>, 1, n), <span class="string">'==='</span>, repmat(<span class="string">'='</span>, 1, 10), <span class="keyword">...</span>
1247                     strjoin(sprintf(<span class="string">'\n'</span>), msg{:}));
1248             <span class="keyword">end</span>
1249             this.substrates(this.substrateMetaboliteLocalIndexs(:, 1), :) = max(0, this.substrates(this.substrateMetaboliteLocalIndexs(:, 1), :));
1250             
1251             <span class="comment">%% update cell volume</span>
1252             this.mass.calcMass();
1253             this.geometry.calculateVolume();
1254         <span class="keyword">end</span>
1255     <span class="keyword">end</span>
1256     
1257     <span class="comment">%evolve state helper methods</span>
1258     methods
1259         <span class="comment">%calculates the growth rate given the supplied flux bounds</span>
1260         <a name="_sub14" href="#_subfunctions" class="code">function [growth, reactionFluxs, fbaReactionFluxs, </a><span class="keyword">...</span>
1261                 reducedCosts, fbaReducedCosts, <span class="keyword">...</span>
1262                 shadowPrices, fbaShadowPrices] = <span class="keyword">...</span>
1263                 calcGrowthRate(this, fluxBounds, fbaObj, fbaSMat)
1264             <span class="comment">% import classes</span>
1265             import edu.stanford.covert.util.ComputationUtil;
1266             import edu.stanford.covert.util.ConstantUtil;
1267             
1268             <span class="comment">%inputs</span>
1269             <span class="keyword">if</span> nargin &lt; 3
1270                 fbaObj = this.fbaObjective;
1271             <span class="keyword">end</span>
1272             <span class="keyword">if</span> nargin &lt; 4
1273                 fbaSMat = this.fbaReactionStoichiometryMatrix;
1274             <span class="keyword">end</span>
1275             
1276             <span class="comment">%flux bounds</span>
1277             loFluxBounds = fluxBounds(:, 1);
1278             upFluxBounds = fluxBounds(:, 2);
1279             
1280             <span class="comment">%real-valued linear programming</span>
1281             loFluxBounds = max(loFluxBounds, -this.realmax);
1282             upFluxBounds = min(upFluxBounds,  this.realmax);
1283             [fbaReactionFluxs, lambda, ~, errFlag, errMsg] = ComputationUtil.linearProgramming(<span class="keyword">...</span>
1284                 <span class="string">'maximize'</span>, fbaObj, fbaSMat, <span class="keyword">...</span>
1285                 this.fbaRightHandSide, loFluxBounds, upFluxBounds, <span class="keyword">...</span>
1286                 <span class="string">'S'</span>, <span class="string">'C'</span>, this.linearProgrammingOptions);
1287             <span class="keyword">if</span> errFlag
1288                 warning(<span class="string">'WholeCell:warning'</span>, <span class="string">'Linear programming error: %s. Returning feasible, but possibly non-optimal solution x=0.'</span>, errMsg);
1289                 fbaReactionFluxs = zeros(size(loFluxBounds));
1290             <span class="keyword">end</span>
1291             
1292             fbaReactionFluxs = max(min(fbaReactionFluxs, upFluxBounds), loFluxBounds);
1293             growth = fbaReactionFluxs(this.fbaReactionIndexs_biomassProduction);
1294             reactionFluxs = zeros(size(this.reactionStoichiometryMatrix, 2), 1);
1295             reactionFluxs(this.reactionIndexs_fba) = fbaReactionFluxs(this.fbaReactionIndexs_metabolicConversion);
1296             
1297             <span class="keyword">if</span> nargout &gt; 3
1298                 fbaReducedCosts = lambda.reducedCosts;
1299                 reducedCosts = zeros(size(this.reactionStoichiometryMatrix, 2), 1);
1300                 reducedCosts(this.reactionIndexs_fba) = fbaReducedCosts(this.fbaReactionIndexs_metabolicConversion);
1301                 
1302                 fbaShadowPrices = lambda.shadowPrices;
1303                 shadowPrices = zeros(size(this.substrates));
1304                 shadowPrices(this.substrateIndexs_fba) = fbaShadowPrices(this.fbaSubstrateIndexs_substrates);
1305             <span class="keyword">end</span>
1306         <span class="keyword">end</span>
1307         
1308         <span class="comment">%Compute reaction flux upper and lower bounds based on</span>
1309         <span class="comment">% 1. Enzyme kinetics         (fbaEnzymeBounds)</span>
1310         <span class="comment">% 2. Enzyme availability     (enzymes)</span>
1311         <span class="comment">% 3. Transport rates         (fbaReactionBounds)</span>
1312         <span class="comment">% 4. Metabolite availability (substrates)</span>
1313         <span class="comment">% 5. Protein availability    (substrates)</span>
1314         <a name="_sub15" href="#_subfunctions" class="code">function bounds = calcFluxBounds(this, substrates, enzymes, fbaReactionBounds, fbaEnzymeBounds, </a><span class="keyword">...</span>
1315                 applyEnzymeKineticBounds, applyEnzymeBounds, applyDirectionalityBounds, <span class="keyword">...</span>
1316                 applyExternalMetaboliteBounds, applyInternalMetaboliteBounds, applyProteinBounds)
1317             <span class="comment">%import classes</span>
1318             import edu.stanford.covert.util.ConstantUtil;
1319             
1320             <span class="comment">%options</span>
1321             <span class="keyword">if</span> nargin &lt; 6,  applyEnzymeKineticBounds      = true; <span class="keyword">end</span>
1322             <span class="keyword">if</span> nargin &lt; 7,  applyEnzymeBounds             = true; <span class="keyword">end</span>
1323             <span class="keyword">if</span> nargin &lt; 8,  applyDirectionalityBounds     = true; <span class="keyword">end</span>
1324             <span class="keyword">if</span> nargin &lt; 9,  applyExternalMetaboliteBounds = true; <span class="keyword">end</span>
1325             <span class="keyword">if</span> nargin &lt; 10, applyInternalMetaboliteBounds = true; <span class="keyword">end</span>
1326             <span class="keyword">if</span> nargin &lt; 11, applyProteinBounds            = true; <span class="keyword">end</span>
1327             
1328             <span class="comment">%numbers</span>
1329             nReactions = size(this.fbaReactionStoichiometryMatrix, 2);
1330             
1331             <span class="comment">%initialize</span>
1332             lowerBounds = -inf(nReactions, 1);
1333             upperBounds =  inf(nReactions, 1);
1334             
1335             <span class="comment">%numbers of enzymes catalyzing each reaction, enzyme kinetics</span>
1336             rxnEnzymes = this.fbaReactionCatalysisMatrix * enzymes;
1337             <span class="keyword">if</span> applyEnzymeKineticBounds
1338                 lowerBounds = max(lowerBounds, fbaEnzymeBounds(:, 1) .* rxnEnzymes);
1339                 upperBounds = min(upperBounds, fbaEnzymeBounds(:, 2) .* rxnEnzymes);
1340             <span class="keyword">end</span>
1341             
1342             <span class="comment">%numbers of enzymes catalyzing each reaction, unkown enzyme kinetics</span>
1343             <span class="keyword">if</span> applyEnzymeBounds
1344                 lowerBounds(any(this.fbaReactionCatalysisMatrix, 2) &amp; rxnEnzymes &lt;= 0) = 0;
1345                 upperBounds(any(this.fbaReactionCatalysisMatrix, 2) &amp; rxnEnzymes &lt;= 0) = 0;
1346             <span class="keyword">end</span>
1347             
1348             <span class="comment">%reaction directionality / thermodynamics</span>
1349             <span class="keyword">if</span> applyDirectionalityBounds
1350                 lowerBounds(this.fbaReactionIndexs_metabolicConversion) = max(lowerBounds(this.fbaReactionIndexs_metabolicConversion), fbaReactionBounds(this.fbaReactionIndexs_metabolicConversion, 1));
1351                 upperBounds(this.fbaReactionIndexs_metabolicConversion) = min(upperBounds(this.fbaReactionIndexs_metabolicConversion), fbaReactionBounds(this.fbaReactionIndexs_metabolicConversion, 2));
1352                 
1353                 lowerBounds(this.fbaReactionIndexs_metaboliteInternalExchange) = max(lowerBounds(this.fbaReactionIndexs_metaboliteInternalExchange), fbaReactionBounds(this.fbaReactionIndexs_metaboliteInternalExchange, 1));
1354                 upperBounds(this.fbaReactionIndexs_metaboliteInternalExchange) = min(upperBounds(this.fbaReactionIndexs_metaboliteInternalExchange), fbaReactionBounds(this.fbaReactionIndexs_metaboliteInternalExchange, 2));
1355                 
1356                 lowerBounds(this.fbaReactionIndexs_biomassExchange) = max(lowerBounds(this.fbaReactionIndexs_biomassExchange), fbaReactionBounds(this.fbaReactionIndexs_biomassExchange, 1));
1357                 upperBounds(this.fbaReactionIndexs_biomassExchange) = min(upperBounds(this.fbaReactionIndexs_biomassExchange), fbaReactionBounds(this.fbaReactionIndexs_biomassExchange, 2));
1358                 
1359                 lowerBounds(this.fbaReactionIndexs_biomassProduction) = max(lowerBounds(this.fbaReactionIndexs_biomassProduction), fbaReactionBounds(this.fbaReactionIndexs_biomassProduction, 1));
1360                 upperBounds(this.fbaReactionIndexs_biomassProduction) = min(upperBounds(this.fbaReactionIndexs_biomassProduction), fbaReactionBounds(this.fbaReactionIndexs_biomassProduction, 2));
1361             <span class="keyword">end</span>
1362             
1363             <span class="comment">%external metabolite availability</span>
1364             <span class="keyword">if</span> applyExternalMetaboliteBounds
1365                 upperBounds(this.fbaReactionIndexs_metaboliteExternalExchange) = min(<span class="keyword">...</span>
1366                     upperBounds(this.fbaReactionIndexs_metaboliteExternalExchange), <span class="keyword">...</span>
1367                     substrates(this.substrateIndexs_externalExchangedMetabolites, this.compartmentIndexs_extracellular) / this.stepSizeSec);
1368                 
1369                 cellDryMass = sum(this.mass.cellDry);
1370                 lowerBounds(this.fbaReactionIndexs_metaboliteExternalExchange) = <span class="keyword">...</span>
1371                     max(lowerBounds(this.fbaReactionIndexs_metaboliteExternalExchange), <span class="keyword">...</span>
1372                     fbaReactionBounds(this.fbaReactionIndexs_metaboliteExternalExchange, 1) * cellDryMass);
1373                 upperBounds(this.fbaReactionIndexs_metaboliteExternalExchange) = <span class="keyword">...</span>
1374                     min(upperBounds(this.fbaReactionIndexs_metaboliteExternalExchange), <span class="keyword">...</span>
1375                     fbaReactionBounds(this.fbaReactionIndexs_metaboliteExternalExchange, 2) * cellDryMass);
1376             <span class="keyword">end</span>
1377             
1378             <span class="comment">%internal metabolite availability</span>
1379             <span class="keyword">if</span> applyInternalMetaboliteBounds
1380                 lowerBounds(this.fbaReactionIndexs_metaboliteInternalLimitedExchange) = max(<span class="keyword">...</span>
1381                     lowerBounds(this.fbaReactionIndexs_metaboliteInternalLimitedExchange), <span class="keyword">...</span>
1382                     -substrates(this.substrateIndexs_internalExchangedLimitedMetabolites) / <span class="keyword">...</span>
1383                     this.stepSizeSec);
1384             <span class="keyword">end</span>
1385             
1386             <span class="comment">%protein monomers and complexes</span>
1387             <span class="keyword">if</span> applyProteinBounds
1388                 limitedReactions = any(any(<span class="keyword">...</span>
1389                     this.reactionStoichiometryMatrix([this.substrateMonomerLocalIndexs; this.substrateComplexLocalIndexs], this.reactionIndexs_fba, :) &amp; <span class="keyword">...</span>
1390                     ~permute(repmat(this.proteinLimitableProteinComposition * substrates(this.substrateIndexs_limitableProteins, :), [1 1 numel(this.reactionIndexs_fba)]), [1 3 2]), <span class="keyword">...</span>
1391                     3), 1);
1392                 lowerBounds(this.fbaReactionIndexs_metabolicConversion(limitedReactions)) = 0;
1393                 upperBounds(this.fbaReactionIndexs_metabolicConversion(limitedReactions)) = 0;
1394             <span class="keyword">end</span>
1395             
1396             <span class="comment">%sanitize linear programming input: replace Infs with large numbers</span>
1397             bounds = [lowerBounds upperBounds];
1398         <span class="keyword">end</span>
1399     <span class="keyword">end</span>
1400 <span class="keyword">end</span></pre></div>
<hr><address>Generated on Thu 19-Jul-2012 18:32:13 by <strong><a href="http://www.artefact.tk/software/matlab/m2html/" title="Matlab Documentation in HTML">m2html</a></strong> &copy; 2005</address>
</body>
</html>